---
title: "Covid-19 Vaccination Atlas: A systems vaccinology analysis"
author: "Wasim Aluísio Prates-Syed"
date: "2024-06-20"
output: 
  html_document: 
    toc: TRUE
    toc_depth: 4
    code_folding: hide
editor_options: 
  markdown: 
    wrap: 72
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval=FALSE, cache = TRUE)
```

# Covid-19 Vaccination Atlas: A systems vaccinology analysis

# Article information

**Authors:** Wasim Aluísio Prates-Syed1,2, Dennyson Leandro Mathias da
Fonseca3, Shahab Zaki Pour3, Aline Lira1,2, Nelson Cortes1, 2, Jaqueline
Dinis Queiroz Silva 1,5, Larissa Vuitika¹, Evellyn Carvalho1,4, Tania
Geraldine Churascari Vinces4, Gerhard Wunderlich4, Ricardo
Durães-Carvalho6, Lorena C. S. Chaves7, Niels O. S. Câmara1, Ester
Cerdeira SABINO8,9, José E. Krieger10, Otávio Cabral-Marques1,3,5,11,∙,
Gustavo Cabral-Miranda1,2,5,8 \*, ∙

**Emails:** [wasim.syed\@usp.br](mailto:wasim.syed@usp.br){.email},
[dennyson\@usp.br](mailto:dennyson@usp.br){.email},
[shahab.178\@gmail.com](mailto:shahab.178@gmail.com){.email},
[cv.tania\@usp.br](mailto:cv.tania@usp.br){.email},
[aline.llira\@usp.br](mailto:aline.llira@usp.br){.email},
[nelson.cortes\@usp.br](mailto:nelson.cortes@usp.br){.email},
[jaquelinedinis\@gmail.com](mailto:jaquelinedinis@gmail.com){.email},
[evelyn.carvalho\@unesp.br](mailto:evelyn.carvalho@unesp.br){.email},
[vuitika\@usp.br](mailto:vuitika@usp.br){.email},
[lorenacschaves\@gmail.com](mailto:lorenacschaves@gmail.com){.email},
[gwunder\@usp.br](mailto:gwunder@usp.br){.email},
[rdcarval\@gmail.com](mailto:rdcarval@gmail.com){.email},
[niels\@icb.usp.br](mailto:niels@icb.usp.br){.email},
[sabinoec\@usp.br](mailto:sabinoec@usp.br){.email},
[j.krieger\@hc.fm.usp.br](mailto:j.krieger@hc.fm.usp.br){.email},
[otavio.cmarques\@usp.br](mailto:otavio.cmarques@usp.br){.email},
[gcabral.miranda\@usp.br](mailto:gcabral.miranda@usp.br){.email}

**Institutions:** 1. Department of Immunology, Institute of Biomedical
Sciences (ICB), University of São Paulo (USP), São Paulo, Brazil. 2. The
Interunits Graduate Program in Biotechnology of the University of São
Paulo, the Butantan Institute and the Technological Research Institute
of the State of São Paulo, Brazil. 3. Interunit Postgraduate Program on
Bioinformatics, Institute of Mathematics and Statistics (IME),
University of Sao Paulo (USP), Sao Paulo, Brazil. 4. Department of
Parasitology, Institute of Biomedical Sciences (ICB), University of São
Paulo (USP), São Paulo, Brazil. 5. The Graduate Program in
Pathophysiology and Toxicology, Department of Clinical and Toxicological
Analyses, School of Pharmaceutical Sciences, University of São Paulo,
São Paulo, Brazil. 6. São Paulo School of Medicine, Department of
Microbiology, Immunology and Parasitology, Federal University of São
Paulo (UNIFESP), São Paulo, SP, Brazil. 7. Department of Microbiology
and Immunology, School of Medicine, Emory University, Atlanta, Georgia,
United States. 8. Institute of Tropical Medicine, Faculty of Medicine of
the University of São Paulo, Brazil 9. Department of Infectious and
Parasitic Diseases, Faculty of Medicine, University of São Paulo, São
Paulo, Brazil 10. Heart Institute, Clinical Hospital, Faculty of
Medicine, University of São Paulo, Brazil; Laboratory of Genetics and
Molecular Cardiology, Clinical Hospital, Faculty of Medicine, University
of São Paulo, Brazil. 11. Department of Medicine, Division of Molecular
Medicine, University of São Paulo School of Medicine, São Paulo, Brazil.

# Abstract:

The rapid development of COVID-19 vaccines has played a crucial role in
mitigating its public health impact. However, a gene-level understanding
of immune system dynamics during vaccination is crucial for
comprehending the various vaccination scenarios, including infection. In
this study, we conducted an integrative analysis of immunological
transcriptomic profiles to delineate the impact of COVID-19 vaccination
on immune responses. Our analysis comprised 681 bulk RNAseq samples from
343 participants, including healthy individuals vaccinated with
different types of COVID-19 vaccines (inactivated virus, protein
subunit, viral-vectored, and mRNA), along with infected patients with or
without prior vaccination history. We identified distinct and
overlapping immune-related genes and pathways associated with COVID-19
vaccination and infection, considering time and regimen. Furthermore, we
expanded our analysis to include datasets from vaccines targeting other
pathogens, such as influenza, smallpox, and yellow fever. Additionally,
we introduce VaxGO, a novel tool for exploring immunological processes
and comparing them across different vaccines. Our work will contribute
to advancing our understanding of the immunological mechanisms
underlying Covid-19 vaccination and inform future developments in
vaccine design and personalized medicine. **Keywords:** transcriptomics,
vaccines, COVID-19, SARS-CoV-2.

Read the recent version here:
(<https://www.medrxiv.org/content/10.1101/2024.05.22.24307755v3>)

# Codes for analyses and figures

# 1. Preparing

## Packages

```{r message=FALSE, warning=FALSE, cache = TRUE}
# install.packages("randomForest")
# install.packages("caret")
# install.packages("rpart")
# install.packages("rpart.plot")
# install.packages("tidymodels")
# install.packages("reticulate")
# install.packages("themis")
# install.packages("ranger")
# install.packages("workflowsets")
# install.packages("glmnet")
# install.packages("kknn")
# install.packages("nnet")
# install.packages("tictoc")
# install.packages("ggExtra")
# install.packages("remotes")
# remotes::install_github("omnideconv/immunedeconv")
# BiocManager::install("multtest")
# install.packages("weights")
# install.packages("vip")
# install.packages("ggstatsplot")
# install.packages("chorddiag")
# devtools::install_github("cmartin/ggConvexHull")
# install.packages("ggbeeswarm")
# install.packages("ggbreak")
library(ggbreak)
library(maditr)
library(readxl)
library(ggConvexHull)
library(randomForest)
library(ranger)
library(caret)
library(reticulate)
library(themis)
library(rpart)
library(rpart.plot)
library(here)
library(ggrepel)
library(ComplexHeatmap)
library(janitor)
library(ggsci)
library(vip)
library(circlize)
library(workflowsets)
library(glmnet)
library(kknn)
library(edgeR)
library(nnet)
library(ggthemes)
library(readr)
library(qvalue)
library(GEOquery)
library(Matrix)
library(RColorBrewer)
library(celldex)
library(biomaRt)
library(org.Hs.eg.db)
library(plotly)
library(DESeq2)
library(msigdbr)
library(clusterProfiler)
library(EnhancedVolcano)
library(corrr)
library(ggcorrplot)
library(FactoMineR)
library(factoextra)
library(esquisse)
library(corto)
library(scales)
library(ggstatsplot)
library(patchwork)
library(vegan)
library(BH)
# library(chorddiag)
library(tidyverse)
library(tidymodels)
library(purrr)
# library(immunedeconv)
tidymodels_prefer()


```

## Standardize figures

```{r cache = TRUE, message=FALSE, warning=FALSE}
### VAX SIG DB ----
ann_vaxsig_colors = list(
  type = c("VLP" = "#7E6148FF",
           "LA" =  "#EFC000FF",
           "CONJ" = "#F39B7FFF", 
           'IN'= "#00A087FF", #1st Gen  vaccines
           'VV' = "#3C5488FF", #3rd Gen vaccines
           'RNA' = "#4DBBD5FF",
           'SU' = "#8491B4FF",
           "IN/SU" = "#8491B4FF",
           "PS" = "#F39B7FFF",
           'I'= "#DC0000FF",
           "H" = "grey95",
           "V-I" = "#ffadc7"),
  Platform = c("VLP" = "#7E6148FF",
           "LA" =  "#EFC000FF",
           "CONJ" = "#F39B7FFF", 
           'IN'= "#00A087FF", #1st Gen  vaccines
           'VV' = "#3C5488FF", #3rd Gen vaccines
           'RNA' = "#4DBBD5FF",
           'SU' = "#8491B4FF",
           'I'= "#DC0000FF",
           "H" = "grey95"),
  target_pathogen_disease = c('MENINGOCOCCUS'	= "#9a998c",
                              'PNEUMOCOCCUS'	= "gray1",
                              'TUBERCULOSIS'	= "#b2675e",
                              'F TULARENSIS'	= "#00A087FF",
                              'LEISHMANIA'	= "#3C5488FF",
                              'MALARIA' = "brown",
                              'EBOV'	= "#d5bf86",
                              'EBOLA'	= "#d5bf86",
                              'HBV'	= "#9DC296",
                              'HBV, HAV'	= "#9DC296",
                              'HEPATITIS A/B'	= "#9DC296",
                              'HIV'	= "#ff6b35",
                              'HPV'	= "#9dcee2",
                              'INFLUENZA'	= "#04395e",
                              'MEASLES'	= "#ff6392",
                              'MMR'	= "#613dc1",
                              'SMALLPOX'	= "#f19c79",
                              'VEEV'	= "#ecbcfd",
                              'VZV'	= "#dc2f02",
                              'VARICELLA ZOSTER'	= "#dc2f02",
                              'YF'	= "#EFC000FF",
                              'YELLOW FEVER'	= "#EFC000FF",
                              "SARS-COV-2" = "#DC0000FF",
                              "SARS-CoV-2" = "#DC0000FF",
                              "SARS-COV-2/BBIBP" = "#00A087FF",
                                "SARS-COV-2/ZF2001" = "#8491B4FF",
                                "SARS-COV-2/BNT" = "#4DBBD5FF",
                                "SARS-COV-2/MO" = "#72ddf7",
                                "SARS-COV-2/CHAD" = "#3C5488FF",
                                "SARS-COV-2/I" = "#DC0000FF"
                              ),
  pathogen = c('MENINGOCOCCUS'	= "#9a998c",
                              'PNEUMOCOCCUS'	= "gray1",
                              'TUBERCULOSIS'	= "#b2675e",
                              'F TULARENSIS'	= "#00A087FF",
                              'LEISHMANIA'	= "#3C5488FF",
                              'MALARIA' = "brown",
                              'EBOV'	= "#d5bf86",
                              'EBOLA'	= "#d5bf86",
                              'HBV'	= "#9DC296",
                              'HBV, HAV'	= "#9DC296",
                              'HEPATITIS A/B'	= "#9DC296",
                              'HIV'	= "#ff6b35",
                              'HPV'	= "#9dcee2",
                              'INFLUENZA'	= "#04395e",
                              'MEASLES'	= "#ff6392",
                              'MMR'	= "#613dc1",
                              'SMALLPOX'	= "#f19c79",
                              'VEEV'	= "#ecbcfd",
                              'VZV'	= "#dc2f02",
                              'VARICELLA ZOSTER'	= "#dc2f02",
                              'YF'	= "#EFC000FF",
                              'YELLOW FEVER'	= "#EFC000FF",
                              "SARS-COV-2" = "#DC0000FF",
                              "SARS-CoV-2" = "#DC0000FF",
                              "SARS-COV-2/BBIBP" = "#00A087FF",
                                "SARS-COV-2/ZF2001" = "#8491B4FF",
                                "SARS-COV-2/BNT" = "#4DBBD5FF",
                                "SARS-COV-2/MO" = "#72ddf7",
                                "SARS-COV-2/CHAD" = "#3C5488FF",
                                "SARS-COV-2/I" = "#DC0000FF"
                              ),
  covid_other = c("COVID" = "#3C5488FF",
                  "OTHER" = "#4DBBD5FF"),
  set = c("COVID-19" = "#3C5488FF",
                  "13Vax Atlas" = "#4DBBD5FF"),
  microbe_type = c("VIRUS" = "#5e60ce",
                   'BACTERIUM' = "#00A087FF",
                   'PARASITE' = "#7E6148FF"),
  
  week = c("0" = "#ECF8FA",
           "1" = "#caf0f8",
           "2" = "#6CD5EA",
           "3" = "#0087BF",
           "4" = "#015BA0", 
           "6" = "#03045e",
           "7" = "#03045e",
           "8" = "#03045e",
           "9" = "#03045e",
           "10" = "#03045e",
           "12" = "#03045e"),
  month = c("1" = "#caf0f8",
            "0" = "#6CD5EA",
            "2" = "#015BA0"),
  vaccine = c("BBIBP" = "#00A087FF",
              "ZF2001" = "#8491B4FF",
              "BNT" = "#4DBBD5FF",
              "MO" = "#72ddf7",
              "ChAd" = "#3C5488FF",
              "I" = "white",
              "H" = "white"),
  dose = c(
    "0" = "white",
    "1" = "#6CD5EA",
    "2" = "#0087BF",
    "3" = "#03045e"),
  day = c(
    "0" = "white",
    "1" = "#caf0f8",
    "2" = "#ade8f4",
    "3" = "#90e0ef",
    "4" = "#6CD5EA",
    "5" = "#48cae4",
    "6" = "#00b4d8",
    "7" = "#0096c7",
    "8" = "#0096c7",
    "9" = "#0096c7",
    "10" = "#0087BF",
    "11" = "#0087BF",
    "12" = "#0087BF",
    "13" = "#0087BF",
    "14" = "#0077b6",
    "21" = "#0077b6",
    "26" = "#015BA0",
    "28" = "#023e8a",
    "51" = "#03045e",
    "60" = "#03045e",
    "70" = "#03045e",
    "84" = "#03045e"),
  # day <- setNames(
  # c(colorRampPalette(c("white", "#023e8a"))(31), rep("#03045e", 70)),
  # as.character(0:100)),
  
  infection = c("0" = "white",
                "1" = "#4DBBD5FF",
                "2" = "#3C5488FF"),
  variant = c("None" = "white",
              "Beta" = "#4DBBD5FF",
              "Omicron" = "#3C5488FF"),
  severity = c("H" = "white",
               "S" = "#DC0000FF",
               "MO" = "#8491B4FF",
               "MI" = "#91D1C2FF",
               "NA" = "white"),
  sex = c("Male" = "#4DBBD5FF",
          "Female" = "#E64B35FF",
          "M/F" = "#8491B4FF"),
  disease_vac = c("I" = "#DC0000FF",
                  "V" = "#4DBBD5FF",
                  "H" = "white"),
  gse_id = c("GSE199750" = "#EFC000FF",
               "GSE201533" = "#CD534CFF",
               "GSE201530" = "#868686FF",
               "GSE206023" = "#7AA6DCFF",
               "GSE189039" = "#0073C2FF"),
  correct = c("Correct" = "#4DBBD5FF",
              "Incorrect" = "black"),
  condition_col = c(
                                "BBIBP (V3, D07)" = "#00A087FF",
                                "BBIBP (V3, D14)" = "#00A087FF",
                                "BBIBP (V3, D28)" = "#00A087FF",
                                "BNT (V1, D6)" = "#4DBBD5FF",
                                "BNT (V2, D1)" = "#4DBBD5FF",
                                "BNT (V3, D1)" = "#4DBBD5FF",
                                "BNT-I (D0-5)"= "#DC0000FF",
                                "BNT-I (D10, hospital)"= "#DC0000FF",
                                "BNT-I (D26, hospital)"= "#DC0000FF",
                                "BNT-I (D51, hospital)"= "#DC0000FF",
                                "BNT-I-BNT (D51, hospital)"= "#DC0000FF",
                                "I (D0-4)"= "#DC0000FF",
                                "I (D10, hospital)"= "#DC0000FF",
                                "I (D26, hospital)"= "#DC0000FF",
                                "I (D51, hospital)"= "#DC0000FF",
                                "I-BNT (D5)"= "#4DBBD5FF", 
                                "I-I (D0-5)" = "#DC0000FF",
                                "BNT-I (D1)"= "#DC0000FF",
                                "BNT-I (D10, mild)"= "#DC0000FF",
                                "BNT-I (D10, severe)"= "#DC0000FF",
                                "BNT-I (D2)"= "#DC0000FF",
                                "BNT-I (D26, mild)"= "#DC0000FF",
                                "BNT-I (D26, severe)"= "#DC0000FF",
                                "BNT-I (D3)"= "#DC0000FF",
                                "BNT-I (D4)"= "#DC0000FF",
                                "BNT-I (D51, severe)"= "#DC0000FF",
                                "BNT-I-BNT (D51, mild)"= "#F5BB00",
                                "BNT-I-BNT (D51, severe)"= "#F5BB00",
                                "BNT-MO (V1, D6)"= "#72ddf7",
                                "BNT-MO (V3, D1)"= "#72ddf7",
                                "ChAd (V1, D3)"= "#3C5488FF",
                                "ChAd (V1, D6)"= "#3C5488FF",
                                "ChAd (V1, D7)"= "#3C5488FF",
                                "ChAd (V2, D1)"= "#3C5488FF",
                                "ChAd (V2, D3)"= "#3C5488FF",
                                "ChAd (V2, D7)"= "#3C5488FF",
                                "ChAd-BNT (V2, D0)"= "#4DBBD5FF",
                                "ChAd-BNT (V2, D3)"= "#4DBBD5FF",
                                "ChAd-BNT (V2, D7)"= "#4DBBD5FF",
                                "ChAd-BNT (V3, D1)"= "#4DBBD5FF",
                                "I (D1)"= "#DC0000FF",
                                "I (D10, moderate)"= "#DC0000FF",
                                "I (D10, severe)"= "#DC0000FF",
                                "I (D26, moderate)"= "#DC0000FF",
                                "I (D26, severe)"= "#DC0000FF",
                                "I (D51, moderate)"= "#DC0000FF",
                                "I (D51, severe)"= "#DC0000FF",
                                "I-BNT-I (D2)"= "#DC0000FF",
                                "I-BNT-I (D5)"= "#DC0000FF",
                                "I-I (D0)"= "#DC0000FF",
                                "I-I (D1)"= "#DC0000FF",
                                "I-I (D2)"= "#DC0000FF",
                                "I-I (D3)"= "#DC0000FF",
                                "I-I (D5)"= "#DC0000FF",
                                "ZF2001 (V3, D07)"= "#8491B4FF",
                                "ZF2001 (V3, D14)"= "#8491B4FF",
                                "ZF2001 (V3, D28)" = "#8491B4FF"
                              )
)



# Not immune processes -----
col_process_notimmune = list(
  "1" = c("Cellular" = "#3C5488FF",
          "Interspecies interaction" = "#00A087FF",
          "Metabolism" = "#4DBBD5FF",
          "Reg. of BP" = "#E64B35FF",
          "."="white"),
  "2" = c("Cellular" = "#3C5488FF",
          "Interspecies interaction" = "#00A087FF",
          "Metabolism" = "#4DBBD5FF",
          "Reg. of BP" = "#E64B35FF",
          "."="white"),
  "3" = c("Cellular" = "#3C5488FF",
          "Interspecies interaction" = "#00A087FF",
          "Metabolism" = "#4DBBD5FF",
          "Reg. of BP" = "#E64B35FF",
          "."="white"),
  "4" = c("Cellular" = "#3C5488FF",
          "Interspecies interaction" = "#00A087FF",
          "Metabolism" = "#4DBBD5FF",
          "Reg. of BP" = "#E64B35FF",
          "."="white"),
  "5" = c("Cellular" = "#3C5488FF",
          "Interspecies interaction" = "#00A087FF",
          "Metabolism" = "#4DBBD5FF",
          "Reg. of BP" = "#E64B35FF",
          "."="white"))


#Immune processes ------
col_process_immune = list(
  "1" = c("ADAPTIVE IMMUNE SYSTEM" = "#6f2dbd",
          "B CELL" = "#7b2cbf",
          "CELLULAR ADAPTIVE IMMUNE SYSTEM" = "#e0aaff",
          "HUMORAL ADAPTIVE IMMUNE SYSTEM" = "#9d4edd",
          "IMMUNOGLOBULIN MEDIATED RESPONSE"= "#5a189a",
          "T CELLS" = "#c77dff",
          "T HELPER CELLS" = "#c77dcc",
          "COMPLEMENT" = "#ffadc7",
          "ANTIVIRAL & IFN" = "#88d4ab",
          "ARPP" = "#14746f",
          "DENDRITIC CELL" = "#036666",
          "EOSINOPHIL" = "#67b99a",
          "INFLAMMATION" = "#99e2b4",
          "INNATE IMMUNE SYSTEM" = "#06d6a0",
          "MAST CELL" = "#56ab91",
          "MONOCYTE" = "#358f80",
          "NEUTROPHIL" = "#78c6a3",
          "NK CELL" = "#469d89",
          "MACROPHAGE" = "#14746f",
          "GENERAL" = "#343a40",
          "ANTIMICROBIAL" = "#48cae4",
          "."="white"),
  "2" = c("ADAPTIVE IMMUNE SYSTEM" = "#6f2dbd",
          "B CELLS" = "#7b2cbf",
          "CELLULAR ADAPTIVE IMMUNE SYSTEM" = "#e0aaff",
          "HUMORAL ADAPTIVE IMMUNE SYSTEM" = "#9d4edd",
          "IMMUNOGLOBULIN MEDIATED RESPONSE"= "#5a189a",
          "T CELL" = "#c77dff",
          "T HELPER CELLS" = "#c77dcc",
          "COMPLEMENT" = "#ffadc7",
          "ANTIVIRAL & IFN" = "#88d4ab",
          "ARPP" = "#14746f",
          "DENDRITIC CELL" = "#036666",
          "EOSINOPHIL" = "#67b99a",
          "INFLAMMATION" = "#99e2b4",
          "INNATE IMMUNE SYSTEM" = "#06d6a0",
          "MAST CELL" = "#56ab91",
          "MONOCYTE" = "#358f80",
          "NEUTROPHIL" = "#78c6a3",
          "NK CELL" = "#469d89",
          "MACROPHAGE" = "#14746f",
          "GENERAL" = "#343a40",
          "ANTIMICROBIAL" = "#48cae4",
          "."="white"),
  "3" = c("ADAPTIVE IMMUNE SYSTEM" = "#6f2dbd",
          "B CELL" = "#7b2cbf",
          "CELLULAR ADAPTIVE IMMUNE SYSTEM" = "#e0aaff",
          "HUMORAL ADAPTIVE IMMUNE SYSTEM" = "#9d4edd",
          "IMMUNOGLOBULIN MEDIATED RESPONSE"= "#5a189a",
          "T CELL" = "#c77dff",
          "T HELPER CELLS" = "#c77dcc",
          "COMPLEMENT" = "#ffadc7",
          "ANTIVIRAL & IFN" = "#88d4ab",
          "ARPP" = "#14746f",
          "DENDRITIC CELL" = "#036666",
          "EOSINOPHIL" = "#67b99a",
          "INFLAMMATION" = "#99e2b4",
          "INNATE IMMUNE SYSTEM" = "#06d6a0",
          "MAST CELL" = "#56ab91",
          "MONOCYTE" = "#358f80",
          "NEUTROPHIL" = "#78c6a3",
          "NK CELL" = "#469d89",
          "MACROPHAGE" = "#14746f",
          "GENERAL" = "#343a40",
          "ANTIMICROBIAL" = "#48cae4",
          "."="white"),
  "4" = c("ADAPTIVE IMMUNE SYSTEM" = "#6f2dbd",
          "B CELL" = "#7b2cbf",
          "CELLULAR ADAPTIVE IMMUNE SYSTEM" = "#e0aaff",
          "HUMORAL ADAPTIVE IMMUNE SYSTEM" = "#9d4edd",
          "IMMUNOGLOBULIN MEDIATED RESPONSE"= "#5a189a",
          "T CELL" = "#c77dff",
          "T HELPER CELLS" = "#c77dcc",
          "COMPLEMENT" = "#ffadc7",
          "ANTIVIRAL & IFN" = "#88d4ab",
          "ARPP" = "#14746f",
          "DENDRITIC CELL" = "#036666",
          "EOSINOPHIL" = "#67b99a",
          "INFLAMMATION" = "#99e2b4",
          "INNATE IMMUNE SYSTEM" = "#06d6a0",
          "MAST CELL" = "#56ab91",
          "MONOCYTE" = "#358f80",
          "NEUTROPHIL" = "#78c6a3",
          "NK CELL" = "#469d89",
          "MACROPHAGE" = "#14746f",
          "GENERAL" = "#343a40",
          "ANTIMICROBIAL" = "#48cae4",
          "."="white"),
  "5" = c("ADAPTIVE IMMUNE SYSTEM" = "#6f2dbd",
          "B CELL" = "#7b2cbf",
          "CELLULAR ADAPTIVE IMMUNE SYSTEM" = "#e0aaff",
          "HUMORAL ADAPTIVE IMMUNE SYSTEM" = "#9d4edd",
          "IMMUNOGLOBULIN MEDIATED RESPONSE"= "#5a189a",
          "T CELL" = "#c77dff",
          "T HELPER CELLS" = "#c77dcc",
          "COMPLEMENT" = "#ffadc7",
          "ANTIVIRAL & IFN" = "#88d4ab",
          "ARPP" = "#14746f",
          "DENDRITIC CELL" = "#036666",
          "EOSINOPHIL" = "#67b99a",
          "INFLAMMATION" = "#99e2b4",
          "INNATE IMMUNE SYSTEM" = "#06d6a0",
          "MAST CELL" = "#56ab91",
          "MONOCYTE" = "#358f80",
          "NEUTROPHIL" = "#78c6a3",
          "NK CELL" = "#469d89",
          "MACROPHAGE" = "#14746f",
          "GENERAL" = "#343a40",
          "ANTIMICROBIAL" = "#48cae4",
          "."="white"),
  "6" = c("ADAPTIVE IMMUNE SYSTEM" = "#6f2dbd",
          "B CELL" = "#7b2cbf",
          "CELLULAR ADAPTIVE IMMUNE SYSTEM" = "#e0aaff",
          "HUMORAL ADAPTIVE IMMUNE SYSTEM" = "#9d4edd",
          "IMMUNOGLOBULIN MEDIATED RESPONSE"= "#5a189a",
          "T CELL" = "#c77dff",
          "T HELPER CELLS" = "#c77dcc",
          "COMPLEMENT" = "#ffadc7",
          "ANTIVIRAL & IFN" = "#88d4ab",
          "ARPP" = "#14746f",
          "DENDRITIC CELL" = "#036666",
          "EOSINOPHIL" = "#67b99a",
          "INFLAMMATION" = "#99e2b4",
          "INNATE IMMUNE SYSTEM" = "#06d6a0",
          "MAST CELL" = "#56ab91",
          "MONOCYTE" = "#358f80",
          "NEUTROPHIL" = "#78c6a3",
          "NK CELL" = "#469d89",
          "MACROPHAGE" = "#14746f",
          "GENERAL" = "#343a40",
          "ANTIMICROBIAL" = "#48cae4",
          "."="white"),
  "7" =c("ADAPTIVE IMMUNE SYSTEM" = "#6f2dbd",
         "B CELL" = "#7b2cbf",
         "CELLULAR ADAPTIVE IMMUNE SYSTEM" = "#e0aaff",
         "HUMORAL ADAPTIVE IMMUNE SYSTEM" = "#9d4edd",
         "IMMUNOGLOBULIN MEDIATED RESPONSE"= "#5a189a",
         "T CELL" = "#c77dff",
         "T HELPER CELLS" = "#c77dcc",
         "COMPLEMENT" = "#ffadc7",
         "ANTIVIRAL & IFN" = "#88d4ab",
         "ARPP" = "#14746f",
         "DENDRITIC CELL" = "#036666",
         "EOSINOPHIL" = "#67b99a",
         "INFLAMMATION" = "#99e2b4",
         "INNATE IMMUNE SYSTEM" = "#06d6a0",
         "MAST CELL" = "#56ab91",
         "MONOCYTE" = "#358f80",
         "NEUTROPHIL" = "#78c6a3",
         "NK CELL" = "#469d89",
         "MACROPHAGE" = "#14746f",
         "GENERAL" = "#343a40",
         "ANTIMICROBIAL" = "#48cae4",
         "."="white"),
  "8" = c("ADAPTIVE IMMUNE SYSTEM" = "#6f2dbd",
          "B CELL" = "#7b2cbf",
          "CELLULAR ADAPTIVE IMMUNE SYSTEM" = "#e0aaff",
          "HUMORAL ADAPTIVE IMMUNE SYSTEM" = "#9d4edd",
          "IMMUNOGLOBULIN MEDIATED RESPONSE"= "#5a189a",
          "T CELL" = "#c77dff",
          "T HELPER CELLS" = "#c77dcc",
          "COMPLEMENT" = "#ffadc7",
          "ANTIVIRAL & IFN" = "#88d4ab",
          "ARPP" = "#14746f",
          "DENDRITIC CELL" = "#036666",
          "EOSINOPHIL" = "#67b99a",
          "INFLAMMATION" = "#99e2b4",
          "INNATE IMMUNE SYSTEM" = "#06d6a0",
          "MAST CELL" = "#56ab91",
          "MONOCYTE" = "#358f80",
          "NEUTROPHIL" = "#78c6a3",
          "NK CELL" = "#469d89",
          "MACROPHAGE" = "#14746f",
          "GENERAL" = "#343a40",
          "ANTIMICROBIAL" = "#48cae4",
          "."="white"),
  "9" = c("ADAPTIVE IMMUNE SYSTEM" = "#6f2dbd",
          "B CELL" = "#7b2cbf",
          "CELLULAR ADAPTIVE IMMUNE SYSTEM" = "#e0aaff",
          "HUMORAL ADAPTIVE IMMUNE SYSTEM" = "#9d4edd",
          "IMMUNOGLOBULIN MEDIATED RESPONSE"= "#5a189a",
          "T CELL" = "#c77dff",
          "T HELPER CELLS" = "#c77dcc",
          "COMPLEMENT" = "#ffadc7",
          "ANTIVIRAL & IFN" = "#88d4ab",
          "ARPP" = "#14746f",
          "DENDRITIC CELL" = "#036666",
          "EOSINOPHIL" = "#67b99a",
          "INFLAMMATION" = "#99e2b4",
          "INNATE IMMUNE SYSTEM" = "#06d6a0",
          "MAST CELL" = "#56ab91",
          "MONOCYTE" = "#358f80",
          "NEUTROPHIL" = "#78c6a3",
          "NK CELL" = "#469d89",
          "MACROPHAGE" = "#14746f",
          "GENERAL" = "#343a40",
          "ANTIMICROBIAL" = "#48cae4",
          "."="white"),
  "10" = c("ADAPTIVE IMMUNE SYSTEM" = "#6f2dbd",
           "B CELL" = "#7b2cbf",
           "CELLULAR ADAPTIVE IMMUNE SYSTEM" = "#e0aaff",
           "HUMORAL ADAPTIVE IMMUNE SYSTEM" = "#9d4edd",
           "IMMUNOGLOBULIN MEDIATED RESPONSE"= "#5a189a",
           "T CELL" = "#c77dff",
           "T HELPER CELLS" = "#c77dcc",
           "COMPLEMENT" = "#ffadc7",
           "ANTIVIRAL & IFN" = "#88d4ab",
           "ARPP" = "#14746f",
           "DENDRITIC CELL" = "#036666",
           "EOSINOPHIL" = "#67b99a",
           "INFLAMMATION" = "#99e2b4",
           "INNATE IMMUNE SYSTEM" = "#06d6a0",
           "MAST CELL" = "#56ab91",
           "MONOCYTE" = "#358f80",
           "NEUTROPHIL" = "#78c6a3",
           "NK CELL" = "#469d89",
           "MACROPHAGE" = "#14746f",
           "GENERAL" = "#343a40",
           "ANTIMICROBIAL" = "#48cae4",
           "."="white"),
  "11" = c("ADAPTIVE IMMUNE SYSTEM" = "#6f2dbd",
           "B CELL" = "#7b2cbf",
           "CELLULAR ADAPTIVE IMMUNE SYSTEM" = "#e0aaff",
           "HUMORAL ADAPTIVE IMMUNE SYSTEM" = "#9d4edd",
           "IMMUNOGLOBULIN MEDIATED RESPONSE"= "#5a189a",
           "T CELL" = "#c77dff",
           "T HELPER CELLS" = "#c77dcc",
           "COMPLEMENT" = "#ffadc7",
           "ANTIVIRAL & IFN" = "#88d4ab",
           "ARPP" = "#14746f",
           "DENDRITIC CELL" = "#036666",
           "EOSINOPHIL" = "#67b99a",
           "INFLAMMATION" = "#99e2b4",
           "INNATE IMMUNE SYSTEM" = "#06d6a0",
           "MAST CELL" = "#56ab91",
           "MONOCYTE" = "#358f80",
           "NEUTROPHIL" = "#78c6a3",
           "NK CELL" = "#469d89",
           "MACROPHAGE" = "#14746f",
           "GENERAL" = "#343a40",
           "ANTIMICROBIAL" = "#48cae4",
           "."="white"),
  "12" = c("ADAPTIVE IMMUNE SYSTEM" = "#6f2dbd",
           "B CELL" = "#7b2cbf",
           "CELLULAR ADAPTIVE IMMUNE SYSTEM" = "#e0aaff",
           "HUMORAL ADAPTIVE IMMUNE SYSTEM" = "#9d4edd",
           "IMMUNOGLOBULIN MEDIATED RESPONSE"= "#5a189a",
           "T CELL" = "#c77dff",
           "T HELPER CELLS" = "#c77dcc",
           "COMPLEMENT" = "#ffadc7",
           "ANTIVIRAL & IFN" = "#88d4ab",
           "ARPP" = "#14746f",
           "DENDRITIC CELL" = "#036666",
           "EOSINOPHIL" = "#67b99a",
           "INFLAMMATION" = "#99e2b4",
           "INNATE IMMUNE SYSTEM" = "#06d6a0",
           "MAST CELL" = "#56ab91",
           "MONOCYTE" = "#358f80",
           "NEUTROPHIL" = "#78c6a3",
           "NK CELL" = "#469d89",
           "MACROPHAGE" = "#14746f",
           "GENERAL" = "#343a40",
           "ANTIMICROBIAL" = "#48cae4",
           "."="white"))

```

# 2. Gene sets

## Gene sets construction

```{r eval=FALSE, echo=TRUE}
library(readr)
library(explore)
library(maditr)
library(GEOquery)
library(Matrix)
library(circlize)
library(RColorBrewer)
library(celldex)
library(biomaRt)
library(org.Hs.eg.db)
library(plotly)
library(DESeq2)
library(msigdbr)
library(ape)
library(GSVA)
library(sva)
library(clusterProfiler)
library(pheatmap)
library(EnhancedVolcano)
library(ggbeeswarm)
library(tidyverse)
library(corrr)
library(ggcorrplot)
library(FactoMineR)
library(factoextra)
library(esquisse)
library(corto)
library(factoextra)
library(reshape2)
library(ComplexHeatmap)
library(janitor)
```


### Blood transcription modules <https://www.nature.com/articles/ni.2789>

```{r}
#Converter ítens de "Module member genes" em linhas em uma tabela longa
btm_annotation_table_long_immune <- BTM_modules %>%
  separate_rows(`Module member genes`, sep = ",") %>% 
  filter(`Module category` == "immune")

#Salvar
write.csv(btm_annotation_table_long_immune, file = "btm_annotation_table_long_immune.csv")

#Display
head(btm_annotation_table_long_immune)

#Calculate Set size
btm_annotation_nested <- btm_annotation_table_long_immune %>%
  group_by(`Module title`) %>% #Somar e agrupar
  mutate(SetSize = n())

#Convert gene symbol column to aggregated rows
btm_annotation_nested <- btm_annotation_nested %>%
  group_by(`Module title`) %>%
  summarize(Genes = paste(`Module member genes`, collapse = ", "), .groups = "drop") %>% 
  left_join(btm_annotation_nested, by = "Module title") %>% 
  filter(Genes != "NA") %>% 
  relocate(Genes, .after="SetSize")

#Salvar
write.csv(btm_annotation_nested, file = "btm_annotation_genes.csv")

#Convert to unique values table
btm_annotation_unique = btm_annotation_nested %>% 
  select(-`Module member genes`)

btm_annotation_unique = btm_annotation_unique %>% 
  distinct() %>% 
  relocate(Genes, .after="SetSize") %>% 
  filter(Genes != "NA")

#salvar
write.csv(btm_annotation_unique, file = "btm_annotation_unique.csv")

#Display
head(btm_annotation_unique)

btm_annotation_table_long_immune
```

### CellMarker

Download database from
"<http://117.50.127.228/CellMarker/CellMarker_download_files/file/Cell_marker_Human.xlsx>"

```{r}
##################### Baixar tabela de marcadores celulares

url <- "http://117.50.127.228/CellMarker/CellMarker_download_files/file/Cell_marker_Human.xlsx" 
f <- tempfile(fileext = ".xlsx")
download.file(url, f)
cell_marker_data <- readxl::read_excel(f, 1)

##################### Sub-agrupar células por sistema imunológico

#Filtrar marcadores e células por tecido (Blood)
cells <- cell_marker_data %>%
    filter(tissue_class == "Blood") %>%
    select(cell_name, marker)

##################### Filtrar células
# Acessar todos os tipos de células no dataset de Blood
cellmarker_celltypes <- unique(cells$cell_name)

# Filtrar apenas os tipos que contêm "B cell" e plasmacell
Bcells_no = cellmarker_celltypes[grep("B cell", cellmarker_celltypes)] #44 outputs
Plasmacells_no = cellmarker_celltypes[grep("plasma", cellmarker_celltypes)] #5 outputs

# Filtrar apenas os tipos que contêm "T cell" e seus subtipos "CD4" e "CD8"
Tcells_no = cellmarker_celltypes[grep("T cell", cellmarker_celltypes)] #93 outputs
TCD4_no = cellmarker_celltypes[grep("CD4", cellmarker_celltypes)] #35 outputs
TCD8_no = cellmarker_celltypes[grep("CD8", cellmarker_celltypes)] #34 outputs
Treg_no = cellmarker_celltypes[grep("Treg", cellmarker_celltypes)] #7 outputs
Thelper_no = cellmarker_celltypes[grep("helper", cellmarker_celltypes)] #17 outputs

# Filtrar outros tipos celulares da resposta imune
Neutrophils_no = cellmarker_celltypes[grep("neutrophil", cellmarker_celltypes)] #7 outputs
Monocyte_no = cellmarker_celltypes[grep("monocyte", cellmarker_celltypes)] #14 outputs
DCells_no = cellmarker_celltypes[grep("dendritic", cellmarker_celltypes)] #25 outputs
Macrophages_no = cellmarker_celltypes[grep("macrophage", cellmarker_celltypes)] #7 outputs
NKcells_no = cellmarker_celltypes[grep("killer", cellmarker_celltypes)] # 11 outputs
Neutrophils_no = cellmarker_celltypes[grep("mastcell", cellmarker_celltypes)] #7 outputs

# Criar variáveis para cada célula para enriquecimento

# Filtrar apenas as células da resposta imune adaptativa
Bcells <- filter(cells, str_detect(cell_name, regex("B cell|plasma cell|plasmablast", ignore_case = TRUE)))
TCD4 <- filter(cells, str_detect(cell_name, regex("CD4", ignore_case = TRUE)))
TCD8 <- filter(cells, str_detect(cell_name, regex("CD8", ignore_case = TRUE)))
Treg <- filter(cells, str_detect(cell_name, regex("Treg", ignore_case = TRUE)))
Thelper <- filter(cells, str_detect(cell_name, regex("helper", ignore_case = TRUE)))


# Filtrar células da resposta imune inata
Neutrophils <- filter(cells, str_detect(cell_name, regex("neutrophil", ignore_case = TRUE)))
Monocyte <- filter(cells, str_detect(cell_name, regex("monocyte", ignore_case = TRUE)))
DCells <- filter(cells, str_detect(cell_name, regex("dendritic", ignore_case = TRUE)))
Macrophages <- filter(cells, str_detect(cell_name, regex("macrophage", ignore_case = TRUE)))
Mastcells <- filter(cells, str_detect(cell_name, regex("mast cell", ignore_case = TRUE)))
NKcells <- filter(cells, str_detect(cell_name, regex("killer", ignore_case = TRUE)))
Platelets <- filter(cells, str_detect(cell_name, regex("Platelet", ignore_case = TRUE)))
Eosinophil <- filter(cells, str_detect(cell_name, regex("Eosinophil", ignore_case = TRUE)))
Basophil <- filter(cells, str_detect(cell_name, regex("Basophil", ignore_case = TRUE)))
Granulocyte <- filter(cells, str_detect(cell_name, regex("Granulocyte", ignore_case = TRUE)))

#Agrupar todos os resultados em Immune cells
Immune_adap = rbind(Bcells, TCD4, TCD8, Treg, Thelper)
Immune_adap$Type = "Adaptive"

Immune_innate = rbind(Neutrophils, Monocyte, DCells, Macrophages, Mastcells, NKcells, Platelets, Eosinophil, Basophil, Granulocyte)
Immune_innate$Type = "Innate"

#Unir tabelas
Immune_cells = rbind(Immune_adap, Immune_innate) 

#Salvar
write.csv(Immune_cells, file = "CellMarker_ImmuneCells.csv")

esquisser(Immune_cells_markersbycells)
```

Top 20 cells by #markers

```{r}

Immune_cells_markersbycells = Immune_cells %>% 
  group_by(cell_name, Type) %>% 
  summarise(n_markers = n()) %>% 
  arrange(desc(n_markers))
Immune_cells_markersbycells_top = Immune_cells_markersbycells[1:20, ]

ggImmune_cells_cellsbyimmune = ggplot(Immune_cells_markersbycells_top) +
  aes(x = reorder(cell_name, n_markers), y = n_markers, fill = Type) +
  scale_fill_manual(values = c('Innate' = "#989898",
                               'Adaptive' = "#6DF8DF")) +
  geom_col() +
  labs(
    x = "#Markers",
    y = "Cells",
    title = "Top 20 cells by #markers"
  ) +
  coord_flip() +
  theme_minimal() +
  geom_text(aes(label = n_markers, y = n_markers), 
            position = position_dodge(width = 0.9), 
            vjust = -0.2, 
            hjust = -0.2, 
            size = 3)

ggsave(ggImmune_cells_cellsbyimmune, file= "CellMarker_Top10_cellsbyn_markers.png", width = 10, height = 5)

print(ggImmune_cells_cellsbyimmune)

```

Number of Cells by immune system

```{r}
Immune_cells_cellsbyimmune = Immune_cells %>% 
  group_by(Type, cell_name) %>% 
  summarise(n_cells = n()) %>% 
  group_by(Type) %>% 
  summarise(n_cells = n())

ggImmune_cells_cellsbyimmune = ggplot(Immune_cells_cellsbyimmune) +
  aes(x = reorder(Type, n_cells), y = n_cells, fill = Type) +
  geom_col() +
  scale_fill_manual(values = c('Innate' = "#989898",
                               'Adaptive' = "#6DF8DF")) +
  labs(
    x = "#Cells",
    y = "Immune system",
    title = "#Cells by immune system"
  ) +
  coord_flip() +
  theme_minimal() +
  geom_text(aes(label = n_cells, y = n_cells), 
            position = position_dodge(width = 0.9), 
            vjust = -0.2, 
            hjust = -0.2, 
            size = 3)
ggsave(ggImmune_cells_cellsbyimmune, file= "CellMarker_nCellsbyimmunesystem.png", width = 10, height = 2)

print(ggImmune_cells_cellsbyimmune)
```

### ImmuneGO

```{r}
organism = "org.Hs.eg.db"
install.packages("GO.db")
BiocManager::install(organism, character.only = TRUE)
library(organism, character.only = TRUE)
library(clusterProfiler)
library(GO.db)

#HGNC symbol to Entrez
x <- org.Hs.egSYMBOL
mapped_genes <- mappedkeys(x)
symbols_entrez <- as.data.frame(x[mapped_genes])

## Entrez to GO
x <- org.Hs.egGO
mapped_genes <- mappedkeys(x)
entrez_geneontology <- as.data.frame(x[mapped_genes])

## GO BP ancestors
go_bp_ancestor = GOBPANCESTOR %>% 
  as.data.frame()
colnames(go_bp_ancestor)[1] <- "go_id"
colnames(go_bp_ancestor)[2] <- "go_ancestor"

## GO ID to terms
go_bp = AnnotationDbi::select(GO.db, columns=c("GOID","TERM"), keys= "BP", keytype= "ONTOLOGY") %>% 
  clean_names() %>% 
  select(go_id = goid, term)

## GO child and ancestor
go_bp_ancestor = go_bp_ancestor %>% 
  inner_join(go_bp %>% rename(go_ancestor = go_id), by = "go_ancestor") %>% 
  rename(term_ancestor = term)

#GO ids to terms
go_bp = AnnotationDbi::select(GO.db, columns=c("GOID","TERM"), keys="BP", keytype="ONTOLOGY") %>% 
  clean_names() %>% 
  rename(go_id = goid) %>% 
  select(-ontology) %>% 
  right_join(entrez_geneontology, by = "go_id") %>% 
  right_join(symbols_entrez, by = "gene_id") %>% 
  clean_names() %>% 
  left_join(go_bp_ancestor, by = "go_id") %>% 
  distinct()

#GO immune process

immunego_go.bp = go_bp %>% 
  filter(term_ancestor == "immune system process") %>% 
  mutate(term = toupper(term))

ImmuneGO_Annotated_Genes_26_2_24 = ImmuneGO_Annotated_GO_8_2_24 %>% 
  select(-"...1") %>% 
  mutate(genes = str_replace_all(genes, "\\s+", "")) %>% 
  separate_rows(genes, sep = ",") %>% 
  full_join(immunego_go.bp %>% select(genes = symbol, process = term), by = join_by("process", "genes")) %>% 
  distinct()

ImmuneGO_Annotated_GO_26_2_24 = ImmuneGO_Annotated_Genes_26_2_24 %>% 
  group_by(process) %>%
  summarize(genes = paste0(genes, collapse = ",")) %>%
  inner_join(ImmuneGO_Annotated_Genes_26_2_24 %>% select(-genes) %>% distinct(), by = "process")

write.csv(ImmuneGO_Annotated_GO_26_2_24, file = "ImmuneGO_Annotated_GO_26_2_24.csv", row.names = F)

write.csv(ImmuneGO_Annotated_Genes_26_2_24, file = "ImmuneGO_Annotated_Genes_26_2_24.csv", row.names = F)

```

Group immune system genes by process and cell

```{r}
ImmuneGO_Annotated_unique = ImmuneGO_Annotated_Genes %>% 
  filter(go_term != "Manual")

#Agrupar genes de todos os processos do sistema imune ----

ImmuneGO_Adaptive <- ImmuneGO_Annotated_unique %>%
  filter(immune_system == "Adaptive") %>% 
  summarize(set_size = n(), 
            genes = paste(genes, collapse = ",")) %>% 
  mutate(process = "ADAPTIVE IMMUNE SYSTEM",
  `Gene set, short` = "Adaptive immune system",
  `Immune system` = "Adaptive",
  `Immune sub-system` = "General",
  `Immune tissue` = "General",
  `GO.TERM` = "Manual") %>% 
  clean_names()

ImmuneGO_Innate <- ImmuneGO_Annotated_unique %>%
  filter(immune_system == "Innate") %>% 
  summarize(set_size = n(), 
            genes = paste(genes, collapse = ",")) %>% 
  mutate(process = "INNATE IMMUNE SYSTEM",
  `Gene set, short` = "Innate immune system",
  `Immune system` = "Innate",
  `Immune sub-system` = "General",
  `Immune tissue` = "General",
  `GO.TERM` = "Manual") %>% 
  clean_names()

ImmuneGO_Complement <- ImmuneGO_Annotated_unique %>%
  filter(immune_system == "Complement") %>% 
  summarize(set_size = n(), 
            genes = paste(genes, collapse = ",")) %>% 
  mutate(process = "COMPLEMENT",
  `Gene set, short` = "Complement immune system",
  `Immune system` = "Complement",
  `Immune sub-system` = "General",
  `Immune tissue` = "General",
  `GO.TERM` = "Manual") %>% 
  clean_names()

ImmuneGO_Inflammation <- ImmuneGO_Annotated_unique %>%
  filter(immune_sub_system == "Inflammation") %>% 
  summarize(set_size = n(), 
            genes = paste(genes, collapse = ",")) %>% 
  mutate(process = "INFLAMMATION",
  `Gene set, short` = "Inflammation",
  `Immune system` = "Innate",
  `Immune sub-system` = "Inflammation",
  `Immune tissue` = "General",
  `GO.TERM` = "Manual") %>% 
  clean_names()

ImmuneGO_Interferon <- ImmuneGO_Annotated_unique %>%
  filter(immune_sub_system %in% c("Antiviral", "Interferon")) %>% 
  summarize(set_size = n(), 
            genes = paste(genes, collapse = ",")) %>% 
  mutate(process = "ANTIVIRAL & IFN",
  `Gene set, short` = "Antiviral and Interferon",
  `Immune system` = "Innate",
  `Immune sub-system` = "Antiviral and Interferon",
  `Immune tissue` = "General",
  `GO.TERM` = "Manual") %>% 
  clean_names()

ImmuneGO_ARPP = ImmuneGO_Annotated_unique %>%
  filter(immune_sub_system %in% c("PRR", "APC")) %>% 
  summarize(set_size = n(), 
            genes = paste(genes, collapse = ",")) %>% 
  mutate(process = "ARPP",
  `Gene set, short` = "Antigen receptors, processing and presentation",
  `Immune system` = "Innate",
  `Immune sub-system` = "APC",
  `Immune tissue` = "APC",
  `GO.TERM` = "Manual") %>% 
  clean_names()


ImmuneGO_Innate_DC <- ImmuneGO_Annotated_unique %>%
  filter(immune_tissue == "Dendritic cell") %>% 
  summarize(set_size = n(), 
            genes = paste(genes, collapse = ",")) %>% 
  mutate(process = "DENDRITIC CELL",
  `Gene set, short` = "Dendritic cell",
  `Immune system` = "Innate",
  `Immune sub-system` = "Cell",
  `Immune tissue` = "APC",
  `GO.TERM` = "Manual") %>% 
  clean_names()

ImmuneGO_Innate_Macrophage <- ImmuneGO_Annotated_unique %>%
  filter(immune_tissue == "Macrophage") %>% 
  summarize(set_size = n(), 
            genes = paste(genes, collapse = ",")) %>% 
  mutate(process = "MACROPHAGE",
  `Gene set, short` = "Macrophage",
  `Immune system` = "Innate",
  `Immune sub-system` = "Cell",
  `Immune tissue` = "APC",
  `GO.TERM` = "Manual") %>% 
  clean_names()

ImmuneGO_Innate_Mastcell <- ImmuneGO_Annotated_unique %>%
  filter(immune_tissue == "Mast cell") %>% 
  summarize(set_size = n(), 
            genes = paste(genes, collapse = ",")) %>% 
  mutate(process = "MAST CELL",
  `Gene set, short` = "Mast cell",
  `Immune system` = "Innate",
  `Immune sub-system` = "Cell",
  `Immune tissue` = "Mast cell",
  `GO.TERM` = "Manual") %>% 
  clean_names()

ImmuneGO_Innate_Neutrophil <- ImmuneGO_Annotated_unique %>%
  filter(immune_tissue == "Neutrophil") %>% 
  summarize(set_size = n(), 
            genes = paste(genes, collapse = ",")) %>% 
  mutate(process = "NEUTROPHIL",
  `Gene set, short` = "Neutrophil",
  `Immune system` = "Innate",
  `Immune sub-system` = "Cell",
  `Immune tissue` = "Neutrophil",
  `GO.TERM` = "Manual") %>% 
  clean_names()

ImmuneGO_Innate_Eosinophil <- ImmuneGO_Annotated_unique %>%
  filter(immune_tissue == "Eosinophil") %>% 
  summarize(set_size = n(), 
            genes = paste(genes, collapse = ",")) %>% 
  mutate(process = "EOSINOPHIL",
  `Gene set, short` = "Eosinophil",
  `Immune system` = "Innate",
  `Immune sub-system` = "Cell",
  `Immune tissue` = "Eosinophil",
  `GO.TERM` = "Manual") %>% 
  clean_names()

ImmuneGO_Innate_NK <- ImmuneGO_Annotated_unique %>%
  filter(immune_tissue == "NK cell") %>% 
  summarize(set_size = n(), 
            genes = paste(genes, collapse = ",")) %>% 
  mutate(process = "NK CELL",
  `Gene set, short` = "Natural Killer Cell",
  `Immune system` = "Innate",
  `Immune sub-system` = "Cell",
  `Immune tissue` = "NK cell",
  `GO.TERM` = "Manual") %>% 
  clean_names()

ImmuneGO_Innate_Monocyte <- ImmuneGO_Annotated_unique %>%
  filter(immune_tissue == "Monocyte") %>% 
  summarize(set_size = n(), 
            genes = paste(genes, collapse = ",")) %>% 
  mutate(process = "MONOCYTE",
  `Gene set, short` = "Monocyte",
  `Immune system` = "Innate",
  `Immune sub-system` = "Cell",
  `Immune tissue` = "APC",
  `GO.TERM` = "Manual") %>% 
  clean_names()

ImmuneGO_Adap_Cellular <- ImmuneGO_Annotated_unique %>% 
  filter(immune_sub_system == "Cellular") %>% 
  summarize(set_size = n(), 
            genes = paste(genes, collapse = ",")) %>% 
  mutate(process = "CELLULAR ADAPTIVE IMMUNE SYSTEM",
  `Gene set, short` = "Cellular adaptive immune system",
  `Immune system` = "Adaptive",
  `Immune sub-system` = "Cellular",
  `Immune tissue` = "Cellular",
  `GO.TERM` = "Manual") %>% 
  clean_names()


ImmuneGO_Adap_Tcell <- ImmuneGO_Annotated_unique %>%
  filter(immune_tissue == "T cell") %>% 
  summarize(set_size = n(), 
            genes = paste(genes, collapse = ",")) %>% 
  mutate(process = "T CELL",
  `Gene set, short` = "T cell",
  `Immune system` = "Adaptive",
  `Immune sub-system` = "Cellular",
  `Immune tissue` = "Cellular",
  `GO.TERM` = "Manual") %>% 
  clean_names()

ImmuneGO_Adap_Thelper <- ImmuneGO_Annotated_unique %>%
  filter(immune_tissue == "T helper") %>% 
  summarize(set_size = n(), 
            genes = paste(genes, collapse = ",")) %>% 
  mutate(process = "T HELPER CELLS",
  `Gene set, short` = "T helper cell",
  `Immune system` = "Adaptive",
  `Immune sub-system` = "Cellular",
  `Immune tissue` = "Cellular",
  `GO.TERM` = "Manual") %>% 
  clean_names()

ImmuneGO_Adap_Humoral <- ImmuneGO_Annotated_unique %>%
  filter(immune_sub_system == "Humoral") %>% 
  summarize(set_size = n(), 
            genes = paste(genes, collapse = ",")) %>% 
  mutate(process = "HUMORAL ADAPTIVE IMMUNE SYSTEM",
  `Gene set, short` = "Humoral adaptive immune system",
  `Immune system` = "Adaptive",
  `Immune sub-system` = "Humoral",
  `Immune tissue` = "Humoral",
  `GO.TERM` = "Manual") %>% 
  clean_names()

ImmuneGO_Adap_Bcell <- ImmuneGO_Annotated_unique %>%
  filter(immune_tissue == "B cell") %>% 
  summarize(set_size = n(), 
            genes = paste(genes, collapse = ",")) %>% 
  mutate(process = "B CELL",
  `Gene set, short` = "B cell",
  `Immune system` = "Adaptive",
  `Immune sub-system` = "Humoral",
  `Immune tissue` = "B cell",
  `GO.TERM` = "Manual") %>% 
  clean_names()

ImmuneGO_Adap_Ig <- ImmuneGO_Annotated_unique %>%
  filter(immune_tissue == "B cell") %>% 
  summarize(set_size = n(), 
            genes = paste(genes, collapse = ",")) %>% 
  mutate(process = "IMMUNOGLOBULIN MEDIATED RESPONSE",
  `Gene set, short` = "Ig-mediated response",
  `Immune system` = "Adaptive",
  `Immune sub-system` = "Humoral",
  `Immune tissue` = "B cell",
  `GO.TERM` = "Manual") %>% 
  clean_names()

ImmuneGO_other <- ImmuneGO_Annotated_unique %>%
  filter(immune_system == "General" |
         immune_tissue == "Leukocyte") %>% 
  summarize(set_size = n(), 
            genes = paste(genes, collapse = ",")) %>% 
  mutate(process = "GENERAL",
  `Gene set, short` = "General",
  `Immune system` = "General",
  `Immune sub-system` = "General",
  `Immune tissue` = "General",
  `GO.TERM` = "Manual") %>% 
  clean_names()


ImmuneGO_antimicrobial <- ImmuneGO_Annotated_unique %>%
  filter(immune_tissue == "Antimicrobial") %>% 
  summarize(set_size = n(), 
            genes = paste(genes, collapse = ",")) %>% 
  mutate(process = "ANTIMICROBIAL",
  `Gene set, short` = "Antimicrobial immune response",
  `Immune system` = "General",
  `Immune sub-system` = "General",
  `Immune tissue` = "General",
  `GO.TERM` = "Manual") %>% 
  clean_names()


# Unite sets -----

ImmuneGO_Annotated_GO = bind_rows(ImmuneGO_Adaptive, 
                            ImmuneGO_Innate, 
                            ImmuneGO_Complement, 
                            ImmuneGO_Inflammation,
                            ImmuneGO_Interferon,
                            ImmuneGO_ARPP,
                            ImmuneGO_Innate_Macrophage,
                            ImmuneGO_Innate_Mastcell,
                            ImmuneGO_Innate_Neutrophil,
                            ImmuneGO_Innate_Eosinophil,
                            ImmuneGO_Innate_NK,
                            ImmuneGO_Innate_Monocyte,
                            ImmuneGO_Adap_Cellular,
                            ImmuneGO_Adap_Tcell,
                            ImmuneGO_Adap_Thelper,
                            ImmuneGO_Adap_Humoral,
                            ImmuneGO_Adap_Bcell,
                            ImmuneGO_Adap_Ig,
                            ImmuneGO_other,
                            ImmuneGO_antimicrobial,
                            ImmuneGO_Innate_DC,
                            ImmuneGO_Annotated_unique) %>% 
  select(process:go_term, set_size:genes) %>% 
  distinct()

ImmuneGO_Annotated_Genes = ImmuneGO_Annotated_GO %>% 
   separate_rows(genes, sep = ",")

# Separate TCR and BCR repertoire genes
TCR_BCR_repertoire = ImmuneGO_Annotated_Genes %>% 
  filter(grepl("^TR", genes) & process == "ADAPTIVE IMMUNE SYSTEM" |
           grepl("^IG", genes) & process == "ADAPTIVE IMMUNE SYSTEM" |
           grepl("^TR", genes) & process == "CELLULAR ADAPTIVE IMMUNE SYSTEM" |
         grepl("^IG", genes) & process == "HUMORAL ADAPTIVE IMMUNE SYSTEM") %>% 
  mutate(process = if_else(process == "CELLULAR ADAPTIVE IMMUNE SYSTEM", "TCR REPERTOIRE", 
                           "BCR REPERTOIRE"),
         gene_set_short = if_else(process == "TCR REPERTOIRE", 
                                  "TCR REPERTOIRE", 
                           "BCR REPERTOIRE"),
         process = if_else(grepl("^TR", genes), "TCR REPERTOIRE", "BCR REPERTOIRE"),
         immune_sub_system = if_else(process == "TCR REPERTOIRE", "Cellular", "Humoral"),
         immune_tissue = if_else(process == "TCR REPERTOIRE", "Cellular", "Humoral"))

immunego_manual = ImmuneGO_Annotated_Genes %>%
  filter(go_term == "Manual") %>% 
  anti_join(TCR_BCR_repertoire %>% select(genes) %>% distinct(), by = "genes") %>% 
  bind_rows()

#Verify if all genes are annotated. If there are no left genes, you are good to go
a = ImmuneGO_Annotated_Genes %>% filter(go_term == "Manual") %>% select(genes) %>% distinct()
ImmuneGO_Annotated_Genes %>% select(process, genes, immune_system) %>% anti_join(a, by = "genes")


#Salvar
write.csv(ImmuneGO_Annotated_GO, file = paste0("ImmuneGO_Annotated_GO_", today(), ".csv"), row.names = F)
saveRDS(ImmuneGO_Annotated_GO, file = paste0("ImmuneGO_Annotated_GO_", today(), ".rds"))
saveRDS(ImmuneGO_Annotated_Genes, file = paste0("ImmuneGO_Annotated_Genes_", today(), ".rds"))

#Display
view(ImmuneGO_Annotated_GO)
view(ImmuneGO_Annotated_Genes)

```

Top 10 gene sets
```{r}
#Top 10 gene sets
ImmuneGO_Annotated_unique_top10 = ImmuneGO_Annotated_GO %>% 
  clean_names() %>% 
  arrange(desc(set_size)) %>% 
  slice_head(n = 20) %>% 
  mutate(immune_system = if_else(gene_set_short == "Adaptive Immune response", "Adaptive", immune_system))

#Gráfico de barras
ImmuneGO_top10_barplot = ggplot(ImmuneGO_Annotated_unique_top10) +
 aes(x = reorder(gene_set_short, set_size), 
     weight = set_size, 
     fill = immune_system) +  # Use reorder para reordenar as barras +
  geom_bar()+
  scale_fill_manual(
    values = c(Adaptive = "#6DF8DF",
    Complement = "#e9edc9",
    General = "#bde0fe",
    Innate = "#989898",
    Undefined = "#edede9")) +
 labs(y = "Set size", title = "ImmuneGO", subtitle = "Top 10 GO, by set size") +
 coord_flip() +
 theme_minimal()

#Salvar
ggsave(ImmuneGO_top10_barplot, file="ImmuneGO_top10_barplot.png", width = 10, height = 6)

#Display
head(ImmuneGO_top10_barplot)
```



### ImmuneGO + BTM

```{r}
btm_annotation_genes = readRDS(here("VaxGO gene sets", "btm_annotation_genes.rds"))
ImmuneGO_Annotated_Genes = readRDS(here("VaxGO gene sets", "ImmuneGO_Annotated_Genes_2024-03-26.rds"))
ImmuneGO_Annotated_GO = read_csv(here("VaxGO gene sets", "ImmuneGO_Annotated_GO_2024_03_26.csv"))

#Unite genes
immunego_btm_genes = ImmuneGO_Annotated_Genes %>% 
              select(genes) %>% 
              distinct() %>% 
              mutate(set = "ImmuneGO") %>% 
  bind_rows(btm_annotation_genes %>% 
              select(genes) %>% 
              distinct() %>% 
              mutate(set = "BTM")) %>% 
  arrange(genes)

immunego_btm_genes_joined = immunego_btm_genes %>% 
  bind_rows(immunego_btm_genes %>% mutate(set = "BTM and ImmuneGO")) %>% 
  distinct()


#Unite subsets
btm_annotation_grouped = btm_annotation_genes %>% 
  group_by(complete_name) %>% 
  rename(process = complete_name) %>% 
  mutate(genes = paste0(genes, collapse = ",")) %>% 
  distinct() %>% 
  mutate(genes = paste0(genes, collapse = ","),
            immune_set = "BTM by Modules") %>% 
    select(process,
           immune_system:cells,
           set_size = module_size,
           genes,
           immune_set)

ImmuneGO_annotation_modules = ImmuneGO_Annotated_GO %>% 
  filter(go_term != "Manual") %>% 
    select(process = gene_set_short, 
           immune_system:immune_tissue,
           set_size, genes) %>% 
    rename(cells = immune_tissue,
           immune_subsystem = immune_sub_system) %>% 
    mutate(immune_set = "ImmuneGO by Modules")

immunego_btm_specific_grouped = btm_annotation_grouped %>% 
  bind_rows(ImmuneGO_annotation_modules)
  
immunego_btm_specific_grouped %>% 
  write_csv(file = here("VaxGO gene sets", "ImmuneGO_BTM_grouped_genesets_specific.csv"))

  
#By immune system
btm_immunesystem = btm_annotation_genes %>% 
  select(immune_system, genes) %>% 
  distinct() %>% 
  rename(process = immune_system) %>% 
  group_by(process) %>% 
  summarize(genes = paste0(genes, collapse = ","),
            set_size = n(),
            process = str_to_upper(process),
            # id = process,
            immune_system = process,
            immune_subsystem = process,
            cells = process,
            complete_name = process,
            annotation_level = "Manual",
            immune_set = "BTM by Immune System") %>% 
  distinct()

#By immune subsystem
btm_immunesubsystem = btm_annotation_genes %>% 
  select(immune_subsystem, genes) %>% 
  distinct() %>% 
  rename(process = immune_subsystem) %>% 
  group_by(process) %>% 
  summarize(genes = paste0(genes, collapse = ","),
            process = str_to_upper(process),
            set_size = n(),
            # id = process,
            immune_system = process,
            immune_subsystem = process,
            cells = process,
            complete_name = process,
            annotation_level = "Manual",
            immune_set = "BTM by Immune Sub-System") %>% 
  distinct() 
#By cell
btm_cells = btm_annotation_genes %>% 
  select(cells, genes) %>%
  distinct() %>% 
  rename(process = cells) %>% 
  group_by(process) %>% 
  summarize(genes = paste0(genes, collapse = ","),
            set_size = n(),
            process = str_to_upper(process),
            # id = process,
            immune_system = process,
            immune_subsystem = process,
            cells = process,
            annotation_level = "Manual",
            complete_name = process,
            immune_set = "BTM by Immune Cell") %>% 
  distinct() 
#Merge
btm_manual = bind_rows(btm_immunesystem,
          btm_immunesubsystem,
          btm_cells) %>% 
  arrange(complete_name) %>% 
  select(complete_name, everything()) %>% 
  distinct()

btm_manual_grouped = bind_rows(btm_manual,
                               btm_annotation_grouped) %>% 
  distinct()


#Merge ImmuneGO and BTM main processes
immunego_btm_manual_grouped = btm_manual %>% 
  select(process, genes, immune_set, set_size) %>% 
  mutate(immune_set = "BTM") %>% 
  bind_rows(ImmuneGO_Annotated_GO %>% 
              filter(go_term == "Manual") %>% 
              select(process, genes, set_size) %>% 
              mutate(immune_set = "ImmuneGO")) %>% 
  arrange(process) %>% 
  mutate(process = case_when(process == "ADAPTIVE" ~ "ADAPTIVE IMMUNE SYSTEM",
                              process == "ALL" ~ "IMMUNE SYSTEM",
                              process == "GENERAL" ~ "LEUKOCYTES",
                              process == "CELLS" ~ "LEUKOCYTES",
                              process == "DCS" ~ "DENDRITIC CELL",
                              process == "APP" ~ "ARPP",
                              process == "CELLULAR" ~ "CELLULAR ADAPTIVE IMMUNE SYSTEM",
                              process == "DCS, MYELOID" ~ "DENDRITIC CELL",
                              process == "GENERIC" ~ "LEUKOCYTES",
                              process == "HUMORAL" ~ "HUMORAL ADAPTIVE IMMUNE SYSTEM",
                              process == "INNATE" ~ "INNATE IMMUNE SYSTEM",
                              process == "INTERFERON" ~ "ANTIVIRAL & IFN",
                              process == "NK" ~ "NK CELL",
                              process == "B CELL, PLASMA CELL" ~ "B CELL",
                              process == "PLASMA CELL" ~ "B CELL",
                              process == "PRR" ~ "ARPP",
                              process == "SIGNALING" ~ "SIGNALING",
                              process == "TH2" ~ "T CELL",
                              process == "TH1" ~ "T CELL",
                              process == "T HELPER CELLS" ~ "T CELL",
                              TRUE ~ process
                              )) %>% 
  distinct() %>% 
  bind_rows(
    immunego_btm_specific_grouped)

immunego_btm_manual_grouped_genes = immunego_btm_manual_grouped %>%
  separate_rows(genes, sep = ",") %>% 
  distinct()

immunego_btm_manual_grouped_genes = immunego_btm_manual_grouped_genes %>% 
  bind_rows(
    immunego_btm_manual_grouped_genes %>% 
      mutate(immune_set = "BTM and ImmuneGO") %>% 
      distinct()
  )
  
  
immunego_btm_manual_grouped %>% 
  write_csv(file = here("VaxGO gene sets", "ImmuneGO_BTM_grouped_genesets.csv"))

immunego_btm_manual_grouped_genes %>% 
  write_csv(file = here("VaxGO gene sets", "ImmuneGO_BTM_grouped_genes.csv"))
  
```
Plot relative frequency of genes by main process
```{r fig.height=5, fig.width=7}

#Get exclusive genes
immunego_only_genes = immunego_btm_manual_grouped_genes %>% 
  filter(immune_set == "ImmuneGO") %>% 
  select(-immune_set, -set_size) %>% 
  anti_join(immunego_btm_manual_grouped_genes %>% 
            filter(immune_set == "BTM"),
            by = "genes") %>% 
  mutate(immune_set = "ImmuneGO only") %>% 
  distinct()

btm_only_genes = immunego_btm_manual_grouped_genes %>% 
  filter(immune_set == "BTM") %>% 
  select(-immune_set, -set_size) %>% 
  anti_join(immunego_btm_manual_grouped_genes %>% 
            filter(immune_set == "ImmuneGO"),
            by = "genes") %>% 
  mutate(immune_set = "BTM only") %>% 
  distinct()

btm_immune_joined_stats = immunego_only_genes %>% 
  bind_rows(btm_only_genes) %>% 
  distinct() %>% 
  mutate(immune_set = "BTM and ImmuneGO") %>% 
  select(process, genes, immune_set) %>% 
  distinct() %>% 
  group_by(process) %>% 
  distinct() %>% 
  mutate(ntotal=n()) %>% 
  ungroup() %>% 
  select(process, ntotal) %>% 
  distinct() %>% 
  inner_join(immunego_only_genes %>% 
               bind_rows(btm_only_genes), by = "process") %>% 
  group_by(process, immune_set) %>% 
  mutate(n = n(),
         genes = paste0(genes, collapse = ",")) %>% 
  select(-genes) %>% 
  distinct() %>% 
  ungroup() %>% 
  mutate(n_perc = (n/ntotal) * 100)

btm_immune_joined_stats_plot = btm_immune_joined_stats %>% 
  filter(!process %in% c("LYMPHOCYTE", 
                         "IMMUNOGLOBULIN MEDIATED RESPONSE",
                         "MYELOID CELL", 
                         "SIGNALING", 
                         "IMMUNE SYSTEM")) %>%
  mutate(process = str_to_title(process),
         process = fct_relevel(process, 
                               "Leukocytes",
                               "Adaptive Immune System",
                               "Innate Immune System",
                               "Complement"	,
                               "Cellular Adaptive Immune System",
                               "Humoral Adaptive Immune System",
                               )) %>% 
 ggplot() +
  aes(x = n_perc, y = fct_rev(process), fill = immune_set) +
  geom_bar(stat = "summary", fun = "sum", position = "fill") +
  scale_fill_manual(values = c("BTM only" = "#5e60ce",
                               "ImmuneGO only" = "#00A087FF")) +
  theme_minimal() +
  theme(legend.position = "right",
        strip.text.x = element_text(color = "black", size = 12), 
        axis.text.x = element_text(color = "black", size = 10),
        axis.text.y = element_text(vjust = 0.5, size = 10, color = "black"),
        axis.line = element_line(size = 0.5, colour = "black", linetype=1),
        axis.ticks = element_line(size = 0.5, color = "black")) +
  labs(x = "%Coverage",
       y = "",
       fill = "Immune set",
       subtitle = "Merged data",
       title = "ImmuneGO and BTM main processes genes") +
  scale_x_continuous(expand = c(0,0)) +
  geom_text(aes(label = paste0(round(n_perc, 0))), 
            position = position_fill(vjust = 0.5), 
            size = 3, 
            color = "white",
            fontface = "bold") 

btm_immune_joined_stats_plot %>% 
  ggsave(filename = here("Figures", "ImmuneGO_BTM_Joined_Stats_barplot.png"),
         width = 10,
         height = 7)

btm_immune_joined_stats_plot
```


### VAX MSigDB


```{r}
#Importar arquivo
VaxSigDB_genesets = VAX_GeneSets_Annotated_26_10_23

# Divida os dados em blocos de informações para cada vacina
vaccines_att <- split(VaxSigDB_genesets$Col2, VaxSigDB_genesets$Col1)

# Combine data
VAX_Genesets <- bind_rows(lapply(vaccines_att, function(x) data.frame(t(x))), .id = "V")

# Name a new index column
colnames(VAX_Genesets) <- paste0("V", seq_len(ncol(VAX_Genesets)))

# Transpose
rownames(VAX_Genesets) <- VAX_Genesets$V1
VAX_Genesets = VAX_Genesets %>% 
  t() %>%
  as.data.frame()

#Order columns
VAX_Genesets = VAX_Genesets %>% 
  select(STANDARD_NAME, everything()) %>% 
  filter(STANDARD_NAME != "STANDARD_NAME") 
rownames(VAX_Genesets) = VAX_Genesets$SYSTEMATIC_NAME
VAX_Genesets = VAX_Genesets %>% select(STANDARD_NAME, 
                                       DESCRIPTION_BRIEF,
                                       DESCRIPTION_FULL,
                                       PMID,
                                       AUTHORS,
                                       SYSTEMATIC_NAME,
                                       GENE_SYMBOLS,
                                       everything())

#Anotar novas colunas
VAX_Genesets <- VAX_Genesets %>%
  mutate(
    SAMPLE_SOURCE = ifelse(
      grepl("pbmc|peripheral blood mononuclear cell|peripheral blood mononuclear cells", DESCRIPTION_FULL, ignore.case = TRUE) |
      grepl("pbmc|peripheral blood mononuclear cell|peripheral blood mononuclear cells", DESCRIPTION_BRIEF, ignore.case = TRUE),
      "PBMC",
      ifelse(
        grepl("whole blood", DESCRIPTION_FULL, ignore.case = TRUE) |
        grepl("whole blood", DESCRIPTION_BRIEF, ignore.case = TRUE) |
        grepl("_BLOOD", STANDARD_NAME, ignore.case = TRUE)  ,
        "WHOLE BLOOD",
        "OTHER")),
    REGULATION = ifelse(
      grepl("up", DESCRIPTION_BRIEF, ignore.case = TRUE),
      "UP",
      ifelse(
        grepl("down", DESCRIPTION_BRIEF, ignore.case = TRUE),
        "DOWN",
        "UNKNOWN")))

#Reordenar colunas
VAX_Genesets = VAX_Genesets %>% select(SAMPLE_SOURCE, REGULATION, everything())

# Salvar
write.csv(VAX_Genesets, file = "VAXSigDB_Annotated.csv") #Continuar a anotação no Google Sheets

head(VAX_Genesets)
```

 Long table with Gene symbols

```{r}
############## Long table with Gene symbols

VAX_Genesets = VAXSigDB_Annotated

VAX_Genes <- VAX_Genesets %>%
  mutate(GENE = GENE_SYMBOLS) %>% 
  separate_rows(GENE, sep = ",") %>% 
  select(GENE, everything()) %>%
  group_by(STANDARD_NAME) %>%
  mutate(SetSize = n()) %>% 
  relocate(SetSize, .after = GENE_SYMBOLS)

VAX_GeneSets_annotated <- VAX_Genes %>%
  select(-GENE) %>% 
  distinct()

#Salvar
write.csv(VAX_Genes, file = "VAX_Genes_Annotated_RAW.csv")
write.csv(VAX_GeneSets_annotated, file = "VAX_Genesets_Annotated_RAW.csv")

#Display
head(VAX_GeneSets_annotated)
```

Filtering

```{r}
unique(VAX_GeneSets_annotated$STUDY_TYPE) 
#"CORRELATION"
#"VACCINE"
#"SUBSET"
#"CHALLENGE"

unique(VAX_GeneSets_annotated$STUDY_SUBTYPE)
# "ANTIBODY"				
# "VAC ONLY"				
# "BOOSTER"				
# "VAC VS CTRL"				
# "SIGNATURE"				
# "ADVERSE EVENT"			
# "COMPARISON"				
# "RESPONDER VS NON RESPONDER"				
# "AGE"				
# "RESPONDER"				
# "ADJUVANT"				
# "NON RESPONDER"				
# "ANTIMICROBIAL ACTIVITY"				
# "CELLS"				
# "CHALLENGE"				
# "TOP DEGS"				
# "PROTECTION"				
# "TRAINING SET"				
# "EXON ANALYSIS"				
# "DOSE"				
# "VAC VS OTHER"


# Subset de vacinas para as seguintes análises comparativas
VAX_GeneSets_Blood_PBMC_VAC = VAX_GeneSets_annotated %>% 
  filter(SAMPLE_SOURCE %in% c("PBMC", "WHOLE BLOOD"), 
         STUDY_TYPE %in% c("VACCINE"))

# Subset de vacinas com resultados de comparações entre vacinas, adjuvantes, idade e sexo
VAX_GeneSets_Blood_PBMC_CORR = VAX_GeneSets_annotated %>% 
  filter(SAMPLE_SOURCE %in% c("PBMC", "WHOLE BLOOD"), 
         STUDY_TYPE %in% c("CORRELATION"))

# Subset com resultados de correlações com resposta inata, celular e humoral
VAX_GeneSets_Blood_PBMC_SUBSETS = VAX_GeneSets_annotated %>% 
  filter(SAMPLE_SOURCE %in% c("PBMC", "WHOLE BLOOD"), 
         STUDY_TYPE %in% c("SUBSET"))


# Subset de vacinas com genes associados a eventos adversos
VAX_GeneSets_Blood_PBMC_CHALLENGE = VAX_GeneSets_annotated %>% 
  filter(SAMPLE_SOURCE %in% c("PBMC", "WHOLE BLOOD"), 
         STUDY_TYPE %in% c("CHALLENGE"))

VAX_GeneSets_Blood_PBMC_VAC.stats = VAX_GeneSets_Blood_PBMC_VAC %>% 
  group_by(VACCINE, PLATFORM, REGULATION, `TARGET_PATHOGEN_DISEASE`, MICROBE_TYPE) %>% 
  summarise(Count = n()) %>% 
  arrange(desc(Count))

```



# 3. Select studies and retrieve data

## Download and process data

Download general data (counts data and metadata). GEO supplementary
files have the counts matrices. Download studies individually. I
recommend downloading large files from the GEO website.

```{r eval=FALSE, message=FALSE, warning=FALSE, echo=TRUE, cache = TRUE}
# Function to read and standardize files
read_standardize <- function(file) {
  files <- read.table(file, header = FALSE, sep = "\t")
  if (ncol(files) >= 2) {
    colnames(file)[1:2] <- c("genes", tools::file_path_sans_ext(basename(file)))
    return(files)
  } else {
    return(NULL)
  }
}
```

GSE189039

```{r eval=FALSE, message=FALSE, warning=FALSE, echo=TRUE, cache = TRUE}
###### GSE189039 ----
#Metadata 
filename = "GSE189039"
GSE189039 <- getGEO("GSE189039", GSEMatrix = TRUE) 
GSE189039 <- GSE189039[[1]]
GSE189039_metadata = GSE189039 %>% 
  pData() %>% 
  select( 
    geo_sample = "geo_accession", 
    subject_id = "title", age = "age:ch1", 
    sex = "gender:ch1", disease_state = "diseas state:ch1", 
    severity = "severity:ch1") 
write_csv(GSE189039_metadata, file = here("raw_data", "GSE189039_metadata.csv"))

#Counts matrix
#Retrieve counts table from supplementary files
getGEOSuppFiles('GSE189039', makeDirectory = T, baseDir = here("raw_data"))
untar(here("raw_data", 'GSE189039_RAW.tar'))
files <- list.files(path = here("raw_data", "GSE189039"), pattern = "*.gz$", full.names = TRUE)
for(files_gz in files) {gunzip(files_gz)}
files_txt_GSE189039 <- list.files(path = here("raw_data", "GSE189039"), pattern = "*.txt$", full.names = TRUE)
list_data_GSE189039 <- lapply(files_txt_GSE189039, read_standardize)
counts_GSE189039 <- list_data_GSE189039 %>%
  purrr::discard(is.null) %>%
  reduce(full_join, by = "genes")
colnames(counts_GSE189039)[-1] <- gsub("^[^_]+_(.*)", "\\1", colnames(counts_GSE189039)[-1])
write_csv(counts_GSE189039, file = here("raw_data", "gse189039_counts.csv"))
```

GSE201533

```{r eval=FALSE, message=FALSE, warning=FALSE, echo=TRUE, cache = TRUE}
###### GSE201533 ------

###### Metadata
GSE201533 <- getGEO("GSE201533", GSEMatrix = TRUE, destdir = here("raw_data", "GSE201533")) 

GSE201533 <- GSE201533[[1]]

GSE201533_metadata = GSE201533 %>% 
  pData() %>% 
  select(
    geo_sample = "geo_accession", 
    sample_id = "title", age = "age:ch1", 
    sex = "gender:ch1", 
    dose_vaccine = "vaccine:ch1")

GSE201533_metadata = GSE201533_metadata %>%
  separate(sample_id, into = c("participant_id", "time"), sep = "-D", remove = F) %>%
  separate(dose_vaccine, into = c("dose", "vaccine"), sep = ",\\s*(?=[A-Z])", extra = "merge") %>%
  mutate(dose = str_extract(dose, "\\d+"),
         vaccine = if_else(vaccine == "ChAdOx1", "ChAd", "BNT"),
         condition = paste0(vaccine, " (V", dose, ", D", time, ")"),
         geo = "GSE201533") 

write_csv(GSE201533_metadata, file = here("raw_data", "GSE201533_metadata.csv"))

##### Counts matrix
getGEOSuppFiles('GSE201533', makeDirectory = T, baseDir = here("raw_data"))
untar(tarfile = here("raw_data", "GSE201533", "GSE201533_RAW.tar"))

list_data_gse201533 <- list.files(path = here("raw_data", 
                                              "GSE201533", 
                                              "GSE201533_RAW"), 
                                  pattern = "*.txt.gz$", 
                                  full.names = TRUE) %>%
  map(~ read.table(.x, header = TRUE, sep = "\t"))
counts_gse201533 <- do.call(cbind, map(list_data_gse201533, ~ .x[, 2]))

# Name columns (samples) and genes
colnames(counts_gse201533) <- paste0("GSM", seq_along(list_data_gse201533) + 6066177)
counts_gse201533 <- bind_cols(geneid = list_data_gse201533[[1]][, 1], counts_gse201533)
counts_gse201533[is.na(counts_gse201533)] <- 0 #Rename column names (ex. "X21.01684.hisat2.bam") to GSM codes, which start with GSM6066178.

counts_gse201533 <- list_data_gse201533 %>%
  purrr::map_dfc(~ .x[, 2]) %>%
  set_names(paste0("Sample_", seq_along(list_data_gse201533))) %>%
  bind_cols(geneid = list_data_gse201533[[1]][, 1], .) %>%
  relocate(geneid) %>%
  replace(is.na(.), 0)

saveRDS(counts_gse201533, file = here("raw_data", "gse201533_counts.rds"))
write.csv(counts_gse201533, file=here("raw_data", "gse201533_counts.csv"))
```

GSE201530

```{r eval=FALSE, message=FALSE, warning=FALSE, echo=TRUE, cache = TRUE}
###### GSE201530 ------
#Metadata
filename = "GSE201530"
GSE201530 <- getGEO("GSE201530", GSEMatrix = TRUE) 
GSE201530 <- GSE201530[[1]]
GSE201530_metadata = GSE201530 %>% 
  pData() %>% 
  select(
    sample = "geo_accession", 
    subject_id = "title", 
    age = "age:ch1", 
    sex = "gender:ch1",
    geographical = "geographical location:ch1",
    disease_state = "disease state:ch1", 
    timepoint = "days after positive pcr results:ch1",
    group = "group (by covid-19 vaccination, prior infection):ch1",
    variant = "omicron sublineage:ch1") %>% 
  separate(timepoint, into = c("delete", "day")) %>%
  select(-delete) %>% 
  mutate(variant = if_else(variant == "--", "None", variant),
         day = if_else(group == "Healthy control", "0", day),
         day = as.double(day)) %>% 
  mutate(condition = case_when(group == "Healthy control" ~ "H",
                               group == "Vaccination, No prior infection" ~ "BNT-I",
                               group == "No vaccination, Prior infection" ~ "I-I",
                               group == "No vaccination, No prior infection" ~ "I",
                               group == "Vaccination, Prior infection" ~ "I-BNT",
                               .default = group),
         condition = paste0(condition, " (D", day, ")"))
write.csv(GSE201530_metadata, file = here("raw_data", "GSE201530_metadata.csv"))

# Counts matrix
getGEOSuppFiles('GSE201530', baseDir = here("raw_data"))
untar(here("raw_data", 'GSE201530_RAW.tar'))
files_txt_GSE201530 <- list.files(path = here("raw_data", "GSE201530"), pattern = "*.txt$", full.names = TRUE)
list_data_GSE201530 <- lapply(files_txt_GSE201530, read_standardize)
counts_gse201530 <- list_data_GSE201530 %>%
  purrr::discard(is.null) %>%
  reduce(full_join, by = "genes")
colnames(counts_gse201530)[-1] <- gsub("^[^_]+_(.*)", "\\1", colnames(counts_gse201530)[-1])

saveRDS(counts_gse201530, file = here("raw_data", "gse201530_counts_ready.rds"))

counts_gse201530 = read_rds(file = here("raw_data", "gse201530_counts.rds"))


```

GSE199750

As we couldn't retrieved all GSE199750 counts matrices through R, we
downloaded the counts table from the Github repository:
<https://bitbucket.org/lynnlab/covirs>.

```{r eval=FALSE, message=FALSE, warning=FALSE, echo=TRUE, cache = TRUE}
###### GSE199750 ----
#Metadata
gse199750 <- getGEO("GSE199750", GSEMatrix = TRUE,  destdir = here("raw_data", "GSE199750")) 
gse199750 <- gse199750[[1]]
gse199750_metadata <- pData(gse199750) %>% 
  select(
    GEO = "geo_accession", 
    SampleN = "sample number:ch1", 
    SubjectID = "subject_id:ch1", 
    Age = "age:ch1", 
    Sex = "Sex:ch1", 
    Vaccine = "vaccine:ch1", 
    FirstSecondVaccine = "first/second vaccine:ch1", 
    ThirdVaccine = "third vaccine:ch1", 
    Timepoint = "timepoint:ch1",
    Tissue = "tissue:ch1", 
    Run = "run:ch1"
  )

#There are two similar columns. The first, "Vaccine," might be filled in, while the second, "FirstSecondVaccine," may not be, and vice versa. Therefore, one should equal the other. Thus, unify the columns.

gse199750_metadata <- gse199750_metadata %>%
  mutate(
    FirstSecondDose = coalesce(FirstSecondVaccine, Vaccine),
    Subject_Timepoint = paste(SubjectID, Timepoint, sep = "_")
  ) %>%
  select(
    Subject_Timepoint, GEO, SampleN, SubjectID, Age, Sex, Timepoint, 
    FirstSecondDose, ThirdVaccine, Tissue, Run
  ) %>%
  mutate(Age_ranges = cut(
    Age, 
    breaks = c(0, 18, 30, 45, 60, 70, 80, Inf), 
    labels = c("<18", "18-29", "30-44", "45-59", "60-69", "70-79", "80>"), 
    right = FALSE
  ) %>% 
  column_to_rownames("Subject_Timepoint")

write.csv(gse199750_metadata, file =here("raw_data", "metadata_gse199750.csv"))
saveRDS(gse199750_metadata, file = here("raw_data", "gse199750_metadata.rds"))

# Counts matrix
raw_RNASeq_counts = readRDS(here("raw_data", "raw_RNASeq_counts.RDS"))

# The counts table provided by the study("https://bitbucket.org/lynnlab/covirs/src/master/raw_RNASeq_counts.RDS") is named with Subject ID + Timepoint and not with the GSM code. Therefore, we will need to process the metadata table. Additionally, it lacks the age column. These data can be manipulated in Excel.
############# Gene annotation -----
library("biomaRt")

mart <- useMart("ensembl")
mart <- useDataset("hsapiens_gene_ensembl", mart)

gene_info <- getBM(attributes = c("entrezgene_id", "hgnc_symbol", "ensembl_gene_id"), 
                   mart = mart) %>% 
  select(ensembl_gene_id, hgnc_symbol) %>% 
  distinct()

raw_RNASeq_counts_ann <- raw_RNASeq_counts %>% 
  rownames_to_column("ensembl_gene_id") %>% 
  left_join(gene_info, by = "ensembl_gene_id") %>% 
  mutate(hgnc_symbol = if_else(hgnc_symbol == "", ensembl_gene_id, hgnc_symbol)) %>% 
  filter(hgnc_symbol != "") %>%
  distinct() %>%
  select(sort(names(.))) %>% 
  select(genes = hgnc_symbol, ensembl_gene_id, everything()) %>% 
  arrange(genes)

#Save
write_csv(raw_RNASeq_counts_ann, file= here("raw_data", "raw_RNASeq_counts_annGenes.csv"))

#Filter duplicated genes. Get only the older ensembl ids
not_interes = raw_RNASeq_counts_ann %>% 
  get_dupes(genes) %>% 
  group_by(genes) %>% 
  arrange(ensembl_gene_id) %>% 
  slice(2) %>% 
  ungroup() %>% 
  select(ensembl_gene_id)

gse199750_counts_ready = raw_RNASeq_counts_ann %>% 
  anti_join(not_interes, by = "ensembl_gene_id") %>% 
  select(-ensembl_gene_id) %>% 
  column_to_rownames("genes")

# Save
saveRDS(gse199750_counts_ready, file= here("raw_data", "gse199750_counts_ready.rds"))

```

GSE206023

```{r eval=FALSE, message=FALSE, warning=FALSE, echo=TRUE, cache = TRUE}
######### GSE206023 ------- 
gse206023 <- getGEO("GSE206023", 
                    GSEMatrix = TRUE,  
                    destdir = here("raw_data")) 

gse206023 <- gse206023[[1]]
gse206023_metadata <- pData(gse206023)
colnames(gse206023_metadata)

gse206023_metadata <- select(gse206023_metadata, "title", "geo_accession", "source_name_ch1", "organism_ch1", "time:ch1", "treatment:ch1"  )

colnames(gse206023_metadata) <-
  c("Sample name",
    "GEO",
    "Sample type",
    "Organism",
    "Timepoints",
    "Vaccine")

write.csv(gse206023_metadata, file= here("raw_data","gse206023_metadata.csv"))
saveRDS(gse206023_metadata, file=here("raw_data","gse206023_metadata.rds"))

#### Import matrix.
#Genes were named incorrectly and represent only 2.000 genes of >30.000. Hence, we externally performed a pseudoalignment to retrieve all genes that were not correctly annotated. 

#Pseudoalignment
gse206023_counts = read_excel(here("raw_data", "GSE206023_counts_Pseudoalignment.xlsx")) 


gse206023_counts_ready %>% 
  rownames_to_column("genes") %>% 
  filter(genes == "XIST")

gse206023_counts_ready = gse206023_counts %>%
  select(!c(1, 2)) %>% 
  drop_na(external_gene_name) %>% 
  column_to_rownames("external_gene_name") %>% 
  round(., 0)

gse206023_counts_ready %>% 
  saveRDS(file = here("raw_data", "gse206023_counts_ready.rds"))
gse206023_counts_ready %>% 
  write_csv(file = here("raw_data", "gse206023_counts_ready.csv"))

#General statistics
GSE206023_pseudoalign_stats <- read_excel("raw_data/GSE206023_pseudoalignment_ST1.xlsx", 
    sheet = "General Statistics")
```

## Combine raw counts

```{r cache = TRUE}
# Counts
####### Wide table
gse189039 = gse189039_counts_ready %>% 
  rownames_to_column("genes") 
gse201530 = gse201530_counts 
gse201533 = gse201533_counts_ready %>% 
  rownames_to_column("genes") 
gse199750 = gse199750_counts_ready %>% 
  as.data.frame() %>% 
  rownames_to_column("genes") 
gse206023 = gse206023_counts_ready %>% 
  rownames_to_column("genes")

all_matrices_counts = gse189039 %>% 
  full_join(gse201530, by = "genes") %>% 
  full_join(gse201533, by = "genes") %>% 
  full_join(gse199750, by = "genes") %>% 
  full_join(gse206023, by = "genes")


all_matrices_counts
saveRDS(all_matrices_counts, file = here("DGE analysis", "all_matrices_counts_raw.rds"))

#Get genes present in all matrices
genes_not_shared = all_matrices_counts %>% 
  column_to_rownames("genes") %>% 
  mutate(across(everything(), ~ if_else(is.na(.), "NA", as.character(.)))) %>% 
  rownames_to_column("genes") %>% 
  pivot_longer(cols = -genes, 
               names_to = "sample",
               values_to = "counts") %>% 
  filter(counts == "NA") %>% 
  select(genes) %>% 
  distinct()


all_matrices_counts_filtered = all_matrices_counts %>% 
  filter(!genes %in% c(genes_not_shared %>% pull(genes)))

saveRDS(all_matrices_counts_filtered, file = here("DGE analysis", "all_matrices_counts.rds"))

```



## Counts normalization
```{r cache = TRUE}
all_matrices_counts = read_rds(file = here("DGE analysis", "all_matrices_counts.rds"))
all_matrices_counts_long = all_matrices_counts %>% 
  pivot_longer(cols = -genes,
               names_to = "sample",
               values_to = "counts")

all_matrices_counts_long_annotated = all_matrices_counts_long %>% 
  inner_join(ann_vaccines_samples, by = "sample")

all_matrices_counts_long_annotated %>% saveRDS(here("DGE analysis", "all_matrices_counts_long_annotated.rds"))
```

#### Transcripts per million (TPM)

```{r}

library("biomaRt")

mart <- useMart("ensembl")
mart <- useDataset("hsapiens_gene_ensembl", mart)

gene_info <- getBM(attributes = c("entrezgene_id", "hgnc_symbol", "ensembl_gene_id", "gene_biotype", "start_position", "end_position"), 
                   mart = mart) 

gene_info = gene_info %>% 
  mutate(gene_length_kb = (end_position - (start_position + 1)) / 1000)


# Exemplo de contagens brutas (genes x amostras)
counts = all_matrices_counts %>% 
   inner_join(gene_info %>% select(genes = hgnc_symbol) %>% distinct(), 
              by = "genes") %>% 
  arrange(genes) %>% 
                    column_to_rownames("genes") %>% 
                    as.matrix() %>% 
                    replace(is.na(.), 0)

dim(counts)

gene_lengths = gene_info %>% 
  inner_join(counts %>% 
            as.data.frame() %>% 
            rownames_to_column("hgnc_symbol") %>% 
            select(hgnc_symbol), 
        by = "hgnc_symbol") %>% 
  select(hgnc_symbol, gene_length_kb) %>% 
  distinct() %>% 
  arrange(hgnc_symbol) %>% 
  select(gene_length_kb) %>% 
  pull()

# 1. Dividir as contagens pelo comprimento do gene em kb
rpk <- counts / gene_lengths

# 2. Calcular o fator de escala para cada amostra
scaling_factors <- colSums(rpk)

# 3. Calcular TPM dividindo por este fator e multiplicando por 1 milhão
tpm <- t(t(rpk) / scaling_factors)

tpm = tpm %>% 
  as.data.frame()

tpm %>% 
  saveRDS(here("DGE analysis", "all_matrices_counts_tpm_normalized.rds"))


all_matrices_counts_tpm_normalized  = readRDS(here("DGE analysis", "all_matrices_counts_tpm_normalized.rds"))
all_matrices_counts_tpm_normalized_long = all_matrices_counts_tpm_normalized %>% 
  rownames_to_column("genes") %>% 
  pivot_longer(cols = -genes,
               names_to = "sample",
               values_to = "counts")

all_matrices_counts_tpm_normalized_long_annotated = all_matrices_counts_tpm_normalized_long %>% 
  inner_join(ann_vaccines_samples, by = "sample")

all_matrices_counts_tpm_normalized_long_annotated %>% saveRDS(here("DGE analysis", "all_matrices_counts_tpm_normalized_long_annotated.rds"))

```

#### Counts per million
```{r}
library(edgeR)

# Get CPM 
cpm_counts <- cpm(all_matrices_counts %>% 
                    column_to_rownames("genes") %>% 
                    as.matrix() %>% 
                    replace(is.na(.), 0))

print(cpm_counts)

cpm_counts %>% 
  saveRDS(here("DGE analysis", "all_matrices_cpm_normalized.rds"))

cpm_counts = readRDS(here("DGE analysis", "all_matrices_cpm_normalized.rds"))
```

#### Log2fc by sample
```{r cache = TRUE}
# Log2FC by sample ------
# all_matrices_counts_tpm_normalized_long_annotated = readRDS(here("DGE analysis", "all_matrices_counts_tpm_normalized_long_annotated.rds"))


##### Vaccination
# GSE206023
counts_control_gse206023 = all_matrices_counts_tpm_normalized_long_annotated %>% 
  filter(gse_id == "GSE206023") %>% 
  filter(!is.na(counts))

counts_gse206023_samples_l2fc = counts_control_gse206023 %>% 
  filter(day == 0) %>% 
  select(genes, participant_id, baseline = counts) %>% 
  inner_join(counts_control_gse206023, by = join_by("genes", "participant_id")) %>% 
  group_by(participant_id) %>% 
  mutate(diff = counts - baseline,
       log2fc = if_else(diff == 0, 0, sign(diff) * log2(abs(diff))),
       log2fc = replace(log2fc, log2fc == -Inf, 0)) %>% 
  select(genes, condition, participant_id, sample, baseline, counts,diff, log2fc) %>% 
  ungroup()

# GSE201533
counts_control_gse201533 = all_matrices_counts_tpm_normalized_long_annotated %>% 
  filter(gse_id == "GSE201533") %>% 
  filter(!is.na(counts))

counts_gse201533_samples_l2fc = counts_control_gse201533 %>% 
  filter(day == 0) %>% 
  select(genes, participant_id, baseline = counts) %>% 
  inner_join(counts_control_gse201533, by = join_by("genes", "participant_id")) %>% 
  group_by(participant_id) %>% 
  mutate(diff = counts - baseline,
       log2fc = if_else(diff == 0, 0, sign(diff) * log2(abs(diff))),
       log2fc = replace(log2fc, log2fc == -Inf, 0)) %>% 
  select(genes, condition, participant_id, sample, baseline, counts,diff, log2fc) %>% 
  ungroup()
  

# GSE199750
counts_control_gse199750 = all_matrices_counts_tpm_normalized_long_annotated %>% 
  filter(gse_id == "GSE199750") %>% 
  filter(!is.na(counts))

counts_gse199750_samples_l2fc = counts_control_gse199750 %>% 
  filter(day == 0) %>% 
  select(genes, participant_id, baseline = counts) %>% 
  inner_join(counts_control_gse199750, by = join_by("genes", "participant_id")) %>% 
  group_by(participant_id) %>% 
  mutate(diff = counts - baseline,
       log2fc = if_else(diff == 0, 0, sign(diff) * log2(abs(diff))),
       log2fc = replace(log2fc, log2fc == -Inf, 0)) %>% 
  select(genes, condition, participant_id, sample, baseline, counts,diff, log2fc) %>% 
  ungroup()

####### Infection

# GSE189039
counts_control_gse189039 = all_matrices_counts_tpm_normalized_long_annotated %>% 
  filter(gse_id == "GSE189039") %>% 
  filter(!is.na(counts))

counts_gse189039_samples_l2fc = counts_control_gse189039 %>% 
  filter(day == 0) %>% 
  select(genes, counts) %>% 
  group_by(genes) %>% 
  summarize(baseline = median(counts)) %>% 
  inner_join(counts_control_gse189039, by = "genes") %>% 
  mutate(diff = counts - baseline,
       log2fc = if_else(diff == 0, 0, sign(diff) * log2(abs(diff))),
       log2fc = replace(log2fc, log2fc == -Inf, 0)) %>% 
  select(genes, condition, sample, baseline, counts,diff, log2fc)

# GSE201530
counts_control_gse201530 = all_matrices_counts_tpm_normalized_long_annotated %>% 
  filter(gse_id == "GSE201530") %>% 
  filter(!is.na(counts))

counts_gse201530_samples_l2fc = counts_control_gse201530 %>% 
  filter(day == 0) %>% 
  select(genes, counts) %>% 
  group_by(genes) %>% 
  summarize(baseline = median(counts)) %>% 
  inner_join(counts_control_gse201530, by = "genes") %>% 
  mutate(diff = counts - baseline,
       log2fc = if_else(diff == 0, 0, sign(diff) * log2(abs(diff))),
       log2fc = replace(log2fc, log2fc == -Inf, 0)) %>% 
  select(genes, condition, sample, baseline, counts,diff, log2fc)


all_samples_l2fc = bind_rows(counts_gse206023_samples_l2fc,
                         counts_gse199750_samples_l2fc,
                         counts_gse201533_samples_l2fc,
                         counts_gse201530_samples_l2fc,
                         counts_gse189039_samples_l2fc)

all_samples_l2fc

all_samples_l2fc %>% 
  saveRDS(file = here("DGE analysis", "all_samples_l2fc.rds"))
```

#### DESeq2 normalization
```{r cache = TRUE}
##### DESEQ2 Normalization
library(DESeq2)

ann_vaccines_samples = read_csv(here("Annotations and metadata", "ann_vaccines_samples.csv"))
all_matrices_counts = readRDS(here("DGE analysis", "all_matrices_counts.rds"))

cvrts = ann_vaccines_samples %>% 
  arrange(condition) %>%
  distinct() %>% 
  column_to_rownames("sample")

exprs_deseq = all_matrices_counts %>% 
  select(c("genes", cvrts %>% rownames_to_column("sample") %>% .$sample)) %>% 
  replace(is.na(.), 0) %>% 
  mutate(across(where(is.numeric), round, 0)) %>% 
  column_to_rownames("genes") 

cvrts = cvrts %>% 
  mutate(condition = as.factor(condition),
         age = as.factor(age),
         sex = as.factor(sex), 
         vaccine = as.factor(vaccine),
         day = as.factor(day),
         gse_id = as.factor(gse_id))
  
#Check dimensions
dim(exprs_deseq)
dim(cvrts)

#Normalize data
dds <- DESeqDataSetFromMatrix(countData = exprs_deseq, 
                              colData = cvrts %>% 
                                select(condition, gse_id) %>% 
                                rownames_to_column("sample"), 
                              design = ~ condition)
dds <- estimateSizeFactors(dds)


normalized_counts <- counts_batch_corrected %>% 
  counts(normalized=TRUE) %>% 
  as.data.frame()

#Remove batch
dds <- DESeqDataSetFromMatrix(countData = exprs_deseq, 
                              colData = cvrts %>% 
                                select(condition, gse_id) %>% 
                                rownames_to_column("sample"), 
                              design = ~ age + sex + condition)
dds <- DESeq(dds)
vsd <- vst(dds, blind=FALSE)
mat <- assay(vsd)
mat <- limma::removeBatchEffect(mat, vsd$gse_id)
assay(vsd) <- mat
counts_batch_corrected <- assay(vsd)


saveRDS(normalized_counts, file = here("DGE analysis", "all_matrices_normalized_counts.rds"))

normalized_counts = read_rds(here("DGE analysis", "all_matrices_normalized_counts.rds"))

```

### PVCA of normalized raw counts

```{r cache = TRUE}
#### PVCA of normalized raw counts -----
# BiocManager::install("ExpressionNormalizationWorkflow")
library(ExpressionNormalizationWorkflow)

ImmuneGO_Annotated_Genes <- readRDS("~/Desktop/Github - Covid Atlas/VaxGO gene sets/ImmuneGO_Annotated_Genes_2024-03-26.rds")
#Expression data
exprs_deseq_immune = exprs_deseq %>% 
  rownames_to_column("genes") %>% 
  filter(genes %in% ImmuneGO_Annotated_Genes$genes) %>% 
  column_to_rownames("genes")

```

Covariants table

```{r cache = TRUE}
cvrts = cvrts %>% 
  rownames_to_column("sample") %>% 
  inner_join(exprs %>% 
               colnames() %>% 
               as.data.frame() %>% 
               select(sample = "."), 
             by = "sample") %>% 
  arrange(sample) %>% 
  column_to_rownames("sample")
cvrts
```

Inputs
```{r cache = TRUE}
# Input
#For vaccination only (no ongoing infection)
cvrts_eff_var =  c("day", 
                   "age", 
                   "sex", 
                   "dose", 
                   "type", 
                   "regimen") 

#For vaccination and infection
cvrts_eff_var = c(
  "day",
  "age",
  "sex",
  "dose",
  "type",
  "previous_vaccination",
  "variant",
  "prior_infection",
  "severity",
  "disease_vac",
  "regimen"
) 


#For vaccination and infection
cvrts_eff_var = c(
  "day",
  "age",
  "sex",
  "dose",
  "type",
  "previous_vaccination",
  "variant",
  # "country_enroll",
  # "time_enroll",
  "prior_infection",
  "severity",
  "disease_vac",
  "regimen"
) 

#Convert to Expression Dataset Object
inpData <- expSetobj(exprs_deseq_immune, cvrts) 

#Threshold
pct_thrsh <- 0.75 
```

Run PVCA analysis

```{r cache = TRUE}
#PVCA analysis
pvca = pvcAnaly(inpData, pct_thrsh, cvrts_eff_var) 

pvca_df = pvca$dat %>% 
  as.data.frame() %>% 
  t() %>%
  as.data.frame() %>% 
  rename(wavgpv = V1) %>% 
  cbind(pvca$label %>% as.data.frame() %>% rename(batch = '.')) %>% 
  as.data.frame() %>% 
  mutate(batch = as.factor(batch),
         wavgpv = as.numeric(wavgpv),
         batch = fct_rev(fct_reorder(batch, wavgpv)),
         batch = fct_relevel(batch, after = Inf)) %>% 
  arrange(batch)

pvca_df %>% saveRDS(here("DGE analysis", "pvca_results_immunegenes.rds"))

pvca_df
```

Barplot: Covariants contributions

```{r cache = TRUE, fig.width= 12, fig.height=10}
# Visualize
pvca_df_plot = ggplot(pvca_df) +
  aes(x = batch, y = wavgpv) +
  geom_col(fill = "#4DBBD5FF") +
  theme_minimal() +
  theme(axis.text.x = element_text(size = 20, angle = 45, vjust = 1, hjust=1, color = "black"),
        axis.text.y = element_text(size = 20, vjust = 1, hjust=1, color = "black")) +
  geom_text(aes(label = round(wavgpv,3), y = wavgpv), 
            position = position_dodge(width = 0.9), 
            vjust = -0.2, 
            hjust = 0.5, 
            size = 4,
            angle = 0) 

ggsave(pvca_df_plot, file = here("Figures", "pvca_raw_normalized_immunegenes_allconditions.png"), width = 12, height = 10)

pvca_df_plot
```

Scatterplot of PVCA

```{r cache = TRUE}
# Visualize
library(ggfortify)
library(plotly)

pca_res <- prcomp(exprs_deseq_immune %>% t(), scale. = F)

pca_res_ann = exprs_deseq_immune %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column("sample") %>% 
  inner_join(cvrts %>% 
               rownames_to_column("sample"),
             by = "sample") %>% 
  distinct()

dim(pca_res)
dim(pca_res_ann)

#Study
normalizedcounts_pca_plot = autoplot(pca_res, data = pca_res_ann) +
  geom_point(aes(fill = gse_id, label = condition), colour="black",pch=21, size = 3) +
  theme_few() +
  scale_fill_aaas()

ggsave(normalizedcounts_pca_plot, file = here("Figures", "covid_vax_pca_samples_study_immune.png"), height = 5, width = 8) 

#Type
normalizedcounts_pca_plot_type = autoplot(pca_res, data = pca_res_ann) +
  geom_point(aes(fill = type, label = condition), colour="black",pch=21, size = 3) +
  theme_few() +
  scale_fill_manual(values = ann_vaxsig_colors$type)

ggsave(normalizedcounts_pca_plot_type, file = here("Figures", "covid_vax_pca_samples_type_immune.png"), height = 5, width = 8) 


#Day
normalizedcounts_pca_plot_day = autoplot(pca_res, data = pca_res_ann %>% 
                                           mutate(day = as.factor(day))) +
  geom_point(aes(fill = day, label = condition), colour="black",pch=21, size = 3) +
  theme_few() +
  scale_fill_manual(values = ann_vaxsig_colors$day)

ggsave(normalizedcounts_pca_plot_day, file = here("Figures", "covid_vax_pca_samples_day_immune.png"), height = 5, width = 8)


#Sex
normalizedcounts_pca_plot_sex = autoplot(pca_res, data = pca_res_ann %>% 
                                           mutate(day = as.factor(day))) +
  geom_point(aes(fill = sex, label = condition), colour="black",pch=21, size = 3) +
  theme_few() +
  scale_fill_manual(values = ann_vaxsig_colors$sex)

ggsave(normalizedcounts_pca_plot_sex, file = here("Figures", "covid_vax_pca_samples_sex_immune.png"), height = 5, width = 8) 

```

# 4. General data analysis


## Metadata

### Annotate metadata

Standardize conditions

```{r eval=FALSE, echo = TRUE, cache = TRUE}
#Metadata ------
gse189039_metadata_2 = GSE189039_metadata_1_ %>% 
  separate(sample, into = c("participant_id"), sep = "_", remove = F) %>% 
  mutate(geo = "GSE189039") %>% 
  rename(day = timepoint_day,
         gsm_id = geo_sample) %>% 
  select(sample, participant_id)

gse199750_metadata_2 = gse199750_metadata_annotated_14_11_23_2_ %>% 
  rename(gsm_id = geo,
         participant_id = subject_id) %>% 
  mutate(geo = "GSE199750") %>% 
  select(sample, participant_id)

gse206023_metadata_3 = gse206023_metadata %>% 
  mutate(day = as.numeric(day),
         geo = "GSE206023") %>%  
  inner_join(gse206023_metadata_ready_2, by = "sample_id") %>% 
  inner_join(ann_vaccines_samples %>% 
            select(condition, sample_id = sample, age, sex, vaccine),
            by = "sample_id") %>% 
  rename(gsm_id = gsm_id.x) %>% 
  select(sample = sample_id, participant_id)

gse201530_metadata_2 = gse201530_metadata %>% 
  rename(gsm_id = geo_sample) %>% 
  separate(subject_id, into = c("variant", "participant_id_num"), sep = "_", remove = F) %>% 
  mutate(participant_id = paste0(variant, "_", participant_id_num)) %>% 
  select(participant_id, sample = subject_id)

gse201533_metadata_2 = GSE201533_metadata %>% 
  mutate(age = as.numeric(age),
         dose = as.numeric(dose),
         gsm_id = geo_sample) %>% 
  select(geo_sample = sample_id, participant_id) %>% 
  rownames_to_column("sample")

```

Unite samples

```{r eval=FALSE, include=TRUE, cache = TRUE}
ann_vaccines_samples_2.1 = ann_vaccines_samples_2 %>% 
  filter(is.na(participant_id)) %>% 
  separate(sample, into = c("participant_id_1", "participant_id_2", "participant_id_3", "delete"), remove = F) %>% 
  mutate(participant_id = paste0(participant_id_1, "_", participant_id_2, "_", participant_id_3)) %>% 
  select(!c(participant_id_1, participant_id_2, participant_id_3)) %>% 
  select(sample, participant_id)

all_participants = bind_rows(gse189039_metadata_2,
          gse199750_metadata_2,
          gse206023_metadata_3,
          gse201533_metadata_2,
          gse201530_metadata_2) %>% 
  bind_rows(ann_vaccines_samples_2.1)

ann_vaccines_samples_2 = ann_vaccines_samples %>% 
  left_join(all_participants, by = "sample") %>% 
  select(condition, sample, age, sex, vaccine:previous_vaccination, participant_id) %>% 
  left_join(ann_vaccines_samples_2.1, by = "sample")


ann_vaccines_samples_participants = ann_vaccines_samples_2 %>% 
  select(-participant_id.y, participant_id = participant_id.x)
  
write.csv(ann_vaccines_samples_participants, file = "ann_vaccines_samples.csv")
  
```



#### Sex imputation
XIST (female-specific) and Y-linked genes IF1AY, KDM5D, UTY, DDX3Y, and RPS4Y1 are indicative of sex
```{r fig.height=5, fig.width=10}
# all_matrices_counts_raw <- readRDS(here("DGE analysis","all_matrices_counts_raw.rds"))
# all_matrices_counts = all_matrices_counts_raw
all_matrices_counts_tpm_normalized_unfiltered <- readRDS(here("DGE analysis","all_matrices_counts_tpm_normalized_unfiltered.rds"))


ann_vaccines_samples_gse206023 = ann_vaccines_samples %>% 
  filter(gse_id == "GSE206023") 

all_matrices_gse206023 = all_matrices_counts_tpm_normalized_unfiltered %>% 
  rownames_to_column("genes") %>%
  select(genes, c(ann_vaccines_samples_gse206023$sample))

seximpute_gse206023_predicted = all_matrices_gse206023 %>% 
  filter(genes %in% c("XIST", "IF1AY", "KDM5D", "UTY", "DDX3Y","RPS4Y1"))  %>% 
  column_to_rownames("genes") %>%
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column("sample") %>% 
  mutate(sex = if_else(XIST != 0 & UTY == 0, "Female", "Male")) %>% 
  mutate(chr_y = if_else(rowSums(select(., DDX3Y, KDM5D, RPS4Y1, UTY)) != 0, "Yes", "No"),
         XIST = if_else(XIST == 0.000000e+00, 0.0, XIST),
         ratio_mf = sum(DDX3Y,KDM5D,RPS4Y1, UTY)/ XIST,
         ratio_mf = if_else(ratio_mf == Inf, 1, ratio_mf)) %>% 
  separate(sample, into = c("participant_id", "day")) %>% 
  pivot_longer(cols = c("XIST", "KDM5D", "UTY", "DDX3Y","RPS4Y1"),
               names_to = "genes",
               values_to = "value") %>% 
  mutate(value_log = if_else(value == 0.000000e+00, 0, log(value))) %>% 
  mutate(day = str_replace(day, "d", "") %>% as.integer()) %>% 
  group_by(genes) %>% 
  mutate(minmax_normalized = (value_log - min(value_log)) / (max(value_log) - min(value_log))) %>% 
    mutate(minmax_normalized = if_else(value_log == 0, 0, minmax_normalized))
  
seximpute_gse206023_plot_2 = seximpute_gse206023_predicted %>% 
   filter(value_log != 0) %>%
   ggplot() +
   aes(
    x = day,
    y = minmax_normalized,
    colour = chr_y,
    label = participant_id,
    group = participant_id
  ) +
  geom_line() +
  geom_point(color = "black", stroke = 1) +
  geom_point(aes(color = chr_y)) +
  scale_color_manual(values = c("Yes" = "#4DBBD5FF",
          "No" = "#E64B35FF")) +
   facet_wrap(~genes, nrow = 1) +
  theme_minimal()+
  theme(axis.text = element_text(color = "black", size = 10),
        legend.text = element_text(color = "black", size = 10),
        axis.line.x  = element_line(size = 1, color ="black"),
        axis.line.y  = element_line(size = 0, color ="black"),
        axis.text.x = element_text(angle = 0, vjust = 1, hjust=0.5),
        axis.ticks = element_line(size = 1, color ="black"),
        panel.grid.major = element_line(linetype = 2, color = "gray75"),
        panel.grid.minor = element_line(linetype = 2, size = 0, color = "gray75"),
        # panel.grid.x = element_line(linetype = 2, color = "gray75"),
        # panel.grid.y = element_line(linetype = 2, color = "gray75"),
        strip.text = element_text(size = 10, color = "black"),
        legend.direction = "horizontal",
        legend.position = "top") +
  scale_y_continuous(expand = c(0,0), limits = c(0, 0.80)) +
  scale_x_continuous(breaks = c(0, 7, 14, 28)) +
  labs(x = "Days",
       # color = "Chromosome Y specific genes present",
       color = "Chromosome Y \nspecific genes present",
       y = "TPM normalized\ncounts",
       title = "")

seximpute_gse206023_plot = seximpute_gse206023_predicted %>%
  # filter(day == 0) %>% 
  mutate(title = paste("Day ", day),
         title = fct_reorder(title, day)) %>% 
ggplot(.) +
  aes(x = genes, y = fct_reorder(participant_id, minmax_normalized), fill = minmax_normalized) +
  geom_tile(color = "white", size = 0.5) +
  theme_minimal() +
  theme(axis.text.y = element_text(color = "black", size = 10),
        axis.line.x  = element_line(size = 1, color ="black"),
        axis.line.y  = element_line(size = 0, color ="black"),
        # axis.line.y  = element_line(size = 1, color ="black"),
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1, size = 10),
        axis.ticks = element_line(size = 1, color ="black"),
        # panel.grid = element_line(linetype = 2, color = "gray75"),
        panel.grid = element_blank(),
        strip.text = element_text(size = 10, color = "black"),
        legend.direction = "horizontal",
        legend.position = "top") +
  scale_x_discrete(expand = c(0,0)) +
  labs(x = "",
       y = "",
       title = "") +
  scale_fill_gradient2(low = "white",
                        name = "TPM normalized\ncounts",
                      high = "#4DBBD5FF") +
  facet_wrap(~title, nrow = 1)


patch = seximpute_gse206023_plot_2 + seximpute_gse206023_plot + 
  plot_layout(widths = c(1, 1),
              guides = "collect"
              ) & theme(legend.position = 'top')
patch
patch %>% 
  ggsave(filename = here("Figures", "seximpute_gse206023_Joined.png"),
         width = 10,
         height = 5)

seximpute_gse206023_plot %>% 
  ggsave(filename = here("Figures", "seximpute_gse206023_plot.png"),
         width = 4,
         height = 4)

seximpute_gse206023_plot_2 %>%
  ggsave(filename = here("Figures", "seximpute_gse206023_plot_lines.png"),
         width = 10,
         height = 5)

seximpute_gse206023_plot_2 %>% ggplotly()

```


```{r fig.height=5, fig.width=5}
ann_vaccines_samples_predicted = ann_vaccines_samples %>%
  left_join(
    seximpute_gse206023_predicted %>% select(participant_id, sex), 
    by = c("participant_id" = "participant_id"), 
    suffix = c("", "_predicted")
  ) %>%
  mutate(
    sex = if_else(gse_id == "GSE206023", sex_predicted, sex)
  ) %>%
  select(-sex_predicted)



ann_vaccines_samples_predicted

ann_vaccines_samples = ann_vaccines_samples_predicted

# ann_vaccines_samples_predicted %>% 
#   saveRDS(file = here("Annotations and metadata", "ann_vaccines_samples_predicted_sex.rds"))

# ann_vaccines_samples %>% 
#   saveRDS(file = here("Annotations and metadata", "ann_vaccines_samples.rds"))

```



#### Age imputation
```{r}

ann_vaccines_samples_gse206023 = ann_vaccines_samples %>% 
  filter(gse_id == "GSE206023") 

all_matrices_tpm_gse206023 = all_matrices_counts_tpm_normalized %>% 
  select(c(ann_vaccines_samples_gse206023$sample))


# BiocManager::install("RNAAgeCalc")
library(RNAAgeCalc)

data(fpkmExample)
head(fpkm)
chronage = data.frame(sampleid = colnames(fpkm), age = c(30,50))
res = predict_age(exprdata = fpkm, tissue = "brain", exprtype = "FPKM", 
                  chronage = chronage)



ann_vaccines_samples %>% 
  arrange(age, gse_id)
```


#### Conditions annotations

```{r eval=FALSE, include=TRUE, cache = TRUE}
library(tidyverse)

ann_vaccines_conditions

demographics = ann_vaccines_samples %>%
  select(vaccine, gse_id, condition_grouped, sample) %>% 
  distinct() %>% 
  group_by(vaccine, gse_id, condition_grouped) %>% 
  summarise(samples = n()) %>% 
  ungroup()

demographics %>% 
  summarize(total = sum(samples))

ann_vaccines_conditions_1 = demographics %>% 
  select(condition = condition_grouped, samples) %>% 
  inner_join(ann_vaccines_conditions %>% 
               select(-samples), 
             by = join_by("condition")) %>% 
  distinct() %>% 
  mutate(week = ceiling(day / 7),
         week = as.factor(week))

write_csv(ann_vaccines_conditions_1, file = here("Annotations and metadata", "ann_vaccines_conditions_23-10-24.csv"))


ann_vaccines_conditions = ann_vaccines_conditions_1
```

### Reorder conditions as factors
```{r}

#Group conditions
ann_vaccines_conditions_grouped = ann_vaccines_conditions %>% 
  mutate(condition_grouped = condition) %>% 
  select(condition, condition_grouped, everything()) %>% 
  mutate(condition_grouped = case_when(condition_grouped %in% c("I (D0)", "I (D1)","I (D2)","I (D3)","I (D4)") ~ "I (D0-4)",
                                       condition_grouped %in% c("I-I (D0)", "I-I (D1)","I-I (D2)","I-I (D3)","I-I (D5)") ~ "I-I (D0-5)",
                                       condition_grouped %in% c("BNT-I (D0)", "BNT-I (D1)","BNT-I (D2)","BNT-I (D3)") ~ "BNT-I (D0-5)",
                                       condition_grouped %in% c("I (D10, moderate)", "I (D10, severe)") ~ "I (D10, hospital)",
                                       condition_grouped %in% c("I (D26, moderate)", "I (D26, severe)") ~ "I (D26, hospital)",
                                       condition_grouped %in% c("I (D51, moderate)", "I (D51, severe)") ~ "I (D51, hospital)",
                                       condition_grouped %in% c("BNT-I (D10, mild)", "BNT-I (D10, severe)") ~ "BNT-I (D10, hospital)",
                                       condition_grouped %in% c("BNT-I (D26, mild)", "BNT-I (D26, severe)") ~ "BNT-I (D26, hospital)",
                                       condition_grouped %in% c("BNT-I (D51, severe)") ~ "BNT-I (D51, hospital)",
                                       condition_grouped %in% c("BNT-I-BNT (D51, mild)", "BNT-I-BNT (D51, severe)") ~ "BNT-I-BNT (D51, hospital)",
                                       condition_grouped == "BNT-MO (V1, D0)" ~ "BNT (V1, D0)",
                                       condition_grouped == "BNT-MO (V1, D6)" ~ "BNT (V1, D6)",
                                       TRUE ~ condition_grouped)) %>% 
  arrange(condition_grouped)


ann_vaccines_conditions_grouped %>% 
  write_csv(file = here("Annotations and metadata", "ann_vaccines_conditions_grouped.csv"))
  
ann_vaccines_conditions = ann_vaccines_conditions_grouped

ann_vaccines_conditions %>%
  write_csv(file = here("Annotations and metadata", "ann_vaccines_conditions.csv"))
  

#Reorder factors
ann_vaccines_conditions_ordered = ann_vaccines_conditions %>% 
  ungroup() %>% 
  mutate(condition = as.character(condition)) %>% 
  mutate(disease_vac = if_else(disease_vac == "Healthy", "H", disease_vac)) %>% 
  mutate(disease_vac = fct_relevel(disease_vac, "H", "I", "V"),
           type = fct_relevel(type, "I", "RNA", "VV", "IN", "SU")) %>% 
  arrange(disease_vac,type, condition)


#Reorder conditions
# ordered_conditions = c("H (D0)",
#                                  ann_vaccines_conditions_ordered %>%
#                                    filter(grepl("^I \\(D[0-9]\\)$", condition)) %>%
#                                    select(condition) %>%
#                                    arrange(condition) %>%
#                                    distinct() %>%
#                                    pull(condition),
#                                  ann_vaccines_conditions_ordered %>%
#                                    filter(str_detect(condition, "^I-I")) %>%
#                                    select(condition) %>%
#                                    arrange(condition) %>%
#                                    distinct() %>%
#                                    pull(condition),
#                                  ann_vaccines_conditions_ordered %>%
#                                     filter(str_detect(condition, "^BNT-I \\(D\\d+\\)")) %>%
#                                     arrange(day, condition) %>%
#                                    select(condition) %>%
#                                     distinct() %>%
#                                     pull(condition),
#                                  "I-BNT (D5)",
#                                  ann_vaccines_conditions_ordered %>%
#                                    filter(!grepl("^I \\(D[0-9]\\)$", condition)) %>%
#                                    filter(str_detect(condition, "^I")) %>%
#                                    filter(!str_detect(condition, "^I-I"),
#                                           condition != "I-BNT (D5)") %>%
#                                    select(condition) %>%
#                                    arrange(condition) %>%
#                                    distinct() %>%
#                                    pull(condition),
#                                  ann_vaccines_conditions_ordered %>%
#                                    filter(disease_vac == "I") %>%
#                                    filter(str_detect(condition, "^BNT-I")) %>%
#                                     filter(!str_detect(condition, "^BNT-I \\(D\\d+\\)")) %>%
#                                     arrange(day, condition) %>%
#                                    select(condition) %>%
#                                     distinct() %>%
#                                     pull(condition),
#                                  ann_vaccines_conditions_ordered %>%
#                                    filter(disease_vac == "V",
#                                           str_detect(condition, "^BNT \\(V1, D\\d+\\)")) %>%
#                                    arrange(day, condition) %>%
#                                    select(condition) %>%
#                                    distinct() %>%
#                                    pull(condition),
#                                  ann_vaccines_conditions_ordered %>%
#                                    filter(disease_vac == "V",
#                                           !str_detect(condition, "^BNT \\(V1, D\\d+\\)"),
#                                           !str_detect(condition, "BBIBP"),
#                                           !str_detect(condition, "ZF2001")) %>%
#                                    arrange(condition) %>%
#                                    select(condition) %>%
#                                    distinct() %>%
#                                    pull(condition),
#                                  ann_vaccines_conditions_ordered %>%
#                                    filter(str_detect(condition, "BBIBP")) %>%
#                                    select(condition) %>%
#                                    arrange(condition) %>%
#                                    distinct() %>%
#                                    pull(condition),
#                                  ann_vaccines_conditions_ordered %>%
#                                    filter(str_detect(condition, "ZF2001")) %>%
#                                    select(condition) %>%
#                                    arrange(condition) %>%
#                                    distinct() %>%
#                                    pull(condition))

#Order conditions manually
ordered_conditions = c(
  "H (D0)",
  "I (D0)",
  "I (D1)",
  "I (D2)",
  "I (D3)",
  "I (D4)",
  "I-I (D0)",
  "I-I (D1)",
  "I-I (D2)",
  "I-I (D3)",
  "I-I (D5)",
  "BNT-I (D0)",
  "BNT-I (D1)",
  "BNT-I (D2)",
  "BNT-I (D3)",
  "I-BNT (D5)",
  "I (D10, moderate)" ,
  "I (D10, severe)"   ,
  "BNT-I (D10, mild)"      ,
  "BNT-I (D10, severe)"    ,
  "I (D26, moderate)" ,
  "I (D26, severe)"   ,
  "BNT-I (D26, mild)"      ,
  "BNT-I (D26, severe)"    ,
  "I (D51, moderate)" ,
  "I (D51, severe)"   ,
  "BNT-I (D51, severe)"    ,
  "BNT-I-BNT (D51, mild)"  ,
  "BNT-I-BNT (D51, severe)",
  "BNT (V1, D0)",
  "BNT (V1, D6)",
  "BNT (V2, D1)",
  "BNT (V3, D1)",
  "BNT-MO (V3, D1)",
  "ChAd (V1, D0)",
  "ChAd (V1, D3)",
  "ChAd (V1, D6)",
  "ChAd (V1, D7)",
  "ChAd (V2, D1)",
  "ChAd (V2, D3)",
  "ChAd (V2, D7)",
  "ChAd-BNT (V2, D0)" ,
  "ChAd-BNT (V2, D3)" ,
  "ChAd-BNT (V2, D7)" ,
  "ChAd-BNT (V3, D1)" ,
  "BBIBP (V3, D0)"  ,
  "BBIBP (V3, D07)" ,
  "BBIBP (V3, D14)" ,
  "BBIBP (V3, D28)" ,
  "ZF2001 (V3, D0)" ,
  "ZF2001 (V3, D07)",
  "ZF2001 (V3, D14)",
  "ZF2001 (V3, D28)"
)

ann_vaccines_conditions_ordered = ann_vaccines_conditions_ordered %>% 
  mutate(condition = fct_relevel(condition, 
                                 ordered_conditions)) %>%
  mutate(type = if_else(type == "H", "I", type),
         disease_vac = if_else(disease_vac == "H", "I", disease_vac)) %>% 
  arrange(condition)

ann_vaccines_conditions_ordered = ann_vaccines_conditions_ordered %>% 
  mutate(condition_grouped = fct_relevel(condition_grouped, c(ann_vaccines_conditions_ordered$condition_grouped %>% 
                                           as.character() %>% 
                                           unique())))

ann_vaccines_conditions_ordered %>%
  saveRDS(file = here("Annotations and metadata", "ann_vaccines_conditions_ordered.rds"))

```


### Plot Samples by condition


Original conditions
```{r fig.height=10, fig.width=20, cache = TRUE}
ann_vaccines_conditions = read_csv(here("Annotations and metadata", "ann_vaccines_conditions_23-10-24.csv"))
ann_vaccines_samples <- read_csv(here("Annotations and metadata","ann_vaccines_samples.csv"))
ann_vaccines_conditions_ordered <- readRDS(here("Annotations and metadata", "ann_vaccines_conditions_ordered.rds"))
#### SAMPLES BY CONDITION -----
samples_by_condition  = ann_vaccines_samples %>% 
  select(sample, condition = condition_grouped) %>% 
  distinct() %>% 
  group_by(condition) %>% 
  summarise(samples = n()) %>% 
  inner_join(ann_vaccines_conditions %>% select(vaccine, condition, type) %>% distinct(), by = "condition") %>% 
  mutate(condition = fct_relevel(condition, c(ann_vaccines_conditions_ordered$condition_grouped %>% 
                                                as.character() %>% 
                                                unique()))) %>% 
  arrange(condition)

samples_by_condition %>% 
  summarize(ntotal = sum(samples))
```


```{r fig.height=5, fig.width=10, cache = TRUE}
samples_by_condition_plot = samples_by_condition %>% 
  ggplot(aes(x = condition, weight = samples, fill = type)) +
  geom_bar(color = "black", 
           size = 0.2) +
  scale_fill_manual(values = ann_vaxsig_colors$type) +
  theme_minimal() +
  theme(legend.position = "top",
        axis.text.x = element_text(size = 10, color = "black", angle = 45, hjust = 1, vjust = 1),
        axis.text.y = element_blank(), 
        legend.text = element_text(size = 10),
        legend.title = element_text(size = 10),
        plot.title = element_text(size = 15),
        axis.title.y = element_text(angle = 180, color = "white"),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        axis.line.y = element_blank(),
        axis.line.x = element_line(size = 0.5, colour = "black", linetype=1),
        axis.ticks.x = element_line(size = 1, color = "black"),
        plot.margin = margin(0, 1, 1, 1, "cm")) +
  guides(fill=guide_legend(nrow=1,byrow=TRUE)) +
  geom_text(aes(label = samples, y = samples),
             fill = "white",
            position = position_dodge(width = 0), 
            vjust = -0.5, 
            size = 3,
            angle = 0) +
  ggtitle("Samples by condition") +
  labs(x = "") +
  scale_y_continuous(expand = c(0,0), limits = c(0, 100))

samples_by_condition_plot
ggsave(samples_by_condition_plot, file = here("Figures", "demographics_samples-by-condition.png"), width = 10, height = 5)
```


```{r fig.height=10, fig.width=10, cache = TRUE}
samples_by_condition_plot_vertical = samples_by_condition %>% 
  ggplot(aes(y = fct_rev(condition), x = samples, fill = type)) +
  geom_col(color = "black", 
           size = 0.2) +
  scale_fill_manual(values = ann_vaxsig_colors$type) +
  theme_minimal() +
  theme(legend.position = "right",
        axis.text.x = element_blank(),
        axis.text.y = element_text(size = 10, color = "black", hjust = 1), 
        legend.text = element_text(size = 10),
        legend.title = element_text(size = 12),
        plot.title = element_text(size = 15),
        axis.title.y = element_text(angle = 180, color = "white"),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        axis.line.y = element_line(size = 0.5, colour = "black", linetype=1),
        axis.line.x = element_blank(),
        axis.ticks.y = element_line(size = 1, color = "black"),
        plot.margin = margin(0, 0, 0, 0, "cm")) +
  guides(fill=guide_legend(ncol=1,byrow=TRUE)) +
  geom_text(aes(label = samples, x = samples),
             fill = "white",
            position = position_dodge(width = 0), 
            hjust = -0.2,
            vjust = 0.5, 
            size = 5,
            angle = 0) +
  ggtitle("Samples by condition") +
  labs(x = "") +
  scale_x_continuous(expand = c(0,0), limits = c(0, 170))

# ggsave(samples_by_condition_plot_vertical, file = here("Figures", "demographics_samples-by-condition_vertical.png"), width = 20, height = 10)

samples_by_condition_plot_vertical
```

### Plot Samples by study
```{r fig.height=3, fig.width=5}
#### SAMPLES BY CONDITION -----
samples_by_study = ann_vaccines_samples %>% 
  select(gse_id) %>% 
  group_by(gse_id) %>% 
  summarise(samples = n()) %>% 
  ungroup() 

samples_by_study_plot = samples_by_study %>% 
  ggplot(aes(x = fct_reorder(gse_id, -samples), weight = samples, fill = gse_id)) +
  geom_bar(color = "black", size = 0.2) +
  scale_fill_manual(name = "GSE ID", 
        values = ann_vaxsig_colors$gse_id) +
  theme_minimal() +
  theme(legend.position = "none",
        axis.text.x = element_text(size = 10, color = "black", angle = 0),
        axis.text.y = element_blank(), 
        legend.text = element_text(size = 10),
        legend.title = element_text(size = 12),
        plot.title = element_text(size = 15),
        axis.title.y = element_text(angle = 180, color = "white"),
        panel.grid.major = element_blank(), 
        axis.line.x = element_line(size = 1, colour = "black", linetype=1),
        axis.line.y = element_blank(),
        axis.ticks.x = element_line(size = 1, color = "black"),
        panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0, "cm")) +
  guides(fill=guide_legend(nrow=1,byrow=TRUE)) +
  geom_text(aes(label = samples, y = samples),
            position = position_dodge(width = 0), 
            vjust = -0.2, 
            size = 4,
            angle = 0) +
  ggtitle("Samples by GSE ID") +
  labs(x = "") +
  # ylim(0, 310) +
  scale_y_continuous(expand = c(0,0), limits = c(0, 330))

ggsave(samples_by_study_plot, file = here("Figures", "demographics_samples-by-study.png"), width = 5, height =3)

samples_by_study_plot
```


### Plot Age by GSE

```{r fig.height=5, fig.width=6, cache = TRUE}

library(ggstatsplot)
# install.packages("Rmpfr")
# library(Rmpfr)

#### AGE BY GSE -------------
age = ann_vaccines_samples %>% 
  select(age, participant_id, gse_id) %>% 
  distinct()

age_stats = ggbetweenstats(
  data = age,
  x = gse_id,
  y = age,
  p.adjust.method = "BH", 
  centrality.plotting = F,
  centrality.point.args = list(size = 1, color = "red"),
  centrality.label.args = list(size = 2, 
                               nudge_x = 0.5, nudge_y = 1, 
                               segment.linetype = 4,
                               min.segment.length = 0),
  point.args = list(position = ggplot2::position_jitterdodge(dodge.width = 0.6, seed = 123),
                    alpha = 0.3, size = 2, stroke = 0, na.rm = TRUE),
  boxplot.args = list(alpha = 1, width = 0.2, mapping = aes(fill = gse_id)),
  violin.args = list(alpha = 0, width = 1),
  type = "np", 
  ggsignif.args = list(textsize = 2, tip_length = 0.0)) +
  # scale_color_jco()+
  scale_fill_manual(
    values = ann_vaxsig_colors$gse_id) +
  scale_color_manual(values = c(GSE199750 = "black",
                                GSE201533= "black",
                                GSE201530= "black",
                                GSE206023 = "black",
                                GSE189039 = "black")) +
  theme_minimal() +
  theme(legend.position = "none",
        axis.text.x = element_text(size = 10,
                                   color = "black",
                                   angle = 0),
        axis.text.y = element_text(size = 10,
                                   color = "black",
                                   angle = 0),
        # legend.text = element_text(size = 10),
        # legend.title = element_text(size = 15),
        plot.title = element_text(size = 12),
        plot.subtitle = element_text(size = 10),
        axis.title.y = element_text(color = "black"),
        panel.grid.major.y = element_line(size = 0.5,
                                          linetype = 2,
                                          color = "gray75"),
        panel.grid.major.x = element_line(size = 0,
                                          linetype = 2,
                                          color = "gray75"),
        axis.line.x = element_line(size = 1,
                                   colour = "black",
                                   linetype = 1),
        axis.line.y = element_line(size = 0,
                                   colour = "white",
                                   linetype = 1),
        axis.ticks.x = element_line(size = 1, color = "black"),
        axis.ticks.y = element_line(size = 0, color = "gray75"),
        panel.grid.minor = element_blank(),
        plot.margin = margin(0, 0, 0, 0, unit = "cm")) +
  labs(y = "Age",
       x = "",
       color = "GSE ID",
       title = "Participants by age") +
  scale_y_continuous(expand = c(0,0), limits = c(0, 120))

extract_stats(age_stats)

ggsave(age_stats, file = here("Figures", "Demographics_age.png") , width = 6, height = 5)
age_stats
```

### Plot Sex by GSE
This figure's text was manipulated further in a design software.
```{r fig.height=5, fig.width=5, cache = TRUE}
#### AGE BY GSE -------------

sex_stats = ann_vaccines_samples %>% 
  select(sex, participant_id, gse_id) %>% 
  distinct() %>% 
  tabyl(gse_id, sex) %>%
  adorn_percentages() %>% 
  as.data.frame() %>% 
  pivot_longer(cols = -gse_id,
               values_to = "value",
               names_to = "sex") %>% 
    filter(value > 0) %>% 
  filter(sex != "Total") %>% 
  mutate(value = value * 100,
         value = round(value, 0)) %>% 
  group_by(gse_id, sex)

sex_stats_annotated = sex_stats %>% 
 ggplot() +
  aes(
    x = gse_id,
    y = value,
    fill = sex,
    weight = value,
    label = value
  ) +
  geom_col(color = "black") +
  geom_label(vjust = -1, size = 4, color = "black", fill = "white") +
  coord_polar(theta = "y") +
  theme_minimal() +
  labs(title = "Sex by study",
       fill = "Sex") +
  theme(axis.title = element_blank(),
        axis.text = element_blank(),
        plot.title = element_text(size = 15, hjust = 0.5),
        strip.text = element_text(size = 12),
        legend.text = element_text(size = 10),
        legend.title = element_text(size = 10),
        legend.position = "top") +
  facet_wrap(vars(gse_id), scales = "free") +
  scale_fill_manual(values = c("Male" = "#4DBBD5FF",
          "Female" = "#E64B35FF",
          "M/F" = "#8491B4FF"))


sex_stats_nonumb = sex_stats %>% 
 ggplot() +
  aes(
    x = gse_id,
    y = value,
    fill = sex,
    weight = value,
    label = value
  ) +
  geom_col(color = "black") +
  coord_polar(theta = "y") +
  theme_minimal() +
  labs(title = "Sex by study",
       fill = "Sex") +
  theme(axis.title = element_blank(),
        axis.text = element_blank(),
        plot.title = element_text(size = 15, hjust = 0.5),
        strip.text = element_text(size = 12),
        legend.text = element_text(size = 10),
        legend.title = element_text(size = 10),
        legend.position = "top") +
  facet_wrap(vars(gse_id), scales = "free") +
  scale_fill_manual(values = c("Male" = "#4DBBD5FF",
          "Female" = "#E64B35FF",
          "M/F" = "#8491B4FF"))

ggsave(sex_stats_annotated,
       file = here("Figures", "Demographics_Sex_annotated.png") ,
       width = 5, height = 5)

ggsave(sex_stats_nonumb,
       file = here("Figures", "Demographics_Sex_nonumbers.png") ,
       width = 5, height = 5)

sex_stats_annotated
sex_stats_nonumb

```

### Plot Participants by day
```{r cache = TRUE}

#Participants by condition
participants_condition = ann_vaccines_samples %>% 
  select(participant_id, condition = condition_grouped, gse_id) %>% 
  group_by(condition, gse_id) %>% 
  summarize(n_participants_condition = n()) %>% 
  ungroup() %>% 
  inner_join(ann_vaccines_conditions, by = join_by("gse_id", "condition")) %>% 
  relocate(.after = week, gse_id)

#Participants by study
participants_study = ann_vaccines_samples %>% 
  select(participant_id, gse_id) %>% 
  group_by(gse_id) %>% 
  summarize(n_participants_study = n()) %>% 
  ungroup() %>% 
  inner_join(ann_vaccines_conditions, by = join_by("gse_id")) %>% 
  relocate(.after = week, gse_id)

#Samples by GSE
samples_by_gse = ann_vaccines_samples %>% 
  ungroup() %>% 
  select(sample, gse_id) %>% 
  distinct() %>% 
  group_by(gse_id) %>% 
  summarise(gse_sample_total = n()) %>% 
  inner_join(ann_vaccines_conditions %>% select(vaccine, condition) %>% distinct(), by = "condition") %>% 
  mutate(condition = fct_relevel(condition, c(ann_vaccines_conditions_ordered$condition_grouped %>% 
                                                as.character() %>% 
                                                unique()))) %>% 
  select(condition, samples, vaccine) %>% 
  distinct() %>% 
  arrange(condition)

#Samples by day
demographics_time = participants_condition %>%
  distinct() %>% 
  select(condition_grouped, day, type) %>% 
  group_by(condition_grouped) %>% 
  summarise(day = median(day) %>% round(., 0),
            type = type) %>% 
  ungroup() %>% 
  arrange(day) %>% 
  mutate(day_f = as.character(day)) 

#Participants by day
participants_condition_day = ann_vaccines_samples %>% 
  select(participant_id, condition = condition_grouped, gse_id, day) %>% 
  group_by(condition, day) %>% 
  summarize(day= median(day),
            n_participants_day = n()) %>% 
  ungroup() %>% 
  inner_join(ann_vaccines_conditions %>% 
               select(-day), by = join_by("condition")) %>% 
  relocate(.after = week, gse_id)
```

```{r fig.height=5, fig.width=8, cache = TRUE}
#Conditions
demographics_time_plot = demographics_time %>% 
  mutate(day_f = fct_relevel(day_f, c(demographics_time$day_f %>% unique()))) %>% 
  ggplot(.) +
  aes(x = day_f, fill = fct_relevel(type, "H", "I", "RNA", "VV", "SU", "IN")) +
  geom_bar(size = 0.3, color = "black") +
  labs(
    x = "Day",
    y = "Samples",
    title = "Samples by day",
    fill = "Type"
  ) +
  theme_minimal() +
  theme(
    legend.position = c(0.95,0.5),
    axis.text.x = element_text(size = 10,
                               color = "black",
                               angle = 0),
    axis.text.y = element_text(size = 10,
                               color = "black",
                               angle = 0),
    legend.text = element_text(size = 10),
    legend.title = element_text(size = 12),
    plot.title = element_text(size = 15),
    axis.title.y = element_text(color = "black"),
    panel.grid.major.y = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
    panel.grid.major.x = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75"),
    axis.line.x = element_line(size = 1,
                               colour = "black",
                               linetype = 1),
    axis.line.y = element_line(size = 0 ,
                               colour = "black",
                               linetype = 1),
    axis.ticks.x = element_line(size = 1, color = "black"),
    axis.ticks.y = element_line(size = 0, color = "black"),
    panel.grid.minor = element_blank(),
    plot.margin = margin(1, 0, 0, 0, "cm")
  ) +
  scale_fill_manual(values = ann_vaxsig_colors$Platform) +
  scale_y_continuous(expand = c(0,0), 
                     breaks = seq(0, 350, by = 50))

demographics_time_plot %>%
  ggsave(file = here("Figures", "Demographics_Time.png"),
         width = 8, height = 5)

demographics_time_plot


```




### Plot Studies designs

Ungrouped
```{r fig.height=6, fig.width=6}
library(ggforce)

demographics_days_samples_doses = ann_vaccines_conditions_ordered %>% 
  ungroup() %>% 
  mutate(condition = as.factor(condition),
         condition = fct_relevel(condition, as.character(unique(ann_vaccines_conditions_ordered$condition))),
         day_f = as.factor(day),
         day_f = fct_reorder(day_f, day),
         dose_f = as.factor(dose),
         dose_f = fct_reorder(dose_f, dose),
         condition = if_else(condition == "BNT (V1, D0)" & 
                               gse_id == "GSE189039", "H (D0)", 
                             condition)) %>%
  separate(condition, into = c("condition_2"), sep = " \\(", remove = F) %>%
  mutate(
    # dose_f = if_else(day_f == "0", "0", dose_f),
         condition_2 = if_else(severity != "H", 
                             paste0(condition_2, " (", severity, ")"),
                             condition_2),
         condition = fct_relevel(condition, as.character(unique(ann_vaccines_conditions_ordered$condition))),
         dose_f = if_else(dose == "0" & infection != "0", "Infection", dose_f)) %>% 
  arrange(condition)

demographics_plot_days_samples_doses = 
  demographics_days_samples_doses %>% 
  mutate(condition_2 = fct_relevel(condition_2,
                                   as.character(unique(demographics_days_samples_doses$condition_2)))) %>%
  ggplot() +
  aes(x = fct_reorder(day_f, day),
      y = fct_rev(condition_2), 
      fill = dose_f,
      size = samples,
      group = dose_f) +
  geom_point(
    position = position_dodge(width = 0.5),
    pch = 21,
    color = "black",
    data = ~ filter(.x, gse_id %in% c("GSE199750", "GSE201533"))
  ) +
  geom_point(
    position = position_identity(),
    pch = 21,
    color = "black",
    data = ~ filter(.x, !gse_id %in% c("GSE199750", "GSE201533"))
  ) +
  scale_fill_manual(values = c(ann_vaxsig_colors$dose, "Infection" = "red")) +
  scale_size_binned(breaks = c(1, 10, 50),
                    range = c(2, 5)) +
  theme_minimal() +
  theme(legend.position = "top",
        strip.text.x = element_text(size = 10, color = "black"),
    axis.text.x = element_text(size = 10,
                               color = "black",
                               angle = 0),
    axis.text.y = element_text(size = 10,
                               color = "black",
                               angle = 0),
    legend.text = element_text(size = 10),
    legend.title = element_text(size = 15),
    plot.title = element_text(size = 15),
    axis.title.y = element_text(color = "black"),
    panel.grid.major.y = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
    panel.grid.major.x = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75"),
    axis.line.x = element_line(size = 1,
                               colour = "black",
                               linetype = 1),
    axis.line.y = element_line(size = 1 ,
                               colour = "black",
                               linetype = 1),
    axis.ticks.x = element_line(size = 1, color = "black"),
    axis.ticks.y = element_line(size = 0, color = "black"),
    panel.grid.minor = element_blank(),
    plot.margin = margin(0, 2, 0, 0, unit = "cm")) +
  labs(x = "Days",
       y = "",
       # title = "Studies design",
       # shape = "Disease or vaccination",
       size = "Samples",
       fill = "Dose") +
  ggforce::facet_col(~fct_relevel(gse_id, "GSE201530", "GSE189039", "GSE199750", "GSE201533"),
                     scales = "free_y", space = "free")


demographics_plot_days_samples_doses

# demographics_plot_days_samples_doses %>%
#   ggsave(filename = here("Figures", "Demographics_studies_design.png"),
#          width = 6,
#          height = 6)


```



Grouped conditions

```{r fig.height=6, fig.width=6}
demographics_days_samples_doses = ann_vaccines_conditions %>% 
  ungroup() %>% 
  mutate(condition = as.factor(condition),
         condition = fct_relevel(condition, as.character(unique(ann_vaccines_conditions_ordered$condition_grouped))),
         day_f = as.character(day),
         dose_f = as.factor(dose),
         dose_f = fct_reorder(dose_f, dose),
         condition = if_else(condition == "BNT (V1, D0)" & 
                               gse_id == "GSE189039", "H (D0)", 
                             condition)) %>%
  separate(condition, into = c("condition_2"), sep = " \\(", remove = F) %>%
  mutate(
    # dose_f = if_else(day_f == "0", "0", dose_f),
         # condition_2 = if_else(severity != "H", 
         #                     paste0(condition_2, " (", severity, ")"),
         #                     condition_2)),
         condition = fct_relevel(condition, as.character(unique(ann_vaccines_conditions_ordered$condition_grouped))),
         dose_f = if_else(dose == "0" & infection != "0", "Infection", dose_f)) %>% 
  arrange(condition)

demographics_plot_days_samples_doses = 
  demographics_days_samples_doses %>%
  ungroup() %>% 
  mutate(condition_2 = fct_relevel(condition_2,
                                   as.character(unique(demographics_days_samples_doses$condition_2)))) %>% 
    mutate(day_f = fct_relevel(day_f, c(demographics_time$day_f %>% unique()))) %>% 
  # group_by(gse_id, dose_f) %>% 
  ggplot(.) +
  aes(x = day_f,
      y = fct_rev(condition_2), 
      fill = dose_f,
      size = samples,
      group = dose_f) +
  geom_point(
    position = position_dodge(width = 0.5),
    pch = 21,
    color = "black",
    data = ~ filter(.x, gse_id %in% c("GSE199750", "GSE201533"))
  ) +
  geom_point(
    position = position_identity(),
    pch = 21,
    color = "black",
    data = ~ filter(.x, !gse_id %in% c("GSE199750", "GSE201533"))
  ) +
  scale_fill_manual(values =  c(ann_vaxsig_colors$dose, "Infection" = "red")) +
  scale_size_binned(breaks = c(1, 10, 50),
                    range = c(2, 5)) +
  theme_minimal() +
    theme(legend.position = "top",
    axis.text.x = element_text(size = 10,
                               color = "black",
                               angle = 0),
    strip.text.x = element_text(size = 12, color = "black"),
    axis.text.y = element_text(size = 10,
                               color = "black",
                               angle = 0),
    legend.text = element_text(size = 10),
    legend.title = element_text(size = 10),
    plot.title = element_text(size = 10),
    axis.title.y = element_text(color = "black"),
    panel.grid.major.y = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
    panel.grid.major.x = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
    axis.line.x = element_line(size = 1,
                               colour = "black",
                               linetype = 1),
    axis.line.y = element_line(size = 1 ,
                               colour = "black",
                               linetype = 1),
    axis.ticks.x = element_line(size = 1, color = "black"),
    axis.ticks.y = element_line(size = 0, color = "black"),
    panel.grid.minor = element_blank(),
    plot.margin = margin(0, 2, 0, 0, unit = "cm")) +
  scale_x_discrete(limits = c(demographics_time$day_f %>% unique())) +
  labs(x = "Days",
       y = "",
       # title = "Studies design - Grouped conditions",
       # shape = "Disease or vaccination",
       size = "Samples",
       fill = "Dose") +
  ggforce::facet_col(~fct_relevel(gse_id, "GSE201530", "GSE189039", "GSE199750", "GSE201533"), scales = "free_y", space = "free") +
  guides(fill = guide_legend(override.aes = list(size=4)))

demographics_plot_days_samples_doses

demographics_plot_days_samples_doses %>%
  ggsave(filename = here("Figures", "Demographics_studies_design_Grouped.png"),
         width = 6,
         height = 6)
```

```{r fig.height=10, fig.width=7}

#Conditions
demographics_time_plot_2 = demographics_time %>% 
  mutate(day_f = fct_relevel(day_f, c(demographics_time$day_f %>% unique()))) %>% 
  ggplot(.) +
  aes(x = day_f, fill = fct_relevel(type, "H", "I", "RNA", "VV", "SU", "IN")) +
  geom_bar(size = 0.2, color = "black") +
  labs(
    x = "",
    y = "Samples",
    title = "Samples by day",
    fill = "Type"
  ) +
  theme_minimal() +
  theme(
    legend.position = c(0.95,0.5),
    axis.text.x = element_text(size = 10,
                               color = "black",
                               angle = 0),
    axis.text.y = element_text(size = 10,
                               color = "black",
                               angle = 0),
    legend.text = element_text(size = 10),
    legend.title = element_text(size = 10),
    plot.title = element_text(size = 10),
    axis.title.y = element_text(color = "black"),
    panel.grid.major.y = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
    panel.grid.major.x = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75"),
    axis.line.x = element_line(size = 1,
                               colour = "black",
                               linetype = 1),
    axis.line.y = element_line(size = 1 ,
                               colour = "black",
                               linetype = 1),
    axis.ticks.x = element_line(size = 1, color = "black"),
    axis.ticks.y = element_line(size = 0, color = "black"),
    panel.grid.minor = element_blank(),
    plot.margin = margin(1, 0, 0, 0, "cm")
  ) +
  scale_fill_manual(values = ann_vaxsig_colors$Platform) +
  scale_y_continuous(expand = c(0,0), 
                     breaks = seq(0, 350, by = 50))


demographics_plot_days_samples_doses_2 = 
  demographics_days_samples_doses %>%
  ungroup() %>% 
  mutate(condition_2 = fct_relevel(condition_2,
                                   as.character(unique(demographics_days_samples_doses$condition_2)))) %>% 
    mutate(day_f = fct_relevel(day_f, c(demographics_time$day_f %>% unique()))) %>% 
  # group_by(gse_id, dose_f) %>% 
  ggplot(.) +
  aes(x = day_f,
      y = fct_rev(condition_2), 
      fill = dose_f,
      size = samples,
      group = dose_f) +
  geom_point(
    position = position_dodge(width = 0.5),
    pch = 21,
    color = "black",
    data = ~ filter(.x, gse_id %in% c("GSE199750", "GSE201533"))
  ) +
  geom_point(
    position = position_identity(),
    pch = 21,
    color = "black",
    data = ~ filter(.x, !gse_id %in% c("GSE199750", "GSE201533"))
  ) +
  scale_fill_manual(values =  c(ann_vaxsig_colors$dose, "Infection" = "red")) +
  scale_size_binned(breaks = c(1, 10, 50),
                    range = c(2, 5)) +
  theme_minimal() +
    theme(legend.position = "right",
    axis.text.x = element_text(size = 10,
                               color = "black",
                               angle = 0),
    strip.text.x = element_text(size = 10, color = "black"),
    axis.text.y = element_text(size = 10,
                               color = "black",
                               angle = 0),
    legend.text = element_text(size = 10),
    legend.title = element_text(size = 10),
    plot.title = element_text(size = 10),
    axis.title.y = element_text(color = "black"),
    panel.grid.major.y = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
    panel.grid.major.x = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
    axis.line.x = element_line(size = 1,
                               colour = "black",
                               linetype = 1),
    axis.line.y = element_line(size = 1 ,
                               colour = "black",
                               linetype = 1),
    axis.ticks.x = element_line(size = 1, color = "black"),
    axis.ticks.y = element_line(size = 0, color = "black"),
    panel.grid.minor = element_blank(),
    plot.margin = margin(0, 2, 0, 0, unit = "cm")) +
  scale_x_discrete(limits = c(demographics_time$day_f %>% unique())) +
  labs(x = "Days",
       y = "",
       # title = "Studies design - Grouped conditions",
       # shape = "Disease or vaccination",
       size = "Samples",
       fill = "Dose") +
  ggforce::facet_col(~fct_relevel(gse_id, "GSE201530", "GSE189039", "GSE199750", "GSE201533"), scales = "free_y", space = "free")  + 
  guides(fill = guide_legend(override.aes = list(size=4)))


 (samples_by_study_plot/ age_stats) 

(demographics_time_plot_2 / demographics_plot_days_samples_doses_2) +
  plot_layout(axes = "collect", guides = "collect", heights = c(1, 2))
```





### Immunological endpoints

```{r}
Serology_data <- read_csv("raw_data/GSE199750/Serology_data.csv")
Serology_data %>%
  clean_names() %>%
  separate(sample_id, into = c("study", "participant_id", "dose")) %>%
  inner_join(ann_vaccines_samples %>% 
               filter(gse_id == "GSE199750") %>% 
               separate(participant_id, into = c("study", "participant_id")) , by = "participant_id") %>% 
  view()
  group_by(dose) %>%
  drop_na(pseudovirus_neutralisation_id50) %>%
  mutate(pseudovirus_neutralisation_id50 = ifelse(pseudovirus_neutralisation_id50 <= 0, 1e-10, pseudovirus_neutralisation_id50)) %>%
  summarise(
    pseudovirus_neutralisation_id50_gmt = log10(prod(pseudovirus_neutralisation_id50)^(1/n()))
  )

```




# 5. Differential Gene Expression Analysis


```{r}
all_matrices_counts = all_matrices_counts %>% 
  column_to_rownames("genes")
```


```{r}
# Get all comparisons
generate_all_DESeq2_results <- function(dds) {
  result_names <- resultsNames(dds)
  all_results <- data.frame()
  
  for (name in result_names) {
    condition_label <- name
      result_df <- results(dds, name = name) %>%
      as.data.frame() %>%
      arrange(padj) %>%
      mutate(condition = condition_label) %>%
      rownames_to_column("genes")
        
    all_results <- bind_rows(all_results, result_df)
  }
  
  return(all_results)
}
```


## GSE206023

```{r}

gse206023_metadata = ann_vaccines_samples %>% 
  filter(gse_id == "GSE206023")

gse206023_counts_ready <- all_matrices_counts %>% 
  select(c(gse206023_metadata$sample))

countData <- gse206023_counts_ready %>%
  select(colnames(gse206023_counts_ready) %>%
      as.data.frame() %>%
      rename(sample = ".") %>%
      arrange(sample) %>%
      pull()
  )


colData <- gse206023_metadata %>% 
  mutate(condition = case_when(condition == "BBIBP (V3, D0)" ~ "H (D0)",
                               condition == "ZF2001 (V3, D0)" ~ "H (D0)",.default = condition),
         condition = as.factor(condition),
         condition = fct_relevel(condition, "H (D0)"),
         participant_id = as_factor(participant_id)) %>% 
  arrange(sample) %>% 
  select(sample, everything())


# DESeq2
dds <- DESeqDataSetFromMatrix(countData = countData, 
                              colData = colData, 
                              design = ~ age + sex + condition)

keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]
dds <- DESeq(dds)

resultsNames(dds)

gse206023_dds = generate_all_DESeq2_results(dds)

gse206023_DEGs =  gse206023_dds %>% 
  mutate(condition = case_when(condition == "condition_BBIBP..V3..D07._vs_H..D0." ~ "BBIBP (V3, D07)", 
                                condition == "condition_BBIBP..V3..D14._vs_H..D0." ~ "BBIBP (V3, D14)", 
                                condition == "condition_BBIBP..V3..D28._vs_H..D0." ~ "BBIBP (V3, D28)", 
                                condition == "condition_ZF2001..V3..D07._vs_H..D0." ~ "ZF2001 (V3, D07)",
                                condition == "condition_ZF2001..V3..D14._vs_H..D0." ~ "ZF2001 (V3, D14)",  
                                condition == "condition_ZF2001..V3..D28._vs_H..D0." ~ "ZF2001 (V3, D28)", 
                               condition == "age" ~ "GSE206023 Ages",
                               condition == "sex_Male_vs_Female" ~ "GSE206023 Sex",
                                .default = condition)) %>% 
  filter(!condition %in% c("Intercept"))

gse206023_DEGs %>% 
  select(condition) %>% 
  distinct()


gse206023_DEGs %>% 
  write_csv(file = here("DGE analysis", "gse206023_DEGs_wald.csv"))

gse206023_DEGs

```

## GSE201533

```{r}
############ Matrix counts ############
gse201533_counts <- as.data.frame(counts_gse201533)
rownames(gse201533_counts) <- gse201533_counts$geneid 
gse201533_counts <- gse201533_counts[, -1]
saveRDS(gse201533_counts, file="gse201533_counts_ready.rds")

########## Metadata #############
gse201533_metadata <- as.data.frame(gse201533_metadata)
rownames(gse201533_metadata) <- gse201533_metadata$GEO 

gse201533_metadata$Timepoint <- sub(".*(0|2|3|4|7).*", "\\1", gse201533_metadata$`Sample name`)

head(gse201533_metadata)

gse201533_metadata <- gse201533_metadata[, -2]
gse201533_metadata <- gse201533_metadata[, -1]

saveRDS(gse201533_metadata, file="gse201533_metadata_ready.rds")
```

DESeq2
```{r}

gse201533_metadata_ready = ann_vaccines_samples %>% 
  filter(gse_id == "GSE201533") 
gse201533_counts_ready <- all_matrices_counts %>% 
  select(c(gse201533_metadata_ready$sample))

countData <- gse201533_counts_ready %>% 
  select(c(gse201533_metadata_ready$sample))
colData <- gse201533_metadata_ready %>% 
  mutate(condition = as.factor(condition),
         condition = fct_relevel(condition, "ChAd (V1, D0)"),
         sex = as.factor(sex),
         age_group = case_when(
    age <= 24 ~ "Young",
    age >= 25 & age <= 64 ~ "Adult",
    age >= 65 ~ "Elderly",
    TRUE ~ NA)) %>% 
  select(sample, everything()) 

# DESeqDataSet
dds <- DESeqDataSetFromMatrix(countData = countData, 
                              colData = colData, 
                              design = ~ age + 
                                # sex + (all samples are Female)
                                condition)
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]

# Execute DESeq
dds <- DESeq(dds)
resultsNames(dds)

gse201533_dds <- generate_all_DESeq2_results(dds)


gse201533_DEGs =  gse201533_dds %>% 
  mutate(condition = case_when(condition == "condition_ChAd..V1..D3._vs_ChAd..V1..D0." ~ "ChAd (V1, D3)", 
                                condition == "condition_ChAd..V1..D7._vs_ChAd..V1..D0." ~ "ChAd (V1, D7)", 
                                condition == "condition_ChAd..V2..D3._vs_ChAd..V1..D0." ~ "ChAd (V2, D3)", 
                                condition == "condition_ChAd..V2..D7._vs_ChAd..V1..D0." ~ "ChAd (V2, D7)",
                                condition == "condition_ChAd.BNT..V2..D3._vs_ChAd..V1..D0." ~ "ChAd-BNT (V2, D3)",  
                                condition == "condition_ChAd.BNT..V2..D7._vs_ChAd..V1..D0." ~ "ChAd-BNT (V2, D7)", 
                               condition == "condition_ChAd.BNT..V2..D0._vs_ChAd..V1..D0." ~ "ChAd-BNT (V2, D0)",
                               condition == "age" ~ "GSE201533 Ages",
                                .default = condition)) %>% 
  filter(!condition %in% c("Intercept"))

gse201533_DEGs %>% 
  select(condition) %>% 
  distinct()


gse201533_DEGs %>% 
  write_csv(here("DGE analysis", "gse201533_DEGs_wald.csv"))

gse201533_DEGs

```


## GSE199750

```{r}
# Required files

gse199750_metadata = ann_vaccines_samples %>% 
  filter(gse_id == "GSE199750")

gse199750_counts_ready <- all_matrices_counts %>% 
  select(c(gse199750_metadata$sample))

countData <- gse199750_counts_ready %>% 
  select(c(gse199750_metadata$sample))

colData <- gse199750_metadata %>% 
  select(sample, everything()) %>% 
  mutate(condition = if_else(day == 0, 
                             "H (D0)", condition),
         condition = as.factor(condition),
         condition = fct_relevel(condition, "H (D0)")) %>% 
  arrange(sample)


###### DESeqDataSet objetct
dds <- DESeqDataSetFromMatrix(countData = countData, 
                              colData = colData, 
                              design = ~ age + sex + condition)


###### Pre-filtering
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]

###### Execute DESeq2
dds <- DESeq(dds)

###### Retrieve comparisons
resultsNames(dds)

gse199750_dds <- generate_all_DESeq2_results(dds)

gse199750_DEGs =  gse199750_dds %>% 
  mutate(condition = case_when(condition == "condition_BNT..V1..D6._vs_H..D0." ~ "BNT (V1, D6)", 
                                condition == "condition_ChAd..V1..D6._vs_H..D0." ~ "ChAd (V1, D6)", 
                                condition == "condition_BNT..V2..D1._vs_H..D0." ~ "BNT (V2, D1)", 
                                condition == "condition_ChAd..V2..D1._vs_H..D0." ~ "ChAd (V2, D1)",
                                condition == "condition_BNT..V3..D1._vs_H..D0." ~ "BNT (V3, D1)",  
                                condition == "condition_BNT.MO..V3..D1._vs_H..D0." ~ "BNT-MO (V3, D1)", 
                                condition == "condition_ChAd.BNT..V3..D1._vs_H..D0." ~ "ChAd-BNT (V3, D1)", 
                               condition == "age" ~ "GSE199750 Ages",
                                condition == "sex_Male_vs_Female" ~ "GSE199750 Sex",
                                .default = condition)) %>% 
  filter(!condition %in% c("Intercept", "condition_BNT.MO..V1..D6._vs_H..D0."))


gse199750_DEGs %>% 
  write_csv(file = here("DGE analysis", "gse199750_DEGs_wald.csv"))

gse199750_DEGs$condition %>% unique()


```

## Infection: GSE189039 and GSE201530

```{r eval=FALSE, echo=TRUE, cache = TRUE}
#Standardize counts tables
############### GSE201530
# Reorder table
infection_counts_ready = gse189039_counts_ready %>% 
  bind_cols(gse201530_counts_ready) %>% 
  rownames_to_column("genes") %>% 
  filter(genes %in% c(all_matrices_counts$genes)) %>% 
  column_to_rownames("genes")

infection_metadata = ann_vaccines_samples %>% 
  select(sample, condition, condition_grouped, age, sex, gse_id, variant, severity, country, enroll_time, sequencer) %>% 
  left_join(colnames(infection_counts_ready) %>% as.data.frame() %>% rename(sample = ".")) %>% 
  filter(sample %in% colnames(infection_counts_ready)) %>% 
  mutate(condition_grouped = if_else(condition_grouped == "BNT (V1, D0)", "H (D0)", condition_grouped),
         condition_grouped = as.factor(condition_grouped),
    condition_grouped = fct_relevel(condition_grouped, "H (D0)"),
    country = as.factor(country),
    variant = fct_relevel(variant, "None", "Beta", "Omicron"),
    enroll_time = as.factor(enroll_time),
    sequencer = as.factor(sequencer),
    sex = as.factor(sex),
    age_f = as.factor(age),
    age_f = fct_reorder(age_f, age),
    age_group = case_when(
    age <= 24 ~ "Young",
    age >= 25 & age <= 64 ~ "Adult",
    age >= 65 ~ "Elderly",
    TRUE ~ NA),
    gse_id = as.factor(gse_id),
    severity = as.factor(severity)) %>% 
  select(sample, condition, condition_grouped, age_f,age_group, age, sex, severity, variant)

infection_counts_ready = infection_counts_ready %>%
  select(infection_metadata$sample)

dim(infection_counts_ready)

infection_metadata = infection_metadata %>% 
  filter(sample %in% c(colnames(infection_counts_ready))) %>% 
  distinct()

countData <- infection_counts_ready
colData <- infection_metadata

###### DESeqDataSet
dds <- DESeqDataSetFromMatrix(countData = countData, 
                              colData = colData, 
                              design = ~ age + 
                                sex +
                                # severity +
                                # variant + 
                                condition_grouped)

keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]

###### Execute DESeq
dds <- DESeq(dds)

resultsNames(dds)
###### overall DEGs
res <- results(dds) %>% 
  as.data.frame() %>% 
  rownames_to_column("genes")


gseinfection_dds <- generate_all_DESeq2_results(dds)


infection_DEGs =  gseinfection_dds %>%
  mutate(condition = case_when(condition == "condition_grouped_BNT.I..D0.5._vs_H..D0." ~ "BNT-I (D0-5)",
                                condition == "condition_grouped_BNT.I..D10..hospital._vs_H..D0." ~ "BNT-I (D10, hospital)",
                                condition == "condition_grouped_BNT.I..D26..hospital._vs_H..D0." ~ "BNT-I (D26, hospital)",
                                condition == "condition_grouped_BNT.I..D51..hospital._vs_H..D0." ~ "BNT-I (D51, hospital)",
                                condition == "condition_grouped_I..D0.4._vs_H..D0." ~ "I (D0-4)",
                                condition == "condition_grouped_I..D10..hospital._vs_H..D0." ~ "I (D10, hospital)",
                                condition == "condition_grouped_I..D26..hospital._vs_H..D0." ~ "I (D26, hospital)",
                                condition == "condition_grouped_I..D51..hospital._vs_H..D0." ~ "I (D51, hospital)",
                                condition == "condition_grouped_I.BNT..D5._vs_H..D0." ~ "I-BNT (D5)",
                                condition == "condition_grouped_I.I..D0.5._vs_H..D0." ~ "I-I (D0-5)",
                                condition == "condition_grouped_BNT.I.BNT..D51..hospital._vs_H..D0." ~ "BNT-I-BNT (D51, hospital)",
                                condition == "age" ~ "Infection Ages",
                                condition == "sex_Male_vs_Female" ~ "Infection Sex",
                                .default = condition)) %>%
  filter(!condition %in% c("Intercept"))

infection_DEGs$condition %>% unique()


# infection_DEGs =  gseinfection_dds %>% 
#   mutate(condition = case_when(condition == "condition_BNT.I..D0._vs_H..D0." ~ "BNT-I (D0)", 
#                                 condition == "condition_BNT.I..D1._vs_H..D0." ~ "BNT-I (D1)", 
#                                 condition == "condition_BNT.I..D2._vs_H..D0." ~ "BNT-I (D2)", 
#                                 condition == "condition_BNT.I..D3._vs_H..D0." ~ "BNT-I (D3)",
#                                 condition == "condition_I.BNT..D5._vs_H..D0." ~ "I-BNT (D5)",  
#                                 condition == "condition_I..D0._vs_H..D0." ~ "I (D0)", 
#                                 condition == "condition_I..D1._vs_H..D0." ~ "I (D1)", 
#                                 condition == "condition_I..D2._vs_H..D0." ~ "I (D2)", 
#                                 condition == "condition_I..D3._vs_H..D0." ~ "I (D3)", 
#                                 condition == "condition_I..D4._vs_H..D0." ~ "I (D4)",
#                                 condition == "condition_I.I..D0._vs_H..D0." ~ "I-I (D0)", 
#                                 condition == "condition_I.I..D1._vs_H..D0." ~ "I-I (D1)", 
#                                 condition == "condition_I.I..D2._vs_H..D0." ~ "I-I (D2)", 
#                                 condition == "condition_I.I..D3._vs_H..D0." ~ "I-I (D3)", 
#                                 condition == "condition_I.I..D5._vs_H..D0." ~ "I-I (D5)", 
#                                 condition == "condition_BNT.I..D10..mild._vs_H..D0." ~ "BNT-I (D10, mild)", 
#                                 condition == "condition_BNT.I..D10..severe._vs_H..D0." ~ "BNT-I (D10, severe)", 
#                                 condition == "condition_BNT.I..D26..mild._vs_H..D0." ~ "BNT-I (D26, mild)", 
#                                 condition == "condition_BNT.I..D26..severe._vs_H..D0." ~ "BNT-I (D26, severe)", 
#                                 condition == "condition_BNT.I..D51..severe._vs_H..D0." ~ "BNT-I (D51, severe)", 
#                                 condition == "condition_BNT.I.BNT..D51..mild._vs_H..D0." ~ "BNT-I-BNT (D51, mild)", 
#                                 condition == "condition_BNT.I.BNT..D51..severe._vs_H..D0." ~ "BNT-I-BNT (D51, severe)", 
#                                 condition == "condition_I..D10..moderate._vs_H..D0." ~ "I (D10, moderate)", 
#                                 condition == "condition_I..D10..severe._vs_H..D0." ~ "I (D10, severe)", 
#                                 condition == "condition_I..D26..moderate._vs_H..D0." ~ "I (D26, moderate)", 
#                                 condition == "condition_I..D26..severe._vs_H..D0." ~ "I (D26, severe)", 
#                                 condition == "condition_I..D51..moderate._vs_H..D0." ~ "I (D51, moderate)", 
#                                 condition == "condition_I..D51..severe._vs_H..D0." ~ "I (D51, severe)", 
#                                condition == "age" ~ "Infection Ages",
#                                 condition == "sex_Male_vs_Female" ~ "Infection Sex",
#                                 .default = condition)) %>% 
#   filter(!condition %in% c("Intercept"))

infection_DEGs %>% 
  write_csv(file = here("DGE analysis", "infection_DEGs_wald_grouped.csv"))
```


## Merge all

```{r}
gse201533_DEGs = read_csv(file = here("DGE analysis", "gse201533_DEGs_wald.csv"))
gse206023_DEGs = read_csv(file = here("DGE analysis", "gse206023_DEGs_wald.csv"))
infection_DEGs = read_csv(file = here("DGE analysis", "infection_DEGs_wald_grouped.csv"))
gse199750_DEGs = read_csv(file = here("DGE analysis", "gse199750_DEGs_wald.csv"))


all_degs = bind_rows(gse199750_DEGs,
                     gse201533_DEGs,
                     gse206023_DEGs,
                     infection_DEGs) %>% 
  mutate(condition = if_else(condition == "ChAd (V3, D1)", "BNT (V3, D1)", condition))


all_degs %>% 
  select(condition) %>% 
  distinct()

#Save
all_degs %>% 
  saveRDS(file = here("DGE analysis", "all_studies_degs_wald.rds"))

all_degs
```

## Weight and adjust p values

```{r}
all_degs = read_rds(file = here("DGE analysis", "all_studies_degs_wald.rds"))
all_degs_annotated = all_degs %>%
  inner_join(ann_vaccines_conditions %>% 
               select(condition, n = samples) %>% 
               distinct(),
             by = "condition") %>%
  mutate(pvalue = if_else(is.na(pvalue), 1, pvalue)) %>% 
  select(-padj) %>% 
  clean_names()

all_degs_annotated$condition %>% unique()

#Weight p values by sample size
all_studies_degs_weighted = all_degs_annotated %>%
  group_by(genes, condition) %>%
  arrange(genes) %>% 
  summarise(
    n = sum(n),
    meta_p = pvalue/sqrt(n),
    log2fold_change = log2fold_change,
    base_mean, lfc_se, stat
  ) %>% 
  distinct() %>% 
  ungroup()

all_studies_degs_weighted = all_studies_degs_weighted %>% 
  mutate(fdr = p.adjust(meta_p, method = "BH"),
         direction = case_when(log2fold_change >= 1 ~ "UP",
                               log2fold_change <= -1 ~ "DOWN",
                               .default = "NEUTRAL"))


all_studies_degs_weighted %>% 
  saveRDS(file = here("DGE analysis", "all_studies_degs_weighted.rds"))
```


# 6. DEGs analysis

## Up and down-regulated DEGs

```{r}
#Required files

#Filter fdr by sample size minimum
all_degs_p_05_vac_infected = read_rds(file = here("DGE analysis", "all_studies_degs_weighted.rds")) %>%
  filter(if_else(n >= 5, fdr <= 0.05, fdr <= (0.05 * 6/n))) 

ann_vaccines_conditions = read_csv(here("Annotations and metadata", "ann_vaccines_conditions.csv"))
ann_vaccines_conditions_ordered = read_rds(here("Annotations and metadata", "ann_vaccines_conditions_ordered.rds"))

ann_vaccines = ann_vaccines_conditions
all_degs = all_degs_p_05_vac_infected %>% 
  group_by(condition, direction) %>% 
  summarise(n_degs = n()) %>% 
  inner_join(ann_vaccines, by="condition") %>% 
  ungroup()

all_degs_total = all_degs %>% 
  pivot_wider(names_from = "direction", values_from = "n_degs") %>% 
  group_by(condition) %>% 
  mutate(TOTAL = sum(DOWN + NEUTRAL + UP)) %>% 
  select(condition, DOWN:TOTAL) %>% 
  distinct()

all_degs_total
```

Horizontal

```{r fig.height=6, fig.width=12}
# Plot 1
ggplot_degs_bars = all_degs %>%
    mutate(condition = fct_relevel(condition, c(ann_vaccines_conditions_ordered$condition_grouped %>% as.character() %>% unique()))) %>% 
  filter(direction != "NEUTRAL") %>% 
 ggplot() +
  aes(x = condition, fill = direction, weight = n_degs) +
  geom_bar(position = "dodge", size = 0.2, color = "black") +
  scale_fill_manual(
    values = c(DOWN = "#4DBBD5FF",
    NEUTRAL = "gray80",
    UP = "#d90429")) +
  theme_minimal() +
   ylim(0, 3000) +
  geom_text(aes(label = n_degs, y = n_degs), 
            position = position_dodge(width = 1), 
            angle = 90,
            vjust = 0.5, 
            hjust = -0.2, 
            size = 3) +
    theme(legend.position = "right",
    strip.text.x = element_text(size = 12, color = "black"),
    legend.text = element_text(size = 10),
    legend.title = element_text(size = 10),
    plot.title = element_text(size = 10),
    axis.title.y = element_text(color = "black"),
    panel.grid.major.y = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
    panel.grid.major.x = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75"),
    axis.line.x = element_line(size = 1,
                               colour = "black",
                               linetype = 1),
    axis.ticks.x = element_line(size = 1, color = "black"),
    axis.line.y = element_line(size = 0 ,
                               colour = "black",
                               linetype = 1),
    axis.ticks.y = element_line(size = 0, color = "black"),
    panel.grid.minor.y = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
    plot.margin = margin(0, 2, 0, 0, unit = "cm"),
    axis.text.x = element_text(angle = 45, vjust = 1, hjust=1, size = 10),
    axis.text.y = element_text(vjust = 0.5, size = 10)) +
  scale_y_continuous(expand = c(0,0), limits = c(0, 3100)) +
  labs(x = "",
       y = "Total",
       fill = "",
       title = "DEGs barplot")


ggsave(ggplot_degs_bars, file = here("Figures", "DEGs_up_down_barplot_pvalue05_1.png"),  width = 12, height = 6)

ggplot_degs_bars

```

Vertical

```{r fig.height=10, fig.width=10}
# Plot 1
ggplot_degs_bars_vertical_nolabels = all_degs %>%
    mutate(condition = fct_relevel(condition, c(ann_vaccines_conditions_ordered$condition_grouped %>% as.character() %>% unique())),
           title = "DEGs by direction") %>% 
  # filter(direction != "NEUTRAL") %>% 
 ggplot() +
  aes(x = n_degs, y = fct_rev(condition), fill = fct_relevel(direction, "UP", "DOWN", "NEUTRAL")) +
  geom_col(position = "dodge", size = 0.5, color = "black", width = 0.8) +
  scale_fill_manual(
    values = c(DOWN = "#4DBBD5FF",
    NEUTRAL = "gray90",
    UP = "#d90429")) +
  theme_minimal() +
      theme(legend.position = "right",
        strip.text.x = element_text(color = "black", size = 12), 
        axis.text.x = element_text(color = "black", size = 10),
        axis.text.y = element_text(vjust = 0.5, size = 10, color = "black"),
        axis.line = element_line(size = 1, colour = "black", linetype=1),
        axis.ticks = element_line(size = 0.5, color = "black"),
        panel.grid.major.y = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.major.x = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.spacing.x = unit(2, "lines")) +
  scale_x_continuous(expand = c(0,0), limits = c(0, 6000), breaks = seq(0, 6000, 1000)) +
  labs(x = "Total",
       y = "",
       fill = "Direction",
       title = "") +
  facet_wrap(~title)


ggsave(ggplot_degs_bars_vertical_nolabels, file = here("Figures", "DEGs_up_down_barplot_vertical_nolabels.png"),  width = 10, height = 10)

ggplot_degs_bars_vertical_nolabels

all_degs

```

No labels


Vertical, facetted
```{r fig.height=10, fig.width=10}

ggplot_degs_bars_directions = all_degs %>%
      mutate(condition = fct_relevel(condition, c(ann_vaccines_conditions_ordered$condition_grouped %>% as.character() %>% unique()))) %>% 
  mutate(direction = case_when(direction == "UP" ~ "Up-regulated",
                               direction == "DOWN" ~ "Down-regulated",
                               TRUE ~ "Neutral")) %>% 
 ggplot() +
  aes(y = fct_rev(condition), fill = type, x = n_degs) +
  geom_col(position = "dodge",
           color = "black",
           size = 0.2) +
  scale_fill_manual(values = ann_vaxsig_colors$type)+
  # scale_fill_manual(
  #   values = c("Down-regulated" = "#4DBBD5FF",
  #   "Neutral" = "gray80",
  #   "Up-regulated" = "#d90429")) +
  theme_minimal() +
  geom_text(aes(label = n_degs, x = n_degs), 
            position = position_dodge(width = 0.9), 
            angle = 0,
            vjust = 0.5, 
            hjust = -0.2, 
            size = 3) +
  scale_x_continuous(expand = c(0,0), limits = c(0, 8000))+
  theme(legend.position = "right",
        strip.text.x = element_text(color = "black", size = 12), 
        axis.text.x = element_text(color = "black", size = 10),
        axis.text.y = element_text(vjust = 0.5, size = 10, color = "black"),
        axis.line = element_line(size = 1, colour = "black", linetype=1),
        axis.ticks.y = element_line(size = 0.5, color = "black"),
        axis.ticks.x = element_line(size = 0, color = "black"),
        panel.grid.major.y = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.major.x = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.spacing.x = unit(2, "lines")) +
  labs(x = "Total",
       y = "",
       fill = "Type",
       title = "Total DEGs, by direction. Threshold = |l2fc| > 1") +
  facet_wrap(~direction)

ggplot_degs_bars_directions %>%
  ggsave(file = here("Figures", "DEGs_L2FC_directions_vertical.png"), width = 10, height = 10)

ggplot_degs_bars_directions
```

Vertical (Up and down)
```{r fig.height=10, fig.width=10}

ggplot_degs_bars_directions_up_down = all_degs %>%
      mutate(condition = fct_relevel(condition, c(ann_vaccines_conditions_ordered$condition_grouped %>% as.character() %>% unique()))) %>% 
  mutate(direction = case_when(direction == "UP" ~ "Up-regulated",
                               direction == "DOWN" ~ "Down-regulated",
                               TRUE ~ "Neutral")) %>% 
  filter(direction != "Neutral") %>% 
 ggplot() +
  aes(y = fct_rev(condition), fill = type, x = n_degs) +
  geom_col(position = "dodge", size = 0.2, color = "black") +
  scale_fill_manual(values = ann_vaxsig_colors$type)+
  # scale_fill_manual(
  #   values = c("Down-regulated" = "#4DBBD5FF",
  #   "Neutral" = "gray80",
  #   "Up-regulated" = "#d90429")) +
  theme_minimal() +
  geom_text(aes(label = n_degs, x = n_degs), 
            position = position_dodge(width = 0.9), 
            angle = 0,
            vjust = 0.5, 
            hjust = -0.2, 
            size = 3) +
  scale_x_continuous(expand = c(0,0), limits = c(0, 3500))+
  theme(legend.position = "right",
        strip.text.x = element_text(color = "black", size = 12), 
        axis.text.x = element_text(color = "black", size = 10),
        axis.text.y = element_text(vjust = 0.5, size = 10, color = "black"),
        axis.line = element_line(size = 1, colour = "black", linetype=1),
        axis.ticks = element_line(size = 0.5, color = "black"),
        panel.grid.minor.x = element_blank(),
        panel.spacing.x = unit(2, "lines")) +
  labs(x = "Total",
       y = "",
       fill = "Direction",
       title = "Total DEGs, by direction. Threshold = |l2fc| > 1") +
  facet_wrap(~direction)

# ggplot_degs_bars_directions_up_down %>% 
  # ggsave(file = here("Figures", "DEGs_L2FC_directions_up_down_vertical.png"), width = 10, height = 10)

ggplot_degs_bars_directions_up_down
```


Stacked
```{r fig.height=10, fig.width=10}

all_degs %>%
      mutate(condition = fct_relevel(condition, c(ann_vaccines_conditions_ordered$condition_grouped %>% as.character() %>% unique()))) %>% 
  mutate(direction = case_when(direction == "UP" ~ "Up-regulated",
                               direction == "DOWN" ~ "Down-regulated",
                               TRUE ~ "Neutral"),
         direction = fct_relevel(direction, "Down-regulated", "Up-regulated")) %>% 
  ggplot() +
  aes(y = fct_rev(condition), fill = fct_rev(direction), x = n_degs) +
  geom_bar(stat = "summary", fun = "sum",
           color = "black",
           size = 0.2) +
  scale_fill_manual(
    values = c("Down-regulated" = "#4DBBD5FF",
    "Neutral" = "gray90",
    "Up-regulated" = "#d90429")) +
  theme_minimal() +
  # geom_text(aes(label = n_degs, x = n_degs), 
  #           position = position_dodge(width = 0.9), 
  #           angle = 0,
  #           vjust = 0.5, 
  #           hjust = -0.2, 
  #           size = 3) +
  scale_x_continuous(expand = c(0,0), limits = c(0, 8000))+
  theme(legend.position = "right",
        strip.text.x = element_text(color = "black", size = 12), 
        axis.text.x = element_text(color = "black", size = 10),
        axis.text.y = element_text(vjust = 0.5, size = 10, color = "black"),
        axis.line.y = element_line(size = 1, colour = "black", linetype=1),
        axis.ticks.y = element_line(size = 0.5, color = "black"),
        axis.ticks.x = element_line(size = 0, color = "black"),
        panel.grid.major.y = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.major.x = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.spacing.x = unit(2, "lines")) +
  labs(x = "Total",
       y = "",
       fill = "Direction",
       title = "Total DEGs, by direction. Threshold = |l2fc| > 1")
  
```



## Gene annotations
### By biotype

```{r fig.height=10, fig.width=10}
# library("biomaRt")
# mart <- useMart("ensembl")
# mart <- useDataset("hsapiens_gene_ensembl", mart)
# gene_info <- getBM(attributes = c("entrezgene_id", "hgnc_symbol", "ensembl_gene_id", "gene_biotype", "start_position", "end_position"),
#                    mart = mart)
# 
# gene_info %>% 
#   saveRDS(file = here("DGE analysis", "Gene_info_biomart.rds"))

gene_info = read_rds(file = here("DGE analysis", "Gene_info_biomart.rds"))

all_degs_totals = all_degs_p_05_vac_infected %>%
  group_by(condition) %>% 
  summarise(n_total = n())
  
all_degs_biotypes = all_degs_p_05_vac_infected %>% 
  inner_join(all_degs_totals, by = "condition") %>% 
  inner_join(gene_info %>% select(genes = hgnc_symbol, gene_biotype), by = "genes") %>% 
  mutate(gene_biotype_grouped = case_when(str_detect(gene_biotype, "IG_") ~ "Protein coding",
                                          str_detect(gene_biotype, "TR_") ~ "Protein coding",
                                          str_detect(gene_biotype, "pseudogene") ~ "Pseudogene",
                                          str_detect(gene_biotype, "protein_coding") ~ "Protein coding",
                                          str_detect(gene_biotype, "Mt") ~ "Mithochondrial RNA",
                                          str_detect(gene_biotype, "ribozyme") ~ "Ribozyme",
                                          str_detect(gene_biotype, "RNA") ~ "Non-coding RNAs",
                                          str_detect(gene_biotype, "TEC") ~ "To be Experimentally Confirmed",
                                          TRUE ~ gene_biotype)) %>% 
  group_by(condition, gene_biotype_grouped) %>% 
  distinct() %>% 
  summarise(n_biotypes = n(),
            n_total = n_total) %>%
  distinct() %>% 
  mutate(n_perc = 100*(n_biotypes/n_total)) %>% 
  inner_join(ann_vaccines, by="condition") %>% 
  ungroup()

all_degs_biotypes_plot_totals = all_degs_biotypes %>% 
  mutate(gene_biotype_grouped = fct_relevel(gene_biotype_grouped, 
                                            "Protein coding", 
                                            "Non-coding RNAs",
                                            "Ribozyme",
                                            "Mithochondrial RNA",
                                            "Pseudogene",
                                            "To be Experimentally Confirmed"
                                            ),
         title = "Total DEGs by biotype",
         condition = fct_relevel(condition, c(ann_vaccines_conditions_ordered$condition_grouped %>% 
                                                as.character() %>% 
                                                unique()))) %>% 
  ggplot(.) +
  aes(x = n_biotypes, y = fct_rev(condition), fill = gene_biotype_grouped) +
  geom_bar(stat = "summary", fun = "sum",
           color ="black",
           size = 0.5,
           width = 0.6) +
  theme_minimal() +
  scale_fill_manual(values = c("Pseudogene" = "#457b9d",
                               "Protein coding" = "gray95",
                               "Non-coding RNAs" ="#EFC000FF")) +
  theme(legend.position = "right",
        strip.text.x = element_text(color = "black", size = 12), 
        axis.text.x = element_text(color = "black", size = 10),
        axis.text.y = element_text(vjust = 0.5, size = 10, color = "black"),
        axis.line = element_line(size = 1, colour = "black", linetype=1),
        axis.ticks = element_line(size = 0.5, color = "black"),
        panel.grid.major.y = element_line(size = 0, 
                                    linetype = 2,
                                    color = "white"),
        panel.grid.major.x = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75")) +
  labs(x = "Total",
       y = "",
       fill = "Gene biotype",
       title = "DEGs by biotype") +
  scale_x_continuous(expand = c(0,0)) +
    facet_wrap(~title, scales= "free", ncol = 1)

all_degs_biotypes_plot_perc = all_degs_biotypes %>% 
  mutate(gene_biotype_grouped = fct_relevel(gene_biotype_grouped, 
                                            "Protein coding", 
                                            "Non-coding RNAs",
                                            "Ribozyme",
                                            "Mithochondrial RNA",
                                            "Pseudogene",
                                            "To be Experimentally Confirmed"
                                            ),
         title = "%DEGs by biotype,\nnon-protein-coding genes",
         condition = fct_relevel(condition, c(ann_vaccines_conditions_ordered$condition_grouped %>% 
                                                as.character() %>% 
                                                unique()))) %>%
  ggplot(.) +
  aes(x = n_perc, y = fct_rev(condition), fill = gene_biotype_grouped) +
  geom_bar(stat = "summary", 
           fun = "sum", 
           # position = "fill",
           color = "black",
           size = 0.5,
           width = 0.6) +
  theme_minimal() +
  scale_fill_manual(values = c("Pseudogene" = "#457b9d",
                               "Protein coding" = "gray95",
                               "Non-coding RNAs" ="#EFC000FF")) +
  theme(legend.position = "right",
        strip.text.x = element_text(color = "black", size = 12), 
        axis.text.x = element_text(color = "black", size = 10),
        axis.text.y = element_text(vjust = 0.5, size = 10, color = "black"),
        axis.line = element_line(size = 1, colour = "black", linetype=1),
        axis.ticks = element_line(size = 0.5, color = "black"),
        panel.grid.major.y = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.major.x = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75")) + 
  labs(x = "%",
       y = "",
       fill = "Gene biotype",
       title = "Percentage of DEGs by biotype") +
  # xlim(0, 7) +
  scale_x_continuous(expand = c(0,0), limits = c(0, 7)) +
  facet_wrap(~title, scales= "free", ncol = 1)

all_degs_biotypes_plot_all = all_degs_biotypes_plot_totals + all_degs_biotypes_plot_perc + 
  plot_layout(axes = "collect", guides = "collect")

all_degs_biotypes_plot_all %>%
  ggsave(file = here("Figures", "DEGs_by_biotype.png"),
         width = 10,
         height = 10)

all_degs_biotypes_plot_all

```



Most of immune genes in ImmuneGO are protein-coding.
```{r fig.height=10, fig.width=10}

immunego_btm_manual_grouped_genes = read_csv(file = here("VaxGO gene sets", "ImmuneGO_BTM_grouped_genes.csv"))
immunego_btm_genes = immunego_btm_manual_grouped_genes %>% 
  select(genes) %>% distinct()

#ImmuneGO

all_degs_biotypes_immune = all_degs_p_05_vac_infected %>% 
  inner_join(all_degs_totals, by = "condition") %>% 
  filter(genes %in% immunego_btm_genes$genes) %>% 
  inner_join(gene_info %>% select(genes = hgnc_symbol, gene_biotype), by = "genes") %>% 
  mutate(gene_biotype_grouped = case_when(str_detect(gene_biotype, "IG_") ~ "Protein coding",
                                          str_detect(gene_biotype, "TR_") ~ "Protein coding",
                                          str_detect(gene_biotype, "pseudogene") ~ "Pseudogene",
                                          str_detect(gene_biotype, "protein_coding") ~ "Protein coding",
                                          str_detect(gene_biotype, "Mt") ~ "Mithochondrial RNA",
                                          str_detect(gene_biotype, "ribozyme") ~ "Ribozyme",
                                          str_detect(gene_biotype, "RNA") ~ "Non-coding RNAs",
                                          str_detect(gene_biotype, "TEC") ~ "To be Experimentally Confirmed",
                                          TRUE ~ gene_biotype)) %>% 
  group_by(condition, gene_biotype_grouped) %>% 
  summarise(n_biotypes = n(),
            n_total = n_total) %>%
  distinct() %>% 
  mutate(n_perc = 100*(n_biotypes/n_total)) %>% 
  inner_join(ann_vaccines, by="condition") %>% 
  ungroup()


all_degs_biotypes_plot_totals_immune = all_degs_biotypes_immune %>% 
  mutate(gene_biotype_grouped = fct_relevel(gene_biotype_grouped, 
                                            "Protein coding", 
                                            "Non-coding RNAs",
                                            "Ribozyme",
                                            "Mithochondrial RNA",
                                            "Pseudogene",
                                            "To be Experimentally Confirmed"
                                            ),
         title = "Total Immune DEGs by biotype",
         condition = fct_relevel(condition, c(ann_vaccines_conditions_ordered$condition_grouped %>% 
                                                as.character() %>% 
                                                unique()))) %>% 
  ggplot(.) +
  aes(x = n_biotypes, y = fct_rev(condition), fill = gene_biotype_grouped) +
  geom_bar(stat = "summary", fun = "sum") +
  theme_minimal() +
    scale_fill_manual(values = c("Pseudogene" = "#F39B7FFF",
                               "Protein coding" = "#118ab2",
                               "Non-coding RNAs" ="#EFC000FF")) +
  theme(legend.position = "right",
        strip.text.x = element_text(color = "black", size = 12), 
        axis.text.x = element_text(color = "black", size = 10),
        axis.text.y = element_text(vjust = 0.5, size = 10, color = "black"),
        axis.line = element_line(size = 1, colour = "black", linetype=1),
        axis.ticks = element_line(size = 0.5, color = "black")) +
  labs(x = "Total",
       y = "",
       fill = "Gene biotype",
       title = "DEGs by biotype") +
  scale_x_continuous(expand = c(0,0)) +
    facet_wrap(~title, scales= "free", ncol = 1)

all_degs_biotypes_plot_perc_immune = all_degs_biotypes_immune %>% 
  mutate(gene_biotype_grouped = fct_relevel(gene_biotype_grouped, 
                                            "Protein coding", 
                                            "Non-coding RNAs",
                                            "Ribozyme",
                                            "Mithochondrial RNA",
                                            "Pseudogene",
                                            "To be Experimentally Confirmed"
                                            ),
         title = "%Immune DEGs by biotype",
         condition = fct_relevel(condition, c(ann_vaccines_conditions_ordered$condition_grouped %>% 
                                                as.character() %>% 
                                                unique()))) %>% 
  ggplot(.) +
  aes(x = n_biotypes, y = fct_rev(condition), fill = gene_biotype_grouped) +
  geom_bar(stat = "summary", fun = "sum", position = "fill") +
  theme_minimal() +
    scale_fill_manual(values = c("Pseudogene" = "#F39B7FFF",
                               "Protein coding" = "#118ab2",
                               "Non-coding RNAs" ="#EFC000FF")) +
  theme(legend.position = "right",
        strip.text.x = element_text(color = "black", size = 12), 
        axis.text.x = element_text(color = "black", size = 10),
        axis.text.y = element_text(vjust = 0.5, size = 10, color = "black"),
        axis.line = element_line(size = 1, colour = "black", linetype=1),
        axis.ticks = element_line(size = 0.5, color = "black")) +
  labs(x = "%",
       y = "",
       fill = "Gene biotype",
       title = "Percentage of DEGs by biotype") +
  scale_x_continuous(expand = c(0,0)) +
  facet_wrap(~title, scales= "free", ncol = 1)


all_degs_biotypes_plot_immune = all_degs_biotypes_plot_perc_immune + all_degs_biotypes_plot_totals_immune + 
   plot_layout(axes = "collect", guides = "collect")

all_degs_biotypes_plot_immune %>%
  ggsave(file = here("Figures", "all_degs_biotypes_plot_immune.png"), width = 10, height = 10)

all_degs_biotypes_plot_immune
```

How many are in?

```{r}

btm_annotation_genes = readRDS(here("VaxGO gene sets", "btm_annotation_genes.rds"))

immunego_btm_genes = ImmuneGO_Annotated_Genes %>% 
              select(genes) %>% 
              distinct() %>% 
              mutate(set = "ImmuneGO") %>% 
  bind_rows(btm_annotation_genes %>% 
              select(genes) %>% 
              distinct() %>% 
              mutate(set = "BTM")) %>% 
  arrange(genes)

immunego_btm_genes_joined = immunego_btm_genes %>% 
  bind_rows(immunego_btm_genes %>% mutate(set = "BTM and ImmuneGO")) %>% 
  distinct()


all_degs_immuneGO_BTM_totals = all_degs_p_05_vac_infected %>% 
  select(genes, condition) %>% 
  left_join(immunego_btm_genes_joined,
            by = "genes") %>% 
  mutate(set = if_else(is.na(set), "Other", set)) %>% 
  group_by(condition, set) %>% 
  summarise(n = n()) %>%
  ungroup() %>% 
  group_by(condition) %>% 
  summarize(ntotal = sum(n),
            n = n,
            set = set) %>% 
  distinct() %>% 
    ungroup() %>% 
  mutate(n_perc = round(100*(n/ntotal), 1)) %>% 
  inner_join(ann_vaccines, by="condition")

all_degs_immuneGO_BTM_reg = all_degs_p_05_vac_infected %>% 
  select(genes, direction, condition) %>% 
  left_join(immunego_btm_genes_joined,
            by = "genes") %>% 
  mutate(set = if_else(is.na(set), "Other", set)) %>% 
  group_by(condition, direction, set) %>% 
  summarise(n = n()) %>%
  ungroup() %>% 
  group_by(condition) %>% 
  summarize(ntotal = sum(n),
            direction = direction,
            n = n,
            set = set) %>% 
  distinct() %>% 
    ungroup() %>% 
  inner_join(ann_vaccines, by="condition")


all_degs_immuneGO_BTM_totals %>% 
  group_by(set) %>% 
  summarize(n_median = median(n_perc))
```


```{r fig.height=10, fig.width=5}
#Percentage total DEGs
all_degs_immuneGO_BTM_plot = all_degs_immuneGO_BTM_totals %>% 
  filter(set != "Other") %>% 
  mutate(condition = fct_relevel(condition, c(ann_vaccines_conditions_ordered$condition_grouped %>% 
                                                as.character() %>% 
                                                unique())),
         title = "Immune sets") %>% 
  ggplot(.) +
  aes(x = n_perc, y = fct_rev(condition), fill = fct_relevel(set, "Other", "BTM and ImmuneGO", "BTM", "ImmuneGO")) +
  geom_bar(stat = "summary", fun = "sum", position = "fill",
           color ="black",
           size = 0.2) +
  scale_fill_manual(values = c("BTM and ImmuneGO" = "#4DBBD5FF",
    "Other" = "gray95",
    "BTM" = "#5e60ce",
    "ImmuneGO" = "#00A087FF")) +
    theme_minimal() +
  theme(legend.position = "right",
        strip.text.x = element_text(color = "black", size = 12), 
        axis.text.x = element_text(color = "black", size = 10),
        axis.text.y = element_text(vjust = 0.5, size = 10, color = "black"),
        axis.line = element_line(size = 1, colour = "black", linetype=1),
        axis.ticks = element_line(size = 0.5, color = "black")) +
  labs(x = "%Coverage",
       y = "",
       fill = "",
       title = "") +
  scale_x_continuous(expand = c(0,0)) +
  geom_text(aes(label = paste0(n_perc)), 
            position = position_fill(vjust = 0.5), 
            size = 3, 
            color = "white") +
  facet_wrap(~title)

all_degs_immuneGO_BTM_plot %>% 
  ggsave(file = here("Figures", "all_degs_immuneGO_BTM_percentage_plot.png"),
         width = 5,
         height = 10)
  
all_degs_immuneGO_BTM_plot
```

Percentage, total, labelled
```{r fig.height=10, fig.width=10}
#Percentage total DEGs only united
all_degs_immuneGO_BTM_united_plot = all_degs_immuneGO_BTM_totals %>% 
  filter(set %in% c("BTM and ImmuneGO", "Other")) %>% 
  mutate(condition = fct_relevel(condition, c(ann_vaccines_conditions_ordered$condition_grouped %>% 
                                                as.character() %>% 
                                                unique())),
         title = "BTM + ImmuneGO sets") %>% 
  ggplot(.) +
  aes(x = n_perc, y = fct_rev(condition), fill = fct_relevel(set, "Other", "BTM and ImmuneGO", "BTM", "ImmuneGO")) +
  geom_bar(stat = "summary", fun = "sum", position = "fill",
           color = "black",
           size = 0.2) +
  scale_fill_manual(values = c("BTM and ImmuneGO" = "#4DBBD5FF",
    "Other" = "gray95",
    "BTM" = "#5e60ce",
    "ImmuneGO" = "#00A087FF")) +
    theme_minimal() +
  theme(legend.position = "right",
        strip.text.x = element_text(color = "black", size = 12), 
        axis.text.x = element_text(color = "black", size = 10),
        axis.text.y = element_text(vjust = 0.5, size = 10, color = "black"),
        axis.line = element_line(size = 1, colour = "black", linetype=1),
        axis.ticks = element_line(size = 0.5, color = "black")) +
  labs(x = "%Coverage",
       y = "",
       fill = "",
       title = "") +
  scale_x_continuous(expand = c(0,0)) +
  geom_text(aes(label = paste0(n_perc)), 
            position = position_fill(vjust = 0.5), 
            size = 3, 
            color = "black") +
  facet_wrap(~title)

all_degs_immuneGO_BTM_united_plot %>% 
  ggsave(file = here("Figures", "all_degs_immuneGO_BTM_united_plot.png"),
         width = 10,
         height = 10)
  
all_degs_immuneGO_BTM_united_plot
```


Unlabelled
```{r}
#Percentage total DEGs only united
all_degs_immuneGO_BTM_united_plot_nolabel = all_degs_immuneGO_BTM_totals %>% 
  filter(set %in% c("BTM and ImmuneGO")) %>% 
  mutate(condition = fct_relevel(condition, c(ann_vaccines_conditions_ordered$condition_grouped %>% 
                                                as.character() %>% 
                                                unique())),
         title = "Immune system") %>% 
  ggplot(.) +
  aes(x = n_perc, y = fct_rev(condition), fill = type) +
  geom_col(color = "black",
           size = 0.5,
           width = 0.6) +
  scale_fill_manual(values = ann_vaxsig_colors$type) +
    theme_minimal() +
  theme(legend.position = "none",
        strip.text.x = element_text(color = "black", size = 12), 
        axis.text.x = element_text(color = "black", size = 10),
        axis.text.y = element_text(vjust = 0.5, size = 10, color = "black"),
        axis.line.x = element_line(size = 1, colour = "black", linetype=1),
        axis.line.y = element_line(size = 1, colour = "black", linetype=1),
        axis.ticks.y = element_line(size = 0.5, color = "black"),
        axis.ticks.x = element_line(size = 0, color = "black"),
        panel.grid.major.y = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.major.x = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.spacing.x = unit(2, "lines")) +
  labs(x = "%Coverage",
       y = "",
       fill = "Immune system",
       title = "") +
  scale_x_continuous(expand = c(0,0), limits = c(0,15)) +
  facet_wrap(~title)

all_degs_immuneGO_BTM_united_plot_nolabel %>% 
  ggsave(file = here("Figures", "all_degs_immuneGO_BTM_united_plot_nolabel.png"),
         width = 10,
         height = 10)
  
all_degs_immuneGO_BTM_united_plot_nolabel
```



Totals by direction, facetted
```{r fig.height=10, fig.width=10}
#Totals by direction
all_degs_immuneGO_BTM_reg_plot = all_degs_immuneGO_BTM_reg %>% 
  group_by(condition, direction) %>% 
  summarize(n_total_direction = sum(n),
            n = n,
            set = set
            ) %>% 
  ungroup() %>% 
  mutate(condition = fct_relevel(condition, c(ann_vaccines_conditions_ordered$condition_grouped %>% 
                                                as.character() %>% 
                                                unique()))) %>% 
  mutate(direction = case_when(direction == "UP" ~ "Up-regulated", 
                               direction == "DOWN" ~ "Down-regulated", 
                               TRUE ~ "Neutral"
                             ),
         ntotalplus = n_total_direction + n) %>% 
  filter(set %in% c("BTM and ImmuneGO", "Other"),
         direction != "Neutral") %>% 
  ggplot(.) +
  aes(x = n, y = fct_rev(condition), fill = fct_relevel(set, "Other", "BTM and ImmuneGO", "BTM", "ImmuneGO")) +
  geom_bar(stat = "summary", fun = "sum",
           size = 0.5,
           width = 0.6,
           color ="black") +  
    scale_fill_manual(values = c("BTM and ImmuneGO" = "#4DBBD5FF",
    "Other" = "gray95",
    "BTM" = "#5e60ce",
    "ImmuneGO" = "#00A087FF")) +
    theme_minimal() +
  theme(legend.position = "right",
        strip.text.x = element_text(color = "black", size = 12), 
        axis.text.x = element_text(color = "black", size = 10),
        axis.text.y = element_text(vjust = 0.5, size = 10, color = "black"),
        axis.line = element_line(size = 1, colour = "black", linetype=1),
        axis.ticks = element_line(size = 0.5, color = "black"),
        panel.grid.major.y = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.major.x = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.spacing.x = unit(2, "lines")) +
  labs(x = "Total",
       y = "",
       fill = "",
       title = "DEGs by Immune role") +
  scale_x_continuous(expand = c(0,0), limits = c(0, 4000)) +
  facet_grid(~direction, scales="free")  +
  geom_text(aes(label = if_else(set == "Other", as.character(n_total_direction), ""),
                x = n_total_direction),
            hjust = -0.2, 
            size = 3,
            color = "black")

all_degs_immuneGO_BTM_reg_plot %>% 
  ggsave(file = here("Figures", "all_degs_immuneGO_BTM_totals_direction_plot.png"),
         width = 10,
         height = 10)

all_degs_immuneGO_BTM_reg_plot
```

Totals by direction, dodged
```{r fig.height=10, fig.width=10}
#Totals by direction

all_degs_immuneGO_BTM_reg_plot_dodged_table = all_degs_immuneGO_BTM_reg %>% 
  group_by(condition, direction) %>% 
  summarize(n_total_direction = sum(n),
            n = n,
            set = set
            ) %>% 
  ungroup() %>% 
  mutate(condition = fct_relevel(condition, c(ann_vaccines_conditions_ordered$condition_grouped %>% 
                                                as.character() %>% 
                                                unique()))) %>% 
  mutate(direction = case_when(direction == "UP" ~ "Up-regulated", 
                               direction == "DOWN" ~ "Down-regulated", 
                               TRUE ~ "Neutral"
                             )) %>% 
  filter(set %in% c("BTM and ImmuneGO", "Other"),
         direction != "Neutral") %>% 
  # Agrupar por 'condition' e 'direction'
  group_by(condition, direction) %>%
  # Calcular o total acumulado para cada grupo
  mutate(cum_n = cumsum(n)) 



all_degs_immuneGO_BTM_reg_plot_dodged = all_degs_immuneGO_BTM_reg_plot_dodged_table %>%
  ggplot(aes(x = cum_n , 
             y = fct_rev(condition), 
             fill = fct_rev(direction)
             )) +
  geom_col(data = . %>% filter(set == "BTM and ImmuneGO"), 
           position = position_dodge(width = 1), alpha = 1, 
           width = 0.6,
           color = "black") +
  geom_col(data = . %>% filter(set == "Other"),
           width = 0.6,
           position = position_dodge(width = 1), alpha = 0.2,
           color = "black") +
  geom_tile(aes(y=NA_integer_, alpha = factor(set), fill = set),) + 
  scale_alpha_manual(name = "Immune", values = c(1, 0.2),
                     guide = guide_legend(override.aes = list(fill = "gray50", color = "black", size = 1) )) + 
    scale_fill_manual(
    values = c("Down-regulated" = "#4DBBD5FF",
    "Neutral" = "gray90",
    "Up-regulated" = "#d90429")) +
  theme_minimal() +
  theme(legend.position = "right",
        strip.text.x = element_text(color = "black", size = 12), 
        axis.text.x = element_text(color = "black", size = 10),
        axis.text.y = element_text(vjust = 0.5, size = 10, color = "black"),
        axis.line = element_line(size = 1, colour = "black", linetype=1),
        axis.ticks = element_line(size = 0.5, color = "black"),
        panel.grid.major.y = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.major.x = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.spacing.x = unit(2, "lines")) +
  labs(x = "Total",
       y = "",
       fill = "Direction",
       title = "DEGs by Immune role") +
  scale_x_continuous(expand = c(0, 0), limits = c(0, 3000), breaks = seq(0, 3000, 500))

  # geom_text(aes(label = if_else(set == "Other", as.character(n_total_direction), ""),
  #               x = n_total_direction),
  #           hjust = -0.2, 
  #           size = 3,
  #           color = "black")

all_degs_immuneGO_BTM_reg_plot_dodged %>% 
  ggsave(file = here("Figures", "all_degs_immuneGO_BTM_totals_direction_plot_dodged.png"),
         width = 10,
         height = 10)

all_degs_immuneGO_BTM_reg_plot_dodged
```

Unite plots
```{r fig.height=10, fig.width=15}
all_degs_immuneGO_BTM_plot + all_degs_immuneGO_BTM_united_plot + all_degs_immuneGO_BTM_reg_plot_dodged +
  plot_layout(guides = "collect", axes = "collect", widths = c(1, 1, 1))
```


```{r}

#ImmuneGO
all_degs_biotypes_immuneBTM = all_degs_p_05_vac_infected %>% 
  inner_join(all_degs_totals, by = "condition") %>% 
  filter(genes %in%  btm_annotation_genes$genes) %>% 
  inner_join(gene_info %>% select(genes = hgnc_symbol, gene_biotype), by = "genes") %>% 
  mutate(gene_biotype_grouped = case_when(str_detect(gene_biotype, "IG_") ~ "Protein coding",
                                          str_detect(gene_biotype, "TR_") ~ "Protein coding",
                                          str_detect(gene_biotype, "pseudogene") ~ "Pseudogene",
                                          str_detect(gene_biotype, "protein_coding") ~ "Protein coding",
                                          str_detect(gene_biotype, "Mt") ~ "Mithochondrial RNA",
                                          str_detect(gene_biotype, "ribozyme") ~ "Ribozyme",
                                          str_detect(gene_biotype, "RNA") ~ "Non-coding RNAs",
                                          str_detect(gene_biotype, "TEC") ~ "To be Experimentally Confirmed",
                                          TRUE ~ gene_biotype)) %>% 
  group_by(condition, gene_biotype_grouped) %>% 
  summarise(n_biotypes = n(),
            n_total = n_total) %>%
  distinct() %>% 
  mutate(n_perc = 100*(n_biotypes/n_total)) %>% 
  inner_join(ann_vaccines, by="condition") %>% 
  ungroup()


all_degs_biotypes_plot_totals_immuneBTM = all_degs_biotypes_immuneBTM %>% 
  mutate(gene_biotype_grouped = fct_relevel(gene_biotype_grouped, 
                                            "Protein coding", 
                                            "Non-coding RNAs",
                                            "Ribozyme",
                                            "Mithochondrial RNA",
                                            "Pseudogene",
                                            "To be Experimentally Confirmed"
                                            ),
         title = "Total Immune (ImmuneGO) DEGs by biotype",
         condition = fct_relevel(condition, c(ann_vaccines_conditions_ordered$condition_grouped %>% 
                                                as.character() %>% 
                                                unique()))) %>% 
  ggplot(.) +
  aes(x = n_biotypes, y = fct_rev(condition), fill = gene_biotype_grouped) +
  geom_bar(stat = "summary", fun = "sum") +
  theme_minimal() +
  scale_fill_simpsons() +
  theme(legend.position = "right",
        strip.text.x = element_text(color = "black", size = 12), 
        axis.text.x = element_text(color = "black", size = 10),
        axis.text.y = element_text(vjust = 0.5, size = 10, color = "black"),
        axis.line = element_line(size = 1, colour = "black", linetype=1),
        axis.ticks = element_line(size = 0.5, color = "black")) +
  labs(x = "Total",
       y = "",
       fill = "Gene biotype",
       title = "DEGs by biotype") +
  scale_x_continuous(expand = c(0,0)) +
    facet_wrap(~title, scales= "free", ncol = 1)

all_degs_biotypes_plot_perc_immuneBTM = all_degs_biotypes_immuneBTM %>% 
  mutate(gene_biotype_grouped = fct_relevel(gene_biotype_grouped, 
                                            "Protein coding", 
                                            "Non-coding RNAs",
                                            "Ribozyme",
                                            "Mithochondrial RNA",
                                            "Pseudogene",
                                            "To be Experimentally Confirmed"
                                            ),
         title = "%Immune DEGs by biotype",
         condition = fct_relevel(condition, c(ann_vaccines_conditions_ordered$condition_grouped %>% 
                                                as.character() %>% 
                                                unique()))) %>% 
  ggplot(.) +
  aes(x = n_biotypes, y = fct_rev(condition), fill = gene_biotype_grouped) +
  geom_bar(stat = "summary", fun = "sum", position = "fill") +
  theme_minimal() +
  scale_fill_simpsons() +
  theme(legend.position = "right",
        strip.text.x = element_text(color = "black", size = 12), 
        axis.text.x = element_text(color = "black", size = 10),
        axis.text.y = element_text(vjust = 0.5, size = 10, color = "black"),
        axis.line = element_line(size = 1, colour = "black", linetype=1),
        axis.ticks = element_line(size = 0.5, color = "black")) +
  labs(x = "%",
       y = "",
       fill = "Gene biotype",
       title = "Percentage of Immune DEGs (BTM) by biotype") +
  scale_x_continuous(expand = c(0,0)) +
  facet_wrap(~title, scales= "free", ncol = 1)


all_degs_biotypes_plot_immuneBTM = all_degs_biotypes_plot_perc_immuneBTM + all_degs_biotypes_plot_totals_immuneBTM + 
   plot_layout(axes = "collect", guides = "collect")

# all_degs_biotypes_plot_immuneBTM %>%
#   ggsave(file = here("Figures", "all_degs_biotypes_plot_immune_BTM.png"), width = 10, height = 10)

all_degs_biotypes_plot_immuneBTM

 
```


## Gene ontology

Retrieving genes by Gene Ontology

```{r}
organism = "org.Hs.eg.db"
conflicted::conflicts_prefer(dplyr::summarize)
# install.packages("GO.db")
# BiocManager::install(organism, character.only = TRUE)
library(organism, character.only = TRUE)
library(clusterProfiler)
library(GO.db)

#HGNC symbol to Entrez
x <- org.Hs.egSYMBOL
mapped_genes <- mappedkeys(x)
symbols_entrez <- as.data.frame(x[mapped_genes])

## Entrez to GO
x <- org.Hs.egGO
mapped_genes <- mappedkeys(x)
entrez_geneontology <- as.data.frame(x[mapped_genes])

## GO BP ancestors
go_bp_ancestor = GOBPANCESTOR %>% 
  as.data.frame()
colnames(go_bp_ancestor)[1] <- "go_id"
colnames(go_bp_ancestor)[2] <- "go_ancestor"

## GO ID to terms
go_bp = AnnotationDbi::select(GO.db, columns=c("GOID","TERM"), keys= "BP", keytype= "ONTOLOGY") %>% 
  clean_names() %>% 
  select(go_id = goid, term)

## GO child and ancestor
go_bp_ancestor = go_bp_ancestor %>% 
  inner_join(go_bp %>% rename(go_ancestor = go_id), by = "go_ancestor") %>% 
  rename(term_ancestor = term)

#GO ids to terms
go_bp = AnnotationDbi::select(GO.db, columns=c("GOID","TERM"), keys="BP", keytype="ONTOLOGY") %>% 
  clean_names() %>% 
  rename(go_id = goid) %>% 
  select(-ontology) %>% 
  right_join(entrez_geneontology, by = "go_id") %>% 
  right_join(symbols_entrez, by = "gene_id") %>% 
  clean_names() %>% 
  left_join(go_bp_ancestor, by = "go_id") %>% 
  distinct()

#Filter main GO BP terms
go_bp_main = go_bp %>%
  filter(go_ancestor %in% c(
      "GO:0040011",
      "GO:0043473",
      "GO:0044419",
      "GO:0009987",
      "GO:0048518",
      "GO:0050789",
      "GO:0065007",
      "GO:0048511",
      "GO:0022414",
      "GO:0002376",
      "GO:0008152")) %>% 
  filter(str_detect(term_ancestor, "cellular process|biological process involved in interspecies interaction between organisms|regulation of biological process|immune system process|No annotation|Not mapped|metabolic process"),
         !term_ancestor == "positive regulation of biological process") %>% 
  mutate(term_ancestor = case_when(
           term_ancestor == "biological process involved in interspecies interaction between organisms" ~ "Interspecies interaction",
           term_ancestor == "biological regulation" ~ "Biological reg.",
           term_ancestor == "cellular process" ~ "Cellular",
           term_ancestor == "immune system process" ~ "Immune system",
           term_ancestor == "regulation of biological process" ~ "Reg. of BP",
           term_ancestor == "locomotion" ~ "Locomotion",
           term_ancestor == "metabolic process" ~ "Metabolism",
           TRUE ~ term_ancestor))

#Add set size column
go_bp_main_2 = go_bp_main %>% 
  group_by(term_ancestor) %>% 
  summarise(set_size = n()) %>% 
  inner_join(go_bp_main) %>% 
  select(term_ancestor, set_size, go_ancestor, symbol) %>% 
  distinct() %>% 
  mutate(annotation = "Major process")

go_bp_child = go_bp_main %>% 
  select(term_ancestor, go_ancestor, symbol) %>% 
  distinct() %>% 
  group_by(term_ancestor) %>% 
  summarise(set_size = n()) %>% 
  inner_join(go_bp_main, by = "term_ancestor") %>% 
  select(term_ancestor, set_size, go_ancestor, symbol) %>% 
  distinct() %>% 
  mutate(annotation = "Minor process")

go_bp_all = bind_rows(go_bp_main_2, go_bp_child)

write_csv(go_bp_all, file = here("VaxGO gene sets", "OtherGOs_Annotated_Genes.csv"))

# Collapse rows 
go_bp_all_comma = go_bp_all %>% 
  group_by(term_ancestor) %>% 
  summarize(genes = paste0(symbol, collapse = ",")) %>% 
  inner_join(go_bp_all, by = "term_ancestor") %>% 
  select(-symbol) %>% 
  distinct() %>% 
  arrange(desc(set_size))

write_csv(go_bp_all_comma, file = here("VaxGO gene sets","OtherGOs_Annotated_GOs.csv"))
```

```{r}
OtherGOs_Annotated_Genes = read_csv(file = here("VaxGO gene sets", "OtherGOs_Annotated_Genes.csv"))
go_bp_all = OtherGOs_Annotated_Genes


# DEGs to Entrez and GO ids --------
degs_GO = all_degs_p_05_vac_infected %>% 
  select(condition, genes, direction) %>% 
  left_join(go_bp_all %>% rename(genes = symbol), by = "genes") %>% 
  mutate(go_ancestor = if_else(is.na(go_ancestor), "Not mapped", go_ancestor),
         term_ancestor = if_else(is.na(term_ancestor), "Not mapped", term_ancestor),
         mapped = if_else(term_ancestor == "Not mapped", "Not mapped", "Mapped"))

# Collapse gene column --------
degs_GO_comma = degs_GO %>% 
  distinct() %>% 
  group_by(condition, term_ancestor) %>% 
  summarize(genes = paste0(genes, collapse = ",")) %>% 
  inner_join(ann_vaccines, by = "condition")

write_csv(degs_GO_comma, file = here("DGE analysis", "degs_annotated_allGOs.csv"))


#DEGs annotated by Main GO BPs ------

main_go = degs_GO %>%
  filter(go_ancestor %in% c(
      "Not mapped",
      "GO:0040011",
      "GO:0043473",
      "GO:0044419",
      "GO:0009987",
      "GO:0048518",
      "GO:0050789",
      "GO:0065007",
      "GO:0048511",
      "GO:0022414",
      "GO:0002376",
      "GO:0008152")) %>% 
  select(condition, genes, direction, term_ancestor) %>% 
  distinct() %>% 
  group_by(condition, direction, term_ancestor) %>% 
  summarise(n = n()) %>% 
  mutate(term_ancestor = if_else(is.na(term_ancestor), "No annotation", term_ancestor))

main_go_2 = degs_GO %>% 
  select(condition, genes, mapped, direction) %>% 
  distinct() %>% 
  group_by(condition, direction) %>% 
  summarise(n_genes_total = n()) %>% 
  inner_join(main_go, by = join_by("condition", direction)) %>% 
  ungroup() %>% 
  select(condition, direction, term_ancestor, n_genes = n, n_genes_total) %>% 
  mutate(n_genes_perc = round(n_genes/n_genes_total * 100, 1),
         term_ancestor = case_when(
           term_ancestor == "biological process involved in interspecies interaction between organisms" ~ "Interspecies interaction",
           term_ancestor == "biological regulation" ~ "Biological reg.",
           term_ancestor == "cellular process" ~ "Cellular",
           term_ancestor == "immune system process" ~ "Immune system",
           term_ancestor == "regulation of biological process" ~ "Reg. of BP",
           term_ancestor == "locomotion" ~ "Locomotion",
           term_ancestor == "metabolic process" ~ "Metabolism",
           TRUE ~ term_ancestor),
         term_ancestor = fct_relevel(term_ancestor, "Not mapped", after = Inf),
         direction = fct_relevel(direction, "UP", "DOWN"))
  
write_csv(main_go_2, file = here("DGE analysis", "degs_annotated_mainGOs_stats.csv"))
```

### Boxplot

```{r fig.height=10, fig.width=15, warning=FALSE}
degs_annotated_mainGOs_stats <- read_csv(here("DGE analysis", "degs_annotated_mainGOs_stats.csv"))
main_go_2 = degs_annotated_mainGOs_stats

# Table
df_day =  degs_annotated_mainGOs_stats %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  select(condition, 
         direction, 
         term_ancestor, 
         n_genes, 
         n_genes_perc,
         day,
         disease_vac) %>% 
  filter(!term_ancestor %in% c("Interspecies interaction","Not mapped", "Reg. of BP")) %>% 
  mutate(day_f = as.factor(day),
         day_f = fct_reorder(day_f, day),
         day_dv = paste(day_f, sep = "",  " (", disease_vac, ")") %>% as.factor(),
         day_dv = fct_reorder(day_dv, day),
         term_ancestor = fct_relevel(term_ancestor, "Immune system", "Cellular", "Metabolism"),
         direction = fct_relevel(direction, "UP", "DOWN"))
df_day
```

```{r fig.height=10, fig.width=20}
all_up_down_day = df_day %>% 
  group_by(disease_vac) %>%
  arrange(desc(n_genes_perc)) %>% 
  ggplot() +
  aes(
    x = day_dv,
    y = n_genes_perc
  ) +
  geom_jitter(aes(colour = disease_vac),
              na.rm = TRUE,
              size = 4) +
  geom_boxplot(aes(fill = disease_vac, color = disease_vac),
               outlier.shape = NA,
               alpha = 0) +
  theme_few() +
  facet_grid(~direction~term_ancestor) +
  theme(axis.text.x = element_text(size = 15, color = "black", angle = 90, vjust = 0.5),
        axis.text.y = element_text(size = 15, color = "black"),
        axis.title.x = element_text(size = 15, color = "black", vjust = -1),
        axis.title.y = element_text(size = 15, color = "black"),
        strip.text.x = element_text(size = 15, face = "bold"),
        strip.text.y = element_text(size = 15, angle = 0,hjust = 0, face = "bold"),
        legend.text = element_text(size = 15, color = "black"),
        legend.title = element_text(size = 15, color = "black"),
        plot.title = element_text(hjust = 0.5,
                                  size = 15,
                                  face = "bold")) +
  labs(x = "Day",
       y = "Number of genes (%)",
       size = "Total number of genes",
       shape = "Type") + 
  scale_fill_npg() + 
  scale_colour_npg() 

all_up_down_day

ggsave(all_up_down_day, width = 20, height = 10, file = here("Figures", "MainGOs_percgenes_boxplot_up_down_day.png"))
```


### Barplot by term

#### Up vs Down vs Neutral
```{r fig.height=30, fig.width=20, message=FALSE, warning=FALSE}
main_go_2_plot_up_down = main_go_2 %>%
  inner_join(ann_vaccines_conditions, by = "condition") %>% 
  mutate(
    term_ancestor = fct_relevel(
      term_ancestor,
      "Immune system",
      "Cellular",
      "Metabolism",
      "Reg. of BP"
    )
  ) %>%
filter(term_ancestor %in% c("Immune system", "Cellular", "Metabolism")) %>% 
mutate(disease_vac = fct_relevel(disease_vac, "I", "V")) %>% 
  arrange(disease_vac) %>% 
  mutate(condition = factor(condition, levels = ann_vaccines_conditions_ordered$condition_grouped %>% as.character() %>% unique())) %>% 
  ggplot() +
 aes(x = fct_rev(factor(condition)), y = n_genes_perc, fill = type) +
 geom_col(size = 0.5, 
          color = "black",
          width = 0.6) +
 coord_flip() +
 theme_minimal() +
   scale_fill_manual(name = "Type",
    values = c('IN'= "#00A087FF", #1st Gen  vaccines
           'VV' = "#3C5488FF", #3rd Gen vaccines
           'RNA' = "#4DBBD5FF",
           'SU' = "#8491B4FF",
           'I'= "#DC0000FF",
           "H" = "grey95")) +
  theme(legend.position = "right",
        strip.text.x = element_text(color = "black", size = 12), 
        axis.text.x = element_text(color = "black", size = 10),
        axis.text.y = element_text(vjust = 0.5, size = 10, color = "black"),
        axis.line = element_line(size = 1, colour = "black", linetype=1),
        axis.ticks = element_line(size = 0.5, color = "black"),
        panel.grid.major.y = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.major.x = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.spacing.x = unit(2, "lines")) +
  geom_text(aes(label = paste0(n_genes_perc, "% (", n_genes, ")"), 
                y = n_genes_perc), 
            hjust = 0, 
            size = 4,
            angle = 0) +
  labs(title = "",
       x = "",
       y = "Number of genes (%)") +
  ylim(0,120) +
  facet_grid(vars(direction),  vars(term_ancestor))

main_go_2_plot_up_down_totals = main_go_2 %>% 
    inner_join(ann_vaccines_conditions, by = "condition") %>% 
    select(condition, n_genes_total, type, direction, disease_vac) %>% 
  distinct() %>% 
  mutate(disease_vac = fct_relevel(disease_vac, "I", "V")) %>% 
  mutate(stat = "Total genes") %>% 
  arrange(disease_vac) %>% 
  mutate(condition = factor(condition, levels = ann_vaccines_conditions_ordered$condition_grouped %>% as.character() %>% unique())) %>%   
  ggplot() +
 aes(x = fct_rev(factor(condition)), y = n_genes_total, fill = type) +
 geom_col(size = 0.5, 
          color = "black",
          width = 0.6) +
 coord_flip() +
 theme_minimal() +
   scale_fill_manual(name = "Type",
    values = c('IN'= "#00A087FF", #1st Gen  vaccines
           'VV' = "#3C5488FF", #3rd Gen vaccines
           'RNA' = "#4DBBD5FF",
           'SU' = "#8491B4FF",
           'I'= "#DC0000FF",
           "H" = "grey95")) +
  theme(legend.position = "right",
        strip.text.x = element_text(color = "black", size = 12), 
        axis.text.x = element_text(color = "black", size = 10),
        axis.text.y = element_text(vjust = 0.5, size = 10, color = "black"),
        axis.line = element_line(size = 1, colour = "black", linetype=1),
        axis.ticks = element_line(size = 0.5, color = "black"),
        panel.grid.major.y = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.major.x = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.spacing.x = unit(2, "lines")) +
  
  geom_text(aes(label = n_genes_total, y = n_genes_total),
            hjust = 0, 
            size = 4,
            angle = 0) +
  labs(title = "",
       x = "",
       y = "Total number of genes") +
  ylim(0, 5000) +
   facet_grid(vars(direction),  vars(stat))

patch_maingo = main_go_2_plot_up_down_totals + main_go_2_plot_up_down +
  plot_layout(widths = c(1, 3),
              guides = "collect",
              axes = "collect") +
  plot_annotation(title = "Main Gene Ontologies (term ancestor), by condition",
                  theme = theme(legend.position = "top",
                                plot.title = element_text(size = 25,
                                                          face = "bold",
                                                     hjust = 0.5)))

patch_maingo
ggsave(patch_maingo, file = here("Figures", "main_go_2_plot_values_UP_DOWN_regulated.png"), 
                                           width = 20, 
                                           height = 30)
```


#### All DEGs
```{r fig.height=10, fig.width=10, message=FALSE, warning=FALSE}
main_go_2_plot_alldirections = main_go_2 %>%
  group_by(condition, term_ancestor) %>%
  summarize(n_genes = sum(n_genes),
            n_genes_total = sum(n_genes_total),
            n_genes_perc = round((n_genes/n_genes_total)*100, 1)) %>% 
inner_join(ann_vaccines_conditions, by = "condition") %>% 
  mutate(
    term_ancestor = fct_relevel(
      term_ancestor,
      "Immune system",
      "Cellular",
      "Metabolism",
      "Reg. of BP"
    )
  ) %>%
filter(term_ancestor %in% c(
  # "Immune system", 
  "Cellular", 
  "Metabolism")) %>% 
  mutate(disease_vac = fct_relevel(disease_vac, "I", "V")) %>% 
  arrange(disease_vac) %>% 
  mutate(condition = factor(condition, levels = ann_vaccines_conditions_ordered$condition_grouped %>% as.character() %>% unique())) %>% 
  ggplot() +
 aes(x = fct_rev(factor(condition)), y = n_genes_perc, fill = type) +
 geom_col(size = 0.5, 
          color = "black",
          width = 0.6) +
 coord_flip() +
 theme_minimal() +
   scale_fill_manual(name = "Type",
    values = c('IN'= "#00A087FF", #1st Gen  vaccines
           'VV' = "#3C5488FF", #3rd Gen vaccines
           'RNA' = "#4DBBD5FF",
           'SU' = "#8491B4FF",
           'I'= "#DC0000FF",
           "H" = "grey90")) +
  theme(legend.position = "right",
        strip.text.x = element_text(color = "black", size = 12), 
        axis.text.x = element_text(color = "black", size = 10),
        axis.text.y = element_text(vjust = 0.5, size = 10, color = "black"),
        axis.line = element_line(size = 1, colour = "black", linetype=1),
        axis.ticks = element_line(size = 0.5, color = "black"),
        panel.grid.major.y = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.major.x = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.spacing.x = unit(2, "lines")) +
  geom_text(aes(label = paste0(n_genes_perc, "% (", n_genes, ")"), 
                y = n_genes_perc), 
            hjust = 0, 
            size = 3,
            angle = 0) +
  labs(title = "Biological processes",
       x = "",
       y = "Number of genes (%)") +
  facet_wrap(~term_ancestor, ncol = 3) +
  scale_y_continuous(expand = c(0,0), limits = c(0,120))

main_go_2_plot_alldirections_totals = main_go_2 %>%
  group_by(condition, term_ancestor) %>% 
  summarize(n_genes = sum(n_genes),
            n_genes_total = sum(n_genes_total),
            n_genes_perc = round((n_genes/n_genes_total)*100, 1)) %>% 
inner_join(ann_vaccines_conditions, by = "condition") %>% 
  mutate(
    term_ancestor = fct_relevel(
      term_ancestor,
      "Immune system",
      "Cellular",
      "Metabolism",
      "Reg. of BP"
    )
  ) %>%
    select(condition, n_genes_total, type, disease_vac) %>% 
  distinct() %>% 
  mutate(disease_vac = fct_relevel(disease_vac, "I", "V")) %>% 
  mutate(stat = "Total genes") %>% 
  arrange(disease_vac) %>% 
  mutate(condition = factor(condition, levels = ann_vaccines_conditions_ordered$condition_grouped %>% as.character() %>% unique())) %>%   
  ggplot() +
 aes(x = fct_rev(factor(condition)), y = n_genes_total, fill = type) +
 geom_col(size = 0.5, 
          color = "black",
          width = 0.6) +
 coord_flip() +
 theme_minimal() +
   scale_fill_manual(name = "Type",
    values = c('IN'= "#00A087FF", #1st Gen  vaccines
           'VV' = "#3C5488FF", #3rd Gen vaccines
           'RNA' = "#4DBBD5FF",
           'SU' = "#8491B4FF",
           'I'= "#DC0000FF",
           "H" = "grey90")) +
   theme(legend.position = "right",
        strip.text.x = element_text(color = "black", size = 12), 
        axis.text.x = element_text(color = "black", size = 10),
        axis.text.y = element_text(vjust = 0.5, size = 10, color = "black"),
        axis.line = element_line(size = 1, colour = "black", linetype=1),
        axis.ticks = element_line(size = 0.5, color = "black"),
        panel.grid.major.y = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.major.x = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.spacing.x = unit(2, "lines")) +
  geom_text(aes(label = n_genes_total, y = n_genes_total),
            hjust = 0, 
            size = 3,
            angle = 0) +
  labs(title = "",
       x = "",
       y = "Total number of genes") +
  facet_wrap(~stat) +
  scale_y_continuous(expand = c(0,0), limits = c(0, 10000))

patch_maingo = main_go_2_plot_alldirections_totals + main_go_2_plot_alldirections +
  plot_layout(widths = c(1, 2),
              axes = "collect",
              guides = "collect") +
  plot_annotation(title = "Main Gene Ontologies (term ancestor), by condition",
                  theme = theme(legend.position = "right",
                                plot.title = element_text(size = 25,
                                                          face = "bold",
                                                     hjust = 0.5)))

patch_maingo
ggsave(patch_maingo, file = here("Figures", "main_go_2_plot_values_AllDirections_regulated.png"), 
                                           width = 15, 
                                           height = 10)
```

No labels
```{r fig.height=10, fig.width=10, message=FALSE, warning=FALSE}
main_go_2_plot_alldirections = main_go_2 %>%
  group_by(condition, term_ancestor) %>%
  summarize(n_genes = sum(n_genes),
            n_genes_total = sum(n_genes_total),
            n_genes_perc = round((n_genes/n_genes_total)*100, 1)) %>% 
inner_join(ann_vaccines_conditions, by = "condition") %>% 
  mutate(
    term_ancestor = fct_relevel(
      term_ancestor,
      "Immune system",
      "Cellular",
      "Metabolism",
      "Reg. of BP"
    )
  ) %>%
filter(term_ancestor %in% c(
  # "Immune system", 
  "Cellular", 
  "Metabolism")) %>% 
  mutate(disease_vac = fct_relevel(disease_vac, "I", "V")) %>% 
  arrange(disease_vac) %>% 
  mutate(condition = factor(condition, levels = ann_vaccines_conditions_ordered$condition_grouped %>% as.character() %>% unique())) %>% 
  ggplot() +
 aes(x = fct_rev(factor(condition)), y = n_genes_perc, fill = type) +
 geom_col(size = 0.5, 
          color = "black",
          width = 0.6) +
 coord_flip() +
 theme_minimal() +
   scale_fill_manual(name = "Type",
    values = c('IN'= "#00A087FF", #1st Gen  vaccines
           'VV' = "#3C5488FF", #3rd Gen vaccines
           'RNA' = "#4DBBD5FF",
           'SU' = "#8491B4FF",
           'I'= "#DC0000FF",
           "H" = "grey90")) +
  theme(legend.position = "right",
        strip.text.x = element_text(color = "black", size = 12), 
        axis.text.x = element_text(color = "black", size = 10),
        axis.text.y = element_text(vjust = 0.5, size = 10, color = "black"),
        axis.line = element_line(size = 1, colour = "black", linetype=1),
        axis.ticks = element_line(size = 0.5, color = "black"),
        panel.grid.major.y = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.major.x = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.spacing.x = unit(2, "lines")) +
  labs(title = "Biological processes",
       x = "",
       y = "Coverage (%)") +
  facet_wrap(~term_ancestor, ncol = 3) +
  scale_y_continuous(expand = c(0,0), limits = c(0,100))

main_go_2_plot_alldirections_totals = main_go_2 %>%
  group_by(condition, term_ancestor) %>% 
  summarize(n_genes = sum(n_genes),
            n_genes_total = sum(n_genes_total),
            n_genes_perc = round((n_genes/n_genes_total)*100, 1)) %>% 
inner_join(ann_vaccines_conditions, by = "condition") %>% 
  mutate(
    term_ancestor = fct_relevel(
      term_ancestor,
      "Immune system",
      "Cellular",
      "Metabolism",
      "Reg. of BP"
    )
  ) %>%
    select(condition, n_genes_total, type, disease_vac) %>% 
  distinct() %>% 
  mutate(disease_vac = fct_relevel(disease_vac, "I", "V")) %>% 
  mutate(stat = "Total genes") %>% 
  arrange(disease_vac) %>% 
  mutate(condition = factor(condition, levels = ann_vaccines_conditions_ordered$condition_grouped %>% as.character() %>% unique())) %>%   
  ggplot() +
 aes(x = fct_rev(factor(condition)), y = n_genes_total, fill = type) +
 geom_col(size = 0.5, 
          color = "black",
          width = 0.6) +
 coord_flip() +
 theme_minimal() +
   scale_fill_manual(name = "Type",
    values = c('IN'= "#00A087FF", #1st Gen  vaccines
           'VV' = "#3C5488FF", #3rd Gen vaccines
           'RNA' = "#4DBBD5FF",
           'SU' = "#8491B4FF",
           'I'= "#DC0000FF",
           "H" = "grey90")) +
   theme(legend.position = "right",
        strip.text.x = element_text(color = "black", size = 12), 
        axis.text.x = element_text(color = "black", size = 10),
        axis.text.y = element_text(vjust = 0.5, size = 10, color = "black"),
        axis.line = element_line(size = 1, colour = "black", linetype=1),
        axis.ticks = element_line(size = 0.5, color = "black"),
        panel.grid.major.y = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.major.x = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.spacing.x = unit(2, "lines")) +
  labs(title = "",
       x = "",
       y = "Total number of genes") +
  facet_wrap(~stat) +
  scale_y_continuous(expand = c(0,0), limits = c(0, 10000))

patch_maingo = main_go_2_plot_alldirections_totals + main_go_2_plot_alldirections +
  plot_layout(widths = c(1, 2),
              axes = "collect",
              guides = "collect") +
  plot_annotation(title = "Main Gene Ontologies (term ancestor), by condition",
                  theme = theme(legend.position = "right",
                                plot.title = element_text(size = 25,
                                                          face = "bold",
                                                     hjust = 0.5)))

patch_maingo
ggsave(patch_maingo, file = here("Figures", "main_go_2_plot_values_AllDirections_regulated_nolabels.png"), 
                                           width = 15, 
                                           height = 10)
```



## Unite with direction and biotype
```{r fig.height=10, fig.width=20}
ggplot_degs_bars_directions_biotype_bps =
  ggplot_degs_bars_vertical_nolabels +
  # all_degs_biotypes_plot_totals + 
  all_degs_biotypes_plot_perc +
  # ggplot_degs_bars_directions_up_down +
  # all_degs_immuneGO_BTM_plot +
  all_degs_immuneGO_BTM_united_plot_nolabel +
  # all_degs_immuneGO_BTM_reg_plot_dodged +
  # all_degs_immuneGO_BTM_united_plot + 
  # all_degs_biotypes_plot_perc_immune + 
  # all_degs_biotypes_plot_totals_immune + 
  # all_degs_biotypes_plot_perc + 
  main_go_2_plot_alldirections + 
  plot_layout(guides = "collect", axes = "collect", axis_titles = "collect", nrow = 1, widths = c(1, 1, 1, 2)) 

ggplot_degs_bars_directions_biotype_bps

ggplot_degs_bars_directions_biotype_bps %>% 
  ggsave(filename = here("Figures", "DEGs_Directions_Biotype_BioProcesses_Up_Down.png"),
         width = 20,
         height = 10)

```

## Correlation analysis - Genes and processes

### Immune genes
```{r fig.height=20, fig.width=15}
library(corrplot)

genes_ImmuneSystem = degs_covid_vax %>%
  inner_join(immunego_btm_genes %>% 
               select(genes) %>% 
               distinct(), by = "genes") %>% 
  select(condition, genes, log2fold_change) %>% 
  distinct() %>% 
  pivot_wider(names_from = "condition",
              values_from = "log2fold_change", values_fn = mean) %>% 
  column_to_rownames("genes") %>% 
  mutate_all(~ replace_na(., 0)) %>% 
  select(any_of(c(ann_vaccines_covid_other_conditions %>% 
                    filter(set == "COVID-19") %>% 
                    arrange(day)  %>% 
                    pull(condition) %>% 
                    as.character())))

correlation_genes_ImmuneSystem = genes_ImmuneSystem %>% 
  cor(method = "spearman") %>% 
  as.data.frame() %>% 
  select(any_of(c(ann_vaccines_conditions_ordered %>% 
                  filter(disease_vac == "V") %>% 
                  arrange(condition, day) %>% 
                  pull(condition_grouped) %>% 
                  as.character(),
                ann_vaccines_conditions_ordered %>% 
                  filter(disease_vac == "I") %>% 
                  arrange(condition, day) %>% 
                  pull(condition_grouped) %>% 
                  as.character())))

p.mat <- cor.mtest(genes_ImmuneSystem, method = 'spearman')
p.mat_long = p.mat %>% 
  as.data.frame() %>%
  rownames_to_column("condition") %>% 
  pivot_longer(cols = -condition,
               names_to = "condition_2",
               values_to =  "pvalue") %>% 
  filter(pvalue < 0.05)



correlation_genes_long = correlation_genes_ImmuneSystem %>% 
  as.data.frame() %>% 
  rownames_to_column("condition") %>% 
  pivot_longer(cols = -condition, 
               names_to = "condition_2",
               values_to = "correlation") %>% 
  inner_join(p.mat_long,
             by = join_by(condition, condition_2))

correlation_genes_filtered_matrix = correlation_genes_long %>% 
  select(-pvalue) %>% 
  pivot_wider(names_from = "condition_2",
              values_from = "correlation") %>% 
  column_to_rownames("condition") %>% 
  mutate_all(~ replace_na(., 0)) %>% 
  as.matrix()

  
```

Clustered
```{r}
# INPUT ------
correlation_genes_filtered_matrix
filename = "ImmuneGenes_CorrelationAnalysis_COVID" 

# Matrix ------
matrix = correlation_genes_filtered_matrix

# Annotations -------
ann_vaccines_covid = ann_vaccines_conditions

# Columns
ann_cols_heatmap = matrix %>% 
  colnames() %>% 
  data.frame() %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_covid, by.x = "condition", by.y = "condition", all.y = F) %>% 
  mutate(day = as.character(day) %>% 
           as.numeric()) %>% 
  select(condition, disease_vac, type, dose, infection, day) %>% 
  distinct() %>% 
  column_to_rownames(var = "condition") %>% 
  arrange(day) %>% 
  mutate(day = as.character(day) %>% fct_reorder(., day))

# Rows
ann_rows_heatmap = matrix %>%  
  as.data.frame() %>% 
  rownames_to_column("condition") %>% 
  distinct() %>% 
  merge(ann_vaccines_covid, by.x = "condition", by.y = "condition", all.y = F) %>% 
  mutate(day = as.character(day) %>% 
           as.numeric()) %>% 
  select(condition, disease_vac, type, dose, infection, day) %>% 
  distinct() %>% 
  column_to_rownames(var = "condition") %>% 
  arrange(day) %>% 
  mutate(day = as.character(day) %>% fct_reorder(., day))

# Arrange cols and rows in matrix
matrix_data = matrix %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>%
  rownames_to_column("condition") %>% 
  inner_join(ann_rows_heatmap %>% 
               rownames_to_column("condition") %>% 
               select(condition, day),
             by = "condition") %>% 
  arrange(day) %>% 
  select(-day) %>% 
  column_to_rownames("condition") %>% 
  mutate_if(is.character, as.numeric) %>%
  replace(is.na(.), 0) %>% 
  as.matrix()

#Check dimensions
dim(matrix_data)
dim(ann_rows_heatmap)
dim(ann_cols_heatmap)

col_fun <- colorRamp2(
  breaks = c(-1, -0.5, 0,0.5, 1), 
  colors = c("darkblue", "#4DBBD5FF", "white", "red", "#DC0000FF")
)

#Heatmap annotation
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "left")

row_ha = HeatmapAnnotation(df = ann_rows_heatmap,
                       col = ann_vaxsig_colors,
                       which= "row",
                       show_annotation_name = F,
                       show_legend = T)


mat = matrix_data
width_image = 35
height_image = 30
width = unit(20, "cm")
height = unit(20, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)
file = paste0("Hagan_", filename)

fileimageheatmap_grouped = paste0(filename, sep ="_", "Complexheatmap_ALL_CLUSTERED.png")

png(file = here("Figures", fileimageheatmap_grouped), 
    width = width_image, 
    height = height_image,
    units="cm", res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "left",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "Pearson\ncorrelation", #Legend
                        col = col_fun, #color
cell_fun = function(j, i, x, y, width, height, fill) {
  # Define um limite próximo de zero
  if(abs(mat[i, j]) > 0.1) {
    grid.text(sprintf("%.2f", mat[i, j]), 
              x, y, 
              gp = gpar(fontsize = 7))
  }
},
                        row_title = "",
                        show_column_names = TRUE,
                        column_title = "Correlation Analysis, p < 0.05.\nValues shown are corr > 0.10",
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = gpar(col = "white", lwd = 2),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(5, "cm"),
                        column_split = NULL, 
                        row_split = NULL, 
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width,
                        height = height 
)

draw(heatmap_plot,
     merge_legend = TRUE,
     annotation_legend_side = "right", 
     heatmap_legend_side = "right")
dev.off()

```

```{r}
# INPUT ------
correlation_genes_filtered_matrix
filename = "ImmuneGenes_CorrelationAnalysis_COVID" 

# Matrix ------
matrix = correlation_genes_filtered_matrix

# Annotations -------
ann_vaccines_covid = ann_vaccines_conditions

# Columns
ann_cols_heatmap = matrix %>% 
  colnames() %>% 
  data.frame() %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_covid, by.x = "condition", by.y = "condition", all.y = F) %>% 
  mutate(day = as.character(day) %>% 
           as.numeric()) %>% 
  select(condition, disease_vac, type, dose, infection, day) %>% 
  distinct() %>% 
  column_to_rownames(var = "condition") %>% 
  arrange(day) %>% 
  mutate(day = as.character(day) %>% fct_reorder(., day))

# Rows
ann_rows_heatmap = matrix %>%  
  as.data.frame() %>% 
  rownames_to_column("condition") %>% 
  distinct() %>% 
  merge(ann_vaccines_covid, by.x = "condition", by.y = "condition", all.y = F) %>% 
  mutate(day = as.character(day) %>% 
           as.numeric()) %>% 
  select(condition, disease_vac, type, dose, infection, day) %>% 
  distinct() %>% 
  column_to_rownames(var = "condition") %>% 
  arrange(day) %>% 
  mutate(day = as.character(day) %>% fct_reorder(., day))

# Arrange cols and rows in matrix
matrix_data = matrix %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>%
  rownames_to_column("condition") %>% 
  inner_join(ann_rows_heatmap %>% 
               rownames_to_column("condition") %>% 
               select(condition, day),
             by = "condition") %>% 
  arrange(day) %>% 
  select(-day) %>% 
  column_to_rownames("condition") %>% 
  mutate_if(is.character, as.numeric) %>%
  replace(is.na(.), 0) %>% 
  as.matrix()

#Check dimensions
dim(matrix_data)
dim(ann_rows_heatmap)
dim(ann_cols_heatmap)

col_fun <- colorRamp2(
  breaks = c(-1, -0.5, 0,0.5, 1), 
  colors = c("darkblue", "#4DBBD5FF", "white", "red", "#DC0000FF")
)

#Heatmap annotation
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "left")

row_ha = HeatmapAnnotation(df = ann_rows_heatmap,
                       col = ann_vaxsig_colors,
                       which= "row",
                       show_annotation_name = F,
                       show_legend = T)


mat = matrix_data
width_image = 35
height_image = 30
width = unit(20, "cm")
height = unit(20, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)
file = paste0("Hagan_", filename)

fileimageheatmap_grouped = paste0(filename, sep ="_", "Complexheatmap_Days.png")

png(file = here("Figures", fileimageheatmap_grouped), 
    width = width_image, 
    height = height_image,
    units="cm", res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "left",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "Pearson\ncorrelation", #Legend
                        col = col_fun, #color
cell_fun = function(j, i, x, y, width, height, fill) {
  # Define um limite próximo de zero
  if(abs(mat[i, j]) > 0.1) {
    grid.text(sprintf("%.2f", mat[i, j]), 
              x, y, 
              gp = gpar(fontsize = 7))
  }
},
                        row_title = "",
                        show_column_names = TRUE,
                        column_title = "Correlation Analysis, p < 0.05.\nValues shown are corr > 0.10",
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = gpar(col = "white", lwd = 2),
                        cluster_rows = F,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(5, "cm"),
                        column_split = NULL, 
                        row_split = NULL, 
                        row_gap = unit(2, "mm"),
                        cluster_columns = F,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width,
                        height = height 
)

draw(heatmap_plot,
     merge_legend = TRUE,
     annotation_legend_side = "right", 
     heatmap_legend_side = "right")
dev.off()

```


Ordered by condition
```{r}
correlation_genes_filtered_matrix
filename = "ImmuneGenes_CorrelationAnalysis_COVID" 

# Matrix ------
matrix = correlation_genes_filtered_matrix

# Annotations -------
ann_vaccines_covid = ann_vaccines_conditions

# Columns
ann_cols_heatmap = matrix %>% 
  colnames() %>% 
  data.frame() %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_covid, by.x = "condition", by.y = "condition", all.y = F) %>% 
  mutate(day = as.character(day) %>% 
           as.numeric()) %>% 
  select(condition, disease_vac, type, dose, infection, day) %>% 
  distinct() %>% 
  arrange(match(condition, c(ann_vaccines_conditions_ordered$condition_grouped %>% as.character()))) %>%  
  column_to_rownames(var = "condition") %>% 
  mutate(day = as.character(day) %>% fct_reorder(., day))

# Rows
ann_rows_heatmap = matrix %>%  
  as.data.frame() %>% 
  rownames_to_column("condition") %>% 
  distinct() %>% 
  merge(ann_vaccines_covid, by.x = "condition", by.y = "condition", all.y = F) %>% 
  mutate(day = as.character(day) %>% 
           as.numeric()) %>% 
  select(condition, disease_vac, type, dose, infection, day) %>% 
  distinct() %>% 
  arrange(match(condition, c(ann_vaccines_conditions_ordered$condition_grouped %>% as.character()))) %>%  
  column_to_rownames(var = "condition") %>% 
  mutate(day = as.character(day) %>% fct_reorder(., day))

# Arrange cols and rows in matrix
matrix_data = matrix %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>%
  rownames_to_column("condition") %>% 
  inner_join(ann_rows_heatmap %>% 
               rownames_to_column("condition") %>% 
               select(condition),
             by = "condition") %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  mutate_if(is.character, as.numeric) %>%
  replace(is.na(.), 0) %>% 
  as.matrix()

#Check dimensions
dim(matrix_data)
dim(ann_rows_heatmap)
dim(ann_cols_heatmap)

col_fun <- colorRamp2(
  breaks = c(-1, -0.5, 0,0.5, 1), 
  colors = c("darkblue", "#4DBBD5FF", "white", "red", "#DC0000FF")
)

#Heatmap annotation
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "left")

row_ha = HeatmapAnnotation(df = ann_rows_heatmap,
                       col = ann_vaxsig_colors,
                       which= "row",
                       show_annotation_name = F,
                       show_legend = T)

mat = matrix_data
width_image = 35
height_image = 30
width = unit(20, "cm")
height = unit(20, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)
file = paste0("Hagan_", filename)

fileimageheatmap_grouped = paste0(filename, sep ="_", "Complexheatmap_Condition.png")

png(file = here("Figures", fileimageheatmap_grouped), 
    width = width_image, 
    height = height_image,
    units="cm", res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "left",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "Pearson\ncorrelation", #Legend
                        col = col_fun, #color
cell_fun = function(j, i, x, y, width, height, fill) {
  # Define um limite próximo de zero
  if(abs(mat[i, j]) > 0.1) {
    grid.text(sprintf("%.2f", mat[i, j]), 
              x, y, 
              gp = gpar(fontsize = 7))
  }
},
                        row_title = "",
                        show_column_names = TRUE,
                        column_title = "Correlation Analysis, p < 0.05.\nValues shown are corr > 0.10",
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = gpar(col = "white", lwd = 2),
                        cluster_rows = F,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(5, "cm"),
                        column_split = NULL, 
                        row_split = NULL, 
                        row_gap = unit(2, "mm"),
                        cluster_columns = F,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width,
                        height = height 
)

draw(heatmap_plot,
     merge_legend = TRUE,
     annotation_legend_side = "right", 
     heatmap_legend_side = "right")
dev.off()

```


## Overlap analysis 
### Shared genes
### Immune related only

```{r}

degs_all_updown = all_degs_p_05_vac_infected

# immunego_btm_genes

shared_genes_infection = degs_all_updown %>%
  select(condition, genes) %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
    filter(disease_vac == "I") %>%
  select(condition, genes, vaccine) %>% 
  inner_join(immunego_btm_genes %>% select(genes) %>% distinct(), by = "genes") %>% 
  group_by(genes) %>% 
  summarise("Infection" = n()) %>% 
  arrange(desc("Infection"), genes) %>% 
  ungroup() %>% 
  pivot_longer(-genes, names_to = "infection_vaccination", values_to = "n_conditions") %>% 
  inner_join(immunego_btm_genes , by = "genes") %>% 
  select(genes, infection_vaccination, n_conditions) %>% 
  mutate(comparison = "Infection only") %>% 
  arrange(desc(n_conditions)) %>% 
  distinct()

shared_genes_vaccines = degs_all_updown %>%
  select(condition, genes) %>% 
  distinct() %>% 
  inner_join(immunego_btm_genes %>% select(genes) %>% distinct(), by = "genes") %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  filter(disease_vac == "V") %>% 
  select(condition, genes, vaccine) %>% 
  group_by(genes) %>% 
  summarise("Vaccination" = n()) %>% 
  arrange(desc("Vaccination"), genes) %>% 
  ungroup() %>% 
  pivot_longer(-genes, names_to = "infection_vaccination", values_to = "n_conditions") %>% 
  inner_join(immunego_btm_genes %>% select(genes) %>% distinct(), by = "genes") %>% 
  select(genes, infection_vaccination,n_conditions) %>% 
  mutate(comparison = "Vaccination only") %>% 
  arrange(desc(n_conditions)) %>% 
  distinct()


# Infection and vaccination
shared_genes_vaccines_infection = shared_genes_vaccines %>% 
  select(genes, n_conditions) %>% 
  inner_join(shared_genes_infection %>% select(genes, n_conditions), by = "genes") %>% 
  rename(Vaccination = n_conditions.x, Infection = n_conditions.y) %>% 
  distinct() %>% 
  pivot_longer(-genes, names_to = "infection_vaccination", values_to = "n_conditions") %>% 
  inner_join(immunego_btm_genes %>% select(genes) %>% distinct(), by = "genes") %>% 
  select(genes, infection_vaccination,n_conditions) %>% 
  mutate(comparison = "Infection and vaccination") %>% 
  distinct()
  

shared_genes_comparisons = bind_rows(shared_genes_infection, 
                                     shared_genes_vaccines, 
                                     shared_genes_vaccines_infection)

write_csv(shared_genes_comparisons, file = here("Figures", "ImmuneGO_BTM_shared_genes_infection_vaccination_comparisons.csv"))

```
Horizontal
```{r fig.height=9, fig.width=10}
# Vaccine and infection ----
shared_genes_infection_vaccination = shared_genes_comparisons %>% 
  filter(comparison == "Infection and vaccination")

shared_genes_infection = shared_genes_comparisons %>% 
  filter(comparison == "Infection only")

shared_genes_vaccines = shared_genes_comparisons %>% 
  filter(comparison == "Vaccination only")


shared_genes_vaccines_infection_plot = shared_genes_infection_vaccination %>% 
  select(genes, infection_vaccination, n_conditions, comparison) %>% 
  distinct() %>%
  slice_head(n=40) %>%  
  ggplot() +
  aes(
    x = reorder(genes, -n_conditions),
    y = n_conditions,
    fill = infection_vaccination
  ) +
  geom_col(color = "black",
           size = 0.2,
           width = 0.7) +
  scale_fill_manual(
    values = c("Vaccination" = "#4DBBD5FF",
              "Infection" = "#DC0000FF")
    )+
  theme_minimal() +
  labs(fill = "Status") +
  xlab("") + 
  ylab("") +
  theme(legend.position = "none",
        strip.text.x = element_text(color = "black", size = 12), 
        axis.text.x = element_text(color = "black", size = 10, angle = 45, vjust = 1, hjust = 1),
        axis.text.y = element_text(vjust = 0.5, size = 10, color = "black"),
        axis.line = element_line(size = 1, colour = "black", linetype=1),
        axis.ticks = element_line(size = 0.5, color = "black"),
        panel.grid.major.y = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.major.x = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.spacing.x = unit(2, "lines"))  +
  scale_y_continuous(expand = c(0,0)) +
  facet_wrap(~comparison)

#Vaccination only -----
shared_genes_vaccines_only = shared_genes_vaccines %>% 
  anti_join(shared_genes_infection, by = "genes") %>% 
  inner_join(immunego_btm_genes %>% select(genes), by = "genes") %>% 
  select(genes, n_conditions) %>% 
  mutate(comparison = "Vaccination only") %>% 
  arrange(desc(n_conditions)) %>% 
  distinct()

shared_genes_vaccines_only_plot = shared_genes_vaccines_only %>% 
  select(genes, n_conditions, comparison) %>% 
  distinct() %>%
  slice_head(n=20) %>% 
  ggplot() +
  aes(x = reorder(genes, -n_conditions), 
      fill = comparison, 
      weight = n_conditions) +
  geom_bar(color = "black",
           size = 0.2,
           width = 0.7) +
  scale_fill_manual(
    values = c("Vaccination only" = "#4DBBD5FF",
              "Infection only" = "#DC0000FF")) +
  theme_minimal() +
  labs(fill = "Vaccination only") +
  xlab("") + 
  ylab("Number of shared conditions") +
  theme(legend.position = "none",
        strip.text.x = element_text(color = "black", size = 12), 
        axis.text.x = element_text(color = "black", size = 10, angle = 45, vjust = 1, hjust = 1),
        axis.text.y = element_text(vjust = 0.5, size = 10, color = "black"),
        axis.line = element_line(size = 1, colour = "black", linetype=1),
        axis.ticks = element_line(size = 0.5, color = "black"),
        panel.grid.major.y = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.major.x = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.spacing.x = unit(2, "lines"))  +
  scale_y_continuous(expand = c(0,0)) +
  facet_wrap(~comparison)


# Infection only
shared_genes_infection_only = shared_genes_infection %>% 
  anti_join(shared_genes_vaccines_only, by = "genes") %>%
  inner_join(ImmuneGO_Annotated_Genes, by = "genes") %>% 
  select(genes, n_conditions) %>% 
  mutate(comparison = "Infection only") %>% 
  arrange(desc(n_conditions))

shared_genes_infection_only_plot = shared_genes_infection_only %>% 
  select(genes, n_conditions, comparison) %>% 
  arrange(desc(n_conditions)) %>% 
  distinct() %>%
  slice_head(n=20) %>% 
  ggplot() +
  aes(x = reorder(genes, -n_conditions), 
      fill = comparison, 
      weight = n_conditions) +
  geom_bar(color = "black",
           size = 0.2,
           width = 0.7) +
  scale_fill_manual(
    values = c("Vaccination only" = "#4DBBD5FF",
              "Infection only" = "#DC0000FF")) +
  theme_minimal() +
  labs(fill = "Infection only") +
  xlab("Genes") + 
  ylab("") +
  theme(legend.position = "none",
        strip.text.x = element_text(color = "black", size = 12), 
        axis.text.x = element_text(color = "black", size = 10, angle = 45, vjust = 1, hjust = 1),
        axis.text.y = element_text(vjust = 0.5, size = 10, color = "black"),
        axis.line = element_line(size = 1, colour = "black", linetype=1),
        axis.ticks = element_line(size = 0.5, color = "black"),
        panel.grid.major.y = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.major.x = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.spacing.x = unit(2, "lines"))  +
  scale_y_continuous(expand = c(0,0)) +
  facet_wrap(~comparison)



patch = shared_genes_vaccines_infection_plot / shared_genes_vaccines_only_plot / shared_genes_infection_only_plot + plot_layout(guides = 'collect') +
  plot_annotation(
  title = "Top 20 shared immune-related genes",
  subtitle = "Infection and vaccination, Vaccination only, Infection only ")

ggsave(patch, file = here("Figures", "Genes_shared_Vaccination_Infection_united.png"), width = 10, height = 9)

patch
```






Vertical
```{r fig.height=5, fig.width=10}
# Vaccine and infection ----
shared_genes_infection_vaccination = shared_genes_comparisons %>% 
  filter(comparison == "Infection and vaccination")

shared_genes_infection = shared_genes_comparisons %>% 
  filter(comparison == "Infection only")

shared_genes_vaccines = shared_genes_comparisons %>% 
  filter(comparison == "Vaccination only")


shared_genes_vaccines_infection_plot = shared_genes_infection_vaccination %>% 
  select(genes, infection_vaccination, n_conditions, comparison) %>% 
  distinct() %>%
  slice_head(n=40) %>%  
  ggplot() +
  aes(
    x = n_conditions,
    y = reorder(genes, n_conditions),
    fill = infection_vaccination
  ) +
 geom_col(color = "black",
           size = 0.2,
           width = 0.6) +
  scale_fill_manual(
    values = c("Vaccination" = "#4DBBD5FF",
              "Infection" = "#DC0000FF")
    )+
  theme_minimal() +
  labs(fill = "Status") +
  xlab("") + 
  ylab("") +
   theme(legend.position = "top",
        strip.text.x = element_text(color = "black", size = 12), 
        axis.text.x = element_text(color = "black", size = 10),
        axis.text.y = element_text(vjust = 0.5, size = 10, color = "black"),
        axis.line = element_line(size = 1, colour = "black", linetype=1),
        axis.ticks = element_line(size = 0.5, color = "black"),
        panel.grid.major.y = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.major.x = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.spacing.x = unit(2, "lines"))  +
  scale_x_continuous(expand = c(0,0)) +
  facet_wrap(~comparison)

#Vaccination only -----
shared_genes_vaccines_only = shared_genes_vaccines %>% 
  anti_join(shared_genes_infection, by = "genes") %>% 
  inner_join(ImmuneGO_Annotated_Genes %>% select(genes), by = "genes") %>% 
  select(genes, n_conditions) %>% 
  mutate(comparison = "Vaccination only") %>% 
  arrange(desc(n_conditions)) %>% 
  distinct()

shared_genes_vaccines_only_plot = shared_genes_vaccines_only %>% 
  select(genes, n_conditions, comparison) %>% 
  distinct() %>%
  slice_head(n=20) %>% 
  ggplot() +
  aes(
    x = n_conditions,
    y = reorder(genes, n_conditions),
    fill = comparison
  ) +
  geom_col(color = "black",
           size = 0.2,
           width = 0.6) +
  scale_fill_manual(
    values = c("Vaccination only" = "#4DBBD5FF",
              "Infection only" = "#DC0000FF")
    )+
guides(fill = "none") +
  theme_minimal() +
  labs(fill = "Status") +
  xlab("") + 
  ylab("") +
   theme(legend.position = "none",
        strip.text.x = element_text(color = "black", size = 12), 
        axis.text.x = element_text(color = "black", size = 10),
        axis.text.y = element_text(vjust = 0.5, size = 10, color = "black"),
        axis.line = element_line(size = 1, colour = "black", linetype=1),
        axis.ticks = element_line(size = 0.5, color = "black"),
        panel.grid.major.y = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.major.x = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.spacing.x = unit(2, "lines"))  +
  scale_x_continuous(expand = c(0,0)) +
  facet_wrap(~comparison)


# Infection only
shared_genes_infection_only = shared_genes_infection %>% 
  anti_join(shared_genes_vaccines_only, by = "genes") %>%
  inner_join(immunego_btm_genes, by = "genes") %>% 
  select(genes, n_conditions) %>% 
  mutate(comparison = "Infection only") %>% 
  arrange(desc(n_conditions))

shared_genes_infection_only_plot = shared_genes_infection_only %>% 
  select(genes, n_conditions, comparison) %>% 
  arrange(desc(n_conditions)) %>% 
  distinct() %>%
  slice_head(n=20) %>% 
  ggplot() +
  aes(
    x = n_conditions,
    y = reorder(genes, n_conditions),
    fill = comparison
  ) +
  geom_col(color = "black",
           size = 0.2,
           width = 0.6) +
   scale_fill_manual(
    values = c("Vaccination only" = "#4DBBD5FF",
              "Infection only" = "#DC0000FF")
    )+
  guides(fill = "none") +
  theme_minimal() +
  labs(fill = "Status") +
  xlab("") + 
  ylab("") +
   theme(legend.position = "none",
         text = element_text(size = 10, color = "black"),
        strip.text.x = element_text(color = "black", size = 12), 
        axis.text.x = element_text(color = "black", size = 10),
        axis.text.y = element_text(vjust = 0.5, size = 10, color = "black"),
        axis.line = element_line(size = 1, colour = "black", linetype=1),
        axis.ticks = element_line(size = 0.5, color = "black"),
        panel.grid.major.y = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.major.x = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.spacing.x = unit(2, "lines"))  +
  scale_x_continuous(expand = c(0,0)) +
  facet_wrap(~comparison)



patch =  shared_genes_vaccines_only_plot + shared_genes_infection_only_plot + shared_genes_vaccines_infection_plot + plot_layout(guides = 'collect') +
  plot_annotation(
  title = "Top 20 shared immune-related genes",
  subtitle = "Infection and vaccination, Vaccination only, Infection only ") & theme(legend.position = "top")
  

ggsave(patch, file = here("Figures", "Genes_shared_Vaccination_Infection_united_vertical.png"), width = 10, height = 5)

patch
```

### Not-immune related only

```{r}
degs_all_updown = all_degs_p_05_vac_infected %>% 
  distinct()

## Processing -----

#Infection -----
shared_genes_infection = degs_all_updown %>%
  select(condition, genes) %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  filter(disease_vac == "I") %>% 
  select(condition, genes, vaccine) %>% 
  anti_join(immunego_btm_genes %>% select(genes) %>% distinct(), by = "genes") %>% 
  group_by(genes) %>% 
  summarise("Infection" = n()) %>% 
  arrange(desc("Infection"), genes) %>% 
  ungroup() %>% 
  pivot_longer(-genes, names_to = "infection_vaccination", values_to = "n_conditions") %>% 
  anti_join(immunego_btm_genes , by = "genes") %>% 
  select(genes, infection_vaccination,n_conditions) %>% 
  mutate(comparison = "Infection only") %>% 
  arrange(desc(n_conditions))

#Vaccine -----
shared_genes_vaccines = degs_all_updown %>%
  select(condition, genes) %>% 
  anti_join(immunego_btm_genes %>% select(genes) %>% distinct(), by = "genes") %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
    filter(disease_vac == "V") %>% 
  select(condition, genes, vaccine) %>% 
  group_by(genes) %>% 
  summarise("Vaccination" = n()) %>% 
  arrange(desc("Vaccination"), genes) %>% 
  ungroup() %>% 
  pivot_longer(-genes, names_to = "infection_vaccination", values_to = "n_conditions") %>% 
  anti_join(immunego_btm_genes %>% select(genes) %>% distinct(), by = "genes") %>% 
  select(genes, infection_vaccination,n_conditions) %>% 
  mutate(comparison = "Vaccination only") %>% 
  arrange(desc(n_conditions))


# Infection and vaccination -----
shared_genes_vaccines_infection = shared_genes_vaccines %>% 
  inner_join(shared_genes_infection, by = "genes") %>% 
  select(genes, Vaccination = n_conditions.x, Infection = n_conditions.y) %>% 
  pivot_longer(-genes, names_to = "infection_vaccination", values_to = "n_conditions") %>% 
  anti_join(immunego_btm_genes , by = "genes") %>% 
  select(genes, infection_vaccination,n_conditions) %>% 
  mutate(comparison = "Infection and vaccination") 

shared_genes_comparisons = bind_rows(shared_genes_infection, 
                                     shared_genes_vaccines, 
                                     shared_genes_vaccines_infection) %>% 
  arrange(desc(n_conditions))

write_csv(shared_genes_comparisons, file = here("DGE analysis", "non_Immune_shared_genes_infection_vaccination_comparisons.csv"))

shared_genes_comparisons
```

```{r fig.height=9, fig.width=10}
# Plotting ------
shared_genes_infection_vaccination = shared_genes_comparisons %>%
  filter(comparison == "Infection and vaccination")

shared_genes_infection = shared_genes_comparisons %>% 
  filter(comparison == "Infection only")

shared_genes_vaccines = shared_genes_comparisons %>% 
  filter(comparison == "Vaccination only")


#Infection and vaccination
shared_genes_vaccines_infection_plot = shared_genes_vaccines_infection %>% 
  select(genes, infection_vaccination, n_conditions, comparison) %>% 
  distinct() %>%
  slice_head(n=40) %>% 
  ggplot() +
  aes(
    x = reorder(genes, -n_conditions),
    y = n_conditions,
    fill = infection_vaccination
  ) +
  geom_col(color = "black",
           size = 0.2,
           width = 0.7) +
  scale_fill_manual(
    values = c("Vaccination" = "#4DBBD5FF",
              "Infection" = "#DC0000FF")) +
  theme_minimal() +
  labs(fill = "Status") +
  xlab("") + 
  ylab("") +
  theme(legend.position = "none",
        strip.text.x = element_text(color = "black", size = 12), 
        axis.text.x = element_text(color = "black", size = 10, angle = 45, hjust = 1, vjust = 1),
        axis.text.y = element_text(vjust = 0.5, size = 10, color = "black"),
        axis.line = element_line(size = 1, colour = "black", linetype=1),
        axis.ticks = element_line(size = 0.5, color = "black"),
        panel.grid.major.y = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.major.x = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.spacing.x = unit(2, "lines"))  +
  scale_y_continuous(expand = c(0,0)) +
  facet_wrap(~comparison)

#Vaccination only -----

shared_genes_vaccines_only_plot = shared_genes_vaccines %>% 
  select(genes, infection_vaccination, n_conditions, comparison) %>% 
  arrange(desc(n_conditions)) %>% 
  distinct() %>%
  slice_head(n=20) %>% 
  ggplot() +
  aes(x = reorder(genes, -n_conditions), 
      fill = comparison, 
      weight = n_conditions) +
  geom_bar(color = "black",
           size = 0.2,
           width = 0.7) +
  scale_fill_manual(
    values = c("Vaccination only" = "#4DBBD5FF",
              "Infection only" = "#DC0000FF")) +
  theme_minimal() +
  labs(fill = "Infection only") +
  xlab("") + 
  ylab("Number of shared conditions") +
  theme(legend.position = "none",
        strip.text.x = element_text(color = "black", size = 12), 
        axis.text.x = element_text(color = "black", size = 10, angle = 45, hjust = 1, vjust = 1),
        axis.text.y = element_text(vjust = 0.5, size = 10, color = "black"),
        axis.line = element_line(size = 1, colour = "black", linetype=1),
        axis.ticks = element_line(size = 0.5, color = "black"),
        panel.grid.major.y = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.major.x = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.spacing.x = unit(2, "lines"))  +
  scale_y_continuous(expand = c(0,0)) +
  facet_wrap(~comparison)

# Infection only -------

shared_genes_infection_only_plot = shared_genes_infection %>% 
  select(genes, infection_vaccination, n_conditions, comparison) %>% 
  arrange(desc(n_conditions)) %>% 
  distinct() %>%
  slice_head(n=20) %>% 
  ggplot() +
  aes(x = reorder(genes, -n_conditions), 
      fill = comparison, 
      weight = n_conditions) +
  geom_bar(color = "black",
           size = 0.2,
           width = 0.7) +
  scale_fill_manual(
    values = c("Vaccination only" = "#4DBBD5FF",
              "Infection only" = "#DC0000FF")) +
  theme_minimal() +
  labs(fill = "Infection only") +
  xlab("Genes") + 
  ylab("") +
  theme(legend.position = "none",
        strip.text.x = element_text(color = "black", size = 12), 
        axis.text.x = element_text(color = "black", size = 10, angle = 45, hjust = 1, vjust = 1),
        axis.text.y = element_text(vjust = 0.5, size = 10, color = "black"),
        axis.line = element_line(size = 1, colour = "black", linetype=1),
        axis.ticks = element_line(size = 0.5, color = "black"),
        panel.grid.major.y = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.major.x = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.spacing.x = unit(2, "lines"))  +
  scale_y_continuous(expand = c(0,0)) +
  facet_wrap(~comparison)

patch = shared_genes_vaccines_infection_plot / shared_genes_vaccines_only_plot / shared_genes_infection_only_plot + plot_layout(guides = 'collect') +
  plot_annotation(
  title = "Top 20 shared non-immune-related genes",
  subtitle = "Infection and vaccination, Vaccination only, Infection only ")


ggsave(patch, file =here("Figures", "NonImmune_Genes_shared_Vaccination_Infection_united.png"), width = 10, height = 9)

patch
```





## Overlapping genes with Immune genes

Input

```{r}
# Gene set data.
ImmuneGO_Annotated_Genes_all = ImmuneGO_Annotated_Genes
ImmuneGO_Annotated_Genes = ImmuneGO_Annotated_Genes_all %>% 
  filter(!process %in% c("TCR REPERTOIRE", "BCR REPERTOIRE"))
ImmuneGO_BTM_grouped_genes <- read_csv("VaxGO gene sets/ImmuneGO_BTM_grouped_genes.csv")

# DEGs table
degs_all_updown = read_rds(file = here("DGE analysis", "all_studies_degs_weighted.rds")) %>% 
  filter(if_else(n >= 5, fdr <= 0.05, fdr <= (0.05 * 6/n))) %>% 
  # filter(direction != "NEUTRAL") %>% 
  # mutate_all(~ replace(., is.na(.), "NA")) %>% 
  # filter(genes %in% btm_annotation_genes$genes) 
  filter(genes %in% ImmuneGO_BTM_grouped_genes$genes)


ann_vaccines = read_csv(here("Annotations and metadata", "ann_vaccines_conditions.csv"))

data = degs_all_updown %>% 
 dplyr::select(process = condition, genes)

filename = "All_covid_ImmuneGO_BTM"
```

Function

```{r}
# Function for overlapping -------
overlap_genes <- function(cond1, cond2, data) {
  genes_cond1 <- data$genes[data$process == cond1]
  genes_cond2 <- data$genes[data$process == cond2]
  
  genes_shared <- intersect(genes_cond1, genes_cond2)
  
  genes_notshared_cond1 <- setdiff(genes_cond1, genes_cond2)
  genes_notshared_cond2 <- setdiff(genes_cond2, genes_cond1)
  
  total_genes_cond1 <- length(genes_cond1)
  total_genes_cond2 <- length(genes_cond2)
  
  percentage_shared_cond1 <- length(genes_shared) / total_genes_cond1 * 100
  percentage_shared_cond2 <- length(genes_shared) / total_genes_cond2 * 100
  
  shared_genes <- data.frame(
    Cond1 = cond1,
    Cond2 = cond2,
    Shared = length(genes_shared),
    NotShared_cond1 = length(genes_notshared_cond1),
    NotShared_cond2 = length(genes_notshared_cond2),
    Total_Genes_Cond1 = total_genes_cond1,
    Total_Genes_Cond2 = total_genes_cond2,
    Genes_Names = paste(genes_shared, collapse = ", "),
    Percentage_Shared_Cond1 = percentage_shared_cond1,
    Percentage_Shared_Cond2 = percentage_shared_cond2
  )
  
  return(shared_genes)
}
```

Perform calculations

```{r}

conflicted::conflicts_prefer(base::intersect)
conflicted::conflicts_prefer(base::setdiff)
conflicted::conflicts_prefer(stats::fisher.test)

# Perform calculations
unique_cond <- unique(data$process) # List sets
shared_genes_df <- data.frame() # Store data

for (i in 1:(length(unique_cond) - 1)) {
  for (j in (i + 1):length(unique_cond)) {
    resultado_temp <- overlap_genes(unique_cond[i], unique_cond[j], data)
    shared_genes_df <- bind_rows(shared_genes_df, resultado_temp)
  }
}

# Fisher's exact test
fisher_exact_test <- function(shared, notshared1, notshared2) {
  cont_table <- matrix(c(shared, notshared1, notshared2, 0), nrow = 2)
  results_fisher <- fisher.test(cont_table)
  return(results_fisher$p.value)
}
shared_genes_df = shared_genes_df %>%
  mutate(pvalue = pmap_dbl(list(Shared, NotShared_cond1, NotShared_cond2), fisher_exact_test)) %>% 
  mutate("-log(p)" = -log10(pvalue))


#Mirror table
shared_genes_df2 = shared_genes_df %>% 
  rename(Cond1_temp = Cond1, Cond2_temp = Cond2) %>%
  rename(Cond1 = Cond2_temp, Cond2 = Cond1_temp)

shared_genes_df_mirror = shared_genes_df %>% 
  bind_rows(shared_genes_df2) %>% 
  mutate(Percentage_Shared_Cond1 = round(Percentage_Shared_Cond1, 2),
         Percentage_Shared_Cond2 = round(Percentage_Shared_Cond2, 2))
  
#Jaccard distance
jaccard.matrix <- data %>%
  mutate(present = 1) %>% 
  distinct() %>% 
  pivot_wider(names_from = genes, values_from = present, values_fill = 0) %>%
  column_to_rownames(var = "process") %>%
  vegdist(binary = TRUE, method = "jaccard", diag = TRUE, upper = TRUE) %>%
  as.matrix() %>%
  {1 - .}

shared_genes_df_mirror = jaccard.matrix %>% 
  as.data.frame() %>% 
  rownames_to_column("Cond1") %>% 
  pivot_longer(cols = -"Cond1",
               names_to = "Cond2",
               values_to = "jaccard_distance") %>% 
  inner_join(shared_genes_df_mirror, by = join_by(Cond1, Cond2)) %>% 
  mutate(jaccard_distance = round(jaccard_distance, 3))

#Save
write_csv(shared_genes_df_mirror, 
          file = here("DGE analysis", paste0(filename, "_sharedgenes_Fisher_Jaccard.csv")))

```

### Heatmap

#### Shared genes

```{r}

filename = "All_covid_ImmuneGO_BTM"

shared_genes_df_mirror = read_csv(file = here("DGE analysis", "All_covid_ImmuneGO_BTM_sharedgenes_Fisher_Jaccard.csv"))

matrix_data = shared_genes_df_mirror %>% 
  dcast(Cond1 ~ Cond2, 
        value.var = "Shared", 
        fun.aggregate = mean) %>% 
  column_to_rownames("Cond1") %>%
  mutate_if(is.character, as.numeric) %>% 
  as.matrix() 

dim(matrix_data)

ann_vaccines_complex = matrix_data %>% 
  as.data.frame() %>%
  rownames_to_column("condition") %>% 
  select(condition) %>% 
  inner_join(ann_vaccines, by = "condition") %>%
  select(condition, vaccine:regimen, week) %>% 
  distinct() 

# merge
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "process") %>% 
  rename(condition = '.') %>% 
  inner_join(ann_vaccines_complex, by = "condition") %>% 
  arrange(condition) %>% 
  relocate(dose, .before=vaccine) %>% 
  column_to_rownames(var= "condition")  %>% 
  mutate(dose = ifelse(is.na(dose), "NA", dose))

# Reorder
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% 
  select(!c(vaccine:regimen, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% 
  as.matrix()

matrix_data_ordered_rows = matrix_data_ordered %>% 
  as.data.frame() %>% 
  rownames_to_column("condition") %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") %>% 
  as.matrix()

#Check
dim(matrix_data_ordered_rows) 
dim(ann_vaccines_heatmap)

col_fun <- colorRamp2(c(0, 500), c("white", "#d90429"))


######## HEATMAP ANNOTATION
ann_vaccines_heatmap = ann_vaccines_heatmap %>% 
  select(type, vaccine, day) %>% 
  mutate(day = as_factor(day),
         day = fct_reorder(day, as.numeric(day)))

ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_vaxsig_colors, 
                       annotation_name_side = "left",
                       show_legend = F)

row_ha = rowAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_vaxsig_colors)



mat = matrix_data_ordered_rows
width_image = 40
height_image = 50
width = unit(20, "cm")
height = unit(20, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)
gp = gpar(fontsize = 5)
rect_gp = gpar(col = "gray1", lwd = 0)



fileimageheatmap_grouped = paste0(filename, sep ="_", "Shared_Heatmap.png")

png(
  file = here("Figures", fileimageheatmap_grouped),
  width = width_image,
  height = height_image,
  units = "cm",
  res = 900
)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "Jaccard distance", #Legend
                        col = col_fun, #color
                         cell_fun = function(j, i, x, y, width, height, fill) {
  if (!is.na(mat[i, j]) && mat[i, j] > 0) {
    grid.text(sprintf("%.f", mat[i, j]), x, y, gp = gpar(fontsize = 5))
  }
},
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 90,
                        column_names_gp = column_names_gp,
                        cluster_rows = TRUE,
                        rect_gp = rect_gp,
                        row_names_gp = row_names_gp,
                        # row_split = 2, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        na_col = "black",
                        column_split = NULL, #Split group clusters
                        #column_km = 3,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "left")

dev.off()
```

#### Jaccard distance

```{r}
matrix_data = shared_genes_df_mirror %>% 
  dcast(Cond1 ~ Cond2, 
        value.var = "jaccard_distance", 
        fun.aggregate = mean) %>% 
  column_to_rownames("Cond1") %>%
  mutate_if(is.character, as.numeric) %>% 
  as.matrix() 

ann_vaccines_complex = matrix_data %>% 
  as.data.frame() %>%
  rownames_to_column("condition") %>% 
  select(condition) %>% 
  inner_join(ann_vaccines, by = "condition") %>%
  select(condition, vaccine:regimen, week) %>% 
  distinct() 


ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "process") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  relocate(dose, .before=vaccine) %>% 
  column_to_rownames(var= "condition")  %>% 
  mutate(dose = ifelse(is.na(dose), "NA", dose))

matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% 
  select(!c(vaccine:regimen, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% 
  as.matrix()

matrix_data_ordered_rows = matrix_data_ordered %>% 
  as.data.frame() %>% 
  rownames_to_column("condition") %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") %>% 
  as.matrix()

dim(matrix_data_ordered)
dim(ann_vaccines_heatmap)

# Quantile interpolation
quartiles <- quantile(matrix_data_ordered, probs = c(0, 0.25, 0.5, 0.75, 1), na.rm = TRUE)
# col_fun <- colorRamp2(quartiles, c("white", "#a8dadc", "#457b9d", "#d90429", "#1d3557"))
col_fun <- colorRamp2(quartiles, c("white", "#a8dadc", "#457b9d", "#EFC000FF", "#d90429"))
# col_fun <- colorRamp2(quartiles, c("white", "#E8E79AFF", "#8CBF9AFF", "#477B95FF", "#011C40FF"))


######## HEATMAP ANNOTATION
ann_vaccines_heatmap = ann_vaccines_heatmap %>% 
  select(type, vaccine, day) %>% 
  mutate(day = as_factor(day),
         day = fct_reorder(day, as.numeric(day)))

ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_vaxsig_colors, 
                       annotation_name_side = "left",
                       show_legend = F)

row_ha = HeatmapAnnotation(df = ann_vaccines_heatmap,
                            which = "row",
                       col = ann_vaxsig_colors, 
                       show_annotation_name = F,
                       show_legend = F)



mat = matrix_data_ordered_rows
width_image = 40
height_image = 40
width = unit(20, "cm")
height = unit(20, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)
gp = gpar(fontsize = 5)
rect_gp = gpar(col = "white", lwd = 2)


fileimageheatmap_grouped = paste0(filename, sep ="_", "Jaccard_Heatmap.png")

png(
  file = here("Figures", fileimageheatmap_grouped),
  width = width_image,
  height = height_image,
  units = "cm",
  res = 900
)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "Jaccard \ndistance", #Legend
                        col = col_fun, #color
                         cell_fun = function(j, i, x, y, width, height, fill) {
  if (!is.na(mat[i, j]) && mat[i, j] > 0.01) {
    grid.text(sprintf("%.2f", mat[i, j]), x, y, gp = gpar(fontsize = 5))
  }
},
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        cluster_rows = TRUE,
                        rect_gp = rect_gp,
                        row_names_gp = row_names_gp,
                        # row_split = 2, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        na_col = "black",
                        column_split = NULL, #Split group clusters
                        #column_km = 3,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "left")

dev.off()
```


Ordered by day
```{r}
# INPUT ------

matrix_data = shared_genes_df_mirror %>% 
  dcast(Cond1 ~ Cond2, 
        value.var = "jaccard_distance", 
        fun.aggregate = mean) %>% 
  column_to_rownames("Cond1") %>%
  mutate_if(is.character, as.numeric) %>% 
  as.matrix() 

ann_vaccines_complex = matrix_data %>% 
  as.data.frame() %>%
  rownames_to_column("condition") %>% 
  select(condition) %>% 
  inner_join(ann_vaccines, by = "condition") %>%
  select(condition, vaccine:regimen, day, week) %>% 
  distinct() 


ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "process") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  relocate(dose, .before=vaccine) %>% 
  column_to_rownames(var= "condition")  %>% 
  mutate(dose = ifelse(is.na(dose), "NA", dose)) %>% 
  arrange(day)

matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(day) %>% 
  select(!c(vaccine:regimen, day, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% 
  as.matrix()

matrix_data_ordered_rows = matrix_data_ordered %>% 
  as.data.frame() %>% 
  rownames_to_column("condition") %>% 
  filter(condition %in% c(matrix_data_ordered %>% colnames())) %>% 
  arrange(match(condition, c(matrix_data_ordered %>% colnames()))) %>% 
  column_to_rownames(var= "condition") %>% 
  as.matrix()



dim(matrix_data_ordered)
dim(ann_vaccines_heatmap)

# Quantile interpolation
quartiles <- quantile(matrix_data_ordered, probs = c(0, 0.25, 0.5, 0.75, 1), na.rm = TRUE)
# col_fun <- colorRamp2(quartiles, c("white", "#a8dadc", "#457b9d", "#d90429", "#1d3557"))
col_fun <- colorRamp2(quartiles, c("white", "#a8dadc", "#457b9d", "#EFC000FF", "#d90429"))
# col_fun <- colorRamp2(quartiles, c("white", "#E8E79AFF", "#8CBF9AFF", "#477B95FF", "#011C40FF"))



######## HEATMAP ANNOTATION
ann_vaccines_heatmap = ann_vaccines_heatmap %>% 
  select(type, vaccine, day) %>% 
  mutate(day = as_factor(day),
         day = fct_reorder(day, as.numeric(day)))

ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_vaxsig_colors, 
                       annotation_name_side = "left",
                       show_legend = F)

row_ha = HeatmapAnnotation(df = ann_vaccines_heatmap,
                            which = "row",
                       col = ann_vaxsig_colors, 
                       show_annotation_name = F,
                       show_legend = F)



mat = matrix_data_ordered_rows
width_image = 40
height_image = 40
width = unit(20, "cm")
height = unit(20, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)
gp = gpar(fontsize = 5)
rect_gp = gpar(col = "white", lwd = 2)


fileimageheatmap_grouped = paste0(filename, sep ="_", "Jaccard_Heatmap_ByDay.png")

png(
  file = here("Figures", fileimageheatmap_grouped),
  width = width_image,
  height = height_image,
  units = "cm",
  res = 900
)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "Jaccard \ndistance", #Legend
                        col = col_fun, #color
                         cell_fun = function(j, i, x, y, width, height, fill) {
  if (!is.na(mat[i, j]) && mat[i, j] > 0.01) {
    grid.text(sprintf("%.2f", mat[i, j]), x, y, gp = gpar(fontsize = 5))
  }
},
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        cluster_rows = F,
                        rect_gp = rect_gp,
                        row_names_gp = row_names_gp,
                        # row_split = 2, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = F,
                        na_col = "black",
                        column_split = NULL, #Split group clusters
                        #column_km = 3,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "left")

dev.off()

```

Ordered by condition
```{r}
# INPUT ------

matrix_data = shared_genes_df_mirror %>% 
  dcast(Cond1 ~ Cond2, 
        value.var = "jaccard_distance", 
        fun.aggregate = mean) %>% 
  column_to_rownames("Cond1") %>%
  mutate_if(is.character, as.numeric) %>% 
  as.matrix() 

ann_vaccines_complex = matrix_data %>% 
  as.data.frame() %>%
  rownames_to_column("condition") %>% 
  select(condition) %>% 
  inner_join(ann_vaccines, by = "condition") %>%
  select(condition, vaccine:regimen, day, week) %>% 
  distinct() 


ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "process") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(match(condition, c(ann_vaccines_conditions_ordered$condition_grouped %>% as.character()))) %>% 
  relocate(dose, .before=vaccine) %>% 
  column_to_rownames(var= "condition")  %>% 
  mutate(dose = ifelse(is.na(dose), "NA", dose)) 

matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(match(condition, c(ann_vaccines_conditions_ordered$condition_grouped %>% as.character()))) %>% 
  select(!c(vaccine:regimen, day, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% 
  as.matrix()

matrix_data_ordered_rows = matrix_data_ordered %>% 
  as.data.frame() %>% 
  rownames_to_column("condition") %>% 
  arrange(match(condition, c(ann_vaccines_conditions_ordered$condition_grouped %>% as.character()))) %>% 
  column_to_rownames(var= "condition") %>% 
  as.matrix()



dim(matrix_data_ordered)
dim(ann_vaccines_heatmap)

# Quantile interpolation
quartiles <- quantile(matrix_data_ordered, probs = c(0, 0.25, 0.5, 0.75, 1), na.rm = TRUE)
# col_fun <- colorRamp2(quartiles, c("white", "#a8dadc", "#457b9d", "#d90429", "#1d3557"))
col_fun <- colorRamp2(quartiles, c("white", "#a8dadc", "#457b9d", "#EFC000FF", "#d90429"))
# col_fun <- colorRamp2(quartiles, c("white", "#E8E79AFF", "#8CBF9AFF", "#477B95FF", "#011C40FF"))



######## HEATMAP ANNOTATION
ann_vaccines_heatmap = ann_vaccines_heatmap %>% 
  select(type, vaccine, day) %>% 
  mutate(day = as_factor(day),
         day = fct_reorder(day, as.numeric(day)))

ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_vaxsig_colors, 
                       annotation_name_side = "left",
                       show_legend = F)

row_ha = HeatmapAnnotation(df = ann_vaccines_heatmap,
                            which = "row",
                       col = ann_vaxsig_colors, 
                       show_annotation_name = F,
                       show_legend = F)



mat = matrix_data_ordered_rows
width_image = 40
height_image = 40
width = unit(20, "cm")
height = unit(20, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)
gp = gpar(fontsize = 5)
rect_gp = gpar(col = "white", lwd = 2)


fileimageheatmap_grouped = paste0(filename, sep ="_", "Jaccard_Heatmap_ByCondition.png")

png(
  file = here("Figures", fileimageheatmap_grouped),
  width = width_image,
  height = height_image,
  units = "cm",
  res = 900
)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "Jaccard \ndistance", #Legend
                        col = col_fun, #color
                         cell_fun = function(j, i, x, y, width, height, fill) {
  if (!is.na(mat[i, j]) && mat[i, j] > 0.01) {
    grid.text(sprintf("%.2f", mat[i, j]), x, y, gp = gpar(fontsize = 5))
  }
},
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        cluster_rows = F,
                        rect_gp = rect_gp,
                        row_names_gp = row_names_gp,
                        # row_split = 2, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = F,
                        na_col = "black",
                        column_split = NULL, #Split group clusters
                        #column_km = 3,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "left")

dev.off()

```
Ordered by condition and clustered by cols
```{r}
# INPUT ------

matrix_data = shared_genes_df_mirror %>% 
  dcast(Cond1 ~ Cond2, 
        value.var = "jaccard_distance", 
        fun.aggregate = mean) %>% 
  column_to_rownames("Cond1") %>%
  mutate_if(is.character, as.numeric) %>% 
  as.matrix() 

ann_vaccines_complex = matrix_data %>% 
  as.data.frame() %>%
  rownames_to_column("condition") %>% 
  select(condition) %>% 
  inner_join(ann_vaccines, by = "condition") %>%
  select(condition, vaccine:regimen, day, week) %>% 
  distinct() 


ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "process") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(match(condition, c(ann_vaccines_conditions_ordered$condition_grouped %>% as.character()))) %>% 
  relocate(dose, .before=vaccine) %>% 
  column_to_rownames(var= "condition")  %>% 
  mutate(dose = ifelse(is.na(dose), "NA", dose)) 

matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(match(condition, c(ann_vaccines_conditions_ordered$condition_grouped %>% as.character()))) %>% 
  select(!c(vaccine:regimen, day, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% 
  as.matrix()

matrix_data_ordered_rows = matrix_data_ordered %>% 
  as.data.frame() %>% 
  rownames_to_column("condition") %>% 
  arrange(match(condition, c(ann_vaccines_conditions_ordered$condition_grouped %>% as.character()))) %>% 
  column_to_rownames(var= "condition") %>% 
  as.matrix()



dim(matrix_data_ordered)
dim(ann_vaccines_heatmap)

# Quantile interpolation
quartiles <- quantile(matrix_data_ordered, probs = c(0, 0.25, 0.5, 0.75, 1), na.rm = TRUE)
# col_fun <- colorRamp2(quartiles, c("white", "#a8dadc", "#457b9d", "#d90429", "#1d3557"))
col_fun <- colorRamp2(quartiles, c("white", "#a8dadc", "#457b9d", "#EFC000FF", "#d90429"))
# col_fun <- colorRamp2(quartiles, c("white", "#E8E79AFF", "#8CBF9AFF", "#477B95FF", "#011C40FF"))



######## HEATMAP ANNOTATION
ann_vaccines_heatmap = ann_vaccines_heatmap %>% 
  select(type, vaccine, day) %>% 
  mutate(day = as_factor(day),
         day = fct_reorder(day, as.numeric(day)))

ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_vaxsig_colors, 
                       annotation_name_side = "left",
                       show_legend = F)

row_ha = HeatmapAnnotation(df = ann_vaccines_heatmap,
                            which = "row",
                       col = ann_vaxsig_colors, 
                       show_annotation_name = F,
                       show_legend = F)



mat = matrix_data_ordered_rows
width_image = 40
height_image = 40
width = unit(20, "cm")
height = unit(20, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)
gp = gpar(fontsize = 5)
rect_gp = gpar(col = "white", lwd = 2)



fileimageheatmap_grouped = paste0(filename, sep ="_", "Jaccard_Heatmap_ByCondition_Colclustered.png")

png(
  file = here("Figures", fileimageheatmap_grouped),
  width = width_image,
  height = height_image,
  units = "cm",
  res = 900
)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "Jaccard \ndistance", #Legend
                        col = col_fun, #color
                         cell_fun = function(j, i, x, y, width, height, fill) {
  if (!is.na(mat[i, j]) && mat[i, j] > 0.01) {
    grid.text(sprintf("%.2f", mat[i, j]), x, y, gp = gpar(fontsize = 5))
  }
},
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        cluster_rows = F,
                        rect_gp = rect_gp,
                        row_names_gp = row_names_gp,
                        # row_split = 2, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = T,
                        na_col = "black",
                        column_split = NULL, #Split group clusters
                        #column_km = 3,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "left")

dev.off()

```


### Chord diagram

#### Shared

##### All conditions (Interactive)

```{r}
# library(chorddiag)

matrix_vaccines <- shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  dplyr::select(Cond1, Cond2, Shared) %>% 
  pivot_wider(names_from = "Cond2", values_from = "Shared", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

types = shared_genes_df_mirror %>% 
  dplyr::select(condition = Cond1) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% 
 dplyr::select(condition, type), by = "condition") %>% 
  dplyr::select(type)

colors = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(ann_vaccines %>% 
 dplyr::select(condition, type), by = "type") %>% 
  inner_join(matrix_vaccines %>% 
               as.data.frame() %>% 
               colnames() %>% as.data.frame() %>% rename(condition = "."),
             by = "condition") %>% 
  pull(color)

# Build the chord diagram:
all_shared = chorddiag(matrix_vaccines,
          groupColors = colors,
          type = "directional",
          width = NULL,
          height = NULL,
          margin = 100,
          showGroupnames = TRUE,
          groupThickness = 0.1,
          groupPadding = 2,
          groupnamePadding = 1,
          groupnameFontsize = 10,
          groupedgeColor = NULL,
          chordedgeColor = NULL,
          categoryNames = NULL,
          categorynamePadding = 100,
          categorynameFontsize = 28,
          showTicks = TRUE,
          tickInterval = NULL,
          ticklabelFontsize = 10,
          fadeLevel = 0.1,
          showTooltips = TRUE,
          showZeroTooltips = TRUE,
          tooltipNames = NULL,
          tooltipUnit = NULL,
          tooltipFontsize = 12,
          tooltipGroupConnector = " &#x25B6; ",
          precision = NULL,
          clickAction = NULL,
          clickGroupAction = NULL
)
all_shared
```

##### Weeks

Prepare tables

```{r fig.height=5, fig.width=5}
#Week 1 ------
links_df_W1 = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(ann_vaccines %>% 
               filter(week == 1) %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(ann_vaccines %>% 
               filter(week == 1) %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, Shared)

mat_week1 = links_df_W1 %>% 
  pivot_wider(names_from = "Cond2", values_from = "Shared", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()
grid.col_w1 = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(ann_vaccines %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_W1,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)
group_w1 = ann_vaccines %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_W1,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)


#Week 2 ------
links_df_W2 = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(ann_vaccines %>% 
               filter(week == 2) %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(ann_vaccines %>% 
               filter(week == 2) %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, Shared)

mat_week2 = links_df_W2 %>% 
  pivot_wider(names_from = "Cond2", values_from = "Shared", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_w2 = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(ann_vaccines %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_W2,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_w2 = ann_vaccines %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_W2,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)


#Week 4 ------
links_df_W4 = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(ann_vaccines %>% 
               filter(week == 4) %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(ann_vaccines %>% 
               filter(week == 4) %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, Shared)

mat_week4 = links_df_W4 %>% 
  pivot_wider(names_from = "Cond2", values_from = "Shared", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_w4 = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(ann_vaccines %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_W4,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_w4 = ann_vaccines %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_W4,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)

#Week 7 ------
links_df_W7 = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(ann_vaccines %>% 
               filter(week >= 7) %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(ann_vaccines %>% 
               filter(week == 7) %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, Shared)

mat_week7 = links_df_W7 %>% 
  pivot_wider(names_from = "Cond2", values_from = "Shared", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_w7 = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(ann_vaccines %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_W7,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_w7 = ann_vaccines %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_W7,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)

```

Plot

```{r fig.height= 5, fig.width=10}
#Plot
png(filename = here("Figures","ChordDiagram_Covid19_Weeks_1-2-4-7.png"), 
    width = 12, height = 14, units = "in", res = 300)

par(mfrow = c(2, 2))
#WEEK 1
chordDiagram(mat_week1, 
             grid.col = grid.col_w1, 
             annotationTrack = "grid", 
             directional = -1,
             big.gap = 6,
             transparency = 0,
             small.gap = 3,
             link.zindex = rank(mat_week1),
             h.ratio = 0.7, 
             group = group_w1,
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_week1))))))
#Labels
circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.8, #Label size
      font = 0.1)},#Label face
  bg.border = NA) 
title("Week 1")

# WEEK 2
chordDiagram(mat_week2, grid.col = grid.col_w2, 
             annotationTrack = "grid", 
            big.gap = 50,
             group = group_w2,
             small.gap = 3,
             link.zindex = rank(mat_week2),
             h.ratio = 0.7, 
             directional = -1,
             transparency = 0,
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_week2))))))
#Labels
circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.8, #Label size
      font = 0.1)},
  bg.border = NA)

title("Week 2")


# WEEK 4
chordDiagram(mat_week4, grid.col = grid.col_w4, 
             annotationTrack = "grid", 
            big.gap = 50,
             group = group_w4,
             small.gap = 3,
             link.zindex = rank(mat_week4),
             h.ratio = 0.7, 
             directional = -1,
             transparency = 0,
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_week4))))))
#Labels
circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.8, #Label size
      font = 0.1)},
  bg.border = NA)

title("Week 4")

# WEEK 7
chordDiagram(mat_week7, grid.col = grid.col_w7,
             annotationTrack = "grid",
            big.gap = 50,
             group = group_w7,
             small.gap = 3,
             link.zindex = rank(mat_week7),
             h.ratio = 0.7,
             directional = -1,
             transparency = 0,
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_week7))))))
#Labels
circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.8, #Label size
      font = 0.1)},
  bg.border = NA)

title("Week 8")

dev.off()
circos.clear()

```

##### Types

Prepare tables

```{r fig.height= 10, fig.width=10}
#Types
###### All -----
links_df = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  dplyr::select(Cond1, Cond2, Shared)

mat_types = links_df %>% 
  pivot_wider(names_from = "Cond2", values_from = "Shared", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_types = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(ann_vaccines %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_types = ann_vaccines %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)

# Vaccination only -------
links_df_vac = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(ann_vaccines %>% 
               filter(disease_vac == "V") %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(ann_vaccines %>% 
               filter(disease_vac == "V") %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, Shared)

mat_types_vac = links_df_vac %>% 
  pivot_wider(names_from = "Cond2", values_from = "Shared", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_types_vac = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(ann_vaccines %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_vac,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_types_vac = ann_vaccines %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_vac,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)

#Infection only
links_df_inf = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(ann_vaccines %>% 
               filter(disease_vac == "I") %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(ann_vaccines %>% 
               filter(disease_vac == "I") %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, Shared)

mat_types_inf = links_df_inf %>% 
  pivot_wider(names_from = "Cond2", values_from = "Shared", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_types_inf = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(ann_vaccines %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_inf,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_types_inf = ann_vaccines %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_inf,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)
```

Plot All conditions and stats

```{r fig.height= 5, fig.width=5}
#Plot

#Types

png(filename = here("Figures","ChordDiagram_Covid19_Types_All.png"), 
    width = 5, height = 5, units = "in", res = 300)

#Infection and vaccination
chordDiagram(mat_types, grid.col = grid.col_types,
             annotationTrack = "grid",  transparency = 0,
             big.gap = 10,
             small.gap = 3,
             directional = -1, 
             group = group_types,
             link.zindex = rank(mat_types),
              h.ratio = 0.7, 
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_types))))))

circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.5, #Label size
      font = 0.1
      )},
  bg.border = NA)
title("Infection and Vaccination")

dev.off()
circos.clear()
```

By state

```{r fig.height= 5, fig.width=15}

png(filename = here("Figures","ChordDiagram_Covid19_Types_Inf_Vac.png"), 
    width = 6, height = 12, units = "in", res = 300)

par(mfrow = c(2, 1))
#Vaccination
chordDiagram(mat_types_vac, grid.col = grid.col_types_vac, 
             annotationTrack = "grid", transparency = 0,
             big.gap = 6,
             small.gap = 3,
                          directional = -1,
             group = group_types,
             link.zindex = rank(mat_types_vac),
             h.ratio = 0.7, 
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_types_vac))))))

circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.5, #Label size
      font = 0.1
      )},
  bg.border = NA) 
title("Vaccination only")

#Infection
chordDiagram(mat_types_inf, grid.col = grid.col_types_inf, 
             annotationTrack = "grid", transparency = 0,
             big.gap = 50,
             small.gap = 3,
            directional = -1,
             group = group_types,
             h.ratio = 0.7, 
             link.zindex = rank(mat_types_inf),
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_types_inf))))))

circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.5, #Label size
      font = 0.1
      )},
  bg.border = NA) 
title("Infection only")
circos.clear()
dev.off()
```

#### Jaccard distance

##### All conditions (Interactive)

```{r}

matrix_vaccines <- shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  dplyr::select(Cond1, Cond2, jaccard_distance) %>% 
  pivot_wider(names_from = "Cond2", values_from = "jaccard_distance", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

types = shared_genes_df_mirror %>% 
  dplyr::select(condition = Cond1) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% 
 dplyr::select(condition, type), by = "condition") %>% 
  dplyr::select(type)

colors = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(ann_vaccines %>% 
 dplyr::select(condition, type), by = "type") %>% 
  inner_join(matrix_vaccines %>% 
               as.data.frame() %>% 
               colnames() %>% as.data.frame() %>% rename(condition = "."),
             by = "condition") %>% 
  pull(color)

# Build the chord diagram:
chorddiag(matrix_vaccines,
          groupColors = colors,
          type = "directional",
          width = NULL,
          height = NULL,
          margin = 100,
          showGroupnames = TRUE,
          groupThickness = 0.1,
          groupPadding = 2,
          groupnamePadding = 1,
          groupnameFontsize = 10,
          groupedgeColor = NULL,
          chordedgeColor = NULL,
          categoryNames = NULL,
          categorynamePadding = 100,
          categorynameFontsize = 28,
          showTicks = TRUE,
          tickInterval = NULL,
          ticklabelFontsize = 10,
          fadeLevel = 0.1,
          showTooltips = TRUE,
          showZeroTooltips = TRUE,
          tooltipNames = NULL,
          tooltipUnit = NULL,
          tooltipFontsize = 12,
          tooltipGroupConnector = " &#x25B6; ",
          precision = NULL,
          clickAction = NULL,
          clickGroupAction = NULL
)
```

#####  Weeks

Prepare tables

```{r fig.height=5, fig.width=5}
#Week 1 ------
links_df_W1 = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(ann_vaccines %>% 
               filter(week == 1) %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(ann_vaccines %>% 
               filter(week == 1) %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, jaccard_distance)

mat_week1 = links_df_W1 %>% 
  pivot_wider(names_from = "Cond2", values_from = "jaccard_distance", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()
grid.col_w1 = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(ann_vaccines %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_W1,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)
group_w1 = ann_vaccines %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_W1,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)


#Week 2 ------
links_df_W2 = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(ann_vaccines %>% 
               filter(week == 2) %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(ann_vaccines %>% 
               filter(week == 2) %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, jaccard_distance)

mat_week2 = links_df_W2 %>% 
  pivot_wider(names_from = "Cond2", values_from = "jaccard_distance", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_w2 = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(ann_vaccines %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_W2,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_w2 = ann_vaccines %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_W2,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)


#Week 4 ------
links_df_W4 = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(ann_vaccines %>% 
               filter(week == 4) %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(ann_vaccines %>% 
               filter(week == 4) %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, jaccard_distance)

mat_week4 = links_df_W4 %>% 
  pivot_wider(names_from = "Cond2", values_from = "jaccard_distance", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_w4 = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(ann_vaccines %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_W4,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_w4 = ann_vaccines %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_W4,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)

#Week 7 ------
links_df_W7 = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(ann_vaccines %>% 
               filter(week == 7) %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(ann_vaccines %>% 
               filter(week == 7) %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, jaccard_distance)

mat_week7 = links_df_W7 %>% 
  pivot_wider(names_from = "Cond2", values_from = "jaccard_distance", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_w7 = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(ann_vaccines %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_W7,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_w7 = ann_vaccines %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_W7,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)

```

Plot

```{r fig.height= 5, fig.width=10}
#Plot
png(filename = here("Figures","ChordDiagram_Covid19_Weeks_1-2-4-7_jaccard.png"), 
    width = 12, height = 14, units = "in", res = 300)

par(mfrow = c(2, 2))
#WEEK 1
chordDiagram(mat_week1, 
             grid.col = grid.col_w1, 
             annotationTrack = "grid", 
             directional = -1,
             big.gap = 6,
             transparency = 0,
             small.gap = 3,
             link.zindex = rank(mat_week1),
             h.ratio = 0.7, 
             group = group_w1,
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_week1))))))
#Labels
circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.8, #Label size
      font = 0.1)},#Label face
  bg.border = NA) 
title("Week 1")

# WEEK 2
chordDiagram(mat_week2, grid.col = grid.col_w2, 
             annotationTrack = "grid", 
            big.gap = 50,
             group = group_w2,
             small.gap = 3,
             link.zindex = rank(mat_week2),
             h.ratio = 0.7, 
             directional = -1,
             transparency = 0,
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_week2))))))
#Labels
circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.8, #Label size
      font = 0.1)},
  bg.border = NA)

title("Week 2")


# WEEK 4
chordDiagram(mat_week4, grid.col = grid.col_w4, 
             annotationTrack = "grid", 
            big.gap = 50,
             group = group_w4,
             small.gap = 3,
             link.zindex = rank(mat_week4),
             h.ratio = 0.7, 
             directional = -1,
             transparency = 0,
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_week4))))))
#Labels
circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.8, #Label size
      font = 0.1)},
  bg.border = NA)

title("Week 4")

# # WEEK 7
# chordDiagram(mat_week7, grid.col = grid.col_w7, 
#              annotationTrack = "grid", 
#             big.gap = 50,
#              group = group_w7,
#              small.gap = 3,
#              link.zindex = rank(mat_week7),
#              h.ratio = 0.7, 
#              directional = -1,
#              transparency = 0,
#              preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_week7))))))
# #Labels
# circos.track(
#   track.index = 1,
#   panel.fun = function(x, y) {
#     circos.text(
#       CELL_META$xcenter,
#       CELL_META$ylim[1],
#       CELL_META$sector.index,
#       facing = "clockwise",
#       niceFacing = TRUE,
#       adj = c(0, 0.5),
#       cex = 0.8, #Label size
#       font = 0.1)},
#   bg.border = NA)
# 
# title("Week 7")

dev.off()
circos.clear()

```

##### Types

Prepare tables

```{r fig.height= 10, fig.width=10}
#Types
###### All -----
links_df = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  dplyr::select(Cond1, Cond2, jaccard_distance)

mat_types = links_df %>% 
  pivot_wider(names_from = "Cond2", values_from = "jaccard_distance", values_fill = list(jaccard_distance = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_types = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(ann_vaccines %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_types = ann_vaccines %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)

# Vaccination only -------
links_df_vac = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(ann_vaccines %>% 
               filter(disease_vac == "V") %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(ann_vaccines %>% 
               filter(disease_vac == "V") %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, jaccard_distance)

mat_types_vac = links_df_vac %>% 
  pivot_wider(names_from = "Cond2", values_from = "jaccard_distance", values_fill = list(jaccard_distance = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_types_vac = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(ann_vaccines %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_vac,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_types_vac = ann_vaccines %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_vac,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)

#Infection only
links_df_inf = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(ann_vaccines %>% 
               filter(disease_vac == "I") %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(ann_vaccines %>% 
               filter(disease_vac == "I") %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, jaccard_distance)

mat_types_inf = links_df_inf %>% 
  pivot_wider(names_from = "Cond2", values_from = "jaccard_distance", values_fill = list(jaccard_distance = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_types_inf = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(ann_vaccines %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_inf,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_types_inf = ann_vaccines %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_inf,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)
```

Plot All conditions and stats

```{r fig.height= 5, fig.width=5}
#Plot

#Types

png(filename = here("Figures","ChordDiagram_Covid19_Types_All_jaccard_distance.png"), 
    width = 5, height = 5, units = "in", res = 300)

#Infection and vaccination
chordDiagram(mat_types, grid.col = grid.col_types,
             annotationTrack = "grid",  transparency = 0,
             big.gap = 10,
             small.gap = 3,
             directional = -1, 
             group = group_types,
             link.zindex = rank(mat_types),
              h.ratio = 0.7, 
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_types))))))

circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.5, #Label size
      font = 0.1
      )},
  bg.border = NA)
title("Infection and Vaccination")

dev.off()
circos.clear()
```

By state

```{r fig.height= 5, fig.width=15}

png(filename = here("Figures","ChordDiagram_Covid19_Types_Inf_Vac_jaccard_distance.png"), 
    width = 6, height = 12, units = "in", res = 300)

par(mfrow = c(2, 1))
#Vaccination
chordDiagram(mat_types_vac, grid.col = grid.col_types_vac, 
             annotationTrack = "grid", transparency = 0,
             big.gap = 6,
             small.gap = 3,
                          directional = -1,
             group = group_types,
             link.zindex = rank(mat_types_vac),
             h.ratio = 0.7, 
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_types_vac))))))

circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.5, #Label size
      font = 0.1
      )},
  bg.border = NA) 
title("Vaccination only")

#Infection
chordDiagram(mat_types_inf, grid.col = grid.col_types_inf, 
             annotationTrack = "grid", transparency = 0,
             big.gap = 50,
             small.gap = 3,
            directional = -1,
             group = group_types,
             h.ratio = 0.7, 
             link.zindex = rank(mat_types_inf),
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_types_inf))))))

circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.5, #Label size
      font = 0.1
      )},
  bg.border = NA) 
title("Infection only")
circos.clear()
dev.off()
```


# 7. Gene set enrichment analysis

Input

```{r}
#Gene sets

all_degs_p_05_vac_infected = readRDS(here("DGE analysis","all_studies_degs_weighted.rds")) %>% 
  filter(if_else(n >= 6, fdr <= 0.05, fdr <= (0.05 * 6/n))) 

ImmuneGO_Annotated_Genes = readRDS(here("VaxGO gene sets", "ImmuneGO_Annotated_Genes_2024-03-26.rds"))

CellMarker_ImmuneCells = read_csv(here("VaxGO gene sets", "CellMarker_ImmuneCells.csv"))


Immune_GO_T2G_general = ImmuneGO_Annotated_Genes %>% 
  filter(go_term == "Manual",
         !process %in% c("ADAPTIVE IMMUNE SYSTEM",
                         "INNATE IMMUNE SYSTEM",
                         "BCR REPERTOIRE",
                         "TCR REPERTOIRE")) %>% 
  select(process, genes)


Immune_GO_T2G_specfic = ImmuneGO_Annotated_Genes %>% 
  filter(!process %in% c("ADAPTIVE IMMUNE SYSTEM",
                         "INNATE IMMUNE SYSTEM",
                         "BCR REPERTOIRE",
                         "TCR REPERTOIRE")) %>% 
  select(process=gene_set_short, genes) 


CellMarker_genes = CellMarker_ImmuneCells %>% 
  select(cell_name, marker) %>% 
  tibble()


TERM2GENE = Immune_GO_T2G_general
geneset_name = "Immune_GO_T2G_general"


```

### Over representation analysis (ORA)

#### Multiple

```{r}
df <- all_degs_p_05_vac_infected

ora_results <- list()
for (condition_2 in unique(df$condition)) {

    genes <- df %>% 
    filter(condition == condition_2) %>% 
    pull(genes)

  go_enrich <- tryCatch({
    enrichGO(gene = genes,
             OrgDb = organism, 
             keyType = 'SYMBOL',
             readable = T,
             ont = "BP",
             pvalueCutoff = 0.05, 
             qvalueCutoff = 0.10) %>% 
      as.data.frame() %>%
      mutate(condition = condition_2)
  }, error = function(e) {
    return(NULL)
  })

  if (!is.null(go_enrich)) {
    ora_results[[condition_2]] <- go_enrich   
  }
}

ora_results_final <- bind_rows(ora_results)

write.csv(ora_results_final, file = "all_degs_other_ORA_2.csv", row.names = F)

```

#### GO Analysis

```{r}
library(GO.db)

go_bp_ancestor =  GOBPANCESTOR %>% 
  as.data.frame()
names(go_bp_ancestor[, 1]) = "go_child"

go = ora_results_final %>% 
  clean_names() %>% 
  rownames_to_column("delete") %>% 
  mutate(log_fdr = -log10(qvalue)) %>% 
  select(condition, go_id = id, description, log_fdr, qvalue, gene_id, gene_ratio, count) %>% 
  filter(qvalue < 0.10) %>%
  group_by(condition) %>% 
  arrange(desc(log_fdr)) %>% 
  slice_head(n = 5)

```

#### Heatmap

```{r}

ann_vaccines

heatmap_data = go
file = "degs_up_down_notimmune_ora_top10"

heatmap_data_log_fdr = heatmap_data %>% 
  select(description, condition, log_fdr)

matrix_data <- dcast(heatmap_data_log_fdr, 
                         description ~ condition, 
                         value.var = "log_fdr", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('description') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

matrix_data_ready =  matrix_data %>% 
  round(2)

ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines, by = "condition")

ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(dose) %>% 
  relocate(dose, .before=vaccine) %>% 
  column_to_rownames(var= "condition")  %>% 
  mutate(dose = ifelse(is.na(dose), "NA", dose))

matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(dose) %>% 
  select(!c("...1":previous_vaccination)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>%
  replace(is.na(.), 0) %>% 
  as.matrix()

dim(matrix_data_ordered) 
dim(ann_vaccines_heatmap) 

ann_vaccines_heatmap = ann_vaccines_heatmap %>% 
  select(type, week, day)


#heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(0, 3), c("white", "#ed6a5a"))

#Parameters
mat = matrix_data_ordered
width_image = 30
height_image = 30
width = unit(15, "cm")
height = unit(20, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plot
fileimageheatmap_ungrouped= paste0(file, sep ="_", "Ungrouped_Complexheatmap.png")

png(file = fileimageheatmap_ungrouped, width= width_image, height=height_image ,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "NES", #Legend
                        cell_fun = function(j, i, x, y, width, height, fill) {
        if(mat[i, j] > 0)
            grid.text(sprintf("%.1f", mat[i, j]), x, y, gp = gpar(fontsize = 8))},
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 10),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        # row_split = NULL, #Split row clusters
                        # row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_split = 3, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()

```

### Gene set enrichment analysis (GSEA)

Function

```{r}
autoGSEA <- function(df, TERM2GENE, geneset_name) {
  resultados <- list()

  condicoes <- df$condition %>% 
    unique() %>% 
    as.character()

  for (condicao in condicoes) {
    
    degs_condicao <- df %>% 
      filter(condition == condicao) %>%
      select(genes, log2fold_change) %>%
      distinct() %>%
      mutate(rank = rank(log2fold_change, ties.method = "random")) %>%
      arrange(desc(rank))

    gene_list_lf2c <- as.vector(degs_condicao$log2fold_change)
    names(gene_list_lf2c) <- degs_condicao$genes
    gene_list_lf2c <- na.omit(gene_list_lf2c)
    gene_list_lf2c <- sort(gene_list_lf2c, decreasing = TRUE)

    auto_gsea <- tryCatch({
      GSEA(geneList = gene_list_lf2c,
           TERM2GENE = TERM2GENE,
           minGSSize = 1,
           maxGSSize = 1000,
           pvalueCutoff = 1,
           pAdjustMethod = "BH") %>%
        as.data.frame() %>%
        arrange(qvalue) %>%
        mutate(condition = condicao,
               gsea_enrichment = geneset_name)
    }, error = function(e) {
      return(NULL)
    })

    if (!is.null(auto_gsea)) {
      resultados[[paste(condicao, geneset_name, sep = "_")]] <- auto_gsea
    }
  }
  

  # Combinar todos os resultados em um único dataframe
  resultado_final <- bind_rows(resultados)

  return(resultado_final)
}
```

#### ImmuneGO
```{r}
# General terms

TERM2GENE = Immune_GO_T2G_general
geneset_name = "Immune_GO_T2G_general"

all_degs_p_05_vac_infected_GSEA_ALL <- all_degs_p_05_vac_infected %>%
  select(condition, genes, log2fold_change) %>% 
  autoGSEA(TERM2GENE = Immune_GO_T2G_general, 
           geneset_name = "Immune_GO_T2G_general") 

all_degs_p_05_vac_infected_GSEA = all_degs_p_05_vac_infected_GSEA_ALL %>% 
  rownames_to_column("delete") %>% 
  clean_names() %>% 
    select(condition, id, everything()) 

all_degs_p_05_vac_infected_GSEA

#Save
write_csv(all_degs_p_05_vac_infected_GSEA, 
          file = here("Gene set enrichment analysis", paste0("all_degs_p_05_vac_infected_", geneset_name, "_GSEA_ALL", ".csv")))

all_degs_p_05_vac_infected_GSEA 

```

```{r}
# Specific terms

TERM2GENE = Immune_GO_T2G_specfic
geneset_name = "Immune_GO_T2G_specfic"

all_degs_p_05_vac_infected_GSEA_ALL <- all_degs_p_05_vac_infected %>%
  select(condition, genes, log2fold_change) %>% 
  autoGSEA(TERM2GENE = Immune_GO_T2G_specfic, 
           geneset_name = "Immune_GO_T2G_specfic") 

all_degs_p_05_vac_infected_GSEA = all_degs_p_05_vac_infected_GSEA_ALL %>% 
  rownames_to_column("delete") %>% 
  clean_names() %>% 
    select(condition, id, everything()) 

all_degs_p_05_vac_infected_GSEA

#Save
write_csv(all_degs_p_05_vac_infected_GSEA, 
          file = here("Gene set enrichment analysis", paste0("all_degs_p_05_vac_infected_", geneset_name, "_GSEA_ALL", ".csv")))

all_degs_p_05_vac_infected_GSEA 
```


#### Bubbleplot (Heatmap with ggplot)

```{r fig.height=8, fig.width=20}
# Function for clustering by day
cluster_by_day <- function(df, day_column = "day", condition_column = "condition") {
  
  # Obter os níveis do fator da coluna de dias
  days <- levels(df[[day_column]])
  
  # Inicializar uma lista para armazenar as ordens
  orders <- list()
  
  # Loop por cada nível (dia) da coluna fator day
  for (day_value in days) {
    # Filtrar os dados para o dia atual
    day_data <- df %>%
      filter(!!sym(day_column) == day_value) %>%
      select(!!sym(condition_column)) %>%
      distinct()
    
    # Criar uma matriz binária onde 1 representa a presença da condição
    if (nrow(day_data) > 1) {
      # Matriz com condições como rótulos e 1 para presença
      presence_matrix <- table(day_data[[condition_column]]) > 0
      dist_matrix <- as.matrix(presence_matrix)
      
      # Aplicar o clustering
      col_clustered <- dist(dist_matrix) %>% hclust()
      
      # Armazenar a ordem dos clusters
      orders[[day_value]] <- col_clustered$labels[col_clustered$order]
    } else {
      # Se houver apenas uma condição, armazenar o rótulo
      orders[[day_value]] <- as.character(day_data[[condition_column]])
    }
  }
  
  # Converter a lista de ordens para um data frame
  order_df <- data.frame(
    day = rep(names(orders), sapply(orders, length)),
    condition = unlist(orders)
  )
  
  return(order_df)
}

```



Main processes
```{r fig.height=8, fig.width=12}
library(ggdendro)
library(cowplot)



all_degs_p_05_vac_infected_Immune_GO_T2G_general_GSEA_ALL <- read_csv("Gene set enrichment analysis/all_degs_p_05_vac_infected_Immune_GO_T2G_general_GSEA_ALL.csv")

df = all_degs_p_05_vac_infected_Immune_GO_T2G_general_GSEA_ALL %>% 
  filter(qvalue<= 0.25) %>% 
  mutate(logq = -log10(qvalue)) %>% 
  inner_join(ann_vaccines_conditions, by = "condition") %>% 
  mutate(day = as.factor(day),
         week = as.factor(week),
         dose = as.factor(dose))

row_clustered = df %>% 
  select(condition, id, nes) %>% 
  distinct() %>% 
  pivot_wider(names_from = "condition",
              values_from= "nes") %>%
  column_to_rownames("id") %>% 
  as.matrix() %>% 
  replace(is.na(.), 0) %>% 
  dist(.) %>% 
  hclust()


result_vac <- cluster_by_day(df %>% filter(disease_vac == "V"), day_column = "day", condition_column = "condition") %>%
  distinct() %>% 
  mutate(day = fct_relevel(day, df$day %>% unique() %>% sort() %>% as.character()))

result_inf_vac2 <- cluster_by_day(df %>% 
                               filter(disease_vac == "I",
                                      dose %in% c("2")), day_column = "day") %>%
  mutate(day = fct_relevel(day, df$day %>% unique() %>% sort() %>% as.character()))

result_inf_vac01 <- cluster_by_day(df %>% 
                               filter(disease_vac == "I",
                                      dose %in% c("0", "1")), day_column = "day") %>%
  mutate(day = fct_relevel(day, df$day %>% unique() %>% sort() %>% as.character()))

column_order <- c(result_vac$condition, result_inf_vac2$condition, result_inf_vac01$condition)


labels_disease_vac <- df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = disease_vac) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$disease_vac) + 
  theme_void() +
  labs(y = "Infection/Vaccination") +
  theme(axis.title.y = element_text(size = 10, hjust = 1),
        legend.position = "none") 

labels_type <- df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = type) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$Platform) + 
  theme_void() +
  labs(y = "Type") +
  theme(axis.title.y = element_text(size = 10, hjust = 1),
        legend.position = "none") 

labels_dose <- df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = dose) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$dose) + 
  theme_void() +
  labs(y = "Dose") +
  theme(axis.title.y = element_text(size = 10, hjust = 1),
        legend.position = "none")

labels_week = df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = week) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$week) + 
  theme_void() +
  labs(y = "Week") +
  theme(axis.title.y = element_text(size = 10, hjust = 1),
        legend.position = "none")

labels_day = df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = day) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$day) + 
  theme_void() +
  labs(y = "Day") +
  theme(axis.title.y = element_text(size = 10, hjust = 1),
        legend.position = "none")
#Plot
dotplot = df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, column_order),
         id = factor(id, levels = row_clustered$labels[row_clustered$order]),
         id = str_to_title(id),
         id = case_when(id == "Humoral Adaptive Immune System" ~ "Humoral Adap.",
                        id == "Cellular Adaptive Immune System" ~ "Cellular Adap.",
                        id == "Complement Immune System" ~ "Complement",
                        id == "Immunoglobulin Mediated Response" ~ "Ig-mediated",
                        id == "Antiviral And Interferon" ~ "Antiviral/IFN",.default = id),
         id = str_to_upper(id)) %>% 
  ggplot(aes(x=condition, y = id, color = nes, size = logq)) +
  geom_point() + 
  theme_minimal() + 
  theme(axis.text = element_text(color = "black", size = 12) ,
        axis.line  = element_blank(),
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1),
        axis.ticks = element_blank(),
        legend.direction = "horizontal",
        legend.position = "bottom") +
  labs(x = "",
       y = "") +
  scale_color_gradient2(low = "#4DBBD5FF",
                        mid = "white",
                        midpoint = 0, 
                        name = "NES",
                      high = "#DC0000FF") +
  scale_y_discrete(position = "right") +
  scale_size(range = c(2, 6))
  

patch_immune_main = 
  labels_disease_vac /
  labels_type /
  labels_dose /
  labels_week/
  labels_day/
  dotplot +
  plot_layout(heights = c(0.3, 0.3, 0.3, 0.3, 0.3, 5),
              guides = 'collect') +
  plot_annotation(title = "GSEA ImmuneGO",
                  subtitle = "Main processes",
                  theme = theme(legend.position = "bottom",
                                plot.margin = margin(0.5, 0, 0.5, 1, unit = "cm"),
                                legend.direction = "horizontal", 
                                legend.box = "horizontal"))


patch_immune_main %>% 
  ggsave(filename = here("Figures", "GSEA_ImmuneGO_MainGOs_Bubbleplot.png"),
         width = 12,
         height = 8)
patch_immune_main
```

Specific all
```{r fig.height=22, fig.width=10}
library(ggdendro)
library(cowplot)

all_degs_p_05_vac_infected_Immune_GO_T2G_specfic_GSEA_ALL = read_csv(here("Gene set enrichment analysis", "all_degs_p_05_vac_infected_Immune_GO_T2G_specfic_GSEA_ALL.csv")) 


df = all_degs_p_05_vac_infected_Immune_GO_T2G_specfic_GSEA_ALL %>% 
  clean_names() %>% 
  filter(qvalue <= 0.25) %>% 
  mutate(logq = -log10(qvalue)) %>% 
  inner_join(ann_vaccines_conditions, by = "condition") %>% 
  mutate(day = as.factor(day),
         week = as.factor(week),
         dose = as.factor(dose))

row_clustered = df %>% 
  select(condition, id, nes) %>% 
  distinct() %>% 
  pivot_wider(names_from = "condition",
              values_from= "nes") %>%
  column_to_rownames("id") %>% 
  as.matrix() %>% 
  replace(is.na(.), 0) %>% 
  dist(.) %>% 
  hclust()

result_vac <- cluster_by_day(df %>% filter(disease_vac == "V"), day_column = "day", condition_column = "condition") %>%
  distinct() %>% 
  mutate(day = fct_relevel(day, df$day %>% unique() %>% sort() %>% as.character()))

result_inf_vac2 <- cluster_by_day(df %>% 
                               filter(disease_vac == "I",
                                      dose %in% c("2")), day_column = "day") %>%
  mutate(day = fct_relevel(day, df$day %>% unique() %>% sort() %>% as.character()))

column_order <- c(result_vac$condition, result_inf_vac2$condition, result_inf_vac01$condition) %>% 
  unique()

result_inf_vac01 <- cluster_by_day(df %>% 
                               filter(disease_vac == "I",
                                      dose %in% c("0", "1")), day_column = "day") %>%
  mutate(day = fct_relevel(day, df$day %>% unique() %>% sort() %>% as.character()))

column_order <- c(result_vac$condition, result_inf_vac2$condition, result_inf_vac01$condition)


labels_disease_vac <- df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = disease_vac) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$disease_vac) + 
  theme_void() +
  labs(y = "Infection/Vac") +
  theme(axis.title.y = element_text(size = 10),
        legend.position = "none") 

labels_type <- df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = type) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$Platform) + 
  theme_void() +
  labs(y = "Type") +
  theme(axis.title.y = element_text(size = 10),
        legend.position = "none") 

labels_dose <- df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = dose) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$dose) + 
  theme_void() +
  labs(y = "Type") +
  theme(axis.title.y = element_text(size = 10),
        legend.position = "none") 

labels_week = df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = week) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$week) + 
  theme_void() +
  labs(y = "Week") +
  theme(axis.title.y = element_text(size = 10),
        legend.position = "none")

labels_day = df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = day) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$day) + 
  theme_void() +
  labs(y = "Day") +
  theme(axis.title.y = element_text(size = 10),
        legend.position = "none")
#Plot
dotplot_specific = df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, column_order),
         id = factor(id, levels = row_clustered$labels[row_clustered$order])) %>% 
  ggplot(aes(x=condition, y = id, color = nes, size = logq)) +
  geom_point() + 
  theme_minimal() + 
  theme(axis.text = element_text(color = "black") ,
        axis.line  = element_blank(),
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1),
        axis.ticks = element_blank(),
        legend.direction = "horizontal",
        legend.position = "bottom") +
  labs(x = "",
       y = "") +
  scale_color_gradient2(low = "#4DBBD5FF",
                        mid = "white",
                        midpoint = 0, 
                        name = "NES",
                      high = "#DC0000FF") +
  scale_y_discrete(position = "right") 
  

patch_immune_specific = 
  labels_disease_vac /
  labels_type /
  labels_dose/
  labels_week/
  labels_day/
  dotplot_specific +
  plot_layout(heights = c(0.05,0.05, 0.05,0.05, 0.05, 5),
              guides = 'collect') +
  plot_annotation(title = "GSEA ImmuneGO",
                  subtitle = "Specific processes",
                  theme = theme(legend.position = "bottom",
                                plot.margin = margin(0.5, 0, 0.5, 1, unit = "cm"),
                                legend.direction = "horizontal", 
                                legend.box = "horizontal"))
patch_immune_specific

patch_immune_specific %>% 
  ggsave(filename = here("Figures", "GSEA_ImmuneGO_Specific_Bubbleplot.png"),
         width = 10,
         height = 22)

```

Specific filtered 
```{r fig.height=8, fig.width=10}
library(ggdendro)
library(cowplot)

all_degs_p_05_vac_infected_Immune_GO_T2G_specfic_GSEA_ALL = read_csv(here("Gene set enrichment analysis", "all_degs_p_05_vac_infected_Immune_GO_T2G_specfic_GSEA_ALL.csv")) 


df = all_degs_p_05_vac_infected_Immune_GO_T2G_specfic_GSEA_ALL %>% 
  clean_names() %>% 
  filter(qvalue <= 0.25) %>% 
  mutate(logq = -log10(qvalue)) %>% 
  inner_join(ann_vaccines_conditions, by = "condition") %>% 
  mutate(day = as.factor(day),
         week = as.factor(week),
         dose = as.factor(dose))

id_filtered = df %>% 
  select(condition, id) %>%
  group_by(id) %>% 
  summarize(n = n()) %>% 
  arrange(-n) %>% 
  filter(n>5)

df = df %>% 
  filter(id %in% c(id_filtered$id))

row_clustered = df %>% 
  select(condition, id, nes) %>% 
  distinct() %>% 
  pivot_wider(names_from = "condition",
              values_from= "nes") %>%
  column_to_rownames("id") %>% 
  as.matrix() %>% 
  replace(is.na(.), 0) %>% 
  dist(.) %>% 
  hclust()

result_vac <- cluster_by_day(df %>% filter(disease_vac == "V"), day_column = "day", condition_column = "condition") %>%
  distinct() %>% 
  mutate(day = fct_relevel(day, df$day %>% unique() %>% sort() %>% as.character()))

result_inf_vac2 <- cluster_by_day(df %>% 
                               filter(disease_vac == "I",
                                      dose %in% c("2")), day_column = "day") %>%
  mutate(day = fct_relevel(day, df$day %>% unique() %>% sort() %>% as.character()))

column_order <- c(result_vac$condition, result_inf_vac2$condition, result_inf_vac01$condition) %>% 
  unique()

result_inf_vac01 <- cluster_by_day(df %>% 
                               filter(disease_vac == "I",
                                      dose %in% c("0", "1")), day_column = "day") %>%
  mutate(day = fct_relevel(day, df$day %>% unique() %>% sort() %>% as.character()))

column_order <- c(result_vac$condition, result_inf_vac2$condition, result_inf_vac01$condition)

labels_type <- df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = type) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$Platform) + 
  theme_void() +
  labs(y = "Type") +
  theme(axis.title.y = element_text(size = 10),
        legend.position = "none") 

labels_disease_vac <- df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = disease_vac) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$disease_vac) + 
  theme_void() +
  labs(y = "Type") +
  theme(axis.title.y = element_text(size = 10),
        legend.position = "none") 

labels_week = df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = week) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$week) + 
  theme_void() +
  labs(y = "Week") +
  theme(axis.title.y = element_text(size = 10),
        legend.position = "none")

labels_day<- df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = day) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$day) + 
  theme_void() +
  labs(y = "Type") +
  theme(axis.title.y = element_text(size = 10),
        legend.position = "none") 
#Plot
dotplot_specific = df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, column_order),
         id = factor(id, levels = row_clustered$labels[row_clustered$order])) %>% 
  ggplot(aes(x=condition, y = id, color = nes, size = logq)) +
  geom_point() + 
  theme_minimal() + 
  theme(axis.text = element_text(color = "black") ,
        axis.line  = element_blank(),
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1),
        axis.ticks = element_blank(),
        legend.direction = "horizontal",
        legend.position = "bottom") +
  labs(x = "",
       y = "") +
  scale_color_gradient2(low = "#4DBBD5FF",
                        mid = "white",
                        midpoint = 0, 
                        name = "NES",
                      high = "#DC0000FF") +
  scale_y_discrete(position = "right")  
  

patch_immune_specific_filtered =  labels_disease_vac/
  labels_type /
  labels_week/
  labels_day/
  dotplot_specific +
  plot_layout(heights = c(0.2, 0.2, 0.2, 0.2, 5),
              guides = 'collect') +
  plot_annotation(title = "GSEA ImmuneGO",
                  subtitle = "Specific processes",
                  theme = theme(legend.position = "bottom",
                                plot.margin = margin(0.5, 0, 0.5, 1, unit = "cm"),
                                legend.direction = "horizontal", 
                                legend.box = "horizontal"))
patch_immune_specific_filtered


```


Specific
```{r fig.height=22, fig.width=10}
library(ggdendro)
library(cowplot)

all_degs_p_05_vac_infected_Immune_GO_T2G_specfic_GSEA_ALL = read_csv(here("Gene set enrichment analysis", "all_degs_p_05_vac_infected_Immune_GO_T2G_specfic_GSEA_ALL.csv")) 


df = all_degs_p_05_vac_infected_Immune_GO_T2G_specfic_GSEA_ALL %>% 
  clean_names() %>% 
  filter(qvalue <= 0.25) %>% 
  mutate(logq = -log10(qvalue)) %>% 
  inner_join(ann_vaccines_conditions, by = "condition") %>% 
  mutate(day = as.factor(day),
         week = as.factor(week),
         dose = as.factor(dose))

row_clustered = df %>% 
  select(condition, id, nes) %>% 
  distinct() %>% 
  pivot_wider(names_from = "condition",
              values_from= "nes") %>%
  column_to_rownames("id") %>% 
  as.matrix() %>% 
  replace(is.na(.), 0) %>% 
  dist(.) %>% 
  hclust()

result_vac <- cluster_by_day(df %>% filter(disease_vac == "V"), day_column = "day", condition_column = "condition") %>%
  distinct() %>% 
  mutate(day = fct_relevel(day, df$day %>% unique() %>% sort() %>% as.character()))

result_inf_vac2 <- cluster_by_day(df %>% 
                               filter(disease_vac == "I",
                                      dose %in% c("2")), day_column = "day") %>%
  mutate(day = fct_relevel(day, df$day %>% unique() %>% sort() %>% as.character()))

column_order <- c(result_vac$condition, result_inf_vac2$condition, result_inf_vac01$condition) %>% 
  unique()

result_inf_vac01 <- cluster_by_day(df %>% 
                               filter(disease_vac == "I",
                                      dose %in% c("0", "1")), day_column = "day") %>%
  mutate(day = fct_relevel(day, df$day %>% unique() %>% sort() %>% as.character()))

column_order <- c(result_vac$condition, result_inf_vac2$condition, result_inf_vac01$condition)

labels_type <- df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = type) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$Platform) + 
  theme_void() +
  labs(y = "Type") +
  theme(axis.title.y = element_text(size = 10),
        legend.position = "none") 

labels_week = df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = week) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$week) + 
  theme_void() +
  labs(y = "Week") +
  theme(axis.title.y = element_text(size = 10),
        legend.position = "none")
#Plot
dotplot_specific = df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, column_order),
         id = factor(id, levels = row_clustered$labels[row_clustered$order])) %>% 
  ggplot(aes(x=condition, y = id, color = nes, size = logq)) +
  geom_point() + 
  theme_minimal() + 
  theme(axis.text = element_text(color = "black") ,
        axis.line  = element_blank(),
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1),
        axis.ticks = element_blank(),
        legend.direction = "horizontal",
        legend.position = "bottom") +
  labs(x = "",
       y = "") +
  scale_color_gradient2(low = "#4DBBD5FF",
                        mid = "white",
                        midpoint = 0, 
                        name = "NES",
                      high = "#DC0000FF") +
  scale_y_discrete(position = "right") 
  

patch_immune_specific = labels_type /
  labels_week/
  dotplot_specific +
  plot_layout(heights = c(0.05, 0.05, 5),
              guides = 'collect') +
  plot_annotation(title = "GSEA ImmuneGO",
                  subtitle = "Specific processes",
                  theme = theme(legend.position = "bottom",
                                plot.margin = margin(0.5, 0, 0.5, 1, unit = "cm"),
                                legend.direction = "horizontal", 
                                legend.box = "horizontal"))
patch_immune_specific
  
  

patch_immune_specific %>% 
  ggsave(filename = here("Figures", "GSEA_ImmuneGO_Specific_Bubbleplot.png"),
         width = 10,
         height = 22)

```


## Cell proportion

<https://www.datacamp.com/tutorial/pca-analysis-r>
<https://rpkgs.datanovia.com/factoextra/index.html>
<http://www.sthda.com/english/articles/31-principal-component-methods-in-r-practical-guide/114-mca-multiple-correspondence-analysis-in-r-essentials/>
<https://statisticsglobe.com/pca-before-k-means-clustering-r>\>




### ssGSEA: CellMarker
```{r}
library(corto)

#Import files
ann_vaccines_samples <- read_csv(here("Annotations and metadata", "ann_vaccines_samples.csv"))

normalized_counts <- readRDS(here("DGE analysis", "all_matrices_normalized_counts.rds")) 

all_samples_l2fc <- readRDS(here("DGE analysis", "all_samples_l2fc.rds")) %>%
  filter(!condition %in% c(ann_vaccines_samples %>% 
           filter(day == 0) %>% 
           pull(condition)))

head(all_samples_l2fc)
```


```{r}
######## CELL MARKER IMMUNE ---------

CellMarker_ImmuneCells = read_csv(here("VaxGO gene sets", "CellMarker_ImmuneCells.csv"))

CellMarker_annotation = CellMarker_ImmuneCells %>% 
 dplyr::select(cell_name, Type) %>% 
  distinct()

CellMarker_genes = CellMarker_ImmuneCells %>% 
 dplyr::select(process = cell_name, genes = marker)

############## Inputs

ssgsea_gene = all_samples_l2fc 
metadata = ann_vaccines_samples 
filename = "covid_all"
go_dataset = "CellMarkerImmune"

#Extract sample names
sample_names = ssgsea_gene %>% 
  as.data.frame() %>% 
  colnames() %>% 
  as.data.frame() %>% 
  rename(sample_names = '.')

#CellMarker immune-------
genelist_CellMarker_genes = CellMarker_genes %>%dplyr::select(process, genes)
genelist_CellMarker_genes = split(genelist_CellMarker_genes$genes, genelist_CellMarker_genes$process) 

nesmat_result = ssgsea(inmat = ssgsea_gene, 
                                        groups = genelist_CellMarker_genes)

nesmat_result_CellMarker =  nesmat_result %>% 
  as.data.frame() %>% 
  t() %>% 
  as.data.frame() %>% 
  cbind(sample_names) %>% 
  select(sample_names, everything()) %>% 
  rownames_to_column("delete") %>% 
  select(-delete) %>% 
  column_to_rownames("sample_names") %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column("process") %>% 
  select(process,everything()) %>% 
  pivot_longer(cols = -process, names_to = "sample", values_to = "nes") %>%
  mutate(pvalue = z2p(nes), #Convert NES into pvalue
         qvalue = p.adjust(pvalue, method = "BH"),
         logq = - log10(qvalue),
         geneset_name = "CellMarker Immune") %>%
  inner_join(metadata, by= "sample") 



#save
write_csv(nesmat_result_CellMarker, file = here("Gene set enrichment analysis", paste0(filename, sep = "_", go_dataset, "_ssgsea_results.csv")))

head(nesmat_result_CellMarker)
```



### immunedeconv

```{r fig.height=6, fig.width=12}
# install.packages("remotes")
# remotes::install_github("omnideconv/immunedeconv")
library(immunedeconv)
library(tidyverse)

all_samples_l2fc = readRDS(file = here("DGE analysis", "all_samples_l2fc.rds"))

all_matrices_counts_l2fc = all_samples_l2fc %>% 
  select(genes, sample, log2fc) %>% 
  pivot_wider(names_from = "sample",
              values_from = "log2fc") %>% 
  column_to_rownames("genes") %>% 
  as.matrix()

```


#### XCell ssGSEA
```{r fig.height=6, fig.width=12}
#XCell (ssGSEA) ----
xcell_results = immunedeconv::deconvolute(all_matrices_counts_l2fc, "xcell") 

xcell_results %>% 
  write_csv(file = here("Cell populations", "all_covid_degs_xcell_results.csv"))
```


```{r fig.height=6, fig.width=12}

xcell_results = read_csv(file = here("Cell populations", "all_covid_degs_xcell_results.csv"))

xcell_results_annotated = xcell_results %>% 
  pivot_longer(cols = -cell_type,
               names_to = "sample",
               values_to = "value") %>% 
  inner_join(ann_vaccines_samples, by = "sample") %>% 
   distinct() %>% 
  filter(condition != "BNT-MO (V1, D0)",
         condition != "BNT-MO (V1, D6)") %>% 
  mutate(disease_vac = if_else(disease_vac == "Healthy", "H", disease_vac)) %>% 
  mutate(disease_vac = fct_relevel(disease_vac, "H", "I", "V"),
           type = fct_relevel(type, "I", "RNA", "VV", "IN", "SU")) %>% 
  inner_join(ann_vaccines_conditions %>% select(condition_grouped = condition), by = join_by("condition_grouped")) %>% 
  select(-condition) %>% 
  rename(condition = condition_grouped) %>% 
  mutate(condition = fct_relevel(condition, 
                                 ann_vaccines_conditions_ordered$condition_grouped %>% as.character()%>% unique())) %>% 
  mutate(type = if_else(condition == "BNT-I (D0-5)", "RNA", type)) %>% 
  filter(day != 0) %>% 
  group_by(condition, cell_type) %>% 
  mutate(minmax_value = (value - min(value)) / 
                             (max(value) - min(value)))

xcell_results_annotated
  
```


```{r fig.height=6, fig.width=10}

xcell_results_annotated_summary = xcell_results_annotated %>% 
  group_by(condition, cell_type) %>% 
  summarize(median = median(minmax_value),
            mean = mean(minmax_value),
            sd = sd(minmax_value)) %>% 
  inner_join(ann_vaccines_conditions %>% 
               select(condition, type, vaccine, disease_vac, day), by = join_by("condition")) %>% 
  ungroup() 

cell_type_selected = "B cell"

#boxplot
xcell_results_annotated_2_plot =  xcell_results_annotated %>% 
  filter(cell_type == cell_type_selected) %>% 
  ggplot() +  
  geom_beeswarm(aes(x = condition, 
                  y = minmax_value), 
                color = "black",
              alpha = 1, 
              size = 3, 
              seed = 123,
              cex = 0.1) +
  geom_beeswarm(aes(x = condition, 
                  y = minmax_value,
                  color = type), 
              alpha = 1, size = 2, seed = 123,
              cex = 0.1) +
  geom_bar(data = xcell_results_annotated_summary %>% 
             filter(cell_type == cell_type_selected), 
           aes(x = condition, y = median, fill = type), 
           stat = "identity", 
           alpha = 0.3) +
  stat_summary(data = xcell_results_annotated_summary %>% 
             filter(cell_type == cell_type_selected,
                    median > 0.001),
             aes(x = condition, y = median),
               fun = "median",
               geom = "crossbar", 
              size = 0.2,
               width = 1,
               colour = "black") +
  scale_fill_manual(values = c("Day 0" = "white", 
                               ann_vaxsig_colors$type)) +
  scale_color_manual(values = c("Day 0" = "gray80", 
                               ann_vaxsig_colors$type)) +
  theme_minimal()+
  theme(legend.position = "top",
        axis.text.x = element_text(size = 10,
                                   color = "black",
                                   vjust = 1,
                                   hjust = 1,
                                   angle = 45),
        axis.text.y = element_text(size = 10,
                                   color = "black",
                                   angle = 0),
        # legend.text = element_text(size = 10),
        # legend.title = element_text(size = 15),
        plot.title = element_text(size = 12),
        plot.subtitle = element_text(size = 10),
        axis.title.y = element_text(color = "black"),
        panel.grid.major.y = element_line(size = 0.5,
                                          linetype = 2,
                                          color = "gray75"),
        panel.grid.major.x = element_line(size = 0,
                                          linetype = 2,
                                          color = "gray75"),
        axis.line.x = element_line(size = 1,
                                   colour = "black",
                                   linetype = 1),
        axis.line.y = element_line(size = 1,
                                   colour = "black",
                                   linetype = 1),
        axis.ticks.x = element_line(size = 1, color = "black"),
        axis.ticks.y = element_line(size = 1, color = "gray75"),
        panel.grid.minor = element_line(size = 0.5,
                                          linetype = 2,
                                          color = "gray75"),
        plot.margin = margin(0, 0, 0, 0, unit = "cm")) +
  # scale_y_continuous(expand = c(0,0)) +
  labs(y = "NES",
       title = "B cell",
       x="",
       subtitle = "By week",
       fill = "Type") +
  guides(color = "none") 


xcell_results_annotated_2_plot
```


```{r}
#Function to create boxplots
create_boxplots <- function(result_df) {
  
  unique_processes <- unique(result_df$cell_type)
  
  purrr::map(unique_processes, function(process_name) {
    cellmarker_boxplot <- result_df %>% 
      filter(cell_type == process_name) %>% 
  ggplot() +
  aes(x = condition, y = value, fill = type) +  
  geom_hline(yintercept = 0, linetype = 2) +
  geom_jitter(aes(color = type),
              alpha = 0.3, size = 0.1) +
  geom_boxplot(outliers = F) +
  theme_few() +
  scale_fill_manual(values = c("Day 0" = "white", 
                               ann_vaxsig_colors$type)) +
    scale_color_manual(values = c("Day 0" = "gray80", 
                               ann_vaxsig_colors$type)) +
  theme(axis.text.x = element_text(angle = 45,
                                   hjust = 1,
                                   vjust = 1),
        plot.title = element_text(hjust = 0.5, face = "bold"),
        plot.margin = margin(0, 0, 0, 2, "cm")) +
  labs(y = "score",
       title = process_name,
       x="",
       subtitle = "By week",
       fill = "Type") +
  guides(color = "none")

    ggsave(cellmarker_boxplot, 
           file = here("Figures", "XCell", paste0("ssgsea_xcell_boxplot_", process_name, ".png")),
           width = 12, height = 6)
  })
}

# Apply to all processes
create_boxplots(xcell_results_annotated)
```



### Visualization

Generate a single boxplot
```{r fig.height=6, fig.width=12}

nesmat_result_CellMarker = read_csv(here("Gene set enrichment analysis", "covid_all_CellMarkerImmune_ssgsea_results.csv"))

unique(nesmat_result_CellMarker$process)

process_selected = "Neutrophil"

nesmat_result_CellMarker_1 = nesmat_result_CellMarker %>% 
  distinct() %>% 
  filter(condition != "BNT-MO (V1, D0)",
         condition != "BNT-MO (V1, D6)") %>% 
  mutate(disease_vac = if_else(disease_vac == "Healthy", "H", disease_vac)) %>% 
  mutate(disease_vac = fct_relevel(disease_vac, "H", "I", "V"),
           type = fct_relevel(type, "I", "RNA", "VV", "IN", "SU")) %>% 
  arrange(disease_vac,type, condition) %>% 
  inner_join(ann_vaccines_conditions %>% select(condition, samples), by = "condition") %>% 
  mutate(condition = factor(condition, levels = c(unique(condition[order(condition)]))),
         condition = paste0(condition, " (n=", samples, ")"))

nesmat_result_CellMarker_2 = nesmat_result_CellMarker_1 %>% 
  mutate(condition = fct_relevel(condition, 
                                 nesmat_result_CellMarker_1 %>% 
                                   filter(str_detect(condition, "^H \\(D0\\)")) %>% 
                                   select(condition) %>% 
                                   distinct() %>% 
                                   pull(condition), 
                                 nesmat_result_CellMarker_1 %>% 
                                   filter(str_detect(condition, "^I")) %>% 
                                   filter(!str_detect(condition, "^I-I")) %>% 
                                   select(condition) %>% 
                                   arrange(condition) %>% 
                                   distinct() %>% 
                                   pull(condition),
                                 nesmat_result_CellMarker_1 %>% 
                                   filter(disease_vac == "I") %>% 
                                   filter(str_detect(condition, "^BNT-I")) %>% 
                                    filter(!str_detect(condition, "^BNT-I \\(D\\d+\\)")) %>% 
                                    arrange(day, condition) %>%
                                   select(condition) %>% 
                                    distinct() %>% 
                                    pull(condition),
                                 nesmat_result_CellMarker_1 %>% 
                                    filter(str_detect(condition, "^BNT-I \\(D\\d+\\)")) %>% 
                                    arrange(day, condition) %>%
                                   select(condition) %>% 
                                    distinct() %>% 
                                    pull(condition),
                                 nesmat_result_CellMarker_1 %>% 
                                   filter(disease_vac == "V",
                                          str_detect(condition, "^BNT \\(V1, D\\d+\\)")) %>% 
                                   arrange(day, condition) %>% 
                                   select(condition) %>% 
                                   distinct() %>% 
                                   pull(condition),
                                 nesmat_result_CellMarker_1 %>% 
                                   filter(disease_vac == "V",
                                          !str_detect(condition, "^BNT \\(V1, D\\d+\\)"),
                                          !str_detect(condition, "BBIBP"),
                                          !str_detect(condition, "ZF2001")) %>% 
                                   arrange(condition) %>% 
                                   select(condition) %>% 
                                   distinct() %>% 
                                   pull(condition),
                                 nesmat_result_CellMarker_1 %>% 
                                   filter(str_detect(condition, "BBIBP")) %>% 
                                   select(condition) %>% 
                                   arrange(condition) %>% 
                                   distinct() %>% 
                                   pull(condition), 
                                 nesmat_result_CellMarker_1 %>% 
                                   filter(str_detect(condition, "ZF2001")) %>% 
                                   select(condition) %>% 
                                   arrange(condition) %>% 
                                   distinct() %>% 
                                   pull(condition))) %>%
  mutate(type = if_else(type == "H", "I", type),
         disease_vac = if_else(disease_vac == "H", "I", disease_vac)) 

nesmat_result_CellMarker_2 %>% 
  select(condition) %>% 
  distinct()

ssgsea_boxplot_by_type 

nesmat_result_CellMarker_2 %>% 
  filter(process == process_selected) %>% 
  ggplot() +
  aes(x = condition, y = nes, fill = if_else(day == 0, "Day 0", type)) +  
  geom_hline(yintercept = 0, linetype = 2) +
  geom_jitter(aes(color = if_else(day == 0, "Day 0", type)),
              alpha = 0.3, size = 0.1) +
  geom_boxplot(outliers = F) +
  theme_few() +
  scale_fill_manual(values = c("Day 0" = "white", 
                               ann_vaxsig_colors$type)) +
    scale_color_manual(values = c("Day 0" = "gray80", 
                               ann_vaxsig_colors$type)) +
  theme(axis.text.x = element_text(angle = 45,
                                   hjust = 1,
                                   vjust = 1),
        plot.title = element_text(hjust = 0.5, face = "bold"),
        plot.margin = margin(0, 0, 0, 2, "cm"),
        panel.grid.major = element_line(size = 0.05),
        panel.grid.minor = element_line(size = 0.05)) +
  labs(y = "NES",
       title = process_selected,
       x="",
       subtitle = "By week",
       fill = "Type") +
  guides(color = "none") +
  facet_wrap(~disease_vac, scales = "free_x")

  ssgsea_boxplot_by_type
```

Generate boxplots for all processes
```{r fig.height=6, fig.width=12}
#Function to create boxplots
create_boxplots <- function(result_df) {
  
  unique_processes <- unique(result_df$process)
  
  purrr::map(unique_processes, function(process_name) {
    cellmarker_boxplot <- result_df %>% 
      filter(process == process_name) %>% 
  ggplot() +
  aes(x = condition, y = nes, fill = if_else(day == 0, "Day 0", type)) +  
  geom_hline(yintercept = 0, linetype = 2) +
  geom_jitter(aes(color = if_else(day == 0, "Day 0", type)),
              alpha = 0.3, size = 0.1) +
  geom_boxplot(outliers = F) +
  theme_few() +
  scale_fill_manual(values = c("Day 0" = "white", 
                               ann_vaxsig_colors$type)) +
    scale_color_manual(values = c("Day 0" = "gray80", 
                               ann_vaxsig_colors$type)) +
  theme(axis.text.x = element_text(angle = 45,
                                   hjust = 1,
                                   vjust = 1),
        plot.title = element_text(hjust = 0.5, face = "bold"),
        plot.margin = margin(0, 0, 0, 2, "cm")) +
  labs(y = "NES",
       title = process_name,
       x="",
       subtitle = "By week",
       fill = "Type") +
  guides(color = "none") +
  facet_wrap(~disease_vac, scales = "free_x")

    ggsave(cellmarker_boxplot, 
           file = here("Figures", paste0("ssgsea_CellMarkerImmune_boxplot_", process_name, ".png")),
           width = 12, height = 6)
  })
}

# Apply to all processes
create_boxplots(nesmat_result_CellMarker_2)
```


```{r fig.height=5, fig.width=5}
#PCA -----
# install.packages("ggfortify")
# library(ggfortify)
nesmat_result_immunego_general_matrix = nesmat_result_immunego_general %>% 
  dplyr::select(sample, process, nes) %>% 
  pivot_wider(names_from = "process",
              values_from = "nes") %>% 
    column_to_rownames("sample")

labels = nesmat_result_immunego_general_matrix %>% 
  rownames_to_column("sample") %>% 
  inner_join(nesmat_result_immunego_general %>% 
               dplyr::select(sample, condition, type, day, week) %>% distinct(),
               by = "sample") 

pca_res <- prcomp(nesmat_result_immunego_general_matrix, scale. = TRUE)

ssgsea_pca_type = autoplot(pca_res,  data = labels) +
  geom_point(aes(fill = type, label = condition), colour="black",pch=21, size = 3) +
  theme_few()+
  scale_fill_manual(values = ann_vaxsig_colors$Platform)

ssgsea_pca_day = autoplot(pca_res,  data = labels) +
  geom_point(aes(fill = day), colour="black",pch=21, size = 3) +
  theme_few()+
  scale_fill_stepsn(colors = c(
    "0" = "white",
    "1" = "#caf0f8",
    "2" = "#0087BF",
    "3" = "#90e0ef",
    "4" = "#6CD5EA",
    "5" = "#48cae4",
    "6" = "#00b4d8",
    "7" = "#0096c7",
    "10" = "#0087BF",
    "11" = "#0087BF",
    "14" = "#0077b6",
    "26" = "#015BA0",
    "28" = "#023e8a",
    "51" = "#03045e"))

```

### Statistical analysis

```{r}
# install.packages("ggstatsplot")
library(ggstatsplot)

ssgsea_interesting = ssgsea_interesting %>% 
  filter(process == "INNATE IMMUNE SYSTEM") %>% 
  distinct()

process = "INNATE IMMUNE SYSTEM"

ggbetweenstats(
  data  = ssgsea_interesting,
  x     = condition,
  y     = nes,
  title =  paste0("ImmuneGO ssGSEA ", filename, process)
)

#Agrupado
grouped_ggbetweenstats(
  data             = ssgsea_interesting,
  x                = condition,
  y                = nes,
  grouping.var     = process,
  ggsignif.args    = list(textsize = 4, tip_length = 0.01),
  p.adjust.method  = "bonferroni",
  palette          = "default_jama",
  package          = "ggsci",
  plotgrid.args    = list(nrow = 1),
  annotation.args  = list(title = paste0("ImmuneGO ssGSEA ", filename))
)

```

#### Kruskal Wallis

```{r}
library(ggsignif)
library(rstatix)

df = test
df$Condition <- as.factor(df$Condition)

stat.test <- df %>%
  t_test(Condition ~ NES) %>%
  add_significance()


test %>%
  filter(Process %in% "INNATE IMMUNE RESPONSE") %>%
  ggboxplot(x = "Process", y = "NES",
          color = "Condition", palette = "jco") +
  facet_wrap(vars(Vaccine)) +
  stat_compare_means(method = "anova", label.y = 4) +      # Add global p-value
  stat_compare_means(label = "p.signif", method = "t.test",
                     ref.group = ".all.")    # Comparação múltipla usando Tukey HSD


# Box plots with p-values
bxp <- ggboxplot(df, x = "Process", y = "NES", fill = "#00AFBB")
stat.test <- stat.test %>% add_xy_position(x = "supp")
bxp + 
  stat_pvalue_manual(stat.test, label = "p") +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1)))

```






# 8. Principal components analysis

Tutorials [Data Camp - PCA Analysis in
R](https://www.datacamp.com/tutorial/pca-analysis-r){.uri}

[FactoExtra](https://rpkgs.datanovia.com/factoextra/index.html){.uri}
[STHDA: Principal Component Methods in R - Practical
guide](http://www.sthda.com/english/articles/31-principal-component-methods-in-r-practical-guide/114-mca-multiple-correspondence-analysis-in-r-essentials/){.uri}

[Statistics globe: PCA before K Means
Clustering](https://statisticsglobe.com/pca-before-k-means-clustering-r){.uri}\>

Packages and colors

```{r}
library(tidyverse)
library(ggrepel) #load before factoextra
library(factoextra)
library(caret)
library(stats)
library(ggfortify)
library(ggplot2)
library(corrplot)
library(cowplot)
library(ggcorrplot)
library(ComplexHeatmap)
library(circlize)
library(ggExtra)
library(FactoMineR)
library(here)

```


## Process data

**Required files**

```{r}
# Genes in long table format
all_degs_p_05_vac_infected = readRDS(here("DGE analysis", "all_studies_degs_weighted.rds")) %>% 
  filter(if_else(n >= 5, fdr <= 0.05, fdr <= (0.05 * 6/n))) 

# Sample or condition annotation
ann_vaccines <- read_csv(here("Annotations and metadata", "ann_vaccines_conditions.csv"))

data_annotation = ann_vaccines %>%
  mutate(day = as.factor(day),
         dose = as.factor(dose)) 
```

**Filtering data**

```{r}
# Gene set data.
ImmuneGO_BTM_grouped_genes <- read_csv("VaxGO gene sets/ImmuneGO_BTM_grouped_genes.csv")

ImmuneGO_Genes = all_degs_p_05_vac_infected %>%
  ungroup() %>% 
  inner_join(ImmuneGO_BTM_grouped_genes %>% select(genes) %>% distinct(), by = "genes") %>%
  filter(if_else(n >= 5, fdr <= 0.05, fdr <= (0.05 * 6/n))) 
  
```



### PCA: COVID-19 infection and vaccination conditions only

**Input**

```{r}
#INPUT
data_genes = ImmuneGO_Genes 
filename = "COVID_Infec_Vac_ImmuneGO_BTM"

#Convert long to wide table format.
#Dataframe with sample annotation

ann_vaccines_pca_matrix = data_genes %>% 
  select(condition, genes, log2fold_change) %>% 
  distinct() %>% 
  pivot_wider(names_from = "condition", 
              values_from = "log2fold_change") %>% 
  replace(is.na(.), 0) %>%
  column_to_rownames("genes") %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column("condition") %>% 
  inner_join(data_annotation %>% 
               select(condition, disease_vac, vaccine, day, week, dose, infection, type), by = "condition") 
dim(ann_vaccines_pca_matrix) #1786 genes


# Convert dataframe to matrix without annotation columns
ann_vaccines_pca_matrix_ready = ann_vaccines_pca_matrix %>% 
  column_to_rownames("condition") %>% 
  select(!disease_vac:type) %>% 
  as.matrix()

dim(ann_vaccines_pca_matrix_ready) #1778 genes

```


#### With normalization
```{r cache = TRUE}
# Delete columns with near 0 variance
nearZeroVarCols <- nearZeroVar(ann_vaccines_pca_matrix_ready, saveMetrics = TRUE)
ann_vaccines_pca_matrix_ready <- ann_vaccines_pca_matrix_ready[, !nearZeroVarCols$nzv]
dim(ann_vaccines_pca_matrix_ready)
pca_res <- prcomp(ann_vaccines_pca_matrix_ready, scale. = T)
# pca_res <- prcomp(ann_vaccines_pca_matrix_ready, scale. = F)

# Correlation plot ----
corr_matrix = cor(ann_vaccines_pca_matrix_ready %>% scale())
# corr_matrix = cor(ann_vaccines_pca_matrix_ready)
var <- get_pca_var(pca_res)

# corrplot = ggcorrplot(corr_matrix, hc.order = TRUE) + 
#   theme(axis.text.x = element_text(angle = 90, size = 5), 
#         axis.text.y = element_text(size = 5))

# ggsave(corrplot, file = paste0(filename, "_corrplot.png"), 
#        width = 20, #Grande 20, pequeno 10
#        height= 20) #Grande 20, pequeno 10
# print(corrplot)

#PCA ----
data.pca = prcomp(corr_matrix) #PCA

#Scree plot ----

scree_plot = fviz_eig(data.pca, 
                      addlabels = TRUE, main = "", ylab = "Explained variances (%)",
                      ylim = c(0, 100)) +
  theme_few() +
  scale_fill_npg()


#Save
ggsave(scree_plot, file = here("Figures", paste0(filename, "_screeplot.png")), width = 5, height = 3) 
scree_plot

```

Loadings plot

```{r fig.height=5, fig.width=5}
# Loadings plot ----
options(ggrepel.max.overlaps = Inf)
PC1_top20_pos = pca_res$rotation %>% 
  as.data.frame() %>% 
  arrange(-PC1) %>% 
  slice_head(n = 5) %>%
  rownames_to_column("genes") %>% 
  pull(genes)

PC1_top20_neg = pca_res$rotation %>% 
  as.data.frame() %>% 
  arrange(PC1) %>% 
  slice_head(n = 5) %>%
  rownames_to_column("genes") %>% 
  pull(genes)

PC2_top20_pos = pca_res$rotation %>% 
  as.data.frame() %>% 
  arrange(PC2) %>% 
  slice_head(n = 5) %>%
  rownames_to_column("genes") %>% 
  pull(genes)

PC2_top20_neg = pca_res$rotation %>% 
  as.data.frame() %>% 
  arrange(-PC2) %>% 
  slice_head(n = 5) %>%
  rownames_to_column("genes") %>% 
  pull(genes)


circle_contrib= fviz_pca_var(pca_res, 
                             col.var = "cos2",
                              gradient.cols = c("#DC0000FF", "black"),
                               col.ind = "black",
                              # select.var= list(cos2 = 20), 
                              select.var = list(name = c(PC1_top20_pos, PC1_top20_neg, PC2_top20_pos, PC2_top20_neg)),
                              repel = T, 
                              labelsize = 5, 
                              col.circle = "gray90") +
  labs(title = "",
       x = "PC1",
       y = "PC2") +
  geom_vline(xintercept = 0, color = "white") +
  geom_hline(yintercept = 0, color = "white") +
  theme_void() +
  theme(panel.grid = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank(),
        panel.grid.major = element_blank(),
        axis.line = element_line(size = 0, colour = "white", linetype=1)) 

#Save
ggsave(circle_contrib, file = here("Figures", paste0(filename, "_circle_contrib_wide.png")), width = 5, height = 5)

circle_contrib


```

KNN

```{r}
#KNN classification -----
#Use the screeplot generated % of explained variance for each dimension


# PC1-PC2 ------
#Determine the number of clusters
pca_scores <- data.frame(pca_res$x[, 1:2])
fviz_nbclust(pca_scores,  
                     FUNcluster = kmeans,
                     method = "wss")

pcas = "PC1-PC2"

set.seed(666) # Set seed for randomization
cluster_model <- kmeans(pca_res$x[, 1:2], centers = 3)  #Set the number of clusters

ann_vaccines_pca_matrix$cluster <- as.factor(cluster_model$cluster)
```


#### Plot PC1-2
```{r fig.height=7, fig.width=7, paged.print=FALSE, cache=TRUE}
# Condition with density plot ------

# ann_vaxsig_colors_linebreak = ann_vaxsig_colors
# names(ann_vaxsig_colors_linebreak$condition_col) <- str_replace(names(ann_vaxsig_colors_linebreak$condition_col), " \\(", "\n(")

pca_plot_knn_cluster_labels = autoplot(pca_res, 
        data = ann_vaccines_pca_matrix)  +
  # stat_ellipse(aes(color = cluster, fill = cluster),
  #              geom = "polygon",
  #              alpha = 0.2,
  #              linetype = 1,
  #              size = 0.3,
  #              type = "t") +
  geom_convexhull(aes(fill = cluster, colour = cluster), alpha = 0.2) +
  labs(title=paste0(filename, "_bycondition"),
       x = "PC1 (71.8%)",
       y = "PC2 (13%)") + 
  # geom_hline(yintercept = 0, size = 0.8, color = "gray50", linetype = 1, alpha = 1) +
  # geom_vline(xintercept = 0, size = 0.8, color = "gray50", linetype = 1, alpha = 1) +
  geom_point(color = "black",
             size = 4,
             stroke = 1) +
  geom_point(aes(fill = condition,
                 color = condition),
             size = 3,
             stroke = 1) +
   scale_color_manual(values = c("1" = "#DC0000FF", 
                               "2" = "#4DBBD5FF", 
                               "3" = "#F5BB00",
                               "4" = "#8491B4FF",
                                ann_vaxsig_colors$condition_col), 
                     guide = "none") +
  scale_fill_manual(values = c("1" = "#DC0000FF", 
                               "2" = "#4DBBD5FF", 
                               "3" = "#F5BB00",
                               "4" = "#8491B4FF",
                               ann_vaxsig_colors$condition_col),
                    guide = "none") + 
  guides(col="none") +
  theme_few() +
  theme(axis.text = element_text(size = 10, color = "black"),
      panel.grid.minor = element_blank(),
      panel.grid.major = element_line(size = 1, 
                                    linetype = 1,
                                    color = "gray90"),
        axis.line = element_line(size = 0, colour = "black", linetype=1),
        # axis.line = element_line(size = 1, colour = "black", linetype=1),
        text = element_text(color = "black", size = 10),
        panel.border = element_rect(fill = NA, color = "gray90", size = 2))   +
  geom_label_repel(aes(label = if_else(cluster != 2 | type == "I" & condition != c("I-BNT (D5)", "BNT-I-BNT (D51, hospital)"), condition, NA)),
                  segment.colour="black",
                  box.padding = 0.5, 
                  # fill = "gray60",
                  # color = "white",
                  size = 3.5)  +
  xlim(-0.3, 0.8)
pca_plot_knn_cluster_labels

ggsave(pca_plot_knn_cluster_labels, 
       file = here(file = "Figures", paste0(filename, pcas, "_KNN_Clustered_labels.png")), width = 7, height = 7)
```




#### Without normalization
```{r cache = TRUE}
# Delete columns with near 0 variance
nearZeroVarCols <- nearZeroVar(ann_vaccines_pca_matrix_ready, saveMetrics = TRUE)
ann_vaccines_pca_matrix_ready <- ann_vaccines_pca_matrix_ready[, !nearZeroVarCols$nzv]
dim(ann_vaccines_pca_matrix_ready)
# pca_res <- prcomp(ann_vaccines_pca_matrix_ready, scale. = T)
pca_res <- prcomp(ann_vaccines_pca_matrix_ready, scale. = F)

# Correlation plot ----
# corr_matrix = cor(ann_vaccines_pca_matrix_ready %>% scale()) 
corr_matrix = cor(ann_vaccines_pca_matrix_ready)
var <- get_pca_var(pca_res)

# corrplot = ggcorrplot(corr_matrix, hc.order = TRUE) + 
#   theme(axis.text.x = element_text(angle = 90, size = 5), 
#         axis.text.y = element_text(size = 5))

# ggsave(corrplot, file = paste0(filename, "_corrplot.png"), 
#        width = 20, #Grande 20, pequeno 10
#        height= 20) #Grande 20, pequeno 10
# print(corrplot)

#PCA ----
data.pca = prcomp(corr_matrix) #PCA

#Scree plot ----

scree_plot = fviz_eig(data.pca, 
                      addlabels = TRUE, main = "", ylab = "Explained variances (%)",
                      ylim = c(0, 100)) +
  theme_few() +
  scale_fill_npg()


#Save
ggsave(scree_plot, file = here("Figures", paste0(filename, "_NoNormal_screeplot.png")), width = 5, height = 3) 
scree_plot

```

Loadings plot

```{r fig.height=5, fig.width=5}
# Loadings plot ----
options(ggrepel.max.overlaps = Inf)
PC1_top20_pos = pca_res$rotation %>% 
  as.data.frame() %>% 
  arrange(-PC1) %>% 
  slice_head(n = 5) %>%
  rownames_to_column("genes") %>% 
  pull(genes)

PC1_top20_neg = pca_res$rotation %>% 
  as.data.frame() %>% 
  arrange(PC1) %>% 
  slice_head(n = 5) %>%
  rownames_to_column("genes") %>% 
  pull(genes)

PC2_top20_pos = pca_res$rotation %>% 
  as.data.frame() %>% 
  arrange(PC2) %>% 
  slice_head(n = 5) %>%
  rownames_to_column("genes") %>% 
  pull(genes)

PC2_top20_neg = pca_res$rotation %>% 
  as.data.frame() %>% 
  arrange(-PC2) %>% 
  slice_head(n = 5) %>%
  rownames_to_column("genes") %>% 
  pull(genes)


circle_contrib= fviz_pca_var(pca_res, 
                             col.var = "cos2",
                              gradient.cols = c("#DC0000FF", "black"),
                               col.ind = "black",
                              # select.var= list(cos2 = 20), 
                              select.var = list(name = c(PC1_top20_pos, PC1_top20_neg, PC2_top20_pos, PC2_top20_neg)),
                              repel = T, 
                              labelsize = 5, 
                              col.circle = "gray90") +
  labs(title = "",
       x = "PC1",
       y = "PC2") +
  geom_vline(xintercept = 0, color = "white") +
  geom_hline(yintercept = 0, color = "white") +
  theme_void() +
  theme(panel.grid = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank(),
        panel.grid.major = element_blank(),
        axis.line = element_line(size = 0, colour = "white", linetype=1)) 

#Save
ggsave(circle_contrib, file = here("Figures", paste0(filename, "_NoNormal__circle_contrib_wide.png")), width = 5, height = 5)

circle_contrib


```

KNN

```{r}
#KNN classification -----
#Use the screeplot generated % of explained variance for each dimension


# PC1-PC2 ------
#Determine the number of clusters
pca_scores <- data.frame(pca_res$x[, 1:2])
fviz_nbclust(pca_scores,  
                     FUNcluster = kmeans,
                     method = "wss")

pcas = "PC1-PC2"

set.seed(666) # Set seed for randomization
cluster_model <- kmeans(pca_res$x[, 1:2], centers = 3)  #Set the number of clusters

ann_vaccines_pca_matrix$cluster <- as.factor(cluster_model$cluster)
```


#### Plot PC1-2
```{r fig.height=7, fig.width=7, paged.print=FALSE, cache=TRUE}
# Condition with density plot ------

# ann_vaxsig_colors_linebreak = ann_vaxsig_colors
# names(ann_vaxsig_colors_linebreak$condition_col) <- str_replace(names(ann_vaxsig_colors_linebreak$condition_col), " \\(", "\n(")

pca_plot_knn_cluster_labels = autoplot(pca_res, 
        data = ann_vaccines_pca_matrix)  +
  # stat_ellipse(aes(color = cluster, fill = cluster),
  #              geom = "polygon",
  #              alpha = 0.2,
  #              linetype = 1,
  #              size = 0.3,
  #              type = "t") +
  geom_convexhull(aes(fill = cluster, colour = cluster), alpha = 0.2) +
  labs(title=paste0(filename, "_bycondition"),
       x = "PC1 (71.8%)",
       y = "PC2 (13%)") + 
  # geom_hline(yintercept = 0, size = 0.8, color = "gray50", linetype = 1, alpha = 1) +
  # geom_vline(xintercept = 0, size = 0.8, color = "gray50", linetype = 1, alpha = 1) +
  geom_point(color = "black",
             size = 4,
             stroke = 1) +
  geom_point(aes(fill = condition,
                 color = condition),
             size = 3,
             stroke = 1) +
   scale_color_manual(values = c("1" = "#DC0000FF", 
                               "2" = "#4DBBD5FF", 
                               "3" = "#F5BB00",
                               "4" = "#8491B4FF",
                                ann_vaxsig_colors$condition_col), 
                     guide = "none") +
  scale_fill_manual(values = c("1" = "#DC0000FF", 
                               "2" = "#4DBBD5FF", 
                               "3" = "#F5BB00",
                               "4" = "#8491B4FF",
                               ann_vaxsig_colors$condition_col),
                    guide = "none") + 
  guides(col="none") +
  theme_few() +
  theme(axis.text = element_text(size = 10, color = "black"),
      panel.grid.minor = element_blank(),
      panel.grid.major = element_line(size = 1, 
                                    linetype = 1,
                                    color = "gray90"),
        axis.line = element_line(size = 0, colour = "black", linetype=1),
        # axis.line = element_line(size = 1, colour = "black", linetype=1),
        text = element_text(color = "black", size = 10),
        panel.border = element_rect(fill = NA, color = "gray90", size = 2))   +
  geom_label_repel(aes(label = if_else(cluster != 2 | type == "I" & condition != c("I-BNT (D5)", "BNT-I-BNT (D51, hospital)"), condition, NA)),
                  segment.colour="black",
                  box.padding = 0.5, 
                  # fill = "gray60",
                  # color = "white",
                  size = 3.5)  +
  xlim(-0.3, 0.8)
pca_plot_knn_cluster_labels

ggsave(pca_plot_knn_cluster_labels, 
       file = here(file = "Figures", paste0(filename, pcas, "_NoNormal__KNN_Clustered_labels.png")), width = 7, height = 7)
```








#### Heatmap of PC1 and PC2 genes

Inputs
```{r}
####### INPUT
PC1 = var$cos2 %>%
  as.data.frame() %>%
  arrange(desc(Dim.1)) %>%
  filter(Dim.1 >= 0.5) %>% 
  #slice_head(n=20) %>% 
  rownames_to_column("genes") %>% 
  distinct() %>% 
  select(genes, cos2 = Dim.1) 

PC1_top20_pos = pca_res$rotation %>% 
  as.data.frame() %>% 
  arrange(-PC1) %>% 
  slice_head(n = 5) %>%
  rownames_to_column("genes") 

PC1_top20_neg = pca_res$rotation %>% 
  as.data.frame() %>% 
  arrange(PC1) %>% 
  slice_head(n = 5) %>%
  rownames_to_column("genes") 

PC2_top20_pos = pca_res$rotation %>% 
  as.data.frame() %>% 
  arrange(PC2) %>% 
  slice_head(n = 5) %>%
  rownames_to_column("genes") 

PC2_top20_neg = pca_res$rotation %>% 
  as.data.frame() %>% 
  arrange(-PC2) %>% 
  slice_head(n = 5) %>%
  rownames_to_column("genes") 



PC2 = var$cos2 %>%
  as.data.frame() %>%
  filter(Dim.2 >= 0.5) %>% 
  #slice_head(n=20) %>% 
  rownames_to_column("genes") %>% 
  distinct() %>% 
  select(genes, cos2 = Dim.2) %>% 
  mutate(dimension = "PC2")

PC1_2 = bind_rows(PC1, PC2) 
PC1_2 %>% 
  select(genes) %>% distinct()

PC1_2 %>% write_csv(file = here("PCA", paste(filename, "PC_1_2_cos2.csv")))

```


#### PC1 genes heatmap

Horizontal
```{r}
# INPUT
ann_genes = ImmuneGO_Annotated_Genes
btm_annotation_genes <- readRDS(here("VaxGO gene sets", "btm_annotation_genes.rds"))
btm_annotation_genes_immune = btm_annotation_genes
pc = "PC1"
# data_2 =  PC1 %>% 
#   slice_max(order_by = cos2, n = 10) %>% 
#   inner_join(data_genes, by = "genes")

data_2 = PC1_top20_pos %>% 
  bind_rows(PC1_top20_neg) %>% 
  select(genes, value= PC1) %>% 
  inner_join(data_genes, by = "genes")

filename_2 = paste0(filename, "_PCA_",pc, "_COVID")

######## Matrix
matrix = data_2 %>% 
  select(genes, condition, log2fold_change) %>%
  pivot_wider(names_from = "condition", 
              values_from = "log2fold_change", 
              values_fn = mean) %>% 
  replace(is.na(.), 0) %>%
  arrange(genes) %>% 
  column_to_rownames(var="genes") %>% 
  as.matrix()

######## Annotations
ann_vaccines_covid = data_2 %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  select(condition, disease_vac, type, day) %>% 
  distinct()

#Columns
ann_cols_heatmap = matrix %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  inner_join(ann_vaccines_covid, by = "condition") %>% 
  distinct() %>% 
  arrange(condition) %>% 
  column_to_rownames("condition")

matrix_data = matrix %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  inner_join(ann_cols_heatmap %>% rownames_to_column("condition"), by = "condition") %>% 
  select(!c(disease_vac:day)) %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>%
  as.matrix()

#Rows Immune ----
ann_rows_immune = matrix_data %>%
  as.data.frame() %>%
  rownames_to_column("genes") %>%
  left_join(ImmuneGO_Annotated_Genes %>% select(genes) %>% distinct() %>% mutate(set = "ImmuneGO"), by = "genes") %>%
  left_join(btm_annotation_genes_immune %>% select(genes) %>% distinct() %>% mutate(set2 = "BTM"), by = "genes") %>%
  select(genes, set, set2) %>%
  distinct() %>%
  group_by(genes) %>%
  arrange(genes, factor(set, levels = c("ImmuneGO", "BTM")), .by_group = TRUE) %>%
  column_to_rownames("genes") %>%
  as.data.frame() %>%
  mutate(set = if_else(is.na(set), set2, set),
         set2 = if_else(set == set2, NA, set2)) %>% 
  replace(is.na(.), "None") %>% 
  select(set2, set)

# Reorder columns (same order as lines in the annotation table)
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_covid, by.y = "condition", all.x=TRUE) %>% 
  arrange(type, condition) %>% 
  select(!c(disease_vac, type, day)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0) %>% 
  as.matrix() 


#Check dimensions
dim(matrix_data_ordered) 
dim(ann_vaccines_heatmap) 
dim(ann_rows_immune)

col_immune_sets = list(
  "set" = c("BTM" = "#5e60ce",
          "ImmuneGO" = "#00A087FF",
          "None"="white"),
  "set2" = c("BTM" = "#5e60ce",
          "ImmuneGO" = "#00A087FF",
          "None"="white"))


ann_cols_heatmap <- ann_cols_heatmap %>%
  mutate(day = factor(day, levels = unique(day[order(day)])))

#Heatmap annotation -----
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "right")

row_ha = HeatmapAnnotation(df = ann_rows_immune,
                            which = "row",
                       col = col_immune_sets, #col_process_immune / col_process_notimmune
                       show_annotation_name = F,
                       show_legend = T,
                       simple_anno_size = unit(2, "mm"))

# Values colors -----
col_fun <- colorRamp2(c(-2, 0, 2), c("#4DBBD5FF", "white", "#ed6a5a"))


# Parameters -----
file = filename_2
mat = matrix_data
width_image = 20
height_image = 15
width = unit(12, "cm")
height = unit(5, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)
rect_gp = gpar(col = "black", lwd = 0.5)

fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap.png")

png(file = here("Figures", fileimageheatmap_grouped), 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        name = "L2FC",
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        col = col_fun, #color
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(2, "cm"),
                        column_split = NULL,
                        row_split = NULL,
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height 
)

draw(heatmap_plot, 
     annotation_legend_side = "right", 
     heatmap_legend_side = "bottom")
dev.off()
```


#### PC2 genes heatmap
```{r}
# INPUT
ann_genes = ImmuneGO_Annotated_Genes
pc = "PC2"
# data_2 =  PC2 %>% 
#   slice_max(order_by = cos2, n = 10) %>% 
#   inner_join(data_genes, by = "genes")

data_2 = PC2_top20_pos %>% 
  bind_rows(PC2_top20_neg) %>% 
  select(genes, value= PC2) %>% 
  inner_join(data_genes, by = "genes")


filename_2 = paste0(filename, "_PCA_",pc, "_COVID")

######## Matrix
matrix = data_2 %>% 
  select(genes, condition, log2fold_change) %>%
  pivot_wider(names_from = "condition", 
              values_from = "log2fold_change", 
              values_fn = mean) %>% 
  replace(is.na(.), 0) %>%
  arrange(genes) %>% 
  column_to_rownames(var="genes") %>% 
  as.matrix()

######## Annotations
ann_vaccines_covid = data_2 %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  select(condition, disease_vac, type, day) %>% 
  distinct()

#Columns
ann_cols_heatmap = matrix %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  inner_join(ann_vaccines_covid, by = "condition") %>% 
  distinct() %>% 
  arrange(condition) %>% 
  column_to_rownames("condition")

matrix_data = matrix %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  inner_join(ann_cols_heatmap %>% rownames_to_column("condition"), by = "condition") %>% 
  select(!c(disease_vac:day)) %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>%
  as.matrix()

#Rows Immune ----
ann_rows_immune = matrix_data %>%
  as.data.frame() %>%
  rownames_to_column("genes") %>%
  left_join(ImmuneGO_Annotated_Genes %>% select(genes) %>% distinct() %>% mutate(set = "ImmuneGO"), by = "genes") %>%
  left_join(btm_annotation_genes_immune %>% select(genes) %>% distinct() %>% mutate(set2 = "BTM"), by = "genes") %>%
  select(genes, set, set2) %>%
  distinct() %>%
  group_by(genes) %>%
  arrange(genes, factor(set, levels = c("ImmuneGO", "BTM")), .by_group = TRUE) %>%
  column_to_rownames("genes") %>%
  as.data.frame() %>%
  mutate(set = if_else(is.na(set), set2, set),
         set2 = if_else(set == set2, NA, set2)) %>% 
  replace(is.na(.), "None") %>% 
  select(set2, set)

# Reorder columns (same order as lines in the annotation table)
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_covid, by.y = "condition", all.x=TRUE) %>% 
  arrange(type, condition) %>% 
  select(!c(disease_vac, type, day)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0) %>% 
  as.matrix() 


#Check dimensions
dim(matrix_data_ordered) 
dim(ann_vaccines_heatmap) 
dim(ann_rows_immune)

col_immune_sets = list(
  "set" = c("BTM" = "#5e60ce",
          "ImmuneGO" = "#00A087FF",
          "None"="white"),
  "set2" = c("BTM" = "#5e60ce",
          "ImmuneGO" = "#00A087FF",
          "None"="white"))

ann_cols_heatmap <- ann_cols_heatmap %>%
  mutate(day = factor(day, levels = unique(day[order(day)])))


#Heatmap annotation -----
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "right")

row_ha = HeatmapAnnotation(df = ann_rows_immune,
                            which = "row",
                       col = col_immune_sets, #col_process_immune / col_process_notimmune
                       show_annotation_name = F,
                       show_legend = T,
                       simple_anno_size = unit(2, "mm"))

# Values colors -----
col_fun <- colorRamp2(c(-2, 0, 2), c("#4DBBD5FF", "white", "#ed6a5a"))


# Parameters -----
file = filename_2
mat = matrix_data
width_image = 15
height_image = 15
width = unit(7, "cm")
height = unit(5, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)
rect_gp = gpar(col = "black", lwd = 0.5)

fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap.png")

png(file = here("Figures", fileimageheatmap_grouped), 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        name = "L2FC",
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        col = col_fun, #color
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(2, "cm"),
                        column_split = NULL,
                        row_split = NULL,
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height 
)

draw(heatmap_plot, 
     annotation_legend_side = "right", 
     heatmap_legend_side = "bottom")
dev.off()

```

```{r}
## Get row and column orders
ht_list = draw(heatmap_plot)
rows_clustered = rownames(mat)[row_order(ht_list)]
cols_clustered = colnames(mat)[column_order(ht_list)]

mat_2 = mat %>% 
  as.data.frame() %>% 
  rownames_to_column("genes") %>% 
  filter(genes %in% c(rev(rows_clustered))) %>% 
  select(genes, c(cols_clustered)) %>% 
  column_to_rownames("genes") %>% 
  as.matrix()

width = unit(15, "cm")
height = unit(5, "cm")

fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap_ForPCA.png")

png(file = here("Figures", fileimageheatmap_grouped), 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot_2 = Heatmap(mat_2,
                        # top_annotation = ha,
                        name = "L2FC",
                        # left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        col = col_fun, #color
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 90,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = F,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(2, "cm"),
                        column_split = NULL,
                        row_split = NULL,
                        row_gap = unit(2, "mm"),
                        cluster_columns = F,
                        column_gap = unit(8, "mm"),
                        column_names_side = "top",
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = F,
                        width = width,
                        height = height 
)

draw(heatmap_plot_2, 
     annotation_legend_side = "right", 
     heatmap_legend_side = "bottom")
dev.off()
```

### PCA: COVID-19 vaccination conditions only

**Required files**

```{r}
# Genes in long table format
all_degs_p_05_vac_infected = readRDS(here("DGE analysis", "all_studies_degs_weighted.rds")) %>% 
  filter(if_else(n >= 5, fdr <= 0.05, fdr <= (0.05 * 6/n))) 

# Sample or condition annotation
ann_vaccines <- read_csv(here("Annotations and metadata", "ann_vaccines_conditions.csv"))

data_annotation = ann_vaccines %>%
  mutate(day = as.factor(day),
         dose = as.factor(dose)) 
```


**Filtering data**

```{r}
# Gene set data.
ImmuneGO_BTM_grouped_genes <- read_csv("VaxGO gene sets/ImmuneGO_BTM_grouped_genes.csv")

ImmuneGO_Genes = all_degs_p_05_vac_infected %>%
  ungroup() %>% 
  inner_join(ImmuneGO_BTM_grouped_genes %>% select(genes) %>% distinct(), by = "genes") %>%
  inner_join(data_annotation, by = "condition") %>% 
  filter(disease_vac == "V")
  
  
```



**Input**

```{r}
#INPUT
data_genes = ImmuneGO_Genes 
filename = "COVID_VacOnly_ImmuneGO_BTM"

#Convert long to wide table format.
#Dataframe with sample annotation

ann_vaccines_pca_matrix = data_genes %>% 
  select(condition, genes, log2fold_change) %>% 
  distinct() %>% 
  pivot_wider(names_from = "condition", 
              values_from = "log2fold_change") %>% 
  replace(is.na(.), 0) %>%
  column_to_rownames("genes") %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column("condition") %>% 
  inner_join(data_annotation %>% 
               select(condition, disease_vac, vaccine, day, week, dose, infection, type), by = "condition") 

dim(ann_vaccines_pca_matrix) #1704 genes


# Convert dataframe to matrix without annotation columns
ann_vaccines_pca_matrix_ready = ann_vaccines_pca_matrix %>% 
  column_to_rownames("condition") %>% 
  select(!disease_vac:type) %>% 
  as.matrix()

dim(ann_vaccines_pca_matrix_ready) #1696 genes

```


```{r cache = TRUE}
# Delete columns with near 0 variance
nearZeroVarCols <- nearZeroVar(ann_vaccines_pca_matrix_ready, saveMetrics = TRUE)
ann_vaccines_pca_matrix_ready <- ann_vaccines_pca_matrix_ready[, !nearZeroVarCols$nzv]
dim(ann_vaccines_pca_matrix_ready)
pca_res <- prcomp(ann_vaccines_pca_matrix_ready, scale. = T)

# Correlation plot ----
corr_matrix = cor(ann_vaccines_pca_matrix_ready %>% scale()) 
var <- get_pca_var(pca_res)

# corrplot = ggcorrplot(corr_matrix, hc.order = TRUE) + 
#   theme(axis.text.x = element_text(angle = 90, size = 5), 
#         axis.text.y = element_text(size = 5))

# ggsave(corrplot, file = paste0(filename, "_corrplot.png"), 
#        width = 20, #Grande 20, pequeno 10
#        height= 20) #Grande 20, pequeno 10
# print(corrplot)

#PCA ----
data.pca = prcomp(corr_matrix) #PCA

#Scree plot ----

scree_plot = fviz_eig(data.pca, 
                      addlabels = TRUE, main = "", ylab = "Explained variances (%)",
                      ylim = c(0, 100)) +
  theme_few() +
  scale_fill_npg()


#Save
ggsave(scree_plot, file = here("Figures", paste0(filename, "_screeplot.png")), width = 5, height = 3) 
scree_plot

```

Loadings plot

```{r fig.height=5, fig.width=5}
# Loadings plot ----
options(ggrepel.max.overlaps = Inf)
PC1_top20_pos = pca_res$rotation %>% 
  as.data.frame() %>% 
  arrange(-PC1) %>% 
  slice_head(n = 5) %>%
  rownames_to_column("genes") %>% 
  pull(genes)

PC1_top20_neg = pca_res$rotation %>% 
  as.data.frame() %>% 
  arrange(PC1) %>% 
  slice_head(n = 5) %>%
  rownames_to_column("genes") %>% 
  pull(genes)

PC2_top20_pos = pca_res$rotation %>% 
  as.data.frame() %>% 
  arrange(PC2) %>% 
  slice_head(n = 5) %>%
  rownames_to_column("genes") %>% 
  pull(genes)

PC2_top20_neg = pca_res$rotation %>% 
  as.data.frame() %>% 
  arrange(-PC2) %>% 
  slice_head(n = 5) %>%
  rownames_to_column("genes") %>% 
  pull(genes)


circle_contrib= fviz_pca_var(pca_res, 
                             col.var = "cos2",
                              gradient.cols = c("#DC0000FF", "black"),
                               col.ind = "black",
                              # select.var= list(cos2 = 20), 
                              select.var = list(name = c(PC1_top20_pos, PC1_top20_neg, PC2_top20_pos, PC2_top20_neg)),
                              repel = T, 
                              labelsize = 5, 
                              col.circle = "gray90") +
  labs(title = "",
       x = "PC1",
       y = "PC2") +
  geom_vline(xintercept = 0, color = "white") +
  geom_hline(yintercept = 0, color = "white") +
  theme_void() +
  theme(panel.grid = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank(),
        panel.grid.major = element_blank(),
        axis.line = element_line(size = 0, colour = "white", linetype=1)) 

#Save
ggsave(circle_contrib, file = here("Figures", paste0(filename, "_circle_contrib_wide.png")), width = 5, height = 5)

circle_contrib


```

KNN

```{r}
#KNN classification -----
#Use the screeplot generated % of explained variance for each dimension
scree_plot #Dim 1 = 68.2%, Dim 2 = 15.9%


# PC1-PC2 ------
#Determine the number of clusters
pca_scores <- data.frame(pca_res$x[, 1:2])
fviz_nbclust(pca_scores,  
                     FUNcluster = kmeans,
                     method = "wss")

pcas = "PC1-PC2"

set.seed(666) # Set seed for randomization
cluster_model <- kmeans(pca_res$x[, 1:2], centers = 4)  #Set the number of clusters

ann_vaccines_pca_matrix$cluster <- as.factor(cluster_model$cluster)
```


#### Plot PC1-2
```{r fig.height=7, fig.width=7, paged.print=FALSE, cache=TRUE}
# Condition with density plot ------

# ann_vaxsig_colors_linebreak = ann_vaxsig_colors
# names(ann_vaxsig_colors_linebreak$condition_col) <- str_replace(names(ann_vaxsig_colors_linebreak$condition_col), " \\(", "\n(")

pca_plot_knn_cluster_labels = autoplot(pca_res, 
        data = ann_vaccines_pca_matrix)  +
  # stat_ellipse(aes(color = cluster, fill = cluster),
  #              geom = "polygon",
  #              alpha = 0.2,
  #              linetype = 1,
  #              size = 0.3,
  #              type = "t") +
  geom_convexhull(aes(fill = cluster, colour = cluster), alpha = 0.2) +
  labs(title=paste0(filename, "_bycondition"),
       x = "PC1 (71.8%)",
       y = "PC2 (13%)") + 
  # geom_hline(yintercept = 0, size = 0.8, color = "gray50", linetype = 1, alpha = 1) +
  # geom_vline(xintercept = 0, size = 0.8, color = "gray50", linetype = 1, alpha = 1) +
  geom_point(color = "black",
             size = 4,
             stroke = 1) +
  geom_point(aes(fill = condition,
                 color = condition),
             size = 3,
             stroke = 1) +
   scale_color_manual(values = c("1" = "#DC0000FF", 
                               "2" = "#4DBBD5FF", 
                               "3" = "#F5BB00",
                               "4" = "#8491B4FF",
                                ann_vaxsig_colors$condition_col), 
                     guide = "none") +
  scale_fill_manual(values = c("1" = "#DC0000FF", 
                               "2" = "#4DBBD5FF", 
                               "3" = "#F5BB00",
                               "4" = "#8491B4FF",
                               ann_vaxsig_colors$condition_col),
                    guide = "none") + 
  guides(col="none") +
  theme_few() +
  theme(axis.text = element_text(size = 10, color = "black"),
      panel.grid.minor = element_blank(),
      panel.grid.major = element_line(size = 1, 
                                    linetype = 1,
                                    color = "gray90"),
        axis.line = element_line(size = 0, colour = "black", linetype=1),
        # axis.line = element_line(size = 1, colour = "black", linetype=1),
        text = element_text(color = "black", size = 10),
        panel.border = element_rect(fill = NA, color = "gray90", size = 2))   +
  geom_label_repel(aes(label = if_else(cluster != 2 | type == "I" & condition != c("I-BNT (D5)", "BNT-I-BNT (D51, hospital)"), condition, NA)),
                  segment.colour="black",
                  box.padding = 0.5, 
                  # fill = "gray60",
                  # color = "white",
                  size = 3.5)  +
  xlim(-0.3, 0.8)
pca_plot_knn_cluster_labels

ggsave(pca_plot_knn_cluster_labels, 
       file = here(file = "Figures", paste0(filename, pcas, "_KNN_Clustered_labels.png")), width = 7, height = 7)
```









#### Heatmap of PC1 and PC2 genes

Inputs
```{r}
####### INPUT
PC1 = var$cos2 %>%
  as.data.frame() %>%
  arrange(desc(Dim.1)) %>%
  filter(Dim.1 >= 0.5) %>% 
  #slice_head(n=20) %>% 
  rownames_to_column("genes") %>% 
  distinct() %>% 
  select(genes, cos2 = Dim.1) 

PC1_top20_pos = pca_res$rotation %>% 
  as.data.frame() %>% 
  arrange(-PC1) %>% 
  slice_head(n = 5) %>%
  rownames_to_column("genes") 

PC1_top20_neg = pca_res$rotation %>% 
  as.data.frame() %>% 
  arrange(PC1) %>% 
  slice_head(n = 5) %>%
  rownames_to_column("genes") 

PC2_top20_pos = pca_res$rotation %>% 
  as.data.frame() %>% 
  arrange(PC2) %>% 
  slice_head(n = 5) %>%
  rownames_to_column("genes") 

PC2_top20_neg = pca_res$rotation %>% 
  as.data.frame() %>% 
  arrange(-PC2) %>% 
  slice_head(n = 5) %>%
  rownames_to_column("genes") 



PC2 = var$cos2 %>%
  as.data.frame() %>%
  filter(Dim.2 >= 0.5) %>% 
  #slice_head(n=20) %>% 
  rownames_to_column("genes") %>% 
  distinct() %>% 
  select(genes, cos2 = Dim.2) %>% 
  mutate(dimension = "PC2")

PC1_2 = bind_rows(PC1, PC2) 
PC1_2 %>% 
  select(genes) %>% distinct()

PC1_2 %>% write_csv(file = here("PCA", paste(filename, "PC_1_2_cos2.csv")))

```


#### PC1 genes heatmap

Horizontal
```{r}
# INPUT
ann_genes = ImmuneGO_Annotated_Genes
btm_annotation_genes <- readRDS(here("VaxGO gene sets", "btm_annotation_genes.rds"))
btm_annotation_genes_immune = btm_annotation_genes
pc = "PC1"
# data_2 =  PC1 %>% 
#   slice_max(order_by = cos2, n = 10) %>% 
#   inner_join(data_genes, by = "genes")

data_2 = PC1_top20_pos %>% 
  bind_rows(PC1_top20_neg) %>% 
  select(genes, value= PC1) %>% 
  inner_join(data_genes, by = "genes")

filename_2 = paste0(filename, "_PCA_",pc, "_COVID")

######## Matrix
matrix = data_2 %>% 
  select(genes, condition, log2fold_change) %>%
  pivot_wider(names_from = "condition", 
              values_from = "log2fold_change", 
              values_fn = mean) %>% 
  replace(is.na(.), 0) %>%
  arrange(genes) %>% 
  column_to_rownames(var="genes") %>% 
  as.matrix()

######## Annotations
ann_vaccines_covid = data_2 %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  select(condition, disease_vac, type, day) %>% 
  distinct()

#Columns
ann_cols_heatmap = matrix %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  inner_join(ann_vaccines_covid, by = "condition") %>% 
  distinct() %>% 
  arrange(condition) %>% 
  column_to_rownames("condition")

matrix_data = matrix %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  inner_join(ann_cols_heatmap %>% rownames_to_column("condition"), by = "condition") %>% 
  select(!c(disease_vac:day)) %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>%
  as.matrix()

#Rows Immune ----
ann_rows_immune = matrix_data %>%
  as.data.frame() %>%
  rownames_to_column("genes") %>%
  left_join(ImmuneGO_Annotated_Genes %>% select(genes) %>% distinct() %>% mutate(set = "ImmuneGO"), by = "genes") %>%
  left_join(btm_annotation_genes_immune %>% select(genes) %>% distinct() %>% mutate(set2 = "BTM"), by = "genes") %>%
  select(genes, set, set2) %>%
  distinct() %>%
  group_by(genes) %>%
  arrange(genes, factor(set, levels = c("ImmuneGO", "BTM")), .by_group = TRUE) %>%
  column_to_rownames("genes") %>%
  as.data.frame() %>%
  mutate(set = if_else(is.na(set), set2, set),
         set2 = if_else(set == set2, NA, set2)) %>% 
  replace(is.na(.), "None") %>% 
  select(set2, set)

# Reorder columns (same order as lines in the annotation table)
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_covid, by.y = "condition", all.x=TRUE) %>% 
  arrange(type, condition) %>% 
  select(!c(disease_vac, type, day)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0) %>% 
  as.matrix() 


#Check dimensions
dim(matrix_data_ordered) 
dim(ann_vaccines_heatmap) 
dim(ann_rows_immune)

col_immune_sets = list(
  "set" = c("BTM" = "#5e60ce",
          "ImmuneGO" = "#00A087FF",
          "None"="white"),
  "set2" = c("BTM" = "#5e60ce",
          "ImmuneGO" = "#00A087FF",
          "None"="white"))


# ann_cols_heatmap <- ann_cols_heatmap %>%
#   mutate(day = factor(day, levels = unique(day[order(day)])))
  

#Heatmap annotation -----
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "right")

row_ha = HeatmapAnnotation(df = ann_rows_immune,
                            which = "row",
                       col = col_immune_sets, #col_process_immune / col_process_notimmune
                       show_annotation_name = F,
                       show_legend = T,
                       simple_anno_size = unit(2, "mm"))

# Values colors -----
col_fun <- colorRamp2(c(-2, 0, 2), c("#4DBBD5FF", "white", "#ed6a5a"))


# Parameters -----
file = filename_2
mat = matrix_data
width_image = 20
height_image = 15
width = unit(5, "cm")
height = unit(5, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)
rect_gp = gpar(col = "black", lwd = 0.5)

fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap.png")

png(file = here("Figures", fileimageheatmap_grouped), 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        name = "L2FC",
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        col = col_fun, #color
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(2, "cm"),
                        column_split = NULL,
                        row_split = NULL,
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height 
)

draw(heatmap_plot, 
     annotation_legend_side = "right", 
     heatmap_legend_side = "bottom")
dev.off()
```


#### PC2 genes heatmap
```{r}
# INPUT
ann_genes = ImmuneGO_Annotated_Genes
pc = "PC2"
# data_2 =  PC2 %>% 
#   slice_max(order_by = cos2, n = 10) %>% 
#   inner_join(data_genes, by = "genes")

data_2 = PC2_top20_pos %>% 
  bind_rows(PC2_top20_neg) %>% 
  select(genes, value= PC2) %>% 
  inner_join(data_genes, by = "genes")


filename_2 = paste0(filename, "_PCA_",pc, "_COVID")

######## Matrix
matrix = data_2 %>% 
  select(genes, condition, log2fold_change) %>%
  pivot_wider(names_from = "condition", 
              values_from = "log2fold_change", 
              values_fn = mean) %>% 
  replace(is.na(.), 0) %>%
  arrange(genes) %>% 
  column_to_rownames(var="genes") %>% 
  as.matrix()

######## Annotations
ann_vaccines_covid = data_2 %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  select(condition, disease_vac, type, day) %>% 
  distinct()

#Columns
ann_cols_heatmap = matrix %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  inner_join(ann_vaccines_covid, by = "condition") %>% 
  distinct() %>% 
  arrange(condition) %>% 
  column_to_rownames("condition")

matrix_data = matrix %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  inner_join(ann_cols_heatmap %>% rownames_to_column("condition"), by = "condition") %>% 
  select(!c(disease_vac:day)) %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>%
  as.matrix()

#Rows Immune ----
ann_rows_immune = matrix_data %>%
  as.data.frame() %>%
  rownames_to_column("genes") %>%
  left_join(ImmuneGO_Annotated_Genes %>% select(genes) %>% distinct() %>% mutate(set = "ImmuneGO"), by = "genes") %>%
  left_join(btm_annotation_genes_immune %>% select(genes) %>% distinct() %>% mutate(set2 = "BTM"), by = "genes") %>%
  select(genes, set, set2) %>%
  distinct() %>%
  group_by(genes) %>%
  arrange(genes, factor(set, levels = c("ImmuneGO", "BTM")), .by_group = TRUE) %>%
  column_to_rownames("genes") %>%
  as.data.frame() %>%
  mutate(set = if_else(is.na(set), set2, set),
         set2 = if_else(set == set2, NA, set2)) %>% 
  replace(is.na(.), "None") %>% 
  select(set2, set)

# Reorder columns (same order as lines in the annotation table)
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_covid, by.y = "condition", all.x=TRUE) %>% 
  arrange(type, condition) %>% 
  select(!c(disease_vac, type, day)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0) %>% 
  as.matrix() 


#Check dimensions
dim(matrix_data_ordered) 
dim(ann_vaccines_heatmap) 
dim(ann_rows_immune)

col_immune_sets = list(
  "set" = c("BTM" = "#5e60ce",
          "ImmuneGO" = "#00A087FF",
          "None"="white"),
  "set2" = c("BTM" = "#5e60ce",
          "ImmuneGO" = "#00A087FF",
          "None"="white"))

# ann_cols_heatmap <- ann_cols_heatmap %>%
#   mutate(day = factor(day, levels = unique(day[order(day)])))


#Heatmap annotation -----
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "right")

row_ha = HeatmapAnnotation(df = ann_rows_immune,
                            which = "row",
                       col = col_immune_sets, #col_process_immune / col_process_notimmune
                       show_annotation_name = F,
                       show_legend = T,
                       simple_anno_size = unit(2, "mm"))

# Values colors -----
col_fun <- colorRamp2(c(-2, 0, 2), c("#4DBBD5FF", "white", "#ed6a5a"))


# Parameters -----
file = filename_2
mat = matrix_data
width_image = 15
height_image = 15
width = unit(5, "cm")
height = unit(5, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)
rect_gp = gpar(col = "black", lwd = 0.5)

fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap.png")

png(file = here("Figures", fileimageheatmap_grouped), 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        name = "L2FC",
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        col = col_fun, #color
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(2, "cm"),
                        column_split = NULL,
                        row_split = NULL,
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height 
)

draw(heatmap_plot, 
     annotation_legend_side = "right", 
     heatmap_legend_side = "bottom")
dev.off()

```

```{r}
## Get row and column orders
ht_list = draw(heatmap_plot)
rows_clustered = rownames(mat)[row_order(ht_list)]
cols_clustered = colnames(mat)[column_order(ht_list)]

mat_2 = mat %>% 
  as.data.frame() %>% 
  rownames_to_column("genes") %>% 
  filter(genes %in% c(rev(rows_clustered))) %>% 
  select(genes, c(cols_clustered)) %>% 
  column_to_rownames("genes") %>% 
  as.matrix()

width = unit(15, "cm")
height = unit(5, "cm")

fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap_ForPCA.png")

png(file = here("Figures", fileimageheatmap_grouped), 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot_2 = Heatmap(mat_2,
                        # top_annotation = ha,
                        name = "L2FC",
                        # left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        col = col_fun, #color
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 90,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = F,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(2, "cm"),
                        column_split = NULL,
                        row_split = NULL,
                        row_gap = unit(2, "mm"),
                        cluster_columns = F,
                        column_gap = unit(8, "mm"),
                        column_names_side = "top",
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = F,
                        width = width,
                        height = height 
)

draw(heatmap_plot_2, 
     annotation_legend_side = "right", 
     heatmap_legend_side = "bottom")
dev.off()
```




### PCA: COVID-19 vaccination ssGSEA

**Required files**

```{r}
xcell_results = read_csv(file = here("Cell populations", "all_covid_degs_xcell_results.csv"))

xcell_results_annotated = xcell_results %>% 
  pivot_longer(cols = -cell_type,
               names_to = "sample",
               values_to = "value") %>% 
  inner_join(ann_vaccines_samples, by = "sample") %>% 
   distinct() %>% 
  filter(condition != "BNT-MO (V1, D0)",
         condition != "BNT-MO (V1, D6)") %>% 
  mutate(disease_vac = if_else(disease_vac == "Healthy", "H", disease_vac)) %>% 
  mutate(disease_vac = fct_relevel(disease_vac, "H", "I", "V"),
           type = fct_relevel(type, "I", "RNA", "VV", "IN", "SU")) %>% 
  inner_join(ann_vaccines_conditions %>% select(condition_grouped = condition), by = join_by("condition_grouped")) %>% 
  select(-condition) %>% 
  rename(condition = condition_grouped) %>% 
  mutate(condition = fct_relevel(condition, 
                                 ann_vaccines_conditions_ordered$condition_grouped %>% as.character()%>% unique())) %>% 
  mutate(type = if_else(condition == "BNT-I (D0-5)", "RNA", type)) %>% 
  filter(day != 0) %>% 
  group_by(condition, cell_type) %>% 
  mutate(minmax_value = (value - min(value)) / 
                             (max(value) - min(value)))

xcell_results_annotated
  
```


**Filtering data**

```{r}
# Gene set data.
ImmuneGO_BTM_grouped_genes <- read_csv("VaxGO gene sets/ImmuneGO_BTM_grouped_genes.csv")

ImmuneGO_Genes = all_degs_p_05_vac_infected %>%
  ungroup() %>% 
  inner_join(ImmuneGO_BTM_grouped_genes %>% select(genes) %>% distinct(), by = "genes") %>%
  inner_join(data_annotation, by = "condition") %>% 
  filter(disease_vac == "V")
  
  
```



**Input**

```{r}
#INPUT
data_genes = ImmuneGO_Genes 
filename = "COVID_VacOnly_ImmuneGO_BTM"

#Convert long to wide table format.
#Dataframe with sample annotation

ann_vaccines_pca_matrix = data_genes %>% 
  select(condition, genes, log2fold_change) %>% 
  distinct() %>% 
  pivot_wider(names_from = "condition", 
              values_from = "log2fold_change") %>% 
  replace(is.na(.), 0) %>%
  column_to_rownames("genes") %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column("condition") %>% 
  inner_join(data_annotation %>% 
               select(condition, disease_vac, vaccine, day, week, dose, infection, type), by = "condition") 

dim(ann_vaccines_pca_matrix) #1704 genes


# Convert dataframe to matrix without annotation columns
ann_vaccines_pca_matrix_ready = ann_vaccines_pca_matrix %>% 
  column_to_rownames("condition") %>% 
  select(!disease_vac:type) %>% 
  as.matrix()

dim(ann_vaccines_pca_matrix_ready) #1696 genes

```


```{r cache = TRUE}
# Delete columns with near 0 variance
nearZeroVarCols <- nearZeroVar(ann_vaccines_pca_matrix_ready, saveMetrics = TRUE)
ann_vaccines_pca_matrix_ready <- ann_vaccines_pca_matrix_ready[, !nearZeroVarCols$nzv]
dim(ann_vaccines_pca_matrix_ready)
pca_res <- prcomp(ann_vaccines_pca_matrix_ready, scale. = T)

# Correlation plot ----
corr_matrix = cor(ann_vaccines_pca_matrix_ready %>% scale()) 
var <- get_pca_var(pca_res)

# corrplot = ggcorrplot(corr_matrix, hc.order = TRUE) + 
#   theme(axis.text.x = element_text(angle = 90, size = 5), 
#         axis.text.y = element_text(size = 5))

# ggsave(corrplot, file = paste0(filename, "_corrplot.png"), 
#        width = 20, #Grande 20, pequeno 10
#        height= 20) #Grande 20, pequeno 10
# print(corrplot)

#PCA ----
data.pca = prcomp(corr_matrix) #PCA

#Scree plot ----

scree_plot = fviz_eig(data.pca, 
                      addlabels = TRUE, main = "", ylab = "Explained variances (%)",
                      ylim = c(0, 100)) +
  theme_few() +
  scale_fill_npg()


#Save
ggsave(scree_plot, file = here("Figures", paste0(filename, "_screeplot.png")), width = 5, height = 3) 
scree_plot

```

Loadings plot

```{r fig.height=5, fig.width=5}
# Loadings plot ----
options(ggrepel.max.overlaps = Inf)
PC1_top20_pos = pca_res$rotation %>% 
  as.data.frame() %>% 
  arrange(-PC1) %>% 
  slice_head(n = 5) %>%
  rownames_to_column("genes") %>% 
  pull(genes)

PC1_top20_neg = pca_res$rotation %>% 
  as.data.frame() %>% 
  arrange(PC1) %>% 
  slice_head(n = 5) %>%
  rownames_to_column("genes") %>% 
  pull(genes)

PC2_top20_pos = pca_res$rotation %>% 
  as.data.frame() %>% 
  arrange(PC2) %>% 
  slice_head(n = 5) %>%
  rownames_to_column("genes") %>% 
  pull(genes)

PC2_top20_neg = pca_res$rotation %>% 
  as.data.frame() %>% 
  arrange(-PC2) %>% 
  slice_head(n = 5) %>%
  rownames_to_column("genes") %>% 
  pull(genes)


circle_contrib= fviz_pca_var(pca_res, 
                             col.var = "cos2",
                              gradient.cols = c("#DC0000FF", "black"),
                               col.ind = "black",
                              # select.var= list(cos2 = 20), 
                              select.var = list(name = c(PC1_top20_pos, PC1_top20_neg, PC2_top20_pos, PC2_top20_neg)),
                              repel = T, 
                              labelsize = 5, 
                              col.circle = "gray90") +
  labs(title = "",
       x = "PC1",
       y = "PC2") +
  geom_vline(xintercept = 0, color = "white") +
  geom_hline(yintercept = 0, color = "white") +
  theme_void() +
  theme(panel.grid = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank(),
        panel.grid.major = element_blank(),
        axis.line = element_line(size = 0, colour = "white", linetype=1)) 

#Save
ggsave(circle_contrib, file = here("Figures", paste0(filename, "_circle_contrib_wide.png")), width = 5, height = 5)

circle_contrib


```

KNN

```{r}
#KNN classification -----
#Use the screeplot generated % of explained variance for each dimension
scree_plot #Dim 1 = 68.2%, Dim 2 = 15.9%


# PC1-PC2 ------
#Determine the number of clusters
pca_scores <- data.frame(pca_res$x[, 1:2])
fviz_nbclust(pca_scores,  
                     FUNcluster = kmeans,
                     method = "wss")

pcas = "PC1-PC2"

set.seed(666) # Set seed for randomization
cluster_model <- kmeans(pca_res$x[, 1:2], centers = 4)  #Set the number of clusters

ann_vaccines_pca_matrix$cluster <- as.factor(cluster_model$cluster)
```


#### Plot PC1-2
```{r fig.height=7, fig.width=7, paged.print=FALSE, cache=TRUE}
# Condition with density plot ------

# ann_vaxsig_colors_linebreak = ann_vaxsig_colors
# names(ann_vaxsig_colors_linebreak$condition_col) <- str_replace(names(ann_vaxsig_colors_linebreak$condition_col), " \\(", "\n(")

pca_plot_knn_cluster_labels = autoplot(pca_res, 
        data = ann_vaccines_pca_matrix)  +
  # stat_ellipse(aes(color = cluster, fill = cluster),
  #              geom = "polygon",
  #              alpha = 0.2,
  #              linetype = 1,
  #              size = 0.3,
  #              type = "t") +
  geom_convexhull(aes(fill = cluster, colour = cluster), alpha = 0.2) +
  labs(title=paste0(filename, "_bycondition"),
       x = "PC1 (71.8%)",
       y = "PC2 (13%)") + 
  # geom_hline(yintercept = 0, size = 0.8, color = "gray50", linetype = 1, alpha = 1) +
  # geom_vline(xintercept = 0, size = 0.8, color = "gray50", linetype = 1, alpha = 1) +
  geom_point(color = "black",
             size = 4,
             stroke = 1) +
  geom_point(aes(fill = condition,
                 color = condition),
             size = 3,
             stroke = 1) +
   scale_color_manual(values = c("1" = "#DC0000FF", 
                               "2" = "#4DBBD5FF", 
                               "3" = "#F5BB00",
                               "4" = "#8491B4FF",
                                ann_vaxsig_colors$condition_col), 
                     guide = "none") +
  scale_fill_manual(values = c("1" = "#DC0000FF", 
                               "2" = "#4DBBD5FF", 
                               "3" = "#F5BB00",
                               "4" = "#8491B4FF",
                               ann_vaxsig_colors$condition_col),
                    guide = "none") + 
  guides(col="none") +
  theme_few() +
  theme(axis.text = element_text(size = 10, color = "black"),
      panel.grid.minor = element_blank(),
      panel.grid.major = element_line(size = 1, 
                                    linetype = 1,
                                    color = "gray90"),
        axis.line = element_line(size = 0, colour = "black", linetype=1),
        # axis.line = element_line(size = 1, colour = "black", linetype=1),
        text = element_text(color = "black", size = 10),
        panel.border = element_rect(fill = NA, color = "gray90", size = 2))   +
  geom_label_repel(aes(label = if_else(cluster != 2 | type == "I" & condition != c("I-BNT (D5)", "BNT-I-BNT (D51, hospital)"), condition, NA)),
                  segment.colour="black",
                  box.padding = 0.5, 
                  # fill = "gray60",
                  # color = "white",
                  size = 3.5)  +
  xlim(-0.3, 0.8)
pca_plot_knn_cluster_labels

ggsave(pca_plot_knn_cluster_labels, 
       file = here(file = "Figures", paste0(filename, pcas, "_KNN_Clustered_labels.png")), width = 7, height = 7)
```









#### Heatmap of PC1 and PC2 genes

Inputs
```{r}
####### INPUT
PC1 = var$cos2 %>%
  as.data.frame() %>%
  arrange(desc(Dim.1)) %>%
  filter(Dim.1 >= 0.5) %>% 
  #slice_head(n=20) %>% 
  rownames_to_column("genes") %>% 
  distinct() %>% 
  select(genes, cos2 = Dim.1) 

PC1_top20_pos = pca_res$rotation %>% 
  as.data.frame() %>% 
  arrange(-PC1) %>% 
  slice_head(n = 5) %>%
  rownames_to_column("genes") 

PC1_top20_neg = pca_res$rotation %>% 
  as.data.frame() %>% 
  arrange(PC1) %>% 
  slice_head(n = 5) %>%
  rownames_to_column("genes") 

PC2_top20_pos = pca_res$rotation %>% 
  as.data.frame() %>% 
  arrange(PC2) %>% 
  slice_head(n = 5) %>%
  rownames_to_column("genes") 

PC2_top20_neg = pca_res$rotation %>% 
  as.data.frame() %>% 
  arrange(-PC2) %>% 
  slice_head(n = 5) %>%
  rownames_to_column("genes") 



PC2 = var$cos2 %>%
  as.data.frame() %>%
  filter(Dim.2 >= 0.5) %>% 
  #slice_head(n=20) %>% 
  rownames_to_column("genes") %>% 
  distinct() %>% 
  select(genes, cos2 = Dim.2) %>% 
  mutate(dimension = "PC2")

PC1_2 = bind_rows(PC1, PC2) 
PC1_2 %>% 
  select(genes) %>% distinct()

PC1_2 %>% write_csv(file = here("PCA", paste(filename, "PC_1_2_cos2.csv")))

```


#### PC1 genes heatmap

Horizontal
```{r}
# INPUT
ann_genes = ImmuneGO_Annotated_Genes
btm_annotation_genes <- readRDS(here("VaxGO gene sets", "btm_annotation_genes.rds"))
btm_annotation_genes_immune = btm_annotation_genes
pc = "PC1"
# data_2 =  PC1 %>% 
#   slice_max(order_by = cos2, n = 10) %>% 
#   inner_join(data_genes, by = "genes")

data_2 = PC1_top20_pos %>% 
  bind_rows(PC1_top20_neg) %>% 
  select(genes, value= PC1) %>% 
  inner_join(data_genes, by = "genes")

filename_2 = paste0(filename, "_PCA_",pc, "_COVID")

######## Matrix
matrix = data_2 %>% 
  select(genes, condition, log2fold_change) %>%
  pivot_wider(names_from = "condition", 
              values_from = "log2fold_change", 
              values_fn = mean) %>% 
  replace(is.na(.), 0) %>%
  arrange(genes) %>% 
  column_to_rownames(var="genes") %>% 
  as.matrix()

######## Annotations
ann_vaccines_covid = data_2 %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  select(condition, disease_vac, type, day) %>% 
  distinct()

#Columns
ann_cols_heatmap = matrix %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  inner_join(ann_vaccines_covid, by = "condition") %>% 
  distinct() %>% 
  arrange(condition) %>% 
  column_to_rownames("condition")

matrix_data = matrix %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  inner_join(ann_cols_heatmap %>% rownames_to_column("condition"), by = "condition") %>% 
  select(!c(disease_vac:day)) %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>%
  as.matrix()

#Rows Immune ----
ann_rows_immune = matrix_data %>%
  as.data.frame() %>%
  rownames_to_column("genes") %>%
  left_join(ImmuneGO_Annotated_Genes %>% select(genes) %>% distinct() %>% mutate(set = "ImmuneGO"), by = "genes") %>%
  left_join(btm_annotation_genes_immune %>% select(genes) %>% distinct() %>% mutate(set2 = "BTM"), by = "genes") %>%
  select(genes, set, set2) %>%
  distinct() %>%
  group_by(genes) %>%
  arrange(genes, factor(set, levels = c("ImmuneGO", "BTM")), .by_group = TRUE) %>%
  column_to_rownames("genes") %>%
  as.data.frame() %>%
  mutate(set = if_else(is.na(set), set2, set),
         set2 = if_else(set == set2, NA, set2)) %>% 
  replace(is.na(.), "None") %>% 
  select(set2, set)

# Reorder columns (same order as lines in the annotation table)
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_covid, by.y = "condition", all.x=TRUE) %>% 
  arrange(type, condition) %>% 
  select(!c(disease_vac, type, day)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0) %>% 
  as.matrix() 


#Check dimensions
dim(matrix_data_ordered) 
dim(ann_vaccines_heatmap) 
dim(ann_rows_immune)

col_immune_sets = list(
  "set" = c("BTM" = "#5e60ce",
          "ImmuneGO" = "#00A087FF",
          "None"="white"),
  "set2" = c("BTM" = "#5e60ce",
          "ImmuneGO" = "#00A087FF",
          "None"="white"))


# ann_cols_heatmap <- ann_cols_heatmap %>%
#   mutate(day = factor(day, levels = unique(day[order(day)])))
  

#Heatmap annotation -----
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "right")

row_ha = HeatmapAnnotation(df = ann_rows_immune,
                            which = "row",
                       col = col_immune_sets, #col_process_immune / col_process_notimmune
                       show_annotation_name = F,
                       show_legend = T,
                       simple_anno_size = unit(2, "mm"))

# Values colors -----
col_fun <- colorRamp2(c(-2, 0, 2), c("#4DBBD5FF", "white", "#ed6a5a"))


# Parameters -----
file = filename_2
mat = matrix_data
width_image = 20
height_image = 15
width = unit(5, "cm")
height = unit(5, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)
rect_gp = gpar(col = "black", lwd = 0.5)

fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap.png")

png(file = here("Figures", fileimageheatmap_grouped), 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        name = "L2FC",
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        col = col_fun, #color
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(2, "cm"),
                        column_split = NULL,
                        row_split = NULL,
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height 
)

draw(heatmap_plot, 
     annotation_legend_side = "right", 
     heatmap_legend_side = "bottom")
dev.off()
```


#### PC2 genes heatmap
```{r}
# INPUT
ann_genes = ImmuneGO_Annotated_Genes
pc = "PC2"
# data_2 =  PC2 %>% 
#   slice_max(order_by = cos2, n = 10) %>% 
#   inner_join(data_genes, by = "genes")

data_2 = PC2_top20_pos %>% 
  bind_rows(PC2_top20_neg) %>% 
  select(genes, value= PC2) %>% 
  inner_join(data_genes, by = "genes")


filename_2 = paste0(filename, "_PCA_",pc, "_COVID")

######## Matrix
matrix = data_2 %>% 
  select(genes, condition, log2fold_change) %>%
  pivot_wider(names_from = "condition", 
              values_from = "log2fold_change", 
              values_fn = mean) %>% 
  replace(is.na(.), 0) %>%
  arrange(genes) %>% 
  column_to_rownames(var="genes") %>% 
  as.matrix()

######## Annotations
ann_vaccines_covid = data_2 %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  select(condition, disease_vac, type, day) %>% 
  distinct()

#Columns
ann_cols_heatmap = matrix %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  inner_join(ann_vaccines_covid, by = "condition") %>% 
  distinct() %>% 
  arrange(condition) %>% 
  column_to_rownames("condition")

matrix_data = matrix %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  inner_join(ann_cols_heatmap %>% rownames_to_column("condition"), by = "condition") %>% 
  select(!c(disease_vac:day)) %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>%
  as.matrix()

#Rows Immune ----
ann_rows_immune = matrix_data %>%
  as.data.frame() %>%
  rownames_to_column("genes") %>%
  left_join(ImmuneGO_Annotated_Genes %>% select(genes) %>% distinct() %>% mutate(set = "ImmuneGO"), by = "genes") %>%
  left_join(btm_annotation_genes_immune %>% select(genes) %>% distinct() %>% mutate(set2 = "BTM"), by = "genes") %>%
  select(genes, set, set2) %>%
  distinct() %>%
  group_by(genes) %>%
  arrange(genes, factor(set, levels = c("ImmuneGO", "BTM")), .by_group = TRUE) %>%
  column_to_rownames("genes") %>%
  as.data.frame() %>%
  mutate(set = if_else(is.na(set), set2, set),
         set2 = if_else(set == set2, NA, set2)) %>% 
  replace(is.na(.), "None") %>% 
  select(set2, set)

# Reorder columns (same order as lines in the annotation table)
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_covid, by.y = "condition", all.x=TRUE) %>% 
  arrange(type, condition) %>% 
  select(!c(disease_vac, type, day)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0) %>% 
  as.matrix() 


#Check dimensions
dim(matrix_data_ordered) 
dim(ann_vaccines_heatmap) 
dim(ann_rows_immune)

col_immune_sets = list(
  "set" = c("BTM" = "#5e60ce",
          "ImmuneGO" = "#00A087FF",
          "None"="white"),
  "set2" = c("BTM" = "#5e60ce",
          "ImmuneGO" = "#00A087FF",
          "None"="white"))

# ann_cols_heatmap <- ann_cols_heatmap %>%
#   mutate(day = factor(day, levels = unique(day[order(day)])))


#Heatmap annotation -----
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "right")

row_ha = HeatmapAnnotation(df = ann_rows_immune,
                            which = "row",
                       col = col_immune_sets, #col_process_immune / col_process_notimmune
                       show_annotation_name = F,
                       show_legend = T,
                       simple_anno_size = unit(2, "mm"))

# Values colors -----
col_fun <- colorRamp2(c(-2, 0, 2), c("#4DBBD5FF", "white", "#ed6a5a"))


# Parameters -----
file = filename_2
mat = matrix_data
width_image = 15
height_image = 15
width = unit(5, "cm")
height = unit(5, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)
rect_gp = gpar(col = "black", lwd = 0.5)

fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap.png")

png(file = here("Figures", fileimageheatmap_grouped), 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        name = "L2FC",
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        col = col_fun, #color
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(2, "cm"),
                        column_split = NULL,
                        row_split = NULL,
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height 
)

draw(heatmap_plot, 
     annotation_legend_side = "right", 
     heatmap_legend_side = "bottom")
dev.off()

```

```{r}
## Get row and column orders
ht_list = draw(heatmap_plot)
rows_clustered = rownames(mat)[row_order(ht_list)]
cols_clustered = colnames(mat)[column_order(ht_list)]

mat_2 = mat %>% 
  as.data.frame() %>% 
  rownames_to_column("genes") %>% 
  filter(genes %in% c(rev(rows_clustered))) %>% 
  select(genes, c(cols_clustered)) %>% 
  column_to_rownames("genes") %>% 
  as.matrix()

width = unit(15, "cm")
height = unit(5, "cm")

fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap_ForPCA.png")

png(file = here("Figures", fileimageheatmap_grouped), 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot_2 = Heatmap(mat_2,
                        # top_annotation = ha,
                        name = "L2FC",
                        # left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        col = col_fun, #color
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 90,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = F,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(2, "cm"),
                        column_split = NULL,
                        row_split = NULL,
                        row_gap = unit(2, "mm"),
                        cluster_columns = F,
                        column_gap = unit(8, "mm"),
                        column_names_side = "top",
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = F,
                        width = width,
                        height = height 
)

draw(heatmap_plot_2, 
     annotation_legend_side = "right", 
     heatmap_legend_side = "bottom")
dev.off()
```

### PCA: COVID-19 Infection conditions only

**Required files**

```{r}
# Genes in long table format
all_degs_p_05_vac_infected = readRDS(here("DGE analysis", "all_studies_degs_weighted.rds")) %>% 
  filter(if_else(n >= 5, fdr <= 0.05, fdr <= (0.05 * 6/n))) 

# Sample or condition annotation
ann_vaccines <- read_csv(here("Annotations and metadata", "ann_vaccines_conditions.csv"))

data_annotation = ann_vaccines %>%
  mutate(day = as.factor(day),
         dose = as.factor(dose)) 
```


**Filtering data**

```{r}
# Gene set data.
ImmuneGO_BTM_grouped_genes <- read_csv("VaxGO gene sets/ImmuneGO_BTM_grouped_genes.csv")

ImmuneGO_Genes = all_degs_p_05_vac_infected %>%
  ungroup() %>% 
  inner_join(ImmuneGO_BTM_grouped_genes %>% select(genes) %>% distinct(), by = "genes") %>%
  inner_join(data_annotation, by = "condition") %>% 
  filter(disease_vac == "I")
  
  
```



**Input**

```{r}
#INPUT
data_genes = ImmuneGO_Genes 
filename = "COVID_InfectionOnly_ImmuneGO_BTM"

#Convert long to wide table format.
#Dataframe with sample annotation

ann_vaccines_pca_matrix = data_genes %>% 
  select(condition, genes, log2fold_change) %>% 
  distinct() %>% 
  pivot_wider(names_from = "condition", 
              values_from = "log2fold_change") %>% 
  replace(is.na(.), 0) %>%
  column_to_rownames("genes") %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column("condition") %>% 
  inner_join(data_annotation %>% 
               select(condition, disease_vac, vaccine, day, week, dose, infection, type), by = "condition") 

dim(ann_vaccines_pca_matrix) #1704 genes


# Convert dataframe to matrix without annotation columns
ann_vaccines_pca_matrix_ready = ann_vaccines_pca_matrix %>% 
  column_to_rownames("condition") %>% 
  select(!disease_vac:type) %>% 
  as.matrix()

dim(ann_vaccines_pca_matrix_ready) #1696 genes

```


```{r cache = TRUE}
# Delete columns with near 0 variance
nearZeroVarCols <- nearZeroVar(ann_vaccines_pca_matrix_ready, saveMetrics = TRUE)
ann_vaccines_pca_matrix_ready <- ann_vaccines_pca_matrix_ready[, !nearZeroVarCols$nzv]
dim(ann_vaccines_pca_matrix_ready)
pca_res <- prcomp(ann_vaccines_pca_matrix_ready, scale. = T)

# Correlation plot ----
corr_matrix = cor(ann_vaccines_pca_matrix_ready %>% scale()) 
var <- get_pca_var(pca_res)

# corrplot = ggcorrplot(corr_matrix, hc.order = TRUE) + 
#   theme(axis.text.x = element_text(angle = 90, size = 5), 
#         axis.text.y = element_text(size = 5))

# ggsave(corrplot, file = paste0(filename, "_corrplot.png"), 
#        width = 20, #Grande 20, pequeno 10
#        height= 20) #Grande 20, pequeno 10
# print(corrplot)

#PCA ----
data.pca = prcomp(corr_matrix) #PCA

#Scree plot ----

scree_plot = fviz_eig(data.pca, 
                      addlabels = TRUE, main = "", ylab = "Explained variances (%)",
                      ylim = c(0, 100)) +
  theme_few() +
  scale_fill_npg()


#Save
ggsave(scree_plot, file = here("Figures", paste0(filename, "_screeplot.png")), width = 5, height = 3) 
scree_plot

```

Loadings plot

```{r fig.height=5, fig.width=5}
# Loadings plot ----
options(ggrepel.max.overlaps = Inf)
PC1_top20_pos = pca_res$rotation %>% 
  as.data.frame() %>% 
  arrange(-PC1) %>% 
  slice_head(n = 5) %>%
  rownames_to_column("genes") %>% 
  pull(genes)

PC1_top20_neg = pca_res$rotation %>% 
  as.data.frame() %>% 
  arrange(PC1) %>% 
  slice_head(n = 5) %>%
  rownames_to_column("genes") %>% 
  pull(genes)

PC2_top20_pos = pca_res$rotation %>% 
  as.data.frame() %>% 
  arrange(PC2) %>% 
  slice_head(n = 5) %>%
  rownames_to_column("genes") %>% 
  pull(genes)

PC2_top20_neg = pca_res$rotation %>% 
  as.data.frame() %>% 
  arrange(-PC2) %>% 
  slice_head(n = 5) %>%
  rownames_to_column("genes") %>% 
  pull(genes)


circle_contrib= fviz_pca_var(pca_res, 
                             col.var = "cos2",
                              gradient.cols = c("#DC0000FF", "black"),
                               col.ind = "black",
                              # select.var= list(cos2 = 20), 
                              select.var = list(name = c(PC1_top20_pos, PC1_top20_neg, PC2_top20_pos, PC2_top20_neg)),
                              repel = T, 
                              labelsize = 5, 
                              col.circle = "gray90") +
  labs(title = "",
       x = "PC1",
       y = "PC2") +
  geom_vline(xintercept = 0, color = "white") +
  geom_hline(yintercept = 0, color = "white") +
  theme_void() +
  theme(panel.grid = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank(),
        panel.grid.major = element_blank(),
        axis.line = element_line(size = 0, colour = "white", linetype=1)) 

#Save
ggsave(circle_contrib, file = here("Figures", paste0(filename, "_circle_contrib_wide.png")), width = 5, height = 5)

circle_contrib


```

KNN

```{r}
#KNN classification -----
#Use the screeplot generated % of explained variance for each dimension
scree_plot #Dim 1 = 68.2%, Dim 2 = 15.9%


# PC1-PC2 ------
#Determine the number of clusters
pca_scores <- data.frame(pca_res$x[, 1:2])
fviz_nbclust(pca_scores,  
                     FUNcluster = kmeans,
                     method = "wss")

pcas = "PC1-PC2"

set.seed(666) # Set seed for randomization
cluster_model <- kmeans(pca_res$x[, 1:2], centers = 3)  #Set the number of clusters

ann_vaccines_pca_matrix$cluster <- as.factor(cluster_model$cluster)
```


#### Plot PC1-2
```{r fig.height=7, fig.width=7, paged.print=FALSE, cache=TRUE}
# Condition with density plot ------

# ann_vaxsig_colors_linebreak = ann_vaxsig_colors
# names(ann_vaxsig_colors_linebreak$condition_col) <- str_replace(names(ann_vaxsig_colors_linebreak$condition_col), " \\(", "\n(")

pca_plot_knn_cluster_labels = autoplot(pca_res, 
        data = ann_vaccines_pca_matrix)  +
  # stat_ellipse(aes(color = cluster, fill = cluster),
  #              geom = "polygon",
  #              alpha = 0.2,
  #              linetype = 1,
  #              size = 0.3,
  #              type = "t") +
  geom_convexhull(aes(fill = cluster, colour = cluster), alpha = 0.2) +
  labs(title=paste0(filename, "_bycondition"),
       x = "PC1 (65.3%)",
       y = "PC2 (19.4%)") + 
  # geom_hline(yintercept = 0, size = 0.8, color = "gray50", linetype = 1, alpha = 1) +
  # geom_vline(xintercept = 0, size = 0.8, color = "gray50", linetype = 1, alpha = 1) +
  geom_point(color = "black",
             size = 4,
             stroke = 1) +
  geom_point(aes(fill = condition,
                 color = condition),
             size = 3,
             stroke = 1) +
   scale_color_manual(values = c("1" = "#DC0000FF", 
                               "2" = "#4DBBD5FF", 
                               "3" = "#F5BB00",
                               "4" = "#8491B4FF",
                                ann_vaxsig_colors$condition_col), 
                     guide = "none") +
  scale_fill_manual(values = c("1" = "#DC0000FF", 
                               "2" = "#4DBBD5FF", 
                               "3" = "#F5BB00",
                               "4" = "#8491B4FF",
                               ann_vaxsig_colors$condition_col),
                    guide = "none") + 
  guides(col="none") +
  theme_few() +
  theme(axis.text = element_text(size = 10, color = "black"),
      panel.grid.minor = element_blank(),
      panel.grid.major = element_line(size = 1, 
                                    linetype = 1,
                                    color = "gray90"),
        axis.line = element_line(size = 0, colour = "black", linetype=1),
        # axis.line = element_line(size = 1, colour = "black", linetype=1),
        text = element_text(color = "black", size = 10),
        panel.border = element_rect(fill = NA, color = "gray90", size = 2))   +
  geom_label_repel(aes(label = if_else(cluster != 2 | type == "I" & condition != c("I-BNT (D5)", "BNT-I-BNT (D51, hospital)"), condition, NA)),
                  segment.colour="black",
                  box.padding = 0.5, 
                  # fill = "gray60",
                  # color = "white",
                  size = 3.5)  +
  xlim(-0.5, 0.8)
pca_plot_knn_cluster_labels

ggsave(pca_plot_knn_cluster_labels, 
       file = here(file = "Figures", paste0(filename, pcas, "_KNN_Clustered_labels.png")), width = 7, height = 7)
```





#### Heatmap of PC1 and PC2 genes

Inputs
```{r}
####### INPUT
PC1 = var$cos2 %>%
  as.data.frame() %>%
  arrange(desc(Dim.1)) %>%
  filter(Dim.1 >= 0.5) %>% 
  #slice_head(n=20) %>% 
  rownames_to_column("genes") %>% 
  distinct() %>% 
  select(genes, cos2 = Dim.1) 

PC1_top20_pos = pca_res$rotation %>% 
  as.data.frame() %>% 
  arrange(-PC1) %>% 
  slice_head(n = 5) %>%
  rownames_to_column("genes") 

PC1_top20_neg = pca_res$rotation %>% 
  as.data.frame() %>% 
  arrange(PC1) %>% 
  slice_head(n = 5) %>%
  rownames_to_column("genes") 

PC2_top20_pos = pca_res$rotation %>% 
  as.data.frame() %>% 
  arrange(PC2) %>% 
  slice_head(n = 5) %>%
  rownames_to_column("genes") 

PC2_top20_neg = pca_res$rotation %>% 
  as.data.frame() %>% 
  arrange(-PC2) %>% 
  slice_head(n = 5) %>%
  rownames_to_column("genes") 



PC2 = var$cos2 %>%
  as.data.frame() %>%
  filter(Dim.2 >= 0.5) %>% 
  #slice_head(n=20) %>% 
  rownames_to_column("genes") %>% 
  distinct() %>% 
  select(genes, cos2 = Dim.2) %>% 
  mutate(dimension = "PC2")

PC1_2 = bind_rows(PC1, PC2) 
PC1_2 %>% 
  select(genes) %>% distinct()

PC1_2 %>% write_csv(file = here("PCA", paste(filename, "PC_1_2_cos2.csv")))

```


#### PC1 genes heatmap

Horizontal
```{r}
# INPUT
ann_genes = ImmuneGO_Annotated_Genes
btm_annotation_genes <- readRDS(here("VaxGO gene sets", "btm_annotation_genes.rds"))
btm_annotation_genes_immune = btm_annotation_genes
pc = "PC1"
# data_2 =  PC1 %>% 
#   slice_max(order_by = cos2, n = 10) %>% 
#   inner_join(data_genes, by = "genes")

data_2 = PC1_top20_pos %>% 
  bind_rows(PC1_top20_neg) %>% 
  select(genes, value= PC1) %>% 
  inner_join(data_genes, by = "genes")

filename_2 = paste0(filename, "_PCA_",pc, "_COVID")

######## Matrix
matrix = data_2 %>% 
  select(genes, condition, log2fold_change) %>%
  pivot_wider(names_from = "condition", 
              values_from = "log2fold_change", 
              values_fn = mean) %>% 
  replace(is.na(.), 0) %>%
  arrange(genes) %>% 
  column_to_rownames(var="genes") %>% 
  as.matrix()

######## Annotations
ann_vaccines_covid = data_2 %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  select(condition, disease_vac, type, day) %>% 
  distinct()

#Columns
ann_cols_heatmap = matrix %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  inner_join(ann_vaccines_covid, by = "condition") %>% 
  distinct() %>% 
  arrange(condition) %>% 
  column_to_rownames("condition")

matrix_data = matrix %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  inner_join(ann_cols_heatmap %>% rownames_to_column("condition"), by = "condition") %>% 
  select(!c(disease_vac:day)) %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>%
  as.matrix()

#Rows Immune ----
ann_rows_immune = matrix_data %>%
  as.data.frame() %>%
  rownames_to_column("genes") %>%
  left_join(ImmuneGO_Annotated_Genes %>% select(genes) %>% distinct() %>% mutate(set = "ImmuneGO"), by = "genes") %>%
  left_join(btm_annotation_genes_immune %>% select(genes) %>% distinct() %>% mutate(set2 = "BTM"), by = "genes") %>%
  select(genes, set, set2) %>%
  distinct() %>%
  group_by(genes) %>%
  arrange(genes, factor(set, levels = c("ImmuneGO", "BTM")), .by_group = TRUE) %>%
  column_to_rownames("genes") %>%
  as.data.frame() %>%
  mutate(set = if_else(is.na(set), set2, set),
         set2 = if_else(set == set2, NA, set2)) %>% 
  replace(is.na(.), "None") %>% 
  select(set2, set)

# Reorder columns (same order as lines in the annotation table)
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_covid, by.y = "condition", all.x=TRUE) %>% 
  arrange(type, condition) %>% 
  select(!c(disease_vac, type, day)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0) %>% 
  as.matrix() 


#Check dimensions
dim(matrix_data_ordered) 
dim(ann_vaccines_heatmap) 
dim(ann_rows_immune)

col_immune_sets = list(
  "set" = c("BTM" = "#5e60ce",
          "ImmuneGO" = "#00A087FF",
          "None"="white"),
  "set2" = c("BTM" = "#5e60ce",
          "ImmuneGO" = "#00A087FF",
          "None"="white"))


# ann_cols_heatmap <- ann_cols_heatmap %>%
#   mutate(day = factor(day, levels = unique(day[order(day)])))
  

#Heatmap annotation -----
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "right")

row_ha = HeatmapAnnotation(df = ann_rows_immune,
                            which = "row",
                       col = col_immune_sets, #col_process_immune / col_process_notimmune
                       show_annotation_name = F,
                       show_legend = T,
                       simple_anno_size = unit(2, "mm"))

# Values colors -----
col_fun <- colorRamp2(c(-2, 0, 2), c("#4DBBD5FF", "white", "#ed6a5a"))


# Parameters -----
file = filename_2
mat = matrix_data
width_image = 20
height_image = 15
width = unit(5, "cm")
height = unit(5, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)
rect_gp = gpar(col = "black", lwd = 0.5)

fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap.png")

png(file = here("Figures", fileimageheatmap_grouped), 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        name = "L2FC",
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        col = col_fun, #color
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(2, "cm"),
                        column_split = NULL,
                        row_split = NULL,
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height 
)

draw(heatmap_plot, 
     annotation_legend_side = "right", 
     heatmap_legend_side = "bottom")
dev.off()
```


#### PC2 genes heatmap
```{r}
# INPUT
ann_genes = ImmuneGO_Annotated_Genes
pc = "PC2"
# data_2 =  PC2 %>% 
#   slice_max(order_by = cos2, n = 10) %>% 
#   inner_join(data_genes, by = "genes")

data_2 = PC2_top20_pos %>% 
  bind_rows(PC2_top20_neg) %>% 
  select(genes, value= PC2) %>% 
  inner_join(data_genes, by = "genes")


filename_2 = paste0(filename, "_PCA_",pc, "_COVID")

######## Matrix
matrix = data_2 %>% 
  select(genes, condition, log2fold_change) %>%
  pivot_wider(names_from = "condition", 
              values_from = "log2fold_change", 
              values_fn = mean) %>% 
  replace(is.na(.), 0) %>%
  arrange(genes) %>% 
  column_to_rownames(var="genes") %>% 
  as.matrix()

######## Annotations
ann_vaccines_covid = data_2 %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  select(condition, disease_vac, type, day) %>% 
  distinct()

#Columns
ann_cols_heatmap = matrix %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  inner_join(ann_vaccines_covid, by = "condition") %>% 
  distinct() %>% 
  arrange(condition) %>% 
  column_to_rownames("condition")

matrix_data = matrix %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  inner_join(ann_cols_heatmap %>% rownames_to_column("condition"), by = "condition") %>% 
  select(!c(disease_vac:day)) %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>%
  as.matrix()

#Rows Immune ----
ann_rows_immune = matrix_data %>%
  as.data.frame() %>%
  rownames_to_column("genes") %>%
  left_join(ImmuneGO_Annotated_Genes %>% select(genes) %>% distinct() %>% mutate(set = "ImmuneGO"), by = "genes") %>%
  left_join(btm_annotation_genes_immune %>% select(genes) %>% distinct() %>% mutate(set2 = "BTM"), by = "genes") %>%
  select(genes, set, set2) %>%
  distinct() %>%
  group_by(genes) %>%
  arrange(genes, factor(set, levels = c("ImmuneGO", "BTM")), .by_group = TRUE) %>%
  column_to_rownames("genes") %>%
  as.data.frame() %>%
  mutate(set = if_else(is.na(set), set2, set),
         set2 = if_else(set == set2, NA, set2)) %>% 
  replace(is.na(.), "None") %>% 
  select(set2, set)

# Reorder columns (same order as lines in the annotation table)
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_covid, by.y = "condition", all.x=TRUE) %>% 
  arrange(type, condition) %>% 
  select(!c(disease_vac, type, day)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0) %>% 
  as.matrix() 


#Check dimensions
dim(matrix_data_ordered) 
dim(ann_vaccines_heatmap) 
dim(ann_rows_immune)

col_immune_sets = list(
  "set" = c("BTM" = "#5e60ce",
          "ImmuneGO" = "#00A087FF",
          "None"="white"),
  "set2" = c("BTM" = "#5e60ce",
          "ImmuneGO" = "#00A087FF",
          "None"="white"))

# ann_cols_heatmap <- ann_cols_heatmap %>%
#   mutate(day = factor(day, levels = unique(day[order(day)])))


#Heatmap annotation -----
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "right")

row_ha = HeatmapAnnotation(df = ann_rows_immune,
                            which = "row",
                       col = col_immune_sets, #col_process_immune / col_process_notimmune
                       show_annotation_name = F,
                       show_legend = T,
                       simple_anno_size = unit(2, "mm"))

# Values colors -----
col_fun <- colorRamp2(c(-2, 0, 2), c("#4DBBD5FF", "white", "#ed6a5a"))


# Parameters -----
file = filename_2
mat = matrix_data
width_image = 15
height_image = 15
width = unit(5, "cm")
height = unit(5, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)
rect_gp = gpar(col = "black", lwd = 0.5)

fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap.png")

png(file = here("Figures", fileimageheatmap_grouped), 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        name = "L2FC",
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        col = col_fun, #color
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(2, "cm"),
                        column_split = NULL,
                        row_split = NULL,
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height 
)

draw(heatmap_plot, 
     annotation_legend_side = "right", 
     heatmap_legend_side = "bottom")
dev.off()

```






# 9. Machine learning descriptors

## Preparing

### Library

```{r message=FALSE, warning=FALSE}
# install.packages("randomForest")
# install.packages("caret")
# install.packages("rpart")
# install.packages("rpart.plot")
# install.packages("tidymodels")
# install.packages("reticulate")
# install.packages("themis")
# install.packages("ranger")
# install.packages("workflowsets")
# install.packages("glmnet")
# install.packages("kknn")
# install.packages("nnet")
# install.packages("tictoc")

library(tidymodels)
tidymodels_prefer()
library(randomForest)
library(ranger)
library(caret)
library(reticulate)
library(themis)
library(tidyverse)
library(rpart)
library(rpart.plot)
library(here)
library(ggrepel)
library(ComplexHeatmap)
library(janitor)
library(ggsci)
library(vip)
library(here)
library(circlize)
library(workflowsets)
library(glmnet)
library(kknn)
library(nnet)
library(tictoc)
library(ggthemes)
tidymodels_prefer()

```

### Import files

```{r}
all_matrices_normalized_counts <- readRDS(here("DGE analysis", "all_matrices_normalized_counts.rds"))
all_matrices_counts_tpm_normalized <- readRDS(here("DGE analysis", "all_matrices_counts_tpm_normalized.rds"))
all_matrices_counts <- readRDS(here("DGE analysis", "all_matrices_counts.rds"))
all_samples_l2fc <- readRDS(here("DGE analysis", "all_samples_l2fc.rds"))
ImmuneGO_BTM_grouped_genes <- read_csv("VaxGO gene sets/ImmuneGO_BTM_grouped_genes.csv")
all_degs_p_05_vac_infected = readRDS(here("DGE analysis", "all_studies_degs_weighted.rds")) %>% 
    filter(if_else(n >= 5, fdr <= 0.05, fdr <= (0.05 * 6/n))) 
ann_vaccines <- read_csv(here("Annotations and metadata", "ann_vaccines_conditions.csv"))

ann_vaccines_samples = read_csv(here("Annotations and metadata", "ann_vaccines_samples.csv"))
PC_1_2_cos2 = read_csv(here("PCA", "PC_1_2_cos2.csv"))


dim(all_matrices_normalized_counts)
```

 

## COVID-19 vaccines vs infection (counts)

### Build models

```{r}
# Input data ----
filename = "ImmuneGO_Genes_Type_including_Infection_Counts"
df = all_matrices_normalized_counts

## Vaccination types
outcomevar = "disease_vac"

ann_vaccines_samples_filtered = ann_vaccines_samples %>% 
  mutate(outcomevar = disease_vac) %>% 
  filter(type != "H")

#How many samples by level?
count_levels = ann_vaccines_samples_filtered %>% 
  dplyr::count(outcomevar)

count_levels

#How many levels?
levels = nrow(count_levels)
order_levels = c("I", #TRUE
                 "V") #FALSE

#Select only important genes
PC1_2 = PC_1_2_cos2
```

Train and test

```{r fig.height=10, fig.width=10}
# Prepare data ------

df_input <- df %>%
  rownames_to_column("genes") %>%
  inner_join(PC1_2 %>% select(genes) %>% distinct(), by = "genes") %>%
  column_to_rownames("genes") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  inner_join(ann_vaccines_samples_filtered %>% select(sample,
                                                      outcomevar), #Outcome
             by = "sample") %>%
  select(sample,
         outcomevar,
         everything()) %>%
  column_to_rownames("sample") %>%
  mutate_if(is.character, factor) %>% 
  mutate(outcomevar = fct_relevel(outcomevar, order_levels))

# glimpse(df_input)

# Split data
set.seed(123)  
split <- initial_split(df_input, prop = 0.6, strata = outcomevar)
trees_train <- training(split)
trees_test <- testing(split)

# Create recipe
tree_rec <- recipe(outcomevar ~ ., data = trees_train) %>% #OUTCOME
  step_dummy(all_nominal(), -all_outcomes()) %>%
  # step_corr(all_numeric_predictors()) %>% 
  step_upsample(outcomevar) #Upsample unbalanced data

tree_prep <- prep(tree_rec)

juiced <- juice(tree_prep)

# Set random forest hyper parameters
tune_spec <- rand_forest(
  mtry = tune(),
  trees = 2000,
  min_n = tune()
) %>%
  set_mode("classification") %>% 
  set_engine("ranger")

# Create workflow
tune_wf <- workflow() %>%
  add_recipe(tree_rec) %>%
  add_model(tune_spec)

#Cross validation
set.seed(234)
trees_folds <- vfold_cv(trees_train)

#Define performance metrics
mer_metrics <- metric_set(yardstick::recall,
                          yardstick::specificity,
                          yardstick::accuracy,
                          yardstick::precision,
                          yardstick::roc_auc,
                          yardstick::pr_auc)

# Parallel computing 
doParallel::registerDoParallel()

set.seed(345)
tune_res <- tune_grid(
  tune_wf,
  metrics = mer_metrics,
  resamples = trees_folds,
  grid = 20
)
```

Assess metrics

```{r fig.height=10, fig.width=10}
# Assess data ----

#Metricss
metrics_tuneres = tune_res %>%
  collect_metrics() %>%
  select(mean, min_n, mtry, .metric) %>%
  pivot_longer(min_n:mtry,
    values_to = "value",
    names_to = "parameter") %>%
  ggplot()+
  aes(x = value, y = mean, colour = .metric, label = value) +
  geom_line() +
  geom_label() +
  scale_color_npg()+
  theme_few() +
  theme(axis.text.x = element_text(color = "black", size = 12)) +
  facet_grid(vars(.metric), vars(parameter), scales = "free")

metrics_tuneres

metrics_tuneres %>% ggsave(file = here("Figures",
                                       paste0(filename, 
                                              outcomevar,
                                              "_tuneres_metrics.png")),
                                        width = 10,
                                        height = 10)

min_n_selec = 5 #Observations by node)
mtry_selec = 6 #Features/genes by tree)
```

Hyperparameter tuning

```{r}
# Tune hyperparameters ----
rf_grid <- grid_regular(
  min_n(range = c(5, 5)),
  mtry(range = c(6, 6)),
  levels = levels
)

set.seed(456)
regular_res <- tune_grid(
  tune_wf,
  metrics = mer_metrics,
  resamples = trees_folds,
  grid = rf_grid
)

# Assess performance -----
regular_res %>%
  collect_metrics()

regular_res %>%
  collect_metrics() %>% 
  write_csv(file = here("Machine learning", paste("metrics_finalmodel_", filename, outcomevar, ".csv")))

#Metrics
regular_res_metrics = regular_res %>%
  collect_metrics() %>%
  select(mean, min_n, mtry, .metric) %>%
  pivot_longer(min_n:mtry,
    values_to = "value",
    names_to = "parameter") %>%
  ggplot()+
  aes(x = value, y = mean, colour = .metric, label = value) +
  geom_line() +
  geom_label(size = 3) +
  scale_color_npg()+
  theme_minimal() +
  facet_grid(vars(.metric), vars(parameter), scales = "free")

regular_res_metrics

# ggsave(regular_res_metrics, file = here("Figures", 
#                                         paste(filename, outcomevar, "metrics", today(), ".png")), 
#        width = 10, height = 10)

# Select best model -----
best_auc <- select_best(regular_res, metric = "pr_auc")
final_rf <- finalize_model(
  tune_spec,
  best_auc
)
```

Important features

```{r fig.height=3, fig.width=5}
final = final_rf %>%
  set_engine("ranger", importance = "permutation") %>%
  fit(outcomevar ~ .,
    data = juice(tree_prep)) %>% 
  vip()

genes_rf = final$data
genes_rf = genes_rf %>% select(genes = Variable, importance = Importance)
genes_rf %>% write_csv(file = here("Figures", paste0("vipgenes_tidymodels_", filename, outcomevar, ".csv")))

#Importance plot
importance_plot = ggplot(genes_rf) +
  aes(x = reorder(genes, importance),
    y = importance,
    fill = importance,
      weight = importance) +
  geom_col() +
  theme_few() +
  xlab("Genes") + 
  ylab("importance") +
  ylim(0, 0.035) +
  geom_text(aes(label = round(importance, 4), y = importance),
            hjust = -0.2, 
            size = 4,
            angle = 0,
            color = "black")+
  scale_fill_gradient(low = "#4DBBD5FF", high = "#DC0000FF") +
  theme(legend.position = "right",
        plot.subtitle = element_text(hjust = 0),
        axis.text.x = element_text(angle = 0, vjust = 0.5, hjust=0.5, size = 12, color = "black"),
        axis.text.y = element_text(size = 12, angle = 0, color = "black"),
        panel.grid.minor.y = element_blank(),
        legend.text = element_text(size=12),
        legend.title = element_text(size = 12, face = "bold")) +
    coord_flip() +
  labs(title = "VIP genes",
       subtitle = "Random forest classification model for \nVaccination vs. Infection",
       x = "",
       y = "Importance",
       color = "Importance")

ggsave(importance_plot,
        file = here("Figures",
                    paste0("vipgenes_tidymodels_", filename, outcomevar,  ".png")),
        width = 5,
        height = 3)

importance_plot

```

Train final model with tuned hyperparameters

```{r}
# Train final model with tuned hyperparameters
final_wf <- workflow() %>%
  add_recipe(tree_rec) %>%
  add_model(final_rf)

final_res <- final_wf %>%
  last_fit(split)

final_res_preds = final_res %>%
  collect_predictions() %>%
  mutate(correct = case_when(
    outcomevar == .pred_class ~ "Correct",
    TRUE ~ "Incorrect"
  )) %>%
  bind_cols(trees_test)


#Confusion matrix
res_conf_mat = conf_mat(final_res_preds, truth = outcomevar...6,
         estimate = .pred_class)

#Accuracy
res_acc = accuracy(final_res_preds, truth = outcomevar...6,
         estimate = .pred_class)

#Specificity
res_spec = spec(final_res_preds, truth = outcomevar...6,
         estimate = .pred_class)

#Precision
res_prec = precision(final_res_preds, truth = outcomevar...6,
         estimate = .pred_class)

#Recall
res_recall = recall(final_res_preds, truth = outcomevar...6,
         estimate = .pred_class)
#Precision recall AUC
res_pr_auc = pr_auc(final_res_preds, outcomevar...6, .pred_I)


#Unite results
final_res_metrics = bind_rows(res_acc,
                         res_recall,
                         res_prec,
                         res_pr_auc,
                         res_spec) %>% 
  select(-".estimator")


final_res_metrics_conf = list(res_conf_mat, 
                         final_res_metrics,
                         final_res_preds)

final_res_metrics_conf

final_res_metrics_conf %>% 
  saveRDS(file = here("Machine learning", paste0("final_res_metrics", filename, outcomevar, ".rds")))


#Plot
final_model_metrics = final_res_metrics %>% 
  rename(metric = .metric,
         estimate = .estimate) %>% 
  mutate(metric = fct_relevel(metric, final_res_metrics$.metric)) %>% 
  ggplot() +
  aes(x = estimate,
    y = fct_rev(metric),
    fill = estimate,
    weight = estimate) +
  geom_col() +
  theme_few() +
  geom_text(aes(label = round(estimate, 3), y = metric),
            hjust = -0.2, 
            size = 4,
            angle = 0,
            color = "black") +
  xlim(0, 1.2) +
  scale_fill_gradient(low = "#4DBBD5FF", high = "#DC0000FF") +
  theme(legend.position = "right",
        plot.subtitle = element_text(hjust = 0),
        axis.text.x = element_text(angle = 0, vjust = 0.5, hjust=0.5, size = 12, color = "black"),
        axis.text.y = element_text(size = 12, angle = 0, color = "black"),
        panel.grid.minor.y = element_blank(),
        legend.text = element_text(size=12),
        legend.title = element_text(size = 12, face = "bold")) +
  labs(title = "Model assessment",
       subtitle = "Random forest classification model for \nVaccination vs. Infection",
       x = "Estimate",
       y = "",
       fill = "Estimate")

ggsave(final_model_metrics,
        file = here("Figures",
                    paste0("final_model_metrics_", filename, outcomevar,  ".png")),
        width = 5,
        height = 3)


#Which are the incorrect?
final_res_preds %>% rownames_to_column("sample") %>%  
  inner_join(ann_vaccines_samples %>% 
               select(condition, sample), by = "sample") %>% 
  filter(correct == "Incorrect") %>% 
  select(sample, condition, everything())
```

Precision recall curve
```{r}
# Precision recall
pr_curve_res = pr_curve(final_res_preds, outcomevar...6, .pred_I) %>% 
  ggplot(aes(x = recall, y = precision)) +
  geom_path(color = "#DC0000FF",
            size = 2) +
  coord_equal() +
  theme_few() +
  labs(title = "Precision-Recall curve",
       subtitle = "Random forest classification \nmodel for Infection vs. Vaccination",
       x = "Recall",
       y = "Precision") 


pr_curve_res %>% 
  ggsave(file = here("Figures",
                    paste0("PR_Curve_tidymodels_", filename, outcomevar,  ".png")),
         width = 4,
         height = 3)

```

Multi ROC
```{r fig.height=5, fig.width=5}
multi_roc = final_res_preds %>% 
  group_by(id) %>%
  roc_curve(outcomevar...9, .pred_H:.pred_VV) %>%
  ggplot(aes(1 - specificity, sensitivity, color = .level)) +
  geom_abline(lty = 2, color = "gray80", size = 0.5) +
  geom_path(show.legend = FALSE, alpha = 0.6, size = 1.2) +
  facet_wrap(~.level, ncol = 3) +
  coord_equal() +
  theme_few() +
    theme(axis.text = element_text(size = 6, color = "black")) +
  labs(title ="Receiver-operator curve",
       subtitle = "Multiclass classification, Randomforest",
       x = "1-Specificity",
       y = "Sensitivity") +
  scale_colour_manual(values = c(ann_vaxsig_colors$type[names(ann_vaxsig_colors$type) != "H"], 
                                 H = "black"))
```

Confusion matrix
```{r fig.height=5, fig.width=5}
conf_matrix_heat = final_res_preds %>% 
  conf_mat(outcomevar...9, .pred_class) %>%
  autoplot(type = "heatmap") +
  scale_fill_gradient(low = "white", high = "#DC0000FF") +
  labs(title ="Confusion matrix",
       subtitle = "Multiclass classification, Randomforest",
       x = "Truth",
       y = "Prediction")

conf_matrix_heat / 
  multi_roc
  
```



```{r}
#Matrix counts of VIP genes
all_matrices_counts_genesRF = all_matrices_counts_long_annotated %>% 
  inner_join(genes_rf, by = "genes")
```


### Relative effects

Function
```{r fig.height=5, fig.width=7}
REffect <- function(data, dep_vars, indep_var, n_boot = 1000, 
                    plot = TRUE, plot_colors = c("red", "darkblue"), 
                    plot_labels = list(y = "Relative effect", x = "Genes"), 
                    plot_theme = theme_few(), plot_legend_position = "top", 
                    plot_legend_labels = c("Vaccination", "Infection")) {
  
  library(npmv)
  library(ggplot2)
  library(reshape2)
  
  if(!all(data[[indep_var]] %in% c(0, 1))) {
    stop("A variável independente precisa conter apenas 0 e 1 (dois grupos).")
  }
  
  n_aab <- nrow(data)
  list_boot_aab <- list()
  
  # Loop de Bootstrap
  for (i in 0:n_boot) {
    if (i == 0) {
      df_sample <- 1:n_aab
    } else {
      df_sample <- sample(1:n_aab, n_aab, replace = TRUE)
    }
    
    formula_str <- paste(dep_vars, collapse = " | ")
    formula <- as.formula(paste(formula_str, "~", indep_var))
    
    manova_aab <- nonpartest(formula, data = data[df_sample, ], permtest = FALSE, plots = FALSE,
                             permreps = 1000, tests = c(1, 0, 0, 0))
    
    if (i == 0) obs_aab <- manova_aab
    
    list_boot_aab[[i + 1]] <- manova_aab
    
    cat("ANOVA/ boot/ aab/ i = ", i, "\n")
  }
  
  obs_manova <- list_boot_aab[[1]]$twogroupreleffects
  row_manova <- rownames(obs_manova)
  
  releffects <- lapply(list_boot_aab[-1], "[[", 2)
  
  df_perc_boot <- lapply(seq_len(nrow(obs_manova)), function(i) {
    q <- sapply(seq_len(ncol(obs_manova)), function(j) {
      v <- sapply(releffects, function(r) {
        row <- rownames(r)
        if(row[1] != row_manova[1]) r <- r[2:1,]
        r[i, j]
      })
      quantile(v, probs = c(0.025, 0.975))
    })
    colnames(q) <- colnames(obs_manova)
    df <- data.frame(Var2 = colnames(q), t(q),
                     group = ifelse(i == 1, row_manova[1], row_manova[2]))
    df$Var2 <- factor(df$Var2, levels = sort(unique(df$Var2)), ordered = TRUE)
    return(df)
  })
  
  names(df_perc_boot) <- row_manova
  
  df <- reshape2::melt(
    cbind(obs_manova, group = row_manova),
    id.vars = "group"
  )
  df$Var1 <- "point_est"
  df <- df[, c(4, 2, 3, 1)]
  colnames(df)[2] <- "Var2"
  
  df <- dplyr::arrange(df, group, desc(value), Var2)
  df$Var2 <- as.character(df$Var2)
  df$Var2 <- factor(df$Var2, levels = unique(df$Var2), ordered = TRUE)
  
  for (k in seq_len(2)) {
    df_perc_boot[[k]]$Var2 <- factor(df_perc_boot[[k]]$Var2, levels = levels(df$Var2), ordered = TRUE)
  }
  
  if (plot) {
    p <- ggplot(data = df, aes(x = Var2, y = value, group = group, color = group)) +
      geom_ribbon(inherit.aes = FALSE, data = df_perc_boot[[1]],
                  aes(x = Var2, ymin = X2.5., ymax = X97.5., group = group), fill = plot_colors[1],
                  alpha = 0.3) +
      geom_ribbon(inherit.aes = FALSE, data = df_perc_boot[[2]],
                  aes(x = Var2, ymin = X2.5., ymax = X97.5., group = group), fill = plot_colors[2],
                  alpha = 0.3) +
      geom_line() +
      geom_point(aes(size = value)) +
      scale_colour_manual("", values = c("1" = plot_colors[2], "0" = plot_colors[1]),
                          labels = plot_legend_labels) +
      plot_theme +
      theme(
        legend.text = element_text(size = 12),
        axis.text.x = element_text(size = 12, angle = 90),
        axis.text.y = element_text(size = 12),
        axis.line.y = element_blank(),
        axis.title = element_text(size = 15),
        panel.grid.major = element_line(),
        legend.position = plot_legend_position,
        legend.box.background = element_rect()
      ) +
      labs(y = plot_labels$y, x = plot_labels$x)
    
    print(p)
  }
  
  return(list(obs_manova = obs_manova, df_perc_boot = df_perc_boot, plot = p))
}
```

Plot
```{r fig.height=5, fig.width=7, message=FALSE, warning=FALSE}
head(all_matrices_counts_genesRF)

covid_vac_releff = all_matrices_counts_genesRF %>% 
  select(genes, sample, counts) %>% 
  pivot_wider(names_from = "genes",
              values_from = "counts") %>% 
  inner_join(all_matrices_counts_genesRF %>% 
               select(sample, disease_vac), by = "sample") %>% 
  mutate(disease_vac = if_else(disease_vac == "I", 1, 0))


releffect_disease_vac = REffect(covid_vac_releff, dep_vars = colnames(covid_vac_releff %>% 
                                              select(-disease_vac, -sample)),
        indep_var = "disease_vac", plot_colors = c("red", "#4DBBD5FF"), 
        n_boot=1000)


plot = releffect_disease_vac$plot +
  scale_colour_manual(values = c("1" = "#DC0000FF", 
                                 "0" = "#4DBBD5FF"),
                      labels = plot_legend_labels) +
  geom_ribbon(inherit.aes = FALSE, data = releffect_disease_vac$df_perc_boot[[1]],
                  aes(x = Var2, ymin = X2.5., ymax = X97.5., group = group), fill = "#4DBBD5FF",
                  alpha = 0.3) +
      geom_ribbon(inherit.aes = FALSE, data = releffect_disease_vac$df_perc_boot[[2]],
                  aes(x = Var2, ymin = X2.5., ymax = X97.5., group = group), fill = "#DC0000FF",
                  alpha = 0.3) +
  theme(
        legend.text = element_text(size = 12),
        axis.text.x = element_text(size = 12, angle = 45, hjust = 1, vjust = 1, color = "black"),
        axis.text.y = element_text(size = 12, color = "black"),
        axis.line.y = element_blank(),
        axis.title = element_text(size = 15),
        panel.grid.major = element_line(),
        legend.position = "top",
        plot.title = element_text(hjust = 0.5, size = 20),
        plot.subtitle = element_text(hjust = 0.5, size = 15),
        legend.box.background = element_blank()
      ) +
      labs(y = "Relative effect", x = "Genes",
           title = "Relative effects of VIP genes",
           subtitle = "Infection vs. Vaccination",
           color = "") +
  ylim(0, 1)

plot

```

### Butterfly plot
```{r}
##### Butterfly plot
all_degs_p_05_vac_infected = read_csv("DGE analysis/all_degs_p_05_vac_infected_19_12_23.csv")
df_aggregated = 

df_wide = all_degs_p_05_vac_infected |>
  inner_join(ImmuneGO_Genes %>% select(genes) %>% distinct(), by = "genes") %>% 
  group_by(genes, disease_vac) |>
  summarise(avg_log2FC = mean(log2fold_change), .groups = 'drop') |>
  pivot_wider(names_from = disease_vac, values_from = avg_log2FC) |>
  mutate(Regulation = case_when(V >= 0 & I <= 0 ~ "no",
                               V >= 0 & I >= 0 ~ "up",
                               V <= 0 & I <= 0 ~ "down",
                               TRUE ~ "ns"))   

butterfly_plot = ggplot(df_wide, aes(x = V, y = I, color = Regulation)) +
  geom_point(aes(size = abs(V),
                 label = genes)) +
  geom_label_repel(data = df_wide %>% filter(V > 2.5 & I > 0 | 
                                               V < -1 & I < -2), aes(label = genes)) +
  theme_minimal() +
  scale_color_manual(values = c("no" = "gray",
                                "ns" = "gray",
                                "down" = "royalblue",
                                "up" = "red")) +
  theme_bw() +
  labs(x = "Vaccination",
       y = "Infection",
       size = "Abs(L2FC)",
       title = "Mean Log2FC")
butterfly_plot
ggplotly(butterfly_plot)

```

### Heatmap by sample, prediction

```{r fig.height=10, fig.width=10}
pred_genes_counts = final_res_preds %>% 
  select(pred_class = .pred_class,
         pred_I = .pred_I,
         pred_V = .pred_V,
         truth = outcomevar...6,
         correct, 
         ACE:ZNF683) %>% 
  pivot_longer(cols = ACE:ZNF683, 
               names_to = "genes",
               values_to = "values") %>% 
  inner_join(genes_rf, by = "genes") %>% 
  inner_join(normalized_counts_long, by = "genes")
  
pred_genes_counts %>% 
  filter(disease_vac != "H",
         disease_vac != "Healthy") %>% 
  ggplot() +
  aes(y = sample,
      x = genes,
      fill = expression) +
  geom_tile() +
  facet_wrap(~disease_vac~type,scales = "free_y")
```


### Heatmap: Gene L2FC per condition

```{r}
# Required files
ImmuneGO_Annotated_Genes = readRDS(here("VaxGO gene sets","ImmuneGO_Annotated_Genes_2024-03-26.rds"))
df = normalized_counts
ann_vaccines <- read_csv(here("Annotations and metadata", "ann_vaccines_19-1-24.csv"))
ann_vaccines_samples <- read_csv(here("Annotations and metadata", "ann_vaccines_samples.csv"))


#Inputs
ImmuneGO_Annotated_Genes_all = ImmuneGO_Annotated_Genes %>% 
  filter(!process %in% c("TCR REPERTOIRE", "BCR REPERTOIRE"))

ImmuneGO_Genes = all_degs_p_05_vac_infected_19_12_23 %>% 
  inner_join(ImmuneGO_Annotated_Genes %>% 
               select(genes, process) %>% 
               distinct() %>% 
               select(genes),
               by = "genes") %>% 
  distinct()

# Matrix
l2fc_imps = genes_rf %>% 
  select(genes = Variable, importance = Importance) %>% 
  inner_join(ImmuneGO_Genes, by = "genes") %>%
  select(genes, condition, log2fold_change) %>% 
  pivot_wider(names_from = condition, values_from = log2fold_change) %>% 
  column_to_rownames("genes") %>% 
  as.matrix() %>% 
  replace(is.na(.), 0)

```

```{r}

#Which VIP genes?
#vipgenes_tidymodels_disease_vac2024_04_24 <- read_csv("vipgenes_tidymodels_disease_vac2024-04-24.csv")
vipgenes_tidymodels_type <- read_csv("vipgenes_tidymodels_type2024-06-21.csv")

vip_genes = vipgenes_tidymodels_type
outcomevar = "_Type_"

rf_genes = vip_genes %>% 
  clean_names() %>% 
  select(genes = variable, importance)

ImmuneGO_Genes = all_degs_p_05_vac_infected_19_12_23 %>% 
  inner_join(rf_genes %>% 
               select(genes) %>% 
               distinct() %>% 
               select(genes),
               by = "genes") %>% 
  distinct()

####### INPUT
data_2 = ImmuneGO_Genes
filename_2 = paste0("RandomForest_ImmuneGO_Genes", outcomevar, today())

######## Matrix
matrix = data_2 %>% 
  select(genes, condition, log2fold_change) %>%
  pivot_wider(names_from = "condition", 
              values_from = "log2fold_change", 
              values_fn = mean) %>% 
  replace(is.na(.), 0) %>%
  arrange(genes) %>% 
  column_to_rownames(var="genes") %>% 
  as.matrix()

######## Annotations
ann_vaccines_covid_other_2 = data_2 %>% 
  select(condition) %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  distinct() %>% 
  select(condition, disease_vac, type, week, day) %>% 
  arrange(week, day) %>% 
  mutate(week = fct_relevel(as.factor(week), sort(levels(week))),
         day = fct_relevel(as.factor(day), sort(levels(day))))

#Columns
ann_cols_heatmap = matrix %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  inner_join(ann_vaccines_covid_other_2, by = "condition") %>% 
  distinct() %>% 
  arrange(condition) %>% 
  column_to_rownames("condition")

matrix_data = matrix %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  inner_join(ann_cols_heatmap %>% rownames_to_column("condition"), by = "condition") %>% 
  select(!c(disease_vac:day)) %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>%
  as.matrix() 

#Rows Immune ----
ann_rows_immune = matrix_data %>%
  as.data.frame() %>%
  rownames_to_column("genes") %>%
  left_join(ImmuneGO_Annotated_Genes %>% filter(go_term == "Manual"), by = "genes") %>%
  select(genes, process) %>%
  distinct() %>%
  group_by(genes) %>%
  arrange(genes, factor(process, levels = c("INNATE IMMUNE SYSTEM", "ADAPTIVE IMMUNE SYSTEM")), .by_group = TRUE) %>%
  mutate(i_process = row_number()) %>%
  pivot_wider(., names_from = "i_process", values_from = "process") %>%
  column_to_rownames("genes") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("index") %>%
  mutate(index = as.numeric(index)) %>%
  arrange(desc(index)) %>%
  column_to_rownames("index") %>%
  t() %>%
  as.data.frame() %>%
  replace(is.na(.), ".") %>% 
  mutate(`1` = if_else(`1` == ".", "INNATE IMMUNE SYSTEM", `1`)) 

#Verificar dimensões do dataset
dim(matrix)
dim(matrix_data)
dim(ann_cols_heatmap)
dim(ann_rows_immune)
```

```{r}
################# Immune 
#Heatmap annotation
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "right")

row_ha = HeatmapAnnotation(df = ann_rows_immune,
                       col = col_process_immune,
                       which= "row",
                       show_annotation_name = F,
                       show_legend = F,
                       simple_anno_size = unit(2, "mm"))

# Values colors
col_fun <- colorRamp2(c(-2, 0, 2), c("#9bc1bc", "white", "#ed6a5a"))

file = filename_2
mat = matrix_data
width_image = 30
height_image = 15
width = unit(15, "cm")
height = unit(5, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)
rect_gp = gpar(col = "gray80", lwd = 0.5)

fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap.png")

png(file = fileimageheatmap_grouped, 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "L2FC",
                        col = col_fun, #color
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(0, "cm"),
                        row_split = NULL,
                        column_split = NULL,
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(2, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height)

heatmap_plot = draw(heatmap_plot, 
     annotation_legend_side = "left", 
     heatmap_legend_side = "left")

dev.off()

```

### Boxplot: VIP gene expression by condition
Import tables
```{r fig.height=10, fig.width=10}
genes_rf = read_csv("Machine learning/vipgenes_tidymodels_ImmuneGO_Genes_Type_including_Infection_type2024-06-27.csv") 
ann_vaccines_conditions = read_csv(here("Annotations and metadata", "ann_vaccines_conditions.csv"))

all_matrices_counts_long_annotated <- readRDS(here("DGE analysis", "all_matrices_counts_long_annotated.rds"))

all_samples_l2fc = read_rds(here("DGE analysis", "all_samples_l2fc.rds"))


all_matrices_counts_genesRF = all_matrices_counts_long_annotated %>% 
  filter(genes %in% genes_rf$Variable) 

all_matrices_counts_genesRF %>% 
  saveRDS(here("Machine learning", "all_matrices_counts_genesRF.rds"))
```


```{r fig.height=10, fig.width=10}
df = all_samples_l2fc
table = df %>% 
  select(-condition) %>% 
  filter(genes == "IL10") %>% 
  #filter(genes %in% genes_rf$Variable) %>% 
  inner_join(ann_vaccines_samples, by = "sample") %>% 
  mutate(disease_vac = if_else(disease_vac == "Healthy", "H", disease_vac)) %>% 
  mutate(disease_vac = fct_relevel(disease_vac, "H", "I", "V"),
           type = fct_relevel(type, "I", "RNA", "VV", "IN", "SU")) %>% 
  arrange(disease_vac,type, condition) %>% 
  inner_join(ann_vaccines_conditions %>% select(condition, samples), by = "condition") %>% 
  mutate(condition = factor(condition, levels = c(unique(condition[order(condition)]))),
         condition = paste0(condition, " (n=", samples, ")")) 
  # mutate(genes = fct_relevel(genes, genes_rf$Variable))

table_ready = table %>% 
  mutate(condition = fct_relevel(condition, 
                                 table %>% 
                                   filter(str_detect(condition, "^H \\(D0\\)")) %>% 
                                   select(condition) %>% 
                                   distinct() %>% 
                                   pull(condition), 
                                 table %>% 
                                   filter(str_detect(condition, "^I")) %>% 
                                   filter(!str_detect(condition, "^I-I")) %>% 
                                   select(condition) %>% 
                                   arrange(condition) %>% 
                                   distinct() %>% 
                                   pull(condition),
                                 table %>% 
                                   filter(disease_vac == "I") %>% 
                                   filter(str_detect(condition, "^BNT-I")) %>% 
                                    filter(!str_detect(condition, "^BNT-I \\(D\\d+\\)")) %>% 
                                    arrange(day, condition) %>%
                                   select(condition) %>% 
                                    distinct() %>% 
                                    pull(condition),
                                 table %>% 
                                    filter(str_detect(condition, "^BNT-I \\(D\\d+\\)")) %>% 
                                    arrange(day, condition) %>%
                                   select(condition) %>% 
                                    distinct() %>% 
                                    pull(condition),
                                 table %>% 
                                   filter(disease_vac == "V",
                                          str_detect(condition, "^BNT \\(V1, D\\d+\\)")) %>% 
                                   arrange(day, condition) %>% 
                                   select(condition) %>% 
                                   distinct() %>% 
                                   pull(condition),
                                 table %>% 
                                   filter(disease_vac == "V",
                                          !str_detect(condition, "^BNT \\(V1, D\\d+\\)"),
                                          !str_detect(condition, "BBIBP"),
                                          !str_detect(condition, "ZF2001")) %>% 
                                   arrange(condition) %>% 
                                   select(condition) %>% 
                                   distinct() %>% 
                                   pull(condition),
                                 table %>% 
                                   filter(str_detect(condition, "BBIBP")) %>% 
                                   select(condition) %>% 
                                   arrange(condition) %>% 
                                   distinct() %>% 
                                   pull(condition), 
                                 table %>% 
                                   filter(str_detect(condition, "ZF2001")) %>% 
                                   select(condition) %>% 
                                   arrange(condition) %>% 
                                   distinct() %>% 
                                   pull(condition))) %>%
  mutate(type = if_else(type == "H", "I", type),
         disease_vac = if_else(disease_vac == "H", "I", disease_vac))
```


```{r fig.height=10, fig.width=10}
#Base value -----
median_h = table_ready %>% 
  select(genes, sample, log2fc, disease_vac) %>% 
  filter(disease_vac == "H") %>% 
  distinct() %>% 
  group_by(genes, disease_vac) %>% 
  summarize(median = median(counts), .groups = 'drop')

median_i = table_ready %>% 
  select(genes, sample, counts, disease_vac) %>% 
  filter(disease_vac == "H") %>% 
  distinct() %>% 
  group_by(genes, disease_vac) %>% 
  summarize(median = median(counts), .groups = 'drop') %>% 
  mutate(disease_vac = "I")

median_v = table_ready %>% 
  select(genes, sample, counts, disease_vac) %>% 
  filter(disease_vac == "H") %>% 
  distinct() %>% 
  group_by(genes, disease_vac) %>% 
  summarize(median = median(counts), .groups = 'drop') %>% 
  mutate(disease_vac = "V")

median = bind_rows(median_i, median_v)
```


```{r fig.height=5, fig.width=12, message=TRUE, warning=FALSE}
#Plots -----

disease_vac_boxplot_samples_VIP = table_ready %>% 
  # filter(genes == "GZMM") %>% 
  ggplot() +
  aes(x = condition, y = log2fc, fill = if_else(day == 0, "Day 0", type)) +  
  geom_hline(yintercept = 0, linetype = 2) +
  scale_y_log10()+
  geom_jitter(aes(color = if_else(day == 0, "Day 0", type)),
              alpha = 0.3, size = 0.1) +
  geom_boxplot(outliers = F) +
  theme_few() +
  scale_fill_manual(values = c("Day 0" = "white", 
                               ann_vaxsig_colors$type)) +
  scale_color_manual(values = c("Day 0" = "gray80", 
                               ann_vaxsig_colors$type)) +
  theme(axis.text.x = element_text(angle = 45,
                                   hjust = 1,
                                   size = 8,
                                   vjust = 1),
        plot.title = element_text(hjust = 0, face = "bold"),
        plot.margin = margin(0, 0, 0, 2, "cm"),
        strip.text.y = element_text(angle = 0),
        panel.grid.major = element_line(size = 0.05),
        panel.grid.minor = element_line(size = 0.05)) +
  labs(y = "Log2 fold-change",
       title = "VIP genes",
       subtitle = "Random forest classification model for Infection vs. Vaccination \n
       Log2FC calculated comparing each sample with their control (D0)",
       x="",
       fill = "Type") +
  guides(color = "none") +
  facet_grid(~genes~disease_vac, scales = "free")

disease_vac_boxplot_samples_VIP %>% 
  ggsave(file = here("Figures" , "ML_infec_vs_vac_boxplot_vipgenes.png"), 
         width = 10,
         height = 5)
```


```{r fig.height=10, fig.width=10}
#Type
vip_genes_boxplot_type = table %>% 
  filter(disease_vac != "H") %>% 
  ggplot() +
  aes(x = type, y = counts, fill = week_fac) +
  geom_boxplot() +
  geom_hline(data = median, aes(yintercept = median), linetype = "dashed", alpha = 0.5) +
  scale_fill_viridis_d(option = "cividis", direction = -1) +
  scale_y_continuous(trans = "log10") +
  labs(
    x = "Type",
    y = "Normalized counts",
    title = "Gene expression of the 5 most important genes",
    fill = "Week"
  ) +
  theme_linedraw() +
  theme(axis.text.y = element_text(hjust = 0.1),
        legend.position = "top") +
  facet_grid(~genes~disease_vac, scales = "free")

vip_genes_boxplot_type 

vip_genes_boxplot_type %>% 
  ggsave(file = here("Figures", paste0("vip_genes_boxplot_type", filename, outcomevar, today(), ".png")), 
                                  width = 10, height = 15)

#Sex
vip_genes_boxplot_type_sex = table %>% 
  mutate(type_sex = paste0(type, "_", sex),
         type_sex = fct_relevel(type_sex, "IN_M/F", "SU_M/F")) %>% 
  filter(disease_vac != "H",
         !type %in% c("IN", "SU")) %>% 
  ggplot() +
  aes(x = type_sex, y = counts, fill = week_fac) +
  geom_boxplot() +
  geom_hline(data = median, aes(yintercept = median), linetype = "dashed", alpha = 0.5) +
  scale_fill_viridis_d(option = "cividis", direction = -1) +
  scale_y_continuous(trans = "log10") +
  labs(
    x = "Type",
    y = "Normalized counts",
    title = "Gene expression of the 5 most important genes",
    fill = "Week"
  ) +
  theme_linedraw() +
  theme(axis.text.y = element_text(hjust = 0.1),
        axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) +
  facet_grid(~genes~disease_vac, scales = "free")

vip_genes_boxplot_type_sex 


```


### PCA with VIP genes
**Input**

```{r}
#INPUT
data_genes = all_matrices_counts_genesRF 
data_annotation = ann_vaccines_samples
filename = "RandomforestGenes_COVID_Infec_Vac"
```

```{r}
#Convert long to wide table format.
#Dataframe with sample annotation

ann_vaccines_pca_matrix = data_genes %>% 
  select(sample, genes, counts) %>% 
  pivot_wider(., names_from = "sample", 
                     values_from = "counts") %>% 
  replace(is.na(.), 0) %>%
  column_to_rownames("genes") %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column("sample") %>% 
  inner_join(data_annotation %>% 
               select(sample, condition, disease_vac, gse_id, vaccine, day, week, dose, infection, type), by = "sample") 

# Convert dataframe to matrix without annotation columns
ann_vaccines_pca_matrix_ready = ann_vaccines_pca_matrix %>% 
  column_to_rownames("sample") %>% 
  select(!condition:type) %>% 
  as.matrix()

# Delete columns with near 0 variance
nearZeroVarCols <- nearZeroVar(ann_vaccines_pca_matrix_ready, saveMetrics = TRUE)
ann_vaccines_pca_matrix_ready <- ann_vaccines_pca_matrix_ready[, !nearZeroVarCols$nzv]
pca_res <- prcomp(ann_vaccines_pca_matrix_ready, scale. = T)

# Correlation plot ----
corr_matrix = cor(ann_vaccines_pca_matrix_ready %>% scale()) 
var <- get_pca_var(pca_res)

# corrplot = ggcorrplot(corr_matrix, hc.order = TRUE) + 
#   theme(axis.text.x = element_text(angle = 90, size = 5), 
#         axis.text.y = element_text(size = 5))
# #Salvar
# ggsave(corrplot, file = paste0(filename, "_corrplot.png"), 
#        width = 20, #Grande 20, pequeno 10
#        height= 20) #Grande 20, pequeno 10
# print(corrplot)

#PCA ----
data.pca = prcomp(corr_matrix) #PCA
summary(corr_matrix) #Retornar PCs

#Scree plot ----
scree_plot = fviz_eig(data.pca, 
                      addlabels = TRUE,
                      ylim = c(0, 70)) +
  geom_col(color = "#00AFBB", fill = "#00AFBB") +
  theme_classic()

#Save
ggsave(scree_plot, file = here("Figures", paste0(filename, "_screeplot.png")), width = 10, height = 3) 
print(scree_plot)

```

```{r}
# Loadings plot ----
options(ggrepel.max.overlaps = Inf)

circle_contrib = fviz_pca_var(pca_res, col.var = "cos2",
                              gradient.cols = c("black", "#DC0000FF"),
                              select.var= list(cos2 = 20), 
                              repel = T, 
                              labelsize = 4,
                              col.circle = NA) +
  xlab("") + 
  ylab("") +
  theme_minimal() +
theme(panel.grid = element_blank(),  # Remover linhas de grade
        axis.text = element_blank(),   # Remover rótulos de texto dos eixos
        axis.ticks = element_blank()) 

#Save
ggsave(circle_contrib, file = here("Figures", paste0(filename, "_circle_contrib_wide.png")), width = 10, height = 5)

circle_contrib
```


```{r}
#KNN classification -----
#Use the screeplot generated % of explained variance for each dimension
scree_plot #Dim 1 = 68.2%, Dim 2 = 15.9%


# PC1-PC2 ------
#Determine the number of clusters
pca_scores <- data.frame(pca_res$x[, 1:2])
fviz_nbclust(pca_scores,  
                     FUNcluster = kmeans,
                     method = "wss")

pcas = "PC1-PC2"

set.seed(666) # Set seed for randomization
cluster_model <- kmeans(pca_res$x[, 1:2], centers = 3)  #Set the number of clusters

ann_vaccines_pca_matrix$cluster <- as.factor(cluster_model$cluster)
```

```{r fig.height=7, fig.width=10, paged.print=FALSE}
# Condition with density plot ------
pca_plot_knn_cluster = autoplot(pca_res, 
        data = ann_vaccines_pca_matrix,
        colour = "gse_id")  +
  stat_ellipse(aes(color = cluster, fill = cluster), 
               geom = "polygon", 
               alpha = 0.2, 
               linetype = 1,
               size = 0.3,
               type = "t") +
  labs(title=paste0(filename, "_bycondition")) + 
  xlab("PC1 (68.2%)") +
  ylab("PC2 (15.9%)") +
    geom_hline(yintercept = 0, size = 0.3, color = "gray50", linetype = 4) +
  geom_vline(xintercept = 0, size = 0.3, color = "gray50", linetype = 4) +
  geom_point(aes(fill = type,
                 color = type),
             stroke = 0.2) +
   scale_color_manual(values = c("1" = "#8491B4FF", 
                                "2" = "#4DBBD5FF", 
                                "3" = "#DC0000FF",
                                ann_vaxsig_colors$type), 
                     guide = "none") +
  scale_fill_manual(values = c("1" = "#8491B4FF", 
                               "2" = "#4DBBD5FF", 
                               "3" = "#DC0000FF",
                               ann_vaxsig_colors$type),
                    guide = "none") + 
  guides(col="none") +
  theme_few() +
  theme(panel.grid.minor = element_blank(),
        text = element_text(color = "black"),
        panel.border = element_blank(),
        axis.line = element_line(colour = "black")) +
  geom_text_repel(aes(label = condition,
                      segment.colour="gray70"),
                  box.padding = 0.3,
                  size = 3) +
  xlim(-0.4, 0.6)
pca_plot_knn_cluster

ggsave(pca_plot_knn_cluster, 
       file = here( "Figures", paste0(filename, pcas, "_KNN_Clustered_labels.png")), width = 10, height = 7)
pca_plot_knn_cluster
# pca_plot_knn_cluster_marginal = ggMarginal(
#   pca_plot_knn_cluster,
#   type = "histogram",
#   groupFill = T,
#   groupColour = T,
#   xparams = list(binwidth = 0.1, size = 0.1),
#   yparams = list(binwidth = 0.1, size = 0.1)
# )
# 
# ggsave(
#   pca_plot_knn_cluster_marginal,
#   file = here(paste0(filename, pcas, "_KNN_Clustered_labels_Marginal.png"), "Figures"),
#   width = 8,
#   height = 5
# )


```


### Heatmap: Gene L2FC per condition

```{r}
# Required files
ImmuneGO_Annotated_Genes = readRDS(here("VaxGO gene sets","ImmuneGO_Annotated_Genes_2024-03-26.rds"))
all_degs_p_05_vac_infected = read_csv(here("DGE analysis", "all_degs_p_05_vac_infected_19_12_23.csv"))
ann_vaccines <- read_csv(here("Annotations and metadata", "ann_vaccines_conditions.csv"))
df = read_csv("Machine Learning/vipgenes_tidymodels_ImmuneGO_Genes_Type_including_Infection_disease_vac.csv")
#Inputs
ImmuneGO_Annotated_Genes_all = ImmuneGO_Annotated_Genes %>% 
  filter(!process %in% c("TCR REPERTOIRE", "BCR REPERTOIRE"))

ImmuneGO_Genes = all_degs_p_05_vac_infected %>% 
  filter(genes%in% df$genes) %>% 
  inner_join(ImmuneGO_Annotated_Genes %>% 
               select(genes, process) %>% 
               distinct() %>% 
               select(genes),
               by = "genes") %>% 
  distinct() %>% 
  select(genes, condition, log2fold_change) %>% 
  pivot_wider(names_from = condition, values_from = log2fold_change)

# Matrix
l2fc_imps = ImmuneGO_Genes %>% 
  column_to_rownames("genes") %>% 
  as.matrix() %>% 
  replace(is.na(.), 0)

outcomevar = "Infection_Vaccine"
####### INPUT
filename_2 = paste0("RandomForest_ImmuneGO_Genes_Conditions_", outcomevar)

######## Matrix
matrix = ImmuneGO_Genes

######## Annotations
ann_vaccines_covid_other_2 = ImmuneGO_Genes %>% 
  column_to_rownames("genes") %>% 
  colnames() %>% 
  as.data.frame() %>% 
  select(condition = ".") %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  distinct() %>% 
  select(condition, disease_vac, type, week, day) %>% 
  arrange(week, day) %>% 
  mutate(week = fct_relevel(as.factor(week), sort(levels(week))),
         day = fct_relevel(as.factor(day), sort(levels(day)))) 

```


```{r}
#Columns
ann_cols_heatmap = matrix %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  inner_join(ann_vaccines_covid_other_2, by = "condition") %>% 
  distinct() %>% 
  arrange(condition) %>% 
  column_to_rownames("condition")

matrix_data = matrix %>% 
  column_to_rownames("genes") %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  inner_join(ann_cols_heatmap %>% rownames_to_column("condition"), by = "condition") %>% 
  select(!c(disease_vac:day)) %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>%
  as.matrix()  %>% 
  replace(is.na(.), 0)

#Rows Immune ----
ann_rows_immune = matrix_data %>%
  as.data.frame() %>%
  rownames_to_column("genes") %>%
  left_join(ImmuneGO_Annotated_Genes %>% filter(go_term == "Manual"), by = "genes") %>%
  select(genes, process) %>%
  distinct() %>%
  group_by(genes) %>%
  arrange(genes, factor(process, levels = c("INNATE IMMUNE SYSTEM", "ADAPTIVE IMMUNE SYSTEM")), .by_group = TRUE) %>%
  mutate(i_process = row_number()) %>%
  pivot_wider(., names_from = "i_process", values_from = "process") %>%
  column_to_rownames("genes") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("index") %>%
  mutate(index = as.numeric(index)) %>%
  arrange(desc(index)) %>%
  column_to_rownames("index") %>%
  t() %>%
  as.data.frame() %>%
  replace(is.na(.), ".") %>% 
  mutate(`1` = if_else(`1` == ".", "INNATE IMMUNE SYSTEM", `1`)) %>% 
  rownames_to_column("genes") %>% 
  mutate(genes = rownames(matrix_data)) %>% 
  column_to_rownames("genes")

#Verificar dimensões do dataset
dim(matrix)
dim(matrix_data)
dim(ann_cols_heatmap)
dim(ann_rows_immune)
```

```{r}
################# Immune 
#Heatmap annotation
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "right")

row_ha = HeatmapAnnotation(df = ann_rows_immune,
                       col = col_process_immune,
                       which= "row",
                       show_annotation_name = F,
                       show_legend = F,
                       simple_anno_size = unit(1, "mm"))

# Values colors
col_fun <- colorRamp2(c(-2, 0, 2), c("#9bc1bc", "white", "#ed6a5a"))

file = filename_2
mat = matrix_data
width_image = 30
height_image = 15
width = unit(15, "cm")
height = unit(6, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)
rect_gp = gpar(col = "gray80", lwd = 0.5)

fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap.png")

png(file = here("Figures", fileimageheatmap_grouped), 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC",
                        col = col_fun, #color
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(0, "cm"),
                        row_split = NULL,
                        column_split = NULL,
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(2, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height)

heatmap_plot = draw(heatmap_plot, 
     annotation_legend_side = "left", 
     heatmap_legend_side = "bottom")

dev.off()

```



### Heatmap: Gene counts per sample
```{r}
# Required files
ImmuneGO_Annotated_Genes = readRDS(here("VaxGO gene sets","ImmuneGO_Annotated_Genes_2024-03-26.rds"))

genes_rf = read_csv("Machine learning/vipgenes_tidymodels_All_coviddisease_vac.csv")
df_samples_l2fc_vipgenes_covid = all_samples_l2fc %>% 
  filter(genes %in% genes_rf$genes)

ann_vaccines_samples <- read_csv(here("Annotations and metadata", "ann_vaccines_samples.csv"))


df_annotated = df_samples_l2fc_vipgenes_covid %>% 
  distinct() %>% 
  inner_join(ann_vaccines_samples %>% select(-condition, -participant_id), by = "sample") %>% 
  distinct() %>% 
  filter(condition != "BNT-MO (V1, D0)",
         condition != "BNT-MO (V1, D6)") %>% 
  mutate(disease_vac = if_else(disease_vac == "Healthy", "H", disease_vac)) %>% 
  mutate(disease_vac = fct_relevel(disease_vac, "H", "I", "V"),
           type = fct_relevel(type, "I", "RNA", "VV", "IN", "SU")) %>% 
  inner_join(ann_vaccines_conditions %>% select(condition, samples), by = "condition") %>% 
  mutate(condition = factor(condition, levels = c(unique(condition[order(condition)]))),
         condition = paste0(condition, " (n=", samples, ")"))
  

df_annotated_2 = df_annotated %>% 
  mutate(condition = fct_relevel(condition, 
                                 df_annotated %>% 
                                   filter(str_detect(condition, "^H \\(D0\\)")) %>% 
                                   select(condition) %>% 
                                   distinct() %>% 
                                   pull(condition), 
                                 df_annotated %>% 
                                   filter(str_detect(condition, "^I")) %>% 
                                   filter(!str_detect(condition, "^I-I")) %>% 
                                   select(condition) %>% 
                                   arrange(condition) %>% 
                                   distinct() %>% 
                                   pull(condition),
                                 df_annotated %>% 
                                   filter(disease_vac == "I") %>% 
                                   filter(str_detect(condition, "^BNT-I")) %>% 
                                    filter(!str_detect(condition, "^BNT-I \\(D\\d+\\)")) %>% 
                                    arrange(day, condition) %>%
                                   select(condition) %>% 
                                    distinct() %>% 
                                    pull(condition),
                                 df_annotated %>% 
                                    filter(str_detect(condition, "^BNT-I \\(D\\d+\\)")) %>% 
                                    arrange(day, condition) %>%
                                   select(condition) %>% 
                                    distinct() %>% 
                                    pull(condition),
                                 df_annotated %>% 
                                   filter(disease_vac == "V",
                                          str_detect(condition, "^BNT \\(V1, D\\d+\\)")) %>% 
                                   arrange(day, condition) %>% 
                                   select(condition) %>% 
                                   distinct() %>% 
                                   pull(condition),
                                 df_annotated %>% 
                                   filter(disease_vac == "V",
                                          !str_detect(condition, "^BNT \\(V1, D\\d+\\)"),
                                          !str_detect(condition, "BBIBP"),
                                          !str_detect(condition, "ZF2001")) %>% 
                                   arrange(condition) %>% 
                                   select(condition) %>% 
                                   distinct() %>% 
                                   pull(condition),
                                 df_annotated %>% 
                                   filter(str_detect(condition, "BBIBP")) %>% 
                                   select(condition) %>% 
                                   arrange(condition) %>% 
                                   distinct() %>% 
                                   pull(condition), 
                                 df_annotated %>% 
                                   filter(str_detect(condition, "ZF2001")) %>% 
                                   select(condition) %>% 
                                   arrange(condition) %>% 
                                   distinct() %>% 
                                   pull(condition))) %>%
  mutate(type = if_else(type == "H", "I", type),
         disease_vac = if_else(disease_vac == "H", "I", disease_vac)) 


```

```{r}
df_annotated_2 %>% 
  filter(day != 0) %>% 
  filter(genes == "ITGAM") %>% 
  ggplot() +
  aes(x = condition, y = log2fc, fill = type) +  
  geom_hline(yintercept = 0, linetype = 2) +
  geom_jitter(aes(color = type),
              alpha = 0.3, size = 0.1) +
  geom_boxplot(outliers = F) +
  theme_few() +
  scale_fill_manual(values = c(ann_vaxsig_colors$type)) +
    scale_color_manual(values = c(ann_vaxsig_colors$type)) +
  theme(axis.text.x = element_text(angle = 45,
                                   hjust = 1,
                                   vjust = 1),
        plot.title = element_text(hjust = 0.5, face = "bold"),
        plot.margin = margin(0, 0, 0, 2, "cm"),
        panel.grid.major = element_line(size = 0.05),
        panel.grid.minor = element_line(size = 0.05)) +
  labs(y = "L2FC",
       x="",
       fill = "Type") +
  guides(color = "none") +
  facet_wrap(~disease_vac, scales = "free_x")
```

Generate boxplots for all genes
```{r fig.height=6, fig.width=12}
#Function to create boxplots
create_boxplots_vipgenes <- function(result_df) {
  
  unique_processes <- unique(result_df$genes)
  
  purrr::map(unique_processes, function(process_name) {
    cellmarker_boxplot <- result_df %>% 
      filter(day != 0) %>% 
  filter(genes == process_name) %>% 
  ggplot() +
  aes(x = condition, y = log2fc, fill = type) +  
  geom_hline(yintercept = 0, linetype = 2) +
  geom_jitter(aes(color = type),
              alpha = 0.3, size = 0.1) +
  geom_boxplot(outliers = F) +
  theme_few() +
  scale_fill_manual(values = c(ann_vaxsig_colors$type)) +
    scale_color_manual(values = c(ann_vaxsig_colors$type)) +
  theme(axis.text.x = element_text(angle = 45,
                                   hjust = 1,
                                   vjust = 1),
        plot.title = element_text(hjust = 0.5, face = "bold"),
        plot.margin = margin(0, 0, 0, 2, "cm"),
        panel.grid.major = element_line(size = 0.05),
        panel.grid.minor = element_line(size = 0.05)) +
  labs(y = "L2FC",
       x="",
       title = process_name,
       subtitle = "VIPgenes - COVID-19 infection vs. Vaccination",
       fill = "Type") +
  guides(color = "none") +
  facet_wrap(~disease_vac, scales = "free_x")

    ggsave(cellmarker_boxplot, 
           file = here("Figures", paste0("MachineLearning_VIPGenes_COVID_VAC_boxplot_", process_name, ".png")),
           width = 12, height = 6)
  })
}

# Apply to all processes
create_boxplots_vipgenes(df_annotated_2)


df_annotated_2$genes %>% unique()
```





```{r}
#Inputs
ImmuneGO_Annotated_Genes_all = ImmuneGO_Annotated_Genes %>% 
  filter(!process %in% c("TCR REPERTOIRE", "BCR REPERTOIRE"))

ImmuneGO_Genes = df %>% 
  inner_join(ImmuneGO_Annotated_Genes %>% 
               select(genes, process) %>% 
               distinct() %>% 
               select(genes),
               by = "genes") %>% 
  distinct() %>% 
  select(genes, sample, log2fc) %>% 
  pivot_wider(names_from = sample, values_from = log2fc)


sample_correct =final_res_preds %>% 
  rownames_to_column("sample") %>%  
  inner_join(ann_vaccines_samples %>% 
               select(condition, sample), by = "sample") %>% 
  select(sample, correct) 

# Matrix
l2fc_imps = ImmuneGO_Genes %>% 
  column_to_rownames("genes") %>% 
  as.matrix() %>% 
  replace(is.na(.), 0)

####### INPUT
data_2 = l2fc_imps
filename_2 = paste0("RandomForest_ImmuneGO_Genes_Samples_", outcomevar)

######## Matrix
matrix = ImmuneGO_Genes

######## Annotations
ann_vaccines_covid_other_2 = ImmuneGO_Genes %>% 
  column_to_rownames("genes") %>% 
  colnames() %>% 
  as.data.frame() %>% 
  select(sample = ".") %>% 
  inner_join(ann_vaccines_samples, by = "sample") %>% 
  distinct() %>% 
  select(sample, disease_vac, type, week, day) %>% 
  arrange(week, day) %>% 
  mutate(week = fct_relevel(as.factor(week), sort(levels(week))),
         day = fct_relevel(as.factor(day), sort(levels(day)))) %>% 
  inner_join(sample_correct, by = "sample")
  

#Columns
ann_cols_heatmap = matrix %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(sample = '.') %>% 
  inner_join(ann_vaccines_covid_other_2, by = "sample") %>% 
  distinct() %>% 
  arrange(sample) %>% 
  column_to_rownames("sample")

matrix_data = matrix %>% 
  column_to_rownames("genes") %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "sample") %>% 
  inner_join(ann_cols_heatmap %>% rownames_to_column("sample"), by = "sample") %>% 
  select(!c(disease_vac:day, correct)) %>% 
  arrange(sample) %>% 
  column_to_rownames("sample") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>%
  as.matrix()  %>% 
  replace(is.na(.), 0)

#Rows Immune ----
ann_rows_immune = matrix_data %>%
  as.data.frame() %>%
  rownames_to_column("genes") %>%
  left_join(ImmuneGO_Annotated_Genes %>% filter(go_term == "Manual"), by = "genes") %>%
  select(genes, process) %>%
  distinct() %>%
  group_by(genes) %>%
  arrange(genes, factor(process, levels = c("INNATE IMMUNE SYSTEM", "ADAPTIVE IMMUNE SYSTEM")), .by_group = TRUE) %>%
  mutate(i_process = row_number()) %>%
  pivot_wider(., names_from = "i_process", values_from = "process") %>%
  column_to_rownames("genes") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("index") %>%
  mutate(index = as.numeric(index)) %>%
  arrange(desc(index)) %>%
  column_to_rownames("index") %>%
  t() %>%
  as.data.frame() %>%
  replace(is.na(.), ".") %>% 
  mutate(`1` = if_else(`1` == ".", "INNATE IMMUNE SYSTEM", `1`)) %>% 
  rownames_to_column("genes") %>% 
  mutate(genes = rownames(matrix_data)) %>% 
  column_to_rownames("genes")

#Verificar dimensões do dataset
dim(matrix_data)
dim(ann_cols_heatmap)
dim(ann_rows_immune)
```

```{r}
################# Immune 
#Heatmap annotation
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "right")

row_ha = HeatmapAnnotation(df = ann_rows_immune,
                       col = col_process_immune,
                       which= "row",
                       show_annotation_name = F,
                       show_legend = F,
                       simple_anno_size = unit(1, "mm"))
min(matrix_data)

max(matrix_data)
# Values colors
col_fun <- colorRamp2(c(-20, 0, 20), c("#9bc1bc", "white", "#ed6a5a"))

file = filename_2
mat = matrix_data
width_image = 60
height_image = 40
width = unit(50, "cm")
height = unit(30, "cm")
column_names_gp = gpar(fontsize = 8)
row_names_gp = gpar(fontsize = 10)
rect_gp = gpar(col = "gray80", lwd = 0.5)

fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap.png")

png(file = here("Figures", fileimageheatmap_grouped), 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "Log10 (raw counts)",
                        col = col_fun, #color
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(0, "cm"),
                        row_split = NULL,
                        column_split = NULL,
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(2, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height)

heatmap_plot = draw(heatmap_plot, 
     annotation_legend_side = "left", 
     heatmap_legend_side = "bottom")

dev.off()

```

## COVID-19 vaccines vs infection (L2FC)

```{r}
ImmuneGO_Annotated_Genes = readRDS(here("VaxGO gene sets","ImmuneGO_Annotated_Genes_2024-03-26.rds"))
ImmuneGO_BTM_grouped_genes <- read_csv("VaxGO gene sets/ImmuneGO_BTM_grouped_genes.csv")

all_samples_l2fc = read_rds(here("DGE analysis", "all_samples_l2fc.rds"))
df_samples_l2fc_vipgenes_covid = all_samples_l2fc 
ann_vaccines_samples <- read_csv(here("Annotations and metadata", "ann_vaccines_samples.csv"))

df_samples_l2fc_vipgenes_covid


# Input data ----
filename = "ImmuneGO_BTM_Genes_Disease_Vac_L2FCvalues_"
df = df_samples_l2fc_vipgenes_covid

## Vaccination types
outcomevar = "disease_vac"

ann_vaccines_samples_filtered = ann_vaccines_samples %>% 
  mutate(outcomevar = disease_vac) %>% 
  filter(type != "H")

#How many samples by level?
count_levels = ann_vaccines_samples_filtered %>% 
  dplyr::count(outcomevar)

count_levels

#How many levels?
levels = nrow(count_levels)
order_levels = c("I", #TRUE
                 "V") #FALSE
```

### Build model

Step_corr:
- threshold = 0.9 -> From 963 genes to 894


```{r fig.height=10, fig.width=10}
# Prepare data ------

df_input <- df %>%
  inner_join(ImmuneGO_BTM_grouped_genes %>% select(genes) %>% distinct(), by = "genes") %>%
  select(genes, sample, log2fc) %>% 
  pivot_wider(names_from = genes,
              values_from = log2fc) %>% 
  inner_join(ann_vaccines_samples_filtered %>% filter(day != 0) %>% 
               select(sample, outcomevar = disease_vac), #Outcome
             by = "sample") %>%
  select(sample,
         outcomevar,
         everything()) %>%
  column_to_rownames("sample") %>%
  mutate_if(is.character, factor) %>% 
  mutate(outcomevar = fct_relevel(outcomevar, order_levels)) %>% 
  replace(is.na(.), 0)

# glimpse(df_input)

# Split data
set.seed(123)  
split <- initial_split(df_input, prop = 0.6, strata = outcomevar)
trees_train <- training(split)
trees_test <- testing(split)

# Create recipe
tree_rec <- recipe(outcomevar ~ ., data = trees_train) %>% #OUTCOME
  step_dummy(all_nominal(), -all_outcomes()) %>%
  step_corr(all_numeric_predictors(), threshold = 0.9) %>% 
  step_upsample(outcomevar) #Upsample unbalanced data

tree_prep <- prep(tree_rec)

juiced <- juice(tree_prep)

# Set random forest hyper parameters
tune_spec <- rand_forest(
  mtry = tune(),
  trees = 2000,
  min_n = tune()
) %>%
  set_mode("classification") %>% 
  set_engine("ranger")

# Create workflow
tune_wf <- workflow() %>%
  add_recipe(tree_rec) %>%
  add_model(tune_spec)

#Cross validation
set.seed(234)
trees_folds <- vfold_cv(trees_train)

#Define performance metrics
mer_metrics <- metric_set(yardstick::recall,
                          yardstick::specificity,
                          yardstick::accuracy,
                          yardstick::precision,
                          yardstick::roc_auc,
                          yardstick::pr_auc)

# Parallel computing 
doParallel::registerDoParallel()

set.seed(345)
tune_res <- tune_grid(
  tune_wf,
  metrics = mer_metrics,
  resamples = trees_folds,
  grid = 20
)
```

Assess metrics

```{r fig.height=10, fig.width=10}
# Assess data ----

#Metricss
metrics_tuneres = tune_res %>%
  collect_metrics() %>%
  select(mean, min_n, mtry, .metric) %>%
  pivot_longer(min_n:mtry,
    values_to = "value",
    names_to = "parameter") %>%
  ggplot()+
  aes(x = value, y = mean, colour = .metric, label = value) +
  geom_line() +
  geom_label() +
  scale_color_npg()+
  theme_few() +
  theme(axis.text.x = element_text(color = "black", size = 12)) +
  facet_grid(vars(.metric), vars(parameter), scales = "free")

metrics_tuneres

metrics_tuneres %>% ggsave(file = here("Figures",
                                       paste0(filename, 
                                              outcomevar,
                                              "_tuneres_metrics.png")),
                                        width = 10,
                                        height = 10)

min_n_selec = 6 #Observations by node)
mtry_selec = 23 #Features/genes by tree)
```

Hyperparameter tuning

```{r}
# Tune hyperparameters ----
rf_grid <- grid_regular(
  min_n(range = c(6, 6)),
  mtry(range = c(23, 23)),
  levels = levels
)

set.seed(456)
regular_res <- tune_grid(
  tune_wf,
  metrics = mer_metrics,
  resamples = trees_folds,
  grid = rf_grid
)

# Assess performance -----
regular_res %>%
  collect_metrics()

regular_res %>%
  collect_metrics() %>% 
  write_csv(file = here("Machine learning", paste("metrics_finalmodel_", filename, outcomevar, ".csv")))

#Metrics
regular_res_metrics = regular_res %>%
  collect_metrics() %>%
  select(mean, min_n, mtry, .metric) %>%
  pivot_longer(min_n:mtry,
    values_to = "value",
    names_to = "parameter") %>%
  ggplot()+
  aes(x = value, y = mean, colour = .metric, label = value) +
  geom_line() +
  geom_label(size = 3) +
  scale_color_npg()+
  theme_minimal() +
  facet_grid(vars(.metric), vars(parameter), scales = "free")

regular_res_metrics

# ggsave(regular_res_metrics, file = here("Figures", 
#                                         paste(filename, outcomevar, "metrics", today(), ".png")), 
#        width = 10, height = 10)

# Select best model -----
best_auc <- select_best(regular_res, metric = "accuracy")
final_rf <- finalize_model(
  tune_spec,
  best_auc
)
```

Important features

```{r fig.height=3, fig.width=5}
final = final_rf %>%
  set_engine("ranger", importance = "permutation") %>%
  fit(outcomevar ~ .,
    data = juice(tree_prep)) %>% 
  vip()

genes_rf = final$data
genes_rf = genes_rf %>% select(genes = Variable, importance = Importance)
genes_rf %>% write_csv(file = here("Figures", paste0("vipgenes_tidymodels_", filename, outcomevar, ".csv")))

#Importance plot
importance_plot = ggplot(genes_rf) +
  aes(x = reorder(genes, importance),
    y = importance,
    fill = importance,
      weight = importance) +
  geom_col() +
  theme_few() +
  xlab("Genes") + 
  ylab("importance") +
  ylim(0, 0.035) +
  geom_text(aes(label = round(importance, 4), y = importance),
            hjust = -0.2, 
            size = 4,
            angle = 0,
            color = "black")+
  scale_fill_gradient(low = "#4DBBD5FF", high = "#DC0000FF") +
  theme(legend.position = "right",
        plot.subtitle = element_text(hjust = 0),
        axis.text.x = element_text(angle = 0, vjust = 0.5, hjust=0.5, size = 12, color = "black"),
        axis.text.y = element_text(size = 12, angle = 0, color = "black"),
        panel.grid.minor.y = element_blank(),
        legend.text = element_text(size=12),
        legend.title = element_text(size = 12, face = "bold")) +
    coord_flip() +
  labs(title = "VIP genes",
       subtitle = "Random forest classification model for \nVaccination vs. Infection",
       x = "",
       y = "Importance",
       color = "Importance")

ggsave(importance_plot,
        file = here("Figures",
                    paste0("vipgenes_tidymodels_", filename, outcomevar,  ".png")),
        width = 5,
        height = 3)

importance_plot

```

Train final model with tuned hyperparameters

```{r}
# Train final model with tuned hyperparameters
final_wf <- workflow() %>%
  add_recipe(tree_rec) %>%
  add_model(final_rf)

final_res <- final_wf %>%
  last_fit(split)

final_res_preds = final_res %>%
  collect_predictions() %>%
  mutate(correct = case_when(
    outcomevar == .pred_class ~ "Correct",
    TRUE ~ "Incorrect"
  )) %>%
  bind_cols(trees_test)


#Confusion matrix
res_conf_mat = conf_mat(final_res_preds, truth = outcomevar...6,
         estimate = .pred_class)

#Accuracy
res_acc = accuracy(final_res_preds, truth = outcomevar...6,
         estimate = .pred_class)

#Specificity
res_spec = spec(final_res_preds, truth = outcomevar...6,
         estimate = .pred_class)

#Precision
res_prec = precision(final_res_preds, truth = outcomevar...6,
         estimate = .pred_class)

#Recall
res_recall = recall(final_res_preds, truth = outcomevar...6,
         estimate = .pred_class)
#Precision recall AUC
res_pr_auc = pr_auc(final_res_preds, outcomevar...6, .pred_I)


#Unite results
final_res_metrics = bind_rows(res_acc,
                         res_recall,
                         res_prec,
                         res_pr_auc,
                         res_spec) %>% 
  select(-".estimator")


final_res_metrics_conf = list(res_conf_mat, 
                         final_res_metrics,
                         final_res_preds)

final_res_metrics_conf

final_res_metrics_conf %>% 
  saveRDS(file = here("Machine learning", paste0("final_res_metrics", filename, outcomevar, ".rds")))


#Plot
final_model_metrics = final_res_metrics %>% 
  rename(metric = .metric,
         estimate = .estimate) %>% 
  mutate(metric = fct_relevel(metric, final_res_metrics$.metric)) %>% 
  ggplot() +
  aes(x = estimate,
    y = fct_rev(metric),
    fill = estimate,
    weight = estimate) +
  geom_col() +
  theme_few() +
  geom_text(aes(label = round(estimate, 3), y = metric),
            hjust = -0.2, 
            size = 4,
            angle = 0,
            color = "black") +
  xlim(0, 1.2) +
  scale_fill_gradient(low = "#4DBBD5FF", high = "#DC0000FF") +
  theme(legend.position = "right",
        plot.subtitle = element_text(hjust = 0),
        axis.text.x = element_text(angle = 0, vjust = 0.5, hjust=0.5, size = 12, color = "black"),
        axis.text.y = element_text(size = 12, angle = 0, color = "black"),
        panel.grid.minor.y = element_blank(),
        legend.text = element_text(size=12),
        legend.title = element_text(size = 12, face = "bold")) +
  labs(title = "Model assessment",
       subtitle = "Random forest classification model for \nVaccination vs. Infection",
       x = "Estimate",
       y = "",
       fill = "Estimate")

ggsave(final_model_metrics,
        file = here("Figures",
                    paste0("final_model_metrics_", filename, outcomevar,  ".png")),
        width = 5,
        height = 3)


#Which are the incorrect?
final_res_preds %>% rownames_to_column("sample") %>%  
  inner_join(ann_vaccines_samples %>% 
               select(condition, sample), by = "sample") %>% 
  filter(correct == "Incorrect") %>% 
  select(sample, condition, everything())
```




Precision recall curve
```{r}
# Precision recall
pr_curve_res = pr_curve(final_res_preds, outcomevar...6, .pred_I) %>% 
  ggplot(aes(x = recall, y = precision)) +
  geom_path(color = "#DC0000FF",
            size = 2) +
  coord_equal() +
  theme_few() +
  labs(title = "Precision-Recall curve",
       subtitle = "Random forest classification \nmodel for Infection vs. Vaccination",
       x = "Recall",
       y = "Precision") 


pr_curve_res %>% 
  ggsave(file = here("Figures",
                    paste0("PR_Curve_tidymodels_", filename, outcomevar,  ".png")),
         width = 4,
         height = 3)

pr_curve_res

```

Multi ROC
```{r fig.height=5, fig.width=5}
roc_plot = final_res_preds %>% 
  roc_curve(outcomevar...9, .pred_I) %>%
  ggplot(aes(1 - specificity, sensitivity)) +
  geom_abline(lty = 2, color = "gray80", size = 0.5) +
  geom_path(show.legend = FALSE, alpha = 0.6, size = 1.2) +
  theme_few() +
    theme(axis.text = element_text(size = 6, color = "black")) +
  labs(title ="Receiver-operator curve",
       subtitle = "Multiclass classification, Randomforest",
       x = "1-Specificity",
       y = "Sensitivity") +
  scale_colour_manual(values = c(ann_vaxsig_colors$type[names(ann_vaxsig_colors$type) != "H"], 
                                 H = "black"))
```

Confusion matrix
```{r fig.height=5, fig.width=5}
conf_matrix_heat = final_res_preds %>% 
  conf_mat(outcomevar...9, .pred_class) %>%
  autoplot(type = "heatmap") +
  scale_fill_gradient(low = "white", high = "#DC0000FF") +
  labs(title ="Confusion matrix",
       subtitle = "Multiclass classification, Randomforest",
       x = "Truth",
       y = "Prediction")

patch = conf_matrix_heat / 
  roc_plot

patch %>% 
  ggsave(file = here("Figures",
                    paste0("ROC_ConfusionMatrix_tidymodels_", filename, outcomevar,  ".png")),
         width = 5,
         height = 5)
  
```



```{r}
#Matrix counts of VIP genes
all_matrices_counts_genesRF = df %>% 
  inner_join(genes_rf, by = "genes")
```


### Relative effects

Function
```{r fig.height=5, fig.width=7}
REffect <- function(data, dep_vars, indep_var, n_boot = 1000, 
                    plot = TRUE, plot_colors = c("red", "darkblue"), 
                    plot_labels = list(y = "Relative effect", x = "Genes"), 
                    plot_theme = theme_few(), plot_legend_position = "top", 
                    plot_legend_labels = c("Vaccination", "Infection")) {
  
  library(npmv)
  library(ggplot2)
  library(reshape2)
  
  if(!all(data[[indep_var]] %in% c(0, 1))) {
    stop("A variável independente precisa conter apenas 0 e 1 (dois grupos).")
  }
  
  n_aab <- nrow(data)
  list_boot_aab <- list()
  
  # Loop de Bootstrap
  for (i in 0:n_boot) {
    if (i == 0) {
      df_sample <- 1:n_aab
    } else {
      df_sample <- sample(1:n_aab, n_aab, replace = TRUE)
    }
    
    formula_str <- paste(dep_vars, collapse = " | ")
    formula <- as.formula(paste(formula_str, "~", indep_var))
    
    manova_aab <- nonpartest(formula, data = data[df_sample, ], permtest = FALSE, plots = FALSE,
                             permreps = 1000, tests = c(1, 0, 0, 0))
    
    if (i == 0) obs_aab <- manova_aab
    
    list_boot_aab[[i + 1]] <- manova_aab
    
    cat("ANOVA/ boot/ aab/ i = ", i, "\n")
  }
  
  obs_manova <- list_boot_aab[[1]]$twogroupreleffects
  row_manova <- rownames(obs_manova)
  
  releffects <- lapply(list_boot_aab[-1], "[[", 2)
  
  df_perc_boot <- lapply(seq_len(nrow(obs_manova)), function(i) {
    q <- sapply(seq_len(ncol(obs_manova)), function(j) {
      v <- sapply(releffects, function(r) {
        row <- rownames(r)
        if(row[1] != row_manova[1]) r <- r[2:1,]
        r[i, j]
      })
      quantile(v, probs = c(0.025, 0.975))
    })
    colnames(q) <- colnames(obs_manova)
    df <- data.frame(Var2 = colnames(q), t(q),
                     group = ifelse(i == 1, row_manova[1], row_manova[2]))
    df$Var2 <- factor(df$Var2, levels = sort(unique(df$Var2)), ordered = TRUE)
    return(df)
  })
  
  names(df_perc_boot) <- row_manova
  
  df <- reshape2::melt(
    cbind(obs_manova, group = row_manova),
    id.vars = "group"
  )
  df$Var1 <- "point_est"
  df <- df[, c(4, 2, 3, 1)]
  colnames(df)[2] <- "Var2"
  
  df <- dplyr::arrange(df, group, desc(value), Var2)
  df$Var2 <- as.character(df$Var2)
  df$Var2 <- factor(df$Var2, levels = unique(df$Var2), ordered = TRUE)
  
  for (k in seq_len(2)) {
    df_perc_boot[[k]]$Var2 <- factor(df_perc_boot[[k]]$Var2, levels = levels(df$Var2), ordered = TRUE)
  }
  
  if (plot) {
    p <- ggplot(data = df, aes(x = Var2, y = value, group = group, color = group)) +
      geom_ribbon(inherit.aes = FALSE, data = df_perc_boot[[1]],
                  aes(x = Var2, ymin = X2.5., ymax = X97.5., group = group), fill = plot_colors[1],
                  alpha = 0.3) +
      geom_ribbon(inherit.aes = FALSE, data = df_perc_boot[[2]],
                  aes(x = Var2, ymin = X2.5., ymax = X97.5., group = group), fill = plot_colors[2],
                  alpha = 0.3) +
      geom_line() +
      geom_point(aes(size = value)) +
      scale_colour_manual("", values = c("1" = plot_colors[2], "0" = plot_colors[1]),
                          labels = plot_legend_labels) +
      plot_theme +
      theme(
        legend.text = element_text(size = 12),
        axis.text.x = element_text(size = 12, angle = 90),
        axis.text.y = element_text(size = 12),
        axis.line.y = element_blank(),
        axis.title = element_text(size = 15),
        panel.grid.major = element_line(),
        legend.position = plot_legend_position,
        legend.box.background = element_rect()
      ) +
      labs(y = plot_labels$y, x = plot_labels$x)
    
    print(p)
  }
  
  return(list(obs_manova = obs_manova, df_perc_boot = df_perc_boot, plot = p))
}
```

Plot
```{r fig.height=5, fig.width=7, message=FALSE, warning=FALSE}
head(all_matrices_counts_genesRF)

covid_vac_releff = all_matrices_counts_genesRF %>% 
  select(genes, sample, log2fc) %>% 
  pivot_wider(names_from = "genes",
              values_from = "log2fc") %>% 
  inner_join(ann_vaccines_samples %>% 
               select(sample, disease_vac), by = "sample") %>% 
  mutate(disease_vac = if_else(disease_vac == "I", 1, 0))


releffect_disease_vac = REffect(covid_vac_releff, dep_vars = colnames(covid_vac_releff %>% 
                                              select(-disease_vac, -sample)),
        indep_var = "disease_vac", plot_colors = c("red", "#4DBBD5FF"), 
        n_boot=1000)


plot = releffect_disease_vac$plot +
  scale_colour_manual(values = c("1" = "#DC0000FF", 
                                 "0" = "#4DBBD5FF"),
                      labels = plot_legend_labels) +
  geom_ribbon(inherit.aes = FALSE, data = releffect_disease_vac$df_perc_boot[[1]],
                  aes(x = Var2, ymin = X2.5., ymax = X97.5., group = group), fill = "#4DBBD5FF",
                  alpha = 0.3) +
      geom_ribbon(inherit.aes = FALSE, data = releffect_disease_vac$df_perc_boot[[2]],
                  aes(x = Var2, ymin = X2.5., ymax = X97.5., group = group), fill = "#DC0000FF",
                  alpha = 0.3) +
  theme(
        legend.text = element_text(size = 12),
        axis.text.x = element_text(size = 12, angle = 45, hjust = 1, vjust = 1, color = "black"),
        axis.text.y = element_text(size = 12, color = "black"),
        axis.line.y = element_blank(),
        axis.title = element_text(size = 15),
        panel.grid.major = element_line(),
        legend.position = "top",
        plot.title = element_text(hjust = 0.5, size = 20),
        plot.subtitle = element_text(hjust = 0.5, size = 15),
        legend.box.background = element_blank()
      ) +
      labs(y = "Relative effect", x = "Genes",
           title = "Relative effects of VIP genes",
           subtitle = "Infection vs. Vaccination",
           color = "") +
  ylim(0, 1)

plot

```

### Butterfly plot
```{r}
##### Butterfly plot
all_degs_p_05_vac_infected = read_csv("DGE analysis/all_degs_p_05_vac_infected_19_12_23.csv")
df_aggregated = 

df_wide = all_degs_p_05_vac_infected |>
  inner_join(ImmuneGO_Genes %>% select(genes) %>% distinct(), by = "genes") %>% 
  group_by(genes, disease_vac) |>
  summarise(avg_log2FC = mean(log2fold_change), .groups = 'drop') |>
  pivot_wider(names_from = disease_vac, values_from = avg_log2FC) |>
  mutate(Regulation = case_when(V >= 0 & I <= 0 ~ "no",
                               V >= 0 & I >= 0 ~ "up",
                               V <= 0 & I <= 0 ~ "down",
                               TRUE ~ "ns"))   

butterfly_plot = ggplot(df_wide, aes(x = V, y = I, color = Regulation)) +
  geom_point(aes(size = abs(V),
                 label = genes)) +
  geom_label_repel(data = df_wide %>% filter(V > 2.5 & I > 0 | 
                                               V < -1 & I < -2), aes(label = genes)) +
  theme_minimal() +
  scale_color_manual(values = c("no" = "gray",
                                "ns" = "gray",
                                "down" = "royalblue",
                                "up" = "red")) +
  theme_bw() +
  labs(x = "Vaccination",
       y = "Infection",
       size = "Abs(L2FC)",
       title = "Mean Log2FC")
butterfly_plot
ggplotly(butterfly_plot)

```


### Prediction samples heatmap

```{r fig.height=10, fig.width=10}
pred_genes_counts = final_res_preds %>% 
  select(pred_class = .pred_class,
         pred_I = .pred_I,
         pred_V = .pred_V,
         truth = outcomevar...6,
         correct, 
         ACE:ZNF683) %>% 
  pivot_longer(cols = ACE:ZNF683, 
               names_to = "genes",
               values_to = "values") %>% 
  inner_join(genes_rf, by = "genes") %>% 
  inner_join(normalized_counts_long, by = "genes")
  
pred_genes_counts %>% 
  filter(disease_vac != "H",
         disease_vac != "Healthy") %>% 
  ggplot() +
  aes(y = sample,
      x = genes,
      fill = expression) +
  geom_tile() +
  facet_wrap(~disease_vac~type,scales = "free_y")
```


###  Heatmap: Gene L2FC per condition

```{r}
# Required files
ImmuneGO_Annotated_Genes = readRDS(here("VaxGO gene sets","ImmuneGO_Annotated_Genes_2024-03-26.rds"))
df = normalized_counts
ann_vaccines <- read_csv(here("Annotations and metadata", "ann_vaccines_conditions.csv"))
ann_vaccines_samples <- read_csv(here("Annotations and metadata", "ann_vaccines_samples.csv"))


#Inputs
ImmuneGO_Annotated_Genes_all = ImmuneGO_Annotated_Genes %>% 
  filter(!process %in% c("TCR REPERTOIRE", "BCR REPERTOIRE"))

ImmuneGO_Genes = all_degs_p_05_vac_infected %>% 
  inner_join(ImmuneGO_Annotated_Genes %>% 
               select(genes, process) %>% 
               distinct() %>% 
               select(genes),
               by = "genes") %>% 
  distinct()

# Matrix
l2fc_imps = genes_rf %>% 
  select(genes, importance) %>% 
  inner_join(ImmuneGO_Genes, by = "genes") %>%
  select(genes, condition, log2fold_change) %>% 
  pivot_wider(names_from = condition, values_from = log2fold_change) %>% 
  column_to_rownames("genes") %>% 
  as.matrix() %>% 
  replace(is.na(.), 0)

```

```{r}

#Which VIP genes?

vip_genes = genes_rf
outcomevar = "_Type_"

rf_genes = vip_genes %>% 
  clean_names() %>% 
  select(genes, importance)

ImmuneGO_Genes = all_degs_p_05_vac_infected %>% 
  inner_join(rf_genes %>% 
               select(genes) %>% 
               distinct() %>% 
               select(genes),
               by = "genes") %>% 
  distinct()

####### INPUT
data_2 = ImmuneGO_Genes
filename_2 = paste0("RandomForest_ImmuneGO_Genes", outcomevar, today())

######## Matrix
matrix = data_2 %>% 
  select(genes, condition, log2fold_change) %>%
  pivot_wider(names_from = "condition", 
              values_from = "log2fold_change", 
              values_fn = mean) %>% 
  replace(is.na(.), 0) %>%
  arrange(genes) %>% 
  column_to_rownames(var="genes") %>% 
  as.matrix()

######## Annotations
ann_vaccines_covid_other_2 = data_2 %>% 
  select(condition) %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  distinct() %>% 
  select(condition, disease_vac, type, week, day) %>% 
  arrange(week, day) %>% 
  mutate(week = fct_relevel(as.factor(week), sort(levels(week))),
         day = fct_relevel(as.factor(day), sort(levels(day))))

#Columns
ann_cols_heatmap = matrix %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  inner_join(ann_vaccines_covid_other_2, by = "condition") %>% 
  distinct() %>% 
  arrange(condition) %>% 
  column_to_rownames("condition")

matrix_data = matrix %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  inner_join(ann_cols_heatmap %>% rownames_to_column("condition"), by = "condition") %>% 
  select(!c(disease_vac:day)) %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>%
  as.matrix() 

#Rows Immune ----
ann_rows_immune = matrix_data %>%
  as.data.frame() %>%
  rownames_to_column("genes") %>%
  left_join(ImmuneGO_Annotated_Genes %>% filter(go_term == "Manual"), by = "genes") %>%
  select(genes, process) %>%
  distinct() %>%
  group_by(genes) %>%
  arrange(genes, factor(process, levels = c("INNATE IMMUNE SYSTEM", "ADAPTIVE IMMUNE SYSTEM")), .by_group = TRUE) %>%
  mutate(i_process = row_number()) %>%
  pivot_wider(., names_from = "i_process", values_from = "process") %>%
  column_to_rownames("genes") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("index") %>%
  mutate(index = as.numeric(index)) %>%
  arrange(desc(index)) %>%
  column_to_rownames("index") %>%
  t() %>%
  as.data.frame() %>%
  replace(is.na(.), ".") %>% 
  mutate(`1` = if_else(`1` == ".", "INNATE IMMUNE SYSTEM", `1`)) 

#Verificar dimensões do dataset
dim(matrix)
dim(matrix_data)
dim(ann_cols_heatmap)
dim(ann_rows_immune)
```

```{r}
################# Immune 
#Heatmap annotation
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "right")

row_ha = HeatmapAnnotation(df = ann_rows_immune,
                       col = col_process_immune,
                       which= "row",
                       show_annotation_name = F,
                       show_legend = F,
                       simple_anno_size = unit(2, "mm"))

# Values colors
col_fun <- colorRamp2(c(-2, 0, 2), c("#9bc1bc", "white", "#ed6a5a"))

file = filename_2
mat = matrix_data
width_image = 30
height_image = 15
width = unit(15, "cm")
height = unit(5, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)
rect_gp = gpar(col = "gray80", lwd = 0.5)

fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap.png")

png(file = fileimageheatmap_grouped, 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "L2FC",
                        col = col_fun, #color
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(0, "cm"),
                        row_split = NULL,
                        column_split = NULL,
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(2, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height)

heatmap_plot = draw(heatmap_plot, 
     annotation_legend_side = "left", 
     heatmap_legend_side = "left")

dev.off()

```

### Boxplot: VIP gene expression by condition
Import tables
```{r fig.height=10, fig.width=10}
genes_rf = read_csv("Machine learning/vipgenes_tidymodels_ImmuneGO_Genes_Type_including_Infection_type2024-06-27.csv") 
ann_vaccines_conditions = read_csv(here("Annotations and metadata", "ann_vaccines_conditions.csv"))

all_matrices_counts_long_annotated <- readRDS("DGE analysis", "all_matrices_counts_long_annotated.rds"))

all_samples_l2fc = read_rds(here("DGE analysis", "all_samples_l2fc.rds"))

all_matrices_counts_genesRF = all_matrices_counts_long_annotated %>% 
  filter(genes %in% genes_rf$Variable) 

all_matrices_counts_genesRF %>% 
  saveRDS(here("Machine learning", "all_matrices_counts_genesRF.rds"))
```


```{r fig.height=10, fig.width=10}
df = all_samples_l2fc
table = df %>% 
  select(-condition) %>% 
  filter(genes == "IL10") %>% 
  #filter(genes %in% genes_rf$Variable) %>% 
  inner_join(ann_vaccines_samples, by = "sample") %>% 
  mutate(disease_vac = if_else(disease_vac == "Healthy", "H", disease_vac)) %>% 
  mutate(disease_vac = fct_relevel(disease_vac, "H", "I", "V"),
           type = fct_relevel(type, "I", "RNA", "VV", "IN", "SU")) %>% 
  arrange(disease_vac,type, condition) %>% 
  inner_join(ann_vaccines_conditions %>% select(condition, samples), by = "condition") %>% 
  mutate(condition = factor(condition, levels = c(unique(condition[order(condition)]))),
         condition = paste0(condition, " (n=", samples, ")")) 
  # mutate(genes = fct_relevel(genes, genes_rf$Variable))

table_ready = table %>% 
  mutate(condition = fct_relevel(condition, 
                                 table %>% 
                                   filter(str_detect(condition, "^H \\(D0\\)")) %>% 
                                   select(condition) %>% 
                                   distinct() %>% 
                                   pull(condition), 
                                 table %>% 
                                   filter(str_detect(condition, "^I")) %>% 
                                   filter(!str_detect(condition, "^I-I")) %>% 
                                   select(condition) %>% 
                                   arrange(condition) %>% 
                                   distinct() %>% 
                                   pull(condition),
                                 table %>% 
                                   filter(disease_vac == "I") %>% 
                                   filter(str_detect(condition, "^BNT-I")) %>% 
                                    filter(!str_detect(condition, "^BNT-I \\(D\\d+\\)")) %>% 
                                    arrange(day, condition) %>%
                                   select(condition) %>% 
                                    distinct() %>% 
                                    pull(condition),
                                 table %>% 
                                    filter(str_detect(condition, "^BNT-I \\(D\\d+\\)")) %>% 
                                    arrange(day, condition) %>%
                                   select(condition) %>% 
                                    distinct() %>% 
                                    pull(condition),
                                 table %>% 
                                   filter(disease_vac == "V",
                                          str_detect(condition, "^BNT \\(V1, D\\d+\\)")) %>% 
                                   arrange(day, condition) %>% 
                                   select(condition) %>% 
                                   distinct() %>% 
                                   pull(condition),
                                 table %>% 
                                   filter(disease_vac == "V",
                                          !str_detect(condition, "^BNT \\(V1, D\\d+\\)"),
                                          !str_detect(condition, "BBIBP"),
                                          !str_detect(condition, "ZF2001")) %>% 
                                   arrange(condition) %>% 
                                   select(condition) %>% 
                                   distinct() %>% 
                                   pull(condition),
                                 table %>% 
                                   filter(str_detect(condition, "BBIBP")) %>% 
                                   select(condition) %>% 
                                   arrange(condition) %>% 
                                   distinct() %>% 
                                   pull(condition), 
                                 table %>% 
                                   filter(str_detect(condition, "ZF2001")) %>% 
                                   select(condition) %>% 
                                   arrange(condition) %>% 
                                   distinct() %>% 
                                   pull(condition))) %>%
  mutate(type = if_else(type == "H", "I", type),
         disease_vac = if_else(disease_vac == "H", "I", disease_vac))
```


```{r fig.height=10, fig.width=10}
#Base value -----
median_h = table_ready %>% 
  select(genes, sample, log2fc, disease_vac) %>% 
  filter(disease_vac == "H") %>% 
  distinct() %>% 
  group_by(genes, disease_vac) %>% 
  summarize(median = median(counts), .groups = 'drop')

median_i = table_ready %>% 
  select(genes, sample, counts, disease_vac) %>% 
  filter(disease_vac == "H") %>% 
  distinct() %>% 
  group_by(genes, disease_vac) %>% 
  summarize(median = median(counts), .groups = 'drop') %>% 
  mutate(disease_vac = "I")

median_v = table_ready %>% 
  select(genes, sample, counts, disease_vac) %>% 
  filter(disease_vac == "H") %>% 
  distinct() %>% 
  group_by(genes, disease_vac) %>% 
  summarize(median = median(counts), .groups = 'drop') %>% 
  mutate(disease_vac = "V")

median = bind_rows(median_i, median_v)
```


```{r fig.height=5, fig.width=12, message=TRUE, warning=FALSE}
#Plots -----

disease_vac_boxplot_samples_VIP = table_ready %>% 
  # filter(genes == "GZMM") %>% 
  ggplot() +
  aes(x = condition, y = log2fc, fill = if_else(day == 0, "Day 0", type)) +  
  geom_hline(yintercept = 0, linetype = 2) +
  scale_y_log10()+
  geom_jitter(aes(color = if_else(day == 0, "Day 0", type)),
              alpha = 0.3, size = 0.1) +
  geom_boxplot(outliers = F) +
  theme_few() +
  scale_fill_manual(values = c("Day 0" = "white", 
                               ann_vaxsig_colors$type)) +
  scale_color_manual(values = c("Day 0" = "gray80", 
                               ann_vaxsig_colors$type)) +
  theme(axis.text.x = element_text(angle = 45,
                                   hjust = 1,
                                   size = 8,
                                   vjust = 1),
        plot.title = element_text(hjust = 0, face = "bold"),
        plot.margin = margin(0, 0, 0, 2, "cm"),
        strip.text.y = element_text(angle = 0),
        panel.grid.major = element_line(size = 0.05),
        panel.grid.minor = element_line(size = 0.05)) +
  labs(y = "Log2 fold-change",
       title = "VIP genes",
       subtitle = "Random forest classification model for Infection vs. Vaccination \n
       Log2FC calculated comparing each sample with their control (D0)",
       x="",
       fill = "Type") +
  guides(color = "none") +
  facet_grid(~genes~disease_vac, scales = "free")

disease_vac_boxplot_samples_VIP %>% 
  ggsave(file = here("Figures" , "ML_infec_vs_vac_boxplot_vipgenes.png"), 
         width = 10,
         height = 5)
```


```{r fig.height=10, fig.width=10}
#Type
vip_genes_boxplot_type = table %>% 
  filter(disease_vac != "H") %>% 
  ggplot() +
  aes(x = type, y = counts, fill = week_fac) +
  geom_boxplot() +
  geom_hline(data = median, aes(yintercept = median), linetype = "dashed", alpha = 0.5) +
  scale_fill_viridis_d(option = "cividis", direction = -1) +
  scale_y_continuous(trans = "log10") +
  labs(
    x = "Type",
    y = "Normalized counts",
    title = "Gene expression of the 5 most important genes",
    fill = "Week"
  ) +
  theme_linedraw() +
  theme(axis.text.y = element_text(hjust = 0.1),
        legend.position = "top") +
  facet_grid(~genes~disease_vac, scales = "free")

vip_genes_boxplot_type 

vip_genes_boxplot_type %>% 
  ggsave(file = here("Figures", paste0("vip_genes_boxplot_type", filename, outcomevar, today(), ".png")), 
                                  width = 10, height = 15)

#Sex
vip_genes_boxplot_type_sex = table %>% 
  mutate(type_sex = paste0(type, "_", sex),
         type_sex = fct_relevel(type_sex, "IN_M/F", "SU_M/F")) %>% 
  filter(disease_vac != "H",
         !type %in% c("IN", "SU")) %>% 
  ggplot() +
  aes(x = type_sex, y = counts, fill = week_fac) +
  geom_boxplot() +
  geom_hline(data = median, aes(yintercept = median), linetype = "dashed", alpha = 0.5) +
  scale_fill_viridis_d(option = "cividis", direction = -1) +
  scale_y_continuous(trans = "log10") +
  labs(
    x = "Type",
    y = "Normalized counts",
    title = "Gene expression of the 5 most important genes",
    fill = "Week"
  ) +
  theme_linedraw() +
  theme(axis.text.y = element_text(hjust = 0.1),
        axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) +
  facet_grid(~genes~disease_vac, scales = "free")

vip_genes_boxplot_type_sex 


```


### PCA with VIP genes
**Input**

```{r}
#INPUT
data_genes = all_matrices_counts_genesRF 
data_annotation = ann_vaccines_samples
filename = "RandomforestGenes_COVID_Infec_Vac"
```

```{r}
#Convert long to wide table format.
#Dataframe with sample annotation

ann_vaccines_pca_matrix = data_genes %>% 
  select(sample, genes, counts) %>% 
  pivot_wider(., names_from = "sample", 
                     values_from = "counts") %>% 
  replace(is.na(.), 0) %>%
  column_to_rownames("genes") %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column("sample") %>% 
  inner_join(data_annotation %>% 
               select(sample, condition, disease_vac, gse_id, vaccine, day, week, dose, infection, type), by = "sample") 

# Convert dataframe to matrix without annotation columns
ann_vaccines_pca_matrix_ready = ann_vaccines_pca_matrix %>% 
  column_to_rownames("sample") %>% 
  select(!condition:type) %>% 
  as.matrix()

# Delete columns with near 0 variance
nearZeroVarCols <- nearZeroVar(ann_vaccines_pca_matrix_ready, saveMetrics = TRUE)
ann_vaccines_pca_matrix_ready <- ann_vaccines_pca_matrix_ready[, !nearZeroVarCols$nzv]
pca_res <- prcomp(ann_vaccines_pca_matrix_ready, scale. = T)

# Correlation plot ----
corr_matrix = cor(ann_vaccines_pca_matrix_ready %>% scale()) 
var <- get_pca_var(pca_res)

# corrplot = ggcorrplot(corr_matrix, hc.order = TRUE) + 
#   theme(axis.text.x = element_text(angle = 90, size = 5), 
#         axis.text.y = element_text(size = 5))
# #Salvar
# ggsave(corrplot, file = paste0(filename, "_corrplot.png"), 
#        width = 20, #Grande 20, pequeno 10
#        height= 20) #Grande 20, pequeno 10
# print(corrplot)

#PCA ----
data.pca = prcomp(corr_matrix) #PCA
summary(corr_matrix) #Retornar PCs

#Scree plot ----
scree_plot = fviz_eig(data.pca, 
                      addlabels = TRUE,
                      ylim = c(0, 70)) +
  geom_col(color = "#00AFBB", fill = "#00AFBB") +
  theme_classic()

#Save
ggsave(scree_plot, file = here("Figures", paste0(filename, "_screeplot.png")), width = 10, height = 3) 
print(scree_plot)

```

Loadings plot

```{r}
# Loadings plot ----
options(ggrepel.max.overlaps = Inf)

circle_contrib = fviz_pca_var(pca_res, col.var = "cos2",
                              gradient.cols = c("black", "#DC0000FF"),
                              select.var= list(cos2 = 20), 
                              repel = T, 
                              labelsize = 4,
                              col.circle = NA) +
  xlab("") + 
  ylab("") +
  theme_minimal() +
theme(panel.grid = element_blank(),  # Remover linhas de grade
        axis.text = element_blank(),   # Remover rótulos de texto dos eixos
        axis.ticks = element_blank()) 

#Save
ggsave(circle_contrib, file = here("Figures", paste0(filename, "_circle_contrib_wide.png")), width = 10, height = 5)

circle_contrib
```

KNN

```{r}
#KNN classification -----
#Use the screeplot generated % of explained variance for each dimension
scree_plot #Dim 1 = 68.2%, Dim 2 = 15.9%


# PC1-PC2 ------
#Determine the number of clusters
pca_scores <- data.frame(pca_res$x[, 1:2])
fviz_nbclust(pca_scores,  
                     FUNcluster = kmeans,
                     method = "wss")

pcas = "PC1-PC2"

set.seed(666) # Set seed for randomization
cluster_model <- kmeans(pca_res$x[, 1:2], centers = 3)  #Set the number of clusters

ann_vaccines_pca_matrix$cluster <- as.factor(cluster_model$cluster)
```

Plot

```{r fig.height=7, fig.width=10, paged.print=FALSE}
# Condition with density plot ------
pca_plot_knn_cluster = autoplot(pca_res, 
        data = ann_vaccines_pca_matrix,
        colour = "gse_id")  +
  stat_ellipse(aes(color = cluster, fill = cluster), 
               geom = "polygon", 
               alpha = 0.2, 
               linetype = 1,
               size = 0.3,
               type = "t") +
  labs(title=paste0(filename, "_bycondition")) + 
  xlab("PC1 (68.2%)") +
  ylab("PC2 (15.9%)") +
    geom_hline(yintercept = 0, size = 0.3, color = "gray50", linetype = 4) +
  geom_vline(xintercept = 0, size = 0.3, color = "gray50", linetype = 4) +
  geom_point(aes(fill = type,
                 color = type),
             stroke = 0.2) +
   scale_color_manual(values = c("1" = "#8491B4FF", 
                                "2" = "#4DBBD5FF", 
                                "3" = "#DC0000FF",
                                ann_vaxsig_colors$type), 
                     guide = "none") +
  scale_fill_manual(values = c("1" = "#8491B4FF", 
                               "2" = "#4DBBD5FF", 
                               "3" = "#DC0000FF",
                               ann_vaxsig_colors$type),
                    guide = "none") + 
  guides(col="none") +
  theme_few() +
  theme(panel.grid.minor = element_blank(),
        text = element_text(color = "black"),
        panel.border = element_blank(),
        axis.line = element_line(colour = "black")) +
  geom_text_repel(aes(label = condition,
                      segment.colour="gray70"),
                  box.padding = 0.3,
                  size = 3) +
  xlim(-0.4, 0.6)
pca_plot_knn_cluster

ggsave(pca_plot_knn_cluster, 
       file = here( "Figures", paste0(filename, pcas, "_KNN_Clustered_labels.png")), width = 10, height = 7)
pca_plot_knn_cluster
# pca_plot_knn_cluster_marginal = ggMarginal(
#   pca_plot_knn_cluster,
#   type = "histogram",
#   groupFill = T,
#   groupColour = T,
#   xparams = list(binwidth = 0.1, size = 0.1),
#   yparams = list(binwidth = 0.1, size = 0.1)
# )
# 
# ggsave(
#   pca_plot_knn_cluster_marginal,
#   file = here(paste0(filename, pcas, "_KNN_Clustered_labels_Marginal.png"), "Figures"),
#   width = 8,
#   height = 5
# )


```


### Heatmap: Gene L2FC per condition

```{r}
# Required files
ImmuneGO_Annotated_Genes = readRDS(here("VaxGO gene sets","ImmuneGO_Annotated_Genes_2024-03-26.rds"))
all_degs_p_05_vac_infected = read_csv(here("DGE analysis", "all_degs_p_05_vac_infected_19_12_23.csv"))
ann_vaccines <- read_csv(here("Annotations and metadata", "ann_vaccines_conditions.csv"))
df = read_csv("Machine Learning/vipgenes_tidymodels_ImmuneGO_Genes_Type_including_Infection_disease_vac.csv")
#Inputs
ImmuneGO_Annotated_Genes_all = ImmuneGO_Annotated_Genes %>% 
  filter(!process %in% c("TCR REPERTOIRE", "BCR REPERTOIRE"))

ImmuneGO_Genes = all_degs_p_05_vac_infected %>% 
  filter(genes%in% df$genes) %>% 
  inner_join(ImmuneGO_Annotated_Genes %>% 
               select(genes, process) %>% 
               distinct() %>% 
               select(genes),
               by = "genes") %>% 
  distinct() %>% 
  select(genes, condition, log2fold_change) %>% 
  pivot_wider(names_from = condition, values_from = log2fold_change)

# Matrix
l2fc_imps = ImmuneGO_Genes %>% 
  column_to_rownames("genes") %>% 
  as.matrix() %>% 
  replace(is.na(.), 0)

outcomevar = "Infection_Vaccine"
####### INPUT
filename_2 = paste0("RandomForest_ImmuneGO_Genes_Conditions_", outcomevar)

######## Matrix
matrix = ImmuneGO_Genes

######## Annotations
ann_vaccines_covid_other_2 = ImmuneGO_Genes %>% 
  column_to_rownames("genes") %>% 
  colnames() %>% 
  as.data.frame() %>% 
  select(condition = ".") %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  distinct() %>% 
  select(condition, disease_vac, type, week, day) %>% 
  arrange(week, day) %>% 
  mutate(week = fct_relevel(as.factor(week), sort(levels(week))),
         day = fct_relevel(as.factor(day), sort(levels(day)))) 

```


```{r}
#Columns
ann_cols_heatmap = matrix %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  inner_join(ann_vaccines_covid_other_2, by = "condition") %>% 
  distinct() %>% 
  arrange(condition) %>% 
  column_to_rownames("condition")

matrix_data = matrix %>% 
  column_to_rownames("genes") %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  inner_join(ann_cols_heatmap %>% rownames_to_column("condition"), by = "condition") %>% 
  select(!c(disease_vac:day)) %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>%
  as.matrix()  %>% 
  replace(is.na(.), 0)

#Rows Immune ----
ann_rows_immune = matrix_data %>%
  as.data.frame() %>%
  rownames_to_column("genes") %>%
  left_join(ImmuneGO_Annotated_Genes %>% filter(go_term == "Manual"), by = "genes") %>%
  select(genes, process) %>%
  distinct() %>%
  group_by(genes) %>%
  arrange(genes, factor(process, levels = c("INNATE IMMUNE SYSTEM", "ADAPTIVE IMMUNE SYSTEM")), .by_group = TRUE) %>%
  mutate(i_process = row_number()) %>%
  pivot_wider(., names_from = "i_process", values_from = "process") %>%
  column_to_rownames("genes") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("index") %>%
  mutate(index = as.numeric(index)) %>%
  arrange(desc(index)) %>%
  column_to_rownames("index") %>%
  t() %>%
  as.data.frame() %>%
  replace(is.na(.), ".") %>% 
  mutate(`1` = if_else(`1` == ".", "INNATE IMMUNE SYSTEM", `1`)) %>% 
  rownames_to_column("genes") %>% 
  mutate(genes = rownames(matrix_data)) %>% 
  column_to_rownames("genes")

#Verificar dimensões do dataset
dim(matrix)
dim(matrix_data)
dim(ann_cols_heatmap)
dim(ann_rows_immune)
```

```{r}
################# Immune 
#Heatmap annotation
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "right")

row_ha = HeatmapAnnotation(df = ann_rows_immune,
                       col = col_process_immune,
                       which= "row",
                       show_annotation_name = F,
                       show_legend = F,
                       simple_anno_size = unit(1, "mm"))

# Values colors
col_fun <- colorRamp2(c(-2, 0, 2), c("#9bc1bc", "white", "#ed6a5a"))

file = filename_2
mat = matrix_data
width_image = 30
height_image = 15
width = unit(15, "cm")
height = unit(6, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)
rect_gp = gpar(col = "gray80", lwd = 0.5)

fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap.png")

png(file = here("Figures", fileimageheatmap_grouped), 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC",
                        col = col_fun, #color
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(0, "cm"),
                        row_split = NULL,
                        column_split = NULL,
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(2, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height)

heatmap_plot = draw(heatmap_plot, 
     annotation_legend_side = "left", 
     heatmap_legend_side = "bottom")

dev.off()

```



### Heatmap: Gene counts per sample
```{r}
# Required files
ImmuneGO_Annotated_Genes = readRDS(here("VaxGO gene sets","ImmuneGO_Annotated_Genes_2024-03-26.rds"))

genes_rf = read_csv("Machine learning/vipgenes_tidymodels_All_coviddisease_vac.csv")
df_samples_l2fc_vipgenes_covid = all_samples_l2fc %>% 
  filter(genes %in% genes_rf$genes)

ann_vaccines_samples <- read_csv(here("Annotations and metadata", "ann_vaccines_samples.csv"))


df_annotated = df_samples_l2fc_vipgenes_covid %>% 
  distinct() %>% 
  inner_join(ann_vaccines_samples %>% select(-condition, -participant_id), by = "sample") %>% 
  distinct() %>% 
  filter(condition != "BNT-MO (V1, D0)",
         condition != "BNT-MO (V1, D6)") %>% 
  mutate(disease_vac = if_else(disease_vac == "Healthy", "H", disease_vac)) %>% 
  mutate(disease_vac = fct_relevel(disease_vac, "H", "I", "V"),
           type = fct_relevel(type, "I", "RNA", "VV", "IN", "SU")) %>% 
  inner_join(ann_vaccines_conditions %>% select(condition, samples), by = "condition") %>% 
  mutate(condition = factor(condition, levels = c(unique(condition[order(condition)]))),
         condition = paste0(condition, " (n=", samples, ")"))
  

df_annotated_2 = df_annotated %>% 
  mutate(condition = fct_relevel(condition, 
                                 df_annotated %>% 
                                   filter(str_detect(condition, "^H \\(D0\\)")) %>% 
                                   select(condition) %>% 
                                   distinct() %>% 
                                   pull(condition), 
                                 df_annotated %>% 
                                   filter(str_detect(condition, "^I")) %>% 
                                   filter(!str_detect(condition, "^I-I")) %>% 
                                   select(condition) %>% 
                                   arrange(condition) %>% 
                                   distinct() %>% 
                                   pull(condition),
                                 df_annotated %>% 
                                   filter(disease_vac == "I") %>% 
                                   filter(str_detect(condition, "^BNT-I")) %>% 
                                    filter(!str_detect(condition, "^BNT-I \\(D\\d+\\)")) %>% 
                                    arrange(day, condition) %>%
                                   select(condition) %>% 
                                    distinct() %>% 
                                    pull(condition),
                                 df_annotated %>% 
                                    filter(str_detect(condition, "^BNT-I \\(D\\d+\\)")) %>% 
                                    arrange(day, condition) %>%
                                   select(condition) %>% 
                                    distinct() %>% 
                                    pull(condition),
                                 df_annotated %>% 
                                   filter(disease_vac == "V",
                                          str_detect(condition, "^BNT \\(V1, D\\d+\\)")) %>% 
                                   arrange(day, condition) %>% 
                                   select(condition) %>% 
                                   distinct() %>% 
                                   pull(condition),
                                 df_annotated %>% 
                                   filter(disease_vac == "V",
                                          !str_detect(condition, "^BNT \\(V1, D\\d+\\)"),
                                          !str_detect(condition, "BBIBP"),
                                          !str_detect(condition, "ZF2001")) %>% 
                                   arrange(condition) %>% 
                                   select(condition) %>% 
                                   distinct() %>% 
                                   pull(condition),
                                 df_annotated %>% 
                                   filter(str_detect(condition, "BBIBP")) %>% 
                                   select(condition) %>% 
                                   arrange(condition) %>% 
                                   distinct() %>% 
                                   pull(condition), 
                                 df_annotated %>% 
                                   filter(str_detect(condition, "ZF2001")) %>% 
                                   select(condition) %>% 
                                   arrange(condition) %>% 
                                   distinct() %>% 
                                   pull(condition))) %>%
  mutate(type = if_else(type == "H", "I", type),
         disease_vac = if_else(disease_vac == "H", "I", disease_vac)) 


```

```{r}
df_annotated_2 %>% 
  filter(day != 0) %>% 
  filter(genes == "ITGAM") %>% 
  ggplot() +
  aes(x = condition, y = log2fc, fill = type) +  
  geom_hline(yintercept = 0, linetype = 2) +
  geom_jitter(aes(color = type),
              alpha = 0.3, size = 0.1) +
  geom_boxplot(outliers = F) +
  theme_few() +
  scale_fill_manual(values = c(ann_vaxsig_colors$type)) +
    scale_color_manual(values = c(ann_vaxsig_colors$type)) +
  theme(axis.text.x = element_text(angle = 45,
                                   hjust = 1,
                                   vjust = 1),
        plot.title = element_text(hjust = 0.5, face = "bold"),
        plot.margin = margin(0, 0, 0, 2, "cm"),
        panel.grid.major = element_line(size = 0.05),
        panel.grid.minor = element_line(size = 0.05)) +
  labs(y = "L2FC",
       x="",
       fill = "Type") +
  guides(color = "none") +
  facet_wrap(~disease_vac, scales = "free_x")
```

Generate boxplots for all genes
```{r fig.height=6, fig.width=12}
#Function to create boxplots
create_boxplots_vipgenes <- function(result_df) {
  
  unique_processes <- unique(result_df$genes)
  
  purrr::map(unique_processes, function(process_name) {
    cellmarker_boxplot <- result_df %>% 
      filter(day != 0) %>% 
  filter(genes == process_name) %>% 
  ggplot() +
  aes(x = condition, y = log2fc, fill = type) +  
  geom_hline(yintercept = 0, linetype = 2) +
  geom_jitter(aes(color = type),
              alpha = 0.3, size = 0.1) +
  geom_boxplot(outliers = F) +
  theme_few() +
  scale_fill_manual(values = c(ann_vaxsig_colors$type)) +
    scale_color_manual(values = c(ann_vaxsig_colors$type)) +
  theme(axis.text.x = element_text(angle = 45,
                                   hjust = 1,
                                   vjust = 1),
        plot.title = element_text(hjust = 0.5, face = "bold"),
        plot.margin = margin(0, 0, 0, 2, "cm"),
        panel.grid.major = element_line(size = 0.05),
        panel.grid.minor = element_line(size = 0.05)) +
  labs(y = "L2FC",
       x="",
       title = process_name,
       subtitle = "VIPgenes - COVID-19 infection vs. Vaccination",
       fill = "Type") +
  guides(color = "none") +
  facet_wrap(~disease_vac, scales = "free_x")

    ggsave(cellmarker_boxplot, 
           file = here("Figures", paste0("MachineLearning_VIPGenes_COVID_VAC_boxplot_", process_name, ".png")),
           width = 12, height = 6)
  })
}

# Apply to all processes
create_boxplots_vipgenes(df_annotated_2)


df_annotated_2$genes %>% unique()
```





```{r}
#Inputs
ImmuneGO_Annotated_Genes_all = ImmuneGO_Annotated_Genes %>% 
  filter(!process %in% c("TCR REPERTOIRE", "BCR REPERTOIRE"))

ImmuneGO_Genes = df %>% 
  inner_join(ImmuneGO_Annotated_Genes %>% 
               select(genes, process) %>% 
               distinct() %>% 
               select(genes),
               by = "genes") %>% 
  distinct() %>% 
  select(genes, sample, log2fc) %>% 
  pivot_wider(names_from = sample, values_from = log2fc)


sample_correct =final_res_preds %>% 
  rownames_to_column("sample") %>%  
  inner_join(ann_vaccines_samples %>% 
               select(condition, sample), by = "sample") %>% 
  select(sample, correct) 

# Matrix
l2fc_imps = ImmuneGO_Genes %>% 
  column_to_rownames("genes") %>% 
  as.matrix() %>% 
  replace(is.na(.), 0)

####### INPUT
data_2 = l2fc_imps
filename_2 = paste0("RandomForest_ImmuneGO_Genes_Samples_", outcomevar)

######## Matrix
matrix = ImmuneGO_Genes

######## Annotations
ann_vaccines_covid_other_2 = ImmuneGO_Genes %>% 
  column_to_rownames("genes") %>% 
  colnames() %>% 
  as.data.frame() %>% 
  select(sample = ".") %>% 
  inner_join(ann_vaccines_samples, by = "sample") %>% 
  distinct() %>% 
  select(sample, disease_vac, type, week, day) %>% 
  arrange(week, day) %>% 
  mutate(week = fct_relevel(as.factor(week), sort(levels(week))),
         day = fct_relevel(as.factor(day), sort(levels(day)))) %>% 
  inner_join(sample_correct, by = "sample")
  

#Columns
ann_cols_heatmap = matrix %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(sample = '.') %>% 
  inner_join(ann_vaccines_covid_other_2, by = "sample") %>% 
  distinct() %>% 
  arrange(sample) %>% 
  column_to_rownames("sample")

matrix_data = matrix %>% 
  column_to_rownames("genes") %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "sample") %>% 
  inner_join(ann_cols_heatmap %>% rownames_to_column("sample"), by = "sample") %>% 
  select(!c(disease_vac:day, correct)) %>% 
  arrange(sample) %>% 
  column_to_rownames("sample") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>%
  as.matrix()  %>% 
  replace(is.na(.), 0)

#Rows Immune ----
ann_rows_immune = matrix_data %>%
  as.data.frame() %>%
  rownames_to_column("genes") %>%
  left_join(ImmuneGO_Annotated_Genes %>% filter(go_term == "Manual"), by = "genes") %>%
  select(genes, process) %>%
  distinct() %>%
  group_by(genes) %>%
  arrange(genes, factor(process, levels = c("INNATE IMMUNE SYSTEM", "ADAPTIVE IMMUNE SYSTEM")), .by_group = TRUE) %>%
  mutate(i_process = row_number()) %>%
  pivot_wider(., names_from = "i_process", values_from = "process") %>%
  column_to_rownames("genes") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("index") %>%
  mutate(index = as.numeric(index)) %>%
  arrange(desc(index)) %>%
  column_to_rownames("index") %>%
  t() %>%
  as.data.frame() %>%
  replace(is.na(.), ".") %>% 
  mutate(`1` = if_else(`1` == ".", "INNATE IMMUNE SYSTEM", `1`)) %>% 
  rownames_to_column("genes") %>% 
  mutate(genes = rownames(matrix_data)) %>% 
  column_to_rownames("genes")

#Verificar dimensões do dataset
dim(matrix_data)
dim(ann_cols_heatmap)
dim(ann_rows_immune)
```

```{r}
################# Immune 
#Heatmap annotation
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "right")

row_ha = HeatmapAnnotation(df = ann_rows_immune,
                       col = col_process_immune,
                       which= "row",
                       show_annotation_name = F,
                       show_legend = F,
                       simple_anno_size = unit(1, "mm"))
min(matrix_data)

max(matrix_data)
# Values colors
col_fun <- colorRamp2(c(-20, 0, 20), c("#9bc1bc", "white", "#ed6a5a"))

file = filename_2
mat = matrix_data
width_image = 60
height_image = 40
width = unit(50, "cm")
height = unit(30, "cm")
column_names_gp = gpar(fontsize = 8)
row_names_gp = gpar(fontsize = 10)
rect_gp = gpar(col = "gray80", lwd = 0.5)

fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap.png")

png(file = here("Figures", fileimageheatmap_grouped), 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "Log10 (raw counts)",
                        col = col_fun, #color
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(0, "cm"),
                        row_split = NULL,
                        column_split = NULL,
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(2, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height)

heatmap_plot = draw(heatmap_plot, 
     annotation_legend_side = "left", 
     heatmap_legend_side = "bottom")

dev.off()

```







##  COVID-19 vaccines types:  All types + Healthy + Infection
```{r}
# Input data ----
filename = "ImmuneGO_Genes_Types_L2FC"
df = all_samples_l2fc
# df = all_matrices_counts_tpm_normalized

## Vaccination types
outcomevar= "type"

#New version
ann_vaccines_samples_filtered = ann_vaccines_samples %>%
  select(condition, outcomevar_selected, type, everything()) %>% 
  mutate(type = if_else(day == 0 & type == "I", "I", type),
         type = if_else(disease_vac == "I", "I", type),
         type = if_else(day == 0 & disease_vac != "I", "H", type),
         type = if_else(vaccine == "H", "H", type),
         # type = fct_relevel(type, "H"),
         outcomevar = type) %>% 
  select(condition, outcomevar, type, vaccine, everything()) %>% 
  filter(type != "H") %>% 
  mutate(type = fct_relevel(type, "I")) %>% 
  mutate(vaccine = if_else(condition == "BNT-MO (V1, D6)", "MO", vaccine)) %>% 
  mutate(type = if_else(type == "I" & vaccine != "I", "V-I", type),
         outcomevar = type)


ann_vaccines_samples_filtered


#How many samples by level?
count_levels = ann_vaccines_samples_filtered %>% 
  dplyr::count(outcomevar)

count_levels 

#How many levels?
levels = nrow(count_levels)


#Select only important genes
COVID_Infec_Vac_ImmuneGO_BTM_PC_1_2_cos2 <- read_csv("PCA/COVID_Infec_Vac_ImmuneGO_BTM PC_1_2_cos2.csv")

PC1_2 = COVID_Infec_Vac_ImmuneGO_BTM_PC_1_2_cos2 %>% 
  filter(cos2>0.75) #232 genes

dim(PC1_2)
```

Train and test

```{r fig.height=10, fig.width=10}
# Prepare data ------

#L2FC
df_input1 <- df %>%
  select(genes, sample, log2fc) %>%
  distinct() %>% 
  inner_join(PC1_2 %>% select(genes) %>% distinct(), by = "genes") %>% 
  distinct() %>% 
  pivot_wider(values_from = "log2fc",
              names_from = "sample",
              values_fn = mean) %>% 
  column_to_rownames("genes") %>%
  t() %>%
  as.data.frame() 


df_input = df_input1 %>% 
  rownames_to_column("sample") %>%
  inner_join(ann_vaccines_samples_filtered %>% select(sample,
                                                      # participant_id,
                                                      day,
                                                      age, 
                                                      sex,
                                                      outcomevar,
                                                      vaccine
                                                      ),
             by = "sample") %>%
  select(sample,
         outcomevar,
         age,
         sex,
         # day,
         # vaccine,
         everything()) %>%
  column_to_rownames("sample") %>%
  mutate_if(is.character, factor) %>% 
  mutate(outcomevar = fct_relevel(outcomevar, "I"))

df_input
```


```{r fig.height=10, fig.width=10}
# glimpse(df_input)

# Split data
set.seed(123)  
split <- initial_split(df_input, prop = 0.6, strata = outcomevar)
trees_train <- training(split)
trees_test <- testing(split)

# Create recipe
tree_rec <- recipe(outcomevar ~ ., data = trees_train) %>% #OUTCOME
  step_dummy(all_nominal(), -all_outcomes()) %>%
  #step_corr(all_numeric_predictors()) %>% 
  step_upsample(outcomevar) #Upsample unbalanced data

tree_prep <- prep(tree_rec)

juiced <- juice(tree_prep) %>% 
  select(!colnames(df_input1), colnames(df_input1))

juiced

# Set random forest hyper parameters
tune_spec <- rand_forest(
  mtry = tune(),
  trees = 2000,
  min_n = tune()
) %>%
  set_mode("classification") %>% 
  set_engine("ranger")

# Create workflow
tune_wf <- workflow() %>%
  add_recipe(tree_rec) %>%
  add_model(tune_spec)

#Cross validation
set.seed(234)
trees_folds <- vfold_cv(trees_train, v = 10, strata = outcomevar)

#Define performance metrics
mer_metrics <- metric_set(yardstick::recall,
                          yardstick::specificity,
                          yardstick::accuracy,
                          yardstick::precision,
                          yardstick::roc_auc,
                          yardstick::pr_auc)

# Parallel computing 
doParallel::registerDoParallel()
conflicted::conflicts_prefer(base::intersect)
conflicted::conflicts_prefer(base::setdiff)

set.seed(345)
tune_res <- tune_grid(
  tune_wf,
  metrics = mer_metrics,
  resamples = trees_folds,
  grid = 20
)
```

Assess metrics

```{r fig.height=10, fig.width=10}
# Assess data ----

#Metrics
metrics_tuneres = tune_res %>%
  collect_metrics() %>%
  select(mean, min_n, mtry, .metric) %>%
  pivot_longer(min_n:mtry,
    values_to = "value",
    names_to = "parameter") %>%
  ggplot()+
  aes(x = value, y = mean, colour = .metric, label = value) +
  geom_line() +
  geom_label() +
  scale_color_npg()+
  theme_few() +
  theme(axis.text.x = element_text(color = "black", size = 12)) +
  facet_grid(vars(.metric), vars(parameter), scales = "free")

metrics_tuneres

metrics_tuneres %>% ggsave(file = here("Figures",
                                       paste0(filename, 
                                              outcomevar,
                                              "_tuneres_metrics.png")),
                                        width = 10,
                                        height = 10)

min_n_selec = c(3, 3) #Observations by node)
mtry_selec = c(16, 16) #Features/genes by tree)
```

Hyperparameter tuning

```{r}
# Tune hyperparameters ----
rf_grid <- grid_regular(
  min_n(range = c(min_n_selec)),
  mtry(range = c(mtry_selec)),
  levels = levels
)

set.seed(456)
regular_res <- tune_grid(
  tune_wf,
  metrics = mer_metrics,
  resamples = trees_folds,
  grid = rf_grid
)

# Assess performance -----
regular_res %>%
  collect_metrics()

regular_res %>%
  collect_metrics() %>% 
  write_csv(file = here("Machine learning", paste("metrics_finalmodel_", filename, outcomevar, ".csv")))

#Metrics
regular_res_metrics = regular_res %>%
  collect_metrics() %>%
  select(mean, min_n, mtry, .metric) %>%
  pivot_longer(min_n:mtry,
    values_to = "value",
    names_to = "parameter") %>%
  ggplot()+
  aes(x = value, y = mean, colour = .metric, label = value) +
  geom_line() +
  geom_label(size = 3) +
  scale_color_npg() +
  theme_minimal() +
  facet_grid(vars(.metric), vars(parameter), scales = "free")

regular_res_metrics

# ggsave(regular_res_metrics, file = here("Figures", 
#                                         paste(filename, outcomevar, "metrics", today(), ".png")), 
#        width = 10, height = 10)
```


```{r}
# Select best model -----
best_auc <- select_best(regular_res, metric = "accuracy")
final_rf <- finalize_model(
  tune_spec,
  best_auc
)

```

Important features

```{r fig.height=5, fig.width=5}
final = final_rf %>%
  set_engine("ranger", importance = "permutation") %>%
  fit(outcomevar ~ .,
    data = juice(tree_prep)) %>% 
  vip(num_features = 20)

genes_rf = final$data
genes_rf = genes_rf %>% select(genes = Variable, importance = Importance)
genes_rf %>% write_csv(file = here("Figures", paste0("vipgenes_tidymodels_", filename, outcomevar, ".csv")))


genes_rf_10 = genes_rf %>% 
  arrange(-importance) %>% 
  slice_head(n = 10)

#Importance plot
importance_plot = ggplot(genes_rf_10) +
  aes(x = reorder(genes, importance),
    y = importance,
    fill = importance,
      weight = importance) +
  geom_col(width = 0.6,
           size = 0.5) +
  theme_few() +
  xlab("Genes") + 
  ylab("importance") +
  ylim(0, 0.16) +
  geom_text(aes(label = round(importance, 4), y = importance),
            hjust = -0.2, 
            size = 4,
            angle = 0,
            color = "black")+
  scale_fill_gradient(low = "#4DBBD5FF", high = "#DC0000FF") +
  theme(legend.position = "none",
        plot.subtitle = element_text(hjust = 0),
        axis.text.x = element_text(angle = 0, vjust = 0.5, hjust=0.5, size = 12, color = "black"),
        axis.text.y = element_text(size = 12, angle = 0, color = "black"),
        panel.grid.minor.y = element_blank(),
        legend.text = element_text(size=12),
        legend.title = element_text(size = 12, face = "bold")) +
    coord_flip() +
  labs(title = "Top 10 VIP genes",
       subtitle = "Random forest classification model for \nVaccine types vs. Naive",
       x = "",
       y = "Importance",
       color = "Importance") 
importance_plot
ggsave(importance_plot,
        file = here("Figures",
                    paste0("vipgenes_tidymodels_", filename, outcomevar,  ".png")),
        width = 5,
        height = 5)

importance_plot

```

 Train final model with tuned hyperparameters

```{r fig.height=3, fig.width=5}
# Train final model with tuned hyperparameters
final_wf <- workflow() %>%
  add_recipe(tree_rec) %>%
  add_model(final_rf)

final_res <- final_wf %>%
  last_fit(split)

final_wf_extracted = workflow() %>%
  add_recipe(tree_rec) %>%
  add_model(final_rf) %>% 
  last_fit(split) %>% 
  extract_workflow()

final_res_preds = final_res %>%
  collect_predictions() %>%
  mutate(correct = case_when(
    outcomevar == .pred_class ~ "Correct",
    TRUE ~ "Incorrect"
  )) %>%
  bind_cols(trees_test)


#Confusion matrix
res_conf_mat = conf_mat(final_res_preds, truth = outcomevar...9,
         estimate = .pred_class)

#Accuracy
res_acc = accuracy(final_res_preds, truth = outcomevar...9,
         estimate = .pred_class)

#Specificity
res_spec = spec(final_res_preds, truth = outcomevar...9,
         estimate = .pred_class)

#Precision
res_prec = precision(final_res_preds, truth = outcomevar...9,
         estimate = .pred_class)

#Recall
res_recall = recall(final_res_preds, truth = outcomevar...9,
         estimate = .pred_class)
#Precision recall AUC
# res_pr_auc = pr_auc(final_res_preds, outcomevar...9, .pred_RNA)


#Unite results
final_res_metrics = bind_rows(res_acc,
                         res_recall,
                         res_prec,
                         # res_pr_auc,
                         res_spec) %>% 
  select(-".estimator")


final_res_metrics_conf = list(res_conf_mat, 
                         final_res_metrics,
                         final_res_preds)

final_res_metrics_conf

final_res_metrics_conf %>% 
  saveRDS(file = here("Machine learning", paste0("final_res_metrics", filename, outcomevar, ".rds")))


#Plot
final_model_metrics = final_res_metrics %>% 
  rename(metric = .metric,
         estimate = .estimate) %>% 
  mutate(metric = fct_relevel(metric, final_res_metrics$.metric)) %>% 
  ggplot() +
  aes(x = estimate,
    y = fct_rev(metric),
    fill = estimate,
    weight = estimate) +
  geom_col(size = 0.5,
           width = 0.6) +
  theme_few() +
  geom_text(aes(label = round(estimate, 3), y = metric),
            hjust = -0.2, 
            size = 4,
            angle = 0,
            color = "black") +
  xlim(0, 1.2) +
  scale_fill_gradient(low = "#4DBBD5FF", high = "#DC0000FF") +
  theme(legend.position = "none",
        plot.subtitle = element_text(hjust = 0),
        axis.text.x = element_text(angle = 0, vjust = 0.5, hjust=0.5, size = 12, color = "black"),
        axis.text.y = element_text(size = 12, angle = 0, color = "black"),
        panel.grid.minor.y = element_blank(),
        legend.text = element_text(size=12),
        legend.title = element_text(size = 12, face = "bold")) +
  labs(title = "Model assessment",
       subtitle = "Random forest classification model for \nVaccine types vs. Naive",
       x = "Estimate",
       y = "",
       fill = "Estimate")

ggsave(final_model_metrics,
        file = here("Figures",
                    paste0("final_model_metrics_", filename, outcomevar,  ".png")),
        width = 5,
        height = 3)


#Which are the incorrect?
final_res_preds %>% rownames_to_column("sample") %>%  
  inner_join(ann_vaccines_samples %>% 
               select(condition, sample), by = "sample") %>% 
  filter(correct == "Incorrect") %>% 
  select(sample, condition, everything()) %>% 
  view()

final_model_metrics

#Matrix counts of VIP genes
all_matrices_counts_genesRF = all_samples_l2fc %>% 
  inner_join(genes_rf %>% 
               select(genes), by = "genes")
```

Multi ROC
```{r fig.height=5, fig.width=7}
multi_roc = final_res_preds %>% 
  group_by(id) %>%
  roc_curve(outcomevar...9, .pred_I:.pred_VV) %>%
  ggplot(aes(1 - specificity, sensitivity, color = .level)) +
  geom_abline(lty = 2, color = "gray80", size = 0.5) +
  geom_path(show.legend = FALSE, alpha = 0.6, size = 1.2) +
  facet_wrap(~.level, ncol = 3) +
  # coord_equal() +
  theme_few() +
    theme(axis.text = element_text(size = 10, color = "black"), 
          panel.spacing.x = unit(2, "lines")) +
  labs(title ="Receiver-operator curve",
       subtitle = "Multiclass classification, Randomforest",
       x = "1-Specificity",
       y = "Sensitivity") +
  scale_colour_manual(values = c(ann_vaxsig_colors$type[names(ann_vaxsig_colors$type) != "H"], 
                                 H = "black"))

multi_roc
```

Confusion matrix
```{r fig.height=4, fig.width=4}
conf_matrix_heat = final_res_preds %>% 
  conf_mat(outcomevar...9, .pred_class) %>%
  autoplot(type = "heatmap") +
    theme_few() +
  theme(axis.text = element_text(size = 10, color = "black"),
        legend.position = "none") +
  scale_fill_gradient(low = "white", high = "#DC0000FF") +
  labs(title ="Confusion matrix",
       subtitle = "Multiclass classification, Randomforest",
       x = "Truth",
       y = "Prediction")

conf_matrix_heat
```


```{r fig.height=5, fig.width=12}
conf_matrix_heat + multi_roc +
  plot_layout(widths = c(1, 2))
  
```

Precision recall curve
```{r}
# Precision recall
pr_curve_res = pr_curve(final_res_preds, outcomevar...9, .pred_I) %>% 
  ggplot(aes(x = recall, y = precision)) +
  geom_path(color = "#DC0000FF",
            size = 2) +
  coord_equal() +
  theme_few() +
  labs(title = "Precision-Recall curve",
       subtitle = "Random forest classification \nmodel for Infection vs. Vaccination",
       x = "Recall",
       y = "Precision") 


pr_curve_res %>% 
  ggsave(file = here("Figures",
                    paste0("PR_Curve_tidymodels_", filename, outcomevar,  ".png")),
         width = 4,
         height = 3)
  
```


Relative effects

Function
```{r fig.height=5, fig.width=7}
REffect <- function(data, dep_vars, indep_var, n_boot = 1000, 
                    plot = TRUE, plot_colors = c("red", "darkblue"), 
                    plot_labels = list(y = "Relative effect", x = "Genes"), 
                    plot_theme = theme_few(), plot_legend_position = "top", 
                    plot_legend_labels = c("Vaccination", "Infection")) {
  
  library(npmv)
  library(ggplot2)
  library(reshape2)
  
  if(!all(data[[indep_var]] %in% c(0, 1))) {
    stop("A variável independente precisa conter apenas 0 e 1 (dois grupos).")
  }
  
  n_aab <- nrow(data)
  list_boot_aab <- list()
  
  # Loop de Bootstrap
  for (i in 0:n_boot) {
    if (i == 0) {
      df_sample <- 1:n_aab
    } else {
      df_sample <- sample(1:n_aab, n_aab, replace = TRUE)
    }
    
    formula_str <- paste(dep_vars, collapse = " | ")
    formula <- as.formula(paste(formula_str, "~", indep_var))
    
    manova_aab <- nonpartest(formula, data = data[df_sample, ], permtest = FALSE, plots = FALSE,
                             permreps = 1000, tests = c(1, 0, 0, 0))
    
    if (i == 0) obs_aab <- manova_aab
    
    list_boot_aab[[i + 1]] <- manova_aab
    
    cat("ANOVA/ boot/ aab/ i = ", i, "\n")
  }
  
  obs_manova <- list_boot_aab[[1]]$twogroupreleffects
  row_manova <- rownames(obs_manova)
  
  releffects <- lapply(list_boot_aab[-1], "[[", 2)
  
  df_perc_boot <- lapply(seq_len(nrow(obs_manova)), function(i) {
    q <- sapply(seq_len(ncol(obs_manova)), function(j) {
      v <- sapply(releffects, function(r) {
        row <- rownames(r)
        if(row[1] != row_manova[1]) r <- r[2:1,]
        r[i, j]
      })
      quantile(v, probs = c(0.025, 0.975))
    })
    colnames(q) <- colnames(obs_manova)
    df <- data.frame(Var2 = colnames(q), t(q),
                     group = ifelse(i == 1, row_manova[1], row_manova[2]))
    df$Var2 <- factor(df$Var2, levels = sort(unique(df$Var2)), ordered = TRUE)
    return(df)
  })
  
  names(df_perc_boot) <- row_manova
  
  df <- reshape2::melt(
    cbind(obs_manova, group = row_manova),
    id.vars = "group"
  )
  df$Var1 <- "point_est"
  df <- df[, c(4, 2, 3, 1)]
  colnames(df)[2] <- "Var2"
  
  df <- dplyr::arrange(df, group, desc(value), Var2)
  df$Var2 <- as.character(df$Var2)
  df$Var2 <- factor(df$Var2, levels = unique(df$Var2), ordered = TRUE)
  
  for (k in seq_len(2)) {
    df_perc_boot[[k]]$Var2 <- factor(df_perc_boot[[k]]$Var2, levels = levels(df$Var2), ordered = TRUE)
  }
  
  if (plot) {
    p <- ggplot(data = df, aes(x = Var2, y = value, group = group, color = group)) +
      geom_ribbon(inherit.aes = FALSE, data = df_perc_boot[[1]],
                  aes(x = Var2, ymin = X2.5., ymax = X97.5., group = group), fill = plot_colors[1],
                  alpha = 0.3) +
      geom_ribbon(inherit.aes = FALSE, data = df_perc_boot[[2]],
                  aes(x = Var2, ymin = X2.5., ymax = X97.5., group = group), fill = plot_colors[2],
                  alpha = 0.3) +
      geom_line() +
      geom_point(aes(size = value)) +
      scale_colour_manual("", values = c("1" = plot_colors[2], "0" = plot_colors[1]),
                          labels = plot_legend_labels) +
      plot_theme +
      theme(
        legend.text = element_text(size = 12),
        axis.text.x = element_text(size = 12, angle = 90),
        axis.text.y = element_text(size = 12),
        axis.line.y = element_blank(),
        axis.title = element_text(size = 15),
        panel.grid.major = element_line(),
        legend.position = plot_legend_position,
        legend.box.background = element_rect()
      ) +
      labs(y = plot_labels$y, x = plot_labels$x)
    
    print(p)
  }
  
  return(list(obs_manova = obs_manova, df_perc_boot = df_perc_boot, plot = p))
}
```

Plot
```{r fig.height=5, fig.width=7, message=FALSE, warning=FALSE}
head(all_matrices_counts_genesRF)

covid_vac_releff = all_matrices_counts_genesRF %>% 
  select(genes, sample, counts) %>% 
  pivot_wider(names_from = "genes",
              values_from = "counts") %>% 
  inner_join(all_matrices_counts_genesRF %>% 
               select(sample, disease_vac), by = "sample") %>% 
  mutate(disease_vac = if_else(disease_vac == "I", 1, 0))


releffect_disease_vac = REffect(covid_vac_releff, dep_vars = colnames(covid_vac_releff %>% 
                                              select(-disease_vac, -sample)),
        indep_var = "disease_vac", plot_colors = c("red", "#4DBBD5FF"), 
        n_boot=1000)


plot = releffect_disease_vac$plot +
  scale_colour_manual(values = c("1" = "#DC0000FF", 
                                 "0" = "#4DBBD5FF"),
                      labels = plot_legend_labels) +
  geom_ribbon(inherit.aes = FALSE, data = releffect_disease_vac$df_perc_boot[[1]],
                  aes(x = Var2, ymin = X2.5., ymax = X97.5., group = group), fill = "#4DBBD5FF",
                  alpha = 0.3) +
      geom_ribbon(inherit.aes = FALSE, data = releffect_disease_vac$df_perc_boot[[2]],
                  aes(x = Var2, ymin = X2.5., ymax = X97.5., group = group), fill = "#DC0000FF",
                  alpha = 0.3) +
  theme(
        legend.text = element_text(size = 12),
        axis.text.x = element_text(size = 12, angle = 45, hjust = 1, vjust = 1, color = "black"),
        axis.text.y = element_text(size = 12, color = "black"),
        axis.line.y = element_blank(),
        axis.title = element_text(size = 15),
        panel.grid.major = element_line(),
        legend.position = "top",
        plot.title = element_text(hjust = 0.5, size = 20),
        plot.subtitle = element_text(hjust = 0.5, size = 15),
        legend.box.background = element_blank()
      ) +
      labs(y = "Relative effect", x = "Genes",
           title = "Relative effects of VIP genes",
           subtitle = "Infection vs. Vaccination",
           color = "") +
  ylim(0, 1)

plot

```



#### Heatmap: Gene L2FC per condition

```{r}
# Required files
ImmuneGO_Annotated_Genes = readRDS(here("VaxGO gene sets","ImmuneGO_Annotated_Genes_2024-03-26.rds"))
all_degs_p_05_vac_infected = readRDS(here("DGE analysis", "all_studies_degs_weighted.rds")) %>% 
  filter(if_else(n >= 5, fdr <= 0.05, fdr <= (0.05 * 6/n))) 

ann_vaccines <- read_csv(here("Annotations and metadata", "ann_vaccines_conditions_23-10-24.csv"))
# df = read_csv("Figures/vipgenes_tidymodels_ImmuneGO_Genes_Types_type.csv") %>% 
#   slice_head(n=10)
#Inputs
ImmuneGO_Annotated_Genes_all = ImmuneGO_Annotated_Genes %>% 
  filter(!process %in% c("TCR REPERTOIRE", "BCR REPERTOIRE"))

ImmuneGO_Genes = all_degs_p_05_vac_infected %>% 
  inner_join(genes_rf %>% 
               slice_head(n = 15) %>% 
               select(genes), by = "genes") %>% 
  # inner_join(ImmuneGO_Annotated_Genes %>% 
  #              select(genes, process) %>% 
  #              distinct() %>% 
  #              select(genes),
  #              by = "genes") %>% 
  distinct() %>% 
  select(genes, condition, log2fold_change) %>% 
  pivot_wider(names_from = condition, values_from = log2fold_change)

# Matrix
l2fc_imps = ImmuneGO_Genes %>% 
  column_to_rownames("genes") %>% 
  as.matrix() %>% 
  replace(is.na(.), 0)


####### INPUT
data_2 = l2fc_imps
outcomevar = "VaccineType"
filename_2 = paste0("RandomForest_ImmuneGO_Genes_Conditions_", outcomevar)

######## Matrix
matrix = ImmuneGO_Genes

######## Annotations
ann_vaccines_covid_other_2 = ImmuneGO_Genes %>% 
  column_to_rownames("genes") %>% 
  colnames() %>% 
  as.data.frame() %>% 
  select(condition = ".") %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  distinct() %>% 
  select(condition, disease_vac, type, week, day) %>% 
  arrange(week, day) %>% 
  mutate(week = fct_relevel(as.factor(week), sort(levels(week))),
         day = fct_relevel(as.factor(day), sort(levels(day)))) 

```


```{r fig.height=10, fig.width=15}
#Columns
ann_cols_heatmap = matrix %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  inner_join(ann_vaccines_covid_other_2, by = "condition") %>% 
  distinct() %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  mutate(day = as.integer(day))

matrix_data = matrix %>% 
  column_to_rownames("genes") %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  inner_join(ann_cols_heatmap %>% rownames_to_column("condition"), by = "condition") %>% 
  select(!c(disease_vac:day)) %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>%
  as.matrix()  %>% 
  replace(is.na(.), 0)

#Rows Immune ----
# ann_rows_immune = matrix_data %>%
#   as.data.frame() %>%
#   rownames_to_column("genes") %>%
#   left_join(ImmuneGO_Annotated_Genes %>% filter(go_term == "Manual"), by = "genes") %>%
#   select(genes, process) %>%
#   distinct() %>%
#   group_by(genes) %>%
#   arrange(genes, factor(process, levels = c("INNATE IMMUNE SYSTEM", "ADAPTIVE IMMUNE SYSTEM")), .by_group = TRUE) %>%
#   mutate(i_process = row_number()) %>%
#   pivot_wider(., names_from = "i_process", values_from = "process") %>%
#   column_to_rownames("genes") %>%
#   t() %>%
#   as.data.frame() %>%
#   rownames_to_column("index") %>%
#   mutate(index = as.numeric(index)) %>%
#   arrange(desc(index)) %>%
#   column_to_rownames("index") %>%
#   t() %>%
#   as.data.frame() %>%
#   replace(is.na(.), ".") %>% 
#   mutate(`1` = if_else(`1` == ".", "INNATE IMMUNE SYSTEM", `1`)) %>% 
#   rownames_to_column("genes") %>% 
#   mutate(genes = rownames(matrix_data)) %>% 
#   column_to_rownames("genes")

#Verificar dimensões do dataset
dim(matrix)
dim(matrix_data)
dim(ann_cols_heatmap)
dim(ann_rows_immune)

################# Immune 
#Heatmap annotation
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "right")

# row_ha = HeatmapAnnotation(df = ann_rows_immune,
#                        col = col_process_immune,
#                        which= "row",
#                        show_annotation_name = F,
#                        show_legend = F,
#                        simple_anno_size = unit(1, "mm"))

# Values colors
col_fun <- colorRamp2(c(-1, 0, 1), c("#4DBBD5FF", "white", "#ed6a5a"))

file = filename_2
mat = matrix_data
width_image = 30
height_image = 15
width = unit(15, "cm")
height = unit(6, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)
rect_gp = gpar(col = "white", lwd = 2)

fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap.png")

png(file = here("Figures", fileimageheatmap_grouped), 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        # left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "L2FC",
                        col = col_fun, #color
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = T,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(0, "cm"),
                        row_split = NULL,
                        column_split = NULL,
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(2, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height)

heatmap_plot = draw(heatmap_plot, 
     annotation_legend_side = "right", 
     heatmap_legend_side = "right")

dev.off()

heatmap_plot

```
#### Heatmap: Gene counts per sample

```{r}
# Required files
# ImmuneGO_Annotated_Genes = readRDS(here("VaxGO gene sets","ImmuneGO_Annotated_Genes_2024-03-26.rds"))
df = all_samples_l2fc %>% 
  inner_join(genes_rf %>% 
               arrange(-importance) %>% 
               slice_head(n=15) %>% 
               select(genes), by = "genes")

ann_vaccines_samples <- read_csv(here("Annotations and metadata", "ann_vaccines_samples.csv"))
ann_vaccines_samples = ann_vaccines_samples %>% 
   mutate(type = if_else(day == 0 & type == "I", "I", type),
         type = if_else(disease_vac == "I", "I", type),
         type = if_else(day == 0 & disease_vac != "I", "H", type),
         type = if_else(vaccine == "H", "H", type),
         # type = fct_relevel(type, "H"),
         outcomevar = type) %>% 
  select(condition, outcomevar, type, vaccine, everything()) %>% 
  filter(type != "H") %>% 
  mutate(type = fct_relevel(type, "I")) %>% 
  mutate(vaccine = if_else(condition == "BNT-MO (V1, D6)", "MO", vaccine)) %>% 
  mutate(type = if_else(type == "I" & vaccine != "I", "V-I", type),
         outcomevar = type)

#Inputs

ImmuneGO_Genes = df %>% 
  # inner_join(ImmuneGO_Annotated_Genes %>% 
  #              select(genes, process) %>% 
  #              distinct() %>% 
  #              select(genes),
  #              by = "genes") %>% 
  distinct() %>% 
  select(genes, sample, log2fc) %>% 
  pivot_wider(names_from = sample, 
              values_from = log2fc,              
              values_fn = mean)


sample_correct =final_res_preds %>% 
  rownames_to_column("sample") %>%  
  inner_join(ann_vaccines_samples %>% 
               select(condition, sample), by = "sample") %>% 
  select(sample, type_pred = .pred_class, correct)

# Matrix
l2fc_imps = ImmuneGO_Genes %>% 
  column_to_rownames("genes") %>% 
  as.matrix() %>% 
  replace(is.na(.), 0)

####### INPUT
data_2 = l2fc_imps
filename_2 = paste0("RandomForest_ImmuneGO_Genes_Samples_", outcomevar)

######## Matrix
matrix = ImmuneGO_Genes

######## Annotations
ann_vaccines_covid_other_2 = ImmuneGO_Genes %>% 
  column_to_rownames("genes") %>% 
  colnames() %>% 
  as.data.frame() %>% 
  select(sample = ".") %>% 
  inner_join(ann_vaccines_samples, by = "sample") %>% 
  distinct() %>% 
  select(sample, condition, gse_id, type, vaccine, variant, dose,  day, sex, age) %>% 
  arrange(dose, day) %>% 
  mutate(day = fct_relevel(as.factor(day), sort(levels(day))),
         type = fct_relevel(type, "H", "I", "RNA", "VV", "SU", "IN")) %>% 
  inner_join(sample_correct, by = "sample") %>% 
  mutate(colnames = paste(condition, sample))

#Columns
ann_cols_heatmap = matrix %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(sample = '.') %>% 
  inner_join(ann_vaccines_covid_other_2, by = "sample") %>% 
  distinct() %>% 
  arrange(dose, day) %>% 
  column_to_rownames("colnames")

matrix_data = matrix %>% 
  column_to_rownames("genes") %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "sample") %>% 
  inner_join(ann_cols_heatmap %>% 
               rownames_to_column("colnames"), by = "sample") %>% 
  arrange(dose, day) %>% 
  select(!c(gse_id:age, sample, type_pred, correct, condition)) %>% 
  column_to_rownames("colnames") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>%
  as.matrix()  %>% 
  replace(is.na(.), 0)

ann_cols_heatmap = ann_cols_heatmap %>% 
  select(!c(condition, sample)) %>% 
  select(correct, 
         # vaccine, 
         type, 
         type_pred,
         # gse_id, 
         dose, 
         day,
         sex, 
         age, 
         variant) %>% 
  mutate(day = as.integer(day))
  

#Verificar dimensões do dataset
dim(matrix_data)
dim(ann_cols_heatmap)
dim(ann_rows_immune)


#Column split
column_split = ann_cols_heatmap$type %>% as.character()

dend1 = cluster_between_groups(matrix_data, column_split)
#Heatmap annotation
ann_vaxsig_colors$age <- circlize::colorRamp2(c(0, 50, 100), c("white", "#4DBBD5FF", "black"))
ann_vaxsig_colors$day <- circlize::colorRamp2(c(0, 50, 100), c("white", "#4DBBD5FF", "black"))
ann_vaxsig_colors$type_pred <- ann_vaxsig_colors$type
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "right")
min(matrix_data)
max(matrix_data)

# Values colors
col_fun <- colorRamp2(c(-25, 0, 25), c("#4DBBD5FF", "white", "#DC0000FF"))

file = filename_2
mat = matrix_data
width_image = 60
height_image = 20
width = unit(50, "cm")
height = unit(6, "cm")
column_names_gp = gpar(fontsize = 5)
row_names_gp = gpar(fontsize = 10)
rect_gp = gpar(col = "white", lwd = 2)

fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap.png")

png(file = here("Figures", fileimageheatmap_grouped), 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        # left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE, 
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "L2FC",
                        col = col_fun, #color
                        show_column_names = T, 
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(1, "cm"), 
                        column_dend_height =  unit(2, "cm"),
                        row_split = NULL,
                        column_split = 6,
                        # column_split = column_split,
                        row_gap = unit(2, "mm"),
                        cluster_columns = dend1,
                        # cluster_columns = F,
                        column_gap = unit(2, "mm"),
                        show_column_dend = T,
                        show_row_dend = T,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height)

heatmap_plot = draw(heatmap_plot, 
     annotation_legend_side = "left", 
     heatmap_legend_side = "left")

dev.off()

```
Clustered by type
```{r}
file = filename_2
mat = matrix_data
width_image = 60
height_image = 20
width = unit(50, "cm")
height = unit(6, "cm")
column_names_gp = gpar(fontsize = 5)
row_names_gp = gpar(fontsize = 10)
rect_gp = gpar(col = "white", lwd = 2)

fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap_Clustered.png")

png(file = here("Figures", fileimageheatmap_grouped), 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        # left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE, 
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "L2FC",
                        col = col_fun, #color
                        show_column_names = T, 
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(1, "cm"), 
                        column_dend_height =  unit(2, "cm"),
                        row_split = NULL,
                        column_split = 5,
                        # column_split = column_split,
                        row_gap = unit(2, "mm"),
                        cluster_columns = T,
                        column_gap = unit(2, "mm"),
                        show_column_dend = T,
                        show_row_dend = T,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height)

heatmap_plot = draw(heatmap_plot, 
     annotation_legend_side = "left", 
     heatmap_legend_side = "left")

dev.off()
```

```{r}
file = filename_2
mat = matrix_data
width_image = 60
height_image = 20
width = unit(50, "cm")
height = unit(6, "cm")
column_names_gp = gpar(fontsize = 5)
row_names_gp = gpar(fontsize = 10)
rect_gp = gpar(col = "white", lwd = 2)

fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap_Clustered_all.png")

png(file = here("Figures", fileimageheatmap_grouped), 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        # left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE, 
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "L2FC",
                        col = col_fun, #color
                        show_column_names = T, 
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(1, "cm"), 
                        column_dend_height =  unit(2, "cm"),
                        row_split = NULL,
                        column_split = 5,
                        # column_split = column_split,
                        row_gap = unit(2, "mm"),
                        cluster_columns = T,
                        column_gap = unit(2, "mm"),
                        show_column_dend = T,
                        show_row_dend = T,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height)

heatmap_plot = draw(heatmap_plot, 
     annotation_legend_side = "left", 
     heatmap_legend_side = "left")

dev.off()
```




### Scatterplot
```{r fig.height=15, fig.width=15, message=FALSE, warning=FALSE}
library(GGally)
genes_rf

selected_genes_rf = df_input %>% 
  select(outcomevar, age, sex, day, vaccine, any_of(genes_rf %>% slice_head(n = 15) %>% 
                                                      pull(genes)))

ggpairs_rf = ggpairs(selected_genes_rf, columns = 6:15, 
        aes(color = outcomevar,
            alpha = 0.5))

ggpairs_rf = ggpairs_rf + 
  scale_color_manual(values = ann_vaxsig_colors$type) +
  scale_fill_manual(values = ann_vaxsig_colors$type) +
  theme_minimal() +
  theme(legend.position = "right",
        strip.text.y = element_text(color = "black", size = 12, face = "bold"),
        strip.text.x = element_text(color = "black", size = 12, face = "bold"),
        axis.text.x = element_text(color = "black", size = 10),
        axis.text.y = element_text(vjust = 0.5,
                                   size = 10,
                                   color = "black"),
        axis.line = element_line(size = 1,
                                 colour = "black",
                                 linetype = 1),
        axis.ticks = element_line(size = 0.5, color = "black"),
        panel.grid.major.y = element_line(size = 0.1,
                                          linetype = 2,
                                          color = "gray75"),
        panel.grid.major.x = element_line(size = 0.1,
                                          linetype = 2,
                                          color = "gray75"),
        panel.spacing.x = unit(2, "lines"))

ggpairs_rf

ggpairs_rf %>% 
  ggsave(filename = here("Figures", paste(file, "ggpairs_rf_selected_genes.png")), width = 15,
         height = 15)
  
```






### PCA with VIP genes
**Input**

```{r}
#INPUT
# data_genes = all_matrices_counts_genesRF 
data_genes = all_samples_l2fc %>% 
  filter(genes %in% c(genes_rf %>% 
                        filter(!genes %in% c("vaccine_BNT", "vaccine_ChAd", "vaccine_I", "vaccine_ZF2001", "day")) %>% 
                        slice_head(n = 10) %>% 
                        pull(genes)))


data_annotation = ann_vaccines_samples_filtered

#Convert long to wide table format.
#Dataframe with sample annotation

ann_vaccines_pca_matrix = data_genes %>% 
  select(sample, genes, log2fc) %>% 
  pivot_wider(names_from = sample, 
              values_from = log2fc,              
              values_fn = mean) %>% 
  replace(is.na(.), 0) %>%
  column_to_rownames("genes") %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column("sample") %>% 
  inner_join(data_annotation %>% 
               select(sample, condition, disease_vac, gse_id, vaccine, day, week, dose, infection, type), by = "sample") 

# Convert dataframe to matrix without annotation columns
ann_vaccines_pca_matrix_ready = ann_vaccines_pca_matrix %>% 
  column_to_rownames("sample") %>% 
  select(!condition:type) %>% 
  as.matrix()

# Delete columns with near 0 variance
nearZeroVarCols <- nearZeroVar(ann_vaccines_pca_matrix_ready, saveMetrics = TRUE)
ann_vaccines_pca_matrix_ready <- ann_vaccines_pca_matrix_ready[, !nearZeroVarCols$nzv]
pca_res <- prcomp(ann_vaccines_pca_matrix_ready, scale. = T)

# Correlation plot ----
corr_matrix = cor(ann_vaccines_pca_matrix_ready %>% scale()) 
var <- get_pca_var(pca_res)

# corrplot = ggcorrplot(corr_matrix, hc.order = TRUE) +
#   theme(axis.text.x = element_text(angle = 90, size = 5),
#         axis.text.y = element_text(size = 5))
# #Salvar
# ggsave(corrplot, file = paste0(filename, "_corrplot.png"),
#        width = 20, #Grande 20, pequeno 10
#        height= 20) #Grande 20, pequeno 10
# print(corrplot)

#PCA ----
data.pca = prcomp(corr_matrix) #PCA
summary(corr_matrix) #Retornar PCs

#Scree plot ----
scree_plot = fviz_eig(data.pca, 
                      addlabels = TRUE,
                      ylim = c(0, 70)) +
  geom_col(color = "#00AFBB", fill = "#00AFBB") +
  theme_classic()

#Save
ggsave(scree_plot, file = here("Figures", paste0(filename, "_screeplot.png")), width = 10, height = 3) 
print(scree_plot)

# Loadings plot ----
options(ggrepel.max.overlaps = Inf)

circle_contrib = fviz_pca_var(pca_res, col.var = "cos2",
                              gradient.cols = c("black", "#DC0000FF"),
                              select.var= list(cos2 = 20), 
                              repel = T, 
                              labelsize = 4,
                              col.circle = NA) +
  xlab("") + 
  ylab("") +
  theme_minimal() +
theme(panel.grid = element_blank(),  # Remover linhas de grade
        axis.text = element_blank(),   # Remover rótulos de texto dos eixos
        axis.ticks = element_blank()) 

#Save
ggsave(circle_contrib, file = here("Figures", paste0(filename, "_circle_contrib_wide.png")), width = 10, height = 5)

circle_contrib

#KNN classification -----
#Use the screeplot generated % of explained variance for each dimension
scree_plot #Dim 1 = 68.2%, Dim 2 = 15.9%


# PC1-PC2 ------
#Determine the number of clusters
pca_scores <- data.frame(pca_res$x[, 1:2])
fviz_nbclust(pca_scores,  
                     FUNcluster = kmeans,
                     method = "wss")

pcas = "PC1-PC2"

set.seed(666) # Set seed for randomization
cluster_model <- kmeans(pca_res$x[, 1:2], centers = 5)  #Set the number of clusters

ann_vaccines_pca_matrix$cluster <- as.factor(cluster_model$cluster)
```

```{r fig.height=7, fig.width=10, paged.print=FALSE}
# Condition with density plot ------
pca_plot_knn_cluster = autoplot(pca_res, 
        data = ann_vaccines_pca_matrix,
        colour = "gse_id")  +
  stat_ellipse(aes(color = cluster, fill = cluster), 
               geom = "polygon", 
               alpha = 0.2, 
               linetype = 1,
               size = 0.3,
               type = "t") +
  labs(title=paste0(filename, "_bycondition")) + 
  xlab("PC1 (44.9%)") +
  ylab("PC2 (15.9%)") +
    geom_hline(yintercept = 0, size = 0.3, color = "gray50", linetype = 4) +
  geom_vline(xintercept = 0, size = 0.3, color = "gray50", linetype = 4) +
  geom_point(aes(fill = type,
                 color = type),
             stroke = 0.2) +
   scale_color_manual(values = c("1" = "#8491B4FF", 
                                "2" = "#4DBBD5FF", 
                                "3" = "#DC0000FF",
                                ann_vaxsig_colors$type), 
                     guide = "none") +
  scale_fill_manual(values = c("1" = "#8491B4FF", 
                               "2" = "#4DBBD5FF", 
                               "3" = "#DC0000FF",
                               ann_vaxsig_colors$type),
                    guide = "none") + 
  guides(col="none") +
  theme_few() +
  theme(axis.text = element_text(size = 10, color = "black"),
      panel.grid.minor = element_blank(),
      panel.grid.major = element_line(size = 1, 
                                    linetype = 1,
                                    color = "gray90"),
        axis.line = element_line(size = 0, colour = "black", linetype=1),
        # axis.line = element_line(size = 1, colour = "black", linetype=1),
        text = element_text(color = "black", size = 10),
        panel.border = element_rect(fill = NA, color = "gray90", size = 2)) 
  # geom_text_repel(aes(label = condition,
  #                     segment.colour="gray70"),
  #                 box.padding = 0.3,
  #                 size = 3) 

ggsave(pca_plot_knn_cluster, 
       file = here( "Figures", paste0(filename, pcas, "_KNN_Clustered_labels.png")), width = 10, height = 7)
pca_plot_knn_cluster
# pca_plot_knn_cluster_marginal = ggMarginal(
#   pca_plot_knn_cluster,
#   type = "histogram",
#   groupFill = T,
#   groupColour = T,
#   xparams = list(binwidth = 0.1, size = 0.1),
#   yparams = list(binwidth = 0.1, size = 0.1)
# )
# 
# ggsave(
#   pca_plot_knn_cluster_marginal,
#   file = here(paste0(filename, pcas, "_KNN_Clustered_labels_Marginal.png"), "Figures"),
#   width = 8,
#   height = 5
# )


```

##  COVID-19 vaccines types:  Vaccine types
```{r}
# Input data ----
filename = "ImmuneGO_Genes_Types_L2FC"
df = all_samples_l2fc
# df = all_matrices_counts_tpm_normalized

## Vaccination types
outcomevar= "type"

#New version
ann_vaccines_samples_filtered = ann_vaccines_samples %>%
  select(condition, outcomevar_selected, type, everything()) %>% 
  mutate(type = if_else(day == 0 & type == "I", "I", type),
         type = if_else(disease_vac == "I", "I", type),
         type = if_else(day == 0 & disease_vac != "I", "H", type),
         type = if_else(vaccine == "H", "H", type),
         # type = fct_relevel(type, "H"),
         outcomevar = type) %>% 
  select(condition, outcomevar, type, vaccine, everything()) %>% 
  filter(type != "H") %>% 
  mutate(type = fct_relevel(type, "I")) %>% 
  mutate(vaccine = if_else(condition == "BNT-MO (V1, D6)", "MO", vaccine)) %>% 
  mutate(type = if_else(type == "I" & vaccine != "I", "V-I", type),
         outcomevar = type)


ann_vaccines_samples_filtered


#How many samples by level?
count_levels = ann_vaccines_samples_filtered %>% 
  dplyr::count(outcomevar)

count_levels 

#How many levels?
levels = nrow(count_levels)


#Select only important genes
COVID_Infec_Vac_ImmuneGO_BTM_PC_1_2_cos2 <- read_csv("PCA/COVID_Infec_Vac_ImmuneGO_BTM PC_1_2_cos2.csv")

PC1_2 = COVID_Infec_Vac_ImmuneGO_BTM_PC_1_2_cos2 %>% 
  filter(cos2>0.75) #232 genes

dim(PC1_2)
```

Train and test

```{r fig.height=10, fig.width=10}
# Prepare data ------

#L2FC
df_input1 <- df %>%
  select(genes, sample, log2fc) %>%
  distinct() %>% 
  inner_join(PC1_2 %>% select(genes) %>% distinct(), by = "genes") %>% 
  distinct() %>% 
  pivot_wider(values_from = "log2fc",
              names_from = "sample",
              values_fn = mean) %>% 
  column_to_rownames("genes") %>%
  t() %>%
  as.data.frame() 


df_input = df_input1 %>% 
  rownames_to_column("sample") %>%
  inner_join(ann_vaccines_samples_filtered %>% select(sample,
                                                      # participant_id,
                                                      day,
                                                      age, 
                                                      sex,
                                                      outcomevar,
                                                      vaccine
                                                      ),
             by = "sample") %>%
  select(sample,
         outcomevar,
         age,
         sex,
         # day,
         # vaccine,
         everything()) %>%
  column_to_rownames("sample") %>%
  mutate_if(is.character, factor) %>% 
  mutate(outcomevar = fct_relevel(outcomevar, "I"))

df_input
```


```{r fig.height=10, fig.width=10}
# glimpse(df_input)

# Split data
set.seed(123)  
split <- initial_split(df_input, prop = 0.6, strata = outcomevar)
trees_train <- training(split)
trees_test <- testing(split)

# Create recipe
tree_rec <- recipe(outcomevar ~ ., data = trees_train) %>% #OUTCOME
  step_dummy(all_nominal(), -all_outcomes()) %>%
  #step_corr(all_numeric_predictors()) %>% 
  step_upsample(outcomevar) #Upsample unbalanced data

tree_prep <- prep(tree_rec)

juiced <- juice(tree_prep) %>% 
  select(!colnames(df_input1), colnames(df_input1))

juiced

# Set random forest hyper parameters
tune_spec <- rand_forest(
  mtry = tune(),
  trees = 2000,
  min_n = tune()
) %>%
  set_mode("classification") %>% 
  set_engine("ranger")

# Create workflow
tune_wf <- workflow() %>%
  add_recipe(tree_rec) %>%
  add_model(tune_spec)

#Cross validation
set.seed(234)
trees_folds <- vfold_cv(trees_train, v = 10, strata = outcomevar)

#Define performance metrics
mer_metrics <- metric_set(yardstick::recall,
                          yardstick::specificity,
                          yardstick::accuracy,
                          yardstick::precision,
                          yardstick::roc_auc,
                          yardstick::pr_auc)

# Parallel computing 
doParallel::registerDoParallel()
conflicted::conflicts_prefer(base::intersect)
conflicted::conflicts_prefer(base::setdiff)

set.seed(345)
tune_res <- tune_grid(
  tune_wf,
  metrics = mer_metrics,
  resamples = trees_folds,
  grid = 20
)
```

Assess metrics

```{r fig.height=10, fig.width=10}
# Assess data ----

#Metrics
metrics_tuneres = tune_res %>%
  collect_metrics() %>%
  select(mean, min_n, mtry, .metric) %>%
  pivot_longer(min_n:mtry,
    values_to = "value",
    names_to = "parameter") %>%
  ggplot()+
  aes(x = value, y = mean, colour = .metric, label = value) +
  geom_line() +
  geom_label() +
  scale_color_npg()+
  theme_few() +
  theme(axis.text.x = element_text(color = "black", size = 12)) +
  facet_grid(vars(.metric), vars(parameter), scales = "free")

metrics_tuneres

metrics_tuneres %>% ggsave(file = here("Figures",
                                       paste0(filename, 
                                              outcomevar,
                                              "_tuneres_metrics.png")),
                                        width = 10,
                                        height = 10)

min_n_selec = c(3, 3) #Observations by node)
mtry_selec = c(16, 16) #Features/genes by tree)
```

Hyperparameter tuning

```{r}
# Tune hyperparameters ----
rf_grid <- grid_regular(
  min_n(range = c(min_n_selec)),
  mtry(range = c(mtry_selec)),
  levels = levels
)

set.seed(456)
regular_res <- tune_grid(
  tune_wf,
  metrics = mer_metrics,
  resamples = trees_folds,
  grid = rf_grid
)

# Assess performance -----
regular_res %>%
  collect_metrics()

regular_res %>%
  collect_metrics() %>% 
  write_csv(file = here("Machine learning", paste("metrics_finalmodel_", filename, outcomevar, ".csv")))

#Metrics
regular_res_metrics = regular_res %>%
  collect_metrics() %>%
  select(mean, min_n, mtry, .metric) %>%
  pivot_longer(min_n:mtry,
    values_to = "value",
    names_to = "parameter") %>%
  ggplot()+
  aes(x = value, y = mean, colour = .metric, label = value) +
  geom_line() +
  geom_label(size = 3) +
  scale_color_npg() +
  theme_minimal() +
  facet_grid(vars(.metric), vars(parameter), scales = "free")

regular_res_metrics

# ggsave(regular_res_metrics, file = here("Figures", 
#                                         paste(filename, outcomevar, "metrics", today(), ".png")), 
#        width = 10, height = 10)
```


```{r}
# Select best model -----
best_auc <- select_best(regular_res, metric = "accuracy")
final_rf <- finalize_model(
  tune_spec,
  best_auc
)

```

Important features

```{r fig.height=5, fig.width=5}
final = final_rf %>%
  set_engine("ranger", importance = "permutation") %>%
  fit(outcomevar ~ .,
    data = juice(tree_prep)) %>% 
  vip(num_features = 20)

genes_rf = final$data
genes_rf = genes_rf %>% select(genes = Variable, importance = Importance)
genes_rf %>% write_csv(file = here("Figures", paste0("vipgenes_tidymodels_", filename, outcomevar, ".csv")))


genes_rf_10 = genes_rf %>% 
  arrange(-importance) %>% 
  slice_head(n = 10)

#Importance plot
importance_plot = ggplot(genes_rf_10) +
  aes(x = reorder(genes, importance),
    y = importance,
    fill = importance,
      weight = importance) +
  geom_col(width = 0.6,
           size = 0.5) +
  theme_few() +
  xlab("Genes") + 
  ylab("importance") +
  ylim(0, 0.16) +
  geom_text(aes(label = round(importance, 4), y = importance),
            hjust = -0.2, 
            size = 4,
            angle = 0,
            color = "black")+
  scale_fill_gradient(low = "#4DBBD5FF", high = "#DC0000FF") +
  theme(legend.position = "none",
        plot.subtitle = element_text(hjust = 0),
        axis.text.x = element_text(angle = 0, vjust = 0.5, hjust=0.5, size = 12, color = "black"),
        axis.text.y = element_text(size = 12, angle = 0, color = "black"),
        panel.grid.minor.y = element_blank(),
        legend.text = element_text(size=12),
        legend.title = element_text(size = 12, face = "bold")) +
    coord_flip() +
  labs(title = "Top 10 VIP genes",
       subtitle = "Random forest classification model for \nVaccine types vs. Naive",
       x = "",
       y = "Importance",
       color = "Importance") 
importance_plot
ggsave(importance_plot,
        file = here("Figures",
                    paste0("vipgenes_tidymodels_", filename, outcomevar,  ".png")),
        width = 5,
        height = 5)

importance_plot

```

 Train final model with tuned hyperparameters

```{r fig.height=3, fig.width=5}
# Train final model with tuned hyperparameters
final_wf <- workflow() %>%
  add_recipe(tree_rec) %>%
  add_model(final_rf)

final_res <- final_wf %>%
  last_fit(split)

final_wf_extracted = workflow() %>%
  add_recipe(tree_rec) %>%
  add_model(final_rf) %>% 
  last_fit(split) %>% 
  extract_workflow()

final_res_preds = final_res %>%
  collect_predictions() %>%
  mutate(correct = case_when(
    outcomevar == .pred_class ~ "Correct",
    TRUE ~ "Incorrect"
  )) %>%
  bind_cols(trees_test)


#Confusion matrix
res_conf_mat = conf_mat(final_res_preds, truth = outcomevar...9,
         estimate = .pred_class)

#Accuracy
res_acc = accuracy(final_res_preds, truth = outcomevar...9,
         estimate = .pred_class)

#Specificity
res_spec = spec(final_res_preds, truth = outcomevar...9,
         estimate = .pred_class)

#Precision
res_prec = precision(final_res_preds, truth = outcomevar...9,
         estimate = .pred_class)

#Recall
res_recall = recall(final_res_preds, truth = outcomevar...9,
         estimate = .pred_class)
#Precision recall AUC
# res_pr_auc = pr_auc(final_res_preds, outcomevar...9, .pred_RNA)


#Unite results
final_res_metrics = bind_rows(res_acc,
                         res_recall,
                         res_prec,
                         # res_pr_auc,
                         res_spec) %>% 
  select(-".estimator")


final_res_metrics_conf = list(res_conf_mat, 
                         final_res_metrics,
                         final_res_preds)

final_res_metrics_conf

final_res_metrics_conf %>% 
  saveRDS(file = here("Machine learning", paste0("final_res_metrics", filename, outcomevar, ".rds")))


#Plot
final_model_metrics = final_res_metrics %>% 
  rename(metric = .metric,
         estimate = .estimate) %>% 
  mutate(metric = fct_relevel(metric, final_res_metrics$.metric)) %>% 
  ggplot() +
  aes(x = estimate,
    y = fct_rev(metric),
    fill = estimate,
    weight = estimate) +
  geom_col(size = 0.5,
           width = 0.6) +
  theme_few() +
  geom_text(aes(label = round(estimate, 3), y = metric),
            hjust = -0.2, 
            size = 4,
            angle = 0,
            color = "black") +
  xlim(0, 1.2) +
  scale_fill_gradient(low = "#4DBBD5FF", high = "#DC0000FF") +
  theme(legend.position = "none",
        plot.subtitle = element_text(hjust = 0),
        axis.text.x = element_text(angle = 0, vjust = 0.5, hjust=0.5, size = 12, color = "black"),
        axis.text.y = element_text(size = 12, angle = 0, color = "black"),
        panel.grid.minor.y = element_blank(),
        legend.text = element_text(size=12),
        legend.title = element_text(size = 12, face = "bold")) +
  labs(title = "Model assessment",
       subtitle = "Random forest classification model for \nVaccine types vs. Naive",
       x = "Estimate",
       y = "",
       fill = "Estimate")

ggsave(final_model_metrics,
        file = here("Figures",
                    paste0("final_model_metrics_", filename, outcomevar,  ".png")),
        width = 5,
        height = 3)


#Which are the incorrect?
final_res_preds %>% rownames_to_column("sample") %>%  
  inner_join(ann_vaccines_samples %>% 
               select(condition, sample), by = "sample") %>% 
  filter(correct == "Incorrect") %>% 
  select(sample, condition, everything()) %>% 
  view()

final_model_metrics

#Matrix counts of VIP genes
all_matrices_counts_genesRF = all_samples_l2fc %>% 
  inner_join(genes_rf %>% 
               select(genes), by = "genes")
```

Multi ROC
```{r fig.height=5, fig.width=7}
multi_roc = final_res_preds %>% 
  group_by(id) %>%
  roc_curve(outcomevar...9, .pred_I:.pred_VV) %>%
  ggplot(aes(1 - specificity, sensitivity, color = .level)) +
  geom_abline(lty = 2, color = "gray80", size = 0.5) +
  geom_path(show.legend = FALSE, alpha = 0.6, size = 1.2) +
  facet_wrap(~.level, ncol = 3) +
  # coord_equal() +
  theme_few() +
    theme(axis.text = element_text(size = 10, color = "black"), 
          panel.spacing.x = unit(2, "lines")) +
  labs(title ="Receiver-operator curve",
       subtitle = "Multiclass classification, Randomforest",
       x = "1-Specificity",
       y = "Sensitivity") +
  scale_colour_manual(values = c(ann_vaxsig_colors$type[names(ann_vaxsig_colors$type) != "H"], 
                                 H = "black"))

multi_roc
```

Confusion matrix
```{r fig.height=4, fig.width=4}
conf_matrix_heat = final_res_preds %>% 
  conf_mat(outcomevar...9, .pred_class) %>%
  autoplot(type = "heatmap") +
    theme_few() +
  theme(axis.text = element_text(size = 10, color = "black"),
        legend.position = "none") +
  scale_fill_gradient(low = "white", high = "#DC0000FF") +
  labs(title ="Confusion matrix",
       subtitle = "Multiclass classification, Randomforest",
       x = "Truth",
       y = "Prediction")

conf_matrix_heat
```


```{r fig.height=5, fig.width=12}
conf_matrix_heat + multi_roc +
  plot_layout(widths = c(1, 2))
  
```

Precision recall curve
```{r}
# Precision recall
pr_curve_res = pr_curve(final_res_preds, outcomevar...9, .pred_I) %>% 
  ggplot(aes(x = recall, y = precision)) +
  geom_path(color = "#DC0000FF",
            size = 2) +
  coord_equal() +
  theme_few() +
  labs(title = "Precision-Recall curve",
       subtitle = "Random forest classification \nmodel for Infection vs. Vaccination",
       x = "Recall",
       y = "Precision") 


pr_curve_res %>% 
  ggsave(file = here("Figures",
                    paste0("PR_Curve_tidymodels_", filename, outcomevar,  ".png")),
         width = 4,
         height = 3)
  
```


Relative effects

Function
```{r fig.height=5, fig.width=7}
REffect <- function(data, dep_vars, indep_var, n_boot = 1000, 
                    plot = TRUE, plot_colors = c("red", "darkblue"), 
                    plot_labels = list(y = "Relative effect", x = "Genes"), 
                    plot_theme = theme_few(), plot_legend_position = "top", 
                    plot_legend_labels = c("Vaccination", "Infection")) {
  
  library(npmv)
  library(ggplot2)
  library(reshape2)
  
  if(!all(data[[indep_var]] %in% c(0, 1))) {
    stop("A variável independente precisa conter apenas 0 e 1 (dois grupos).")
  }
  
  n_aab <- nrow(data)
  list_boot_aab <- list()
  
  # Loop de Bootstrap
  for (i in 0:n_boot) {
    if (i == 0) {
      df_sample <- 1:n_aab
    } else {
      df_sample <- sample(1:n_aab, n_aab, replace = TRUE)
    }
    
    formula_str <- paste(dep_vars, collapse = " | ")
    formula <- as.formula(paste(formula_str, "~", indep_var))
    
    manova_aab <- nonpartest(formula, data = data[df_sample, ], permtest = FALSE, plots = FALSE,
                             permreps = 1000, tests = c(1, 0, 0, 0))
    
    if (i == 0) obs_aab <- manova_aab
    
    list_boot_aab[[i + 1]] <- manova_aab
    
    cat("ANOVA/ boot/ aab/ i = ", i, "\n")
  }
  
  obs_manova <- list_boot_aab[[1]]$twogroupreleffects
  row_manova <- rownames(obs_manova)
  
  releffects <- lapply(list_boot_aab[-1], "[[", 2)
  
  df_perc_boot <- lapply(seq_len(nrow(obs_manova)), function(i) {
    q <- sapply(seq_len(ncol(obs_manova)), function(j) {
      v <- sapply(releffects, function(r) {
        row <- rownames(r)
        if(row[1] != row_manova[1]) r <- r[2:1,]
        r[i, j]
      })
      quantile(v, probs = c(0.025, 0.975))
    })
    colnames(q) <- colnames(obs_manova)
    df <- data.frame(Var2 = colnames(q), t(q),
                     group = ifelse(i == 1, row_manova[1], row_manova[2]))
    df$Var2 <- factor(df$Var2, levels = sort(unique(df$Var2)), ordered = TRUE)
    return(df)
  })
  
  names(df_perc_boot) <- row_manova
  
  df <- reshape2::melt(
    cbind(obs_manova, group = row_manova),
    id.vars = "group"
  )
  df$Var1 <- "point_est"
  df <- df[, c(4, 2, 3, 1)]
  colnames(df)[2] <- "Var2"
  
  df <- dplyr::arrange(df, group, desc(value), Var2)
  df$Var2 <- as.character(df$Var2)
  df$Var2 <- factor(df$Var2, levels = unique(df$Var2), ordered = TRUE)
  
  for (k in seq_len(2)) {
    df_perc_boot[[k]]$Var2 <- factor(df_perc_boot[[k]]$Var2, levels = levels(df$Var2), ordered = TRUE)
  }
  
  if (plot) {
    p <- ggplot(data = df, aes(x = Var2, y = value, group = group, color = group)) +
      geom_ribbon(inherit.aes = FALSE, data = df_perc_boot[[1]],
                  aes(x = Var2, ymin = X2.5., ymax = X97.5., group = group), fill = plot_colors[1],
                  alpha = 0.3) +
      geom_ribbon(inherit.aes = FALSE, data = df_perc_boot[[2]],
                  aes(x = Var2, ymin = X2.5., ymax = X97.5., group = group), fill = plot_colors[2],
                  alpha = 0.3) +
      geom_line() +
      geom_point(aes(size = value)) +
      scale_colour_manual("", values = c("1" = plot_colors[2], "0" = plot_colors[1]),
                          labels = plot_legend_labels) +
      plot_theme +
      theme(
        legend.text = element_text(size = 12),
        axis.text.x = element_text(size = 12, angle = 90),
        axis.text.y = element_text(size = 12),
        axis.line.y = element_blank(),
        axis.title = element_text(size = 15),
        panel.grid.major = element_line(),
        legend.position = plot_legend_position,
        legend.box.background = element_rect()
      ) +
      labs(y = plot_labels$y, x = plot_labels$x)
    
    print(p)
  }
  
  return(list(obs_manova = obs_manova, df_perc_boot = df_perc_boot, plot = p))
}
```

Plot
```{r fig.height=5, fig.width=7, message=FALSE, warning=FALSE}
head(all_matrices_counts_genesRF)

covid_vac_releff = all_matrices_counts_genesRF %>% 
  select(genes, sample, counts) %>% 
  pivot_wider(names_from = "genes",
              values_from = "counts") %>% 
  inner_join(all_matrices_counts_genesRF %>% 
               select(sample, disease_vac), by = "sample") %>% 
  mutate(disease_vac = if_else(disease_vac == "I", 1, 0))


releffect_disease_vac = REffect(covid_vac_releff, dep_vars = colnames(covid_vac_releff %>% 
                                              select(-disease_vac, -sample)),
        indep_var = "disease_vac", plot_colors = c("red", "#4DBBD5FF"), 
        n_boot=1000)


plot = releffect_disease_vac$plot +
  scale_colour_manual(values = c("1" = "#DC0000FF", 
                                 "0" = "#4DBBD5FF"),
                      labels = plot_legend_labels) +
  geom_ribbon(inherit.aes = FALSE, data = releffect_disease_vac$df_perc_boot[[1]],
                  aes(x = Var2, ymin = X2.5., ymax = X97.5., group = group), fill = "#4DBBD5FF",
                  alpha = 0.3) +
      geom_ribbon(inherit.aes = FALSE, data = releffect_disease_vac$df_perc_boot[[2]],
                  aes(x = Var2, ymin = X2.5., ymax = X97.5., group = group), fill = "#DC0000FF",
                  alpha = 0.3) +
  theme(
        legend.text = element_text(size = 12),
        axis.text.x = element_text(size = 12, angle = 45, hjust = 1, vjust = 1, color = "black"),
        axis.text.y = element_text(size = 12, color = "black"),
        axis.line.y = element_blank(),
        axis.title = element_text(size = 15),
        panel.grid.major = element_line(),
        legend.position = "top",
        plot.title = element_text(hjust = 0.5, size = 20),
        plot.subtitle = element_text(hjust = 0.5, size = 15),
        legend.box.background = element_blank()
      ) +
      labs(y = "Relative effect", x = "Genes",
           title = "Relative effects of VIP genes",
           subtitle = "Infection vs. Vaccination",
           color = "") +
  ylim(0, 1)

plot

```



#### Heatmap: Gene L2FC per condition

```{r}
# Required files
ImmuneGO_Annotated_Genes = readRDS(here("VaxGO gene sets","ImmuneGO_Annotated_Genes_2024-03-26.rds"))
all_degs_p_05_vac_infected = readRDS(here("DGE analysis", "all_studies_degs_weighted.rds")) %>% 
  filter(if_else(n >= 5, fdr <= 0.05, fdr <= (0.05 * 6/n))) 

ann_vaccines <- read_csv(here("Annotations and metadata", "ann_vaccines_conditions_23-10-24.csv"))
# df = read_csv("Figures/vipgenes_tidymodels_ImmuneGO_Genes_Types_type.csv") %>% 
#   slice_head(n=10)
#Inputs
ImmuneGO_Annotated_Genes_all = ImmuneGO_Annotated_Genes %>% 
  filter(!process %in% c("TCR REPERTOIRE", "BCR REPERTOIRE"))

ImmuneGO_Genes = all_degs_p_05_vac_infected %>% 
  inner_join(genes_rf %>% 
               slice_head(n = 15) %>% 
               select(genes), by = "genes") %>% 
  inner_join(ann_vaccines %>% filter(disease_vac != "I"), by = "condition") %>% 
  # inner_join(ImmuneGO_Annotated_Genes %>% 
  #              select(genes, process) %>% 
  #              distinct() %>% 
  #              select(genes),
  #              by = "genes") %>% 
  distinct() %>% 
  select(genes, condition, log2fold_change) %>% 
  pivot_wider(names_from = condition, values_from = log2fold_change)

# Matrix
l2fc_imps = ImmuneGO_Genes %>% 
  column_to_rownames("genes") %>% 
  as.matrix() %>% 
  replace(is.na(.), 0)


####### INPUT
data_2 = l2fc_imps
outcomevar = "VaccineType"
filename_2 = paste0("RandomForest_ImmuneGO_Genes_Conditions_", outcomevar)

######## Matrix
matrix = ImmuneGO_Genes

######## Annotations
ann_vaccines_covid_other_2 = ImmuneGO_Genes %>% 
  column_to_rownames("genes") %>% 
  colnames() %>% 
  as.data.frame() %>% 
  select(condition = ".") %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  distinct() %>% 
  select(condition, disease_vac, type, week, day) %>% 
  arrange(week, day) %>% 
  mutate(week = fct_relevel(as.factor(week), sort(levels(week))),
         day = fct_relevel(as.factor(day), sort(levels(day)))) 

```


```{r fig.height=10, fig.width=15}
#Columns
ann_cols_heatmap = matrix %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  inner_join(ann_vaccines_covid_other_2, by = "condition") %>% 
  distinct() %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  mutate(day = as.integer(day))

matrix_data = matrix %>% 
  column_to_rownames("genes") %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  inner_join(ann_cols_heatmap %>% rownames_to_column("condition"), by = "condition") %>% 
  select(!c(disease_vac:day)) %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>%
  as.matrix()  %>% 
  replace(is.na(.), 0)

#Rows Immune ----
# ann_rows_immune = matrix_data %>%
#   as.data.frame() %>%
#   rownames_to_column("genes") %>%
#   left_join(ImmuneGO_Annotated_Genes %>% filter(go_term == "Manual"), by = "genes") %>%
#   select(genes, process) %>%
#   distinct() %>%
#   group_by(genes) %>%
#   arrange(genes, factor(process, levels = c("INNATE IMMUNE SYSTEM", "ADAPTIVE IMMUNE SYSTEM")), .by_group = TRUE) %>%
#   mutate(i_process = row_number()) %>%
#   pivot_wider(., names_from = "i_process", values_from = "process") %>%
#   column_to_rownames("genes") %>%
#   t() %>%
#   as.data.frame() %>%
#   rownames_to_column("index") %>%
#   mutate(index = as.numeric(index)) %>%
#   arrange(desc(index)) %>%
#   column_to_rownames("index") %>%
#   t() %>%
#   as.data.frame() %>%
#   replace(is.na(.), ".") %>% 
#   mutate(`1` = if_else(`1` == ".", "INNATE IMMUNE SYSTEM", `1`)) %>% 
#   rownames_to_column("genes") %>% 
#   mutate(genes = rownames(matrix_data)) %>% 
#   column_to_rownames("genes")

#Verificar dimensões do dataset
dim(matrix)
dim(matrix_data)
dim(ann_cols_heatmap)
dim(ann_rows_immune)

################# Immune 
#Heatmap annotation
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "right")

# row_ha = HeatmapAnnotation(df = ann_rows_immune,
#                        col = col_process_immune,
#                        which= "row",
#                        show_annotation_name = F,
#                        show_legend = F,
#                        simple_anno_size = unit(1, "mm"))

# Values colors
col_fun <- colorRamp2(c(-1, 0, 1), c("#4DBBD5FF", "white", "#ed6a5a"))

file = filename_2
mat = matrix_data
width_image = 30
height_image = 15
width = unit(6, "cm")
height = unit(6, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)
rect_gp = gpar(col = "white", lwd = 2)

fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap.png")

png(file = here("Figures", fileimageheatmap_grouped), 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        # left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "L2FC",
                        col = col_fun, #color
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = T,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(0, "cm"),
                        row_split = NULL,
                        column_split = NULL,
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(2, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height)

heatmap_plot = draw(heatmap_plot, 
     annotation_legend_side = "right", 
     heatmap_legend_side = "right")

dev.off()

heatmap_plot

```
#### Heatmap: Gene counts per sample

```{r}
# Required files
# ImmuneGO_Annotated_Genes = readRDS(here("VaxGO gene sets","ImmuneGO_Annotated_Genes_2024-03-26.rds"))
df = all_samples_l2fc %>% 
  inner_join(genes_rf %>% 
               arrange(-importance) %>% 
               slice_head(n=15) %>% 
               select(genes), by = "genes")

ann_vaccines_samples <- read_csv(here("Annotations and metadata", "ann_vaccines_samples.csv"))
ann_vaccines_samples = ann_vaccines_samples %>% 
   mutate(type = if_else(day == 0 & type == "I", "I", type),
         type = if_else(disease_vac == "I", "I", type),
         type = if_else(day == 0 & disease_vac != "I", "H", type),
         type = if_else(vaccine == "H", "H", type),
         # type = fct_relevel(type, "H"),
         outcomevar = type) %>% 
  select(condition, outcomevar, type, vaccine, everything()) %>% 
  filter(type != "H") %>% 
  filter(type != "I") %>% 
  mutate(type = fct_relevel(type, "I")) %>% 
  mutate(vaccine = if_else(condition == "BNT-MO (V1, D6)", "MO", vaccine)) %>% 
  mutate(type = if_else(type == "I" & vaccine != "I", "V-I", type),
         outcomevar = type) 

#Inputs

ImmuneGO_Genes = df %>% 
  # inner_join(ImmuneGO_Annotated_Genes %>% 
  #              select(genes, process) %>% 
  #              distinct() %>% 
  #              select(genes),
  #              by = "genes") %>% 
  distinct() %>% 
  select(genes, sample, log2fc) %>% 
  pivot_wider(names_from = sample, 
              values_from = log2fc,              
              values_fn = mean)


sample_correct =final_res_preds %>% 
  rownames_to_column("sample") %>%  
  inner_join(ann_vaccines_samples %>% 
               select(condition, sample), by = "sample") %>% 
  select(sample, type_pred = .pred_class, correct)

# Matrix
l2fc_imps = ImmuneGO_Genes %>% 
  column_to_rownames("genes") %>% 
  as.matrix() %>% 
  replace(is.na(.), 0)

####### INPUT
data_2 = l2fc_imps
filename_2 = paste0("RandomForest_ImmuneGO_Genes_Samples_", outcomevar)

######## Matrix
matrix = ImmuneGO_Genes

######## Annotations
ann_vaccines_covid_other_2 = ImmuneGO_Genes %>% 
  column_to_rownames("genes") %>% 
  colnames() %>% 
  as.data.frame() %>% 
  select(sample = ".") %>% 
  inner_join(ann_vaccines_samples, by = "sample") %>% 
  distinct() %>% 
  select(sample, condition, gse_id, type, vaccine, variant, dose,  day, sex, age) %>% 
  arrange(dose, day) %>% 
  mutate(day = fct_relevel(as.factor(day), sort(levels(day))),
         type = fct_relevel(type, "H", "I", "RNA", "VV", "SU", "IN")) %>% 
  inner_join(sample_correct, by = "sample") %>% 
  mutate(colnames = paste(condition, sample))

#Columns
ann_cols_heatmap = matrix %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(sample = '.') %>% 
  inner_join(ann_vaccines_covid_other_2, by = "sample") %>% 
  distinct() %>% 
  arrange(dose, day) %>% 
  column_to_rownames("colnames")

matrix_data = matrix %>% 
  column_to_rownames("genes") %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "sample") %>% 
  inner_join(ann_cols_heatmap %>% 
               rownames_to_column("colnames"), by = "sample") %>% 
  arrange(dose, day) %>% 
  select(!c(gse_id:age, sample, type_pred, correct, condition)) %>% 
  column_to_rownames("colnames") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>%
  as.matrix()  %>% 
  replace(is.na(.), 0)

ann_cols_heatmap = ann_cols_heatmap %>% 
  select(!c(condition, sample)) %>% 
  select(correct, 
         # vaccine, 
         type, 
         type_pred,
         # gse_id, 
         dose, 
         day,
         sex, 
         age) %>% 
  mutate(day = as.integer(day))
  

#Verificar dimensões do dataset
dim(matrix_data)
dim(ann_cols_heatmap)
dim(ann_rows_immune)


#Column split
column_split = ann_cols_heatmap$type %>% as.character()

dend1 = cluster_between_groups(matrix_data, column_split)
#Heatmap annotation
ann_vaxsig_colors$age <- circlize::colorRamp2(c(0, 50, 100), c("white", "#4DBBD5FF", "black"))
# ann_vaxsig_colors$day <- circlize::colorRamp2(c(0, 50, 100), c("white", "#4DBBD5FF", "black"))
ann_vaxsig_colors$type_pred <- ann_vaxsig_colors$type
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "right")
min(matrix_data)
max(matrix_data)

# Values colors
col_fun <- colorRamp2(c(-25, 0, 25), c("#4DBBD5FF", "white", "#DC0000FF"))

file = filename_2
mat = matrix_data
width_image = 60
height_image = 20
width = unit(50, "cm")
height = unit(6, "cm")
column_names_gp = gpar(fontsize = 5)
row_names_gp = gpar(fontsize = 10)
rect_gp = gpar(col = "white", lwd = 2)

fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap.png")

png(file = here("Figures", fileimageheatmap_grouped), 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        # left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE, 
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "L2FC",
                        col = col_fun, #color
                        show_column_names = T, 
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(1, "cm"), 
                        column_dend_height =  unit(2, "cm"),
                        row_split = NULL,
                        column_split = 6,
                        # column_split = column_split,
                        row_gap = unit(2, "mm"),
                        cluster_columns = dend1,
                        # cluster_columns = F,
                        column_gap = unit(2, "mm"),
                        show_column_dend = T,
                        show_row_dend = T,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height)

heatmap_plot = draw(heatmap_plot, 
     annotation_legend_side = "left", 
     heatmap_legend_side = "left")

dev.off()

```
Clustered by type
```{r}
file = filename_2
mat = matrix_data
width_image = 60
height_image = 20
width = unit(50, "cm")
height = unit(6, "cm")
column_names_gp = gpar(fontsize = 5)
row_names_gp = gpar(fontsize = 10)
rect_gp = gpar(col = "white", lwd = 2)

fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap_Clustered.png")

png(file = here("Figures", fileimageheatmap_grouped), 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        # left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE, 
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "L2FC",
                        col = col_fun, #color
                        show_column_names = T, 
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(1, "cm"), 
                        column_dend_height =  unit(2, "cm"),
                        row_split = NULL,
                        column_split = 5,
                        # column_split = column_split,
                        row_gap = unit(2, "mm"),
                        cluster_columns = T,
                        column_gap = unit(2, "mm"),
                        show_column_dend = T,
                        show_row_dend = T,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height)

heatmap_plot = draw(heatmap_plot, 
     annotation_legend_side = "left", 
     heatmap_legend_side = "left")

dev.off()
```

```{r}
file = filename_2
mat = matrix_data
width_image = 60
height_image = 20
width = unit(50, "cm")
height = unit(6, "cm")
column_names_gp = gpar(fontsize = 5)
row_names_gp = gpar(fontsize = 10)
rect_gp = gpar(col = "white", lwd = 2)

fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap_Clustered_all.png")

png(file = here("Figures", fileimageheatmap_grouped), 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        # left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE, 
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "L2FC",
                        col = col_fun, #color
                        show_column_names = T, 
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(1, "cm"), 
                        column_dend_height =  unit(2, "cm"),
                        row_split = NULL,
                        column_split = 5,
                        # column_split = column_split,
                        row_gap = unit(2, "mm"),
                        cluster_columns = T,
                        column_gap = unit(2, "mm"),
                        show_column_dend = T,
                        show_row_dend = T,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height)

heatmap_plot = draw(heatmap_plot, 
     annotation_legend_side = "left", 
     heatmap_legend_side = "left")

dev.off()
```




### Scatterplot
```{r fig.height=15, fig.width=15, message=FALSE, warning=FALSE}
library(GGally)
genes_rf

selected_genes_rf = df_input %>% 
  select(outcomevar, age, sex, day, any_of(genes_rf %>% slice_head(n = 15) %>% 
                                                      pull(genes)))

ggpairs_rf = ggpairs(selected_genes_rf, columns = 5:10, 
        aes(color = outcomevar,
            alpha = 0.5))

ggpairs_rf = ggpairs_rf + 
  scale_color_manual(values = ann_vaxsig_colors$type) +
  scale_fill_manual(values = ann_vaxsig_colors$type) +
  theme_minimal() +
  theme(legend.position = "right",
        strip.text.y = element_text(color = "black", size = 12, face = "bold"),
        strip.text.x = element_text(color = "black", size = 12, face = "bold"),
        axis.text.x = element_text(color = "black", size = 10),
        axis.text.y = element_text(vjust = 0.5,
                                   size = 10,
                                   color = "black"),
        axis.line = element_line(size = 1,
                                 colour = "black",
                                 linetype = 1),
        axis.ticks = element_line(size = 0.5, color = "black"),
        panel.grid.major.y = element_line(size = 0.1,
                                          linetype = 2,
                                          color = "gray75"),
        panel.grid.major.x = element_line(size = 0.1,
                                          linetype = 2,
                                          color = "gray75"),
        panel.spacing.x = unit(2, "lines"))

ggpairs_rf

ggpairs_rf %>% 
  ggsave(filename = here("Figures", paste(file, "ggpairs_rf_selected_genes.png")), width = 15,
         height = 15)
  
```



#### Boxplot of VIP genes
```{r}
selected_genes_rf = df_input %>% 
  select(outcomevar, age, sex, day, any_of(genes_rf %>% slice_head(n = 15) %>% 
                                                      pull(genes)))

selected_genes_rf %>% 
  esquisser()

```







### PCA of samples with VIP genes
**Input**

```{r}
#INPUT
# data_genes = all_matrices_counts_genesRF 
data_genes = all_samples_l2fc %>% 
  filter(genes %in% c(genes_rf %>% 
                        filter(!genes %in% c("vaccine_BNT", "vaccine_ChAd", "vaccine_I", "vaccine_ZF2001", "day")) %>% 
                        # slice_head(n = ) %>% 
                        pull(genes)))


data_annotation = ann_vaccines_samples_filtered

#Convert long to wide table format.
#Dataframe with sample annotation

ann_vaccines_pca_matrix = data_genes %>% 
  select(sample, genes, log2fc) %>% 
  pivot_wider(names_from = sample, 
              values_from = log2fc,              
              values_fn = mean) %>% 
  replace(is.na(.), 0) %>%
  column_to_rownames("genes") %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column("sample") %>% 
  inner_join(data_annotation %>% 
               select(sample, condition, disease_vac, gse_id, vaccine, day, week, dose, infection, type), by = "sample") 

# Convert dataframe to matrix without annotation columns
ann_vaccines_pca_matrix_ready = ann_vaccines_pca_matrix %>% 
  column_to_rownames("sample") %>% 
  select(!condition:type) %>% 
  as.matrix()

# Delete columns with near 0 variance
# nearZeroVarCols <- nearZeroVar(ann_vaccines_pca_matrix_ready, saveMetrics = TRUE)
# ann_vaccines_pca_matrix_ready <- ann_vaccines_pca_matrix_ready[, !nearZeroVarCols$nzv]
pca_res <- prcomp(ann_vaccines_pca_matrix_ready, scale. = F)

# Correlation plot ----
# corr_matrix = cor(ann_vaccines_pca_matrix_ready %>% scale()) 
corr_matrix = cor(ann_vaccines_pca_matrix_ready)

var <- get_pca_var(pca_res)

# corrplot = ggcorrplot(corr_matrix, hc.order = TRUE) +
#   theme(axis.text.x = element_text(angle = 90, size = 5),
#         axis.text.y = element_text(size = 5))
# #Salvar
# ggsave(corrplot, file = paste0(filename, "_corrplot.png"),
#        width = 20, #Grande 20, pequeno 10
#        height= 20) #Grande 20, pequeno 10
# print(corrplot)

#PCA ----
data.pca = prcomp(corr_matrix) #PCA
summary(corr_matrix) #Retornar PCs

#Scree plot ----
scree_plot = fviz_eig(data.pca, 
                      addlabels = TRUE,
                      ylim = c(0, 70)) +
  geom_col(color = "#00AFBB", fill = "#00AFBB") +
  theme_classic()

#Save
ggsave(scree_plot, file = here("Figures", paste0(filename, "_screeplot.png")), width = 10, height = 3) 
print(scree_plot)

# Loadings plot ----
options(ggrepel.max.overlaps = Inf)

circle_contrib = fviz_pca_var(pca_res, col.var = "cos2",
                              gradient.cols = c("black", "#DC0000FF"),
                              select.var= list(cos2 = 20), 
                              repel = T, 
                              labelsize = 4,
                              col.circle = NA) +
  xlab("") + 
  ylab("") +
  theme_minimal() +
theme(panel.grid = element_blank(),  # Remover linhas de grade
        axis.text = element_blank(),   # Remover rótulos de texto dos eixos
        axis.ticks = element_blank()) 

#Save
ggsave(circle_contrib, file = here("Figures", paste0(filename, "_circle_contrib_wide.png")), width = 10, height = 5)

circle_contrib

#KNN classification -----
#Use the screeplot generated % of explained variance for each dimension
scree_plot #Dim 1 = 68.2%, Dim 2 = 15.9%


# PC1-PC2 ------
#Determine the number of clusters
pca_scores <- data.frame(pca_res$x[, 1:2])
fviz_nbclust(pca_scores,  
                     FUNcluster = kmeans,
                     method = "wss")

pcas = "PC1-PC2"

set.seed(666) # Set seed for randomization
cluster_model <- kmeans(pca_res$x[, 1:2], centers = 3)  #Set the number of clusters

ann_vaccines_pca_matrix$cluster <- as.factor(cluster_model$cluster)
```

```{r fig.height=7, fig.width=10, paged.print=FALSE}
# Condition with density plot ------
pca_plot_knn_cluster = autoplot(pca_res, 
        data = ann_vaccines_pca_matrix,
        colour = "gse_id")  +
  stat_ellipse(aes(color = cluster, fill = cluster), 
               geom = "polygon", 
               alpha = 0.2, 
               linetype = 1,
               size = 0.3,
               type = "t") +
  labs(title=paste0(filename, "_bycondition")) + 
  # xlab("PC1 (44.9%)") +
  # ylab("PC2 (15.9%)") +
    geom_hline(yintercept = 0, size = 0.3, color = "gray50", linetype = 4) +
  geom_vline(xintercept = 0, size = 0.3, color = "gray50", linetype = 4) +
  geom_point(aes(fill = type,
                 color = type),
             stroke = 0.2) +
   scale_color_manual(values = c("1" = "#8491B4FF", 
                                "2" = "#4DBBD5FF", 
                                "3" = "#DC0000FF",
                                ann_vaxsig_colors$type), 
                     guide = "none") +
  scale_fill_manual(values = c("1" = "#8491B4FF", 
                               "2" = "#4DBBD5FF", 
                               "3" = "#DC0000FF",
                               ann_vaxsig_colors$type),
                    guide = "none") + 
  guides(col="none") +
  theme_few() +
  theme(axis.text = element_text(size = 10, color = "black"),
      panel.grid.minor = element_blank(),
      panel.grid.major = element_line(size = 1, 
                                    linetype = 1,
                                    color = "gray90"),
        axis.line = element_line(size = 0, colour = "black", linetype=1),
        # axis.line = element_line(size = 1, colour = "black", linetype=1),
        text = element_text(color = "black", size = 10),
        panel.border = element_rect(fill = NA, color = "gray90", size = 2)) 
  # geom_text_repel(aes(label = condition,
  #                     segment.colour="gray70"),
  #                 box.padding = 0.3,
  #                 size = 3) 

ggsave(pca_plot_knn_cluster, 
       file = here( "Figures", paste0(filename, pcas, "_KNN_Clustered_labels.png")), width = 10, height = 7)
pca_plot_knn_cluster
# pca_plot_knn_cluster_marginal = ggMarginal(
#   pca_plot_knn_cluster,
#   type = "histogram",
#   groupFill = T,
#   groupColour = T,
#   xparams = list(binwidth = 0.1, size = 0.1),
#   yparams = list(binwidth = 0.1, size = 0.1)
# )
# 
# ggsave(
#   pca_plot_knn_cluster_marginal,
#   file = here(paste0(filename, pcas, "_KNN_Clustered_labels_Marginal.png"), "Figures"),
#   width = 8,
#   height = 5
# )


```

### PCA of conditions with VIP genes
**Input**

```{r}
# Genes in long table format
all_degs_p_05_vac_infected = readRDS(here("DGE analysis", "all_studies_degs_weighted.rds")) %>% 
  filter(if_else(n >= 5, fdr <= 0.05, fdr <= (0.05 * 6/n))) 

# Sample or condition annotation
ann_vaccines <- read_csv(here("Annotations and metadata", "ann_vaccines_conditions.csv"))

data_annotation = ann_vaccines %>%
  mutate(day = as.factor(day),
         dose = as.factor(dose)) 


#INPUT
# data_genes = all_matrices_counts_genesRF 
data_genes = all_degs_p_05_vac_infected %>% 
  filter(genes %in% c(genes_rf %>% 
                        filter(!genes %in% c("vaccine_BNT", "vaccine_ChAd", "vaccine_I", "vaccine_ZF2001", "day")) %>% 
                        # slice_head(n = 10) %>% 
                        pull(genes)))


data_annotation = ann_vaccines_samples_filtered

#Convert long to wide table format.
#Dataframe with condition annotation

ann_vaccines_pca_matrix = data_genes %>% 
  select(condition, genes, log2fold_change) %>% 
  pivot_wider(names_from =condition, 
              values_from = log2fold_change,              
              values_fn = mean) %>% 
  replace(is.na(.), 0) %>%
  column_to_rownames("genes") %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column("condition") %>% 
  inner_join(data_annotation %>% 
               select(condition, condition, disease_vac, gse_id, vaccine, day, week, dose, infection, type), by = "condition") %>% 
  distinct()

# Convert dataframe to matrix without annotation columns
ann_vaccines_pca_matrix_ready = ann_vaccines_pca_matrix %>% 
  column_to_rownames("condition") %>% 
  select(!disease_vac:type) %>% 
  as.matrix()

# Delete columns with near 0 variance
nearZeroVarCols <- nearZeroVar(ann_vaccines_pca_matrix_ready, saveMetrics = TRUE)
ann_vaccines_pca_matrix_ready <- ann_vaccines_pca_matrix_ready[, !nearZeroVarCols$nzv]
pca_res <- prcomp(ann_vaccines_pca_matrix_ready, scale. = T)

# Correlation plot ----
# corr_matrix = cor(ann_vaccines_pca_matrix_ready %>% scale()) 
corr_matrix = cor(ann_vaccines_pca_matrix_ready)

var <- get_pca_var(pca_res)

# corrplot = ggcorrplot(corr_matrix, hc.order = TRUE) +
#   theme(axis.text.x = element_text(angle = 90, size = 5),
#         axis.text.y = element_text(size = 5))
# #Salvar
# ggsave(corrplot, file = paste0(filename, "_corrplot.png"),
#        width = 20, #Grande 20, pequeno 10
#        height= 20) #Grande 20, pequeno 10
# print(corrplot)

#PCA ----
data.pca = prcomp(corr_matrix) #PCA
summary(corr_matrix) #Retornar PCs

#Scree plot ----
scree_plot = fviz_eig(data.pca, 
                      addlabels = TRUE,
                      ylim = c(0, 70)) +
  geom_col(color = "#00AFBB", fill = "#00AFBB") +
  theme_classic()

#Save
ggsave(scree_plot, file = here("Figures", paste0(filename, "CONDITION_screeplot.png")), width = 10, height = 3) 
print(scree_plot)

# Loadings plot ----
options(ggrepel.max.overlaps = Inf)

circle_contrib = fviz_pca_var(pca_res, col.var = "cos2",
                              gradient.cols = c("black", "#DC0000FF"),
                              select.var= list(cos2 = 20), 
                              repel = T, 
                              labelsize = 4,
                              col.circle = NA) +
  xlab("") + 
  ylab("") +
  theme_minimal() +
theme(panel.grid = element_blank(),  # Remover linhas de grade
        axis.text = element_blank(),   # Remover rótulos de texto dos eixos
        axis.ticks = element_blank()) 

#Save
ggsave(circle_contrib, file = here("Figures", paste0(filename, "CONDITION_circle_contrib_wide.png")), width = 10, height = 5)

circle_contrib

#KNN classification -----
#Use the screeplot generated % of explained variance for each dimension


# PC1-PC2 ------
#Determine the number of clusters
pca_scores <- data.frame(pca_res$x[, 1:2])
fviz_nbclust(pca_scores,  
                     FUNcluster = kmeans,
                     method = "wss")

pcas = "PC1-PC2"

set.seed(666) # Set seed for randomization
cluster_model <- kmeans(pca_res$x[, 1:2], centers = 3)  #Set the number of clusters

ann_vaccines_pca_matrix$cluster <- as.factor(cluster_model$cluster)
```

```{r fig.height=7, fig.width=10, paged.print=FALSE}
# Condition with density plot ------
pca_plot_knn_cluster = autoplot(pca_res, 
        data = ann_vaccines_pca_matrix,
        colour = "gse_id")  +
  # stat_ellipse(aes(color = cluster, fill = cluster), 
  #              geom = "polygon", 
  #              alpha = 0.2, 
  #              linetype = 1,
  #              size = 0.3,
  #              type = "t") +
    geom_convexhull(aes(fill = cluster, colour = cluster), alpha = 0.2) +
  labs(title=paste0(filename, "_bycondition")) + 
  # xlab("PC1 (44.9%)") +
  # ylab("PC2 (15.9%)") +
    geom_hline(yintercept = 0, size = 0.3, color = "gray50", linetype = 4) +
  geom_vline(xintercept = 0, size = 0.3, color = "gray50", linetype = 4) +
  geom_point(aes(fill = type,
                 color = type),
             stroke = 0.2) +
   scale_color_manual(values = c("1" = "#8491B4FF", 
                                "2" = "#4DBBD5FF", 
                                "3" = "#DC0000FF",
                                ann_vaxsig_colors$type), 
                     guide = "none") +
  scale_fill_manual(values = c("1" = "#8491B4FF", 
                               "2" = "#4DBBD5FF", 
                               "3" = "#DC0000FF",
                               ann_vaxsig_colors$type),
                    guide = "none") + 
  guides(col="none") +
  theme_few() +
  theme(axis.text = element_text(size = 10, color = "black"),
      panel.grid.minor = element_blank(),
      panel.grid.major = element_line(size = 1, 
                                    linetype = 1,
                                    color = "gray90"),
        axis.line = element_line(size = 0, colour = "black", linetype=1),
        # axis.line = element_line(size = 1, colour = "black", linetype=1),
        text = element_text(color = "black", size = 10),
        panel.border = element_rect(fill = NA, color = "gray90", size = 2)) 
  # geom_text_repel(aes(label = condition,
  #                     segment.colour="gray70"),
  #                 box.padding = 0.3,
  #                 size = 3) 

ggsave(pca_plot_knn_cluster, 
       file = here( "Figures", paste0(filename, pcas, "CONDITION_KNN_Clustered_labels.png")), width = 10, height = 7)
pca_plot_knn_cluster
# pca_plot_knn_cluster_marginal = ggMarginal(
#   pca_plot_knn_cluster,
#   type = "histogram",
#   groupFill = T,
#   groupColour = T,
#   xparams = list(binwidth = 0.1, size = 0.1),
#   yparams = list(binwidth = 0.1, size = 0.1)
# )
# 
# ggsave(
#   pca_plot_knn_cluster_marginal,
#   file = here(paste0(filename, pcas, "_KNN_Clustered_labels_Marginal.png"), "Figures"),
#   width = 8,
#   height = 5
# )


```








##  COVID-19 vaccines:  Infection vs. infection with previous vaccination

```{r message=FALSE, warning=FALSE}
# install.packages("randomForest")
# install.packages("caret")
# install.packages("rpart")
# install.packages("rpart.plot")
# install.packages("tidymodels")
# install.packages("reticulate")
# install.packages("themis")
# install.packages("ranger")
# install.packages("workflowsets")
# install.packages("glmnet")
# install.packages("kknn")
# install.packages("nnet")
# install.packages("tictoc")

library(tidymodels)
tidymodels_prefer()
library(randomForest)
library(ranger)
library(caret)
library(reticulate)
library(themis)
library(tidyverse)
library(rpart)
library(rpart.plot)
library(here)
library(ggrepel)
library(ComplexHeatmap)
library(janitor)
library(ggsci)
library(vip)
library(here)
library(circlize)
library(workflowsets)
library(glmnet)
library(kknn)
library(nnet)
library(tictoc)
library(ggthemes)
tidymodels_prefer()

```

### Import files

```{r}
all_matrices_normalized_counts <- readRDS(here("DGE analysis", "all_matrices_normalized_counts.rds"))
all_matrices_counts_tpm_normalized <- readRDS(here("DGE analysis", "all_matrices_counts_tpm_normalized.rds"))
all_matrices_counts <- readRDS(here("DGE analysis", "all_matrices_counts.rds"))
all_samples_l2fc <- readRDS(here("DGE analysis", "all_samples_l2fc.rds"))
ImmuneGO_BTM_grouped_genes <- read_csv("VaxGO gene sets/ImmuneGO_BTM_grouped_genes.csv")
all_degs_p_05_vac_infected = readRDS(here("DGE analysis", "all_studies_degs_weighted.rds")) %>% 
    filter(if_else(n >= 5, fdr <= 0.05, fdr <= (0.05 * 6/n))) 
ann_vaccines <- read_csv(here("Annotations and metadata", "ann_vaccines_conditions.csv"))

ann_vaccines_samples = read_csv(here("Annotations and metadata", "ann_vaccines_samples.csv"))
PC_1_2_cos2 = read_csv(here("PCA", "PC_1_2_cos2.csv"))


dim(all_matrices_normalized_counts)
```


```{r}
# Input data ----
filename = "ImmuneGO_Genes_Vaccine_Types_L2FC"
df = all_samples_l2fc
# df = all_matrices_counts_tpm_normalized

## Vaccination types
outcomevar_selected= "type"

#New version
ann_vaccines_samples_filtered = ann_vaccines_samples %>%
  select(condition, outcomevar_selected, type, everything()) %>% 
  mutate(type = if_else(day == 0 & type == "I", "I", type),
         type = if_else(disease_vac == "I", "I", type),
         type = if_else(day == 0 & disease_vac != "I", "H", type),
         type = if_else(vaccine == "H", "H", type),
         # type = fct_relevel(type, "H"),
         outcomevar = type) %>% 
  select(condition, outcomevar, type, vaccine, everything()) %>% 
  filter(type != "H") %>% 
  mutate(type = fct_relevel(type, "I")) %>% 
  mutate(vaccine = if_else(condition == "BNT-MO (V1, D6)", "MO", vaccine)) %>% 
  mutate(type = if_else(type == "I" & vaccine != "I", "V-I", type),
         outcomevar = type) %>% 
  filter(!outcomevar %in% c("I"))


ann_vaccines_samples_filtered


#How many samples by level?
count_levels = ann_vaccines_samples_filtered %>% 
  dplyr::count(outcomevar)

count_levels 

#How many levels?
levels = nrow(count_levels)


#Select only important genes
COVID_Infec_Vac_ImmuneGO_BTM_PC_1_2_cos2 <- read_csv("PCA/COVID_Infec_Vac_ImmuneGO_BTM PC_1_2_cos2.csv")

PC1_2 = COVID_Infec_Vac_ImmuneGO_BTM_PC_1_2_cos2 %>% 
  filter(cos2>0.75) #232 genes

dim(PC1_2)
```

Train and test

```{r fig.height=10, fig.width=10}
# Prepare data ------

#L2FC
df_input1 <- df %>%
  select(genes, sample, log2fc) %>%
  distinct() %>% 
  inner_join(PC1_2 %>% select(genes) %>% distinct(), by = "genes") %>% 
  distinct() %>% 
  pivot_wider(values_from = "log2fc",
              names_from = "sample",
              values_fn = mean) %>% 
  column_to_rownames("genes") %>%
  t() %>%
  as.data.frame() 


df_input = df_input1 %>% 
  rownames_to_column("sample") %>%
  inner_join(ann_vaccines_samples_filtered %>% select(sample,
                                                      # participant_id,
                                                      day,
                                                      age, 
                                                      sex,
                                                      outcomevar,
                                                      # vaccine
                                                      ),
             by = "sample") %>%
  select(sample,
         outcomevar,
         age,
         sex,
         day,
         # vaccine,
         everything()) %>%
  column_to_rownames("sample") %>%
  mutate_if(is.character, factor) %>% 
  mutate(outcomevar = fct_relevel(outcomevar, "I"))

df_input
```


```{r fig.height=10, fig.width=10}
# glimpse(df_input)

# Split data
set.seed(123)  
split <- initial_split(df_input, prop = 0.6, strata = outcomevar)
trees_train <- training(split)
trees_test <- testing(split)

# Create recipe
tree_rec <- recipe(outcomevar ~ ., data = trees_train) %>% #OUTCOME
  step_dummy(all_nominal(), -all_outcomes()) %>%
  #step_corr(all_numeric_predictors()) %>% 
  step_upsample(outcomevar) #Upsample unbalanced data

tree_prep <- prep(tree_rec)

juiced <- juice(tree_prep) %>% 
  select(!colnames(df_input1), colnames(df_input1))

juiced

# Set random forest hyper parameters
tune_spec <- rand_forest(
  mtry = tune(),
  trees = 2000,
  min_n = tune()
) %>%
  set_mode("classification") %>% 
  set_engine("ranger")

# Create workflow
tune_wf <- workflow() %>%
  add_recipe(tree_rec) %>%
  add_model(tune_spec)

#Cross validation
set.seed(234)
trees_folds <- vfold_cv(trees_train, v = 10, strata = outcomevar)

#Define performance metrics
mer_metrics <- metric_set(yardstick::recall,
                          yardstick::specificity,
                          yardstick::accuracy,
                          yardstick::precision,
                          yardstick::roc_auc,
                          yardstick::pr_auc)

# Parallel computing 
doParallel::registerDoParallel()
conflicted::conflicts_prefer(base::intersect)
conflicted::conflicts_prefer(base::setdiff)

set.seed(345)
tune_res <- tune_grid(
  tune_wf,
  metrics = mer_metrics,
  resamples = trees_folds,
  grid = 20
)
```

Assess metrics

```{r fig.height=10, fig.width=10}
# Assess data ----

#Metrics
metrics_tuneres = tune_res %>%
  collect_metrics() %>%
  select(mean, min_n, mtry, .metric) %>%
  pivot_longer(min_n:mtry,
    values_to = "value",
    names_to = "parameter") %>%
  ggplot()+
  aes(x = value, y = mean, colour = .metric, label = value) +
  geom_line() +
  geom_label() +
  scale_color_npg()+
  theme_few() +
  theme(axis.text.x = element_text(color = "black", size = 12)) +
  facet_grid(vars(.metric), vars(parameter), scales = "free")

metrics_tuneres

metrics_tuneres %>% ggsave(file = here("Figures",
                                       paste0(filename, 
                                              outcomevar,
                                              "_tuneres_metrics.png")),
                                        width = 10,
                                        height = 10)

min_n_selec = c(3, 3) #Observations by node)
mtry_selec = c(36, 36) #Features/genes by tree)
```

Hyperparameter tuning

```{r}
# Tune hyperparameters ----
rf_grid <- grid_regular(
  min_n(range = c(min_n_selec)),
  mtry(range = c(mtry_selec)),
  levels = levels
)

set.seed(456)
regular_res <- tune_grid(
  tune_wf,
  metrics = mer_metrics,
  resamples = trees_folds,
  grid = rf_grid
)

# Assess performance -----
regular_res %>%
  collect_metrics()

regular_res %>%
  collect_metrics() %>% 
  write_csv(file = here("Machine learning", paste("metrics_finalmodel_", filename, outcomevar, ".csv")))

#Metrics
regular_res_metrics = regular_res %>%
  collect_metrics() %>%
  select(mean, min_n, mtry, .metric) %>%
  pivot_longer(min_n:mtry,
    values_to = "value",
    names_to = "parameter") %>%
  ggplot()+
  aes(x = value, y = mean, colour = .metric, label = value) +
  geom_line() +
  geom_label(size = 3) +
  scale_color_npg() +
  theme_minimal() +
  facet_grid(vars(.metric), vars(parameter), scales = "free")

regular_res_metrics

# ggsave(regular_res_metrics, file = here("Figures", 
#                                         paste(filename, outcomevar, "metrics", today(), ".png")), 
#        width = 10, height = 10)
```


```{r}
# Select best model -----
best_auc <- select_best(regular_res, metric = "accuracy")
final_rf <- finalize_model(
  tune_spec,
  best_auc
)

```

Important features

```{r fig.height=5, fig.width=5}
final = final_rf %>%
  set_engine("ranger", importance = "permutation") %>%
  fit(outcomevar ~ .,
    data = juice(tree_prep)) %>% 
  vip(num_features = 100)

genes_rf = final$data
genes_rf = genes_rf %>% select(genes = Variable, importance = Importance)
genes_rf %>% write_csv(file = here("Figures", paste0("vipgenes_tidymodels_", filename, outcomevar, ".csv")))


genes_rf_10 = genes_rf %>% 
  arrange(-importance) %>% 
  slice_head(n = 20)

#Importance plot
importance_plot = ggplot(genes_rf_10) +
  aes(x = reorder(genes, importance),
    y = importance,
    fill = importance,
      weight = importance) +
  geom_col(width = 0.6,
           size = 0.5) +
  theme_few() +
  xlab("Genes") + 
  ylab("importance") +
  ylim(0, 0.16) +
  geom_text(aes(label = round(importance, 4), y = importance),
            hjust = -0.2, 
            size = 4,
            angle = 0,
            color = "black")+
  scale_fill_gradient(low = "#4DBBD5FF", high = "#DC0000FF") +
  theme(legend.position = "none",
        plot.subtitle = element_text(hjust = 0),
        axis.text.x = element_text(angle = 0, vjust = 0.5, hjust=0.5, size = 12, color = "black"),
        axis.text.y = element_text(size = 12, angle = 0, color = "black"),
        panel.grid.minor.y = element_blank(),
        legend.text = element_text(size=12),
        legend.title = element_text(size = 12, face = "bold")) +
    coord_flip() +
  labs(title = "Top 10 VIP genes",
       subtitle = "Random forest classification model for \nVaccine types vs. Naive",
       x = "",
       y = "Importance",
       color = "Importance") 
importance_plot
ggsave(importance_plot,
        file = here("Figures",
                    paste0("vipgenes_tidymodels_", filename, outcomevar,  ".png")),
        width = 5,
        height = 5)

importance_plot

```

 Train final model with tuned hyperparameters

```{r fig.height=3, fig.width=5}
# Train final model with tuned hyperparameters
final_wf <- workflow() %>%
  add_recipe(tree_rec) %>%
  add_model(final_rf)

final_res <- final_wf %>%
  last_fit(split)

final_wf_extracted = workflow() %>%
  add_recipe(tree_rec) %>%
  add_model(final_rf) %>% 
  last_fit(split) %>% 
  extract_workflow()

final_res_preds = final_res %>%
  collect_predictions() %>%
  mutate(correct = case_when(
    outcomevar == .pred_class ~ "Correct",
    TRUE ~ "Incorrect"
  )) %>%
  bind_cols(trees_test)


#Confusion matrix
res_conf_mat = conf_mat(final_res_preds, truth = outcomevar...9,
         estimate = .pred_class)

#Accuracy
res_acc = accuracy(final_res_preds, truth = outcomevar...9,
         estimate = .pred_class)

#Specificity
res_spec = spec(final_res_preds, truth = outcomevar...9,
         estimate = .pred_class)

#Precision
res_prec = precision(final_res_preds, truth = outcomevar...9,
         estimate = .pred_class)

#Recall
res_recall = recall(final_res_preds, truth = outcomevar...9,
         estimate = .pred_class)
#Precision recall AUC
# res_pr_auc = pr_auc(final_res_preds, outcomevar...9, .pred_RNA)


#Unite results
final_res_metrics = bind_rows(res_acc,
                         res_recall,
                         res_prec,
                         # res_pr_auc,
                         res_spec) %>% 
  select(-".estimator")


final_res_metrics_conf = list(res_conf_mat, 
                         final_res_metrics,
                         final_res_preds)

final_res_metrics_conf

final_res_metrics_conf %>% 
  saveRDS(file = here("Machine learning", paste0("final_res_metrics", filename, outcomevar, ".rds")))


#Plot
final_model_metrics = final_res_metrics %>% 
  rename(metric = .metric,
         estimate = .estimate) %>% 
  mutate(metric = fct_relevel(metric, final_res_metrics$.metric)) %>% 
  ggplot() +
  aes(x = estimate,
    y = fct_rev(metric),
    fill = estimate,
    weight = estimate) +
  geom_col(size = 0.5,
           width = 0.6) +
  theme_few() +
  geom_text(aes(label = round(estimate, 3), y = metric),
            hjust = -0.2, 
            size = 4,
            angle = 0,
            color = "black") +
  xlim(0, 1.2) +
  scale_fill_gradient(low = "#4DBBD5FF", high = "#DC0000FF") +
  theme(legend.position = "none",
        plot.subtitle = element_text(hjust = 0),
        axis.text.x = element_text(angle = 0, vjust = 0.5, hjust=0.5, size = 12, color = "black"),
        axis.text.y = element_text(size = 12, angle = 0, color = "black"),
        panel.grid.minor.y = element_blank(),
        legend.text = element_text(size=12),
        legend.title = element_text(size = 12, face = "bold")) +
  labs(title = "Model assessment",
       subtitle = "Random forest classification model for \nVaccine types vs. Naive",
       x = "Estimate",
       y = "",
       fill = "Estimate")

ggsave(final_model_metrics,
        file = here("Figures",
                    paste0("final_model_metrics_", filename, outcomevar,  ".png")),
        width = 5,
        height = 3)


#Which are the incorrect?
final_res_preds %>% rownames_to_column("sample") %>%  
  inner_join(ann_vaccines_samples %>% 
               select(condition, sample), by = "sample") %>% 
  filter(correct == "Incorrect") %>% 
  select(sample, condition, everything()) %>% 
  view()

final_model_metrics

#Matrix counts of VIP genes
all_matrices_counts_genesRF = all_samples_l2fc %>% 
  inner_join(genes_rf %>% 
               select(genes), by = "genes")
```

Multi ROC
```{r fig.height=5, fig.width=7}
multi_roc = final_res_preds %>% 
  group_by(id) %>%
  roc_curve(outcomevar...12, .pred_IN:.pred_VV) %>%
  ggplot(aes(1 - specificity, sensitivity, color = .level)) +
  geom_abline(lty = 2, color = "gray80", size = 0.5) +
  geom_path(show.legend = FALSE, alpha = 0.6, size = 1.2) +
  facet_wrap(~.level, ncol = 3) +
  # coord_equal() +
  theme_few() +
    theme(axis.text = element_text(size = 10, color = "black"), 
          panel.spacing.x = unit(2, "lines")) +
  labs(title ="Receiver-operator curve",
       subtitle = "Multiclass classification, Randomforest",
       x = "1-Specificity",
       y = "Sensitivity") +
  scale_colour_manual(values = c(ann_vaxsig_colors$type[names(ann_vaxsig_colors$type) != "H"], 
                                 H = "black"))

multi_roc
```

Confusion matrix
```{r fig.height=4, fig.width=4}
conf_matrix_heat = final_res_preds %>% 
  conf_mat(outcomevar...12, .pred_class) %>%
  autoplot(type = "heatmap") +
    theme_few() +
  theme(axis.text = element_text(size = 10, color = "black"),
        legend.position = "none") +
  scale_fill_gradient(low = "white", high = "#DC0000FF") +
  labs(title ="Confusion matrix",
       subtitle = "Multiclass classification, Randomforest",
       x = "Truth",
       y = "Prediction")

conf_matrix_heat
```


```{r fig.height=5, fig.width=12}
conf_matrix_heat + multi_roc +
  plot_layout(widths = c(1, 2))
  
```


Relative effects

Function
```{r fig.height=5, fig.width=7}
REffect <- function(data, dep_vars, indep_var, n_boot = 1000, 
                    plot = TRUE, plot_colors = c("red", "darkblue"), 
                    plot_labels = list(y = "Relative effect", x = "Genes"), 
                    plot_theme = theme_few(), plot_legend_position = "top", 
                    plot_legend_labels = c("Vaccination", "Infection")) {
  
  library(npmv)
  library(ggplot2)
  library(reshape2)
  
  if(!all(data[[indep_var]] %in% c(0, 1))) {
    stop("A variável independente precisa conter apenas 0 e 1 (dois grupos).")
  }
  
  n_aab <- nrow(data)
  list_boot_aab <- list()
  
  # Loop de Bootstrap
  for (i in 0:n_boot) {
    if (i == 0) {
      df_sample <- 1:n_aab
    } else {
      df_sample <- sample(1:n_aab, n_aab, replace = TRUE)
    }
    
    formula_str <- paste(dep_vars, collapse = " | ")
    formula <- as.formula(paste(formula_str, "~", indep_var))
    
    manova_aab <- nonpartest(formula, data = data[df_sample, ], permtest = FALSE, plots = FALSE,
                             permreps = 1000, tests = c(1, 0, 0, 0))
    
    if (i == 0) obs_aab <- manova_aab
    
    list_boot_aab[[i + 1]] <- manova_aab
    
    cat("ANOVA/ boot/ aab/ i = ", i, "\n")
  }
  
  obs_manova <- list_boot_aab[[1]]$twogroupreleffects
  row_manova <- rownames(obs_manova)
  
  releffects <- lapply(list_boot_aab[-1], "[[", 2)
  
  df_perc_boot <- lapply(seq_len(nrow(obs_manova)), function(i) {
    q <- sapply(seq_len(ncol(obs_manova)), function(j) {
      v <- sapply(releffects, function(r) {
        row <- rownames(r)
        if(row[1] != row_manova[1]) r <- r[2:1,]
        r[i, j]
      })
      quantile(v, probs = c(0.025, 0.975))
    })
    colnames(q) <- colnames(obs_manova)
    df <- data.frame(Var2 = colnames(q), t(q),
                     group = ifelse(i == 1, row_manova[1], row_manova[2]))
    df$Var2 <- factor(df$Var2, levels = sort(unique(df$Var2)), ordered = TRUE)
    return(df)
  })
  
  names(df_perc_boot) <- row_manova
  
  df <- reshape2::melt(
    cbind(obs_manova, group = row_manova),
    id.vars = "group"
  )
  df$Var1 <- "point_est"
  df <- df[, c(4, 2, 3, 1)]
  colnames(df)[2] <- "Var2"
  
  df <- dplyr::arrange(df, group, desc(value), Var2)
  df$Var2 <- as.character(df$Var2)
  df$Var2 <- factor(df$Var2, levels = unique(df$Var2), ordered = TRUE)
  
  for (k in seq_len(2)) {
    df_perc_boot[[k]]$Var2 <- factor(df_perc_boot[[k]]$Var2, levels = levels(df$Var2), ordered = TRUE)
  }
  
  if (plot) {
    p <- ggplot(data = df, aes(x = Var2, y = value, group = group, color = group)) +
      geom_ribbon(inherit.aes = FALSE, data = df_perc_boot[[1]],
                  aes(x = Var2, ymin = X2.5., ymax = X97.5., group = group), fill = plot_colors[1],
                  alpha = 0.3) +
      geom_ribbon(inherit.aes = FALSE, data = df_perc_boot[[2]],
                  aes(x = Var2, ymin = X2.5., ymax = X97.5., group = group), fill = plot_colors[2],
                  alpha = 0.3) +
      geom_line() +
      geom_point(aes(size = value)) +
      scale_colour_manual("", values = c("1" = plot_colors[2], "0" = plot_colors[1]),
                          labels = plot_legend_labels) +
      plot_theme +
      theme(
        legend.text = element_text(size = 12),
        axis.text.x = element_text(size = 12, angle = 90),
        axis.text.y = element_text(size = 12),
        axis.line.y = element_blank(),
        axis.title = element_text(size = 15),
        panel.grid.major = element_line(),
        legend.position = plot_legend_position,
        legend.box.background = element_rect()
      ) +
      labs(y = plot_labels$y, x = plot_labels$x)
    
    print(p)
  }
  
  return(list(obs_manova = obs_manova, df_perc_boot = df_perc_boot, plot = p))
}
```

Plot
```{r fig.height=5, fig.width=7, message=FALSE, warning=FALSE}
head(all_matrices_counts_genesRF)

covid_vac_releff = all_matrices_counts_genesRF %>% 
  select(genes, sample, counts) %>% 
  pivot_wider(names_from = "genes",
              values_from = "counts") %>% 
  inner_join(all_matrices_counts_genesRF %>% 
               select(sample, disease_vac), by = "sample") %>% 
  mutate(disease_vac = if_else(disease_vac == "I", 1, 0))


releffect_disease_vac = REffect(covid_vac_releff, dep_vars = colnames(covid_vac_releff %>% 
                                              select(-disease_vac, -sample)),
        indep_var = "disease_vac", plot_colors = c("red", "#4DBBD5FF"), 
        n_boot=1000)


plot = releffect_disease_vac$plot +
  scale_colour_manual(values = c("1" = "#DC0000FF", 
                                 "0" = "#4DBBD5FF"),
                      labels = plot_legend_labels) +
  geom_ribbon(inherit.aes = FALSE, data = releffect_disease_vac$df_perc_boot[[1]],
                  aes(x = Var2, ymin = X2.5., ymax = X97.5., group = group), fill = "#4DBBD5FF",
                  alpha = 0.3) +
      geom_ribbon(inherit.aes = FALSE, data = releffect_disease_vac$df_perc_boot[[2]],
                  aes(x = Var2, ymin = X2.5., ymax = X97.5., group = group), fill = "#DC0000FF",
                  alpha = 0.3) +
  theme(
        legend.text = element_text(size = 12),
        axis.text.x = element_text(size = 12, angle = 45, hjust = 1, vjust = 1, color = "black"),
        axis.text.y = element_text(size = 12, color = "black"),
        axis.line.y = element_blank(),
        axis.title = element_text(size = 15),
        panel.grid.major = element_line(),
        legend.position = "top",
        plot.title = element_text(hjust = 0.5, size = 20),
        plot.subtitle = element_text(hjust = 0.5, size = 15),
        legend.box.background = element_blank()
      ) +
      labs(y = "Relative effect", x = "Genes",
           title = "Relative effects of VIP genes",
           subtitle = "Infection vs. Vaccination",
           color = "") +
  ylim(0, 1)

plot

```



#### Heatmap: Gene L2FC per condition



```{r}
# Required files
ImmuneGO_Annotated_Genes = readRDS(here("VaxGO gene sets","ImmuneGO_Annotated_Genes_2024-03-26.rds"))
all_degs_p_05_vac_infected = readRDS(here("DGE analysis", "all_studies_degs_weighted.rds")) %>% 
  filter(if_else(n >= 5, fdr <= 0.05, fdr <= (0.05 * 6/n))) 

ann_vaccines <- read_csv(here("Annotations and metadata", "ann_vaccines_conditions_23-10-24.csv"))
# df = read_csv("Figures/vipgenes_tidymodels_ImmuneGO_Genes_Types_type.csv") %>% 
#   slice_head(n=10)
#Inputs
ImmuneGO_Annotated_Genes_all = ImmuneGO_Annotated_Genes %>% 
  filter(!process %in% c("TCR REPERTOIRE", "BCR REPERTOIRE"))

ImmuneGO_Genes = all_degs_p_05_vac_infected %>% 
  inner_join(genes_rf %>% 
               slice_head(n = 15) %>% 
               select(genes), by = "genes") %>% 
  # inner_join(ImmuneGO_Annotated_Genes %>% 
  #              select(genes, process) %>% 
  #              distinct() %>% 
  #              select(genes),
  #              by = "genes") %>% 
  distinct() %>% 
  select(genes, condition, log2fold_change) %>% 
  pivot_wider(names_from = condition, values_from = log2fold_change)

# Matrix
l2fc_imps = ImmuneGO_Genes %>% 
  column_to_rownames("genes") %>% 
  as.matrix() %>% 
  replace(is.na(.), 0)


####### INPUT
data_2 = l2fc_imps
outcomevar = "VaccineType"
filename_2 = paste0("RandomForest_ImmuneGO_Genes_Conditions_", outcomevar)

######## Matrix
matrix = ImmuneGO_Genes

######## Annotations
ann_vaccines_covid_other_2 = ImmuneGO_Genes %>% 
  column_to_rownames("genes") %>% 
  colnames() %>% 
  as.data.frame() %>% 
  select(condition = ".") %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  distinct() %>% 
  select(condition, disease_vac, type, week, day) %>% 
  arrange(week, day) %>% 
  mutate(week = fct_relevel(as.factor(week), sort(levels(week))),
         day = fct_relevel(as.factor(day), sort(levels(day)))) 

```


```{r fig.height=10, fig.width=15}
#Columns
ann_cols_heatmap = matrix %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  inner_join(ann_vaccines_covid_other_2, by = "condition") %>% 
  distinct() %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  mutate(day = as.integer(day))

matrix_data = matrix %>% 
  column_to_rownames("genes") %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  inner_join(ann_cols_heatmap %>% rownames_to_column("condition"), by = "condition") %>% 
  select(!c(disease_vac:day)) %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>%
  as.matrix()  %>% 
  replace(is.na(.), 0)

#Rows Immune ----
# ann_rows_immune = matrix_data %>%
#   as.data.frame() %>%
#   rownames_to_column("genes") %>%
#   left_join(ImmuneGO_Annotated_Genes %>% filter(go_term == "Manual"), by = "genes") %>%
#   select(genes, process) %>%
#   distinct() %>%
#   group_by(genes) %>%
#   arrange(genes, factor(process, levels = c("INNATE IMMUNE SYSTEM", "ADAPTIVE IMMUNE SYSTEM")), .by_group = TRUE) %>%
#   mutate(i_process = row_number()) %>%
#   pivot_wider(., names_from = "i_process", values_from = "process") %>%
#   column_to_rownames("genes") %>%
#   t() %>%
#   as.data.frame() %>%
#   rownames_to_column("index") %>%
#   mutate(index = as.numeric(index)) %>%
#   arrange(desc(index)) %>%
#   column_to_rownames("index") %>%
#   t() %>%
#   as.data.frame() %>%
#   replace(is.na(.), ".") %>% 
#   mutate(`1` = if_else(`1` == ".", "INNATE IMMUNE SYSTEM", `1`)) %>% 
#   rownames_to_column("genes") %>% 
#   mutate(genes = rownames(matrix_data)) %>% 
#   column_to_rownames("genes")

#Verificar dimensões do dataset
dim(matrix)
dim(matrix_data)
dim(ann_cols_heatmap)
dim(ann_rows_immune)

################# Immune 
#Heatmap annotation
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "right")

# row_ha = HeatmapAnnotation(df = ann_rows_immune,
#                        col = col_process_immune,
#                        which= "row",
#                        show_annotation_name = F,
#                        show_legend = F,
#                        simple_anno_size = unit(1, "mm"))

# Values colors
col_fun <- colorRamp2(c(-1, 0, 1), c("#4DBBD5FF", "white", "#ed6a5a"))

file = filename_2
mat = matrix_data
width_image = 30
height_image = 15
width = unit(15, "cm")
height = unit(6, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)
rect_gp = gpar(col = "white", lwd = 2)

fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap.png")

png(file = here("Figures", fileimageheatmap_grouped), 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        # left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "L2FC",
                        col = col_fun, #color
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = T,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(0, "cm"),
                        row_split = NULL,
                        column_split = NULL,
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(2, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height)

heatmap_plot = draw(heatmap_plot, 
     annotation_legend_side = "right", 
     heatmap_legend_side = "right")

dev.off()

heatmap_plot

```
#### Heatmap: Gene counts per sample

```{r}
# Required files
# ImmuneGO_Annotated_Genes = readRDS(here("VaxGO gene sets","ImmuneGO_Annotated_Genes_2024-03-26.rds"))
df = all_samples_l2fc %>% 
  inner_join(genes_rf %>% 
               arrange(-importance) %>% 
               slice_head(n=15) %>% 
               select(genes), by = "genes")

ann_vaccines_samples <- read_csv(here("Annotations and metadata", "ann_vaccines_samples.csv"))
ann_vaccines_samples = ann_vaccines_samples %>% 
   mutate(type = if_else(day == 0 & type == "I", "I", type),
         type = if_else(disease_vac == "I", "I", type),
         type = if_else(day == 0 & disease_vac != "I", "H", type),
         type = if_else(vaccine == "H", "H", type),
         # type = fct_relevel(type, "H"),
         outcomevar = type) %>% 
  select(condition, outcomevar, type, vaccine, everything()) %>% 
  filter(type != "H") %>% 
  mutate(type = fct_relevel(type, "I")) %>% 
  mutate(vaccine = if_else(condition == "BNT-MO (V1, D6)", "MO", vaccine)) %>% 
  mutate(type = if_else(type == "I" & vaccine != "I", "V-I", type),
         outcomevar = type)

#Inputs

ImmuneGO_Genes = df %>% 
  # inner_join(ImmuneGO_Annotated_Genes %>% 
  #              select(genes, process) %>% 
  #              distinct() %>% 
  #              select(genes),
  #              by = "genes") %>% 
  distinct() %>% 
  select(genes, sample, log2fc) %>% 
  pivot_wider(names_from = sample, 
              values_from = log2fc,              
              values_fn = mean)


sample_correct =final_res_preds %>% 
  rownames_to_column("sample") %>%  
  inner_join(ann_vaccines_samples %>% 
               select(condition, sample), by = "sample") %>% 
  select(sample, type_pred = .pred_class, correct)

# Matrix
l2fc_imps = ImmuneGO_Genes %>% 
  column_to_rownames("genes") %>% 
  as.matrix() %>% 
  replace(is.na(.), 0)

####### INPUT
data_2 = l2fc_imps
filename_2 = paste0("RandomForest_ImmuneGO_Genes_Samples_", outcomevar)

######## Matrix
matrix = ImmuneGO_Genes

######## Annotations
ann_vaccines_covid_other_2 = ImmuneGO_Genes %>% 
  column_to_rownames("genes") %>% 
  colnames() %>% 
  as.data.frame() %>% 
  select(sample = ".") %>% 
  inner_join(ann_vaccines_samples, by = "sample") %>% 
  distinct() %>% 
  select(sample, condition, gse_id, type, vaccine, variant, dose,  day, sex, age) %>% 
  arrange(dose, day) %>% 
  mutate(day = fct_relevel(as.factor(day), sort(levels(day))),
         type = fct_relevel(type, "H", "I", "RNA", "VV", "SU", "IN")) %>% 
  inner_join(sample_correct, by = "sample") %>% 
  mutate(colnames = paste(condition, sample))

#Columns
ann_cols_heatmap = matrix %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(sample = '.') %>% 
  inner_join(ann_vaccines_covid_other_2, by = "sample") %>% 
  distinct() %>% 
  arrange(dose, day) %>% 
  column_to_rownames("colnames")

matrix_data = matrix %>% 
  column_to_rownames("genes") %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "sample") %>% 
  inner_join(ann_cols_heatmap %>% 
               rownames_to_column("colnames"), by = "sample") %>% 
  arrange(dose, day) %>% 
  select(!c(gse_id:age, sample, type_pred, correct, condition)) %>% 
  column_to_rownames("colnames") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>%
  as.matrix()  %>% 
  replace(is.na(.), 0)

ann_cols_heatmap = ann_cols_heatmap %>% 
  select(!c(condition, sample)) %>% 
  select(correct, 
         # vaccine, 
         type, 
         type_pred,
         # gse_id, 
         dose, 
         day,
         sex, 
         age, 
         variant) %>% 
  mutate(day = as.integer(day))
  

#Verificar dimensões do dataset
dim(matrix_data)
dim(ann_cols_heatmap)
dim(ann_rows_immune)


#Column split
column_split = ann_cols_heatmap$type %>% as.character()

dend1 = cluster_between_groups(matrix_data, column_split)
#Heatmap annotation
ann_vaxsig_colors$age <- circlize::colorRamp2(c(0, 50, 100), c("white", "#4DBBD5FF", "black"))
ann_vaxsig_colors$day <- circlize::colorRamp2(c(0, 50, 100), c("white", "#4DBBD5FF", "black"))
ann_vaxsig_colors$type_pred <- ann_vaxsig_colors$type
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "right")
min(matrix_data)
max(matrix_data)

# Values colors
col_fun <- colorRamp2(c(-25, 0, 25), c("#4DBBD5FF", "white", "#DC0000FF"))

file = filename_2
mat = matrix_data
width_image = 60
height_image = 20
width = unit(50, "cm")
height = unit(6, "cm")
column_names_gp = gpar(fontsize = 5)
row_names_gp = gpar(fontsize = 10)
rect_gp = gpar(col = "white", lwd = 2)

fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap.png")

png(file = here("Figures", fileimageheatmap_grouped), 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        # left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE, 
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "L2FC",
                        col = col_fun, #color
                        show_column_names = T, 
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(1, "cm"), 
                        column_dend_height =  unit(2, "cm"),
                        row_split = NULL,
                        column_split = 6,
                        # column_split = column_split,
                        row_gap = unit(2, "mm"),
                        cluster_columns = dend1,
                        # cluster_columns = F,
                        column_gap = unit(2, "mm"),
                        show_column_dend = T,
                        show_row_dend = T,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height)

heatmap_plot = draw(heatmap_plot, 
     annotation_legend_side = "left", 
     heatmap_legend_side = "left")

dev.off()

```
Clustered by type
```{r}
file = filename_2
mat = matrix_data
width_image = 60
height_image = 20
width = unit(50, "cm")
height = unit(6, "cm")
column_names_gp = gpar(fontsize = 5)
row_names_gp = gpar(fontsize = 10)
rect_gp = gpar(col = "white", lwd = 2)

fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap_Clustered.png")

png(file = here("Figures", fileimageheatmap_grouped), 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        # left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE, 
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "L2FC",
                        col = col_fun, #color
                        show_column_names = T, 
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(1, "cm"), 
                        column_dend_height =  unit(2, "cm"),
                        row_split = NULL,
                        column_split = 5,
                        # column_split = column_split,
                        row_gap = unit(2, "mm"),
                        cluster_columns = T,
                        column_gap = unit(2, "mm"),
                        show_column_dend = T,
                        show_row_dend = T,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height)

heatmap_plot = draw(heatmap_plot, 
     annotation_legend_side = "left", 
     heatmap_legend_side = "left")

dev.off()
```

```{r}
file = filename_2
mat = matrix_data
width_image = 60
height_image = 20
width = unit(50, "cm")
height = unit(6, "cm")
column_names_gp = gpar(fontsize = 5)
row_names_gp = gpar(fontsize = 10)
rect_gp = gpar(col = "white", lwd = 2)

fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap_Clustered_all.png")

png(file = here("Figures", fileimageheatmap_grouped), 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        # left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE, 
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "L2FC",
                        col = col_fun, #color
                        show_column_names = T, 
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(1, "cm"), 
                        column_dend_height =  unit(2, "cm"),
                        row_split = NULL,
                        column_split = 5,
                        # column_split = column_split,
                        row_gap = unit(2, "mm"),
                        cluster_columns = T,
                        column_gap = unit(2, "mm"),
                        show_column_dend = T,
                        show_row_dend = T,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height)

heatmap_plot = draw(heatmap_plot, 
     annotation_legend_side = "left", 
     heatmap_legend_side = "left")

dev.off()
```




### Scatterplot
```{r fig.height=15, fig.width=15, message=FALSE, warning=FALSE}
library(GGally)
genes_rf

selected_genes_rf = df_input %>% 
  select(outcomevar, age, sex, day, vaccine, any_of(genes_rf %>% slice_head(n = 15) %>% 
                                                      pull(genes)))

ggpairs_rf = ggpairs(selected_genes_rf, columns = 6:15, 
        aes(color = outcomevar,
            alpha = 0.5))

ggpairs_rf = ggpairs_rf + 
  scale_color_manual(values = ann_vaxsig_colors$type) +
  scale_fill_manual(values = ann_vaxsig_colors$type) +
  theme_minimal() +
  theme(legend.position = "right",
        strip.text.y = element_text(color = "black", size = 12, face = "bold"),
        strip.text.x = element_text(color = "black", size = 12, face = "bold"),
        axis.text.x = element_text(color = "black", size = 10),
        axis.text.y = element_text(vjust = 0.5,
                                   size = 10,
                                   color = "black"),
        axis.line = element_line(size = 1,
                                 colour = "black",
                                 linetype = 1),
        axis.ticks = element_line(size = 0.5, color = "black"),
        panel.grid.major.y = element_line(size = 0.1,
                                          linetype = 2,
                                          color = "gray75"),
        panel.grid.major.x = element_line(size = 0.1,
                                          linetype = 2,
                                          color = "gray75"),
        panel.spacing.x = unit(2, "lines"))

ggpairs_rf

ggpairs_rf %>% 
  ggsave(filename = here("Figures", paste(file, "ggpairs_rf_selected_genes.png")), width = 15,
         height = 15)
  
```






### PCA with VIP genes
**Input**

```{r}
#INPUT
# data_genes = all_matrices_counts_genesRF 
data_genes = all_samples_l2fc %>% 
  filter(genes %in% c(genes_rf %>% 
                        filter(!genes %in% c("vaccine_BNT", "vaccine_ChAd", "vaccine_I", "vaccine_ZF2001", "day")) %>% 
                        slice_head(n = 10) %>% 
                        pull(genes)))


data_annotation = ann_vaccines_samples_filtered

#Convert long to wide table format.
#Dataframe with sample annotation

ann_vaccines_pca_matrix = data_genes %>% 
  select(sample, genes, log2fc) %>% 
  pivot_wider(names_from = sample, 
              values_from = log2fc,              
              values_fn = mean) %>% 
  replace(is.na(.), 0) %>%
  column_to_rownames("genes") %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column("sample") %>% 
  inner_join(data_annotation %>% 
               select(sample, condition, disease_vac, gse_id, vaccine, day, week, dose, infection, type), by = "sample") 

# Convert dataframe to matrix without annotation columns
ann_vaccines_pca_matrix_ready = ann_vaccines_pca_matrix %>% 
  column_to_rownames("sample") %>% 
  select(!condition:type) %>% 
  as.matrix()

# Delete columns with near 0 variance
nearZeroVarCols <- nearZeroVar(ann_vaccines_pca_matrix_ready, saveMetrics = TRUE)
ann_vaccines_pca_matrix_ready <- ann_vaccines_pca_matrix_ready[, !nearZeroVarCols$nzv]
pca_res <- prcomp(ann_vaccines_pca_matrix_ready, scale. = T)

# Correlation plot ----
corr_matrix = cor(ann_vaccines_pca_matrix_ready %>% scale()) 
var <- get_pca_var(pca_res)

# corrplot = ggcorrplot(corr_matrix, hc.order = TRUE) +
#   theme(axis.text.x = element_text(angle = 90, size = 5),
#         axis.text.y = element_text(size = 5))
# #Salvar
# ggsave(corrplot, file = paste0(filename, "_corrplot.png"),
#        width = 20, #Grande 20, pequeno 10
#        height= 20) #Grande 20, pequeno 10
# print(corrplot)

#PCA ----
data.pca = prcomp(corr_matrix) #PCA
summary(corr_matrix) #Retornar PCs

#Scree plot ----
scree_plot = fviz_eig(data.pca, 
                      addlabels = TRUE,
                      ylim = c(0, 70)) +
  geom_col(color = "#00AFBB", fill = "#00AFBB") +
  theme_classic()

#Save
ggsave(scree_plot, file = here("Figures", paste0(filename, "_screeplot.png")), width = 10, height = 3) 
print(scree_plot)

# Loadings plot ----
options(ggrepel.max.overlaps = Inf)

circle_contrib = fviz_pca_var(pca_res, col.var = "cos2",
                              gradient.cols = c("black", "#DC0000FF"),
                              select.var= list(cos2 = 20), 
                              repel = T, 
                              labelsize = 4,
                              col.circle = NA) +
  xlab("") + 
  ylab("") +
  theme_minimal() +
theme(panel.grid = element_blank(),  # Remover linhas de grade
        axis.text = element_blank(),   # Remover rótulos de texto dos eixos
        axis.ticks = element_blank()) 

#Save
ggsave(circle_contrib, file = here("Figures", paste0(filename, "_circle_contrib_wide.png")), width = 10, height = 5)

circle_contrib

#KNN classification -----
#Use the screeplot generated % of explained variance for each dimension
scree_plot #Dim 1 = 68.2%, Dim 2 = 15.9%


# PC1-PC2 ------
#Determine the number of clusters
pca_scores <- data.frame(pca_res$x[, 1:2])
fviz_nbclust(pca_scores,  
                     FUNcluster = kmeans,
                     method = "wss")

pcas = "PC1-PC2"

set.seed(666) # Set seed for randomization
cluster_model <- kmeans(pca_res$x[, 1:2], centers = 5)  #Set the number of clusters

ann_vaccines_pca_matrix$cluster <- as.factor(cluster_model$cluster)
```

```{r fig.height=7, fig.width=10, paged.print=FALSE}
# Condition with density plot ------
pca_plot_knn_cluster = autoplot(pca_res, 
        data = ann_vaccines_pca_matrix,
        colour = "gse_id")  +
  stat_ellipse(aes(color = cluster, fill = cluster), 
               geom = "polygon", 
               alpha = 0.2, 
               linetype = 1,
               size = 0.3,
               type = "t") +
  labs(title=paste0(filename, "_bycondition")) + 
  xlab("PC1 (44.9%)") +
  ylab("PC2 (15.9%)") +
    geom_hline(yintercept = 0, size = 0.3, color = "gray50", linetype = 4) +
  geom_vline(xintercept = 0, size = 0.3, color = "gray50", linetype = 4) +
  geom_point(aes(fill = type,
                 color = type),
             stroke = 0.2) +
   scale_color_manual(values = c("1" = "#8491B4FF", 
                                "2" = "#4DBBD5FF", 
                                "3" = "#DC0000FF",
                                ann_vaxsig_colors$type), 
                     guide = "none") +
  scale_fill_manual(values = c("1" = "#8491B4FF", 
                               "2" = "#4DBBD5FF", 
                               "3" = "#DC0000FF",
                               ann_vaxsig_colors$type),
                    guide = "none") + 
  guides(col="none") +
  theme_few() +
  theme(axis.text = element_text(size = 10, color = "black"),
      panel.grid.minor = element_blank(),
      panel.grid.major = element_line(size = 1, 
                                    linetype = 1,
                                    color = "gray90"),
        axis.line = element_line(size = 0, colour = "black", linetype=1),
        # axis.line = element_line(size = 1, colour = "black", linetype=1),
        text = element_text(color = "black", size = 10),
        panel.border = element_rect(fill = NA, color = "gray90", size = 2)) 
  # geom_text_repel(aes(label = condition,
  #                     segment.colour="gray70"),
  #                 box.padding = 0.3,
  #                 size = 3) 

ggsave(pca_plot_knn_cluster, 
       file = here( "Figures", paste0(filename, pcas, "_KNN_Clustered_labels.png")), width = 10, height = 7)
pca_plot_knn_cluster
# pca_plot_knn_cluster_marginal = ggMarginal(
#   pca_plot_knn_cluster,
#   type = "histogram",
#   groupFill = T,
#   groupColour = T,
#   xparams = list(binwidth = 0.1, size = 0.1),
#   yparams = list(binwidth = 0.1, size = 0.1)
# )
# 
# ggsave(
#   pca_plot_knn_cluster_marginal,
#   file = here(paste0(filename, pcas, "_KNN_Clustered_labels_Marginal.png"), "Figures"),
#   width = 8,
#   height = 5
# )


```








# 10. Heatmaps - Genes - Immune sets

Import files

```{r }
############## Import files

# ImmuneGO 
ImmuneGO_Annotated_GO = read_csv(here("VaxGO gene sets", "ImmuneGO_Annotated_GO_2024_03_26.csv"))
ImmuneGO_Annotated_Genes = readRDS(here("VaxGO gene sets",  "ImmuneGO_Annotated_Genes_2024-03-26.rds"))

# Informations about gene sets
ImmuneGO = ImmuneGO_Annotated_GO %>% 
  select(process:immune_tissue, set_size) %>% 
  distinct()

# Filter only manual annotated gene sets
immunego_genes = ImmuneGO_Annotated_Genes %>%
  filter(go_term == "Manual") %>% 
  select(process,genes)

ann_vaccines = read_csv(here("Annotations and metadata", "ann_vaccines_conditions.csv")) %>% 
  mutate(week = fct_reorder(as.factor(week), week),
         day = fct_reorder(as.factor(day), day),
         dose = fct_reorder(as.factor(dose), dose))

# DEGs table
data.immunego = readRDS(here("DGE analysis", "all_studies_degs_weighted.rds")) %>% 
  filter(if_else(n >= 5, fdr <= 0.05, fdr <= (0.05 * 6/n))) 

#Colors
ann_colors = ann_vaxsig_colors 
```

Process data

```{r }
############## Process data

# All gene sets
Immune_GO_BTMgenes_All = data.immunego %>% 
  inner_join(ImmuneGO_BTM_genes, by = "genes") %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  clean_names()


# All gene sets
Immune_GO_genes_All = data.immunego %>% 
  merge(immunego_genes, by = "genes", all.x = F, all.y=F) %>% 
  merge(ann_vaccines, by = "condition", all.y=F) %>% 
  clean_names()

###### Other immune gene sets
Immune_GO_genes_general = Immune_GO_genes_All %>%
  filter(process == "GENERAL")
Immune_GO_genes_antimicrobial = Immune_GO_genes_All %>%
  filter(process == "ANTIMICROBIAL")
Immune_GO_genes_Innate = Immune_GO_genes_All %>%
 filter(process == "INNATE IMMUNE SYSTEM")
Immune_GO_genes_Adaptive = Immune_GO_genes_All %>%
 filter(process == "ADAPTIVE IMMUNE SYSTEM")

# Innate immune system
Immune_GO_genes_Inflammation = Immune_GO_genes_All %>%
  filter(process == "INFLAMMATION")
Immune_GO_genes_ARPP = Immune_GO_genes_All %>%
  filter(process == "ARPP")
Immune_GO_genes_IFN = Immune_GO_genes_All %>%
  filter(process == "ANTIVIRAL & IFN")
Immune_GO_genes_Neutrophil = Immune_GO_genes_All %>%
  filter(process == "NEUTROPHIL")
Immune_GO_genes_Eosinophil = Immune_GO_genes_All %>%
  filter(process == "EOSINOPHIL")
Immune_GO_genes_Complement = Immune_GO_genes_All %>%
  filter(process == "COMPLEMENT")
Immune_GO_genes_Monocytes = Immune_GO_genes_All %>%
  filter(process == "MONOCYTE")
Immune_GO_genes_DC = Immune_GO_genes_All %>%
  filter(process == "DENDRITIC CELL")
Immune_GO_genes_Macro = Immune_GO_genes_All %>%
  filter(process == "MACROPHAGE")
Immune_GO_genes_NK = Immune_GO_genes_All %>%
  filter(process == "NK CELL")
Immune_GO_genes_Mast = Immune_GO_genes_All %>%
  filter(process == "MAST CELL") %>%
  distinct()

# Adaptive immune system
Immune_GO_genes_Cellular = Immune_GO_genes_All %>%
  filter(process == "CELLULAR ADAPTIVE IMMUNE SYSTEM")
Immune_GO_genes_Humoral = Immune_GO_genes_All %>%
  filter(process == "HUMORAL ADAPTIVE IMMUNE SYSTEM")
Immune_GO_genes_Tcells = Immune_GO_genes_All %>%
  filter(process == "T CELL")
Immune_GO_genes_Th = Immune_GO_genes_All %>%
  filter(process == "T HELPER CELLS")
Immune_GO_genes_Bcells = Immune_GO_genes_All %>%
  filter(process == "B CELL")
Immune_GO_genes_Ig = Immune_GO_genes_All %>%
  filter(process == "IMMUNOGLOBULIN MEDIATED RESPONSE") %>%
  distinct()

```

### Interactive heatmap version - All immmune genes

```{r}
library(InteractiveComplexHeatmap)
library(maditr)

####### Input
heatmap_data = Immune_GO_BTMgenes_All
file = "Immune_GO_BTM_genes_All"

###### Process data

heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA") %>% 
  select(genes, condition, log2fold_change)

# Convert to wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

matrix_data_ready =  matrix_data %>% 
  round(2)

# Check if all vaccines are in the matrix and join annotations
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

# Join annotations
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 


#Rows Immune ----
ann_rows_immune = matrix_data %>%
  as.data.frame() %>%
  rownames_to_column("genes") %>%
  left_join(ImmuneGO_Annotated_Genes %>%
              filter(go_term == "Manual",
                     !process %in% c("TCR REPERTOIRE", "BCR REPERTOIRE")), by = "genes") %>%
  select(genes, process) %>%
  distinct() %>%
  group_by(genes) %>%
  arrange(genes, factor(process, levels = c("INNATE IMMUNE SYSTEM", "ADAPTIVE IMMUNE SYSTEM")), .by_group = TRUE) %>%
  mutate(i_process = row_number()) %>%
  pivot_wider(., names_from = "i_process", values_from = "process") %>%
  column_to_rownames("genes") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("index") %>%
  mutate(index = as.numeric(index)) %>%
  arrange(desc(index)) %>%
  column_to_rownames("index") %>%
  t() %>%
  as.data.frame() %>%
  replace(is.na(.), ".")

# Reorder columns (same order as lines in the annotation table)
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% 
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0) %>% 
  as.matrix() 


#Check dimensions
dim(matrix_data_ordered) 
dim(ann_vaccines_heatmap) 
dim(ann_rows_immune)
```


```{r}
#Create heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")

row_ha = HeatmapAnnotation(df = ann_rows_immune,
                           col = col_immune_sets,
                       col = col_process_immune,
                       which= "row",
                       simple_anno_size = unit(2, "mm"),
                       show_annotation_name = F,
                       show_legend = F)

col_fun <- colorRamp2(c(-2, 0, 2), 
                      c("#9bc1bc", "white", "#ed6a5a"))

#Parameters
mat = matrix_data_ordered
width_image = 30
height_image = 30
width = unit(20, "cm")
height = unit(20, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plot and save
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "L2FC", #Legend
                        col = col_fun,
                        column_names_rot = 45,
                        show_row_dend = F,
                        row_split = 2, #Split group clusters
                        show_row_names = F,
                        column_names_gp = column_names_gp,
                        width = width,
                        height = height,
                        row_names_gp = row_names_gp)

ht_list = draw(heatmap_plot_all, 
               annotation_legend_side = "right", 
               heatmap_legend_side = "right")


draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

Interactive
```{r}
#Parameters
mat = matrix_data_ordered
width_image = 40
height_image = 90
width = unit(30, "cm")
height = unit(80, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "L2FC", #Legend
                        col = col_fun,
                        show_row_dend = F,
                        row_split = 2, #Split group clusters
                        show_row_names = F,
                        column_names_gp = column_names_gp,
                        row_names_gp = row_names_gp)

ht_list = draw(heatmap_plot_all, 
               annotation_legend_side = "right", 
               heatmap_legend_side = "right")
htShiny(ht_list)

htShiny(ht_list, save = "Interactive_CovidOnly_Heatmap")

```

### General

```{r}
####### Input
heatmap_data = Immune_GO_genes_general
file = "Immune_GO_genes_general"

###### Process data

heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Convert to wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

matrix_data_ready =  matrix_data %>% 
  round(2)

# Check if all vaccines are in the matrix and join annotations
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

# Join annotations
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 

# Reorder columns (same order as lines in the annotation table)
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% 
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Check dimensions
dim(matrix_data_ordered) 
dim(ann_vaccines_heatmap) 


#Create heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), 
                      c("#9bc1bc", "white", "#ed6a5a"))


#Parameters
mat = matrix_data_ordered
width_image = 40
height_image = 90
width = unit(30, "cm")
height = unit(80, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 3)


#Plot and save
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 10),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        #column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### Antimicrobial

```{r}
####### Input

heatmap_data = Immune_GO_genes_antimicrobial
file = "Immune_GO_genes_antimicrobial"

heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

matrix_data_ready =  matrix_data %>% 
  round(2)

ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 

matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% 
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0) %>% 
  as.matrix() 

dim(matrix_data_ordered)
dim(ann_vaccines_heatmap)

ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), 
                      c("#9bc1bc", "white", "#ed6a5a"))

mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(20, "cm")
width_image = 40
height_image = 30
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)

fileimageheatmap_ungrouped = here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))
png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)
heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 10),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, 
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### Complement

```{r}
heatmap_data = Immune_GO_genes_Complement
file = "Immune_GO_genes_Complement"

heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

matrix_data_ready =  matrix_data %>% 
  round(2)

ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 

matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% 
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0) %>% 
  as.matrix() 

dim(matrix_data_ordered)
dim(ann_vaccines_heatmap)


ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#9bc1bc", "white", "#ed6a5a"))


mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(15, "cm")
width_image = 40
height_image = 25
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)

fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))
png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)
heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 10),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### Inflammation

```{r}
heatmap_data = Immune_GO_genes_Inflammation
file = "Immune_GO_genes_Inflammation"
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

matrix_data_ready =  matrix_data %>% 
  round(2)

ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 

matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% 
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0) %>% 
  as.matrix() 

dim(matrix_data_ordered) 
dim(ann_vaccines_heatmap)

ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), 
                      c("#9bc1bc", "white", "#ed6a5a"))

mat = matrix_data_ordered
width_image = 40
height_image = 20
width = unit(30, "cm")
height = unit(10, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)

fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, 
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### ARPP

```{r}
####### Input
#Troque pelo geneset de interesse

heatmap_data = Immune_GO_genes_ARPP
file = "Immune_GO_genes_ARPP"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Converter para wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#9bc1bc", "white", "#ed6a5a"))


#Parameters
mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(50, "cm")
width_image = 40
height_image = 60
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        # column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### AV and IFN

```{r}
####### Input
#Troque pelo geneset de interesse

heatmap_data = Immune_GO_genes_IFN
file = "Immune_GO_genes_IFN"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Converter para wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#9bc1bc", "white", "#ed6a5a"))


#Parameters
mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(50, "cm")
width_image = 40
height_image = 60
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        # column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### Neutrophil

```{r}
####### Input
#Troque pelo geneset de interesse

heatmap_data = Immune_GO_genes_Neutrophil
file = "Immune_GO_genes_Neutrophil"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Converter para wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#9bc1bc", "white", "#ed6a5a"))


#Parameters
mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(40, "cm")
width_image = 40
height_image = 50
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plotar e salvar imagem
fileimageheatmap_ungrouped = here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        # column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### Eosinophil

```{r}
####### Input
#Troque pelo geneset de interesse

heatmap_data = Immune_GO_genes_Eosinophil
file = "Immune_GO_genes_Eosinophil"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Converter para wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#9bc1bc", "white", "#ed6a5a"))


#Parameters
mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(10, "cm")
width_image = 40
height_image = 20
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        # column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### Monocyte

```{r}
####### Input
#Troque pelo geneset de interesse

heatmap_data = Immune_GO_genes_Monocytes
file = "Immune_GO_genes_Monocytes"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Converter para wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#9bc1bc", "white", "#ed6a5a"))


#Parameters
mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(15, "cm")
width_image = 40
height_image = 25
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        #column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### Dendritic cells

```{r}
####### Input
#Troque pelo geneset de interesse

heatmap_data = Immune_GO_genes_DC
file = "Immune_GO_genes_DC"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Converter para wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#9bc1bc", "white", "#ed6a5a"))


#Parameters
mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(15, "cm")
width_image = 40
height_image = 25
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 10),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        #column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### Macrophage

```{r}
####### Input
#Troque pelo geneset de interesse

heatmap_data = Immune_GO_genes_Macro
file = "Immune_GO_genes_Macro"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Converter para wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#9bc1bc", "white", "#ed6a5a"))


#Parameters
mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(15, "cm")
width_image = 40
height_image = 25
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 10),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        #column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### NK

```{r}
####### Input
#Troque pelo geneset de interesse

heatmap_data = Immune_GO_genes_NK
file = "Immune_GO_genes_NK"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Converter para wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#9bc1bc", "white", "#ed6a5a"))


#Parameters
mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(15, "cm")
width_image = 40
height_image = 25
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 10),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        #column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### Mast cell

```{r}
####### Input
#Troque pelo geneset de interesse

heatmap_data = Immune_GO_genes_Mast
file = "Immune_GO_genes_Mast"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Converter para wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#9bc1bc", "white", "#ed6a5a"))


#Parameters
mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(10, "cm")
width_image = 40
height_image = 20
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 10),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        #column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

###Cellular adaptive immune system

```{r}
####### Input
#Troque pelo geneset de interesse

heatmap_data = Immune_GO_genes_Cellular
file = "Immune_GO_genes_Cellular"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Converter para wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#9bc1bc", "white", "#ed6a5a"))


#Parameters
mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(70, "cm")
width_image = 40
height_image = 80
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 5)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 10),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        #column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

###T Cell

```{r}
####### Input
heatmap_data = Immune_GO_genes_Tcells
file = "Immune_GO_genes_Tcells"

heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

matrix_data_ready =  matrix_data %>% 
  round(2)

ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 

matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% 
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0) %>% 
  as.matrix() 

dim(matrix_data_ordered) 
dim(ann_vaccines_heatmap) 


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#9bc1bc", "white", "#ed6a5a"))


#Parameters
mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(80, "cm")
width_image = 40
height_image = 90
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 10),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        #column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### Th Cell

```{r}
####### Input
#Troque pelo geneset de interesse

heatmap_data = Immune_GO_genes_Th
file = "Immune_GO_genes_Th"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Converter para wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#9bc1bc", "white", "#ed6a5a"))


#Parameters
mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(10, "cm")
width_image = 40
height_image = 20
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 10),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        #column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### Humoral adaptive immune system

```{r}
####### Input
#Troque pelo geneset de interesse

heatmap_data = Immune_GO_genes_Humoral
file = "Immune_GO_genes_Humoral"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Converter para wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#9bc1bc", "white", "#ed6a5a"))


#Parameters
mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(70, "cm")
width_image = 40
height_image = 80
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 10),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        #column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### B cells

```{r}
####### Input
#Troque pelo geneset de interesse

heatmap_data = Immune_GO_genes_Bcells
file = "Immune_GO_genes_Bcells"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Converter para wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#9bc1bc", "white", "#ed6a5a"))


#Parameters
mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(50, "cm")
width_image = 40
height_image = 60
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 10),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        #column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### Ig mediated response

```{r}
####### Input
#Troque pelo geneset de interesse

heatmap_data = Immune_GO_genes_Ig
file = "Immune_GO_genes_Ig"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Converter para wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#9bc1bc", "white", "#ed6a5a"))


#Parameters
mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(50, "cm")
width_image = 40
height_image = 60
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 10),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        #column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

# 11. Gene networks

```{r}
ImmuneGO_Annotated_Genes = readRDS(here("VaxGO gene sets", "ImmuneGO_Annotated_Genes_2024-03-26.rds"))
OtherGOs_Annotated_Genes = read_csv(here("VaxGO gene sets", "OtherGOs_Annotated_Genes.csv")) %>% 
  rename(genes = symbol)
  
degs_all_updown = read_csv(here("DGE analysis", "degs_updown_filtered_p_05_vac_infected.csv")) %>% 
  filter(direction != "NEUTRAL")
```

## COVID immune genes

### Chord diagram

Input
```{r}
All_covid_sharedgenes_Fisher_Jaccard <- read_csv("DGE analysis/All_covid_sharedgenes_Fisher_Jaccard.csv")

shared_genes_df_mirror = All_covid_sharedgenes_Fisher_Jaccard
```


#### Shared
```{r}
#Filters 
filter_shared = 1
filter_pvalue = 0.10

library(chorddiag)

matrix_vaccines <- shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  dplyr::select(Cond1, Cond2, Shared) %>% 
  pivot_wider(names_from = "Cond2", values_from = "Shared", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

types = shared_genes_df_mirror %>% 
  dplyr::select(condition = Cond1) %>% 
  distinct() %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(condition, type), by = "condition") %>% 
  dplyr::select(type)

colors = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(condition, type), by = "type") %>% 
  inner_join(matrix_vaccines %>% 
               as.data.frame() %>% 
               colnames() %>% as.data.frame() %>% rename(condition = "."),
             by = "condition") %>% 
  pull(color)

# Build the chord diagram:
all_shared = chorddiag(matrix_vaccines,
          groupColors = colors,
          type = "directional",
          width = NULL,
          height = NULL,
          margin = 100,
          showGroupnames = TRUE,
          groupThickness = 0.1,
          groupPadding = 2,
          groupnamePadding = 1,
          groupnameFontsize = 10,
          groupedgeColor = NULL,
          chordedgeColor = NULL,
          categoryNames = NULL,
          categorynamePadding = 100,
          categorynameFontsize = 28,
          showTicks = TRUE,
          tickInterval = NULL,
          ticklabelFontsize = 10,
          fadeLevel = 0.1,
          showTooltips = TRUE,
          showZeroTooltips = TRUE,
          tooltipNames = NULL,
          tooltipUnit = NULL,
          tooltipFontsize = 12,
          tooltipGroupConnector = " &#x25B6; ",
          precision = NULL,
          clickAction = NULL,
          clickGroupAction = NULL
)
all_shared
```

```{r}
matrix_vaccines_week1 <- shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(annotation_conditions %>% 
               filter(week == 1) %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(annotation_conditions %>% 
               filter(week == 1) %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, Shared) %>% 
  pivot_wider(names_from = "Cond2", values_from = "Shared", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

types = shared_genes_df_mirror %>% 
  dplyr::select(condition = Cond1) %>% 
  distinct() %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(condition, type), by = "condition") %>% 
  dplyr::select(type)

colors = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(condition, type), by = "type") %>% 
  inner_join(matrix_vaccines_week1 %>% 
               as.data.frame() %>% 
               colnames() %>% as.data.frame() %>% rename(condition = "."),
             by = "condition") %>% 
  pull(color)

# Build the chord diagram:
week1 = chorddiag(matrix_vaccines_week1,
          groupColors = colors,
          type = "directional",
          width = NULL,
          height = NULL,
          margin = 100,
          showGroupnames = TRUE,
          groupThickness = 0.1,
          groupPadding = 2,
          groupnamePadding = 1,
          groupnameFontsize = 10,
          groupedgeColor = NULL,
          chordedgeColor = NULL,
          categoryNames = NULL,
          categorynamePadding = 100,
          categorynameFontsize = 28,
          showTicks = TRUE,
          tickInterval = NULL,
          ticklabelFontsize = 10,
          fadeLevel = 0.1,
          showTooltips = TRUE,
          showZeroTooltips = TRUE,
          tooltipNames = NULL,
          tooltipUnit = NULL,
          tooltipFontsize = 12,
          tooltipGroupConnector = " &#x25B6; ",
          precision = NULL,
          clickAction = NULL,
          clickGroupAction = NULL
)
week1
```


###### Weeks
Prepare tables
```{r fig.height=5, fig.width=5}
#Week 1 ------
links_df_W1 = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(annotation_conditions %>% 
               filter(week == 1) %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(annotation_conditions %>% 
               filter(week == 1) %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, Shared)

mat_week1 = links_df_W1 %>% 
  pivot_wider(names_from = "Cond2", values_from = "Shared", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()
grid.col_w1 = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_W1,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)
group_w1 = annotation_conditions %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_W1,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)


#Week 2 ------
links_df_W2 = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(annotation_conditions %>% 
               filter(week == 2) %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(annotation_conditions %>% 
               filter(week == 2) %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, Shared)

mat_week2 = links_df_W2 %>% 
  pivot_wider(names_from = "Cond2", values_from = "Shared", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_w2 = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_W2,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_w2 = annotation_conditions %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_W2,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)


#Week 4 ------
links_df_W4 = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(annotation_conditions %>% 
               filter(week == 4) %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(annotation_conditions %>% 
               filter(week == 4) %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, Shared)

mat_week4 = links_df_W4 %>% 
  pivot_wider(names_from = "Cond2", values_from = "Shared", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_w4 = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_W4,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_w4 = annotation_conditions %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_W4,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)

#Week 8 ------
links_df_W8 = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(annotation_conditions %>% 
               filter(week == 8) %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(annotation_conditions %>% 
               filter(week == 8) %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, Shared)

mat_week8 = links_df_W8 %>% 
  pivot_wider(names_from = "Cond2", values_from = "Shared", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_w8 = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_W8,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_w8 = annotation_conditions %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_W8,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)

```
Plot
```{r fig.height= 5, fig.width=10}
#Plot
png(filename = here("Figures","ChordDiagram_Covid19_Weeks_1-2-4-7.png"), 
    width = 12, height = 14, units = "in", res = 300)

par(mfrow = c(2, 2))
#WEEK 1
chordDiagram(mat_week1, 
             grid.col = grid.col_w1, 
             annotationTrack = "grid", 
             directional = -1,
             big.gap = 6,
             transparency = 0,
             small.gap = 3,
             link.zindex = rank(mat_week1),
             h.ratio = 0.7, 
             group = group_w1,
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_week1))))))
#Labels
circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.8, #Label size
      font = 0.1)},#Label face
  bg.border = NA) 
title("Week 1")

# WEEK 2
chordDiagram(mat_week2, grid.col = grid.col_w2, 
             annotationTrack = "grid", 
            big.gap = 50,
             group = group_w2,
             small.gap = 3,
             link.zindex = rank(mat_week2),
             h.ratio = 0.7, 
             directional = -1,
             transparency = 0,
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_week2))))))
#Labels
circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.8, #Label size
      font = 0.1)},
  bg.border = NA)

title("Week 2")


# WEEK 4
chordDiagram(mat_week4, grid.col = grid.col_w4, 
             annotationTrack = "grid", 
            big.gap = 50,
             group = group_w4,
             small.gap = 3,
             link.zindex = rank(mat_week4),
             h.ratio = 0.7, 
             directional = -1,
             transparency = 0,
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_week4))))))
#Labels
circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.8, #Label size
      font = 0.1)},
  bg.border = NA)

title("Week 4")

# WEEK 8
chordDiagram(mat_week8, grid.col = grid.col_w8, 
             annotationTrack = "grid", 
            big.gap = 50,
             group = group_w8,
             small.gap = 3,
             link.zindex = rank(mat_week8),
             h.ratio = 0.7, 
             directional = -1,
             transparency = 0,
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_week8))))))
#Labels
circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.8, #Label size
      font = 0.1)},
  bg.border = NA)

title("Week 8")

dev.off()
circos.clear()

```

##### Types
Prepare tables
```{r fig.height= 10, fig.width=10}
#Types
###### All -----
links_df = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  dplyr::select(Cond1, Cond2, Shared)

mat_types = links_df %>% 
  pivot_wider(names_from = "Cond2", values_from = "Shared", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_types = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_types = annotation_conditions %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)

# Vaccination only -------
links_df_vac = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(annotation_conditions %>% 
               filter(disease_vac == "V") %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(annotation_conditions %>% 
               filter(disease_vac == "V") %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, Shared)

mat_types_vac = links_df_vac %>% 
  pivot_wider(names_from = "Cond2", values_from = "Shared", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_types_vac = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_vac,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_types_vac = annotation_conditions %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_vac,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)

#Infection only
links_df_inf = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(annotation_conditions %>% 
               filter(disease_vac == "I") %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(annotation_conditions %>% 
               filter(disease_vac == "I") %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, Shared)

mat_types_inf = links_df_inf %>% 
  pivot_wider(names_from = "Cond2", values_from = "Shared", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_types_inf = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_inf,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_types_inf = annotation_conditions %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_inf,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)
```

Plot
All conditions and stats
```{r fig.height= 5, fig.width=5}
#Plot

#Types

png(filename = here("Figures","ChordDiagram_Covid19_Types_All.png"), 
    width = 5, height = 5, units = "in", res = 300)

#Infection and vaccination
chordDiagram(mat_types, grid.col = grid.col_types,
             annotationTrack = "grid",  transparency = 0,
             big.gap = 10,
             small.gap = 3,
             directional = -1, 
             group = group_types,
             link.zindex = rank(mat_types),
              h.ratio = 0.7, 
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_types))))))

circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.5, #Label size
      font = 0.1
      )},
  bg.border = NA)
title("Infection and Vaccination")

dev.off()
circos.clear()
```

By state
```{r fig.height= 5, fig.width=15}

png(filename = here("Figures","ChordDiagram_Covid19_Types_Inf_Vac.png"), 
    width = 6, height = 12, units = "in", res = 300)

par(mfrow = c(2, 1))
#Vaccination
chordDiagram(mat_types_vac, grid.col = grid.col_types_vac, 
             annotationTrack = "grid", transparency = 0,
             big.gap = 6,
             small.gap = 3,
                          directional = -1,
             group = group_types,
             link.zindex = rank(mat_types_vac),
             h.ratio = 0.7, 
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_types_vac))))))

circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.5, #Label size
      font = 0.1
      )},
  bg.border = NA) 
title("Vaccination only")

#Infection
chordDiagram(mat_types_inf, grid.col = grid.col_types_inf, 
             annotationTrack = "grid", transparency = 0,
             big.gap = 50,
             small.gap = 3,
            directional = -1,
             group = group_types,
             h.ratio = 0.7, 
             link.zindex = rank(mat_types_inf),
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_types_inf))))))

circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.5, #Label size
      font = 0.1
      )},
  bg.border = NA) 
title("Infection only")
circos.clear()
dev.off()
```


#### Jaccard distance
##### Interactive
```{r}

matrix_vaccines <- shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  dplyr::select(Cond1, Cond2, jaccard_distance) %>% 
  pivot_wider(names_from = "Cond2", values_from = "jaccard_distance", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

types = shared_genes_df_mirror %>% 
  dplyr::select(condition = Cond1) %>% 
  distinct() %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(condition, type), by = "condition") %>% 
  dplyr::select(type)

colors = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(condition, type), by = "type") %>% 
  inner_join(matrix_vaccines %>% 
               as.data.frame() %>% 
               colnames() %>% as.data.frame() %>% rename(condition = "."),
             by = "condition") %>% 
  pull(color)

# Build the chord diagram:
chorddiag(matrix_vaccines,
          groupColors = colors,
          type = "directional",
          width = NULL,
          height = NULL,
          margin = 100,
          showGroupnames = TRUE,
          groupThickness = 0.1,
          groupPadding = 2,
          groupnamePadding = 1,
          groupnameFontsize = 10,
          groupedgeColor = NULL,
          chordedgeColor = NULL,
          categoryNames = NULL,
          categorynamePadding = 100,
          categorynameFontsize = 28,
          showTicks = TRUE,
          tickInterval = NULL,
          ticklabelFontsize = 10,
          fadeLevel = 0.1,
          showTooltips = TRUE,
          showZeroTooltips = TRUE,
          tooltipNames = NULL,
          tooltipUnit = NULL,
          tooltipFontsize = 12,
          tooltipGroupConnector = " &#x25B6; ",
          precision = NULL,
          clickAction = NULL,
          clickGroupAction = NULL
)
```

##### Static
Prepare tables
```{r fig.height=5, fig.width=5}
#Week 1 ------
links_df_W1 = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(annotation_conditions %>% 
               filter(week == 1) %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(annotation_conditions %>% 
               filter(week == 1) %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, jaccard_distance)

mat_week1 = links_df_W1 %>% 
  pivot_wider(names_from = "Cond2", values_from = "jaccard_distance", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()
grid.col_w1 = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_W1,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)
group_w1 = annotation_conditions %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_W1,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)


#Week 2 ------
links_df_W2 = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(annotation_conditions %>% 
               filter(week == 2) %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(annotation_conditions %>% 
               filter(week == 2) %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, jaccard_distance)

mat_week2 = links_df_W2 %>% 
  pivot_wider(names_from = "Cond2", values_from = "jaccard_distance", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_w2 = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_W2,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_w2 = annotation_conditions %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_W2,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)


#Week 4 ------
links_df_W4 = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(annotation_conditions %>% 
               filter(week == 4) %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(annotation_conditions %>% 
               filter(week == 4) %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, jaccard_distance)

mat_week4 = links_df_W4 %>% 
  pivot_wider(names_from = "Cond2", values_from = "jaccard_distance", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_w4 = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_W4,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_w4 = annotation_conditions %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_W4,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)

#Week 8 ------
links_df_W8 = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(annotation_conditions %>% 
               filter(week == 8) %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(annotation_conditions %>% 
               filter(week == 8) %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, jaccard_distance)

mat_week8 = links_df_W8 %>% 
  pivot_wider(names_from = "Cond2", values_from = "jaccard_distance", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_w8 = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_W8,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_w8 = annotation_conditions %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_W8,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)

```
Plot
```{r fig.height= 5, fig.width=10}
#Plot
png(filename = here("Figures","ChordDiagram_Covid19_Weeks_1-2-4-7_jaccard.png"), 
    width = 12, height = 14, units = "in", res = 300)

par(mfrow = c(2, 2))
#WEEK 1
chordDiagram(mat_week1, 
             grid.col = grid.col_w1, 
             annotationTrack = "grid", 
             directional = -1,
             big.gap = 6,
             transparency = 0,
             small.gap = 3,
             link.zindex = rank(mat_week1),
             h.ratio = 0.7, 
             group = group_w1,
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_week1))))))
#Labels
circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.8, #Label size
      font = 0.1)},#Label face
  bg.border = NA) 
title("Week 1")

# WEEK 2
chordDiagram(mat_week2, grid.col = grid.col_w2, 
             annotationTrack = "grid", 
            big.gap = 50,
             group = group_w2,
             small.gap = 3,
             link.zindex = rank(mat_week2),
             h.ratio = 0.7, 
             directional = -1,
             transparency = 0,
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_week2))))))
#Labels
circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.8, #Label size
      font = 0.1)},
  bg.border = NA)

title("Week 2")


# WEEK 4
chordDiagram(mat_week4, grid.col = grid.col_w4, 
             annotationTrack = "grid", 
            big.gap = 50,
             group = group_w4,
             small.gap = 3,
             link.zindex = rank(mat_week4),
             h.ratio = 0.7, 
             directional = -1,
             transparency = 0,
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_week4))))))
#Labels
circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.8, #Label size
      font = 0.1)},
  bg.border = NA)

title("Week 4")

# WEEK 8
chordDiagram(mat_week8, grid.col = grid.col_w8, 
             annotationTrack = "grid", 
            big.gap = 50,
             group = group_w8,
             small.gap = 3,
             link.zindex = rank(mat_week8),
             h.ratio = 0.7, 
             directional = -1,
             transparency = 0,
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_week8))))))
#Labels
circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.8, #Label size
      font = 0.1)},
  bg.border = NA)

title("Week 8")

dev.off()
circos.clear()

```

##### Types
Prepare tables
```{r fig.height= 10, fig.width=10}
#Types
###### All -----
links_df = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  dplyr::select(Cond1, Cond2, jaccard_distance)

mat_types = links_df %>% 
  pivot_wider(names_from = "Cond2", values_from = "jaccard_distance", values_fill = list(jaccard_distance = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_types = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_types = annotation_conditions %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)

# Vaccination only -------
links_df_vac = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(annotation_conditions %>% 
               filter(disease_vac == "V") %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(annotation_conditions %>% 
               filter(disease_vac == "V") %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, jaccard_distance)

mat_types_vac = links_df_vac %>% 
  pivot_wider(names_from = "Cond2", values_from = "jaccard_distance", values_fill = list(jaccard_distance = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_types_vac = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_vac,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_types_vac = annotation_conditions %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_vac,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)

#Infection only
links_df_inf = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(annotation_conditions %>% 
               filter(disease_vac == "I") %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(annotation_conditions %>% 
               filter(disease_vac == "I") %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, jaccard_distance)

mat_types_inf = links_df_inf %>% 
  pivot_wider(names_from = "Cond2", values_from = "jaccard_distance", values_fill = list(jaccard_distance = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_types_inf = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_inf,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_types_inf = annotation_conditions %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_inf,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)
```

Plot
All conditions and stats
```{r fig.height= 5, fig.width=5}
#Plot

#Types

png(filename = here("Figures","ChordDiagram_Covid19_Types_All_jaccard_distance.png"), 
    width = 5, height = 5, units = "in", res = 300)

#Infection and vaccination
chordDiagram(mat_types, grid.col = grid.col_types,
             annotationTrack = "grid",  transparency = 0,
             big.gap = 10,
             small.gap = 3,
             directional = -1, 
             group = group_types,
             link.zindex = rank(mat_types),
              h.ratio = 0.7, 
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_types))))))

circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.5, #Label size
      font = 0.1
      )},
  bg.border = NA)
title("Infection and Vaccination")

dev.off()
circos.clear()
```

By state
```{r fig.height= 5, fig.width=15}

png(filename = here("Figures","ChordDiagram_Covid19_Types_Inf_Vac_jaccard_distance.png"), 
    width = 6, height = 12, units = "in", res = 300)

par(mfrow = c(2, 1))
#Vaccination
chordDiagram(mat_types_vac, grid.col = grid.col_types_vac, 
             annotationTrack = "grid", transparency = 0,
             big.gap = 6,
             small.gap = 3,
                          directional = -1,
             group = group_types,
             link.zindex = rank(mat_types_vac),
             h.ratio = 0.7, 
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_types_vac))))))

circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.5, #Label size
      font = 0.1
      )},
  bg.border = NA) 
title("Vaccination only")

#Infection
chordDiagram(mat_types_inf, grid.col = grid.col_types_inf, 
             annotationTrack = "grid", transparency = 0,
             big.gap = 50,
             small.gap = 3,
            directional = -1,
             group = group_types,
             h.ratio = 0.7, 
             link.zindex = rank(mat_types_inf),
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_types_inf))))))

circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.5, #Label size
      font = 0.1
      )},
  bg.border = NA) 
title("Infection only")
circos.clear()
dev.off()
```
## COVID and other Vaccines - immune genes

### Chord diagram

Input
```{r}
degs_all_updown_covid_other_immune_shared <- read_csv("VaxGO gene sets/degs_all_updown_covid_other_immune_shared.csv")

All_covid_sharedgenes_Fisher_Jaccard = degs_all_updown_covid_other_immune_shared

shared_genes_df_mirror = All_covid_sharedgenes_Fisher_Jaccard
```


##### Shared
```{r}
#Filters 
filter_shared = 1
filter_pvalue = 0.10

library(chorddiag)

matrix_vaccines <- shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  dplyr::select(Cond1, Cond2, Shared) %>% 
  pivot_wider(names_from = "Cond2", values_from = "Shared", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

types = shared_genes_df_mirror %>% 
  dplyr::select(condition = Cond1) %>% 
  distinct() %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(condition, type), by = "condition") %>% 
  dplyr::select(type)

colors = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(condition, type), by = "type") %>% 
  inner_join(matrix_vaccines %>% 
               as.data.frame() %>% 
               colnames() %>% as.data.frame() %>% rename(condition = "."),
             by = "condition") %>% 
  pull(color)

# Build the chord diagram:
all_shared = chorddiag(matrix_vaccines,
          groupColors = colors,
          type = "directional",
          width = NULL,
          height = NULL,
          margin = 100,
          showGroupnames = TRUE,
          groupThickness = 0.1,
          groupPadding = 2,
          groupnamePadding = 1,
          groupnameFontsize = 10,
          groupedgeColor = NULL,
          chordedgeColor = NULL,
          categoryNames = NULL,
          categorynamePadding = 100,
          categorynameFontsize = 28,
          showTicks = TRUE,
          tickInterval = NULL,
          ticklabelFontsize = 10,
          fadeLevel = 0.1,
          showTooltips = TRUE,
          showZeroTooltips = TRUE,
          tooltipNames = NULL,
          tooltipUnit = NULL,
          tooltipFontsize = 12,
          tooltipGroupConnector = " &#x25B6; ",
          precision = NULL,
          clickAction = NULL,
          clickGroupAction = NULL
)
all_shared
```

```{r}
matrix_vaccines_week1 <- shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(annotation_conditions %>% 
               filter(week == 1) %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(annotation_conditions %>% 
               filter(week == 1) %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, Shared) %>% 
  pivot_wider(names_from = "Cond2", values_from = "Shared", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

types = shared_genes_df_mirror %>% 
  dplyr::select(condition = Cond1) %>% 
  distinct() %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(condition, type), by = "condition") %>% 
  dplyr::select(type)

colors = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(condition, type), by = "type") %>% 
  inner_join(matrix_vaccines_week1 %>% 
               as.data.frame() %>% 
               colnames() %>% as.data.frame() %>% rename(condition = "."),
             by = "condition") %>% 
  pull(color)

# Build the chord diagram:
week1 = chorddiag(matrix_vaccines_week1,
          groupColors = colors,
          type = "directional",
          width = NULL,
          height = NULL,
          margin = 100,
          showGroupnames = TRUE,
          groupThickness = 0.1,
          groupPadding = 2,
          groupnamePadding = 1,
          groupnameFontsize = 10,
          groupedgeColor = NULL,
          chordedgeColor = NULL,
          categoryNames = NULL,
          categorynamePadding = 100,
          categorynameFontsize = 28,
          showTicks = TRUE,
          tickInterval = NULL,
          ticklabelFontsize = 10,
          fadeLevel = 0.1,
          showTooltips = TRUE,
          showZeroTooltips = TRUE,
          tooltipNames = NULL,
          tooltipUnit = NULL,
          tooltipFontsize = 12,
          tooltipGroupConnector = " &#x25B6; ",
          precision = NULL,
          clickAction = NULL,
          clickGroupAction = NULL
)
week1
```


###### Weeks
Prepare tables
```{r fig.height=5, fig.width=5}
#Week 1 ------
links_df_W1 = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(annotation_conditions %>% 
               filter(week == 1) %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(annotation_conditions %>% 
               filter(week == 1) %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, Shared)

mat_week1 = links_df_W1 %>% 
  pivot_wider(names_from = "Cond2", values_from = "Shared", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()
grid.col_w1 = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_W1,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)
group_w1 = annotation_conditions %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_W1,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)


#Week 2 ------
links_df_W2 = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(annotation_conditions %>% 
               filter(week == 2) %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(annotation_conditions %>% 
               filter(week == 2) %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, Shared)

mat_week2 = links_df_W2 %>% 
  pivot_wider(names_from = "Cond2", values_from = "Shared", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_w2 = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_W2,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_w2 = annotation_conditions %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_W2,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)


#Week 4 ------
links_df_W4 = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(annotation_conditions %>% 
               filter(week == 4) %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(annotation_conditions %>% 
               filter(week == 4) %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, Shared)

mat_week4 = links_df_W4 %>% 
  pivot_wider(names_from = "Cond2", values_from = "Shared", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_w4 = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_W4,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_w4 = annotation_conditions %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_W4,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)

#Week 8 ------
links_df_W8 = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(annotation_conditions %>% 
               filter(week == 8) %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(annotation_conditions %>% 
               filter(week == 8) %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, Shared)

mat_week8 = links_df_W8 %>% 
  pivot_wider(names_from = "Cond2", values_from = "Shared", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_w8 = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_W8,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_w8 = annotation_conditions %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_W8,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)

```
Plot
```{r fig.height= 5, fig.width=10}
#Plot
png(filename = here("Figures","ChordDiagram_Covid19_Weeks_1-2-4-7.png"), 
    width = 12, height = 14, units = "in", res = 300)

par(mfrow = c(2, 2))
#WEEK 1
chordDiagram(mat_week1, 
             grid.col = grid.col_w1, 
             annotationTrack = "grid", 
             directional = -1,
             big.gap = 6,
             transparency = 0,
             small.gap = 3,
             link.zindex = rank(mat_week1),
             h.ratio = 0.7, 
             group = group_w1,
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_week1))))))
#Labels
circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.8, #Label size
      font = 0.1)},#Label face
  bg.border = NA) 
title("Week 1")

# WEEK 2
chordDiagram(mat_week2, grid.col = grid.col_w2, 
             annotationTrack = "grid", 
            big.gap = 50,
             group = group_w2,
             small.gap = 3,
             link.zindex = rank(mat_week2),
             h.ratio = 0.7, 
             directional = -1,
             transparency = 0,
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_week2))))))
#Labels
circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.8, #Label size
      font = 0.1)},
  bg.border = NA)

title("Week 2")


# WEEK 4
chordDiagram(mat_week4, grid.col = grid.col_w4, 
             annotationTrack = "grid", 
            big.gap = 50,
             group = group_w4,
             small.gap = 3,
             link.zindex = rank(mat_week4),
             h.ratio = 0.7, 
             directional = -1,
             transparency = 0,
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_week4))))))
#Labels
circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.8, #Label size
      font = 0.1)},
  bg.border = NA)

title("Week 4")

# WEEK 8
chordDiagram(mat_week8, grid.col = grid.col_w8, 
             annotationTrack = "grid", 
            big.gap = 50,
             group = group_w8,
             small.gap = 3,
             link.zindex = rank(mat_week8),
             h.ratio = 0.7, 
             directional = -1,
             transparency = 0,
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_week8))))))
#Labels
circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.8, #Label size
      font = 0.1)},
  bg.border = NA)

title("Week 8")

dev.off()
circos.clear()

```

###### Types
Prepare tables
```{r fig.height= 10, fig.width=10}
#Types
###### All -----
links_df = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  dplyr::select(Cond1, Cond2, Shared)

mat_types = links_df %>% 
  pivot_wider(names_from = "Cond2", values_from = "Shared", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_types = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_types = annotation_conditions %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)

# Vaccination only -------
links_df_vac = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(annotation_conditions %>% 
               filter(disease_vac == "V") %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(annotation_conditions %>% 
               filter(disease_vac == "V") %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, Shared)

mat_types_vac = links_df_vac %>% 
  pivot_wider(names_from = "Cond2", values_from = "Shared", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_types_vac = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_vac,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_types_vac = annotation_conditions %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_vac,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)

#Infection only
links_df_inf = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(annotation_conditions %>% 
               filter(disease_vac == "I") %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(annotation_conditions %>% 
               filter(disease_vac == "I") %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, Shared)

mat_types_inf = links_df_inf %>% 
  pivot_wider(names_from = "Cond2", values_from = "Shared", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_types_inf = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_inf,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_types_inf = annotation_conditions %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_inf,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)
```

Plot
All conditions and stats
```{r fig.height= 5, fig.width=5}
#Plot

#Types

png(filename = here("Figures","ChordDiagram_Covid19_Types_All.png"), 
    width = 5, height = 5, units = "in", res = 300)

#Infection and vaccination
chordDiagram(mat_types, grid.col = grid.col_types,
             annotationTrack = "grid",  transparency = 0,
             big.gap = 10,
             small.gap = 3,
             directional = -1, 
             group = group_types,
             link.zindex = rank(mat_types),
              h.ratio = 0.7, 
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_types))))))

circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.5, #Label size
      font = 0.1
      )},
  bg.border = NA)
title("Infection and Vaccination")

dev.off()
circos.clear()
```

By state
```{r fig.height= 5, fig.width=15}

png(filename = here("Figures","ChordDiagram_Covid19_Types_Inf_Vac.png"), 
    width = 6, height = 12, units = "in", res = 300)

par(mfrow = c(2, 1))
#Vaccination
chordDiagram(mat_types_vac, grid.col = grid.col_types_vac, 
             annotationTrack = "grid", transparency = 0,
             big.gap = 6,
             small.gap = 3,
                          directional = -1,
             group = group_types,
             link.zindex = rank(mat_types_vac),
             h.ratio = 0.7, 
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_types_vac))))))

circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.5, #Label size
      font = 0.1
      )},
  bg.border = NA) 
title("Vaccination only")

#Infection
chordDiagram(mat_types_inf, grid.col = grid.col_types_inf, 
             annotationTrack = "grid", transparency = 0,
             big.gap = 50,
             small.gap = 3,
            directional = -1,
             group = group_types,
             h.ratio = 0.7, 
             link.zindex = rank(mat_types_inf),
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_types_inf))))))

circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.5, #Label size
      font = 0.1
      )},
  bg.border = NA) 
title("Infection only")
circos.clear()
dev.off()
```


##### Jaccard distance
###### Interactive
```{r}

matrix_vaccines <- shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  dplyr::select(Cond1, Cond2, jaccard_distance) %>% 
  pivot_wider(names_from = "Cond2", values_from = "jaccard_distance", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

types = shared_genes_df_mirror %>% 
  dplyr::select(condition = Cond1) %>% 
  distinct() %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(condition, type), by = "condition") %>% 
  dplyr::select(type)

colors = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(condition, type), by = "type") %>% 
  inner_join(matrix_vaccines %>% 
               as.data.frame() %>% 
               colnames() %>% as.data.frame() %>% rename(condition = "."),
             by = "condition") %>% 
  pull(color)

# Build the chord diagram:
chorddiag(matrix_vaccines,
          groupColors = colors,
          type = "directional",
          width = NULL,
          height = NULL,
          margin = 100,
          showGroupnames = TRUE,
          groupThickness = 0.1,
          groupPadding = 2,
          groupnamePadding = 1,
          groupnameFontsize = 10,
          groupedgeColor = NULL,
          chordedgeColor = NULL,
          categoryNames = NULL,
          categorynamePadding = 100,
          categorynameFontsize = 28,
          showTicks = TRUE,
          tickInterval = NULL,
          ticklabelFontsize = 10,
          fadeLevel = 0.1,
          showTooltips = TRUE,
          showZeroTooltips = TRUE,
          tooltipNames = NULL,
          tooltipUnit = NULL,
          tooltipFontsize = 12,
          tooltipGroupConnector = " &#x25B6; ",
          precision = NULL,
          clickAction = NULL,
          clickGroupAction = NULL
)
```

###### Static
Prepare tables
```{r fig.height=5, fig.width=5}
#Week 1 ------
links_df_W1 = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(annotation_conditions %>% 
               filter(week == 1) %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(annotation_conditions %>% 
               filter(week == 1) %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, jaccard_distance)

mat_week1 = links_df_W1 %>% 
  pivot_wider(names_from = "Cond2", values_from = "jaccard_distance", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()
grid.col_w1 = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_W1,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)
group_w1 = annotation_conditions %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_W1,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)


#Week 2 ------
links_df_W2 = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(annotation_conditions %>% 
               filter(week == 2) %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(annotation_conditions %>% 
               filter(week == 2) %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, jaccard_distance)

mat_week2 = links_df_W2 %>% 
  pivot_wider(names_from = "Cond2", values_from = "jaccard_distance", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_w2 = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_W2,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_w2 = annotation_conditions %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_W2,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)


#Week 4 ------
links_df_W4 = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(annotation_conditions %>% 
               filter(week == 4) %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(annotation_conditions %>% 
               filter(week == 4) %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, jaccard_distance)

mat_week4 = links_df_W4 %>% 
  pivot_wider(names_from = "Cond2", values_from = "jaccard_distance", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_w4 = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_W4,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_w4 = annotation_conditions %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_W4,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)

#Week 8 ------
links_df_W8 = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(annotation_conditions %>% 
               filter(week == 8) %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(annotation_conditions %>% 
               filter(week == 8) %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, jaccard_distance)

mat_week8 = links_df_W8 %>% 
  pivot_wider(names_from = "Cond2", values_from = "jaccard_distance", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_w8 = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_W8,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_w8 = annotation_conditions %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_W8,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)

```
Plot
```{r fig.height= 5, fig.width=10}
#Plot
png(filename = here("Figures","ChordDiagram_Covid19_Weeks_1-2-4-7_jaccard.png"), 
    width = 12, height = 14, units = "in", res = 300)

par(mfrow = c(2, 2))
#WEEK 1
chordDiagram(mat_week1, 
             grid.col = grid.col_w1, 
             annotationTrack = "grid", 
             directional = -1,
             big.gap = 6,
             transparency = 0,
             small.gap = 3,
             link.zindex = rank(mat_week1),
             h.ratio = 0.7, 
             group = group_w1,
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_week1))))))
#Labels
circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.8, #Label size
      font = 0.1)},#Label face
  bg.border = NA) 
title("Week 1")

# WEEK 2
chordDiagram(mat_week2, grid.col = grid.col_w2, 
             annotationTrack = "grid", 
            big.gap = 50,
             group = group_w2,
             small.gap = 3,
             link.zindex = rank(mat_week2),
             h.ratio = 0.7, 
             directional = -1,
             transparency = 0,
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_week2))))))
#Labels
circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.8, #Label size
      font = 0.1)},
  bg.border = NA)

title("Week 2")


# WEEK 4
chordDiagram(mat_week4, grid.col = grid.col_w4, 
             annotationTrack = "grid", 
            big.gap = 50,
             group = group_w4,
             small.gap = 3,
             link.zindex = rank(mat_week4),
             h.ratio = 0.7, 
             directional = -1,
             transparency = 0,
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_week4))))))
#Labels
circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.8, #Label size
      font = 0.1)},
  bg.border = NA)

title("Week 4")

# WEEK 8
chordDiagram(mat_week8, grid.col = grid.col_w8, 
             annotationTrack = "grid", 
            big.gap = 50,
             group = group_w8,
             small.gap = 3,
             link.zindex = rank(mat_week8),
             h.ratio = 0.7, 
             directional = -1,
             transparency = 0,
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_week8))))))
#Labels
circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.8, #Label size
      font = 0.1)},
  bg.border = NA)

title("Week 8")

dev.off()
circos.clear()

```

###### Types
Prepare tables
```{r fig.height= 10, fig.width=10}
#Types
###### All -----
links_df = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  dplyr::select(Cond1, Cond2, jaccard_distance)

mat_types = links_df %>% 
  pivot_wider(names_from = "Cond2", values_from = "jaccard_distance", values_fill = list(jaccard_distance = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_types = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_types = annotation_conditions %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)

# Vaccination only -------
links_df_vac = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(annotation_conditions %>% 
               filter(disease_vac == "V") %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(annotation_conditions %>% 
               filter(disease_vac == "V") %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, jaccard_distance)

mat_types_vac = links_df_vac %>% 
  pivot_wider(names_from = "Cond2", values_from = "jaccard_distance", values_fill = list(jaccard_distance = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_types_vac = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_vac,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_types_vac = annotation_conditions %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_vac,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)

#Infection only
links_df_inf = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(annotation_conditions %>% 
               filter(disease_vac == "I") %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(annotation_conditions %>% 
               filter(disease_vac == "I") %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, jaccard_distance)

mat_types_inf = links_df_inf %>% 
  pivot_wider(names_from = "Cond2", values_from = "jaccard_distance", values_fill = list(jaccard_distance = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_types_inf = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_inf,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_types_inf = annotation_conditions %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_inf,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)
```

Plot
All conditions and stats
```{r fig.height= 5, fig.width=5}
#Plot

#Types

png(filename = here("Figures","ChordDiagram_Covid19_Types_All_jaccard_distance.png"), 
    width = 5, height = 5, units = "in", res = 300)

#Infection and vaccination
chordDiagram(mat_types, grid.col = grid.col_types,
             annotationTrack = "grid",  transparency = 0,
             big.gap = 10,
             small.gap = 3,
             directional = -1, 
             group = group_types,
             link.zindex = rank(mat_types),
              h.ratio = 0.7, 
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_types))))))

circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.5, #Label size
      font = 0.1
      )},
  bg.border = NA)
title("Infection and Vaccination")

dev.off()
circos.clear()
```

By state
```{r fig.height= 5, fig.width=15}

png(filename = here("Figures","ChordDiagram_Covid19_Types_Inf_Vac_jaccard_distance.png"), 
    width = 6, height = 12, units = "in", res = 300)

par(mfrow = c(2, 1))
#Vaccination
chordDiagram(mat_types_vac, grid.col = grid.col_types_vac, 
             annotationTrack = "grid", transparency = 0,
             big.gap = 6,
             small.gap = 3,
                          directional = -1,
             group = group_types,
             link.zindex = rank(mat_types_vac),
             h.ratio = 0.7, 
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_types_vac))))))

circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.5, #Label size
      font = 0.1
      )},
  bg.border = NA) 
title("Vaccination only")

#Infection
chordDiagram(mat_types_inf, grid.col = grid.col_types_inf, 
             annotationTrack = "grid", transparency = 0,
             big.gap = 50,
             small.gap = 3,
            directional = -1,
             group = group_types,
             h.ratio = 0.7, 
             link.zindex = rank(mat_types_inf),
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_types_inf))))))

circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.5, #Label size
      font = 0.1
      )},
  bg.border = NA) 
title("Infection only")
circos.clear()
dev.off()
```


## Cytoscape 
### COVID-19 non-immune genes

```{r}
ImmuneGO_Annotated_Genes
OtherGOs_Annotated_Genes
degs_all_updown
filename = "covid_vaccines_infection_NonImmune"


##### Table source - target - annotations ------

# Not immune
degs_all_updown_covid_notimmune = degs_all_updown %>% 
  as.data.frame() %>% 
  select(condition:direction) %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  inner_join(OtherGOs_Annotated_Genes %>% select(genes) %>% distinct(), by = "genes") %>% 
  distinct() %>% 
  select(condition, genes, regulation = direction)  %>% 
  distinct() %>% 
    filter(regulation %in% c("UP", "DOWN")) %>% 
  mutate(regulation_numeric = case_when(
    regulation %in% c("UP") ~ 1,
    regulation %in% c("DOWN") ~ -1,
    TRUE ~ as.numeric(regulation)
  )) %>% 
  distinct()


# Not immune genes
notimmune_source1 = degs_all_updown_covid_notimmune %>% 
  inner_join(OtherGOs_Annotated_Genes %>% filter(annotation == "Major process", term != "Immune system"), by = "genes") %>% 
  select(source = condition, target = genes, value = regulation_numeric) %>% 
  distinct() %>% 
  left_join(ann_vaccines_covid_other %>% rename(source = condition), by = "source") %>% 
    mutate(covid_other = if_else(is.na(covid_other), "OTHER", covid_other),
           annotation = paste0("Condition", covid_other))

notimmune_source2 = degs_all_updown_covid_notimmune %>% 
  inner_join(OtherGOs_Annotated_Genes %>% filter(annotation == "Major process", term != "Immune system"), by = "genes") %>% 
  select(source = term, target = genes, value = regulation_numeric) %>% 
  distinct() %>% 
  mutate(value = 0, 
         annotation = "Other processes")

not_immune = bind_rows(notimmune_source1, notimmune_source2) %>% mutate(id = source)

write.csv(not_immune, file = paste0(filename, "Network.csv"), row.names = F)

##### Tables nodes and edges ------

edges = not_immune %>% 
  select(source, target, value) %>% 
  mutate(type = "Directed")

nodes = not_immune %>% 
  select(source, annotation, vaccine:infection) %>% 
  distinct() %>% 
  mutate(id = source)

write.csv(edges, file = paste0(filename, "Network_edges.csv"), row.names = F)
write.csv(nodes, file = paste0(filename, "Network_nodes.csv"), row.names = F)

```

### COVID immune and non immune genes

```{r}
ImmuneGO_Annotated_Genes
OtherGOs_Annotated_Genes
degs_all_updown
filename = "covid_vaccines_infection_Immune_Notimmune"


##### Table source - target - annotations ------

#Immune
degs_all_updown_covid_immune = degs_all_updown %>% 
  as.data.frame() %>% 
  select(condition:direction) %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  inner_join(ImmuneGO_Annotated_Genes %>% select(genes) %>% distinct(), by = "genes") %>% 
  distinct() %>% 
  select(condition, genes, regulation = direction)  %>% 
  distinct() %>% 
  filter(regulation %in% c("UP", "DOWN")) %>% 
  mutate(regulation_numeric = case_when(
    regulation %in% c("UP") ~ 1,
    regulation %in% c("DOWN") ~ -1,
    TRUE ~ as.numeric(regulation)
  )) %>% 
  distinct()

# Not immune
degs_all_updown_covid_notimmune = degs_all_updown %>% 
  as.data.frame() %>% 
  select(condition:direction) %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  inner_join(OtherGOs_Annotated_Genes %>% filter(term != "Immune system") %>% select(genes) %>% distinct(), by = "genes") %>% 
  distinct() %>% 
  select(condition, genes, regulation = direction)  %>% 
  distinct() %>% 
    filter(regulation %in% c("UP", "DOWN")) %>% 
  mutate(regulation_numeric = case_when(
    regulation %in% c("UP") ~ 1,
    regulation %in% c("DOWN") ~ -1,
    TRUE ~ as.numeric(regulation)
  )) %>% 
  distinct()

# Immune
immune_source1 = degs_all_updown_covid_immune %>% 
  inner_join(ImmuneGO_Annotated_Genes %>% filter(go_term == "Manual", 
                                                 !(process %in% c("TCR REPERTOIRE", "BCR REPERTOIRE"))), 
             by = "genes") %>% 
  select(source = condition, target = genes, value = regulation_numeric) %>% 
  distinct() %>% 
  left_join(ann_vaccines_covid_other %>% rename(source = condition), by = "source") %>% 
  mutate(covid_other = if_else(is.na(covid_other), "OTHER", covid_other),
         annotation = paste0("Condition", covid_other))

immune_source2 = degs_all_updown_covid_immune %>% 
  inner_join(ImmuneGO_Annotated_Genes %>% filter(go_term == "Manual", 
                                                 !(process %in% c("TCR REPERTOIRE", "BCR REPERTOIRE"))), 
             by = "genes") %>% 
  select(source = process, target = genes) %>% 
  distinct() %>% 
  mutate(value = 0, 
         annotation = "Immune system")

immune = bind_rows(immune_source1, immune_source2) %>% mutate(id = source)


# Not immune genes
notimmune_source1 = degs_all_updown_covid_notimmune %>% 
  inner_join(OtherGOs_Annotated_Genes %>% filter(annotation == "Major process", term != "Immune system"), by = "genes") %>% 
  select(source = condition, target = genes, value = regulation_numeric) %>% 
  distinct() %>% 
  left_join(ann_vaccines_covid_other %>% rename(source = condition), by = "source") %>% 
    mutate(covid_other = if_else(is.na(covid_other), "OTHER", covid_other),
           annotation = paste0("Condition", covid_other))

notimmune_source2 = degs_all_updown_covid_notimmune %>% 
  inner_join(OtherGOs_Annotated_Genes %>% filter(annotation == "Major process", term != "Immune system"), by = "genes") %>% 
  select(source = term, target = genes, value = regulation_numeric) %>% 
  distinct() %>% 
  mutate(value = 0, 
         annotation = "Other processes")

not_immune = bind_rows(notimmune_source1, notimmune_source2) %>% mutate(id = source)
immune_notimmune = bind_rows(not_immune, immune)
write.csv(immune_notimmune, file = paste0(filename, "Network_immune_notimmune.csv"), row.names = F)

##### Tables nodes and edges ------

edges = immune_notimmune %>% 
  select(source, target, value) %>% 
  mutate(type = "Directed")

nodes = immune_notimmune %>% 
  select(source, annotation, vaccine:infection) %>% 
  distinct() %>% 
  mutate(id = source)

write.csv(edges, file = paste0(filename, "Network_immune_notimmune_edges.csv"), row.names = F)
write.csv(nodes, file = paste0(filename, "Network_immune_notimmune_nodes.csv"), row.names = F)

```

### COVID-COVID Network by statistics only (TOTAL = UP AND DOWN)

```{r}
# Vaccines and infection -----
ann_vaccines_covid_other

degs_all_updown_covid_other_immune_shared
degs_all_updown_covid_other_nonimmune_shared

filename = "degs_all_updown_covid-covid_immune_shared"

edges = degs_all_updown_covid_other_immune_shared %>% 
  select(-"...1") %>% 
  inner_join(ann_vaccines_covid_other %>% filter(covid_other == "COVID"), by = join_by(Cond1 == condition)) %>% 
  inner_join(ann_vaccines_covid_other %>% filter(covid_other == "COVID") %>% select(condition), by = join_by(Cond2 == condition)) %>% 
  clean_names()

edges_mirror = edges %>% 
  select(cond1:qvalue) %>% 
  rename(cond2.1 = cond1, cond1.2 = cond2) %>% 
  rename(cond2 = cond2.1, cond1 = cond1.2) %>% 
  select(cond1, cond2, everything()) %>% 
  inner_join(ann_vaccines_covid_other %>% filter(covid_other == "COVID"), by = join_by(cond1 == condition))

edges_flourish = bind_rows(edges, edges_mirror)

nodes = ann_vaccines_covid_other 
edges_nodes = edges


my_colors <- ann_vaccines_covid_other %>% 
 mutate(color = case_when(
    type == "VLP" ~ "#7E6148FF",
    type == "LA" ~ "#E64B35FF",
    type == "CONJ" ~ "#F39B7FFF",
    type == "IN" ~ "#00A087FF",
    type == "VV" ~ "#3C5488FF",
    type == "RNA" ~ "#4DBBD5FF",
    type == "SU" ~ "#8491B4FF",
    type == "I" ~ "#DC0000FF",
    TRUE ~ NA_character_),
    color_covid_other = if_else(covid_other == "COVID", "#56cfe1", "#343a40"))

my_colors %>% select(condition, color, color_covid_other) %>% 
  write_tsv("covid_others_colors.tsv")


write_tsv(edges_nodes, file = paste0(filename, "Network_Covid_Other_sharedgenes_edges_and_nodes.tsv"))
write_tsv(edges_flourish, file = paste0(filename, "Network_Covid_Other_sharedgenes_edges.tsv"))
write.csv(edges_flourish, file = paste0(filename, "Network_Covid_Other_sharedgenes_edges.csv"), row.names = F)
write_tsv(nodes, file = paste0(filename, "Network_Covid_Other_sharedgenes_nodes.tsv"))

# Only Covid vaccination (previous or no previous infection) vs other vaccines-----
ann_vaccines_covid_other

edges = degs_all_updown_covid_other_immune_shared %>% 
  select(-"...1") %>% 
  inner_join(ann_vaccines_covid_other %>% filter(covid_other == "COVID", !regimen %in% c("I", "I-I")), by = join_by(Cond1 == condition)) %>% 
  inner_join(ann_vaccines_covid_other %>% filter(covid_other == "OTHER") %>% select(condition), by = join_by(Cond2 == condition)) %>% 
  clean_names()

edges_mirror = edges %>% 
  select(cond1:qvalue) %>% 
  rename(cond2.1 = cond1, cond1.2 = cond2) %>% 
  rename(cond2 = cond2.1, cond1 = cond1.2) %>% 
  select(cond1, cond2, everything()) %>% 
  inner_join(ann_vaccines_covid_other %>% filter(covid_other == "OTHER"), by = join_by(cond1 == condition))

edges_flourish = bind_rows(edges, edges_mirror)

nodes = ann_vaccines_covid_other 
my_colors <- ann_vaccines_covid_other %>% 
 mutate(color = case_when(
    type == "VLP" ~ "#D7B0EE",
    type == "LA" ~ "#ffb703",
    type == "CONJ" ~ "#015BA0",
    type == "IN" ~ "#76c893",
    type == "VV" ~ "#5e60ce",
    type == "RNA" ~ "#56cfe1",
    type == "SU" ~ "#b5e48c",
    type == "I" ~ "#f72585",
    TRUE ~ NA_character_),
    color_covid_other = if_else(covid_other == "COVID", "#56cfe1", "#343a40"))

my_colors %>% select(condition, color, color_covid_other) %>% 
  write_tsv("covid_others_colors.tsv")

write_tsv(edges_flourish, file = paste0(filename, "Network_Covid_Other_Vaccinesonly_sharedgenes_edges.tsv"))
write.csv(edges_flourish, file = paste0(filename, "Network_Covid_Other_Vaccinesonly_sharedgenes_edges.csv"), row.names = F)
write_tsv(nodes, file = paste0(filename, "Network_Covid_Other_sharedgenes_nodes.tsv"))

```

### COVID-COVID Network (UP + DOWN)

```{r}
# Vaccines and infection -----
ann_vaccines_covid_other

degs_all_updown_covid_immune

filename = "degs_all_updown_covid-covid_immune_shared_UPandDOWN_"

edges 

degs_all_updown_covid_immune %>% 
  inner_join(degs_all_updown_covid_immune %>% select(target = condition, genes, regulation), by = "genes") %>% 
  select(source = condition, genes, regulation)
  inner_join(ann_vaccines_covid_other %>% filter(covid_other == "COVID"), by = join_by(Cond1 == condition)) %>% 
  inner_join(ann_vaccines_covid_other %>% filter(covid_other == "COVID") %>% select(condition), by = join_by(Cond2 == condition)) %>% 
  clean_names()
```

```{r}
edges_mirror = edges %>% 
  select(cond1:qvalue) %>% 
  rename(cond2.1 = cond1, cond1.2 = cond2) %>% 
  rename(cond2 = cond2.1, cond1 = cond1.2) %>% 
  select(cond1, cond2, everything()) %>% 
  inner_join(ann_vaccines_covid_other %>% filter(covid_other == "COVID"), by = join_by(cond1 == condition))

edges_flourish = bind_rows(edges, edges_mirror)

nodes = ann_vaccines_covid_other 
edges_nodes = edges


my_colors <- ann_vaccines_covid_other %>% 
 mutate(color = case_when(
    type == "VLP" ~ "#7E6148FF",
    type == "LA" ~ "#E64B35FF",
    type == "CONJ" ~ "#F39B7FFF",
    type == "IN" ~ "#00A087FF",
    type == "VV" ~ "#3C5488FF",
    type == "RNA" ~ "#4DBBD5FF",
    type == "SU" ~ "#8491B4FF",
    type == "I" ~ "#DC0000FF",
    TRUE ~ NA_character_),
    color_covid_other = if_else(covid_other == "COVID", "#56cfe1", "#343a40"))

my_colors %>% select(condition, color, color_covid_other) %>% 
  write_tsv("covid_others_colors.tsv")


write_tsv(edges_nodes, file = paste0(filename, "Network_Covid_Other_sharedgenes_edges_and_nodes.tsv"))
write_tsv(edges_flourish, file = paste0(filename, "Network_Covid_Other_sharedgenes_edges.tsv"))
write.csv(edges_flourish, file = paste0(filename, "Network_Covid_Other_sharedgenes_edges.csv"), row.names = F)
write_tsv(nodes, file = paste0(filename, "Network_Covid_Other_sharedgenes_nodes.tsv"))

# Only Covid vaccination (previous or no previous infection) vs other vaccines-----
ann_vaccines_covid_other

edges = degs_all_updown_covid_other_immune_shared %>% 
  select(-"...1") %>% 
  inner_join(ann_vaccines_covid_other %>% filter(covid_other == "COVID", !regimen %in% c("I", "I-I")), by = join_by(Cond1 == condition)) %>% 
  inner_join(ann_vaccines_covid_other %>% filter(covid_other == "OTHER") %>% select(condition), by = join_by(Cond2 == condition)) %>% 
  clean_names()

edges_mirror = edges %>% 
  select(cond1:qvalue) %>% 
  rename(cond2.1 = cond1, cond1.2 = cond2) %>% 
  rename(cond2 = cond2.1, cond1 = cond1.2) %>% 
  select(cond1, cond2, everything()) %>% 
  inner_join(ann_vaccines_covid_other %>% filter(covid_other == "OTHER"), by = join_by(cond1 == condition))

edges_flourish = bind_rows(edges, edges_mirror)

nodes = ann_vaccines_covid_other 
my_colors <- ann_vaccines_covid_other %>% 
 mutate(color = case_when(
    type == "VLP" ~ "#D7B0EE",
    type == "LA" ~ "#ffb703",
    type == "CONJ" ~ "#015BA0",
    type == "IN" ~ "#76c893",
    type == "VV" ~ "#5e60ce",
    type == "RNA" ~ "#56cfe1",
    type == "SU" ~ "#b5e48c",
    type == "I" ~ "#f72585",
    TRUE ~ NA_character_),
    color_covid_other = if_else(covid_other == "COVID", "#56cfe1", "#343a40"))

my_colors %>% select(condition, color, color_covid_other) %>% 
  write_tsv("covid_others_colors.tsv")

write_tsv(edges_flourish, file = paste0(filename, "Network_Covid_Other_Vaccinesonly_sharedgenes_edges.tsv"))
write.csv(edges_flourish, file = paste0(filename, "Network_Covid_Other_Vaccinesonly_sharedgenes_edges.csv"), row.names = F)
write_tsv(nodes, file = paste0(filename, "Network_Covid_Other_sharedgenes_nodes.tsv"))

```

COVID and other vaccines

```{r}

ImmuneGO_Annotated_Genes
OtherGOs_Annotated_Genes


##### Table source - target - annotations ------


# Immune
immune_source1 = covid_other_immunego_genes %>% 
  inner_join(ImmuneGO_Annotated_Genes %>% filter(go_term == "Manual", 
                                                 !(process %in% c("TCR REPERTOIRE", "BCR REPERTOIRE"))), 
             by = "genes") %>% 
  select(source = condition, target = genes, value = regulation_numeric) %>% 
  distinct() %>% 
  left_join(ann_vaccines_covid_other %>% rename(source = condition), by = "source") %>% 
  mutate(covid_other = if_else(is.na(covid_other), "OTHER", covid_other),
         annotation = paste0("Condition", covid_other))

immune_source2 = covid_other_immunego_genes %>% 
  inner_join(ImmuneGO_Annotated_Genes %>% filter(go_term == "Manual", 
                                                 !(process %in% c("TCR REPERTOIRE", "BCR REPERTOIRE"))), 
             by = "genes") %>% 
  select(source = process, target = genes) %>% 
  distinct() %>% 
  mutate(value = 0, 
         annotation = "Immune system")

immune = bind_rows(immune_source1, immune_source2) %>% mutate(id = source)


# Not immune genes
notimmune_source1 = covid_other_notimmune_genes %>% 
  inner_join(OtherGOs_Annotated_Genes %>% filter(annotation == "Major process"), by = "genes") %>% 
  select(source = condition, target = genes, value = regulation_numeric) %>% 
  distinct() %>% 
  left_join(ann_vaccines_covid_other %>% rename(source = condition), by = "source") %>% 
    mutate(covid_other = if_else(is.na(covid_other), "OTHER", covid_other),
           annotation = paste0("Condition", covid_other))

notimmune_source2 = covid_other_notimmune_genes %>% 
  inner_join(OtherGOs_Annotated_Genes %>% filter(annotation == "Major process") , by = "genes") %>% 
  select(source = term, target = genes, value = regulation_numeric) %>% 
  distinct() %>% 
  mutate(value = 0, 
         annotation = "Other processes")

not_immune = bind_rows(notimmune_source1, notimmune_source2) %>% mutate(id = source)
immune_notimmune = bind_rows(not_immune, immune)
view(immune_notimmune)
write.csv(immune_notimmune, file = "Network_immune_notimmune.csv", row.names = F)

##### Tables nodes and edges ------

edges = immune_notimmune %>% 
  select(source, target, value) %>% 
  mutate(type = "Directed")

nodes = immune_notimmune %>% 
  select(source, annotation, vaccine:infection) %>% 
  distinct() %>% 
  mutate(id = source)

write.csv(edges, file = "Network_immune_notimmune_edges.csv", row.names = F)
write.csv(nodes, file = "Network_immune_notimmune_nodes.csv", row.names = F)
```

## Chord plot

### COVID-19 - All conditions
#### Genes shared by process (immune and non immune)
```{r}

all_covid_vax_atlas =  readRDS(here("DGE analysis", "all_studies_degs_weighted.rds")) %>% 
  filter(if_else(n >= 5, fdr <= 0.05, fdr <= (0.05 * 6/n))) 

degs_all_updown_covid_vax_other = read_csv(here("VaxGO gene sets","degs_all_updown_covid_vax_other.csv")) %>% 
  select(-"...1")

ann_vaccines_vaxsig_covid <- read_csv(here("Annotations and metadata", "ann_vaccines_vaxsig_covid.csv"))

annotation_conditions = read_csv(here("Annotations and metadata", "ann_vaccines_conditions.csv"), 
    col_types = cols(...1 = col_skip())) %>% 
    mutate(type = if_else(condition == "BNT−I (D0−5)", "I", type))

ImmuneGO_Annotated_GO = read_rds(here("VaxGO gene sets", "ImmuneGO_Annotated_Genes_2024-03-26.rds"))

OtherGOs_Annotated_Genes <- read_csv(here("VaxGO gene sets", "OtherGOs_Annotated_Genes.csv")) %>% 
  select(process = term_ancestor, genes = symbol) %>% 
  filter(process %in% c("Cellular",
                        "Immune system",
                        "Metabolism"))

ImmuneGO_genes_selected = ImmuneGO_BTM_grouped_genes %>% 
  filter(immune_set == "BTM and ImmuneGO") %>% 
  mutate(process = "Immune system") %>% 
  distinct() %>% 
  # filter(is.na(immune_system)) %>% 
  # filter(process %in% c("ADAPTIVE IMMUNE SYSTEM",
  #                       "IMMUNE SYSTEM",
  #                       "INNATE IMMUNE SYSTEM",
  #                       "ANTIMICROBIAL",
  #                       "LEUKOCYTES",
  #                       "COMPLEMENT")) %>% 
  bind_rows(OtherGOs_Annotated_Genes %>% 
              filter(!process == "Immune system"))

ImmuneGO_genes_selected$process %>% unique()

conflicted::conflicts_prefer(dplyr::count)
immune_genes_covid = all_covid_vax_atlas  %>% 
  inner_join(ImmuneGO_genes_selected %>% 
               select(genes), by = "genes") %>% 
  inner_join(annotation_conditions %>% 
               select( condition),
             by = "condition") %>% 
    select(condition1 = condition, genes) %>% 
  distinct()

immune_genes_covid_shared_process = immune_genes_covid %>% 
  rename(condition2 = condition1) %>% 
  inner_join(immune_genes_covid %>% 
               select(condition1, genes), by = "genes") %>% 
    filter(condition1 != condition2) %>% 
  inner_join(ImmuneGO_genes_selected %>% 
               select(genes, process), by = "genes") %>% 
  distinct() %>% 
  group_by(condition1, condition2, process) %>% 
  summarise(n = n(),
            genes = paste0(genes, collapse = ",")) %>% 
  ungroup()

immune_genes_covid_shared_process$process %>% unique()


immune_genes_covid_shared_process_genes = immune_genes_covid_shared_process %>% 
  mutate(condition_grouped = paste0(condition1, "__", condition2)) %>% 
  separate_rows(genes, sep = ",") %>% 
  select(condition = condition_grouped, genes) %>% 
  distinct()

#Input
data = immune_genes_covid_shared_process_genes %>%
  select(condition, genes)
Immune_GO_T2G = ImmuneGO_genes_selected %>% 
  select(process, genes)
  
filename = "All_covid_13Vax_overlapping_ImmuneSystem"
```

```{r}

# ORA function
perform_ORA <- function(df, Immune_GO_T2G) {
  resultados <- list()
  
  # Obter condições únicas na coluna "condition"
  condicoes <- df$condition %>% 
    unique() %>% 
    as.character()
  
  for (condicao in condicoes) {
    # Selecionar DEGs para a condição atual
    degs_condicao <- df %>% 
      filter(condition == condicao) %>%
      select(genes) %>%
      distinct()

    # Preparar gene list
    gene_list_ora <- degs_condicao$genes

    ############ IMMUNE GO
    immune_GO_ORA <- tryCatch({
      enricher(gene_list_ora, 
               TERM2GENE = Immune_GO_T2G, 
               minGSSize = 1,
               maxGSSize = 1000000,
               pvalueCutoff = 1,
               pAdjustMethod = "BH") %>% 
        as.data.frame() %>% 
        mutate(condition = condicao,
               ora_enrichment = "ImmuneGO")
    }, error = function(e) {
      return(NULL)  # Retorna NULL caso ocorra o erro
    })

    if (!is.null(immune_GO_ORA)) {
      resultados[[paste(condicao, "ImmuneGO", sep = "_")]] <- immune_GO_ORA
    }
  }

  # Combinar todos os resultados em um único dataframe
  resultado_final <- bind_rows(resultados)

  return(resultado_final)
}

Immune_GO_T2G = ImmuneGO_genes_selected %>% 
                               select(process, genes)
df_resultados <- perform_ORA(immune_genes_covid_shared_process_genes, 
                             Immune_GO_T2G)

ImmuneGO_genes_selected$process %>% unique()

df_clean = df_resultados %>% 
  clean_names() %>% 
  select(condition, count, process = id,
         pvalue,
         p_adjust,
         genes = gene_id) %>% 
  separate(condition, into = c("condition1", "condition2"), sep = "__")

write.csv(df_clean, file = here("VaxGO gene sets", paste0(filename, "_ORA_filtered.csv")))

df_clean 
```



```{r}

df_vaccines <- df_clean %>% 
  filter( p_adjust <= 0.05) %>% 
  mutate(count = if_else(count >0, 1, 0)) %>% 
  dplyr::select(Cond1 = condition1, 
                Cond2 = condition2, 
                Shared = count, 
                process)

df_vaccines$process %>% unique()

group_types = annotation_conditions %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(df_vaccines,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)

types = df_vaccines %>% 
  dplyr::select(condition = Cond1) %>% 
  distinct() %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(condition, type), by = "condition") %>% 
  dplyr::select(type)

colors = ann_vaxsig_colors$Platform %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(condition, type), by = "type") %>% 
  select(condition, color)

col_immune = c("ADAPTIVE IMMUNE SYSTEM" = "#6817c5",
          "COMPLEMENT" = "#2a9d8f",
          "INNATE IMMUNE SYSTEM" = "#0099ff",
          "IMMUNE SYSTEM" = "#435804",
          "ANTIMICROBIAL" = "#38b000",
          
          #General processes
          "Immune system" = "blue",
          "Cellular" = "green",
          "Metabolism" = "red")

colors_grid = df_vaccines %>%
  rename(condition = Cond1) %>% 
 inner_join(colors, by = "condition") %>% 
 pull(color, condition)

col_links = df_vaccines %>% 
  inner_join(col_immune %>% 
  as.data.frame() %>% 
  rownames_to_column("process") %>% 
  rename("color" = "."), by = "process") %>% 
  pull(color)
```

Select links
```{r}
###### Select links

# Week 1
selected_links_W1 = df_vaccines %>% 
  ungroup() %>% 
  select(Cond1, Cond2) %>% 
  inner_join(annotation_conditions %>% 
              select(Cond1 = condition, week) %>% 
               filter(week == 1),
             by = "Cond1") %>% 
  select(Cond1) %>% 
  mutate(selected_Cond1 = Cond1) %>% 
  distinct()

df_vaccines_selected_W1 = df_vaccines %>% 
  left_join(selected_links_W1 %>% 
              rename(Cond2 = Cond1,
                     selected_Cond2 = selected_Cond1),
             by = "Cond2") %>% 
  left_join(selected_links_W1,
             by = "Cond1") %>% 
  mutate(selected_TF = if_else(!is.na(selected_Cond1), T, F),
         Shared = if_else(!is.na(selected_Cond2), Shared, 0))

selected_links_vector_W1 = df_vaccines_selected_W1 %>% 
  pull(selected_TF)

df_W1 = df_vaccines_selected_W1 %>% 
  select(Cond1:Shared)

# Week 2

selected_links_W2 = df_vaccines %>% 
  ungroup() %>% 
  select(Cond1, Cond2) %>% 
  inner_join(annotation_conditions %>% 
              select(Cond1 = condition, week) %>% 
               filter(week == 2),
             by = "Cond1") %>% 
  select(Cond1) %>% 
  mutate(selected_Cond1 = Cond1) %>% 
  distinct()

df_vaccines_selected_W2 = df_vaccines %>% 
  left_join(selected_links_W2 %>% 
              rename(Cond2 = Cond1,
                     selected_Cond2 = selected_Cond1),
             by = "Cond2") %>% 
  left_join(selected_links_W2,
             by = "Cond1") %>% 
  mutate(selected_TF = if_else(!is.na(selected_Cond1), T, F),
         Shared = if_else(!is.na(selected_Cond2), Shared, 0))

selected_links_vector_W2 = df_vaccines_selected_W2 %>% 
  pull(selected_TF)

df_W2 = df_vaccines_selected_W2 %>% 
  select(Cond1:Shared)


# Week 2>

selected_links_W2plus = df_vaccines %>% 
  ungroup() %>% 
  select(Cond1, Cond2) %>% 
  inner_join(annotation_conditions %>% 
              select(Cond1 = condition, week) %>% 
               filter(week > 2),
             by = "Cond1") %>% 
  select(Cond1) %>% 
  mutate(selected_Cond1 = Cond1) %>% 
  distinct()

df_vaccines_selected_W2plus = df_vaccines %>% 
  left_join(selected_links_W2 %>% 
              rename(Cond2 = Cond1,
                     selected_Cond2 = selected_Cond1),
             by = "Cond2") %>% 
  left_join(selected_links_W2plus,
             by = "Cond1") %>% 
  mutate(selected_TF = if_else(!is.na(selected_Cond1), T, F),
         Shared = if_else(!is.na(selected_Cond2), Shared, 0))

selected_links_vector_W2plus = df_vaccines_selected_W2plus %>% 
  pull(selected_TF)

df_W2plus = df_vaccines_selected_W2plus %>% 
  select(Cond1:Shared)

# Week 4

selected_links_W4 = df_vaccines %>% 
  ungroup() %>% 
  select(Cond1, Cond2) %>% 
  inner_join(annotation_conditions %>% 
              select(Cond1 = condition, week) %>% 
               filter(week == 4),
             by = "Cond1") %>% 
  select(Cond1) %>% 
  mutate(selected_Cond1 = Cond1) %>% 
  distinct()

df_vaccines_selected_W4 = df_vaccines %>% 
  left_join(selected_links_W4 %>% 
              rename(Cond2 = Cond1,
                     selected_Cond2 = selected_Cond1),
             by = "Cond2") %>% 
  left_join(selected_links_W4,
             by = "Cond1") %>% 
  mutate(selected_TF = if_else(!is.na(selected_Cond1), T, F),
         Shared = if_else(!is.na(selected_Cond2), Shared, 0))

selected_links_vector_W4 = df_vaccines_selected_W4 %>% 
  pull(selected_TF)

df_W4 = df_vaccines_selected_W4 %>% 
  select(Cond1:Shared)

# Week 7

selected_links_W7 = df_vaccines %>% 
  ungroup() %>% 
  select(Cond1, Cond2) %>% 
  inner_join(annotation_conditions %>% 
              select(Cond1 = condition, week) %>% 
               filter(week >= 7),
             by = "Cond1") %>% 
  select(Cond1) %>% 
  mutate(selected_Cond1 = Cond1) %>% 
  distinct()

df_vaccines_selected_W7 = df_vaccines %>% 
  left_join(selected_links_W7 %>% 
              rename(Cond2 = Cond1,
                     selected_Cond2 = selected_Cond1),
             by = "Cond2") %>% 
  left_join(selected_links_W7,
             by = "Cond1") %>% 
  mutate(selected_TF = if_else(!is.na(selected_Cond1), T, F),
         Shared = if_else(!is.na(selected_Cond2), Shared, 0))

selected_links_vector_W7 = df_vaccines_selected_W7 %>% 
  pull(selected_TF)

df_W7 = df_vaccines_selected_W7 %>% 
  select(Cond1:Shared)


#All
selected_links_all = df_vaccines %>% 
  ungroup() %>% 
  select(Cond1, Cond2) %>% 
  inner_join(annotation_conditions %>% 
              select(Cond1 = condition, week),
             by = "Cond1") %>% 
  select(Cond1) %>% 
  mutate(selected_Cond1 = Cond1) %>% 
  distinct()

df_vaccines_selected_all = df_vaccines %>% 
  left_join(selected_links_all %>% 
              rename(Cond2 = Cond1,
                     selected_Cond2 = selected_Cond1),
             by = "Cond2") %>% 
  left_join(selected_links_all,
             by = "Cond1") %>% 
  mutate(selected_TF = if_else(!is.na(selected_Cond1), T, F),
         Shared = if_else(!is.na(selected_Cond2), Shared, 0))

selected_links_vector_all = df_vaccines_selected_all %>% 
  pull(selected_TF)

df_all = df_vaccines_selected_all %>% 
  select(Cond1:Shared)
```


#### All and by week

All
```{r fig.height=10, fig.width=10}
#Plot
pdf(file = here("Figures","ChordDiagram_Covid19_AllCovid_By_Process_Immune_PAdjusted_All.pdf"), 
    width = 10, 
    height = 10)
#Week 1
chordDiagram(df_all, 
             grid.col = colors_grid, 
             col = col_links,
             link.sort = TRUE,
             link.decreasing = TRUE,
             annotationTrack = "grid", 
             directional = 1,
             scale = F,
             big.gap = 3,
             transparency = 0,
             small.gap = 0.5,
             h.ratio = 0.7,
             link.visible = selected_links_vector_all,
             group = group_types,
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(unique(df_vaccines$Cond1)))))
             )
circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.8, #Label size
      font = 0.1)},#Label face
  bg.border = NA) 
title("All")

dev.off()
circos.clear()
```

Weeks 1-7
```{r fig.height=10, fig.width=10}

#Plot
pdf(file = here("Figures","ChordDiagram_Covid19_AllCovid_By_Process_Weeks_Immune_PAdjusted.pdf"), 
    width = 20, 
    height = 5)

par(mfrow = c(1, 4))
#Week 1
chordDiagram(df_W1, 
             grid.col = colors_grid, 
             col = col_links,
             link.sort = TRUE,
             link.decreasing = TRUE,
             annotationTrack = "grid", 
             directional = 0,
             scale = TRUE,
             big.gap = 3,
             transparency = 0.3,
             small.gap = 0.5,
             h.ratio = 0.7,
             link.visible = selected_links_vector_W1,
             group = group_types,
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(unique(df_vaccines$Cond1)))))
             )
circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.8, #Label size
      font = 0.1)},#Label face
  bg.border = NA) 
title("Week 1")


#Week 2
chordDiagram(df_W2, 
             grid.col = colors_grid, 
             col = col_links,
             link.sort = TRUE,
             link.decreasing = TRUE,
             scale = TRUE,
             annotationTrack = "grid", 
             directional = 0,
             big.gap = 3,
             transparency = 0,
             small.gap = 0.5,
             h.ratio = 0.7,
             link.visible = selected_links_vector_W2,
             group = group_types,
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(unique(df_vaccines$Cond1)))))
             )
circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.8, #Label size
      font = 0.1)},#Label face
  bg.border = NA) 
title("Week 2")

#Week 4
chordDiagram(df_W4, 
             grid.col = colors_grid, 
             col = col_links,
             link.sort = TRUE,
             link.decreasing = TRUE,
             scale = TRUE,
             annotationTrack = "grid", 
             directional = 0,
             big.gap = 3,
             transparency = 0,
             small.gap = 0.5,
             h.ratio = 0.7,
             link.visible = selected_links_vector_W4,
             group = group_types,
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(unique(df_vaccines$Cond1)))))
             )
circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.8, #Label size
      font = 0.1)},#Label face
  bg.border = NA) 
title("Week 4")


#Week 8
chordDiagram(df_W7,
             grid.col = colors_grid,
             col = col_links,
             # link.sort = TRUE,
             # link.decreasing = TRUE,
             annotationTrack = "grid",
             scale = TRUE,
             directional = 0,
             big.gap = 3,
             transparency = 0,
             small.gap = 0.5,
             h.ratio = 0.7,
             link.visible = selected_links_vector_W7,
             group = group_types,
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(unique(df_vaccines$Cond1)))))
             )
circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.8, #Label size
      font = 0.1)},#Label face
  bg.border = NA)
title("Week 8")

dev.off()
circos.clear()
```
Weeks 1 and 2+
```{r fig.height=10, fig.width=10}

#Plot
pdf(file = here("Figures","ChordDiagram_Covid19_AllCovid_By_Process_Weeks_1_2_more_Immune_PAdjusted.pdf"), 
    width = 15, 
    height = 10)

par(mfrow = c(1, 2))
#Week 1
chordDiagram(df_W1, 
             grid.col = colors_grid, 
             col = col_links,
             link.sort = F,
             link.decreasing = F,
             annotationTrack = "grid", 
             directional = 1,
             scale = F, 
             big.gap = 3,
             transparency = 0,
             small.gap = 0.5,
             h.ratio = 0.7,
             link.visible = selected_links_vector_W1,
             group = group_types,
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(unique(df_vaccines$Cond1)))))
             )
circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.8, #Label size
      font = 0.1)},#Label face
  bg.border = NA) 
title("Week 1")


#Week 2
chordDiagram(df_W2plus, 
             grid.col = colors_grid, 
             col = col_links,
             link.sort = F,
             link.decreasing = F,
             scale = F,
             annotationTrack = "grid", 
             directional = 1,
             big.gap = 3,
             transparency = 0,
             small.gap = 0.5,
             h.ratio = 0.7,
             link.visible = selected_links_vector_W2plus,
             group = group_types,
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(unique(df_vaccines$Cond1)))))
             )
circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.8, #Label size
      font = 0.1)},#Label face
  bg.border = NA) 
title("Week 2+")

dev.off()
circos.clear()
```


### COVID-19 VS OTHER 

#### Genes shared by process
```{r}
degs_all_updown_covid_vax_other = read_csv("Example/degs_all_updown_covid_vax_other.csv") %>% 
  select(-"...1")
ann_vaccines_vaxsig_covid <- read_csv("Example/ann_vaccines_vaxsig_covid.csv") %>% 
  filter(disease_vac != "I")

degs_all_updown_covid_vax_other = degs_all_updown_covid_vax_other %>% 
  inner_join(ann_vaccines_vaxsig_covid %>% 
               select(condition),
             by = "condition") 

ImmuneGO_genes_selected = ImmuneGO_Annotated_GO %>% 
  filter(go_term == "Manual",
         process %in% c("ADAPTIVE IMMUNE SYSTEM",
                        "INNATE IMMUNE SYSTEM",
                        "ANTIMICROBIAL",
                        "GENERAL",
                        "COMPLEMENT")) %>% 
  separate_rows(genes,sep = ",")


immune_genes_covid = degs_all_updown_covid_vax_other  %>% 
  inner_join(ImmuneGO_genes_selected %>% 
               select(genes), by = "genes") %>% 
  select(condition1 = condition, genes)

immune_genes_covid_shared_process = immune_genes_covid %>% 
  rename(condition2 = condition1) %>% 
  inner_join(immune_genes_covid %>% 
               select(condition1, genes), by = "genes") %>% 
    filter(condition1 != condition2) %>% 
  inner_join(ImmuneGO_genes_selected %>% 
               select(genes, process), by = "genes") %>% 
  distinct() %>% 
  group_by(condition1, condition2, process) %>% 
  summarise(n = n())
```


```{r}
df_vaccines <- immune_genes_covid_shared_process %>% 
  dplyr::select(Cond1 = condition1, 
                Cond2 = condition2, 
                Shared = n, 
                process) 

group_types = ann_vaccines_vaxsig_covid %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(df_vaccines,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)

types = df_vaccines %>% 
  dplyr::select(condition = Cond1) %>% 
  distinct() %>% 
  inner_join(ann_vaccines_vaxsig_covid %>% 
 dplyr::select(condition, type), by = "condition") %>% 
  dplyr::select(type)

colors = ann_vaxsig_colors$Platform %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(ann_vaccines_vaxsig_covid %>% 
 dplyr::select(condition, type), by = "type") %>% 
  select(condition, color)

colors_grid = df_vaccines %>%
  rename(condition = Cond1) %>% 
 inner_join(colors, by = "condition") %>% 
 pull(color, condition)

col_links = df_vaccines %>% 
  inner_join(col_immune %>% 
  as.data.frame() %>% 
  rownames_to_column("process") %>% 
  rename("color" = "."), by = "process") %>% 
  pull(color)
```

All

```{r fig.height=10, fig.width=10}

#Plot
png(filename = here("Figures","ChordDiagram_Covid19_Other_By_Process.png"),
    width = 10, height = 10, units = "in", res = 300)

chordDiagram(df_vaccines, 
             grid.col = colors_grid, 
             col = col_links, scale = TRUE,
             # link.sort = TRUE, 
             # link.decreasing = TRUE,
             annotationTrack = "grid", 
             directional = 1,
             big.gap = 2,
             transparency = 0,
             small.gap = 0.5,
             h.ratio = 0.7,
             link.sort = TRUE, 
             link.decreasing = TRUE,
             group = group_types,
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(unique(df_vaccines$Cond1)))))
             )
circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.8, #Label size
      font = 0.1)},#Label face
  bg.border = NA) 
title("All")

dev.off()
circos.clear()
```


Select links
```{r}
###### Select links

# Week 1

selected_links_W1 = df_vaccines %>% 
  ungroup() %>% 
  select(Cond1, Cond2) %>% 
  inner_join(annotation_conditions %>% 
              select(Cond1 = condition, week) %>% 
               filter(week == 1),
             by = "Cond1") %>% 
  select(Cond1) %>% 
  mutate(selected_Cond1 = Cond1) %>% 
  distinct()

df_vaccines_selected_W1 = df_vaccines %>% 
  left_join(selected_links_W1 %>% 
              rename(Cond2 = Cond1,
                     selected_Cond2 = selected_Cond1),
             by = "Cond2") %>% 
  left_join(selected_links_W1,
             by = "Cond1") %>% 
  mutate(selected_TF = if_else(!is.na(selected_Cond1), T, F),
         Shared = if_else(!is.na(selected_Cond2), Shared, 0))

selected_links_vector_W1 = df_vaccines_selected_W1 %>% 
  pull(selected_TF)

df_W1 = df_vaccines_selected_W1 %>% 
  select(Cond1:Shared)

# Week 2

selected_links_W2 = df_vaccines %>% 
  ungroup() %>% 
  select(Cond1, Cond2) %>% 
  inner_join(annotation_conditions %>% 
              select(Cond1 = condition, week) %>% 
               filter(week == 2),
             by = "Cond1") %>% 
  select(Cond1) %>% 
  mutate(selected_Cond1 = Cond1) %>% 
  distinct()

df_vaccines_selected_W2 = df_vaccines %>% 
  left_join(selected_links_W2 %>% 
              rename(Cond2 = Cond1,
                     selected_Cond2 = selected_Cond1),
             by = "Cond2") %>% 
  left_join(selected_links_W2,
             by = "Cond1") %>% 
  mutate(selected_TF = if_else(!is.na(selected_Cond1), T, F),
         Shared = if_else(!is.na(selected_Cond2), Shared, 0))

selected_links_vector_W2 = df_vaccines_selected_W2 %>% 
  pull(selected_TF)

df_W2 = df_vaccines_selected_W2 %>% 
  select(Cond1:Shared)

# Week 4

selected_links_W4 = df_vaccines %>% 
  ungroup() %>% 
  select(Cond1, Cond2) %>% 
  inner_join(annotation_conditions %>% 
              select(Cond1 = condition, week) %>% 
               filter(week == 4),
             by = "Cond1") %>% 
  select(Cond1) %>% 
  mutate(selected_Cond1 = Cond1) %>% 
  distinct()

df_vaccines_selected_W4 = df_vaccines %>% 
  left_join(selected_links_W4 %>% 
              rename(Cond2 = Cond1,
                     selected_Cond2 = selected_Cond1),
             by = "Cond2") %>% 
  left_join(selected_links_W4,
             by = "Cond1") %>% 
  mutate(selected_TF = if_else(!is.na(selected_Cond1), T, F),
         Shared = if_else(!is.na(selected_Cond2), Shared, 0))

selected_links_vector_W4 = df_vaccines_selected_W4 %>% 
  pull(selected_TF)

df_W4 = df_vaccines_selected_W4 %>% 
  select(Cond1:Shared)

# Week 7

selected_links_W7 = df_vaccines %>% 
  ungroup() %>% 
  select(Cond1, Cond2) %>% 
  inner_join(annotation_conditions %>% 
              select(Cond1 = condition, week) %>% 
               filter(week == 7),
             by = "Cond1") %>% 
  select(Cond1) %>% 
  mutate(selected_Cond1 = Cond1) %>% 
  distinct()

df_vaccines_selected_W7 = df_vaccines %>% 
  left_join(selected_links_W7 %>% 
              rename(Cond2 = Cond1,
                     selected_Cond2 = selected_Cond1),
             by = "Cond2") %>% 
  left_join(selected_links_W7,
             by = "Cond1") %>% 
  mutate(selected_TF = if_else(!is.na(selected_Cond1), T, F),
         Shared = if_else(!is.na(selected_Cond2), Shared, 0))

selected_links_vector_W7 = df_vaccines_selected_W7 %>% 
  pull(selected_TF)

df_W7 = df_vaccines_selected_W7 %>% 
  select(Cond1:Shared)
```


By week

```{r fig.height=10, fig.width=10}

#Plot
png(filename = here("Figures","ChordDiagram_Covid19_Other_By_Process.png"),
    width = 20, height = 20, units = "in", res = 600)

# par(mfrow = c(2, 2))
#Week 1
chordDiagram(df_W1, 
             grid.col = colors_grid, 
             col = col_links,
             # link.sort = TRUE, 
             # link.decreasing = TRUE,
             annotationTrack = "grid", 
             directional = 1,
             big.gap = 6,
             transparency = 0,
             small.gap = 3,
             h.ratio = 0.7,
             link.visible = selected_links_vector_W1,
             group = group_types,
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(unique(df_vaccines$Cond1)))))
             )
circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 1),
      cex = 0.8, #Label size
      font = 0.1)},#Label face
  bg.border = NA) 
title("Week 1")


#Week 2
chordDiagram(df_W2, 
             grid.col = colors_grid, 
             col = col_links,
             # link.sort = TRUE, 
             # link.decreasing = TRUE,
             annotationTrack = "grid", 
             directional = 1,
             big.gap = 6,
             transparency = 0,
             small.gap = 3,
             h.ratio = 0.7,
             link.visible = selected_links_vector_W2,
             group = group_types,
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(unique(df_vaccines$Cond1)))))
             )
circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 1),
      cex = 0.8, #Label size
      font = 0.1)},#Label face
  bg.border = NA) 
title("Week 2")

#Week 4
chordDiagram(df_W4, 
             grid.col = colors_grid, 
             col = col_links,
             # link.sort = TRUE,
             # link.decreasing = TRUE,
             annotationTrack = "grid", 
             directional = 1,
             big.gap = 6,
             transparency = 0,
             small.gap = 3,
             h.ratio = 0.7,
             link.visible = selected_links_vector_W4,
             group = group_types,
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(unique(df_vaccines$Cond1)))))
             )
circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 1),
      cex = 0.8, #Label size
      font = 0.1)},#Label face
  bg.border = NA) 
title("Week 4")


#Week 7
chordDiagram(df_W7, 
             grid.col = colors_grid, 
             col = col_links,
             # link.sort = TRUE, 
             # link.decreasing = TRUE,
             annotationTrack = "grid", 
             directional = 1,
             big.gap = 6,
             transparency = 0,
             small.gap = 3,
             h.ratio = 0.7,
             link.visible = selected_links_vector_W7,
             group = group_types,
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(unique(df_vaccines$Cond1)))))
             )
circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 1),
      cex = 0.8, #Label size
      font = 0.1)},#Label face
  bg.border = NA) 
title("Week 7")

dev.off()
circos.clear()
```


#### Jaccard distance
```{r}
ann_vaccines_vaxsig_covid <- read_csv("Example/ann_vaccines_vaxsig_covid.csv")

VAXSigDB_sharedgenes <- read_csv(file = here("Example", "degs_all_updown_covid_vax_other_sharedgenes_Fisher_Jaccard.csv"))

annotation_conditions = ann_vaccines_vaxsig_covid %>% 
  filter(disease_vac != "I")
```

#####  Types
Prepare tables
```{r fig.height= 10, fig.width=10}
#Types
###### All -----
links_df = shared_genes_df_mirror %>% 
  dplyr::select(Cond1, Cond2, jaccard_distance) %>% 
  inner_join(annotation_conditions %>% 
               select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(annotation_conditions %>% 
               select(Cond2 = condition),
             by = "Cond2") 

mat_types = links_df %>% 
  pivot_wider(names_from = "Cond2", values_from = "jaccard_distance", values_fill = list(jaccard_distance = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix() %>% 
  replace(is.na(.), 0)

grid.col_types = ann_vaxsig_colors$Platform %>%
  as.data.frame() %>%
  rownames_to_column(color_flourish) %>%
  rename(color = ".") %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df %>% 
               select(Cond1) %>% 
               distinct(),
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_types = annotation_conditions %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)

```

Plot
All conditions and stats
```{r fig.height= 5, fig.width=5}
#Plot

#Types

png(filename = here("Figures","ChordDiagram_Covid19_Other_Vaccines_Types_All_jaccard_distance.png"), 
    width = 10, height = 10, units = "in", res = 600)

#All
chordDiagram(mat_types, 
             grid.col = grid.col_types,
             annotationTrack = "grid",  
             transparency = 0,
             link.sort = TRUE, link.decreasing = TRUE,
             scale=TRUE,
             big.gap = 3,
             small.gap = 0.5,
             directional = -1, 
             group = group_types,
             link.zindex = rank(mat_types),
             h.ratio = 0.7, 
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_types))))))

circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.5, #Label size
      font = 0.1
      )},
  bg.border = NA)
title("Vax MsigDB collection - Jaccard distance")

dev.off()
circos.clear()
```

#### Shared total genes
```{r, cache = TRUE}
ann_vaccines_vaxsig_covid <- read_csv("Example/ann_vaccines_vaxsig_covid.csv")

VAXSigDB_sharedgenes <- read_csv(file = here("Example", "degs_all_updown_covid_vax_other_sharedgenes_Fisher_Jaccard.csv"))

annotation_conditions = ann_vaccines_vaxsig_covid %>% 
  filter(disease_vac != "I")
```

##### Types
Prepare tables
```{r fig.height= 10, fig.width=10, cache = TRUE}
#Types
###### All -----
links_df = shared_genes_df_mirror %>% 
  dplyr::select(Cond1, Cond2, Shared) %>% 
  inner_join(annotation_conditions %>% 
               select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(annotation_conditions %>% 
               select(Cond2 = condition),
             by = "Cond2") 

mat_types = links_df %>% 
  pivot_wider(names_from = "Cond2", values_from = "Shared", values_fill = list(jaccard_distance = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix() %>% 
  replace(is.na(.), 0)

grid.col_types = ann_vaxsig_colors$Platform %>%
  as.data.frame() %>%
  rownames_to_column(color_flourish) %>%
  rename(color = ".") %>% 
  inner_join(annotation_conditions %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df %>% 
               select(Cond1) %>% 
               distinct(),
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_types = annotation_conditions %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)

```

Plot
All conditions and stats
```{r fig.height= 5, fig.width=5, cache = TRUE}
#Plot

#Types

png(filename = here("Figures","ChordDiagram_Covid19_Other_Vaccines_Types_All_SharedTotal.png"), 
    width = 10, height = 10, units = "in", res = 600)

#All
chordDiagram(mat_types, 
             grid.col = grid.col_types,
             annotationTrack = "grid",  
             transparency = 0,
             link.sort = TRUE, link.decreasing = TRUE,
             scale=TRUE,
             big.gap = 3,
             small.gap = 0.5,
             directional = -1, 
             group = group_types,
             link.zindex = rank(mat_types),
             h.ratio = 0.7, 
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_types))))))

circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.5, #Label size
      font = 0.1
      )},
  bg.border = NA)
title("Vax MsigDB collection - Shared total")

dev.off()
circos.clear()
```




### Shared genes (Cytoscape and Flourish)

```{r cache = TRUE}
# Vaccines and infection -----
ann_vaccines_covid_other

degs_all_updown_covid_other_immune_shared
degs_all_updown_covid_other_nonimmune_shared

filename = "degs_all_updown_covid_other_immune_shared"

edges = degs_all_updown_covid_other_immune_shared %>% 
  select(-"...1") %>% 
  inner_join(ann_vaccines_covid_other %>% filter(covid_other == "COVID"), by = join_by(Cond1 == condition)) %>% 
  inner_join(ann_vaccines_covid_other %>% filter(covid_other == "OTHER") %>% select(condition), by = join_by(Cond2 == condition)) %>% 
  clean_names()

edges_mirror = edges %>% 
  select(cond1:qvalue) %>% 
  rename(cond2.1 = cond1, cond1.2 = cond2) %>% 
  rename(cond2 = cond2.1, cond1 = cond1.2) %>% 
  select(cond1, cond2, everything()) %>% 
  inner_join(ann_vaccines_covid_other %>% filter(covid_other == "OTHER"), by = join_by(cond1 == condition))

edges_flourish = bind_rows(edges, edges_mirror)

nodes = ann_vaccines_covid_other 
edges_nodes = edges


my_colors <- ann_vaccines_covid_other %>% 
 mutate(color = case_when(
    type == "VLP" ~ "#7E6148FF",
    type == "LA" ~ "#E64B35FF",
    type == "CONJ" ~ "#F39B7FFF",
    type == "IN" ~ "#00A087FF",
    type == "VV" ~ "#3C5488FF",
    type == "RNA" ~ "#4DBBD5FF",
    type == "SU" ~ "#8491B4FF",
    type == "I" ~ "#DC0000FF",
    TRUE ~ NA_character_),
    color_covid_other = if_else(covid_other == "COVID", "#56cfe1", "#343a40"))

my_colors %>% select(condition, color, color_covid_other) %>% 
  write_tsv("covid_others_colors.tsv")


write_tsv(edges_nodes, file = paste0(filename, "Network_Covid_Other_sharedgenes_edges_and_nodes.tsv"))
write_tsv(edges_flourish, file = paste0(filename, "Network_Covid_Other_sharedgenes_edges.tsv"))
write.csv(edges_flourish, file = paste0(filename, "Network_Covid_Other_sharedgenes_edges.csv"), row.names = F)
write_tsv(nodes, file = paste0(filename, "Network_Covid_Other_sharedgenes_nodes.tsv"))

# Only Covid vaccination (previous or no previous infection) vs other vaccines-----
ann_vaccines_covid_other

edges = degs_all_updown_covid_other_immune_shared %>% 
  select(-"...1") %>% 
  inner_join(ann_vaccines_covid_other %>% filter(covid_other == "COVID", !regimen %in% c("I", "I-I")), by = join_by(Cond1 == condition)) %>% 
  inner_join(ann_vaccines_covid_other %>% filter(covid_other == "OTHER") %>% select(condition), by = join_by(Cond2 == condition)) %>% 
  clean_names()

edges_mirror = edges %>% 
  select(cond1:qvalue) %>% 
  rename(cond2.1 = cond1, cond1.2 = cond2) %>% 
  rename(cond2 = cond2.1, cond1 = cond1.2) %>% 
  select(cond1, cond2, everything()) %>% 
  inner_join(ann_vaccines_covid_other %>% filter(covid_other == "OTHER"), by = join_by(cond1 == condition))

edges_flourish = bind_rows(edges, edges_mirror)

nodes = ann_vaccines_covid_other 
my_colors <- ann_vaccines_covid_other %>% 
 mutate(color = case_when(
    type == "VLP" ~ "#7E6148FF",
    type == "LA" ~ "#c8b6ff",
    type == "CONJ" ~ "#F39B7FFF",
    type == "IN" ~ "#00A087FF",
    type == "VV" ~ "#3C5488FF",
    type == "RNA" ~ "#4DBBD5FF",
    type == "SU" ~ "#8491B4FF",
    type == "I" ~ "#DC0000FF",
    TRUE ~ NA_character_),
    color_covid_other = if_else(covid_other == "COVID", "#56cfe1", "#343a40"))

my_colors %>% select(condition, color, color_covid_other) %>% 
  write_tsv("covid_others_colors.tsv")

#Copiar e colar direto no flourish
my_colors %>% select(condition, color) %>% mutate(combined = paste(condition, color, sep = ": ")) %>% select(combined)

write_tsv(edges_flourish, file = paste0(filename, "Network_Covid_Other_Vaccinesonly_sharedgenes_edges.tsv"))
write.csv(edges_flourish, file = paste0(filename, "Network_Covid_Other_Vaccinesonly_sharedgenes_edges.csv"), row.names = F)
write_tsv(nodes, file = paste0(filename, "Network_Covid_Other_sharedgenes_nodes.tsv"))


```




# 12. 13 Vax Atlas


## Standardize tables
```{r}
IS2_FDR05_genes <- readRDS(here("13VaxAtlas", "IS2_FDR05_genes.rds"))

hagan_vaxatlas = IS2_FDR05_genes %>% 
  clean_names() %>% 
  rename(day = time_post_last_vax,
         genes = gene, type = vaccine_type, 
         wavg_fc = weighted_avg_fc,
         vaccine = pvt) %>% 
  mutate(day = if_else(day <= 1, 1, day),
         week = ceiling(day / 7),
         day_f = as_factor(day),
         day_f = fct_reorder(day_f, day),
         day = day_f,
         type = case_when(type == "LV" ~ "LA",
                          type == "IN/RP" ~ "IN/SU",
                          type == "RP" ~ "SU",
                          type == "CJ" ~"CONJ",
                          type == "RVV" ~ "VV",
                          .default = type),
         log2fold_change = wavg_fc,
         # log2fold_change = if_else(wavg_fc < 0, 
         #                  abs(wavg_fc) %>% log2(), 
         #                  log2(wavg_fc)),
         direction = case_when(log2fold_change >=1 ~ "UP",
                               log2fold_change <= -1 ~"DOWN",
                               .default = "NEUTRAL"),
         condition = paste0(pathogen,"/", type, " (D", day, ")"))

hagan_vaxatlas_annotation = hagan_vaxatlas %>% 
  select(day, pathogen, type, condition, vaccine, week) %>% 
  mutate(set = "13Vax Atlas",
         dose = "1",
         week = as.factor(week),
         disease_vac = "V") %>% 
  distinct()


all_degs_p_05_vac_infected = readRDS(here("DGE analysis", "all_studies_degs_weighted.rds")) %>% 
  filter(if_else(n >= 5, fdr <= 0.05, fdr <= (0.05 * 6/n))) 

degs_all_updown = all_degs_p_05_vac_infected

ann_vaccines = read_csv(here("Annotations and metadata", "ann_vaccines_conditions.csv"))

ann_vaccines_conditions = ann_vaccines

ann_vaccines_covid_other_conditions = ann_vaccines_conditions %>%
  select(condition, vaccine, day, week, dose, infection, type, disease_vac) %>% 
  mutate(day = as.factor(day),
         week = as.factor(week),
         dose = as.factor(dose),
         set = "COVID-19") %>% 
  bind_rows(hagan_vaxatlas_annotation)

# ann_vaccines_covid_other_conditions %>% 
#   write_csv(file = here("Annotations and metadata", "ann_vaccines_covid_other_conditions.csv"))

# Annotation ------
filename = "All_covid_13Vax"

degs_covid = degs_all_updown %>% 
  select(condition, genes, log2fold_change) #16303 unique genes

degs_hagan = hagan_vaxatlas %>%
  filter(fdr <= 0.05) %>% 
  select(condition, genes, log2fold_change) #8,829 unique genes


common_genes = degs_covid %>% 
  select(genes) %>% distinct() %>% 
  inner_join(degs_hagan %>% select(genes) %>% distinct()) #8045
  

degs_covid_vax = degs_covid %>% 
  bind_rows(degs_hagan) 
  # filter(genes %in% c(common_genes$genes %>% unique())) #8045

degs_covid_vax$genes %>% unique() %>% length()

degs_covid_vax %>% 
  saveRDS(file = here("VaxGO gene sets", "degs_covid_vaxhagan.rds"))

```


Inputs
```{r}
#Input
degs_covid_vax = read_rds(file = here("VaxGO gene sets", "degs_covid_vaxhagan.rds")) %>% 
  mutate(direction = case_when(log2fold_change >= 1 ~ "UP",
                               log2fold_change <= -1 ~ "DOWN",
                               TRUE ~ "NEUTRAL"))

ImmuneGO_BTM_grouped_genes <- read_csv("VaxGO gene sets/ImmuneGO_BTM_grouped_genes.csv")
immunego_btm_genes = ImmuneGO_BTM_grouped_genes

all_degs_p_05_vac_infected = readRDS(here("DGE analysis", "all_studies_degs_weighted.rds")) %>% 
  filter(if_else(n >= 5, fdr <= 0.05, fdr <= (0.05 * 6/n))) 


ann_vaccines = read_csv(here("Annotations and metadata", "ann_vaccines_conditions.csv"))
ann_vaccines_conditions_ordered <- readRDS(here("Annotations and metadata", "ann_vaccines_conditions_ordered.rds"))
ann_vaccines_conditions = ann_vaccines

ann_vaccines_covid_other_conditions = read_csv(file = here("Annotations and metadata", "ann_vaccines_covid_other_conditions.csv")) %>% 
  mutate(day = factor(day) %>%
           fct_reorder(as.numeric(as.character(day))),
        dose = factor(dose) %>% 
           fct_reorder(as.numeric(as.character(dose))),
        week = factor(week) %>% 
           fct_reorder(as.numeric(as.character(week))))


ann_vaccines_covid_other_conditions_ordered = ann_vaccines_covid_other_conditions %>% 
  mutate(day = as.character(day) %>% 
           as.integer()) %>% 
  arrange(vaccine, day)

ann_vaccines_covid_other_conditions_ordered = ann_vaccines_covid_other_conditions %>% 
  mutate(condition = fct_relevel(condition, 
                            unique(ann_vaccines_covid_other_conditions_ordered$condition)))

ann_vaccines_covid_other_conditions_ordered$condition %>% levels()
```





## DEGs analysis

### Up and down-regulated DEGs

```{r}

all_degs = degs_covid_vax %>%
  mutate(direction = case_when(log2fold_change >= 1 ~ "UP",
                               log2fold_change <= -1 ~ "DOWN",
                               TRUE ~ "NEUTRAL")) %>% 
  group_by(condition, direction) %>% 
  summarise(n_degs = n()) %>% 
  inner_join(ann_vaccines_covid_other_conditions %>% 
               filter(set == "13Vax Atlas"), by="condition") %>% 

  ungroup() %>% 
  mutate(condition = fct_relevel(condition, ann_vaccines_covid_other_conditions_ordered$condition %>% levels())) %>% 
  arrange(condition)

all_degs_total = all_degs %>% 
  pivot_wider(names_from = "direction", values_from = "n_degs") %>% 
  mutate_all(~replace(., is.na(.), 0)) %>% 
  group_by(condition) %>% 
  mutate(TOTAL = sum(DOWN + NEUTRAL + UP)) %>% 
  select(condition, DOWN:TOTAL) %>% 
  distinct() %>% 
  ungroup() %>% 
  mutate(condition = fct_relevel(condition, ann_vaccines_covid_other_conditions_ordered$condition %>% levels())) %>% 
  arrange(condition)

all_degs_total
```


Vertical, facetted
```{r fig.height=10, fig.width=10}

ggplot_degs_bars_directions = all_degs %>%
  mutate(direction = case_when(direction == "UP" ~ "Up-regulated",
                               direction == "DOWN" ~ "Down-regulated",
                               TRUE ~ "Neutral")) %>% 
  filter(type != "IN/SU") %>% 
 ggplot() +
  aes(y = fct_rev(condition), fill = type, x = n_degs) +
  geom_col(position = "dodge",
           color = "black",
           size = 0.2) +
  scale_fill_manual(values = ann_vaxsig_colors$type)+
  # scale_fill_manual(
  #   values = c("Down-regulated" = "#4DBBD5FF",
  #   "Neutral" = "gray80",
  #   "Up-regulated" = "#d90429")) +
  theme_minimal() +
  # geom_text(aes(label = n_degs, x = n_degs), 
  #           position = position_dodge(width = 0.9), 
  #           angle = 0,
  #           vjust = 0.5, 
  #           hjust = -0.2, 
  #           size = 3) +
  scale_x_continuous(expand = c(0,0)
                     # limits = c(0, 8000)
                     )+
  theme(legend.position = "right",
        strip.text.x = element_text(color = "black", size = 12), 
        axis.text.x = element_text(color = "black", size = 10),
        axis.text.y = element_text(vjust = 0.5, size = 10, color = "black"),
        axis.line = element_line(size = 1, colour = "black", linetype=1),
        axis.ticks.y = element_line(size = 0.5, color = "black"),
        axis.ticks.x = element_line(size = 0, color = "black"),
        panel.grid.major.y = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.major.x = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.spacing.x = unit(2, "lines")) +
  labs(x = "Total",
       y = "",
       fill = "Type",
       title = "Total DEGs, by direction. Threshold = |l2fc| > 1") +
  facet_wrap(~direction, scales = "free_x")

ggplot_degs_bars_directions %>%
  ggsave(file = here("Figures", "OtherVaccines_DEGs_L2FC_directions_vertical.png"), width = 10, height = 10)

ggplot_degs_bars_directions
```



Unite with SARS

```{r fig.height=5, fig.width=30}
all_degs = degs_covid_vax %>%
  mutate(direction = case_when(log2fold_change >= 1 ~ "UP",
                               log2fold_change <= -1 ~ "DOWN",
                               TRUE ~ "NEUTRAL")) %>% 
  group_by(condition, direction) %>% 
  summarise(n_degs = n()) %>% 
  inner_join(ann_vaccines_covid_other_conditions , by="condition") %>% 

  ungroup() %>% 
  mutate(condition = fct_relevel(condition, ann_vaccines_covid_other_conditions_ordered$condition %>% levels())) %>% 
  arrange(condition)

all_degs_total = all_degs %>% 
  pivot_wider(names_from = "direction", values_from = "n_degs") %>% 
  mutate_all(~replace(., is.na(.), 0)) %>% 
  group_by(condition) %>% 
  mutate(TOTAL = sum(DOWN + NEUTRAL + UP)) %>% 
  select(condition, DOWN:TOTAL) %>% 
  distinct() %>% 
  ungroup() %>% 
  mutate(condition = fct_relevel(condition, ann_vaccines_covid_other_conditions_ordered$condition %>% levels())) %>% 
  arrange(condition)


all_degs = all_degs %>% 
  inner_join(all_degs_total %>% 
               select(condition, total = TOTAL),
             by = "condition")
all_degs

degs_tbl = all_degs %>% 
  filter(!condition %in% c("ChAd-BNT (V2, D0)",
                           "BNT-I-BNT (D51, hospital)")) %>% 
  select(condition, type, dose, day, pathogen, everything())  

degs_BNT_MO_before <- degs_tbl %>%
  filter(condition %in% c("BNT (V2, D1)", "BNT (V1, D6)")) %>% 
  mutate(condition = if_else(condition == "BNT (V1, D6)", "BNT-MO (V1, D6)", "BNT-MO (V2, D1)"),
         vaccine = "MO")

degs_tbl_I <- degs_tbl %>%
  filter(disease_vac == "I") %>% 
  select(condition, pathogen, everything())

degs_tbl = degs_tbl %>% 
  filter(!condition %in% c(degs_tbl_I$condition %>% unique()))

degs_tbl_2 = bind_rows(degs_tbl,
            degs_BNT_MO_before,
            degs_tbl_I) %>% 
    mutate(day = as.character(day),
           day = as.numeric(day),
         pathogen = if_else(is.na(pathogen) & disease_vac == "V", 
                            paste0("SARS-CoV-2/", vaccine), 
                            pathogen),
         pathogen = str_to_upper(pathogen),
         pathogen = case_when(str_detect(condition, "BNT-I.*hospital") ~ "SARS-COV-2/BNT-I Hospital",
                              str_detect(condition, "I.*hospital") ~ "SARS-COV-2/I Hospital", 
                              condition == "BNT-I (D0-5)" ~ "SARS-COV-2/BNT-I", 
                              condition == "I (D0-4)" ~ "SARS-COV-2/I", 
                              condition == "I-BNT (D5)" ~ "SARS-COV-2/I-BNT", 
                              condition == "I-I (D0-5)" ~ "SARS-COV-2/I-I", 
                              .default = pathogen)) %>% 
    arrange(day) %>% 
    mutate(type_grouped = case_when(type == "CONJ" ~ "Conjugated/PS",
                                 type == "PS" ~ "Conjugated/PS",
                                 type == "VV" ~ "Genetic",
                                 type == "RNA" ~ "Genetic",
                                 type == "LA" ~ "Live attenuated",
                                 type == "IN" ~ "Inactivated",
                                 type == "SU" ~ "Subunit",
                                 disease_vac == "I" ~ "Infection",
                                 TRUE ~ type),
         type = case_when(disease_vac == "I" ~ "I",
                                 TRUE ~ type),
         type_grouped = as.factor(type_grouped),
         type_grouped = fct_relevel(type_grouped, 
                            "Infection",
                            "Live Attenuated",
                            "Inactivated",
                            "Conjugated/PS",
                            "Subunit",
                            "Genetic")) %>% 
  select(condition, type, type_grouped, dose, day, pathogen, everything()) 

degs_tbl_day0 <- degs_tbl_2 %>%
  select(condition, type, type_grouped, dose, pathogen, day, total) %>% 
  mutate(day = 0, 
         total = 0, 
         dose = "0",
         condition = paste0(pathogen, " D(0)")) %>% 
  distinct()

degs_tbl_2 = degs_tbl_2 %>% 
  bind_rows(degs_tbl_day0) %>% 
  arrange(condition)

degs_tbl_2 %>% 
  select(condition, type, type_grouped, dose, pathogen, day, total) %>% 
  distinct() %>% 
  arrange(pathogen, day)


day_filter = 30
time_bytype_DEGs_covid = degs_tbl_2 %>%
  filter(day <= day_filter, 
         type != "IN/SU") %>% 
  mutate(dose = as.factor(dose),
         dose = fct_relevel(dose, "0", "1", "2", "3")) %>%
  mutate(type_grouped = str_to_upper(type_grouped),
         type_grouped = as.factor(type_grouped),
         type_grouped = fct_relevel(type_grouped, 
                            "INFECTION",
                            "LIVE ATTENUATED",
                            "INACTIVATED",
                            "CONJUGATED/PS",
                            "SUBUNIT",
                            "GENETIC"),
         title = "Total DEGs") %>% 

  arrange(condition) %>% 
  distinct() %>% 
  ggplot() +
  aes(x = day, 
      y = total, 
    group = pathogen
      ) +
  geom_line(
  aes(color = pathogen,
      group = pathogen
    ),
  size = 1) +
  geom_point(aes(shape = dose,
                 color = pathogen),
             size = 3.5,
             color = "black",
             alpha = 1) +
  geom_point(aes(shape = dose,
                 color = pathogen),             
             size = 3,
             alpha = 0.7) +
  # geom_hline(yintercept = 0, linetype = 1, alpha = 1) +
  scale_color_manual(values = c(ann_vaxsig_colors$type, 
                                ann_vaxsig_colors$target_pathogen_disease,
                                ann_vaxsig_colors$dose,
                                "SARS-COV-2/BBIBP" = "#00A087FF",
                                "SARS-COV-2/ZF2001" = "#8491B4FF",
                                "SARS-COV-2/BNT" = "#4DBBD5FF",
                                "SARS-COV-2/MO" = "blue",
                                "SARS-COV-2/CHAD" = "#3C5488FF",
                                "SARS-COV-2/I" = "red",
                                "SARS-COV-2/I Hospital" = "red",
                                "SARS-COV-2/I-I" = "purple",
                                "SARS-COV-2/BNT-I" = "pink",
                                "SARS-COV-2/BNT-I Hospital" = "pink",
                                "SARS-COV-2/I-BNT" = "#Af101e"
                                
                                )) +
  scale_shape_manual(values = c("0" = 15, 
                                "1" = 16, 
                                "2" = 17, 
                                "3" = 18)) +
  scale_alpha_manual(values = c("COVID-19" = 1, 
                                "13Vax Atlas" = 1), guide = "none") +
  theme_minimal() +
  labs(x = "Day",
       y = "Total",
       color = "Pathogen",
       shape = "Dose",
       alpha = "COVID-19 or Other",
       # title = "Temporal dynamics of immunization", 
       # subtitle = "Immune System process",
       linetype = "Set") +
  facet_grid(~title~type_grouped, scales = "free_x") +
  theme(legend.position = "right",
        strip.text.x = element_text(color = "black", size = 12,
                                    face = "bold"), 
        strip.text.y = element_text(color = "black", 
                                    size = 12, 
                                    hjust = 0,
                                    angle = 360,
                                    face = "bold"), 
        axis.title.x = element_blank(),
        axis.text.x = element_blank(),
        axis.text.y = element_text(vjust = 0.5, size = 12, color = "black"),
        axis.line.y = element_line(size = 1, colour = "black", linetype=1),
        axis.line.x = element_blank(),
        axis.ticks.y = element_line(size = 0.5, color = "black"),
        axis.ticks.x = element_blank(),
        panel.grid.major.y = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.minor.y = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.major.x = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.minor.x = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75")) +
    scale_x_continuous(breaks = c(0, 3, 7, 14, 21, 28, 30))


time_bytype_DEGs_covid 

# time_bytype_DEGs_covid %>%
#   ggsave(filename = here("Figures", "OtherVaccines_COVID_DEGs_Linechart_FacetByType.png"),
#          width = 30,
#          height = 5)

```





### By biotype

```{r fig.height=10, fig.width=10}
# library("biomaRt")
# mart <- useMart("ensembl")
# mart <- useDataset("hsapiens_gene_ensembl", mart)
# gene_info <- getBM(attributes = c("entrezgene_id", "hgnc_symbol", "ensembl_gene_id", "gene_biotype", "start_position", "end_position"),
#                    mart = mart)
# 
# gene_info %>% 
#   saveRDS(file = here("DGE analysis", "Gene_info_biomart.rds"))

gene_info = read_rds(file = here("DGE analysis", "Gene_info_biomart.rds"))

all_degs_totals = degs_covid_vax %>%
  group_by(condition) %>% 
  summarise(n_total = n())
  
all_degs_biotypes = degs_covid_vax %>% 
  inner_join(all_degs_totals, by = "condition") %>% 
  inner_join(gene_info %>% select(genes = hgnc_symbol, gene_biotype), by = "genes") %>% 
  mutate(gene_biotype_grouped = case_when(str_detect(gene_biotype, "IG_") ~ "Protein coding",
                                          str_detect(gene_biotype, "TR_") ~ "Protein coding",
                                          str_detect(gene_biotype, "pseudogene") ~ "Pseudogene",
                                          str_detect(gene_biotype, "protein_coding") ~ "Protein coding",
                                          str_detect(gene_biotype, "Mt") ~ "Mithochondrial RNA",
                                          str_detect(gene_biotype, "ribozyme") ~ "Ribozyme",
                                          str_detect(gene_biotype, "RNA") ~ "Non-coding RNAs",
                                          str_detect(gene_biotype, "TEC") ~ "To be Experimentally Confirmed",
                                          TRUE ~ gene_biotype)) %>% 
  group_by(condition, gene_biotype_grouped) %>% 
  distinct() %>% 
  summarise(n_biotypes = n(),
            n_total = n_total) %>%
  distinct() %>% 
  mutate(n_perc = 100*(n_biotypes/n_total)) %>% 
  inner_join(ann_vaccines_covid_other_conditions %>% 
               filter(set == "13Vax Atlas"), by="condition") %>% 
  ungroup()

all_degs_biotypes %>% 
  filter(gene_biotype_grouped != "Protein coding") %>% 
  summarize(mean = median(n_perc))

all_degs_biotypes_ordered = all_degs_biotypes %>% 
  mutate(day = as.character(day) %>% 
           as.integer()) %>% 
  arrange(vaccine, day)


all_degs_biotypes_plot_totals = all_degs_biotypes %>% 
    filter(type != "IN/SU") %>% 
  mutate(condition = as.factor(condition),
         pathogen = str_to_upper(pathogen),
         condition = fct_relevel(condition, 
                                 unique(all_degs_biotypes_ordered$condition))) %>% 
  mutate(gene_biotype_grouped = fct_relevel(gene_biotype_grouped, 
                                            "Protein coding", 
                                            "Non-coding RNAs",
                                            "Ribozyme",
                                            "Mithochondrial RNA",
                                            "Pseudogene",
                                            "To be Experimentally Confirmed"
                                            ),
         title = "Total DEGs by biotype") %>% 
  ggplot(.) +
  aes(x = n_biotypes, y = fct_rev(condition), fill = type) +
  geom_bar(stat = "summary", fun = "sum",
           color ="black",
           size = 0.5,
           width = 0.6) +
  theme_minimal() +
  scale_fill_manual(values = ann_vaxsig_colors$type) +
  theme(legend.position = "right",
        strip.text.x = element_text(color = "black", size = 12), 
        axis.text.x = element_text(color = "black", size = 10),
        axis.text.y = element_text(vjust = 0.5, size = 10, color = "black"),
        axis.line = element_line(size = 1, colour = "black", linetype=1),
        axis.ticks = element_line(size = 0.5, color = "black"),
        panel.grid.major.y = element_line(size = 0, 
                                    linetype = 2,
                                    color = "white"),
        panel.grid.major.x = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75")) +
  labs(x = "Total",
       y = "",
       fill = "Type",
       title = "Total DEGs") +
  scale_x_continuous(expand = c(0,0)) +
    facet_wrap(~title, scales= "free", ncol = 1)

all_degs_biotypes_plot_totals %>%
  ggsave(file = here("Figures", "DEGs_by_biotype_OtherVaccines.png"),
         width = 10,
         height = 10)

all_degs_biotypes_plot_totals

```
### By immune genes

```{r}
ImmuneGO_BTM_grouped_genes <- read_csv("VaxGO gene sets/ImmuneGO_BTM_grouped_genes.csv")

all_degs_immuneGO_BTM_totals_covid = degs_covid_vax %>% 
  select(genes, condition) %>% 
  left_join(ImmuneGO_BTM_grouped_genes %>% 
               select(genes) %>% 
               distinct() %>% 
              mutate(set = "Immune"),
               by = "genes") %>% 
  mutate(set = if_else(is.na(set), "Other", set)) %>% 
  group_by(condition,set) %>% 
  summarise(n = n()) %>%
  ungroup() %>% 
  group_by(condition) %>% 
  summarize(ntotal = sum(n),
            n = n,
            set = set) %>% 
  distinct() %>% 
    ungroup() %>%  
  mutate(n_perc = round(100*(n/ntotal), 1)) %>% 
  inner_join(ann_vaccines_covid_other_conditions_ordered %>% 
               select(-set),
             by="condition")

all_degs_immuneGO_BTM_reg_covid = degs_covid_vax %>% 
  select(genes, direction, condition) %>% 
  inner_join(ImmuneGO_BTM_grouped_genes %>% 
               select(genes) %>% 
               distinct() %>% 
               mutate(set = "Immune"),
               by = "genes") %>% 
  mutate(set = if_else(is.na(set), "Other", set)) %>% 
  group_by(condition, direction, set) %>% 
  summarise(n = n()) %>%
  ungroup() %>% 
  group_by(condition) %>% 
  summarize(ntotal = sum(n),
            direction = direction,
            n = n,
            set = set) %>% 
  distinct() %>% 
    ungroup() %>% 
  inner_join(ann_vaccines_covid_other_conditions_ordered %>% 
               select(-set), by="condition")

all_degs_immuneGO_BTM_totals_covid

immune_others_covid = all_degs_immuneGO_BTM_totals_covid %>%
  filter(set == "Immune") %>% 
  rename(immune = set) %>% 
  ungroup() %>% 
  arrange(condition) %>% 
  mutate(condition = fct_relevel(condition,
                                 ann_vaccines_covid_other_conditions_ordered$condition %>% 
                                   levels()),
         day = as.character(day) %>% 
           as.integer(),
         pathogen = str_to_upper(pathogen))

immune_others_covid_tbl = immune_others_covid %>% 
  filter(!condition %in% c("ChAd-BNT (V2, D0)",
                           "BNT-I-BNT (D51, hospital)")) %>% 
  select(condition, type, dose, day, pathogen, everything())  

immune_others_covid_tbl_BNT_MO_before <- immune_others_covid_tbl %>%
  filter(condition %in% c("BNT (V2, D1)", "BNT (V1, D6)")) %>% 
  mutate(condition = if_else(condition == "BNT (V1, D6)", "BNT-MO (V1, D6)", "BNT-MO (V2, D1)"),
         vaccine = "MO")

immune_others_covid_tbl_I <- immune_others_covid_tbl %>%
  filter(disease_vac == "I") %>% 
  select(condition, pathogen, everything())

immune_others_covid_tbl = immune_others_covid_tbl %>% 
  filter(!condition %in% c(immune_others_covid_tbl_I$condition %>% unique()))

immune_others_covid_tbl_2 = bind_rows(immune_others_covid_tbl,
            immune_others_covid_tbl_BNT_MO_before,
            immune_others_covid_tbl_I) %>% 
    mutate(day = as.character(day),
           day = as.numeric(day),
         pathogen = if_else(is.na(pathogen) & disease_vac == "V", 
                            paste0("SARS-CoV-2/", vaccine), 
                            pathogen),
         pathogen = str_to_upper(pathogen),
         pathogen = case_when(str_detect(condition, "BNT-I.*hospital") ~ "SARS-COV-2/BNT-I Hospital",
                              str_detect(condition, "I.*hospital") ~ "SARS-COV-2/I Hospital", 
                              condition == "BNT-I (D0-5)" ~ "SARS-COV-2/BNT-I", 
                              condition == "I (D0-4)" ~ "SARS-COV-2/I", 
                              condition == "I-BNT (D5)" ~ "SARS-COV-2/I-BNT", 
                              condition == "I-I (D0-5)" ~ "SARS-COV-2/I-I", 
                              .default = pathogen)) %>% 
    arrange(day) %>% 
    mutate(type_grouped = case_when(type == "CONJ" ~ "Conjugated/PS",
                                 type == "PS" ~ "Conjugated/PS",
                                 type == "VV" ~ "Genetic",
                                 type == "RNA" ~ "Genetic",
                                 type == "LA" ~ "Live attenuated",
                                 type == "IN" ~ "Inactivated",
                                 type == "SU" ~ "Subunit",
                                 disease_vac == "I" ~ "Infection",
                                 TRUE ~ type),
         type = case_when(disease_vac == "I" ~ "I",
                                 TRUE ~ type),
         type_grouped = as.factor(type_grouped),
         type_grouped = fct_relevel(type_grouped, 
                            "Infection",
                            "Live Attenuated",
                            "Inactivated",
                            "Conjugated/PS",
                            "Subunit",
                            "Genetic")) %>% 
  select(condition, type, type_grouped, dose, day, pathogen, everything()) 

immune_others_covid_day0 <- immune_others_covid_tbl_2 %>%
  select(condition, type, type_grouped, dose, pathogen, day,immune, n_perc) %>% 
  mutate(day = 0, 
         n_perc = 0, 
         dose = "0",
         condition = paste0(pathogen, " D(0)")) %>% 
  distinct()

immune_others_covid_tbl_2 = immune_others_covid_tbl_2 %>% 
  bind_rows(immune_others_covid_day0) %>% 
  arrange(condition)

immune_others_covid_tbl_2 %>% 
  select(condition, type, type_grouped, dose, pathogen, day, immune, n_perc) %>% 
  distinct() %>% 
  arrange(pathogen, day)

```

Separated
```{r fig.height=5, fig.width=30}

day_filter = 30
time_bytype_Immune_COVID = immune_others_covid_tbl_2 %>%
  filter(day <= day_filter, 
         type != "IN/SU") %>% 
  mutate(dose = as.factor(dose),
         dose = fct_relevel(dose, "0", "1", "2", "3")) %>%
  mutate(type_grouped = str_to_upper(type_grouped),
         type_grouped = as.factor(type_grouped),
         type_grouped = fct_relevel(type_grouped, 
                            "INFECTION",
                            "LIVE ATTENUATED",
                            "INACTIVATED",
                            "CONJUGATED/PS",
                            "SUBUNIT",
                            "GENETIC"),
         title = immune) %>% 

  arrange(condition) %>% 
  distinct() %>% 
  ggplot() +
  aes(x = day, 
      y = n_perc, 
     # label = condition,
    group = pathogen
      ) +
  geom_line(
  aes(color = pathogen,
      group = pathogen
    ),
  size = 1) +
  geom_point(aes(shape = dose,
                 color = pathogen),
             size = 3.5,
             color = "black",
             alpha = 1) +
  geom_point(aes(shape = dose,
                 color = pathogen),             
             size = 3,
             alpha = 0.7) +
  # geom_hline(yintercept = 0, linetype = 1, alpha = 1) +
  scale_color_manual(values = c(ann_vaxsig_colors$type, 
                                ann_vaxsig_colors$target_pathogen_disease,
                                ann_vaxsig_colors$dose,
                                "SARS-COV-2/BBIBP" = "#00A087FF",
                                "SARS-COV-2/ZF2001" = "#8491B4FF",
                                "SARS-COV-2/BNT" = "#4DBBD5FF",
                                "SARS-COV-2/MO" = "blue",
                                "SARS-COV-2/CHAD" = "#3C5488FF",
                                "SARS-COV-2/I" = "red",
                                "SARS-COV-2/I Hospital" = "red",
                                "SARS-COV-2/I-I" = "purple",
                                "SARS-COV-2/BNT-I" = "pink",
                                "SARS-COV-2/BNT-I Hospital" = "pink",
                                "SARS-COV-2/I-BNT" = "#Af101e"
                                
                                )) +
  scale_shape_manual(values = c("0" = 15, 
                                "1" = 16, 
                                "2" = 17, 
                                "3" = 18)) +
  scale_alpha_manual(values = c("COVID-19" = 1, "13Vax Atlas" = 1), guide = "none") +
  theme_minimal() +
  labs(x = "Day",
       y = "Total",
       color = "Pathogen",
       shape = "Dose",
       alpha = "COVID-19 or Other",
       # title = "Temporal dynamics of immunization", 
       # subtitle = "Immune System process",
       linetype = "Set") +
  facet_grid(~title~type_grouped, scales = "free_x") +
  theme(legend.position = "right",
        strip.text.x = element_blank(),
        strip.text.y = element_text(color = "black", 
                                    size = 12, 
                                    hjust = 0,
                                    angle = 360,
                                    face = "bold"), 
        axis.title.x = element_blank(),
        axis.text.x = element_blank(),
        axis.text.y = element_text(vjust = 0.5, size = 12, color = "black"),
        axis.line.y = element_line(size = 1, colour = "black", linetype=1),
        axis.line.x = element_blank(),
        axis.ticks.y = element_line(size = 0.5, color = "black"),
        axis.ticks.x = element_blank(),
        panel.grid.major.y = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.minor.y = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.major.x = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.minor.x = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75")) +
    scale_x_continuous(breaks = c(0, 3, 7, 14, 21, 28, 30)) 


```


### DEGs by Main GO
```{r}
OtherGOs_Annotated_Genes = read_csv(file = here("VaxGO gene sets", "OtherGOs_Annotated_Genes.csv"))
go_bp_all = OtherGOs_Annotated_Genes


# DEGs to Entrez and GO ids --------
degs_GO = degs_covid_vax %>% 
  select(condition, genes, direction) %>% 
  left_join(go_bp_all %>% rename(genes = symbol), by = "genes") %>% 
  mutate(go_ancestor = if_else(is.na(go_ancestor), "Not mapped", go_ancestor),
         term_ancestor = if_else(is.na(term_ancestor), "Not mapped", term_ancestor),
         mapped = if_else(term_ancestor == "Not mapped", "Not mapped", "Mapped"))

# Collapse gene column --------
degs_GO_comma = degs_GO %>% 
  distinct() %>% 
  group_by(condition, term_ancestor) %>% 
  summarize(genes = paste0(genes, collapse = ",")) %>% 
  inner_join(ann_vaccines, by = "condition")

write_csv(degs_GO_comma, file = here("DGE analysis", "OtherVaccines_degs_annotated_allGOs.csv"))


#DEGs annotated by Main GO BPs ------

main_go = degs_GO %>%
  filter(go_ancestor %in% c(
      "Not mapped",
      "GO:0040011",
      "GO:0043473",
      "GO:0044419",
      "GO:0009987",
      "GO:0048518",
      "GO:0050789",
      "GO:0065007",
      "GO:0048511",
      "GO:0022414",
      "GO:0002376",
      "GO:0008152")) %>% 
  select(condition, genes, direction, term_ancestor) %>% 
  distinct() %>% 
  group_by(condition, direction, term_ancestor) %>% 
  summarise(n = n()) %>% 
  mutate(term_ancestor = if_else(is.na(term_ancestor), "No annotation", term_ancestor))

main_go_2 = degs_GO %>% 
  select(condition, genes, mapped, direction) %>% 
  distinct() %>% 
  group_by(condition, direction) %>% 
  summarise(n_genes_total = n()) %>% 
  inner_join(main_go, by = join_by("condition", direction)) %>% 
  ungroup() %>% 
  select(condition, direction, term_ancestor, n_genes = n, n_genes_total) %>% 
  mutate(n_genes_perc = round(n_genes/n_genes_total * 100, 1),
         term_ancestor = case_when(
           term_ancestor == "biological process involved in interspecies interaction between organisms" ~ "Interspecies interaction",
           term_ancestor == "biological regulation" ~ "Biological reg.",
           term_ancestor == "cellular process" ~ "Cellular",
           term_ancestor == "immune system process" ~ "Immune system",
           term_ancestor == "regulation of biological process" ~ "Reg. of BP",
           term_ancestor == "locomotion" ~ "Locomotion",
           term_ancestor == "metabolic process" ~ "Metabolism",
           TRUE ~ term_ancestor),
         term_ancestor = fct_relevel(term_ancestor, "Not mapped", after = Inf),
         direction = fct_relevel(direction, "UP", "DOWN"))
  
write_csv(main_go_2, file = here("DGE analysis", "OtherVaccines_degs_annotated_mainGOs_stats.csv"))
```



### By Metabolism and Cellular processes

```{r fig.height=10, fig.width=10, message=FALSE, warning=FALSE}
main_go_2 = read_csv(file = here("DGE analysis", "OtherVaccines_degs_annotated_mainGOs_stats.csv")) %>% 
  filter(condition %in% c(ann_vaccines_covid_other_conditions_ordered %>% 
           pull(condition)))

main_go_2covid = main_go_2 %>%
  group_by(condition, term_ancestor) %>%
  summarize(n_genes = sum(n_genes),
            n_genes_total = sum(n_genes_total),
            n_genes_perc = round((n_genes/n_genes_total)*100, 1)) %>% 
  ungroup() %>% 
  inner_join(ann_vaccines_covid_other_conditions_ordered, by = "condition") %>% 
  mutate(term_ancestor = fct_relevel(term_ancestor,
                                     "Immune system",
                                     "Cellular",
                                     "Metabolism",
                                     "Reg. of BP")) %>%
  filter(term_ancestor %in% c("Cellular", "Metabolism")) %>% 
  arrange(condition) %>% 
  filter(type != "IN/SU") %>% 
  mutate(condition = fct_relevel(condition,
                                 ann_vaccines_covid_other_conditions_ordered$condition %>% 
                                   levels()),
         day = as.character(day) %>% 
           as.integer(),
         pathogen = str_to_upper(pathogen))



degs_notimmune_tbl = main_go_2covid %>% 
  filter(!condition %in% c("ChAd-BNT (V2, D0)",
                           "BNT-I-BNT (D51, hospital)")) %>% 
  select(condition, type, dose, day, pathogen, everything())  

degs_notimmune_tbl_BNT_MO_before <- degs_notimmune_tbl %>%
  filter(condition %in% c("BNT (V2, D1)", "BNT (V1, D6)")) %>% 
  mutate(condition = if_else(condition == "BNT (V1, D6)", "BNT-MO (V1, D6)", "BNT-MO (V2, D1)"),
         vaccine = "MO")

degs_notimmune_tbl_I <- degs_notimmune_tbl %>%
  filter(disease_vac == "I") %>% 
  select(condition, pathogen, everything())

degs_notimmune_tbl = degs_notimmune_tbl %>% 
  filter(!condition %in% c(degs_notimmune_tbl_I$condition %>% unique()))

degs_notimmune_tbl_2 = bind_rows(degs_notimmune_tbl,
            degs_notimmune_tbl_BNT_MO_before,
            degs_notimmune_tbl_I) %>% 
    mutate(day = as.character(day),
           day = as.numeric(day),
         pathogen = if_else(is.na(pathogen) & disease_vac == "V", 
                            paste0("SARS-CoV-2/", vaccine), 
                            pathogen),
         pathogen = str_to_upper(pathogen),
         pathogen = case_when(str_detect(condition, "BNT-I.*hospital") ~ "SARS-COV-2/BNT-I Hospital",
                              str_detect(condition, "I.*hospital") ~ "SARS-COV-2/I Hospital", 
                              condition == "BNT-I (D0-5)" ~ "SARS-COV-2/BNT-I", 
                              condition == "I (D0-4)" ~ "SARS-COV-2/I", 
                              condition == "I-BNT (D5)" ~ "SARS-COV-2/I-BNT", 
                              condition == "I-I (D0-5)" ~ "SARS-COV-2/I-I", 
                              .default = pathogen)) %>% 
    arrange(day) %>% 
    mutate(type_grouped = case_when(type == "CONJ" ~ "Conjugated/PS",
                                 type == "PS" ~ "Conjugated/PS",
                                 type == "VV" ~ "Genetic",
                                 type == "RNA" ~ "Genetic",
                                 type == "LA" ~ "Live attenuated",
                                 type == "IN" ~ "Inactivated",
                                 type == "SU" ~ "Subunit",
                                 disease_vac == "I" ~ "Infection",
                                 TRUE ~ type),
         type = case_when(disease_vac == "I" ~ "I",
                                 TRUE ~ type),
         type_grouped = as.factor(type_grouped),
         type_grouped = fct_relevel(type_grouped, 
                            "Infection",
                            "Live Attenuated",
                            "Inactivated",
                            "Conjugated/PS",
                            "Subunit",
                            "Genetic")) %>% 
  select(condition, type, type_grouped, dose, day, pathogen, everything()) 

degs_notimmune_tbl_day0 <- degs_notimmune_tbl_2 %>%
  select(condition, type, type_grouped, dose, pathogen, day,term_ancestor, n_genes_perc) %>% 
  mutate(day = 0, 
         n_genes_perc = 0, 
         dose = "0",
         condition = paste0(pathogen, " D(0)")) %>% 
  distinct()

degs_notimmune_tbl_2 = degs_notimmune_tbl_2 %>% 
  bind_rows(degs_notimmune_tbl_day0) %>% 
  arrange(condition)

degs_notimmune_tbl_2 %>% 
  select(condition, type, type_grouped, dose, pathogen, day, term_ancestor, n_genes_perc) %>% 
  distinct() %>% 
  arrange(pathogen, day)
```


```{r fig.height=10, fig.width=30}
day_filter = 30
time_bytype_MainGOs_COVID = degs_notimmune_tbl_2 %>%
  filter(day <= day_filter, 
         type != "IN/SU") %>% 
  mutate(dose = as.factor(dose),
         dose = fct_relevel(dose, "0", "1", "2", "3")) %>%
  mutate(type_grouped = str_to_upper(type_grouped),
         type_grouped = as.factor(type_grouped),
         type_grouped = fct_relevel(type_grouped, 
                            "INFECTION",
                            "LIVE ATTENUATED",
                            "INACTIVATED",
                            "CONJUGATED/PS",
                            "SUBUNIT",
                            "GENETIC"),
         title = term_ancestor) %>% 

  arrange(condition) %>% 
  distinct() %>% 
  ggplot() +
  aes(x = day, 
      y = n_genes_perc, 
     # label = condition,
    group = pathogen
      ) +
  geom_line(
  aes(color = pathogen,
      group = pathogen
    ),
  size = 1) +
  geom_point(aes(shape = dose,
                 color = pathogen),
             size = 3.5,
             color = "black",
             alpha = 1) +
  geom_point(aes(shape = dose,
                 color = pathogen),             
             size = 3,
             alpha = 0.7) +
  # geom_hline(yintercept = 0, linetype = 1, alpha = 1) +
  scale_color_manual(values = c(ann_vaxsig_colors$type, 
                                ann_vaxsig_colors$target_pathogen_disease,
                                ann_vaxsig_colors$dose,
                                "SARS-COV-2/BBIBP" = "#00A087FF",
                                "SARS-COV-2/ZF2001" = "#8491B4FF",
                                "SARS-COV-2/BNT" = "#4DBBD5FF",
                                "SARS-COV-2/MO" = "blue",
                                "SARS-COV-2/CHAD" = "#3C5488FF",
                                "SARS-COV-2/I" = "red",
                                "SARS-COV-2/I Hospital" = "red",
                                "SARS-COV-2/I-I" = "purple",
                                "SARS-COV-2/BNT-I" = "pink",
                                "SARS-COV-2/BNT-I Hospital" = "pink",
                                "SARS-COV-2/I-BNT" = "#Af101e"
                                
                                )) +
  scale_shape_manual(values = c("0" = 15, 
                                "1" = 16, 
                                "2" = 17, 
                                "3" = 18)) +
  scale_alpha_manual(values = c("COVID-19" = 1, "13Vax Atlas" = 1), guide = "none") +
  theme_minimal() +
  labs(x = "Day",
       y = "Total",
       color = "Pathogen",
       shape = "Dose",
       alpha = "COVID-19 or Other",
       # title = "Temporal dynamics of immunization", 
       # subtitle = "Immune System process",
       linetype = "Set") +
  facet_grid(~title~type_grouped, scales = "free_x") +
  theme(legend.position = "right",
        strip.text.x = element_blank(), 
        strip.text.y = element_text(color = "black", 
                                    size = 12, 
                                    hjust = 0,
                                    angle = 360,
                                    face = "bold"), 
        axis.text.x = element_text(color = "black", size = 12),
        axis.text.y = element_text(vjust = 0.5, size = 12, color = "black"),
        axis.line = element_line(size = 1, colour = "black", linetype=1),
        axis.ticks = element_line(size = 0.5, color = "black"),
        panel.grid.major.y = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.minor.y = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.major.x = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.minor.x = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75")) +
    scale_x_continuous(breaks = c(0, 3, 7, 14, 21, 28, 30)) 
 

time_bytype_MainGOs_COVID 

time_bytype_MainGOs_COVID %>%
  ggsave(filename = here("Figures", "OtherVaccines_COVID_Metabolism_Cellular_Linechart_FacetByType.png"),
         width = 30,
         height = 7)

```





### Merge plots


```{r fig.height=7, fig.width=25}
othervaccines_degs_lineplots = (time_bytype_DEGs_covid / 
   time_bytype_Immune_COVID / 
   time_bytype_MainGOs_COVID)  +
  plot_layout(guides = "collect", 
               axes = "collect", 
               axis_titles = "collect", 
               heights = c(1, 1, 2))

othervaccines_degs_lineplots
  
othervaccines_degs_lineplots %>%
  ggsave(filename = here("Figures", "OtherVaccines_DEGs_lineplots.png"),
         height = 7,
         width = 30)
othervaccines_degs_lineplots
```








## Overlapping genes

```{r}

data = degs_covid_vax %>%
  filter(genes %in% immunego_btm_genes$genes) %>% 
  rename(process = condition) %>% 
  select(process, genes)
  
```

Function for overlapping
```{r}

# Function for overlapping -------
overlap_genes <- function(cond1, cond2, data) {
  genes_cond1 <- data$genes[data$process == cond1]
  genes_cond2 <- data$genes[data$process == cond2]
  
  genes_shared <- intersect(genes_cond1, genes_cond2)
  
  genes_notshared_cond1 <- setdiff(genes_cond1, genes_cond2)
  genes_notshared_cond2 <- setdiff(genes_cond2, genes_cond1)
  
  total_genes_cond1 <- length(genes_cond1)
  total_genes_cond2 <- length(genes_cond2)
  
  percentage_shared_cond1 <- length(genes_shared) / total_genes_cond1 * 100
  percentage_shared_cond2 <- length(genes_shared) / total_genes_cond2 * 100
  
  shared_genes <- data.frame(
    Cond1 = cond1,
    Cond2 = cond2,
    Shared = length(genes_shared),
    NotShared_cond1 = length(genes_notshared_cond1),
    NotShared_cond2 = length(genes_notshared_cond2),
    Total_Genes_Cond1 = total_genes_cond1,
    Total_Genes_Cond2 = total_genes_cond2,
    Genes_Names = paste(genes_shared, collapse = ", "),
    Percentage_Shared_Cond1 = percentage_shared_cond1,
    Percentage_Shared_Cond2 = percentage_shared_cond2
  )
  
  return(shared_genes)
}
conflicted::conflicts_prefer(base::intersect)
conflicted::conflicts_prefer(base::setdiff)
conflicted::conflicts_prefer(janitor::fisher.test)
```

Perform calculations
```{r}
# Perform calculations
unique_cond <- unique(data$process) # List sets
shared_genes_df <- data.frame() # Store data

for (i in 1:(length(unique_cond) - 1)) {
  for (j in (i + 1):length(unique_cond)) {
    resultado_temp <- overlap_genes(unique_cond[i], unique_cond[j], data)
    shared_genes_df <- bind_rows(shared_genes_df, resultado_temp)
  }
}

# Fisher's exact test
fisher_exact_test <- function(shared, notshared1, notshared2) {
  cont_table <- matrix(c(shared, notshared1, notshared2, 0), nrow = 2)
  results_fisher <- fisher.test(cont_table)
  return(results_fisher$p.value)
}
shared_genes_df = shared_genes_df %>%
  mutate(pvalue = pmap_dbl(list(Shared, NotShared_cond1, NotShared_cond2), fisher_exact_test)) %>% 
  mutate("-log(p)" = -log10(pvalue),
         padj = p.adjust(pvalue))


#Mirror table
shared_genes_df2 = shared_genes_df %>% 
  rename(Cond1_temp = Cond1, Cond2_temp = Cond2) %>%
  rename(Cond1 = Cond2_temp, Cond2 = Cond1_temp)

shared_genes_df_mirror = shared_genes_df %>% 
  bind_rows(shared_genes_df2) %>% 
  mutate(Percentage_Shared_Cond1 = round(Percentage_Shared_Cond1, 2),
         Percentage_Shared_Cond2 = round(Percentage_Shared_Cond2, 2))
  
#Jaccard distance
jaccard.matrix <- data %>%
  mutate(present = 1) %>% 
  distinct() %>% 
  pivot_wider(names_from = genes, values_from = present, values_fill = 0) %>%
  column_to_rownames(var = "process") %>%
  vegdist(binary = TRUE, method = "jaccard", diag = TRUE, upper = TRUE) %>%
  as.matrix() %>%
  {1 - .}

shared_genes_df_mirror = jaccard.matrix %>% 
  as.data.frame() %>% 
  rownames_to_column("Cond1") %>% 
  pivot_longer(cols = -"Cond1",
               names_to = "Cond2",
               values_to = "jaccard_distance") %>% 
  inner_join(shared_genes_df_mirror, by = join_by(Cond1, Cond2)) %>% 
  mutate(jaccard_distance = round(jaccard_distance, 3))

#Save
write_csv(shared_genes_df_mirror, 
          file = here("VaxGO gene sets", paste0(filename, "_sharedgenes_Fisher_Jaccard.csv")))
```


### Tree
```{r fig.height=20, fig.width=5}
library(ggdendro)
# Convert to "dist" class for clustering
# Make the matrix symmetric

jaccard.matrix <- as.dist(jaccard.matrix)

# Perform hierarchical clustering
hc <- hclust(jaccard_dist, method = "complete")

dendro_data <- as.dendrogram(hc) %>% 
  dendro_data()

ggplot(segment(dendro_data)) +
  geom_segment(aes(x = y, y = x, xend = yend, yend = xend)) + # Flip x and y
  scale_x_reverse() +                                         # Reverse x-axis for better appearance
  labs(title = "Hierarchical Clustering Dendrogram",
       x = "Jaccard Distance", y = "") +
  theme_minimal() +
  geom_text(data = label(dendro_data),                       # Add labels
            aes(x = y, y = x, label = label), 
            hjust = 0,  
            nudge_x = 0.005,                                  # Adjust position of labels
            size = 3) + 
  theme(axis.text.y = element_text(size = 10),
        axis.text.x = element_text(size = 10),
        plot.margin = margin(0, 5, 0, 0, unit = "cm")) +
  coord_cartesian(clip = "off")
```



### Heatmap

Clustered
```{r}
# INPUT ------
degs_all_updown_covid_other_immune_shared = read_csv(here("VaxGO gene sets", "ImmuneGO_Genes_JaccardDistance_Hagan_sharedgenes_Fisher_Jaccard.csv"))

table = degs_all_updown_covid_other_immune_shared
filename = "ImmuneGO_Genes_JaccardDistance_Hagan" 

data = table %>% 
  filter(pvalue <0.25) %>% 
  inner_join(hagan_vaxatlas_annotation %>% 
               select(Cond1 = condition, everything()), by = "Cond1") %>% 
  inner_join(ann_vaccines_conditions %>% 
               select(Cond2 = condition, everything()), by = "Cond2") %>% 
  clean_names() %>% 
  rename(vac_ref = cond1, vac_condition = cond2)

# Matrix ------
matrix = data %>%
  filter(pvalue <= 0.10) %>% 
  select(vac_condition, vac_ref, jaccard_distance) %>% 
  pivot_wider(names_from = vac_condition, values_from = jaccard_distance, values_fn = mean) %>% 
  replace(is.na(.), 0) %>%
  column_to_rownames(var="vac_ref") %>%
  as.matrix()

matrix = matrix*100

# Annotations -------
ann_vaccines_covid = ann_vaccines_conditions 
ann_vaccines_other = hagan_vaxatlas_annotation 

# Columns
ann_cols_heatmap = matrix %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "process") %>% 
  rename(vac_condition = '.') %>% 
  merge(ann_vaccines_covid, by.x = "vac_condition", by.y = "condition", all.y = F) %>% 
  select(vac_condition, type, week, day) %>% #ordenar colunas para anotação
  distinct() %>% 
  column_to_rownames(var = "vac_condition") %>% 
  mutate_all(as.factor)

# Rows
ann_rows_heatmap = matrix %>%  
  as.data.frame() %>% 
  rownames_to_column("condition") %>% 
  distinct() %>% 
  inner_join(ann_vaccines_other, by = "condition") %>% 
  mutate_all(as.factor) %>%
  arrange(condition) %>%
  distinct() %>% 
  column_to_rownames(var= "condition") %>% 
  select(type, week, day)

# Arrange cols and rows in matrix
matrix_data = matrix %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>%
  rownames_to_column("condition") %>% 
  inner_join(ann_rows_heatmap %>% 
               rownames_to_column("condition") %>% 
               select(condition),
             by = "condition") %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  mutate_if(is.character, as.numeric) %>%
  replace(is.na(.), 0) %>% 
  as.matrix()

#Check dimensions
dim(matrix_data)
dim(ann_rows_heatmap)
dim(ann_cols_heatmap)

#Heatmap annotation
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "left")

row_ha = HeatmapAnnotation(df = ann_rows_heatmap,
                       col = ann_vaxsig_colors,
                       which= "row",
                       show_annotation_name = F,
                       show_legend = T)

# Quantile interpolation
quartiles <- quantile(matrix_data, probs = c(0, 0.25, 0.5, 0.75, 1), na.rm = TRUE)
# col_fun <- colorRamp2(quartiles, c("white", "#a8dadc", "#457b9d", "#d90429", "#1d3557"))
col_fun <- colorRamp2(quartiles, c("white", "#a8dadc", "#457b9d", "#EFC000FF", "#d90429"))
# col_fun <- colorRamp2(quartiles, c("white", "#E8E79AFF", "#8CBF9AFF", "#477B95FF", "#011C40FF"))


mat = matrix_data
width_image = 30
height_image = 40
width = unit(20, "cm")
height = unit(25, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)
file = paste0("Hagan_", filename)

fileimageheatmap_grouped = paste0(filename, sep ="_", "Complexheatmap_ALL_CLUSTERED.png")

png(file = here("Figures", fileimageheatmap_grouped), 
    width = width_image, 
    height = height_image,
    units="cm", res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "left",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "Jaccard \ndistance", #Legend
                        col = col_fun, #color
                        cell_fun = function(j, i, x, y, width, height, fill) {
                                            if(mat[i, j] > 0)
                                            grid.text(sprintf("%.f", mat[i, j]), 
                                                      x, y, 
                                                      gp = gpar(fontsize = 8))},
                        row_title = "",
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = gpar(col = "white", lwd = 2),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp, #Grande = 6, Pequeno 15
                        row_dend_width = unit(5, "cm"), #Altura do dendrograma
                        column_split = NULL, #Split column clusters 
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, # 20 para pequeno e grande
                        height = height # 20 para pequeno, 60 para grande
)

draw(heatmap_plot,
     merge_legend = TRUE,
     annotation_legend_side = "right", 
     heatmap_legend_side = "right")
dev.off()

```

Ordered by day
```{r}
# INPUT ------
degs_all_updown_covid_other_immune_shared = read_csv(here("VaxGO gene sets", "ImmuneGO_Genes_JaccardDistance_Hagan_sharedgenes_Fisher_Jaccard.csv"))

table = degs_all_updown_covid_other_immune_shared
filename = "ImmuneGO_Genes_JaccardDistance_Hagan" 

data = table %>% 
  filter(pvalue <0.25) %>% 
  inner_join(hagan_vaxatlas_annotation %>% 
               select(Cond1 = condition, everything()), by = "Cond1") %>% 
  inner_join(ann_vaccines_conditions %>% 
               select(Cond2 = condition, everything()), by = "Cond2") %>% 
  clean_names() %>% 
  rename(vac_ref = cond1, vac_condition = cond2)

# Matrix ------
matrix = data %>%
  filter(pvalue <= 0.10) %>% 
  select(vac_condition, vac_ref, jaccard_distance) %>% 
  pivot_wider(names_from = vac_condition, values_from = jaccard_distance, values_fn = mean) %>% 
  replace(is.na(.), 0) %>%
  column_to_rownames(var="vac_ref") %>%
  as.matrix()

matrix = matrix*100

# Annotations -------
ann_vaccines_covid = ann_vaccines_conditions 
ann_vaccines_other = hagan_vaxatlas_annotation 

# Columns
ann_cols_heatmap = matrix %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "process") %>% 
  rename(vac_condition = '.') %>% 
  merge(ann_vaccines_covid, by.x = "vac_condition", by.y = "condition", all.y = F) %>% 
  select(vac_condition, type, week, day) %>% #ordenar colunas para anotação
  distinct() %>% 
  column_to_rownames(var = "vac_condition") %>% 
  mutate_all(as.factor) %>% 
  arrange(day)

# Rows
ann_rows_heatmap = matrix %>%  
  as.data.frame() %>% 
  rownames_to_column("condition") %>% 
  distinct() %>% 
  inner_join(ann_vaccines_other, by = "condition") %>% 
  mutate_all(as.factor) %>%
  arrange(condition) %>%
  distinct() %>% 
  column_to_rownames(var= "condition") %>% 
  select(type, week, day) %>% 
  arrange(day)

# Arrange cols and rows in matrix
matrix_data = matrix %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  inner_join(ann_cols_heatmap %>% rownames_to_column("condition") %>% select(condition, day), by = "condition") %>% 
  arrange(day) %>%  
  select(-day) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>%
  rownames_to_column("condition") %>% 
  inner_join(ann_rows_heatmap %>% 
               rownames_to_column("condition") %>% 
               select(condition, day),
             by = "condition") %>% 
  arrange(day) %>% 
    select(-day) %>% 
  column_to_rownames("condition") %>% 
  mutate_if(is.character, as.numeric) %>%
  replace(is.na(.), 0) %>% 
  as.matrix()


ann_rows_heatmap = ann_rows_heatmap %>%
  rownames_to_column("condition") %>%
  mutate(condition = factor(condition, levels = unique(matrix_data %>% rownames()))) %>%
  arrange(condition) %>% 
  column_to_rownames("condition")
  

#Check dimensions
dim(matrix_data)
dim(ann_rows_heatmap)
dim(ann_cols_heatmap)

#Heatmap annotation
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "left")

row_ha = HeatmapAnnotation(df = ann_rows_heatmap,
                       col = ann_vaxsig_colors,
                       which= "row",
                       show_annotation_name = F,
                       show_legend = T)

# Quantile interpolation
quartiles <- quantile(matrix_data, probs = c(0, 0.25, 0.5, 0.75, 1), na.rm = TRUE)
# col_fun <- colorRamp2(quartiles, c("white", "#a8dadc", "#457b9d", "#d90429", "#1d3557"))
col_fun <- colorRamp2(quartiles, c("white", "#a8dadc", "#457b9d", "#EFC000FF", "#d90429"))
# col_fun <- colorRamp2(quartiles, c("white", "#E8E79AFF", "#8CBF9AFF", "#477B95FF", "#011C40FF"))


mat = matrix_data
width_image = 30
height_image = 40
width = unit(20, "cm")
height = unit(25, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)
file = paste0("Hagan_", filename)

fileimageheatmap_grouped = paste0(filename, sep ="_", "Complexheatmap_ByDay.png")

png(file = here("Figures", fileimageheatmap_grouped), 
    width = width_image, 
    height = height_image,
    units="cm", res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "left",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(title = "Jaccard \ndistance", direction = "vertical"),
                        col = col_fun, #color
                        cell_fun = function(j, i, x, y, width, height, fill) {
                                            if(mat[i, j] > 0)
                                            grid.text(sprintf("%.f", mat[i, j]), 
                                                      x, y, 
                                                      gp = gpar(fontsize = 8))},
                        row_title = "",
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = gpar(col = "white", lwd = 2),
                        cluster_rows = F,
                        row_names_gp = row_names_gp, #Grande = 6, Pequeno 15
                        row_dend_width = unit(5, "cm"), #Altura do dendrograma
                        column_split = NULL, #Split column clusters 
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = F,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, # 20 para pequeno e grande
                        height = height # 20 para pequeno, 60 para grande
)

draw(heatmap_plot,
     merge_legend = TRUE,
     annotation_legend_side = "right", 
     heatmap_legend_side = "right")
dev.off()

```
Ordered by condition
```{r}
# INPUT ------
degs_all_updown_covid_other_immune_shared = read_csv(here("VaxGO gene sets", "ImmuneGO_Genes_JaccardDistance_Hagan_sharedgenes_Fisher_Jaccard.csv"))

table = degs_all_updown_covid_other_immune_shared
filename = "ImmuneGO_Genes_JaccardDistance_Hagan" 

data = table %>% 
  filter(pvalue <0.25) %>% 
  inner_join(hagan_vaxatlas_annotation %>% 
               select(Cond1 = condition, everything()), by = "Cond1") %>% 
  inner_join(ann_vaccines_conditions %>% 
               select(Cond2 = condition, everything()), by = "Cond2") %>% 
  clean_names() %>% 
  rename(vac_ref = cond1, vac_condition = cond2)

# Matrix ------
matrix = data %>%
  filter(pvalue <= 0.10) %>% 
  select(vac_condition, vac_ref, jaccard_distance) %>% 
  pivot_wider(names_from = vac_condition, values_from = jaccard_distance, values_fn = mean) %>% 
  replace(is.na(.), 0) %>%
  column_to_rownames(var="vac_ref") %>%
  as.matrix()

matrix = matrix*100

# Annotations -------
ann_vaccines_covid = ann_vaccines_conditions 
ann_vaccines_other = hagan_vaxatlas_annotation 

# Columns
ann_cols_heatmap = matrix %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "process") %>% 
  rename(vac_condition = '.') %>% 
  merge(ann_vaccines_covid, by.x = "vac_condition", by.y = "condition", all.y = F) %>% 
 arrange(match(vac_condition, c(ann_vaccines_conditions_ordered$condition_grouped %>% as.character()))) %>%   
  distinct() %>% 
  column_to_rownames(var = "vac_condition") %>% 
  mutate_all(as.factor)  %>% 
  select(disease_vac, type, dose, week, day)

# Rows
ann_rows_heatmap = matrix %>%  
  as.data.frame() %>% 
  rownames_to_column("condition") %>% 
  distinct() %>% 
  inner_join(ann_vaccines_other, by = "condition") %>% 
  mutate_all(as.factor) %>%
  arrange(condition) %>%
  distinct() %>% 
  column_to_rownames(var= "condition") %>% 
  select(type, week, day) %>% 
  arrange(day)

# Arrange cols and rows in matrix
matrix_data = matrix %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  inner_join(ann_cols_heatmap %>% rownames_to_column("condition") %>% select(condition, day), by = "condition") %>% 
  arrange(match(condition, c(ann_vaccines_conditions_ordered$condition_grouped %>% as.character()))) %>% 
  select(-day) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>%
  rownames_to_column("condition") %>% 
  inner_join(ann_rows_heatmap %>% 
               rownames_to_column("condition") %>% 
               select(condition, day),
             by = "condition") %>% 
  arrange(day) %>% 
    select(-day) %>% 
  column_to_rownames("condition") %>% 
  mutate_if(is.character, as.numeric) %>%
  replace(is.na(.), 0) %>% 
  as.matrix()


ann_rows_heatmap = ann_rows_heatmap %>%
  rownames_to_column("condition") %>%
  mutate(condition = factor(condition, levels = unique(matrix_data %>% rownames()))) %>%
  arrange(condition) %>% 
  column_to_rownames("condition")

#Check dimensions
dim(matrix_data)
dim(ann_rows_heatmap)
dim(ann_cols_heatmap)

#Heatmap annotation
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "left")

row_ha = HeatmapAnnotation(df = ann_rows_heatmap,
                       col = ann_vaxsig_colors,
                       which= "row",
                       show_annotation_name = F,
                       show_legend = T)

# Quantile interpolation
quartiles <- quantile(matrix_data, probs = c(0, 0.25, 0.5, 0.75, 1), na.rm = TRUE)
# col_fun <- colorRamp2(quartiles, c("white", "#a8dadc", "#457b9d", "#d90429", "#1d3557"))
col_fun <- colorRamp2(quartiles, c("white", "#a8dadc", "#457b9d", "#EFC000FF", "#d90429"))
# col_fun <- colorRamp2(quartiles, c("white", "#E8E79AFF", "#8CBF9AFF", "#477B95FF", "#011C40FF"))


mat = matrix_data
width_image = 30
height_image = 40
width = unit(20, "cm")
height = unit(25, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)
file = paste0("Hagan_", filename)

fileimageheatmap_grouped = paste0(filename, sep ="_", "Complexheatmap_Condition.png")

png(file = here("Figures", fileimageheatmap_grouped), 
    width = width_image, 
    height = height_image,
    units="cm", res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "left",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(title = "Jaccard \ndistance", direction = "vertical"),
                        col = col_fun, #color
                        cell_fun = function(j, i, x, y, width, height, fill) {
                                            if(mat[i, j] > 0)
                                            grid.text(sprintf("%.f", mat[i, j]), 
                                                      x, y, 
                                                      gp = gpar(fontsize = 8))},
                        row_title = "",
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = gpar(col = "white", lwd = 2),
                        cluster_rows = F,
                        row_names_gp = row_names_gp, #Grande = 6, Pequeno 15
                        row_dend_width = unit(5, "cm"), #Altura do dendrograma
                        column_split = NULL, #Split column clusters 
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = F,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, # 20 para pequeno e grande
                        height = height # 20 para pequeno, 60 para grande
)

draw(heatmap_plot,
     merge_legend = TRUE,
     annotation_legend_side = "right", 
     heatmap_legend_side = "right")
dev.off()

```



## Gene set enrichment analysis

Input

```{r}
#Gene sets

ImmuneGO_BTM_grouped_genes <- read_csv("VaxGO gene sets/ImmuneGO_BTM_grouped_genes.csv") %>% 
  mutate(annotation = if_else(is.na(immune_system), "Manual", "Specific"))

ImmuneGO_BTM_general = ImmuneGO_BTM_grouped_genes %>% 
  filter(annotation == "Manual") %>% 
  select(process, genes)

ImmuneGO_BTM_specific = ImmuneGO_BTM_grouped_genes %>%
  filter(annotation == "Specific") %>% 
  select(process, genes)

filename = "All_covid_13Vax"
```


#### ImmuneGO-BTM

```{r}
autoGSEA <- function(df, TERM2GENE, geneset_name) {
  resultados <- list()

  condicoes <- df$condition %>% 
    unique() %>% 
    as.character()

  for (condicao in condicoes) {
    
    degs_condicao <- df %>% 
      filter(condition == condicao) %>%
      select(genes, log2fold_change) %>%
      distinct() %>%
      mutate(rank = rank(log2fold_change, ties.method = "random")) %>%
      arrange(desc(rank))

    gene_list_lf2c <- as.vector(degs_condicao$log2fold_change)
    names(gene_list_lf2c) <- degs_condicao$genes
    gene_list_lf2c <- na.omit(gene_list_lf2c)
    gene_list_lf2c <- sort(gene_list_lf2c, decreasing = TRUE)

    auto_gsea <- tryCatch({
      GSEA(geneList = gene_list_lf2c,
           TERM2GENE = TERM2GENE,
           minGSSize = 1,
           maxGSSize = 1000000,
           pvalueCutoff = 1,
           pAdjustMethod = "BH") %>%
        as.data.frame() %>%
        arrange(qvalue) %>%
        mutate(condition = condicao,
               gsea_enrichment = geneset_name)
    }, error = function(e) {
      return(NULL)
    })

    if (!is.null(auto_gsea)) {
      resultados[[paste(condicao, geneset_name, sep = "_")]] <- auto_gsea
    }
  }
  

  resultado_final <- bind_rows(resultados)

  return(resultado_final)
}
```

```{r}
# Perform GSEA

#General modules
geneset_name = "ImmuneGO_BTM_general"
filename = "All_covid_13Vax"

degs_covid_vax_GSEA_ALL <- degs_covid_vax %>%
  select(condition, genes, log2fold_change) %>% 
  autoGSEA(TERM2GENE = ImmuneGO_BTM_general, 
           geneset_name = "ImmuneGO_BTM_general") 

degs_covid_vax_GSEA = degs_covid_vax_GSEA_ALL %>% 
  rownames_to_column("delete") %>% 
  clean_names() %>% 
    select(condition, id, everything()) 

#Save
write_csv(degs_covid_vax_GSEA, 
          file = here("Gene set enrichment analysis", paste0(filename, geneset_name, "_GSEA_ALL", ".csv")))

# Specific modules

geneset_name = "ImmuneGO_BTM_specific"
filename = "All_covid_13Vax"

degs_covid_vax_GSEA_specific_ALL <- degs_covid_vax %>%
  select(condition, genes, log2fold_change) %>%
  autoGSEA(TERM2GENE = ImmuneGO_BTM_specific,
           geneset_name = "ImmuneGO_BTM_specific")

degs_covid_vax_GSEA_specific = degs_covid_vax_GSEA_specific_ALL %>%
  rownames_to_column("delete") %>%
  clean_names() %>%
    select(condition, id, everything())

#Save
write_csv(degs_covid_vax_GSEA_specific,
          file = here("Gene set enrichment analysis", paste0(filename, geneset_name, "_GSEA_ALL", ".csv")))

degs_covid_vax_GSEA_specific 

```


```{r fig.height=8, fig.width=12}
library(ggdendro)
library(cowplot)

All_covid_13VaxImmune_GO_T2G_general_GSEA_ALL <- read_csv("Gene set enrichment analysis/All_covid_13VaxImmuneGO_BTM_general_GSEA_ALL.csv")

All_covid_13VaxImmune_GO_T2G_specfic_GSEA_ALL <- read_csv("Gene set enrichment analysis/All_covid_13VaxImmuneGO_BTM_specific_GSEA_ALL.csv")

hagan_vaxatlas_annotation = hagan_vaxatlas_annotation %>% 
  mutate(disease_vac = "V",
         week = as.factor(week))

ann_vaccines_covid_other_conditions = ann_vaccines_conditions %>%
  select(condition, vaccine, day, week, dose, infection, type, disease_vac) %>% 
  mutate(day = as.factor(day),
         week = as.factor(week),
         dose = as.factor(dose),
         set = "COVID-19") %>% 
  bind_rows(hagan_vaxatlas_annotation)

```


```{r fig.height=8, fig.width=20}
# Function for clustering by day
cluster_by_day <- function(df, day_column = "day", condition_column = "condition") {
  
  days <- levels(df[[day_column]])
  orders <- list()
  
  for (day_value in days) {
    day_data <- df %>%
      filter(!!sym(day_column) == day_value) %>%
      select(!!sym(condition_column)) %>%
      distinct()
    
    if (nrow(day_data) > 1) {
      presence_matrix <- table(day_data[[condition_column]]) > 0
      dist_matrix <- as.matrix(presence_matrix)
      col_clustered <- dist(dist_matrix) %>% hclust()
      
      orders[[day_value]] <- col_clustered$labels[col_clustered$order]
    } else {
      orders[[day_value]] <- as.character(day_data[[condition_column]])
    }
  }
  
  order_df <- data.frame(
    day = rep(names(orders), sapply(orders, length)),
    condition = unlist(orders)
  )
  
  return(order_df)
}

```

##### COVID and Other

###### Main processes
By day
```{r fig.height=15, fig.width=30}
df = All_covid_13VaxImmune_GO_T2G_general_GSEA_ALL %>% 
  filter(qvalue<= 0.25) %>%
  inner_join(ann_vaccines_covid_other_conditions, by = "condition") %>% 
    mutate(logq = -log10(qvalue)) 
  
df =  df %>% 
    mutate(day = fct_relevel(day, df %>% 
                               select(day) %>% 
                               distinct() %>% 
                               mutate(day = as.character(day),
                                      day = as.numeric(day)) %>% 
                               arrange(day) %>% 
                               pull() %>% 
                               as.character()))


row_clustered = df %>% 
  select(condition, id, nes) %>% 
  distinct() %>% 
  pivot_wider(names_from = "condition",
              values_from= "nes") %>%
  column_to_rownames("id") %>% 
  as.matrix() %>% 
  replace(is.na(.), 0) %>% 
  dist(.) %>% 
  hclust()

row_shared =  df %>%
  filter(id %in% row_clustered$labels, 
         p_adjust <= 0.10) %>% 
  mutate(id = str_to_title(id),
         id = case_when(id == "Humoral Adaptive Immune System" ~ "Humoral Adap.",
                        id == "Cellular Adaptive Immune System" ~ "Cellular Adap.",
                        id == "Adaptive Immune System" ~ "Adap.",
                        id == "Innate Immune System" ~ "Innate",
                        id == "Complement Immune System" ~ "Complement",
                        id == "Immunoglobulin Mediated Response" ~ "Ig-mediated",
                        id == "Antiviral And Interferon" ~ "Antiviral/IFN",
                        .default = id)) %>% 
  group_by(id) %>% 
  summarize(n = n()) %>% 
  ungroup() %>% 
  arrange(-n)

result_vac <- cluster_by_day(df %>% filter(disease_vac == "V"), day_column = "day", condition_column = "condition") %>%
  distinct() %>% 
  mutate(day = fct_relevel(day, df$day %>% unique() %>% sort() %>% as.character()))

result_inf_vac2 <- cluster_by_day(df %>% 
                               filter(disease_vac == "I",
                                      dose %in% c("2")), day_column = "day") %>%
  mutate(day = fct_relevel(day, df$day %>% unique() %>% sort() %>% as.character()))

result_inf_vac01 <- cluster_by_day(df %>% 
                               filter(disease_vac == "I",
                                      dose %in% c("0", "1")), day_column = "day") %>%
  mutate(day = fct_relevel(day, df$day %>% unique() %>% sort() %>% as.character()))

column_order <- c(result_vac$condition, result_inf_vac2$condition, result_inf_vac01$condition)

labels_set <- df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = set) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$set) + 
  theme_void() +
  labs(y = "Set") +
  theme(axis.title.y = element_text(size = 15, hjust = 1),
        legend.position = "right") 

labels_disease_vac <- df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = disease_vac) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$disease_vac) + 
  theme_void() +
  labs(y = "Disease_vac") +
  theme(axis.title.y = element_text(size = 15, hjust = 1),
        legend.position = "right") 

labels_type <- df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = type) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$Platform) + 
  theme_void() +
  labs(y = "Type") +
  theme(axis.title.y = element_text(size = 15, hjust = 1),
        legend.position = "right") 

labels_dose <- df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = dose) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$dose) + 
  theme_void() +
  labs(y = "Dose") +
  theme(axis.title.y = element_text(size = 15, hjust = 1),
        legend.position = "right") 

labels_week = df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = week) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$week) + 
  theme_void() +
  labs(y = "Week") +
  theme(axis.title.y = element_text(size = 15, hjust = 1),
        legend.position = "right")

labels_day = df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = day) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$day) + 
  theme_void() +
  labs(y = "Day") +
  theme(axis.title.y = element_text(size = 15, hjust = 1),
        legend.position = "right")


#Plot



dotplot = df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(signific = case_when(p_adjust <= 0.10 ~ "*",
                              TRUE ~ "")) %>% 
  mutate(condition = factor(condition, column_order),
         # id = factor(id, levels = row_clustered$labels[row_clustered$order]),
         id = str_to_title(id),
         id = case_when(id == "Humoral Adaptive Immune System" ~ "Humoral Adap.",
                        id == "Cellular Adaptive Immune System" ~ "Cellular Adap.",
                        id == "Adaptive Immune System" ~ "Adap.",
                        id == "Innate Immune System" ~ "Innate",
                        id == "Complement Immune System" ~ "Complement",
                        id == "Immunoglobulin Mediated Response" ~ "Ig-mediated",
                        id == "Antiviral And Interferon" ~ "Antiviral/IFN",.default = id),
         id = fct_relevel(id, c(row_shared$id))) %>% 
  ggplot(aes(x=condition, y = fct_rev(id), 
             fill = nes, 
             color = nes
             )) +
  geom_tile(color = "white",
            size = 0.5) +
  geom_text(aes(label = signific), size = 6, vjust = 0.5, color = "black") +
  theme_minimal() + 
  theme(axis.text = element_text(color = "black", size = 15),
        axis.line.x  = element_line(size = 1, color ="black"),
        # axis.line.y  = element_line(size = 1, color ="black"),
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1),
        axis.ticks = element_line(size = 1, color ="black"),
        # panel.grid = element_line(linetype = 2, color = "gray75"),
        panel.grid = element_blank(),
        legend.direction = "horizontal",
        legend.position = "bottom") +
  labs(x = "",
       y = "") +
  scale_fill_gradient2(low = "#4DBBD5FF",
                        mid = "white",
                        midpoint = 0, 
                        name = "NES",
                      high = "#DC0000FF") +
  # scale_color_gradient2(low = "blue",
  #                       mid = "white",
  #                       midpoint = 0, 
  #                       name = "NES",
  #                     high = "#DC0000FF") +
  scale_y_discrete(position = "right") 


patch_immune_main = 
  (labels_set /
  labels_disease_vac /
  labels_type /
  labels_dose/
  labels_day/
  dotplot) +
  plot_layout(heights = c( 0.3, 0.3, 0.3, 0.3, 0.3, 6),
              guides = 'collect',
              axes = "collect") +
  plot_annotation(title = "GSEA ImmuneGO + BTM",
                  subtitle = "Main processes, p-adjusted value <= 0.25",
                  theme = theme(legend.position = "right",
                                plot.title = element_text(hjust = 0.5, size = 20),
                                plot.subtitle = element_text(hjust = 0.5, size = 20),
                                plot.margin = margin(5, 1, 5, 1, unit = "cm"),
                                legend.direction = "vertical", 
                                legend.box = "vertical"))
  

patch_immune_main %>%
  ggsave(filename = here("Figures", paste0(filename, "_GSEA_ImmuneGO_BTM_MainGOs_Bubbleplot.png")),
         width = 30,
         height = 15)
patch_immune_main

```

By condition
By day
```{r fig.height=15, fig.width=30}
df = All_covid_13VaxImmune_GO_T2G_general_GSEA_ALL %>% 
  filter(qvalue<= 0.25) %>%
  inner_join(ann_vaccines_covid_other_conditions, by = "condition") %>% 
    mutate(logq = -log10(qvalue)) 
  
df =  df %>% 
    mutate(day = fct_relevel(day, df %>% 
                               select(day) %>% 
                               distinct() %>% 
                               mutate(day = as.character(day),
                                      day = as.numeric(day)) %>% 
                               arrange(day) %>% 
                               pull() %>% 
                               as.character()))


row_clustered = df %>% 
  select(condition, id, nes) %>% 
  distinct() %>% 
  pivot_wider(names_from = "condition",
              values_from= "nes") %>%
  column_to_rownames("id") %>% 
  as.matrix() %>% 
  replace(is.na(.), 0) %>% 
  dist(.) %>% 
  hclust()

row_shared =  df %>%
  filter(id %in% row_clustered$labels, 
         p_adjust <= 0.10) %>% 
  mutate(id = str_to_title(id),
         id = case_when(id == "Humoral Adaptive Immune System" ~ "Humoral Adap.",
                        id == "Cellular Adaptive Immune System" ~ "Cellular Adap.",
                        id == "Adaptive Immune System" ~ "Adap.",
                        id == "Innate Immune System" ~ "Innate",
                        id == "Complement Immune System" ~ "Complement",
                        id == "Immunoglobulin Mediated Response" ~ "Ig-mediated",
                        id == "Antiviral And Interferon" ~ "Antiviral/IFN",
                        .default = id)) %>% 
  group_by(id) %>% 
  summarize(n = n()) %>% 
  ungroup() %>% 
  arrange(-n)

result_vac <- cluster_by_day(df %>% filter(disease_vac == "V"), day_column = "day", condition_column = "condition") %>%
  distinct() %>% 
  mutate(day = fct_relevel(day, df$day %>% unique() %>% sort() %>% as.character()))

result_inf_vac2 <- cluster_by_day(df %>% 
                               filter(disease_vac == "I",
                                      dose %in% c("2")), day_column = "day") %>%
  mutate(day = fct_relevel(day, df$day %>% unique() %>% sort() %>% as.character()))

result_inf_vac01 <- cluster_by_day(df %>% 
                               filter(disease_vac == "I",
                                      dose %in% c("0", "1")), day_column = "day") %>%
  mutate(day = fct_relevel(day, df$day %>% unique() %>% sort() %>% as.character()))

column_order <- c(result_vac$condition, result_inf_vac2$condition, result_inf_vac01$condition)

labels_set <- df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = set) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$set) + 
  theme_void() +
  labs(y = "Set") +
  theme(axis.title.y = element_text(size = 15, hjust = 1),
        legend.position = "right") 

labels_disease_vac <- df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = disease_vac) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$disease_vac) + 
  theme_void() +
  labs(y = "Disease_vac") +
  theme(axis.title.y = element_text(size = 15, hjust = 1),
        legend.position = "right") 

labels_type <- df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = type) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$Platform) + 
  theme_void() +
  labs(y = "Type") +
  theme(axis.title.y = element_text(size = 15, hjust = 1),
        legend.position = "right") 

labels_dose <- df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = dose) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$dose) + 
  theme_void() +
  labs(y = "Dose") +
  theme(axis.title.y = element_text(size = 15, hjust = 1),
        legend.position = "right") 

labels_week = df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = week) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$week) + 
  theme_void() +
  labs(y = "Week") +
  theme(axis.title.y = element_text(size = 15, hjust = 1),
        legend.position = "right")

labels_day = df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = day) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$day) + 
  theme_void() +
  labs(y = "Day") +
  theme(axis.title.y = element_text(size = 15, hjust = 1),
        legend.position = "right")


#Plot



dotplot = df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(signific = case_when(p_adjust <= 0.10 ~ "*",
                              TRUE ~ "")) %>% 
  mutate(condition = factor(condition, column_order),
         # id = factor(id, levels = row_clustered$labels[row_clustered$order]),
         id = str_to_title(id),
         id = case_when(id == "Humoral Adaptive Immune System" ~ "Humoral Adap.",
                        id == "Cellular Adaptive Immune System" ~ "Cellular Adap.",
                        id == "Adaptive Immune System" ~ "Adap.",
                        id == "Innate Immune System" ~ "Innate",
                        id == "Complement Immune System" ~ "Complement",
                        id == "Immunoglobulin Mediated Response" ~ "Ig-mediated",
                        id == "Antiviral And Interferon" ~ "Antiviral/IFN",.default = id),
         id = fct_relevel(id, c(row_shared$id))) %>% 
  ggplot(aes(x=condition, y = fct_rev(id), 
             fill = nes, 
             color = nes
             )) +
  geom_tile(color = "white",
            size = 0.5) +
  geom_text(aes(label = signific), size = 6, vjust = 0.5, color = "black") +
  theme_minimal() + 
  theme(axis.text = element_text(color = "black", size = 15),
        axis.line.x  = element_line(size = 1, color ="black"),
        # axis.line.y  = element_line(size = 1, color ="black"),
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1),
        axis.ticks = element_line(size = 1, color ="black"),
        # panel.grid = element_line(linetype = 2, color = "gray75"),
        panel.grid = element_blank(),
        legend.direction = "horizontal",
        legend.position = "bottom") +
  labs(x = "",
       y = "") +
  scale_fill_gradient2(low = "#4DBBD5FF",
                        mid = "white",
                        midpoint = 0, 
                        name = "NES",
                      high = "#DC0000FF") +
  # scale_color_gradient2(low = "blue",
  #                       mid = "white",
  #                       midpoint = 0, 
  #                       name = "NES",
  #                     high = "#DC0000FF") +
  scale_y_discrete(position = "right") 


patch_immune_main = 
  (labels_set /
  labels_disease_vac /
  labels_type /
  labels_dose/
  labels_day/
  dotplot) +
  plot_layout(heights = c( 0.3, 0.3, 0.3, 0.3, 0.3, 6),
              guides = 'collect',
              axes = "collect") +
  plot_annotation(title = "GSEA ImmuneGO + BTM",
                  subtitle = "Main processes, p-adjusted value <= 0.25",
                  theme = theme(legend.position = "right",
                                plot.title = element_text(hjust = 0.5, size = 20),
                                plot.subtitle = element_text(hjust = 0.5, size = 20),
                                plot.margin = margin(5, 1, 5, 1, unit = "cm"),
                                legend.direction = "vertical", 
                                legend.box = "vertical"))
  

patch_immune_main %>%
  ggsave(filename = here("Figures", paste0(filename, "_GSEA_ImmuneGO_BTM_MainGOs_Bubbleplot_Condition.png")),
         width = 30,
         height = 15)
patch_immune_main

```

###### Main Processes GSEA Timeline

```{r fig.height=10, fig.width=10}

ann_vaccines = ann_vaccines_covid_other_conditions

gsea_l2fc_tbl = All_covid_13VaxImmune_GO_T2G_general_GSEA_ALL %>% 
  inner_join(ann_vaccines, by = "condition")  %>% 
  filter(!condition %in% c("ChAd-BNT (V2, D0)",
                           "BNT-I-BNT (D51, hospital)")) %>% 
  select(condition, type, dose, day, pathogen, everything())  %>% 
  select(-delete)

gsea_l2fc_BNT_MO_before <- gsea_l2fc_tbl %>%
  filter(condition %in% c("BNT (V2, D1)", "BNT (V1, D6)")) %>% 
  mutate(condition = if_else(condition == "BNT (V1, D6)", "BNT-MO (V1, D6)", "BNT-MO (V2, D1)"),
         vaccine = "MO")

gsea_l2fc_I <- gsea_l2fc_tbl %>%
  filter(disease_vac == "I") %>% 
  select(condition, pathogen, everything())

gsea_l2fc_tbl = gsea_l2fc_tbl %>% 
  filter(!condition %in% c(gsea_l2fc_I$condition %>% unique()))

gsea_l2fc_tbl_2 = bind_rows(gsea_l2fc_tbl,
            gsea_l2fc_BNT_MO_before,
            gsea_l2fc_I) %>% 
    mutate(day = as.character(day),
           day = as.numeric(day),
         pathogen = if_else(is.na(pathogen) & disease_vac == "V", 
                            paste0("SARS-CoV-2/", vaccine), 
                            pathogen),
         pathogen = str_to_upper(pathogen),
         pathogen = case_when(str_detect(condition, "BNT-I.*hospital") ~ "SARS-COV-2/BNT-I Hospital",
                              str_detect(condition, "I.*hospital") ~ "SARS-COV-2/I Hospital", 
                              condition == "BNT-I (D0-5)" ~ "SARS-COV-2/BNT-I", 
                              condition == "I (D0-4)" ~ "SARS-COV-2/I", 
                              condition == "I-BNT (D5)" ~ "SARS-COV-2/I-BNT", 
                              condition == "I-I (D0-5)" ~ "SARS-COV-2/I-I", 
                              .default = pathogen)) %>% 
    arrange(day) %>% 
    mutate(type_grouped = case_when(type == "CONJ" ~ "Conjugated/PS",
                                 type == "PS" ~ "Conjugated/PS",
                                 type == "VV" ~ "Genetic",
                                 type == "RNA" ~ "Genetic",
                                 type == "LA" ~ "Live attenuated",
                                 type == "IN" ~ "Inactivated",
                                 type == "SU" ~ "Subunit",
                                 disease_vac == "I" ~ "Infection",
                                 TRUE ~ type),
         type = case_when(disease_vac == "I" ~ "I",
                                 TRUE ~ type),
         type_grouped = as.factor(type_grouped),
         type_grouped = fct_relevel(type_grouped, 
                            "Infection",
                            "Live Attenuated",
                            "Inactivated",
                            "Conjugated/PS",
                            "Subunit",
                            "Genetic"),
         id = case_when(id == "HUMORAL ADAPTIVE IMMUNE SYSTEM" ~ "HUMORAL ADAP.",
                        id == "CELLULAR ADAPTIVE IMMUNE SYSTEM" ~ "CELULLAR ADAP.",
                        id == "ADAPTIVE IMMUNE SYSTEM" ~ "ADAPTIVE",
                        id == "INNATE IMMUNE SYSTEM" ~ "INNATE",
                        id == "IMMUNOGLOBULIN MEDIATED RESPONSE" ~ "IMMUNOGLOBULIN\nMEDIATED RESPONSE",
                        TRUE ~ id)) %>% 
  select(condition, type, type_grouped, dose, day, pathogen, everything()) 

gsea_l2fc_day0 <- gsea_l2fc_tbl_2 %>%
  select(condition, type, type_grouped, dose, pathogen, day, id) %>% 
  mutate(day = 0, 
         nes = 0, 
         dose = "0",
         condition = paste0(pathogen, " D(0)")) %>% 
  distinct()

gsea_l2fc_tbl_2 = gsea_l2fc_tbl_2 %>% 
  bind_rows(gsea_l2fc_day0) %>% 
  arrange(condition)

gsea_l2fc_tbl_2 %>% 
  select(condition, type, type_grouped, dose, pathogen, day) %>% 
  distinct() %>% 
  arrange(pathogen, day)
  

```


```{r fig.height=10, fig.width=10}
idorder = c(
  "IMMUNE SYSTEM",
  "ANTIMICROBIAL",
  "LEUKOCYTES",
  "COMPLEMENT",
  "INNATE IMMUNE SYSTEM",
  "INNATE",
  "INFLAMMATION",
  "SIGNALING",
  "ANTIVIRAL & IFN",
  "NEUTROPHIL",
  "MAST CELL",
  "EOSINOPHIL",
  "MAST CELL",
  "MYELOID CELL",
  "ARPP",
  "MONOCYTE",
  "MACROPHAGE",
  "DENDRITIC CELL",
  "NK CELL",
  "ADAPTIVE IMMUNE SYSTEM",
  "ADAPTIVE",
  "LYMPHOCYTE",
  "CELLULAR ADAPTIVE IMMUNE SYSTEM",
  "CELULLAR ADAP.",
  "T CELL",
  "HUMORAL ADAPTIVE IMMUNE SYSTEM",
  "HUMORAL ADAP.",
  "B CELL",
  "IMMUNOGLOBULIN MEDIATED RESPONSE",
  "IMMUNOGLOBULIN\nMEDIATED RESPONSE"
)



```

```{r fig.height=20, fig.width=15}
day_filter = 30
all_together = gsea_l2fc_tbl_2 %>%
  filter(day <= day_filter, 
         type != "IN/SU") %>% 
  mutate(id = fct_relevel(id, c(idorder)),
         dose = as.factor(dose),
         dose = fct_relevel(dose, "0", "1", "2", "3")) %>%
  mutate(type_grouped = str_to_upper(type_grouped),
         type_grouped = as.factor(type_grouped),
         type_grouped = fct_relevel(type_grouped, 
                            "INFECTION",
                            "LIVE ATTENUATED",
                            "INACTIVATED",
                            "CONJUGATED/PS",
                            "SUBUNIT",
                            "GENETIC")) %>% 
  
  ggplot() +
  aes(x = day, y = nes, group = vaccine) +
  geom_line(
  aes(color = pathogen,
      group = pathogen
    ),
  size = 1) +
  geom_point(aes(shape = dose,
                 color = pathogen),
             size = 3.5,
             color = "black",
             alpha = 1) +
  geom_point(aes(shape = dose,
                 color = pathogen),             
             size = 3,
             alpha = 0.7) +
  geom_hline(yintercept = 0, linetype = 1, alpha = 1) +
  scale_color_manual(values = c(ann_vaxsig_colors$type, 
                                ann_vaxsig_colors$target_pathogen_disease,
                                ann_vaxsig_colors$dose,
                                "SARS-COV-2/BBIBP" = "#00A087FF",
                                "SARS-COV-2/ZF2001" = "#8491B4FF",
                                "SARS-COV-2/BNT" = "#4DBBD5FF",
                                "SARS-COV-2/MO" = "#72ddf7",
                                "SARS-COV-2/CHAD" = "#3C5488FF",
                                "SARS-COV-2/I" = "red",
                                "SARS-COV-2/BNT-I" = "#ff006e",
                                "SARS-COV-2/I-BNT" = "#Af101e"
                                )) +
  # Manual shape scale
    scale_shape_manual(values = c("0" = 15, 
                                "1" = 16, 
                                "2" = 17, 
                                "3" = 18)) +
  theme_minimal()+
  # ggthemes::theme_few() +
  theme(legend.position = "top",
        strip.text.x = element_text(color = "black", size = 12), 
        axis.text.x = element_text(color = "black", size = 10),
        axis.text.y = element_text(vjust = 0.5, size = 10, color = "black"),
        axis.line = element_line(size = 1, colour = "black", linetype=1),
        axis.ticks = element_line(size = 0.5, color = "black"),
        panel.grid.major.y = element_line(size = 0, 
                                    linetype = 2,
                                    color = "white"),
        panel.grid.major.x = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75")) +
  labs(x = "Day",
       y = "NES",
       color = "Pathogen",
       shape = "Type",
       alpha = "COVID-19 or Other",
       title = "Temporal dynamics of immunization, GSEA of Immune processes (BTM + ImmuneGO)") +
    facet_wrap(~id, 
               scales = "free",
               nrow = 6)

all_together

all_together %>%
  ggsave(filename = here("Figures", "GSEA_Linechart_ByImmuneProcess_AllVaccines_Infection.png"),
         width = 15,
         height = 20)
```



Separated
Main processes
```{r fig.height=8, fig.width=25}

process_filter = c("IMMUNE SYSTEM",
                   "COMPLEMENT",
                   "INNATE",
                   "ADAPTIVE",
                   "HUMORAL ADAP.",
                   "CELLULAR ADAP.",
                   "IMMUNOGLOBULIN\nMEDIATED RESPONSE")
day_filter = 30
time_bytype_main = gsea_l2fc_tbl_2 %>%
  filter(day <= day_filter, 
         type != "IN/SU",
         id %in% process_filter) %>% 
  mutate(id = fct_relevel(id, c(idorder)),
         dose = as.factor(dose),
         dose = fct_relevel(dose, "0", "1", "2", "3")) %>%
  mutate(type_grouped = str_to_upper(type_grouped),
         type_grouped = as.factor(type_grouped),
         type_grouped = fct_relevel(type_grouped, 
                            "INFECTION",
                            "LIVE ATTENUATED",
                            "INACTIVATED",
                            "CONJUGATED/PS",
                            "SUBUNIT",
                            "GENETIC")) %>% 

  arrange(condition) %>% 
  distinct() %>% 
  ggplot() +
  aes(x = day, 
      y = nes, 
     # label = condition,
    group = pathogen
      ) +
  geom_line(
  aes(color = pathogen,
      group = pathogen
    ),
  size = 1) +
  geom_point(aes(shape = dose,
                 color = pathogen),
             size = 3.5,
             color = "black",
             alpha = 1) +
  geom_point(aes(shape = dose,
                 color = pathogen),             
             size = 3,
             alpha = 0.7) +
  geom_hline(yintercept = 0, linetype = 1, alpha = 1) +
  scale_color_manual(values = c(ann_vaxsig_colors$type, 
                                ann_vaxsig_colors$target_pathogen_disease,
                                ann_vaxsig_colors$dose,
                                "SARS-COV-2/BBIBP" = "#00A087FF",
                                "SARS-COV-2/ZF2001" = "#8491B4FF",
                                "SARS-COV-2/BNT" = "#4DBBD5FF",
                                "SARS-COV-2/MO" = "#72ddf7",
                                "SARS-COV-2/CHAD" = "#3C5488FF",
                                "SARS-COV-2/I" = "red",
                                "SARS-COV-2/I Hospital" = "red",
                                "SARS-COV-2/I-I" = "purple",
                                "SARS-COV-2/BNT-I" = "pink",
                                "SARS-COV-2/BNT-I Hospital" = "pink",
                                "SARS-COV-2/I-BNT" = "#Af101e"
                                
                                )) +
  scale_shape_manual(values = c("0" = 15, 
                                "1" = 16, 
                                "2" = 17, 
                                "3" = 18)) +
  scale_alpha_manual(values = c("COVID-19" = 1, "13Vax Atlas" = 1), guide = "none") +
  scale_x_continuous(breaks = c(0, 3, 7, 14, 21, 28, 30)) +
  theme_minimal() +

  labs(x = "Day",
       y = "NES",
       color = "Pathogen",
       shape = "Dose",
       alpha = "COVID-19 or Other",
       title = "Temporal dynamics of immunization", 
       subtitle = "Immune System process",
       linetype = "Set") +
  facet_grid(~id~type_grouped, scales = "free") +
  theme(legend.position = "right",
        strip.text.x = element_text(color = "black", 
                                    size = 10,
                                    face = "bold"), 
        strip.text.y = element_text(color = "black", 
                                    size = 10, 
                                    hjust = 0,
                                    angle = 360,
                                    face = "bold"), 
        axis.text.x = element_text(color = "black", size = 10),
        axis.text.y = element_text(vjust = 0.5, size = 10, color = "black"),
        axis.line = element_line(size = 1, colour = "black", linetype=1),
        axis.ticks = element_line(size = 0.5, color = "black"),
        panel.grid.major.y = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.minor.y = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.major.x = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.minor.x = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75")) 

time_bytype_main 

time_bytype_main %>%
  ggsave(filename = here("Figures", "GSEA_Linechart_MainProcesses_AllVaccines_Infection_FacetByType.png"),
         width = 25,
         height = 8)

# all_together_sarsfocus %>% ggplotly()
```

Subprocesses
```{r fig.height=25, fig.width=15}
process_filter = c(
  "ANTIMICROBIAL",
  "LEUKOCYTES",
  "INFLAMMATION",
  "SIGNALING",
  "ANTIVIRAL & IFN",
  "NEUTROPHIL",
  "ARPP",
  "MONOCYTE",
  "MACROPHAGE",
  "DENDRITIC CELL",
  "NK CELL",
  "T CELL",
  "B CELL"
)

day_filter = 30
time_bytype_sub = gsea_l2fc_tbl_2 %>%
  filter(day <= day_filter, 
         type != "IN/SU",
         id %in% process_filter) %>% 
  mutate(id = fct_relevel(id, c(idorder)),
         dose = as.factor(dose),
         dose = fct_relevel(dose, "0", "1", "2", "3")) %>% 
  mutate(type_grouped = str_to_upper(type_grouped),
         type_grouped = as.factor(type_grouped),
         type_grouped = fct_relevel(type_grouped, 
                            "INFECTION",
                            "LIVE ATTENUATED",
                            "INACTIVATED",
                            "CONJUGATED/PS",
                            "SUBUNIT",
                            "GENETIC")) %>% 
  arrange(condition) %>% 
  distinct() %>% 
  ggplot() +
  aes(x = day, 
      y = nes, 
     # label = condition,
    group = pathogen
      ) +
  geom_line(
  aes(color = pathogen,
      group = pathogen
    ),
  size = 0.7) +
  geom_point(aes(shape = dose,
                 color = pathogen),
             size = 3.5,
             color = "black",
             alpha = 1) +
  geom_point(aes(shape = dose,
                 color = pathogen),             
             size = 3,
             alpha = 0.7) +
  geom_hline(yintercept = 0, linetype = 1, alpha = 1) +
  scale_color_manual(values = c(ann_vaxsig_colors$type, 
                                ann_vaxsig_colors$target_pathogen_disease,
                                ann_vaxsig_colors$dose,
                                "SARS-COV-2/BBIBP" = "#00A087FF",
                                "SARS-COV-2/ZF2001" = "#8491B4FF",
                                "SARS-COV-2/BNT" = "#4DBBD5FF",
                                "SARS-COV-2/MO" = "#72ddf7",
                                "SARS-COV-2/CHAD" = "#3C5488FF",
                                "SARS-COV-2/I" = "red",
                                "SARS-COV-2/I Hospital" = "red",
                                "SARS-COV-2/I-I" = "purple",
                                "SARS-COV-2/BNT-I" = "pink",
                                "SARS-COV-2/BNT-I Hospital" = "pink",
                                "SARS-COV-2/I-BNT" = "#Af101e"
                                
                                )) +
  scale_shape_manual(values = c("0" = 15, 
                                "1" = 16, 
                                "2" = 17, 
                                "3" = 18)) +
  scale_alpha_manual(values = c("COVID-19" = 1, "13Vax Atlas" = 1), guide = "none") +
  scale_x_continuous(breaks = c(0, 3, 7, 14, 21, 28, 30)) +
  theme_minimal() +

  labs(x = "Day",
       y = "NES",
       color = "Pathogen",
       shape = "Dose",
       alpha = "COVID-19 or Other",
       title = "Temporal dynamics of immunization", 
       subtitle = "Other processes",
       linetype = "Set") +
  facet_grid(~id~type_grouped, scales = "free") +
  theme(legend.position = "right",
        strip.text.x = element_text(color = "black", 
                                    size = 10,
                                    face = "bold"), 
        strip.text.y = element_text(color = "black", 
                                    size = 10, 
                                    hjust = 0,
                                    angle = 360,
                                    face = "bold"), 
        axis.text.x = element_text(color = "black", size = 10),
        axis.text.y = element_text(vjust = 0.5, size = 10, color = "black"),
        axis.line = element_line(size = 1, colour = "black", linetype=1),
        axis.ticks = element_line(size = 0.5, color = "black"),
        panel.grid.major.y = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.minor.y = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.major.x = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.minor.x = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75"))  

time_bytype_sub 

time_bytype_sub %>%
  ggsave(filename = here("Figures", "GSEA_Linechart_SubProcesses_AllVaccines_Infection_FacetByType.png"),
         width = 15,
         height = 20)
```

###### COVID-19 ONLY
Separated
All processes
```{r fig.height=17, fig.width=12}

process_filter = c("IMMUNE SYSTEM",
                   "COMPLEMENT",
                   "INNATE",
                   "ADAPTIVE",
                   "HUMORAL ADAP.",
                   "CELLULAR ADAP.",
                   "IMMUNOGLOBULIN\nMEDIATED RESPONSE")
day_filter = 30
time_bytype_all_COVID = gsea_l2fc_tbl_2 %>%
  filter(day <= day_filter, 
         type != "IN/SU",
         # id %in% process_filter,
         str_detect(pathogen, "SARS")) %>% 
  mutate(id = fct_relevel(id, c(idorder)),
         dose = as.factor(dose),
         dose = fct_relevel(dose, "0", "1", "2", "3")) %>%
  mutate(type_grouped = str_to_upper(type_grouped),
         type_grouped = as.factor(type_grouped),
         type_grouped = fct_relevel(type_grouped, 
                            "INFECTION",
                            "LIVE ATTENUATED",
                            "INACTIVATED",
                            "CONJUGATED/PS",
                            "SUBUNIT",
                            "GENETIC")) %>% 

  arrange(condition) %>% 
  distinct() %>% 
  ggplot() +
  aes(x = day, 
      y = nes, 
     # label = condition,
    group = pathogen
      ) +
  geom_line(
  aes(color = pathogen,
      group = pathogen
    # alpha = set
    # position = position_jitter(width = 0, height = 0.2),
    # linetype = if_else(set == "COVID-19", "dashed",  "solid"),
    # size = if_else(set == "COVID-19", 1,  0.5)
    ),
  size = 1) +
  geom_point(aes(shape = dose,
                 color = pathogen),
             size = 3.5,
             color = "black",
             alpha = 1) +
  geom_point(aes(shape = dose,
                 color = pathogen),             
             size = 3,
             alpha = 0.7) +
  geom_hline(yintercept = 0, linetype = 1, alpha = 1) +
  # geom_hline(yintercept = 1, linetype = "dashed", alpha = 0.8) +
  # geom_hline(yintercept = -1, linetype = "dashed", alpha = 0.8) +
  # Manual color scale for pathogen and custom colors for specific vaccines
  scale_color_manual(values = c(ann_vaxsig_colors$type, 
                                ann_vaxsig_colors$target_pathogen_disease,
                                ann_vaxsig_colors$dose,
                                "SARS-COV-2/BBIBP" = "#00A087FF",
                                "SARS-COV-2/ZF2001" = "#8491B4FF",
                                "SARS-COV-2/BNT" = "#4DBBD5FF",
                                "SARS-COV-2/MO" = "#72ddf7",
                                "SARS-COV-2/CHAD" = "#3C5488FF",
                                "SARS-COV-2/I" = "red",
                                "SARS-COV-2/I Hospital" = "red",
                                "SARS-COV-2/I-I" = "purple",
                                "SARS-COV-2/BNT-I" = "pink",
                                "SARS-COV-2/BNT-I Hospital" = "pink",
                                "SARS-COV-2/I-BNT" = "#Af101e"
                                
                                )) +
  scale_shape_manual(values = c("0" = 15, 
                                "1" = 16, 
                                "2" = 17, 
                                "3" = 18)) +
  scale_alpha_manual(values = c("COVID-19" = 1, "13Vax Atlas" = 1), guide = "none") +
  scale_x_continuous(breaks = c(0, 3, 7, 14, 21, 28, 30)) +
  theme_minimal() +
  labs(x = "Day",
       y = "NES",
       color = "Pathogen",
       shape = "Dose",
       alpha = "COVID-19 or Other",
       title = "Temporal dynamics of immunization", 
       subtitle = "Immune System process",
       linetype = "Set") +
  facet_wrap(~id, scales = "free",
             ncol = 4) +
  theme(legend.position = "right",
        strip.text.x = element_text(color = "black", 
                                    size = 10,
                                    face = "bold"), 
        strip.text.y = element_text(color = "black", 
                                    size = 10, 
                                    hjust = 0,
                                    angle = 360,
                                    face = "bold"), 
        axis.text.x = element_text(color = "black", size = 10),
        axis.text.y = element_text(vjust = 0.5, size = 10, color = "black"),
        axis.line = element_line(size = 1, colour = "black", linetype=1),
        axis.ticks = element_line(size = 0.5, color = "black"),
        panel.grid.major.y = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.minor.y = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.major.x = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.minor.x = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75")) 

time_bytype_all_COVID 

time_bytype_all_COVID %>%
  ggsave(filename = here("Figures", "GSEA_Linechart_AllProcesses_COVID_Infection_FacetByType.png"),
         width = 12,
         height = 17)

# all_together_sarsfocus %>% ggplotly()
```

Main
```{r fig.height=6, fig.width=20}

process_filter = c("IMMUNE SYSTEM",
                   "COMPLEMENT",
                   "INNATE",
                   "ADAPTIVE",
                   "HUMORAL ADAP.",
                   "CELLULAR ADAP.",
                   "IMMUNOGLOBULIN\nMEDIATED RESPONSE")
day_filter = 30
time_bytype_main_COVID = gsea_l2fc_tbl_2 %>%
  filter(day <= day_filter, 
         type != "IN/SU",
         id %in% process_filter,
         str_detect(pathogen, "SARS")) %>% 
  mutate(id = fct_relevel(id, c(idorder)),
         dose = as.factor(dose),
         dose = fct_relevel(dose, "0", "1", "2", "3")) %>%
  mutate(type_grouped = str_to_upper(type_grouped),
         type_grouped = as.factor(type_grouped),
         type_grouped = fct_relevel(type_grouped, 
                            "INFECTION",
                            "LIVE ATTENUATED",
                            "INACTIVATED",
                            "CONJUGATED/PS",
                            "SUBUNIT",
                            "GENETIC")) %>% 

  arrange(condition) %>% 
  distinct() %>% 
  ggplot() +
  aes(x = day, 
      y = nes, 
     # label = condition,
    group = pathogen
      ) +
  geom_line(
  aes(color = pathogen,
      group = pathogen
    # alpha = set
    # position = position_jitter(width = 0, height = 0.2),
    # linetype = if_else(set == "COVID-19", "dashed",  "solid"),
    # size = if_else(set == "COVID-19", 1,  0.5)
    ),
  size = 1) +
  geom_point(aes(shape = dose,
                 color = pathogen),
             size = 3.5,
             color = "black",
             alpha = 1) +
  geom_point(aes(shape = dose,
                 color = pathogen),             
             size = 3,
             alpha = 0.7) +
  geom_hline(yintercept = 0, linetype = 1, alpha = 1) +
  # geom_hline(yintercept = 1, linetype = "dashed", alpha = 0.8) +
  # geom_hline(yintercept = -1, linetype = "dashed", alpha = 0.8) +
  # Manual color scale for pathogen and custom colors for specific vaccines
  scale_color_manual(values = c(ann_vaxsig_colors$type, 
                                ann_vaxsig_colors$target_pathogen_disease,
                                ann_vaxsig_colors$dose,
                                "SARS-COV-2/BBIBP" = "#00A087FF",
                                "SARS-COV-2/ZF2001" = "#8491B4FF",
                                "SARS-COV-2/BNT" = "#4DBBD5FF",
                                "SARS-COV-2/MO" = "#72ddf7",
                                "SARS-COV-2/CHAD" = "#3C5488FF",
                                "SARS-COV-2/I" = "red",
                                "SARS-COV-2/I Hospital" = "red",
                                "SARS-COV-2/I-I" = "purple",
                                "SARS-COV-2/BNT-I" = "pink",
                                "SARS-COV-2/BNT-I Hospital" = "pink",
                                "SARS-COV-2/I-BNT" = "#Af101e"
                                
                                )) +
  scale_shape_manual(values = c("0" = 15, 
                                "1" = 16, 
                                "2" = 17, 
                                "3" = 18)) +
  scale_alpha_manual(values = c("COVID-19" = 1, "13Vax Atlas" = 1), guide = "none") +
  scale_x_continuous(breaks = c(0, 3, 7, 14, 21, 28, 30)) +
  theme_minimal() +
  labs(x = "Day",
       y = "NES",
       color = "Pathogen",
       shape = "Dose",
       alpha = "COVID-19 or Other",
       title = "Temporal dynamics of immunization", 
       subtitle = "Immune System process",
       linetype = "Set") +
  facet_wrap(~id, scales = "free",
             nrow = 1) +
  theme(legend.position = "right",
        strip.text.x = element_text(color = "black", 
                                    size = 10,
                                    face = "bold"), 
        strip.text.y = element_text(color = "black", 
                                    size = 10, 
                                    hjust = 0,
                                    angle = 360,
                                    face = "bold"), 
        axis.text.x = element_text(color = "black", size = 10),
        axis.text.y = element_text(vjust = 0.5, size = 10, color = "black"),
        axis.line = element_line(size = 1, colour = "black", linetype=1),
        axis.ticks = element_line(size = 0.5, color = "black"),
        panel.grid.major.y = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.minor.y = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.major.x = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.minor.x = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75"),
        plot.margin = margin(2, 0, 2, 0, unit = "cm")) 

time_bytype_main_COVID 

time_bytype_main_COVID %>%
  ggsave(filename = here("Figures", "GSEA_Linechart_MainProcesses_COVID_Infection_FacetByType.png"),
         width = 20,
         height = 6)

# all_together_sarsfocus %>% ggplotly()
```

Subprocesses
```{r fig.height=20, fig.width=15}
process_filter = c(
  "ANTIMICROBIAL",
  "LEUKOCYTES",
  "INFLAMMATION",
  "SIGNALING",
  "ANTIVIRAL & IFN",
  "NEUTROPHIL",
  "ARPP",
  "MONOCYTE",
  "MACROPHAGE",
  "DENDRITIC CELL",
  "NK CELL",
  "T CELL",
  "B CELL"
)

day_filter = 30
time_bytype_sub = gsea_l2fc_tbl_2 %>%
  filter(day <= day_filter, 
         type != "IN/SU",
         id %in% process_filter) %>% 
  mutate(id = fct_relevel(id, c(idorder)),
         dose = as.factor(dose),
         dose = fct_relevel(dose, "0", "1", "2", "3")) %>% 
  mutate(type_grouped = str_to_upper(type_grouped),
         type_grouped = as.factor(type_grouped),
         type_grouped = fct_relevel(type_grouped, 
                            "INFECTION",
                            "LIVE ATTENUATED",
                            "INACTIVATED",
                            "CONJUGATED/PS",
                            "SUBUNIT",
                            "GENETIC")) %>% 
  arrange(condition) %>% 
  distinct() %>% 
  ggplot() +
  aes(x = day, 
      y = nes, 
     # label = condition,
    group = pathogen
      ) +
  geom_line(
  aes(color = pathogen,
      group = pathogen
    # alpha = set
    # position = position_jitter(width = 0, height = 0.2),
    # linetype = if_else(set == "COVID-19", "dashed",  "solid"),
    # size = if_else(set == "COVID-19", 1,  0.5)
    ),
  size = 0.7) +
  geom_point(aes(shape = dose,
                 color = pathogen),
             size = 3.5,
             color = "black",
             alpha = 1) +
  geom_point(aes(shape = dose,
                 color = pathogen),             
             size = 3,
             alpha = 0.7) +
  geom_hline(yintercept = 0, linetype = 1, alpha = 1) +
  # geom_hline(yintercept = 1, linetype = "dashed", alpha = 0.8) +
  # geom_hline(yintercept = -1, linetype = "dashed", alpha = 0.8) +
  # Manual color scale for pathogen and custom colors for specific vaccines
  scale_color_manual(values = c(ann_vaxsig_colors$type, 
                                ann_vaxsig_colors$target_pathogen_disease,
                                ann_vaxsig_colors$dose,
                                "SARS-COV-2/BBIBP" = "#00A087FF",
                                "SARS-COV-2/ZF2001" = "#8491B4FF",
                                "SARS-COV-2/BNT" = "#4DBBD5FF",
                                "SARS-COV-2/MO" = "#72ddf7",
                                "SARS-COV-2/CHAD" = "#3C5488FF",
                                "SARS-COV-2/I" = "red",
                                "SARS-COV-2/I Hospital" = "red",
                                "SARS-COV-2/I-I" = "purple",
                                "SARS-COV-2/BNT-I" = "pink",
                                "SARS-COV-2/BNT-I Hospital" = "pink",
                                "SARS-COV-2/I-BNT" = "#Af101e"
                                
                                )) +
  scale_shape_manual(values = c("0" = 15, 
                                "1" = 16, 
                                "2" = 17, 
                                "3" = 18)) +
  scale_alpha_manual(values = c("COVID-19" = 1, "13Vax Atlas" = 1), guide = "none") +
  scale_x_continuous(breaks = c(0, 3, 7, 14, 21, 28, 30)) +
  theme_minimal() +
  # geom_text_repel(mapping = aes(label = condition),
  #                 size = 3,
  #                 box.padding = 0.5,
  #                 # nudge_y = 0.5,
  #                 # nudge_x = 0,
  #                 min.segment.length = 0,
  #                 data = gsea_l2fc_tbl_2 %>%
  #                   filter(id %in% process_filter,
  #                          day > 0,
  #                          condition %in% c("I-I (D0-5)",
  #                                           "I (D0-4)"))) +
  # geom_text_repel(label = "Hosp.",
  #                 size = 3,
  #                 box.padding = 0.5,
  #                 min.segment.length = 0, 
  #                 # nudge_y = 0.5,
  #                 # nudge_x = -1,
  #                 data = gsea_l2fc_tbl_2 %>%
  #                   filter(id %in% process_filter,
  #                          day <= day_filter,
  #                          day > 0,
  #                          pathogen %in% c("SARS-COV-2/BNT-I Hospital",
  #                                          "SARS-COV-2/I Hospital"))) +
  labs(x = "Day",
       y = "NES",
       color = "Pathogen",
       shape = "Dose",
       alpha = "COVID-19 or Other",
       title = "Temporal dynamics of immunization", 
       subtitle = "Other processes",
       linetype = "Set") +
  facet_grid(~id~type_grouped, scales = "free") +
  theme(legend.position = "right",
        strip.text.x = element_text(color = "black", 
                                    size = 10,
                                    face = "bold"), 
        strip.text.y = element_text(color = "black", 
                                    size = 10, 
                                    hjust = 0,
                                    angle = 360,
                                    face = "bold"), 
        axis.text.x = element_text(color = "black", size = 10),
        axis.text.y = element_text(vjust = 0.5, size = 10, color = "black"),
        axis.line = element_line(size = 1, colour = "black", linetype=1),
        axis.ticks = element_line(size = 0.5, color = "black"),
        panel.grid.major.y = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.minor.y = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.major.x = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.minor.x = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75"))  

time_bytype_sub 

time_bytype_sub %>%
  ggsave(filename = here("Figures", "GSEA_Linechart_SubProcesses_AllVaccines_Infection_FacetByType.png"),
         width = 15,
         height = 15)
```




###### Correlation analysis - Genes and processes
By process
```{r fig.height=20, fig.width=25}
library(corrplot)
process_filter = "IMMUNE SYSTEM"
genes_ImmuneSystem = degs_covid_vax %>%
  inner_join(immunego_btm_genes %>% 
               filter(is.na(immune_system)) %>% 
               select(process, genes) %>% 
               distinct(), by = "genes") %>% 
  filter(process == process_filter) %>% 
  select(condition, genes, log2fold_change) %>% 
  distinct() %>% 
  pivot_wider(names_from = "condition",
              values_from = "log2fold_change", values_fn = mean) %>% 
  column_to_rownames("genes") %>% 
  mutate_all(~ replace_na(., 0)) %>% 
  select(any_of(c(ann_vaccines_covid_other_conditions %>% 
                    arrange(day)  %>% 
                    pull(condition) %>% 
                    as.character())))

correlation_genes_ImmuneSystem = genes_ImmuneSystem %>% 
  cor(method = "spearman")

p.mat <- cor.mtest(genes_ImmuneSystem,
                   method = "spearman")

col<- colorRampPalette(c("#4DBBD5FF", "white", "#DC0000FF"))(40)
correlation_genes_ImmuneSystem_plot = corrplot(correlation_genes_ImmuneSystem, 
         p.mat = p.mat, 
         title = process_filter,
         sig.level = 0.10,
         pch.col = "gray80",
         # insig = "blank",
         diag = F,
         col = col,
         method= c("color"),
         # type="upper",
         order="hclust",
         tl.col="black",
         mar=c(0,0,5,0),
         # tl.cex = 1,
         tl.srt = 45)

```


```{r}
# Function to generate and save correlation plots for each process
generate_correlation_for_process <- function(process_filter) {
  # Filter data for the current process
  genes_data <- degs_covid_vax %>%
    inner_join(immunego_btm_genes %>% 
                 filter(is.na(immune_system)) %>% 
                 select(process, genes) %>% 
                 distinct(), by = "genes") %>% 
    filter(process == process_filter) %>% 
    select(condition, genes, log2fold_change) %>% 
    distinct() %>% 
    pivot_wider(names_from = "condition",
                values_from = "log2fold_change", values_fn = mean) %>% 
    column_to_rownames("genes") %>% 
    mutate_all(~ replace_na(., 0)) %>% 
    select(any_of(c(ann_vaccines_covid_other_conditions %>% 
                      arrange(day)  %>% 
                      pull(condition) %>% 
                      as.character())))
  
  # Compute the correlation matrix
  correlation_matrix <- cor(genes_data)
  p.mat <- cor.mtest(genes_data)
  
  # Color setup
  col<- colorRampPalette(c("#4DBBD5FF", "white", "#DC0000FF"))(40)
  
  # Generate and save the plot to a file
  png(filename = here("Figures", paste0(filename, "correlation_", process_filter, ".png")), 
      width = 30, height = 20, units = "in", res = 300)
  corrplot(correlation_matrix, 
           p.mat = p.mat, 
           title = process_filter,
           sig.level = 0.10,
           pch.col = "gray80",
           diag = FALSE,
           col = col,
           method= "color",
           order="hclust",
           tl.col="black",
           mar=c(0, 0, 5, 0),
           tl.srt = 45)
  dev.off()
}

# Iterate over each unique process and generate plots
unique_processes <- immunego_btm_genes %>% 
               filter(is.na(immune_system)) %>% 
               select(process) %>% 
               distinct() %>%  
  pull() %>% unique()

purrr::walk(unique_processes, generate_correlation_for_process)
```


By genes
```{r fig.height=20, fig.width=15}
library(corrplot)

genes_ImmuneSystem = degs_covid_vax %>%
  inner_join(immunego_btm_genes %>% 
               select(genes) %>% 
               distinct(), by = "genes") %>% 
  select(condition, genes, log2fold_change) %>% 
  distinct() %>% 
  pivot_wider(names_from = "condition",
              values_from = "log2fold_change", values_fn = mean) %>% 
  column_to_rownames("genes") %>% 
  mutate_all(~ replace_na(., 0)) %>% 
  select(any_of(c(ann_vaccines_covid_other_conditions %>% 
                    arrange(day)  %>% 
                    pull(condition) %>% 
                    as.character())))

correlation_genes_ImmuneSystem = genes_ImmuneSystem %>% 
  cor(method = "spearman") %>% 
  as.data.frame() %>% 
  select(any_of(c(ann_vaccines_conditions_ordered %>% 
                  filter(disease_vac == "V") %>% 
                  arrange(condition, day) %>% 
                  pull(condition_grouped) %>% 
                  as.character(),
                ann_vaccines_conditions_ordered %>% 
                  filter(disease_vac == "I") %>% 
                  arrange(condition, day) %>% 
                  pull(condition_grouped) %>% 
                  as.character()))) %>% 
  rownames_to_column("condition") %>% 
  filter(!condition %in% c(ann_vaccines_conditions_ordered %>% 
                  arrange(condition, day) %>% 
                  pull(condition_grouped) %>% 
                  as.character())) %>% 
  inner_join(ann_vaccines_covid_other_conditions %>%
                select(condition, day) %>%
                mutate(day_numeric = as.numeric(gsub("D", "", day))) %>%  # Convert "D1" to 1, "D10" to 10, etc.
                arrange(condition, day_numeric) %>%
                mutate(day = factor(day, levels = unique(day[order(day_numeric)]))) %>%  # Re-level `day` based on sorted numeric days
                select(-day), 
                           by = "condition") %>% 
  arrange(day_numeric) %>% 
  select(-day_numeric) %>%
  column_to_rownames("condition") %>% 
  as.matrix()
  

col<- colorRampPalette(c("#4DBBD5FF", "white", "#DC0000FF"))(11)
correlation_genes_ImmuneSystem_plot = corrplot(correlation_genes_ImmuneSystem, 
         # p.mat = p.mat,
         title = "All immune genes",
         # sig.level = 0.10,
         pch.col = "gray80",
         # insig = "blank",
         diag = F,
         col = col,
         method= c("color"),
         # type="upper",
         # order="hclust",
         tl.col="black",
         mar=c(0,0,5,0),
         tl.cex = 1,
         tl.srt = 45)

```


Clustered
```{r}
# INPUT ------
correlation_genes_ImmuneSystem

table = degs_all_updown_covid_other_immune_shared
filename = "ImmuneGO_Genes_CorrelationAnalysis_COVID_Other" 

# Matrix ------
matrix = correlation_genes_ImmuneSystem

# Annotations -------
ann_vaccines_covid = ann_vaccines_covid_other_conditions %>% 
  filter(set == "COVID-19")
ann_vaccines_other = ann_vaccines_covid_other_conditions %>% 
  filter(set == "13Vax Atlas")

# Columns
ann_cols_heatmap = matrix %>% 
  colnames() %>% 
  data.frame() %>% 
  rename(vac_condition = '.') %>% 
  merge(ann_vaccines_covid, by.x = "vac_condition", by.y = "condition", all.y = F) %>% 
  mutate(day = as.character(day) %>% 
           as.numeric(),
         pathogen = if_else(is.na(pathogen),"SARS-COV-2", pathogen) %>% 
           str_to_upper()) %>% 
  select(vac_condition, type, pathogen, day) %>% 
  distinct() %>% 
  column_to_rownames(var = "vac_condition") %>% 
  arrange(day) %>% 
  mutate(day = as.character(day) %>% fct_reorder(., day))

# Rows
ann_rows_heatmap = matrix %>%  
  as.data.frame() %>% 
  rownames_to_column("condition") %>% 
  distinct() %>% 
  inner_join(ann_vaccines_other, by = "condition") %>% 
  mutate(day = as.character(day) %>% 
           as.numeric(),
         pathogen = if_else(is.na(pathogen),"SARS-COV-2", pathogen) %>% 
           str_to_upper()) %>% 
  arrange(day) %>%
  mutate(day = as.character(day) %>% fct_reorder(., day)) %>% 
  distinct() %>% 
  column_to_rownames(var= "condition") %>% 
  select(type, pathogen, day)

# Arrange cols and rows in matrix
matrix_data = matrix %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>%
  rownames_to_column("condition") %>% 
  inner_join(ann_rows_heatmap %>% 
               rownames_to_column("condition") %>% 
               select(condition, day),
             by = "condition") %>% 
  arrange(day) %>% 
  select(-day) %>% 
  column_to_rownames("condition") %>% 
  mutate_if(is.character, as.numeric) %>%
  replace(is.na(.), 0) %>% 
  as.matrix()

#Check dimensions
dim(matrix_data)
dim(ann_rows_heatmap)
dim(ann_cols_heatmap)

col_fun <- colorRamp2(
  breaks = c(-1, -0.5, 0,0.5, 1), 
  colors = c("darkblue", "#4DBBD5FF", "white", "red", "#DC0000FF")
)

#Heatmap annotation
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "left")

row_ha = HeatmapAnnotation(df = ann_rows_heatmap,
                       col = ann_vaxsig_colors,
                       which= "row",
                       show_annotation_name = F,
                       show_legend = T)


mat = matrix_data
width_image = 35
height_image = 40
width = unit(20, "cm")
height = unit(25, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)
file = paste0("Hagan_", filename)

fileimageheatmap_grouped = paste0(filename, sep ="_", "Complexheatmap_ALL_CLUSTERED.png")

png(file = here("Figures", fileimageheatmap_grouped), 
    width = width_image, 
    height = height_image,
    units="cm", res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "left",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "Spearman\ncorrelation", #Legend
                        col = col_fun, #color
cell_fun = function(j, i, x, y, width, height, fill) {
  # Define um limite próximo de zero
  if(abs(mat[i, j]) > 0.1) {
    grid.text(sprintf("%.2f", mat[i, j]), 
              x, y, 
              gp = gpar(fontsize = 7))
  }
},
                        row_title = "",
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = gpar(col = "white", lwd = 2),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(5, "cm"),
                        column_split = NULL, 
                        row_split = NULL, 
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width,
                        height = height 
)

draw(heatmap_plot,
     merge_legend = TRUE,
     annotation_legend_side = "right", 
     heatmap_legend_side = "right")
dev.off()

```

```{r}
# INPUT ------
correlation_genes_ImmuneSystem

table = degs_all_updown_covid_other_immune_shared
filename = "ImmuneGO_Genes_CorrelationAnalysis_COVID_Other" 

# Matrix ------
matrix = correlation_genes_ImmuneSystem

# Annotations -------
ann_vaccines_covid = ann_vaccines_covid_other_conditions %>% 
  filter(set == "COVID-19")
ann_vaccines_other = ann_vaccines_covid_other_conditions %>% 
  filter(set == "13Vax Atlas")

# Columns
ann_cols_heatmap = matrix %>% 
  colnames() %>% 
  data.frame() %>% 
  rename(vac_condition = '.') %>% 
  merge(ann_vaccines_covid, by.x = "vac_condition", by.y = "condition", all.y = F) %>% 
  mutate(day = as.character(day) %>% 
           as.numeric(),
         pathogen = if_else(is.na(pathogen),"SARS-COV-2", pathogen) %>% 
           str_to_upper()) %>% 
  select(vac_condition, type, pathogen, day) %>% 
  distinct() %>% 
  column_to_rownames(var = "vac_condition") %>% 
  arrange(day) %>% 
  mutate(day = as.character(day) %>% fct_reorder(., day))

# Rows
ann_rows_heatmap = matrix %>%  
  as.data.frame() %>% 
  rownames_to_column("condition") %>% 
  distinct() %>% 
  inner_join(ann_vaccines_other, by = "condition") %>% 
  mutate(day = as.character(day) %>% 
           as.numeric(),
         pathogen = if_else(is.na(pathogen),"SARS-COV-2", pathogen) %>% 
           str_to_upper()) %>% 
  arrange(day) %>%
  mutate(day = as.character(day) %>% fct_reorder(., day)) %>% 
  distinct() %>% 
  column_to_rownames(var= "condition") %>% 
  select(type, pathogen, day)

# Arrange cols and rows in matrix
matrix_data = matrix %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  inner_join(ann_cols_heatmap %>% 
               rownames_to_column("condition") %>% 
               select(condition, day),
             by = "condition") %>% 
  arrange(day) %>% 
  select(-day) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>%
  rownames_to_column("condition") %>% 
  inner_join(ann_rows_heatmap %>% 
               rownames_to_column("condition") %>% 
               select(condition, day),
             by = "condition") %>% 
  arrange(day) %>% 
  select(-day) %>% 
  column_to_rownames("condition") %>% 
  mutate_if(is.character, as.numeric) %>%
  replace(is.na(.), 0) %>% 
  as.matrix()

#Check dimensions
dim(matrix_data)
dim(ann_rows_heatmap)
dim(ann_cols_heatmap)

col_fun <- colorRamp2(
  breaks = c(-1, -0.5, 0,0.5, 1), 
  colors = c("darkblue", "#4DBBD5FF", "white", "red", "#DC0000FF")
)

#Heatmap annotation
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "right")

row_ha = HeatmapAnnotation(df = ann_rows_heatmap,
                       col = ann_vaxsig_colors,
                       which= "row",
                       show_annotation_name = F,
                       show_legend = T)


mat = matrix_data
width_image = 35
height_image = 40
width = unit(20, "cm")
height = unit(25, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)
file = paste0("Hagan_", filename)

fileimageheatmap_grouped = paste0(filename, sep ="_", "Complexheatmap_Days.png")

png(file = here("Figures", fileimageheatmap_grouped), 
    width = width_image, 
    height = height_image,
    units="cm", res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "left",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "Spearman\ncorrelation", #Legend
                        col = col_fun, #color
cell_fun = function(j, i, x, y, width, height, fill) {
  if(abs(mat[i, j]) > 0.05) {
    grid.text(sprintf("%.2f", mat[i, j]), 
              x, y, 
              gp = gpar(fontsize = 7))
  }
},
                        row_title = "",
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = gpar(col = "white", lwd = 2),
                        cluster_rows = F,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(5, "cm"),
                        column_split = NULL, 
                        row_split = NULL, 
                        row_gap = unit(2, "mm"),
                        cluster_columns = F,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width,
                        height = height 
)

draw(heatmap_plot,
     merge_legend = TRUE,
     annotation_legend_side = "right", 
     heatmap_legend_side = "right")
dev.off()

```




###### Specific processes

```{r fig.height=45, fig.width=25}
df = All_covid_13VaxImmune_GO_T2G_specfic_GSEA_ALL %>% 
  filter(p_adjust <= 0.25) %>%
  inner_join(ann_vaccines_covid_other_conditions, by = "condition") %>% 
    mutate(logpadj = -log10(p_adjust)) 

btm_annotation_immune_table_labelled
  
df =  df %>% 
    mutate(day = fct_relevel(day, df %>% 
                               select(day) %>% 
                               distinct() %>% 
                               mutate(day = as.character(day),
                                      day = as.numeric(day)) %>% 
                               arrange(day) %>% 
                               pull() %>% 
                               as.character()))


row_clustered = df %>% 
  select(condition, id, nes) %>% 
  distinct() %>% 
  pivot_wider(names_from = "condition",
              values_from= "nes") %>%
  column_to_rownames("id") %>% 
  as.matrix() %>% 
  replace(is.na(.), 0) %>% 
  dist(.) %>% 
  hclust()

row_shared =  df %>%
  filter(p_adjust <= 0.10) %>% 
  mutate(id = factor(id, levels = row_clustered$labels[row_clustered$order])) %>% 
  group_by(id) %>% 
  summarize(n = n()) %>% 
  ungroup() %>% 
  arrange(-n) %>% 
  filter(n >= 15)


df = df %>% 
  filter(id %in% c(row_shared$id))


result_vac <- cluster_by_day(df %>% filter(disease_vac == "V"), day_column = "day", condition_column = "condition") %>%
  distinct() %>% 
  mutate(day = fct_relevel(day, df$day %>% unique() %>% sort() %>% as.character()))

result_inf_vac2 <- cluster_by_day(df %>% 
                               filter(disease_vac == "I",
                                      dose %in% c("2")), day_column = "day") %>%
  mutate(day = fct_relevel(day, df$day %>% unique() %>% sort() %>% as.character()))

result_inf_vac01 <- cluster_by_day(df %>% 
                               filter(disease_vac == "I",
                                      dose %in% c("0", "1")), day_column = "day") %>%
  mutate(day = fct_relevel(day, df$day %>% unique() %>% sort() %>% as.character()))

column_order <- c(result_vac$condition, result_inf_vac2$condition, result_inf_vac01$condition)

labels_set <- df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = set) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$set) + 
  theme_void() +
  labs(y = "Set") +
  theme(axis.title.y = element_text(size = 12, hjust = 1),
        legend.position = "right") 

labels_disease_vac <- df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = disease_vac) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$disease_vac) + 
  theme_void() +
  labs(y = "Infection/Vaccination") +
  theme(axis.title.y = element_text(size = 12, hjust = 1),
        legend.position = "right") 

labels_type <- df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = type) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$Platform) + 
  theme_void() +
  labs(y = "Type") +
  theme(axis.title.y = element_text(size = 12, hjust = 1),
        legend.position = "right") 

labels_dose <- df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = dose) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$dose) + 
  theme_void() +
  labs(y = "Dose") +
  theme(axis.title.y = element_text(size = 12, hjust = 1),
        legend.position = "right") 

labels_week = df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = week) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$week) + 
  theme_void() +
  labs(y = "Week") +
  theme(axis.title.y = element_text(size = 12, hjust = 1),
        legend.position = "right")

labels_day = df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = day) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$day) + 
  theme_void() +
  labs(y = "Day") +
  theme(axis.title.y = element_text(size = 12, hjust = 1),
        legend.position = "right")


#Plot
dotplot = df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(signific = case_when(p_adjust <= 0.10 ~ "*",
                              TRUE ~ "")) %>% 
  mutate(condition = factor(condition, column_order)) %>% 
  ggplot(aes(x=condition, y = fct_rev(id), 
             fill = nes, 
             color = nes
             )) +
  geom_tile(color = "white",
            size = 0.5) +
  geom_text(aes(label = signific), size = 6, vjust = 0.5, color = "black") +
  theme_minimal() + 
  theme(axis.text = element_text(color = "black", size = 15),
        axis.line.x  = element_line(size = 1, color ="black"),
        # axis.line.y  = element_line(size = 1, color ="black"),
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1),
        axis.ticks = element_line(size = 1, color ="black"),
        # panel.grid = element_line(linetype = 2, color = "gray75"),
        panel.grid = element_blank(),
        legend.direction = "horizontal",
        legend.position = "bottom") +
  labs(x = "",
       y = "") +
  scale_fill_gradient2(low = "#4DBBD5FF",
                        mid = "white",
                        midpoint = 0, 
                        name = "NES",
                      high = "#DC0000FF") +
  # scale_color_gradient2(low = "blue",
  #                       mid = "white",
  #                       midpoint = 0, 
  #                       name = "NES",
  #                     high = "#DC0000FF") +
  scale_y_discrete(position = "right")
  

patch_immune_main = 
  labels_set /
  labels_disease_vac /
  labels_type /
  labels_dose/
  labels_day/
  dotplot +
  plot_layout(heights = c(0.05, 0.05, 0.05, 0.05, 0.05, 5),
              guides = 'collect') +
  plot_annotation(title = "GSEA ImmuneGO_BTM",
                  subtitle = "Specific processes",
                  theme = theme(legend.position = "right",
                                plot.title = element_text(hjust = 0.5, size = 20),
                                plot.subtitle = element_text(hjust = 0.5, size = 20),
                                plot.margin = margin(5, 1, 5, 1, unit = "cm"),
                                legend.direction = "vertical", 
                                legend.box = "vertical"))

patch_immune_main %>% 
  ggsave(filename = here("Figures", paste0(filename, "_GSEA_ImmuneGO_Specific_Heatmap.png")),
         width = 30,
         height = 50, limitsize = F)
patch_immune_main

```

Specific filtered

```{r fig.height=15, fig.width=30}
df = All_covid_13VaxImmune_GO_T2G_specfic_GSEA_ALL %>% 
  inner_join(ann_vaccines_covid_other_conditions, by = "condition") %>% 
    mutate(logpadj = -log10(p_adjust))

#Filter most shared modules
filtered_mods = df %>% 
  filter(p_adjust <= 0.25) %>%
  group_by(id, disease_vac) %>% 
  summarize(n = n()) %>% 
  filter(disease_vac == "V" & n > 15 | id == "Plasma cells, immunoglobulins") %>% 
  .$id

df = df %>% 
  filter(id %in% c(filtered_mods))
  
df =  df %>% 
    mutate(day = fct_relevel(day, df %>% 
                               select(day) %>% 
                               distinct() %>% 
                               mutate(day = as.character(day),
                                      day = as.numeric(day)) %>% 
                               arrange(day) %>% 
                               pull() %>% 
                               as.character()))


row_clustered = df %>% 
  filter(disease_vac == "V") %>% 
  select(condition, id, nes) %>% 
  distinct() %>% 
  pivot_wider(names_from = "condition",
              values_from= "nes") %>%
  column_to_rownames("id") %>% 
  as.matrix() %>% 
  replace(is.na(.), 0) %>% 
  dist(.) %>% 
  hclust()

result_vac <- cluster_by_day(df %>% filter(disease_vac == "V"), day_column = "day", condition_column = "condition") %>%
  distinct() %>% 
  mutate(day = fct_relevel(day, df$day %>% unique() %>% sort() %>% as.character()))

result_inf_vac01 <- cluster_by_day(df %>% 
                               filter(disease_vac == "I",
                                      dose %in% c("0", "1")), day_column = "day") %>%
  mutate(day = fct_relevel(day, df$day %>% unique() %>% sort() %>% as.character()))

result_inf_vac2 <- cluster_by_day(df %>% 
                               filter(disease_vac == "I",
                                      dose %in% c("2")), day_column = "day") %>%
  mutate(day = fct_relevel(day, df$day %>% unique() %>% sort() %>% as.character()))

column_order <- c(result_vac$condition, result_inf_vac2$condition, result_inf_vac01$condition)

labels_set <- df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = set) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$set) + 
  theme_void() +
  labs(y = "Set") +
  theme(axis.title.y = element_text(size = 12, hjust = 1),
        legend.position = "right") 

labels_disease_vac <- df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = disease_vac) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$disease_vac) + 
  theme_void() +
  labs(y = "Infection/Vaccination") +
  theme(axis.title.y = element_text(size = 12, hjust = 1),
        legend.position = "right") 

labels_type <- df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = type) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$Platform) + 
  theme_void() +
  labs(y = "Type") +
  theme(axis.title.y = element_text(size = 12, hjust = 1),
        legend.position = "right") 

labels_dose <- df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = dose) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$dose) + 
  theme_void() +
  labs(y = "Dose") +
  theme(axis.title.y = element_text(size = 12, hjust = 1),
        legend.position = "right") 

labels_week = df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = week) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$week) + 
  theme_void() +
  labs(y = "Week") +
  theme(axis.title.y = element_text(size = 12, hjust = 1),
        legend.position = "right")

labels_day = df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = day) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$day) + 
  theme_void() +
  labs(y = "Day") +
  theme(axis.title.y = element_text(size = 12, hjust = 1),
        legend.position = "right")


#Plot
dotplot = df %>% 
  mutate(signific = case_when(p_adjust <= 0.10 ~ "*",
                              TRUE ~ "")) %>% 
  # filter(p_adjust <= 0.25) %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, column_order),
         id = factor(id, levels = row_clustered$labels[row_clustered$order]),
         id = str_to_title(id),
         id = case_when(id == "Humoral Adaptive Immune System" ~ "Humoral Adap.",
                        id == "Cellular Adaptive Immune System" ~ "Cellular Adap.",
                        id == "Complement Immune System" ~ "Complement",
                        id == "Immunoglobulin Mediated Response" ~ "Ig-mediated",
                        id == "Antiviral And Interferon" ~ "Antiviral/IFN",.default = id)) %>% 
  ggplot(aes(x=condition,
             y = id,
             fill = nes, 
             # size = logq
             )) +
  geom_tile() + 
  geom_text(aes(label = signific), size = 4, vjust = 0.5) +
  theme_minimal() + 
  theme(axis.text = element_text(color = "black", size = 12) ,
        axis.line  = element_blank(),
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1),
        axis.ticks = element_blank(),
        legend.direction = "horizontal",
        legend.position = "bottom", 
        panel.grid = element_blank() ) +
  labs(x = "",
       y = "") +
  scale_fill_gradient2(low = "#4DBBD5FF",
                        mid = "white",
                        midpoint = 0, 
                        name = "NES",
                      high = "#DC0000FF") +
  scale_y_discrete(position = "right") 
  # scale_size(range = c(2, 6))
  

patch_immune_main = 
  labels_set /
  labels_disease_vac /
  labels_type /
  labels_dose/
  labels_day/
  dotplot +
  plot_layout(heights = c(0.2, 0.2, 0.2, 0.2, 0.2, 5),
              guides = 'collect') +
  plot_annotation(title = "GSEA ImmuneGO",
                  subtitle = "Specific processes",
                  theme = theme(legend.position = "right",
                                plot.title = element_text(hjust = 0.5, size = 20),
                                plot.subtitle = element_text(hjust = 0.5, size = 20),
                                plot.margin = margin(5, 1, 5, 1, unit = "cm"),
                                legend.direction = "vertical", 
                                legend.box = "vertical"))

patch_immune_main %>% 
  ggsave(filename = here("Figures", paste0(filename, "_GSEA_ImmuneGO_Specific_Filtered_Bubbleplot.png")),
         width = 30,
         height = 15)
patch_immune_main

```


##### COVID only
By day
```{r fig.height=15, fig.width=15}
df = All_covid_13VaxImmune_GO_T2G_general_GSEA_ALL %>% 
  filter(p_adjust<= 0.25) %>%
  inner_join(ann_vaccines_covid_other_conditions, by = "condition") %>% 
  filter(set == "COVID-19")

df =  df %>% 
    mutate(day = fct_relevel(day, df %>% 
                               select(day) %>% 
                               distinct() %>% 
                               mutate(day = as.character(day),
                                      day = as.numeric(day)) %>% 
                               arrange(day) %>% 
                               pull() %>% 
                               as.character()))


row_clustered = df %>% 
  select(condition, id, nes) %>% 
  distinct() %>% 
  pivot_wider(names_from = "condition",
              values_from= "nes") %>%
  column_to_rownames("id") %>% 
  as.matrix() %>% 
  replace(is.na(.), 0) %>% 
  dist(.) %>% 
  hclust()

row_shared =  df %>%
  filter(id %in% row_clustered$labels,
         p_adjust <= 0.10) %>% 
  mutate(condition = factor(condition, column_order),
         id = factor(id, levels = row_clustered$labels[row_clustered$order]),
         id = str_to_title(id),
         id = case_when(id == "Humoral Adaptive Immune System" ~ "Humoral Adap.",
                        id == "Cellular Adaptive Immune System" ~ "Cellular Adap.",
                        id == "Adaptive Immune System" ~ "Adap.",
                        id == "Innate Immune System" ~ "Innate",
                        id == "Complement Immune System" ~ "Complement",
                        id == "Immunoglobulin Mediated Response" ~ "Ig-mediated",
                        id == "Antiviral And Interferon" ~ "Antiviral/IFN",.default = id)) %>% 
  group_by(id) %>% 
  summarize(n = n()) %>% 
  ungroup() %>% 
  arrange(-n)

result_vac <- cluster_by_day(df %>% filter(disease_vac == "V"), day_column = "day", condition_column = "condition") %>%
  distinct() %>% 
  mutate(day = fct_relevel(day, df$day %>% unique() %>% sort() %>% as.character()))

result_inf_vac2 <- cluster_by_day(df %>% 
                               filter(disease_vac == "I",
                                      dose %in% c("2")), day_column = "day") %>%
  mutate(day = fct_relevel(day, df$day %>% unique() %>% sort() %>% as.character()))

result_inf_vac01 <- cluster_by_day(df %>% 
                               filter(disease_vac == "I",
                                      dose %in% c("0", "1")), day_column = "day") %>%
  mutate(day = fct_relevel(day, df$day %>% unique() %>% sort() %>% as.character()))

column_order <- c(result_vac$condition, result_inf_vac2$condition, result_inf_vac01$condition)

labels_set <- df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = set) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$set) + 
  theme_void() +
  labs(y = "Set") +
  theme(axis.title.y = element_text(size = 12, hjust = 1),
        legend.position = "right") 

labels_disease_vac <- df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = disease_vac) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$disease_vac) + 
  theme_void() +
  labs(y = "disease_vac") +
  theme(axis.title.y = element_text(size = 12, hjust = 1),
        legend.position = "right") 

labels_type <- df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = type) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$Platform) + 
  theme_void() +
  labs(y = "Type") +
  theme(axis.title.y = element_text(size = 12, hjust = 1),
        legend.position = "right") 

labels_dose <- df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = dose) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$dose) + 
  theme_void() +
  labs(y = "Dose") +
  theme(axis.title.y = element_text(size = 12, hjust = 1),
        legend.position = "right") 

labels_week = df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = week) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$week) + 
  theme_void() +
  labs(y = "Week") +
  theme(axis.title.y = element_text(size = 12, hjust = 1),
        legend.position = "right")

labels_day = df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = day) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$day) + 
  theme_void() +
  labs(y = "Day") +
  theme(axis.title.y = element_text(size = 12, hjust = 1),
        legend.position = "right")


#Plot

dotplot = df %>% 
  filter(id %in% row_clustered$labels) %>% 
  mutate(signific = case_when(p_adjust <= 0.10 ~ "*",
                              TRUE ~ "")) %>% 
  mutate(condition = factor(condition, column_order),
         # id = factor(id, levels = row_clustered$labels[row_clustered$order]),
         id = str_to_title(id),
         id = case_when(id == "Humoral Adaptive Immune System" ~ "Humoral Adap.",
                        id == "Cellular Adaptive Immune System" ~ "Cellular Adap.",
                        id == "Adaptive Immune System" ~ "Adap.",
                        id == "Innate Immune System" ~ "Innate",
                        id == "Complement Immune System" ~ "Complement",
                        id == "Immunoglobulin Mediated Response" ~ "Ig-mediated",
                        id == "Antiviral And Interferon" ~ "Antiviral/IFN",.default = id),
         id = fct_relevel(id, c(row_shared$id))) %>% 
  ggplot(aes(x=condition, y = fct_rev(id), 
             fill = nes
             )) +
  geom_tile(color = "white",
            size = 0.5) +
  geom_text(aes(label = signific), size = 6, vjust = 0.5, color = "black") +
  theme_minimal() + 
  theme(axis.text = element_text(color = "black", size = 15),
        axis.line.x  = element_line(size = 1, color ="black"),
        # axis.line.y  = element_line(size = 1, color ="black"),
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1),
        axis.ticks = element_line(size = 1, color ="black"),
        # panel.grid = element_line(linetype = 2, color = "gray75"),
        panel.grid = element_blank(),
        legend.direction = "vertical",
        legend.position = "bottom") +
  labs(x = "",
       y = "") +
  scale_fill_gradient2(low = "#4DBBD5FF",
                        mid = "white",
                        midpoint = 0, 
                        name = "NES",
                      high = "#DC0000FF") +
  # scale_color_gradient2(low = "blue",
  #                       mid = "white",
  #                       midpoint = 0, 
  #                       name = "NES",
  #                     high = "#DC0000FF") +
  scale_y_discrete(position = "right") 


patch_immune_main = 
  (labels_disease_vac /
  labels_type /
  labels_dose/
  labels_day/
  dotplot) +
  plot_layout(heights = c(0.2, 0.2, 0.2, 0.2, 5),
              guides = 'collect',
              axes = "collect") +
  plot_annotation(title = "GSEA ImmuneGO + BTM",
                  subtitle = "Main processes, p-adjusted value <= 0.25",
                  theme = theme(legend.position = "right",
                                plot.title = element_text(hjust = 0.5, size = 20),
                                plot.subtitle = element_text(hjust = 0.5, size = 20),
                                plot.margin = margin(5, 1, 5, 1, unit = "cm"),
                                legend.direction = "vertical", 
                                legend.box = "vertical"))
  
  patch_immune_main

patch_immune_main %>%
  ggsave(filename = here("Figures", paste0(filename, "_GSEA_ImmuneGO_BTM_MainGOs_Bubbleplot_COVIDONLY.png")),
         width = 15,
         height = 15)
patch_immune_main

```



## Gene analysis
```{r fig.height=18, fig.width=25}

degs_covid_vax_immune = degs_covid_vax %>%
  filter(genes %in% immunego_btm_genes$genes) %>% 
  select(condition, genes, log2fold_change)


topdegs_all = degs_covid_vax_immune %>% 
  group_by(genes) %>% 
  summarize(n = n()) %>% 
  arrange(-n) %>% 
  slice_head(n = 20)

topdegs_vac = degs_covid_vax_immune %>%
  inner_join(ann_vaccines_covid_other_conditions %>% 
               filter(dose != "0") %>% 
               select(condition), 
             by = "condition") %>% 
  group_by(genes) %>% 
  summarize(n = n()) %>% 
  arrange(-n) %>% 
  slice_head(n = 20)

topdegs_inf = degs_covid_vax_immune %>%
  inner_join(ann_vaccines_covid_other_conditions %>% 
               filter(disease_vac == "I") %>% 
               select(condition), 
             by = "condition") %>% 
  group_by(genes) %>% 
  summarize(n = n()) %>% 
  arrange(-n) %>% 
  slice_head(n = 20)
```

Vaccination
```{r fig.height=18, fig.width=25}
hagan_vaxatlas_annotation = hagan_vaxatlas_annotation %>% 
  mutate(week = as.factor(week))

ann_vaccines_covid_other_conditions = ann_vaccines_conditions %>%
  select(condition, vaccine, day, week, dose, infection, type, disease_vac) %>% 
  mutate(day = as.factor(day),
         week = as.factor(week),
         dose = as.factor(dose),
         set = "COVID-19") %>% 
  bind_rows(hagan_vaxatlas_annotation) 
  

df = degs_covid_vax_immune %>% 
  inner_join(ann_vaccines_covid_other_conditions, by = "condition") %>% 
  filter(genes %in% topdegs_vac$genes) %>% 
  ungroup()

df = df %>% 
    mutate(day = fct_relevel(day, df %>% select(day) %>% distinct() %>% 
                               mutate(day = as.character(day),
                                      day = as.numeric(day)) %>% 
                               arrange(day) %>% 
                               distinct() %>% pull() %>% as.character()))

row_clustered = df %>% 
  select(condition, genes, log2fold_change) %>% 
  distinct() %>% 
  pivot_wider(names_from = "condition",
              values_from= "log2fold_change", values_fn = mean) %>%
  column_to_rownames("genes") %>% 
  as.matrix() %>% 
  replace(is.na(.), 0) %>% 
  dist(.) %>% 
  hclust()

result_vac <- cluster_by_day(df %>% 
                               filter(dose != "0"), day_column = "day") %>%
  mutate(day = fct_relevel(day, df$day %>% unique() %>% sort() %>% as.character()))

result_inf <- cluster_by_day(df %>% 
                               filter(dose == "0"), day_column = "day") %>%
  mutate(day = fct_relevel(day, df$day %>% unique() %>% sort() %>% as.character()))


# Ordem final das colunas
column_order <- c(result_vac$condition, result_inf$condition)

labels_set <- degs_covid_vax_immune %>% 
  full_join(ann_vaccines_covid_other_conditions, by = "condition") %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = set) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$set) + 
  theme_void() +
  labs(y = "Set") +
  theme(axis.title.y = element_text(size = 10, hjust = 1),
        legend.direction = "horizontal",
        legend.position = "bottom") 

labels_disease_vac <- degs_covid_vax_immune %>% 
  full_join(ann_vaccines_covid_other_conditions, by = "condition") %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = disease_vac) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$disease_vac) + 
  theme_void() +
  labs(y = "Infection/Vaccination") +
  theme(axis.title.y = element_text(size = 10, hjust = 1),
        legend.direction = "horizontal",
        legend.position = "bottom") 

labels_type <-degs_covid_vax_immune %>% 
  full_join(ann_vaccines_covid_other_conditions, by = "condition") %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = type) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$type) + 
  theme_void() +
  labs(y = "Type") +
  theme(axis.title.y = element_text(size = 10, hjust = 1),
        legend.direction = "horizontal",
        legend.position = "bottom") 

labels_dose <- degs_covid_vax_immune %>% 
  full_join(ann_vaccines_covid_other_conditions, by = "condition") %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = dose) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$dose) + 
  theme_void() +
  labs(y = "Dose") +
  theme(axis.title.y = element_text(size = 10, hjust = 1),
        legend.direction = "horizontal",
        legend.position = "bottom") 

labels_week = degs_covid_vax_immune %>% 
  full_join(ann_vaccines_covid_other_conditions, by = "condition") %>% 
  mutate(condition = factor(condition, column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = week) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$week) + 
  theme_void() +
  labs(y = "Week") +
  theme(axis.title.y = element_text(size = 10, hjust = 1),
        legend.direction = "horizontal",
        legend.position = "bottom")

labels_day = degs_covid_vax_immune %>% 
  full_join(ann_vaccines_covid_other_conditions, by = "condition") %>% 
  mutate(condition = factor(condition, column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = day) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$day) + 
  theme_void() +
  labs(y = "Day") +
  theme(axis.title.y = element_text(size = 10, hjust = 1),
        legend.direction = "horizontal",
        legend.position = "bottom")


#Plot
dotplot = df %>% 
  filter(genes %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, column_order),
         genes = factor(genes, levels = row_clustered$labels[row_clustered$order])) %>% 
  ggplot(aes(x=condition, y = genes, fill = log2fold_change)) +
  geom_tile() + 
  theme_minimal() + 
  theme(axis.text = element_text(color = "black", size = 12) ,
        axis.line  = element_blank(),
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1),
        axis.ticks = element_blank(),
        legend.direction = "horizontal",
        legend.position = "bottom") +
  labs(x = "",
       y = "") +
  scale_fill_gradient2(low = "#4DBBD5FF",
                        mid = "white",
                        midpoint = 0, 
                        name = "NES", 
                       limits = c(-2, 2),     # Definir os limites
    oob = scales::squish,
                      high = "#DC0000FF") +
  scale_y_discrete(position = "right")
  

patch_immune_genes_vac = 
  labels_set /
  labels_disease_vac /
  labels_type /
  labels_dose/
  labels_week/
  labels_day/
  dotplot +
  plot_layout(heights = c(0.3, 0.3, 0.3, 0.3, 0.3, 0.3, 5),
              guides = 'collect') +
  plot_annotation(title = "Top 20 genes shared",
                  subtitle = "At least one vaccine dose",
                  theme = theme(legend.position = "bottom",
                                plot.title = element_text(hjust = 0.5, size = 15),
                                plot.subtitle = element_text(hjust = 0.5, size = 15),
                                plot.margin = margin(10, 0.5, 10, 1, unit = "cm"),
                                legend.direction = "horizontal", 
                                legend.box = "horizontal"))
patch_immune_genes_vac
patch_immune_genes_vac %>% 
  ggsave(filename = here("Figures", paste0(filename, "_Top20DEGs_1dose_Bubbleplot.png")),
         width = 25,
         height = 18)

```


Infection
```{r fig.height=18, fig.width=25}
hagan_vaxatlas_annotation = hagan_vaxatlas_annotation %>% 
  mutate(week = as.factor(week))

ann_vaccines_covid_other_conditions = ann_vaccines_conditions %>%
  select(condition, vaccine, day, week, dose, infection, type, disease_vac) %>% 
  mutate(day = as.factor(day),
         week = as.factor(week),
         dose = as.factor(dose),
         set = "COVID-19") %>% 
  bind_rows(hagan_vaxatlas_annotation)
  

df = degs_covid_vax_immune %>% 
  inner_join(ann_vaccines_covid_other_conditions, by = "condition") %>% 
  filter(genes %in% topdegs_inf$genes) %>% 
  ungroup()

df = df %>% 
    mutate(day = fct_relevel(day, df %>% select(day) %>% distinct() %>% 
                               mutate(day = as.character(day),
                                      day = as.numeric(day)) %>% 
                               arrange(day) %>% 
                               distinct() %>% pull() %>% as.character()))

row_clustered = df %>% 
  select(condition, genes, log2fold_change) %>% 
  distinct() %>% 
  pivot_wider(names_from = "condition",
              values_from= "log2fold_change", values_fn = mean) %>%
  column_to_rownames("genes") %>% 
  as.matrix() %>% 
  replace(is.na(.), 0) %>% 
  dist(.) %>% 
  hclust()

result_vac <- cluster_by_day(df %>% 
                               filter(dose != "0"), day_column = "day") %>%
  mutate(day = fct_relevel(day, df$day %>% unique() %>% sort() %>% as.character()))

result_inf <- cluster_by_day(df %>% 
                               filter(dose == "0"), day_column = "day") %>%
  mutate(day = fct_relevel(day, df$day %>% unique() %>% sort() %>% as.character()))


# Ordem final das colunas
column_order <- c(result_vac$condition, result_inf$condition)


labels_set <- degs_covid_vax_immune %>% 
  full_join(ann_vaccines_covid_other_conditions, by = "condition") %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = set) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$set) + 
  theme_void() +
  labs(y = "Set") +
  theme(axis.title.y = element_text(size = 12, hjust = 1),
        legend.direction = "horizontal",
        legend.position = "bottom") 

labels_disease_vac <- degs_covid_vax_immune %>% 
  full_join(ann_vaccines_covid_other_conditions, by = "condition") %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = disease_vac) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$disease_vac) + 
  theme_void() +
  labs(y = "Infection/Vaccination") +
  theme(axis.title.y = element_text(size = 12, hjust = 1),
        legend.direction = "horizontal",
        legend.position = "bottom") 

labels_type <-degs_covid_vax_immune %>% 
  full_join(ann_vaccines_covid_other_conditions, by = "condition") %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = type) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$type) + 
  theme_void() +
  labs(y = "Type") +
  theme(axis.title.y = element_text(size = 12, hjust = 1),
        legend.direction = "horizontal",
        legend.position = "bottom") 

labels_dose <- degs_covid_vax_immune %>% 
  full_join(ann_vaccines_covid_other_conditions, by = "condition") %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = dose) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$dose) + 
  theme_void() +
  labs(y = "Dose") +
  theme(axis.title.y = element_text(size = 12, hjust = 1),
        legend.direction = "horizontal",
        legend.position = "bottom") 

labels_week = degs_covid_vax_immune %>% 
  full_join(ann_vaccines_covid_other_conditions, by = "condition") %>% 
  mutate(condition = factor(condition, column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = week) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$week) + 
  theme_void() +
  labs(y = "Week") +
  theme(axis.title.y = element_text(size = 12, hjust = 1),
        legend.direction = "horizontal",
        legend.position = "bottom")

labels_day = degs_covid_vax_immune %>% 
  full_join(ann_vaccines_covid_other_conditions, by = "condition") %>% 
  mutate(condition = factor(condition, column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = day) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$day) + 
  theme_void() +
  labs(y = "Day") +
  theme(axis.title.y = element_text(size = 12, hjust = 1),
        legend.direction = "horizontal",
        legend.position = "bottom")


#Plot
dotplot_vac = df %>% 
  filter(genes %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, column_order),
         genes = factor(genes, levels = row_clustered$labels[row_clustered$order])) %>% 
  ggplot(aes(x=condition, y = genes, fill = log2fold_change)) +
  geom_tile() + 
  theme_minimal() + 
  theme(axis.text = element_text(color = "black", size = 12) ,
        axis.line  = element_blank(),
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1),
        axis.ticks = element_blank(),
        legend.direction = "horizontal",
        legend.position = "bottom") +
  labs(x = "",
       y = "") +
  scale_fill_gradient2(low = "#4DBBD5FF",
                        mid = "white",
                        midpoint = 0, 
                        name = "L2FC", 
                       limits = c(-2, 2),     # Definir os limites
    oob = scales::squish,
                      high = "#DC0000FF") +
  scale_y_discrete(position = "right")
  

patch_immune_genes_infec = 
  labels_set /
  labels_disease_vac /
  labels_type /
  labels_dose/
  labels_week/
  labels_day/
  dotplot_vac+
  plot_layout(heights = c(0.2, 0.2, 0.2, 0.2, 0.2, 0.2, 5),
              guides = 'collect',
              axes = "collect") +
  plot_annotation(title = "Top 20 genes shared",
                  subtitle = "Among infection",
                  theme = theme(legend.position = "bottom",
                                plot.title = element_text(hjust = 0.5, size = 15),
                                plot.subtitle = element_text(hjust = 0.5, size = 15),
                                plot.margin = margin(10, 0.5, 10, 1, unit = "cm"),
                                legend.direction = "horizontal", 
                                legend.box = "horizontal"))
patch_immune_genes_infec
patch_immune_genes_infec %>% 
  ggsave(filename = here("Figures", paste0(filename, "_Top20DEGs_infec_Bubbleplot.png")),
         width = 25,
         height = 18)

```


#### Antibodies
```{r fig.height=10, fig.width=25}
selected_degs = degs_covid_vax_immune %>% 
  filter(genes %in% c(immunego_btm_genes %>% 
           filter(process == "plasma cells, immunoglobulins") %>% 
           select(genes) %>% 
           pull()))

hagan_vaxatlas_annotation = hagan_vaxatlas_annotation %>% 
  mutate(week = as.factor(week))

ann_vaccines_covid_other_conditions = ann_vaccines_conditions %>%
  select(condition, vaccine, day, week, dose, infection, type, disease_vac) %>% 
  mutate(day = as.factor(day),
         week = as.factor(week),
         dose = as.factor(dose),
         set = "COVID-19") %>% 
  bind_rows(hagan_vaxatlas_annotation)
  

df = degs_covid_vax %>% 
  inner_join(ann_vaccines_covid_other_conditions, by = "condition") %>% 
  filter(genes %in% selected_degs$genes) %>% 
  ungroup()

df = df %>% 
    mutate(day = fct_relevel(day, df %>% select(day) %>% distinct() %>% 
                               mutate(day = as.character(day),
                                      day = as.numeric(day)) %>% 
                               arrange(day) %>% 
                               distinct() %>% pull() %>% as.character()))

row_clustered = df %>% 
  select(condition, genes, log2fold_change) %>% 
  distinct() %>% 
  pivot_wider(names_from = "condition",
              values_from= "log2fold_change", values_fn = mean) %>%
  column_to_rownames("genes") %>% 
  as.matrix() %>% 
  replace(is.na(.), 0) %>% 
  dist(.) %>% 
  hclust()

result_vac <- cluster_by_day(df %>% 
                               filter(dose != "0"), day_column = "day") %>%
  mutate(day = fct_relevel(day, df$day %>% unique() %>% sort() %>% as.character()))

result_inf <- cluster_by_day(df %>% 
                               filter(dose == "0"), day_column = "day") %>%
  mutate(day = fct_relevel(day, df$day %>% unique() %>% sort() %>% as.character()))


# Ordem final das colunas
column_order <- c(result_vac$condition, result_inf$condition)


labels_set <- degs_covid_vax %>% 
  full_join(ann_vaccines_covid_other_conditions, by = "condition") %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = set) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$set) + 
  theme_void() +
  labs(y = "Set") +
  theme(axis.title.y = element_text(size = 12, hjust = 1),
        legend.direction = "horizontal",
        legend.position = "bottom") 

labels_disease_vac <- degs_covid_vax %>% 
  full_join(ann_vaccines_covid_other_conditions, by = "condition") %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = disease_vac) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$disease_vac) + 
  theme_void() +
  labs(y = "Infection/Vaccination") +
  theme(axis.title.y = element_text(size = 12, hjust = 1),
        legend.direction = "horizontal",
        legend.position = "bottom") 

labels_type <-degs_covid_vax %>% 
  full_join(ann_vaccines_covid_other_conditions, by = "condition") %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = type) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$type) + 
  theme_void() +
  labs(y = "Type") +
  theme(axis.title.y = element_text(size = 12, hjust = 1),
        legend.direction = "horizontal",
        legend.position = "bottom") 

labels_dose <- degs_covid_vax %>% 
  full_join(ann_vaccines_covid_other_conditions, by = "condition") %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = dose) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$dose) + 
  theme_void() +
  labs(y = "Dose") +
  theme(axis.title.y = element_text(size = 12, hjust = 1),
        legend.direction = "horizontal",
        legend.position = "bottom") 

labels_week = degs_covid_vax %>% 
  full_join(ann_vaccines_covid_other_conditions, by = "condition") %>% 
  mutate(condition = factor(condition, column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = week) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$week) + 
  theme_void() +
  labs(y = "Week") +
  theme(axis.title.y = element_text(size = 12, hjust = 1),
        legend.direction = "horizontal",
        legend.position = "bottom")

labels_day = degs_covid_vax %>% 
  full_join(ann_vaccines_covid_other_conditions, by = "condition") %>% 
  mutate(condition = factor(condition, column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = day) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$day) + 
  theme_void() +
  labs(y = "Day") +
  theme(axis.title.y = element_text(size = 12, hjust = 1),
        legend.direction = "horizontal",
        legend.position = "bottom")


#Plot
dotplot_vac = df %>% 
  filter(genes %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, column_order),
         genes = factor(genes, levels = row_clustered$labels[row_clustered$order])) %>% 
  ggplot(aes(x=condition, y = genes, fill = log2fold_change)) +
  geom_tile() + 
  theme_minimal() + 
  theme(axis.text = element_text(color = "black", size = 12) ,
        axis.line  = element_blank(),
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1),
        axis.ticks = element_blank(),
        legend.direction = "horizontal",
        legend.position = "bottom") +
  labs(x = "",
       y = "") +
  scale_fill_gradient2(low = "#4DBBD5FF",
                        mid = "white",
                        midpoint = 0, 
                        name = "L2FC", 
                       limits = c(-2, 2),     # Definir os limites
    oob = scales::squish,
                      high = "#DC0000FF") +
  scale_y_discrete(position = "right")
  

patch_immune_genes_infec = 
  labels_set /
  labels_disease_vac /
  labels_type /
  labels_dose/
  labels_week/
  labels_day/
  dotplot_vac+
  plot_layout(heights = c(0.3, 0.3, 0.3, 0.3, 0.3, 0.3, 5),
              guides = 'collect',
              axes = "collect") +
  plot_annotation(title = "Plasma cells, immunoglobulins (M156.1)",
                  theme = theme(legend.position = "bottom",
                                plot.title = element_text(hjust = 0.5, size = 15),
                                plot.subtitle = element_text(hjust = 0.5, size = 15),
                                plot.margin = margin(1, 0.5, 1, 1, unit = "cm"),
                                legend.direction = "horizontal", 
                                legend.box = "horizontal"))
patch_immune_genes_infec
patch_immune_genes_infec %>% 
  ggsave(filename = here("Figures", paste0(filename, "_Infec_Immunoglobulins.png")),
         width = 25,
         height = 10)
```

#### Antiviral
```{r fig.height=10, fig.width=25}
selected_degs = ImmuneGO_BTM_grouped_genes %>% 
  filter(str_detect(process, "antiviral IFN signature"))%>% 
  select(genes)

hagan_vaxatlas_annotation = hagan_vaxatlas_annotation %>% 
  mutate(week = as.factor(week))

ann_vaccines_covid_other_conditions = ann_vaccines_conditions %>%
  select(condition, vaccine, day, week, dose, infection, type, disease_vac) %>% 
  mutate(day = as.factor(day),
         week = as.factor(week),
         dose = as.factor(dose),
         set = "COVID-19") %>% 
  bind_rows(hagan_vaxatlas_annotation)
  

df = degs_covid_vax %>% 
  inner_join(ann_vaccines_covid_other_conditions, by = "condition") %>% 
  filter(genes %in% selected_degs$genes) %>% 
  ungroup()

df = df %>% 
    mutate(day = fct_relevel(day, df %>% select(day) %>% distinct() %>% 
                               mutate(day = as.character(day),
                                      day = as.numeric(day)) %>% 
                               arrange(day) %>% 
                               distinct() %>% pull() %>% as.character()))

row_clustered = df %>% 
  select(condition, genes, log2fold_change) %>% 
  distinct() %>% 
  pivot_wider(names_from = "condition",
              values_from= "log2fold_change", values_fn = mean) %>%
  column_to_rownames("genes") %>% 
  as.matrix() %>% 
  replace(is.na(.), 0) %>% 
  dist(.) %>% 
  hclust()

result_vac <- cluster_by_day(df %>% 
                               filter(dose != "0"), day_column = "day") %>%
  mutate(day = fct_relevel(day, df$day %>% unique() %>% sort() %>% as.character()))

result_inf <- cluster_by_day(df %>% 
                               filter(dose == "0"), day_column = "day") %>%
  mutate(day = fct_relevel(day, df$day %>% unique() %>% sort() %>% as.character()))


# Ordem final das colunas
column_order <- c(result_vac$condition, result_inf$condition)


labels_set <- degs_covid_vax_immune %>% 
  full_join(ann_vaccines_covid_other_conditions, by = "condition") %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = set) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$set) + 
  theme_void() +
  labs(y = "Set") +
  theme(axis.title.y = element_text(size = 12, hjust = 1),
        legend.direction = "horizontal",
        legend.position = "bottom") 

labels_disease_vac <- degs_covid_vax_immune %>% 
  full_join(ann_vaccines_covid_other_conditions, by = "condition") %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = disease_vac) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$disease_vac) + 
  theme_void() +
  labs(y = "Infection/Vaccination") +
  theme(axis.title.y = element_text(size = 12, hjust = 1),
        legend.direction = "horizontal",
        legend.position = "bottom") 

labels_type <-degs_covid_vax_immune %>% 
  full_join(ann_vaccines_covid_other_conditions, by = "condition") %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = type) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$type) + 
  theme_void() +
  labs(y = "Type") +
  theme(axis.title.y = element_text(size = 12, hjust = 1),
        legend.direction = "horizontal",
        legend.position = "bottom") 

labels_dose <- degs_covid_vax_immune %>% 
  full_join(ann_vaccines_covid_other_conditions, by = "condition") %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = dose) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$dose) + 
  theme_void() +
  labs(y = "Dose") +
  theme(axis.title.y = element_text(size = 12, hjust = 1),
        legend.direction = "horizontal",
        legend.position = "bottom") 

labels_week = degs_covid_vax_immune %>% 
  full_join(ann_vaccines_covid_other_conditions, by = "condition") %>% 
  mutate(condition = factor(condition, column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = week) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$week) + 
  theme_void() +
  labs(y = "Week") +
  theme(axis.title.y = element_text(size = 12, hjust = 1),
        legend.direction = "horizontal",
        legend.position = "bottom")

labels_day = degs_covid_vax_immune %>% 
  full_join(ann_vaccines_covid_other_conditions, by = "condition") %>% 
  mutate(condition = factor(condition, column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = day) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$day) + 
  theme_void() +
  labs(y = "Day") +
  theme(axis.title.y = element_text(size = 12, hjust = 1),
        legend.direction = "horizontal",
        legend.position = "bottom")


#Plot
dotplot_vac = df %>% 
  filter(genes %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, column_order),
         genes = factor(genes, levels = row_clustered$labels[row_clustered$order])) %>% 
  ggplot(aes(x=condition, y = genes, fill = log2fold_change)) +
  geom_tile() + 
  theme_minimal() + 
  theme(axis.text = element_text(color = "black", size = 12) ,
        axis.line  = element_blank(),
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1),
        axis.ticks = element_blank(),
        legend.direction = "horizontal",
        legend.position = "bottom") +
  labs(x = "",
       y = "") +
  scale_fill_gradient2(low = "#4DBBD5FF",
                        mid = "white",
                        midpoint = 0, 
                        name = "L2FC", 
                       limits = c(-2, 2),     # Definir os limites
    oob = scales::squish,
                      high = "#DC0000FF") +
  scale_y_discrete(position = "right")
  

patch_immune_genes_infec = 
  labels_set /
  labels_disease_vac /
  labels_type /
  labels_dose/
  labels_week/
  labels_day/
  dotplot_vac+
  plot_layout(heights = c(0.3, 0.3, 0.3, 0.3, 0.3, 0.3, 5),
              guides = 'collect',
              axes = "collect") +
  plot_annotation(title = "Antiviral IFN siganture (M75)",
                  theme = theme(legend.position = "bottom",
                                plot.title = element_text(hjust = 0.5, size = 15),
                                plot.subtitle = element_text(hjust = 0.5, size = 15),
                                plot.margin = margin(1, 0.5, 1, 1, unit = "cm"),
                                legend.direction = "horizontal", 
                                legend.box = "horizontal"))
patch_immune_genes_infec
patch_immune_genes_infec %>% 
  ggsave(filename = here("Figures", paste0(filename, "_Infec_antiviral.png")),
         width = 25,
         height = 18)
```







## Principal components analysis

```{r}
library(tidyverse)
library(ggrepel) #load before factoextra
library(factoextra)
library(caret)
library(stats)
library(ggfortify)
library(ggplot2)
library(corrplot)
library(cowplot)
library(ggcorrplot)
library(ComplexHeatmap)
library(circlize)
library(ggExtra)
library(FactoMineR)
library(here)
library(PCAtools)
library(ggplotify)

```

**Required files**

```{r}
# Genes in long table format

degs_covid_vax = readRDS(file = here("VaxGO gene sets", "degs_covid_vaxhagan.rds"))

# Gene set annotation.Used only to merge and filter genes.
ImmuneGO_Annotated_Genes = readRDS(here("VaxGO gene sets", "ImmuneGO_Annotated_Genes_2024-03-26.rds"))
ImmuneGO_BTM_grouped_genes <- read_csv("VaxGO gene sets/ImmuneGO_BTM_grouped_genes.csv")

btm_annotation_genes_immune = ImmuneGO_BTM_grouped_genes %>% 
  select(genes) %>% distinct()

# Sample or condition annotation
ann_vaccines = ann_vaccines_covid_other_conditions

data_annotation = ann_vaccines %>%
  mutate(day = as.factor(day),
         dose = as.factor(dose)) 
```

**Filtering data**

```{r}
# Gene set data.
# ImmuneGO_Annotated_Genes_all = ImmuneGO_Annotated_Genes
# ImmuneGO_Annotated_Genes = ImmuneGO_Annotated_Genes_all %>% 
#   filter(!process %in% c("TCR REPERTOIRE", "BCR REPERTOIRE"))

ImmuneGO_Genes = degs_covid_vax %>%
  inner_join(btm_annotation_genes_immune, by = "genes") %>% 
  # inner_join(ImmuneGO_Annotated_Genes %>% 
  #              select(genes, process) %>% 
  #              distinct() %>% 
  #              select(genes),
  #              by = "genes") %>% 
  distinct() %>% 
  ungroup()

#INPUT
data_genes = ImmuneGO_Genes 
filename = "13Vax_Covid_BTM_ImmuneGO"

#Convert long to wide table format.
#Dataframe with sample annotation

ann_vaccines_pca_matrix = data_genes %>% 
  dplyr::summarise(n = dplyr::n(), .by = c(genes, condition)) |>
  dplyr::filter(n > 1L) %>% 
  select(genes) %>% 
  inner_join(data_genes, by = "genes") %>% 
  select(condition, genes, log2fold_change) %>% 
  distinct() %>% 
  pivot_wider(names_from = "condition", 
              values_from = "log2fold_change", values_fn = mean) %>% 
  replace(is.na(.), 0) %>%
  column_to_rownames("genes") %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column("condition") %>% 
  inner_join(data_annotation %>% 
               select(condition, disease_vac, vaccine, day, week, dose, type, set), by = "condition") 

# Convert dataframe to matrix without annotation columns
ann_vaccines_pca_matrix_ready = ann_vaccines_pca_matrix %>% 
  column_to_rownames("condition") %>% 
  select(!disease_vac:set)


```

### With infection
```{r fig.height=5, fig.width=10, cache = TRUE}
##### PCA using PCAtools
# BiocManager::install('PCAtools')

library(PCAtools)

#Matrix (genes, sample)
p_vacinfec_matrix = ann_vaccines_pca_matrix_ready %>% 
  t() %>% 
  as.data.frame()
#Annotation 
p_vacinfec_metadata = data_annotation %>% 
  inner_join(p_vacinfec_matrix %>% t() %>% 
               as.data.frame() %>% 
               rownames_to_column("condition") %>% select("condition"), 
             by = "condition") %>% 
  column_to_rownames("condition") %>% 
  select(disease_vac, type, vaccine, dose, day, week, set)

p_vacinfec_matrix = p_vacinfec_matrix%>% 
  select(rownames(p_vacinfec_metadata)) %>% 
  as.matrix()


#PCA
p_vacinfec <- pca(mat = p_vacinfec_matrix, 
         metadata = p_vacinfec_metadata, 
         removeVar = 0.1)

options(ggrepel.max.overlaps = Inf)
#Plot
conflicted::conflicts_prefer(PCAtools::biplot)

#Screeplot

p_vacinfec_horn <- parallelPCA(mat)
p_vacinfec_elbow <- findElbowPoint(p_vacinfec$variance)
conflicted::conflicts_prefer(PCAtools::screeplot)

p_vacinfec_screeplot = screeplot(p_vacinfec,
          components = getComponents(p_vacinfec, 1:20),
          vline = c(p_vacinfec_horn$n, p_vacinfec_elbow)) +
  geom_label(aes(x = p_vacinfec_horn$n + 1, y = 50,
                 label = 'Horn\'s', vjust = -1, size = 8)) +
  geom_label(aes(x = p_vacinfec_elbow + 1, y = 50,
                 label = 'Elbow method', vjust = -1, size = 8)) 


p_vacinfec_screeplot %>% 
  ggsave(filename = here("Figures", paste0(filename, "_PCA_WithInfection_Screeplot.png")), 
         width = 10,
         height = 5)


p_vacinfec_screeplot
```


```{r fig.height=10, fig.width=10, message=FALSE, warning=FALSE, cache=FALSE}
#Plot all paired components 
p_vacinfec_pairsplot_type = pairsplot(p_vacinfec,
          components = getComponents(p_vacinfec, c(1:6)),
          triangle = T, 
          trianglelabSize = 12,
          hline = 0, vline = 0,
          pointSize = 2,
          gridlines.major = FALSE, 
          gridlines.minor = FALSE,
          colby = 'type',
          colkey = ann_vaxsig_colors$type,
          title = 'Pairs plot, by vaccine type', returnPlot = T,
          plotaxes = FALSE)  +
    theme(axis.text = element_text(size = 10, color = "black"),
      panel.grid.minor = element_blank(),
      panel.grid.major = element_line(size = 1, 
                                    linetype = 1,
                                    color = "gray90"),
        axis.line = element_line(size = 0, colour = "black", linetype=1),
        # axis.line = element_line(size = 1, colour = "black", linetype=1),
        text = element_text(color = "black", size = 10),
        panel.border = element_rect(fill = NA, color = "gray90", size = 2)) 

p_vacinfec_pairsplot_type %>% 
   ggsave(filename = here("Figures", paste0(filename, "_PCA_WithInfection_Pairsplot_TYPE.png")), 
         width = 10,
         height = 10)

p_vacinfec_pairsplot_type

p_vacinfec_pairsplot_disease_vac = pairsplot(p_vacinfec,
          components = getComponents(p_vacinfec, c(1:6)),
          triangle = T, 
          trianglelabSize = 12,
          hline = 0, vline = 0,
          pointSize = 2,
          gridlines.major = FALSE, 
          gridlines.minor = FALSE,
          colby = 'day',
          colkey = ann_vaxsig_colors$day,
          title = 'Pairs plot, by day', returnPlot = T,
          plotaxes = FALSE) +
  theme(axis.text = element_text(size = 10, color = "black"),
      panel.grid.minor = element_blank(),
      panel.grid.major = element_line(size = 1, 
                                    linetype = 1,
                                    color = "gray90"),
        axis.line = element_line(size = 0, colour = "black", linetype=1),
        # axis.line = element_line(size = 1, colour = "black", linetype=1),
        text = element_text(color = "black", size = 10),
        panel.border = element_rect(fill = NA, color = "gray90", size = 2)) 
p_vacinfec_pairsplot_disease_vac

p_vacinfec_pairsplot_disease_vac %>% 
ggsave(filename = here("Figures", paste0(filename, "_PCA_WithInfection_Pairsplot_Day.png")), 
         width = 10,
         height = 10)

```


```{r fig.height=5, fig.width=10, cache = TRUE}
#Loadings plot
p_vacinfec_loadingsplot = plotloadings(p_vacinfec, 
                                       labSize = 3,  
                                       shapeSizeRange = 5,
                                       titleLabSize = 22,
                            components = 1:10,
                            title = "Loadings plot, Vaccination and infection",
                            col = c(low = '#4DBBD5FF', mid = "white", high = '#DC0000FF') )

#Meta variables
p_vacinfec_metavars = eigencorplot(p_vacinfec,
             metavars = colnames(p_vacinfec_metadata), 
             main = "Meta variables, Vaccination and infection",
             col = c(low = '#4DBBD5FF', mid = "white", high = '#DC0000FF'))

p_vacinfec_metavars
```


```{r fig.height=15, fig.width=10, cache = TRUE}
library(ggplotify)
library(cowplot)


vacinfec_metavars_loadings <- plot_grid(p_vacinfec_loadingsplot,
      as.grob(p_vacinfec_metavars),
      ncol = 1,
      labels = c('D', 'E'),
      label_fontface = 'bold',
      # label_size = 22,
      align = 'v',
      rel_widths = c(1, 1.5),
      rel_heights = c(1, 1))

vacinfec_metavars_loadings %>% 
  ggsave(filename = here("Figures", paste0(filename, "_PCA_WithInfection_Metavars_Loadings.png")), 
         width = 10,
         height = 15)


vacinfec_metavars_loadings

```


```{r fig.height=10, fig.width=10}
vac_p_biplot_encircle_PC1_PC2 = biplot(p_vacinfec,  
                               x = 'PC1', 
                               y = 'PC2',
                               ntopLoadings = 10,
                               lengthLoadingsArrowsFactor = 2,
                               alphaLoadingsArrow = 0.5,
                               legendPosition = 'right',
                      labSize = 3,
                      showLoadings = F, 
                      lab = NULL,
                      pointSize = 5,
                      encircle = T,
                      encircleFill = T,
                      encircleAlpha = 0.1,
                      encircleLineSize = 5, 
                      colby = 'type',
                      hline = 0, 
                      vline = 0,
                      max.overlaps = Inf,
                      colkey = ann_vaxsig_colors$type)

vac_p_biplot_encircle_PC1_PC2

vac_p_biplot_encircle_PC1_PC2_day = biplot(p_vacinfec,  
                               x = 'PC2', 
                               y = 'PC1',
                               ntopLoadings = 10,
                               lengthLoadingsArrowsFactor = 2,
                               alphaLoadingsArrow = 0.5,
                               legendPosition = 'right',
                      labSize = 3,
                      showLoadings = F, 
                      lab = NULL,
                      pointSize = 5,
                      encircle = F,
                      encircleFill = T,
                      encircleAlpha = 0.1,
                      encircleLineSize = 5, 
                      colby = 'day',
                      hline = 0, 
                      vline = 0,
                      max.overlaps = Inf,
                      colkey = ann_vaxsig_colors$day)

vac_p_biplot_encircle_PC1_PC2_day

vac_p_biplot_encircle_PC1_PC3 = biplot(p_vacinfec,  
                               x = 'PC1', 
                               y = 'PC3',
                               ntopLoadings = 10,
                               lengthLoadingsArrowsFactor = 2,
                               alphaLoadingsArrow = 0.5,
                               legendPosition = 'right',
                      labSize = 3,
                      showLoadings = F, 
                      lab = NULL,
                      pointSize = 5,
                      encircle = T,
                      encircleFill = T,
                      encircleAlpha = 0.1,
                      encircleLineSize = 5, 
                      colby = 'type',
                      hline = 0, 
                      vline = 0,
                      max.overlaps = Inf,
                      colkey = ann_vaxsig_colors$type)

vac_p_biplot_encircle_PC1_PC3



```




### Only vaccines
```{r fig.height=10, fig.width=10, message=TRUE, warning=TRUE, cache=FALSE}
#Matrix (genes, sample)
p_vac_matrix = ann_vaccines_pca_matrix_ready %>% 
  t() %>% 
  as.data.frame()
#Annotation 
p_vac_metadata = data_annotation %>% 
  inner_join(p_vac_matrix %>% t() %>% 
               as.data.frame() %>% 
               rownames_to_column("condition") %>% select("condition"), 
             by = "condition") %>% 
  column_to_rownames("condition") %>% 
  filter(dose != "0") %>% 
  select(disease_vac, type, vaccine, dose, day, week)

p_vac_matrix = p_vac_matrix%>% 
  select(rownames(p_vac_metadata)) %>% 
  as.matrix()
#PCA
vac_p <- pca(mat = p_vac_matrix, 
         metadata = p_vac_metadata, 
         removeVar = 0.1)

options(ggrepel.max.overlaps = Inf)
#Plot


# install.packages("ggalt")
library(ggalt)
conflicted::conflicts_prefer(PCAtools::biplot)
vac_p_biplot_encircle = biplot(vac_p,
                      labSize = 3,
                      showLoadings = T, 
                      lab = NULL,
                      pointSize = 5,
                      encircle = TRUE, 
                      encircleFill = FALSE,
                      encircleAlpha = 1, 
                      encircleLineSize = 4,
                      colby = 'type',
                      hline = 0, 
                      vline = 0,
                      max.overlaps = Inf,
                      colkey = ann_vaxsig_colors$type)

vac_p_biplot = biplot(vac_p,
                      labSize = 3,
                      showLoadings = T, 
                      lab = paste0(vac_p$metadata$vaccine," (D" ,vac_p$metadata$day, ")"),
                      pointSize = 5,
                      colby = 'type',
                      hline = 0, 
                      vline = 0,
                      max.overlaps = Inf,
                      colkey = ann_vaxsig_colors$type)

vac_p_biplot
```


```{r fig.height=5, fig.width=10, message=TRUE, warning=TRUE, cache=FALSE}
#Screeplot
#Determine optimal number of PCs
vac_horn <- parallelPCA(mat)
vac_elbow <- findElbowPoint(vac_p$variance)
conflicted::conflicts_prefer(PCAtools::screeplot)
vac_screeplot = screeplot(vac_p,
          components = getComponents(vac_p, 1:20),
          vline = c(vac_horn$n, vac_elbow)) +
  geom_label(aes(x = vac_horn$n + 1, y = 50,
                 label = 'Horn\'s', vjust = -1, size = 8)) +
  geom_label(aes(x = vac_elbow + 1, y = 50,
                 label = 'Elbow method', vjust = -1, size = 8))

vac_screeplot %>% 
    ggsave(filename = here("Figures", paste0(filename, "_PCA_VaccinesOnly_Screeplot.png")), 
         width = 10,
         height = 5)

vac_screeplot
```


```{r fig.height=20, fig.width=20, message=FALSE, warning=FALSE, cache=FALSE}
#Plot all paired components 
vac_pairsplot_type = pairsplot(vac_p,
          components = getComponents(vac_p, c(1:10)),
          triangle = T, 
          trianglelabSize = 12,
          hline = 0, vline = 0,
          pointSize = 2,
          gridlines.major = FALSE, 
          gridlines.minor = FALSE,
          colby = 'type',
          colkey = ann_vaxsig_colors$type,
          title = 'Pairs plot, by type', returnPlot = T,
          plotaxes = FALSE) 

vac_pairsplot_type %>% 
ggsave(filename = here("Figures", paste0(filename, "_PCA_VaccinesOnly_Pairsplot_TYPE.png")), 
         width = 20,
         height = 20)

vac_pairsplot_disease_vac = pairsplot(vac_p,
          components = getComponents(vac_p, c(1:10)),
          triangle = T, 
          trianglelabSize = 12,
          hline = 0, vline = 0,
          pointSize = 2,
          gridlines.major = FALSE, 
          gridlines.minor = FALSE,
          colby = 'disease_vac',
          colkey = ann_vaxsig_colors$disease_vac,
          title = 'Pairs plot, by infection or vaccination', returnPlot = T,
          plotaxes = FALSE) 


vac_pairsplot_disease_vac %>% 
ggsave(filename = here("Figures", paste0(filename, "_PCA_VaccinesOnly_Pairsplot_DISEASE_VAC.png")), 
         width = 20,
         height = 20)

vac_pairsplot_type
vac_pairsplot_disease_vac

```


```{r fig.height=15, fig.width=10, cache = TRUE}
#Loadings plot
vac_loadingsplot = plotloadings(vac_p, labSize = 3,  shapeSizeRange = 5,
                            components = 1:10, 
                            title = "Loadings plot, Vaccination only",
             col = c(low = '#4DBBD5FF', mid = "white", high = '#DC0000FF') )

#Meta variables
vac_metavars = eigencorplot(vac_p,
             metavars = colnames(p_vac_metadata), 
             main = "Meta variables, Vaccination only",
             col = c(low = '#4DBBD5FF', mid = "white", high = '#DC0000FF')) 

vac_metavars_loadings <- plot_grid(vac_loadingsplot,
      as.grob(vac_metavars),
      ncol = 1,
      labels = c('D', 'E'),
      label_fontface = 'bold',
      # label_size = 22,
      align = 'v',
      rel_widths = c(1, 1.5),
      rel_heights = c(1, 1))

vac_metavars_loadings %>% 
  ggsave(filename = here("Figures", paste0(filename, "_PCA_VaccinesOnly_Metavars_Loadings.png")), 
         width = 10,
         height = 15)

vac_metavars_loadings
```



```{r fig.height=10, fig.width=15}


p_vac_matrix_d1 = degs_covid_vax %>% 
  filter(genes %in% c(btm_genes %>% select(genes) %>% distinct() %>% pull(genes))) %>% 
  filter(condition %in% c(data_annotation %>% 
           filter(week %in% c("1")) %>% 
           pull(condition))) 

p_vac_matrix_d1 = p_vac_matrix_d1 %>% 
  pivot_wider(names_from = "genes",
              values_from = "log2fold_change", values_fn = mean, values_fill = 0) %>% 
  column_to_rownames("condition") %>% 
  t() 

corr <- round(cor(p_vac_matrix_d1, method = "spearman"), 1)

corr_plot_vac = ggcorrplot(corr, 
           method = "circle")
corr_plot_vac


dm<-dist(p_vac_matrix %>% t()) 
hc<-hclust(dm, method="complete")

par(mar=c(10,10,10,20)) 
plot(hc) 
```








## Cell proportion

<https://www.datacamp.com/tutorial/pca-analysis-r>
<https://rpkgs.datanovia.com/factoextra/index.html>
<http://www.sthda.com/english/articles/31-principal-component-methods-in-r-practical-guide/114-mca-multiple-correspondence-analysis-in-r-essentials/>
<https://statisticsglobe.com/pca-before-k-means-clustering-r>\>




### COVID-19 immunedeconv/XCell

```{r fig.height=6, fig.width=12}
all_matrices_counts <- readRDS(here("DGE analysis", "all_matrices_counts.rds"))

all_samples_l2fc_matrix = all_samples_l2fc %>% 
  select(genes, sample, log2fc) %>% 
  pivot_wider(names_from = "sample",
              values_from = "log2fc") %>% 
  column_to_rownames("genes") %>% 
  as.matrix()

library(immunedeconv)
library(tidyverse)
tidymodels_prefer()
```


```{r fig.height=6, fig.width=12}
#XCell (ssGSEA) ----
xcell_results = immunedeconv::deconvolute(all_samples_l2fc_matrix, "xcell") 

xcell_results %>% 
  write_csv(file = here("Cell populations", "all_covid_degs_xcell_results.csv"))

xcell_results_annotated = xcell_results %>% 
  pivot_longer(cols = -cell_type,
               names_to = "sample",
               values_to = "value") %>% 
  inner_join(ann_vaccines_samples, by = "sample") %>% 
   distinct() %>% 
  filter(condition != "BNT-MO (V1, D0)",
         condition != "BNT-MO (V1, D6)") %>% 
  mutate(disease_vac = if_else(disease_vac == "Healthy", "H", disease_vac)) %>% 
  mutate(disease_vac = fct_relevel(disease_vac, "H", "I", "V"),
           type = fct_relevel(type, "I", "RNA", "VV", "IN", "SU")) %>% 
  arrange(disease_vac,type, condition) %>% 
  inner_join(ann_vaccines_conditions %>% select(condition, samples), by = "condition")



#boxplot
xcell_results_annotated_2_plot =  xcell_results_annotated %>% 
  mutate(condition = fct_relevel(condition, 
                                 c(ann_vaccines_conditions_ordered$condition %>% as.character() %>% unique()))) %>% 
  filter(cell_type == "B cell") %>% 
  ggplot() +
  aes(x = condition, y = value, fill = if_else(day == 0, "Day 0", type)) +  
  geom_hline(yintercept = 0, linetype = 2) +
  geom_jitter(aes(color = if_else(day == 0, "Day 0", type)),
              alpha = 0.3, size = 0.1) +
  geom_boxplot(outliers = F) +
  theme_few() +
  scale_fill_manual(values = c("Day 0" = "white", 
                               ann_vaxsig_colors$type)) +
    scale_color_manual(values = c("Day 0" = "gray80", 
                               ann_vaxsig_colors$type)) +
  theme(axis.text.x = element_text(angle = 45,
                                   hjust = 1,
                                   vjust = 1),
        plot.title = element_text(hjust = 0.5, face = "bold"),
        plot.margin = margin(0, 0, 0, 2, "cm"),
        panel.grid.major = element_line(size = 0.05),
        panel.grid.minor = element_line(size = 0.05)) +
  labs(y = "Score",
       title = "B cell",
       x="",
       subtitle = "By week",
       fill = "Type") +
  guides(color = "none") +
  facet_wrap(~disease_vac, scales = "free_x")

xcell_results_annotated_2_plot

```


```{r}
#Function to create boxplots
create_boxplots <- function(result_df) {
  
  unique_processes <- unique(result_df$cell_type)
  
  purrr::map(unique_processes, function(process_name) {
    cellmarker_boxplot <- result_df %>% 
      filter(cell_type == process_name) %>% 
  ggplot() +
  aes(x = condition, y = value, fill = if_else(day == 0, "Day 0", type)) +  
  geom_hline(yintercept = 0, linetype = 2) +
  geom_jitter(aes(color = if_else(day == 0, "Day 0", type)),
              alpha = 0.3, size = 0.1) +
  geom_boxplot(outliers = F) +
  theme_few() +
  scale_fill_manual(values = c("Day 0" = "white", 
                               ann_vaxsig_colors$type)) +
    scale_color_manual(values = c("Day 0" = "gray80", 
                               ann_vaxsig_colors$type)) +
  theme(axis.text.x = element_text(angle = 45,
                                   hjust = 1,
                                   vjust = 1),
        plot.title = element_text(hjust = 0.5, face = "bold"),
        plot.margin = margin(0, 0, 0, 2, "cm")) +
  labs(y = "score",
       title = process_name,
       x="",
       subtitle = "By week",
       fill = "Type") +
  guides(color = "none") +
  facet_wrap(~disease_vac, scales = "free_x")

    ggsave(cellmarker_boxplot, 
           file = here("Figures", paste0("ssgsea_xcell_boxplot_", process_name, ".png")),
           width = 12, height = 6)
  })
}

# Apply to all processes
create_boxplots(xcell_results_annotated)
```


### 13 Vax and Covid-19 immunedeconv/XCell

```{r fig.height=6, fig.width=12}
degs_covid_vax

degs_covid_vax_l2fc_matrix = degs_covid_vax %>% 
  select(genes, condition, log2fold_change) %>% 
  distinct() %>% 
  pivot_wider(names_from = "condition",
              values_from = "log2fold_change", values_fn = mean, values_fill = 0) %>% 
  column_to_rownames("genes") %>% 
  as.matrix()


```


```{r fig.height=10, fig.width=25}

library(immunedeconv)
#XCell (ssGSEA) ----
xcell_results = immunedeconv::deconvolute(degs_covid_vax_l2fc_matrix, "xcell") 

xcell_results %>% 
  write_csv(file = here("Cell populations", "degs_covid_vax_l2fc_matrix_xcell_results.csv"))
```

##### Heatmap
```{r fig.height=12, fig.width=35}

ann_vaccines = ann_vaccines_covid_other_conditions %>% 
  mutate(pathogen = if_else(is.na(pathogen), "SARS-COV-2", pathogen),
         pathogen = str_to_upper(pathogen))

xcell_results = read_csv(file = here("Cell populations", "degs_covid_vax_l2fc_matrix_xcell_results.csv"))
filename = "All_covid_13Vax"
xcell_results_annotated = xcell_results %>% 
  pivot_longer(cols = -cell_type,
               names_to = "condition",
               values_to = "value") %>% 
  inner_join(ann_vaccines_covid_other_conditions, by = "condition") %>% 
   distinct() %>% 
  filter(condition != "BNT-MO (V1, D0)",
         condition != "BNT-MO (V1, D6)") %>% 
  mutate(disease_vac = if_else(disease_vac == "Healthy", "H", disease_vac)) %>% 
  mutate(disease_vac = fct_relevel(disease_vac, "H", "I", "V"),
           type = fct_relevel(type, "I", "RNA", "VV", "IN", "SU")) %>% 
  arrange(disease_vac,type, condition) %>% 
  inner_join(ann_vaccines_covid_other_conditions %>% select(condition), by = "condition") %>% 
  mutate(value_minmax = (value - min(value)) / 
                             (max(value) - min(value)))

hagan_vaxatlas_annotation = hagan_vaxatlas_annotation %>% 
  mutate(week = as.factor(week))

ann_vaccines_covid_other_conditions = ann_vaccines %>%
  select(condition, vaccine, pathogen, day, week, dose, infection, type, disease_vac, set) %>% 
  mutate(day = as.factor(day),
         week = as.factor(week),
         dose = as.factor(dose))
  

df = xcell_results_annotated %>% 
  filter(day != "0") %>% 
  select(cell_type, condition, value_minmax) %>% 
  pivot_wider(names_from = "cell_type",
              values_from = "value_minmax") %>% 
  column_to_rownames("condition") %>% 
  as.matrix() %>%
  as.data.frame() %>% 
  rownames_to_column("condition") %>% 
  pivot_longer(cols = -condition, 
               names_to = "cell_type",
               values_to = "value_minmax") %>% 
  inner_join(ann_vaccines_covid_other_conditions, by = "condition") %>% 
  # filter(day != "0")
    # filter(!cell_type %in% c("stroma score", "microenvironment score",
    #                          "Cancer associated fibroblast",
    #                          "Common lymphoid progenitor"))
    filter(cell_type %in% c("T cell CD8+",
                            "T cell CD4+ Th1",
                            "T cell CD4+ Th2" ,
                            "immune score",
                          # "T cell CD4+ (non-regulatory)",
                          "T cell regulatory (Tregs)",
                          "Neutrophil",
                          "Macrophage",
                          "Monocyte",
                          "B cell",
                          "Myeloid dendritic cell",
                          "NK cell",
                          "B cell plasma"))

df = df %>% mutate(day = fct_relevel(day, df %>% 
                               select(day) %>% 
                               distinct() %>% 
                               mutate(day = as.character(day),
                                      day = as.numeric(day)) %>% 
                               arrange(day) %>% 
                               pull() %>% 
                               as.character()))

row_clustered = df %>% 
  select(condition, cell_type, value_minmax) %>% 
  distinct() %>% 
  pivot_wider(names_from = "condition",
              values_from= "value_minmax", values_fn = mean) %>%
  column_to_rownames("cell_type") %>% 
  as.matrix() %>% 
  replace(is.na(.), 0) %>% 
  dist(.) %>% 
  hclust()

row_shared =  df %>%
  filter(cell_type %in% row_clustered$labels,
         value_minmax > 0.0000001) %>% 
  group_by(cell_type) %>% 
  summarize(n = n()) %>% 
  ungroup() %>% 
  arrange(-n)

result_vac <- cluster_by_day(df %>% filter(disease_vac == "V"), day_column = "day", condition_column = "condition") %>%
  distinct() %>% 
  mutate(day = fct_relevel(day, df$day %>% unique() %>% sort() %>% as.character()))

result_inf_vac01 <- cluster_by_day(df %>% 
                               filter(disease_vac == "I",
                                      dose %in% c("0", "1")), day_column = "day") %>%
  mutate(day = fct_relevel(day, df$day %>% unique() %>% sort() %>% as.character()))

result_inf_vac2 <- cluster_by_day(df %>% 
                               filter(disease_vac == "I",
                                      dose %in% c("2")), day_column = "day") %>%
  mutate(day = fct_relevel(day, df$day %>% unique() %>% sort() %>% as.character()))

column_order <- c(result_vac$condition, result_inf_vac2$condition, result_inf_vac01$condition) %>% 
  unique()

labels_set <- df %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = set) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$set) + 
  theme_void() +
  labs(y = "Set") +
  theme(axis.title.y = element_text(size = 15, hjust = 1),
        legend.direction = "vertical",
        legend.position = "right") 

labels_disease_vac <- df %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = disease_vac) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$disease_vac) + 
  theme_void() +
  labs(y = "disease_vac") +
  theme(axis.title.y = element_text(size = 15, hjust = 1),
        legend.direction = "vertical",
        legend.position = "right") 

labels_type <-df %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = type) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$type) + 
  theme_void() +
  labs(y = "Type") +
  theme(axis.title.y = element_text(size = 15, hjust = 1),
        legend.direction = "vertical",
        legend.position = "right") 

labels_dose <- df %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = dose) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$dose) + 
  theme_void() +
  labs(y = "Dose") +
  theme(axis.title.y = element_text(size = 15, hjust = 1),
        legend.direction = "vertical",
        legend.position = "right") 

labels_pathogen = df %>%
  mutate(condition = factor(condition, column_order)) %>%
  ggplot() +
  aes(x = condition, y = 1, fill = pathogen) +
  geom_tile() +
  scale_fill_manual(values = ann_vaxsig_colors$pathogen) +
  theme_void() +
  labs(y = "Week") +
  theme(axis.title.y = element_text(size = 15, hjust = 1),
        legend.direction = "vertical",
        legend.position = "right")

labels_day = df %>% 
  mutate(condition = factor(condition, column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = day) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$day) + 
  theme_void() +
  labs(y = "Day") +
  theme(axis.title.y = element_text(size = 15, hjust = 1),
        legend.direction = "horizontal",
        legend.position = "none")

#Plot
quartiles <- quantile(df$value_minmax, probs = c(0, 0.25, 0.5, 0.75, 1), na.rm = TRUE)


dotplot = df %>% 
  filter(cell_type %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, column_order),
         cell_type = factor(cell_type, levels = c(row_clustered$labels[row_clustered$order]))
         # cell_type = factor(cell_type, levels = c(row_shared$cell_type)) %>% 
         #   fct_rev()
         ) %>% 
  ggplot(aes(x=condition, y = cell_type, fill = value_minmax)) +
  geom_tile(color = "white",
            size = 2) + 
  theme_minimal() + 
  theme(axis.text = element_text(color = "black", size = 15) ,
        axis.line.x  = element_line(size = 1, color = "black"),
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1),
        axis.ticks = element_line(size = 1, color = "black"),
        legend.direction = "vertical",
        legend.position = "right") +
  labs(x = "",
       y = "") +
  scale_y_discrete(position = "right") +
  scale_fill_gradientn(
    colors = c("white", "#4DBBD5FF", "#EFC000FF", "#F39C12", "#d90429"),  # cores para os quartis
    values = c(0, 0.10, 0.25, 0.5, 0.75, 1),                       # limites dos quartis
    name = "Min-Max Normalized\nscore",
    limits = c(0, 1),
    oob = scales::squish
  ) 
  # scale_fill_gradientn(colors = c("white", "#4DBBD5FF", "#EFC000FF", "#DC0000FF"),
  #                      values = scales::rescale(c(quartiles)),
  #                      limits = c(min(df$value_minmax, na.rm = TRUE), max(df$value_minmax, na.rm = TRUE)),
  #                      oob = scales::squish,
  #                      name = "Regularized\nscore") 
#   scale_fill_gradientn(
#   colors = c("white", "#4DBBD5FF", "#EFC000FF", "#d90429"),  # cores para cada quartil
#   values = scales::rescale(c(0, 0.25, 0.5, 0.75, 1)),     # divisão para quartis
#   name = "Regularized\nscore",
#   limits = c(0, 1),
#   oob = scales::squish
# )
patch_xcell_vac = 
  labels_set /
  labels_disease_vac /
  labels_pathogen/
  labels_type /
  labels_dose/
  # labels_week/
  labels_day/
  dotplot +
  plot_layout(heights = c(0.2, 0.2, 0.2, 0.2, 0.2, 0.2, 3),
              guides = 'collect') +
  plot_annotation(
    # title = "Top 20 genes shared",
    #               subtitle = "At least one vaccine dose",
                  theme = theme(legend.position = "right",
                                plot.title = element_text(hjust = 0.5, size = 15),
                                plot.subtitle = element_text(hjust = 0.5, size = 15),
                                plot.margin = margin(5, 1, 1, 1, unit = "cm"),
                                legend.direction = "horizontal", 
                                legend.box = "vertical"))
patch_xcell_vac
patch_xcell_vac %>%
  ggsave(filename = here("Figures", paste0(filename, "_ssGSEA_XCell_Heatmap_filtered.png")),
         width = 35,
         height = 12)
```




```{r}
#Function to create boxplots
create_colplots <- function(result_df) {
  
  unique_processes <- unique(result_df$cell_type)
  
  purrr::map(unique_processes, function(process_name) {
    cellmarker_boxplot <- result_df %>% 
      filter(cell_type == process_name) %>% 
  ggplot() +
  aes(x = condition, y = value, fill = if_else(day == 0, "Day 0", type)) +  
  geom_hline(yintercept = 0, linetype = 2) +
  geom_col() +
  theme_few() +
  scale_fill_manual(values = c("Day 0" = "white", 
                               ann_vaxsig_colors$type)) +
    scale_color_manual(values = c("Day 0" = "gray80", 
                               ann_vaxsig_colors$type)) +
  theme(axis.text.x = element_text(angle = 45,
                                   hjust = 1,
                                   size = 10,
                                   vjust = 1),
        plot.title = element_text(hjust = 0.5, face = "bold"),
        plot.margin = margin(0, 0, 0, 2, "cm"),
        panel.grid.major = element_line(size = 0.05),
        panel.grid.minor = element_line(size = 0.05)) +
  labs(y = "Score",
       title = process_name,
       x="",
       subtitle = "By week",
       fill = "Type") +
  guides(color = "none") +
  ggforce::facet_row(~type, scales = "free_x", space = "free")

    ggsave(cellmarker_boxplot, 
           file = here("Figures","XCell", paste0("ssgsea_xcell_barplot_AllCovid13Vax_", process_name, ".png")),
           width = 25, height = 6)
  })
}

# Apply to all processes
create_colplots(xcell_results_annotated %>% 
  mutate(condition = fct_relevel(condition, c(ann_vaccines_covid_other_conditions %>%
                                                mutate(day = as.character(day),
                                                       day = as.integer(day)) %>% 
                                                arrange(disease_vac, vaccine, day) %>% pull(condition)))))
```

###### Timeline

```{r fig.height=10, fig.width=10}
ann_vaccines = ann_vaccines_covid_other_conditions

gsea_l2fc_tbl = xcell_results_annotated %>% 
  filter(!condition %in% c("ChAd-BNT (V2, D0)",
                           "BNT-I-BNT (D51, hospital)")) %>% 
  select(condition, type, dose, day, pathogen, everything())

gsea_l2fc_BNT_MO_before <- gsea_l2fc_tbl %>%
  filter(condition %in% c("BNT (V2, D1)", "BNT (V1, D6)")) %>% 
  mutate(condition = if_else(condition == "BNT (V1, D6)", "BNT-MO (V1, D6)", "BNT-MO (V2, D1)"),
         vaccine = "MO")

gsea_l2fc_I <- gsea_l2fc_tbl %>%
  filter(disease_vac == "I") %>% 
  select(condition, pathogen, everything())

gsea_l2fc_tbl = gsea_l2fc_tbl %>% 
  filter(!condition %in% c(gsea_l2fc_I$condition %>% unique()))

gsea_l2fc_tbl_2 = bind_rows(gsea_l2fc_tbl,
            gsea_l2fc_BNT_MO_before,
            gsea_l2fc_I) %>% 
    mutate(day = as.character(day),
           day = as.numeric(day),
         pathogen = if_else(is.na(pathogen) & disease_vac == "V", 
                            paste0("SARS-CoV-2/", vaccine), 
                            pathogen),
         pathogen = str_to_upper(pathogen),
         pathogen = case_when(str_detect(condition, "BNT-I.*hospital") ~ "SARS-COV-2/BNT-I Hospital",
                              str_detect(condition, "I.*hospital") ~ "SARS-COV-2/I Hospital", 
                              condition == "BNT-I (D0-5)" ~ "SARS-COV-2/BNT-I", 
                              condition == "I (D0-4)" ~ "SARS-COV-2/I", 
                              condition == "I-BNT (D5)" ~ "SARS-COV-2/I-BNT", 
                              condition == "I-I (D0-5)" ~ "SARS-COV-2/I-I", 
                              .default = pathogen)) %>% 
    arrange(day) %>% 
    mutate(type_grouped = case_when(type == "CONJ" ~ "Conjugated/PS",
                                 type == "PS" ~ "Conjugated/PS",
                                 type == "VV" ~ "Genetic",
                                 type == "RNA" ~ "Genetic",
                                 type == "LA" ~ "Live attenuated",
                                 type == "IN" ~ "Inactivated",
                                 type == "SU" ~ "Subunit",
                                 disease_vac == "I" ~ "Infection",
                                 TRUE ~ type),
         type = case_when(disease_vac == "I" ~ "I",
                                 TRUE ~ type),
         type_grouped = as.factor(type_grouped),
         type_grouped = fct_relevel(type_grouped, 
                            "Infection",
                            "Live Attenuated",
                            "Inactivated",
                            "Conjugated/PS",
                            "Subunit",
                            "Genetic")) %>% 
  select(condition, type, type_grouped, dose, day, pathogen, everything()) 

gsea_l2fc_day0 <- gsea_l2fc_tbl_2 %>%
  select(condition, type, type_grouped, dose, pathogen, day, cell_type) %>% 
  mutate(day = 0, 
         value_minmax = 0, 
         dose = "0",
         condition = paste0(pathogen, " D(0)")) %>% 
  distinct()

gsea_l2fc_tbl_2 = gsea_l2fc_tbl_2 %>% 
  bind_rows(gsea_l2fc_day0) %>% 
  arrange(condition)

gsea_l2fc_tbl_2 %>% 
  select(condition, type, type_grouped, dose, pathogen, day) %>% 
  distinct() %>% 
  arrange(pathogen, day)
  

```



```{r fig.height=20, fig.width=15}
day_filter = 30
idorder = c("immune score",
            "Neutrophil",
            "Macrophage",
            "Monocyte",
            "Myeloid dendritic cell",
            "NK cell",
            "T cell CD8+",
            "T cell CD4+ Th1",
            "T cell CD4+ Th2" ,
            # "T cell CD4+ (non-regulatory)",
            "T cell regulatory (Tregs)",
            "B cell",
            "B cell plasma")

all_together = gsea_l2fc_tbl_2 %>%
  filter(cell_type %in% c(idorder)) %>% 
  filter(day <= day_filter, 
         type != "IN/SU") %>% 
  mutate(cell_type = fct_relevel(cell_type, c(idorder)),
         dose = as.factor(dose),
         dose = fct_relevel(dose, "0", "1", "2", "3")) %>%
  mutate(type_grouped = str_to_upper(type_grouped),
         type_grouped = as.factor(type_grouped),
         type_grouped = fct_relevel(type_grouped, 
                            "INFECTION",
                            "LIVE ATTENUATED",
                            "INACTIVATED",
                            "CONJUGATED/PS",
                            "SUBUNIT",
                            "GENETIC")) %>% 
  
  ggplot() +
  aes(x = day, y = value_minmax, group = vaccine) +
  geom_line(
  aes(color = pathogen,
      group = pathogen
    ),
  size = 1) +
  geom_point(aes(shape = dose,
                 color = pathogen),
             size = 3.5,
             color = "black",
             alpha = 1) +
  geom_point(aes(shape = dose,
                 color = pathogen),             
             size = 3,
             alpha = 0.7) +
  geom_hline(yintercept = 0, linetype = 1, alpha = 1) +
  scale_color_manual(values = c(ann_vaxsig_colors$type, 
                                ann_vaxsig_colors$target_pathogen_disease,
                                ann_vaxsig_colors$dose,
                                "SARS-COV-2/BBIBP" = "#00A087FF",
                                "SARS-COV-2/ZF2001" = "#8491B4FF",
                                "SARS-COV-2/BNT" = "#4DBBD5FF",
                                "SARS-COV-2/MO" = "#72ddf7",
                                "SARS-COV-2/CHAD" = "#3C5488FF",
                                "SARS-COV-2/I" = "red",
                                "SARS-COV-2/BNT-I" = "#ff006e",
                                "SARS-COV-2/I-BNT" = "#Af101e"
                                )) +
  # Manual shape scale
    scale_shape_manual(values = c("0" = 15, 
                                "1" = 16, 
                                "2" = 17, 
                                "3" = 18)) +
  theme_minimal()+
  # ggthemes::theme_few() +
  theme(legend.position = "top",
        strip.text.x = element_text(color = "black", size = 12), 
        axis.text.x = element_text(color = "black", size = 10),
        axis.text.y = element_text(vjust = 0.5, size = 10, color = "black"),
        axis.line = element_line(size = 1, colour = "black", linetype=1),
        axis.ticks = element_line(size = 0.5, color = "black"),
        panel.grid.major.y = element_line(size = 0, 
                                    linetype = 2,
                                    color = "white"),
        panel.grid.major.x = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75")) +
  labs(x = "Day",
       y = "Min-Max Normalized score",
       color = "Pathogen",
       shape = "Type",
       alpha = "COVID-19 or Other",
       title = "Temporal dynamics of immunization, GSEA of Immune processes (BTM + ImmuneGO)") +
    facet_wrap(~cell_type, 
               scales = "free",
               nrow = 6)

all_together

all_together %>%
  ggsave(filename = here("Figures", "XCell_Linechart_ByImmuneProcess_AllVaccines_Infection.png"),
         width = 15,
         height = 20)
```



Separated
Main processes
```{r fig.height=8, fig.width=25}

process_filter = c("IMMUNE SYSTEM",
                   "COMPLEMENT",
                   "INNATE",
                   "ADAPTIVE",
                   "HUMORAL ADAP.",
                   "CELLULAR ADAP.",
                   "IMMUNOGLOBULIN\nMEDIATED RESPONSE")
day_filter = 30
time_bytype_main = gsea_l2fc_tbl_2 %>%
  filter(day <= day_filter, 
         type != "IN/SU",
         id %in% process_filter) %>% 
  mutate(id = fct_relevel(id, c(idorder)),
         dose = as.factor(dose),
         dose = fct_relevel(dose, "0", "1", "2", "3")) %>%
  mutate(type_grouped = str_to_upper(type_grouped),
         type_grouped = as.factor(type_grouped),
         type_grouped = fct_relevel(type_grouped, 
                            "INFECTION",
                            "LIVE ATTENUATED",
                            "INACTIVATED",
                            "CONJUGATED/PS",
                            "SUBUNIT",
                            "GENETIC")) %>% 

  arrange(condition) %>% 
  distinct() %>% 
  ggplot() +
  aes(x = day, 
      y = nes, 
     # label = condition,
    group = pathogen
      ) +
  geom_line(
  aes(color = pathogen,
      group = pathogen
    ),
  size = 1) +
  geom_point(aes(shape = dose,
                 color = pathogen),
             size = 3.5,
             color = "black",
             alpha = 1) +
  geom_point(aes(shape = dose,
                 color = pathogen),             
             size = 3,
             alpha = 0.7) +
  geom_hline(yintercept = 0, linetype = 1, alpha = 1) +
  scale_color_manual(values = c(ann_vaxsig_colors$type, 
                                ann_vaxsig_colors$target_pathogen_disease,
                                ann_vaxsig_colors$dose,
                                "SARS-COV-2/BBIBP" = "#00A087FF",
                                "SARS-COV-2/ZF2001" = "#8491B4FF",
                                "SARS-COV-2/BNT" = "#4DBBD5FF",
                                "SARS-COV-2/MO" = "#72ddf7",
                                "SARS-COV-2/CHAD" = "#3C5488FF",
                                "SARS-COV-2/I" = "red",
                                "SARS-COV-2/I Hospital" = "red",
                                "SARS-COV-2/I-I" = "purple",
                                "SARS-COV-2/BNT-I" = "pink",
                                "SARS-COV-2/BNT-I Hospital" = "pink",
                                "SARS-COV-2/I-BNT" = "#Af101e"
                                
                                )) +
  scale_shape_manual(values = c("0" = 15, 
                                "1" = 16, 
                                "2" = 17, 
                                "3" = 18)) +
  scale_alpha_manual(values = c("COVID-19" = 1, "13Vax Atlas" = 1), guide = "none") +
  scale_x_continuous(breaks = c(0, 3, 7, 14, 21, 28, 30)) +
  theme_minimal() +

  labs(x = "Day",
       y = "NES",
       color = "Pathogen",
       shape = "Dose",
       alpha = "COVID-19 or Other",
       title = "Temporal dynamics of immunization", 
       subtitle = "Immune System process",
       linetype = "Set") +
  facet_grid(~id~type_grouped, scales = "free") +
  theme(legend.position = "right",
        strip.text.x = element_text(color = "black", 
                                    size = 10,
                                    face = "bold"), 
        strip.text.y = element_text(color = "black", 
                                    size = 10, 
                                    hjust = 0,
                                    angle = 360,
                                    face = "bold"), 
        axis.text.x = element_text(color = "black", size = 10),
        axis.text.y = element_text(vjust = 0.5, size = 10, color = "black"),
        axis.line = element_line(size = 1, colour = "black", linetype=1),
        axis.ticks = element_line(size = 0.5, color = "black"),
        panel.grid.major.y = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.minor.y = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.major.x = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.minor.x = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75")) 

time_bytype_main 

time_bytype_main %>%
  ggsave(filename = here("Figures", "GSEA_Linechart_MainProcesses_AllVaccines_Infection_FacetByType.png"),
         width = 25,
         height = 8)

# all_together_sarsfocus %>% ggplotly()
```

Subprocesses
```{r fig.height=25, fig.width=15}
process_filter = c(
  "ANTIMICROBIAL",
  "LEUKOCYTES",
  "INFLAMMATION",
  "SIGNALING",
  "ANTIVIRAL & IFN",
  "NEUTROPHIL",
  "ARPP",
  "MONOCYTE",
  "MACROPHAGE",
  "DENDRITIC CELL",
  "NK CELL",
  "T CELL",
  "B CELL"
)

day_filter = 30
time_bytype_sub = gsea_l2fc_tbl_2 %>%
  filter(day <= day_filter, 
         type != "IN/SU",
         id %in% process_filter) %>% 
  mutate(id = fct_relevel(id, c(idorder)),
         dose = as.factor(dose),
         dose = fct_relevel(dose, "0", "1", "2", "3")) %>% 
  mutate(type_grouped = str_to_upper(type_grouped),
         type_grouped = as.factor(type_grouped),
         type_grouped = fct_relevel(type_grouped, 
                            "INFECTION",
                            "LIVE ATTENUATED",
                            "INACTIVATED",
                            "CONJUGATED/PS",
                            "SUBUNIT",
                            "GENETIC")) %>% 
  arrange(condition) %>% 
  distinct() %>% 
  ggplot() +
  aes(x = day, 
      y = nes, 
     # label = condition,
    group = pathogen
      ) +
  geom_line(
  aes(color = pathogen,
      group = pathogen
    ),
  size = 0.7) +
  geom_point(aes(shape = dose,
                 color = pathogen),
             size = 3.5,
             color = "black",
             alpha = 1) +
  geom_point(aes(shape = dose,
                 color = pathogen),             
             size = 3,
             alpha = 0.7) +
  geom_hline(yintercept = 0, linetype = 1, alpha = 1) +
  scale_color_manual(values = c(ann_vaxsig_colors$type, 
                                ann_vaxsig_colors$target_pathogen_disease,
                                ann_vaxsig_colors$dose,
                                "SARS-COV-2/BBIBP" = "#00A087FF",
                                "SARS-COV-2/ZF2001" = "#8491B4FF",
                                "SARS-COV-2/BNT" = "#4DBBD5FF",
                                "SARS-COV-2/MO" = "#72ddf7",
                                "SARS-COV-2/CHAD" = "#3C5488FF",
                                "SARS-COV-2/I" = "red",
                                "SARS-COV-2/I Hospital" = "red",
                                "SARS-COV-2/I-I" = "purple",
                                "SARS-COV-2/BNT-I" = "pink",
                                "SARS-COV-2/BNT-I Hospital" = "pink",
                                "SARS-COV-2/I-BNT" = "#Af101e"
                                
                                )) +
  scale_shape_manual(values = c("0" = 15, 
                                "1" = 16, 
                                "2" = 17, 
                                "3" = 18)) +
  scale_alpha_manual(values = c("COVID-19" = 1, "13Vax Atlas" = 1), guide = "none") +
  scale_x_continuous(breaks = c(0, 3, 7, 14, 21, 28, 30)) +
  theme_minimal() +

  labs(x = "Day",
       y = "NES",
       color = "Pathogen",
       shape = "Dose",
       alpha = "COVID-19 or Other",
       title = "Temporal dynamics of immunization", 
       subtitle = "Other processes",
       linetype = "Set") +
  facet_grid(~id~type_grouped, scales = "free") +
  theme(legend.position = "right",
        strip.text.x = element_text(color = "black", 
                                    size = 10,
                                    face = "bold"), 
        strip.text.y = element_text(color = "black", 
                                    size = 10, 
                                    hjust = 0,
                                    angle = 360,
                                    face = "bold"), 
        axis.text.x = element_text(color = "black", size = 10),
        axis.text.y = element_text(vjust = 0.5, size = 10, color = "black"),
        axis.line = element_line(size = 1, colour = "black", linetype=1),
        axis.ticks = element_line(size = 0.5, color = "black"),
        panel.grid.major.y = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.minor.y = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.major.x = element_line(size = 0.5, 
                                    linetype = 2,
                                    color = "gray75"),
        panel.grid.minor.x = element_line(size = 0, 
                                    linetype = 2,
                                    color = "gray75"))  

time_bytype_sub 

time_bytype_sub %>%
  ggsave(filename = here("Figures", "GSEA_Linechart_SubProcesses_AllVaccines_Infection_FacetByType.png"),
         width = 15,
         height = 20)
```



##### PCA

```{r}
library(tidyverse)
library(ggrepel) #load before factoextra
library(factoextra)
library(caret)
library(stats)
library(ggfortify)
library(ggplot2)
library(corrplot)
library(cowplot)
library(ggcorrplot)
library(ComplexHeatmap)
library(circlize)
library(ggExtra)
library(FactoMineR)
library(here)
library(PCAtools)
library(ggplotify)

```

**Required files**

```{r}
# Genes in long table format
degs_covid_vax_l2fc_matrix_xcell_results <- read_csv("Cell populations/degs_covid_vax_l2fc_matrix_xcell_results.csv")

# Gene set annotation.Used only to merge and filter genes.
ImmuneGO_BTM_grouped_genes <- read_csv("VaxGO gene sets/ImmuneGO_BTM_grouped_genes.csv")

btm_annotation_genes_immune = ImmuneGO_BTM_grouped_genes %>% 
  select(genes) %>% distinct()

# Sample or condition annotation
ann_vaccines_covid_other_conditions = read_csv(file = here("Annotations and metadata", "ann_vaccines_covid_other_conditions.csv")) %>% 
  mutate(day = factor(day) %>%
           fct_reorder(as.numeric(as.character(day))),
        dose = factor(dose) %>% 
           fct_reorder(as.numeric(as.character(dose))),
        week = factor(week) %>% 
           fct_reorder(as.numeric(as.character(week))))

data_annotation = ann_vaccines_covid_other_conditions

```

**Filtering data**

```{r}
# Gene set data.
# ImmuneGO_Annotated_Genes_all = ImmuneGO_Annotated_Genes
# ImmuneGO_Annotated_Genes = ImmuneGO_Annotated_Genes_all %>% 
#   filter(!process %in% c("TCR REPERTOIRE", "BCR REPERTOIRE"))

#INPUT
data_genes = degs_covid_vax_l2fc_matrix_xcell_results %>% 
  distinct()
filename = "13Vax_Covid_BTM_ImmuneGO_PCA_XCell"

#Convert long to wide table format.
#Dataframe with sample annotation

ann_vaccines_pca_matrix = data_genes %>% 
  replace(is.na(.), 0) %>%
  column_to_rownames("cell_type") %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column("condition") %>% 
  inner_join(data_annotation %>% 
               select(condition, disease_vac, vaccine, day, week, dose, type, set), by = "condition") 

# Convert dataframe to matrix without annotation columns
ann_vaccines_pca_matrix_ready = ann_vaccines_pca_matrix %>% 
  column_to_rownames("condition") %>% 
  select(!disease_vac:set) %>% 
  select(!c("stroma score", "microenvironment score", "Cancer associated fibroblast"))


```


```{r fig.height=5, fig.width=10, cache = TRUE}
##### PCA using PCAtools
# BiocManager::install('PCAtools')

library(PCAtools)

#Matrix (genes, sample)
p_vacinfec_matrix = ann_vaccines_pca_matrix_ready %>% 
  t() %>% 
  as.data.frame()
#Annotation 
p_vacinfec_metadata = data_annotation %>% 
  inner_join(p_vacinfec_matrix %>% t() %>% 
               as.data.frame() %>% 
               rownames_to_column("condition") %>% select("condition"), 
             by = "condition") %>% 
  column_to_rownames("condition") %>% 
  select(disease_vac, type, vaccine, dose, day, week, set)

p_vacinfec_matrix = p_vacinfec_matrix%>% 
  select(rownames(p_vacinfec_metadata)) %>% 
  as.matrix()


#PCA
p_vacinfec <- pca(mat = p_vacinfec_matrix, 
         metadata = p_vacinfec_metadata, 
         removeVar = 0.1)

options(ggrepel.max.overlaps = Inf)
#Plot
conflicted::conflicts_prefer(PCAtools::biplot)

#Screeplot

p_vacinfec_horn <- parallelPCA(mat)
p_vacinfec_elbow <- findElbowPoint(p_vacinfec$variance)
conflicted::conflicts_prefer(PCAtools::screeplot)

p_vacinfec_screeplot = screeplot(p_vacinfec,
          components = getComponents(p_vacinfec, 1:20),
          vline = c(p_vacinfec_horn$n, p_vacinfec_elbow)) +
  geom_label(aes(x = p_vacinfec_horn$n + 1, y = 50,
                 label = 'Horn\'s', vjust = -1, size = 8)) +
  geom_label(aes(x = p_vacinfec_elbow + 1, y = 50,
                 label = 'Elbow method', vjust = -1, size = 8)) 


p_vacinfec_screeplot %>% 
  ggsave(filename = here("Figures", paste0(filename, "_PCA_WithInfection_Screeplot.png")), 
         width = 10,
         height = 5)


p_vacinfec_screeplot
```


```{r fig.height=10, fig.width=10, message=FALSE, warning=FALSE, cache=FALSE}
#Plot all paired components 
p_vacinfec_pairsplot_type = pairsplot(p_vacinfec,
          components = getComponents(p_vacinfec, c(1:6)),
          triangle = T, 
          trianglelabSize = 12,
          hline = 0, vline = 0,
          pointSize = 2,
          gridlines.major = FALSE, 
          gridlines.minor = FALSE,
          colby = 'type',
          colkey = ann_vaxsig_colors$type,
          title = 'Pairs plot, by vaccine type', returnPlot = T,
          plotaxes = FALSE)  +
    theme(axis.text = element_text(size = 10, color = "black"),
      panel.grid.minor = element_blank(),
      panel.grid.major = element_line(size = 1, 
                                    linetype = 1,
                                    color = "gray90"),
        axis.line = element_line(size = 0, colour = "black", linetype=1),
        # axis.line = element_line(size = 1, colour = "black", linetype=1),
        text = element_text(color = "black", size = 10),
        panel.border = element_rect(fill = NA, color = "gray90", size = 2)) 

p_vacinfec_pairsplot_type %>% 
   ggsave(filename = here("Figures", paste0(filename, "_PCA_WithInfection_Pairsplot_TYPE.png")), 
         width = 10,
         height = 10)

p_vacinfec_pairsplot_type

p_vacinfec_pairsplot_disease_vac = pairsplot(p_vacinfec,
          components = getComponents(p_vacinfec, c(1:6)),
          triangle = T, 
          trianglelabSize = 12,
          hline = 0, vline = 0,
          pointSize = 2,
          gridlines.major = FALSE, 
          gridlines.minor = FALSE,
          colby = 'day',
          colkey = ann_vaxsig_colors$day,
          title = 'Pairs plot, by day', returnPlot = T,
          plotaxes = FALSE) +
  theme(axis.text = element_text(size = 10, color = "black"),
      panel.grid.minor = element_blank(),
      panel.grid.major = element_line(size = 1, 
                                    linetype = 1,
                                    color = "gray90"),
        axis.line = element_line(size = 0, colour = "black", linetype=1),
        # axis.line = element_line(size = 1, colour = "black", linetype=1),
        text = element_text(color = "black", size = 10),
        panel.border = element_rect(fill = NA, color = "gray90", size = 2)) 
p_vacinfec_pairsplot_disease_vac

p_vacinfec_pairsplot_disease_vac %>% 
ggsave(filename = here("Figures", paste0(filename, "_PCA_WithInfection_Pairsplot_Day.png")), 
         width = 10,
         height = 10)

```


```{r fig.height=5, fig.width=10, cache = TRUE}
#Loadings plot
p_vacinfec_loadingsplot = plotloadings(p_vacinfec, 
                                       labSize = 3,  
                                       shapeSizeRange = 5,
                                       titleLabSize = 22,
                            components = 1:10,
                            title = "Loadings plot, Vaccination and infection",
                            col = c(low = '#4DBBD5FF', mid = "white", high = '#DC0000FF') )

#Meta variables
p_vacinfec_metavars = eigencorplot(p_vacinfec,
             metavars = colnames(p_vacinfec_metadata), 
             main = "Meta variables, Vaccination and infection",
             col = c(low = '#4DBBD5FF', mid = "white", high = '#DC0000FF'))

p_vacinfec_metavars
```


```{r fig.height=15, fig.width=10, cache = TRUE}
library(ggplotify)
library(cowplot)


vacinfec_metavars_loadings <- plot_grid(p_vacinfec_loadingsplot,
      as.grob(p_vacinfec_metavars),
      ncol = 1,
      labels = c('D', 'E'),
      label_fontface = 'bold',
      # label_size = 22,
      align = 'v',
      rel_widths = c(1, 1.5),
      rel_heights = c(1, 1))

vacinfec_metavars_loadings %>% 
  ggsave(filename = here("Figures", paste0(filename, "_PCA_WithInfection_Metavars_Loadings.png")), 
         width = 10,
         height = 15)


vacinfec_metavars_loadings

```


```{r fig.height=10, fig.width=10}
vac_p_biplot_encircle_PC2_PC4 = biplot(p_vacinfec,  
                               x = 'PC2', 
                               y = 'PC4',
                               ntopLoadings = 10,
                               lengthLoadingsArrowsFactor = 2,
                               alphaLoadingsArrow = 0.5,
                               legendPosition = 'right',
                      labSize = 3,
                      showLoadings = T, 
                      lab = NULL,
                      pointSize = 5,
                      encircle = F,
                      encircleFill = T,
                      encircleAlpha = 0.1,
                      encircleLineSize = 5, 
                      colby = 'type',
                      hline = 0, 
                      vline = 0,
                      max.overlaps = Inf,
                      colkey = ann_vaxsig_colors$type)

vac_p_biplot_encircle_PC2_PC4

vac_p_biplot_encircle_PC4_PC7 = biplot(p_vacinfec,  
                               x = 'PC4', 
                               y = 'PC7',
                               ntopLoadings = 10,
                               lengthLoadingsArrowsFactor = 2,
                               alphaLoadingsArrow = 0.5,
                               legendPosition = 'right',
                      labSize = 3,
                      showLoadings =F, 
                      lab = p_vacinfec$yvars,
                      pointSize = 5,
                      encircle = F,
                      encircleFill = T,
                      encircleAlpha = 0.1,
                      encircleLineSize = 5, 
                      colby = 'type',
                      hline = 0, 
                      vline = 0,
                      max.overlaps = Inf,
                      colkey = ann_vaxsig_colors$type)

vac_p_biplot_encircle_PC4_PC7

vac_p_biplot_encircle_PC4_PC7_day = biplot(p_vacinfec,  
                               x = 'PC4', 
                               y = 'PC7',
                               ntopLoadings = 10,
                               lengthLoadingsArrowsFactor = 2,
                               alphaLoadingsArrow = 0.5,
                               legendPosition = 'right',
                      labSize = 3,
                      showLoadings = T, 
                      lab = NULL,
                      pointSize = 5,
                      encircle = F,
                      encircleFill = T,
                      encircleAlpha = 0.1,
                      encircleLineSize = 5, 
                      colby = 'day',
                      hline = 0, 
                      vline = 0,
                      max.overlaps = Inf,
                      colkey = ann_vaxsig_colors$day)

vac_p_biplot_encircle_PC4_PC7_day

vac_p_biplot_encircle_PC4_PC7 %>% ggplotly()

```







### COVID-19 only
```{r fig.height=10, fig.width=25}

library(immunedeconv)
#XCell (ssGSEA) ----
xcell_results = immunedeconv::deconvolute(degs_covid_vax_l2fc_matrix, "xcell") 

xcell_results %>% 
  write_csv(file = here("Cell populations", "degs_covid_vax_l2fc_matrix_xcell_results.csv"))
```

##### Heatmap
```{r fig.height=10, fig.width=15}

ann_vaccines = ann_vaccines_covid_other_conditions %>% 
  mutate(pathogen = if_else(is.na(pathogen), "SARS-COV-2", pathogen),
         pathogen = str_to_upper(pathogen)) %>% 
  filter(set == "COVID-19")

xcell_results = read_csv(file = here("Cell populations", "degs_covid_vax_l2fc_matrix_xcell_results.csv"))
filename = "COVID"
xcell_results_annotated = xcell_results %>% 
  pivot_longer(cols = -cell_type,
               names_to = "condition",
               values_to = "value") %>% 
  inner_join(ann_vaccines_covid_other_conditions, by = "condition") %>% 
   distinct() %>% 
  filter(condition != "BNT-MO (V1, D0)",
         condition != "BNT-MO (V1, D6)") %>% 
  mutate(disease_vac = if_else(disease_vac == "Healthy", "H", disease_vac)) %>% 
  mutate(disease_vac = fct_relevel(disease_vac, "H", "I", "V"),
           type = fct_relevel(type, "I", "RNA", "VV", "IN", "SU")) %>% 
  arrange(disease_vac,type, condition) %>% 
  inner_join(ann_vaccines_covid_other_conditions %>% select(condition), by = "condition") %>% 
  mutate(value_minmax = (value - min(value)) / 
                             (max(value) - min(value)))

hagan_vaxatlas_annotation = hagan_vaxatlas_annotation %>% 
  mutate(week = as.factor(week))

ann_vaccines_covid_other_conditions = ann_vaccines %>%
  select(condition, vaccine, pathogen, day, week, dose, infection, type, disease_vac, set) %>% 
  mutate(day = as.factor(day),
         week = as.factor(week),
         dose = as.factor(dose))
  

df = xcell_results_annotated %>% 
  filter(day != "0") %>% 
  select(cell_type, condition, value_minmax) %>% 
  pivot_wider(names_from = "cell_type",
              values_from = "value_minmax") %>% 
  column_to_rownames("condition") %>% 
  as.matrix() %>%
  as.data.frame() %>% 
  rownames_to_column("condition") %>% 
  pivot_longer(cols = -condition, 
               names_to = "cell_type",
               values_to = "value_minmax") %>% 
  inner_join(ann_vaccines_covid_other_conditions, by = "condition") %>% 
  # filter(day != "0")
    # filter(!cell_type %in% c("stroma score", "microenvironment score",
    #                          "Cancer associated fibroblast",
    #                          "Common lymphoid progenitor"))
    filter(cell_type %in% c("T cell CD8+",
                            "T cell CD4+ Th1",
                            "T cell CD4+ Th2" ,
                            "immune score",
                          # "T cell CD4+ (non-regulatory)",
                          "T cell regulatory (Tregs)",
                          "Neutrophil",
                          "Macrophage",
                          "Monocyte",
                          "B cell",
                          "Myeloid dendritic cell",
                          "NK cell",
                          "B cell plasma"))

df = df %>% mutate(day = fct_relevel(day, df %>% 
                               select(day) %>% 
                               distinct() %>% 
                               mutate(day = as.character(day),
                                      day = as.numeric(day)) %>% 
                               arrange(day) %>% 
                               pull() %>% 
                               as.character()))

row_clustered = df %>% 
  select(condition, cell_type, value_minmax) %>% 
  distinct() %>% 
  pivot_wider(names_from = "condition",
              values_from= "value_minmax", values_fn = mean) %>%
  column_to_rownames("cell_type") %>% 
  as.matrix() %>% 
  replace(is.na(.), 0) %>% 
  dist(.) %>% 
  hclust()

row_shared =  df %>%
  filter(cell_type %in% row_clustered$labels,
         value_minmax > 0.0000001) %>% 
  group_by(cell_type) %>% 
  summarize(n = n()) %>% 
  ungroup() %>% 
  arrange(-n)

result_vac <- cluster_by_day(df %>% filter(disease_vac == "V"), day_column = "day", condition_column = "condition") %>%
  distinct() %>% 
  mutate(day = fct_relevel(day, df$day %>% unique() %>% sort() %>% as.character()))

result_inf_vac01 <- cluster_by_day(df %>% 
                               filter(disease_vac == "I",
                                      dose %in% c("0", "1")), day_column = "day") %>%
  mutate(day = fct_relevel(day, df$day %>% unique() %>% sort() %>% as.character()))

result_inf_vac2 <- cluster_by_day(df %>% 
                               filter(disease_vac == "I",
                                      dose %in% c("2")), day_column = "day") %>%
  mutate(day = fct_relevel(day, df$day %>% unique() %>% sort() %>% as.character()))

column_order <- c(result_vac$condition, result_inf_vac2$condition, result_inf_vac01$condition) %>% 
  unique()

labels_set <- df %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = set) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$set) + 
  theme_void() +
  labs(y = "Set") +
  theme(axis.title.y = element_text(size = 15, hjust = 1),
        legend.direction = "vertical",
        legend.position = "right") 

labels_disease_vac <- df %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = disease_vac) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$disease_vac) + 
  theme_void() +
  labs(y = "disease_vac") +
  theme(axis.title.y = element_text(size = 15, hjust = 1),
        legend.direction = "vertical",
        legend.position = "right") 

labels_type <-df %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = type) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$type) + 
  theme_void() +
  labs(y = "Type") +
  theme(axis.title.y = element_text(size = 15, hjust = 1),
        legend.direction = "vertical",
        legend.position = "right") 

labels_dose <- df %>% 
  mutate(condition = factor(condition, levels = column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = dose) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$dose) + 
  theme_void() +
  labs(y = "Dose") +
  theme(axis.title.y = element_text(size = 15, hjust = 1),
        legend.direction = "vertical",
        legend.position = "right") 

labels_pathogen = df %>%
  mutate(condition = factor(condition, column_order)) %>%
  ggplot() +
  aes(x = condition, y = 1, fill = pathogen) +
  geom_tile() +
  scale_fill_manual(values = ann_vaxsig_colors$pathogen) +
  theme_void() +
  labs(y = "Week") +
  theme(axis.title.y = element_text(size = 15, hjust = 1),
        legend.direction = "vertical",
        legend.position = "right")

labels_day = df %>% 
  mutate(condition = factor(condition, column_order)) %>% 
  ggplot() +
  aes(x = condition, y = 1, fill = day) + 
  geom_tile() + 
  scale_fill_manual(values = ann_vaxsig_colors$day) + 
  theme_void() +
  labs(y = "Day") +
  theme(axis.title.y = element_text(size = 15, hjust = 1),
        legend.direction = "horizontal",
        legend.position = "none")

#Plot
quartiles <- quantile(df$value_minmax, probs = c(0, 0.25, 0.5, 0.75, 1), na.rm = TRUE)


dotplot = df %>% 
  filter(cell_type %in% row_clustered$labels) %>% 
  mutate(condition = factor(condition, column_order),
         cell_type = factor(cell_type, levels = c(row_clustered$labels[row_clustered$order]))
         # cell_type = factor(cell_type, levels = c(row_shared$cell_type)) %>% 
         #   fct_rev()
         ) %>% 
  ggplot(aes(x=condition, y = cell_type, fill = value_minmax)) +
  geom_tile(color = "white",
            size = 2) + 
  theme_minimal() + 
  theme(axis.text = element_text(color = "black", size = 15) ,
        axis.line.x  = element_line(size = 1, color = "black"),
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1),
        axis.ticks = element_line(size = 1, color = "black"),
        legend.direction = "vertical",
        legend.position = "right") +
  labs(x = "",
       y = "") +
  scale_y_discrete(position = "right") +
  scale_fill_gradientn(
    colors = c("white", "#4DBBD5FF", "#EFC000FF", "#F39C12", "#d90429"),  # cores para os quartis
    values = c(0, 0.10, 0.25, 0.5, 0.75, 1),                       # limites dos quartis
    name = "Min-Max Normalized\nscore",
    limits = c(0, 1),
    oob = scales::squish
  ) 
  # scale_fill_gradientn(colors = c("white", "#4DBBD5FF", "#EFC000FF", "#DC0000FF"),
  #                      values = scales::rescale(c(quartiles)),
  #                      limits = c(min(df$value_minmax, na.rm = TRUE), max(df$value_minmax, na.rm = TRUE)),
  #                      oob = scales::squish,
  #                      name = "Regularized\nscore") 
#   scale_fill_gradientn(
#   colors = c("white", "#4DBBD5FF", "#EFC000FF", "#d90429"),  # cores para cada quartil
#   values = scales::rescale(c(0, 0.25, 0.5, 0.75, 1)),     # divisão para quartis
#   name = "Regularized\nscore",
#   limits = c(0, 1),
#   oob = scales::squish
# )
patch_xcell_vac = 
  labels_disease_vac /
  labels_type /
  labels_dose/
  # labels_week/
  labels_day/
  dotplot +
  plot_layout(heights = c(0.2, 0.2, 0.2, 0.2, 3),
              guides = 'collect') +
  plot_annotation(
    # title = "Top 20 genes shared",
    #               subtitle = "At least one vaccine dose",
                  theme = theme(legend.position = "right",
                                plot.title = element_text(hjust = 0.5, size = 15),
                                plot.subtitle = element_text(hjust = 0.5, size = 15),
                                plot.margin = margin(5, 1, 1, 1, unit = "cm"),
                                legend.direction = "horizontal", 
                                legend.box = "vertical"))
patch_xcell_vac
patch_xcell_vac %>%
  ggsave(filename = here("Figures", paste0(filename, "_ssGSEA_XCell_Heatmap_filtered.png")),
         width = 15,
         height = 10)
```






## Heatmaps - Genes - Covid and other 13 vaccines

Import files

```{r }
############## Import files

# ImmuneGO 
ImmuneGO_Annotated_GO = read_csv(here("VaxGO gene sets", "ImmuneGO_Annotated_GO_2024_03_26.csv"))
ImmuneGO_Annotated_Genes = readRDS(here("VaxGO gene sets",  "ImmuneGO_Annotated_Genes_2024-03-26.rds"))
ImmuneGO_BTM_grouped_genes <- read_csv("VaxGO gene sets/ImmuneGO_BTM_grouped_genes.csv")

# Informations about gene sets
ImmuneGO = ImmuneGO_BTM_grouped_genes %>% 
  mutate(annotation = if_else(is.na(immune_system),"Manual", "Specific"),
         immune_system = if_else(is.na(immune_system), process, immune_system)) %>% 
  distinct()

# Filter only manual annotated gene sets
immunego_genes = ImmuneGO %>%
  filter(annotation == "Manual") %>% 
  select(process,genes)

ann_vaccines = ann_vaccines_covid_other_conditions

# DEGs table
data = degs_covid_vax %>% arrange(condition)

#Colors
ann_colors = ann_vaxsig_colors 
```


### ImmuneGO genes
Process data

```{r }
############## Process data

# All gene sets
Immune_GO_genes_All = data %>% 
  inner_join(immunego_genes, by = "genes") %>% 
  merge(ann_vaccines_covid_other_conditions, by = "condition", all.y=F) %>% 
  clean_names()

Immune_GO_genes_All$process %>% unique()

###### Other immune gene sets
Immune_GO_genes_general = Immune_GO_genes_All %>%
  filter(process == "IMMUNE SYSTEM") %>% 
  distinct()
Immune_GO_genes_general = Immune_GO_genes_All %>%
  filter(process == "LEUKOCYTES") %>% 
  distinct()
Immune_GO_genes_antimicrobial = Immune_GO_genes_All %>%
  filter(process == "ANTIMICROBIAL") %>% 
  distinct()
Immune_GO_genes_Innate = Immune_GO_genes_All %>%
 filter(process == "INNATE IMMUNE SYSTEM") %>% 
  distinct()
Immune_GO_genes_Adaptive = Immune_GO_genes_All %>%
 filter(process == "ADAPTIVE IMMUNE SYSTEM") %>% 
  distinct()
Immune_GO_genes_Signaling = Immune_GO_genes_All %>%
 filter(process == "SIGNALING") %>% 
  distinct()

# Innate immune system
Immune_GO_genes_Inflammation = Immune_GO_genes_All %>%
  filter(process == "INFLAMMATION") %>% 
  distinct()
Immune_GO_genes_ARPP = Immune_GO_genes_All %>%
  filter(process == "ARPP") %>% 
  distinct()
Immune_GO_genes_IFN = Immune_GO_genes_All %>%
  filter(process == "ANTIVIRAL & IFN") %>% 
  distinct()
Immune_GO_genes_Neutrophil = Immune_GO_genes_All %>%
  filter(process == "NEUTROPHIL") %>% 
  distinct()
Immune_GO_genes_Myeloid = Immune_GO_genes_All %>%
  filter(process == "MYELOID CELL") %>% 
  distinct()
Immune_GO_genes_Eosinophil = Immune_GO_genes_All %>%
  filter(process == "EOSINOPHIL") %>% 
  distinct()
Immune_GO_genes_Complement = Immune_GO_genes_All %>%
  filter(process == "COMPLEMENT") %>% 
  distinct()
Immune_GO_genes_Monocytes = Immune_GO_genes_All %>%
  filter(process == "MONOCYTE") %>% 
  distinct()
Immune_GO_genes_DC = Immune_GO_genes_All %>%
  filter(process == "DENDRITIC CELL") %>% 
  distinct()
Immune_GO_genes_Macro = Immune_GO_genes_All %>%
  filter(process == "MACROPHAGE") %>% 
  distinct()
Immune_GO_genes_NK = Immune_GO_genes_All %>%
  filter(process == "NK CELL") %>% 
  distinct()
Immune_GO_genes_Mast = Immune_GO_genes_All %>%
  filter(process == "MAST CELL") %>%
  distinct()

# Adaptive immune system
Immune_GO_genes_Cellular = Immune_GO_genes_All %>%
  filter(process == "CELLULAR ADAPTIVE IMMUNE SYSTEM")  %>% 
  distinct()
Immune_GO_genes_Humoral = Immune_GO_genes_All %>%
  filter(process == "HUMORAL ADAPTIVE IMMUNE SYSTEM") %>% 
  distinct()
Immune_GO_genes_Tcells = Immune_GO_genes_All %>%
  filter(process == "T CELL") %>% 
  distinct()
# Immune_GO_genes_Th = Immune_GO_genes_All %>%
#   filter(process == "T HELPER CELLS") %>% 
#   distinct()
Immune_GO_genes_Bcells = Immune_GO_genes_All %>%
  filter(process == "B CELL") %>% 
  distinct()
Immune_GO_genes_Ig = Immune_GO_genes_All %>%
  filter(process == "IMMUNOGLOBULIN MEDIATED RESPONSE") %>%
  distinct() %>% 
  distinct()

```

### Interactive heatmap version - All immmune genes

```{r}
library(InteractiveComplexHeatmap)
library(maditr)

####### Input
heatmap_data = Immune_GO_genes_All
file = "Immune_GO_genes_All_Covid_13Vaccines"

###### Process data

heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Convert to wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0) 

matrix_data_ready =  matrix_data %>% 
  round(2)

# Check if all vaccines are in the matrix and join annotations
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines_covid_other_conditions %>% select(condition, set, disease_vac, type, week, day), by = "condition")

# Join annotations
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(day) %>% 
  column_to_rownames(var= "condition")  %>% 
  mutate(day = factor(day) %>%
           fct_reorder(as.numeric(as.character(day))),
         week = factor(week) %>%
           fct_reorder(as.numeric(as.character(week))))


#Rows Immune ----
ann_rows_immune = matrix_data %>%
  as.data.frame() %>%
  rownames_to_column("genes") %>%
  left_join(ImmuneGO_Annotated_Genes %>%
              filter(go_term == "Manual",
                     !process %in% c("TCR REPERTOIRE", "BCR REPERTOIRE")), by = "genes") %>%
  select(genes, process) %>%
  distinct() %>%
  group_by(genes) %>%
  arrange(genes, factor(process, levels = c("INNATE IMMUNE SYSTEM", "ADAPTIVE IMMUNE SYSTEM")), .by_group = TRUE) %>%
  mutate(i_process = row_number()) %>%
  pivot_wider(., names_from = "i_process", values_from = "process") %>%
  column_to_rownames("genes") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("index") %>%
  mutate(index = as.numeric(index)) %>%
  arrange(desc(index)) %>%
  column_to_rownames("index") %>%
  t() %>%
  as.data.frame() %>%
  replace(is.na(.), ".")

# Reorder columns (same order as lines in the annotation table)
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(day) %>% 
  select(!c(set, disease_vac, type, week, day)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0) %>% 
  as.matrix() 


#Check dimensions
dim(matrix_data_ordered) 
dim(ann_vaccines_heatmap) 
dim(ann_rows_immune)
```

By type
```{r}
column_split = ann_vaccines_heatmap$type %>% as.character()

dend1 = cluster_between_groups(matrix_data_ordered, column_split)


#Create heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")

row_ha = HeatmapAnnotation(df = ann_rows_immune,
                       col = col_process_immune,
                       which= "row",
                       simple_anno_size = unit(2, "mm"),
                       show_annotation_name = F,
                       show_legend = F)

col_fun <- colorRamp2(c(-1, 0, 1), 
                      c("#4DBBD5FF", "white", "#DC0000FF"))

#Parameters
mat = matrix_data_ordered
width_image = 50
height_image = 30
width = unit(40, "cm")
height = unit(20, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plot and save
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap_ByType.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        # left_annotation = row_ha,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "L2FC", #Legend
                        col = col_fun, 
                        cluster_columns = dend1,
                        column_names_rot = 45,
                        show_row_dend = F,
                        row_split = NULL, #Split group clusters
                        show_row_names = F,
                        column_names_gp = column_names_gp,
                        width = width,
                        height = height,
                        row_names_gp = row_names_gp)

ht_list = draw(heatmap_plot_all, 
               annotation_legend_side = "right", 
               heatmap_legend_side = "right")


draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```


By day
```{r}

#Create heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")

row_ha = HeatmapAnnotation(df = ann_rows_immune,
                       col = col_process_immune,
                       which= "row",
                       simple_anno_size = unit(2, "mm"),
                       show_annotation_name = F,
                       show_legend = F)

col_fun <- colorRamp2(c(-1, 0, 1), 
                      c("#4DBBD5FF", "white", "#DC0000FF"))

#Parameters
mat = matrix_data_ordered
width_image = 50
height_image = 30
width = unit(40, "cm")
height = unit(20, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plot and save
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap_ByDay.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "L2FC", #Legend
                        col = col_fun, 
                        cluster_columns = F,
                        column_names_rot = 45,
                        show_row_dend = T,
                        row_split = 2, #Split group clusters
                        show_row_names = F,
                        column_names_gp = column_names_gp,
                        width = width,
                        height = height,
                        row_names_gp = row_names_gp)

ht_list = draw(heatmap_plot_all, 
               annotation_legend_side = "right", 
               heatmap_legend_side = "right")


draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

No infection
```{r}
library(InteractiveComplexHeatmap)
library(maditr)

####### Input
heatmap_data = Immune_GO_genes_All
file = "Immune_GO_genes_All_Covid_13Vaccines"

###### Process data

heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Convert to wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

matrix_data_ready =  matrix_data %>% 
  round(2)

# Check if all vaccines are in the matrix and join annotations
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines_covid_other_conditions %>% select(condition, set, disease_vac, type, week, day), by = "condition") %>% 
  filter(disease_vac != "I")  


# Join annotations
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  inner_join(ann_vaccines_complex, by = "condition") %>% 
  arrange(day) %>% 
  column_to_rownames(var= "condition")

# Reorder columns (same order as lines in the annotation table)
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  inner_join(ann_vaccines_complex, by = "condition") %>% 
  arrange(day) %>% 
  select(!c(set, disease_vac, type, week, day)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0) %>% 
  as.matrix() 


#Check dimensions
dim(matrix_data_ordered) 
dim(ann_vaccines_heatmap) 
dim(ann_rows_immune)
```

By day
```{r}
#Create heatmap annotation


ha = HeatmapAnnotation(df = ann_vaccines_heatmap,
                       col = ann_colors, 
                       annotation_name_side = "left")

row_ha = HeatmapAnnotation(df = ann_rows_immune,
                       col = col_process_immune,
                       which= "row",
                       simple_anno_size = unit(2, "mm"),
                       show_annotation_name = F,
                       show_legend = F)

col_fun <- colorRamp2(c(-0.5, 0, 0.5), 
                      c("#4DBBD5FF", "white", "#DC0000FF"))

#Parameters
mat = matrix_data_ordered
width_image = 50
height_image = 30
width = unit(40, "cm")
height = unit(20, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plot and save
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap_DAY_OnlyInfection.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        # left_annotation = row_ha,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "L2FC", #Legend
                        col = col_fun, 
                        cluster_columns = F,
                        column_names_rot = 45,
                        show_row_dend = T,
                        row_split = NULL, #Split group clusters
                        show_row_names = F,
                        column_names_gp = column_names_gp,
                        width = width,
                        height = height,
                        row_names_gp = row_names_gp)

ht_list = draw(heatmap_plot_all, 
               annotation_legend_side = "right", 
               heatmap_legend_side = "right")


draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

By Type
```{r}
#Create heatmap annotation
column_split = ann_vaccines_heatmap$type %>% as.character()

dend1 = cluster_between_groups(matrix_data_ordered, column_split)


ha = HeatmapAnnotation(df = ann_vaccines_heatmap,
                       col = ann_colors, 
                       annotation_name_side = "left")

row_ha = HeatmapAnnotation(df = ann_rows_immune,
                       col = col_process_immune,
                       which= "row",
                       simple_anno_size = unit(2, "mm"),
                       show_annotation_name = F,
                       show_legend = F)

col_fun <- colorRamp2(c(-0.5, 0, 0.5), 
                      c("#4DBBD5FF", "white", "#DC0000FF"))

#Parameters
mat = matrix_data_ordered
width_image = 50
height_image = 30
width = unit(40, "cm")
height = unit(20, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plot and save
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap_Type_OnlyInfection.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        # left_annotation = row_ha,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "L2FC", #Legend
                        col = col_fun, 
                        cluster_columns = dend1,
                        column_names_rot = 45,
                        show_row_dend = T,
                        row_split = NULL, #Split group clusters
                        show_row_names = F,
                        column_names_gp = column_names_gp,
                        width = width,
                        height = height,
                        row_names_gp = row_names_gp)

ht_list = draw(heatmap_plot_all, 
               annotation_legend_side = "right", 
               heatmap_legend_side = "right")


draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```



Interactive
```{r}
column_split = ann_vaccines_heatmap$day %>% as.character()

dend1 = cluster_between_groups(matrix_data_ordered, column_split)


#Parameters
mat = matrix_data_ordered
width_image = 40
height_image = 90
width = unit(30, "cm")
height = unit(80, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "L2FC", #Legend
                        col = col_fun,
                        cluster_columns = F,
                        show_row_dend = F,
                        row_split = 2, #Split group clusters
                        show_row_names = F,
                        column_names_gp = column_names_gp,
                        row_names_gp = row_names_gp)

ht_list = draw(heatmap_plot_all, 
               annotation_legend_side = "right", 
               heatmap_legend_side = "right")
htShiny(ht_list)

htShiny(ht_list, save = "Interactive_Covidand13Vax_Heatmap")

```

### General

```{r}
####### Input
heatmap_data = Immune_GO_genes_general
file = "Immune_GO_genes_general_Covid_13Vaccines"

###### Process data

heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Convert to wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

matrix_data_ready =  matrix_data %>% 
  round(2)

# Check if all vaccines are in the matrix and join annotations
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, day), by = "condition")

# Join annotations
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 

# Reorder columns (same order as lines in the annotation table)
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(day) %>% 
  select(!c(disease_vac, type, day)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Check dimensions
dim(matrix_data_ordered) 
dim(ann_vaccines_heatmap) 


#Create heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), 
                      c("#4DBBD5FF", "white", "#DC0000FF"))


#Parameters
mat = matrix_data_ordered
width_image = 40
height_image = 90
width = unit(30, "cm")
height = unit(80, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 3)


#Plot and save
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 10),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = F,
                        #column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### Antimicrobial
By day
```{r}
####### Input

heatmap_data = Immune_GO_genes_antimicrobial
file = "Immune_GO_genes_antimicrobial"

heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

matrix_data_ready =  matrix_data %>% 
  round(2)

ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, set, disease_vac, type, day), by = "condition")

ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(day) %>% 
  column_to_rownames(var= "condition") 

matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(day) %>% 
  select(!c(set, disease_vac, type,  day)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0) %>% 
  as.matrix() 

dim(matrix_data_ordered)
dim(ann_vaccines_heatmap)

ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-1, 0, 1), 
                      c("#4DBBD5FF", "white", "#DC0000FF"))

mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(20, "cm")
width_image = 40
height_image = 30
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)

fileimageheatmap_ungrouped = here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))
png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)
heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 10),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = T,
                        row_names_gp = row_names_gp,
                        row_split = NULL, 
                        row_gap = unit(2, "mm"),
                        cluster_columns = F,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

Clustered all
```{r}
####### Input

heatmap_data = Immune_GO_genes_antimicrobial
file = "Immune_GO_genes_antimicrobial"

heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

matrix_data_ready =  matrix_data %>% 
  round(2)

ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, set, disease_vac, type, day), by = "condition")

ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(day) %>% 
  column_to_rownames(var= "condition") 

matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(day) %>% 
  select(!c(set, disease_vac, type,  day)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0) %>% 
  as.matrix() 

dim(matrix_data_ordered)
dim(ann_vaccines_heatmap)

ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), 
                      c("#4DBBD5FF", "white", "#DC0000FF"))

mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(20, "cm")
width_image = 40
height_image = 30
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)

fileimageheatmap_ungrouped = here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap_Clustered.png"))
png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)
heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 10),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, 
                        row_gap = unit(2, "mm"),
                        cluster_columns = T,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### Complement

```{r}
heatmap_data = Immune_GO_genes_Complement
file = "Immune_GO_genes_Complement"

heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

matrix_data_ready =  matrix_data %>% 
  round(2)

ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, set, disease_vac, type, day), by = "condition")

ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  inner_join(ann_vaccines_complex, by = "condition") %>% 
  arrange(day) %>% 
  column_to_rownames(var= "condition") 

matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  inner_join(ann_vaccines_complex, by = "condition") %>% 
  arrange(day) %>% 
  select(!c(set, disease_vac, type, day)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0) %>% 
  as.matrix() 

dim(matrix_data_ordered)
dim(ann_vaccines_heatmap)


ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#4DBBD5FF", "white", "#DC0000FF"))


mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(15, "cm")
width_image = 40
height_image = 25
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)

fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))
png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)
heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 10),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = F,
                        column_split = NULL, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### Inflammation

```{r}
heatmap_data = Immune_GO_genes_Inflammation
file = "Immune_GO_genes_Inflammation"
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

matrix_data_ready =  matrix_data %>% 
  round(2)

ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, set, disease_vac, type, day), by = "condition")

ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(day) %>% 
  column_to_rownames(var= "condition") 

matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(day) %>% 
  select(!c(set, disease_vac, type, day)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0) %>% 
  as.matrix() 

dim(matrix_data_ordered) 
dim(ann_vaccines_heatmap)

ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), 
                      c("#4DBBD5FF", "white", "#DC0000FF"))

mat = matrix_data_ordered
width_image = 40
height_image = 30
width = unit(30, "cm")
height = unit(20, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)

fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, 
                        row_gap = unit(2, "mm"),
                        cluster_columns = F,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### ARPP

```{r}
####### Input
#Troque pelo geneset de interesse

heatmap_data = Immune_GO_genes_ARPP
file = "Immune_GO_genes_ARPP"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Converter para wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, set, disease_vac, type, day), by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(day) %>% 
  column_to_rownames(var= "condition") 

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(day) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(set, disease_vac, type, day)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#4DBBD5FF", "white", "#DC0000FF"))


#Parameters
mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(55, "cm")
width_image = 40
height_image = 65
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 5)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = F,
                        # column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### AV and IFN

```{r}
####### Input
#Troque pelo geneset de interesse

heatmap_data = Immune_GO_genes_IFN
file = "Immune_GO_genes_IFN"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Converter para wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(5)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, set, disease_vac, type, day), by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(day) %>% 
  column_to_rownames(var= "condition") 

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(day) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(set, disease_vac, type, day)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


# col_fun <- colorRamp2(c(-2, -1, 0, 1, 2), c("#8491B4FF", "#4DBBD5FF", "white", "#F39B7FFF", "#DC0000FF"))
col_fun <- colorRamp2(c(-2, 0, 2), c("#4DBBD5FF", "white", "#DC0000FF"))

#Parameters
mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(50, "cm")
width_image = 40
height_image = 60
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = F,
                        # column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### Neutrophil

```{r}
####### Input
#Troque pelo geneset de interesse

heatmap_data = Immune_GO_genes_Neutrophil
file = "Immune_GO_genes_Neutrophil"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Converter para wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, set, disease_vac, type, day), by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(day) %>% 
  column_to_rownames(var= "condition") 

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(day) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(set, disease_vac, type, day)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#4DBBD5FF", "white", "#DC0000FF"))


#Parameters
mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(40, "cm")
width_image = 40
height_image = 50
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 5)


#Plotar e salvar imagem
fileimageheatmap_ungrouped = here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = F,
                        # column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### Eosinophil

```{r}
####### Input
#Troque pelo geneset de interesse

heatmap_data = Immune_GO_genes_Eosinophil
file = "Immune_GO_genes_Eosinophil"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Converter para wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, set, disease_vac, type, day), by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(day) %>% 
  column_to_rownames(var= "condition") 

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(day) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(set, disease_vac, type, day)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#4DBBD5FF", "white", "#DC0000FF"))


#Parameters
mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(10, "cm")
width_image = 40
height_image = 20
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = F,
                        # column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### Monocyte

```{r}
####### Input
#Troque pelo geneset de interesse

heatmap_data = Immune_GO_genes_Monocytes
file = "Immune_GO_genes_Monocytes"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Converter para wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, set, disease_vac, type, day), by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(day) %>% 
  column_to_rownames(var= "condition") 

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(day) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(set, disease_vac, type, day)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#4DBBD5FF", "white", "#DC0000FF"))


#Parameters
mat = matrix_data_ordered
width = unit(35, "cm")
height = unit(50, "cm")
width_image = 45
height_image = 60
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 5)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = F,
                        #column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### Dendritic cells

```{r}
####### Input
#Troque pelo geneset de interesse

heatmap_data = Immune_GO_genes_DC
file = "Immune_GO_genes_DC"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Converter para wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, set, disease_vac, type, day), by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(day) %>% 
  column_to_rownames(var= "condition") 

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(day) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(set, disease_vac, type, day)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#4DBBD5FF", "white", "#DC0000FF"))


#Parameters
mat = matrix_data_ordered
width = unit(35, "cm")
height = unit(50, "cm")
width_image = 45
height_image = 60
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 5)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 10),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = F,
                        #column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### Macrophage

```{r}
####### Input
#Troque pelo geneset de interesse

heatmap_data = Immune_GO_genes_Macro
file = "Immune_GO_genes_Macro"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Converter para wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, set, disease_vac, type, day), by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(day) %>% 
  column_to_rownames(var= "condition") 

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(day) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(set, disease_vac, type, day)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#4DBBD5FF", "white", "#DC0000FF"))


#Parameters
mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(15, "cm")
width_image = 40
height_image = 25
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 10),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = F,
                        #column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### NK

```{r}
####### Input
#Troque pelo geneset de interesse

heatmap_data = Immune_GO_genes_NK
file = "Immune_GO_genes_NK"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Converter para wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, set, disease_vac, type, day), by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(day) %>% 
  column_to_rownames(var= "condition") 

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(day) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(set, disease_vac, type, day)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#4DBBD5FF", "white", "#DC0000FF"))


#Parameters
mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(40, "cm")
width_image = 40
height_image = 50
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 10),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = F,
                        #column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### Mast cell

```{r}
####### Input
#Troque pelo geneset de interesse

heatmap_data = Immune_GO_genes_Mast
file = "Immune_GO_genes_Mast"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Converter para wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, set, disease_vac, type, day), by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(day) %>% 
  column_to_rownames(var= "condition") 

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(day) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(set, disease_vac, type, day)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#4DBBD5FF", "white", "#DC0000FF"))


#Parameters
mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(10, "cm")
width_image = 40
height_image = 20
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 10),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = F,
                        #column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

###Cellular adaptive immune system

```{r}
####### Input
#Troque pelo geneset de interesse

heatmap_data = Immune_GO_genes_Cellular
file = "Immune_GO_genes_Cellular"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Converter para wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, set, disease_vac, type, day), by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(day) %>% 
  column_to_rownames(var= "condition") 

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(day) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(set, disease_vac, type, day)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#4DBBD5FF", "white", "#DC0000FF"))


#Parameters
mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(70, "cm")
width_image = 40
height_image = 80
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 5)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 10),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = F,
                        #column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

###T Cell

```{r}
####### Input
heatmap_data = Immune_GO_genes_Tcells
file = "Immune_GO_genes_Tcells"

heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

matrix_data_ready =  matrix_data %>% 
  round(2)

ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, set, disease_vac, type, day), by = "condition")

ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(day) %>% 
  column_to_rownames(var= "condition") 

matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(day) %>% 
  select(!c(set, disease_vac, type, day)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0) %>% 
  as.matrix() 

dim(matrix_data_ordered) 
dim(ann_vaccines_heatmap) 


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#4DBBD5FF", "white", "#DC0000FF"))


#Parameters
mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(80, "cm")
width_image = 40
height_image = 90
column_names_gp = gpar(fontsize = 5)
row_names_gp = gpar(fontsize = 5)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = F,
                        #column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### Th Cell

```{r}
####### Input
#Troque pelo geneset de interesse

heatmap_data = Immune_GO_genes_Th
file = "Immune_GO_genes_Th"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Converter para wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#4DBBD5FF", "white", "#DC0000FF"))


#Parameters
mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(10, "cm")
width_image = 40
height_image = 20
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 10),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        #column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### Humoral adaptive immune system

```{r}
####### Input
#Troque pelo geneset de interesse

heatmap_data = Immune_GO_genes_Humoral
file = "Immune_GO_genes_Humoral"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Converter para wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#4DBBD5FF", "white", "#DC0000FF"))


#Parameters
mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(70, "cm")
width_image = 40
height_image = 80
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 10),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        #column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### B cells

```{r}
####### Input
#Troque pelo geneset de interesse

heatmap_data = Immune_GO_genes_Bcells
file = "Immune_GO_genes_Bcells"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Converter para wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#4DBBD5FF", "white", "#DC0000FF"))


#Parameters
mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(50, "cm")
width_image = 40
height_image = 60
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 10),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        #column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### Ig mediated response

```{r}
####### Input
#Troque pelo geneset de interesse

heatmap_data = Immune_GO_genes_Ig
file = "Immune_GO_genes_Ig"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Converter para wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, set, disease_vac, type, dose, day), by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(day) %>% 
  column_to_rownames(var= "condition") 

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(day) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(set, disease_vac, dose, type, day)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#4DBBD5FF", "white", "#DC0000FF"))


#Parameters
mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(50, "cm")
width_image = 40
height_image = 60
column_names_gp = gpar(fontsize = 7)
row_names_gp = gpar(fontsize = 10)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = F,
                        #column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

## All genes
### Immune genes
Process data

```{r }
############## Process data

# All gene sets
AllGenes = data %>% 
  full_join(immunego_genes, by = "genes") %>% 
  merge(ann_vaccines_covid_other_conditions, by = "condition", all.y=F) %>% 
  clean_names()

```

```{r}
library(InteractiveComplexHeatmap)
library(maditr)

####### Input
heatmap_data = AllGenes
file = "AllGenes_Covid_13Vaccines"

###### Process data

heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA") %>% 
  group_by(genes) %>%
  filter(n_distinct(condition) >= 15 & dose != 0) %>%
  ungroup() %>% 
  select(genes, condition, log2fold_change) 

# Convert to wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

matrix_data_ready =  matrix_data %>% 
  round(2)

# Check if all vaccines are in the matrix and join annotations
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines_covid_other_conditions %>% select(condition, set, disease_vac, type, week, day), by = "condition")

# Join annotations
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(day) %>% 
  column_to_rownames(var= "condition") 


#Rows Immune ----
ann_rows_immune = matrix_data %>%
  as.data.frame() %>%
  rownames_to_column("genes") %>%
  left_join(ImmuneGO_Annotated_Genes %>% select(genes) %>% distinct() %>% mutate(Immune = "ImmuneGO"), by = "genes") %>%
  left_join(btm_annotation_genes_immune %>% select(genes) %>% distinct() %>% mutate(Immune2 = "BTM"), by = "genes") %>%
  select(genes, Immune, Immune2) %>%
  distinct() %>%
  group_by(genes) %>%
  arrange(genes, factor(Immune, levels = c("ImmuneGO", "BTM")), .by_group = TRUE) %>%
  column_to_rownames("genes") %>%
  as.data.frame() %>%
  mutate(Immune = if_else(is.na(Immune), Immune2, Immune),
         Immune2 = if_else(Immune == Immune2, NA, Immune2)) %>% 
  replace(is.na(.), "None")

# Reorder columns (same order as lines in the annotation table)
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(day) %>% 
  select(!c(set, disease_vac, type, week, day)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0) %>% 
  as.matrix() 


#Check dimensions
dim(matrix_data_ordered) 
dim(ann_vaccines_heatmap) 
dim(ann_rows_immune)


col_immune_sets = list(
  "Immune" = c("BTM" = "#5e60ce",
          "ImmuneGO" = "#00A087FF",
          "None"="white"),
  "Immune2" = c("BTM" = "#5e60ce",
          "ImmuneGO" = "#00A087FF",
          "None"="white"))

```


```{r}
#Create heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_vaxsig_colors, 
                       annotation_name_side = "left")

row_ha = HeatmapAnnotation(df = ann_rows_immune,
                       col = col_immune_sets,
                       which= "row",
                       simple_anno_size = unit(2, "mm"),
                       show_annotation_name = F,
                       show_legend = T)

col_fun <- colorRamp2(c(-1, 0, 1), 
                      c("#4DBBD5FF", "white", "#DC0000FF"))

#Parameters
mat = matrix_data_ordered
width_image = 50
height_image = 30
width = unit(40, "cm")
height = unit(20, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plot and save
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "L2FC", #Legend
                        col = col_fun, cluster_columns = F,
                        column_names_rot = 45,
                        show_row_dend = F,
                        row_split = 2, #Split group clusters
                        show_row_names = F,
                        column_names_gp = column_names_gp,
                        width = width,
                        height = height,
                        row_names_gp = row_names_gp)

ht_list = draw(heatmap_plot_all, 
               annotation_legend_side = "right", 
               heatmap_legend_side = "right")


draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

Interactive
```{r}
#Parameters
mat = matrix_data_ordered
width_image = 40
height_image = 90
width = unit(30, "cm")
height = unit(80, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "L2FC", #Legend
                        col = col_fun,
                        column_names_rot = 45,
                        cluster_columns = F,
                        show_row_dend = F,
                        row_split = NULL, #Split group clusters
                        show_row_names = F,
                        column_names_gp = column_names_gp,
                        row_names_gp = row_names_gp)

ht_list = draw(heatmap_plot_all, 
               annotation_legend_side = "right", 
               heatmap_legend_side = "right")
htShiny(ht_list)

htShiny(ht_list, save = "Interactive_Covidand13Vax_Heatmap")

```

#### Get Cluster 1 Immune genes
Get row clusters

```{r}
r.dend <- row_dend(heatmap_plot_all)  #Extract row dendrogram
rcl.list <- row_order(heatmap_plot_all)  #Extract clusters (output is a list)
lapply(rcl.list, function(x) length(x))  #check/confirm size cluster

gene_cluster <- NULL
for (i in 1:length(rcl.list)) {
  a <- row.names(mat[row_order(heatmap_plot_all)[[i]], ])
  gene_cluster <- rbind(gene_cluster, data.frame(genes = a, cluster = paste0("cluster ", i), gene_set = "All genes, shared > 16 conditions"))
}
write.csv(gene_cluster, file = here("DGE analysis", paste0(file, "_clustered_genes.csv")), row.names = F)
```

```{r}
####### INPUT
data_2 = heatmap_data %>% 
  inner_join(gene_cluster %>% filter(cluster == "cluster 1") %>% select(genes), by = "genes")
filename_2 = paste0("Immune_GO_genes_cluster1", file)

# Convert to wide table
matrix_data <- dcast(data_2, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

matrix_data_ready =  matrix_data %>% 
  round(2)

# Check if all vaccines are in the matrix and join annotations
ann_vaccines_complex = data_2 %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines_covid_other_conditions %>% select(condition, set, disease_vac, type, week, day), by = "condition")

# Join annotations
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(day) %>% 
  column_to_rownames(var= "condition") 


#Rows Immune ----
ann_rows_immune = matrix_data %>%
  as.data.frame() %>%
  rownames_to_column("genes") %>%
  left_join(ImmuneGO_Annotated_Genes %>% select(genes) %>% distinct() %>% mutate(set = "ImmuneGO"), by = "genes") %>%
  left_join(btm_annotation_genes_immune %>% select(genes) %>% distinct() %>% mutate(set2 = "BTM"), by = "genes") %>%
  select(genes, set, set2) %>%
  distinct() %>%
  group_by(genes) %>%
  arrange(genes, factor(set, levels = c("ImmuneGO", "BTM")), .by_group = TRUE) %>%
  column_to_rownames("genes") %>%
  as.data.frame() %>%
  mutate(set = if_else(is.na(set), set2, set),
         set2 = if_else(set == set2, NA, set2)) %>% 
  replace(is.na(.), "None")

# Reorder columns (same order as lines in the annotation table)
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(day) %>% 
  select(!c(set, disease_vac, type, week, day)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0) %>% 
  as.matrix() 


#Check dimensions
dim(matrix_data_ordered) 
dim(ann_vaccines_heatmap) 
dim(ann_rows_immune)


col_immune_sets = list(
  "set" = c("BTM" = "#F5BB00",
          "ImmuneGO" = "#7E6148FF",
          "None"="white"),
  "set2" = c("BTM" = "#F5BB00",
          "ImmuneGO" = "#7E6148FF",
          "None"="white"))


#Create heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_vaxsig_colors, 
                       annotation_name_side = "left")

row_ha = HeatmapAnnotation(df = ann_rows_immune,
                       col = col_immune_sets,
                       which= "row",
                       simple_anno_size = unit(2, "mm"),
                       show_annotation_name = F,
                       show_legend = T)

col_fun <- colorRamp2(c(-1, 0, 1), 
                      c("#9bc1bc", "white", "#ed6a5a"))

#Parameters
mat = matrix_data_ordered
width_image = 50
height_image = 30
width = unit(40, "cm")
height = unit(20, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plot and save
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Cluster1_BYDAY_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "L2FC", #Legend
                        col = col_fun, 
                        cluster_columns = F,
                        column_names_rot = 45,
                        show_row_dend = T,
                        row_split = 2, #Split group clusters
                        show_row_names = T,
                        column_names_gp = column_names_gp,
                        width = width,
                        height = height,
                        row_names_gp = row_names_gp)

ht_list = draw(heatmap_plot_all, 
               annotation_legend_side = "right", 
               heatmap_legend_side = "right")


draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()

```

By type and day
```{r}
####### INPUT
data_2 = heatmap_data %>% 
  inner_join(gene_cluster %>% filter(cluster == "cluster 1") %>% select(genes), by = "genes")
filename_2 = paste0("Immune_GO_genes_cluster1", file)

# Convert to wide table
matrix_data <- dcast(data_2, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

matrix_data_ready =  matrix_data %>% 
  round(2)

# Check if all vaccines are in the matrix and join annotations
ann_vaccines_complex = data_2 %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines_covid_other_conditions %>% select(condition, set, disease_vac, type, week, day), by = "condition")

# Join annotations
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(type, condition) %>%  
  column_to_rownames(var= "condition") 


#Rows Immune ----
ann_rows_immune = matrix_data %>%
  as.data.frame() %>%
  rownames_to_column("genes") %>%
  left_join(ImmuneGO_Annotated_Genes %>% select(genes) %>% distinct() %>% mutate(set = "ImmuneGO"), by = "genes") %>%
  left_join(btm_annotation_genes_immune %>% select(genes) %>% distinct() %>% mutate(set2 = "BTM"), by = "genes") %>%
  select(genes, set, set2) %>%
  distinct() %>%
  group_by(genes) %>%
  arrange(genes, factor(set, levels = c("ImmuneGO", "BTM")), .by_group = TRUE) %>%
  column_to_rownames("genes") %>%
  as.data.frame() %>%
  mutate(set = if_else(is.na(set), set2, set),
         set2 = if_else(set == set2, NA, set2)) %>% 
  replace(is.na(.), "None")

# Reorder columns (same order as lines in the annotation table)
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(type, condition) %>% 
  select(!c(set, disease_vac, type, week, day)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0) %>% 
  as.matrix() 


#Check dimensions
dim(matrix_data_ordered) 
dim(ann_vaccines_heatmap) 
dim(ann_rows_immune)


col_immune_sets = list(
  "set" = c("BTM" = "#F5BB00",
          "ImmuneGO" = "#7E6148FF",
          "None"="white"),
  "set2" = c("BTM" = "#F5BB00",
          "ImmuneGO" = "#7E6148FF",
          "None"="white"))



#Create heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_vaxsig_colors, 
                       annotation_name_side = "left")

row_ha = HeatmapAnnotation(df = ann_rows_immune,
                       col = col_immune_sets,
                       which= "row",
                       simple_anno_size = unit(2, "mm"),
                       show_annotation_name = F,
                       show_legend = T)

col_fun <- colorRamp2(c(-2, 0, 2), 
                      c("#9bc1bc", "white", "#ed6a5a"))

#Parameters
mat = matrix_data_ordered
width_image = 50
height_image = 30
width = unit(40, "cm")
height = unit(20, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plot and save
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Cluster1__BYType_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "L2FC", #Legend
                        col = col_fun, 
                        cluster_columns = F,
                        column_names_rot = 45,
                        show_row_dend = T,
                        row_split = NULL, #Split group clusters
                        show_row_names = T,
                        column_names_gp = column_names_gp,
                        width = width,
                        height = height,
                        row_names_gp = row_names_gp)

ht_list = draw(heatmap_plot_all, 
               annotation_legend_side = "right", 
               heatmap_legend_side = "right")


draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()

```




# MSigDB Vax: COVID-19 and other vaccines

```{r}
# Input files ------
immunego = ImmuneGO_Annotated_Genes # ImmuneGO Annotation
degs_updown_p_05_vac_infected = readRDS(here("DGE analysis", "all_studies_degs_weighted.rds")) # DEGs
ann_vaccines_covid_other = read_csv(here("Annotations and metadata", "ann_vaccines_vaxsig_covid.csv")) #Conditions annotation
VAX_Genes_Annotated_RAW_2 = read_csv(here("VaxGO gene sets", "VaxSigDB_Gene_sets_Annotated_RAW.csv")) %>% 
  separate_rows(GENE_SYMBOLS)
ann_vaccines = read_csv(here("Annotations and metadata", "ann_vaccines_conditions.csv"))


# Processing
VAX_Genes_Annotated_RAW_3 = VAX_Genes_Annotated_RAW_2 %>% 
  clean_names() %>% 
  filter(study_subtype == "VAC ONLY",
         sample_source == "PBMC") %>% 
  mutate(condition = paste0(vaccine, " (", date_time, ")")) %>% 
  select(condition, everything())

degs_all_updown_vax = VAX_Genes_Annotated_RAW_3 %>%
  clean_names() %>% 
  mutate(condition = paste0(vaccine, " (", date_time, ")")) %>% 
  select(condition, genes = gene_symbols) %>% 
  distinct()

degs_all_updown_covid= degs_updown_p_05_vac_infected %>% 
  as.data.frame() %>% 
  select(genes, condition:direction) %>% 
  filter(direction %in% c("UP", "DOWN")) %>% 
  inner_join(ann_vaccines, by = "condition") %>%
  distinct() %>% 
  select(condition, genes)

degs_all_updown_covid_other = bind_rows(degs_all_updown_vax, degs_all_updown_covid) %>% 
  distinct()


degs_all_updown_covid_other_immune = degs_all_updown_covid_other %>% 
  inner_join(ImmuneGO_Annotated_Genes %>% select(genes) %>% distinct(), by = "genes") %>% 
  distinct()

degs_all_updown_covid_other_nonimmune = degs_all_updown_covid_other %>% 
  anti_join(ImmuneGO_Annotated_Genes %>% select(genes) %>% distinct(), by = "genes") %>% 
  distinct()

write.csv(degs_all_updown_covid_other, file = here("VaxGO gene sets", "degs_all_updown_covid_vax_other.csv"))
write.csv(degs_all_updown_covid_other_immune, file = here("VaxGO gene sets","degs_all_updown_covid_vax_other_immune.csv"))
write.csv(degs_all_updown_covid_other_nonimmune, file = here("VaxGO gene sets","degs_all_updown_covid_vax_other_nonimmune.csv"))


degs_covid = degs_updown_p_05_vac_infected %>% 
  select(condition, genes, direction) %>% 
  filter(direction != "NEUTRAL") %>% 
  mutate(regulation_numeric = if_else(direction == "UP", "1", "-1"),
         regulation_numeric = as.numeric(regulation_numeric))


degs_vax = VAX_Genes_Annotated_RAW_2 %>% 
  filter(STUDY_SUBTYPE == "VAC ONLY") %>% 
  mutate(condition_vaccine = paste0(VACCINE, " (", `DATE-TIME`, ")")) %>% 
  select(condition_vaccine, SYSTEMATIC_NAME, VACCINE, everything()) %>% 
  separate_rows(GENE_SYMBOLS, sep = ",")
degs_vax %>% saveRDS(file = here("VaxGO gene sets","VAXSigDB_Genes.rds"))


degs_vaxsig = degs_vax %>% 
  filter(STUDY_SUBTYPE == "VAC ONLY") %>% 
  select(condition = condition_vaccine, genes = GENE_SYMBOLS, regulation = REGULATION) %>% 
  mutate(regulation_numeric = if_else(regulation == "UP", 1, -1))


degs_covid_vax = degs_covid %>% 
  bind_rows(degs_vaxsig) %>% 
  select(-regulation) %>% 
  inner_join(ImmuneGO_Annotated_Genes %>%
               filter(process != c("TCR REPERTOIRE", "BCR REPERTOIRE")) %>% 
               select(genes) %>%
               distinct(), 
             by = "genes")

degs_covid_vax %>% 
  saveRDS(file = here("VaxGO gene sets", "degs_covid_vaxsig_up_down_immune.rds"))

```

### Gene overlap analysis

```{r}
#Input
data = degs_all_updown_covid_other_immune %>% 
  rename(process = condition) %>% 
  select(process, genes)

filename = "degs_all_updown_covid_other_immune"
  
```

Function for overlapping
```{r}

# Function for overlapping -------
overlap_genes <- function(cond1, cond2, data) {
  genes_cond1 <- data$genes[data$process == cond1]
  genes_cond2 <- data$genes[data$process == cond2]
  
  genes_shared <- intersect(genes_cond1, genes_cond2)
  
  genes_notshared_cond1 <- setdiff(genes_cond1, genes_cond2)
  genes_notshared_cond2 <- setdiff(genes_cond2, genes_cond1)
  
  total_genes_cond1 <- length(genes_cond1)
  total_genes_cond2 <- length(genes_cond2)
  
  percentage_shared_cond1 <- length(genes_shared) / total_genes_cond1 * 100
  percentage_shared_cond2 <- length(genes_shared) / total_genes_cond2 * 100
  
  shared_genes <- data.frame(
    Cond1 = cond1,
    Cond2 = cond2,
    Shared = length(genes_shared),
    NotShared_cond1 = length(genes_notshared_cond1),
    NotShared_cond2 = length(genes_notshared_cond2),
    Total_Genes_Cond1 = total_genes_cond1,
    Total_Genes_Cond2 = total_genes_cond2,
    Genes_Names = paste(genes_shared, collapse = ", "),
    Percentage_Shared_Cond1 = percentage_shared_cond1,
    Percentage_Shared_Cond2 = percentage_shared_cond2
  )
  
  return(shared_genes)
}
conflicted::conflicts_prefer(base::intersect)
conflicted::conflicts_prefer(base::setdiff)
conflicted::conflicts_prefer(janitor::fisher.test)
```

Perform calculations
```{r}
# Perform calculations
unique_cond <- unique(data$process) # List sets
shared_genes_df <- data.frame() # Store data

for (i in 1:(length(unique_cond) - 1)) {
  for (j in (i + 1):length(unique_cond)) {
    resultado_temp <- overlap_genes(unique_cond[i], unique_cond[j], data)
    shared_genes_df <- bind_rows(shared_genes_df, resultado_temp)
  }
}

# Fisher's exact test
fisher_exact_test <- function(shared, notshared1, notshared2) {
  cont_table <- matrix(c(shared, notshared1, notshared2, 0), nrow = 2)
  results_fisher <- fisher.test(cont_table)
  return(results_fisher$p.value)
}
shared_genes_df = shared_genes_df %>%
  mutate(pvalue = pmap_dbl(list(Shared, NotShared_cond1, NotShared_cond2), fisher_exact_test)) %>% 
  mutate("-log(p)" = -log10(pvalue),
         padj = p.adjust(pvalue))


#Mirror table
shared_genes_df2 = shared_genes_df %>% 
  rename(Cond1_temp = Cond1, Cond2_temp = Cond2) %>%
  rename(Cond1 = Cond2_temp, Cond2 = Cond1_temp)

shared_genes_df_mirror = shared_genes_df %>% 
  bind_rows(shared_genes_df2) %>% 
  mutate(Percentage_Shared_Cond1 = round(Percentage_Shared_Cond1, 2),
         Percentage_Shared_Cond2 = round(Percentage_Shared_Cond2, 2))
  
#Jaccard distance
jaccard.matrix <- data %>%
  mutate(present = 1) %>% 
  distinct() %>% 
  pivot_wider(names_from = genes, values_from = present, values_fill = 0) %>%
  column_to_rownames(var = "process") %>%
  vegdist(binary = TRUE, method = "jaccard", diag = TRUE, upper = TRUE) %>%
  as.matrix() %>%
  {1 - .}

shared_genes_df_mirror = jaccard.matrix %>% 
  as.data.frame() %>% 
  rownames_to_column("Cond1") %>% 
  pivot_longer(cols = -"Cond1",
               names_to = "Cond2",
               values_to = "jaccard_distance") %>% 
  inner_join(shared_genes_df_mirror, by = join_by(Cond1, Cond2)) %>% 
  mutate(jaccard_distance = round(jaccard_distance, 3))

#Save
write_csv(shared_genes_df_mirror, 
          file = here("VaxGO gene sets", paste0(filename, "_sharedgenes_Fisher_Jaccard.csv")))
```


### Other comparisons

```{r}
############# VACCINE
#Use "vax_degs_up_VAXSigDB_sharedgenes_ann.csv" for comparing UP degs with UP degs from other conditions

#VAC ONLY
shared_genes_ann_vac = shared_genes_df_mirror %>%
  select(vac_condition, vaccine_condition, vac_ref, Cond2:sample_source, pmid, gene_symbols:participants) %>% 
  filter(study_type == "VACCINE", 
         sample_source == "PBMC",
         !(vaccine_condition %in% c("BBIBP", "ZF2001"))) 

vac_only = shared_genes_ann_vac %>%
  filter(study_subtype == "VAC ONLY")

```

### Heatmap

```{r}
# INPUT ------

degs_all_updown_covid_other_immune_shared = read_csv(here("VaxGO gene sets", "degs_all_updown_covid_other_immune_sharedgenes_Fisher_Jaccard.csv"))

table = degs_all_updown_covid_other_immune_shared
filename = "ImmuneGO_Genes_JaccardDistance" 

data = table %>% 
  filter(pvalue <0.25) %>% 
  inner_join(ann_vaccines_covid_other %>% 
               select(Cond1 = condition, everything()), by = "Cond1") %>% 
  inner_join(ann_vaccines %>% 
               select(Cond2 = condition, everything()), by = "Cond2") %>% 
  clean_names() %>% 
  rename(vac_ref = cond1, vac_condition = cond2)

# Matrix ------
matrix = data %>%
  filter(pvalue <= 0.10) %>% 
  select(vac_condition, vac_ref, jaccard_distance) %>% 
  pivot_wider(names_from = vac_condition, values_from = jaccard_distance, values_fn = mean) %>% 
  replace(is.na(.), 0) %>%
  column_to_rownames(var="vac_ref") %>%
  as.matrix()

matrix = matrix*100

# Annotations -------
ann_vaccines_covid = ann_vaccines_covid_other %>% 
  filter(covid_other == "COVID")
ann_vaccines_other = ann_vaccines_covid_other %>% 
  filter(covid_other == "OTHER")

# Columns
ann_cols_heatmap = matrix %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "process") %>% 
  rename(vac_condition = '.') %>% 
  merge(ann_vaccines_covid, by.x = "vac_condition", by.y = "condition", all.y = F) %>% 
  select(vac_condition, type, vaccine, week) %>% #ordenar colunas para anotação
  distinct() %>% 
  column_to_rownames(var = "vac_condition") %>% 
  mutate_all(as.factor)

# Rows
ann_rows_heatmap = matrix %>%  
  as.data.frame() %>% 
  rownames_to_column("condition") %>% 
  distinct() %>% 
  inner_join(ann_vaccines_other, by = "condition") %>% 
  mutate_all(as.factor) %>%
  arrange(condition) %>%
  distinct() %>% 
  column_to_rownames(var= "condition") %>% 
  select(vaccine:infection) %>% 
  select(microbe_type, type, week)

# Arrange cols and rows in matrix
matrix_data = matrix %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>%
  rownames_to_column("condition") %>% 
  inner_join(ann_rows_heatmap %>% 
               rownames_to_column("condition") %>% 
               select(condition),
             by = "condition") %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  mutate_if(is.character, as.numeric) %>%
  replace(is.na(.), 0) %>% 
  as.matrix()

#Check dimensions
dim(matrix_data)
dim(ann_rows_heatmap)
dim(ann_cols_heatmap)

```

Plot
```{r}

#Heatmap annotation
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "left")

row_ha = HeatmapAnnotation(df = ann_rows_heatmap,
                       col = ann_vaxsig_colors,
                       which= "row",
                       show_annotation_name = F,
                       show_legend = F)

# Values colors
col_fun <- colorRamp2(c(0, max(matrix_data)), c("white", "#ed6a5a"))


mat = matrix_data
width_image = 30
height_image = 25
width = unit(20, "cm")
height = unit(15, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)
file = paste0("VAXSig_", filename)

fileimageheatmap_grouped = paste0(filename, sep ="_", "Complexheatmap_ALL_CLUSTERED.png")

png(file = here("Figures", fileimageheatmap_grouped), 
    width = width_image, 
    height = height_image,
    units="cm", res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "left",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(title = "Jaccard distance", direction = "vertical"),
                        col = col_fun, #color
                        cell_fun = function(j, i, x, y, width, height, fill) {
                                            if(mat[i, j] > 0)
                                            grid.text(sprintf("%.f", mat[i, j]), 
                                                      x, y, 
                                                      gp = gpar(fontsize = 8))},
                        row_title = "",
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp, #Grande = 6, Pequeno 15
                        row_dend_width = unit(5, "cm"), #Altura do dendrograma
                        column_split = NULL, #Split column clusters 
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, # 20 para pequeno e grande
                        height = height # 20 para pequeno, 60 para grande
)

draw(heatmap_plot,
     merge_legend = TRUE,
     annotation_legend_side = "right", 
     heatmap_legend_side = "right")
dev.off()

```

##ORA COVID vs. VAXGO with ImmuneGO

```{r}
#Input
degs_all_updown_covid = degs_updown_p_05_vac_infected %>% 
   as.data.frame() %>% 
   select(genes, condition:direction) %>% 
   filter(direction %in% c("UP", "DOWN")) %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
   distinct() %>% 
  select(condition, genes)

VAX_Genes_Annotated_RAW = VAX_Genes_Annotated_RAW_2 %>% 
  clean_names() %>% 
  mutate(condition = paste0(vaccine, " (", date_time, ")"))

ann_vaccines_covid_other 

degs_all_updown_vax_immunego = VAX_Genes_Annotated_RAW %>%
  clean_names() %>% 
  mutate(condition = paste0(vaccine, " (", date_time, ")")) %>% 
  select(condition, genes = gene_symbols) %>% 
  inner_join(ImmuneGO_Annotated_Genes %>% 
               select(genes), by = "genes") %>% 
  distinct()

degs_all_updown_covid_other = bind_rows(degs_all_updown_vax_immunego, degs_all_updown_covid) %>% 
  distinct()

Immune_GO_T2G = ImmuneGO_Annotated_Genes %>% 
  filter(go_term == "Manual",
         !process %in% c("TCR REPERTOIRE", "BCR REPERTOIRE", "ADAPTIVE IMMUNE SYSTEM", "INNATE IMMUNE SYSTEM")) %>% 
  select(process, genes)
```

```{r}
# ORA function
perform_ORA <- function(df, Immune_GO_T2G) {
  resultados <- list()
  
  # Obter condições únicas na coluna "condition"
  condicoes <- df$condition %>% 
    unique() %>% 
    as.character()
  
  for (condicao in condicoes) {
    # Selecionar DEGs para a condição atual
    degs_condicao <- df %>% 
      filter(condition == condicao) %>%
      select(genes) %>%
      distinct()

    # Preparar gene list
    gene_list_ora <- degs_condicao$genes

    ############ IMMUNE GO
    immune_GO_ORA <- tryCatch({
      enricher(gene_list_ora, 
               TERM2GENE = Immune_GO_T2G, 
               minGSSize = 1,
               maxGSSize = 1000,
               pvalueCutoff = 0.25,
               pAdjustMethod = "BH") %>% 
        as.data.frame() %>% 
        arrange(qvalue) %>% 
        mutate(condition = condicao,
               ora_enrichment = "ImmuneGO")
    }, error = function(e) {
      return(NULL)  # Retorna NULL caso ocorra o erro
    })

    if (!is.null(immune_GO_ORA)) {
      resultados[[paste(condicao, "ImmuneGO", sep = "_")]] <- immune_GO_ORA
    }
  }

  # Combinar todos os resultados em um único dataframe
  resultado_final <- bind_rows(resultados)

  return(resultado_final)
}

df_resultados <- perform_ORA(degs_all_updown_covid_other, Immune_GO_T2G)
print(df_resultados)

#Anotar
immune_GO_ORA_covid_other = df_resultados %>% 
  clean_names() %>% 
  select(process = id, gene_ratio:ora_enrichment) %>% 
  inner_join(ann_vaccines_covid_other, by = "condition") %>% 
  distinct() %>% 
  mutate(log_q = -log10(qvalue))

write.csv(immune_GO_ORA_covid_other, file = here("VaxGO gene sets", "COVID_VAX_immune_GO_ORA.csv"))

```

#### Heatmap

```{r}
####### INPUT
data_2 = immune_GO_ORA_covid_other
  
filename_2 = "immune_GO_ORA_covid_other"

######## Matrix
matrix = data_2 %>% 
  filter(qvalue <= 0.25) %>% 
  select(condition, process, log_q) %>% #Percentage_Shared_Cond1
  pivot_wider(names_from = "condition", 
              values_from = "log_q",  #Percentage_Shared_Cond1
              values_fn = mean) %>% 
  replace(is.na(.), 0) %>% 
  column_to_rownames(var="process") %>% 
  as.matrix() %>% 
  round(2)

######## Annotations
ann_cols = matrix %>% 
  as.data.frame() %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column("condition") %>% 
  select(condition) %>% 
  inner_join(ann_vaccines_covid_other, by = "condition") %>% 
  select(condition, covid_other, microbe_type, disease_vac, type, week) %>% 
  distinct() %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") 

matrix_data = matrix %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame()%>%  
  mutate_if(is.character, as.numeric) %>%
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Check dimensions
dim(matrix_data)
dim(ann_cols)
```

```{r}
################# Plotar

#Heatmap annotation
ha = HeatmapAnnotation(df = ann_cols, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "right")


# Values colors
col_fun <- colorRamp2(c(0, 2), c("white", "#ED6A5A"))


# Parameters
file = filename_2
mat = matrix_data
width_image = 40
height_image = 20
width = unit(25, "cm")
height = unit(7, "cm")
column_names_gp = gpar(fontsize = 9)
row_names_gp = gpar(fontsize = 9)
rect_gp = gpar(col = "gray90", lwd = 0.5)

###### CLUSTERIZED ALL VS ALL

fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap.png")

png(file = here("Figures", fileimageheatmap_grouped), width = width_image, height = height_image, units="cm", res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        row_names_side = "left",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical",
                                                    title = "-log(FDR)"),
                        col = col_fun, #color
                        cell_fun = function(j, i, x, y, w, h, fill) {
                              if (mat[i, j] >= 2) {
                                grid.text("***", x, y, gp = gpar(fontsize = 8))
                              } else if (mat[i, j] > 1.5) {
                                grid.text("**", x, y, gp = gpar(fontsize = 9))
                              } else if (mat[i, j] >= 1) {
                                grid.text("*", x, y, gp = gpar(fontsize = 8))
                              } else {
                                grid.text("", x, y)
                              }
                            },
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp, #Grande = 6, Pequeno 15
                        row_dend_width = unit(5, "cm"), #Altura do dendrograma
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        width = width, # 20 para pequeno e grande
                        height = height # 20 para pequeno, 60 para grande
)

draw(heatmap_plot, 
     annotation_legend_side = "left", 
     heatmap_legend_side = "left",
          merge_legend = TRUE)
dev.off()


```

## COVID-19 genes vs. VaxGO genes with ImmuneGO genes

```{r}
# VAX Genes  ------
VAX_Genes_Annotated_RAW = VAX_Genes_Annotated_RAW %>% 
  clean_names() %>% 
  mutate(condition = paste0(vaccine, " (", date_time, ")"))

#ImmuneGO
degs_all_updown_vax_immunego = VAX_Genes_Annotated_RAW %>%
  clean_names() %>% 
  mutate(condition = paste0(vaccine, " (", date_time, ")")) %>% 
  select(condition, genes = gene_symbols, regulation) %>% 
  inner_join(ImmuneGO_Annotated_Genes %>% 
               select(genes), by = "genes") %>% 
  distinct()

# Covid-19 conditions ----
degs_all_updown = degs_updown_p_05_vac_infected

#Immune
degs_all_updown_covid_immune = degs_all_updown %>% 
  as.data.frame() %>% 
  select(genes,  condition:fdr) %>% 
  mutate(direction = case_when(log2fold_change >= 1 ~ "UP",
                               log2fold_change <= -1 ~ "DOWN",
                               TRUE ~ "NEUTRAL")) %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  inner_join(ImmuneGO_Annotated_Genes %>% select(genes) %>% distinct(), by = "genes") %>% 
  distinct() %>% 
  filter(vaccine != "I", 
         gse_id != "GSE189039") %>% 
  select(condition, genes, regulation = direction)  %>% 
  distinct()


# Bind Immune ------
degs_all_updown_covid_other = bind_rows(degs_all_updown_vax_immunego, degs_all_updown_covid_immune) %>% 
  distinct()


covid_other_immunego_genes = degs_all_updown_covid_other %>% 
  distinct() %>% 
  filter(regulation %in% c("UP", "DOWN")) %>% 
  mutate(regulation_numeric = case_when(
    regulation %in% c("UP") ~ 1,
    regulation %in% c("DOWN") ~ -1,
    TRUE ~ as.numeric(regulation)
  )) %>% 
  distinct()

saveRDS(covid_other_immunego_genes, file = "covid_other_immunego_genes.rds")

```

```{r}
###### Gene sets Immune GO ------

Immune_GO_genes_All = covid_other_immunego_genes 

# Immune_GO_genes_Innate = Immune_GO_genes_All %>%
#    filter(process == "INNATE IMMUNE SYSTEM")
# Immune_GO_genes_Adaptive = Immune_GO_genes_All %>%
#    filter(process == "ADAPTIVE IMMUNE SYSTEM")
# Immune_GO_genes_Complement = Immune_GO_genes_All %>%
#    filter(process == "COMPLEMENT IMMUNE SYSTEM")
# 
# # Sistema inato
# Immune_GO_genes_Inflammation = Immune_GO_genes_All %>%
#   filter(process == "INFLAMMATION")
# Immune_GO_genes_ARPP = Immune_GO_genes_All %>%
#   filter(process == "ARPP")
# Immune_GO_genes_IFN = Immune_GO_genes_All %>%
#   filter(process == "ANTIVIRAL AND INTERFERON")
# Immune_GO_genes_Neutrophil = Immune_GO_genes_All %>%
#   filter(process == "NEUTROPHIL")
# Immune_GO_genes_Eosinophil = Immune_GO_genes_All %>%
#   filter(process == "EOSINOPHIL")
# Immune_GO_genes_Complement = Immune_GO_genes_All %>%
#   filter(process == "COMPLEMENT IMMUNE SYSTEM")
# Immune_GO_genes_Monocytes = Immune_GO_genes_All %>%
#   filter(process == "MONOCYTES")
# Immune_GO_genes_DC = Immune_GO_genes_All %>%
#   filter(process == "DENDRITIC CELLS")
# Immune_GO_genes_Macro = Immune_GO_genes_All %>%
#   filter(process == "MACROPHAGES")
# Immune_GO_genes_NK = Immune_GO_genes_All %>%
#   filter(process == "NK CELL")
# 
# # Sistema adaptativo
# Immune_GO_genes_Cellular = Immune_GO_genes_All %>%
#   filter(process == "CELLULAR ADAPTIVE IMMUNE SYSTEM")
# Immune_GO_genes_Humoral = Immune_GO_genes_All %>%
#   filter(process == "HUMORAL ADAPTIVE IMMUNE SYSTEM")
# Immune_GO_genes_Tcells = Immune_GO_genes_All %>%
#   filter(process == "T CELLS")
# Immune_GO_genes_Bcells = Immune_GO_genes_All %>%
#   filter(process == "B CELLS")
# Immune_GO_genes_Ig = Immune_GO_genes_All %>%
#   filter(process == "IMMUNOGLOBULIN MEDIATED IMMUNE RESPONSE")

```

#### Heatmap

```{r}
####### INPUT
data_2 = covid_other_immunego_genes
filename_2 = paste0("Immune_GO_genes_All", "_VAX_COVID")

######## Matrix
matrix = data_2 %>% 
  select(genes, condition, regulation_numeric) %>%
  pivot_wider(names_from = "condition", 
              values_from = "regulation_numeric", 
              values_fn = mean) %>% 
  replace(is.na(.), 0) %>%
  arrange(genes) %>% 
  column_to_rownames(var="genes") %>% 
  as.matrix()

######## Annotations
ann_vaccines_covid_other_2 = data_2 %>% 
  inner_join(ann_vaccines_covid_other, by = "condition") %>% 
  select(condition, covid_other, microbe_type, disease_vac, type, week) %>% 
  distinct()

#Columns
ann_cols_heatmap = matrix %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  inner_join(ann_vaccines_covid_other_2, by = "condition") %>% 
  distinct() %>% 
  arrange(condition) %>% 
  column_to_rownames("condition")

matrix_data = matrix %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  inner_join(ann_cols_heatmap %>% rownames_to_column("condition"), by = "condition") %>% 
  select(!c(covid_other:week)) %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  filter(rowSums(. != 0) > 1) %>% 
  mutate_if(is.character, as.numeric) %>%
  as.matrix() 

#Rows Immune ----
ann_rows_immune = matrix_data %>%
  as.data.frame() %>%
  rownames_to_column("genes") %>%
  left_join(ImmuneGO_Annotated_Genes %>%
              filter(go_term == "Manual",
                     !process %in% c("TCR REPERTOIRE", "BCR REPERTOIRE")), by = "genes") %>%
  select(genes, process) %>%
  distinct() %>%
  group_by(genes) %>%
  arrange(genes, factor(process, levels = c("INNATE IMMUNE SYSTEM", "ADAPTIVE IMMUNE SYSTEM")), .by_group = TRUE) %>%
  mutate(i_process = row_number()) %>%
  pivot_wider(., names_from = "i_process", values_from = "process") %>%
  column_to_rownames("genes") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("index") %>%
  mutate(index = as.numeric(index)) %>%
  arrange(desc(index)) %>%
  column_to_rownames("index") %>%
  t() %>%
  as.data.frame() %>%
  replace(is.na(.), ".")


#Check dimensions
dim(matrix_data)
dim(ann_cols_heatmap)
dim(ann_rows_immune)

```

Sizes: 1. Cellular adaptive immune system: width = unit(25, "cm"),
height = unit(60, "cm") #254 genes 2. Humoral adaptive immune system:
width = unit(25, "cm"), height = unit(30, "cm") #51 genes 3. T cells:
width = unit(25, "cm"), height = unit(60, "cm") #221 genes 4. B cells:
width = unit(25, "cm"), height = unit(50, "cm") #146 genes 5. Ig
mediated: width = unit(25, "cm"), height = unit(10, "cm") #146 genes 6.
Inflammation: width = unit(25, "cm"), height = unit(10, "cm") 7. ARPP:
width = unit(25, "cm"), height = unit(30, "cm") 60 genes 8. IFN: width =
unit(25, "cm"), height = unit(30, "cm") 93 genes 9. DC: width = unit(25,
"cm"), height = unit(20, "cm") #29 genes 10. Macrophages: width =
unit(25, "cm"), height = unit(25, "cm") \# 52 genes 11. Neutrophils:
width = unit(25, "cm"), height = unit(40, "cm") \# 70 genes 12.
Eosinophil: width = unit(25, "cm"), height = unit(10, "cm") \# 42 genes
13. Complement: width = unit(25, "cm"), height = unit(10, "cm") \# 40
genes 13. Monocytes: width = unit(25, "cm"), height = unit(15, "cm") \#
48 genes 15. All immune genes: width_image = 60, height_image = 70,
width = unit(30, "cm"), height = unit(60, "cm"), row_split = 7,
column_split = 2

More refs:
<https://debrowser.readthedocs.io/en/master/heatmap/heatmap.html>

#### All genes

```{r}
################# Immune 
#Heatmap annotation
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "left")

row_ha = HeatmapAnnotation(df = ann_rows_immune,
                       col = col_process_immune,
                       which= "row",
                       show_annotation_name = F,
                       show_legend = F,
                       simple_anno_size = unit(2, "mm")
                       )

# Values colors
col_fun <- colorRamp2(c(-1, 0, 1), c("#9bc1bc", "white", "#ed6a5a"))

file = filename_2
mat = matrix_data
width_image = 60
height_image = 50
width = unit(30, "cm")
height = unit(40, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 0)
rect_gp = gpar(col = "gray80", lwd = 0)
row_split = 3
column_split = 3


fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap.png")

png(file = here("Figures", fileimageheatmap_grouped), 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        col = col_fun, #color
                        name = "L2FC",
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(5, "cm"),
                        row_split = row_split,
                        column_split = column_split,
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "right",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        clustering_distance_columns = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height,
                        use_raster = F)

heatmap_plot = draw(heatmap_plot)

dev.off()

```
#### Interactive heatmap
```{r}
library(InteractiveComplexHeatmap)

####### Input
heatmap_data = Immune_GO_genes_All
file = "Immune_GO_genes_All"

###### Process data

heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Convert to wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

matrix_data_ready =  matrix_data %>% 
  round(2)

# Check if all vaccines are in the matrix and join annotations
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

# Join annotations
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 


#Rows Immune ----
ann_rows_immune = matrix_data %>%
  as.data.frame() %>%
  rownames_to_column("genes") %>%
  left_join(ImmuneGO_Annotated_Genes %>%
              filter(go_term == "Manual",
                     !process %in% c("TCR REPERTOIRE", "BCR REPERTOIRE")), by = "genes") %>%
  select(genes, process) %>%
  distinct() %>%
  group_by(genes) %>%
  arrange(genes, factor(process, levels = c("INNATE IMMUNE SYSTEM", "ADAPTIVE IMMUNE SYSTEM")), .by_group = TRUE) %>%
  mutate(i_process = row_number()) %>%
  pivot_wider(., names_from = "i_process", values_from = "process") %>%
  column_to_rownames("genes") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("index") %>%
  mutate(index = as.numeric(index)) %>%
  arrange(desc(index)) %>%
  column_to_rownames("index") %>%
  t() %>%
  as.data.frame() %>%
  replace(is.na(.), ".")

# Reorder columns (same order as lines in the annotation table)
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% 
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0) %>% 
  as.matrix() 


#Check dimensions
dim(matrix_data_ordered) 
dim(ann_vaccines_heatmap) 
dim(ann_rows_immune)

#Create heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")

row_ha = HeatmapAnnotation(df = ann_rows_immune,
                       col = col_process_immune,
                       which= "row",
                       show_annotation_name = F,
                       show_legend = F)

col_fun <- colorRamp2(c(-2, 0, 2), 
                      c("#9bc1bc", "white", "#ed6a5a"))


#Parameters
#Parameters
mat = matrix_data_ordered
width_image = 40
height_image = 90
width = unit(30, "cm")
height = unit(80, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_heatmap_legend = TRUE, 
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "L2FC", #Legend
                        col = col_fun,
                        show_row_dend = F,
                        row_split = 2, #Split group clusters
                        show_row_names = T,
                        column_names_gp = column_names_gp,
                        row_names_gp = row_names_gp)

ht_list = draw(heatmap_plot_all, 
               annotation_legend_side = "right", 
               heatmap_legend_side = "right")
htShiny(ht_list)

htShiny(ht_list, save = "Interactive_Covid_Other_Vax_Heatmap")

```



#### Get Cluster 1 Immune genes
Get row clusters

```{r}
r.dend <- row_dend(heatmap_plot)  #Extract row dendrogram
rcl.list <- row_order(heatmap_plot)  #Extract clusters (output is a list)
lapply(rcl.list, function(x) length(x))  #check/confirm size cluster

gene_cluster <- NULL
for (i in 1:length(rcl.list)) {
  a <- row.names(mat[row_order(heatmap_plot)[[i]], ])
  gene_cluster <- rbind(gene_cluster, data.frame(genes = a, cluster = paste0("cluster ", i), gene_set = "Immune genes"))
}
write.csv(gene_cluster, file = here("DGE analysis", paste0(file, "_clustered_genes.csv")), row.names = F)
```

```{r}
####### INPUT
data_2 = Immune_GO_genes_All %>% 
  inner_join(gene_cluster %>% filter(cluster == "cluster 1") %>% select(genes), by = "genes")
filename_2 = paste0("Immune_GO_genes_cluster1", "_VAX_COVID")

######## Matrix
matrix = data_2 %>% 
  select(genes, condition, regulation_numeric) %>%
  pivot_wider(names_from = "condition", 
              values_from = "regulation_numeric", 
              values_fn = mean) %>% 
  replace(is.na(.), 0) %>%
  arrange(genes) %>% 
  column_to_rownames(var="genes") %>% 
  as.matrix()

######## Annotations
ann_vaccines_covid_other_2 = data_2 %>% 
  inner_join(ann_vaccines_covid_other, by = "condition") %>% 
  select(condition, covid_other, microbe_type, disease_vac, type, week) %>% 
  distinct()

#Columns
ann_cols_heatmap = matrix %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  inner_join(ann_vaccines_covid_other_2, by = "condition") %>% 
  distinct() %>% 
  arrange(condition) %>% 
  column_to_rownames("condition")

matrix_data = matrix %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  inner_join(ann_cols_heatmap %>% rownames_to_column("condition"), by = "condition") %>% 
  select(!c(covid_other:week)) %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>%
  as.matrix() 

#Rows Immune ----
ann_rows_immune = matrix_data %>%
  as.data.frame() %>%
  rownames_to_column("genes") %>%
  left_join(ImmuneGO_Annotated_Genes %>% filter(go_term == "Manual"), by = "genes") %>%
  select(genes, process) %>%
  distinct() %>%
  group_by(genes) %>%
  arrange(genes, factor(process, levels = c("INNATE IMMUNE SYSTEM", "ADAPTIVE IMMUNE SYSTEM")), .by_group = TRUE) %>%
  mutate(i_process = row_number()) %>%
  pivot_wider(., names_from = "i_process", values_from = "process") %>%
  column_to_rownames("genes") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("index") %>%
  mutate(index = as.numeric(index)) %>%
  arrange(desc(index)) %>%
  column_to_rownames("index") %>%
  t() %>%
  as.data.frame() %>%
  replace(is.na(.), ".") %>% 
  mutate(`1` = if_else(`1` == ".", "INNATE IMMUNE SYSTEM", `1`))

dim(matrix)
dim(matrix_data)
dim(ann_cols_heatmap)
dim(ann_rows_immune)

```

More refs:
<https://debrowser.readthedocs.io/en/master/heatmap/heatmap.html>

```{r}
################# Immune 
#Heatmap annotation
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "right")

row_ha = row_ha = HeatmapAnnotation(df = ann_rows_immune,
                       col = col_process_immune,
                       which= "row",
                       show_annotation_name = F,
                       show_legend = F,
                       simple_anno_size = unit(2, "mm"))

# Values colors
col_fun <- colorRamp2(c(-1, 0, 1), c("#9bc1bc", "white", "#ed6a5a"))

file = filename_2
mat = matrix_data
width_image = 50
height_image = 50
width = unit(30, "cm")
height = unit(30, "cm")
column_names_gp = gpar(fontsize = 12)
row_names_gp = gpar(fontsize = 12)
rect_gp = gpar(col = "gray80", lwd = 0.5)

fileimageheatmap_grouped = here("Figures", paste0(file, sep ="_", "Complexheatmap.png"))

png(file = fileimageheatmap_grouped, 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                       left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        col = col_fun, #color
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                       name = "L2FC",
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(0, "cm"),
                        row_split = NULL,
                        column_split = 2,
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height)

heatmap_plot = draw(heatmap_plot, 
     annotation_legend_side = "left", 
     heatmap_legend_side = "left",
     merge_legend = TRUE)

heatmap_plot

dev.off()

```

#### 50 most shared genes

```{r fig.height=10, fig.width=10}
#### Immune -----

degs_all_updown_covid_vax_other_immune = degs_all_updown_covid_other_immune

stats_1 = degs_all_updown_covid_vax_other_immune %>% 
  inner_join(ann_vaccines_covid_other, by = "condition") %>% 
  group_by(genes, covid_other) %>% 
  summarise(n_conditions = n()) %>% 
  filter(all(c("COVID", "OTHER") %in% covid_other)) %>% 
  arrange(desc(n_conditions), genes) %>% 
  group_by(genes) %>% 
  pivot_wider(., names_from = "covid_other", values_from = n_conditions) %>% 
  arrange(desc(OTHER)) %>% 
  ungroup() %>% 
  slice_head(n = 50) %>% 
  pivot_longer(-genes, names_to = "covid_other", values_to = "n_conditions")

stats_1 %>% 
  select(genes) %>% 
  distinct()

ggplot_stats1 = stats_1 %>% 
  ggplot() +
  aes(x = reorder(genes, n_conditions), fill = covid_other, weight = n_conditions) +
  geom_bar() +
  theme_minimal() +
  scale_fill_manual(
    values = ann_vaxsig_colors$covid_other) +
  theme_minimal() +
  labs(fill = "Covid or non-covid",
       title = "Genes shared",
       x = "",
       y = "Number of shared conditions") +
  theme(legend.position = "top",
        axis.text.x = element_text(vjust = 0, hjust=0, size = 12),
        axis.text.y = element_text(size = 12, angle = 0, color = "black"),
        panel.grid.minor.y = element_blank(),
        legend.text = element_text(size=12),
        legend.title = element_text(size = 12, face = "bold")) +
  coord_flip() +
    ylim(-0.1, 22)
ggplot_stats1


ggsave(ggplot_stats1, file = here("Figures", "Covid_Vax_genes_shared_column.png"), width = 10, height = 10)


# Subset the 50 most shared genes
covid_other_immune_genes_50_mostshared_all = stats_1 %>% 
  select(genes) %>% 
  distinct() %>% 
  inner_join(degs_all_updown_covid_vax_other_immune %>% 
               select(condition, genes), by = "genes")
```




