---
title: "Covid-19 Vaccination Atlas"
author: "Wasim Aluísio Prates-Syed"
date: "2024-06-20"
output: 
  html_document: 
    toc: TRUE
    toc_depth: 4
    code_folding: hide
editor_options: 
  markdown: 
    wrap: 72
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval=FALSE, cache = TRUE)
```

# Covid-19 Vaccination Atlas: A systems vaccinology analysis

# Article information

**Authors:** Wasim Aluísio Prates-Syed1,2, Dennyson Leandro Mathias da
Fonseca3, Shahab Zaki Pour3, Aline Lira1,2, Nelson Cortes1, 2, Jaqueline
Dinis Queiroz Silva 1,5, Larissa Vuitika¹, Evellyn Carvalho1,4, Tania
Geraldine Churascari Vinces4, Gerhard Wunderlich4, Ricardo
Durães-Carvalho6, Lorena C. S. Chaves7, Niels O. S. Câmara1, Ester
Cerdeira SABINO8,9, José E. Krieger10, Otávio Cabral-Marques1,3,5,11,∙,
Gustavo Cabral-Miranda1,2,5,8 \*, ∙

**Emails:** [wasim.syed\@usp.br](mailto:wasim.syed@usp.br){.email},
[dennyson\@usp.br](mailto:dennyson@usp.br){.email},
[shahab.178\@gmail.com](mailto:shahab.178@gmail.com){.email},
[cv.tania\@usp.br](mailto:cv.tania@usp.br){.email},
[aline.llira\@usp.br](mailto:aline.llira@usp.br){.email},
[nelson.cortes\@usp.br](mailto:nelson.cortes@usp.br){.email},
[jaquelinedinis\@gmail.com](mailto:jaquelinedinis@gmail.com){.email},
[evelyn.carvalho\@unesp.br](mailto:evelyn.carvalho@unesp.br){.email},
[vuitika\@usp.br](mailto:vuitika@usp.br){.email},
[lorenacschaves\@gmail.com](mailto:lorenacschaves@gmail.com){.email},
[gwunder\@usp.br](mailto:gwunder@usp.br){.email},
[rdcarval\@gmail.com](mailto:rdcarval@gmail.com){.email},
[niels\@icb.usp.br](mailto:niels@icb.usp.br){.email},
[sabinoec\@usp.br](mailto:sabinoec@usp.br){.email},
[j.krieger\@hc.fm.usp.br](mailto:j.krieger@hc.fm.usp.br){.email},
[otavio.cmarques\@usp.br](mailto:otavio.cmarques@usp.br){.email},
[gcabral.miranda\@usp.br](mailto:gcabral.miranda@usp.br){.email}

**Institutions:** 1. Department of Immunology, Institute of Biomedical
Sciences (ICB), University of São Paulo (USP), São Paulo, Brazil. 2. The
Interunits Graduate Program in Biotechnology of the University of São
Paulo, the Butantan Institute and the Technological Research Institute
of the State of São Paulo, Brazil. 3. Interunit Postgraduate Program on
Bioinformatics, Institute of Mathematics and Statistics (IME),
University of Sao Paulo (USP), Sao Paulo, Brazil. 4. Department of
Parasitology, Institute of Biomedical Sciences (ICB), University of São
Paulo (USP), São Paulo, Brazil. 5. The Graduate Program in
Pathophysiology and Toxicology, Department of Clinical and Toxicological
Analyses, School of Pharmaceutical Sciences, University of São Paulo,
São Paulo, Brazil. 6. São Paulo School of Medicine, Department of
Microbiology, Immunology and Parasitology, Federal University of São
Paulo (UNIFESP), São Paulo, SP, Brazil. 7. Department of Microbiology
and Immunology, School of Medicine, Emory University, Atlanta, Georgia,
United States. 8. Institute of Tropical Medicine, Faculty of Medicine of
the University of São Paulo, Brazil 9. Department of Infectious and
Parasitic Diseases, Faculty of Medicine, University of São Paulo, São
Paulo, Brazil 10. Heart Institute, Clinical Hospital, Faculty of
Medicine, University of São Paulo, Brazil; Laboratory of Genetics and
Molecular Cardiology, Clinical Hospital, Faculty of Medicine, University
of São Paulo, Brazil. 11. Department of Medicine, Division of Molecular
Medicine, University of São Paulo School of Medicine, São Paulo, Brazil.

# Abstract:

The rapid development of COVID-19 vaccines has played a crucial role in
mitigating its public health impact. However, a gene-level understanding
of immune system dynamics during vaccination is crucial for
comprehending the various vaccination scenarios, including infection. In
this study, we conducted an integrative analysis of immunological
transcriptomic profiles to delineate the impact of COVID-19 vaccination
on immune responses. Our analysis comprised 681 bulk RNAseq samples from
343 participants, including healthy individuals vaccinated with
different types of COVID-19 vaccines (inactivated virus, protein
subunit, viral-vectored, and mRNA), along with infected patients with or
without prior vaccination history. We identified distinct and
overlapping immune-related genes and pathways associated with COVID-19
vaccination and infection, considering time and regimen. Furthermore, we
expanded our analysis to include datasets from vaccines targeting other
pathogens, such as influenza, smallpox, and yellow fever. Additionally,
we introduce VaxGO, a novel tool for exploring immunological processes
and comparing them across different vaccines. Our work will contribute
to advancing our understanding of the immunological mechanisms
underlying Covid-19 vaccination and inform future developments in
vaccine design and personalized medicine. **Keywords:** transcriptomics,
vaccines, COVID-19, SARS-CoV-2.

Read the recent version here:
(<https://www.medrxiv.org/content/10.1101/2024.05.22.24307755v3>)

# Codes for analyses and figures

# 1. Preparing

## 1.1. Packages

```{r message=FALSE, warning=FALSE, cache = TRUE}
# install.packages("randomForest")
# install.packages("caret")
# install.packages("rpart")
# install.packages("rpart.plot")
# install.packages("tidymodels")
# install.packages("reticulate")
# install.packages("themis")
# install.packages("ranger")
# install.packages("workflowsets")
# install.packages("glmnet")
# install.packages("kknn")
# install.packages("nnet")
# install.packages("tictoc")

library(randomForest)
library(ranger)
library(caret)
library(reticulate)
library(themis)
library(rpart)
library(rpart.plot)
library(here)
library(ggrepel)
library(ComplexHeatmap)
library(janitor)
library(ggsci)
library(vip)
library(here)
library(circlize)
library(workflowsets)
library(glmnet)
library(kknn)
library(nnet)
library(tictoc)
library(ggthemes)
library(readr)
library(qvalue)
library(explore)
library(maditr)
library(GEOquery)
library(Matrix)
library(pheatmap)
library(RColorBrewer)
library(celldex)
library(biomaRt)
library(org.Hs.eg.db)
library(plotly)
library(DESeq2)
library(msigdbr)
library(clusterProfiler)
library(EnhancedVolcano)
library(ggbeeswarm)
library(corrr)
library(ggcorrplot)
library(FactoMineR)
library(factoextra)
library(esquisse)
library(corto)
library(circlize)
library(ComplexHeatmap)
library(janitor)
library(scales)
library(ggstatsplot)
library(patchwork)
library(ggsci)
library(here)
library(vegan)
library(BH)
library(chorddiag)
library(tidyverse)
library(tidymodels)
tidymodels_prefer()
```

## 1.2. Standardize figures

```{r cache = TRUE, message=FALSE, warning=FALSE}
### VAX SIG DB ----
ann_vaxsig_colors = list(
  type = c("VLP" = "#7E6148FF",
           "LA" =  "#E64B35FF",
           "CONJ" = "#F39B7FFF", 
           'IN'= "#00A087FF", #1st Gen  vaccines
           'VV' = "#3C5488FF", #3rd Gen vaccines
           'RNA' = "#4DBBD5FF",
           'SU' = "#8491B4FF",
           'I'= "#DC0000FF",
           "H" = "grey95"),
  target_pathogen_disease = c('MENINGOCOCCUS'	= "#00A087FF",
                              'PNEUMOCOCCUS'	= "#44C199",
                              'TUBERCULOSIS'	= "#99e2b4",
                              'F TULARENSIS'	= "#4DBBD5FF",
                              'LEISHMANIA'	= "#118ab2",
                              'MALARIA' = "#015BA0",
                              'EBOV'	= "#5e60ce",
                              'HBV'	= "#D7B0EE",
                              'HBV, HAV'	= "#D7B0EE",
                              'HIV'	= "#b5179e",
                              'HPV'	= "#c77dff",
                              'INFLUENZA'	= "#deaaff",
                              'MEASLES'	= "#9d4edd",
                              'MMR'	= "#7b2cbf",
                              'SMALLPOX'	= "#e5b3fe",
                              'VEEV'	= "#ecbcfd",
                              'VZV'	= "#c8b6ff",
                              'YF'	= "#b8c0ff"),
  covid_other = c("COVID" = "#DC0000FF",
                  "OTHER" = "#4DBBD5FF"),
  microbe_type = c("VIRUS" = "#5e60ce",
                   'BACTERIUM' = "#00A087FF",
                   'PARASITE' = "#7E6148FF"),
  
  week = c("0" = "#ECF8FA",
           "1" = "#caf0f8",
           "2" = "#6CD5EA",
           "3" = "#0087BF",
           "4" = "#015BA0", 
           "6" = "#03045e",
           "7" = "#03045e",
           "8" = "#03045e"),
  month = c("1" = "#caf0f8",
            "0" = "#6CD5EA",
            "2" = "#015BA0"),
  vaccine = c("BBIBP" = "#00A087FF",
              "ZF2001" = "#8491B4FF",
              "BNT" = "#4DBBD5FF",
              "MO" = "#72ddf7",
              "ChAd" = "#3C5488FF",
              "I" = "grey95",
              "H" = "grey95"),
  dose = c(
    "0" = "grey95",
    "1" = "#91D1C2A0",
    "2" = "#00A087FF",
    "3" = "#3C5488FF"),
  day = c(
    "0" = "white",
    "1" = "#caf0f8",
    "2" = "#ade8f4",
    "3" = "#90e0ef",
    "4" = "#6CD5EA",
    "5" = "#48cae4",
    "6" = "#00b4d8",
    "7" = "#0096c7",
    "10" = "#0087BF",
    "11" = "#0087BF",
    "14" = "#0077b6",
    "26" = "#015BA0",
    "28" = "#023e8a",
    "51" = "#03045e"),
  infection = c("0" = "white",
                "1" = "#00A087FF",
                "2" = "#3C5488FF"),
  variant = c("None" = "grey95",
              "Beta" = "#00A087FF",
              "Omicron" = "#3C5488FF"),
  severity = c("H" = "grey95",
               "S" = "#DC0000FF",
               "MO" = "#8491B4FF",
               "MI" = "#91D1C2FF",
               "NA" = "grey95"),
  sex = c("Male" = "#4DBBD5FF",
          "Female" = "#E64B35FF",
          "M/F" = "#8491B4FF"),
  disease_vac = c("I" = "#DC0000FF",
                  "V" = "#4DBBD5FF")
)


# Not immune processes -----
col_process_notimmune = list(
  "1" = c("Cellular" = "#3C5488FF",
          "Interspecies interaction" = "#00A087FF",
          "Metabolism" = "#4DBBD5FF",
          "Reg. of BP" = "#E64B35FF",
          "."="white"),
  "2" = c("Cellular" = "#3C5488FF",
          "Interspecies interaction" = "#00A087FF",
          "Metabolism" = "#4DBBD5FF",
          "Reg. of BP" = "#E64B35FF",
          "."="white"),
  "3" = c("Cellular" = "#3C5488FF",
          "Interspecies interaction" = "#00A087FF",
          "Metabolism" = "#4DBBD5FF",
          "Reg. of BP" = "#E64B35FF",
          "."="white"),
  "4" = c("Cellular" = "#3C5488FF",
          "Interspecies interaction" = "#00A087FF",
          "Metabolism" = "#4DBBD5FF",
          "Reg. of BP" = "#E64B35FF",
          "."="white"),
  "5" = c("Cellular" = "#3C5488FF",
          "Interspecies interaction" = "#00A087FF",
          "Metabolism" = "#4DBBD5FF",
          "Reg. of BP" = "#E64B35FF",
          "."="white"))


#Immune processes ------
col_process_immune = list(
  "1" = c("ADAPTIVE IMMUNE SYSTEM" = "#6f2dbd",
          "B CELL" = "#7b2cbf",
          "CELLULAR ADAPTIVE IMMUNE SYSTEM" = "#e0aaff",
          "HUMORAL ADAPTIVE IMMUNE SYSTEM" = "#9d4edd",
          "IMMUNOGLOBULIN MEDIATED RESPONSE"= "#5a189a",
          "T CELLS" = "#c77dff",
          "T HELPER CELLS" = "#c77dcc",
          "COMPLEMENT" = "#ffadc7",
          "ANTIVIRAL & IFN" = "#88d4ab",
          "ARPP" = "#14746f",
          "DENDRITIC CELL" = "#036666",
          "EOSINOPHIL" = "#67b99a",
          "INFLAMMATION" = "#99e2b4",
          "INNATE IMMUNE SYSTEM" = "#06d6a0",
          "MAST CELL" = "#56ab91",
          "MONOCYTE" = "#358f80",
          "NEUTROPHIL" = "#78c6a3",
          "NK CELL" = "#469d89",
          "MACROPHAGE" = "#14746f",
          "GENERAL" = "#343a40",
          "ANTIMICROBIAL" = "#48cae4",
          "."="white"),
  "2" = c("ADAPTIVE IMMUNE SYSTEM" = "#6f2dbd",
          "B CELLS" = "#7b2cbf",
          "CELLULAR ADAPTIVE IMMUNE SYSTEM" = "#e0aaff",
          "HUMORAL ADAPTIVE IMMUNE SYSTEM" = "#9d4edd",
          "IMMUNOGLOBULIN MEDIATED RESPONSE"= "#5a189a",
          "T CELL" = "#c77dff",
          "T HELPER CELLS" = "#c77dcc",
          "COMPLEMENT" = "#ffadc7",
          "ANTIVIRAL & IFN" = "#88d4ab",
          "ARPP" = "#14746f",
          "DENDRITIC CELL" = "#036666",
          "EOSINOPHIL" = "#67b99a",
          "INFLAMMATION" = "#99e2b4",
          "INNATE IMMUNE SYSTEM" = "#06d6a0",
          "MAST CELL" = "#56ab91",
          "MONOCYTE" = "#358f80",
          "NEUTROPHIL" = "#78c6a3",
          "NK CELL" = "#469d89",
          "MACROPHAGE" = "#14746f",
          "GENERAL" = "#343a40",
          "ANTIMICROBIAL" = "#48cae4",
          "."="white"),
  "3" = c("ADAPTIVE IMMUNE SYSTEM" = "#6f2dbd",
          "B CELL" = "#7b2cbf",
          "CELLULAR ADAPTIVE IMMUNE SYSTEM" = "#e0aaff",
          "HUMORAL ADAPTIVE IMMUNE SYSTEM" = "#9d4edd",
          "IMMUNOGLOBULIN MEDIATED RESPONSE"= "#5a189a",
          "T CELL" = "#c77dff",
          "T HELPER CELLS" = "#c77dcc",
          "COMPLEMENT" = "#ffadc7",
          "ANTIVIRAL & IFN" = "#88d4ab",
          "ARPP" = "#14746f",
          "DENDRITIC CELL" = "#036666",
          "EOSINOPHIL" = "#67b99a",
          "INFLAMMATION" = "#99e2b4",
          "INNATE IMMUNE SYSTEM" = "#06d6a0",
          "MAST CELL" = "#56ab91",
          "MONOCYTE" = "#358f80",
          "NEUTROPHIL" = "#78c6a3",
          "NK CELL" = "#469d89",
          "MACROPHAGE" = "#14746f",
          "GENERAL" = "#343a40",
          "ANTIMICROBIAL" = "#48cae4",
          "."="white"),
  "4" = c("ADAPTIVE IMMUNE SYSTEM" = "#6f2dbd",
          "B CELL" = "#7b2cbf",
          "CELLULAR ADAPTIVE IMMUNE SYSTEM" = "#e0aaff",
          "HUMORAL ADAPTIVE IMMUNE SYSTEM" = "#9d4edd",
          "IMMUNOGLOBULIN MEDIATED RESPONSE"= "#5a189a",
          "T CELL" = "#c77dff",
          "T HELPER CELLS" = "#c77dcc",
          "COMPLEMENT" = "#ffadc7",
          "ANTIVIRAL & IFN" = "#88d4ab",
          "ARPP" = "#14746f",
          "DENDRITIC CELL" = "#036666",
          "EOSINOPHIL" = "#67b99a",
          "INFLAMMATION" = "#99e2b4",
          "INNATE IMMUNE SYSTEM" = "#06d6a0",
          "MAST CELL" = "#56ab91",
          "MONOCYTE" = "#358f80",
          "NEUTROPHIL" = "#78c6a3",
          "NK CELL" = "#469d89",
          "MACROPHAGE" = "#14746f",
          "GENERAL" = "#343a40",
          "ANTIMICROBIAL" = "#48cae4",
          "."="white"),
  "5" = c("ADAPTIVE IMMUNE SYSTEM" = "#6f2dbd",
          "B CELL" = "#7b2cbf",
          "CELLULAR ADAPTIVE IMMUNE SYSTEM" = "#e0aaff",
          "HUMORAL ADAPTIVE IMMUNE SYSTEM" = "#9d4edd",
          "IMMUNOGLOBULIN MEDIATED RESPONSE"= "#5a189a",
          "T CELL" = "#c77dff",
          "T HELPER CELLS" = "#c77dcc",
          "COMPLEMENT" = "#ffadc7",
          "ANTIVIRAL & IFN" = "#88d4ab",
          "ARPP" = "#14746f",
          "DENDRITIC CELL" = "#036666",
          "EOSINOPHIL" = "#67b99a",
          "INFLAMMATION" = "#99e2b4",
          "INNATE IMMUNE SYSTEM" = "#06d6a0",
          "MAST CELL" = "#56ab91",
          "MONOCYTE" = "#358f80",
          "NEUTROPHIL" = "#78c6a3",
          "NK CELL" = "#469d89",
          "MACROPHAGE" = "#14746f",
          "GENERAL" = "#343a40",
          "ANTIMICROBIAL" = "#48cae4",
          "."="white"),
  "6" = c("ADAPTIVE IMMUNE SYSTEM" = "#6f2dbd",
          "B CELL" = "#7b2cbf",
          "CELLULAR ADAPTIVE IMMUNE SYSTEM" = "#e0aaff",
          "HUMORAL ADAPTIVE IMMUNE SYSTEM" = "#9d4edd",
          "IMMUNOGLOBULIN MEDIATED RESPONSE"= "#5a189a",
          "T CELL" = "#c77dff",
          "T HELPER CELLS" = "#c77dcc",
          "COMPLEMENT" = "#ffadc7",
          "ANTIVIRAL & IFN" = "#88d4ab",
          "ARPP" = "#14746f",
          "DENDRITIC CELL" = "#036666",
          "EOSINOPHIL" = "#67b99a",
          "INFLAMMATION" = "#99e2b4",
          "INNATE IMMUNE SYSTEM" = "#06d6a0",
          "MAST CELL" = "#56ab91",
          "MONOCYTE" = "#358f80",
          "NEUTROPHIL" = "#78c6a3",
          "NK CELL" = "#469d89",
          "MACROPHAGE" = "#14746f",
          "GENERAL" = "#343a40",
          "ANTIMICROBIAL" = "#48cae4",
          "."="white"),
  "7" =c("ADAPTIVE IMMUNE SYSTEM" = "#6f2dbd",
         "B CELL" = "#7b2cbf",
         "CELLULAR ADAPTIVE IMMUNE SYSTEM" = "#e0aaff",
         "HUMORAL ADAPTIVE IMMUNE SYSTEM" = "#9d4edd",
         "IMMUNOGLOBULIN MEDIATED RESPONSE"= "#5a189a",
         "T CELL" = "#c77dff",
         "T HELPER CELLS" = "#c77dcc",
         "COMPLEMENT" = "#ffadc7",
         "ANTIVIRAL & IFN" = "#88d4ab",
         "ARPP" = "#14746f",
         "DENDRITIC CELL" = "#036666",
         "EOSINOPHIL" = "#67b99a",
         "INFLAMMATION" = "#99e2b4",
         "INNATE IMMUNE SYSTEM" = "#06d6a0",
         "MAST CELL" = "#56ab91",
         "MONOCYTE" = "#358f80",
         "NEUTROPHIL" = "#78c6a3",
         "NK CELL" = "#469d89",
         "MACROPHAGE" = "#14746f",
         "GENERAL" = "#343a40",
         "ANTIMICROBIAL" = "#48cae4",
         "."="white"),
  "8" = c("ADAPTIVE IMMUNE SYSTEM" = "#6f2dbd",
          "B CELL" = "#7b2cbf",
          "CELLULAR ADAPTIVE IMMUNE SYSTEM" = "#e0aaff",
          "HUMORAL ADAPTIVE IMMUNE SYSTEM" = "#9d4edd",
          "IMMUNOGLOBULIN MEDIATED RESPONSE"= "#5a189a",
          "T CELL" = "#c77dff",
          "T HELPER CELLS" = "#c77dcc",
          "COMPLEMENT" = "#ffadc7",
          "ANTIVIRAL & IFN" = "#88d4ab",
          "ARPP" = "#14746f",
          "DENDRITIC CELL" = "#036666",
          "EOSINOPHIL" = "#67b99a",
          "INFLAMMATION" = "#99e2b4",
          "INNATE IMMUNE SYSTEM" = "#06d6a0",
          "MAST CELL" = "#56ab91",
          "MONOCYTE" = "#358f80",
          "NEUTROPHIL" = "#78c6a3",
          "NK CELL" = "#469d89",
          "MACROPHAGE" = "#14746f",
          "GENERAL" = "#343a40",
          "ANTIMICROBIAL" = "#48cae4",
          "."="white"),
  "9" = c("ADAPTIVE IMMUNE SYSTEM" = "#6f2dbd",
          "B CELL" = "#7b2cbf",
          "CELLULAR ADAPTIVE IMMUNE SYSTEM" = "#e0aaff",
          "HUMORAL ADAPTIVE IMMUNE SYSTEM" = "#9d4edd",
          "IMMUNOGLOBULIN MEDIATED RESPONSE"= "#5a189a",
          "T CELL" = "#c77dff",
          "T HELPER CELLS" = "#c77dcc",
          "COMPLEMENT" = "#ffadc7",
          "ANTIVIRAL & IFN" = "#88d4ab",
          "ARPP" = "#14746f",
          "DENDRITIC CELL" = "#036666",
          "EOSINOPHIL" = "#67b99a",
          "INFLAMMATION" = "#99e2b4",
          "INNATE IMMUNE SYSTEM" = "#06d6a0",
          "MAST CELL" = "#56ab91",
          "MONOCYTE" = "#358f80",
          "NEUTROPHIL" = "#78c6a3",
          "NK CELL" = "#469d89",
          "MACROPHAGE" = "#14746f",
          "GENERAL" = "#343a40",
          "ANTIMICROBIAL" = "#48cae4",
          "."="white"),
  "10" = c("ADAPTIVE IMMUNE SYSTEM" = "#6f2dbd",
           "B CELL" = "#7b2cbf",
           "CELLULAR ADAPTIVE IMMUNE SYSTEM" = "#e0aaff",
           "HUMORAL ADAPTIVE IMMUNE SYSTEM" = "#9d4edd",
           "IMMUNOGLOBULIN MEDIATED RESPONSE"= "#5a189a",
           "T CELL" = "#c77dff",
           "T HELPER CELLS" = "#c77dcc",
           "COMPLEMENT" = "#ffadc7",
           "ANTIVIRAL & IFN" = "#88d4ab",
           "ARPP" = "#14746f",
           "DENDRITIC CELL" = "#036666",
           "EOSINOPHIL" = "#67b99a",
           "INFLAMMATION" = "#99e2b4",
           "INNATE IMMUNE SYSTEM" = "#06d6a0",
           "MAST CELL" = "#56ab91",
           "MONOCYTE" = "#358f80",
           "NEUTROPHIL" = "#78c6a3",
           "NK CELL" = "#469d89",
           "MACROPHAGE" = "#14746f",
           "GENERAL" = "#343a40",
           "ANTIMICROBIAL" = "#48cae4",
           "."="white"),
  "11" = c("ADAPTIVE IMMUNE SYSTEM" = "#6f2dbd",
           "B CELL" = "#7b2cbf",
           "CELLULAR ADAPTIVE IMMUNE SYSTEM" = "#e0aaff",
           "HUMORAL ADAPTIVE IMMUNE SYSTEM" = "#9d4edd",
           "IMMUNOGLOBULIN MEDIATED RESPONSE"= "#5a189a",
           "T CELL" = "#c77dff",
           "T HELPER CELLS" = "#c77dcc",
           "COMPLEMENT" = "#ffadc7",
           "ANTIVIRAL & IFN" = "#88d4ab",
           "ARPP" = "#14746f",
           "DENDRITIC CELL" = "#036666",
           "EOSINOPHIL" = "#67b99a",
           "INFLAMMATION" = "#99e2b4",
           "INNATE IMMUNE SYSTEM" = "#06d6a0",
           "MAST CELL" = "#56ab91",
           "MONOCYTE" = "#358f80",
           "NEUTROPHIL" = "#78c6a3",
           "NK CELL" = "#469d89",
           "MACROPHAGE" = "#14746f",
           "GENERAL" = "#343a40",
           "ANTIMICROBIAL" = "#48cae4",
           "."="white"),
  "12" = c("ADAPTIVE IMMUNE SYSTEM" = "#6f2dbd",
           "B CELL" = "#7b2cbf",
           "CELLULAR ADAPTIVE IMMUNE SYSTEM" = "#e0aaff",
           "HUMORAL ADAPTIVE IMMUNE SYSTEM" = "#9d4edd",
           "IMMUNOGLOBULIN MEDIATED RESPONSE"= "#5a189a",
           "T CELL" = "#c77dff",
           "T HELPER CELLS" = "#c77dcc",
           "COMPLEMENT" = "#ffadc7",
           "ANTIVIRAL & IFN" = "#88d4ab",
           "ARPP" = "#14746f",
           "DENDRITIC CELL" = "#036666",
           "EOSINOPHIL" = "#67b99a",
           "INFLAMMATION" = "#99e2b4",
           "INNATE IMMUNE SYSTEM" = "#06d6a0",
           "MAST CELL" = "#56ab91",
           "MONOCYTE" = "#358f80",
           "NEUTROPHIL" = "#78c6a3",
           "NK CELL" = "#469d89",
           "MACROPHAGE" = "#14746f",
           "GENERAL" = "#343a40",
           "ANTIMICROBIAL" = "#48cae4",
           "."="white"))

```

# 2. Select studies and retrieve data

## 2.1. Download and process data

Download general data (counts data and metadata). GEO supplementary
files have the counts matrices. Download studies individually. I
recommend downloading large files from the GEO website.

```{r eval=FALSE, message=FALSE, warning=FALSE, echo=TRUE, cache = TRUE}
# Function to read and standardize files
read_standardize <- function(file) {
  files <- read.table(file, header = FALSE, sep = "\t")
  if (ncol(files) >= 2) {
    colnames(file)[1:2] <- c("genes", tools::file_path_sans_ext(basename(file)))
    return(files)
  } else {
    return(NULL)
  }
}
```

GSE189039

```{r eval=FALSE, message=FALSE, warning=FALSE, echo=TRUE, cache = TRUE}
###### GSE189039 ----
#Metadata 
filename = "GSE189039"
GSE189039 <- getGEO("GSE189039", GSEMatrix = TRUE) 
GSE189039 <- GSE189039[[1]]
GSE189039_metadata = GSE189039 %>% 
  pData() %>% 
  select( 
    geo_sample = "geo_accession", 
    subject_id = "title", age = "age:ch1", 
    sex = "gender:ch1", disease_state = "diseas state:ch1", 
    severity = "severity:ch1") 
write_csv(GSE189039_metadata, file = here("raw_data", "GSE189039_metadata.csv"))

#Counts matrix
#Retrieve counts table from supplementary files
getGEOSuppFiles('GSE189039', makeDirectory = T, baseDir = here("raw_data"))
untar(here("raw_data", 'GSE189039_RAW.tar'))
files <- list.files(path = here("raw_data", "GSE189039"), pattern = "*.gz$", full.names = TRUE)
for(files_gz in files) {gunzip(files_gz)}
files_txt_GSE189039 <- list.files(path = here("raw_data", "GSE189039"), pattern = "*.txt$", full.names = TRUE)
list_data_GSE189039 <- lapply(files_txt_GSE189039, read_standardize)
counts_GSE189039 <- list_data_GSE189039 %>%
  purrr::discard(is.null) %>%
  reduce(full_join, by = "genes")
colnames(counts_GSE189039)[-1] <- gsub("^[^_]+_(.*)", "\\1", colnames(counts_GSE189039)[-1])
write_csv(counts_GSE189039, file = here("raw_data", "gse189039_counts.csv"))
```

GSE201533

```{r eval=FALSE, message=FALSE, warning=FALSE, echo=TRUE, cache = TRUE}
###### GSE201533 ------

###### Metadata
GSE201533 <- getGEO("GSE201533", GSEMatrix = TRUE, destdir = here("raw_data", "GSE201533")) 

GSE201533 <- GSE201533[[1]]

GSE201533_metadata = GSE201533 %>% 
  pData() %>% 
  select(
    geo_sample = "geo_accession", 
    sample_id = "title", age = "age:ch1", 
    sex = "gender:ch1", 
    dose_vaccine = "vaccine:ch1")

GSE201533_metadata = GSE201533_metadata %>%
  separate(sample_id, into = c("participant_id", "time"), sep = "-D", remove = F) %>%
  separate(dose_vaccine, into = c("dose", "vaccine"), sep = ",\\s*(?=[A-Z])", extra = "merge") %>%
  mutate(dose = str_extract(dose, "\\d+"),
         vaccine = if_else(vaccine == "ChAdOx1", "ChAd", "BNT"),
         condition = paste0(vaccine, " (V", dose, ", D", time, ")"),
         geo = "GSE201533") 

write_csv(GSE201533_metadata, file = here("raw_data", "GSE201533_metadata.csv"))

##### Counts matrix
getGEOSuppFiles('GSE201533', makeDirectory = T, baseDir = here("raw_data"))
untar(tarfile = here("raw_data", "GSE201533", "GSE201533_RAW.tar"))

list_data_gse201533 <- list.files(path = here("raw_data", 
                                              "GSE201533", 
                                              "GSE201533_RAW"), 
                                  pattern = "*.txt.gz$", 
                                  full.names = TRUE) %>%
  map(~ read.table(.x, header = TRUE, sep = "\t"))
counts_gse201533 <- do.call(cbind, map(list_data_gse201533, ~ .x[, 2]))

# Name columns (samples) and genes
colnames(counts_gse201533) <- paste0("GSM", seq_along(list_data_gse201533) + 6066177)
counts_gse201533 <- bind_cols(geneid = list_data_gse201533[[1]][, 1], counts_gse201533)
counts_gse201533[is.na(counts_gse201533)] <- 0 #Rename column names (ex. "X21.01684.hisat2.bam") to GSM codes, which start with GSM6066178.

counts_gse201533 <- list_data_gse201533 %>%
  purrr::map_dfc(~ .x[, 2]) %>%
  set_names(paste0("Sample_", seq_along(list_data_gse201533))) %>%
  bind_cols(geneid = list_data_gse201533[[1]][, 1], .) %>%
  relocate(geneid) %>%
  replace(is.na(.), 0)

saveRDS(counts_gse201533, file = here("raw_data", "gse201533_counts.rds"))
write.csv(counts_gse201533, file=here("raw_data", "gse201533_counts.csv"))
```

GSE201530

```{r eval=FALSE, message=FALSE, warning=FALSE, echo=TRUE, cache = TRUE}
###### GSE201530 ------
#Metadata
filename = "GSE201530"
GSE201530 <- getGEO("GSE201530", GSEMatrix = TRUE) 
GSE201530 <- GSE201530[[1]]
GSE201530_metadata = GSE201530 %>% 
  pData() %>% 
  select(
    geo_sample = "geo_accession", 
    subject_id = "title", 
    age = "age:ch1", 
    sex = "gender:ch1", 
    disease_state = "disease state:ch1", 
    timepoint = "days after positive pcr results:ch1",
    group = "group (by covid-19 vaccination, prior infection):ch1",
    variant = "omicron sublineage:ch1") 
write.csv(GSE201530_metadata, file = here("raw_data", "GSE201530_metadata.csv"))

# Counts matrix
getGEOSuppFiles('GSE201530', baseDir = here("raw_data"))
untar(here("raw_data", 'GSE201530_RAW.tar'))
files_txt_GSE201530 <- list.files(path = here("raw_data", "GSE201530"), pattern = "*.txt$", full.names = TRUE)
list_data_GSE201530 <- lapply(files_txt_GSE201530, read_standardize)
counts_gse201530 <- list_data_GSE201530 %>%
  purrr::discard(is.null) %>%
  reduce(full_join, by = "genes")
colnames(counts_gse201530)[-1] <- gsub("^[^_]+_(.*)", "\\1", colnames(counts_gse201530)[-1])
write_csv(counts_gse201530, file = here("raw_data", "gse201530_counts.csv"))
```

GSE199750

As we couldn't retrieved all GSE199750 counts matrices through R, we
downloaded the counts table from the Github repository:
<https://bitbucket.org/lynnlab/covirs>.

```{r eval=FALSE, message=FALSE, warning=FALSE, echo=TRUE, cache = TRUE}
###### GSE199750 ----
#Metadata
gse199750 <- getGEO("GSE199750", GSEMatrix = TRUE,  destdir = here("raw_data", "GSE199750")) 
gse199750 <- gse199750[[1]]
gse199750_metadata <- pData(gse199750) %>% 
  select(
    GEO = "geo_accession", 
    SampleN = "sample number:ch1", 
    SubjectID = "subject_id:ch1", 
    Age = "age:ch1", 
    Sex = "Sex:ch1", 
    Vaccine = "vaccine:ch1", 
    FirstSecondVaccine = "first/second vaccine:ch1", 
    ThirdVaccine = "third vaccine:ch1", 
    Timepoint = "timepoint:ch1",
    Tissue = "tissue:ch1", 
    Run = "run:ch1"
  )

#There are two similar columns. The first, "Vaccine," might be filled in, while the second, "FirstSecondVaccine," may not be, and vice versa. Therefore, one should equal the other. Thus, unify the columns.

gse199750_metadata <- gse199750_metadata %>%
  mutate(
    FirstSecondDose = coalesce(FirstSecondVaccine, Vaccine),
    Subject_Timepoint = paste(SubjectID, Timepoint, sep = "_")
  ) %>%
  select(
    Subject_Timepoint, GEO, SampleN, SubjectID, Age, Sex, Timepoint, 
    FirstSecondDose, ThirdVaccine, Tissue, Run
  ) %>%
  mutate(Age_ranges = cut(
    Age, 
    breaks = c(0, 18, 30, 45, 60, 70, 80, Inf), 
    labels = c("<18", "18-29", "30-44", "45-59", "60-69", "70-79", "80>"), 
    right = FALSE
  ) %>% 
  column_to_rownames("Subject_Timepoint")

write.csv(gse199750_metadata, file =here("raw_data", "metadata_gse199750.csv"))
saveRDS(gse199750_metadata, file = here("raw_data", "gse199750_metadata.rds"))

# Counts matrix
raw_RNASeq_counts = readRDS(here("raw_data", "raw_RNASeq_counts.RDS"))

# The counts table provided by the study("https://bitbucket.org/lynnlab/covirs/src/master/raw_RNASeq_counts.RDS") is named with Subject ID + Timepoint and not with the GSM code. Therefore, we will need to process the metadata table. Additionally, it lacks the age column. These data can be manipulated in Excel.
############# Gene annotation -----
library("biomaRt")

mart <- useMart("ensembl")
mart <- useDataset("hsapiens_gene_ensembl", mart)

gene_info <- getBM(attributes = c("entrezgene_id", "hgnc_symbol", "ensembl_gene_id"), 
                   mart = mart) %>% 
  select(ensembl_gene_id, hgnc_symbol) %>% 
  distinct()

raw_RNASeq_counts_ann <- raw_RNASeq_counts %>% 
  rownames_to_column("ensembl_gene_id") %>% 
  inner_join(gene_info, by = "ensembl_gene_id") %>% 
  filter(hgnc_symbol != "") %>% 
  distinct() %>%
  select(sort(names(.))) %>% 
  select(genes = hgnc_symbol, ensembl_gene_id, everything()) %>% 
  arrange(genes)

#Save
write_csv(raw_RNASeq_counts_ann, file= here("raw_data", "raw_RNASeq_counts_annGenes.csv"))

#Filter duplicated genes. Get only the older ensembl ids

not_interes = raw_RNASeq_counts_ann %>% 
  get_dupes(genes) %>% 
  group_by(genes) %>% 
  arrange(ensembl_gene_id) %>% 
  slice(2) %>% 
  ungroup() %>% 
  select(ensembl_gene_id)

gse199750_counts_ready = raw_RNASeq_counts_ann %>% 
  anti_join(not_interes, by = "ensembl_gene_id") %>% 
  select(-ensembl_gene_id) %>% 
  column_to_rownames("genes")

# Save
saveRDS(gse199750_counts_ready, file= here("raw_data", "gse199750_counts_ready.rds"))

```

GSE206023

```{r eval=FALSE, message=FALSE, warning=FALSE, echo=TRUE, cache = TRUE}
######### GSE206023 ------- 
gse206023 <- getGEO("GSE206023", 
                    GSEMatrix = TRUE,  
                    destdir = here("raw_data")) 

gse206023 <- gse206023[[1]]
gse206023_metadata <- pData(gse206023)
colnames(gse206023_metadata)

gse206023_metadata <- select(gse206023_metadata, "title", "geo_accession", "source_name_ch1", "organism_ch1", "time:ch1", "treatment:ch1"  )

colnames(gse206023_metadata) <-
  c("Sample name",
    "GEO",
    "Sample type",
    "Organism",
    "Timepoints",
    "Vaccine")

write.csv(gse206023_metadata, file= here("raw_data","gse206023_metadata.csv"))
saveRDS(gse206023_metadata, file=here("raw_data","gse206023_metadata.rds"))

#### Import matrix.
#Genes were named incorrectly and represent only 2.000 genes of >30.000. Hence, we externally performed a pseudoalignment to retrieve all genes that were not correctly annotated. 
gse206023_counts <- read.delim(file=here("raw_data", "GSE206023_tpm_of_48_samples.txt"), 
                               header = TRUE, sep = "\t", dec = ".")

#Pseudoalignment
gse206023_counts = read_excel(here("raw_data", "GSE206023_counts_Pseudoalignment.xlsx"))

gse206023_counts_ready = gse206023_counts %>% 
  select(!c(1, 2)) %>% 
  drop_na(external_gene_name) %>% 
  column_to_rownames("external_gene_name")

gse206023_counts_ready %>% 
  saveRDS(file = here("raw_data", "gse206023_counts_ready.rds"))
gse206023_counts_ready %>% 
  write_csv(file = here("raw_data", "gse206023_counts_ready.csv"))

#General statistics
GSE206023_pseudoalign_stats <- read_excel("raw_data/GSE206023_pseudoalignment_ST1.xlsx", 
    sheet = "General Statistics")
```

# Principal Variance Components Analysis

## Counts normalization

```{r}
##### DESEQ2 Normalization
library(DESeq2)

ann_vaccines_samples = read_csv(here("Annotations and metadata", "ann_vaccines_samples.csv"))
all_matrices_counts = readRDS(here("DGE analysis", "all_matrices_counts.rds"))

cvrts = ann_vaccines_samples %>% 
  arrange(condition) %>%
  distinct() %>% 
  column_to_rownames("sample")

exprs_deseq = all_matrices_counts %>% 
  column_to_rownames("genes") %>% 
  replace(is.na(.), 0) %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column("sample") %>% 
  arrange(sample) %>% 
  inner_join(cvrts %>% 
               rownames_to_column("sample"),
               select(sample) %>% 
               distinct(), 
             by = "sample") %>% 
  column_to_rownames("sample") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(., . == -Inf, 0) %>% 
  replace(., . == Inf, 0) %>% 
  replace(is.na(.), 0)


cvrts = cvrts %>% 
  mutate(condition = as.factor(condition),
         age = as.factor(age),
         sex = as.factor(sex),
         vaccine = as.factor(vaccine),
         day = as.factor(day),
         gse_id = as.factor(gse_id))
  
#Check dimensions
dim(exprs)
dim(cvrts)

#Normalize data
dds <- DESeqDataSetFromMatrix(countData = exprs_deseq, 
                              colData = cvrts, 
                              design = ~ condition + gse_id)

dds <- estimateSizeFactors(dds)

normalized_counts <- dds %>% 
  counts(normalized=TRUE) %>% 
  as.data.frame()

saveRDS(normalized_counts, file = here("DGE analysis", "all_matrices_normalized_counts.rds"))

normalized_counts
```

## PVCA of normalized raw counts

```{r cache = TRUE}
#### PVCA of normalized raw counts -----
# BiocManager::install("ExpressionNormalizationWorkflow")
library(ExpressionNormalizationWorkflow)

#Expression data
exprs = all_matrices_normalized_counts %>% 
  replace(is.na(.), 0) %>% 
  replace(., . == -Inf, 0) %>% 
  replace(., . == Inf, 0) %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column("sample") %>% 
  arrange(sample) %>% 
  inner_join(cvrts %>% 
               #filter(disease_vac == "V") %>% 
               rownames_to_column("sample"),
               select(sample) %>% 
               distinct(), 
             by = "sample") %>% 
  column_to_rownames("sample") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(., . == -Inf, 0) %>% 
  replace(., . == Inf, 0) %>% 
  replace(is.na(.), 0)

exprs
```

Covariants table

```{r cache = TRUE}
cvrts = cvrts %>% 
  rownames_to_column("sample") %>% 
  inner_join(exprs %>% 
               colnames() %>% 
               as.data.frame() %>% 
               select(sample = "."), 
             by = "sample") %>% 
  arrange(sample) %>% 
  column_to_rownames("sample")
cvrts
```

Inputs
```{r cache = TRUE}
# Input
#For vaccination only (no ongoing infection)
cvrts_eff_var =  c("day", 
                   "age", 
                   "sex", 
                   "dose", 
                   "type", 
                   "regimen") 

#For vaccination and infection
cvrts_eff_var = c(
  "day",
  "age",
  "sex",
  "dose",
  "type",
  "previous_vaccination",
  "variant",
  "prior_infection",
  "severity",
  "disease_vac",
  "regimen"
) 

#Convert to Expression Dataset Object
inpData <- expSetobj(exprs, cvrts) 

#Threshold
pct_thrsh <- 0.75 
```

Run PVCA analysis

```{r cache = TRUE}
#PVCA analysis
pvca = pvcAnaly(inpData, pct_thrsh, cvrts_eff_var) 

pvca_df = pvca$dat %>% 
  as.data.frame() %>% 
  t() %>%
  as.data.frame() %>% 
  rename(wavgpv = V1) %>% 
  cbind(pvca$label %>% as.data.frame() %>% rename(batch = '.')) %>% 
  as.data.frame() %>% 
  mutate(batch = as.factor(batch),
         wavgpv = as.numeric(wavgpv),
         batch = fct_rev(fct_reorder(batch, wavgpv))) %>% 
  arrange(batch, wavgpv)

pvca_df %>% saveRDS(here("DGE analysis", "pvca_results.rds"))

pvca_df
```

Barplot: Covariants contributions

```{r cache = TRUE}
# Visualize
pvca_df_plot = ggplot(pvca_df) +
  aes(x = batch, y = wavgpv) +
  geom_col(fill = "#5e60ce") +
  theme_minimal() +
  theme(axis.text.x = element_text(size = 20, angle = 45, vjust = 1, hjust=1, color = "black"),
        axis.text.y = element_text(size = 20, vjust = 1, hjust=1, color = "black")) +
  geom_text(aes(label = round(wavgpv,3), y = wavgpv), 
            position = position_dodge(width = 0.9), 
            vjust = -0.2, 
            hjust = 0.5, 
            size = 4,
            angle = 0) 

ggsave(pvca_df_plot, file = here("Figures", "pvca_raw_normalized_immunegenes_allconditions.png"), width = 12, height = 10)

pvca_df_plot
```

Scatterplot of PVCA

```{r cache = TRUE}
# Visualize
library(ggfortify)
library(plotly)

pca_res <- prcomp(normalized_counts %>% t() %>% as.data.frame() %>% select(-day), scale. = F)

pca_res_ann = normalized_counts %>% 
  t() %>% 
  as.data.frame() %>% 
    select(-day) %>% 
  rownames_to_column("sample") %>% 
  inner_join(cvrts %>% 
               rownames_to_column("sample"),
             by = "sample") %>% 
  distinct()

dim(pca_res)
dim(pca_res_ann)

normalizedcounts_pca_plot = autoplot(pca_res, data = pca_res_ann)+
  geom_point(aes(fill = gse_id, label = condition), colour="black",pch=21, size = 3) +
  theme_few() +
  scale_fill_aaas()

ggsave(normalizedcounts_pca_plot, file = here("Figures", "covid_vax_pca_samples.png"))

normalizedcounts_pca_plot %>% ggplotly()

```

# 3. General data analysis

## 3.1. Metadata

Merge metadata

```{r eval=FALSE, include=TRUE}
#Metadata ------
gse189039_metadata_2 = GSE189039_metadata_1_ %>% 
  separate(sample, into = c("participant_id"), sep = "_", remove = F) %>% 
  mutate(geo = "GSE189039") %>% 
  rename(day = timepoint_day,
         gsm_id = geo_sample) %>% 
  select(sample, participant_id)

gse199750_metadata_2 = gse199750_metadata_annotated_14_11_23_2_ %>% 
  rename(gsm_id = geo,
         participant_id = subject_id) %>% 
  mutate(geo = "GSE199750") %>% 
  select(sample, participant_id)

gse206023_metadata_3 = gse206023_metadata %>% 
  mutate(day = as.numeric(day),
         geo = "GSE206023") %>%  
  inner_join(gse206023_metadata_ready_2, by = "sample_id") %>% 
  inner_join(ann_vaccines_samples %>% 
            select(condition, sample_id = sample, age, sex, vaccine),
            by = "sample_id") %>% 
  rename(gsm_id = gsm_id.x) %>% 
  select(sample = sample_id, participant_id)

gse201530_metadata_2 = gse201530_metadata %>% 
  rename(gsm_id = geo_sample) %>% 
  separate(subject_id, into = c("variant", "participant_id_num"), sep = "_", remove = F) %>% 
  mutate(participant_id = paste0(variant, "_", participant_id_num)) %>% 
  select(participant_id, sample = subject_id)

gse201533_metadata_2 = GSE201533_metadata %>% 
  mutate(age = as.numeric(age),
         dose = as.numeric(dose),
         gsm_id = geo_sample) %>% 
  select(geo_sample = sample_id, participant_id) %>% 
  rownames_to_column("sample")
```

Annotate samples

```{r eval=FALSE, include=TRUE}
ann_vaccines_samples_2.1 = ann_vaccines_samples_2 %>% 
  filter(is.na(participant_id)) %>% 
  separate(sample, into = c("participant_id_1", "participant_id_2", "participant_id_3", "delete"), remove = F) %>% 
  mutate(participant_id = paste0(participant_id_1, "_", participant_id_2, "_", participant_id_3)) %>% 
  select(!c(participant_id_1, participant_id_2, participant_id_3)) %>% 
  select(sample, participant_id)

all_participants = bind_rows(gse189039_metadata_2,
          gse199750_metadata_2,
          gse206023_metadata_3,
          gse201533_metadata_2,
          gse201530_metadata_2) %>% 
  bind_rows(ann_vaccines_samples_2.1)

ann_vaccines_samples_2 = ann_vaccines_samples %>% 
  left_join(all_participants, by = "sample") %>% 
  select(condition, sample, age, sex, vaccine:previous_vaccination, participant_id) %>% 
  left_join(ann_vaccines_samples_2.1, by = "sample")


ann_vaccines_samples_participants = ann_vaccines_samples_2 %>% 
  select(-participant_id.y, participant_id = participant_id.x)
  

write.csv(ann_vaccines_samples_participants, file = "ann_vaccines_samples.csv")
```

Conditions annotations

```{r eval=FALSE, include=TRUE}
library(tidyverse)

demographics = ann_vaccines_samples_4_12_23 %>%
  select(vaccine, gse_id, condition, sample) %>% 
  distinct() %>% 
  group_by(vaccine, gse_id, condition) %>% 
  summarise(samples = n()) %>% 
  ungroup()

ann_vaccines = demographics %>% 
  select(condition, samples) %>% 
  inner_join(ann_vaccines_18_1_24, by = "condition") %>% 
  distinct() %>% 
  mutate(week = as.factor(week))

ann_vaccines_2 = ann_vaccines %>% 
  group_by(vaccine) %>% 
  summarise(samples_byvaccine = sum(samples)) %>% 
  ungroup() %>% 
  inner_join(ann_vaccines, by = "vaccine") 


 ann_vaccines_3 = ann_vaccines_2 %>% 
  group_by(gse_id) %>% 
  summarise(samples_bygse = sum(samples)) %>% 
  ungroup() %>% 
  inner_join(ann_vaccines_2, by = "gse_id") 

write.csv(ann_vaccines_3, file = "ann_vaccines_19-1-24.csv")
```

Samples by condition

```{r fig.height=10, fig.width=20}

ann_vaccines = read_csv(here("Annotations and metadata", "ann_vaccines_conditions.csv"))

ann_vaccines_samples_participants = read_csv(here("Annotations and metadata", "ann_vaccines_samples.csv")) 

#### SAMPLES BY CONDITION -----
samples_by_condition  = ann_vaccines %>% 
  select(samples, condition, vaccine) %>% 
  distinct() %>% 
  group_by(condition) %>% 
  summarise(samples = sum(samples)) %>% 
  inner_join(ann_vaccines %>% select(vaccine, condition) %>% distinct(), by = "condition")


samples_by_condition_plot = samples_by_condition %>% 
  ggplot(aes(x = fct_reorder(condition, -samples), weight = samples, fill = vaccine)) +
  geom_bar() +
  scale_fill_manual(name = "Vaccine or infection", labels = c("BBIBP", 
                                                                "BNT", 
                                                                "ChAd",
                                                                "Healthy",
                                                                "Infection",
                                                                "MO",
                                                                "ZF2001"),
    values = c("BBIBP" = "#00A087FF",
                "ZF2001" = "#8491B4FF",
                "BNT" = "#4DBBD5FF",
                "MO" = "#72ddf7",
                "ChAd" = "#3C5488FF",
                "I" = "#DC0000FF",
                "H" = "grey80")) +
  theme_minimal() +
  theme(legend.position = "inside",
        axis.text.x = element_text(size = 15, color = "black", angle = 45, hjust = 1, vjust = 1.1),
        axis.text.y = element_blank(), 
        axis.title.y = element_text(angle = 180, color = "white"),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        plot.margin = margin(2, 2, 2, 2, "cm")) +
  geom_label(aes(label = samples, y = samples),
             fill = "white",
            position = position_dodge(width = 0), 
            vjust = -0.5, 
            # hjust = 1, 
            size = 4,
            angle = 0) +
  ggtitle("Samples by condition") +
  labs(x = "") +
  ylim(0, 80)

ggsave(samples_by_condition_plot, file = here("Figures", "demographics_samples-by-condition.png"), width = 20, height = 10)

samples_by_condition_plot
```

### 3.1.2. Age by GSE

```{r fig.height=8, fig.width=10}
#### AGE BY GSE -------------
age = ann_vaccines_samples_participants %>% 
  select(age, participant_id, gse_id) %>% 
  distinct()

age_stats = ggbetweenstats(
  data = age,
  x = gse_id,
  y = age,
  type = "np", 
  p.adjust.method = "BH"
) +
  ggplot2::scale_color_manual(
    values = c("GSE199750" = "#80ffdb",
               "GSE201533" = "#5e60ce", 
               "GSE201530" = "#7400b8",
               "GSE206023" = "#4361ee",
               "GSE189039" = "#48bfe3")) + 
  theme_few() +
  ylim(0, 110) +
  theme(legend.position = "right",
        axis.text.x = element_text(size = 10, color = "black", angle = 0),
        axis.text.y = element_text(size = 10, color = "black", angle = 0)) +
  labs(y = "Age",
       x = "")

ggsave(age_stats, file = here("Figures", "Demographics_age.png") , width = 10, height = 8)

age_stats
```

### 3.1.3. Participants by condition

```{r}
#### Participants by condition ------

ann_vaccines_samples %>% 
  select(condition, disease_vac) %>% 
  group_by(condition) %>% 
  summarise(samples = n()) %>% 
  inner_join(ann_vaccines_condition, by = "condition") %>% 
  write_csv(here("Annotations and metadata", "ann_vaccines_conditions.csv"))


participants = Studies_Participants %>% 
  select(doses_abrev, participants, vac_infection) %>% 
  distinct() %>% 
  group_by(dose) %>% 
  summarise(n_total_participants = sum(participants)) %>% 
  inner_join(Studies_Participants, by = "doses_abrev") %>% 
  rename(condition = doses_abrev) %>% 
    select(condition, n_total_participants, vac_infection) %>% 
  distinct()

p = participants %>%
 filter(!is.na(condition)) %>% 
  ggplot(aes(x = fct_reorder(condition, desc(n_total_participants)), 
             weight = n_total_participants)) +
  geom_bar(fill = "#5e60ce") +
  theme_minimal() +
  scale_fill_manual(values = "#5e60ce") +
  theme(legend.position = "right",
        axis.text.x = element_text(size = 16, color = "black", angle = 45, vjust = 1, hjust = 1),
        axis.text.y = element_blank(),
        panel.grid.minor = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.major.x = element_line(color = "gray70",
                                          size = 0.1,
                                          linetype = 1)) +
  geom_text(aes(label = n_total_participants, y = n_total_participants), 
            position = position_dodge(width = 0), 
            vjust = -0.2, 
            hjust = 0.5, 
            size = 5,
            angle = 0) +
  ylim(0, 70)+
  ggtitle("Participants by condition") +
  labs(x = "Conditions",
       y = "Participants")
p
ggsave(p, file = "demographics_participants-by-condition.png", width = 12, height = 6)

```

# 3. Differential Gene Expression Analysis

## GSE201530

Metadata: A tabela de counts fornecida pelo estudo é nomeada com o
Subject ID + Timepoint e não com o codigo GSM. Por isso, teremos de
tratar a tabela de metadados. Além disso, ela não possui a coluna de
idade. É possível manipular estes dados no excel.

```{r eval=FALSE, include=FALSE}
#Standardize counts tables
############### GSE201530
# Reorder table
gse201530_counts_ready <- counts_gse201530 %>%
  column_to_rownames("genes") %>% 
  t() %>%
  as.data.frame() %>% 
  rownames_to_column("sample") %>% 
  arrange(sample) %>% 
  column_to_rownames("sample") %>% 
  t() %>% 
  as.data.frame() %>%
  rownames_to_column("genes") %>% 
  filter(!grepl("_", genes)) %>% 
  column_to_rownames("genes")

saveRDS(gse201530_counts_ready, file = "gse201530_counts_ready.rds")
```

```{r}
###DESEQ2 - PF versus AZ, PF em diferentes tempos, AZ em diferentes tempos

####Definindo design

###### Padronize as variáveis 
countData <- gse201530_counts_ready
colData <- GSE201530_metadata %>%
  mutate(
    timepoint_day = factor(timepoint_day),
    age = factor(age),
    sex = factor(sex),
    disease_vac = factor(disease_vac),
    dose = factor(dose),
    disease_time = paste0(disease_vac, "_D", timepoint_day)) %>% 
  select(subject_id, everything())


###### Crie o objeto DESeqDataSet ######
# O design deve incluir todos os fatores que podem influenciar significativamente os dados. Coloque por último o fator de interesse principal - no caso, o fator de primeira ou segunda doses, pois nesta primeira etapa não serão analisados os efeitos de uma terceira dose de vacina heteróloga.

dds <- DESeqDataSetFromMatrix(countData = countData, 
                              colData = colData, 
                              design = ~ disease_time)

###### Pré-filtragem ######
#Mantenha os genes com pelo menos 10 leituras no total.

keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]

###### Executar DESeq ######

#Executar DESeq com método Wald
dds <- DESeq(dds)

#Salvar
write_rds(dds,file="gse201530_dds_DESeq2_Wald_disease_vac_time.rds")

###### Obter comparações realizadas ######
resultsNames(dds)

dds = gse201530_dds_DESeq2_Wald_disease_vac_timepoint_day

###### Obtendo os DEGs totais ######
res <- results(dds) %>% 
  as.data.frame() %>% 
  filter(pvalue < 0.05)

#Salvar tabela
write.csv(res, file = "gse201530_dds_DESeq2_Wald_disease_vac_AND_timepoint_day_ALLDEGS.rds", row.names = TRUE)

```

Contrasts

```{r}
#H-BNT-I
# contrast = c("disease_time", "H.BNT.I_D1", "H_D0")
# contrast = c("disease_time", "H.BNT.I_D2", "H_D0") (NAO DEU)
# contrast = c("disease_time", "H.BNT.I_D3", "H_D0")
# contrast = c("disease_time", "H.BNT.I_D4", "H_D0")

#I-BNT-I
# contrast = c("disease_time", "I.BNT.I_D2", "H_D0")
# contrast = c("disease_time", "I.BNT.I_D5", "H_D0")

#H-I
# contrast = c("disease_time", "H.I_D1", "H_D0")

#H-I-I
# contrast = c("disease_time", "H.I.I_D2", "H_D0")
#I-I
# contrast = c("disease_time", "I.I_D1", "H_D0")
# contrast = c("disease_time", "I.I_D3", "H_D0")
# contrast = c("disease_time", "I.I_D5", "H_D0")
```

**H_BNT_I**

```{r}
######### H_BNT_I

#H_BNT_I_D1
gse201530_H_BNT_I_D1 = results(dds, contrast = c("disease_time", "H.BNT.I_D1", "H_D0")) %>% 
  as.data.frame() %>% 
  arrange(padj) %>% 
  mutate(condition = "H-BNT-I (D1)",
         day = 1,
         vaccine = "H-BNT-I",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse201530_H_BNT_I_D1, file = "gse201530_H_BNT_I_D1_DEGs_notfiltered.csv", row.names = TRUE)

#H_BNT_I_D3
gse201530_H_BNT_I_D3 = results(dds, contrast = c("disease_time", "H.BNT.I_D3", "H_D0")) %>%
  as.data.frame() %>%
  arrange(padj) %>%
  mutate(condition = "H-BNT-I (D3)",
         day = 3,
         vaccine = "H-BNT-I",
         direction = ifelse(log2FoldChange >= 1, "UP",
                            ifelse(log2FoldChange <= -1, "DOWN",
                                   "NEUTRAL"))) %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse201530_H_BNT_I_D3, file = "gse201530_H_BNT_I_D3_DEGs.csv", row.names = TRUE)

#H_BNT_I_D4
gse201530_H_BNT_I_D4 = results(dds, contrast = c("disease_time", "H.BNT.I_D4", "H_D0")) %>%
  as.data.frame() %>%
  filter(padj < 0.05) %>%
  arrange(padj) %>%
  mutate(condition = "H-BNT-I (D4)",
         day = 4,
         vaccine = "H-BNT-I",
         direction = ifelse(log2FoldChange >= 1, "UP",
                            ifelse(log2FoldChange <= -1, "DOWN",
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse201530_H_BNT_I_D4, file = "gse201530_H_BNT_I_D4_DEGs.csv", row.names = TRUE)

gse201530_H_BNT_I = bind_rows(gse201530_H_BNT_I_D1,
                            gse201530_H_BNT_I_D3,
                            gse201530_H_BNT_I_D4)

write.csv(gse201530_H_BNT_I, "gse201530_H_BNT_I.csv")

```

**I_BNT_I**

```{r}
######### I_BNT_I

#D2
gse201530_I_BNT_I_D2 = results(dds, contrast = c("disease_time", "I.BNT.I_D2", "H_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "I-BNT-I (D2)",
         day = 2,
         vaccine = "I-BNT-I",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse201530_I_BNT_I_D2, file = "gse201530_I_BNT_I_D2_DEGs.csv")


#D5
gse201530_I_BNT_I_D5 = results(dds, contrast = c("disease_time", "I.BNT.I_D5", "H_D0")) %>%
  as.data.frame() %>%
  filter(padj < 0.05) %>%
  arrange(padj) %>%
  mutate(condition = "I-BNT-I (D5)",
         day = 5,
         vaccine = "I-BNT-I",
         direction = ifelse(log2FoldChange >= 1, "UP",
                            ifelse(log2FoldChange <= -1, "DOWN",
                                   "NEUTRAL"))) %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse201530_I_BNT_I_D5, file = "gse201530_H_BNT_I_D5_DEGs.csv")

gse201530_I_BNT_I = bind_rows(gse201530_I_BNT_I_D2,
                            gse201530_I_BNT_I_D5)

write.csv(gse201530_I_BNT_I, "gse201530_I_BNT_I.csv")

```

*H-I*

```{r}
######### H-I

#D1
gse201530_H_I_D1 = results(dds, contrast = c("disease_time", "H.I_D1", "H_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "H-I (D1)",
         day = 1,
         vaccine = "H-I",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse201530_H_I_D1, file = "gse201530_H_I_D1_DEGs-PF.csv")


#D5
gse201530_H_I_D5 = results(dds, contrast = c("disease_time", "I.BNT.I_D5", "H_D0")) %>%
  as.data.frame() %>%
  filter(padj < 0.05) %>%
  arrange(padj) %>%
  mutate(condition = "H-I (D5)",
         day = 5,
         vaccine = "H-I",
         direction = ifelse(log2FoldChange >= 1, "UP",
                            ifelse(log2FoldChange <= -1, "DOWN",
                                   "NEUTRAL"))) %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse201530_H_I_D5, file = "gse201530_H_I_D5_DEGs-PF.csv")

gse201530_H_I = bind_rows(gse201530_I_BNT_I_D2,
                            gse201530_I_BNT_I_D5)

write.csv(gse201530_H_I, "gse201530_H_I.csv")

```

*H-I-I*

```{r}

######### H-I-I
#D2
gse201530_H_I_I_D2 = results(dds, contrast = c("disease_time", "H.I.I_D2", "H_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "H-I-I (D2)",
         day = 2,
         vaccine = "H-I-I",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse201530_H_I_I_D2, file = "gse201530_H_I_I_D2_DEGs-PF.csv")

```

*I-I*

```{r}

######### I-I
#D1
gse201530_I_I_D1 = results(dds, contrast = c("disease_time", "I.I_D1", "H_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "I-I (D1)",
         day = 1,
         vaccine = "I-I",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse201530_I_I_D1, file = "gse201530_I_I_D1_DEGs-PF.csv")


#D3
gse201530_I_I_D3 = results(dds, contrast = c("disease_time", "I.I_D3", "H_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "I-I (D3)",
         day = 3,
         vaccine = "I-I",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse201530_I_I_D3, file = "gse201530_I_I_D3_DEGs-PF.csv")


#D5
gse201530_I_I_D5 = results(dds, contrast = c("disease_time", "I.I_D5", "H_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "I-I (D5)",
         day = 5,
         vaccine = "I-I",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse201530_I_I_D5, file = "gse201530_I_I_D5_DEGs-PF.csv")

gse201530_I_I = bind_rows(gse201530_I_I_D1,
                          gse201530_I_I_D3,
                          gse201530_I_I_D5)

```

**Unir raw datasets**

```{r}
########## GSE201530

gse201530_DEGs = bind_rows(gse201530_I_I,
                           gse201530_H_I_I_D2,
                           gse201530_H_I,
                           gse201530_I_BNT_I,
                           gse201530_H_BNT_I) %>% 
  mutate(study = "GSE201530")
write.csv(gse201530_DEGs, file = "gse201530_DEGs_27-11-23.csv")

GSE201530_metadata_conditions = gse201530_DEGs %>% 
  select(condition:vaccine, study) %>% 
  distinct()


#Annotations
ann_vaccines_lower = ann_vaccines %>% 
  clean_names()

ann_vaccines_27_11_23 = bind_rows(ann_vaccines_lower, 
                                  GSE201530_metadata_conditions)

write.csv(ann_vaccines_27_11_23, file = 'ann_vaccines_27_11_23.csv')


```

## GSE189039

```{r eval=FALSE, include=FALSE}
############### GSE189039
# Reorder table
gse189039_counts_ready <- counts_GSE189039 %>%
  column_to_rownames("genes") %>% 
  t() %>%
  as.data.frame() %>% 
  rownames_to_column("sample") %>% 
  arrange(sample) %>% 
  column_to_rownames("sample") %>% 
  t() %>% 
  as.data.frame() %>%
  rownames_to_column("genes") %>% 
  filter(!grepl("_", genes)) %>% 
  column_to_rownames("genes")

saveRDS(gse189039_counts_ready, file = "gse189039_counts_ready.rds")

```

```{r}

###### Padronize as variáveis 
countData <- gse189039_counts_ready
colData <- GSE189039_metadata %>%
  mutate(disease_vac = ifelse(disease_vac == "I", "H-I",
                              ifelse(disease_vac == "BNT-I", "H-BNT-I", disease_vac)),
         disease_vac_severity_time = paste0(disease_vac, "_", severity, "_D", timepoint_day),
         disease_vac_severity_time = factor(disease_vac_severity_time)) %>% 
  select(subject_id, everything())


###### Crie o objeto DESeqDataSet ######
dds <- DESeqDataSetFromMatrix(countData = countData, 
                              colData = colData, 
                              design = ~ disease_vac_severity_time)

keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]

###### Executar DESeq ######
dds <- DESeq(dds)
write_rds(dds,file="gse189039_dds_DESeq2_Wald_disease_vac_severity_time.rds")

###### Obter comparações realizadas ######
resultsNames(dds)

###### Obtendo os DEGs totais ######
res <- results(dds) %>% 
  as.data.frame() %>% 
  filter(pvalue < 0.05)

#Salvar tabela
write.csv(res, file = "gse189039_dds_DESeq2_Wald_disease_vac_severity_time_ALLDEGS.rds")

```

### Infection

```{r}
## H-I
# MODERATE
# contrast = c("disease_vac_severity_time", "H.I_moderate_D10", "H_naive_vac_D0")
# contrast = c("disease_vac_severity_time", "H.I_moderate_D26", "H_naive_vac_D0")
# contrast = c("disease_vac_severity_time", "H.I_moderate_D51", "H_naive_vac_D0")
```

```{r}
# SEVERE
# contrast = c("disease_vac_severity_time", "H.I_severe_D10", "H_naive_vac_D0")
# contrast = c("disease_vac_severity_time", "H.I_severe_D26", "H_naive_vac_D0")
# contrast = c("disease_vac_severity_time", "H.I_severe_D51", "H_naive_vac_D0")
```

*H_I*

```{r}
######### H_I

#H_I_D10
gse189039_H_I_moderate_D10 = results(dds, contrast = c("disease_vac_severity_time", "H.I_moderate_D10", "H_naive_vac_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "H-I (D10, moderate)",
         day = 10,
         vaccine = "H-I",
         severity = "moderate",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse189039_H_I_moderate_D10, file = "gse189039_H_I_moderate_D10_DEGs.csv")

#H_I_D26
gse189039_H_I_moderate_D26 = results(dds, contrast = c("disease_vac_severity_time", "H.I_moderate_D26", "H_naive_vac_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "H-I (D26, moderate)",
         day = 26,
         severity = "moderate",
         vaccine = "H-I",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse189039_H_I_moderate_D26, file = "gse189039_H_I_moderate_D26_DEGs.csv")

#H_I_D51
gse189039_H_I_moderate_D51 = results(dds, contrast = c("disease_vac_severity_time", "H.I_moderate_D51", "H_naive_vac_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "H-I (D51, moderate)",
         day = 51,
         vaccine = "H-I",
         severity = "moderate",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse189039_H_I_moderate_D51, file = "gse189039_H_I_moderate_D51_DEGs.csv")

#H_I_severe_D10
gse189039_H_I_severe_D10 = results(dds, contrast = c("disease_vac_severity_time", "H.I_severe_D10", "H_naive_vac_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "H-I (D10, severe)",
         day = 10,
         vaccine = "H-I",
         severity = "severe",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse189039_H_I_severe_D10, file = "gse189039_H_I_severe_D10_DEGs.csv")



#H_I_severe_D26
gse189039_H_I_severe_D26 = results(dds, contrast = c("disease_vac_severity_time", "H.I_severe_D26", "H_naive_vac_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "H-I (D26, severe)",
         day = 26,
         vaccine = "H-I",
         severity = "severe",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse189039_H_I_severe_D26, file = "gse189039_H_I_severe_D26_DEGs.csv")

#H_I_severe_D51
gse189039_H_I_severe_D51 = results(dds, contrast = c("disease_vac_severity_time", "H.I_severe_D51", "H_naive_vac_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "H-I (D51, severe)",
         day = 51,
         vaccine = "H-I",
         severity = "severe",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse189039_H_I_severe_D51, file = "gse189039_H_I_severe_D51_DEGs.csv")

gse189039_H_I = bind_rows(gse189039_H_I_moderate_D10,
                          gse189039_H_I_severe_D10,
                            gse189039_H_I_moderate_D26,
                          gse189039_H_I_severe_D26,
                            gse189039_H_I_moderate_D51,
                          gse189039_H_I_severe_D51)

write.csv(gse189039_H_I, "gse189039_H_I_DEGS.csv")

```

### Vaccination

```{r}
## H-BNT-I
# MILD
# contrast = c("disease_vac_severity_time", "H.BNT.I_mild_D10", "H_naive_vac_D0")
# contrast = c("disease_vac_severity_time", "H.BNT.I_mild_D26", "H_naive_vac_D0")

# SEVERE
# contrast = c("disease_vac_severity_time", "H.BNT.I_severe_D10", "H_naive_vac_D0")
# contrast = c("disease_vac_severity_time", "H.BNT.I_severe_D26", "H_naive_vac_D0")

## BNT-I-BNT
# contrast = c("disease_vac_severity_time", "BNT.I.BNT_severe_D51", "H_naive_vac_D0")
# contrast = c("disease_vac_severity_time", "BNT.I.BNT_mild_D51", "H_naive_vac_D0") 

```

*H_BNT_I*

```{r}
######### H-BNT-I

#H_BNT-I

#MILD
gse189039_H_BNT_I_mild_D10 = results(dds, contrast = c("disease_vac_severity_time", "H.BNT.I_mild_D10", "H_naive_vac_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "H-BNT-I (D10, mild)",
         day = 10,
         vaccine = "H-BNT-I",
         severity = "mild",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse189039_H_BNT_I_mild_D10, file = "gse189039_H_BNT_I_mild_D10_DEGs.csv")


gse189039_H_BNT_I_mild_D26 = results(dds, contrast = c("disease_vac_severity_time", "H.BNT.I_mild_D26", "H_naive_vac_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "H-BNT-I (D26, mild)",
         day = 26,
         vaccine = "H-BNT-I",
         severity = "mild",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse189039_H_BNT_I_mild_D26, file = "gse189039_H_BNT_I_mild_D26_DEGs.csv")


#SEVERE

gse189039_H_BNT_I_severe_D10 = results(dds, contrast = c("disease_vac_severity_time", "H.BNT.I_severe_D10", "H_naive_vac_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "H-BNT-I (D10, severe)",
         day = 10,
         vaccine = "H-BNT-I",
         severity = "severe",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse189039_H_BNT_I_severe_D10, file = "gse189039_H_BNT_I_severe_D10_DEGs.csv")


gse189039_H_BNT_I_severe_D26 = results(dds, contrast = c("disease_vac_severity_time", "H.BNT.I_severe_D26", "H_naive_vac_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "H-BNT-I (D26, severe)",
         day = 26,
         vaccine = "H-BNT-I",
         severity = "severe",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse189039_H_BNT_I_severe_D26, file = "gse189039_H_BNT_I_severe_D26_DEGs.csv")

#UNIR
gse189039_H_BNT_I = bind_rows(gse189039_H_BNT_I_mild_D10,
                              gse189039_H_BNT_I_severe_D10,
                              gse189039_H_BNT_I_mild_D26,
                              gse189039_H_BNT_I_severe_D26)
#Salvar
write.csv(gse189039_H_BNT_I, file = "gse189039_H_BNT_I_DEGS.csv")

```

*BNT-I-BNT*

```{r}
######### BNT-I-BNT	

# D51
gse189039_BNT_I_BNT_severe_D51 = results(dds, contrast = c("disease_vac_severity_time",
                                                         "BNT.I.BNT_severe_D51", 
                                                         "H_naive_vac_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "BNT-I-BNT (D51, severe)",
         day = 51,
         vaccine = "BNT-I-BNT",
         severity = "severe",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse189039_BNT_I_BNT_severe_D51, file = "gse189039_BNT_I_BNT_severe_D51_DEGs.csv")


#Aqui tive de usar o argumento name, pois contrast não estava funcionando (?)
gse189039_BNT_I_BNT_mild_D51 = results(dds, name= "disease_vac_severity_time_H_naive_vac_D0_vs_BNT.I.BNT_mild_D51") %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  mutate(log2FoldChange = ifelse(log2FoldChange > 0, -log2FoldChange, abs(log2FoldChange))) %>% 
  arrange(padj) %>% 
  mutate(condition = "BNT-I-BNT (D51, mild)",
         day = 51,
         vaccine = "BNT-I-BNT",
         severity = "mild",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes") 

#Salvar tabela
write.csv(gse189039_BNT_I_BNT_mild_D51, file = "gse189039_BNT_I_BNT_mild_D51_DEGs.csv")

#Unir
gse189039_BNT_I_BNT = bind_rows(gse189039_BNT_I_BNT_mild_D51,
                                gse189039_BNT_I_BNT_severe_D51)


write.csv(gse189039_BNT_I_BNT, file = "gse189039_BNT_I_BNT_DEGS.csv")

```

*UNIR*

```{r}
gse189039_DEGs = bind_rows(gse189039_BNT_I_BNT,
                           gse189039_H_I,
                           gse189039_H_BNT_I) %>% 
  mutate(study = "GSE189039")



##### TODAS AS DEGS

gse_201530_189039 = bind_rows(gse201530_DEGs,
                              gse189039_DEGs) %>% 
  select(genes:condition, direction) %>% 
  merge(ann_vaccines, by = "condition", all.y = F) %>% 
  clean_names()

degs_allstudies_updown_p_05 = degs_allstudies_updown_p_05 %>% 
  clean_names() %>% 
  select(genes:direction, -vaccine, -gse_id) %>% 
  merge(ann_vaccines, by = "condition", all.y = F) %>% 
  select(-participants)


all_degs_p_05_vac_infected = bind_rows(degs_allstudies_updown_p_05,
                                       gse_201530_189039) %>% 
  mutate_all(~ replace(., is.na(.), "NA")) %>% 
  mutate(regimen = case_when(
regimen == "BNT-I-BNT" ~ "V-I-V",
regimen == "H-BNT-I" ~ "V-I",
regimen == "VAC-I" ~ "V-I",
regimen == "I-VAC-I" ~ "I-V-I",
regimen == "H-V-I" ~ "V-I",
regimen == "H-I-I" ~ "I-I",
regimen == "H-I" ~ "I",
TRUE ~ as.character(regimen)))

#Salvar
write.csv(gse_201530_189039, file = "gse_201530_189039_DEGs_27-11-23.csv")

#Salvar
write.csv(all_degs_p_05_vac_infected, file = 'all_degs_p_05_vac_infected_27-11-23.csv')
```

## GSE206023

```{r}
##Defina as variáveis
countData <- gse206023_counts_ready
colData <- gse206023_metadata #coloque metadata_ready caso faça o upload direto

## Padronize as variáveis. Para que o objeto DESeq seja criado, os valores da matriz de contagem devem ser inteiros. 

# Verifique se os valores em countData são números inteiros
if (!all(is.integer(countData))) {
  # Se os valores não forem inteiros, faça a conversão
  countData <- as.matrix(round(countData))
}

# Defina os fatores. 
colData$Timepointf <- factor(colData$Timepoints)
colData$Vaccinef <- factor(colData$Vaccine)

# Crie o objeto DESeqDataSet. O design deve incluir todos os fatores que podem influenciar significativamente os dados. Coloque por último o fator de interesse principal, que no caso considera a interação entre fatores, já que são comparadas vacinas em diferentes timepoints. 

dds <- DESeqDataSetFromMatrix(countData = countData, colData = colData, design = ~Timepointf + Vaccinef + Timepointf:Vaccinef)


#Pré-filtragem. Mantenha os genes com pelo menos 10 leituras no total.
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]

# Executar DESeq
dds <- DESeq(dds)
write_rds(dds,file="gse206023_dds_DESeq2_time_vaccine.rds")

# Obtendo os genes diferencialmente expressos:
res <- results(dds)
res

# Para obter informações sobre quais variáveis e testes foram usados e quantos genes foram significativamente up (LFC > 0) e down (LFC < 0)regulados:
summary (res)

#Salvar tabela deseq
write_rds(res,file="gse206023_res_DESeq2_time_vaccine.rds")

```

## GSE201533 (ORGANIZAR)

```{r}
############ Matrix counts ############

gse201533_counts <- as.data.frame(counts_gse201533)
rownames(gse201533_counts) <- gse201533_counts$geneid 
gse201533_counts <- gse201533_counts[, -1]
saveRDS(gse201533_counts, file="gse201533_counts_ready.rds")

########## Metadata #############
gse201533_metadata <- as.data.frame(gse201533_metadata)
rownames(gse201533_metadata) <- gse201533_metadata$GEO 

#A coluna de Sample name corresponde aos timepoints da coleta de amostra pós vacinacao. Por isso, vamos extrair somente o numero do timepoint (0,2,3,4,7)

gse201533_metadata$Timepoint <- sub(".*(0|2|3|4|7).*", "\\1", gse201533_metadata$`Sample name`)

head(gse201533_metadata)

gse201533_metadata <- gse201533_metadata[, -2]
gse201533_metadata <- gse201533_metadata[, -1]

saveRDS(gse201533_metadata, file="gse201533_metadata_ready.rds")
```

Caso a tabela apresente valores NA na coluna de hgnc_symbol, execute
este comando

```{r}
#########Caso a tabela apresente valores NA na coluna de hgnc_symbol, execute este comando
any_na <- any(is.na(gse201533_counts$geneid))
print(any_na) 
any_duplicates <- any(duplicated(gse201533_counts$geneid))
print(any_duplicates)
rows_with_na <- which(is.na(gse201533_counts$geneid))
print(rows_with_na)
raw_RNASeq_counts_annGenes <- raw_RNASeq_counts_annGenes[complete.cases(raw_RNASeq_counts_annGenes$hgnc_symbol), ]
any_na <- any(is.na(raw_RNASeq_counts_annGenes$hgnc_symbol))
print(any_na)
```

```{r}
countData <- gse201533_counts_ready
colData <- gse201533_metadata_ready 

if (!all(is.integer(countData))) {
  countData <- as.matrix(round(countData))
}

colData$Dosef <- factor(colData$Dose)
colData$Vaccinef <- factor(colData$Vaccine)
colData$Timepointf <- factor(colData$Timepoint)


# DESeqDataSet

dds <- DESeqDataSetFromMatrix(countData = countData, colData = colData, design = ~Timepointf + Dosef + Vaccinef)
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]

# Executar DESeq
dds <- DESeq(dds)
write_rds(dds,file="gse201533_dds_DESeq2_time_vaccine.rds")

# Obtendo os genes diferencialmente expressos:
res <- results(dds)

# Para obter informações sobre quais variáveis e testes foram usados e quantos genes foram significativamente up (LFC > 0) e down (LFC < 0)regulados:
summary (res)

#Salvar tabela deseq
write_rds(res,file="gse201533_res_DESeq2_vaccine.rds")

```

## GSE199750

```{r}
# Required files
countData <- gse199750_counts_ready
colData <- gse199750_Sample_meta_data_RNASeq
colData = colData[order(row.names(colData)), ]

#Factors
colData$Timepointf <- factor(colData$Timepoint)
colData$Sexf <- factor(colData$Sex)
colData$FirstSecondDosef <- factor(colData$first.second.vaccine)
colData$ThirdDosef <- factor(colData$third.vaccine)
colData$Group_I_II = factor(paste0(colData$FirstSecondDosef, colData$Timepointf))
colData$AllVaccines = factor(paste0(colData$Timepointf, sep="_", colData$FirstSecondDosef, sep="_3A_", colData$ThirdDosef))


###### DESeqDataSet objetct
dds <- DESeqDataSetFromMatrix(countData = countData, 
                              colData = colData, 
                              design = ~ AllVaccines)


###### Pre-filtering
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]
```

Execute DESeq2

```{r}
###### Execute DESeq2
dds <- DESeq(dds, test = "Wald")
write_rds(dds,file="gse199750_dds_DESeq2_AllVaccines.rds")

###### Retrieve comparisons
resultsNames(dds)
```

#### BNT

```{r}
######### PFIZER ######### 

# BNT V1
res_1_0_bnt <- results(dds, contrast= c("Group_I_II", "BNT162b2V1", "BNT162b2V0"))
res_1_0_bnt <- subset(res_1_0_bnt, padj < 0.05) #excluir genes com padj > 5%
res_1_0_bnt <- subset(res_1_0_bnt, log2FoldChange > 1.0 | log2FoldChange < -1.0)
res_1_0_bnt <- res_1_0_bnt[order(res_1_0_bnt$padj),]  #ordenar por padj
res_up <- subset(res, log2FoldChange > 1.0 )
res_down <- subset(res, log2FoldChange < -1.0 )
res_1_0_bnt_degs_table <- data.frame(Gene = rownames(res_1_0_bnt),
                          log2FoldChange = res_1_0_bnt$log2FoldChange,
                          pvalue = res_1_0_bnt$pvalue,
                          padj = res_1_0_bnt$padj)
write.csv(res_1_0_bnt_degs_table, file = "GSE199750_T0-T1_DEGs-PF.csv", row.names = TRUE)
saveRDS(res_1_0_bnt_degs_table, file = "GSE199750_T0-T1_DEGs-PF.rds")

# BNT V2
res_2_0_bnt <- results(dds, contrast= c("Group_I_II","BNT162b2V2A","BNT162b2V0"))
res_2_0_bnt <- subset(res_2_0_bnt, padj < 0.05) #excluir genes com padj > 5%
res_2_0_bnt <- subset(res_2_0_bnt, log2FoldChange > 1.0 | log2FoldChange < -1.0)
res_2_0_bnt <- res_2_0_bnt[order(res_2_0_bnt$padj),]  #ordenar por padj
res_up <- subset(res, log2FoldChange > 1.0 )
res_down <- subset(res, log2FoldChange < -1.0 )
res_2_0_bnt_degs_table <- data.frame(Gene = rownames(res_2_0_bnt),
                          log2FoldChange = res_2_0_bnt$log2FoldChange,
                          pvalue = res_2_0_bnt$pvalue,
                          padj = res_2_0_bnt$padj)
write.csv(res_2_0_bnt_degs_table, file = "GSE199750_T0-T2_DEGs-PF.csv", row.names = TRUE)
saveRDS(res_2_0_bnt_degs_table, file = "GSE199750_T0-T2_DEGs-PF.rds")

# BNT V3
resultsNames(dds)
res_3_0_bnt_bnt <- results(dds, contrast= c("AllVaccines", "V3A_BNT162b2_3A_BNT162b2", "V0_BNT162b2_3A_BNT162b2"))
res_3_0_bnt_bnt <- subset(res_3_0_bnt_bnt, padj < 0.05) #excluir genes com padj > 5%
res_3_0_bnt_bnt <- subset(res_3_0_bnt_bnt, log2FoldChange > 1.0 | log2FoldChange < -1.0)
res_3_0_bnt_bnt <- res_3_0_bnt_bnt[order(res_3_0_bnt_bnt$padj),]  #ordenar por padj
res_3_0_bnt_bnt_up <- subset(res_3_0_bnt_bnt, log2FoldChange > 1.0 )
res_3_0_bnt_bnt_down <- subset(res_3_0_bnt_bnt, log2FoldChange < -1.0 )
res_3_0_bnt_bnt_degs_table <- data.frame(res_3_0_bnt_bnt)
#Salvar tabela
write.csv(res_3_0_bnt_bnt_degs_table, file = "GSE199750_T3-T0_DEGs_2PF-1PF.csv", row.names = TRUE)

# MO V3
res_3_0_bnt_mo <- results(dds, contrast= c("AllVaccines", "V3A_BNT162b2_3A_mRNA-1273", "V0_BNT162b2_3A_BNT162b2"))
res_3_0_bnt_mo <- subset(res_3_0_bnt_mo, padj < 0.05) #excluir genes com padj > 5%
res_3_0_bnt_mo <- subset(res_3_0_bnt_mo, log2FoldChange > 1.0 | log2FoldChange < -1.0)
res_3_0_bnt_mo <- res_3_0_bnt_mo[order(res_3_0_bnt_mo$padj),]  #ordenar por padj
res_3_0_bnt_mo_up <- subset(res_3_0_bnt_mo, log2FoldChange > 1.0 )
res_3_0_bnt_mo_down <- subset(res_3_0_bnt_mo, log2FoldChange < -1.0 )
res_3_0_bnt_mo_degs_table <- data.frame(res_3_0_bnt_mo)
write.csv(res_3_0_bnt_mo_degs_table, file = "GSE199750_T3-T0_DEGs_2PF-1MO.csv", row.names = TRUE)

```

#### ChAd

```{r}
resultsNames(dds)
# ChAd V1
res_1_0_chad <- results(dds, contrast= c("Group_I_II","ChAdOx1V1","ChAdOx1V0"))
res_1_0_chad <- subset(res_1_0_chad, padj < 0.05)
res_1_0_chad <- subset(res_1_0_chad, log2FoldChange > 1.0 | log2FoldChange < -1.0)
res_1_0_chad <- res_1_0_chad[order(res_1_0_chad$padj),]  #ordenar por padj
res_up <- subset(res, log2FoldChange > 1.0 )
res_down <- subset(res, log2FoldChange < -1.0 )
res_1_0_chad_degs_table <- data.frame(Gene = rownames(res_1_0_chad),
                          log2FoldChange = res_1_0_chad$log2FoldChange,
                          pvalue = res_1_0_chad$pvalue,
                          padj = res_1_0_chad$padj)
write.csv(res_1_0_chad_degs_table, file = "GSE199750_T0-T1_DEGs-AZ.csv", row.names = TRUE)

# ChAd V2
res_2_0_chad <- results(dds, contrast= c("Group_I_II","ChAdOx1V2A","ChAdOx1V0"))
res_2_0_chad <- subset(res_2_0_chad, padj < 0.05) #excluir genes com padj > 5%
res_2_0_chad <- subset(res_2_0_chad, log2FoldChange > 1.0 | log2FoldChange < -1.0)
res_2_0_chad <- res_2_0_chad[order(res_2_0_chad$padj),]  #ordenar por padj
summary(res_2_0_chad) #Up e Down
res_2_0_chad_degs_table <- data.frame(Gene = rownames(res_2_0_chad),
                          log2FoldChange = res_2_0_chad$log2FoldChange,
                          pvalue = res_2_0_chad$pvalue,
                          padj = res_2_0_chad$padj)
write.csv(res_2_0_chad_degs_table, file = "GSE199750_T0-T2_DEGs-AZ.csv", row.names = FALSE)
saveRDS(res_2_0_chad_degs_table, file = "GSE199750_T0-T2_DEGs-AZ.rds")

#ChAd-BNT V3

res_3_0_chad_bnt <- results(dds, contrast= c("AllVaccines", "V3A_ChAdOx1_3A_BNT162b2", "V0_ChAdOx1_3A_Missing"))
res_3_0_chad_bnt <- subset(res_3_0_chad_bnt, padj < 0.05) #excluir genes com padj > 5%
res_3_0_chad_bnt <- subset(res_3_0_chad_bnt, log2FoldChange > 1.0 | log2FoldChange < -1.0)
res_3_0_chad_bnt <- res_3_0_chad_bnt[order(res_3_0_chad_bnt$padj),]  #ordenar por padj
res_3_0_chad_bnt_up <- subset(res_3_0_chad_bnt, log2FoldChange > 1.0 )
res_3_0_chad_bnt_down <- subset(res_3_0_chad_bnt, log2FoldChange < -1.0 )
res_3_0_chad_bnt_degs_table <- data.frame(res_3_0_chad_bnt)
write.csv(res_3_0_chad_bnt_degs_table, file = "GSE199750_T3-T0_DEGs_2AZ-1PF.csv", row.names = TRUE)
```

## Combine raw counts

```{r}
# Counts
gse189039 = gse189039_counts_ready %>% 
  rownames_to_column("genes") 
gse201530 = gse201530_counts_ready %>% 
  rownames_to_column("genes") 
gse201533 = gse201533_counts_ready %>% 
  rownames_to_column("genes") 
gse199750 = gse199750_counts_ready %>% 
  as.data.frame() %>% 
  rownames_to_column("genes") 
gse206023 = gse206023_counts_ready %>% 
  select(genes, everything(), -"...1")

all_matrices_counts = gse189039 %>% 
  full_join(gse201530, by = "genes") %>% 
  full_join(gse201533, by = "genes") %>% 
  full_join(gse199750, by = "genes") %>% 
  full_join(gse206023, by = "genes")

write_tsv(all_matrices_counts, file = "all_matrices_counts.tsv")

```

# 4. DEGs analysis

## 4.1. Up and down-regulated DEGs

```{r}
#Required files
all_degs_p_05_vac_infected = read_csv(here("DGE analysis", "all_degs_p_05_vac_infected_19_12_23.csv"))

ann_vaccines = read_csv(here("Annotations and metadata", "ann_vaccines_conditions.csv"))

all_degs = all_degs_p_05_vac_infected %>% 
  group_by(condition, direction) %>% 
  summarise(n_degs = n()) %>% 
  inner_join(ann_vaccines, by="condition")

all_degs_wide = all_degs %>% 
  pivot_wider(names_from = "direction", values_from = "n_degs") %>% 
  mutate(total = sum(DOWN + NEUTRAL + UP))

write.csv(all_degs_wide, file = here("DGE analysis", "all_degs_wide.csv"))
```

By vaccination or disease

```{r fig.height=6, fig.width=10}
# Plot 1
ggplot_degs_bars = all_degs %>%
 filter(direction != "NEUTRAL") %>%
 ggplot() +
  aes(x = reorder(condition, day), fill = direction, weight = n_degs) +
  geom_bar(position = "dodge") +
  scale_fill_manual(
    values = c(DOWN = "#0E252D",
    NEUTRAL = "#00C19F",
    UP = "#3399FF")) +
  theme_few() +
  facet_wrap(vars(disease_vac), scales = "free", nrow = 2) +
  geom_text(aes(label = n_degs, y = n_degs), 
            position = position_dodge(width = 0.9), 
            vjust = -1, 
            hjust = 0.5, 
            size = 2) +
  ylim(0,3000) +
  theme(legend.position = "top",
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1, size = 7),
        axis.text.y = element_text(vjust = 0.5)) +
  ggtitle("DEGs - all studies, p<0.05") +
  labs(x = "")

ggsave(ggplot_degs_bars, file = here("Figures", "DEGs_up_down_barplot_pvalue05_1.png"),  width = 10, height = 6)

ggplot_degs_bars
```

By type

```{r fig.height=6, fig.width=10}
#Plot 2
ggplot_degs_bars_2 = all_degs %>%
 filter(!(direction %in% "NEUTRAL")) %>%
 ggplot(aes(x = condition, fill = direction, weight = n_degs,
            label = condition)) +
  geom_bar(position = "dodge") +
  scale_fill_manual(
    values = c(DOWN = "#0E252D",
    NEUTRAL = "#00C19F",
    UP = "#3399FF")) +
  theme_minimal() +
  facet_wrap(vars(vaccine), scales = "free", nrow = 2) +
  theme(legend.position = "top",
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1, size = 7),
        axis.text.y = element_text(vjust = 0.5)) +
  ggtitle("DEGs - all studies, p<0.05")

ggsave(ggplot_degs_bars_2, file = here("Figures", "DEGs_up_down_barplot_pvalue05_2.png"),  width = 10, height = 6)

ggplot_degs_bars_2 %>% 
  ggplotly()
```

## 4.2. Overlapping genes

### 4.2.1. Input

```{r}
# DEGs table
degs_all_updown = read_csv(here("DGE analysis","all_degs_p_05_vac_infected_19_12_23.csv")) %>% 
  filter(direction != "NEUTRAL") %>% 
  mutate_all(~ replace(., is.na(.), "NA"))

ann_vaccines = read_csv(here("Annotations and metadata", "ann_vaccines_conditions.csv"))

# Only up and down
data = degs_all_updown %>% 
  filter(padj < 0.05,
         log2fold_change > 1 | log2fold_change < -1) %>% 
 dplyr::select(process = condition, genes)

filename = "All_covid"
```

### 4.2.2. Function

```{r}
# Function for overlapping -------
overlap_genes <- function(cond1, cond2, data) {
  genes_cond1 <- data$genes[data$process == cond1]
  genes_cond2 <- data$genes[data$process == cond2]
  
  genes_shared <- intersect(genes_cond1, genes_cond2)
  
  genes_notshared_cond1 <- setdiff(genes_cond1, genes_cond2)
  genes_notshared_cond2 <- setdiff(genes_cond2, genes_cond1)
  
  total_genes_cond1 <- length(genes_cond1)
  total_genes_cond2 <- length(genes_cond2)
  
  percentage_shared_cond1 <- length(genes_shared) / total_genes_cond1 * 100
  percentage_shared_cond2 <- length(genes_shared) / total_genes_cond2 * 100
  
  shared_genes <- data.frame(
    Cond1 = cond1,
    Cond2 = cond2,
    Shared = length(genes_shared),
    NotShared_cond1 = length(genes_notshared_cond1),
    NotShared_cond2 = length(genes_notshared_cond2),
    Total_Genes_Cond1 = total_genes_cond1,
    Total_Genes_Cond2 = total_genes_cond2,
    Genes_Names = paste(genes_shared, collapse = ", "),
    Percentage_Shared_Cond1 = percentage_shared_cond1,
    Percentage_Shared_Cond2 = percentage_shared_cond2
  )
  
  return(shared_genes)
}
```

### 4.2.3. Perform calculations

```{r}
# Perform calculations
unique_cond <- unique(data$process) # List sets
shared_genes_df <- data.frame() # Store data

for (i in 1:(length(unique_cond) - 1)) {
  for (j in (i + 1):length(unique_cond)) {
    resultado_temp <- overlap_genes(unique_cond[i], unique_cond[j], data)
    shared_genes_df <- bind_rows(shared_genes_df, resultado_temp)
  }
}

# Fisher's exact test
fisher_exact_test <- function(shared, notshared1, notshared2) {
  cont_table <- matrix(c(shared, notshared1, notshared2, 0), nrow = 2)
  results_fisher <- fisher.test(cont_table)
  return(results_fisher$p.value)
}
shared_genes_df = shared_genes_df %>%
  mutate(pvalue = pmap_dbl(list(Shared, NotShared_cond1, NotShared_cond2), fisher_exact_test)) %>% 
  mutate("-log(p)" = -log10(pvalue))


#Mirror table
shared_genes_df2 = shared_genes_df %>% 
  rename(Cond1_temp = Cond1, Cond2_temp = Cond2) %>%
  rename(Cond1 = Cond2_temp, Cond2 = Cond1_temp)

shared_genes_df_mirror = shared_genes_df %>% 
  bind_rows(shared_genes_df2) %>% 
  mutate(Percentage_Shared_Cond1 = round(Percentage_Shared_Cond1, 2),
         Percentage_Shared_Cond2 = round(Percentage_Shared_Cond2, 2))
  
#Jaccard distance
jaccard.matrix <- data %>%
  mutate(present = 1) %>% 
  distinct() %>% 
  pivot_wider(names_from = genes, values_from = present, values_fill = 0) %>%
  column_to_rownames(var = "process") %>%
  vegdist(binary = TRUE, method = "jaccard", diag = TRUE, upper = TRUE) %>%
  as.matrix() %>%
  {1 - .}

shared_genes_df_mirror = jaccard.matrix %>% 
  as.data.frame() %>% 
  rownames_to_column("Cond1") %>% 
  pivot_longer(cols = -"Cond1",
               names_to = "Cond2",
               values_to = "jaccard_distance") %>% 
  inner_join(shared_genes_df_mirror, by = join_by(Cond1, Cond2)) %>% 
  mutate(jaccard_distance = round(jaccard_distance, 3))

#Save
write_csv(shared_genes_df_mirror, 
          file = here("DGE analysis", paste0(filename, "_sharedgenes_Fisher_Jaccard.csv")))

```

### 4.2.4. Heatmap

#### 4.2.4.1. Shared genes

```{r}
filename = "All_covid_Shared"
#Anotar colunas
# Converter para wide table
matrix_data = shared_genes_df_mirror %>% 
  dcast(Cond1 ~ Cond2, 
        value.var = "Shared", 
        fun.aggregate = mean) %>% 
  column_to_rownames("Cond1") %>%
  mutate_if(is.character, as.numeric) %>% 
  as.matrix() 

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = matrix_data %>% 
  as.data.frame() %>%
  rownames_to_column("condition") %>% 
  select(condition) %>% 
  inner_join(ann_vaccines, by = "condition") %>%
  select(condition, vaccine:regimen, week) %>% 
  distinct() 

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "process") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  relocate(dose, .before=vaccine) %>% 
  column_to_rownames(var= "condition")  %>% 
  mutate(dose = ifelse(is.na(dose), "NA", dose))

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(vaccine:regimen)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  as.matrix()

matrix_data_ordered_rows = matrix_data_ordered %>% 
  as.data.frame() %>% 
  rownames_to_column("condition") %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") %>% 
  as.matrix()

#Verificar se as linhas sao iguais------
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas

#Espectro de cores
col_fun <- colorRamp2(c(0, 500), c("white", "#d90429"))

```

```{r}
######## HEATMAP ANNOTATION
ann_vaccines_heatmap = ann_vaccines_heatmap %>% 
  select(type, vaccine, day) %>% 
  mutate(day = as_factor(day),
         day = fct_reorder(day, as.numeric(day)))

ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_vaxsig_colors, 
                       annotation_name_side = "left",
                       show_legend = F)

row_ha = rowAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_vaxsig_colors)



mat = matrix_data_ordered_rows
width_image = 40
height_image = 40
width = unit(20, "cm")
height = unit(20, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)
gp = gpar(fontsize = 5)
rect_gp = gpar(col = "gray1", lwd = 0)



fileimageheatmap_grouped = paste0(filename, sep ="_", "Heatmap.png")

png(
  file = here("Figures", fileimageheatmap_grouped),
  width = width_image,
  height = height_image,
  units = "cm",
  res = 900
)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "Shared DEGs", #Legend
                        col = col_fun, #color
                         cell_fun = function(j, i, x, y, width, height, fill) {
  if (!is.na(mat[i, j]) && mat[i, j] > 0) {
    grid.text(sprintf("%.f", mat[i, j]), x, y, gp = gpar(fontsize = 5))
  }
},
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 90,
                        column_names_gp = column_names_gp,
                        cluster_rows = TRUE,
                        rect_gp = rect_gp,
                        row_names_gp = row_names_gp,
                        # row_split = 2, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        na_col = "black",
                        column_split = NULL, #Split group clusters
                        #column_km = 3,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "left")

dev.off()
```

#### 4.2.4.2. Jaccard distance

```{r}
filename = "All_covid_Jaccarddist"
#Anotar colunas
# Converter para wide table
matrix_data = shared_genes_df_mirror %>% 
  dcast(Cond1 ~ Cond2, 
        value.var = "jaccard_distance", 
        fun.aggregate = mean) %>% 
  column_to_rownames("Cond1") %>%
  mutate_if(is.character, as.numeric) %>% 
  as.matrix() 

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = matrix_data %>% 
  as.data.frame() %>%
  rownames_to_column("condition") %>% 
  select(condition) %>% 
  inner_join(ann_vaccines, by = "condition") %>%
  select(condition, vaccine:regimen, week) %>% 
  distinct() 

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "process") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  relocate(dose, .before=vaccine) %>% 
  column_to_rownames(var= "condition")  %>% 
  mutate(dose = ifelse(is.na(dose), "NA", dose))

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(vaccine:regimen)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  as.matrix()

matrix_data_ordered_rows = matrix_data_ordered %>% 
  as.data.frame() %>% 
  rownames_to_column("condition") %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") %>% 
  as.matrix()

#Verificar se as linhas sao iguais------
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas
# dim(matrix_data_rows_cols_ordered) #Usar se ordenar por colunas e linhas
# dim(ann_genesets_heatmap) #Anotações sobre as linhas

#Definir cores do gráfico
# Defina os valores mínimos e máximos para o espectro de cores

#Espectro de cores
col_fun <- colorRamp2(c(0, 0.50), c("white", "#d90429")) #Immune

```

```{r}

######## HEATMAP ANNOTATION
ann_vaccines_heatmap = ann_vaccines_heatmap %>% 
  select(type, vaccine, day) %>% 
  mutate(day = as_factor(day),
         day = fct_reorder(day, as.numeric(day)))

ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_vaxsig_colors, 
                       annotation_name_side = "left",
                       show_legend = F)

row_ha = rowAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_vaxsig_colors)



mat = matrix_data_ordered_rows
width_image = 40
height_image = 40
width = unit(20, "cm")
height = unit(20, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)
gp = gpar(fontsize = 5)
rect_gp = gpar(col = "gray1", lwd = 0)



fileimageheatmap_grouped = paste0(filename, sep ="_", "Heatmap.png")

png(
  file = here("Figures", fileimageheatmap_grouped),
  width = width_image,
  height = height_image,
  units = "cm",
  res = 900
)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "Shared DEGs", #Legend
                        col = col_fun, #color
                         cell_fun = function(j, i, x, y, width, height, fill) {
  if (!is.na(mat[i, j]) && mat[i, j] > 0.01) {
    grid.text(sprintf("%.2f", mat[i, j]), x, y, gp = gpar(fontsize = 5))
  }
},
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 90,
                        column_names_gp = column_names_gp,
                        cluster_rows = TRUE,
                        rect_gp = rect_gp,
                        row_names_gp = row_names_gp,
                        # row_split = 2, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        na_col = "black",
                        column_split = NULL, #Split group clusters
                        #column_km = 3,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "left")

dev.off()
```

### 4.2.5. Chord diagram

#### 4.2.5.1. Shared

##### 4.2.5.1.1. All conditions (Interactive)

```{r}
library(chorddiag)

matrix_vaccines <- shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  dplyr::select(Cond1, Cond2, Shared) %>% 
  pivot_wider(names_from = "Cond2", values_from = "Shared", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

types = shared_genes_df_mirror %>% 
  dplyr::select(condition = Cond1) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% 
 dplyr::select(condition, type), by = "condition") %>% 
  dplyr::select(type)

colors = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(ann_vaccines %>% 
 dplyr::select(condition, type), by = "type") %>% 
  inner_join(matrix_vaccines %>% 
               as.data.frame() %>% 
               colnames() %>% as.data.frame() %>% rename(condition = "."),
             by = "condition") %>% 
  pull(color)

# Build the chord diagram:
all_shared = chorddiag(matrix_vaccines,
          groupColors = colors,
          type = "directional",
          width = NULL,
          height = NULL,
          margin = 100,
          showGroupnames = TRUE,
          groupThickness = 0.1,
          groupPadding = 2,
          groupnamePadding = 1,
          groupnameFontsize = 10,
          groupedgeColor = NULL,
          chordedgeColor = NULL,
          categoryNames = NULL,
          categorynamePadding = 100,
          categorynameFontsize = 28,
          showTicks = TRUE,
          tickInterval = NULL,
          ticklabelFontsize = 10,
          fadeLevel = 0.1,
          showTooltips = TRUE,
          showZeroTooltips = TRUE,
          tooltipNames = NULL,
          tooltipUnit = NULL,
          tooltipFontsize = 12,
          tooltipGroupConnector = " &#x25B6; ",
          precision = NULL,
          clickAction = NULL,
          clickGroupAction = NULL
)
all_shared
```

##### 4.2.5.1.2. Weeks

Prepare tables

```{r fig.height=5, fig.width=5}
#Week 1 ------
links_df_W1 = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(ann_vaccines %>% 
               filter(week == 1) %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(ann_vaccines %>% 
               filter(week == 1) %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, Shared)

mat_week1 = links_df_W1 %>% 
  pivot_wider(names_from = "Cond2", values_from = "Shared", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()
grid.col_w1 = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(ann_vaccines %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_W1,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)
group_w1 = ann_vaccines %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_W1,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)


#Week 2 ------
links_df_W2 = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(ann_vaccines %>% 
               filter(week == 2) %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(ann_vaccines %>% 
               filter(week == 2) %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, Shared)

mat_week2 = links_df_W2 %>% 
  pivot_wider(names_from = "Cond2", values_from = "Shared", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_w2 = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(ann_vaccines %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_W2,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_w2 = ann_vaccines %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_W2,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)


#Week 4 ------
links_df_W4 = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(ann_vaccines %>% 
               filter(week == 4) %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(ann_vaccines %>% 
               filter(week == 4) %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, Shared)

mat_week4 = links_df_W4 %>% 
  pivot_wider(names_from = "Cond2", values_from = "Shared", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_w4 = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(ann_vaccines %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_W4,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_w4 = ann_vaccines %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_W4,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)

#Week 7 ------
links_df_W7 = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(ann_vaccines %>% 
               filter(week == 7) %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(ann_vaccines %>% 
               filter(week == 7) %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, Shared)

mat_week7 = links_df_W7 %>% 
  pivot_wider(names_from = "Cond2", values_from = "Shared", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_w7 = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(ann_vaccines %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_W7,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_w7 = ann_vaccines %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_W7,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)

```

Plot

```{r fig.height= 5, fig.width=10}
#Plot
png(filename = here("Figures","ChordDiagram_Covid19_Weeks_1-2-4-7.png"), 
    width = 12, height = 14, units = "in", res = 300)

par(mfrow = c(2, 2))
#WEEK 1
chordDiagram(mat_week1, 
             grid.col = grid.col_w1, 
             annotationTrack = "grid", 
             directional = -1,
             big.gap = 6,
             transparency = 0,
             small.gap = 3,
             link.zindex = rank(mat_week1),
             h.ratio = 0.7, 
             group = group_w1,
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_week1))))))
#Labels
circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.8, #Label size
      font = 0.1)},#Label face
  bg.border = NA) 
title("Week 1")

# WEEK 2
chordDiagram(mat_week2, grid.col = grid.col_w2, 
             annotationTrack = "grid", 
            big.gap = 50,
             group = group_w2,
             small.gap = 3,
             link.zindex = rank(mat_week2),
             h.ratio = 0.7, 
             directional = -1,
             transparency = 0,
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_week2))))))
#Labels
circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.8, #Label size
      font = 0.1)},
  bg.border = NA)

title("Week 2")


# WEEK 4
chordDiagram(mat_week4, grid.col = grid.col_w4, 
             annotationTrack = "grid", 
            big.gap = 50,
             group = group_w4,
             small.gap = 3,
             link.zindex = rank(mat_week4),
             h.ratio = 0.7, 
             directional = -1,
             transparency = 0,
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_week4))))))
#Labels
circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.8, #Label size
      font = 0.1)},
  bg.border = NA)

title("Week 4")

# WEEK 7
chordDiagram(mat_week7, grid.col = grid.col_w7, 
             annotationTrack = "grid", 
            big.gap = 50,
             group = group_w7,
             small.gap = 3,
             link.zindex = rank(mat_week7),
             h.ratio = 0.7, 
             directional = -1,
             transparency = 0,
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_week7))))))
#Labels
circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.8, #Label size
      font = 0.1)},
  bg.border = NA)

title("Week 7")

dev.off()
circos.clear()

```

##### 4.2.5.1.3. Types

Prepare tables

```{r fig.height= 10, fig.width=10}
#Types
###### All -----
links_df = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  dplyr::select(Cond1, Cond2, Shared)

mat_types = links_df %>% 
  pivot_wider(names_from = "Cond2", values_from = "Shared", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_types = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(ann_vaccines %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_types = ann_vaccines %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)

# Vaccination only -------
links_df_vac = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(ann_vaccines %>% 
               filter(disease_vac == "V") %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(ann_vaccines %>% 
               filter(disease_vac == "V") %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, Shared)

mat_types_vac = links_df_vac %>% 
  pivot_wider(names_from = "Cond2", values_from = "Shared", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_types_vac = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(ann_vaccines %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_vac,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_types_vac = ann_vaccines %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_vac,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)

#Infection only
links_df_inf = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(ann_vaccines %>% 
               filter(disease_vac == "I") %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(ann_vaccines %>% 
               filter(disease_vac == "I") %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, Shared)

mat_types_inf = links_df_inf %>% 
  pivot_wider(names_from = "Cond2", values_from = "Shared", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_types_inf = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(ann_vaccines %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_inf,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_types_inf = ann_vaccines %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_inf,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)
```

Plot All conditions and stats

```{r fig.height= 5, fig.width=5}
#Plot

#Types

png(filename = here("Figures","ChordDiagram_Covid19_Types_All.png"), 
    width = 5, height = 5, units = "in", res = 300)

#Infection and vaccination
chordDiagram(mat_types, grid.col = grid.col_types,
             annotationTrack = "grid",  transparency = 0,
             big.gap = 10,
             small.gap = 3,
             directional = -1, 
             group = group_types,
             link.zindex = rank(mat_types),
              h.ratio = 0.7, 
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_types))))))

circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.5, #Label size
      font = 0.1
      )},
  bg.border = NA)
title("Infection and Vaccination")

dev.off()
circos.clear()
```

By state

```{r fig.height= 5, fig.width=15}

png(filename = here("Figures","ChordDiagram_Covid19_Types_Inf_Vac.png"), 
    width = 6, height = 12, units = "in", res = 300)

par(mfrow = c(2, 1))
#Vaccination
chordDiagram(mat_types_vac, grid.col = grid.col_types_vac, 
             annotationTrack = "grid", transparency = 0,
             big.gap = 6,
             small.gap = 3,
                          directional = -1,
             group = group_types,
             link.zindex = rank(mat_types_vac),
             h.ratio = 0.7, 
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_types_vac))))))

circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.5, #Label size
      font = 0.1
      )},
  bg.border = NA) 
title("Vaccination only")

#Infection
chordDiagram(mat_types_inf, grid.col = grid.col_types_inf, 
             annotationTrack = "grid", transparency = 0,
             big.gap = 50,
             small.gap = 3,
            directional = -1,
             group = group_types,
             h.ratio = 0.7, 
             link.zindex = rank(mat_types_inf),
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_types_inf))))))

circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.5, #Label size
      font = 0.1
      )},
  bg.border = NA) 
title("Infection only")
circos.clear()
dev.off()
```

#### 4.2.5.2. Jaccard distance

##### 4.2.5.2.1. All conditions (Interactive)

```{r}

matrix_vaccines <- shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  dplyr::select(Cond1, Cond2, jaccard_distance) %>% 
  pivot_wider(names_from = "Cond2", values_from = "jaccard_distance", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

types = shared_genes_df_mirror %>% 
  dplyr::select(condition = Cond1) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% 
 dplyr::select(condition, type), by = "condition") %>% 
  dplyr::select(type)

colors = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(ann_vaccines %>% 
 dplyr::select(condition, type), by = "type") %>% 
  inner_join(matrix_vaccines %>% 
               as.data.frame() %>% 
               colnames() %>% as.data.frame() %>% rename(condition = "."),
             by = "condition") %>% 
  pull(color)

# Build the chord diagram:
chorddiag(matrix_vaccines,
          groupColors = colors,
          type = "directional",
          width = NULL,
          height = NULL,
          margin = 100,
          showGroupnames = TRUE,
          groupThickness = 0.1,
          groupPadding = 2,
          groupnamePadding = 1,
          groupnameFontsize = 10,
          groupedgeColor = NULL,
          chordedgeColor = NULL,
          categoryNames = NULL,
          categorynamePadding = 100,
          categorynameFontsize = 28,
          showTicks = TRUE,
          tickInterval = NULL,
          ticklabelFontsize = 10,
          fadeLevel = 0.1,
          showTooltips = TRUE,
          showZeroTooltips = TRUE,
          tooltipNames = NULL,
          tooltipUnit = NULL,
          tooltipFontsize = 12,
          tooltipGroupConnector = " &#x25B6; ",
          precision = NULL,
          clickAction = NULL,
          clickGroupAction = NULL
)
```

##### 4.2.5.2.2. Weeks

Prepare tables

```{r fig.height=5, fig.width=5}
#Week 1 ------
links_df_W1 = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(ann_vaccines %>% 
               filter(week == 1) %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(ann_vaccines %>% 
               filter(week == 1) %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, jaccard_distance)

mat_week1 = links_df_W1 %>% 
  pivot_wider(names_from = "Cond2", values_from = "jaccard_distance", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()
grid.col_w1 = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(ann_vaccines %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_W1,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)
group_w1 = ann_vaccines %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_W1,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)


#Week 2 ------
links_df_W2 = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(ann_vaccines %>% 
               filter(week == 2) %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(ann_vaccines %>% 
               filter(week == 2) %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, jaccard_distance)

mat_week2 = links_df_W2 %>% 
  pivot_wider(names_from = "Cond2", values_from = "jaccard_distance", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_w2 = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(ann_vaccines %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_W2,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_w2 = ann_vaccines %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_W2,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)


#Week 4 ------
links_df_W4 = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(ann_vaccines %>% 
               filter(week == 4) %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(ann_vaccines %>% 
               filter(week == 4) %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, jaccard_distance)

mat_week4 = links_df_W4 %>% 
  pivot_wider(names_from = "Cond2", values_from = "jaccard_distance", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_w4 = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(ann_vaccines %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_W4,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_w4 = ann_vaccines %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_W4,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)

#Week 7 ------
links_df_W7 = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(ann_vaccines %>% 
               filter(week == 7) %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(ann_vaccines %>% 
               filter(week == 7) %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, jaccard_distance)

mat_week7 = links_df_W7 %>% 
  pivot_wider(names_from = "Cond2", values_from = "jaccard_distance", values_fill = list(Shared = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_w7 = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(ann_vaccines %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_W7,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_w7 = ann_vaccines %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_W7,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)

```

Plot

```{r fig.height= 5, fig.width=10}
#Plot
png(filename = here("Figures","ChordDiagram_Covid19_Weeks_1-2-4-7_jaccard.png"), 
    width = 12, height = 14, units = "in", res = 300)

par(mfrow = c(2, 2))
#WEEK 1
chordDiagram(mat_week1, 
             grid.col = grid.col_w1, 
             annotationTrack = "grid", 
             directional = -1,
             big.gap = 6,
             transparency = 0,
             small.gap = 3,
             link.zindex = rank(mat_week1),
             h.ratio = 0.7, 
             group = group_w1,
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_week1))))))
#Labels
circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.8, #Label size
      font = 0.1)},#Label face
  bg.border = NA) 
title("Week 1")

# WEEK 2
chordDiagram(mat_week2, grid.col = grid.col_w2, 
             annotationTrack = "grid", 
            big.gap = 50,
             group = group_w2,
             small.gap = 3,
             link.zindex = rank(mat_week2),
             h.ratio = 0.7, 
             directional = -1,
             transparency = 0,
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_week2))))))
#Labels
circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.8, #Label size
      font = 0.1)},
  bg.border = NA)

title("Week 2")


# WEEK 4
chordDiagram(mat_week4, grid.col = grid.col_w4, 
             annotationTrack = "grid", 
            big.gap = 50,
             group = group_w4,
             small.gap = 3,
             link.zindex = rank(mat_week4),
             h.ratio = 0.7, 
             directional = -1,
             transparency = 0,
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_week4))))))
#Labels
circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.8, #Label size
      font = 0.1)},
  bg.border = NA)

title("Week 4")

# WEEK 7
chordDiagram(mat_week7, grid.col = grid.col_w7, 
             annotationTrack = "grid", 
            big.gap = 50,
             group = group_w7,
             small.gap = 3,
             link.zindex = rank(mat_week7),
             h.ratio = 0.7, 
             directional = -1,
             transparency = 0,
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_week7))))))
#Labels
circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.8, #Label size
      font = 0.1)},
  bg.border = NA)

title("Week 7")

dev.off()
circos.clear()

```

##### 4.2.5.2.3. Types

Prepare tables

```{r fig.height= 10, fig.width=10}
#Types
###### All -----
links_df = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  dplyr::select(Cond1, Cond2, jaccard_distance)

mat_types = links_df %>% 
  pivot_wider(names_from = "Cond2", values_from = "jaccard_distance", values_fill = list(jaccard_distance = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_types = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(ann_vaccines %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_types = ann_vaccines %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)

# Vaccination only -------
links_df_vac = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(ann_vaccines %>% 
               filter(disease_vac == "V") %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(ann_vaccines %>% 
               filter(disease_vac == "V") %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, jaccard_distance)

mat_types_vac = links_df_vac %>% 
  pivot_wider(names_from = "Cond2", values_from = "jaccard_distance", values_fill = list(jaccard_distance = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_types_vac = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(ann_vaccines %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_vac,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_types_vac = ann_vaccines %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_vac,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)

#Infection only
links_df_inf = shared_genes_df_mirror %>% 
  filter(pvalue < 0.05,
         Shared > 1) %>% 
  inner_join(ann_vaccines %>% 
               filter(disease_vac == "I") %>% 
               dplyr::select(Cond1 = condition),
             by = "Cond1") %>% 
  inner_join(ann_vaccines %>% 
               filter(disease_vac == "I") %>% 
               dplyr::select(Cond2 = condition),
             by = "Cond2") %>% 
  dplyr::select(Cond1, Cond2, jaccard_distance)

mat_types_inf = links_df_inf %>% 
  pivot_wider(names_from = "Cond2", values_from = "jaccard_distance", values_fill = list(jaccard_distance = NA)) %>% 
  column_to_rownames("Cond1") %>%
  dplyr::select(sort(names(.))) %>% 
  as.matrix()

grid.col_types_inf = ann_vaxsig_colors$type %>%
  as.data.frame() %>%
  rownames_to_column("type") %>%
  rename(color = ".") %>% 
  inner_join(ann_vaccines %>% 
 dplyr::select(Cond1 = condition, type), by = "type") %>% 
  inner_join(links_df_inf,
             by = "Cond1") %>% 
  dplyr::select(Cond1, color) %>% 
  distinct() %>% 
  pull(color, Cond1)

group_types_inf = ann_vaccines %>% 
  dplyr::select(Cond1 = condition, type) %>% 
  inner_join(links_df_inf,
             by = "Cond1") %>% 
  dplyr::select(Cond1, type) %>% 
  distinct() %>% 
  pull(type, Cond1)
```

Plot All conditions and stats

```{r fig.height= 5, fig.width=5}
#Plot

#Types

png(filename = here("Figures","ChordDiagram_Covid19_Types_All_jaccard_distance.png"), 
    width = 5, height = 5, units = "in", res = 300)

#Infection and vaccination
chordDiagram(mat_types, grid.col = grid.col_types,
             annotationTrack = "grid",  transparency = 0,
             big.gap = 10,
             small.gap = 3,
             directional = -1, 
             group = group_types,
             link.zindex = rank(mat_types),
              h.ratio = 0.7, 
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_types))))))

circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.5, #Label size
      font = 0.1
      )},
  bg.border = NA)
title("Infection and Vaccination")

dev.off()
circos.clear()
```

By state

```{r fig.height= 5, fig.width=15}

png(filename = here("Figures","ChordDiagram_Covid19_Types_Inf_Vac_jaccard_distance.png"), 
    width = 6, height = 12, units = "in", res = 300)

par(mfrow = c(2, 1))
#Vaccination
chordDiagram(mat_types_vac, grid.col = grid.col_types_vac, 
             annotationTrack = "grid", transparency = 0,
             big.gap = 6,
             small.gap = 3,
                          directional = -1,
             group = group_types,
             link.zindex = rank(mat_types_vac),
             h.ratio = 0.7, 
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_types_vac))))))

circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.5, #Label size
      font = 0.1
      )},
  bg.border = NA) 
title("Vaccination only")

#Infection
chordDiagram(mat_types_inf, grid.col = grid.col_types_inf, 
             annotationTrack = "grid", transparency = 0,
             big.gap = 50,
             small.gap = 3,
            directional = -1,
             group = group_types,
             h.ratio = 0.7, 
             link.zindex = rank(mat_types_inf),
             preAllocateTracks = list(track.height = 0.5 * max(strwidth(unlist(dimnames(mat_types_inf))))))

circos.track(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.5, #Label size
      font = 0.1
      )},
  bg.border = NA) 
title("Infection only")
circos.clear()
dev.off()
```

## 4.3. Dotplot

```{r}
#### Identidade e -log(q)
# install.packages("ggdendro")
library(ggdendro)
library(cowplot)
library(ggtree)
library(patchwork) 

#Filtrar valores com -log(p) >= 1
sharedDEGs_UP_DOWN_fisher_filtered <- sharedDEGs_UP_DOWN_fisher %>%
  mutate(`Significance -log(p)` = cut(`-log(p)`, 
                        breaks = c(-Inf, 1, 1.3, 2, Inf), 
                        labels = c("<1", "1-1.3", "1.3-2", ">2"))) %>%
  filter(Percentage_Shared_Cond1 >= 1, Percentage_Shared_Cond2 >= 1, `Significance -log(p)` != "<1")

#Salvar
write.csv(sharedDEGs_UP_DOWN_fisher_filtered, file = "sharedDEGs_UP_DOWN_fisher_filtered.csv")

#Visualizar
plot = ggplot(sharedDEGs_UP_DOWN_fisher, aes(x = Cond1, y = Cond2, color = Percentage_Shared_Cond1, size = Percentage_Shared_Cond1)) + 
  geom_point() +
  scale_size_continuous(range = c(2, 8)) +
  geom_text(aes(label = sprintf("%.f", Percentage_Shared_Cond1)), vjust = 0.5, hjust = 0.5, size = 2, color = "black")  +
  cowplot::theme_cowplot() + 
  theme(axis.line = element_blank(),
        axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1, size = 10),
        axis.text.y = element_text(size = 10)) +
  scale_color_gradient(low = "white", high = "#ef233c", limits = c(0, 60)) +
  ggtitle("%Identity between vaccines, Up and Down-Regulated, pvalue < 0.05" ) +
  ylab('Cond2') +
  theme(axis.ticks = element_blank()) +
  guides(color = FALSE, size = FALSE)

#Salvar
ggsave(plot, file = here("Figures","sharedDEGs_UP_DOWN_fisher.png"))

#Display
plot
```

```{r}

##### Unir countdata de todos os datasets

gse189039_counts_ready_long = gse189039_counts_ready %>% 
  rownames_to_column("genes") %>% 
  pivot_longer(-genes, 
               names_to = "condition", 
               values_to = "counts") %>% 
  mutate(study = "GSE189039")

gse201530_counts_ready_long = gse201530_counts_ready %>% 
    as.data.frame() %>% 
    rownames_to_column("genes") %>% 
  pivot_longer(-genes, 
               names_to = "condition", 
               values_to = "counts") %>% 
    mutate(study = "GSE201530")

gse201533_counts_ready_long = gse201533_counts_ready %>% 
    as.data.frame() %>% 
    rownames_to_column("genes") %>% 
  pivot_longer(-genes, 
               names_to = "condition", 
               values_to = "counts")  %>% 
    mutate(study = "GSE201533")

gse199750_counts_ready_long = gse199750_counts_ready %>% 
  as.data.frame() %>% 
  rownames_to_column("genes") %>% 
  pivot_longer(-genes, 
               names_to = "condition", 
               values_to = "counts") %>% 
    mutate(study = "GSE199750")

gse_counts_all_long = bind_rows(gse189039_counts_ready_long, 
                           gse201530_counts_ready_long,
                           gse201533_counts_ready_long,
                           gse199750_counts_ready_long)

gse_counts_all_matrix = gse_counts_all_long %>% 
  select(-study) %>% 
  pivot_wider(names_from = "condition", values_from = "counts")


write_csv(gse_counts_all_long, file = here("DGE analysis", "gse_counts_all_long.csv"))
```

## 4.4. Immune and non-immune DEGs

```{r}
immunego = ImmuneGO_Annotated_GO_8_2_24 %>% 
  mutate(genes = str_split(genes, ",")) %>%
  unnest(genes)
#### Not immune

all_degs_infected_notimmune = all_degs_p_05_vac_infected_19_12_23 %>%
  anti_join(immunego %>%
              select(genes) %>%
              distinct(),
            by = "genes") %>%
  distinct() %>% 
  mutate(bioprocess = "Other")

all_degs_infected_notimmune %>% 
  tabyl(condition, bioprocess)

write_csv(all_degs_infected_notimmune, file = here("DGE analysis", "all_degs_p_05_vac_infected_notimmune.csv"))
```

```{r}
#### Immune
all_degs_vac_infected_immune = all_degs_p_05_vac_infected_19_12_23 %>% 
  inner_join(immunego %>% 
              select(genes) %>% 
              distinct(),
            by = "genes") %>% 
  distinct() %>% 
  mutate(bioprocess = "Immune")

all_degs_p_05_vac_infected_immune %>% 
  tabyl(condition, bioprocess) 

write_csv(all_degs_p_05_vac_infected_immune, file = here("DGE analysis", "all_degs_p_05_vac_infected_immune.csv"))
```

```{r}
##### Immune and other
all_degs_processes_immune_nonimmune = bind_rows(all_degs_vac_infected_immune, 
                               all_degs_infected_notimmune)

all_degs_processes_immune_nonimmune %>%
  tabyl(condition, bioprocess) %>% 
  adorn_totals("col") %>% 
  adorn_percentages() %>% 
  adorn_pct_formatting() 

write_csv(all_degs_processes_immune_nonimmune, file = here("DGE analysis", "all_degs_processes_immune_nonimmune.csv"))
```

### 4.4.1. Retrieving genes by Gene Ontology

```{r}
# organism = "org.Hs.eg.db"
# install.packages("GO.db")
# BiocManager::install(organism, character.only = TRUE)
library(organism, character.only = TRUE)
library(clusterProfiler)
library(GO.db)

#HGNC symbol to Entrez
x <- org.Hs.egSYMBOL
mapped_genes <- mappedkeys(x)
symbols_entrez <- as.data.frame(x[mapped_genes])

## Entrez to GO
x <- org.Hs.egGO
mapped_genes <- mappedkeys(x)
entrez_geneontology <- as.data.frame(x[mapped_genes])

## GO BP ancestors
go_bp_ancestor = GOBPANCESTOR %>% 
  as.data.frame()
colnames(go_bp_ancestor)[1] <- "go_id"
colnames(go_bp_ancestor)[2] <- "go_ancestor"

## GO ID to terms
go_bp = AnnotationDbi::select(GO.db, columns=c("GOID","TERM"), keys= "BP", keytype= "ONTOLOGY") %>% 
  clean_names() %>% 
  select(go_id = goid, term)

## GO child and ancestor
go_bp_ancestor = go_bp_ancestor %>% 
  inner_join(go_bp %>% rename(go_ancestor = go_id), by = "go_ancestor") %>% 
  rename(term_ancestor = term)

#GO ids to terms
go_bp = AnnotationDbi::select(GO.db, columns=c("GOID","TERM"), keys="BP", keytype="ONTOLOGY") %>% 
  clean_names() %>% 
  rename(go_id = goid) %>% 
  select(-ontology) %>% 
  right_join(entrez_geneontology, by = "go_id") %>% 
  right_join(symbols_entrez, by = "gene_id") %>% 
  clean_names() %>% 
  left_join(go_bp_ancestor, by = "go_id") %>% 
  distinct()

#Filter main GO BP terms
go_bp_main = go_bp %>%
  filter(go_ancestor %in% c(
      "GO:0040011",
      "GO:0043473",
      "GO:0044419",
      "GO:0009987",
      "GO:0048518",
      "GO:0050789",
      "GO:0065007",
      "GO:0048511",
      "GO:0022414",
      "GO:0002376",
      "GO:0008152")) %>% 
  filter(str_detect(term_ancestor, "cellular process|biological process involved in interspecies interaction between organisms|regulation of biological process|immune system process|No annotation|Not mapped|metabolic process"),
         !term_ancestor == "positive regulation of biological process") %>% 
  mutate(term_ancestor = case_when(
           term_ancestor == "biological process involved in interspecies interaction between organisms" ~ "Interspecies interaction",
           term_ancestor == "biological regulation" ~ "Biological reg.",
           term_ancestor == "cellular process" ~ "Cellular",
           term_ancestor == "immune system process" ~ "Immune system",
           term_ancestor == "regulation of biological process" ~ "Reg. of BP",
           term_ancestor == "locomotion" ~ "Locomotion",
           term_ancestor == "metabolic process" ~ "Metabolism",
           TRUE ~ term_ancestor))

#Add set size column
go_bp_main_2 = go_bp_main %>% 
  group_by(term_ancestor) %>% 
  summarise(set_size = n()) %>% 
  inner_join(go_bp_main) %>% 
  select(term = term_ancestor, set_size, go_id = go_ancestor, symbol) %>% 
  distinct() %>% 
  mutate(annotation = "Major process")

go_bp_child = go_bp_main %>% 
  select(term, go_id, symbol) %>% 
  distinct() %>% 
  group_by(term) %>% 
  summarise(set_size = n()) %>% 
  inner_join(go_bp_main, by = "term") %>% 
  select(term, set_size, go_id, symbol) %>% 
  distinct() %>% 
  mutate(annotation = "Minor process")

go_bp_all = bind_rows(go_bp_main_2, go_bp_child)

write_csv(go_bp_all, file = here("VaxGO gene sets", "OtherGOs_Annotated_Genes.csv"))

# Collapse rows 
go_bp_all_comma = go_bp_all %>% 
  group_by(term) %>% 
  summarize(genes = paste0(genes, collapse = ",")) %>% 
  inner_join(go_bp_all, by = "term") %>% 
  select(-symbol) %>% 
  distinct() %>% 
  arrange(desc(set_size))

write_csv(go_bp_all_comma, file = here("VaxGO gene sets","OtherGOs_Annotated_GOs.csv"))

go_bp_all
go_bp_all_comma
```

```{r}
# DEGs to Entrez and GO ids --------
degs_GO = all_degs_p_05_vac_infected %>% 
  filter(direction != "NEUTRAL") %>% 
  select(condition, genes, direction) %>% 
  left_join(go_bp %>% rename(genes = symbol, entrez = gene_id), by = "genes") %>% 
  mutate(go_id = if_else(is.na(go_id), "Not mapped", go_id),
         term = if_else(is.na(term), "Not mapped", term),
         evidence = if_else(is.na(evidence), "Not mapped", evidence),
         ontology = if_else(is.na(ontology), "Not mapped", ontology),
         mapped = if_else(ontology == "Not mapped", "Not mapped", "Mapped"),
         go_ancestor = if_else(is.na(go_ancestor), "Not mapped", go_ancestor),
         term_ancestor = if_else(is.na(term_ancestor), "Not mapped", term_ancestor)) %>%
  filter(ontology %in% c("BP", "Not mapped")) %>% 
  select(-evidence) 

# Collapse gene column --------
degs_GO_comma = degs_GO %>% 
  select(-term_ancestor, -go_ancestor) %>% 
  distinct() %>% 
  group_by(condition, term) %>% 
  summarize(genes = paste0(genes, collapse = ",")) %>% 
  inner_join(ann_vaccines, by = "condition")

write_csv(degs_GO_comma, file = here("DGE analysis", "degs_annotated_allGOs.csv"))


#DEGs annotated by Main GO BPs ------

main_go = degs_GO %>%
  filter(go_ancestor %in% c(
      "Not mapped",
      "GO:0040011",
      "GO:0043473",
      "GO:0044419",
      "GO:0009987",
      "GO:0048518",
      "GO:0050789",
      "GO:0065007",
      "GO:0048511",
      "GO:0022414",
      "GO:0002376",
      "GO:0008152")) %>% 
  select(condition, genes, direction, term_ancestor) %>% 
  distinct() %>% 
  group_by(condition, direction, term_ancestor) %>% 
  summarise(n = n()) %>% 
  mutate(term_ancestor = if_else(is.na(term_ancestor), "No annotation", term_ancestor))

main_go_2 = degs_GO %>% 
  select(condition, genes, mapped, direction) %>% 
  distinct() %>% 
  group_by(condition, direction) %>% 
  summarise(n_genes_total = n()) %>% 
  inner_join(main_go, by = join_by("condition", direction)) %>% 
  ungroup() %>% 
  select(condition, direction, term_ancestor, n_genes = n, n_genes_total) %>%
  filter(str_detect(term_ancestor, "cellular process|biological process involved in interspecies interaction between organisms|regulation of biological process|immune system process|No annotation|Not mapped|metabolic process"),
         !term_ancestor == "positive regulation of biological process") %>% 
  mutate(n_genes_perc = round(n_genes/n_genes_total * 100, 1),
         term_ancestor = case_when(
           term_ancestor == "biological process involved in interspecies interaction between organisms" ~ "Interspecies interaction",
           term_ancestor == "biological regulation" ~ "Biological reg.",
           term_ancestor == "cellular process" ~ "Cellular",
           term_ancestor == "immune system process" ~ "Immune system",
           term_ancestor == "regulation of biological process" ~ "Reg. of BP",
           term_ancestor == "locomotion" ~ "Locomotion",
           term_ancestor == "metabolic process" ~ "Metabolism",
           TRUE ~ term_ancestor),
         term_ancestor = fct_relevel(term_ancestor, "Not mapped", after = Inf),
         direction = fct_relevel(direction, "UP", "DOWN")) %>% 
  inner_join(ann_vaccines_19_1_24 %>% select(condition, type, disease_vac), by = "condition")
  
write_csv(main_go_2, file = here("DGE analysis", "degs_annotated_mainGOs_stats.csv"))
```

###4.4.2.1. Boxplot

```{r fig.height=10, fig.width=15, warning=FALSE}
degs_annotated_mainGOs_stats <- read_csv(here("DGE analysis", "degs_annotated_mainGOs_stats.csv"))

# Table
df =  degs_annotated_mainGOs_stats %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  select(condition, 
         direction, 
         term_ancestor, 
         n_genes, 
         n_genes_perc,
         type = type.x,
         week,
         disease_vac = disease_vac.x) %>% 
  filter(!term_ancestor %in% c("Not mapped", "Interspecies interaction")) %>% 
  mutate(week = as.factor(week),
         week_dv = paste(week, sep = "",  " (", disease_vac, ")"),
         term_ancestor = if_else(term_ancestor == "Interspecies interaction", 
                                 "Interspecies", 
                                 term_ancestor),
         term_ancestor = fct_relevel(term_ancestor, "Immune system", "Interpecies", "Cellular", "Metabolism"))

# UP
up = df %>% 
  filter(direction == "UP") %>% 
  group_by(disease_vac) %>%
  arrange(desc(n_genes_perc)) %>% 
  ggplot() +
  aes(
    x = week_dv,
    y = n_genes_perc,
    label = condition) +
  geom_jitter(aes(colour = disease_vac,
                 shape = type,
                 size = n_genes),
             na.rm = T) +
    geom_boxplot(aes(fill = disease_vac, color = disease_vac),
               outlier.shape = NA,
               alpha = 0) +

  theme_minimal() +
  facet_wrap(~term_ancestor, nrow=1) +
  theme(axis.text.x = element_text(size = 15, color = "black", angle = 90, vjust = 0.5),
        axis.text.y = element_text(size = 15, color = "black"),
        axis.title.x = element_text(size = 15, color = "black", vjust = -1),
        axis.title.y = element_text(size = 15, color = "black"),
        strip.text.x = element_text(size = 15, face = "bold"),
        legend.text = element_text(size=15, color = "black"),
        legend.title = element_text(size=15, color = "black"),
        plot.title = element_text(hjust = 0.5,
                                  size = 15,
                                  face = "bold"))+
  labs(title = "Up regulated genes",
       x = "Week",
       y = "Number of genes (%)",
       size = "Total number of genes",
       shape = "Type")+ 
  scale_fill_npg() + 
  scale_colour_npg()+
  scale_size_continuous(range=c(1,3), breaks=c(100,500,1000))
up
  
down = df %>% 
  filter(direction == "DOWN") %>% 
  group_by(disease_vac) %>%
  arrange(desc(n_genes_perc)) %>% 
  ggplot() +
  aes(
    x = week_dv,
    y = n_genes_perc,
    label = condition) +
  geom_jitter(aes(colour = disease_vac,
                 shape = type,
                 size = n_genes),
             na.rm = T) +
    geom_boxplot(aes(fill = disease_vac, color = disease_vac),
               outlier.shape = NA,
               alpha = 0) +
  theme_minimal() +
  facet_wrap(~term_ancestor, nrow=1) +
  theme(axis.text.x = element_text(size = 15, color = "black", angle = 90, vjust = 0.5),
        axis.text.y = element_text(size = 15, color = "black"),
        axis.title.x = element_text(size = 15, color = "black", vjust = -1),
        axis.title.y = element_text(size = 15, color = "black"),
        strip.text.x = element_text(size = 15, face = "bold"),
        legend.text = element_text(size=15, color = "black"),
        legend.title = element_text(size=15, color = "black"),
        plot.title = element_text(hjust = 0.5,
                                  size = 15,
                                  face = "bold"))+
  labs(title = "Down regulated genes",
       x = "Week",
       y = "Number of genes (%)",
       size = "Total number of genes",
       shape = "Type")+ 
  scale_fill_npg() + 
  scale_colour_npg()+
  scale_size_continuous(range=c(1.5,3), breaks=c(100,500,1000))

down

patch = up/down + plot_layout(guides = "collect")

ggsave(patch, width = 15, height = 10, file = here("Figures", "MainGOs_percgenes_boxplot_up_down.png"))

patch
```

###4.4.2.2. Barplot by term

```{r fig.height=15, fig.width=20, message=FALSE, warning=FALSE}
### Bar plot by term (United) -----

#UP regulated
main_go_2_plot_up = main_go_2 %>% 
  mutate(direction = if_else(direction == "UP", "↑", "↓"),
         term_ancestor = fct_relevel(term_ancestor, "Immune system", "Cellular", "Metabolism", "Reg. of BP")) %>%
  filter(direction == "↑", 
         term_ancestor %in% c("Immune system", "Cellular", "Metabolism", "Reg. of BP")) %>% 
  ggplot() +
 aes(x = forcats::fct_rev(factor(condition)), y = n_genes_perc, fill = type) +
 geom_col() +
 coord_flip() +
 theme_minimal() +
 theme(legend.position = "top") +
 facet_grid(vars(direction),  vars(term_ancestor)) +
   scale_fill_manual(name = "Type",
    values = c('IN'= "#00A087FF", #1st Gen  vaccines
           'VV' = "#3C5488FF", #3rd Gen vaccines
           'RNA' = "#4DBBD5FF",
           'SU' = "#8491B4FF",
           'I'= "#DC0000FF",
           "H" = "grey95")) +
  theme(legend.position = "right",
          legend.title = element_text(size = 15),
          legend.text =element_text(size=15),
        axis.text.x = element_text(size = 12, color = "black", angle = 0, hjust = 1),
        axis.text.y = element_text(size = 12, color = "black", angle = 0, hjust = 1),
        strip.text.x = element_text(size = 20, color = "black", hjust = 0.5, face = "bold"),
        strip.text.y = element_text(size = 20, angle = 0, color = "white", face = "bold"),
        strip.background.x = element_rect(
          fill = "white", 
          color = "white", 
          linewidth = 1 ),
        strip.background.y = element_rect(
          fill = "#DC0000FF", 
          color = "#DC0000FF", 
          linewidth = 1 )
        ) +
  geom_text(aes(label = paste0(n_genes_perc, "% (", n_genes, ")"), 
                y = n_genes_perc), 
            hjust = 0, 
            size = 4,
            angle = 0) +
  labs(title = "",
       x = "",
       y = "") +
  ylim(0,120)

#DOWN regulated
main_go_2_plot_down = main_go_2 %>% 
  mutate(direction = if_else(direction == "UP", "↑", "↓"),
         term_ancestor = fct_relevel(term_ancestor, "Immune system", "Cellular", "Metabolism", "Reg. of BP")) %>%
  filter(direction == "↓", 
         term_ancestor %in% c("Immune system", "Cellular", "Metabolism", "Reg. of BP")) %>% 
  ggplot() +
 aes(x = forcats::fct_rev(factor(condition)), y = n_genes_perc, fill = type) +
 geom_col() +
 coord_flip() +
 theme_minimal() +
 theme(legend.position = "top") +
 facet_grid(vars(direction),  vars(term_ancestor)) +
   scale_fill_manual(name = "Type",
    values = c('IN'= "#00A087FF", #1st Gen  vaccines
           'VV' = "#3C5488FF", #3rd Gen vaccines
           'RNA' = "#4DBBD5FF",
           'SU' = "#8491B4FF",
           'I'= "#DC0000FF",
           "H" = "grey95")) +
  theme(legend.position = "right",
          legend.title = element_text(size = 15),
          legend.text =element_text(size=15),
        axis.text.x = element_text(size = 12, color = "black", angle = 0, hjust = 1),
        axis.text.y = element_text(size = 12, color = "black", angle = 0, hjust = 1),
        strip.text.x = element_blank(),
        strip.text.y = element_text(size = 20, angle = 0, color = "black", face = "bold"),
        strip.background.x = element_blank(),
        strip.background.y = element_rect(
          fill = "#4DBBD5FF", 
          color = "#4DBBD5FF", 
          linewidth = 1)
        ) +
  geom_text(aes(label = paste0(n_genes_perc, "% (", n_genes, ")"), 
                y = n_genes_perc), 
            hjust = 0, 
            size = 4,
            angle = 0) +
  labs(x = "",
       y = "Number of genes (%)") +
  ylim(0,120)

patch = (main_go_2_plot_up/main_go_2_plot_down) + plot_layout(guides = "collect")

ggsave(patch, file = here("Figures", "main_go_2_plot_values_UP_DOWN_regulated.png"), width = 20, height = 15)

patch

```

```{r fig.height=15, fig.width=8}
### Totals ------
main_go_2_plot_totals_values = main_go_2 %>%
  select(condition, n_genes_total, type, direction) %>% 
  distinct() %>% 
  mutate(stat = "Total genes",
         direction = if_else(direction == "UP", "↑", "↓")) %>% 

  ggplot() +
 aes(x = forcats::fct_rev(factor(condition)), y = n_genes_total, fill = type) +
 geom_col() +
 coord_flip() +
 theme_minimal() +
 theme(legend.position = "top") +
 facet_grid(direction~ stat, scales = "free") +
   scale_fill_manual(
    values = c('IN'= "#00A087FF", #1st Gen  vaccines
           'VV' = "#3C5488FF", #3rd Gen vaccines
           'RNA' = "#4DBBD5FF",
           'SU' = "#8491B4FF",
           'I'= "#DC0000FF",
           "H" = "grey95")) +
    theme(legend.position = "right",
          legend.title = element_text(size = 15),
          legend.text =element_text(size=15),
        axis.text.x = element_text(size = 12, color = "black", angle = 0, hjust = 1),
        axis.text.y = element_text(size = 12, color = "black", angle = 0, hjust = 1),
        strip.text.x = element_text(size = 12, color = "white", hjust = 0.5),
        strip.text.y = element_text(size = 15, angle = 0),
        strip.background.x = element_rect(
          fill = "black", 
          color = "black", 
          linewidth = 1 ),
        strip.background.y = element_rect(
          fill = "gray70", 
          color = "gray70", 
          linewidth = 1 )) +
  geom_text(aes(label = n_genes_total, y = n_genes_total),
            hjust = 0, 
            size = 3,
            angle = 0) +
  ylim(0, 2600) +
  labs(x = "", 
       y = "Total genes",
       fill = "Type")

main_go_2_plot_totals_values
ggsave(main_go_2_plot_totals_values, file = "main_go_2_plot_totals_values.png", width = 8, height = 15)
```

```{r fig.height=7, fig.width=15}
### Immune in totals ------

#UP
immune_other_plot  = main_go_2 %>% 
  mutate(direction = if_else(direction == "UP", "Upregulated", "Downregulated"),
         term_ancestor = fct_relevel(term_ancestor, "Immune system", "Cellular", "Metabolism", "Reg. of BP")) %>%
  filter(term_ancestor %in% c("Immune system")) %>% 
  rename(n_genes_immune = n_genes) %>% 
  pivot_longer(names_to = "totals", values_to = "total_number", cols = c(n_genes_immune, n_genes_total)) %>% 
  select(condition, direction, term_ancestor, n_genes_perc, totals, total_number) %>% 
  group_by(condition) %>% 
  mutate(immune_genes = if_else(totals == "n_genes_immune", total_number, NA),
         total_degs = if_else(totals == "n_genes_total", total_number, NA),
         totals = fct_rev(factor(totals)),
         direction = fct_rev(direction)) %>% 
  fill(total_degs, .direction = "up") %>% 
ggplot() +
 aes(x = fct_rev(factor(condition)), y = total_number, fill = totals) +
 geom_col() +
 coord_flip() +
 theme_minimal() +
  scale_fill_manual(name = "",
    values = c("n_genes_total" = "gray90",
              "n_genes_immune" = "#4DBBD5FF"),
    labels = c('Total', 'Immune')) +
  facet_grid(~direction) +
  theme(legend.position = "right", 
        legend.title = element_text(size = 10),
        legend.text = element_text(size=10),
        axis.title.x = element_text(size = 10, color = "black"),
        axis.text.x = element_text(size = 10, color = "black", angle = 0, hjust = 1),
        axis.text.y = element_text(size = 10, color = "black", angle = 0, hjust = 1),
                strip.text.x = element_text(size = 12, color = "white", hjust = 0.5),
        strip.text.y = element_text(size = 15, angle = 0),
        strip.background.x = element_rect(
          fill = "black", 
          color = "black", 
          linewidth = 1 )) +
  geom_text(aes(label = if_else(!is.na(immune_genes), paste0(immune_genes, " (", n_genes_perc, "% of ", total_degs,")") , NA), 
                y = total_degs + immune_genes), 
            hjust = -0.05, 
            size = 3,
            angle = 0) +
  labs(title = "",
       fill = "Gene type",
       x = "",
       y = "Number of genes") +
  ylim(0, 4000)

ggsave(immune_other_plot, file = "immune_other_plot.png", width = 15, height = 7)
immune_other_plot

```

### 4.4.3 Shared genes

#### 4.4.3.1. Immune related only

```{r}

degs_all_updown = all_degs_p_05_vac_infected %>% 
  filter(direction != "NEUTRAL") %>% 
  distinct()

ImmuneGO_Annotated_Genes = ImmuneGO_Annotated_Genes %>% 
  filter(!(process %in% c("TCR REPERTOIRE", "BCR REPERTOIRE")))



shared_genes_infection = degs_all_updown %>%
  filter(disease_vac == "I") %>% 
  select(condition, genes) %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  select(condition, genes, vaccine) %>% 
  inner_join(ImmuneGO_Annotated_Genes %>% select(genes) %>% distinct(), by = "genes") %>% 
  group_by(genes) %>% 
  summarise("Infection" = n()) %>% 
  arrange(desc("Infection"), genes) %>% 
  ungroup() %>% 
  pivot_longer(-genes, names_to = "infection_vaccination", values_to = "n_conditions") %>% 
  inner_join(ImmuneGO_Annotated_Genes , by = "genes") %>% 
  select(genes, infection_vaccination, n_conditions) %>% 
  mutate(comparison = "Infection only") %>% 
  arrange(desc(n_conditions)) %>% 
  distinct()

shared_genes_vaccines = degs_all_updown %>%
  filter(disease_vac == "V") %>% 
  select(condition, genes) %>% 
  distinct() %>% 
  inner_join(ImmuneGO_Annotated_Genes %>% select(genes) %>% distinct(), by = "genes") %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  select(condition, genes, vaccine) %>% 
  group_by(genes) %>% 
  summarise("Vaccination" = n()) %>% 
  arrange(desc("Vaccination"), genes) %>% 
  ungroup() %>% 
  pivot_longer(-genes, names_to = "infection_vaccination", values_to = "n_conditions") %>% 
  inner_join(ImmuneGO_Annotated_Genes , by = "genes") %>% 
  select(genes, infection_vaccination,n_conditions) %>% 
  mutate(comparison = "Vaccination only") %>% 
  arrange(desc(n_conditions)) %>% 
  distinct()


# Infection and vaccination
shared_genes_vaccines_infection = shared_genes_vaccines %>% 
  select(genes, n_conditions) %>% 
  inner_join(shared_genes_infection %>% select(genes, n_conditions), by = "genes") %>% 
  rename(Vaccination = n_conditions.x, Infection = n_conditions.y) %>% 
  distinct() %>% 
  pivot_longer(-genes, names_to = "infection_vaccination", values_to = "n_conditions") %>% 
  inner_join(ImmuneGO_Annotated_Genes , by = "genes") %>% 
  select(genes, infection_vaccination,n_conditions) %>% 
  mutate(comparison = "Infection and vaccination") %>% 
  distinct()
  

shared_genes_comparisons = bind_rows(shared_genes_infection, 
                                     shared_genes_vaccines, 
                                     shared_genes_vaccines_infection)

write_csv(shared_genes_comparisons, file = here("Figures", "shared_genes_infection_vaccination_comparisons.csv"))

```

```{r fig.height=9, fig.width=10}
# Vaccine and infection ----
shared_genes_infection_vaccination = shared_genes_comparisons %>% 
  filter(comparison == "Infection and vaccination")

shared_genes_infection = shared_genes_comparisons %>% 
  filter(comparison == "Infection only")

shared_genes_vaccines = shared_genes_comparisons %>% 
  filter(comparison == "Vaccination only")


shared_genes_vaccines_infection_plot = shared_genes_infection_vaccination %>% 
  select(genes, infection_vaccination, n_conditions, comparison) %>% 
  distinct() %>%
  slice_head(n=40) %>%  
  ggplot() +
  aes(
    x = reorder(genes, -n_conditions),
    y = n_conditions,
    fill = infection_vaccination
  ) +
  geom_col() +
  scale_fill_manual(
    values = c("Vaccination" = "#4DBBD5FF",
              "Infection" = "#DC0000FF")
    )+
  theme_minimal() +
  labs(fill = "Status") +
  xlab("") + 
  ylab("") +
  theme(legend.position = "right",
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1, size = 10, color = "black"),
        axis.text.y = element_text(size = 10, angle = 90, color = "black"),
        panel.grid.minor.y = element_blank(),
        legend.text = element_text(size=10),
        legend.title = element_text(size = 10, face = "bold")) 

#Vaccination only -----
shared_genes_vaccines_only = shared_genes_vaccines %>% 
  anti_join(shared_genes_infection, by = "genes") %>% 
  inner_join(ImmuneGO_Annotated_Genes %>% select(genes), by = "genes") %>% 
  select(genes, n_conditions) %>% 
  mutate(comparison = "Vaccination only") %>% 
  arrange(desc(n_conditions)) %>% 
  distinct()

shared_genes_vaccines_only_plot = shared_genes_vaccines_only %>% 
  select(genes, n_conditions, comparison) %>% 
  distinct() %>%
  slice_head(n=20) %>% 
  ggplot() +
  aes(x = reorder(genes, -n_conditions), 
      fill = comparison, 
      weight = n_conditions) +
  geom_bar() +
  scale_fill_manual(
    values = c("Vaccination only" = "#4DBBD5FF",
              "Infection only" = "#DC0000FF")) +
  theme_minimal() +
  labs(fill = "Vaccination only") +
  xlab("") + 
  ylab("Number of shared conditions") +
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1, size = 10, color = "black"),
        axis.text.y = element_text(size = 10, angle = 0, color = "black"),
        panel.grid.minor.y = element_blank(),
        legend.text = element_text(size=10),
        legend.title = element_text(size = 10, face = "bold")) 


# Infection only
shared_genes_infection_only = shared_genes_infection %>% 
  anti_join(shared_genes_vaccines_only, by = "genes") %>%
  inner_join(ImmuneGO_Annotated_Genes, by = "genes") %>% 
  select(genes, n_conditions) %>% 
  mutate(comparison = "Infection only") %>% 
  arrange(desc(n_conditions))

shared_genes_infection_only_plot = shared_genes_infection_only %>% 
  select(genes, n_conditions, comparison) %>% 
  arrange(desc(n_conditions)) %>% 
  distinct() %>%
  slice_head(n=20) %>% 
  ggplot() +
  aes(x = reorder(genes, -n_conditions), 
      fill = comparison, 
      weight = n_conditions) +
  geom_bar() +
  scale_fill_manual(
    values = c("Vaccination only" = "#4DBBD5FF",
              "Infection only" = "#DC0000FF")) +
  theme_minimal() +
  labs(fill = "Infection only") +
  xlab("Genes") + 
  ylab("") +
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1, size = 10, color = "black"),
        axis.text.y = element_text(size = 10, angle = 0, color = "black"),
        panel.grid.minor.y = element_blank(),
        legend.text = element_text(size=10),
        legend.title = element_text(size = 10, face = "bold"))



patch = shared_genes_vaccines_infection_plot / shared_genes_vaccines_only_plot / shared_genes_infection_only_plot + plot_layout(guides = 'collect') +
  plot_annotation(
  title = "Top 20 shared immune-related genes",
  subtitle = "Infection and vaccination, Vaccination only, Infection only ")

ggsave(patch, file = here("Figures", "Genes_shared_Vaccination_Infection_united.png"), width = 10, height = 9)

patch
```

#### 4.4.3.1. Not-immune related only

```{r}
degs_all_updown = all_degs_p_05_vac_infected %>% 
  filter(direction != "NEUTRAL") %>% 
  distinct()

ImmuneGO_Annotated_Genes = ImmuneGO_Annotated_Genes

## Processing -----

#Infection -----
shared_genes_infection = degs_all_updown %>%
  filter(disease_vac == "I") %>% 
  select(condition, genes) %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  select(condition, genes, vaccine) %>% 
  anti_join(ImmuneGO_Annotated_Genes %>% select(genes) %>% distinct(), by = "genes") %>% 
  group_by(genes) %>% 
  summarise("Infection" = n()) %>% 
  arrange(desc("Infection"), genes) %>% 
  ungroup() %>% 
  pivot_longer(-genes, names_to = "infection_vaccination", values_to = "n_conditions") %>% 
  anti_join(ImmuneGO_Annotated_Genes , by = "genes") %>% 
  select(genes, infection_vaccination,n_conditions) %>% 
  mutate(comparison = "Infection only") %>% 
  arrange(desc(n_conditions))

#Vaccine -----
shared_genes_vaccines = degs_all_updown %>%
  filter(disease_vac == "V") %>% 
  select(condition, genes) %>% 
  anti_join(ImmuneGO_Annotated_Genes %>% select(genes) %>% distinct(), by = "genes") %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  select(condition, genes, vaccine) %>% 
  group_by(genes) %>% 
  summarise("Vaccination" = n()) %>% 
  arrange(desc("Vaccination"), genes) %>% 
  ungroup() %>% 
  pivot_longer(-genes, names_to = "infection_vaccination", values_to = "n_conditions") %>% 
  anti_join(ImmuneGO_Annotated_Genes %>% select(genes) %>% distinct(), by = "genes") %>% 
  select(genes, infection_vaccination,n_conditions) %>% 
  mutate(comparison = "Vaccination only") %>% 
  arrange(desc(n_conditions))


# Infection and vaccination -----
shared_genes_vaccines_infection = shared_genes_vaccines %>% 
  inner_join(shared_genes_infection, by = "genes") %>% 
  select(genes, Vaccination = n_conditions.x, Infection = n_conditions.y) %>% 
  pivot_longer(-genes, names_to = "infection_vaccination", values_to = "n_conditions") %>% 
  anti_join(ImmuneGO_Annotated_Genes , by = "genes") %>% 
  select(genes, infection_vaccination,n_conditions) %>% 
  mutate(comparison = "Infection and vaccination") 

shared_genes_comparisons = bind_rows(shared_genes_infection, 
                                     shared_genes_vaccines, 
                                     shared_genes_vaccines_infection) %>% 
  arrange(desc(n_conditions))

write_csv(shared_genes_comparisons, file = here("DGE analysis", "non_Immune_shared_genes_infection_vaccination_comparisons.csv"))

shared_genes_comparisons
```

```{r fig.height=9, fig.width=10}
# Plotting ------
shared_genes_infection_vaccination = shared_genes_comparisons %>%
  filter(comparison == "Infection and vaccination")

shared_genes_infection = shared_genes_comparisons %>% 
  filter(comparison == "Infection only")

shared_genes_vaccines = shared_genes_comparisons %>% 
  filter(comparison == "Vaccination only")


#Infection and vaccination
shared_genes_vaccines_infection_plot = shared_genes_vaccines_infection %>% 
  select(genes, infection_vaccination, n_conditions) %>% 
  distinct() %>%
  slice_head(n=40) %>% 
  ggplot() +
  aes(
    x = reorder(genes, -n_conditions),
    y = n_conditions,
    fill = infection_vaccination
  ) +
  geom_col() +
  scale_fill_manual(
    values = c("Vaccination" = "#4DBBD5FF",
              "Infection" = "#DC0000FF")) +
  theme_minimal() +
  labs(fill = "Status") +
  xlab("") + 
  ylab("") +
  theme(legend.position = "right",
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1, size = 10, color = "black"),
        axis.text.y = element_text(size = 10, angle = 0, color = "black"),
        panel.grid.minor.y = element_blank(),
        legend.text = element_text(size=10),
        legend.title = element_text(size = 10, face = "bold"))

#Vaccination only -----

shared_genes_vaccines_only_plot = shared_genes_vaccines %>% 
  select(genes, infection_vaccination, n_conditions, comparison) %>% 
  arrange(desc(n_conditions)) %>% 
  distinct() %>%
  slice_head(n=20) %>% 
  ggplot() +
  aes(x = reorder(genes, -n_conditions), 
      fill = comparison, 
      weight = n_conditions) +
  geom_bar() +
  scale_fill_manual(
    values = c("Vaccination only" = "#4DBBD5FF",
              "Infection only" = "#DC0000FF")) +
  theme_minimal() +
  labs(fill = "Infection only") +
  xlab("") + 
  ylab("Number of shared conditions") +
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1, size = 10, color = "black"),
        axis.text.y = element_text(size = 10, angle = 0, color = "black"),
        panel.grid.minor.y = element_blank(),
        legend.text = element_text(size=10),
        legend.title = element_text(size = 10, face = "bold")) 



# Infection only -------

shared_genes_infection_only_plot = shared_genes_infection %>% 
  select(genes, infection_vaccination, n_conditions, comparison) %>% 
  arrange(desc(n_conditions)) %>% 
  distinct() %>%
  slice_head(n=20) %>% 
  ggplot() +
  aes(x = reorder(genes, -n_conditions), 
      fill = comparison, 
      weight = n_conditions) +
  geom_bar() +
  scale_fill_manual(
    values = c("Vaccination only" = "#4DBBD5FF",
              "Infection only" = "#DC0000FF")) +
  theme_minimal() +
  labs(fill = "Infection only") +
  xlab("Genes") + 
  ylab("") +
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1, size = 10, color = "black"),
        axis.text.y = element_text(size = 10, angle = 0, color = "black"),
        panel.grid.minor.y = element_blank(),
        legend.text = element_text(size=10),
        legend.title = element_text(size = 10, face = "bold"))

patch = shared_genes_vaccines_infection_plot / shared_genes_vaccines_only_plot / shared_genes_infection_only_plot + plot_layout(guides = 'collect') +
  plot_annotation(
  title = "Top 20 shared non-immune-related genes",
  subtitle = "Infection and vaccination, Vaccination only, Infection only ")


ggsave(patch, file =here("Figures", "NonImmune_Genes_shared_Vaccination_Infection_united.png"), width = 10, height = 9)

patch
```

# 5. Gene set enrichment analysis

O ORA e GSEA podem ser pouco sensíveis quando o pvalue/padj \< 5%. Na
documentação, recomendam usar o cutoff de padj (FDR) de 25% para incluir
gene sets que estejam 75% corretos - isto é, em 4 observações, 3 estão
corretas -, o que aumenta o número de resultados interessantes. Para
verificar se são realmente significantes, visualize no heatmap com
labels de \*significancia. Referencia:
Link{<https://www.gsea-msigdb.org/gsea/doc/GSEAUserGuideFrame.html>
<https://yulab-smu.top/biomedical-knowledge-mining-book/universal-api.html>

### 5.1. Input

```{r}
######### INPUT AQUI ######### 

#Data
all_degs_p_05_vac_infected = all_degs_p_05_vac_infected_19_12_23 %>% 
  select(!c("...1", "...12")) %>% 
  distinct()

#Gene sets
ImmuneGO_Annotated_Genes = readRDS(here("VaxGO gene sets", "ImmuneGO_Annotated_Genes_2024-03-26.rds"))

CellMarker_ImmuneCells = read_csv(here("VaxGO gene sets", "CellMarker_ImmuneCells.csv"))


Immune_GO_T2G_general = ImmuneGO_Annotated_Genes %>% 
  filter(go_term == "Manual",
         !process %in% c("ADAPTIVE IMMUNE SYSTEM",
                         "INNATE IMMUNE SYSTEM",
                         "BCR REPERTOIRE",
                         "TCR REPERTOIRE")) %>% 
  select(process, genes)


Immune_GO_T2G_specfic = ImmuneGO_Annotated_Genes %>% 
  filter(!process %in% c("ADAPTIVE IMMUNE SYSTEM",
                         "INNATE IMMUNE SYSTEM",
                         "BCR REPERTOIRE",
                         "TCR REPERTOIRE")) %>% 
  select(process=gene_set_short, genes) 


CellMarker_genes = CellMarker_ImmuneCells %>% 
  select(cell_name, marker) %>% 
  tibble()


TERM2GENE = Immune_GO_T2G_general
geneset_name = "Immune_GO_T2G_general"


```

### 5.2. ORA

#### 5.2.1. Multiple

```{r}
# Filtrar os genes para todas as condições
df <- all_degs_p_05_vac_infected_notimmune %>% 
  filter(direction != "NEUTRAL")

ora_results <- list()
for (condition_2 in unique(df$condition)) {

    genes <- df %>% 
    filter(condition == condition_2) %>% 
    pull(genes)

  go_enrich <- tryCatch({
    enrichGO(gene = genes,
             OrgDb = organism, 
             keyType = 'SYMBOL',
             readable = T,
             ont = "BP",
             pvalueCutoff = 0.05, 
             qvalueCutoff = 0.10) %>% 
      as.data.frame() %>%
      mutate(condition = condition_2)
  }, error = function(e) {
    return(NULL)
  })

  if (!is.null(go_enrich)) {
    ora_results[[condition_2]] <- go_enrich   
  }
}

ora_results_final <- bind_rows(ora_results)

write.csv(ora_results_final, file = "all_degs_other_ORA_2.csv", row.names = F)

```

#### 5.2.2. GO Analysis

```{r}
library(GO.db)
columns(GO.db)

go_bp_ancestor =  GOBPANCESTOR %>% 
  as.data.frame()
names(go_bp_ancestor[, 1]) = "go_child"

go = ora_results_final %>% 
  clean_names() %>% 
  rownames_to_column("delete") %>% 
  mutate(log_fdr = -log10(qvalue)) %>% 
  select(condition, go_id = id, description, log_fdr, qvalue, gene_id, gene_ratio, count) %>% 
  filter(qvalue < 0.10) %>%
  group_by(condition) %>% 
  arrange(desc(log_fdr)) %>% 
  slice_head(n = 5)

```

#### 5.2.3. Heatmap

```{r}

# Anotações das amostras/condições (Vacinas)
ann_vaccines

heatmap_data = go
file = "degs_up_down_notimmune_ora_top10"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_log_fdr = heatmap_data %>% 
  select(description, condition, log_fdr)

# Converter para wide table
matrix_data <- dcast(heatmap_data_log_fdr, 
                         description ~ condition, 
                         value.var = "log_fdr", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('description') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines, by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(dose) %>% 
  relocate(dose, .before=vaccine) %>% 
  column_to_rownames(var= "condition")  %>% 
  mutate(dose = ifelse(is.na(dose), "NA", dose))

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(dose) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c("...1":previous_vaccination)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() #%>% 
  #scale() #normalizar

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Ordenar colunas
ann_vaccines_heatmap = ann_vaccines_heatmap %>% 
  select(type, week, day)


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(0, 3), c("white", "#ed6a5a"))

#Parameters
mat = matrix_data_ordered
width_image = 30
height_image = 30
width = unit(15, "cm")
height = unit(20, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= paste0(file, sep ="_", "Ungrouped_Complexheatmap.png")

png(file = fileimageheatmap_ungrouped, width= width_image, height=height_image ,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "NES", #Legend
                        cell_fun = function(j, i, x, y, width, height, fill) {
        if(mat[i, j] > 0)
            grid.text(sprintf("%.1f", mat[i, j]), x, y, gp = gpar(fontsize = 8))},
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 10),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        # row_split = NULL, #Split row clusters
                        # row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_split = 3, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()

```

### 5.3. GSEA

#### 5.3.1. Multiple

```{r}
autoGSEA <- function(df, TERM2GENE, geneset_name) {
  resultados <- list()

  condicoes <- df$condition %>% 
    unique() %>% 
    as.character()

  for (condicao in condicoes) {
    # Selecionar DEGs para a condição atual
    degs_condicao <- df %>% 
      filter(condition == condicao) %>%
      select(genes, log2fold_change) %>%
      distinct() %>%
      mutate(rank = rank(log2fold_change, ties.method = "random")) %>%
      arrange(desc(rank))

    # Garantir que log2fold_change seja um vetor
    gene_list_lf2c <- as.vector(degs_condicao$log2fold_change)
    names(gene_list_lf2c) <- degs_condicao$genes
    gene_list_lf2c <- na.omit(gene_list_lf2c)
    gene_list_lf2c <- sort(gene_list_lf2c, decreasing = TRUE)

    ############ IMMUNE GO
    auto_gsea <- tryCatch({
      GSEA(geneList = gene_list_lf2c,
           TERM2GENE = TERM2GENE,
           ont=
           minGSSize = 1,
           maxGSSize = 1000,
           pvalueCutoff = 0.25,
           pAdjustMethod = "BH") %>%
        as.data.frame() %>%
        arrange(qvalue) %>%
        mutate(condition = condicao,
               gsea_enrichment = "ImmuneGO")
    }, error = function(e) {
      return(NULL)  # Retorna NULL caso ocorra o erro
    })

    if (!is.null(auto_gsea)) {
      resultados[[paste(condicao, geneset_name, sep = "_")]] <- auto_gsea
    }
  }
  

  # Combinar todos os resultados em um único dataframe
  resultado_final <- bind_rows(resultados)

  return(resultado_final)
}
```

```{r}
# Perform GSEA
all_degs_p_05_vac_infected_GSEA_ALL <- all_degs_p_05_vac_infected %>%
  select(condition, genes, log2fold_change) %>% 
  autoGSEA(TERM2GENE = TERM2GENE, 
           geneset_name = geneset_name) 

all_degs_p_05_vac_infected_GSEA = all_degs_p_05_vac_infected_GSEA_ALL %>% 
  rownames_to_column("delete") %>% 
  select(condition, ID, everything()) %>% 
  clean_names()

view(all_degs_p_05_vac_infected_GSEA)

#Save
write_csv(all_degs_p_05_vac_infected_GSEA, 
          file = here("Gene set enrichment analysis", paste0("all_degs_p_05_vac_infected_", geneset_name, "_GSEA_ALL_",today(), ".csv")))

view(all_degs_p_05_vac_infected_GSEA_ALL)

```

#### 5.3.2. Shared DEGs Heatmap

```{r}
#Abrir tabela de GSEA
dados = DEGs_All_studies_ImmuneGO_GSEA_1_

dados_long <- dados %>%
  mutate(core_enrichment = strsplit(core_enrichment, "/")) %>%  # Divida a coluna em listas de genes
  unnest(core_enrichment)  # Transforme a lista em várias linhas

#Filtrar genes por categorias
genes_adaptive <- dados_long %>%
  filter(`Immune system` == "Adaptive")
```

Calculando media, soma etc. por categorias em uma tabela longa

```{r}
dados = DEGs_All_studies_ImmuneGO_GSEA_1_
# Selecione as colunas de interesse
names(dados)

# Calcular a soma dos valores da coluna NES para cada combinação de Vaccine e Immune system
resultado <- dados %>%
  group_by(Vaccine, `Immune system`) %>%
  summarise(`NES sum` = sum(NES))

degs_all = DEGs_All_studies_DEGs_All_Studies
degs_all_long <- pivot_longer(degs_all, cols = everything(), names_to = "Vaccine", values_to = "Genes")

#Salvar
write.csv(degs_all_long, file="degs_all_long.csv")
```

Transformando tabela longa em tabela wide para pca e heatmap

Shared Genes

```{r}
#Selecionar colunas de comparação e valores. Trocar value.var pela variável de interesse ("Shared", "-log(p)")

sharedDEGs_DOWN_fisher_sig = sharedDEGs_DOWN_fisher %>% filter()

matrix_data <- dcast(sharedDEGs_DOWN_fisher, `Cond1` ~ `Cond2`, value.var = "Percentage_Shared_Cond1", fun.aggregate = mean)

#Converter coluna 1 em rownames e excluir
matrix_data$Cond1 = as.factor(matrix_data$Cond1)
rownames(matrix_data) <- matrix_data$Cond1
matrix_data <- matrix_data %>% select(-`Cond1`)

#Arredondar valores para facilitar visualização
matrix_data <- round(matrix_data, 1)
matrix_data <- replace(matrix_data, matrix_data == "NaN", 0)

#Salvar
write.csv(matrix_data, file = "heatmap_sharedDEGs_UP_shared_percent.csv")

#Visualizar
heatmaply(matrix_data, 
          main = "Shared DEGs - UP-DOWN - %identity", 
          cellnote = matrix_data, 
          cellnote_size = 8,
          cellnote_textposition = "middle right",
  scale_fill_gradient_fun = ggplot2::scale_fill_gradient2(
    low = "white", 
    high = "red", 
    midpoint = 1, 
    limits = c(0, 60)
  ))
```

Down vs UP only

```{r}
####### DOWN vs UP only ######## 
filtered_data <- shared_genes_df_mirror %>%
  filter((grepl("UP", Cond1) & grepl("DOWN", Cond2)) | (grepl("DOWN", Cond1) & grepl("UP", Cond2)))

matrix_data <- dcast(filtered_data, `Cond1` ~ `Cond2`, value.var = "Percentage_Shared_Cond1", fun.aggregate = mean)

#Converter coluna 1 em rownames e excluir
rownames(matrix_data) <- matrix_data$`Cond1`
matrix_data <- matrix_data %>% select(-`Cond1`)

#Arredondar valores para facilitar visualização
matrix_data <- round(matrix_data, 1)
matrix_data <- replace(matrix_data, matrix_data == "NaN", 0)

#Arredondar valores para facilitar visualização
matrix_data <- round(matrix_data, 1)

#Salvar
write.csv(matrix_data, file = "heatmap_sharedDEGs_UP_vs_DOWN_shared_percent.csv")

#Visualizar
heatmaply(matrix_data, 
          main = "Shared DEGs - UP vs. DOWN - %identity", 
          cellnote = matrix_data, 
          cellnote_size = 8,
          cellnote_textposition = "middle right",
  scale_fill_gradient_fun = ggplot2::scale_fill_gradient2(
    low = "white", 
    high = "red", 
    midpoint = 1, 
    limits = c(0, 100)
  )
)

```

#### 5.3.3. Barplot: Innate

```{r}
#Definir conjunto de dados
Immune_GO_GSEA$`-log(qvalue)` = as.numeric(Immune_GO_GSEA$`-log(qvalue)`)

# Filtrar as linhas onde a coluna "Gene set" está em valores_desejados
GO_set_innate <- Immune_GO_GSEA %>% filter(`Gene set short` %in% c("INNATE IMMUNE RESPONSE")) %>% mutate(`-log(qvalue)` = as.numeric(`-log(qvalue)`))

GO_set_innate = merge(GO_set_innate, ann_vaccines_complex, by.x = "Vaccine", by.y = "Condition")
esquisser(GO_set_innate)

GO_set_innate_bars = ggplot(GO_set_innate) +
  aes(
    x = Day,
    fill = `Vaccine group`,
    colour = Vaccine,
    weight = NES
  ) +
  geom_bar(position = "dodge") +
  scale_fill_viridis_d(option = "magma", direction = 1) +
  scale_color_viridis_d(option = "magma", direction = 1) +
  labs(y = "-LOG(Q)") +
  theme_minimal() +
  facet_wrap(vars(Dose))

ggsave(GO_set_innate_bars, file = "GO_set_innate_bars.png")

```

```{r}
#Definir conjunto de dados
Immune_GO_GSEA$`-log(qvalue)` = as.numeric(Immune_GO_GSEA$`-log(qvalue)`)

# Filtrar as linhas onde a coluna "Gene set" está em valores_desejados
GO_set_adap <- Immune_GO_GSEA %>% filter(`Gene set short` %in% c("ADAPTIVE IMMUNE RESPONSE")) %>% mutate(`-log(qvalue)` = as.numeric(`-log(qvalue)`))

GO_set_adap = merge(GO_set_adap, ann_vaccines_complex, by.x = "Vaccine", by.y = "Condition")
esquisser(GO_set_adap)

GO_set_adap_bars = ggplot(GO_set_adap) +
  aes(
    x = Day,
    fill = `Vaccine group`,
    colour = Vaccine,
    weight = NES
  ) +
  geom_bar(position = "dodge") +
  scale_fill_viridis_d(option = "magma", direction = 1) +
  scale_color_viridis_d(option = "magma", direction = 1) +
  labs(y = "-LOG(Q)") +
  theme_minimal() +
  facet_wrap(vars(Dose))

ggsave(GO_set_adap_bars, file = "GO_set_adap_bars.png")

```

```{r}
GO_set_humoral <- Immune_GO_GSEA %>% filter(`Gene set short` %in% c("innate immune response", "adaptive immune response", "immunoglobulin mediated immune response", "T cell activation"))



#Definir nome para arquivos gerados
file_bar <- "Immune_GO_general"
fileimage <- paste(file_bar, "_barplot.png", sep = "") 

#Input

barras = 	GO_set_general #Trocar

#Criar gráfico de barras
GO_bars <- ggplot(barras, aes(x = NES, y = `Vaccine`, fill = `-log(qvalue)`)) +
  geom_col(position = "dodge") +
  geom_text(aes(label = round(`-log(qvalue)`, 2)),
            position = position_dodge(width = 0), hjust = 0, vjust = 0, size = 3, angle = 0) +
  theme_minimal() +
  labs(title = fileimage, size = 5,
       x = "NES",
       y = "Vaccine") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 6),
        legend.position = "right") +  # Posição da legenda à direita
  facet_wrap(~`Gene set, short`, ncol = 2) +
  scale_fill_gradient(
    low = "gray90",
    high = "red",# Cor para valores ausentes (NA)
    guide = guide_colorbar(  # Personalizar a legenda de cores
      title = "-log(qvalue)",  # Título da legenda
      title.position = "top",  # Posição do título
      title.hjust = 0.5  # Ajuste de alinhamento do título
    )
  )

GO_bars_bg = GO_bars + theme(
  plot.background = element_rect(fill = "white"),  # Fundo branco
  panel.background = element_rect(fill = "white"),  # Fundo branco
  legend.key = element_rect(fill = "white"),  # Fundo branco para a legenda
  legend.background = element_rect(fill = "white"),  # Fundo branco para a legenda
  legend.title = element_text(color = "black")  # Cor do título da legenda
)

# Salvar
ggsave(GO_bars_bg, file=fileimage, width = 10, height = 5, units = "in") #Trocar

# Display
print(GO_bars_bg)
```

#### 5.3.4. Heatmap: Gene L2FC by immune process

```{r}
#Padronizar tabela
Immune_allgenes_allprocesses = Immune_allgenes_l2fc %>% 
  rename(Condition = study, Genes = genes) %>% filter(Vaccine != "NA")

#General
Immune_GO_general_innate = Immune_allgenes_allprocesses %>% 
  filter(Process == "innate immune response") %>% 
  filter(Condition != "NA")

Immune_GO_general_adaptive = Immune_allgenes_allprocesses %>% 
  rename(Condition = study, Genes = genes) %>% 
  filter(Process == "adaptive immune response",
         Condition != "NA"))

\#Humoral
Immune_GO_humoral_humoral = Immune_GO_humoral %>% 
  rename(Condition = study, Genes = genes) %>% 
  filter(Process == "humoral immune response",
         Condition != "NA"))

Immune_GO_humoral_IgMediated = Immune_GO_humoral %>% 
  rename(Condition = study, Genes = genes) %>% 
  filter(Process == "immunoglobulin mediated immune response", 
         Condition != "NA"))

#Cellular
Immune_GO_adaptive_cellular_tcellactivation = Immune_GO_cellular %>% rename(Condition = study, Genes = genes) %>% filter(Process == "T cell activation") %>% filter(Condition != "NA")
```

Criando matriz para heatmap

```{r}
#Input
heatmap_data = Immune_GO_general_adaptive
file= "Immune_GO_general_adaptive"

#Excluir linhas com Target ou Source == NA
heatmap_data = filter(heatmap_data, Vaccine != "NA")

#Selecionar colunas de comparação e valores. Trocar value.var pela variável de interesse
names(heatmap_data)
matrix_data <- dcast(heatmap_data, `Genes` ~ `Condition`, value.var = "log2FoldChange", fun.aggregate = mean)

#Converter coluna 1 em rownames e excluir
rownames(matrix_data) <- matrix_data$`Genes`
matrix_data <- matrix_data %>% select(-`Genes`)

#Arredondar valores para facilitar visualização
matrix_data <- round(matrix_data, 1)
matrix_data <- replace(matrix_data, matrix_data == "NaN", 0)


#Definir nome para arquivo geral
write.csv(matrix_data, file=paste(file, "heatmap.csv",sep="_")) 

```

```{r}

#Definir matriz de dados
matrix_data = as.matrix(matrix_data)

cellmarker_studies

# Criar tabela de anotações para as vacinas (só rodar uma vez)
# rownames(ann_vaccines) = rownames(t(matrix_data)) # Obter condições e transpor para linhas
# ann_vaccines = select(ann_vaccines, -ann_vaccines)
#ann_vaccines = data.frame(ann_vaccines) # Converter em dataframe

# Agrupamento
# ann_vaccines$Vaccine = NA
# ann_vaccines$Dose= NA
# ann_vaccines$Day = NA
# #Anotar manualmente usando editData
# ann_vaccines = editData(ann_vaccines) 

#Cores

# # Defina os níveis do fator Vaccine para corresponderem às cores
# ann_vaccines$Vaccine <- factor(ann_vaccines$Vaccine)
# ann_vaccines$Dose = factor(ann_vaccines$Dose)
# ann_vaccines$Day <- factor(ann_vaccines$Day)

#Defina a ordem das anotações no gráfico
# ann_vaccines = select(ann_vaccines, Day, Dose, Vaccine)

# # Defina as cores correspondentes em annotation_colors
# ann_colors <- list(
#   Vaccine = c(
#   "BBIBP" = "#F46D43",
#   "BNT" = "#0096C7",
#   "ChAd" = "#99d98c",
#   "ChAd-BNT" = "#76c893",
#   "ZF2001" = "#ffd500"),
#   Dose = c(
#     "1" = "#ffea00",
#     "2" = "#ff7b00",
#     "3" = "#e01e37"
#   ),
#   Day = c(
#     "01" = "#a5ffd6",
#     "03" = "#99d98c",
#     "06" = "#00bbf9",
#     "07" = "#4cc9f0",
#     "14" = "#7b2cbf",
#     "28" = "#b5179e"))


# Defina os valores mínimos e máximos para o espectro de cores
valores_minimos <- -1  #  Substitua pelo valor mínimo desejado
valores_maximos <- 3   # Substitua pelo valor máximo desejado

# min(matrix_data)
# max(matrix_data)



# Crie o gráfico de calor com configurações personalizadas
p = pheatmap::pheatmap(
matrix_data,
annotation_colors = ann_colors,
annotation_col = ann_vaccines,
cutree_cols = 5,
cutree_rows = 7,
width = 100,  # Largura da figura (ajuste conforme necessário)
height = 1000,  # Altura da figura (ajuste conforme necessário)
fontsize_row = 2,  # Tamanho da fonte para rótulos de linhas
fontsize_col = 3,  # Tamanho da fonte para rótulos de colunas
cluster_rows = TRUE,  # Impede o agrupamento de linhas
cluster_cols = F,  # Impede o agrupamento de colunas
main = "Immune_GO_adaptive_cellular_tcellactivation",
angle_col = 90,  # Rotação dos rótulos de coluna (0 para horizontal)
cellheight = 2,  # Altura das células (ajuste conforme necessário)
legend = TRUE,
border = TRUE,
border_color = "black",
show_colnames = TRUE,
show_rownames = TRUE,
color = colorRampPalette(c("blue", "white", "red"))(100),  # Esquema de cores
breaks = c(seq(valores_minimos, -0.001, length.out = 50), seq(0.001, valores_maximos, length.out = 50)))


# Salve o gráfico como um arquivo PNG
png(file= fileimage, width=2000, height=2000, res=315, pointsize=10)
print(p)  # Substitua 'p' pelo objeto 'pheatmap'
dev.off()  # Finalize a gravação do arquivo


```

#### 5.3.5. Gráfico de barras

```{r}

#Definir conjunto de dados
gene_set = ImmuneGO_Annotated_Genes

#Definir nome para arquivos gerados
file <- "Immune_GO"
fileimage <- paste(file, "_barplot.png", sep = "") 


#Separar por vacina
# BBIBP = filter(gene_set, Vaccine == "BBIBP")
# ZF2001 = filter(gene_set, Vaccine == "ZF2001")
# BNT = filter(gene_set, Vaccine == "BNT")
# ChAd = filter(gene_set, Vaccine == "ChAd")

#Input
barras = gene_set #Trocar

#Criar gráfico de barras
genes_bars <- ggplot(barras, aes(x = Genes, y = log2FoldChange, fill = Condition)) +
  geom_col(position = "dodge") +
  geom_text(aes(label = round(log2FoldChange, 1)),
            position = position_dodge(width = 0), hjust = 0, vjust = 0, size = 1.5, angle = 0) +
  theme_minimal() +
  labs(title = "Innate immune response associated genes - BBIBP", size = 5, #Trocar
       x = "Gene",
       y = "L2FC") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, size = 6),
        legend.position = "none") + 
  facet_wrap(~Condition, ncol = 1)

genes_bars

ggsave(genes_bars, file=fileimage, width = 8, height = 16, units = "in") #Trocar
```

## 5.4. ssGSEA

### 5.4.1. Process data

```{r}
library(corto)
#Required files

CellMarker_ImmuneCells
ann_vaccines_samples = read_csv(here("Annotations and metadata", "ann_vaccines_samples.csv"))

#Filter
gse199750_counts_ready_2 = gse199750_counts_ready %>% as.data.frame() %>% rownames_to_column("genes") %>% pivot_longer(!genes, names_to = "sample", values_to = "counts")
gse189039_counts_ready_2 = gse189039_counts_ready %>% as.data.frame() %>% rownames_to_column("genes") %>% pivot_longer(!genes, names_to = "sample", values_to = "counts")
gse201530_counts_ready_2 = gse201530_counts_ready %>% as.data.frame() %>% rownames_to_column("genes") %>% pivot_longer(!genes, names_to = "sample", values_to = "counts")
gse201533_counts_ready_2 = gse201533_counts_ready %>% as.data.frame() %>% rownames_to_column("genes") %>% pivot_longer(!genes, names_to = "sample", values_to = "counts")
gse206023_counts_ready_2 = gse206023_counts_ready %>% as.data.frame() %>% rownames_to_column("genes") %>% pivot_longer(!genes, names_to = "sample", values_to = "counts")

all_counts_matrices = bind_rows(gse199750_counts_ready_2,
                                gse189039_counts_ready_2,
                                gse201530_counts_ready_2,
                                gse201533_counts_ready_2,
                                gse206023_counts_ready_2)


all_counts_matrices_wide = all_counts_matrices %>% 
  pivot_wider(names_from = "sample", values_from = "counts") %>% 
  column_to_rownames("genes")

write.csv(all_counts_matrices_wide, file = "all_counts_matrices_wide.csv")

```

### 5.4.2. Perform analysis

```{r}
############## Inputs
ssgsea_gene = all_counts_matrices_wide 
metadata = ann_vaccines_samples 
filename = "all_counts_matrices_"
go_dataset = "CellMarker ImmuneCells"
go_enrich = CellMarker_ImmuneCells
```

```{r}
#Extrair sample names. O ssGSEA perde o nome no resultado.
sample_names = ssgsea_gene %>% 
  as.data.frame() %>% 
  colnames() %>% 
  as.data.frame() %>% 
  rename(sample_names = '.')

#Tabela com genes de interesse (geral) convertida em listas 

genelist = go_enrich %>%
  select(process, genes) 

genelist = split(genelist$genes, genelist$process) 

############## Análise 

# Run ssGSEA

nesmat = ssgsea(ssgsea_gene, genelist)

#Padronizar resultado
nesmat.result = nesmat %>% 
  as.data.frame() %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column("sample") %>% 
  relocate(sample, .before = everything()) %>% 
  cbind(sample_names) %>% 
  select(-sample) %>% 
  column_to_rownames("sample_names") %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column("process") %>% 
  relocate(process, .before=everything()) %>% 
  pivot_longer(cols = -process, names_to = "sample", values_to = "nes") %>%
  mutate(pvalue = z2p(nes), #Converter NES em pvalue
         qvalue = p.adjust(pvalue, method = "BH"), #ajustar para qvalue 
         logq = - log10(qvalue)) %>%  #Calcular -log(q)
  inner_join(metadata, by= "sample") %>% #Unir com metadata
  inner_join(go_enrich, by = "process") %>% 
  select(!genes) %>% 
  distinct() 
#Salvar
write_csv(nesmat.result, file = here("Gene set enrichment analysis", paste0(filename, sep = "_", go_dataset, "_ssgsea_result.csv")))

nesmat.result
```

### 5.4.3. Visualization

```{r}
#Filter
ssgsea_interesting = nesmat.result %>% 
  filter(qvalue <= 0.10) %>% 
  distinct()

#Boxplot (NES)
NES_plot = ggplot(ssgsea_interesting) +
  aes(x = condition, y = nes, fill = vaccine) +
  geom_boxplot() +
  scale_fill_hue(direction = 1) +
  theme_minimal() +
  facet_grid(vars(process), vars(vaccine), scales = "free_x") +
 labs(title = paste0(filename, "_ImmuneGO ssGSEA"), 
      fill = "vaccine") +
 theme_bw() +
 theme(plot.title = element_text(size = 20L, face = "bold", hjust = 0.5),
       axis.text.x = element_text(angle = 45, vjust = 1, hjust=1, size = 7)) 

ggsave(NES_plot, file= paste0(filename, "_ssGSEA_NES_qvalue010.png"), width = 10, height = 20)

NES_plot
```

```{r}
ssgsea_interesting.unified = ssgsea_results_unified %>% 
  filter(!condition %in% c("BNT-MO (V1, D0)", "BNT-MO (V2, D0)")) %>% 
  mutate(day = as.factor(day),
         dose = as.factor(dose))

ssgsea_interesting.general.unified = ssgsea_interesting.unified %>% 
  filter(process %in% c("ADAPTIVE IMMUNE SYSTEM", 
                        "CELLULAR ADAPTIVE IMMUNE SYSTEM", 
                        "INNATE IMMUNE SYSTEM", 
                        "INFLAMMATION"),
         pvalue <= 0.10)

ssgsea_interesting.specific.unified = ssgsea_interesting.unified %>% 
  filter(!process %in% c("ADAPTIVE IMMUNE SYSTEM", 
                         "CELLULAR ADAPTIVE IMMUNE SYSTEM", 
                         "INNATE IMMUNE SYSTEM", 
                         "INFLAMMATION"),
         pvalue <= 0.10)


data = ssgsea_results_unified
filename.unified = "ssgsea_results_unified"
```

```{r}
#All conditions divided by dose, time on x-axis

# Filtrar os dados antes de passar para ggplot
filtered_data <- data %>%
  group_by(cell_name, Vaccine, condition) %>%
  summarise(n = n()) %>%
  filter(n < 5) %>%
  pull(condition)


NES_plot_unified_dayx <- ggplot(data) +
  aes(x = day, y = nes, fill = Vaccine) +
  geom_boxplot(outlier.alpha = 1) +
  geom_point(data = subset(data, day %in% filtered_data),
             aes(x = day, y = nes, color = Vaccine),
             position = position_jitter(width = 0.2), alpha = 1) +
  scale_fill_manual(values = vaxcolors) +
  scale_color_manual(values = c(BBIBP = "#ffd000", 
                                ZF2001 = "#b5179e",
                                BNT = "#56cfe1",
                                ChAd = "#80ffdb" ,
                                "ChAd-BNT" = "#72efdd")) +
  labs(
    title = "ssGSEA - All studies",
    subtitle = "General ImmuneGO, by dose",
    caption = "Vaccine",
    fill = "Vaccines"
  ) + 
  theme_minimal() +
  facet_grid(vars(Type), vars(dose), scales = "free_x") +
  theme(plot.title = element_text(size = 20L, face = "bold", hjust = 0.5),
        strip.text.y = element_text(angle = 0, vjust = 1, hjust = 1, size = 5),
        strip.text.x = element_text(size = 10))

#Salvar
ggsave(NES_plot_unified_dayx, file = paste0(filename.unified, "_ssGSEA_NES_qvalue010_2.png"), width = 10, height = 10)

print(NES_plot_unified_dayx)

```

```{r}
#All conditions divided by vaccine, time on x-axis

# Criar o gráfico
NES_plot_unified_dayx_vax <- ggplot(data) +
  aes(x = condition, y = nes, fill = dose) +
  geom_boxplot(outlier.alpha = 0.5) +
  geom_point(data = subset(data, condition %in% filtered_data),
             aes(x = condition, y = nes, color = "black"),
             position = position_jitter(width = 0.2), alpha = 1) +
  scale_fill_manual(values = dose_colors) +
  scale_color_manual(values = dose_colors) +
  labs(
    title = "ssGSEA - All studies",
    subtitle = "General ImmuneGO, by dose",
    caption = "Vaccine",
    fill = "Vaccines"
  ) +
  theme_minimal() +
  facet_grid(vars(process), vars(vaccine), scales = "free") +
  theme(
    plot.title = element_text(size = 20L, face = "bold", hjust = 0.5),
    axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1, size = 7),
    strip.text.y = element_text(angle = 0, vjust = 1, hjust = 1, size = 7)  # Rotação do texto da facet row
  )

#Salvar
ggsave(NES_plot_unified_dayx_vax, file = paste0(filename.unified, "_ssGSEA_NES_qvalue010_3.png"), width = 20, height = 20)

print(NES_plot_unified_dayx_vax)
```

### 5.4.4. Statistical analysis

```{r}
install.packages("ggstatsplot")
library(ggstatsplot)

ssgsea_interesting = ssgsea_interesting %>% 
  filter(process == "INNATE IMMUNE SYSTEM") %>% 
  distinct()

process = "INNATE IMMUNE SYSTEM"


ggbetweenstats(
  data  = ssgsea_interesting,
  x     = condition,
  y     = nes,
  title =  paste0("ImmuneGO ssGSEA ", filename, process)
)

#Agrupado
grouped_ggbetweenstats(
  data             = ssgsea_interesting,
  x                = condition,
  y                = nes,
  grouping.var     = process,
  ggsignif.args    = list(textsize = 4, tip_length = 0.01),
  p.adjust.method  = "bonferroni",
  palette          = "default_jama",
  package          = "ggsci",
  plotgrid.args    = list(nrow = 1),
  annotation.args  = list(title = paste0("ImmuneGO ssGSEA ", filename))
)

```

#### 5.4.3.2. ANOVA

```{r}
esquisser(test)

library(ggsignif)
library(rstatix)

df = test
df$Condition <- as.factor(df$Condition)

stat.test <- df %>%
  t_test(Condition ~ NES) %>%
  add_significance()


test %>%
  filter(Process %in% "INNATE IMMUNE RESPONSE") %>%
  ggboxplot(x = "Process", y = "NES",
          color = "Condition", palette = "jco") +
  facet_wrap(vars(Vaccine)) +
  stat_compare_means(method = "anova", label.y = 4) +      # Add global p-value
  stat_compare_means(label = "p.signif", method = "t.test",
                     ref.group = ".all.")    # Comparação múltipla usando Tukey HSD


# Box plots with p-values
bxp <- ggboxplot(df, x = "Process", y = "NES", fill = "#00AFBB")
stat.test <- stat.test %>% add_xy_position(x = "supp")
bxp + 
  stat_pvalue_manual(stat.test, label = "p") +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1)))

```

# 6. Principal components analysis

Tutorials [Data Camp - PCA Analysis in
R](https://www.datacamp.com/tutorial/pca-analysis-r){.uri}

[FactoExtra](https://rpkgs.datanovia.com/factoextra/index.html){.uri}
[STHDA: Principal Component Methods in R - Practical
guide](http://www.sthda.com/english/articles/31-principal-component-methods-in-r-practical-guide/114-mca-multiple-correspondence-analysis-in-r-essentials/){.uri}

[Statistics globe: PCA before K Means
Clustering](https://statisticsglobe.com/pca-before-k-means-clustering-r){.uri}\>

## 6.1. Packages and colors

```{r}
library(tidyverse)
library(ggrepel) #load before factoextra
library(factoextra)
library(caret)
library(stats)
library(ggfortify)
library(ggplot2)
library(corrplot)
library(cowplot)
library(ggcorrplot)
library(ComplexHeatmap)
library(circlize)
library(ggExtra)
library(FactoMineR)
library(here)

condition_col = c(
                                "BBIBP (V3, D07)" = "#00A087FF",
                                "BBIBP (V3, D14)" = "#00A087FF",
                                "BBIBP (V3, D28)" = "#00A087FF",
                                "BNT (V1, D6)" = "#4DBBD5FF",
                                "BNT (V2, D1)" = "#4DBBD5FF",
                                "BNT (V3, D1)" = "#4DBBD5FF",
                                "BNT-I (D1)"= "#DC0000FF",
                                "BNT-I (D10, mild)"= "#DC0000FF",
                                "BNT-I (D10, severe)"= "#DC0000FF",
                                "BNT-I (D2)"= "#DC0000FF",
                                "BNT-I (D26, mild)"= "#DC0000FF",
                                "BNT-I (D26, severe)"= "#DC0000FF",
                                "BNT-I (D3)"= "#DC0000FF",
                                "BNT-I (D4)"= "#DC0000FF",
                                "BNT-I (D51, severe)"= "#DC0000FF",
                                "BNT-I-BNT (D51, mild)"= "#F5BB00",
                                "BNT-I-BNT (D51, severe)"= "#F5BB00",
                                "BNT-MO (V1, D6)"= "#72ddf7",
                                "BNT-MO (V3, D1)"= "#72ddf7",
                                "ChAd (V1, D3)"= "#3C5488FF",
                                "ChAd (V1, D6)"= "#3C5488FF",
                                "ChAd (V1, D7)"= "#3C5488FF",
                                "ChAd (V2, D1)"= "#3C5488FF",
                                "ChAd (V2, D3)"= "#3C5488FF",
                                "ChAd (V2, D7)"= "#3C5488FF",
                                "ChAd-BNT (V2, D0)"= "#4DBBD5FF",
                                "ChAd-BNT (V2, D3)"= "#4DBBD5FF",
                                "ChAd-BNT (V2, D7)"= "#4DBBD5FF",
                                "ChAd-BNT (V3, D1)"= "#4DBBD5FF",
                                "I (D1)"= "#DC0000FF",
                                "I (D10, moderate)"= "#DC0000FF",
                                "I (D10, severe)"= "#DC0000FF",
                                "I (D26, moderate)"= "#DC0000FF",
                                "I (D26, severe)"= "#DC0000FF",
                                "I (D51, moderate)"= "#DC0000FF",
                                "I (D51, severe)"= "#DC0000FF",
                                "I-BNT-I (D2)"= "#DC0000FF",
                                "I-BNT-I (D5)"= "#DC0000FF",
                                "I-I (D0)"= "#DC0000FF",
                                "I-I (D1)"= "#DC0000FF",
                                "I-I (D2)"= "#DC0000FF",
                                "I-I (D3)"= "#DC0000FF",
                                "I-I (D5)"= "#DC0000FF",
                                "ZF2001 (V3, D07)"= "#8491B4FF",
                                "ZF2001 (V3, D14)"= "#8491B4FF",
                                "ZF2001 (V3, D28)" = "#8491B4FF"
                              )



col_immune = c("ADAPTIVE" = "#bc3908",
          "B CELL" = "#e76f51",
          "HUMORAL" = "#ee8959",
          "IMMUNOGLOBULIN-MEDIATED"= "#f4a261",
          "CELLULAR" = "#e9c46a",
          "T CELLS" = "#ECDDA3",
          "T HELPER CELLS" = "#EDE9BF",
          "COMPLEMENT" = "#734f5a",
          "ANTIVIRAL & IFN" = "#D2E1DF",
          "ARPP" = "#BCD2CF",
          "DENDRITIC CELL" = "#1E5750",
          "EOSINOPHIL" = "#183431",
          "INFLAMMATION" = "#247A70",
          "INNATE" = "#2a9d8f",
          "MAST CELL" = "#A5C2BF",
          "MONOCYTE" = "#7CB6AF",
          "NEUTROPHIL" = "#264653",
          "NK CELL" = "#125961",
          "MACROPHAGE" = "#288C62",
          "GENERAL" = "#babd8d",
          "ANTIMICROBIAL" = "#219ebc"
  )

ann_vaxsig_colors = list(
  Platform = c("VLP" = "#7E6148FF",
           "LA" =  "#E64B35FF",
           "CONJ" = "#F39B7FFF", 
           'IN'= "#00A087FF", #1st Gen  vaccines
           'VV' = "#3C5488FF", #3rd Gen vaccines
           'RNA' = "#4DBBD5FF",
           'SU' = "#8491B4FF",
           'I'= "#DC0000FF",
           "H" = "grey95"),
  week = c("0" = "#ECF8FA",
           "1" = "#caf0f8",
           "2" = "#6CD5EA",
           "3" = "#0087BF",
           "4" = "#015BA0", 
           "6" = "#03045e",
           "7" = "#03045e",
           "8" = "#03045e"),
  month = c("1" = "#caf0f8",
            "0" = "#6CD5EA",
            "2" = "#015BA0"),
  dose = c(
    "0" = "grey95",
    "1" = "#91D1C2A0",
    "2" = "#00A087FF",
    "3" = "#3C5488FF"),
  date = c(
    "H" = "#F39B7FFF",
    "D" = "#DF8F44FF",
    "M" = "#7E6148FF",
    "ANY" = "black"),
  day = c(
    "0" = "white",
    "1" = "#caf0f8",
    "2" = "#ade8f4",
    "3" = "#90e0ef",
    "4" = "#6CD5EA",
    "5" = "#48cae4",
    "6" = "#00b4d8",
    "7" = "#0096c7",
    "10" = "#0087BF",
    "11" = "#0087BF",
    "14" = "#0077b6",
    "26" = "#015BA0",
    "28" = "#023e8a",
    "51" = "#03045e"),
  "Pathogen" = c('MENINGOCOCCUS'	= "#374E55FF",
                              'PNEUMOCOCCUS'	= "#697B7F",
                              'TUBERCULOSIS'	= "#9BA7AA",
                              'F TULARENSIS'	= "#CDD3D5",
                              'LEISHMANIA'	= "#DF8F44FF",
                              'MALARIA' = "#EFC7A2",
                              'EBOV'	= "#142F36",
                              'HBV'	= "#275E6B",
                              'HBV, HAV'	= "#275E6B",
                              'HIV'	= "#3A8DA0",
                              'HPV'	= "#4DBBD5",
                              'INFLUENZA'	= "#A6DDEA",
                              'MEASLES'	= "#D3EEF5",
                              'MMR'	= "#a2d2ff",
                              'SMALLPOX'	= "#bde0fe",
                              'VEEV'	= "#b8c0ff",
                              'VZV'	= "#94d2bd",
                              'YF'	= "#118ab2"),
  
  "Microbe type" = c("VIRUS" = "#4DBBD5FF",
                   'BACTERIUM' = "#374E55FF",
                   'PARASITE' = "#DF8F44FF"))
```

## 6.2. Process data

**Required files**

```{r}
# Genes in long table format
all_degs_p_05_vac_infected_19_12_23 = read_csv(here("DGE analysis", "all_degs_p_05_vac_infected_19_12_23.csv"))

# Gene set annotation.Used only to merge and filter genes.
ImmuneGO_Annotated_Genes = readRDS(here("VaxGO gene sets", "ImmuneGO_Annotated_Genes_2024-03-26.rds"))

# Sample or condition annotation
ann_vaccines_19_1_24 <-
  read_csv(here("Annotations and metadata", "ann_vaccines_19-1-24.csv"))

data_annotation = ann_vaccines_19_1_24 %>%
  mutate(day = as.factor(day),
         dose = as.factor(dose)) 
```

**Filtering data**

```{r}
# Gene set data.
ImmuneGO_Annotated_Genes_all = ImmuneGO_Annotated_Genes
ImmuneGO_Annotated_Genes = ImmuneGO_Annotated_Genes_all %>%
  filter(!process %in% c("TCR REPERTOIRE", "BCR REPERTOIRE"))

ImmuneGO_Genes = all_degs_p_05_vac_infected_19_12_23 %>%
  inner_join(ImmuneGO_Annotated_Genes %>% 
               select(genes, process) %>% 
               distinct() %>% 
               select(genes),
               by = "genes") %>% 
  distinct()
```

**Input**

```{r}
#INPUT
data_genes = ImmuneGO_Genes 
filename = "ImmuneGO_Genes"
```

```{r}
#Convert long to wide table format.
#Dataframe with sample annotation
ann_vaccines_pca_matrix = data_genes %>% 
  mutate(log2fold_change = as.numeric(log2fold_change)) %>% 
  select(condition, genes, log2fold_change) %>% 
  pivot_wider(., names_from = "condition", 
                     values_from = "log2fold_change") %>% 
  replace(is.na(.), 0) %>%
  column_to_rownames("genes") %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column("condition") %>% 
  inner_join(data_annotation %>% 
               select(condition, disease_vac, vaccine, day, week, dose, infection, type), by = "condition") 

# Convert dataframe to matrix without annotation columns
ann_vaccines_pca_matrix_ready = ann_vaccines_pca_matrix %>% 
  column_to_rownames("condition") %>% 
  select(!disease_vac:type) %>% 
  as.matrix()

```

## 6.3. PCA: COVID-19 conditions only

```{r}
# Delete columns with near 0 variance
nearZeroVarCols <- nearZeroVar(ann_vaccines_pca_matrix_ready, saveMetrics = TRUE)
ann_vaccines_pca_matrix_ready <- ann_vaccines_pca_matrix_ready[, !nearZeroVarCols$nzv]
pca_res <- prcomp(ann_vaccines_pca_matrix_ready, scale. = T)

# Correlation plot ----
corr_matrix = cor(ann_vaccines_pca_matrix_ready %>% scale()) 
var <- get_pca_var(pca_res)

# corrplot = ggcorrplot(corr_matrix, hc.order = TRUE) + 
#   theme(axis.text.x = element_text(angle = 90, size = 5), 
#         axis.text.y = element_text(size = 5))
# #Salvar
# ggsave(corrplot, file = paste0(filename, "_corrplot.png"), 
#        width = 20, #Grande 20, pequeno 10
#        height= 20) #Grande 20, pequeno 10
# print(corrplot)

#PCA ----
data.pca = prcomp(corr_matrix) #PCA
summary(corr_matrix) #Retornar PCs

#Scree plot ----
scree_plot = fviz_eig(data.pca, 
                      addlabels = TRUE,
                      ylim = c(0, 70)) +
  geom_col(color = "#00AFBB", fill = "#00AFBB") +
  theme_classic()

#Save
ggsave(scree_plot, file = here(paste0(filename, "_screeplot.png"), "Figures"), width = 10, height = 3) 
print(scree_plot)

```

## 6.3.1. Loadings plot

```{r}
# Loadings plot ----
options(ggrepel.max.overlaps = Inf)

circle_contrib= fviz_pca_var(pca_res, col.var = "cos2",
                              gradient.cols = c("black", "#DC0000FF"),
                              select.var= list(cos2 = 20), 
                              repel = T, 
                              labelsize = 4,
                              col.circle = NA) +
  xlab("") + 
  ylab("") +
  theme_minimal() +
theme(panel.grid = element_blank(),  # Remover linhas de grade
        axis.text = element_blank(),   # Remover rótulos de texto dos eixos
        axis.ticks = element_blank()) 

#Save
ggsave(circle_contrib, file = here(paste0(filename, "_circle_contrib_wide.png"), "Figures"), width = 10, height = 5)
```

## 6.3.2. KNN

```{r}
#KNN classification -----
#Use the screeplot generated % of explained variance for each dimension
scree_plot #Dim 1 = 68.2%, Dim 2 = 15.9%


# PC1-PC2 ------
#Determine the number of clusters
pca_scores <- data.frame(pca_res$x[, 1:2])
fviz_nbclust(pca_scores,  
                     FUNcluster = kmeans,
                     method = "wss")

pcas = "PC1-PC2"

set.seed(666) # Set seed for randomization
cluster_model <- kmeans(pca_res$x[, 1:2], centers = 3)  #Set the number of clusters

ann_vaccines_pca_matrix$cluster <- as.factor(cluster_model$cluster)
```

## 6.3.3. Plot

```{r}
# Condition with density plot ------
pca_plot_knn_cluster = autoplot(pca_res, 
        data = ann_vaccines_pca_matrix, 
        colour = 'cluster')  +
  stat_ellipse(aes(color = cluster, fill = cluster), 
               geom = "polygon", 
               alpha = 0.2, 
               linetype = 1,
               size = 0.3,
               type = "t") +
  labs(title=paste0(filename, "_bycondition")) + 
  xlab("PC1 (68.2%)") +
  ylab("PC2 (15.9%)") +
  
  scale_color_manual(values = c("1" = "#DC0000FF", 
                                "2" = "#4DBBD5FF", 
                                "3" = "#8491B4FF", 
                                condition_col)) +  # Ellipse line colors
  scale_fill_manual(values = c("1" = "#DC0000FF", 
                                "2" = "#4DBBD5FF", 
                                "3" = "#8491B4FF")) +  # Cluster fill colors
  theme_minimal() +
  guides(col="none") +
  theme(panel.grid = element_blank(),
        text = element_text(color = "black")) +
  geom_text_repel(aes(label = condition,
                      segment.colour="gray70"),
                  box.padding = 0.3,
                  size = 3) 


ggsave(pca_plot_knn_cluster, 
       file = here(paste0(filename, pcas, "_KNN_Clustered_labels.png"), "Figures"), width = 10, height = 5)

pca_plot_knn_cluster_marginal = ggMarginal(
  pca_plot_knn_cluster,
  type = "histogram",
  groupFill = T,
  groupColour = T,
  xparams = list(binwidth = 0.1, size = 0.1),
  yparams = list(binwidth = 0.1, size = 0.1)
)

ggsave(
  pca_plot_knn_cluster_marginal,
  file = here(paste0(filename, pcas, "_KNN_Clustered_labels_Marginal.png"), "Figures"),
  width = 8,
  height = 5
)


```

## 6.3.4. Heatmap

### PC1 and PC2 genes

```{r}
####### INPUT
PC1 = var$cos2 %>%
  as.data.frame() %>%
  arrange(desc(Dim.1)) %>%
  filter(Dim.1 >= 0.5) %>% 
  #slice_head(n=20) %>% 
  rownames_to_column("genes") %>% 
  distinct() %>% 
  select(genes, cos2 = Dim.1) %>% 
    mutate(dimension = "PC1")

PC2 = var$cos2 %>%
  as.data.frame() %>%
  filter(Dim.2 >= 0.5) %>% 
  #slice_head(n=20) %>% 
  rownames_to_column("genes") %>% 
  distinct() %>% 
  select(genes, cos2 = Dim.2) %>% 
  mutate(dimension = "PC2")

PC1_2 = bind_rows(PC1, PC2) 
PC1_2 %>% 
  select(genes) %>% distinct()

PC1_2 %>% write.csv(file = "PC_1_2_cos2.csv", row.names = F)


# Gene annotation
ImmuneGO_Annotated_Genes
OtherGOs_Annotated_Genes

# INPUT

ann_genes = ImmuneGO_Annotated_Genes
pc = "PC2"
data_2 =  PC2 %>% 
  slice_max(order_by = cos2, n = 20) %>% 
  inner_join(data_genes, by = "genes")
filename_2 = paste0(filename, "_PCA_",pc, "_COVID")

######## Matrix
matrix = data_2 %>% 
  select(genes, condition, log2fold_change) %>%
  pivot_wider(names_from = "condition", 
              values_from = "log2fold_change", 
              values_fn = mean) %>% 
  replace(is.na(.), 0) %>%
  arrange(genes) %>% 
  column_to_rownames(var="genes") %>% 
  as.matrix()

######## Annotations
ann_vaccines_covid = data_2 %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  select(condition, disease_vac, type, day) %>% 
  distinct()

#Columns
ann_cols_heatmap = matrix %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  inner_join(ann_vaccines_covid, by = "condition") %>% 
  distinct() %>% 
  arrange(condition) %>% 
  column_to_rownames("condition")

matrix_data = matrix %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  inner_join(ann_cols_heatmap %>% rownames_to_column("condition"), by = "condition") %>% 
  select(!c(disease_vac:day)) %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>%
  as.matrix()

#Rows Immune -----
ann_rows_immune = matrix_data %>%
  as.data.frame() %>%
  rownames_to_column("genes") %>%
  inner_join(ann_genes, by = "genes") %>%
  filter(go_term == "Manual") %>%
  select(genes, process) %>%
  distinct() %>%
  group_by(genes) %>%
  arrange(genes, factor(process, levels = c("INNATE IMMUNE SYSTEM", "ADAPTIVE IMMUNE SYSTEM", "OTHER", "ANTIMICROBIAL IMMUNE RESPONSE")), .by_group = TRUE) %>%
  mutate(i_process = row_number()) %>%
  pivot_wider(., names_from = "i_process", values_from = "process") %>%
  column_to_rownames("genes") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("index") %>%
  arrange(desc(as.numeric(index))) %>%
  column_to_rownames("index") %>%
  t() %>%
  as.data.frame() %>%
  replace(is.na(.), ".")

#Verificar dimensões do dataset
dim(matrix_data)
dim(ann_rows_immune)
dim(ann_cols_heatmap)
```

```{r}
#Heatmap annotation -----
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "right")

row_ha = HeatmapAnnotation(df = ann_rows_immune, 
                            which = "row",
                       col = col_process_immune, #col_process_immune / col_process_notimmune 
                       show_annotation_name = F,
                       show_legend = F)

# Values colors -----
col_fun <- colorRamp2(c(-2, 0, 2), c("#9bc1bc", "white", "#ed6a5a"))


# Parameters -----
file = filename_2
mat = matrix_data
width_image = 30
height_image = 30
width = unit(15, "cm")
height = unit(10, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)
rect_gp = gpar(col = "gray80", lwd = 0.5)

fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap.png")

png(file = fileimageheatmap_grouped, 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        col = col_fun, #color
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(2, "cm"),
                        column_split = 3,
                        row_split = 2,
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height 
)

draw(heatmap_plot, 
     annotation_legend_side = "left", 
     heatmap_legend_side = "left")
dev.off()

```

### Statistics

```{r}
stats_1 = covid_other_immunego_genes %>% 
  inner_join(ann_vaccines_covid_other, by = "condition") %>% 
  group_by(genes, covid_other) %>% 
  summarise(n_conditions = n()) %>% 
  filter(all(c("COVID", "OTHER") %in% covid_other)) %>% 
  arrange(desc(n_conditions), genes) %>% 
  ungroup()


ggplot_stats1 = ggplot(stats_1) +
  aes(x = reorder(genes, n_conditions), fill = covid_other, weight = n_conditions) +
  geom_bar() +
  theme_minimal() +
  scale_fill_manual(
    values = c(COVID = "#56cfe1",
    OTHER = "#343a40")
  ) +
  theme_minimal() +
  labs(fill = "Covid or non-covid") +
  xlab("Genes") + 
  ylab("Number of shared conditions") +
  theme(legend.position = "right",
        axis.text.x = element_text(angle = 45, vjust = 1.5, hjust=1, size = 8),
        axis.text.y = element_text(size = 7, angle = 0),
        panel.grid.minor.y = element_blank(),
        legend.text = element_text(size=12),
        legend.title = element_text(size = 12, face = "bold")) +
  # geom_text(aes(label = n_conditions, y = n_conditions), 
  #           position = position_dodge(width = 0.9), 
  #           vjust = 0.2, 
  #           hjust = -0.2, 
  #           size = 1,
  #           angle = 90) +
  ggtitle("Genes shared") +
  coord_flip()

ggsave(ggplot_stats1, file = "Covid_Vax_genes_shared_column.png", width = 10, height = 20)
```


```{r}
stats_2 = matrix_data %>%
  as.data.frame() %>%
  rownames_to_column("genes") %>%
  inner_join(ImmuneGO_Annotated_Genes, by = "genes") %>%
  filter(go_term == "Manual",
         !(process %in% c("ADAPTIVE IMMUNE SYSTEM",
                          "INNATE IMMUNE SYSTEM"))) %>%
  select(genes, process, immune_system) %>%
  distinct() %>%
  group_by(genes, immune_system) %>% 
  summarise(n_processes = n()) %>% 
  arrange(desc(n_processes)) %>% 
  ungroup() %>% 
  filter(n_processes >1)
  
ggplot_stats2 = ggplot(stats_2) +
  aes(x = reorder(genes, n_processes), fill = immune_system, weight = n_processes) +
  geom_bar() +
  theme_minimal() +
  scale_fill_manual(
    values = c(Adaptive = "#6f2dbd",
    Innate = "#06d6a0",
    Complement = "#ffadc7")
  ) +
  theme_minimal() +
  labs(fill = "Immune system") +
  xlab("Genes") + 
  ylab("Number of shared conditions") +
  theme(legend.position = "right",
        axis.text.x = element_text(angle = 45, vjust = 1.5, hjust=1, size = 8),
        axis.text.y = element_text(size = 7, angle = 0),
        panel.grid.minor.y = element_blank(),
        strip.text.x.top = element_text(size = 10, face = "bold"),
        legend.text = element_text(size=12),
        legend.title = element_text(size = 12, face = "bold")) +
  # geom_text(aes(label = n_conditions, y = n_conditions), 
  #           position = position_dodge(width = 0.9), 
  #           vjust = 0.2, 
  #           hjust = -0.2, 
  #           size = 1,
  #           angle = 90) +
  ggtitle("Genes shared") +
  coord_flip() +
  facet_wrap(~immune_system, scales = "free_y")

ggsave(ggplot_stats2, file = "ImmuneGO_genes_shared_major_process.png", width = 6, height = 20)

```


### 6.3.4.2 Gene expression by gender

#### PC1 genes

```{r}
#### PC1 -------
PC1 = var$cos2 %>%
  as.data.frame() %>%
  arrange(desc(Dim.1)) %>%
  slice_head(n = 5) %>%
  rownames_to_column("genes") %>% 
  distinct() %>% 
  select(genes, Dim.1)

PC1_genes_counts = normalized_counts %>% 
  rownames_to_column("genes") %>% 
  inner_join(PC1, by = "genes") %>% 
  pivot_longer(cols = -c("genes", "Dim.1"), names_to = "sample", values_to = "counts") %>% 
  mutate(counts = round(counts, 0)) %>% 
  select(genes, sample, counts, Dim.1) %>% 
  inner_join(ann_vaccines_samples, by = "sample") %>% 
  filter(gse_id != "GSE206023") %>% 
  group_by(condition) %>%
  filter(any(sex == 'Male') & any(sex == 'Female'),
         condition != "H (D0)") %>% 
  ungroup() %>% 
  mutate(age_inter = cut(
    age,
    breaks = c(10, 20, 30, 40, 50, 60, 70, 80, 90, 100),
    labels = c(
      "10-20",
      "21-30",
      "31-40",
      "41-50",
      "51-60",
      "61-70",
      "71-80",
      "81-90",
      "91-100"
    )
  ))
  
write_csv(PC1_genes_counts, file = here("PCA", "PC1_genes_samples_counts.csv"))

PC1_genes_counts
```

Sex

```{r fig.height=20, fig.width=20}
#Sex ----
pc1_5genes  = ggplot(PC1_genes_counts) +
  aes(x = condition, y = counts, fill = sex, colour = sex) +
  geom_boxplot() +
  scale_fill_hue(h = c(180, 270)) +
  scale_color_hue(h = c(180, 270)) +
  scale_y_continuous(trans = "log2") +
  theme_light() +
  labs(fill = "Sex") +
  theme(legend.position = "right") +
  facet_wrap(vars(genes), scales = "free", ncol =1) +
  theme(axis.text.x = element_text(size = 10, color = "black", angle = 45, vjust = 1, hjust = 1),
        axis.text.y = element_text(size = 10, color = "black"),
        strip.background = element_rect(fill = "black"),
        strip.text = element_text(color = "white", face = "bold", size = 15))
```

Age

```{r fig.height=20, fig.width=20}
#Age ----

pc1_5genes_age  = ggplot(PC1_genes_counts) +
  aes(x = condition, y = counts, fill = age_inter) +
  geom_boxplot() +
  scale_fill_hue(h = c(180, 360)) +
  scale_color_hue(h = c(180, 360)) +
  scale_y_continuous(trans = "log2") +
  theme_light() +
  labs(fill = "Age") +
  theme(legend.position = "right") +
  theme(axis.text.x = element_text(size = 10, color = "black", angle = 45, vjust = 1, hjust = 1),
        axis.text.y = element_text(size = 10, color = "black"),
        strip.background = element_rect(fill = "black"),
        strip.text = element_text(color = "white", face = "bold", size = 15)) +
  facet_wrap(vars(genes), scales = "free", ncol = 1)
```

Unite

```{r fig.height=20, fig.width=20}
patch_pc1_genes = pc1_5genes_age + pc1_5genes + plot_layout(guides = "collect")

ggsave(patch_pc1_genes, file = here("Figures","pc1_5genes_age_sex.png"), width = 20, height = 20)

patch_pc1_genes

```

#### PC2 genes

```{r fig.height=20, fig.width=20}
#### PC2 -------
PC2 = var$cos2 %>%
  as.data.frame() %>%
  arrange(desc(Dim.2)) %>%
  slice_head(n = 5) %>%
  rownames_to_column("genes") %>% 
  distinct() %>% 
  select(genes, Dim.1)

PC2_genes_counts = normalized_counts %>% 
  rownames_to_column("genes") %>% 
  inner_join(PC2, by = "genes") %>% 
  pivot_longer(cols = -c("genes", "Dim.1"), names_to = "sample", values_to = "counts") %>% 
  mutate(counts = round(counts, 0)) %>% 
  select(genes, sample, counts, Dim.1) %>% 
  inner_join(ann_vaccines_samples, by = "sample") %>% 
  filter(gse_id != "GSE206023") %>% 
  group_by(condition) %>%
  filter(any(sex == 'Male') & any(sex == 'Female'),
         condition != "H (D0)") %>% 
  ungroup() %>% 
  mutate(age_inter = cut(
    age,
    breaks = c(10, 20, 30, 40, 50, 60, 70, 80, 90, 100), 
    labels = c(
      "10-20",
      "21-30",
      "31-40",
      "41-50",
      "51-60",
      "61-70",
      "71-80",
      "81-90",
      "91-100"
    )
  ))

write_csv(PC2_genes_counts, file = here("PCA","PC2_genes_samples_counts.csv"))
```

Sex

```{r fig.height=20, fig.width=20}
#Sex ----
pc2_5genes  = ggplot(PC2_genes_counts) +
  aes(x = condition, y = counts, fill = sex, colour = sex) +
  geom_boxplot() +
  scale_fill_hue(h = c(180, 270)) +
  scale_color_hue(h = c(180, 270)) +
  scale_y_continuous(trans = "log2") +
  theme_light() +
  labs(fill = "Sex") +
  theme(legend.position = "right") +
  facet_wrap(vars(genes), scales = "free", ncol =1) +
  theme(axis.text.x = element_text(size = 10, color = "black", angle = 45, vjust = 1, hjust = 1),
        axis.text.y = element_text(size = 10, color = "black"),
        strip.background = element_rect(fill = "black"),
        strip.text = element_text(color = "white", face = "bold", size = 15))
```

Age

```{r fig.height=20, fig.width=20}
#Age ----

pc2_5genes_age  = ggplot(PC2_genes_counts) +
  aes(x = condition, y = counts, fill = age_inter, colour = age_inter) +
  geom_boxplot() +
  scale_fill_hue(h = c(180, 360)) +
  scale_color_hue(h = c(180, 360)) +
  scale_y_continuous(trans = "log2") +
  theme_light() +
  theme(legend.position = "right") +
  labs(fill = "Age") +
  theme(axis.text.x = element_text(size = 10, color = "black", angle = 45, vjust = 1, hjust = 1),
        axis.text.y = element_text(size = 10, color = "black"),
        strip.background = element_rect(fill = "black"),
        strip.text = element_text(color = "white", face = "bold", size = 15)) +
  facet_wrap(vars(genes), scales = "free", ncol = 1)
```

Unite

```{r fig.height=20, fig.width=20}
patch_pc2_genes = pc2_5genes_age + pc2_5genes + plot_layout(guides = "collect")

ggsave(patch_pc2_genes, file = here("Figures","pc2_5genes_age_sex.png"), width = 20, height = 20)

patch_pc2_genes
```




## 6.4. PCA: COVID-19 and others, only up and down regulated genes

```{r}

# Gene set annotation.Used only to merge and filter genes.
ImmuneGO_Annotated_Genes
ImmuneGO_Genes = degs_covid_vax %>% 
  select(-direction)


# Sample or condition annotation
data_annotation = VAXSigDB_Annotated %>% 
  clean_names() %>% 
  select(condition = vaccine_condition, type = platform, microbe_type, target_pathogen_disease, time) %>% 
  mutate(time = if_else(time == "ANY", "0", time),
         time = if_else(time == "1-3", "2", time),
         time = as.numeric(time)) %>% 
  bind_rows(ann_vaccines_covid_other %>% rename(time = day_total)) %>% 
  select(condition, type, target_pathogen_disease, time)

VAXSigDB_Annotated_days = VAXSigDB_Annotated %>% 
  select(VACCINE_CONDITION, `DATE-TIME`:TIME) %>% 
  mutate(TIME = case_when(TIME == "ANY" ~ "0",
                          TIME == "1-3" ~ "2", 
                          TRUE ~ TIME),
         TIME = as.numeric(TIME)) %>% 
 mutate(day = case_when(DATE == "H" ~ TIME/24,
                        DATE == "M" ~ TIME * 30,
                        DATE == "W" ~ TIME * 7,
                        TRUE ~ TIME)) %>% 
  select(condition = VACCINE_CONDITION, day) %>% 
  distinct()

sars = data_annotation %>% 
  filter(target_pathogen_disease == "SARS") %>% 
  mutate(target_pathogen_disease = "COVID-19") %>% 
  rename(day = time)

data_annotation = data_annotation %>% 
  left_join(VAXSigDB_Annotated_days, by = "condition") %>% 
  filter(day < 100,
         day > 0) %>% 
  distinct() %>% 
  select(-time) %>% 
  bind_rows(sars) %>% 
  filter(type != "I") %>% 
  mutate(target_pathogen_disease = if_else(target_pathogen_disease == "COVID-19", "COVID-19", "OTHER"))

```

```{r}
#INPUT
data_genes = ImmuneGO_Genes 
filename = "ImmuneGO_Genes_UP_DOWN_COVID_OTHER"
```

```{r Create dataframe with annotations and matrix for PCA}
#Convert long to wide table format.
#Dataframe with sample annotation
ann_vaccines_pca_matrix = data_genes %>% 
  mutate(regulation_numeric = as.numeric(regulation_numeric)) %>% 
  select(genes, condition, regulation_numeric) %>% 
  distinct() %>%
  pivot_wider(names_from = "condition", 
              values_from = "regulation_numeric",
              values_fn = mean) %>% 
  replace(is.na(.), 0) %>%
  column_to_rownames("genes") %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column("condition") %>% 
  inner_join(data_annotation, by = "condition") %>% 
  distinct() 

# Convert dataframe to matrix without annotation columns
ann_vaccines_pca_matrix_ready = ann_vaccines_pca_matrix %>% 
  column_to_rownames("condition") %>% 
  select(!c(type, target_pathogen_disease, day)) %>% 
  as.matrix()

```

```{r PCA}
# Delete columns with near 0 variance
nearZeroVarCols <- nearZeroVar(ann_vaccines_pca_matrix_ready, saveMetrics = TRUE)
ann_vaccines_pca_matrix_ready <- ann_vaccines_pca_matrix_ready[, !nearZeroVarCols$nzv]
pca_res <- prcomp(ann_vaccines_pca_matrix_ready)

# Correlation plot ----
corr_matrix = cor(ann_vaccines_pca_matrix_ready) 
var <- get_pca_var(pca_res)

#PCA ----
#data.pca = prcomp(corr_matrix) #PCA
summary(corr_matrix) #Retornar PCs

#Scree plot ----
scree_plot = fviz_eig(data.pca, 
                      addlabels = TRUE,
                      ylim = c(0, 80)) +
  geom_col(color = "#00AFBB", fill = "#00AFBB") +
  theme_classic()

#Save
ggsave(scree_plot, file = paste0(filename, "_screeplot.png"), width = 10, height = 3) 
print(scree_plot)

```

```{r Loadings plot}
# Loadings plot ----
options(ggrepel.max.overlaps = Inf)

# Filter top vars for each PC
cos2_df <- get_pca_var(pca_res)$cos2
cos2_df <- as.data.frame(cos2_df)
cos2_df$variable <- rownames(cos2_df)
top_n_cos2 <- function(df, n = 10) {
  df %>%
    arrange(desc(cos2)) %>%
    slice_head(n = n)
}
top_cos2_Dim1 <- top_n_cos2(cos2_df %>% select(variable, cos2 = Dim.1))
top_cos2_Dim2 <- top_n_cos2(cos2_df %>% select(variable, cos2 = Dim.2))
top_cos2_combined <- bind_rows(top_cos2_Dim1, top_cos2_Dim2) %>%
  distinct(variable, .keep_all = TRUE)



circle_contrib= fviz_pca_var(pca_res, col.var = "cos2",
                              gradient.cols = c("black", "#DC0000FF"),
                              select.var = list(name = top_cos2_combined$variable),
                              repel = T, 
                              labelsize = 4,
                              col.circle = NA) +
  xlab("") + 
  ylab("") +
  theme_minimal() +
theme(panel.grid = element_blank(),  # Remover linhas de grade
        axis.text = element_blank(),   # Remover rótulos de texto dos eixos
        axis.ticks = element_blank()) 

circle_contrib


#Save
ggsave(circle_contrib, file = paste0(filename, "_circle_contrib_wide.png"), width = 10, height = 5)

```

```{r KNN classification}
#KNN classification -----
#Use the screeplot generated % of explained variance for each dimension
scree_plot #Dim 1 = 58.8%, Dim 2 = 10.5%


# PC1-PC2 ------
#Determine the number of clusters
pca_scores <- data.frame(pca_res$x[, 1:2])
fviz_nbclust(pca_scores,  
                     FUNcluster = kmeans,
                     method = "wss")

pcas = "PC1-PC2"

set.seed(666) # Set seed for randomization
cluster_model <- kmeans(pca_res$x[, 1:2], centers = 2)  #Set the number of clusters

ann_vaccines_pca_matrix$cluster <- as.factor(cluster_model$cluster)
```

```{r}
# Condition with density plot ------
pca_plot_knn_cluster = autoplot(pca_res, 
        data = ann_vaccines_pca_matrix, 
        size = 2,
        shape = 'type',
        color = "day")  +
  stat_ellipse(aes(fill = cluster), 
               geom = "path", 
               alpha = 0.1, 
               linetype = 1,
               size = 1,
               type = "t") +
  labs(title=paste0(filename, "_bycondition")) + 
  xlab("PC1 (58.8%)") +
  ylab("PC2 (10.5%)") +
  scale_color_stepsn(colors = c("#DC0000FF", "#F39B7FFF", "#4DBBD5FF"),
                     limits = c(0, 14), 
                     na.value = "#4DBBD5FF") +
  scale_shape_manual(values = c("VLP" = 1,
           "LA" =  5,
           "CONJ" = 3,
           'IN'= 4, #1st Gen  vaccines
           'VV' = 2, #3rd Gen vaccines
           'RNA' = 6,
           'SU' = 7,
           'I'= 8)) +
  theme_minimal() +
  theme(panel.grid = element_blank(),
        text = element_text(color = "black")) 

pca_plot_knn_cluster 

ggsave(pca_plot_knn_cluster, file = paste0(filename, pcas, "_KNN_Clustered_labels.png"), width = 10, height = 5)
```

```{r}
pca_plot_knn_cluster_interactive = autoplot(pca_res, 
        data = ann_vaccines_pca_matrix, 
        shape = 'type',
        color = "day")  +
  labs(title=paste0(filename, "_bycondition")) + 
  xlab("PC1 (58.8%)") +
  ylab("PC2 (10.5%)") +
  scale_color_gradientn(colors = c("darkblue", "purple"), limits = c(0, 14), na.value = "purple") +
  scale_shape_manual(values = c("VLP" = 1,
                                 "LA" =  2,
                                 "CONJ" = 3, 
                                 'IN'= 4, #1st Gen  vaccines
                                 'VV' = 5, #3rd Gen vaccines
                                 'RNA' = 6,
                                 'SU' = 7)) +
  theme_minimal() +
  theme(panel.grid.minor = element_blank(),
        text = element_text(color = "black")) 

pca_plot_knn_cluster_interactive %>% ggplotly()
```

## Heatmap

### Genes annotated

```{r}
####### INPUT
PC1 = var$cos2 %>%
  as.data.frame() %>%
  arrange(desc(Dim.1)) %>%
  #slice_head(n=20) %>% 
  rownames_to_column("genes") %>% 
  distinct() %>% 
  select(genes, cos2 = Dim.1) %>% 
    mutate(dimension = "PC1")

PC2 = var$cos2 %>%
  as.data.frame() %>%
  filter(Dim.2 >= 0.5) %>% 
  #slice_head(n=20) %>% 
  rownames_to_column("genes") %>% 
  distinct() %>% 
  select(genes, cos2 = Dim.2) %>% 
  mutate(dimension = "PC2")

PC1_2 = bind_rows(PC1, PC2) 
PC1_2 %>% 
  select(genes) %>% distinct()

PC1_2 %>% write.csv(file = "PC_1_2_cos2.csv", row.names = F)


# Gene annotation
ImmuneGO_Annotated_Genes
OtherGOs_Annotated_Genes

# INPUT

ann_genes = ImmuneGO_Annotated_Genes
pc = "PC2"
data_2 =  PC2 %>% 
  slice_max(order_by = cos2, n = 20) %>% 
  inner_join(data_genes, by = "genes")
filename_2 = paste0(filename, "_PCA_",pc, "_COVID")

######## Matrix
matrix = data_2 %>% 
  select(genes, condition, log2fold_change) %>%
  pivot_wider(names_from = "condition", 
              values_from = "log2fold_change", 
              values_fn = mean) %>% 
  replace(is.na(.), 0) %>%
  arrange(genes) %>% 
  column_to_rownames(var="genes") %>% 
  as.matrix()

######## Annotations
ann_vaccines_covid = data_2 %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  select(condition, disease_vac, type, day) %>% 
  distinct()

#Columns
ann_cols_heatmap = matrix %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  inner_join(ann_vaccines_covid, by = "condition") %>% 
  distinct() %>% 
  arrange(condition) %>% 
  column_to_rownames("condition")

matrix_data = matrix %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  inner_join(ann_cols_heatmap %>% rownames_to_column("condition"), by = "condition") %>% 
  select(!c(disease_vac:day)) %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>%
  as.matrix()

#Rows Immune -----
ann_rows_immune = matrix_data %>%
  as.data.frame() %>%
  rownames_to_column("genes") %>%
  inner_join(ann_genes, by = "genes") %>%
  filter(go_term == "Manual") %>%
  select(genes, process) %>%
  distinct() %>%
  group_by(genes) %>%
  arrange(genes, factor(process, levels = c("INNATE IMMUNE SYSTEM", "ADAPTIVE IMMUNE SYSTEM", "OTHER", "ANTIMICROBIAL IMMUNE RESPONSE")), .by_group = TRUE) %>%
  mutate(i_process = row_number()) %>%
  pivot_wider(., names_from = "i_process", values_from = "process") %>%
  column_to_rownames("genes") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("index") %>%
  arrange(desc(as.numeric(index))) %>%
  column_to_rownames("index") %>%
  t() %>%
  as.data.frame() %>%
  replace(is.na(.), ".")

# Rows Not immune -----
# ann_rows_notimmune = matrix_data %>%
#   as.data.frame() %>%
#   rownames_to_column("genes") %>%
#   left_join(ann_genes %>% 
#                select(process = term, genes = symbol, annotation) %>% 
#                filter(annotation == "Major process", 
#                       process != "Immune system"), by = "genes") %>% 
#   select(genes, process) %>%
#   distinct() %>%
#   group_by(genes) %>%
#   #arrange(genes, factor(process, levels = c("INNATE IMMUNE SYSTEM", "ADAPTIVE IMMUNE SYSTEM", "OTHER", "ANTIMICROBIAL IMMUNE RESPONSE")), .by_group = TRUE) %>% 
#   mutate(i_process = row_number()) %>% 
#   pivot_wider(., names_from = "i_process", values_from = "process") %>% 
#   column_to_rownames("genes") %>% 
#   t() %>% 
#   as.data.frame() %>% 
#   rownames_to_column("index") %>% 
#   arrange(desc(index)) %>% 
#   column_to_rownames("index") %>% 
#   t() %>% 
#   as.data.frame() %>% 
#   replace(is.na(.), ".")

#Verificar dimensões do dataset
dim(matrix_data)
dim(ann_rows_immune)
dim(ann_cols_heatmap)
```

### Plot

```{r}
#Heatmap annotation -----
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "right")

row_ha = HeatmapAnnotation(df = ann_rows_immune, 
                            which = "row",
                       col = col_process_immune, #col_process_immune / col_process_notimmune 
                       show_annotation_name = F,
                       show_legend = F)

# Values colors -----
col_fun <- colorRamp2(c(-2, 0, 2), c("#9bc1bc", "white", "#ed6a5a"))


# Parameters -----
file = filename_2
mat = matrix_data
width_image = 30
height_image = 30
width = unit(15, "cm")
height = unit(10, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)
rect_gp = gpar(col = "gray80", lwd = 0.5)

fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap.png")

png(file = fileimageheatmap_grouped, 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        col = col_fun, #color
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(2, "cm"),
                        column_split = 3,
                        row_split = 2,
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height 
)

draw(heatmap_plot, 
     annotation_legend_side = "left", 
     heatmap_legend_side = "left")
dev.off()

```

### STATS

```{r}
stats_1 = covid_other_immunego_genes %>% 
  inner_join(ann_vaccines_covid_other, by = "condition") %>% 
  group_by(genes, covid_other) %>% 
  summarise(n_conditions = n()) %>% 
  filter(all(c("COVID", "OTHER") %in% covid_other)) %>% 
  arrange(desc(n_conditions), genes) %>% 
  ungroup()


ggplot_stats1 = ggplot(stats_1) +
  aes(x = reorder(genes, n_conditions), fill = covid_other, weight = n_conditions) +
  geom_bar() +
  theme_minimal() +
  scale_fill_manual(
    values = c(COVID = "#56cfe1",
    OTHER = "#343a40")
  ) +
  theme_minimal() +
  labs(fill = "Covid or non-covid") +
  xlab("Genes") + 
  ylab("Number of shared conditions") +
  theme(legend.position = "right",
        axis.text.x = element_text(angle = 45, vjust = 1.5, hjust=1, size = 8),
        axis.text.y = element_text(size = 7, angle = 0),
        panel.grid.minor.y = element_blank(),
        legend.text = element_text(size=12),
        legend.title = element_text(size = 12, face = "bold")) +
  # geom_text(aes(label = n_conditions, y = n_conditions), 
  #           position = position_dodge(width = 0.9), 
  #           vjust = 0.2, 
  #           hjust = -0.2, 
  #           size = 1,
  #           angle = 90) +
  ggtitle("Genes shared") +
  coord_flip()

ggsave(ggplot_stats1, file = "Covid_Vax_genes_shared_column.png", width = 10, height = 20)




stats_2 = matrix_data %>%
  as.data.frame() %>%
  rownames_to_column("genes") %>%
  inner_join(ImmuneGO_Annotated_Genes, by = "genes") %>%
  filter(go_term == "Manual",
         !(process %in% c("ADAPTIVE IMMUNE SYSTEM",
                          "INNATE IMMUNE SYSTEM"))) %>%
  select(genes, process, immune_system) %>%
  distinct() %>%
  group_by(genes, immune_system) %>% 
  summarise(n_processes = n()) %>% 
  arrange(desc(n_processes)) %>% 
  ungroup() %>% 
  filter(n_processes >1)
  
ggplot_stats2 = ggplot(stats_2) +
  aes(x = reorder(genes, n_processes), fill = immune_system, weight = n_processes) +
  geom_bar() +
  theme_minimal() +
  scale_fill_manual(
    values = c(Adaptive = "#6f2dbd",
    Innate = "#06d6a0",
    Complement = "#ffadc7")
  ) +
  theme_minimal() +
  labs(fill = "Immune system") +
  xlab("Genes") + 
  ylab("Number of shared conditions") +
  theme(legend.position = "right",
        axis.text.x = element_text(angle = 45, vjust = 1.5, hjust=1, size = 8),
        axis.text.y = element_text(size = 7, angle = 0),
        panel.grid.minor.y = element_blank(),
        strip.text.x.top = element_text(size = 10, face = "bold"),
        legend.text = element_text(size=12),
        legend.title = element_text(size = 12, face = "bold")) +
  # geom_text(aes(label = n_conditions, y = n_conditions), 
  #           position = position_dodge(width = 0.9), 
  #           vjust = 0.2, 
  #           hjust = -0.2, 
  #           size = 1,
  #           angle = 90) +
  ggtitle("Genes shared") +
  coord_flip() +
  facet_wrap(~immune_system, scales = "free_y")

ggsave(ggplot_stats2, file = "ImmuneGO_genes_shared_major_process.png", width = 6, height = 20)

```

## ssGSEA PCA

<https://www.datacamp.com/tutorial/pca-analysis-r>
<https://rpkgs.datanovia.com/factoextra/index.html>
<http://www.sthda.com/english/articles/31-principal-component-methods-in-r-practical-guide/114-mca-multiple-correspondence-analysis-in-r-essentials/>
<https://statisticsglobe.com/pca-before-k-means-clustering-r>\>

#### By GO term

```{r}
ssgsea_results_unified = ImmuneGO_ssgsea_results_unified_26_12_23

data_ImmuneGO = ssgsea_results_unified %>% 
  filter(!process %in% c("TCR REPERTOIRE", "BCR REPERTOIRE"),
         vaccine != "NA")
filename = "ssgsea_results_unified"

pca_ann = data_ImmuneGO %>% 
  select(sample, condition, vaccine, gse_id:sex)

#Converter de long para wide
matrix_genes = data_ImmuneGO %>% 
  select(sample, process, nes)

matrix_genes <- dcast(matrix_genes, `sample` ~ `process`, 
                     value.var = "nes", 
                     fun.aggregate = mean)

matrix_genes <- replace(matrix_genes, matrix_genes == "NaN", 0) 

#Converter coluna 1 em rownames e excluir
matrix_data_pca = matrix_genes %>% 
  as.data.frame()

matrix_data_pca_ready = matrix_data_pca %>%  
  column_to_rownames("sample")

ann_vaccines_pca_matrix = matrix_data_pca %>% 
  merge(pca_ann, by.y = "sample", by.x = "sample", all.x = T, all.y = F) %>% 
  distinct()

ann_vaccines_pca_matrix_ready = ann_vaccines_pca_matrix %>% 
  select(sample, condition, vaccine, day, dose) %>% 
  mutate(dose = as.factor(dose),
         day = as.factor(day))

# Verifique quais colunas têm variância muito baixa
nearZeroVarCols <- nearZeroVar(matrix_data_pca_ready, saveMetrics = TRUE)
matrix_data_pca_ready <- matrix_data_pca_ready[, !nearZeroVarCols$nzv]
pca_res <- prcomp(matrix_data_pca_ready, scale. = TRUE)

# Crie o gráfico de PCA

###### Condition
pca_plot_condition = autoplot(pca_res, 
                    data = ann_vaccines_pca_matrix_ready, 
                    colour = 'condition', 
                    label = 0) + 
  labs(title=paste0(filename, "_bycondition")) + #scale fill manual
  scale_fill_continuous(type = "viridis") +
  theme_minimal()


###### Vaccine
pca_plot_vac = autoplot(pca_res, 
                    data = ann_vaccines_pca_matrix_ready, 
                    colour = 'vaccine', 
                    label = 0) + 
  labs(title=paste0(filename, "_byvaccine")) + #scale fill manual
  scale_color_manual(
    values = c(
  "BBIBP" = "#ffd000",
  "BNT" = "#56cfe1",
  "ChAd" = "#80ffdb",
  "ChAd-BNT" = "#0077b6",
  "ZF2001" = "#b5179e")) +
  theme_minimal()

###### Day
pca_plot_day = autoplot(pca_res, 
                    data = ann_vaccines_pca_matrix_ready, 
                    colour = 'day', 
                    label = 0) + 
  labs(title=paste0(filename, "_day")) + #scale fill manual
   scale_color_manual(
     values = c(
       "0" = "grey90",
    "1" = "#caf0f8",
    "3" = "#90e0ef",
    "6" = "#00b4d8",
    "7" = "#0077b6",
    "14" = "#023e8a",
    "28" = "#03045e")) +
  theme_minimal()

###### Dose

pca_plot_dose = autoplot(pca_res, 
                    data = ann_vaccines_pca_matrix_ready, 
                    colour = 'dose', 
                    label = 0) + 
  labs(title=paste0(filename, "_dose")) +
  scale_color_manual(
    values = c("0" = "#caf0f8", 
               "1" = "#56cfe1", 
               "2" = "#5978d4",
               "3" = "#b5179e"))+
  theme_minimal()

# Exiba o gráfico
print(pca_plot_condition)
print(pca_plot_day)
print(pca_plot_dose)
print(pca_plot_vac)

ggsave(pca_plot_condition, filename = paste0(filename, "CONDITION_PCA_Immune_GO_GSEA_not-ann.png"), width = 10, height = 8)
ggsave(pca_plot_vac, filename = paste0(filename, "_VACCINE_PCA_Immune_GO_GSEA_not-ann.png"), width = 10, height = 8)
ggsave(pca_plot_day, filename = paste0(filename, "_DAY_PCA_Immune_GO_GSEA_not-ann.png"), width = 10, height = 8)
ggsave(pca_plot_dose, filename = paste0(filename, "_DOSE_PCA_Immune_GO_GSEA_not-ann.png"), width = 10, height = 8)

```

```{r}
############### KNN classification

#Determinar o número de clusters para KNN
pca_scores <- data.frame(pca_res$x[, 1:2])
fviz_nbclust(pca_scores,  
                     FUNcluster = kmeans,
                     method = "wss")

set.seed(123)                             # Set seed for randomization
kmeans_clust <- kmeans(pca_scores,        # Perform k-means clustering
                        centers = 2) # Definir numero de clusters

#Visualizar clusters
ggp2 <- fviz_pca_ind(pca_res,
                   habillage = kmeans_clust$cluster,
                   repel = TRUE,
                   addEllipses = TRUE,
                   ellipse.type = "t",
                   label = "none",
                   labelsize = 0) +
  guides(color = guide_legend(override.aes = list(label = ""))) +
  ggtitle(paste0(filename,  "_KNN_Clustered.png")) +
  scale_color_brewer(palette="Set1")

print(ggp2)

ggsave(ggp2, file = paste0(filename, "_KNN_Clustered.png"), width = 5, height = 4)


# Clusterizar por grupo

#Vaccine
pca_group_vac <- fviz_pca_ind(pca_res,
                   habillage = ann_vaccines_pca_matrix_ready$vaccine,
                   repel = TRUE,
                   addEllipses = TRUE,
                   ellipse.type = "t",
                   label = "none",
                   labelsize = 0) +
  guides(color = guide_legend(override.aes = list(label = ""))) +
  ggtitle(paste0(filename,  "_Vac_KNN_Clustered.png")) +
  scale_color_brewer(palette="Set1")

#Dose
pca_group_dose <- fviz_pca_ind(pca_res,
                   habillage = ann_vaccines_pca_matrix_ready$dose,
                   repel = TRUE,
                   addEllipses = TRUE,
                   ellipse.type = "t",
                   label = "none",
                   labelsize = 0) +
  guides(color = guide_legend(override.aes = list(label = ""))) +
  ggtitle(paste0(filename,  "_Dose_KNN_Clustered.png")) +
  scale_color_brewer(palette="Set1")

#Day
pca_group_day <- fviz_pca_ind(pca_res,
                   habillage = ann_vaccines_pca_matrix_ready$day,
                   repel = TRUE,
                   addEllipses = TRUE,
                   ellipse.type = "t",
                   label = "none",
                   labelsize = 0) +
  guides(color = guide_legend(override.aes = list(label = ""))) +
  ggtitle(paste0(filename,  "_Day_KNN_Clustered.png")) +
  scale_color_brewer(palette="Set1")


#visualizar
pca_group_vac
pca_group_dose
pca_group_day

#Salvar
ggsave(pca_group_day, file = paste0(filename, "_Day_Clustered.png"), width = 5, height = 4)
ggsave(pca_group_dose, file = paste0(filename, "_Dose_Clustered.png"), width = 5, height = 4)
ggsave(pca_group_vac, file = paste0(filename, "_Vac_Clustered.png"), width = 5, height = 4)
```

```{r}
#Biplot
biplot = fviz_pca_biplot(pca_res,              # Visualize clusters in biplot
                      col.var = "black",
                      alpha.var = 0.5,
                      habillage = kmeans_clust$cluster,
                      repel = TRUE,
                      addEllipses = TRUE,
                      ellipse.type = "convex",
                      labelsize = 3,
                      label = "var",
                      palette = "Set1")
biplot

ggsave(biplot, file = paste0(filename, "_BIPLOT_KNN_Clustered.png"), width = 10, height = 8)
```

```{r}
########Correlation plot
corr_matrix = cor(matrix_data_pca_ready) 

#Plot
corrplot = ggcorrplot(corr_matrix, hc.order = TRUE) + 
  theme(axis.text.x = element_text(angle = 90, size = 5), 
        axis.text.y = element_text(size = 5))

corrplot
```

```{r}
#Salvar
ggsave(corrplot, file = paste0(filename, "_corrplot.png"), 
       width = 4, #Grande 20, pequeno 10
       height= 4) #Grande 20, pequeno 10
print(corrplot)
data.pca <- princomp(corr_matrix) #PCA
summary(data.pca) #Retornar PCs


#########Scree plot
data.pca = princomp(corr_matrix) #PCA
summary(data.pca) #Retornar PCs

#########Scree plot
scree_plot = fviz_eig(data.pca, 
                      addlabels = TRUE,
                      ylim = c(0, 70)) +
  geom_col(color = "#00AFBB", fill = "#00AFBB") +
  theme_classic()

#Salvar
ggsave(scree_plot, file = paste0(filename, "_GSEA_screeplot.png"), width = 10, height = 3) 
print(scree_plot)


#Scree plot
loadings = data.frame(data.pca$loadings[, 1:3])
loadings$Genes = rownames(loadings)
loadings_1 = arrange(loadings, desc(Comp.1))
loadings_2 = arrange(loadings, desc(Comp.2))
loadings_3 = arrange(loadings, desc(Comp.3))

#Plot
loadings_1_plot = ggplot(loadings_1, aes(x = reorder(Genes, -Comp.1), y=Comp.1, fill = Comp.1)) + 
  ggtitle(paste0("Comp1-Comp2 Genes", filename)) + 
  ylab("Comp1") +
  xlab("Gene sets") +
  geom_bar(stat = "identity") + 
  theme_light() +
  theme(axis.text.x = element_text(vjust = 1, hjust = 1, angle = 45, size = 8), 
        axis.text.y = element_text(size = 5),
        plot.title = element_text(size = 15L, hjust = 0.5)) + 
  scale_fill_continuous(type = "viridis") #cores

print(loadings_1_plot)


loadings_2_plot = ggplot(loadings_2, aes(x = reorder(Genes, -Comp.2), y=Comp.2, fill = Comp.2)) + 
  ggtitle(paste0("Comp2-Comp3 Genes_", filename)) + 
  ylab("Comp2") +
  xlab("Gene sets") +
  geom_bar(stat = "identity") + 
  theme_light() +
  theme(axis.text.x = element_text(vjust = 1, hjust = 1, angle = 45, size = 8), 
        axis.text.y = element_text(size = 5),
        plot.title = element_text(size = 15L, hjust = 0.5)) + 
  scale_fill_continuous(type = "viridis") #cores

print(loadings_2_plot)


#Salvar
ggsave(loadings_1_plot, 
       file = paste0(filename, "_GSEA_genescomp1-2.png"), 
       width = 10, height = 5)

#Salvar
ggsave(loadings_2_plot, 
       file = paste0(filename, "_GSEA_genescomp2-3.png"), 
       width = 10, height = 5)


# Graph of the variables
fviz_pca_var_genes = fviz_pca_var(data.pca, 
                                  col.ind = "cos2",
                                  gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"))
fviz_pca_var_genes

ggsave(fviz_pca_var_genes, file = paste0(filename, "_fviz_pca_var_GSEA.png"), width = 10)

cos2.1 = fviz_cos2(data.pca, choice = "var", axes = 1:2)
cos2.2 = fviz_cos2(data.pca, choice = "var", axes = 2:3)

ggsave(cos2.1, file = paste0(filename, "_cos2_GSEA_1.png"), width = 10)
ggsave(cos2.2, file = paste0(filename, "_cos2_GSEA_2.png"), width = 10)

```

# 7. VAXGO

## 7.1. VaxGO gene sets construction

```{r eval=FALSE, include=FALSE}
library(readr)
library(explore)
library(maditr)
library(GEOquery)
library(Matrix)
library(circlize)
library(RColorBrewer)
library(celldex)
library(biomaRt)
library(org.Hs.eg.db)
library(plotly)
library(DESeq2)
library(msigdbr)
library(ape)
library(GSVA)
library(sva)
library(clusterProfiler)
library(pheatmap)
library(EnhancedVolcano)
library(metaRNASeq)
library(ggbeeswarm)
library(tidyverse)
library(corrr)
library(ggcorrplot)
library(FactoMineR)
library(factoextra)
library(esquisse)
library(corto)
library(factoextra)
library(reshape2)
library(ComplexHeatmap)
library(janitor)
```


## 7.1. Gene set databases

### 7.1.1. Blood transcription modules <https://www.nature.com/articles/ni.2789>

```{r}
#Converter ítens de "Module member genes" em linhas em uma tabela longa
btm_annotation_table_long_immune <- BTM_modules %>%
  separate_rows(`Module member genes`, sep = ",") %>% 
  filter(`Module category` == "immune")

#Salvar
write.csv(btm_annotation_table_long_immune, file = "btm_annotation_table_long_immune.csv")

#Display
head(btm_annotation_table_long_immune)

#Calculate Set size
btm_annotation_nested <- btm_annotation_table_long_immune %>%
  group_by(`Module title`) %>% #Somar e agrupar
  mutate(SetSize = n())

#Convert gene symbol column to aggregated rows
btm_annotation_nested <- btm_annotation_nested %>%
  group_by(`Module title`) %>%
  summarize(Genes = paste(`Module member genes`, collapse = ", "), .groups = "drop") %>% 
  left_join(btm_annotation_nested, by = "Module title") %>% 
  filter(Genes != "NA") %>% 
  relocate(Genes, .after="SetSize")

#Salvar
write.csv(btm_annotation_nested, file = "btm_annotation_genes.csv")

#Convert to unique values table
btm_annotation_unique = btm_annotation_nested %>% 
  select(-`Module member genes`)

btm_annotation_unique = btm_annotation_unique %>% 
  distinct() %>% 
  relocate(Genes, .after="SetSize") %>% 
  filter(Genes != "NA")

#salvar
write.csv(btm_annotation_unique, file = "btm_annotation_unique.csv")

#Display
head(btm_annotation_unique)
```

### 7.1.2. CellMarker

Baixar tabela dos sites
"<http://117.50.127.228/CellMarker/CellMarker_download_files/file/Cell_marker_Human.xlsx>"
ou
"<http://117.50.127.228/CellMarker/CellMarker_download_files/file/Cell_marker_Human.xlsx>"

```{r}
##################### Baixar tabela de marcadores celulares

url <- "http://117.50.127.228/CellMarker/CellMarker_download_files/file/Cell_marker_Human.xlsx" #or http://117.50.127.228/CellMarker/CellMarker_download_files/file/Cell_marker_Human.xlsx 
f <- tempfile(fileext = ".xlsx")
download.file(url, f)
cell_marker_data <- readxl::read_excel(f, 1)

##################### Sub-agrupar células por sistema imunológico

#Filtrar marcadores e células por tecido (Blood)
cells <- cell_marker_data %>%
    filter(tissue_class == "Blood") %>%
    select(cell_name, marker)

##################### Filtrar células
# Acessar todos os tipos de células no dataset de Blood
cellmarker_celltypes <- unique(cells$cell_name)

# Filtrar apenas os tipos que contêm "B cell" e plasmacell
Bcells_no = cellmarker_celltypes[grep("B cell", cellmarker_celltypes)] #44 outputs
Plasmacells_no = cellmarker_celltypes[grep("plasma", cellmarker_celltypes)] #5 outputs

# Filtrar apenas os tipos que contêm "T cell" e seus subtipos "CD4" e "CD8"
Tcells_no = cellmarker_celltypes[grep("T cell", cellmarker_celltypes)] #93 outputs
TCD4_no = cellmarker_celltypes[grep("CD4", cellmarker_celltypes)] #35 outputs
TCD8_no = cellmarker_celltypes[grep("CD8", cellmarker_celltypes)] #34 outputs
Treg_no = cellmarker_celltypes[grep("Treg", cellmarker_celltypes)] #7 outputs
Thelper_no = cellmarker_celltypes[grep("helper", cellmarker_celltypes)] #17 outputs

# Filtrar outros tipos celulares da resposta imune
Neutrophils_no = cellmarker_celltypes[grep("neutrophil", cellmarker_celltypes)] #7 outputs
Monocyte_no = cellmarker_celltypes[grep("monocyte", cellmarker_celltypes)] #14 outputs
DCells_no = cellmarker_celltypes[grep("dendritic", cellmarker_celltypes)] #25 outputs
Macrophages_no = cellmarker_celltypes[grep("macrophage", cellmarker_celltypes)] #7 outputs
NKcells_no = cellmarker_celltypes[grep("killer", cellmarker_celltypes)] # 11 outputs
Neutrophils_no = cellmarker_celltypes[grep("mastcell", cellmarker_celltypes)] #7 outputs

# Criar variáveis para cada célula para enriquecimento

# Filtrar apenas as células da resposta imune adaptativa
Bcells <- filter(cells, str_detect(cell_name, regex("B cell|plasma cell|plasmablast", ignore_case = TRUE)))
TCD4 <- filter(cells, str_detect(cell_name, regex("CD4", ignore_case = TRUE)))
TCD8 <- filter(cells, str_detect(cell_name, regex("CD8", ignore_case = TRUE)))
Treg <- filter(cells, str_detect(cell_name, regex("Treg", ignore_case = TRUE)))
Thelper <- filter(cells, str_detect(cell_name, regex("helper", ignore_case = TRUE)))


# Filtrar células da resposta imune inata
Neutrophils <- filter(cells, str_detect(cell_name, regex("neutrophil", ignore_case = TRUE)))
Monocyte <- filter(cells, str_detect(cell_name, regex("monocyte", ignore_case = TRUE)))
DCells <- filter(cells, str_detect(cell_name, regex("dendritic", ignore_case = TRUE)))
Macrophages <- filter(cells, str_detect(cell_name, regex("macrophage", ignore_case = TRUE)))
Mastcells <- filter(cells, str_detect(cell_name, regex("mast cell", ignore_case = TRUE)))
NKcells <- filter(cells, str_detect(cell_name, regex("killer", ignore_case = TRUE)))
Platelets <- filter(cells, str_detect(cell_name, regex("Platelet", ignore_case = TRUE)))
Eosinophil <- filter(cells, str_detect(cell_name, regex("Eosinophil", ignore_case = TRUE)))
Basophil <- filter(cells, str_detect(cell_name, regex("Basophil", ignore_case = TRUE)))
Granulocyte <- filter(cells, str_detect(cell_name, regex("Granulocyte", ignore_case = TRUE)))

#Agrupar todos os resultados em Immune cells
Immune_adap = rbind(Bcells, TCD4, TCD8, Treg, Thelper)
Immune_adap$Type = "Adaptive"

Immune_innate = rbind(Neutrophils, Monocyte, DCells, Macrophages, Mastcells, NKcells, Platelets, Eosinophil, Basophil, Granulocyte)
Immune_innate$Type = "Innate"

#Unir tabelas
Immune_cells = rbind(Immune_adap, Immune_innate) 

#Salvar
write.csv(Immune_cells, file = "CellMarker_ImmuneCells.csv")

esquisser(Immune_cells_markersbycells)
```

Top 20 cells by #markers

```{r}

Immune_cells_markersbycells = Immune_cells %>% 
  group_by(cell_name, Type) %>% 
  summarise(n_markers = n()) %>% 
  arrange(desc(n_markers))
Immune_cells_markersbycells_top = Immune_cells_markersbycells[1:20, ]

ggImmune_cells_cellsbyimmune = ggplot(Immune_cells_markersbycells_top) +
  aes(x = reorder(cell_name, n_markers), y = n_markers, fill = Type) +
  scale_fill_manual(values = c('Innate' = "#989898",
                               'Adaptive' = "#6DF8DF")) +
  geom_col() +
  labs(
    x = "#Markers",
    y = "Cells",
    title = "Top 20 cells by #markers"
  ) +
  coord_flip() +
  theme_minimal() +
  geom_text(aes(label = n_markers, y = n_markers), 
            position = position_dodge(width = 0.9), 
            vjust = -0.2, 
            hjust = -0.2, 
            size = 3)

ggsave(ggImmune_cells_cellsbyimmune, file= "CellMarker_Top10_cellsbyn_markers.png", width = 10, height = 5)

print(ggImmune_cells_cellsbyimmune)

```

Number of Cells by immune system

```{r}
Immune_cells_cellsbyimmune = Immune_cells %>% 
  group_by(Type, cell_name) %>% 
  summarise(n_cells = n()) %>% 
  group_by(Type) %>% 
  summarise(n_cells = n())

ggImmune_cells_cellsbyimmune = ggplot(Immune_cells_cellsbyimmune) +
  aes(x = reorder(Type, n_cells), y = n_cells, fill = Type) +
  geom_col() +
  scale_fill_manual(values = c('Innate' = "#989898",
                               'Adaptive' = "#6DF8DF")) +
  labs(
    x = "#Cells",
    y = "Immune system",
    title = "#Cells by immune system"
  ) +
  coord_flip() +
  theme_minimal() +
  geom_text(aes(label = n_cells, y = n_cells), 
            position = position_dodge(width = 0.9), 
            vjust = -0.2, 
            hjust = -0.2, 
            size = 3)
ggsave(ggImmune_cells_cellsbyimmune, file= "CellMarker_nCellsbyimmunesystem.png", width = 10, height = 2)

print(ggImmune_cells_cellsbyimmune)
```

### 7.1.3. ImmuneGO

```{r}
organism = "org.Hs.eg.db"
install.packages("GO.db")
BiocManager::install(organism, character.only = TRUE)
library(organism, character.only = TRUE)
library(clusterProfiler)
library(GO.db)

#HGNC symbol to Entrez
x <- org.Hs.egSYMBOL
mapped_genes <- mappedkeys(x)
symbols_entrez <- as.data.frame(x[mapped_genes])

## Entrez to GO
x <- org.Hs.egGO
mapped_genes <- mappedkeys(x)
entrez_geneontology <- as.data.frame(x[mapped_genes])

## GO BP ancestors
go_bp_ancestor = GOBPANCESTOR %>% 
  as.data.frame()
colnames(go_bp_ancestor)[1] <- "go_id"
colnames(go_bp_ancestor)[2] <- "go_ancestor"

## GO ID to terms
go_bp = AnnotationDbi::select(GO.db, columns=c("GOID","TERM"), keys= "BP", keytype= "ONTOLOGY") %>% 
  clean_names() %>% 
  select(go_id = goid, term)

## GO child and ancestor
go_bp_ancestor = go_bp_ancestor %>% 
  inner_join(go_bp %>% rename(go_ancestor = go_id), by = "go_ancestor") %>% 
  rename(term_ancestor = term)

#GO ids to terms
go_bp = AnnotationDbi::select(GO.db, columns=c("GOID","TERM"), keys="BP", keytype="ONTOLOGY") %>% 
  clean_names() %>% 
  rename(go_id = goid) %>% 
  select(-ontology) %>% 
  right_join(entrez_geneontology, by = "go_id") %>% 
  right_join(symbols_entrez, by = "gene_id") %>% 
  clean_names() %>% 
  left_join(go_bp_ancestor, by = "go_id") %>% 
  distinct()

#GO immune process

immunego_go.bp = go_bp %>% 
  filter(term_ancestor == "immune system process") %>% 
  mutate(term = toupper(term))

ImmuneGO_Annotated_Genes_26_2_24 = ImmuneGO_Annotated_GO_8_2_24 %>% 
  select(-"...1") %>% 
  mutate(genes = str_replace_all(genes, "\\s+", "")) %>% 
  separate_rows(genes, sep = ",") %>% 
  full_join(immunego_go.bp %>% select(genes = symbol, process = term), by = join_by("process", "genes")) %>% 
  distinct()

ImmuneGO_Annotated_GO_26_2_24 = ImmuneGO_Annotated_Genes_26_2_24 %>% 
  group_by(process) %>%
  summarize(genes = paste0(genes, collapse = ",")) %>%
  inner_join(ImmuneGO_Annotated_Genes_26_2_24 %>% select(-genes) %>% distinct(), by = "process")

write.csv(ImmuneGO_Annotated_GO_26_2_24, file = "ImmuneGO_Annotated_GO_26_2_24.csv", row.names = F)

write.csv(ImmuneGO_Annotated_Genes_26_2_24, file = "ImmuneGO_Annotated_Genes_26_2_24.csv", row.names = F)

```

Group immune system genes by process and cell

```{r}
ImmuneGO_Annotated_unique = ImmuneGO_Annotated_Genes %>% 
  filter(go_term != "Manual")

#Agrupar genes de todos os processos do sistema imune ----

ImmuneGO_Adaptive <- ImmuneGO_Annotated_unique %>%
  filter(immune_system == "Adaptive") %>% 
  summarize(set_size = n(), 
            genes = paste(genes, collapse = ",")) %>% 
  mutate(process = "ADAPTIVE IMMUNE SYSTEM",
  `Gene set, short` = "Adaptive immune system",
  `Immune system` = "Adaptive",
  `Immune sub-system` = "General",
  `Immune tissue` = "General",
  `GO.TERM` = "Manual") %>% 
  clean_names()

ImmuneGO_Innate <- ImmuneGO_Annotated_unique %>%
  filter(immune_system == "Innate") %>% 
  summarize(set_size = n(), 
            genes = paste(genes, collapse = ",")) %>% 
  mutate(process = "INNATE IMMUNE SYSTEM",
  `Gene set, short` = "Innate immune system",
  `Immune system` = "Innate",
  `Immune sub-system` = "General",
  `Immune tissue` = "General",
  `GO.TERM` = "Manual") %>% 
  clean_names()

ImmuneGO_Complement <- ImmuneGO_Annotated_unique %>%
  filter(immune_system == "Complement") %>% 
  summarize(set_size = n(), 
            genes = paste(genes, collapse = ",")) %>% 
  mutate(process = "COMPLEMENT",
  `Gene set, short` = "Complement immune system",
  `Immune system` = "Complement",
  `Immune sub-system` = "General",
  `Immune tissue` = "General",
  `GO.TERM` = "Manual") %>% 
  clean_names()

ImmuneGO_Inflammation <- ImmuneGO_Annotated_unique %>%
  filter(immune_sub_system == "Inflammation") %>% 
  summarize(set_size = n(), 
            genes = paste(genes, collapse = ",")) %>% 
  mutate(process = "INFLAMMATION",
  `Gene set, short` = "Inflammation",
  `Immune system` = "Innate",
  `Immune sub-system` = "Inflammation",
  `Immune tissue` = "General",
  `GO.TERM` = "Manual") %>% 
  clean_names()

ImmuneGO_Interferon <- ImmuneGO_Annotated_unique %>%
  filter(immune_sub_system %in% c("Antiviral", "Interferon")) %>% 
  summarize(set_size = n(), 
            genes = paste(genes, collapse = ",")) %>% 
  mutate(process = "ANTIVIRAL & IFN",
  `Gene set, short` = "Antiviral and Interferon",
  `Immune system` = "Innate",
  `Immune sub-system` = "Antiviral and Interferon",
  `Immune tissue` = "General",
  `GO.TERM` = "Manual") %>% 
  clean_names()

ImmuneGO_ARPP = ImmuneGO_Annotated_unique %>%
  filter(immune_sub_system %in% c("PRR", "APC")) %>% 
  summarize(set_size = n(), 
            genes = paste(genes, collapse = ",")) %>% 
  mutate(process = "ARPP",
  `Gene set, short` = "Antigen receptors, processing and presentation",
  `Immune system` = "Innate",
  `Immune sub-system` = "APC",
  `Immune tissue` = "APC",
  `GO.TERM` = "Manual") %>% 
  clean_names()


ImmuneGO_Innate_DC <- ImmuneGO_Annotated_unique %>%
  filter(immune_tissue == "Dendritic cell") %>% 
  summarize(set_size = n(), 
            genes = paste(genes, collapse = ",")) %>% 
  mutate(process = "DENDRITIC CELL",
  `Gene set, short` = "Dendritic cell",
  `Immune system` = "Innate",
  `Immune sub-system` = "Cell",
  `Immune tissue` = "APC",
  `GO.TERM` = "Manual") %>% 
  clean_names()

ImmuneGO_Innate_Macrophage <- ImmuneGO_Annotated_unique %>%
  filter(immune_tissue == "Macrophage") %>% 
  summarize(set_size = n(), 
            genes = paste(genes, collapse = ",")) %>% 
  mutate(process = "MACROPHAGE",
  `Gene set, short` = "Macrophage",
  `Immune system` = "Innate",
  `Immune sub-system` = "Cell",
  `Immune tissue` = "APC",
  `GO.TERM` = "Manual") %>% 
  clean_names()

ImmuneGO_Innate_Mastcell <- ImmuneGO_Annotated_unique %>%
  filter(immune_tissue == "Mast cell") %>% 
  summarize(set_size = n(), 
            genes = paste(genes, collapse = ",")) %>% 
  mutate(process = "MAST CELL",
  `Gene set, short` = "Mast cell",
  `Immune system` = "Innate",
  `Immune sub-system` = "Cell",
  `Immune tissue` = "Mast cell",
  `GO.TERM` = "Manual") %>% 
  clean_names()

ImmuneGO_Innate_Neutrophil <- ImmuneGO_Annotated_unique %>%
  filter(immune_tissue == "Neutrophil") %>% 
  summarize(set_size = n(), 
            genes = paste(genes, collapse = ",")) %>% 
  mutate(process = "NEUTROPHIL",
  `Gene set, short` = "Neutrophil",
  `Immune system` = "Innate",
  `Immune sub-system` = "Cell",
  `Immune tissue` = "Neutrophil",
  `GO.TERM` = "Manual") %>% 
  clean_names()

ImmuneGO_Innate_Eosinophil <- ImmuneGO_Annotated_unique %>%
  filter(immune_tissue == "Eosinophil") %>% 
  summarize(set_size = n(), 
            genes = paste(genes, collapse = ",")) %>% 
  mutate(process = "EOSINOPHIL",
  `Gene set, short` = "Eosinophil",
  `Immune system` = "Innate",
  `Immune sub-system` = "Cell",
  `Immune tissue` = "Eosinophil",
  `GO.TERM` = "Manual") %>% 
  clean_names()

ImmuneGO_Innate_NK <- ImmuneGO_Annotated_unique %>%
  filter(immune_tissue == "NK cell") %>% 
  summarize(set_size = n(), 
            genes = paste(genes, collapse = ",")) %>% 
  mutate(process = "NK CELL",
  `Gene set, short` = "Natural Killer Cell",
  `Immune system` = "Innate",
  `Immune sub-system` = "Cell",
  `Immune tissue` = "NK cell",
  `GO.TERM` = "Manual") %>% 
  clean_names()

ImmuneGO_Innate_Monocyte <- ImmuneGO_Annotated_unique %>%
  filter(immune_tissue == "Monocyte") %>% 
  summarize(set_size = n(), 
            genes = paste(genes, collapse = ",")) %>% 
  mutate(process = "MONOCYTE",
  `Gene set, short` = "Monocyte",
  `Immune system` = "Innate",
  `Immune sub-system` = "Cell",
  `Immune tissue` = "APC",
  `GO.TERM` = "Manual") %>% 
  clean_names()

ImmuneGO_Adap_Cellular <- ImmuneGO_Annotated_unique %>% 
  filter(immune_sub_system == "Cellular") %>% 
  summarize(set_size = n(), 
            genes = paste(genes, collapse = ",")) %>% 
  mutate(process = "CELLULAR ADAPTIVE IMMUNE SYSTEM",
  `Gene set, short` = "Cellular adaptive immune system",
  `Immune system` = "Adaptive",
  `Immune sub-system` = "Cellular",
  `Immune tissue` = "Cellular",
  `GO.TERM` = "Manual") %>% 
  clean_names()


ImmuneGO_Adap_Tcell <- ImmuneGO_Annotated_unique %>%
  filter(immune_tissue == "T cell") %>% 
  summarize(set_size = n(), 
            genes = paste(genes, collapse = ",")) %>% 
  mutate(process = "T CELL",
  `Gene set, short` = "T cell",
  `Immune system` = "Adaptive",
  `Immune sub-system` = "Cellular",
  `Immune tissue` = "Cellular",
  `GO.TERM` = "Manual") %>% 
  clean_names()

ImmuneGO_Adap_Thelper <- ImmuneGO_Annotated_unique %>%
  filter(immune_tissue == "T helper") %>% 
  summarize(set_size = n(), 
            genes = paste(genes, collapse = ",")) %>% 
  mutate(process = "T HELPER CELLS",
  `Gene set, short` = "T helper cell",
  `Immune system` = "Adaptive",
  `Immune sub-system` = "Cellular",
  `Immune tissue` = "Cellular",
  `GO.TERM` = "Manual") %>% 
  clean_names()

ImmuneGO_Adap_Humoral <- ImmuneGO_Annotated_unique %>%
  filter(immune_sub_system == "Humoral") %>% 
  summarize(set_size = n(), 
            genes = paste(genes, collapse = ",")) %>% 
  mutate(process = "HUMORAL ADAPTIVE IMMUNE SYSTEM",
  `Gene set, short` = "Humoral adaptive immune system",
  `Immune system` = "Adaptive",
  `Immune sub-system` = "Humoral",
  `Immune tissue` = "Humoral",
  `GO.TERM` = "Manual") %>% 
  clean_names()

ImmuneGO_Adap_Bcell <- ImmuneGO_Annotated_unique %>%
  filter(immune_tissue == "B cell") %>% 
  summarize(set_size = n(), 
            genes = paste(genes, collapse = ",")) %>% 
  mutate(process = "B CELL",
  `Gene set, short` = "B cell",
  `Immune system` = "Adaptive",
  `Immune sub-system` = "Humoral",
  `Immune tissue` = "B cell",
  `GO.TERM` = "Manual") %>% 
  clean_names()

ImmuneGO_Adap_Ig <- ImmuneGO_Annotated_unique %>%
  filter(immune_tissue == "B cell") %>% 
  summarize(set_size = n(), 
            genes = paste(genes, collapse = ",")) %>% 
  mutate(process = "IMMUNOGLOBULIN MEDIATED RESPONSE",
  `Gene set, short` = "Ig-mediated response",
  `Immune system` = "Adaptive",
  `Immune sub-system` = "Humoral",
  `Immune tissue` = "B cell",
  `GO.TERM` = "Manual") %>% 
  clean_names()

ImmuneGO_other <- ImmuneGO_Annotated_unique %>%
  filter(immune_system == "General" |
         immune_tissue == "Leukocyte") %>% 
  summarize(set_size = n(), 
            genes = paste(genes, collapse = ",")) %>% 
  mutate(process = "GENERAL",
  `Gene set, short` = "General",
  `Immune system` = "General",
  `Immune sub-system` = "General",
  `Immune tissue` = "General",
  `GO.TERM` = "Manual") %>% 
  clean_names()


ImmuneGO_antimicrobial <- ImmuneGO_Annotated_unique %>%
  filter(immune_tissue == "Antimicrobial") %>% 
  summarize(set_size = n(), 
            genes = paste(genes, collapse = ",")) %>% 
  mutate(process = "ANTIMICROBIAL",
  `Gene set, short` = "Antimicrobial immune response",
  `Immune system` = "General",
  `Immune sub-system` = "General",
  `Immune tissue` = "General",
  `GO.TERM` = "Manual") %>% 
  clean_names()


# Unite sets -----

ImmuneGO_Annotated_GO = bind_rows(ImmuneGO_Adaptive, 
                            ImmuneGO_Innate, 
                            ImmuneGO_Complement, 
                            ImmuneGO_Inflammation,
                            ImmuneGO_Interferon,
                            ImmuneGO_ARPP,
                            ImmuneGO_Innate_Macrophage,
                            ImmuneGO_Innate_Mastcell,
                            ImmuneGO_Innate_Neutrophil,
                            ImmuneGO_Innate_Eosinophil,
                            ImmuneGO_Innate_NK,
                            ImmuneGO_Innate_Monocyte,
                            ImmuneGO_Adap_Cellular,
                            ImmuneGO_Adap_Tcell,
                            ImmuneGO_Adap_Thelper,
                            ImmuneGO_Adap_Humoral,
                            ImmuneGO_Adap_Bcell,
                            ImmuneGO_Adap_Ig,
                            ImmuneGO_other,
                            ImmuneGO_antimicrobial,
                            ImmuneGO_Innate_DC,
                            ImmuneGO_Annotated_unique) %>% 
  select(process:go_term, set_size:genes) %>% 
  distinct()

ImmuneGO_Annotated_Genes = ImmuneGO_Annotated_GO %>% 
   separate_rows(genes, sep = ",")

# Separate TCR and BCR repertoire genes
TCR_BCR_repertoire = ImmuneGO_Annotated_Genes %>% 
  filter(grepl("^TR", genes) & process == "ADAPTIVE IMMUNE SYSTEM" |
           grepl("^IG", genes) & process == "ADAPTIVE IMMUNE SYSTEM" |
           grepl("^TR", genes) & process == "CELLULAR ADAPTIVE IMMUNE SYSTEM" |
         grepl("^IG", genes) & process == "HUMORAL ADAPTIVE IMMUNE SYSTEM") %>% 
  mutate(process = if_else(process == "CELLULAR ADAPTIVE IMMUNE SYSTEM", "TCR REPERTOIRE", 
                           "BCR REPERTOIRE"),
         gene_set_short = if_else(process == "TCR REPERTOIRE", 
                                  "TCR REPERTOIRE", 
                           "BCR REPERTOIRE"),
         process = if_else(grepl("^TR", genes), "TCR REPERTOIRE", "BCR REPERTOIRE"),
         immune_sub_system = if_else(process == "TCR REPERTOIRE", "Cellular", "Humoral"),
         immune_tissue = if_else(process == "TCR REPERTOIRE", "Cellular", "Humoral"))

immunego_manual = ImmuneGO_Annotated_Genes %>%
  filter(go_term == "Manual") %>% 
  anti_join(TCR_BCR_repertoire %>% select(genes) %>% distinct(), by = "genes") %>% 
  bind_rows()

#Verify if all genes are annotated. If there are no left genes, you are good to go
a = ImmuneGO_Annotated_Genes %>% filter(go_term == "Manual") %>% select(genes) %>% distinct()
ImmuneGO_Annotated_Genes %>% select(process, genes, immune_system) %>% anti_join(a, by = "genes")


#Salvar
write.csv(ImmuneGO_Annotated_GO, file = paste0("ImmuneGO_Annotated_GO_", today(), ".csv"), row.names = F)
saveRDS(ImmuneGO_Annotated_GO, file = paste0("ImmuneGO_Annotated_GO_", today(), ".rds"))
saveRDS(ImmuneGO_Annotated_Genes, file = paste0("ImmuneGO_Annotated_Genes_", today(), ".rds"))

#Display
view(ImmuneGO_Annotated_GO)
view(ImmuneGO_Annotated_Genes)

```

Top 10 gene sets
```{r}
#Top 10 gene sets
ImmuneGO_Annotated_unique_top10 = ImmuneGO_Annotated_GO %>% 
  clean_names() %>% 
  arrange(desc(set_size)) %>% 
  slice_head(n = 20) %>% 
  mutate(immune_system = if_else(gene_set_short == "Adaptive Immune response", "Adaptive", immune_system))

#Gráfico de barras
ImmuneGO_top10_barplot = ggplot(ImmuneGO_Annotated_unique_top10) +
 aes(x = reorder(gene_set_short, set_size), 
     weight = set_size, 
     fill = immune_system) +  # Use reorder para reordenar as barras +
  geom_bar()+
  scale_fill_manual(
    values = c(Adaptive = "#6DF8DF",
    Complement = "#e9edc9",
    General = "#bde0fe",
    Innate = "#989898",
    Undefined = "#edede9")) +
 labs(y = "Set size", title = "ImmuneGO", subtitle = "Top 10 GO, by set size") +
 coord_flip() +
 theme_minimal()

#Salvar
ggsave(ImmuneGO_top10_barplot, file="ImmuneGO_top10_barplot.png", width = 10, height = 6)

#Display
head(ImmuneGO_top10_barplot)
```

#### Dataset representation

```{r}
#Genes representados e não representados

ImmuneGO_Annotated_Genes = ImmuneGO_Annotated_Genes_Nested_Interesting %>% 
  select(genes, process) 

ImmuneGO_Annotated_Genes_2 = ImmuneGO_Annotated_Genes_Nested_Interesting %>% 
  select(genes) %>% 
  distinct() %>% 
  mutate(process = "All") %>% 
  bind_rows(ImmuneGO_Annotated_Genes_Nested_Interesting)


ImmuneGO_Annotated_Genes_Interesting_studies = ImmuneGO_Annotated_Genes_2 %>% 
  merge(genes_raw_combined, all.x = T, all.y =F) %>% 
  select(study, process, genes) %>% 
  group_by(study, process) %>% 
  mutate(process = case_when(
  process == "ANTIVIRAL AND INTERFERON" ~ "ANTIVIRAL/IFN",
  process == "INNATE IMMUNE SYSTEM" ~ "INNATE",
  process == "HUMORAL ADAPTIVE IMMUNE SYSTEM" ~ "HUMORAL",
  process == "CELLULAR ADAPTIVE IMMUNE SYSTEM" ~ "CELLULAR",
  process == "ADAPTIVE IMMUNE SYSTEM" ~ "ADAPTIVE",
  process == "IMMUNOGLOBULIN MEDIATED IMMUNE RESPONSE" ~ "IMMUNOGLOBULIN-MEDIATED",
  TRUE ~ process)) %>% 
  mutate(in_out = 1) %>% 
  filter(!is.na(study)) %>% 
  mutate(process = factor(process, levels = c('INNATE', 
                                               'ADAPTIVE', 
                                               'CELLULAR', 
                                               'HUMORAL')),
         study = factor(study, levels = c('GSE199750', 
                                          'GSE201533',
                                          'GSE201530',
                                          'GSE189039',
                                          "GSE206023")))


######## ALL

ImmuneGO_Interesting_Genes_plot.higher <- ImmuneGO_Annotated_Genes_Interesting_studies %>%
  filter(process %in% c("ADAPTIVE", "INNATE", "HUMORAL", "CELLULAR")) %>%
  ggplot() +
  aes(x = study, y = genes, fill = study) +  # Reordenar os levels de study
  geom_tile() +
  scale_fill_manual(
    values = c("GSE199750" = "#80ffdb",
               "GSE201533" = "#5e60ce", 
               "GSE201530" = "#7400b8",
               "GSE206023" = "#4361ee",
               "GSE189039" = "#48bfe3")) +
  coord_flip() +
  theme_minimal() +
  facet_wrap(vars(process), scales = "free_x", ncol = 1) +
  labs(
    x = "GSE_ID",
    y = "#Genes",
    title = "#Genes represented in ImmuneGO, by study") + 
  theme(axis.text.x = element_text(angle = 90, size = 3, hjust = 1),
        axis.text.y = element_text(size = 20),
        strip.text = element_text(size= 20)) +
    scale_x_discrete(limits = rev(levels(ImmuneGO_Annotated_Genes_Interesting_studies$study)))

#Salvar
ggsave(ImmuneGO_Interesting_Genes_plot.higher, 
       file = "ImmuneGO_Interesting_Genes_Adaptive_Innate_plot_label.png", 
       width = 20, height = 20)


#Sem label
ImmuneGO_Interesting_Genes_plot.nolabel = ImmuneGO_Annotated_Genes_Interesting_studies %>%
  filter(process %in% c("ADAPTIVE", "INNATE", "HUMORAL", "CELLULAR")) %>% 
  ggplot() +
  aes(x = study, y = genes, fill = study) +  # Reordenar os levels de study
  geom_tile() +
  scale_fill_manual(
    values = c("GSE199750" = "#80ffdb",
               "GSE201533" = "#5e60ce", 
               "GSE201530" = "#7400b8",
               "GSE206023" = "#4361ee",
               "GSE189039" = "#48bfe3"))  +
  coord_flip() +
  theme_minimal() +
  facet_wrap(vars(process), scales = "free_x", ncol = 1)  +
  labs(
    x = "GSE_ID",
    y = "#Genes",
    title = "#Genes represented in ImmuneGO, by study") + 
  theme(axis.text.x = element_text(angle = 90, size = 0, hjust = 1),
        axis.text.y = element_text(size = 20),
        strip.text = element_text(size= 20)) +
  scale_x_discrete(limits = rev(levels(ImmuneGO_Annotated_Genes_Interesting_studies$study)))

#Salvar
ggsave(ImmuneGO_Interesting_Genes_plot.nolabel, 
       file = "ImmuneGO_Interesting_Genes_Adaptive_Innate_plot_nolabel.png", 
       width = 20, height = 20)

############### Grouped

ImmuneGO_Annotated_Genes_Nested_Interesting_grouped = ImmuneGO_Annotated_Genes_Nested_Interesting %>% 
  filter(process %in% c("INNATE IMMUNE SYSTEM", 
                        "ADAPTIVE IMMUNE SYSTEM", 
                        "CELLULAR ADAPTIVE IMMUNE SYSTEM", 
                        "HUMORAL ADAPTIVE IMMUNE SYSTEM"))

ImmuneGO_genes_indatasets.merge_grouped = genes_raw_combined %>% 
  group_by(study, genes) %>% 
  merge(ImmuneGO_Annotated_Genes_Nested_Interesting_grouped, by.x = "genes", all.x = F, all.y = T) %>%
  select(genes, study, process) %>% 
  distinct() %>% 
  group_by(study, process) %>%  
  select(-genes) %>% 
  summarise(n_genes_in = n()) %>% 
  merge(ImmuneGO_Annotated_Genes_Nested_Interesting_grouped, by = "process", all.y = F) %>% 
  select(study, process, set_size, n_genes_in) %>%
  distinct() %>% 
  filter(study != "NA")  %>% 
  summarise(study,
            process,
            set_size,
            n_genes_in,
            prop_in = (n_genes_in/set_size)*100,
            prop_in = round(prop_in, 1),
            study,
            process) %>% 
  mutate(process = case_when(
            process == "ANTIVIRAL AND INTERFERON" ~ "ANTIVIRAL/IFN",
            process == "INNATE IMMUNE SYSTEM" ~ "INNATE",
            process == "HUMORAL ADAPTIVE IMMUNE SYSTEM" ~ "HUMORAL",
            process == "CELLULAR ADAPTIVE IMMUNE SYSTEM" ~ "CELLULAR",
            process == "ADAPTIVE IMMUNE SYSTEM" ~ "ADAPTIVE",
            process == "IMMUNOGLOBULIN MEDIATED IMMUNE RESPONSE" ~ "IMMUNOGLOBULIN-MEDIATED",
            TRUE ~ process),
         process = factor(process, levels = c('INNATE', 
                                               'ADAPTIVE', 
                                               'CELLULAR', 
                                               'HUMORAL')),
         study = factor(study, levels = unique(study)))

#Plotar

ggImmuneGO_genes_indatasets.stats = ggplot(ImmuneGO_genes_indatasets.merge_grouped) +
  aes(x = reorder(study, prop_in), fill = study, weight = prop_in) +
  geom_bar() +
  scale_fill_manual(
    values = c("GSE199750" = "#80ffdb",
               "GSE201533" = "#5e60ce", 
               "GSE201530" = "#7400b8",
               "GSE206023" = "#4361ee",
               "GSE189039" = "#48bfe3")) +
  coord_flip() +
  theme_minimal() +
  geom_text(aes(label = prop_in, y = prop_in), 
            position = position_dodge(width = 0.9), 
            vjust = -0, 
            hjust = -0.2, 
            size = 3) +
  labs(
    x = "GSE_ID",
    y = "#Genes",
    title = "#Genes represented in ImmuneGO, by study"
  ) +
  facet_wrap(vars(process), ncol = 1) +
  theme(strip.text = element_text(size=8))
  
ggsave(ggImmuneGO_genes_indatasets.stats, file = "ImmuneGO_Genesrepresented_bystudy.png", width = 6, height = 10)

```

### VAX MSigDB


```{r}
#Importar arquivo
VaxSigDB_genesets = VAX_GeneSets_Annotated_26_10_23

# Divida os dados em blocos de informações para cada vacina
vaccines_att <- split(VaxSigDB_genesets$Col2, VaxSigDB_genesets$Col1)

# Combine data
VAX_Genesets <- bind_rows(lapply(vaccines_att, function(x) data.frame(t(x))), .id = "V")

# Name a new index column
colnames(VAX_Genesets) <- paste0("V", seq_len(ncol(VAX_Genesets)))

# Transpose
rownames(VAX_Genesets) <- VAX_Genesets$V1
VAX_Genesets = VAX_Genesets %>% 
  t() %>%
  as.data.frame()

#Order columns
VAX_Genesets = VAX_Genesets %>% 
  select(STANDARD_NAME, everything()) %>% 
  filter(STANDARD_NAME != "STANDARD_NAME") 
rownames(VAX_Genesets) = VAX_Genesets$SYSTEMATIC_NAME
VAX_Genesets = VAX_Genesets %>% select(STANDARD_NAME, 
                                       DESCRIPTION_BRIEF,
                                       DESCRIPTION_FULL,
                                       PMID,
                                       AUTHORS,
                                       SYSTEMATIC_NAME,
                                       GENE_SYMBOLS,
                                       everything())

#Anotar novas colunas
VAX_Genesets <- VAX_Genesets %>%
  mutate(
    SAMPLE_SOURCE = ifelse(
      grepl("pbmc|peripheral blood mononuclear cell|peripheral blood mononuclear cells", DESCRIPTION_FULL, ignore.case = TRUE) |
      grepl("pbmc|peripheral blood mononuclear cell|peripheral blood mononuclear cells", DESCRIPTION_BRIEF, ignore.case = TRUE),
      "PBMC",
      ifelse(
        grepl("whole blood", DESCRIPTION_FULL, ignore.case = TRUE) |
        grepl("whole blood", DESCRIPTION_BRIEF, ignore.case = TRUE) |
        grepl("_BLOOD", STANDARD_NAME, ignore.case = TRUE)  ,
        "WHOLE BLOOD",
        "OTHER")),
    REGULATION = ifelse(
      grepl("up", DESCRIPTION_BRIEF, ignore.case = TRUE),
      "UP",
      ifelse(
        grepl("down", DESCRIPTION_BRIEF, ignore.case = TRUE),
        "DOWN",
        "UNKNOWN")))

#Reordenar colunas
VAX_Genesets = VAX_Genesets %>% select(SAMPLE_SOURCE, REGULATION, everything())

# Salvar
write.csv(VAX_Genesets, file = "VAXSigDB_Annotated.csv") #Continuar a anotação no Google Sheets

head(VAX_Genesets)
```

 Long table with Gene symbols

```{r}
############## Long table with Gene symbols

VAX_Genesets = VAXSigDB_Annotated

VAX_Genes <- VAX_Genesets %>%
  mutate(GENE = GENE_SYMBOLS) %>% 
  separate_rows(GENE, sep = ",") %>% 
  select(GENE, everything()) %>%
  group_by(STANDARD_NAME) %>%
  mutate(SetSize = n()) %>% 
  relocate(SetSize, .after = GENE_SYMBOLS)

VAX_GeneSets_annotated <- VAX_Genes %>%
  select(-GENE) %>% 
  distinct()

#Salvar
write.csv(VAX_Genes, file = "VAX_Genes_Annotated_RAW.csv")
write.csv(VAX_GeneSets_annotated, file = "VAX_Genesets_Annotated_RAW.csv")

#Display
head(VAX_GeneSets_annotated)
```

Filtering

```{r}
unique(VAX_GeneSets_annotated$STUDY_TYPE) 
#"CORRELATION"
#"VACCINE"
#"SUBSET"
#"CHALLENGE"

unique(VAX_GeneSets_annotated$STUDY_SUBTYPE)
# "ANTIBODY"				
# "VAC ONLY"				
# "BOOSTER"				
# "VAC VS CTRL"				
# "SIGNATURE"				
# "ADVERSE EVENT"			
# "COMPARISON"				
# "RESPONDER VS NON RESPONDER"				
# "AGE"				
# "RESPONDER"				
# "ADJUVANT"				
# "NON RESPONDER"				
# "ANTIMICROBIAL ACTIVITY"				
# "CELLS"				
# "CHALLENGE"				
# "TOP DEGS"				
# "PROTECTION"				
# "TRAINING SET"				
# "EXON ANALYSIS"				
# "DOSE"				
# "VAC VS OTHER"


# Subset de vacinas para as seguintes análises comparativas
VAX_GeneSets_Blood_PBMC_VAC = VAX_GeneSets_annotated %>% 
  filter(SAMPLE_SOURCE %in% c("PBMC", "WHOLE BLOOD"), 
         STUDY_TYPE %in% c("VACCINE"))

# Subset de vacinas com resultados de comparações entre vacinas, adjuvantes, idade e sexo
VAX_GeneSets_Blood_PBMC_CORR = VAX_GeneSets_annotated %>% 
  filter(SAMPLE_SOURCE %in% c("PBMC", "WHOLE BLOOD"), 
         STUDY_TYPE %in% c("CORRELATION"))

# Subset com resultados de correlações com resposta inata, celular e humoral
VAX_GeneSets_Blood_PBMC_SUBSETS = VAX_GeneSets_annotated %>% 
  filter(SAMPLE_SOURCE %in% c("PBMC", "WHOLE BLOOD"), 
         STUDY_TYPE %in% c("SUBSET"))


# Subset de vacinas com genes associados a eventos adversos
VAX_GeneSets_Blood_PBMC_CHALLENGE = VAX_GeneSets_annotated %>% 
  filter(SAMPLE_SOURCE %in% c("PBMC", "WHOLE BLOOD"), 
         STUDY_TYPE %in% c("CHALLENGE"))

VAX_GeneSets_Blood_PBMC_VAC.stats = VAX_GeneSets_Blood_PBMC_VAC %>% 
  group_by(VACCINE, PLATFORM, REGULATION, `TARGET_PATHOGEN_DISEASE`, MICROBE_TYPE) %>% 
  summarise(Count = n()) %>% 
  arrange(desc(Count))

```

# Representação de genes por gene set para cada condição

## Gráfico de linhas

```{r}

Immune_GO_GSEA

################## Innate
Immune_GO_innate = Immune_GO_GSEA %>% filter(`Immune system` == "Innate")
write.csv(Immune_GO_innate, file = "Immune_GO_innate_l2fc.csv")

Immune_GO_innate_SUM <- Immune_GO_innate %>% 
  rename(Condition = study) %>% 
  group_by(Condition, Process) %>%
  mutate(Soma = n()) %>% select(Vaccine, Condition, Process, `Immune group`, `Immune response gene set, specific`, `Immune system`, Soma) %>% filter(Vaccine != "NA")
write.csv(Immune_GO_innate_SUM, file="Immune_GO_innate_SUM.csv")


################## Adaptive
#Humoral
Immune_GO_humoral = Immune_allgenes %>% filter(`Immune system` == "Humoral")
write.csv(Immune_GO_humoral, file = "Immune_GO_humoral_l2fc.csv")

Immune_GO_humoral_SUM <- Immune_GO_humoral %>% 
  rename(Condition = study) %>% 
  group_by(Condition, Process) %>%
  mutate(Soma = n()) %>% select(Vaccine, Condition, Process, `Immune group`, `Immune response gene set, specific`, `Immune system`, Soma) %>% filter(Vaccine != "NA")
write.csv(Immune_GO_humoral_SUM, file="Immune_GO_humoral_SUM.csv")

#Cellular
Immune_GO_cellular = Immune_allgenes %>% filter(`Immune system` == "Cellular")
write.csv(Immune_GO_cellular, file = "Immune_GO_cellular_l2fc.csv")

Immune_GO_cellular_SUM <- Immune_GO_cellular %>% 
  rename(Condition = study) %>% 
  group_by(Condition, Process) %>%
  mutate(Soma = n()) %>% select(Vaccine, Condition, Process, `Immune group`, `Immune response gene set, specific`, `Immune system`, Soma) %>% filter(Vaccine != "NA")
write.csv(Immune_GO_cellular_SUM, file="Immune_GO_cellular_SUM.csv")

################## General
Immune_GO_general = Immune_allgenes %>% filter(`Immune system` == "General")
write.csv(Immune_GO_general, file = "Immune_GO_general_l2fc.csv")

Immune_GO_general_SUM <- Immune_GO_general %>% 
  rename(Condition = study) %>% 
  group_by(Condition, Process) %>%
  mutate(Soma = n()) %>% select(Vaccine, Condition, Process, `Immune group`, `Immune response gene set, specific`, `Immune system`, Soma) %>% filter(Vaccine != "NA")
write.csv(Immune_GO_general_SUM, file = "Immune_GO_general_SUM.csv")

################## All immune processes and genes
Immune_allgenes_vaccines = Immune_allgenes_l2fc %>% rename(Condition = study, Genes = genes) %>% filter(Vaccine != "NA")
ann_vaccines_pca_matrix = merge(ann_vaccines_pca, matrix_data_pca, by.x = "Condition", all.x = F, all.y = F)
ann_vaccines_pca_matrix_long = ann_vaccines_pca_matrix %>% gather(Genes, L2FC, -Condition, -Vaccine, -Day, -Dose)

Immune_allgenes_vaccines_SUM <- Immune_allgenes_vaccines %>%
  group_by(Condition, Process) %>% #Somar e agrupar
  summarize(Soma = n())

Immune_allgenes_vaccines_SUM = merge(ann_vaccines_pca, Immune_allgenes_vaccines_SUM, by.x = "Condition", all.x = F, all.y = F) # Anotar vacinas
write.csv(Immune_allgenes_vaccines_SUM, file="Immune_allgenes_vaccines_SUM.csv") #Salvar

```


# COVID-19 and other vaccines

## VENN DIAGRAM

```{r}
# Input files ------

immunego = ImmuneGO_Annotated_Genes_28_2_24 # ImmuneGO Annotation
degs_updown_p_05_vac_infected = all_degs_p_05_vac_infected_19_12_23 # DEGs
ann_vaccines_covid_other = ann_vaccines_vaxsig_covid #Conditions annotation
VAX_Genes_Annotated_RAW_2 = VAX_Genes_Annotated_RAW # VAX MSigDB DEGs 


# Processing
VAX_Genes_Annotated_RAW_3 = VAX_Genes_Annotated_RAW_2 %>% 
  clean_names() %>% 
  filter(study_subtype == "VAC ONLY",
         sample_source == "PBMC") %>% 
  mutate(condition = paste0(vaccine, " (", date_time, ")")) %>% 
  select(condition, everything())

degs_all_updown_vax = VAX_Genes_Annotated_RAW_3 %>%
  clean_names() %>% 
  mutate(condition = paste0(vaccine, " (", date_time, ")")) %>% 
  select(condition, genes = gene) %>% 
  distinct()

degs_all_updown_covid= degs_updown_p_05_vac_infected %>% 
  as.data.frame() %>% 
  select(condition:direction) %>% 
  filter(direction %in% c("UP", "DOWN")) %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  distinct() %>% 
  select(condition, genes)

degs_all_updown_covid_other = bind_rows(degs_all_updown_vax, degs_all_updown_covid) %>% 
  distinct()

degs_all_updown_covid_other = degs_all_updown_covid_vax_other

degs_all_updown_covid_other_immune = degs_all_updown_covid_other %>% 
  inner_join(ImmuneGO_Annotated_Genes %>% select(genes) %>% distinct(), by = "genes") %>% 
  distinct()

degs_all_updown_covid_other_nonimmune = degs_all_updown_covid_other %>% 
  anti_join(ImmuneGO_Annotated_Genes %>% select(genes) %>% distinct(), by = "genes") %>% 
  distinct()

write.csv(degs_all_updown_covid_other, file = "degs_all_updown_covid_vax_other.csv")
write.csv(degs_all_updown_covid_other_immune, file = "degs_all_updown_covid_vax_other_immune.csv")
write.csv(degs_all_updown_covid_other_nonimmune, file = "degs_all_updown_covid_vax_other_nonimmune.csv")




degs_covid = all_degs_p_05_vac_infected_19_12_23 %>% 
  select(condition, genes, direction) %>% 
  filter(direction != "NEUTRAL") %>% 
  mutate(regulation_numeric = if_else(direction == "UP", "1", "-1"),
         regulation_numeric = as.numeric(regulation_numeric))


degs_vax = VAXSigDB_degs_all %>% 
  filter(STUDY_SUBTYPE == "VAC ONLY") %>% 
  mutate(condition_vaccine = paste0(VACCINE, " (", `DATE-TIME`, ")")) %>% 
  select(condition_vaccine, SYSTEMATIC_NAME, VACCINE, everything()) %>% 
  separate_rows(GENE_SYMBOLS, sep = ",")
degs_vax %>% saveRDS(file = "VAXSigDB_Genes.rds")


degs_vaxsig = degs_vax %>% 
  filter(STUDY_SUBTYPE == "VAC ONLY") %>% 
  select(condition = condition_vaccine, genes = GENE_SYMBOLS, regulation = REGULATION) %>% 
  mutate(regulation_numeric = if_else(regulation == "UP", 1, -1))


degs_covid_vax = degs_covid %>% 
  bind_rows(degs_vaxsig) %>% 
  select(-regulation) %>% 
  inner_join(ImmuneGO_Annotated_Genes %>%
               filter(process != c("TCR REPERTOIRE", "BCR REPERTOIRE")) %>% 
               select(genes) %>%
               distinct(), 
             by = "genes")


degs_covid_vax %>% 
  saveRDS(file = here("VaxGO gene sets", "degs_covid_vaxsig_up_down_immune.rds"))

```

```{r}
#Input
data = degs_all_updown_covid_other_nonimmune %>% 
  rename(geneset = condition) %>% 
  select(geneset, genes)

filename = "degs_all_updown_covid_other_nonimmune"
  
```

```{r}
# Função para calcular a sobreposição entre pares de vacinas e a porcentagem de genes compartilhados
overlap_genes <- function(cond1, cond2, data) {
  genes_cond1 <- data$genes[data$geneset == cond1] #Trocar coluna
  genes_cond2 <- data$genes[data$geneset == cond2] #Trocar coluna
  genes_shared <- intersect(genes_cond1, genes_cond2)
  genes_notshared_cond1 <- setdiff(genes_cond1, genes_cond2)
  genes_notshared_cond2 <- setdiff(genes_cond2, genes_cond1)
  
  total_genes_cond1 <- length(genes_cond1)
  total_genes_cond2 <- length(genes_cond2)
  
  percentage_shared_cond1 <- length(genes_shared) / total_genes_cond1 * 100
  percentage_shared_cond2 <- length(genes_shared) / total_genes_cond2 * 100
  
  shared_genes <- data.frame(
    Cond1 = cond1,
    Cond2 = cond2,
    Shared = length(genes_shared),
    NotShared_cond1 = length(genes_notshared_cond1),
    NotShared_cond2 = length(genes_notshared_cond2),
    Total_Genes_Cond1 = total_genes_cond1,
    Total_Genes_Cond2 = total_genes_cond2,
    Genes_Names = paste(genes_shared, collapse = ","),
    Percentage_Shared_Cond1 = percentage_shared_cond1,
    Percentage_Shared_Cond2 = percentage_shared_cond2
  )
  
  return(shared_genes)
}
# Obtenha uma lista de todas as vacinas únicas
unique_cond <- unique(data$geneset)
shared_genes_df <- data.frame()

for (i in 1:(length(unique_cond) - 1)) {
  for (j in (i + 1):length(unique_cond)) {
    resultado_temp <- overlap_genes(unique_cond[i], unique_cond[j], data)
    shared_genes_df <- bind_rows(shared_genes_df, resultado_temp)
  }
}

# Fisher's exact test and Benjamini-Hochberg multiple tests
fisher_exact_test <- function(shared, notshared1, notshared2) {
  cont_table <- matrix(c(shared, notshared1, notshared2, 0), nrow = 2)
  results_fisher <- fisher.test(cont_table)
  return(results_fisher$p.value)
}

shared_genes_df$pvalue <- mapply(fisher_exact_test, 
                                    shared_genes_df$Shared, 
                                    shared_genes_df$NotShared_cond1, 
                                    shared_genes_df$NotShared_cond2)

shared_genes_df <- shared_genes_df %>%
  filter(pvalue <= 1) %>% 
  mutate(qvalue = qvalue(pvalue)$pvalue)

#Duplicar e trocar colunas (Vacina 1 e Vacina 2)
shared_genes_df2 = shared_genes_df
shared_genes_df2[, c("Cond1", "Cond2")] <- shared_genes_df2[, c("Cond2", "Cond1")]
shared_genes_df_mirror = rbind(shared_genes_df, shared_genes_df2)

shared_genes_df_mirror[is.na(shared_genes_df_mirror)] <- 0
shared_genes_df_mirror<- replace(shared_genes_df_mirror, shared_genes_df_mirror == "Inf", 0)

shared_genes_df_mirror = shared_genes_df_mirror %>%
  mutate(Percentage_Shared_Cond1 = round(Percentage_Shared_Cond1, 2),
         Percentage_Shared_Cond2 = round(Percentage_Shared_Cond2, 2))

#Salvar resultados
write.csv(shared_genes_df_mirror, file = paste0(filename, "_shared.csv"))

```

### Other comparisons

```{r}
############# VACCINE
#Use "vax_degs_up_VAXSigDB_sharedgenes_ann.csv" for comparing UP degs with UP degs from other conditions

#VAC ONLY
shared_genes_ann_vac = shared_genes_df_mirror_ann %>%
  select(vac_condition, vaccine_condition, vac_ref, Cond2:sample_source, pmid, gene_symbols:participants) %>% 
  filter(study_type == "VACCINE", 
         sample_source == "PBMC",
         !(vaccine_condition %in% c("BBIBP", "ZF2001"))) 

# vac_only_up = shared_genes_ann_vac %>%
#   filter(study_subtype == "VAC ONLY",
#          regulation == 'UP')

vac_only_down = shared_genes_ann_vac %>%
  filter(study_subtype == "VAC ONLY",
         regulation == 'DOWN')

#RESPONDER

vac_responder_up = shared_genes_ann_vac %>%
  filter(study_subtype == "RESPONDER VS NON RESPONDER",
         regulation == 'UP')

vac_responder_down = shared_genes_ann_vac %>% 
  filter(study_subtype == "RESPONDER VS NON RESPONDER",
         regulation == 'DOWN')

# CELLULAR RESPONSE

vac_responder_cell_up = shared_genes_ann_vac %>% 
  filter(study_subtype == "CELL",
         regulation == 'UP')

vac_responder_cell_down = shared_genes_ann_vac %>% 
  filter(study_subtype == "CELL",
         regulation == 'DOWN')

############# CORRELATION
shared_genes_ann_corr = shared_genes_df_mirror_ann %>% 
  select(vac_condition, vaccine_condition, vac_ref, Cond2:sample_source, pmid, gene_symbols:previous_vaccinationarticipants) %>% 
  filter(study_type == "CORRELATION", 
         sample_source == "PBMC",
         !(vaccine_condition %in% c("BBIBP", "ZF2001")))

# ADVERSE EVENT
corr_AE_up = shared_genes_ann_corr %>%
  filter(study_subtype == "ADVERSE EVENT",
         regulation == 'UP')

corr_AE_down = shared_genes_ann_corr %>%
  filter(study_subtype == "ADVERSE EVENT",
         regulation == 'DOWN')

# ANTIBODY
corr_antibody_up = shared_genes_ann_corr %>%
  filter(study_subtype == "ANTIBODY",
         regulation == 'UP')

corr_antibody_down = shared_genes_ann_corr %>% 
  filter(study_subtype == "ANTIBODY",
         regulation == 'POSITIVE')

corr_antibody_positive = shared_genes_ann_corr %>%
  filter(study_subtype == "ANTIBODY",
         regulation == 'NEGATIVE')

corr_antibody_negative = shared_genes_ann_corr %>% 
  filter(study_subtype == "ANTIBODY",
         regulation == 'NEGATIVE')


# CELL
corr_cell_up = shared_genes_ann_corr %>%
  filter(study_subtype == "CELL",
         regulation == 'UP')

corr_cell_down = shared_genes_ann_corr %>% 
  filter(study_subtype == "CELL",
         regulation == 'POSITIVE')

corr_cell_positive = shared_genes_ann_corr %>% 
  filter(study_subtype == "CELL",
         regulation == 'UP')

corr_cell_negative = shared_genes_ann_corr %>% 
  filter(study_subtype == "CELL",
         regulation == 'NEGATIVE')

```

### Heatmap

```{r}
table = degs_all_updown_covid_other_immune_shared
filename = "ImmuneGO_Genes" 

# table = degs_all_updown_covid_other_nonimmune_shared
# filename = "NonImmune_Genes" 

# INPUT ------
data = table %>% 
  filter(qvalue <0.25) %>% 
  inner_join(ann_vaccines_covid_other %>% 
               select(Cond1 = condition, everything()), by = "Cond1") %>% 
  inner_join(ann_vaccines %>% 
               select(Cond2 = condition, everything()), by = "Cond2") %>% 
  clean_names() %>% 
  rename(vac_ref = cond1, vac_condition = cond2)

# Matrix ------
matrix = data %>%
  filter(qvalue <= 0.10) %>% 
  select(vac_condition, vac_ref, shared) %>% 
  pivot_wider(names_from = vac_condition, values_from = shared, values_fn = mean) %>% 
  replace(is.na(.), 0) %>%
  column_to_rownames(var="vac_ref") %>%
  as.matrix() %>%
  round(0)

# Annotations -------
ann_vaccines_covid = ann_vaccines_covid_other %>% 
  filter(covid_other == "COVID")
ann_vaccines_other = ann_vaccines_covid_other %>% 
  filter(covid_other == "OTHER")

# Columns
ann_cols_heatmap = matrix %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "process") %>% 
  rename(vac_condition = '.') %>% 
  merge(ann_vaccines_covid, by.x = "vac_condition", by.y = "condition", all.y = F) %>% 
  select(vac_condition, type, vaccine, week) %>% #ordenar colunas para anotação
  distinct() %>% 
  column_to_rownames(var = "vac_condition") %>% 
  mutate_all(as.factor)

# Rows
ann_rows_heatmap = matrix %>%  
  as.data.frame() %>% 
  rownames_to_column("condition") %>% 
  distinct() %>% 
  inner_join(ann_vaccines_other, by = "condition") %>% 
  mutate_all(as.factor) %>%
  arrange(condition) %>%
  distinct() %>% 
  column_to_rownames(var= "condition") %>% 
  select(vaccine:infection) %>% 
  select(microbe_type, type, week)

# Arrange cols and rows in matrix
matrix_data = matrix %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>%
  rownames_to_column("condition") %>% 
  inner_join(ann_rows_heatmap %>% 
               rownames_to_column("condition") %>% 
               select(condition),
             by = "condition") %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  mutate_if(is.character, as.numeric) %>%
  replace(is.na(.), 0) %>% 
  as.matrix()


#Verificar dimensões do dataset
dim(matrix_data)
dim(ann_rows_heatmap)
dim(ann_cols_heatmap)

```

```{r}

#Heatmap annotation
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "left")

row_ha = rowAnnotation(df =  ann_rows_heatmap, 
                       col = ann_vaxsig_colors)

# Values colors
col_fun <- colorRamp2(c(0, max(matrix_data)), c("white", "#ed6a5a"))


mat = matrix_data
width_image = 40
height_image = 30
width = unit(20, "cm")
height = unit(15, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)
file = paste0("VAXSig_", filename_2)

fileimageheatmap_grouped = paste0(filename, sep ="_", "Complexheatmap_ALL_CLUSTERED.png")

png(file = fileimageheatmap_grouped, 
    width = width_image, 
    height = height_image,
    units="cm", res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "left",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(title = "Shared", direction = "vertical"),
                        col = col_fun, #color
                        cell_fun = function(j, i, x, y, width, height, fill) {
                                            if(mat[i, j] > 0)
                                            grid.text(sprintf("%.f", mat[i, j]), 
                                                      x, y, 
                                                      gp = gpar(fontsize = 8))},
                        row_title = "",
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 90,
                        column_names_gp = column_names_gp,
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp, #Grande = 6, Pequeno 15
                        row_dend_width = unit(5, "cm"), #Altura do dendrograma
                        column_split = NULL, #Split column clusters 
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, # 20 para pequeno e grande
                        height = height # 20 para pequeno, 60 para grande
)

draw(heatmap_plot,
     merge_legend = TRUE,
     annotation_legend_side = "right", 
     heatmap_legend_side = "right")
dev.off()

```

### ORA COVID vs. VAXGO com ImmuneGO

*Qual a relação entre os genes de cada dataset?*

```{r}
#Input
degs_all_updown_covid = degs_updown_p_05_vac_infected %>% 
   as.data.frame() %>% 
   select(condition:direction) %>% 
   filter(direction %in% c("UP", "DOWN")) %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
   distinct() %>% 
  select(condition, genes)

VAX_Genes_Annotated_RAW = VAX_Genes_Annotated_RAW %>% 
  clean_names() %>% 
  mutate(condition = paste0(vaccine, " (", date_time, ")"))

ann_vaccines_covid_other 

degs_all_updown_vax_immunego = VAX_Genes_Annotated_RAW %>%
  clean_names() %>% 
  mutate(condition = paste0(vaccine, " (", date_time, ")")) %>% 
  select(condition, genes = gene) %>% 
  inner_join(ImmuneGO_Annotated_Genes_8_1_24 %>% 
               select(genes), by = "genes") %>% 
  distinct()

degs_all_updown_covid_other = bind_rows(degs_all_updown_vax_immunego, degs_all_updown_covid) %>% 
  distinct()

Immune_GO_T2G = ImmuneGO_Annotated_Genes %>% 
  filter(go_term == "Manual",
         !process %in% c("TCR REPERTOIRE", "BCR REPERTOIRE", "ADAPTIVE IMMUNE SYSTEM", "INNATE IMMUNE SYSTEM")) %>% 
  select(process, genes)
```

```{r}
# Função para realizar a análise GSEA para uma condição específica
perform_ORA <- function(df, Immune_GO_T2G) {
  resultados <- list()
  
  # Obter condições únicas na coluna "condition"
  condicoes <- df$condition %>% 
    unique() %>% 
    as.character()
  
  for (condicao in condicoes) {
    # Selecionar DEGs para a condição atual
    degs_condicao <- df %>% 
      filter(condition == condicao) %>%
      select(genes) %>%
      distinct()

    # Preparar gene list
    gene_list_ora <- degs_condicao$genes

    ############ IMMUNE GO
    immune_GO_ORA <- tryCatch({
      enricher(gene_list_ora, 
               TERM2GENE = Immune_GO_T2G, 
               minGSSize = 1,
               maxGSSize = 1000,
               pvalueCutoff = 0.25,
               pAdjustMethod = "BH") %>% 
        as.data.frame() %>% 
        arrange(qvalue) %>% 
        mutate(condition = condicao,
               ora_enrichment = "ImmuneGO")
    }, error = function(e) {
      return(NULL)  # Retorna NULL caso ocorra o erro
    })

    if (!is.null(immune_GO_ORA)) {
      resultados[[paste(condicao, "ImmuneGO", sep = "_")]] <- immune_GO_ORA
    }
  }

  # Combinar todos os resultados em um único dataframe
  resultado_final <- bind_rows(resultados)

  return(resultado_final)
}

df_resultados <- perform_ORA(degs_all_updown_covid_other, Immune_GO_T2G)
print(df_resultados)

#Anotar
immune_GO_ORA_covid_other = df_resultados %>% 
  clean_names() %>% 
  select(process = id, gene_ratio:ora_enrichment) %>% 
  inner_join(ann_vaccines_covid_other, by = "condition") %>% 
  distinct() %>% 
  mutate(log_q = -log10(qvalue))

write.csv(immune_GO_ORA_covid_other, file = "COVID_VAX_immune_GO_ORA.csv")

view(immune_GO_ORA_covid_other)
```

#### Heatmap

```{r}
####### INPUT
data_2 = immune_GO_ORA_covid_other
  
filename_2 = "immune_GO_ORA_covid_other"

######## Matrix
matrix = data_2 %>% 
  filter(qvalue <= 0.25) %>% 
  select(condition, process, log_q) %>% #Percentage_Shared_Cond1
  pivot_wider(names_from = "condition", 
              values_from = "log_q",  #Percentage_Shared_Cond1
              values_fn = mean) %>% 
  replace(is.na(.), 0) %>% 
  column_to_rownames(var="process") %>% 
  as.matrix() %>% 
  round(2)

######## Annotations
ann_cols = matrix %>% 
  as.data.frame() %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column("condition") %>% 
  select(condition) %>% 
  inner_join(ann_vaccines_covid_other, by = "condition") %>% 
  select(condition, covid_other, microbe_type, disease_vac, type, week) %>% 
  distinct() %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") 

matrix_data = matrix %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame()%>%  
  mutate_if(is.character, as.numeric) %>%
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar dimensões do dataset
dim(matrix_data)
dim(ann_cols)
```

```{r}
################# Plotar

#Heatmap annotation
ha = HeatmapAnnotation(df = ann_cols, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "right")


# Values colors
col_fun <- colorRamp2(c(0, 2), c("white", "#ED6A5A"))


# Parameters
file = filename_2
mat = matrix_data
width_image = 50
height_image = 20
width = unit(25, "cm")
height = unit(7, "cm")
column_names_gp = gpar(fontsize = 9)
row_names_gp = gpar(fontsize = 9)
rect_gp = gpar(col = "gray90", lwd = 0.5)

###### CLUSTERIZED ALL VS ALL

fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap.png")

png(file = fileimageheatmap_grouped, width = width_image, height = height_image, units="cm", res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        row_names_side = "left",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical",
                                                    title = "-log(FDR)"),
                        col = col_fun, #color
                        cell_fun = function(j, i, x, y, w, h, fill) {
                              if (mat[i, j] >= 2) {
                                grid.text("***", x, y, gp = gpar(fontsize = 8))
                              } else if (mat[i, j] > 1.5) {
                                grid.text("**", x, y, gp = gpar(fontsize = 9))
                              } else if (mat[i, j] >= 1) {
                                grid.text("*", x, y, gp = gpar(fontsize = 8))
                              } else {
                                grid.text("", x, y)
                              }
                            },
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp, #Grande = 6, Pequeno 15
                        row_dend_width = unit(5, "cm"), #Altura do dendrograma
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        width = width, # 20 para pequeno e grande
                        height = height # 20 para pequeno, 60 para grande
)

draw(heatmap_plot, 
     annotation_legend_side = "left", 
     heatmap_legend_side = "left",
          merge_legend = TRUE)
dev.off()


```

## GENES COVID vs. VAXGO com ImmuneGO

*Qual a relação entre os genes de cada dataset?*

```{r}
# Gene sets for annotation
OtherGOs_Annotated_Genes = OtherGOs_Annotated_Genes %>% 
  filter(annotation == "Major process",
         term != "Immune system") %>% 
    anti_join(ImmuneGO_Annotated_Genes %>% 
                select(genes) %>% distinct(), 
              by = "genes")

# VAX Genes  ------
VAX_Genes_Annotated_RAW = VAX_Genes_Annotated_RAW %>% 
  clean_names() %>% 
  mutate(condition = paste0(vaccine, " (", date_time, ")"))

#ImmuneGO
degs_all_updown_vax_immunego = VAX_Genes_Annotated_RAW %>%
  clean_names() %>% 
  mutate(condition = paste0(vaccine, " (", date_time, ")")) %>% 
  select(condition, genes = gene, regulation) %>% 
  inner_join(ImmuneGO_Annotated_Genes %>% 
               select(genes), by = "genes") %>% 
  distinct()

# Not Immune
degs_all_updown_vax_notimmune = VAX_Genes_Annotated_RAW %>%
  clean_names() %>% 
  mutate(condition = paste0(vaccine, " (", date_time, ")")) %>% 
  select(condition, genes = gene, regulation) %>% 
  inner_join(OtherGOs_Annotated_Genes %>% 
               select(genes), by = "genes") %>% 
  distinct()

# Covid-19 conditions ----
degs_all_updown = degs_updown_p_05_vac_infected

#Immune
degs_all_updown_covid_immune = degs_all_updown %>% 
  as.data.frame() %>% 
  select(condition:padj) %>% 
  mutate(direction = case_when(log2fold_change >= 1 ~ "UP",
                               log2fold_change <= -1 ~ "DOWN",
                               TRUE ~ "NEUTRAL")) %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  inner_join(ImmuneGO_Annotated_Genes %>% select(genes) %>% distinct(), by = "genes") %>% 
  distinct() %>% 
  filter(vaccine != "I", 
         gse_id != "GSE189039") %>% 
  select(condition, genes, regulation = direction)  %>% 
  distinct()

# Not immune
degs_all_updown_covid_notimmune = degs_all_updown %>% 
  as.data.frame() %>% 
  select(condition:padj) %>% 
    mutate(direction = case_when(log2fold_change >= 1 ~ "UP",
                               log2fold_change <= -1 ~ "DOWN",
                               TRUE ~ "NEUTRAL")) %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  inner_join(OtherGOs_Annotated_Genes %>% select(genes) %>% distinct(), by = "genes") %>% 
  distinct() %>% 
  filter(vaccine != "I", 
         gse_id != "GSE189039") %>% 
  # select(condition, genes, regulation = direction)  %>% 
  distinct()


# Bind Immune ------
degs_all_updown_covid_other = bind_rows(degs_all_updown_vax_immunego, degs_all_updown_covid_immune) %>% 
  distinct()


covid_other_immunego_genes = degs_all_updown_covid_other %>% 
  distinct() %>% 
  filter(regulation %in% c("UP", "DOWN")) %>% 
  mutate(regulation_numeric = case_when(
    regulation %in% c("UP") ~ 1,
    regulation %in% c("DOWN") ~ -1,
    TRUE ~ as.numeric(regulation)
  )) %>% 
  distinct()

saveRDS(covid_other_immunego_genes, file = "covid_other_immunego_genes.rds")



# Bind not-immune ----
degs_all_updown_covid_other_notimmune = bind_rows(degs_all_updown_vax_notimmune, degs_all_updown_covid_notimmune) %>% 
  distinct()

covid_other_notimmune_genes = degs_all_updown_covid_other_notimmune %>% 
  anti_join(ImmuneGO_Annotated_Genes %>% select(genes) %>% distinct(), by = "genes") %>% 
  inner_join(OtherGOs_Annotated_Genes %>% select(process = term, genes) %>% filter(), by = "genes") %>% 
  distinct() %>% 
  filter(regulation %in% c("UP", "DOWN")) %>% 
  mutate(regulation_numeric = case_when(
    regulation %in% c("UP") ~ 1,
    regulation %in% c("DOWN") ~ -1,
    TRUE ~ as.numeric(regulation))) %>% 
  distinct()

saveRDS(covid_other_notimmune_genes, file = "covid_other_notimmune_genes.rds")


```

```{r}
###### Gene sets Immune GO ------

Immune_GO_genes_All = covid_other_immunego_genes 

Immune_GO_genes_Innate = Immune_GO_genes_All %>%
   filter(process == "INNATE IMMUNE SYSTEM")
Immune_GO_genes_Adaptive = Immune_GO_genes_All %>%
   filter(process == "ADAPTIVE IMMUNE SYSTEM")
Immune_GO_genes_Complement = Immune_GO_genes_All %>%
   filter(process == "COMPLEMENT IMMUNE SYSTEM")

# Sistema inato
Immune_GO_genes_Inflammation = Immune_GO_genes_All %>%
  filter(process == "INFLAMMATION")
Immune_GO_genes_ARPP = Immune_GO_genes_All %>%
  filter(process == "ARPP")
Immune_GO_genes_IFN = Immune_GO_genes_All %>%
  filter(process == "ANTIVIRAL AND INTERFERON")
Immune_GO_genes_Neutrophil = Immune_GO_genes_All %>%
  filter(process == "NEUTROPHIL")
Immune_GO_genes_Eosinophil = Immune_GO_genes_All %>%
  filter(process == "EOSINOPHIL")
Immune_GO_genes_Complement = Immune_GO_genes_All %>%
  filter(process == "COMPLEMENT IMMUNE SYSTEM")
Immune_GO_genes_Monocytes = Immune_GO_genes_All %>%
  filter(process == "MONOCYTES")
Immune_GO_genes_DC = Immune_GO_genes_All %>%
  filter(process == "DENDRITIC CELLS")
Immune_GO_genes_Macro = Immune_GO_genes_All %>%
  filter(process == "MACROPHAGES")
Immune_GO_genes_NK = Immune_GO_genes_All %>%
  filter(process == "NK CELL")

# Sistema adaptativo
Immune_GO_genes_Cellular = Immune_GO_genes_All %>%
  filter(process == "CELLULAR ADAPTIVE IMMUNE SYSTEM")
Immune_GO_genes_Humoral = Immune_GO_genes_All %>%
  filter(process == "HUMORAL ADAPTIVE IMMUNE SYSTEM")
Immune_GO_genes_Tcells = Immune_GO_genes_All %>%
  filter(process == "T CELLS")
Immune_GO_genes_Bcells = Immune_GO_genes_All %>%
  filter(process == "B CELLS")
Immune_GO_genes_Ig = Immune_GO_genes_All %>%
  filter(process == "IMMUNOGLOBULIN MEDIATED IMMUNE RESPONSE")

```

#### Heatmap

```{r}
####### INPUT
data_2 = Immune_GO_genes_All
filename_2 = paste0("Immune_GO_genes_All", "_VAX_COVID")

######## Matrix
matrix = data_2 %>% 
  select(genes, condition, regulation_numeric) %>%
  pivot_wider(names_from = "condition", 
              values_from = "regulation_numeric", 
              values_fn = mean) %>% 
  replace(is.na(.), 0) %>%
  arrange(genes) %>% 
  column_to_rownames(var="genes") %>% 
  as.matrix()

######## Annotations
ann_vaccines_covid_other_2 = data_2 %>% 
  inner_join(ann_vaccines_covid_other, by = "condition") %>% 
  select(condition, covid_other, microbe_type, disease_vac, type, week) %>% 
  distinct()

#Columns
ann_cols_heatmap = matrix %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  inner_join(ann_vaccines_covid_other_2, by = "condition") %>% 
  distinct() %>% 
  arrange(condition) %>% 
  column_to_rownames("condition")

matrix_data = matrix %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  inner_join(ann_cols_heatmap %>% rownames_to_column("condition"), by = "condition") %>% 
  select(!c(covid_other:week)) %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  filter(rowSums(. != 0) > 1) %>% 
  mutate_if(is.character, as.numeric) %>%
  as.matrix() 

#Rows Immune ----
ann_rows_immune = matrix_data %>%
  as.data.frame() %>%
  rownames_to_column("genes") %>%
  left_join(ImmuneGO_Annotated_Genes %>%
              filter(go_term == "Manual",
                     !process %in% c("TCR REPERTOIRE", "BCR REPERTOIRE")), by = "genes") %>%
  select(genes, process) %>%
  distinct() %>%
  group_by(genes) %>%
  arrange(genes, factor(process, levels = c("INNATE IMMUNE SYSTEM", "ADAPTIVE IMMUNE SYSTEM")), .by_group = TRUE) %>%
  mutate(i_process = row_number()) %>%
  pivot_wider(., names_from = "i_process", values_from = "process") %>%
  column_to_rownames("genes") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("index") %>%
  mutate(index = as.numeric(index)) %>%
  arrange(desc(index)) %>%
  column_to_rownames("index") %>%
  t() %>%
  as.data.frame() %>%
  replace(is.na(.), ".")


# Rows Not immune -----
# ann_rows_notimmune = matrix_data %>%
#   as.data.frame() %>%
#   rownames_to_column("genes") %>%
#   left_join(ann_genes %>%
#                select(process = term, genes = symbol, annotation) %>%
#                filter(annotation == "Major process",
#                       process != "Immune system"), by = "genes") %>%
#   select(genes, process) %>%
#   distinct() %>%
#   group_by(genes) %>%
#   mutate(i_process = row_number()) %>%
#   pivot_wider(., names_from = "i_process", values_from = "process") %>%
#   column_to_rownames("genes") %>%
#   t() %>%
#   as.data.frame() %>%
#   rownames_to_column("index") %>%
#   arrange(desc(index)) %>%
#   column_to_rownames("index") %>%
#   t() %>%
#   as.data.frame() %>%
#   replace(is.na(.), ".")


#Verificar dimensões do dataset
dim(matrix_data)
dim(ann_cols_heatmap)
# dim(ann_rows_notimmune)
dim(ann_rows_immune)

```

Sizes: 1. Cellular adaptive immune system: width = unit(25, "cm"),
height = unit(60, "cm") #254 genes 2. Humoral adaptive immune system:
width = unit(25, "cm"), height = unit(30, "cm") #51 genes 3. T cells:
width = unit(25, "cm"), height = unit(60, "cm") #221 genes 4. B cells:
width = unit(25, "cm"), height = unit(50, "cm") #146 genes 5. Ig
mediated: width = unit(25, "cm"), height = unit(10, "cm") #146 genes 6.
Inflammation: width = unit(25, "cm"), height = unit(10, "cm") 7. ARPP:
width = unit(25, "cm"), height = unit(30, "cm") 60 genes 8. IFN: width =
unit(25, "cm"), height = unit(30, "cm") 93 genes 9. DC: width = unit(25,
"cm"), height = unit(20, "cm") #29 genes 10. Macrophages: width =
unit(25, "cm"), height = unit(25, "cm") \# 52 genes 11. Neutrophils:
width = unit(25, "cm"), height = unit(40, "cm") \# 70 genes 12.
Eosinophil: width = unit(25, "cm"), height = unit(10, "cm") \# 42 genes
13. Complement: width = unit(25, "cm"), height = unit(10, "cm") \# 40
genes 13. Monocytes: width = unit(25, "cm"), height = unit(15, "cm") \#
48 genes 15. All immune genes: width_image = 60, height_image = 70,
width = unit(30, "cm"), height = unit(60, "cm"), row_split = 7,
column_split = 2

More refs:
<https://debrowser.readthedocs.io/en/master/heatmap/heatmap.html>

### All genes

```{r}
################# Immune 
#Heatmap annotation
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "left")

row_ha = HeatmapAnnotation(df = ann_rows_immune,
                       col = col_process_immune,
                       which= "row",
                       show_annotation_name = F,
                       show_legend = F)

# Values colors
col_fun <- colorRamp2(c(-1, 0, 1), c("#9bc1bc", "white", "#ed6a5a"))

file = filename_2
mat = matrix_data
width_image = 60
height_image = 50
width = unit(30, "cm")
height = unit(40, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 0)
rect_gp = gpar(col = "gray80", lwd = 0)
row_split = 2
column_split = 2


fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap.png")

png(file = fileimageheatmap_grouped, 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        col = col_fun, #color
                        name = "L2FC",
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(5, "cm"),
                        row_split = row_split,
                        column_split = column_split,
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "right",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        clustering_distance_columns = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height,
                        use_raster = F)

heatmap_plot = draw(heatmap_plot)

dev.off()

```

Get row clusters

```{r}
r.dend <- row_dend(heatmap_plot)  #Extract row dendrogram
rcl.list <- row_order(heatmap_plot)  #Extract clusters (output is a list)
lapply(rcl.list, function(x) length(x))  #check/confirm size cluster

gene_cluster <- NULL
for (i in 1:length(rcl.list)) {
  a <- row.names(mat[row_order(heatmap_plot)[[i]], ])
  gene_cluster <- rbind(gene_cluster, data.frame(genes = a, cluster = paste0("cluster ", i), gene_set = "Immune genes"))
}
write.csv(gene_cluster, file = paste0(file, "_clustered_genes.csv"), row.names = F)
```

### Cluster 1 Immune genes

```{r}
####### INPUT
data_2 = Immune_GO_genes_All %>% 
  inner_join(gene_cluster %>% filter(cluster == "cluster 1") %>% select(genes), by = "genes")
filename_2 = paste0("Immune_GO_genes_cluster1", "_VAX_COVID")
```

```{r}
######## Matrix
matrix = data_2 %>% 
  select(genes, condition, regulation_numeric) %>%
  pivot_wider(names_from = "condition", 
              values_from = "regulation_numeric", 
              values_fn = mean) %>% 
  replace(is.na(.), 0) %>%
  arrange(genes) %>% 
  column_to_rownames(var="genes") %>% 
  as.matrix()

######## Annotations
ann_vaccines_covid_other_2 = data_2 %>% 
  inner_join(ann_vaccines_covid_other, by = "condition") %>% 
  select(condition, covid_other, microbe_type, disease_vac, type, week) %>% 
  distinct()

#Columns
ann_cols_heatmap = matrix %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  inner_join(ann_vaccines_covid_other_2, by = "condition") %>% 
  distinct() %>% 
  arrange(condition) %>% 
  column_to_rownames("condition")

matrix_data = matrix %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  inner_join(ann_cols_heatmap %>% rownames_to_column("condition"), by = "condition") %>% 
  select(!c(covid_other:week)) %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>%
  as.matrix() 

#Rows Immune ----
ann_rows_immune = matrix_data %>%
  as.data.frame() %>%
  rownames_to_column("genes") %>%
  left_join(ImmuneGO_Annotated_Genes %>% filter(go_term == "Manual"), by = "genes") %>%
  select(genes, process) %>%
  distinct() %>%
  group_by(genes) %>%
  arrange(genes, factor(process, levels = c("INNATE IMMUNE SYSTEM", "ADAPTIVE IMMUNE SYSTEM")), .by_group = TRUE) %>%
  mutate(i_process = row_number()) %>%
  pivot_wider(., names_from = "i_process", values_from = "process") %>%
  column_to_rownames("genes") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("index") %>%
  mutate(index = as.numeric(index)) %>%
  arrange(desc(index)) %>%
  column_to_rownames("index") %>%
  t() %>%
  as.data.frame() %>%
  replace(is.na(.), ".") %>% 
  mutate(`1` = if_else(`1` == ".", "INNATE IMMUNE SYSTEM", `1`))


#Verificar dimensões do dataset
dim(matrix)
dim(matrix_data)
dim(ann_cols_heatmap)
dim(ann_rows_immune)

```

More refs:
<https://debrowser.readthedocs.io/en/master/heatmap/heatmap.html>

```{r}
################# Immune 
#Heatmap annotation
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "right")

row_ha = row_ha = HeatmapAnnotation(df = ann_rows_immune,
                       col = col_process_immune,
                       which= "row",
                       show_annotation_name = F,
                       show_legend = F)

# Values colors
col_fun <- colorRamp2(c(-1, 0, 1), c("#9bc1bc", "white", "#ed6a5a"))

file = filename_2
mat = matrix_data
width_image = 50
height_image = 50
width = unit(30, "cm")
height = unit(30, "cm")
column_names_gp = gpar(fontsize = 12)
row_names_gp = gpar(fontsize = 12)
rect_gp = gpar(col = "gray80", lwd = 0.5)

fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap.png")

png(file = fileimageheatmap_grouped, 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                       left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        col = col_fun, #color
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                       name = "L2FC",
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(0, "cm"),
                        row_split = NULL,
                        column_split = 2,
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height)

heatmap_plot = draw(heatmap_plot, 
     annotation_legend_side = "left", 
     heatmap_legend_side = "left",
     merge_legend = TRUE)

heatmap_plot

dev.off()

```

### Inflammation

```{r}
################# Immune 
#Heatmap annotation
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "left")

row_ha = HeatmapAnnotation(df = ann_rows_immune,
                       col = col_process_immune,
                       which= "row",
                       show_annotation_name = F,
                       show_legend = F)

# Values colors
col_fun <- colorRamp2(c(-1, 0, 1), c("#9bc1bc", "white", "#ed6a5a"))

file = filename_2
mat = matrix_data
width_image = 60
height_image = 50
width = unit(30, "cm")
height = unit(40, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 0)
rect_gp = gpar(col = "gray80", lwd = 0)
row_split = 2
column_split = 2


fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap.png")

png(file = fileimageheatmap_grouped, 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        col = col_fun, #color
                        name = "L2FC",
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(5, "cm"),
                        row_split = row_split,
                        column_split = column_split,
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "right",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        clustering_distance_columns = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height,
                        use_raster = F)

heatmap_plot = draw(heatmap_plot)

dev.off()

```

#### Not immune

```{r}
################# NOT Immune 
#Heatmap annotation
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "right")

row_ha = HeatmapAnnotation(df = ann_rows_notimmune,
                       col = col_process_notimmune,
                       which= "row",
                       show_annotation_name = F,
                       show_legend = F)

# Values colors
col_fun <- colorRamp2(c(-1, 0, 1), c("#9bc1bc", "white", "#ed6a5a"))

file = filename_2
mat = matrix_data
width_image = 60
height_image = 50
width = unit(30, "cm")
height = unit(40, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 0)
rect_gp = gpar(col = "gray80", lwd = 0)
row_split = 7

fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap.png")

png(file = fileimageheatmap_grouped, 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        col = col_fun, #color
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(5, "cm"),
                        row_split = row_split,
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = T,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        clustering_distance_columns = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height,
                        use_raster = F)

heatmap_plot = draw(heatmap_plot)

heatmap_plot

dev.off()
```

# Random forest genes

## 50 most shared genes

```{r}
#### Immune -----
ann_vaccines_covid_other = ann_vaccines_vaxsig_covid

stats_1 = degs_all_updown_covid_vax_other_immune %>% 
  inner_join(ann_vaccines_covid_other, by = "condition") %>% 
  group_by(genes, covid_other) %>% 
  summarise(n_conditions = n()) %>% 
  filter(all(c("COVID", "OTHER") %in% covid_other)) %>% 
  arrange(desc(n_conditions), genes) %>% 
  group_by(genes) %>% 
  pivot_wider(., names_from = "covid_other", values_from = n_conditions) %>% 
  arrange(desc(OTHER)) %>% 
  ungroup() %>% 
  slice_head(n = 50) %>% 
  pivot_longer(-genes, names_to = "covid_other", values_to = "n_conditions")


stats_1 = degs_all_updown_covid_vax_other_immune %>% 
  inner_join(ann_vaccines_covid_other, by = "condition") %>% 
  group_by(genes, covid_other) %>% 
  summarise(n_conditions = n()) %>% 
  filter(all(c("COVID", "OTHER") %in% covid_other)) %>% 
  arrange(desc(n_conditions), genes) %>% 
  group_by(genes) %>% 
  pivot_wider(., names_from = "covid_other", values_from = n_conditions) %>% 
  arrange(desc(OTHER)) %>% 
  ungroup() %>% 
  slice_head(n = 50) %>% 
  pivot_longer(-genes, names_to = "covid_other", values_to = "n_conditions")

stats_1 %>% 
  select(genes) %>% 
  distinct()

ggplot_stats1 = ggplot(stats_1) +
  aes(x = reorder(genes, n_conditions), fill = covid_other, weight = n_conditions) +
  geom_bar() +
  theme_minimal() +
  scale_fill_manual(
    values = c(COVID = "#56cfe1",
    OTHER = "#343a40")
  ) +
  theme_minimal() +
  labs(fill = "Covid or non-covid") +
  xlab("Genes") + 
  ylab("Number of shared conditions") +
  theme(legend.position = "right",
        axis.text.x = element_text(vjust = 1.5, hjust=1, size = 12),
        axis.text.y = element_text(size = 12, angle = 0, color = "black"),
        panel.grid.minor.y = element_blank(),
        legend.text = element_text(size=12),
        legend.title = element_text(size = 12, face = "bold")) +
  # geom_text(aes(label = n_conditions, y = n_conditions), 
  #           position = position_dodge(width = 0.9), 
  #           vjust = 0.2, 
  #           hjust = -0.2, 
  #           size = 1,
  #           angle = 90) +
  ggtitle("Genes shared") +
  coord_flip()

ggsave(ggplot_stats1, file = "Covid_Vax_genes_shared_column.png", width = 10, height = 10)


# Subset the 50 most shared genes
covid_other_immune_genes_50_mostshared_all = stats_1 %>% 
  select(genes) %>% 
  distinct() %>% 
  inner_join(degs_all_updown_covid_vax_other_immune %>% 
               select(condition, genes), by = "genes")
```

```{r}
#### Not Immune -----
stats_2 = covid_other_notimmune_genes %>% 
  select(-process) %>% 
  distinct() %>% 
  inner_join(ann_vaccines_covid_other, by = "condition") %>% 
  group_by(genes, covid_other) %>% 
  summarise(n_conditions = n()) %>% 
  filter(all(c("COVID", "OTHER") %in% covid_other)) %>% 
  arrange(desc(n_conditions), genes) %>% 
  group_by(genes) %>% 
  pivot_wider(., names_from = "covid_other", values_from = n_conditions) %>% 
  arrange(desc(OTHER)) %>% 
  ungroup() %>% 
  slice_head(n = 50) %>% 
  pivot_longer(-genes, names_to = "covid_other", values_to = "n_conditions")

ggplot_stats2 = ggplot(stats_2) +
  aes(x = reorder(genes, n_conditions), fill = covid_other, weight = n_conditions) +
  geom_bar() +
  theme_minimal() +
  scale_fill_manual(
    values = c(COVID = "#56cfe1",
    OTHER = "#343a40")
  ) +
  theme_minimal() +
  labs(fill = "Covid or non-covid") +
  xlab("Genes") + 
  ylab("Number of shared conditions") +
  theme(legend.position = "right",
        axis.text.x = element_text(vjust = 1.5, hjust=1, size = 12),
        axis.text.y = element_text(size = 12, angle = 0, color = "black"),
        panel.grid.minor.y = element_blank(),
        legend.text = element_text(size=12),
        legend.title = element_text(size = 12, face = "bold")) +
  # geom_text(aes(label = n_conditions, y = n_conditions), 
  #           position = position_dodge(width = 0.9), 
  #           vjust = 0.2, 
  #           hjust = -0.2, 
  #           size = 1,
  #           angle = 90) +
  ggtitle("Genes shared") +
  coord_flip()

ggsave(ggplot_stats2, file = "Covid_Vax_Not-Immune_genes_shared_column.png", width = 10, height = 10)

# Subset the 50 most shared genes
covid_other_notimmune_genes_50_mostshared_all = stats_2 %>% 
  select(genes) %>% 
  inner_join(covid_other_notimmune_genes, by = "genes") 

covid_other_notimmune_genes_50_mostshared_all %>% select(genes) %>% distinct()

```

#### Heatmap

```{r}
####### INPUT
data_2 = covid_other_immune_genes_50_mostshared_all
filename_2 = paste0("mmune_GO_genes_All_50", "_VAX_COVID")

######## Matrix
matrix = data_2 %>% 
  select(genes, condition, regulation_numeric) %>%
  pivot_wider(names_from = "condition", 
              values_from = "regulation_numeric", 
              values_fn = mean) %>% 
  replace(is.na(.), 0) %>%
  arrange(genes) %>% 
  column_to_rownames(var="genes") %>% 
  as.matrix()

######## Annotations
ann_vaccines_covid_other_2 = data_2 %>% 
  inner_join(ann_vaccines_covid_other, by = "condition") %>% 
  select(condition, covid_other, microbe_type, disease_vac, type, week) %>% 
  distinct()

#Columns
ann_cols_heatmap = matrix %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  inner_join(ann_vaccines_covid_other_2, by = "condition") %>% 
  distinct() %>% 
  arrange(condition) %>% 
  column_to_rownames("condition")

matrix_data = matrix %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  inner_join(ann_cols_heatmap %>% rownames_to_column("condition"), by = "condition") %>% 
  select(!c(covid_other:week)) %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>%
  as.matrix() 

#Rows Immune ----
ann_rows_immune = matrix_data %>%
  as.data.frame() %>%
  rownames_to_column("genes") %>%
  left_join(ImmuneGO_Annotated_Genes %>% filter(go_term == "Manual"), by = "genes") %>%
  select(genes, process) %>%
  distinct() %>%
  group_by(genes) %>%
  arrange(genes, factor(process, levels = c("INNATE IMMUNE SYSTEM", "ADAPTIVE IMMUNE SYSTEM")), .by_group = TRUE) %>%
  mutate(i_process = row_number()) %>%
  pivot_wider(., names_from = "i_process", values_from = "process") %>%
  column_to_rownames("genes") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("index") %>%
  mutate(index = as.numeric(index)) %>%
  arrange(desc(index)) %>%
  column_to_rownames("index") %>%
  t() %>%
  as.data.frame() %>%
  replace(is.na(.), ".") %>% 
  mutate(`1` = if_else(`1` == ".", "INNATE IMMUNE SYSTEM", `1`))


# Rows Not immune -----
ann_rows_notimmune = matrix_data %>%
  as.data.frame() %>%
  rownames_to_column("genes") %>%
  left_join(ann_genes %>%
               select(process = term, genes = symbol, annotation) %>%
               filter(annotation == "Major process",
                      process != "Immune system"), by = "genes") %>%
  select(genes, process) %>%
  distinct() %>%
  group_by(genes) %>%
  mutate(i_process = row_number()) %>%
  pivot_wider(., names_from = "i_process", values_from = "process") %>%
  column_to_rownames("genes") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("index") %>%
  arrange(desc(index)) %>%
  column_to_rownames("index") %>%
  t() %>%
  as.data.frame() %>%
  replace(is.na(.), ".")


#Verificar dimensões do dataset
dim(matrix)
dim(matrix_data)
dim(ann_cols_heatmap)
dim(ann_rows_notimmune)
dim(ann_rows_immune)

```

More refs:
<https://debrowser.readthedocs.io/en/master/heatmap/heatmap.html>

```{r}
################# Immune 
#Heatmap annotation
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "right")

row_ha = row_ha = HeatmapAnnotation(df = ann_rows_immune,
                       col = col_process_immune,
                       which= "row",
                       show_annotation_name = F,
                       show_legend = F)

# Values colors
col_fun <- colorRamp2(c(-1, 0, 1), c("#9bc1bc", "white", "#ed6a5a"))

file = filename_2
mat = matrix_data
width_image = 50
height_image = 50
width = unit(30, "cm")
height = unit(30, "cm")
column_names_gp = gpar(fontsize = 12)
row_names_gp = gpar(fontsize = 8)
rect_gp = gpar(col = "gray80", lwd = 0.5)

fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap.png")

png(file = fileimageheatmap_grouped, 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                       left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        col = col_fun, #color
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(0, "cm"),
                        row_split = NULL,
                        column_split = 2,
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height)

heatmap_plot = draw(heatmap_plot, 
     annotation_legend_side = "left", 
     heatmap_legend_side = "left")

heatmap_plot

dev.off()

```

```{r}
################# NOT Immune 
#Heatmap annotation
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "right")

row_ha = HeatmapAnnotation(df = ann_rows_notimmune,
                       col = col_process_notimmune,
                       which= "row",
                       show_annotation_name = F,
                       show_legend = F)

# Values colors
col_fun <- colorRamp2(c(-1, 0, 1), c("#9bc1bc", "white", "#ed6a5a"))

file = filename_2
mat = matrix_data
width_image = 50
height_image = 50
width = unit(30, "cm")
height = unit(30, "cm")
column_names_gp = gpar(fontsize = 12)
row_names_gp = gpar(fontsize = 8)
rect_gp = gpar(col = "gray80", lwd = 0.5)
row_split = NULL
column_split = 3

fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap.png")

png(file = fileimageheatmap_grouped, 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        col = col_fun, #color
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(0, "cm"),
                        row_split = row_split,
                        column_split = column_split,
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        clustering_distance_columns = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height,
                        use_raster = F)

heatmap_plot = draw(heatmap_plot)

heatmap_plot

dev.off()
```

# 8. Machine learning descriptors

## 8.1 Preparing

## 8.1.1. Library

```{r message=FALSE, warning=FALSE}
# install.packages("randomForest")
# install.packages("caret")
# install.packages("rpart")
# install.packages("rpart.plot")
# install.packages("tidymodels")
# install.packages("reticulate")
# install.packages("themis")
# install.packages("ranger")
# install.packages("workflowsets")
# install.packages("glmnet")
# install.packages("kknn")
# install.packages("nnet")
# install.packages("tictoc")

library(tidymodels)
tidymodels_prefer()
library(randomForest)
library(ranger)
library(caret)
library(reticulate)
library(themis)
library(tidyverse)
library(rpart)
library(rpart.plot)
library(here)
library(ggrepel)
library(ComplexHeatmap)
library(janitor)
library(ggsci)
library(vip)
library(here)
library(circlize)
library(workflowsets)
library(glmnet)
library(kknn)
library(nnet)
library(tictoc)
library(ggthemes)

```

## 8.1.2. Import files

```{r}
all_matrices_normalized_counts <- readRDS(here("DGE analysis", "all_matrices_normalized_counts.rds"))
ann_vaccines_samples = read_csv(here("Annotations and metadata", "ann_vaccines_samples.csv"))
PC_1_2_cos2 = read_csv(here("PCA", "PC_1_2_cos2.csv"))
```

## 8.2. COVID-19 vaccines and infection: Random forest

```{r}
# Input data ----
filename = "ImmuneGO_Genes_Type_including_Infection_"
df = all_matrices_normalized_counts

## Vaccination types
outcomevar = "type"

ann_vaccines_samples_filtered = ann_vaccines_samples %>% 
  mutate(outcomevar = type) %>% 
  filter(type != "H")

#How many samples by level?
count_levels = ann_vaccines_samples_filtered %>% 
  count(outcomevar)

#How many levels?
levels = nrow(count_levels)

#Select only important genes
PC1_2 = PC_1_2_cos2
```

### 8.2.1. Train and test

```{r fig.height=10, fig.width=10}
# Prepare data ------

df_input <- df %>%
  rownames_to_column("genes") %>%
  inner_join(PC1_2 %>% select(genes) %>% distinct(), by = "genes") %>%
  column_to_rownames("genes") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  inner_join(ann_vaccines_samples_filtered %>% select(sample,
                                                      outcomevar), #Outcome
             by = "sample") %>%
  select(sample,
         outcomevar,
         everything()) %>%
  column_to_rownames("sample") %>%
  mutate_if(is.character, factor)

glimpse(df_input)

# Split data
set.seed(123)  
split <- initial_split(df_input, prop = 0.7, strata = outcomevar)
trees_train <- training(split)
trees_test <- testing(split)

# Create recipe
tree_rec <- recipe(outcomevar ~ ., data = trees_train) %>% #OUTCOME
  step_dummy(all_nominal(), -all_outcomes()) %>%
  step_upsample(outcomevar) #Upsample unbalanced data (OUTCOME)

tree_prep <- prep(tree_rec)

juiced <- juice(tree_prep)

# Set random forest hyper parameters
tune_spec <- rand_forest(
  mtry = tune(),
  trees = 2000,
  min_n = tune()
) %>%
  set_mode("classification") %>% 
  set_engine("ranger")

# Create workflow
tune_wf <- workflow() %>%
  add_recipe(tree_rec) %>%
  add_model(tune_spec)

#Cross validation
set.seed(234)
trees_folds <- vfold_cv(trees_train)


#Define performance metrics
mer_metrics <- metric_set(yardstick::recall,
                          yardstick::specificity,
                          yardstick::accuracy,
                          yardstick::precision,
                          yardstick::roc_auc,
                          yardstick::pr_auc)

# Parallel computing 
doParallel::registerDoParallel()

set.seed(345)
tune_res <- tune_grid(
  tune_wf,
  metrics = mer_metrics,
  resamples = trees_folds,
  grid = 20
)
```

### 8.2.2. Assess metrics

```{r fig.height=10, fig.width=10}
# Assess data ----

# recall
metrics_tuneres_recall = tune_res %>%
  collect_metrics() %>%
  filter(.metric == "recall") %>%
  select(mean, min_n, mtry) %>%
  pivot_longer(min_n:mtry,
    values_to = "value",
    names_to = "parameter"
  ) %>%
  ggplot(aes(value, mean, color = parameter)) +
  geom_line() +
  geom_point(show.legend = FALSE) +
  geom_text_repel(aes(label = value))+
  facet_wrap(~parameter, scales = "free_x") +
  labs(x = NULL, y = "recall")

metrics_tuneres_recall

#Accuracy
metrics_tuneres = tune_res %>%
  collect_metrics() %>%
  select(mean, min_n, mtry, .metric) %>%
  pivot_longer(min_n:mtry,
    values_to = "value",
    names_to = "parameter") %>%
  ggplot()+
  aes(x = value, y = mean, colour = .metric, label = value) +
  geom_line() +
  geom_label() +
  scale_color_npg()+
  theme_minimal() +
  facet_grid(vars(.metric), vars(parameter), scales = "free")

metrics_tuneres

metrics_tuneres %>% ggsave(file = here("Figures",
                                       paste0(filename, 
                                              outcomevar, today(),
                                              "_tuneres_metrics.png")),
                                        width = 10,
                                        height = 10)

#Min_n = 6,8
#mtry = 6,14
```

### 8.2.3. Hyperparameter tuning

```{r}
# Tune hyperparameters ----
rf_grid <- grid_regular(
  mtry(range = c(6, 14)),
  min_n(range = c(6, 8)),
  levels = levels
)

set.seed(456)
regular_res <- tune_grid(
  tune_wf,
  metrics = mer_metrics,
  resamples = trees_folds,
  grid = rf_grid
)

# Assess performance -----
regular_res %>%
  collect_metrics()

regular_res %>%
  collect_metrics() %>% 
  write.csv(file = paste("metrics_finalmodel_", filename, outcomevar, today(), ".csv"), row.names = F)

#Metrics
regular_res_metrics = regular_res %>%
  collect_metrics() %>%
  select(mean, min_n, mtry, .metric) %>%
  pivot_longer(min_n:mtry,
    values_to = "value",
    names_to = "parameter") %>%
  ggplot()+
  aes(x = value, y = mean, colour = .metric, label = value) +
  geom_line() +
  geom_label(size = 3) +
  scale_color_npg()+
  theme_minimal() +
  facet_grid(vars(.metric), vars(parameter), scales = "free")

regular_res_metrics

ggsave(regular_res_metrics, file = here("Figures", 
                                        paste(filename, outcomevar, "metrics", today(), ".png")), 
       width = 10, height = 10)

# Select best model -----
best_auc <- select_best(regular_res, metric = "pr_auc")
final_rf <- finalize_model(
  tune_spec,
  best_auc
)
```

### 8.2.4. Important features

```{r}
final = final_rf %>%
  set_engine("ranger", importance = "permutation") %>%
  fit(outcomevar ~ .,
    data = juice(tree_prep)
  ) %>% 
  vip(num_features = 10)
final
final %>% ggsave(file = paste0("vipgenes_tidymodels_", filename, outcomevar, today(), ".png"))

genes_rf = final$data
genes_rf = genes_rf %>% select(genes = Variable, importance = Importance)
genes_rf %>% write.csv(file = paste0("vipgenes_tidymodels_", filename, outcomevar, today(), ".csv"), row.names = F)

#Importance plot
importance_plot = ggplot(genes_rf) +
  aes(x = reorder(Variable, Importance),
    y = Importance,
    color = Importance,
      weight = Importance) +
  geom_segment(aes(x=Variable, xend=Variable, y=0, yend=Importance)) +
  geom_point(shape = "circle", size = 5) +
  theme_minimal() +
  xlab("Genes") + 
  ylab("Importance") +
  ylim(0, 0.05) +
  geom_text(aes(label = round(Importance, 4), y = Importance),
            hjust = -0.5, 
            size = 4,
            angle = 0,
            color = "black")+
  scale_color_gradient(low = "#4DBBD5FF", high = "#DC0000FF") +
  theme(legend.position = "right",
        axis.text.x = element_text(angle = 90, vjust = 1, hjust=1, size = 12, color = "black"),
        axis.text.y = element_text(size = 12, angle = 0, color = "black"),
        panel.grid.minor.y = element_blank(),
        legend.text = element_text(size=12),
        legend.title = element_text(size = 12, face = "bold")) +
    coord_flip()
importance_plot


ggsave(importance_plot,
        file = here("Figures",
                    paste0("vipgenes_tidymodels_", filename, outcomevar, today(), ".png")),
        width = 8,
        height = 5)

importance_plot

```

### 8.2.5. Train final model with tuned hyperparameters

```{r}
# Train final model with tuned hyperparameters
final_wf <- workflow() %>%
  add_recipe(tree_rec) %>%
  add_model(final_rf)

final_res <- final_wf %>%
  last_fit(split)

final_res_preds = final_res %>%
  collect_predictions() %>%
  mutate(correct = case_when(
    outcomevar == .pred_class ~ "Correct",
    TRUE ~ "Incorrect"
  )) %>%
  bind_cols(trees_test)

#Confusion matrix
res_conf_mat = conf_mat(final_res_preds, truth = outcomevar...12,
         estimate = .pred_class)

#Accuracy
res_acc = accuracy(final_res_preds, truth = outcomevar...12,
         estimate = .pred_class)

#Sensitivity
res_sens = sens(final_res_preds, truth = outcomevar...12,
         estimate = .pred_class)

#Specificity
res_spec = spec(final_res_preds, truth = outcomevar...12,
         estimate = .pred_class)

#Precision
res_prec = precision(final_res_preds, truth = outcomevar...12,
         estimate = .pred_class)

#Recall
res_recall = recall(final_res_preds, truth = outcomevar...12,
         estimate = .pred_class)

final_res_metrics = bind_rows(res_acc,
                         res_recall,
                         res_prec,
                         res_spec,
                         res_sens)


final_res_metrics_conf = list(res_conf_mat, 
                         final_res_metrics)

final_res_metrics_conf

final_res_metrics_conf %>% 
  saveRDS(file = paste0("final_res_metrics", filename, outcomevar, today(), ".rds"))

```

### 8.2.6. Heatmap: Gene L2FC per condition

```{r}
# Required files
ImmuneGO_Annotated_Genes = readRDS(here("VaxGO gene sets","ImmuneGO_Annotated_Genes_2024-03-26.rds"))
all_degs_p_05_vac_infected_19_12_23 = read_csv(here("DGE analysis", "all_degs_p_05_vac_infected_19_12_23.csv"))
ann_vaccines <- read_csv(here("Annotations and metadata", "ann_vaccines_19-1-24.csv"))

#Inputs
ImmuneGO_Annotated_Genes_all = ImmuneGO_Annotated_Genes %>% 
  filter(!process %in% c("TCR REPERTOIRE", "BCR REPERTOIRE"))

ImmuneGO_Genes = all_degs_p_05_vac_infected_19_12_23 %>% 
  inner_join(ImmuneGO_Annotated_Genes %>% 
               select(genes, process) %>% 
               distinct() %>% 
               select(genes),
               by = "genes") %>% 
  distinct()

# Matrix
l2fc_imps = genes_rf %>% 
  select(genes = Variable, importance = Importance) %>% 
  inner_join(ImmuneGO_Genes, by = "genes") %>%
  select(genes, condition, log2fold_change) %>% 
  pivot_wider(names_from = condition, values_from = log2fold_change) %>% 
  column_to_rownames("genes") %>% 
  as.matrix() %>% 
  replace(is.na(.), 0)

```

```{r}

#Which VIP genes?
#vipgenes_tidymodels_disease_vac2024_04_24 <- read_csv("vipgenes_tidymodels_disease_vac2024-04-24.csv")
vipgenes_tidymodels_type <- read_csv("vipgenes_tidymodels_type2024-06-21.csv")

vip_genes = vipgenes_tidymodels_type
outcomevar = "_Type_"

rf_genes = vip_genes %>% 
  clean_names() %>% 
  select(genes = variable, importance)

ImmuneGO_Genes = all_degs_p_05_vac_infected_19_12_23 %>% 
  inner_join(rf_genes %>% 
               select(genes) %>% 
               distinct() %>% 
               select(genes),
               by = "genes") %>% 
  distinct()

####### INPUT
data_2 = ImmuneGO_Genes
filename_2 = paste0("RandomForest_ImmuneGO_Genes", outcomevar, today())

######## Matrix
matrix = data_2 %>% 
  select(genes, condition, log2fold_change) %>%
  pivot_wider(names_from = "condition", 
              values_from = "log2fold_change", 
              values_fn = mean) %>% 
  replace(is.na(.), 0) %>%
  arrange(genes) %>% 
  column_to_rownames(var="genes") %>% 
  as.matrix()

######## Annotations
ann_vaccines_covid_other_2 = data_2 %>% 
  select(condition) %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  distinct() %>% 
  select(condition, disease_vac, type, week, day) %>% 
  arrange(week, day) %>% 
  mutate(week = fct_relevel(as.factor(week), sort(levels(week))),
         day = fct_relevel(as.factor(day), sort(levels(day))))

#Columns
ann_cols_heatmap = matrix %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  inner_join(ann_vaccines_covid_other_2, by = "condition") %>% 
  distinct() %>% 
  arrange(condition) %>% 
  column_to_rownames("condition")

matrix_data = matrix %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  inner_join(ann_cols_heatmap %>% rownames_to_column("condition"), by = "condition") %>% 
  select(!c(disease_vac:day)) %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>%
  as.matrix() 

#Rows Immune ----
ann_rows_immune = matrix_data %>%
  as.data.frame() %>%
  rownames_to_column("genes") %>%
  left_join(ImmuneGO_Annotated_Genes %>% filter(go_term == "Manual"), by = "genes") %>%
  select(genes, process) %>%
  distinct() %>%
  group_by(genes) %>%
  arrange(genes, factor(process, levels = c("INNATE IMMUNE SYSTEM", "ADAPTIVE IMMUNE SYSTEM")), .by_group = TRUE) %>%
  mutate(i_process = row_number()) %>%
  pivot_wider(., names_from = "i_process", values_from = "process") %>%
  column_to_rownames("genes") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("index") %>%
  mutate(index = as.numeric(index)) %>%
  arrange(desc(index)) %>%
  column_to_rownames("index") %>%
  t() %>%
  as.data.frame() %>%
  replace(is.na(.), ".") %>% 
  mutate(`1` = if_else(`1` == ".", "INNATE IMMUNE SYSTEM", `1`)) 

#Verificar dimensões do dataset
dim(matrix)
dim(matrix_data)
dim(ann_cols_heatmap)
dim(ann_rows_immune)
```

```{r}
################# Immune 
#Heatmap annotation
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "right")

row_ha = HeatmapAnnotation(df = ann_rows_immune,
                       col = col_process_immune,
                       which= "row",
                       show_annotation_name = F,
                       show_legend = F,
                       simple_anno_size = unit(2, "mm"))

# Values colors
col_fun <- colorRamp2(c(-2, 0, 2), c("#9bc1bc", "white", "#ed6a5a"))

file = filename_2
mat = matrix_data
width_image = 30
height_image = 15
width = unit(15, "cm")
height = unit(5, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)
rect_gp = gpar(col = "gray80", lwd = 0.5)

fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap.png")

png(file = fileimageheatmap_grouped, 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "L2FC",
                        col = col_fun, #color
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(0, "cm"),
                        row_split = NULL,
                        column_split = NULL,
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(2, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height)

heatmap_plot = draw(heatmap_plot, 
     annotation_legend_side = "left", 
     heatmap_legend_side = "left")

dev.off()

```

### 8.2.7. Boxplot: VIP gene expression by condition

```{r fig.height=10, fig.width=10}

table = df %>% 
  rownames_to_column("genes") %>% 
  pivot_longer(cols = -genes,
               names_to = "sample",
               values_to = "counts") %>% 
  filter(genes %in% c("GATA3", "ZNF3", "KMT2A", "ASXL1", "SP100", "GZMM","ITGAM", "ACTG1", "LGALS3", "STAT5B")) %>% 
  inner_join(ann_vaccines_samples_4_12_23, by = "sample") %>% 
  mutate(week_fac = as_factor(week),
         week_fac = fct_reorder(week_fac, week),
         type = fct_relevel(type, "H",  "I","RNA", "VV", "IN", "SU"))


#Base value -----
median_h = table %>% 
  select(genes, sample, counts, disease_vac) %>% 
  filter(disease_vac == "H") %>% 
  distinct() %>% 
  group_by(genes, disease_vac) %>% 
  summarize(median = median(counts), .groups = 'drop')

median_i = table %>% 
  select(genes, sample, counts, disease_vac) %>% 
  filter(disease_vac == "H") %>% 
  distinct() %>% 
  group_by(genes, disease_vac) %>% 
  summarize(median = median(counts), .groups = 'drop') %>% 
  mutate(disease_vac = "I")

median_v = table %>% 
  select(genes, sample, counts, disease_vac) %>% 
  filter(disease_vac == "H") %>% 
  distinct() %>% 
  group_by(genes, disease_vac) %>% 
  summarize(median = median(counts), .groups = 'drop') %>% 
  mutate(disease_vac = "V")

median = bind_rows(median_i, median_v)
  
  
  
#Plots -----
#Week
vip_genes_boxplot_week = table %>% 
  ggplot() +
  aes(x = week_fac, y = counts, fill = type,label = condition) +
  geom_boxplot() +
  scale_y_log10()+
  scale_fill_brewer(palette = "Set2", direction = 1) +
  scale_color_brewer(palette = "Set2", direction = 1) +
  theme_light() +
  facet_wrap(vars(genes), scales = "free", ncol = 1)

#Type
vip_genes_boxplot_type = table %>% 
  filter(disease_vac != "H") %>% 
  ggplot() +
  aes(x = type, y = counts, fill = week_fac) +
  geom_boxplot() +
  geom_hline(data = median, aes(yintercept = median), linetype = "dashed", alpha = 0.5) +
  scale_fill_viridis_d(option = "cividis", direction = -1) +
  scale_y_continuous(trans = "log10") +
  labs(
    x = "Type",
    y = "Normalized counts",
    title = "Gene expression of the 5 most important genes",
    fill = "Week"
  ) +
  theme_linedraw() +
  theme(axis.text.y = element_text(hjust = 0.1),
        legend.position = "top") +
  facet_grid(~genes~disease_vac, scales = "free")

vip_genes_boxplot_type 

vip_genes_boxplot_type %>% 
  ggsave(file = here("Figures", paste0("vip_genes_boxplot_type", filename, outcomevar, today(), ".png")), 
                                  width = 10, height = 15)

#Sex
vip_genes_boxplot_type_sex = table %>% 
  mutate(type_sex = paste0(type, "_", sex),
         type_sex = fct_relevel(type_sex, "IN_M/F", "SU_M/F")) %>% 
  filter(disease_vac != "H",
         !type %in% c("IN", "SU")) %>% 
  ggplot() +
  aes(x = type_sex, y = counts, fill = week_fac) +
  geom_boxplot() +
  geom_hline(data = median, aes(yintercept = median), linetype = "dashed", alpha = 0.5) +
  scale_fill_viridis_d(option = "cividis", direction = -1) +
  scale_y_continuous(trans = "log10") +
  labs(
    x = "Type",
    y = "Normalized counts",
    title = "Gene expression of the 5 most important genes",
    fill = "Week"
  ) +
  theme_linedraw() +
  theme(axis.text.y = element_text(hjust = 0.1),
        axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) +
  facet_grid(~genes~disease_vac, scales = "free")

vip_genes_boxplot_type_sex 


```



## 8.3. COVID-19 and Infection - Time and dose adjusted

```{r}
# Input data ----

filename = "ImmuneGO_Genes_Type_including_Infection_"
df = all_matrices_normalized_counts

## Vaccination types
outcomevar = "type_dose_time"

ann_vaccines_samples_filtered = ann_vaccines_samples %>% 
  filter(type != "H") %>% 
  mutate(outcomevar = paste0(type, "_", dose, "_", week))

#How many samples by level?
count_levels = ann_vaccines_samples_filtered %>% 
  count(outcomevar)
count_levels
#How many levels?
levels = nrow(count_levels)
levels
#Select only important genes
PC1_2 = PC_1_2_cos2
```

### 8.3.1. Train and test

```{r}
# Prepare data ------

#Filter genes with most importance in PCAs

df_input <- df %>%
  rownames_to_column("genes") %>%
  inner_join(PC1_2 %>% select(genes) %>% distinct(), by = "genes") %>%
  column_to_rownames("genes") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  inner_join(ann_vaccines_samples_filtered %>% 
               select(sample,
                      outcomevar), #Outcome
             by = "sample") %>%
  select(sample,
         outcomevar,
         everything()) %>%
  column_to_rownames("sample") %>%
  mutate_if(is.character, factor)

glimpse(df_input)

# Split data
set.seed(123)  
split <- initial_split(df_input, prop = 0.7, strata = outcomevar)
trees_train <- training(split)
trees_test <- testing(split)

# Create recipe
tree_rec <- recipe(outcomevar ~ ., data = trees_train) %>% #OUTCOME
  step_dummy(all_nominal(), -all_outcomes()) %>%
  step_corr(all_numeric_predictors()) %>% 
  step_upsample(outcomevar) #Upsample unbalanced data (OUTCOME)

tree_prep <- prep(tree_rec)

juiced <- juice(tree_prep)

# Set random forest hyper parameters
tune_spec <- rand_forest(
  mtry = tune(),
  trees = 2000,
  min_n = tune()
) %>%
  set_mode("classification") %>% 
  set_engine("ranger")

# Create workflow
tune_wf <- workflow() %>%
  add_recipe(tree_rec) %>%
  add_model(tune_spec)

#Cross validation
set.seed(234)
trees_folds <- vfold_cv(trees_train)


#Define performance metrics
mer_metrics <- metric_set(roc_auc, 
                          pr_auc,
                          yardstick::recall,
                          yardstick::accuracy)

# Parallel computing 
doParallel::registerDoParallel()

set.seed(345)
tune_res <- tune_grid(
  tune_wf,
  metrics = mer_metrics,
  resamples = trees_folds,
  grid = 20
)

# Assess data ----
# accuracy
metrics_tuneres = tune_res %>%
  collect_metrics() %>%
  filter(.metric == "accuracy") %>%
  select(mean, min_n, mtry) %>%
  pivot_longer(min_n:mtry,
    values_to = "value",
    names_to = "parameter"
  ) %>%
  ggplot(aes(value, mean, color = parameter)) +
  geom_line() +
  geom_point(show.legend = FALSE) +
  geom_text_repel(aes(label = value))+
  facet_wrap(~parameter, scales = "free_x") +
  labs(x = NULL, y = "accuracy")

metrics_tuneres

ggsave(metrics_tuneres, file = paste0(filename, outcomevar, "_tidymodels_tune_res.png"), width = 10, height = 6)
```

### 8.3.2. Hyperparameter tuning

```{r}
# Tune hyperparameters ----
rf_grid <- grid_regular(
  mtry(range = c(44, 45)),
  min_n(range = c(16, 17)),
  levels = levels
)

set.seed(456)
regular_res <- tune_grid(
  tune_wf,
  metrics = mer_metrics,
  resamples = trees_folds,
  grid = rf_grid
)

# Assess performance -----
regular_res %>%
  collect_metrics()

regular_res %>%
  collect_metrics() %>% 
  write.csv(file = paste("metrics_finalmodel_", filename, outcomevar, today(), ".csv"), row.names = F)

#ROC AUC
roc_auc = regular_res %>%
  collect_metrics() %>%
  filter(.metric == "roc_auc") %>%
  mutate(min_n = factor(min_n)) %>%
  ggplot(aes(mtry, mean, color = min_n)) +
  geom_line(alpha = 0.5, size = 1.5) +
  geom_point() +
  labs(y = "AUC")

ggsave(roc_auc, file = paste("roc_auc_", filename, outcomevar, today(), ".png"), width = 10, height = 6)

# Select best model -----
best_auc <- select_best(regular_res, metric = "accuracy")
final_rf <- finalize_model(
  tune_spec,
  best_auc
)
```

### 8.3.3. Important features

```{r}
final = final_rf %>%
  set_engine("ranger", importance = "permutation") %>%
  fit(outcomevar ~ .,
    data = juice(tree_prep)
  ) %>% 
  vip()

final

final %>% ggsave(file = paste0("vipgenes_tidymodels_", outcomevar, today(), ".png"))

genes_rf = final$data
genes_rf
genes_rf %>% write.csv(file = paste0("vipgenes_tidymodels_", filename, outcomevar, today(), ".csv"), row.names = F)

importance_plot = ggplot(genes_rf) +
  aes(x = reorder(Variable, Importance),
    y = Importance,
    color = Importance,
      weight = Importance) +
  geom_segment(aes(x=Variable, xend=Variable, y=0, yend=Importance)) +
  geom_point(shape = "circle", size = 5) +
  theme_minimal() +
  xlab("Genes") + 
  ylab("Importance") +
  ylim(0.03, 0.06) +
  geom_text(aes(label = round(Importance, 4), y = Importance),
            hjust = -0.5, 
            size = 4,
            angle = 0,
            color = "black")+
  scale_color_gradient(low = "#4DBBD5FF", high = "#DC0000FF") +
  theme(legend.position = "right",
        axis.text.x = element_text(angle = 90, vjust = 1, hjust=1, size = 12, color = "black"),
        axis.text.y = element_text(size = 12, angle = 0, color = "black"),
        panel.grid.minor.y = element_blank(),
        legend.text = element_text(size=12),
        legend.title = element_text(size = 12, face = "bold")) +
    coord_flip()

ggsave(importance_plot, file = paste0("vipgenes_tidymodels_", filename, outcomevar, today(), ".png"), width = 8, height = 5)

importance_plot

```

### 8.3.4. Train final model with tuned hyperparameters

```{r}
# Train final model with tuned hyperparameters
final_wf <- workflow() %>%
  add_recipe(tree_rec) %>%
  add_model(final_rf)

final_res <- final_wf %>%
  last_fit(split)

final_res %>%
  collect_metrics()

final_res %>%
  collect_metrics() %>% 
write.csv(file = paste0("rf_metrics", filename, outcomevar, today(), ".csv"), row.names = F)

final_res_preds = final_res %>%
  collect_predictions() %>%
  mutate(correct = case_when(
    outcomevar == .pred_class ~ "Correct",
    TRUE ~ "Incorrect"
  )) %>%
  bind_cols(trees_test)

#Confusion matrix
conf_mat(final_res_preds, truth = fgr,
         estimate = .pred_class)

#Accuracy
accuracy(final_res_preds, truth = fgr,
         estimate = .pred_class)

#Sensitivity
sens(fgr_results, truth = fgr,
         estimate = .pred_class)

#Specificity
spec(fgr_results, truth = fgr,
         estimate = .pred_class)

#Precision
precision(fgr_results, truth = fgr,
         estimate = .pred_class)

#Recall
recall(fgr_results, truth = fgr,
         estimate = .pred_class)

#ROC AUC
roc_plot = final_res_preds %>%
  roc_curve(truth = fgr, .pred_1) %>%
  autoplot()


final_res_acc_values = final_res_preds %>% 
  rename(outcomevar = starts_with("outcomevar")) %>% 
  group_by(correct, outcomevar1) %>% 
  tabyl(correct, outcomevar1) %>% 
  as.data.frame() %>% 
  group_by(correct) %>% 
  pivot_longer(-correct, values_to = "value", names_to = "outcomevar") %>% 
  group_by(outcomevar) %>% 
  arrange(outcomevar) %>% 
  mutate(perc = round(value/sum(value) * 100, 0))

final_res_acc_values %>% 
  write.csv(file = paste0("accuracy", filename, outcomevar, today(), ".csv"), row.names = F)

```

## 8.4. COVID-19 vaccines and infection: Workflowsets

```{r}
# Input data ----

filename = "ImmuneGO_Genes_Type_including_Infection_"
df = all_matrices_normalized_counts

## Vaccination types
outcomevar = "type"

ann_vaccines_samples_filtered = ann_vaccines_samples %>% 
  mutate(outcomevar = type) %>% 
  filter(type != "H") %>% 
  select(-type)

#How many samples by level?
count_levels = ann_vaccines_samples_filtered %>% 
  count(outcomevar)

count_levels

#How many levels?
levels = nrow(count_levels)

#Select only important genes
PC1_2 = PC_1_2_cos2

#Prepare data
df_input <- df %>%
  rownames_to_column("genes") %>%
  inner_join(PC1_2 %>% select(genes) %>% distinct(), by = "genes") %>%
  column_to_rownames("genes") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  inner_join(ann_vaccines_samples_filtered %>% select(sample,
                                                      outcomevar), #Outcome
             by = "sample") %>%
  select(sample,
         outcomevar,
         everything()) %>%
  column_to_rownames("sample") %>%
  mutate_if(is.character, factor) %>% 
  select(-"DOCK8", -"IL18BP")

glimpse(df_input)
```

### 8.4.1. Train and test

```{r}
# Prepare data ------

# Split data
set.seed(123)  
split <- initial_split(df_input, prop = 0.7, strata = "outcomevar")
train <- training(split)
test <- testing(split)
cv_folds <- vfold_cv(data = train, v = 5) 

#Recipe
base_rec <- recipe(outcomevar ~ ., data = train) %>%
 step_dummy(all_nominal(), -all_outcomes())
  # step_upsample(outcomevar) 

#Prep
base_prep <- prep(base_rec)
juiced <- juice(base_prep)


# Definir os modelos ----
# glm_model <- logistic_reg() %>%
#   set_engine("glm")

rf_model <- rand_forest(trees = 2000, 
                        min_n = tune(), 
                        mtry = tune()) %>%
  set_engine("ranger") %>%
  set_mode("classification")

tree_model <- decision_tree(cost_complexity = tune(), 
                            min_n = tune()) %>% 
  set_engine("rpart") %>%
  set_mode("classification")

knn_model <-  nearest_neighbor(neighbors = tune(), 
                               dist_power = tune(), 
                               weight_func = tune()) %>% 
  set_engine("kknn") %>%
  set_mode("classification")

nn_model <- mlp(hidden_units = tune(), 
                penalty = tune()) %>%
  set_engine("nnet") %>%
  set_mode("classification")

xgb_spec <- 
   boost_tree(tree_depth = tune(), 
              learn_rate = tune(), 
              loss_reduction = tune(), 
              min_n = tune(), 
              sample_size = tune(), 
              trees = tune()) %>% 
   set_engine("xgboost") %>% 
   set_mode("classification")

rf_wf <- workflow() %>%
  add_model(rf_model) %>%
  add_recipe(base_rec)

tree_wf <- workflow() %>%
  add_model(tree_model) %>%
  add_recipe(base_rec)

knn_wf <- workflow() %>%
  add_model(knn_model) %>%
  add_recipe(base_rec)

nn_wf <- workflow() %>%
  add_model(nn_model) %>%
  add_recipe(base_rec)

# Workflowsets ----
# link{https://www.tmwr.org/workflow-sets}
wf_set <- workflow_set(
  preproc = list(base_rec),
  models = list(rf = rf_model, 
                tree = tree_model, 
                knn = knn_model, 
                nn = nn_model)
)

#Tuning
grid_ctrl <-
   control_grid(
      save_pred = TRUE,
      parallel_over = "everything",
      save_workflow = TRUE
   )

grid_results <-
   wf_set %>%
   workflow_map(
      seed = 1503,
      resamples = cv_folds,
      grid = 10,
      control = grid_ctrl
   )

#Evaluate
grid_results %>% 
   rank_results() %>% 
   filter(.metric == "accuracy") %>% 
   select(model, .config, accuracy, rank)



## plot each of the model's performance and ROC
autoplot(
   grid_results,
   rank_metric = "accuracy",  # <- how to order models
   metric = "accuracy",       # <- which metric to visualize
   select_best = TRUE     # <- one point per workflow
) +
   geom_text(aes(y = mean - 1/2, label = wflow_id), angle = 90, hjust = 1) +
   lims(y = c(3.5, 9.5)) +
   theme(legend.position = "none")

## Look at the model metrics for each of the models
collect_metrics(fit_wf) 
 
#Specific model
autoplot(grid_results, id = "rf", metric = "accuracy")
```

### 8.4.2. Hyperparameter tuning

```{r}
# Tune hyperparameters ----
rf_grid <- grid_regular(
  mtry(range = c(2, 6)),
  min_n(range = c(6, 14)),
  levels = levels
)

set.seed(456)
regular_res <- tune_grid(
  tune_wf,
  resamples = trees_folds,
  grid = rf_grid
)

# Assess performance -----

regular_res %>%
  collect_metrics()

regular_res %>%
  collect_metrics() %>% 
  write.csv(file = paste("metrics_finalmodel_", filename, outcomevar, today(), ".csv"), row.names = F)

roc_auc = regular_res %>%
  collect_metrics() %>%
  filter(.metric == "roc_auc") %>%
  mutate(min_n = factor(min_n)) %>%
  ggplot(aes(mtry, mean, color = min_n)) +
  geom_line(alpha = 0.5, size = 1.5) +
  geom_point() +
  labs(y = "AUC") 
ggsave(roc_auc, file = paste("roc_auc_", filename, outcomevar, today(), ".png"), width = 10, height = 6)

# Select best model -----
best_auc <- select_best(regular_res, metric = "roc_auc")
final_rf <- finalize_model(
  tune_spec,
  best_auc
)
```

### 8.4.3. Important features

```{r}
final = final_rf %>%
  set_engine("ranger", importance = "permutation") %>%
  fit(outcomevar ~ .,
    data = juice(tree_prep)
  ) %>% 
  vip()

final

final %>% ggsave(file = paste0("vipgenes_tidymodels_", outcomevar, today(), ".png"))

genes_rf = final$data
genes_rf
genes_rf %>% write.csv(file = paste0("vipgenes_tidymodels_", filename, outcomevar, today(), ".csv"), row.names = F)

importance_plot = ggplot(genes_rf) +
  aes(x = reorder(Variable, Importance),
    y = Importance,
    color = Importance,
      weight = Importance) +
  geom_segment(aes(x=Variable, xend=Variable, y=0, yend=Importance)) +
  geom_point(shape = "circle", size = 5) +
  theme_minimal() +
  xlab("Genes") + 
  ylab("Importance") +
  geom_text(aes(label = round(Importance, 4), y = Importance),
            hjust = -0.5, 
            size = 4,
            angle = 0,
            color = "black")+
  scale_color_gradient(low = "#4DBBD5FF", high = "#DC0000FF") +
  theme(legend.position = "right",
        axis.text.x = element_text(angle = 90, vjust = 1, hjust=1, size = 12, color = "black"),
        axis.text.y = element_text(size = 12, angle = 0, color = "black"),
        panel.grid.minor.y = element_blank(),
        legend.text = element_text(size=12),
        legend.title = element_text(size = 12, face = "bold")) +
    coord_flip()

ggsave(importance_plot, file = paste0("vipgenes_tidymodels_", filename, outcomevar, today(), ".png"), width = 8, height = 5)

importance_plot

```

### 8.4.4. Train final model with tuned hyperparameters

```{r}
# Train final model with tuned hyperparameters
final_wf <- workflow() %>%
  add_recipe(tree_rec) %>%
  add_model(final_rf)

final_res <- final_wf %>%
  last_fit(split)

final_res %>%
  collect_metrics()

final_res %>%
  collect_metrics() %>% 
write.csv(file = paste0("rf_metrics", filename, outcomevar, today(), ".csv"), row.names = F)

final_res_preds = final_res %>%
  collect_predictions() %>%
  mutate(correct = case_when(
    outcomevar == .pred_class ~ "Correct",
    TRUE ~ "Incorrect"
  )) %>%
  bind_cols(trees_test)


final_res_acc_values = final_res_preds %>% 
  rename(outcomevar = starts_with("outcomevar")) %>% 
  group_by(correct, outcomevar1) %>% 
  tabyl(correct, outcomevar1) %>% 
  as.data.frame() %>% 
  group_by(correct) %>% 
  pivot_longer(-correct, values_to = "value", names_to = "outcomevar") %>% 
  group_by(outcomevar) %>% 
  arrange(outcomevar) %>% 
  mutate(perc = round(value/sum(value) * 100, 0))

final_res_acc_values %>% 
  write.csv(file = paste0("accuracy", filename, outcomevar, today(), ".csv"), row.names = F)

```

### 8.4.5 Heatmap: Gene L2FC per condition

```{r}
# Required files
ImmuneGO_Annotated_Genes = readRDS(here("VaxGO gene sets","ImmuneGO_Annotated_Genes_2024-03-26.rds"))
all_degs_p_05_vac_infected_19_12_23 = read_csv(here("DGE analysis", "all_degs_p_05_vac_infected_19_12_23.csv"))
ann_vaccines <- read_csv(here("Annotations and metadata", "ann_vaccines_19-1-24.csv"))

#Inputs
ImmuneGO_Annotated_Genes_all = ImmuneGO_Annotated_Genes %>% 
  filter(!process %in% c("TCR REPERTOIRE", "BCR REPERTOIRE"))

ImmuneGO_Genes = all_degs_p_05_vac_infected_19_12_23 %>% 
  inner_join(ImmuneGO_Annotated_Genes %>% 
               select(genes, process) %>% 
               distinct() %>% 
               select(genes),
               by = "genes") %>% 
  distinct()

# Matrix
l2fc_imps = genes_rf %>% 
  select(genes = Variable, importance = Importance) %>% 
  inner_join(ImmuneGO_Genes, by = "genes") %>%
  select(genes, condition, log2fold_change) %>% 
  pivot_wider(names_from = condition, values_from = log2fold_change) %>% 
  column_to_rownames("genes") %>% 
  as.matrix() %>% 
  replace(is.na(.), 0)

```

```{r}

#Which VIP genes?
#vipgenes_tidymodels_disease_vac2024_04_24 <- read_csv("vipgenes_tidymodels_disease_vac2024-04-24.csv")
vipgenes_tidymodels_type <- read_csv("vipgenes_tidymodels_type2024-06-21.csv")

vip_genes = vipgenes_tidymodels_type
outcomevar = "_Type_"

rf_genes = vip_genes %>% 
  clean_names() %>% 
  select(genes = variable, importance)

ImmuneGO_Genes = all_degs_p_05_vac_infected_19_12_23 %>% 
  inner_join(rf_genes %>% 
               select(genes) %>% 
               distinct() %>% 
               select(genes),
               by = "genes") %>% 
  distinct()

####### INPUT
data_2 = ImmuneGO_Genes
filename_2 = paste0("RandomForest_ImmuneGO_Genes", outcomevar, today())

######## Matrix
matrix = data_2 %>% 
  select(genes, condition, log2fold_change) %>%
  pivot_wider(names_from = "condition", 
              values_from = "log2fold_change", 
              values_fn = mean) %>% 
  replace(is.na(.), 0) %>%
  arrange(genes) %>% 
  column_to_rownames(var="genes") %>% 
  as.matrix()

######## Annotations
ann_vaccines_covid_other_2 = data_2 %>% 
  select(condition) %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  distinct() %>% 
  select(condition, disease_vac, type, week, day) %>% 
  arrange(week, day) %>% 
  mutate(week = fct_relevel(as.factor(week), sort(levels(week))),
         day = fct_relevel(as.factor(day), sort(levels(day))))

#Columns
ann_cols_heatmap = matrix %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  inner_join(ann_vaccines_covid_other_2, by = "condition") %>% 
  distinct() %>% 
  arrange(condition) %>% 
  column_to_rownames("condition")

matrix_data = matrix %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  inner_join(ann_cols_heatmap %>% rownames_to_column("condition"), by = "condition") %>% 
  select(!c(disease_vac:day)) %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>%
  as.matrix() 

#Rows Immune ----
ann_rows_immune = matrix_data %>%
  as.data.frame() %>%
  rownames_to_column("genes") %>%
  left_join(ImmuneGO_Annotated_Genes %>% filter(go_term == "Manual"), by = "genes") %>%
  select(genes, process) %>%
  distinct() %>%
  group_by(genes) %>%
  arrange(genes, factor(process, levels = c("INNATE IMMUNE SYSTEM", "ADAPTIVE IMMUNE SYSTEM")), .by_group = TRUE) %>%
  mutate(i_process = row_number()) %>%
  pivot_wider(., names_from = "i_process", values_from = "process") %>%
  column_to_rownames("genes") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("index") %>%
  mutate(index = as.numeric(index)) %>%
  arrange(desc(index)) %>%
  column_to_rownames("index") %>%
  t() %>%
  as.data.frame() %>%
  replace(is.na(.), ".") %>% 
  mutate(`1` = if_else(`1` == ".", "INNATE IMMUNE SYSTEM", `1`)) 

#Verificar dimensões do dataset
dim(matrix)
dim(matrix_data)
dim(ann_cols_heatmap)
dim(ann_rows_immune)
```

```{r}
################# Immune 
#Heatmap annotation
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "right")

row_ha = HeatmapAnnotation(df = ann_rows_immune,
                       col = col_process_immune,
                       which= "row",
                       show_annotation_name = F,
                       show_legend = F,
                       simple_anno_size = unit(2, "mm"))

# Values colors
col_fun <- colorRamp2(c(-2, 0, 2), c("#9bc1bc", "white", "#ed6a5a"))

file = filename_2
mat = matrix_data
width_image = 30
height_image = 15
width = unit(15, "cm")
height = unit(5, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)
rect_gp = gpar(col = "gray80", lwd = 0.5)

fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap.png")

png(file = fileimageheatmap_grouped, 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "L2FC",
                        col = col_fun, #color
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(0, "cm"),
                        row_split = NULL,
                        column_split = NULL,
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(2, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height)

heatmap_plot = draw(heatmap_plot, 
     annotation_legend_side = "left", 
     heatmap_legend_side = "left")

dev.off()

```

### 8.4.6. Heatmap: Random forest genes in COVID and other vaccines

```{r}
covid_other_immunego_genes <- readRDS(here("VaxGO gene sets", "covid_other_immunego_genes.rds"))
ann_vaccines_covid_other = read_csv(here("Annotations and metadata", "ann_vaccines_vaxsig_covid.csv"))

####### INPUT
data_2 = covid_other_immunego_genes %>% 
  inner_join(rf_genes, by = "genes")
filename_2 = paste0("RF_Genes", "_VAX_COVID")
```

```{r}
######## Matrix
matrix = data_2 %>% 
  select(genes, condition, regulation_numeric) %>%
  pivot_wider(names_from = "condition", 
              values_from = "regulation_numeric", 
              values_fn = mean) %>% 
  replace(is.na(.), 0) %>%
  arrange(genes) %>% 
  column_to_rownames(var="genes") %>% 
  as.matrix()

######## Annotations
ann_vaccines_covid_other_2 = data_2 %>% 
  inner_join(ann_vaccines_covid_other, by = "condition") %>% 
  select(condition, covid_other, microbe_type, disease_vac, type, week) %>% 
  distinct()

#Columns
ann_cols_heatmap = matrix %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  inner_join(ann_vaccines_covid_other_2, by = "condition") %>% 
  distinct() %>% 
  arrange(condition) %>% 
  column_to_rownames("condition")

matrix_data = matrix %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  inner_join(ann_cols_heatmap %>% rownames_to_column("condition"), by = "condition") %>% 
  select(!c(covid_other:week)) %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>%
  as.matrix() 

#Rows Immune ----
ann_rows_immune = matrix_data %>%
  as.data.frame() %>%
  rownames_to_column("genes") %>%
  left_join(ImmuneGO_Annotated_Genes %>%
              filter(go_term == "Manual",
                     !process %in% c("TCR REPERTOIRE", "BCR REPERTOIRE")), by = "genes") %>%
  select(genes, process) %>%
  distinct() %>%
  group_by(genes) %>%
  arrange(genes, factor(process, levels = c("INNATE IMMUNE SYSTEM", "ADAPTIVE IMMUNE SYSTEM")), .by_group = TRUE) %>%
  mutate(i_process = row_number()) %>%
  pivot_wider(., names_from = "i_process", values_from = "process") %>%
  column_to_rownames("genes") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("index") %>%
  mutate(index = as.numeric(index)) %>%
  arrange(desc(index)) %>%
  column_to_rownames("index") %>%
  t() %>%
  as.data.frame() %>%
  replace(is.na(.), ".")

#Verificar dimensões do dataset
dim(matrix_data)
dim(ann_cols_heatmap)
dim(ann_rows_immune)

```

```{r}
#Heatmap annotation
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "left")

row_ha = HeatmapAnnotation(df = ann_rows_immune,
                       col = col_process_immune,
                       which= "row",
                       show_annotation_name = F,
                       show_legend = F)

# Values colors
col_fun <- colorRamp2(c(-1, 0, 1), c("#9bc1bc", "white", "#ed6a5a"))

file = filename_2
mat = matrix_data
width_image = 30
height_image = 10
width = unit(5, "cm")
height = unit(2, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)
rect_gp = gpar(col = "gray80", lwd = 0)
row_split = NULL
column_split = NULL


fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap.png")

png(file = fileimageheatmap_grouped, 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        col = col_fun, #color
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(5, "cm"),
                        row_split = row_split,
                        column_split = column_split,
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "right",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        clustering_distance_columns = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height,
                        use_raster = F)

heatmap_plot = draw(heatmap_plot)

dev.off()

```

## 8.5. COVID-19: only up and down

```{r}
# Input data ----

filename = "ImmuneGO_Genes_Type_including_Infection_"
df = all_matrices_normalized_counts

## Disease vs Vaccination
# outcomevar = "disease_vac"
# levels = 2 

## Vaccination types
outcomevar = "type"

ann_vaccines_samples_filtered = ann_vaccines_samples %>% 
  mutate(outcomevar = type) %>% 
  filter(type != "H")

#Vaccine type, week, infection
  # mutate(outcomevar = paste0(if_else(disease_vac == "I", "Infection", "Vaccination"), "_", type, " (V", dose, ", W", week, ")")) %>% 

#How many samples by level?
count_levels = ann_vaccines_samples_filtered %>% 
  count(outcomevar)

#How many levels?
levels = nrow(count_levels)

#Select only important genes
PC1_2 = PC_1_2_cos2
```

### 8.5.1. Train and test

```{r}
# Prepare data ------

df_input <- df %>%
  rownames_to_column("genes") %>%
  inner_join(PC1_2 %>% select(genes) %>% distinct(), by = "genes") %>%
  column_to_rownames("genes") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  inner_join(ann_vaccines_samples_filtered %>% select(sample,
                                                      outcomevar), #Outcome
             by = "sample") %>%
  select(sample,
         outcomevar,
         everything()) %>%
  column_to_rownames("sample") %>%
  mutate_if(is.character, factor)

glimpse(df_input)

# Split data
set.seed(123)  
split <- initial_split(df_input, prop = 0.7, strata = outcomevar)
trees_train <- training(split)
trees_test <- testing(split)

# Create recipe
tree_rec <- recipe(outcomevar ~ ., data = trees_train) %>% #OUTCOME
  step_dummy(all_nominal(), -all_outcomes()) %>%
  step_upsample(outcomevar) #Upsample unbalanced data (OUTCOME)

tree_prep <- prep(tree_rec)

juiced <- juice(tree_prep)

# Set random forest hyper parameters
tune_spec <- rand_forest(
  mtry = tune(),
  trees = 2000,
  min_n = tune()
) %>%
  set_mode("classification") %>% 
  set_engine("ranger")

# Create workflow
tune_wf <- workflow() %>%
  add_recipe(tree_rec) %>%
  add_model(tune_spec)

#Cross validation
set.seed(234)
trees_folds <- vfold_cv(trees_train)

# Parallel computing 
doParallel::registerDoParallel()

set.seed(345)
tune_res <- tune_grid(
  tune_wf,
  resamples = trees_folds,
  grid = 20
)

# Assess data ----
# AUC
metrics_tuneres = tune_res %>%
  collect_metrics() %>%
  filter(.metric == "roc_auc") %>%
  select(mean, min_n, mtry) %>%
  pivot_longer(min_n:mtry,
    values_to = "value",
    names_to = "parameter"
  ) %>%
  ggplot(aes(value, mean, color = parameter)) +
  geom_line() +
  geom_point(show.legend = FALSE) +
  geom_text_repel(aes(label = value))+
  facet_wrap(~parameter, scales = "free_x") +
  labs(x = NULL, y = "AUC")

metrics_tuneres

ggsave(metrics_tuneres, file = paste0(filename, outcomevar, "_tidymodels_tune_res.png"), width = 10, height = 6)
```

### 8.5.2. Hyperparameter tuning

```{r}
# Tune hyperparameters ----
rf_grid <- grid_regular(
  mtry(range = c(2, 6)),
  min_n(range = c(6, 14)),
  levels = levels
)

set.seed(456)
regular_res <- tune_grid(
  tune_wf,
  resamples = trees_folds,
  grid = rf_grid
)

# Assess performance -----

regular_res %>%
  collect_metrics()

regular_res %>%
  collect_metrics() %>% 
  write.csv(file = paste("metrics_finalmodel_", filename, outcomevar, today(), ".csv"), row.names = F)

roc_auc = regular_res %>%
  collect_metrics() %>%
  filter(.metric == "roc_auc") %>%
  mutate(min_n = factor(min_n)) %>%
  ggplot(aes(mtry, mean, color = min_n)) +
  geom_line(alpha = 0.5, size = 1.5) +
  geom_point() +
  labs(y = "AUC") 
ggsave(roc_auc, file = paste("roc_auc_", filename, outcomevar, today(), ".png"), width = 10, height = 6)

# Select best model -----
best_auc <- select_best(regular_res, metric = "roc_auc")
final_rf <- finalize_model(
  tune_spec,
  best_auc
)
```

### 8.5.3. Important features

```{r}
final = final_rf %>%
  set_engine("ranger", importance = "permutation") %>%
  fit(outcomevar ~ .,
    data = juice(tree_prep)
  ) %>% 
  vip()

final

final %>% ggsave(file = paste0("vipgenes_tidymodels_", outcomevar, today(), ".png"))

genes_rf = final$data
genes_rf
genes_rf %>% write.csv(file = paste0("vipgenes_tidymodels_", outcomevar, today(), ".csv"), row.names = F)

importance_plot = ggplot(genes_rf) +
  aes(x = reorder(Variable, Importance),
    y = Importance,
    color = Importance,
      weight = Importance) +
  geom_segment(aes(x=Variable, xend=Variable, y=0, yend=Importance)) +
  geom_point(shape = "circle", size = 5) +
  theme_minimal() +
  xlab("Genes") + 
  ylab("Importance") +
  geom_text(aes(label = round(Importance, 4), y = Importance),
            hjust = -0.5, 
            size = 4,
            angle = 0,
            color = "black")+
  scale_color_gradient(low = "#4DBBD5FF", high = "#DC0000FF") +
  theme(legend.position = "right",
        axis.text.x = element_text(angle = 90, vjust = 1, hjust=1, size = 12, color = "black"),
        axis.text.y = element_text(size = 12, angle = 0, color = "black"),
        panel.grid.minor.y = element_blank(),
        legend.text = element_text(size=12),
        legend.title = element_text(size = 12, face = "bold")) +
    coord_flip()

ggsave(importance_plot, file = paste0("vipgenes_tidymodels_", filename, outcomevar, today(), ".png"), width = 8, height = 5)

importance_plot

```

### 8.5.4. Train final model with tuned hyperparameters

```{r}
# Train final model with tuned hyperparameters
final_wf <- workflow() %>%
  add_recipe(tree_rec) %>%
  add_model(final_rf)

final_res <- final_wf %>%
  last_fit(split)

final_res %>%
  collect_metrics()

final_res %>%
  collect_metrics() %>% 
write.csv(file = paste0("rf_metrics", filename, outcomevar, today(), ".csv"), row.names = F)

final_res_preds = final_res %>%
  collect_predictions() %>%
  mutate(correct = case_when(
    outcomevar == .pred_class ~ "Correct",
    TRUE ~ "Incorrect"
  )) %>%
  bind_cols(trees_test)


final_res_acc_values = final_res_preds %>% 
  rename(outcomevar = starts_with("outcomevar")) %>% 
  group_by(correct, outcomevar1) %>% 
  tabyl(correct, outcomevar1) %>% 
  as.data.frame() %>% 
  group_by(correct) %>% 
  pivot_longer(-correct, values_to = "value", names_to = "outcomevar") %>% 
  group_by(outcomevar) %>% 
  arrange(outcomevar) %>% 
  mutate(perc = round(value/sum(value) * 100, 0))

final_res_acc_values %>% 
  write.csv(file = paste0("accuracy",filename,  outcomevar, today(), ".csv"), row.names = F)

```

### 8.5.5. Heatmap: Gene L2FC per condition

```{r}
# Required files
ImmuneGO_Annotated_Genes = readRDS(here("VaxGO gene sets","ImmuneGO_Annotated_Genes_2024-03-26.rds"))
all_degs_p_05_vac_infected_19_12_23 = read_csv(here("DGE analysis", "all_degs_p_05_vac_infected_19_12_23.csv"))
ann_vaccines <- read_csv(here("Annotations and metadata", "ann_vaccines_19-1-24.csv"))

#Inputs
ImmuneGO_Annotated_Genes_all = ImmuneGO_Annotated_Genes %>% 
  filter(!process %in% c("TCR REPERTOIRE", "BCR REPERTOIRE"))

ImmuneGO_Genes = all_degs_p_05_vac_infected_19_12_23 %>% 
  inner_join(ImmuneGO_Annotated_Genes %>% 
               select(genes, process) %>% 
               distinct() %>% 
               select(genes),
               by = "genes") %>% 
  distinct()

# Matrix
l2fc_imps = genes_rf %>% 
  select(genes = Variable, importance = Importance) %>% 
  inner_join(ImmuneGO_Genes, by = "genes") %>%
  select(genes, condition, log2fold_change) %>% 
  pivot_wider(names_from = condition, values_from = log2fold_change) %>% 
  column_to_rownames("genes") %>% 
  as.matrix() %>% 
  replace(is.na(.), 0)

```

```{r}

#Which VIP genes?
#vipgenes_tidymodels_disease_vac2024_04_24 <- read_csv("vipgenes_tidymodels_disease_vac2024-04-24.csv")
vipgenes_tidymodels_type <- read_csv("vipgenes_tidymodels_type2024-06-21.csv")

vip_genes = vipgenes_tidymodels_type
outcomevar = "_Type_"

rf_genes = vip_genes %>% 
  clean_names() %>% 
  select(genes = variable, importance)

ImmuneGO_Genes = all_degs_p_05_vac_infected_19_12_23 %>% 
  inner_join(rf_genes %>% 
               select(genes) %>% 
               distinct() %>% 
               select(genes),
               by = "genes") %>% 
  distinct()

####### INPUT
data_2 = ImmuneGO_Genes
filename_2 = paste0("RandomForest_ImmuneGO_Genes", outcomevar, today())

######## Matrix
matrix = data_2 %>% 
  select(genes, condition, log2fold_change) %>%
  pivot_wider(names_from = "condition", 
              values_from = "log2fold_change", 
              values_fn = mean) %>% 
  replace(is.na(.), 0) %>%
  arrange(genes) %>% 
  column_to_rownames(var="genes") %>% 
  as.matrix()

######## Annotations
ann_vaccines_covid_other_2 = data_2 %>% 
  select(condition) %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  distinct() %>% 
  select(condition, disease_vac, type, week, day) %>% 
  arrange(week, day) %>% 
  mutate(week = fct_relevel(as.factor(week), sort(levels(week))),
         day = fct_relevel(as.factor(day), sort(levels(day))))

#Columns
ann_cols_heatmap = matrix %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  inner_join(ann_vaccines_covid_other_2, by = "condition") %>% 
  distinct() %>% 
  arrange(condition) %>% 
  column_to_rownames("condition")

matrix_data = matrix %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  inner_join(ann_cols_heatmap %>% rownames_to_column("condition"), by = "condition") %>% 
  select(!c(disease_vac:day)) %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>%
  as.matrix() 

#Rows Immune ----
ann_rows_immune = matrix_data %>%
  as.data.frame() %>%
  rownames_to_column("genes") %>%
  left_join(ImmuneGO_Annotated_Genes %>% filter(go_term == "Manual"), by = "genes") %>%
  select(genes, process) %>%
  distinct() %>%
  group_by(genes) %>%
  arrange(genes, factor(process, levels = c("INNATE IMMUNE SYSTEM", "ADAPTIVE IMMUNE SYSTEM")), .by_group = TRUE) %>%
  mutate(i_process = row_number()) %>%
  pivot_wider(., names_from = "i_process", values_from = "process") %>%
  column_to_rownames("genes") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("index") %>%
  mutate(index = as.numeric(index)) %>%
  arrange(desc(index)) %>%
  column_to_rownames("index") %>%
  t() %>%
  as.data.frame() %>%
  replace(is.na(.), ".") %>% 
  mutate(`1` = if_else(`1` == ".", "INNATE IMMUNE SYSTEM", `1`)) 

#Verificar dimensões do dataset
dim(matrix)
dim(matrix_data)
dim(ann_cols_heatmap)
dim(ann_rows_immune)
```

```{r}
################# Immune 
#Heatmap annotation
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "right")

row_ha = HeatmapAnnotation(df = ann_rows_immune,
                       col = col_process_immune,
                       which= "row",
                       show_annotation_name = F,
                       show_legend = F,
                       simple_anno_size = unit(2, "mm"))

# Values colors
col_fun <- colorRamp2(c(-2, 0, 2), c("#9bc1bc", "white", "#ed6a5a"))

file = filename_2
mat = matrix_data
width_image = 30
height_image = 15
width = unit(15, "cm")
height = unit(5, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)
rect_gp = gpar(col = "gray80", lwd = 0.5)

fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap.png")

png(file = fileimageheatmap_grouped, 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "L2FC",
                        col = col_fun, #color
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(0, "cm"),
                        row_split = NULL,
                        column_split = NULL,
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(2, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height)

heatmap_plot = draw(heatmap_plot, 
     annotation_legend_side = "left", 
     heatmap_legend_side = "left")

dev.off()

```

###8.5.6. Heatmap: Random forest genes in COVID and other vaccines

```{r}
covid_other_immunego_genes <- readRDS(here("VaxGO gene sets", "covid_other_immunego_genes.rds"))
ann_vaccines_covid_other = read_csv(here("Annotations and metadata", "ann_vaccines_vaxsig_covid.csv"))

####### INPUT
data_2 = covid_other_immunego_genes %>% 
  inner_join(rf_genes, by = "genes")
filename_2 = paste0("RF_Genes", "_VAX_COVID")
```

```{r}
######## Matrix
matrix = data_2 %>% 
  select(genes, condition, regulation_numeric) %>%
  pivot_wider(names_from = "condition", 
              values_from = "regulation_numeric", 
              values_fn = mean) %>% 
  replace(is.na(.), 0) %>%
  arrange(genes) %>% 
  column_to_rownames(var="genes") %>% 
  as.matrix()

######## Annotations
ann_vaccines_covid_other_2 = data_2 %>% 
  inner_join(ann_vaccines_covid_other, by = "condition") %>% 
  select(condition, covid_other, microbe_type, disease_vac, type, week) %>% 
  distinct()

#Columns
ann_cols_heatmap = matrix %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  inner_join(ann_vaccines_covid_other_2, by = "condition") %>% 
  distinct() %>% 
  arrange(condition) %>% 
  column_to_rownames("condition")

matrix_data = matrix %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  inner_join(ann_cols_heatmap %>% rownames_to_column("condition"), by = "condition") %>% 
  select(!c(covid_other:week)) %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>%
  as.matrix() 

#Rows Immune ----
ann_rows_immune = matrix_data %>%
  as.data.frame() %>%
  rownames_to_column("genes") %>%
  left_join(ImmuneGO_Annotated_Genes %>%
              filter(go_term == "Manual",
                     !process %in% c("TCR REPERTOIRE", "BCR REPERTOIRE")), by = "genes") %>%
  select(genes, process) %>%
  distinct() %>%
  group_by(genes) %>%
  arrange(genes, factor(process, levels = c("INNATE IMMUNE SYSTEM", "ADAPTIVE IMMUNE SYSTEM")), .by_group = TRUE) %>%
  mutate(i_process = row_number()) %>%
  pivot_wider(., names_from = "i_process", values_from = "process") %>%
  column_to_rownames("genes") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("index") %>%
  mutate(index = as.numeric(index)) %>%
  arrange(desc(index)) %>%
  column_to_rownames("index") %>%
  t() %>%
  as.data.frame() %>%
  replace(is.na(.), ".")

#Verificar dimensões do dataset
dim(matrix_data)
dim(ann_cols_heatmap)
dim(ann_rows_immune)

```

```{r}
#Heatmap annotation
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "left")

row_ha = HeatmapAnnotation(df = ann_rows_immune,
                       col = col_process_immune,
                       which= "row",
                       show_annotation_name = F,
                       show_legend = F)

# Values colors
col_fun <- colorRamp2(c(-1, 0, 1), c("#9bc1bc", "white", "#ed6a5a"))

file = filename_2
mat = matrix_data
width_image = 30
height_image = 10
width = unit(5, "cm")
height = unit(2, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)
rect_gp = gpar(col = "gray80", lwd = 0)
row_split = NULL
column_split = NULL


fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap.png")

png(file = fileimageheatmap_grouped, 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        col = col_fun, #color
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(5, "cm"),
                        row_split = row_split,
                        column_split = column_split,
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "right",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        clustering_distance_columns = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height,
                        use_raster = F)

heatmap_plot = draw(heatmap_plot)

dev.off()

```

## 8.6. COVID-19 and others: up and down

```{r}
# Input data ----
filename = "COVID_Others_ImmuneGO_Genes_Type_including_Infection_"
degs_covid_vax = readRDS("~/Desktop/Github - Covid Atlas/VaxGO gene sets/degs_covid_vaxsig_up_down_immune.rds")
data_annotation = read_csv(here("Annotations and metadata", "ann_vaccines_vaxsig_covid.csv")) %>% 
  select(condition, target_pathogen_disease, type, week)

df = degs_covid_vax %>% 
  select(condition, genes, regulation_numeric) %>% 
  distinct() %>% 
  pivot_wider(names_from = genes,
              values_from = regulation_numeric,
              values_fn = mean) %>% 
  inner_join(data_annotation , by = "condition")

## Vaccination types
outcomevar = "type"

ann_vaccines_samples_filtered = df %>% 
  mutate(outcomevar = type) %>% 
  select(-type) %>% 
  select(target_pathogen_disease, week, outcomevar, condition, everything())
  #filter(!outcomevar %in% c("CONJ", "VLP", "SU", "I")) #Remove classes with few rows

#How many samples by level?
count_levels = ann_vaccines_samples_filtered %>% 
  count(outcomevar)

count_levels

#How many levels?
levels = nrow(count_levels)
```

### 8.6.1. Train and test

```{r}
# Prepare data ------

df_input <- ann_vaccines_samples_filtered %>% 
  distinct() %>% 
  column_to_rownames("condition") %>%
  mutate_if(is.character, factor) %>% 
  mutate(week = as.factor(week),
         week = fct_relevel(week, -week)) %>% 
  replace(is.na(.), 0)

glimpse(df_input)

# Split data
set.seed(123)  
split <- initial_split(df_input, prop = 0.6, strata = outcomevar)
train <- training(split)
test <- testing(split)

# Create recipe
base_rec <- recipe(outcomevar ~ ., data = train) %>%
  step_dummy(all_nominal(), -all_outcomes())

base_rec <- prep(base_rec)
juiced <- juice(base_rec)
juiced %>% 
  glimpse()

#Cross validation
set.seed(764)
cv_folds <- vfold_cv(
  data = train, 
  v = 5
  ) 
```

### 8.6.2. Workflowsets

```{r}
#Recipe
base_rec <- recipe(outcomevar ~ ., data = train) %>%
  step_dummy(all_nominal(), -all_outcomes()) %>%
  step_upsample(outcomevar)

# Definir os modelos
glm_model <- logistic_reg() %>%
  set_engine("glm")

rf_model <- rand_forest(trees = tune(), 
                        min_n = tune(), 
                        mtry = tune()) %>%
  set_engine("ranger") %>%
  set_mode("classification")

tree_model <- decision_tree() %>%
  set_engine("rpart") %>%
  set_mode("classification")

knn_model <- nearest_neighbor(neighbors = tune()) %>%
  set_engine("kknn") %>%
  set_mode("classification")

nn_model <- mlp(hidden_units = tune(), 
                penalty = tune()) %>%
  set_engine("nnet") %>%
  set_mode("classification")

# Criar os workflows
glm_wf <- workflow() %>%
  add_model(glm_model) %>%
  add_recipe(base_rec)

rf_wf <- workflow() %>%
  add_model(rf_model) %>%
  add_recipe(base_rec)

tree_wf <- workflow() %>%
  add_model(tree_model) %>%
  add_recipe(base_rec)

knn_wf <- workflow() %>%
  add_model(knn_model) %>%
  add_recipe(base_rec)

nn_wf <- workflow() %>%
  add_model(nn_model) %>%
  add_recipe(base_rec)

# Criar o conjunto de workflows
wf_set <- workflow_set(
  preproc = list(base_rec),
  models = list(glm = glm_model, 
                rf = rf_model, 
                tree = tree_model, 
                knn = knn_model, 
                nn = nn_model)
)

#Tuning
doParallel::registerDoParallel(cores = 10)
 
tic()
 
fit_wf <- wf_set %>%  
  workflow_map(
    seed = 44, 
    fn = "tune_grid",
    grid = 10,           ## parameters to pass to tune grid
    resamples = cv_folds
  )
 
toc()
 
doParallel::stopImplicitCluster()
 
fit_wf

#Evaluate
## plot each of the model's performance and ROC
autoplot(fit_wf)
 
## Look at the model metrics for each of the models
collect_metrics(fit_wf) 
 
## Rank the results based on model accuracy
rank_results(fit_wf, rank_metric = "accuracy", select_best = TRUE)

```

## 8.7. COVID-19 and others: 13 Vaccines atlas dataset

# 9. Heatmaps - Genes - ImmuneGO

Import files

```{r }
############## Import files

# ImmuneGO 
ImmuneGO_Annotated_GO = read_csv(here("VaxGO gene sets", "ImmuneGO_Annotated_GO_2024_03_26.csv"))
ImmuneGO_Annotated_Genes = readRDS(here("VaxGO gene sets",  "ImmuneGO_Annotated_Genes_2024-03-26.rds"))

# Informations about gene sets
ImmuneGO = ImmuneGO_Annotated_GO %>% 
  select(process:immune_tissue, set_size) %>% 
  distinct()

# Filter only manual annotated gene sets
immunego_genes = ImmuneGO_Annotated_Genes %>%
  filter(go_term == "Manual") %>% 
  select(process,genes)
ann_vaccines = read_csv(here("Annotations and metadata", "ann_vaccines_conditions.csv")) %>% 
  mutate(week = fct_reorder(as.factor(week), week),
         day = fct_reorder(as.factor(day), day),
         dose = fct_reorder(as.factor(dose), dose))

# DEGs table
data.immunego = read_csv(here("DGE analysis", "all_degs_p_05_vac_infected_19_12_23.csv")) %>% 
  select(condition:padj)

#Colors
ann_colors = ann_vaxsig_colors 
```

Process data

```{r }
############## Process data

# All gene sets
Immune_GO_genes_All = data.immunego %>% 
  merge(immunego_genes, by = "genes", all.x = F, all.y=F) %>% 
  merge(ann_vaccines, by = "condition", all.y=F) %>% 
  clean_names()

###### Other immune gene sets
Immune_GO_genes_general = Immune_GO_genes_All %>%
  filter(process == "GENERAL")
Immune_GO_genes_antimicrobial = Immune_GO_genes_All %>%
  filter(process == "ANTIMICROBIAL")
Immune_GO_genes_Innate = Immune_GO_genes_All %>%
 filter(process == "INNATE IMMUNE SYSTEM")
Immune_GO_genes_Adaptive = Immune_GO_genes_All %>%
 filter(process == "ADAPTIVE IMMUNE SYSTEM")

# Innate immune system
Immune_GO_genes_Inflammation = Immune_GO_genes_All %>%
  filter(process == "INFLAMMATION")
Immune_GO_genes_ARPP = Immune_GO_genes_All %>%
  filter(process == "ARPP")
Immune_GO_genes_IFN = Immune_GO_genes_All %>%
  filter(process == "ANTIVIRAL & IFN")
Immune_GO_genes_Neutrophil = Immune_GO_genes_All %>%
  filter(process == "NEUTROPHIL")
Immune_GO_genes_Eosinophil = Immune_GO_genes_All %>%
  filter(process == "EOSINOPHIL")
Immune_GO_genes_Complement = Immune_GO_genes_All %>%
  filter(process == "COMPLEMENT")
Immune_GO_genes_Monocytes = Immune_GO_genes_All %>%
  filter(process == "MONOCYTE")
Immune_GO_genes_DC = Immune_GO_genes_All %>%
  filter(process == "DENDRITIC CELL")
Immune_GO_genes_Macro = Immune_GO_genes_All %>%
  filter(process == "MACROPHAGE")
Immune_GO_genes_NK = Immune_GO_genes_All %>%
  filter(process == "NK CELL")
Immune_GO_genes_Mast = Immune_GO_genes_All %>%
  filter(process == "MAST CELL") %>%
  distinct()

# Adaptive immune system
Immune_GO_genes_Cellular = Immune_GO_genes_All %>%
  filter(process == "CELLULAR ADAPTIVE IMMUNE SYSTEM")
Immune_GO_genes_Humoral = Immune_GO_genes_All %>%
  filter(process == "HUMORAL ADAPTIVE IMMUNE SYSTEM")
Immune_GO_genes_Tcells = Immune_GO_genes_All %>%
  filter(process == "T CELL")
Immune_GO_genes_Th = Immune_GO_genes_All %>%
  filter(process == "T HELPER CELLS")
Immune_GO_genes_Bcells = Immune_GO_genes_All %>%
  filter(process == "B CELL")
Immune_GO_genes_Ig = Immune_GO_genes_All %>%
  filter(process == "IMMUNOGLOBULIN MEDIATED RESPONSE") %>%
  distinct()

```

### Interactive heatmap version - All immmune genes

```{r}
library(InteractiveComplexHeatmap)

####### Input
heatmap_data = Immune_GO_genes_All
file = "Immune_GO_genes_All"

###### Process data

heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Convert to wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

matrix_data_ready =  matrix_data %>% 
  round(2)

# Check if all vaccines are in the matrix and join annotations
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

# Join annotations
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 


#Rows Immune ----
ann_rows_immune = matrix_data %>%
  as.data.frame() %>%
  rownames_to_column("genes") %>%
  left_join(ImmuneGO_Annotated_Genes %>%
              filter(go_term == "Manual",
                     !process %in% c("TCR REPERTOIRE", "BCR REPERTOIRE")), by = "genes") %>%
  select(genes, process) %>%
  distinct() %>%
  group_by(genes) %>%
  arrange(genes, factor(process, levels = c("INNATE IMMUNE SYSTEM", "ADAPTIVE IMMUNE SYSTEM")), .by_group = TRUE) %>%
  mutate(i_process = row_number()) %>%
  pivot_wider(., names_from = "i_process", values_from = "process") %>%
  column_to_rownames("genes") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("index") %>%
  mutate(index = as.numeric(index)) %>%
  arrange(desc(index)) %>%
  column_to_rownames("index") %>%
  t() %>%
  as.data.frame() %>%
  replace(is.na(.), ".")

# Reorder columns (same order as lines in the annotation table)
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% 
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0) %>% 
  as.matrix() 


#Check dimensions
dim(matrix_data_ordered) 
dim(ann_vaccines_heatmap) 
dim(ann_rows_immune)

#Create heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")

row_ha = HeatmapAnnotation(df = ann_rows_immune,
                       col = col_process_immune,
                       which= "row",
                       show_annotation_name = F,
                       show_legend = F)

col_fun <- colorRamp2(c(-2, 0, 2), 
                      c("#9bc1bc", "white", "#ed6a5a"))


#Parameters
mat = matrix_data_ordered
width_image = 40
height_image = 90
width = unit(30, "cm")
height = unit(80, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "vertical"),
                        name = "L2FC", #Legend
                        col = col_fun,
                        show_row_dend = F,
                        row_split = 2, #Split group clusters
                        show_row_names = F,
                        column_names_gp = column_names_gp,
                        row_names_gp = row_names_gp)

ht_list = draw(heatmap_plot_all, 
               annotation_legend_side = "right", 
               heatmap_legend_side = "right")
htShiny(ht_list)

htShiny(ht_list, save = "Interactive_CovidOnly_Heatmap")

```

### 9.1. General

```{r}
####### Input
heatmap_data = Immune_GO_genes_general
file = "Immune_GO_genes_general"

###### Process data

heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Convert to wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

matrix_data_ready =  matrix_data %>% 
  round(2)

# Check if all vaccines are in the matrix and join annotations
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

# Join annotations
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 

# Reorder columns (same order as lines in the annotation table)
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% 
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Check dimensions
dim(matrix_data_ordered) 
dim(ann_vaccines_heatmap) 


#Create heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), 
                      c("#9bc1bc", "white", "#ed6a5a"))


#Parameters
mat = matrix_data_ordered
width_image = 40
height_image = 90
width = unit(30, "cm")
height = unit(80, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 3)


#Plot and save
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 10),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        #column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### 9.2. Antimicrobial

```{r}
####### Input

heatmap_data = Immune_GO_genes_antimicrobial
file = "Immune_GO_genes_antimicrobial"

heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

matrix_data_ready =  matrix_data %>% 
  round(2)

ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 

matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% 
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0) %>% 
  as.matrix() 

dim(matrix_data_ordered)
dim(ann_vaccines_heatmap)

ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), 
                      c("#9bc1bc", "white", "#ed6a5a"))

mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(20, "cm")
width_image = 40
height_image = 30
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)

fileimageheatmap_ungrouped = here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))
png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)
heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 10),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, 
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### 9.3. Complement

```{r}
heatmap_data = Immune_GO_genes_Complement
file = "Immune_GO_genes_Complement"

heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

matrix_data_ready =  matrix_data %>% 
  round(2)

ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 

matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% 
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0) %>% 
  as.matrix() 

dim(matrix_data_ordered)
dim(ann_vaccines_heatmap)


ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#9bc1bc", "white", "#ed6a5a"))


mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(15, "cm")
width_image = 40
height_image = 25
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)

fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))
png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)
heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 10),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### 9.4. Inflammation

```{r}
heatmap_data = Immune_GO_genes_Inflammation
file = "Immune_GO_genes_Inflammation"
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

matrix_data_ready =  matrix_data %>% 
  round(2)

ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 

matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% 
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0) %>% 
  as.matrix() 

dim(matrix_data_ordered) 
dim(ann_vaccines_heatmap)

ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), 
                      c("#9bc1bc", "white", "#ed6a5a"))

mat = matrix_data_ordered
width_image = 40
height_image = 20
width = unit(30, "cm")
height = unit(10, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)

fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, 
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### 9.5. ARPP

```{r}
####### Input
#Troque pelo geneset de interesse

heatmap_data = Immune_GO_genes_ARPP
file = "Immune_GO_genes_ARPP"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Converter para wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#9bc1bc", "white", "#ed6a5a"))


#Parameters
mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(50, "cm")
width_image = 40
height_image = 60
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        # column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### 9.6. AV and IFN

```{r}
####### Input
#Troque pelo geneset de interesse

heatmap_data = Immune_GO_genes_IFN
file = "Immune_GO_genes_IFN"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Converter para wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#9bc1bc", "white", "#ed6a5a"))


#Parameters
mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(50, "cm")
width_image = 40
height_image = 60
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        # column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### 9.7. Neutrophil

```{r}
####### Input
#Troque pelo geneset de interesse

heatmap_data = Immune_GO_genes_Neutrophil
file = "Immune_GO_genes_Neutrophil"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Converter para wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#9bc1bc", "white", "#ed6a5a"))


#Parameters
mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(40, "cm")
width_image = 40
height_image = 50
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plotar e salvar imagem
fileimageheatmap_ungrouped = here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        # column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### 9.8. Eosinophil

```{r}
####### Input
#Troque pelo geneset de interesse

heatmap_data = Immune_GO_genes_Eosinophil
file = "Immune_GO_genes_Eosinophil"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Converter para wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#9bc1bc", "white", "#ed6a5a"))


#Parameters
mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(10, "cm")
width_image = 40
height_image = 20
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        # column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### 9.9. Monocyte

```{r}
####### Input
#Troque pelo geneset de interesse

heatmap_data = Immune_GO_genes_Monocytes
file = "Immune_GO_genes_Monocytes"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Converter para wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#9bc1bc", "white", "#ed6a5a"))


#Parameters
mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(15, "cm")
width_image = 40
height_image = 25
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        #column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### 9.10. Dendritic cells

```{r}
####### Input
#Troque pelo geneset de interesse

heatmap_data = Immune_GO_genes_DC
file = "Immune_GO_genes_DC"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Converter para wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#9bc1bc", "white", "#ed6a5a"))


#Parameters
mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(15, "cm")
width_image = 40
height_image = 25
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 10),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        #column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### 9.11. Macrophage

```{r}
####### Input
#Troque pelo geneset de interesse

heatmap_data = Immune_GO_genes_Macro
file = "Immune_GO_genes_Macro"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Converter para wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#9bc1bc", "white", "#ed6a5a"))


#Parameters
mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(15, "cm")
width_image = 40
height_image = 25
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 10),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        #column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### 9.12. NK

```{r}
####### Input
#Troque pelo geneset de interesse

heatmap_data = Immune_GO_genes_NK
file = "Immune_GO_genes_NK"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Converter para wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#9bc1bc", "white", "#ed6a5a"))


#Parameters
mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(15, "cm")
width_image = 40
height_image = 25
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 10),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        #column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### 9.13. Mast cell

```{r}
####### Input
#Troque pelo geneset de interesse

heatmap_data = Immune_GO_genes_Mast
file = "Immune_GO_genes_Mast"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Converter para wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#9bc1bc", "white", "#ed6a5a"))


#Parameters
mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(10, "cm")
width_image = 40
height_image = 20
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 10),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        #column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### 9.14. Cellular adaptive immune system

```{r}
####### Input
#Troque pelo geneset de interesse

heatmap_data = Immune_GO_genes_Cellular
file = "Immune_GO_genes_Cellular"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Converter para wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#9bc1bc", "white", "#ed6a5a"))


#Parameters
mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(70, "cm")
width_image = 40
height_image = 80
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 5)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 10),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        #column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### 9.15. T Cell

```{r}
####### Input
#Troque pelo geneset de interesse

heatmap_data = Immune_GO_genes_Tcells
file = "Immune_GO_genes_Tcells"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Converter para wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#9bc1bc", "white", "#ed6a5a"))


#Parameters
mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(80, "cm")
width_image = 40
height_image = 90
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 10),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        #column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### 9.16. Th Cell

```{r}
####### Input
#Troque pelo geneset de interesse

heatmap_data = Immune_GO_genes_Th
file = "Immune_GO_genes_Th"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Converter para wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#9bc1bc", "white", "#ed6a5a"))


#Parameters
mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(10, "cm")
width_image = 40
height_image = 20
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 10),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        #column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### 9.17. Humoral adaptive immune system

```{r}
####### Input
#Troque pelo geneset de interesse

heatmap_data = Immune_GO_genes_Humoral
file = "Immune_GO_genes_Humoral"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Converter para wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#9bc1bc", "white", "#ed6a5a"))


#Parameters
mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(70, "cm")
width_image = 40
height_image = 80
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 10),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        #column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### 9.18. B cells

```{r}
####### Input
#Troque pelo geneset de interesse

heatmap_data = Immune_GO_genes_Bcells
file = "Immune_GO_genes_Bcells"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Converter para wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#9bc1bc", "white", "#ed6a5a"))


#Parameters
mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(50, "cm")
width_image = 40
height_image = 60
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 10),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        #column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

### 9.19. Ig mediated response

```{r}
####### Input
#Troque pelo geneset de interesse

heatmap_data = Immune_GO_genes_Ig
file = "Immune_GO_genes_Ig"

########## Processar dados e criar matriz

#Excluir linhas com Target ou Source == NA
heatmap_data_l2fc= heatmap_data %>% 
  filter(condition != "NA", 
         process != "NA") %>% 
  select(genes, condition, log2fold_change)

# Converter para wide table
matrix_data <- dcast(heatmap_data_l2fc, 
                         genes ~ condition, 
                         value.var = "log2fold_change", 
                         fun.aggregate = mean) %>% 
  column_to_rownames('genes') %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(is.na(.), 0)

#Arredondar valores para facilitar visualização
matrix_data_ready =  matrix_data %>% 
  round(2)

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = heatmap_data %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines %>% select(condition, disease_vac, type, week), by = "condition")

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") 

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(disease_vac, type, week)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  replace(is.na(.), 0) %>% 
  as.matrix() 

#Verificar se as linhas sao iguais
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas


#Criar heatmap annotation
ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")


col_fun <- colorRamp2(c(-2, 0, 2), c("#9bc1bc", "white", "#ed6a5a"))


#Parameters
mat = matrix_data_ordered
width = unit(30, "cm")
height = unit(50, "cm")
width_image = 40
height_image = 60
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)


#Plotar e salvar imagem
fileimageheatmap_ungrouped= here("Figures", paste0(file, sep ="_", "Ungrouped_Complexheatmap.png"))

png(file = fileimageheatmap_ungrouped, width=width_image, height=height_image,units="cm",res=800)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "L2FC", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 10),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        #column_split = 2, #Split group clusters
                        #column_km = 2,
                        column_gap = unit(8, "mm"),
                        show_column_dend = TRUE,
                        show_row_dend = TRUE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```

# 10. Gene networks

```{r}
ImmuneGO_Annotated_Genes = readRDS(here("VaxGO gene sets", "ImmuneGO_Annotated_Genes_2024-03-26.rds"))
OtherGOs_Annotated_Genes = read_csv(here("VaxGO gene sets", "OtherGOs_Annotated_Genes.csv")) %>% 
  rename(genes = symbol)
  
degs_all_updown = read_csv(here("DGE analysis", "degs_updown_filtered_p_05_vac_infected.csv")) %>% 
  filter(direction != "NEUTRAL")
```

## 10.1. COVID immune genes - ImmuneGO set

```{r}
ImmuneGO_Annotated_Genes
OtherGOs_Annotated_Genes
degs_all_updown
filename = "covid_vaccines_infection_Immune_"


##### Table source - target - annotations ------

#Immune
degs_all_updown_covid_immune = degs_all_updown %>% 
  as.data.frame() %>% 
  select(condition:direction) %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  inner_join(ImmuneGO_Annotated_Genes %>% select(genes) %>% distinct(), by = "genes") %>% 
  distinct() %>% 
  select(condition, genes, regulation = direction)  %>% 
  distinct() %>% 
  filter(regulation %in% c("UP", "DOWN")) %>% 
  mutate(regulation_numeric = case_when(
    regulation %in% c("UP") ~ 1,
    regulation %in% c("DOWN") ~ -1,
    TRUE ~ as.numeric(regulation))) %>% 
  distinct()

# Immune
immune_source1 = degs_all_updown_covid_immune %>% 
  inner_join(ImmuneGO_Annotated_Genes %>% filter(go_term == "Manual", 
                                                 !(process %in% c("TCR REPERTOIRE", "BCR REPERTOIRE"))), 
             by = "genes") %>% 
  select(source = condition, target = genes, value = regulation_numeric) %>% 
  distinct() %>% 
  left_join(ann_vaccines_covid_other %>% rename(source = condition), by = "source") %>% 
  mutate(covid_other = if_else(is.na(covid_other), "OTHER", covid_other),
         annotation = paste0("Condition", covid_other))

immune_source2 = degs_all_updown_covid_immune %>% 
  inner_join(ImmuneGO_Annotated_Genes %>% filter(go_term == "Manual", 
                                                 !(process %in% c("TCR REPERTOIRE", "BCR REPERTOIRE"))), 
             by = "genes") %>% 
  select(source = process, target = genes) %>% 
  distinct() %>% 
  mutate(value = 0, 
         annotation = "Immune system",
         type = "Immune system")

immune = bind_rows(immune_source1, immune_source2) %>% mutate(id = source)

immune =  immune %>% 
   mutate(source = str_replace(source, ",", " "))

write.csv(immune, file = paste0(filename, "Network.csv"), row.names = F)
write_tsv(immune, file = paste0(filename, "Network.tsv"))

##### Tables nodes and edges ------

edges = immune %>% 
  select(source, target, value) %>% 
  mutate(type = "Directed")

nodes = immune %>% 
  select(source, annotation, vaccine:infection) %>% 
  distinct() %>% 
  mutate(id = source)

write.csv(edges, file = paste0(filename, "Network_edges.csv"), row.names = F)
write.csv(nodes, file = paste0(filename, "Network_nodes.csv"), row.names = F)
```

## 10.2. Immune GOs - Up and down

```{r}
##### Table source - target - annotations ------

filename = "COVID_ImmuneGOs_up_down_counts_"

#Immune
degs_all_updown_covid_immune = degs_all_updown %>% 
  as.data.frame() %>% 
  select(condition:direction) %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  inner_join(ImmuneGO_Annotated_Genes %>% select(genes) %>% distinct(), by = "genes") %>% 
  distinct() %>% 
  select(condition, genes, regulation = direction)  %>% 
  distinct() %>% 
  filter(regulation %in% c("UP", "DOWN")) %>% 
  mutate(regulation_numeric = case_when(
    regulation %in% c("UP") ~ 1,
    regulation %in% c("DOWN") ~ -1,
    TRUE ~ as.numeric(regulation))) %>% 
  distinct()

# Immune
immune_GOs_up_down = degs_all_updown_covid_immune %>% 
  inner_join(ImmuneGO_Annotated_Genes %>% filter(go_term == "Manual", 
                                                 !(process %in% c("TCR REPERTOIRE", "BCR REPERTOIRE"))), 
             by = "genes") %>% 
  select(source = condition, target = process, value = regulation) %>% 
  group_by(source, value) %>% 
  count(target) %>% 
  distinct() %>% 
  left_join(ann_vaccines_covid_other %>% rename(source = condition), by = "source") %>% 
  mutate(covid_other = if_else(is.na(covid_other), "OTHER", covid_other),
         annotation = paste0("Condition", covid_other))

immune =  immune_GOs_up_down %>% 
   mutate(source = str_replace(source, ",", " "))

write.csv(immune, file = paste0(filename, "Network.csv"), row.names = F)
write_tsv(immune, file = paste0(filename, "Network.tsv"))

##### Tables nodes and edges ------

edges = immune %>% 
  select(source, target, value) %>% 
  mutate(type = "Directed")

nodes = immune %>% 
  select(source, annotation, vaccine:infection) %>% 
  distinct() %>% 
  mutate(id = source)

write.csv(edges, file = paste0(filename, "Network_edges.csv"), row.names = F)
write.csv(nodes, file = paste0(filename, "Network_nodes.csv"), row.names = F)
```

## 10.3. COVID non immune genes

```{r}
ImmuneGO_Annotated_Genes
OtherGOs_Annotated_Genes
degs_all_updown
filename = "covid_vaccines_infection_NonImmune"


##### Table source - target - annotations ------

# Not immune
degs_all_updown_covid_notimmune = degs_all_updown %>% 
  as.data.frame() %>% 
  select(condition:direction) %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  inner_join(OtherGOs_Annotated_Genes %>% select(genes) %>% distinct(), by = "genes") %>% 
  distinct() %>% 
  select(condition, genes, regulation = direction)  %>% 
  distinct() %>% 
    filter(regulation %in% c("UP", "DOWN")) %>% 
  mutate(regulation_numeric = case_when(
    regulation %in% c("UP") ~ 1,
    regulation %in% c("DOWN") ~ -1,
    TRUE ~ as.numeric(regulation)
  )) %>% 
  distinct()


# Not immune genes
notimmune_source1 = degs_all_updown_covid_notimmune %>% 
  inner_join(OtherGOs_Annotated_Genes %>% filter(annotation == "Major process", term != "Immune system"), by = "genes") %>% 
  select(source = condition, target = genes, value = regulation_numeric) %>% 
  distinct() %>% 
  left_join(ann_vaccines_covid_other %>% rename(source = condition), by = "source") %>% 
    mutate(covid_other = if_else(is.na(covid_other), "OTHER", covid_other),
           annotation = paste0("Condition", covid_other))

notimmune_source2 = degs_all_updown_covid_notimmune %>% 
  inner_join(OtherGOs_Annotated_Genes %>% filter(annotation == "Major process", term != "Immune system"), by = "genes") %>% 
  select(source = term, target = genes, value = regulation_numeric) %>% 
  distinct() %>% 
  mutate(value = 0, 
         annotation = "Other processes")

not_immune = bind_rows(notimmune_source1, notimmune_source2) %>% mutate(id = source)

write.csv(not_immune, file = paste0(filename, "Network.csv"), row.names = F)

##### Tables nodes and edges ------

edges = not_immune %>% 
  select(source, target, value) %>% 
  mutate(type = "Directed")

nodes = not_immune %>% 
  select(source, annotation, vaccine:infection) %>% 
  distinct() %>% 
  mutate(id = source)

write.csv(edges, file = paste0(filename, "Network_edges.csv"), row.names = F)
write.csv(nodes, file = paste0(filename, "Network_nodes.csv"), row.names = F)

```

## 10.4. COVID immune and non immune genes

```{r}
ImmuneGO_Annotated_Genes
OtherGOs_Annotated_Genes
degs_all_updown
filename = "covid_vaccines_infection_Immune_Notimmune"


##### Table source - target - annotations ------

#Immune
degs_all_updown_covid_immune = degs_all_updown %>% 
  as.data.frame() %>% 
  select(condition:direction) %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  inner_join(ImmuneGO_Annotated_Genes %>% select(genes) %>% distinct(), by = "genes") %>% 
  distinct() %>% 
  select(condition, genes, regulation = direction)  %>% 
  distinct() %>% 
  filter(regulation %in% c("UP", "DOWN")) %>% 
  mutate(regulation_numeric = case_when(
    regulation %in% c("UP") ~ 1,
    regulation %in% c("DOWN") ~ -1,
    TRUE ~ as.numeric(regulation)
  )) %>% 
  distinct()

# Not immune
degs_all_updown_covid_notimmune = degs_all_updown %>% 
  as.data.frame() %>% 
  select(condition:direction) %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  inner_join(OtherGOs_Annotated_Genes %>% filter(term != "Immune system") %>% select(genes) %>% distinct(), by = "genes") %>% 
  distinct() %>% 
  select(condition, genes, regulation = direction)  %>% 
  distinct() %>% 
    filter(regulation %in% c("UP", "DOWN")) %>% 
  mutate(regulation_numeric = case_when(
    regulation %in% c("UP") ~ 1,
    regulation %in% c("DOWN") ~ -1,
    TRUE ~ as.numeric(regulation)
  )) %>% 
  distinct()

# Immune
immune_source1 = degs_all_updown_covid_immune %>% 
  inner_join(ImmuneGO_Annotated_Genes %>% filter(go_term == "Manual", 
                                                 !(process %in% c("TCR REPERTOIRE", "BCR REPERTOIRE"))), 
             by = "genes") %>% 
  select(source = condition, target = genes, value = regulation_numeric) %>% 
  distinct() %>% 
  left_join(ann_vaccines_covid_other %>% rename(source = condition), by = "source") %>% 
  mutate(covid_other = if_else(is.na(covid_other), "OTHER", covid_other),
         annotation = paste0("Condition", covid_other))

immune_source2 = degs_all_updown_covid_immune %>% 
  inner_join(ImmuneGO_Annotated_Genes %>% filter(go_term == "Manual", 
                                                 !(process %in% c("TCR REPERTOIRE", "BCR REPERTOIRE"))), 
             by = "genes") %>% 
  select(source = process, target = genes) %>% 
  distinct() %>% 
  mutate(value = 0, 
         annotation = "Immune system")

immune = bind_rows(immune_source1, immune_source2) %>% mutate(id = source)


# Not immune genes
notimmune_source1 = degs_all_updown_covid_notimmune %>% 
  inner_join(OtherGOs_Annotated_Genes %>% filter(annotation == "Major process", term != "Immune system"), by = "genes") %>% 
  select(source = condition, target = genes, value = regulation_numeric) %>% 
  distinct() %>% 
  left_join(ann_vaccines_covid_other %>% rename(source = condition), by = "source") %>% 
    mutate(covid_other = if_else(is.na(covid_other), "OTHER", covid_other),
           annotation = paste0("Condition", covid_other))

notimmune_source2 = degs_all_updown_covid_notimmune %>% 
  inner_join(OtherGOs_Annotated_Genes %>% filter(annotation == "Major process", term != "Immune system"), by = "genes") %>% 
  select(source = term, target = genes, value = regulation_numeric) %>% 
  distinct() %>% 
  mutate(value = 0, 
         annotation = "Other processes")

not_immune = bind_rows(notimmune_source1, notimmune_source2) %>% mutate(id = source)
immune_notimmune = bind_rows(not_immune, immune)
write.csv(immune_notimmune, file = paste0(filename, "Network_immune_notimmune.csv"), row.names = F)

##### Tables nodes and edges ------

edges = immune_notimmune %>% 
  select(source, target, value) %>% 
  mutate(type = "Directed")

nodes = immune_notimmune %>% 
  select(source, annotation, vaccine:infection) %>% 
  distinct() %>% 
  mutate(id = source)

write.csv(edges, file = paste0(filename, "Network_immune_notimmune_edges.csv"), row.names = F)
write.csv(nodes, file = paste0(filename, "Network_immune_notimmune_nodes.csv"), row.names = F)

```

### 10.4.1. COVID-COVID Network by statistics only (TOTAL = UP AND DOWN)

```{r}
# Vaccines and infection -----
ann_vaccines_covid_other

degs_all_updown_covid_other_immune_shared
degs_all_updown_covid_other_nonimmune_shared

filename = "degs_all_updown_covid-covid_immune_shared"

edges = degs_all_updown_covid_other_immune_shared %>% 
  select(-"...1") %>% 
  inner_join(ann_vaccines_covid_other %>% filter(covid_other == "COVID"), by = join_by(Cond1 == condition)) %>% 
  inner_join(ann_vaccines_covid_other %>% filter(covid_other == "COVID") %>% select(condition), by = join_by(Cond2 == condition)) %>% 
  clean_names()

edges_mirror = edges %>% 
  select(cond1:qvalue) %>% 
  rename(cond2.1 = cond1, cond1.2 = cond2) %>% 
  rename(cond2 = cond2.1, cond1 = cond1.2) %>% 
  select(cond1, cond2, everything()) %>% 
  inner_join(ann_vaccines_covid_other %>% filter(covid_other == "COVID"), by = join_by(cond1 == condition))

edges_flourish = bind_rows(edges, edges_mirror)

nodes = ann_vaccines_covid_other 
edges_nodes = edges


my_colors <- ann_vaccines_covid_other %>% 
 mutate(color = case_when(
    type == "VLP" ~ "#7E6148FF",
    type == "LA" ~ "#E64B35FF",
    type == "CONJ" ~ "#F39B7FFF",
    type == "IN" ~ "#00A087FF",
    type == "VV" ~ "#3C5488FF",
    type == "RNA" ~ "#4DBBD5FF",
    type == "SU" ~ "#8491B4FF",
    type == "I" ~ "#DC0000FF",
    TRUE ~ NA_character_),
    color_covid_other = if_else(covid_other == "COVID", "#56cfe1", "#343a40"))

my_colors %>% select(condition, color, color_covid_other) %>% 
  write_tsv("covid_others_colors.tsv")


write_tsv(edges_nodes, file = paste0(filename, "Network_Covid_Other_sharedgenes_edges_and_nodes.tsv"))
write_tsv(edges_flourish, file = paste0(filename, "Network_Covid_Other_sharedgenes_edges.tsv"))
write.csv(edges_flourish, file = paste0(filename, "Network_Covid_Other_sharedgenes_edges.csv"), row.names = F)
write_tsv(nodes, file = paste0(filename, "Network_Covid_Other_sharedgenes_nodes.tsv"))

# Only Covid vaccination (previous or no previous infection) vs other vaccines-----
ann_vaccines_covid_other

edges = degs_all_updown_covid_other_immune_shared %>% 
  select(-"...1") %>% 
  inner_join(ann_vaccines_covid_other %>% filter(covid_other == "COVID", !regimen %in% c("I", "I-I")), by = join_by(Cond1 == condition)) %>% 
  inner_join(ann_vaccines_covid_other %>% filter(covid_other == "OTHER") %>% select(condition), by = join_by(Cond2 == condition)) %>% 
  clean_names()

edges_mirror = edges %>% 
  select(cond1:qvalue) %>% 
  rename(cond2.1 = cond1, cond1.2 = cond2) %>% 
  rename(cond2 = cond2.1, cond1 = cond1.2) %>% 
  select(cond1, cond2, everything()) %>% 
  inner_join(ann_vaccines_covid_other %>% filter(covid_other == "OTHER"), by = join_by(cond1 == condition))

edges_flourish = bind_rows(edges, edges_mirror)

nodes = ann_vaccines_covid_other 
my_colors <- ann_vaccines_covid_other %>% 
 mutate(color = case_when(
    type == "VLP" ~ "#D7B0EE",
    type == "LA" ~ "#ffb703",
    type == "CONJ" ~ "#015BA0",
    type == "IN" ~ "#76c893",
    type == "VV" ~ "#5e60ce",
    type == "RNA" ~ "#56cfe1",
    type == "SU" ~ "#b5e48c",
    type == "I" ~ "#f72585",
    TRUE ~ NA_character_),
    color_covid_other = if_else(covid_other == "COVID", "#56cfe1", "#343a40"))

my_colors %>% select(condition, color, color_covid_other) %>% 
  write_tsv("covid_others_colors.tsv")

write_tsv(edges_flourish, file = paste0(filename, "Network_Covid_Other_Vaccinesonly_sharedgenes_edges.tsv"))
write.csv(edges_flourish, file = paste0(filename, "Network_Covid_Other_Vaccinesonly_sharedgenes_edges.csv"), row.names = F)
write_tsv(nodes, file = paste0(filename, "Network_Covid_Other_sharedgenes_nodes.tsv"))

```

### 10.4.2. COVID-COVID Network (UP + DOWN)

```{r}
# Vaccines and infection -----
ann_vaccines_covid_other

degs_all_updown_covid_immune

filename = "degs_all_updown_covid-covid_immune_shared_UPandDOWN_"

edges 

degs_all_updown_covid_immune %>% 
  inner_join(degs_all_updown_covid_immune %>% select(target = condition, genes, regulation), by = "genes") %>% 
  select(source = condition, genes, regulation)
  inner_join(ann_vaccines_covid_other %>% filter(covid_other == "COVID"), by = join_by(Cond1 == condition)) %>% 
  inner_join(ann_vaccines_covid_other %>% filter(covid_other == "COVID") %>% select(condition), by = join_by(Cond2 == condition)) %>% 
  clean_names()
```

```{r}
edges_mirror = edges %>% 
  select(cond1:qvalue) %>% 
  rename(cond2.1 = cond1, cond1.2 = cond2) %>% 
  rename(cond2 = cond2.1, cond1 = cond1.2) %>% 
  select(cond1, cond2, everything()) %>% 
  inner_join(ann_vaccines_covid_other %>% filter(covid_other == "COVID"), by = join_by(cond1 == condition))

edges_flourish = bind_rows(edges, edges_mirror)

nodes = ann_vaccines_covid_other 
edges_nodes = edges


my_colors <- ann_vaccines_covid_other %>% 
 mutate(color = case_when(
    type == "VLP" ~ "#7E6148FF",
    type == "LA" ~ "#E64B35FF",
    type == "CONJ" ~ "#F39B7FFF",
    type == "IN" ~ "#00A087FF",
    type == "VV" ~ "#3C5488FF",
    type == "RNA" ~ "#4DBBD5FF",
    type == "SU" ~ "#8491B4FF",
    type == "I" ~ "#DC0000FF",
    TRUE ~ NA_character_),
    color_covid_other = if_else(covid_other == "COVID", "#56cfe1", "#343a40"))

my_colors %>% select(condition, color, color_covid_other) %>% 
  write_tsv("covid_others_colors.tsv")


write_tsv(edges_nodes, file = paste0(filename, "Network_Covid_Other_sharedgenes_edges_and_nodes.tsv"))
write_tsv(edges_flourish, file = paste0(filename, "Network_Covid_Other_sharedgenes_edges.tsv"))
write.csv(edges_flourish, file = paste0(filename, "Network_Covid_Other_sharedgenes_edges.csv"), row.names = F)
write_tsv(nodes, file = paste0(filename, "Network_Covid_Other_sharedgenes_nodes.tsv"))

# Only Covid vaccination (previous or no previous infection) vs other vaccines-----
ann_vaccines_covid_other

edges = degs_all_updown_covid_other_immune_shared %>% 
  select(-"...1") %>% 
  inner_join(ann_vaccines_covid_other %>% filter(covid_other == "COVID", !regimen %in% c("I", "I-I")), by = join_by(Cond1 == condition)) %>% 
  inner_join(ann_vaccines_covid_other %>% filter(covid_other == "OTHER") %>% select(condition), by = join_by(Cond2 == condition)) %>% 
  clean_names()

edges_mirror = edges %>% 
  select(cond1:qvalue) %>% 
  rename(cond2.1 = cond1, cond1.2 = cond2) %>% 
  rename(cond2 = cond2.1, cond1 = cond1.2) %>% 
  select(cond1, cond2, everything()) %>% 
  inner_join(ann_vaccines_covid_other %>% filter(covid_other == "OTHER"), by = join_by(cond1 == condition))

edges_flourish = bind_rows(edges, edges_mirror)

nodes = ann_vaccines_covid_other 
my_colors <- ann_vaccines_covid_other %>% 
 mutate(color = case_when(
    type == "VLP" ~ "#D7B0EE",
    type == "LA" ~ "#ffb703",
    type == "CONJ" ~ "#015BA0",
    type == "IN" ~ "#76c893",
    type == "VV" ~ "#5e60ce",
    type == "RNA" ~ "#56cfe1",
    type == "SU" ~ "#b5e48c",
    type == "I" ~ "#f72585",
    TRUE ~ NA_character_),
    color_covid_other = if_else(covid_other == "COVID", "#56cfe1", "#343a40"))

my_colors %>% select(condition, color, color_covid_other) %>% 
  write_tsv("covid_others_colors.tsv")

write_tsv(edges_flourish, file = paste0(filename, "Network_Covid_Other_Vaccinesonly_sharedgenes_edges.tsv"))
write.csv(edges_flourish, file = paste0(filename, "Network_Covid_Other_Vaccinesonly_sharedgenes_edges.csv"), row.names = F)
write_tsv(nodes, file = paste0(filename, "Network_Covid_Other_sharedgenes_nodes.tsv"))

```

COVID and other vaccines

```{r}

ImmuneGO_Annotated_Genes
OtherGOs_Annotated_Genes


##### Table source - target - annotations ------


# Immune
immune_source1 = covid_other_immunego_genes %>% 
  inner_join(ImmuneGO_Annotated_Genes %>% filter(go_term == "Manual", 
                                                 !(process %in% c("TCR REPERTOIRE", "BCR REPERTOIRE"))), 
             by = "genes") %>% 
  select(source = condition, target = genes, value = regulation_numeric) %>% 
  distinct() %>% 
  left_join(ann_vaccines_covid_other %>% rename(source = condition), by = "source") %>% 
  mutate(covid_other = if_else(is.na(covid_other), "OTHER", covid_other),
         annotation = paste0("Condition", covid_other))

immune_source2 = covid_other_immunego_genes %>% 
  inner_join(ImmuneGO_Annotated_Genes %>% filter(go_term == "Manual", 
                                                 !(process %in% c("TCR REPERTOIRE", "BCR REPERTOIRE"))), 
             by = "genes") %>% 
  select(source = process, target = genes) %>% 
  distinct() %>% 
  mutate(value = 0, 
         annotation = "Immune system")

immune = bind_rows(immune_source1, immune_source2) %>% mutate(id = source)


# Not immune genes
notimmune_source1 = covid_other_notimmune_genes %>% 
  inner_join(OtherGOs_Annotated_Genes %>% filter(annotation == "Major process"), by = "genes") %>% 
  select(source = condition, target = genes, value = regulation_numeric) %>% 
  distinct() %>% 
  left_join(ann_vaccines_covid_other %>% rename(source = condition), by = "source") %>% 
    mutate(covid_other = if_else(is.na(covid_other), "OTHER", covid_other),
           annotation = paste0("Condition", covid_other))

notimmune_source2 = covid_other_notimmune_genes %>% 
  inner_join(OtherGOs_Annotated_Genes %>% filter(annotation == "Major process") , by = "genes") %>% 
  select(source = term, target = genes, value = regulation_numeric) %>% 
  distinct() %>% 
  mutate(value = 0, 
         annotation = "Other processes")

not_immune = bind_rows(notimmune_source1, notimmune_source2) %>% mutate(id = source)
immune_notimmune = bind_rows(not_immune, immune)
view(immune_notimmune)
write.csv(immune_notimmune, file = "Network_immune_notimmune.csv", row.names = F)

##### Tables nodes and edges ------

edges = immune_notimmune %>% 
  select(source, target, value) %>% 
  mutate(type = "Directed")

nodes = immune_notimmune %>% 
  select(source, annotation, vaccine:infection) %>% 
  distinct() %>% 
  mutate(id = source)

write.csv(edges, file = "Network_immune_notimmune_edges.csv", row.names = F)
write.csv(nodes, file = "Network_immune_notimmune_nodes.csv", row.names = F)
```

## 10.5. Shared genes (Cytoscape and Flourish)

```{r}
# Vaccines and infection -----
ann_vaccines_covid_other

degs_all_updown_covid_other_immune_shared
degs_all_updown_covid_other_nonimmune_shared

filename = "degs_all_updown_covid_other_immune_shared"

edges = degs_all_updown_covid_other_immune_shared %>% 
  select(-"...1") %>% 
  inner_join(ann_vaccines_covid_other %>% filter(covid_other == "COVID"), by = join_by(Cond1 == condition)) %>% 
  inner_join(ann_vaccines_covid_other %>% filter(covid_other == "OTHER") %>% select(condition), by = join_by(Cond2 == condition)) %>% 
  clean_names()

edges_mirror = edges %>% 
  select(cond1:qvalue) %>% 
  rename(cond2.1 = cond1, cond1.2 = cond2) %>% 
  rename(cond2 = cond2.1, cond1 = cond1.2) %>% 
  select(cond1, cond2, everything()) %>% 
  inner_join(ann_vaccines_covid_other %>% filter(covid_other == "OTHER"), by = join_by(cond1 == condition))

edges_flourish = bind_rows(edges, edges_mirror)

nodes = ann_vaccines_covid_other 
edges_nodes = edges


my_colors <- ann_vaccines_covid_other %>% 
 mutate(color = case_when(
    type == "VLP" ~ "#7E6148FF",
    type == "LA" ~ "#E64B35FF",
    type == "CONJ" ~ "#F39B7FFF",
    type == "IN" ~ "#00A087FF",
    type == "VV" ~ "#3C5488FF",
    type == "RNA" ~ "#4DBBD5FF",
    type == "SU" ~ "#8491B4FF",
    type == "I" ~ "#DC0000FF",
    TRUE ~ NA_character_),
    color_covid_other = if_else(covid_other == "COVID", "#56cfe1", "#343a40"))

my_colors %>% select(condition, color, color_covid_other) %>% 
  write_tsv("covid_others_colors.tsv")


write_tsv(edges_nodes, file = paste0(filename, "Network_Covid_Other_sharedgenes_edges_and_nodes.tsv"))
write_tsv(edges_flourish, file = paste0(filename, "Network_Covid_Other_sharedgenes_edges.tsv"))
write.csv(edges_flourish, file = paste0(filename, "Network_Covid_Other_sharedgenes_edges.csv"), row.names = F)
write_tsv(nodes, file = paste0(filename, "Network_Covid_Other_sharedgenes_nodes.tsv"))

# Only Covid vaccination (previous or no previous infection) vs other vaccines-----
ann_vaccines_covid_other

edges = degs_all_updown_covid_other_immune_shared %>% 
  select(-"...1") %>% 
  inner_join(ann_vaccines_covid_other %>% filter(covid_other == "COVID", !regimen %in% c("I", "I-I")), by = join_by(Cond1 == condition)) %>% 
  inner_join(ann_vaccines_covid_other %>% filter(covid_other == "OTHER") %>% select(condition), by = join_by(Cond2 == condition)) %>% 
  clean_names()

edges_mirror = edges %>% 
  select(cond1:qvalue) %>% 
  rename(cond2.1 = cond1, cond1.2 = cond2) %>% 
  rename(cond2 = cond2.1, cond1 = cond1.2) %>% 
  select(cond1, cond2, everything()) %>% 
  inner_join(ann_vaccines_covid_other %>% filter(covid_other == "OTHER"), by = join_by(cond1 == condition))

edges_flourish = bind_rows(edges, edges_mirror)

nodes = ann_vaccines_covid_other 
my_colors <- ann_vaccines_covid_other %>% 
 mutate(color = case_when(
    type == "VLP" ~ "#7E6148FF",
    type == "LA" ~ "#c8b6ff",
    type == "CONJ" ~ "#F39B7FFF",
    type == "IN" ~ "#00A087FF",
    type == "VV" ~ "#3C5488FF",
    type == "RNA" ~ "#4DBBD5FF",
    type == "SU" ~ "#8491B4FF",
    type == "I" ~ "#DC0000FF",
    TRUE ~ NA_character_),
    color_covid_other = if_else(covid_other == "COVID", "#56cfe1", "#343a40"))

my_colors %>% select(condition, color, color_covid_other) %>% 
  write_tsv("covid_others_colors.tsv")

#Copiar e colar direto no flourish
my_colors %>% select(condition, color) %>% mutate(combined = paste(condition, color, sep = ": ")) %>% select(combined)

write_tsv(edges_flourish, file = paste0(filename, "Network_Covid_Other_Vaccinesonly_sharedgenes_edges.tsv"))
write.csv(edges_flourish, file = paste0(filename, "Network_Covid_Other_Vaccinesonly_sharedgenes_edges.csv"), row.names = F)
write_tsv(nodes, file = paste0(filename, "Network_Covid_Other_sharedgenes_nodes.tsv"))


```
