---
title: "Differential gene expression analysis"
output: html_notebook
---
# Bibliotecas

```{r}
library(readr)
library(qvalue)
library(explore)
library(maditr)
library(GEOquery)
library(Matrix)
library(pheatmap)
library(RColorBrewer)
library(celldex)
library(biomaRt)
library(org.Hs.eg.db)
library(plotly)
library(DESeq2)
library(msigdbr)
library(clusterProfiler)
library(EnhancedVolcano)
library(ggbeeswarm)
library(tidyverse)
library(corrr)
library(ggcorrplot)
library(FactoMineR)
library(factoextra)
library(esquisse)
library(corto)
library(circlize)
library(ComplexHeatmap)
library(janitor)
library(scales)
library(ggstatsplot)

```


# 1. Preparação dos dados

## 2. Definir padrões de figuras

####Cores
```{r}
# Definir variáveis de cores
vaxcolors = c(
  "BBIBP" = "#ffd000",
  "BNT" = "#56cfe1",
  "ChAd" = "#80ffdb",
  "ChAd-BNT" = "#72efdd",
  "ZF2001" = "#b5179e")

dose_colors = c(
    "1" = "#56cfe1",
    "2" = "#5978d4",
    "3" = "#b5179e")

day_colors = c(
    "1" = "#caf0f8",
    "3" = "#90e0ef",
    "6" = "#00b4d8",
    "7" = "#0077b6",
    "14" = "#023e8a",
    "28" = "#03045e")

type_colors = c('IN'= "#ffd000", #First generation vaccines
           'LA' = "#ffff3f",
           'CONJ' = "#f4a261",
           'VLP' = "#da1e37", #Second generation vaccines
           'SU' = "#ff4d6d",
           'PEP' = "#f48498",
           'VV' = "#72efdd", #Third generation vaccines
           'RNA' = "#56cfe1",
           'VV-RNA' = "#5e60ce",
           'IN-SU' = "#b5179e")

regimen_colors = c('HO' = "#56cfe1",
              'HE' = "#5e60ce")

immune_system_colors = c(
    "Innate" = "#faedcd",
    "Adaptive" = "#ffb5a7",
    "General" = "#99d98c",
    "Other" = "#e5e5e5",
    "Complement" = "#7400b8")

immune_subsystem_colors = c(
    "Adaptive" = "#ffb5a7",
    "Cellular" = "#ff758f",
    "Humoral" = "#ff758f",
    "Antibacterial" = "#4895ef",
    "Antifungal" = "#4361ee",
    "Antimicrobial" = "#7400b8",
    "General" = "#99d98c",
    "Antigen presentation" = "#e4c4a5",
    "APC" = "#ccd5ae",
    "Eosinophil" = "#faedcd",
    "Granulocyte" = "#faedcd",
    "Leukocyte" = "#ffb5a7",
    "Neutrophil" = "#ccd5ae",
    "Innate immune response" = "#faedcd",
    "Interferon" = "#a4c3b2",
    "Antiviral" = "#60495a",
    "Pattern recognition receptor" = "#f4f4d5",
    "Natural killer" = "#fedd96",
    "Mucosa" = "#c3dfe0",
    "Erythrocyte" = "#4895ef")



########## Colors dictionary

#Vaccines
ann_colors = list(
  vaccine = c(
  "BBIBP" = "#dee2e6",
  "BNT" = "#56cfe1",
  "ChAd" = "#80ffdb",
  "ChAd-BNT" = "#5e60ce",
  "ZF2001" = "#b5179e",
  "INFECTION" = "#ff4d6d",
  "INFECTION-VACCINE"= "#ffccd5"),
  dose = c(
    "NA" = "white",
    "0" = "white",
    "1" = "#56cfe1",
    "2" = "#5978d4",
    "3" = "#b5179e"),
  day = c(
    "0" = "white",
    "1" = "#caf0f8",
    "2" = "#ade8f4",
    "3" = "#90e0ef",
    "4" = "#6CD5EA",
    "5" = "#48cae4",
    "6" = "#00b4d8",
    "7" = "#0096c7",
    "10" = "#0087BF",
    "14" = "#0077b6",
    "26" = "#015BA0",
    "28" = "#023e8a",
    "51" = "#03045e"),
  type = c('IN'= "#e5e5e5", #1st Gen  vaccines
           'SU' = "#00a896",
           'VV' = "#72efdd", #3rd Gen vaccines
           'RNA' = "#56cfe1",
           'VV-RNA' = "#5e60ce", #Heterologous
           'IN-SU' = "#b5179e",
           'INFECTION'= "#da1e37"), #Infection
  regimen = c('HO' = "grey90",
              'HE' = "#8d99ae",
              "V-I-V" = "#36248f",
              "V-I" = "#D7B0EE",
              "I-V-I" = "#7400b8",
              "I-I" = "#5390d9",
              "I" = "#64dfdf"),
  infection = c("none" = "white",
                "1" = "#56cfe1",
                "2" = "#5e60ce"),
  variant = c("NA" = "white",
              "Beta" = "#72efdd",
              "Omicron" = "#5e60ce"),
  severity = c("A (9%) MI (81%) MO (0) S (4%)" = "#caf0f8",
               "A (29%) MI (57%) MO (14) S (0%)" = "#56cfe1",
               "A (0%) MI (89%) MO (11%) S (0%)" = "#72efdd",
               "A (9%) MI (47%) MO (21%) S (3%)" = "#64dfdf",
               "S" = "#5e60ce",
               "MO" = "#64dfdf",
               "MI" = "#72efdd",
               "NA" = "white"),
  gse_id = c("GSE199750" = "#80ffdb",
               "GSE201533" = "#5e60ce", 
               "GSE201530" = "#7400b8",
               "GSE206023" = "#4361ee",
               "GSE189039" = "#48bfe3"))

#Rows
# ann_genesets_heatmap
ann_colors_rows = list(
  `immune_system` = c(
    "Innate" = "#faedcd",
    "Adaptive" = "#ffb5a7",
    "General" = "#99d98c",
    "Other" = "#e5e5e5",
    "Complement" = "#7400b8",
    "No enrichment" = "gray90"
  ),
  `immune_sub_system` = c(
    "Adaptive" = "#ffb5a7",
    "Cellular" = "#ff758f",
    "Humoral" = "#ff758f",
    "Antibacterial" = "#4895ef",
    "Antifungal" = "#4361ee",
    "Antimicrobial" = "#7400b8",
    "General" = "#99d98c",
    "Antigen presentation" = "#e4c4a5",
    "APC" = "#ccd5ae",
    "Eosinophil" = "#faedcd",
    "Granulocyte" = "#faedcd",
    "Leukocyte" = "#ffb5a7",
    "Neutrophil" = "#ccd5ae",
    "Innate immune response" = "#faedcd",
    "Interferon" = "#a4c3b2",
    "Antiviral" = "#60495a",
    "Pattern recognition receptor" = "#f4f4d5",
    "Natural killer" = "#fedd96",
    "Mucosa" = "#c3dfe0",
    "Erythrocyte" = "#4895ef",
    "No enrichment" = "gray90"
  )
)
  

```



# 3. Seleção de estudos


## 3.1 Baixar dados

```{r eval=FALSE, include=FALSE}
#Baixar dados de uma lista de estudos

############ Obter Tabela de counts
#GEO supplementary files. Baixar individualmente. Arquivos grandes devem ser baixados direto do navegador.
getGEOSuppFiles('GSE189039') # Baixar tabela de counts
untar('GSE189039_RAW.tar') # Descompactar tar zipped
#gunzip("GSE174621_RAW.tar") # Descompactar gunzip (.gz)


############ Metadados

filename = "GSE189039"
GSE189039 <- getGEO("GSE189039", GSEMatrix = TRUE) #Use GSEMatrix = TRUE para obter metadados de cada amostra
GSE189039 <- GSE189039[[1]] #Analisar o primeiro objeto da lista. 

#############Analisar informações sobre amostras. 

GSE189039_metadata = GSE189039 %>% 
  pData() %>% #Informações sobre amostras 
  select( #Selecionar e renomear colunas de interesse
    geo_sample = "geo_accession", 
    subject_id = "title", age = "age:ch1", 
    sex = "gender:ch1", disease_state = "diseas state:ch1", 
    severity = "severity:ch1") 

write.csv(GSE189039_metadata, file = "GSE189039_metadata.csv") #Terminar de anotar no excel



###### GSE201533
GSE201533 <- getGEO("GSE201533", GSEMatrix = TRUE) #Use GSEMatrix = TRUE para obter metadados de cada amostra

GSE201533 <- GSE201533[[1]] #Analisar o primeiro objeto da lista. 

#############Analisar informações sobre amostras. 

GSE201533_metadata = GSE201533 %>% 
  pData() %>% #Informações sobre amostras 
  select( #Selecionar e renomear colunas de interesse
    geo_sample = "geo_accession", 
    sample_id = "title", age = "age:ch1", 
    sex = "gender:ch1", 
    dose_vaccine = "vaccine:ch1")

GSE201533_metadata = GSE201533_metadata %>%
  separate(sample_id, into = c("participant_id", "time"), sep = "-D", remove = F) %>%
  separate(dose_vaccine, into = c("dose", "vaccine"), sep = ",\\s*(?=[A-Z])", extra = "merge") %>%
  mutate(dose = str_extract(dose, "\\d+"),
         vaccine = if_else(vaccine == "ChAdOx1", "ChAd", "BNT"),
         condition = paste0(vaccine, " (V", dose, ", D", time, ")"),
         geo = "GSE201533") 

write.csv(GSE201533_metadata, file = "GSE201533_metadata.csv")

```


```{r eval=FALSE, include=FALSE}
############ Obter Tabela de counts
#GEO supplementary files. Baixar individualmente. Arquivos grandes devem ser baixados direto do navegador.
getGEOSuppFiles('GSE201530') # Baixar tabela de counts
untar('GSE201530_RAW.tar') # Descompactar tar zipped

############ Metadados

filename = "GSE201530"
GSE201530 <- getGEO("GSE201530", GSEMatrix = TRUE) #Use GSEMatrix = TRUE para obter metadados de cada amostra
GSE201530 <- GSE201530[[1]] #Analisar o primeiro objeto da lista. 

#############Analisar informações sobre amostras. 

GSE201530_metadata = GSE201530 %>% 
  pData() %>% #Informações sobre amostras 
  select(
    geo_sample = "geo_accession", 
    subject_id = "title", 
    age = "age:ch1", 
    sex = "gender:ch1", 
    disease_state = "disease state:ch1", 
    timepoint = "days after positive pcr results:ch1",
    group = "group (by covid-19 vaccination, prior infection):ch1",
    variant = "omicron sublineage:ch1") 

write.csv(GSE201530_metadata, file = "GSE201530_metadata.csv")

#Anotar no excel com dados do artigo
```

*Ler arquivos e transformar em tabela*

Abra a lista do estudo e analise suas informações. Ao descompactar, surgem diversos arquivos ".featureCounts.txt.gz", que devem ser descompactados. Para descompactar a lista de arquivos, use um loop.

```{r eval=FALSE, include=FALSE}
#Descompactar todos os arquivos
diretorio <- "~/Desktop/RNA seq - Wasim/GSE189039"
arquivos <- list.files(path = diretorio, pattern = "*.gz$", full.names = TRUE)
for(arquivo_gz in arquivos) {gunzip(arquivo_gz)}

#Ler arquivos do estudo específico
diretorio_GSE201530 <- "~/Desktop/RNA seq - Wasim/GSE201530"
diretorio_GSE189039 <- "~/Desktop/RNA seq - Wasim/GSE189039"

arquivos_txt_GSE201530 <- list.files(path = diretorio_GSE201530, pattern = "*.txt$", full.names = TRUE)
arquivos_txt_GSE189039 <- list.files(path = diretorio_GSE189039, pattern = "*.txt$", full.names = TRUE)

# Função para ler e ajustar os arquivos
ler_e_ajustar <- function(arquivo) {
  dados <- read.table(arquivo, header = FALSE, sep = "\t")
  if (ncol(dados) >= 2) {
    colnames(dados)[1:2] <- c("genes", tools::file_path_sans_ext(basename(arquivo)))
    return(dados)
  } else {
    return(NULL)
  }
}

# Ler e ajustar todos os arquivos na lista
lista_dados_GSE201530 <- lapply(arquivos_txt_GSE201530, ler_e_ajustar)
lista_dados_GSE189039 <- lapply(arquivos_txt_GSE189039, ler_e_ajustar)

# Filtrar elementos NULL da lista e unir os dataframes
counts_gse201530 <- lista_dados_GSE201530 %>%
  purrr::discard(is.null) %>%
  reduce(full_join, by = "genes")

counts_GSE189039 <- lista_dados_GSE189039 %>%
  purrr::discard(is.null) %>%
  reduce(full_join, by = "genes")

# Renomear as colunas exceto a primeira ("genes")
colnames(counts_gse201530)[-1] <- gsub("^[^_]+_(.*)", "\\1", colnames(counts_gse201530)[-1])
colnames(counts_GSE189039)[-1] <- gsub("^[^_]+_(.*)", "\\1", colnames(counts_GSE189039)[-1])


# Ler e ajustar todos os arquivos na lista
lista_dados_GSE201530 <- lapply(arquivos_txt_GSE201530, ler_e_ajustar)
lista_dados_GSE189039 <- lapply(arquivos_txt_GSE189039, ler_e_ajustar)

# Filtrar elementos NULL da lista e unir os dataframes
counts_gse201530 <- lista_dados_GSE201530 %>%
  purrr::discard(is.null) %>%
  reduce(full_join, by = "genes")

counts_GSE189039 <- lista_dados_GSE189039 %>%
  purrr::discard(is.null) %>%
  reduce(full_join, by = "genes")

# Renomear as colunas exceto a primeira ("genes")
colnames(counts_gse201530)[-1] <- gsub("^[^_]+_(.*)", "\\1", colnames(counts_gse201530)[-1])
colnames(counts_GSE189039)[-1] <- gsub("^[^_]+_(.*)", "\\1", colnames(counts_GSE189039)[-1])

#Salvar
write.csv(counts_gse201530, file = "gse201530_counts.csv")
write.csv(counts_GSE189039, file = "gse189039_counts.csv")

```



# 4. Análise de dados


## 4.1.1 Demographics

Tratamento de metadados
```{r}
#Metadata ------
gse189039_metadata_2 = GSE189039_metadata_1_ %>% 
  separate(sample, into = c("participant_id"), sep = "_", remove = F) %>% 
  mutate(geo = "GSE189039") %>% 
  rename(day = timepoint_day,
         gsm_id = geo_sample) %>% 
  select(sample, participant_id)

gse199750_metadata_2 = gse199750_metadata_annotated_14_11_23_2_ %>% 
  rename(gsm_id = geo,
         participant_id = subject_id) %>% 
  mutate(geo = "GSE199750") %>% 
  select(sample, participant_id)

gse206023_metadata_3 = gse206023_metadata %>% 
  mutate(day = as.numeric(day),
         geo = "GSE206023") %>%  
  inner_join(gse206023_metadata_ready_2, by = "sample_id") %>% 
  inner_join(ann_vaccines_samples %>% 
            select(condition, sample_id = sample, age, sex, vaccine),
            by = "sample_id") %>% 
  rename(gsm_id = gsm_id.x) %>% 
  select(sample = sample_id, participant_id)

gse201530_metadata_2 = gse201530_metadata %>% 
  rename(gsm_id = geo_sample) %>% 
  separate(subject_id, into = c("variant", "participant_id_num"), sep = "_", remove = F) %>% 
  mutate(participant_id = paste0(variant, "_", participant_id_num)) %>% 
  select(participant_id, sample = subject_id)


gse201533_metadata_2 = GSE201533_metadata %>% 
  mutate(age = as.numeric(age),
         dose = as.numeric(dose),
         gsm_id = geo_sample) %>% 
  select(geo_sample = sample_id, participant_id) %>% 
  rownames_to_column("sample")


ann_vaccines_samples_2.1 = ann_vaccines_samples_2 %>% 
  filter(is.na(participant_id)) %>% 
  separate(sample, into = c("participant_id_1", "participant_id_2", "participant_id_3", "delete"), remove = F) %>% 
  mutate(participant_id = paste0(participant_id_1, "_", participant_id_2, "_", participant_id_3)) %>% 
  select(!c(participant_id_1, participant_id_2, participant_id_3)) %>% 
  select(sample, participant_id)

all_participants = bind_rows(gse189039_metadata_2,
          gse199750_metadata_2,
          gse206023_metadata_3,
          gse201533_metadata_2,
          gse201530_metadata_2) %>% 
  bind_rows(ann_vaccines_samples_2.1)

ann_vaccines_samples_2 = ann_vaccines_samples %>% 
  left_join(all_participants, by = "sample") %>% 
  select(condition, sample, age, sex, vaccine:previous_vaccination, participant_id) %>% 
  left_join(ann_vaccines_samples_2.1, by = "sample")


ann_vaccines_samples_participants = ann_vaccines_samples_2 %>% 
  select(-participant_id.y, participant_id = participant_id.x)
  

write.csv(ann_vaccines_samples_participants, file = "ann_vaccines_samples.csv")
```




```{r}
library(tidyverse)

demographics = ann_vaccines_samples_4_12_23 %>%
  select(vaccine, gse_id, condition, sample) %>% 
  distinct() %>% 
  group_by(vaccine, gse_id, condition) %>% 
  summarise(samples = n()) %>% 
  ungroup()

ann_vaccines = demographics %>% 
  select(condition, samples) %>% 
  inner_join(ann_vaccines_18_1_24, by = "condition") %>% 
  distinct() %>% 
  mutate(week = as.factor(week))

ann_vaccines_2 = ann_vaccines %>% 
  group_by(vaccine) %>% 
  summarise(samples_byvaccine = sum(samples)) %>% 
  ungroup() %>% 
  inner_join(ann_vaccines, by = "vaccine") 


 ann_vaccines_3 = ann_vaccines_2 %>% 
  group_by(gse_id) %>% 
  summarise(samples_bygse = sum(samples)) %>% 
  ungroup() %>% 
  inner_join(ann_vaccines_2, by = "gse_id") 

write.csv(ann_vaccines_3, file = "ann_vaccines_19-1-24.csv")
```

```{r}
ann_vaccines = ann_vaccines_19_1_24 


#### SAMPLES BY CONDITION -----
samples_by_condition  = ann_vaccines %>% 
  select(samples, condition, vaccine) %>% 
  distinct() %>% 
  group_by(condition) %>% 
  summarise(samples = sum(samples)) %>% 
  inner_join(ann_vaccines %>% select(vaccine, condition) %>% distinct(), by = "condition")


a = samples_by_condition %>% 
  ggplot(aes(x = fct_reorder(condition, samples), weight = samples, fill = vaccine)) +
  geom_bar() +
  scale_fill_manual(name = "Vaccine or infection", labels = c("BBIBP", 
                                                                "BNT", 
                                                                "ChAd",
                                                                "Healthy",
                                                                "Infection",
                                                                "MO",
                                                                "ZF2001"),
    values = c("BBIBP" = "#76c893",
                "ZF2001" = "#b5e48c",
                "BNT" = "#56cfe1",
                "MO" = "#72ddf7",
                "ChAd" = "#5e60ce",
                "I" = "#f72585",
                "H" = "grey80")) +
  theme_minimal() +
  theme(legend.position = "right",
        axis.text.x = element_blank(),
        axis.text.y = element_text(size = 11, color = "black", angle = 0),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank()) +
  geom_text(aes(label = samples, y = samples), 
            position = position_dodge(width = 0), 
            vjust = 0.5, 
            hjust = -0.2, 
            size = 4,
            angle = 0) +
  coord_flip() +
  ggtitle("Samples by condition") +
  labs(x = "") +
  ylim(0, 80)
a
ggsave(a, file = "demographics_samples-by-condition.png", width = 6, height = 10)

```


```{r}
#### AGE BY GSE -------------
age = ann_vaccines_samples_participants %>% 
  select(age, participant_id, gse_id) %>% 
  distinct()

age = ggbetweenstats(
  data = age,
  x = gse_id,
  y = age
) +
  ggplot2::scale_color_manual(
    values = c("GSE199750" = "#80ffdb",
               "GSE201533" = "#5e60ce", 
               "GSE201530" = "#7400b8",
               "GSE206023" = "#4361ee",
               "GSE189039" = "#48bfe3")) + 
  theme_minimal() +
  theme(legend.position = "right",
        axis.text.x = element_text(size = 10, color = "black", angle = 0),
        axis.text.y = element_text(size = 10, color = "black", angle = 0)) +
  labs(y = "Age")

ggsave(age, file = "Demographics_age.png", width = 10, height = 6)
```


```{r}
#### Participants by condition ------
Studies_Participants = Studies_Participants %>% 
  clean_names() 


participants = Studies_Participants %>% 
  clean_names() %>% 
  select(doses_abrev, participants, vac_infection) %>% 
  distinct() %>% 
  group_by(doses_abrev) %>% 
  summarise(n_total_participants = sum(participants)) %>% 
  inner_join(Studies_Participants, by = "doses_abrev") %>% 
  rename(condition = doses_abrev) %>% 
    select(condition, n_total_participants, vac_infection) %>% 
  distinct()


p = participants %>%
 filter(!is.na(condition)) %>% 
  ggplot(aes(x = fct_reorder(condition, desc(n_total_participants)), 
             weight = n_total_participants)) +
  geom_bar(fill = "#5e60ce") +
  theme_minimal() +
  scale_fill_manual(values = "#5e60ce") +
  theme(legend.position = "right",
        axis.text.x = element_text(size = 16, color = "black", angle = 45, vjust = 1, hjust = 1),
        axis.text.y = element_blank(),
        panel.grid.minor = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.major.x = element_line(color = "gray70",
                                          size = 0.1,
                                          linetype = 1)) +
  geom_text(aes(label = n_total_participants, y = n_total_participants), 
            position = position_dodge(width = 0), 
            vjust = -0.2, 
            hjust = 0.5, 
            size = 5,
            angle = 0) +
  ylim(0, 70)+
  ggtitle("Participants by condition") +
  labs(x = "Conditions",
       y = "Participants")
p
ggsave(p, file = "demographics_participants-by-condition.png", width = 12, height = 6)

```


## 4.1.2 Análise de expressão genica diferencial

*Padronizando tabela de counts*

```{r eval=FALSE, include=FALSE}

############### GSE201530
# Ordenar tabela
gse201530_counts_ready <- counts_gse201530 %>%
  column_to_rownames("genes") %>% 
  t() %>%
  as.data.frame() %>% 
  rownames_to_column("sample") %>% 
  arrange(sample) %>% 
  column_to_rownames("sample") %>% 
  t() %>% 
  as.data.frame() %>%
  rownames_to_column("genes") %>% 
  filter(!grepl("_", genes)) %>% 
  column_to_rownames("genes")

# Salvar arquivo com rownames
saveRDS(gse201530_counts_ready, file = "gse201530_counts_ready.rds")


############### GSE189039
# Ordenar tabela
gse189039_counts_ready <- counts_GSE189039 %>%
  column_to_rownames("genes") %>% 
  t() %>%
  as.data.frame() %>% 
  rownames_to_column("sample") %>% 
  arrange(sample) %>% 
  column_to_rownames("sample") %>% 
  t() %>% 
  as.data.frame() %>%
  rownames_to_column("genes") %>% 
  filter(!grepl("_", genes)) %>% 
  column_to_rownames("genes")

# Salvar arquivo com rownames
saveRDS(gse189039_counts_ready, file = "gse189039_counts_ready.rds")

GSE189039_metadata

```

*Metadata*

A tabela de counts fornecida pelo estudo é nomeada com o Subject ID + Timepoint e não com o codigo GSM. Por isso, teremos de tratar a tabela de metadados. Além disso, ela não possui a coluna de idade. É possível manipular estes dados no excel.


*DESEQ2*

*Preparando dados*


## GSE201530

```{r}
###DESEQ2 - PF versus AZ, PF em diferentes tempos, AZ em diferentes tempos

####Definindo design

###### Padronize as variáveis 
countData <- gse201530_counts_ready
colData <- GSE201530_metadata %>%
  mutate(
    timepoint_day = factor(timepoint_day),
    age = factor(age),
    sex = factor(sex),
    disease_vac = factor(disease_vac),
    dose = factor(dose),
    disease_time = paste0(disease_vac, "_D", timepoint_day)) %>% 
  select(subject_id, everything())


###### Crie o objeto DESeqDataSet ######
# O design deve incluir todos os fatores que podem influenciar significativamente os dados. Coloque por último o fator de interesse principal - no caso, o fator de primeira ou segunda doses, pois nesta primeira etapa não serão analisados os efeitos de uma terceira dose de vacina heteróloga.

dds <- DESeqDataSetFromMatrix(countData = countData, 
                              colData = colData, 
                              design = ~ disease_time)

###### Pré-filtragem ######
#Mantenha os genes com pelo menos 10 leituras no total.

keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]

```

*Executando DESEQ2*

```{r}
###### Executar DESeq ######

#Executar DESeq com método Wald
dds <- DESeq(dds)

#Salvar
write_rds(dds,file="gse201530_dds_DESeq2_Wald_disease_vac_time.rds")

###### Obter comparações realizadas ######
resultsNames(dds)

dds = gse201530_dds_DESeq2_Wald_disease_vac_timepoint_day

###### Obtendo os DEGs totais ######
res <- results(dds) %>% 
  as.data.frame() %>% 
  filter(pvalue < 0.05)

#Salvar tabela
write.csv(res, file = "gse201530_dds_DESeq2_Wald_disease_vac_AND_timepoint_day_ALLDEGS.rds", row.names = TRUE)

```

*Comparando tempos e condições*

Nesta comparação, são encontradas DEGs de cada vacina individual nos diferentes tempos. Para comparar uma vacina com a outra em diferentes tempos, é necessário usar os argumentos "contrast" ou "name".

```{r}
#H-BNT-I
# contrast = c("disease_time", "H.BNT.I_D1", "H_D0")
# contrast = c("disease_time", "H.BNT.I_D2", "H_D0") (NAO DEU)
# contrast = c("disease_time", "H.BNT.I_D3", "H_D0")
# contrast = c("disease_time", "H.BNT.I_D4", "H_D0")

#I-BNT-I
# contrast = c("disease_time", "I.BNT.I_D2", "H_D0")
# contrast = c("disease_time", "I.BNT.I_D5", "H_D0")

#H-I
# contrast = c("disease_time", "H.I_D1", "H_D0")

#H-I-I
# contrast = c("disease_time", "H.I.I_D2", "H_D0")
#I-I
# contrast = c("disease_time", "I.I_D1", "H_D0")
# contrast = c("disease_time", "I.I_D3", "H_D0")
# contrast = c("disease_time", "I.I_D5", "H_D0")
```

*H_BNT_I* 
```{r}
######### H_BNT_I

#H_BNT_I_D1
gse201530_H_BNT_I_D1 = results(dds, contrast = c("disease_time", "H.BNT.I_D1", "H_D0")) %>% 
  as.data.frame() %>% 
  arrange(padj) %>% 
  mutate(condition = "H-BNT-I (D1)",
         day = 1,
         vaccine = "H-BNT-I",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse201530_H_BNT_I_D1, file = "gse201530_H_BNT_I_D1_DEGs_notfiltered.csv", row.names = TRUE)

#H_BNT_I_D3
gse201530_H_BNT_I_D3 = results(dds, contrast = c("disease_time", "H.BNT.I_D3", "H_D0")) %>%
  as.data.frame() %>%
  arrange(padj) %>%
  mutate(condition = "H-BNT-I (D3)",
         day = 3,
         vaccine = "H-BNT-I",
         direction = ifelse(log2FoldChange >= 1, "UP",
                            ifelse(log2FoldChange <= -1, "DOWN",
                                   "NEUTRAL"))) %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse201530_H_BNT_I_D3, file = "gse201530_H_BNT_I_D3_DEGs.csv", row.names = TRUE)

#H_BNT_I_D4
gse201530_H_BNT_I_D4 = results(dds, contrast = c("disease_time", "H.BNT.I_D4", "H_D0")) %>%
  as.data.frame() %>%
  filter(padj < 0.05) %>%
  arrange(padj) %>%
  mutate(condition = "H-BNT-I (D4)",
         day = 4,
         vaccine = "H-BNT-I",
         direction = ifelse(log2FoldChange >= 1, "UP",
                            ifelse(log2FoldChange <= -1, "DOWN",
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse201530_H_BNT_I_D4, file = "gse201530_H_BNT_I_D4_DEGs.csv", row.names = TRUE)

gse201530_H_BNT_I = bind_rows(gse201530_H_BNT_I_D1,
                            gse201530_H_BNT_I_D3,
                            gse201530_H_BNT_I_D4)

write.csv(gse201530_H_BNT_I, "gse201530_H_BNT_I.csv")

```


*I_BNT_I* 
```{r}
######### I_BNT_I

#D2
gse201530_I_BNT_I_D2 = results(dds, contrast = c("disease_time", "I.BNT.I_D2", "H_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "I-BNT-I (D2)",
         day = 2,
         vaccine = "I-BNT-I",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse201530_I_BNT_I_D2, file = "gse201530_I_BNT_I_D2_DEGs.csv")


#D5
gse201530_I_BNT_I_D5 = results(dds, contrast = c("disease_time", "I.BNT.I_D5", "H_D0")) %>%
  as.data.frame() %>%
  filter(padj < 0.05) %>%
  arrange(padj) %>%
  mutate(condition = "I-BNT-I (D5)",
         day = 5,
         vaccine = "I-BNT-I",
         direction = ifelse(log2FoldChange >= 1, "UP",
                            ifelse(log2FoldChange <= -1, "DOWN",
                                   "NEUTRAL"))) %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse201530_I_BNT_I_D5, file = "gse201530_H_BNT_I_D5_DEGs.csv")

gse201530_I_BNT_I = bind_rows(gse201530_I_BNT_I_D2,
                            gse201530_I_BNT_I_D5)

write.csv(gse201530_I_BNT_I, "gse201530_I_BNT_I.csv")

```


*H-I* 
```{r}
######### H-I

#D1
gse201530_H_I_D1 = results(dds, contrast = c("disease_time", "H.I_D1", "H_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "H-I (D1)",
         day = 1,
         vaccine = "H-I",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse201530_H_I_D1, file = "gse201530_H_I_D1_DEGs-PF.csv")


#D5
gse201530_H_I_D5 = results(dds, contrast = c("disease_time", "I.BNT.I_D5", "H_D0")) %>%
  as.data.frame() %>%
  filter(padj < 0.05) %>%
  arrange(padj) %>%
  mutate(condition = "H-I (D5)",
         day = 5,
         vaccine = "H-I",
         direction = ifelse(log2FoldChange >= 1, "UP",
                            ifelse(log2FoldChange <= -1, "DOWN",
                                   "NEUTRAL"))) %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse201530_H_I_D5, file = "gse201530_H_I_D5_DEGs-PF.csv")

gse201530_H_I = bind_rows(gse201530_I_BNT_I_D2,
                            gse201530_I_BNT_I_D5)

write.csv(gse201530_H_I, "gse201530_H_I.csv")

```

*H-I-I* 
```{r}

######### H-I-I
#D2
gse201530_H_I_I_D2 = results(dds, contrast = c("disease_time", "H.I.I_D2", "H_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "H-I-I (D2)",
         day = 2,
         vaccine = "H-I-I",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse201530_H_I_I_D2, file = "gse201530_H_I_I_D2_DEGs-PF.csv")

```

*I-I*
```{r}

######### I-I
#D1
gse201530_I_I_D1 = results(dds, contrast = c("disease_time", "I.I_D1", "H_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "I-I (D1)",
         day = 1,
         vaccine = "I-I",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse201530_I_I_D1, file = "gse201530_I_I_D1_DEGs-PF.csv")


#D3
gse201530_I_I_D3 = results(dds, contrast = c("disease_time", "I.I_D3", "H_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "I-I (D3)",
         day = 3,
         vaccine = "I-I",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse201530_I_I_D3, file = "gse201530_I_I_D3_DEGs-PF.csv")


#D5
gse201530_I_I_D5 = results(dds, contrast = c("disease_time", "I.I_D5", "H_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "I-I (D5)",
         day = 5,
         vaccine = "I-I",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse201530_I_I_D5, file = "gse201530_I_I_D5_DEGs-PF.csv")

gse201530_I_I = bind_rows(gse201530_I_I_D1,
                          gse201530_I_I_D3,
                          gse201530_I_I_D5)

```

Unir raw datasets 
```{r}
########## GSE201530

gse201530_DEGs = bind_rows(gse201530_I_I,
                           gse201530_H_I_I_D2,
                           gse201530_H_I,
                           gse201530_I_BNT_I,
                           gse201530_H_BNT_I) %>% 
  mutate(study = "GSE201530")

#Salvar
write.csv(gse201530_DEGs, file = "gse201530_DEGs_27-11-23.csv")

GSE201530_metadata_conditions = gse201530_DEGs %>% 
  select(condition:vaccine, study) %>% 
  distinct()


#Anotações
ann_vaccines_lower = ann_vaccines %>% 
  clean_names()

ann_vaccines_27_11_23 = bind_rows(ann_vaccines_lower, 
                                  GSE201530_metadata_conditions)

#Salvar
write.csv(ann_vaccines_27_11_23, file = 'ann_vaccines_27_11_23.csv')


```






## GSE189039

```{r}
###DESEQ2 - PF versus AZ, PF em diferentes tempos, AZ em diferentes tempos

####Definindo design

###### Padronize as variáveis 
countData <- gse189039_counts_ready
colData <- GSE189039_metadata %>%
  mutate(disease_vac = ifelse(disease_vac == "I", "H-I",
                              ifelse(disease_vac == "BNT-I", "H-BNT-I", disease_vac)),
         disease_vac_severity_time = paste0(disease_vac, "_", severity, "_D", timepoint_day),
         disease_vac_severity_time = factor(disease_vac_severity_time)) %>% 
  select(subject_id, everything())


###### Crie o objeto DESeqDataSet ######
dds <- DESeqDataSetFromMatrix(countData = countData, 
                              colData = colData, 
                              design = ~ disease_vac_severity_time)

###### Pré-filtragem ######
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]

###### Executar DESeq ######

#Método Wald
dds <- DESeq(dds)

#Salvar
write_rds(dds,file="gse189039_dds_DESeq2_Wald_disease_vac_severity_time.rds")

###### Obter comparações realizadas ######
resultsNames(dds)

###### Obtendo os DEGs totais ######
res <- results(dds) %>% 
  as.data.frame() %>% 
  filter(pvalue < 0.05)

#Salvar tabela
write.csv(res, file = "gse189039_dds_DESeq2_Wald_disease_vac_severity_time_ALLDEGS.rds")

```

*Comparando tempos e condições*

Nesta comparação, são encontradas DEGs de cada vacina individual nos diferentes tempos. Para comparar uma vacina com a outra em diferentes tempos, é necessário usar os argumentos "contrast" ou "name".

```{r}

## H-I
# MODERATE
# contrast = c("disease_vac_severity_time", "H.I_moderate_D10", "H_naive_vac_D0")
# contrast = c("disease_vac_severity_time", "H.I_moderate_D26", "H_naive_vac_D0")
# contrast = c("disease_vac_severity_time", "H.I_moderate_D51", "H_naive_vac_D0")

# SEVERE
# contrast = c("disease_vac_severity_time", "H.I_severe_D10", "H_naive_vac_D0")
# contrast = c("disease_vac_severity_time", "H.I_severe_D26", "H_naive_vac_D0")
# contrast = c("disease_vac_severity_time", "H.I_severe_D51", "H_naive_vac_D0")


## H-BNT-I
# MILD
# contrast = c("disease_vac_severity_time", "H.BNT.I_mild_D10", "H_naive_vac_D0")
# contrast = c("disease_vac_severity_time", "H.BNT.I_mild_D26", "H_naive_vac_D0")

# SEVERE
# contrast = c("disease_vac_severity_time", "H.BNT.I_severe_D10", "H_naive_vac_D0")
# contrast = c("disease_vac_severity_time", "H.BNT.I_severe_D26", "H_naive_vac_D0")

## BNT-I-BNT
# contrast = c("disease_vac_severity_time", "BNT.I.BNT_severe_D51", "H_naive_vac_D0")
contrast = c("disease_vac_severity_time", "BNT.I.BNT_mild_D51", "H_naive_vac_D0") 

```

*H_I* 
```{r}
######### H_I

#H_I_D10
gse189039_H_I_moderate_D10 = results(dds, contrast = c("disease_vac_severity_time", "H.I_moderate_D10", "H_naive_vac_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "H-I (D10, moderate)",
         day = 10,
         vaccine = "H-I",
         severity = "moderate",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse189039_H_I_moderate_D10, file = "gse189039_H_I_moderate_D10_DEGs.csv")

#H_I_D26
gse189039_H_I_moderate_D26 = results(dds, contrast = c("disease_vac_severity_time", "H.I_moderate_D26", "H_naive_vac_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "H-I (D26, moderate)",
         day = 26,
         severity = "moderate",
         vaccine = "H-I",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse189039_H_I_moderate_D26, file = "gse189039_H_I_moderate_D26_DEGs.csv")

#H_I_D51
gse189039_H_I_moderate_D51 = results(dds, contrast = c("disease_vac_severity_time", "H.I_moderate_D51", "H_naive_vac_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "H-I (D51, moderate)",
         day = 51,
         vaccine = "H-I",
         severity = "moderate",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse189039_H_I_moderate_D51, file = "gse189039_H_I_moderate_D51_DEGs.csv")

#H_I_severe_D10
gse189039_H_I_severe_D10 = results(dds, contrast = c("disease_vac_severity_time", "H.I_severe_D10", "H_naive_vac_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "H-I (D10, severe)",
         day = 10,
         vaccine = "H-I",
         severity = "severe",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse189039_H_I_severe_D10, file = "gse189039_H_I_severe_D10_DEGs.csv")



#H_I_severe_D26
gse189039_H_I_severe_D26 = results(dds, contrast = c("disease_vac_severity_time", "H.I_severe_D26", "H_naive_vac_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "H-I (D26, severe)",
         day = 26,
         vaccine = "H-I",
         severity = "severe",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse189039_H_I_severe_D26, file = "gse189039_H_I_severe_D26_DEGs.csv")

#H_I_severe_D51
gse189039_H_I_severe_D51 = results(dds, contrast = c("disease_vac_severity_time", "H.I_severe_D51", "H_naive_vac_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "H-I (D51, severe)",
         day = 51,
         vaccine = "H-I",
         severity = "severe",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse189039_H_I_severe_D51, file = "gse189039_H_I_severe_D51_DEGs.csv")

gse189039_H_I = bind_rows(gse189039_H_I_moderate_D10,
                          gse189039_H_I_severe_D10,
                            gse189039_H_I_moderate_D26,
                          gse189039_H_I_severe_D26,
                            gse189039_H_I_moderate_D51,
                          gse189039_H_I_severe_D51)

write.csv(gse189039_H_I, "gse189039_H_I_DEGS.csv")

```


*H_BNT_I* 
```{r}
######### H-BNT-I

#H_BNT-I

#MILD
gse189039_H_BNT_I_mild_D10 = results(dds, contrast = c("disease_vac_severity_time", "H.BNT.I_mild_D10", "H_naive_vac_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "H-BNT-I (D10, mild)",
         day = 10,
         vaccine = "H-BNT-I",
         severity = "mild",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse189039_H_BNT_I_mild_D10, file = "gse189039_H_BNT_I_mild_D10_DEGs.csv")


gse189039_H_BNT_I_mild_D26 = results(dds, contrast = c("disease_vac_severity_time", "H.BNT.I_mild_D26", "H_naive_vac_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "H-BNT-I (D26, mild)",
         day = 26,
         vaccine = "H-BNT-I",
         severity = "mild",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse189039_H_BNT_I_mild_D26, file = "gse189039_H_BNT_I_mild_D26_DEGs.csv")


#SEVERE

gse189039_H_BNT_I_severe_D10 = results(dds, contrast = c("disease_vac_severity_time", "H.BNT.I_severe_D10", "H_naive_vac_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "H-BNT-I (D10, severe)",
         day = 10,
         vaccine = "H-BNT-I",
         severity = "severe",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse189039_H_BNT_I_severe_D10, file = "gse189039_H_BNT_I_severe_D10_DEGs.csv")


gse189039_H_BNT_I_severe_D26 = results(dds, contrast = c("disease_vac_severity_time", "H.BNT.I_severe_D26", "H_naive_vac_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "H-BNT-I (D26, severe)",
         day = 26,
         vaccine = "H-BNT-I",
         severity = "severe",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse189039_H_BNT_I_severe_D26, file = "gse189039_H_BNT_I_severe_D26_DEGs.csv")

#UNIR
gse189039_H_BNT_I = bind_rows(gse189039_H_BNT_I_mild_D10,
                              gse189039_H_BNT_I_severe_D10,
                              gse189039_H_BNT_I_mild_D26,
                              gse189039_H_BNT_I_severe_D26)
#Salvar
write.csv(gse189039_H_BNT_I, file = "gse189039_H_BNT_I_DEGS.csv")

```


*BNT-I-BNT*
```{r}
######### BNT-I-BNT	

# D51
gse189039_BNT_I_BNT_severe_D51 = results(dds, contrast = c("disease_vac_severity_time",
                                                         "BNT.I.BNT_severe_D51", 
                                                         "H_naive_vac_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "BNT-I-BNT (D51, severe)",
         day = 51,
         vaccine = "BNT-I-BNT",
         severity = "severe",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse189039_BNT_I_BNT_severe_D51, file = "gse189039_BNT_I_BNT_severe_D51_DEGs.csv")


#Aqui tive de usar o argumento name, pois contrast não estava funcionando (?)
gse189039_BNT_I_BNT_mild_D51 = results(dds, name= "disease_vac_severity_time_H_naive_vac_D0_vs_BNT.I.BNT_mild_D51") %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  mutate(log2FoldChange = ifelse(log2FoldChange > 0, -log2FoldChange, abs(log2FoldChange))) %>% 
  arrange(padj) %>% 
  mutate(condition = "BNT-I-BNT (D51, mild)",
         day = 51,
         vaccine = "BNT-I-BNT",
         severity = "mild",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes") 

#Salvar tabela
write.csv(gse189039_BNT_I_BNT_mild_D51, file = "gse189039_BNT_I_BNT_mild_D51_DEGs.csv")

#Unir
gse189039_BNT_I_BNT = bind_rows(gse189039_BNT_I_BNT_mild_D51,
                                gse189039_BNT_I_BNT_severe_D51)


write.csv(gse189039_BNT_I_BNT, file = "gse189039_BNT_I_BNT_DEGS.csv")

```

*UNIR*
```{r}
gse189039_DEGs = bind_rows(gse189039_BNT_I_BNT,
                           gse189039_H_I,
                           gse189039_H_BNT_I) %>% 
  mutate(study = "GSE189039")



##### TODAS AS DEGS

gse_201530_189039 = bind_rows(gse201530_DEGs,
                              gse189039_DEGs) %>% 
  select(genes:condition, direction) %>% 
  merge(ann_vaccines, by = "condition", all.y = F) %>% 
  clean_names()

degs_allstudies_updown_p_05 = degs_allstudies_updown_p_05 %>% 
  clean_names() %>% 
  select(genes:direction, -vaccine, -gse_id) %>% 
  merge(ann_vaccines, by = "condition", all.y = F) %>% 
  select(-participants)


all_degs_p_05_vac_infected = bind_rows(degs_allstudies_updown_p_05,
                                       gse_201530_189039) %>% 
  mutate_all(~ replace(., is.na(.), "NA")) %>% 
  mutate(regimen = case_when(
regimen == "BNT-I-BNT" ~ "V-I-V",
regimen == "H-BNT-I" ~ "V-I",
regimen == "VAC-I" ~ "V-I",
regimen == "I-VAC-I" ~ "I-V-I",
regimen == "H-V-I" ~ "V-I",
regimen == "H-I-I" ~ "I-I",
regimen == "H-I" ~ "I",
TRUE ~ as.character(regimen)))

#Salvar
write.csv(gse_201530_189039, file = "gse_201530_189039_DEGs_27-11-23.csv")

#Salvar
write.csv(all_degs_p_05_vac_infected, file = 'all_degs_p_05_vac_infected_27-11-23.csv')
```

Comparar numero de participantes
```{r}

ann_vaccines = ann_vaccines_29_11_23 %>% 
  clean_names() %>% 
  mutate(condition = factor(condition, levels = rev(condition)))


plot = ann_vaccines %>% 
  ggplot() +
  aes(x = condition, 
      y = participants, 
      fill = gse_id) +
  geom_col() +
  labs(title = "Participants by condition") +
  coord_flip() +
  theme_minimal() +
  ylim(0, 70) +
  scale_fill_manual(
    values = c("GSE199750" = "#80ffdb",
               "GSE201533" = "#5e60ce", 
               "GSE201530" = "#7400b8",
               "GSE206023" = "#4361ee",
               "GSE189039" = "#48bfe3")) +
  geom_text(aes(label = participants, y = participants), 
            position = position_dodge(width = 0.9), 
            vjust = 0.5, 
            hjust = -0.2, 
            size = 3)

ggsave(plot, file = "Samples_per_condition_by_GSE.png", width = 6, height = 10)
ggsave(plot, file = "Samples_per_condition_by_GSE_2.png", width = 10, height = 6)

```

Comparar sexo
```{r}

esquisser(Demographics)

Sex = Demographics %>%
 filter(!(Sub %in% "Mean")) %>%
 ggplot() +
  aes(x = Study, y = Value, fill = Sub) +
  geom_col() +
  scale_fill_manual(
    values = c(Female = "#BA27FF",
    Male = "#1EC3FF")) +
  theme_minimal() +
  facet_wrap(vars(Category)) +
  geom_text(aes(label = Value, y = Value), 
            vjust = 2, 
            hjust = 0.5, 
            size = 4) +
  xlab("GSE ID") +
  ylab("Percentage") +
  labs(title = "Sex",
       fill = "Sex")

ggsave(Sex, file = "Demographics_Sex_1.png", width = 10, height = 6)
ggsave(Sex, file = "Demographics_Sex_2.png", width = 6, height = 10)



Age = Demographics %>%
 filter(!(Category %in% "Sex")) %>%
 ggplot() +
  aes(x = Study, y = Value, fill = Sub) +
  geom_col() +
  scale_fill_manual(
    values = c(Mean = "#1EC3FF")) +
  theme_minimal() +
  facet_wrap(vars(Category)) +
  geom_text(aes(label = Value, y = Value), 
            vjust = 2, 
            hjust = 0.5, 
            size = 4) +
  xlab("GSE ID") +
  ylab("Age") +
  labs(fill = "Age")

ggsave(Age, file = "Demographics_Age_1.png", width = 10, height = 6)
ggsave(Age, file = "Demographics_Age_2.png", width = 6, height = 10)

```


```{r}
ann_vaccines = ann_vaccines_29_11_23 %>% 
  clean_names() %>% 
  mutate(condition = factor(condition, levels = rev(condition)))

plot = ann_vaccines %>% 
  ggplot() +
  aes(x = condition, 
      y = participants, 
      fill = gse_id) +
  geom_col() +
  labs(title = "Participants by condition") +
  coord_flip() +
  theme_minimal() +
  ylim(0, 70) +
  scale_fill_manual(
    values = c("GSE199750" = "#80ffdb",
               "GSE201533" = "#5e60ce", 
               "GSE201530" = "#7400b8",
               "GSE206023" = "#4361ee",
               "GSE189039" = "#48bfe3")) +
  geom_text(aes(label = participants, y = participants), 
            position = position_dodge(width = 0.9), 
            vjust = 0.5, 
            hjust = -0.2, 
            size = 3)

ggsave(plot, file = "Samples_per_condition_by_GSE.png", width = 6, height = 10)
ggsave(plot, file = "Samples_per_condition_by_GSE_2.png", width = 10, height = 6)


```



Combinar tabela de genes sequenciados (raw)

```{r}
#TABELA DE GENES SEQUENCIADOS

gse201530_genes = gse201530_counts_ready %>%
  rownames_to_column("genes") %>% 
  select(genes) %>% 
  mutate(study = "GSE201530")

gse199750_genes = gse199750_counts_ready %>%
  as.data.frame() %>% 
  rownames_to_column("genes") %>% 
  select(genes) %>% 
  mutate(study = "GSE199750")

gse189039_genes = gse189039_counts_ready %>%
  as.data.frame() %>% 
  rownames_to_column("genes") %>% 
  select(genes) %>% 
  mutate(study = "GSE189039")

gse201533_genes = gse201533_counts_ready %>%
  as.data.frame() %>% 
  rownames_to_column("genes") %>% 
  select(genes) %>% 
  mutate(study = "GSE201533")

gse206023_genes = gse206023_counts_long_annotated_14_11_23 %>% 
  select(genes) %>% 
  distinct() %>% 
  mutate(study = "GSE206023")


genes_raw_combined = bind_rows(gse206023_genes,
                               gse201533_genes,
                               gse189039_genes,
                               gse199750_genes,
                               gse201530_genes)


write.csv(genes_raw_combined, file = "genes_raw_combined.csv")

```


# 5. Visualização de dados

## DEGs plot

```{r}

ann_vaccines = ann_vaccines_27_11_23 %>% clean_names()

all_degs = all_degs_p_05_vac_infected %>% 
  group_by(condition, direction) %>% 
  summarise(DEGs = n()) %>% 
  merge(ann_vaccines, by="condition", all.y = F)

all_degs_wide = all_degs %>% 
  pivot_wider(names_from = "direction", values_from = "DEGs") %>% 
  mutate(TOTAL = sum(DOWN + NEUTRAL + UP)) %>% 
  merge(ann_vaccines, by="condition", all.y = F)

write.csv(all_degs_wide, file = "all_degs_wide.csv")



# Plot 1
ggplot_degs_bars = all_degs %>%
 filter(!(direction %in% "NEUTRAL")) %>%
 ggplot() +
  aes(x = reorder(condition, day), fill = direction, weight = DEGs) +
  geom_bar(position = "dodge") +
  scale_fill_manual(
    values = c(DOWN = "#0E252D",
    NEUTRAL = "#00C19F",
    UP = "#3399FF")) +
  theme_minimal() +
  facet_wrap(vars(vac_infec), scales = "free", nrow = 2) +
  geom_text(aes(label = DEGs, y = DEGs), 
            position = position_dodge(width = 0.9), 
            vjust = -1, 
            hjust = 0.5, 
            size = 2) +
  theme(legend.position = "top",
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1, size = 7),
        axis.text.y = element_text(vjust = 2)) +
  ggtitle("DEGs - all studies, p<0.05")

ggsave(ggplot_degs_bars, file = "DEGs_up_down_barplot_pvalue05_1.png",  width = 10, height = 6) 


esquisser(all_degs)

#Plot 2
ggplot_degs_bars_2 = all_degs %>%
 filter(!(direction %in% "NEUTRAL")) %>%
 ggplot(aes(x = condition, fill = direction, weight = DEGs)) +
  geom_bar(position = "dodge") +
  scale_fill_manual(
    values = c(DOWN = "#0E252D",
    NEUTRAL = "#00C19F",
    UP = "#3399FF")) +
  theme_minimal() +
  facet_wrap(vars(vaccine), scales = "free", nrow = 2) +
  geom_text(aes(label = DEGs, y = DEGs), 
            position = position_dodge(width = 0.9), 
            vjust = -1, 
            hjust = 0.5, 
            size = 2) +
  theme(legend.position = "top",
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1, size = 7),
        axis.text.y = element_text(vjust = 2)) +
  ggtitle("DEGs - all studies, p<0.05")

ggsave(ggplot_degs_bars_2, file = "DEGs_up_down_barplot_pvalue05_2.png",  width = 10, height = 6) 



#Plot 3
ggplot_degs_bars_3 = all_degs %>%
 filter(!(direction %in% "NEUTRAL")) %>%
 ggplot() +
  aes(x = condition, fill = direction, weight = DEGs) +
  geom_bar(position = "dodge") +
  scale_fill_manual(
    values = c(DOWN = "#0E252D",
    NEUTRAL = "#00C19F",
    UP = "#3399FF")) +
  theme_minimal() +
  facet_wrap(vars(vaccine), scales = "free", nrow = 7) +
  geom_text(aes(label = DEGs, y = DEGs), 
            position = position_dodge(width = 0.9), 
            vjust = -1, 
            hjust = 0.5, 
            size = 2) +
  theme(legend.position = "top",
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1, size = 7),
        axis.text.y = element_text(vjust = 2)) +
  ggtitle("DEGs - all studies, p<0.05")

ggsave(ggplot_degs_bars_3, file = "DEGs_up_down_barplot_pvalue05_3.png",  width = 5, height = 20)


#Plot 4
ggplot_degs_bars_4 = all_degs %>%
 filter(direction !="NEUTRAL",
        vac_infec != "vaccine") %>%
 ggplot() +
  aes(x = condition, fill = direction, weight = DEGs) +
  geom_bar(position = "dodge") +
  scale_fill_manual(
    values = c(DOWN = "#0E252D",
    NEUTRAL = "#00C19F",
    UP = "#3399FF")) +
  theme_minimal() +
  facet_grid(vars(prior_infection), vars(previous_vaccination), scales = "free") +
  geom_text(aes(label = DEGs, y = DEGs), 
            position = position_dodge(width = 0.9), 
            vjust = -1, 
            hjust = 0.5, 
            size = 2) +
  theme(legend.position = "top",
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1, size = 7),
        axis.text.y = element_text(vjust = 2)) +
  ggtitle("DEGs - all studies, p<0.05")

ggsave(ggplot_degs_bars_4, file = "DEGs_up_down_barplot_pvalue05_4.png",  width = 10, height = 6)

```

## 5.1. Volcanoplot

Os volcanoplots das duas vacinas são os mesmos, mas espelhados. Isso significa que definir o nivel de referencia no fator de vacinas é importante para determinar as DEGs up e down. \#### ASTRAZENECA

*Astrazeneca* 

```{r}

#All
volcano_res <- EnhancedVolcano(res, 
                                    lab = rownames(res),
                                    x = 'log2FoldChange',
                                    y = 'pvalue',
                                    title='Differential expression of genes - All timepoints - ChAdOx1')
# Save the plot as a PNG file
png("gse199750_volcanoplot_DEGs_allfactores_AZ.png", width = 800, height = 600)
print(volcano_res_chad)
dev.off()

#T1-T0
volcano_res_chad_1_0 <- EnhancedVolcano(res_1_0_chad, 
                                    lab = rownames(res_1_0_chad),
                                    x = 'log2FoldChange',
                                    y = 'pvalue',
                                    title='Differential expression of genes - D1 vs. D0 - ChAdOx1')
# Save the plot as a PNG file
png("gse199750_volcanoplot_DEGs_D0-D1_AZ.png", width = 800, height = 600)
print(volcano_res_chad_1_0)
dev.off()

#T2-T0
volcano_res_chad_2_0 <- EnhancedVolcano(res_time0_time2_chad, 
                                    lab = rownames(res_time0_time2_chad),
                                    x = 'log2FoldChange',
                                    y = 'pvalue',
                                    title='Differential expression of genes - D2 vs. D0 - ChAdOx1')
# Save the plot as a PNG file
png("gse199750_volcanoplot_DEGs_D0-D2_AZ.png", width = 800, height = 600)
print(volcano_res_chad_2_0)
dev.off()

#Plotar mais de um gráfico
plot_grid(volcano_res_chad, volcano_res_chad_1_0, volcano_res_chad_2_0, ncol = 3) 

```


## 5.2. Diagrama de Venn com Heatmap

*Tabela de overlaps de DEGs com teste exato de fisher*

Determinar a proporção de DEGs compartilhadas entre condições diferentes e sua significância. O código abaixo deve ter como input uma tabela longa chamada "degs_all" (coluna 1: condicoes/vacinas, coluna 2: DEGs). O output será uma tabela longa com as colunas Vacina 1, Vacina 2, shared degs, not shared vacina 1, not shared vacina 2, lista com nomes das degs, valor p de fisher (teste exato de fisher).




```{r}

# DEGs table
degs_all_updown = all_degs_p_05_vac_infected_19_12_23 %>% 
  filter(direction != "NEUTRAL") %>% 
  mutate_all(~ replace(., is.na(.), "NA"))

# Annotation table
immunego_genes = ImmuneGO_Annotated_Genes_26_2_24_2_ 

# DOWN-UP ALL
degs_venn_updown = degs_all_updown %>% 
  filter(direction %in% c("UP", "DOWN")) %>%
  mutate(comparison = "UP_AND_DOWN") %>% 
  select(condition, genes)

degs_updown_immune = degs_all_updown %>% 
  filter(direction %in% c("UP", "DOWN")) %>%
  mutate(comparison = "UP_AND_DOWN") %>% 
  select(condition, genes) %>% 
  inner_join(immunego_genes %>% select(genes), by = "genes") %>% 
  distinct()

degs_updown_nonimmune = degs_all_updown %>% 
  filter(direction %in% c("UP", "DOWN")) %>%
  mutate(comparison = "UP_AND_DOWN") %>% 
  select(condition, genes) %>% 
  anti_join(immunego_genes %>% select(genes), by = "genes") %>% 
  distinct()

#INPUT
data = degs_updown_nonimmune 
filename = "degs_updown_nonimmune_SharedDEGS_Venn"

```


```{r}
# Função para calcular a sobreposição entre pares de vacinas e a porcentagem de genes compartilhados ----
overlap_genes <- function(cond1, cond2, data) {
  genes_cond1 <- data$genes[data$condition == cond1] #Trocar coluna
  genes_cond2 <- data$genes[data$condition == cond2] #Trocar coluna
  genes_shared <- intersect(genes_cond1, genes_cond2)
  genes_notshared_cond1 <- setdiff(genes_cond1, genes_cond2)
  genes_notshared_cond2 <- setdiff(genes_cond2, genes_cond1)
  
  total_genes_cond1 <- length(genes_cond1)
  total_genes_cond2 <- length(genes_cond2)
  
  percentage_shared_cond1 <- length(genes_shared) / total_genes_cond1 * 100
  percentage_shared_cond2 <- length(genes_shared) / total_genes_cond2 * 100
  
  shared_genes <- data.frame(
    Cond1 = cond1,
    Cond2 = cond2,
    Shared = length(genes_shared),
    NotShared_cond1 = length(genes_notshared_cond1),
    NotShared_cond2 = length(genes_notshared_cond2),
    Total_Genes_Cond1 = total_genes_cond1,
    Total_Genes_Cond2 = total_genes_cond2,
    Genes_Names = paste(genes_shared, collapse = ", "),
    Percentage_Shared_Cond1 = percentage_shared_cond1,
    Percentage_Shared_Cond2 = percentage_shared_cond2
  )
  
  return(shared_genes)
}
# Obtenha uma lista de todas as vacinas únicas
unique_cond <- unique(data$condition)  #trocar "study" pelo nome da coluna de interesse

# Inicialize um dataframe para armazenar os resultados
shared_genes_df <- data.frame()

# Calcular a sobreposição e porcentagem para cada par de vacinas
for (i in 1:(length(unique_cond) - 1)) {
  for (j in (i + 1):length(unique_cond)) {
    resultado_temp <- overlap_genes(unique_cond[i], unique_cond[j], data)
    shared_genes_df <- bind_rows(shared_genes_df, resultado_temp)
  }
}

# Função para realizar o teste exato de Fisher
fisher_exact_test <- function(shared, notshared1, notshared2) {
  cont_table <- matrix(c(shared, notshared1, notshared2, 0), nrow = 2)
  results_fisher <- fisher.test(cont_table)
  return(results_fisher$p.value)
}

# Aplicar a função a cada linha da tabela de resultados
shared_genes_df$pvalue <- mapply(fisher_exact_test, 
                                    shared_genes_df$Shared, 
                                    shared_genes_df$NotShared_cond1, 
                                    shared_genes_df$NotShared_cond2)


# Calcular qvalue (BH) e -log(FDR)
shared_genes_df <- shared_genes_df %>%
  mutate(qvalue = qvalue(pvalue)$qvalues)

# Calcular -log(q)
shared_genes_df$`log_q`= -log(shared_genes_df$qvalue, 10)


#Duplicar e trocar colunas (Vacina 1 e Vacina 2)
shared_genes_df2 = shared_genes_df
shared_genes_df2[, c("Cond1", "Cond2")] <- shared_genes_df2[, c("Cond2", "Cond1")]

#Unir datasets com os dados espelhados
shared_genes_df_mirror = rbind(shared_genes_df, shared_genes_df2)

#Converter NA, NaNe inf em 0
shared_genes_df_mirror[is.na(shared_genes_df_mirror)] <- 0
shared_genes_df_mirror<- replace(shared_genes_df_mirror, shared_genes_df_mirror == "Inf", 0)

#Arredondar porcentagens
shared_genes_df_mirror = shared_genes_df_mirror %>%
  mutate(Percentage_Shared_Cond1 = round(Percentage_Shared_Cond1, 2),
         Percentage_Shared_Cond2 = round(Percentage_Shared_Cond2, 2))

#Salvar resultados
write.csv(shared_genes_df_mirror, file = paste0("shared_", filename, "_fisher.csv")) #Alterar

```


*pHeatmap*

```{r}

#Anotar colunas
# Converter para wide table
matrix_data = shared_genes_df_mirror %>% 
  dcast(Cond1 ~ Cond2, 
        value.var = "Shared", 
        fun.aggregate = mean) %>% 
  column_to_rownames("Cond1") %>%
  mutate_if(is.character, as.numeric) %>% 
  as.matrix() 

# Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex = matrix_data %>% 
  as.data.frame() %>%
  rownames_to_column("condition") %>% 
  select(condition) %>% 
  inner_join(ann_vaccines, by = "condition") %>%
  select(condition, vaccine:regimen, week) %>% 
  distinct() 

# Unir tabela de anotações
ann_vaccines_heatmap = matrix_data %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "process") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange(condition) %>% 
  relocate(dose, .before=vaccine) %>% 
  column_to_rownames(var= "condition")  %>% 
  mutate(dose = ifelse(is.na(dose), "NA", dose))

# Ordenar colunas da matriz para a mesma ordem das linhas da tabela de anotações
matrix_data_ordered = matrix_data %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  merge(ann_vaccines_complex, by.y = "condition", all.x=TRUE) %>% 
  arrange(condition) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  select(!c(vaccine:regimen)) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% #Converter em numéricos
  as.matrix()

matrix_data_ordered_rows = matrix_data_ordered %>% 
  as.data.frame() %>% 
  rownames_to_column("condition") %>% 
  arrange(condition) %>% 
  column_to_rownames(var= "condition") %>% 
  as.matrix()

#Verificar se as linhas sao iguais------
dim(matrix_data_ordered) #Usar se for ordenar por colunas
dim(ann_vaccines_heatmap) #Anotações sobre as colunas
# dim(matrix_data_rows_cols_ordered) #Usar se ordenar por colunas e linhas
# dim(ann_genesets_heatmap) #Anotações sobre as linhas

```


4. Cores.
Defina as cores de cada elemento da tabela das anotações das colunas e/ou linhas.

```{r}
########## Definir Cores dos grupos
#Vaccines
ann_colors = list(
  vaccine = c(
  "BBIBP" = "#76c893",
  "ZF2001" = "#b5e48c",
  "BNT" = "#56cfe1",
  "MO" = "#72ddf7",
  "ChAd" = "#5e60ce",
  "I" = "grey95",
  "H" = "grey95"),
  dose = c(
    "0" = "grey95",
    "1" = "#bbdefb",
    "2" = "#1e88e5",
    "3" = "#0d47a1"),
  day = c(
    "0" = "white",
    "1" = "#e3f2fd",
    "2" = "#ade8f4",
    "3" = "#90e0ef",
    "4" = "#6CD5EA",
    "5" = "#48cae4",
    "6" = "#00b4d8",
    "7" = "#0096c7",
    "10" = "#0087BF",
    "11" = "#0087BF",
    "14" = "#0077b6",
    "26" = "#015BA0",
    "28" = "#023e8a",
    "51" = "#03045e"),
  week = c("1" = "#e3f2fd",
           "2" = "#6CD5EA",
           "3" = "#0087BF",
           "4" = "#015BA0", 
           "7" = "#03045e"),
  type = c('IN'= "#76c893", #1st Gen  vaccines
           'VV' = "#5e60ce", #3rd Gen vaccines
           'RNA' = "#56cfe1",
           'SU' = "#b5e48c",
           'I'= "#f72585",
           "H" = "grey95"), #Infection
  regimen = c('HO' = "grey90",
              'HE' = "#8d99ae",
              "V-I-V" = "#36248f",
              "V-I" = "#D7B0EE",
              "I-V-I" = "#7400b8",
              "I-I" = "#5390d9",
              "I" = "#64dfdf",
              "H" = "grey95"),
  infection = c("0" = "#64dfdf",
                "1" = "#f72585",
                "2" = "#a4161a"),
  variant = c("None" = "grey95",
              "Beta" = "#72efdd",
              "Omicron" = "#5e60ce"),
  severity = c("H" = "grey95",
               "S" = "#0d47a1",
               "MO" = "#1e88e5",
               "MI" = "#bbdefb",
               "NA" = "grey95"),
  sex = c("Male" = "#64dfdf",
          "Female" = "#5e60ce",
          "M/F" = "#D7B0EE"))


#Definir cores do gráfico
# Defina os valores mínimos e máximos para o espectro de cores
min(matrix_data_ordered_rows)
max(matrix_data_ordered_rows)

#Espectro de cores
col_fun <- colorRamp2(c(0, 500), c("white", "#d90429")) #Immune

col_fun <- colorRamp2(c(0, 100, 1000, 2000), c("white", "#f098a7","#ef233c","#d90429")) #Not immune

```


```{r}

######## HEATMAP ANNOTATION
ann_vaccines_heatmap = ann_vaccines_heatmap %>% 
  select(type, vaccine, day)

ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")

row_ha = rowAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors)



mat = matrix_data_ordered_rows
width_image = 40
height_image = 40
width = unit(20, "cm")
height = unit(20, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)
gp = gpar(fontsize = 5)
rect_gp = gpar(col = "gray1", lwd = 0)



fileimageheatmap_grouped= paste0(filename, sep ="_", "Heatmap.png")

png(file = fileimageheatmap_grouped, width= width_image, height= height_image, units = "cm", res=900)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "Shared DEGs", #Legend
                        col = col_fun, #color
                        cell_fun = function(j, i, x, y, width, height, fill) 
                          {grid.text(sprintf("%.f", mat[i, j]), x, y, gp = gp)},
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 90,
                        column_names_gp = column_names_gp,
                        cluster_rows = TRUE,
                        rect_gp = rect_gp,
                        row_names_gp = row_names_gp,
                        # row_split = 2, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        na_col = "black",
                        column_split = NULL, #Split group clusters
                        #column_km = 3,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = width, 
                        height = height
)

draw(heatmap_plot_all, annotation_legend_side = "left")

dev.off()
```

*Chord diagram*

```{r}
#Chord diagram
install.packages("circlize")
library(circlize)

# sharedDEGs_UPDOWN_fisher_pvalue10
# sharedDEGs_UP_fisher_p_10_pvalue10
# sharedDEGs_DOWN_fisher_p_10_pvalue10

#Input
shared_genes_df_mirror_p10 = sharedDEGs_UP_fisher_p_10_pvalue10


#All doses
circos_alldoses = shared_genes_df_mirror_p10 %>% 
  select(Cond1, Cond2, Shared) %>% 
  mutate(Shared = as.numeric(Shared)) %>%  
  merge(ann_vaccines, by.x = "Cond1", by.y="Condition", all.x=T, all.y=F) %>%
  merge(ann_vaccines, by.x = "Cond2", by.y="Condition", all.x=T, all.y=F) %>% 
  select(Cond1, Cond2, Shared,Dose.x, Dose.y) %>% 
  rename(cond1_dose = Dose.x,
         cond2_dose = Dose.y) %>% 
  filter(Shared != 0)

circos_alldoses_mirror = circos_alldoses %>% 
  mutate(Cond1A = Cond2,
         Cond2A = Cond1) %>% 
  select(Cond1A, Cond2A, Shared, cond1_dose, cond2_dose) %>% 
  rename(Cond1 = Cond1A,
         Cond2 = Cond2A) %>% 
  rbind(circos_alldoses)

write.csv(circos_alldoses_mirror, file = "Circos_DOWN_Alldoses_pvalue10_dose1.csv")


#Dose 1
circos_dose1 = shared_genes_df_mirror_p10 %>% 
  select(Cond1, Cond2, Shared) %>% 
  mutate(Shared = as.numeric(Shared)) %>%  
  merge(ann_vaccines, by.x = "Cond1", by.y="Condition", all.x=T, all.y=F) %>%
  merge(ann_vaccines, by.x = "Cond2", by.y="Condition", all.x=T, all.y=F) %>% 
  select(Cond1, Cond2, Shared,Dose.x, Dose.y) %>% 
  rename(cond1_dose = Dose.x,
         cond2_dose = Dose.y) %>% 
  filter(cond1_dose == 1,
         cond2_dose == 1)

circos_dose1_mirror = circos_dose1 %>% 
  mutate(Cond1A = Cond2,
         Cond2A = Cond1) %>% 
  select(Cond1A, Cond2A, Shared, cond1_dose, cond2_dose) %>% 
  rename(Cond1 = Cond1A,
         Cond2 = Cond2A) %>% 
  rbind(circos_dose1)

write.csv(circos_dose1_mirror, file = "Circos_DOWN_pvalue10_dose1.csv")

#Dose 2
circos_dose2 = shared_genes_df_mirror_p10 %>% 
  select(Cond1, Cond2, Shared) %>% 
  mutate(Shared = as.numeric(Shared)) %>%  
  merge(ann_vaccines, by.x = "Cond1", by.y="Condition", all.x=T, all.y=F) %>%
  merge(ann_vaccines, by.x = "Cond2", by.y="Condition", all.x=T, all.y=F) %>% 
  select(Cond1, Cond2, Shared,Dose.x, Dose.y) %>% 
  rename(cond1_dose = Dose.x,
         cond2_dose = Dose.y) %>% 
  filter(cond1_dose == 2,
         cond2_dose == 2)
circos_dose2_mirror = circos_dose2 %>% 
  mutate(Cond1A = Cond2,
         Cond2A = Cond1) %>% 
  select(Cond1A, Cond2A, Shared, cond1_dose, cond2_dose) %>% 
  rename(Cond1 = Cond1A,
         Cond2 = Cond2A) %>% 
  rbind(circos_dose2) 

write.csv(circos_dose2_mirror, file = "Circos_DOWN_pvalue10_dose2.csv")

#Dose 3
circos_dose3 = shared_genes_df_mirror_p10 %>% 
  select(Cond1, Cond2, Shared) %>% 
  mutate(Shared = as.numeric(Shared)) %>%  
  merge(ann_vaccines, by.x = "Cond1", by.y="Condition", all.x=T, all.y=F) %>%
  merge(ann_vaccines, by.x = "Cond2", by.y="Condition", all.x=T, all.y=F) %>% 
  select(Cond1, Cond2, Shared,Dose.x, Dose.y) %>% 
  rename(cond1_dose = Dose.x,
         cond2_dose = Dose.y) %>% 
  filter(cond1_dose == 3,
         cond2_dose == 3)

 circos_dose3_mirror = circos_dose3 %>% 
  mutate(Cond1A = Cond2,
         Cond2A = Cond1) %>% 
  select(Cond1A, Cond2A, Shared, cond1_dose, cond2_dose) %>% 
  rename(Cond1 = Cond1A,
         Cond2 = Cond2A) %>% 
  rbind(circos_dose3) 

write.csv(circos_dose3_mirror, file = "Circos_DOWN_pvalue10_dose3.csv")

```

### Visualização

Dotplot

```{r}
#### Identidade e -log(q)
install.packages("ggdendro")
library(ggdendro)
library(cowplot)
library(ggtree)
library(patchwork) 

#Filtrar valores com -log(p) >= 1
sharedDEGs_UP_DOWN_fisher_filtered <- sharedDEGs_UP_DOWN_fisher %>%
  mutate(`Significance -log(p)` = cut(`-log(p)`, 
                        breaks = c(-Inf, 1, 1.3, 2, Inf), 
                        labels = c("<1", "1-1.3", "1.3-2", ">2"))) %>%
  filter(Percentage_Shared_Cond1 >= 1, Percentage_Shared_Cond2 >= 1, `Significance -log(p)` != "<1")

#Salvar
write.csv(sharedDEGs_UP_DOWN_fisher_filtered, file = "sharedDEGs_UP_DOWN_fisher_filtered.csv")

#Visualizar
plot = ggplot(sharedDEGs_UP_DOWN_fisher, aes(x = Cond1, y = Cond2, color = Percentage_Shared_Cond1, size = Percentage_Shared_Cond1)) + 
  geom_point() +
  scale_size_continuous(range = c(2, 8)) +
  geom_text(aes(label = sprintf("%.f", Percentage_Shared_Cond1)), vjust = 0.5, hjust = 0.5, size = 2, color = "black")  +
  cowplot::theme_cowplot() + 
  theme(axis.line = element_blank(),
        axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1, size = 10),
        axis.text.y = element_text(size = 10)) +
  scale_color_gradient(low = "white", high = "#ef233c", limits = c(0, 60)) +
  ggtitle("%Identity between vaccines, Up and Down-Regulated, pvalue < 0.05" ) +
  ylab('Cond2') +
  theme(axis.ticks = element_blank()) +
  guides(color = FALSE, size = FALSE)

#Salvar
ggsave(plot, file = "sharedDEGs_UP_DOWN_fisher.png")

#Display
plot
```

```{r}

##### Unir countdata de todos os datasets

gse189039_counts_ready_long = gse189039_counts_ready %>% 
  rownames_to_column("genes") %>% 
  pivot_longer(-genes, 
               names_to = "condition", 
               values_to = "counts") %>% 
  mutate(study = "GSE189039")

gse201530_counts_ready_long = gse201530_counts_ready %>% 
    as.data.frame() %>% 
    rownames_to_column("genes") %>% 
  pivot_longer(-genes, 
               names_to = "condition", 
               values_to = "counts") %>% 
    mutate(study = "GSE201530")

gse201533_counts_ready_long = gse201533_counts_ready %>% 
    as.data.frame() %>% 
    rownames_to_column("genes") %>% 
  pivot_longer(-genes, 
               names_to = "condition", 
               values_to = "counts")  %>% 
    mutate(study = "GSE201533")

gse199750_counts_ready_long = gse199750_counts_ready %>% 
  as.data.frame() %>% 
  rownames_to_column("genes") %>% 
  pivot_longer(-genes, 
               names_to = "condition", 
               values_to = "counts") %>% 
    mutate(study = "GSE199750")

gse_counts_all_long = bind_rows(gse189039_counts_ready_long, 
                           gse201530_counts_ready_long,
                           gse201533_counts_ready_long,
                           gse199750_counts_ready_long)

gse_counts_all_matrix = gse_counts_all_long %>% 
  select(-study) %>% 
  pivot_wider(names_from = "condition", values_from = "counts")


write.csv(gse_counts_all_long, file = "gse_counts_all_long.csv")

# Unir metadata 


```


### DEGs quantification analysis

```{r}
degs_all_updown = all_degs_p_05_vac_infected_29_11_23 %>% 
  mutate(direction = case_when(log2fold_change >= 1 ~ "UP",
                               log2fold_change <= -1 ~ "DOWN",
                               TRUE ~ "NEUTRAL"))

#All DEGs, not Immune filtered
plot_degs = degs_all_updown %>% 
  group_by(gse_id, condition, direction) %>% 
  summarise(n_regulated = n()) %>%
  ungroup() %>% 
  inner_join(ann_vaccines, by = "condition")

ggplot_degs = plot_degs %>%
 filter(!(direction %in% "NEUTRAL")) %>%
 ggplot() +
  aes(x = condition, fill = direction, weight = n_regulated) +
  geom_bar(position = "dodge") +
  scale_fill_manual(
    values = c(DOWN = "grey80",
    UP = "#5e60ce")
  ) +
  theme_void() +
  ylab("") +
  theme(legend.position = "right",
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1, size = 10),
        axis.text.y = element_text(size = 20, color = "white", angle = 0, hjust = -1)) +
  geom_text(aes(label = n_regulated, y = n_regulated), 
            position = position_dodge(width = 0.9), 
            vjust = 0.2, 
            hjust = -0.2, 
            size = 3,
            angle = 90) +
  ylim(0,3000)+
  ggtitle("DEGs - Up and down regulated, by vaccine, pvalue <0.05") +
  facet_wrap(~disease_vac, scales = "free_x", nrow = 2)

print(ggplot_degs)


ggsave(ggplot_degs, file = "DEGs_up_down_barplot_pvalue05.png",  width = 10, height = 8) 


```


DEGs immune, cell cycle, and other
```{r}
immunego = ImmuneGO_Annotated_GO_8_2_24 %>% 
  mutate(genes = str_split(genes, ",")) %>%
  unnest(genes)

immunego_2 = ImmuneGO_QuickGO_Unreviewed_19_2_24 %>% 
  select(genes = symbol) %>% 
  distinct()

cellcycleGO = QuickGO_CellCycle %>% 
  clean_names() %>% 
  rename(genes = symbol)


all_degs_p_05_vac_infected_29_11_23 = all_degs_p_05_vac_infected_29_11_23 %>% 
  select(-"...1")

#### Not immune

all_degs_infected_notimmune = all_degs_p_05_vac_infected_29_11_23 %>%
  anti_join(immunego_2 %>%
              select(genes) %>%
              distinct(),
            by = "genes") %>%
  distinct() %>% 
  mutate(bioprocess = "Other")

all_degs_infected_notimmune %>% 
  tabyl(condition, bioprocess)

write.csv(all_degs_infected_notimmune, file = "all_degs_p_05_vac_infected_notimmune.csv", row.names = F)
```


```{r}
#### Immune
all_degs_p_05_vac_infected_immune = all_degs_p_05_vac_infected_29_11_23 %>% 
  inner_join(immunego_2 %>% 
              select(genes) %>% 
              distinct(),
            by = "genes") %>% 
  distinct() %>% 
  mutate(bioprocess = "Immune")

all_degs_p_05_vac_infected_immune %>% 
  tabyl(condition, bioprocess) 
  

write.csv(all_degs_p_05_vac_infected_immune, file = "all_degs_p_05_vac_infected_immune.csv", row.names = F)
```


```{r}
##### Immune and other
all_degs_processes_immune_nonimmune = bind_rows(all_degs_p_05_vac_infected_immune, 
                               all_degs_p_05_vac_infected_notimmune)

all_degs_processes_immune_nonimmune %>%
  tabyl(condition, bioprocess) %>% 
  adorn_totals("col") %>% 
  adorn_percentages() %>% 
  adorn_pct_formatting() 

write.csv(all_degs_processes_immune_nonimmune, file = "all_degs_processes_immune_nonimmune.csv", row.names = F)
```




```{r}
organism = "org.Hs.eg.db"
install.packages("GO.db")
BiocManager::install(organism, character.only = TRUE)
library(organism, character.only = TRUE)
library(clusterProfiler)
library(GO.db)

#HGNC symbol to Entrez
x <- org.Hs.egSYMBOL
mapped_genes <- mappedkeys(x)
symbols_entrez <- as.data.frame(x[mapped_genes])

## Entrez to GO
x <- org.Hs.egGO
mapped_genes <- mappedkeys(x)
entrez_geneontology <- as.data.frame(x[mapped_genes])

## GO BP ancestors
go_bp_ancestor = GOBPANCESTOR %>% 
  as.data.frame()
colnames(go_bp_ancestor)[1] <- "go_id"
colnames(go_bp_ancestor)[2] <- "go_ancestor"

## GO ID to terms
go_bp = AnnotationDbi::select(GO.db, columns=c("GOID","TERM"), keys= "BP", keytype= "ONTOLOGY") %>% 
  clean_names() %>% 
  select(go_id = goid, term)

## GO child and ancestor
go_bp_ancestor = go_bp_ancestor %>% 
  inner_join(go_bp %>% rename(go_ancestor = go_id), by = "go_ancestor") %>% 
  rename(term_ancestor = term)

#GO ids to terms
go_bp = AnnotationDbi::select(GO.db, columns=c("GOID","TERM"), keys="BP", keytype="ONTOLOGY") %>% 
  clean_names() %>% 
  rename(go_id = goid) %>% 
  select(-ontology) %>% 
  right_join(entrez_geneontology, by = "go_id") %>% 
  right_join(symbols_entrez, by = "gene_id") %>% 
  clean_names() %>% 
  left_join(go_bp_ancestor, by = "go_id") %>% 
  distinct()

#Filter main GO BP terms
go_bp_main = go_bp %>%
  filter(go_ancestor %in% c(
      "GO:0040011",
      "GO:0043473",
      "GO:0044419",
      "GO:0009987",
      "GO:0048518",
      "GO:0050789",
      "GO:0065007",
      "GO:0048511",
      "GO:0022414",
      "GO:0002376",
      "GO:0008152")) %>% 
  filter(str_detect(term_ancestor, "cellular process|biological process involved in interspecies interaction between organisms|regulation of biological process|immune system process|No annotation|Not mapped|metabolic process"),
         !term_ancestor == "positive regulation of biological process") %>% 
  mutate(term_ancestor = case_when(
           term_ancestor == "biological process involved in interspecies interaction between organisms" ~ "Interspecies interaction",
           term_ancestor == "biological regulation" ~ "Biological reg.",
           term_ancestor == "cellular process" ~ "Cellular",
           term_ancestor == "immune system process" ~ "Immune system",
           term_ancestor == "regulation of biological process" ~ "Reg. of BP",
           term_ancestor == "locomotion" ~ "Locomotion",
           term_ancestor == "metabolic process" ~ "Metabolism",
           TRUE ~ term_ancestor))

#Add set size column
go_bp_main_2 = go_bp_main %>% 
  group_by(term_ancestor) %>% 
  summarise(set_size = n()) %>% 
  inner_join(go_bp_main) %>% 
  select(term = term_ancestor, set_size, go_id = go_ancestor, symbol) %>% 
  distinct() %>% 
  mutate(annotation = "Major process")

go_bp_child = go_bp_main %>% 
  select(term, go_id, symbol) %>% 
  distinct() %>% 
  group_by(term) %>% 
  summarise(set_size = n()) %>% 
  inner_join(go_bp_main, by = "term") %>% 
  select(term, set_size, go_id, symbol) %>% 
  distinct() %>% 
  mutate(annotation = "Minor process")

go_bp_all = bind_rows(go_bp_main_2, go_bp_child)

write.csv(go_bp_all, file = "OtherGOs_Annotated_Genes.csv", row.names = F)

# Collapse rows 
go_bp_all_comma = go_bp_all %>% 
  group_by(term) %>% 
  summarize(genes = paste0(genes, collapse = ",")) %>% 
  inner_join(go_bp_all, by = "term") %>% 
  select(-symbol) %>% 
  distinct() %>% 
  arrange(desc(set_size))

write.csv(go_bp_all_comma, file = "OtherGOs_Annotated_GOs.csv", row.names = F)
```


```{r}
# DEGs to Entrez and GO ids --------
degs_GO = all_degs_p_05_vac_infected_29_11_23 %>% 
  filter(direction != "NEUTRAL") %>% 
  select(condition, genes, direction) %>% 
  left_join(go_bp %>% rename(genes = symbol, entrez = gene_id), by = "genes") %>% 
  mutate(go_id = if_else(is.na(go_id), "Not mapped", go_id),
         term = if_else(is.na(term), "Not mapped", term),
         evidence = if_else(is.na(evidence), "Not mapped", evidence),
         ontology = if_else(is.na(ontology), "Not mapped", ontology),
         mapped = if_else(ontology == "Not mapped", "Not mapped", "Mapped"),
         go_ancestor = if_else(is.na(go_ancestor), "Not mapped", go_ancestor),
         term_ancestor = if_else(is.na(term_ancestor), "Not mapped", term_ancestor)) %>%
  filter(ontology %in% c("BP", "Not mapped")) %>% 
  select(-evidence) 

# Collapse gene column --------
degs_GO_comma = degs_GO %>% 
  select(-term_ancestor, -go_ancestor) %>% 
  distinct() %>% 
  group_by(condition, term) %>% 
  summarize(genes = paste0(genes, collapse = ",")) %>% 
  inner_join(ann_vaccines, by = "condition")

write.csv(degs_GO_comma, file = "degs_annotated_allGOs.csv", row.names = F) #Save


#DEGs annotated by Main GO BPs ------

main_go = degs_GO %>%
  filter(go_ancestor %in% c(
      "Not mapped",
      "GO:0040011",
      "GO:0043473",
      "GO:0044419",
      "GO:0009987",
      "GO:0048518",
      "GO:0050789",
      "GO:0065007",
      "GO:0048511",
      "GO:0022414",
      "GO:0002376",
      "GO:0008152")) %>% 
  select(condition, genes, direction, term_ancestor) %>% 
  distinct() %>% 
  group_by(condition, direction, term_ancestor) %>% 
  summarise(n = n()) %>% 
  mutate(term_ancestor = if_else(is.na(term_ancestor), "No annotation", term_ancestor))

main_go_2 = degs_GO %>% 
  select(condition, genes, mapped, direction) %>% 
  distinct() %>% 
  group_by(condition, direction) %>% 
  summarise(n_genes_total = n()) %>% 
  inner_join(main_go, by = join_by("condition", direction)) %>% 
  ungroup() %>% 
  select(condition, direction, term_ancestor, n_genes = n, n_genes_total) %>%
  filter(str_detect(term_ancestor, "cellular process|biological process involved in interspecies interaction between organisms|regulation of biological process|immune system process|No annotation|Not mapped|metabolic process"),
         !term_ancestor == "positive regulation of biological process") %>% 
  mutate(n_genes_perc = round(n_genes/n_genes_total * 100, 1),
         term_ancestor = case_when(
           term_ancestor == "biological process involved in interspecies interaction between organisms" ~ "Interspecies interaction",
           term_ancestor == "biological regulation" ~ "Biological reg.",
           term_ancestor == "cellular process" ~ "Cellular",
           term_ancestor == "immune system process" ~ "Immune system",
           term_ancestor == "regulation of biological process" ~ "Reg. of BP",
           term_ancestor == "locomotion" ~ "Locomotion",
           term_ancestor == "metabolic process" ~ "Metabolism",
           TRUE ~ term_ancestor),
         term_ancestor = fct_relevel(term_ancestor, "Not mapped", after = Inf),
         direction = fct_relevel(direction, "UP", "DOWN")) %>% 
  inner_join(ann_vaccines_19_1_24 %>% select(condition, type, disease_vac), by = "condition")
  
write.csv(main_go_2, file = "degs_annotated_mainGOs_stats.csv", row.names = F)





```


```{r}
### Bar plot by term -----

main_go_2_plot = main_go_2 %>% 
  mutate(direction = if_else(direction == "UP", "↑", "↓")) %>% 
  ggplot() +
 aes(x = forcats::fct_rev(factor(condition)), y = n_genes_perc, fill = type) +
 geom_col() +
 coord_flip() +
 theme_minimal() +
 theme(legend.position = "top") +
 facet_grid(vars(direction),  vars(term_ancestor)) +
   scale_fill_manual(name = "Type",
    values = c('IN'= "#76c893", #1st Gen  vaccines
           'VV' = "#5e60ce", #3rd Gen vaccines
           'RNA' = "#56cfe1",
           'SU' = "#b5e48c",
           'I'= "#f72585",
           "H" = "grey95")) +
  theme(legend.position = "right",
          legend.title = element_text(size = 15),
          legend.text =element_text(size=15),
        axis.text.x = element_text(size = 12, color = "black", angle = 0, hjust = 1),
        axis.text.y = element_text(size = 12, color = "black", angle = 0, hjust = 1),
        strip.text.x = element_text(size = 12, color = "white", hjust = 0.5),
        strip.text.y = element_text(size = 15, angle = 0),
        strip.background.x = element_rect(
          fill = "black", 
          color = "black", 
          linewidth = 1 ),
        strip.background.y = element_rect(
          fill = "gray70", 
          color = "gray70", 
          linewidth = 1 )
        ) +
  geom_text(aes(label = paste0(n_genes_perc, "% (", n_genes, ")"), 
                y = n_genes_perc), 
            hjust = 0, 
            size = 3,
            angle = 0) +
  labs(title = "DEGs by Biological Process",
       x = "",
       y = "Number of genes (%)") +
  ylim(0,120)


ggsave(main_go_2_plot, file = "main_go_2_plot_values.png", width = 20, height = 15)


### Totals ------

main_go_2_plot_totals_values = main_go_2 %>%
  select(condition, n_genes_total, type, direction) %>% 
  distinct() %>% 
  mutate(stat = "Total genes",
         direction = if_else(direction == "UP", "↑", "↓")) %>% 
  ggplot() +
 aes(x = forcats::fct_rev(factor(condition)), y = n_genes_total, fill = type) +
 geom_col() +
 coord_flip() +
 theme_minimal() +
 theme(legend.position = "top") +
 facet_grid(direction~ stat, scales = "free") +
   scale_fill_manual(
    values = c('IN'= "#76c893", #1st Gen  vaccines
           'VV' = "#5e60ce", #3rd Gen vaccines
           'RNA' = "#56cfe1",
           'SU' = "#b5e48c",
           'I'= "#f72585",
           "H" = "grey95")) +
    theme(legend.position = "right",
          legend.title = element_text(size = 15),
          legend.text =element_text(size=15),
        axis.text.x = element_text(size = 12, color = "black", angle = 0, hjust = 1),
        axis.text.y = element_text(size = 12, color = "black", angle = 0, hjust = 1),
        strip.text.x = element_text(size = 12, color = "white", hjust = 0.5),
        strip.text.y = element_text(size = 15, angle = 0),
        strip.background.x = element_rect(
          fill = "black", 
          color = "black", 
          linewidth = 1 ),
        strip.background.y = element_rect(
          fill = "gray70", 
          color = "gray70", 
          linewidth = 1 )) +
  geom_text(aes(label = n_genes_total, y = n_genes_total),
            hjust = 0, 
            size = 3,
            angle = 0) +
  ylim(0, 2600) +
  labs(x = "", 
       y = "Total genes",
       fill = "Type")


ggsave(main_go_2_plot_totals_values, file = "main_go_2_plot_totals_values.png", width = 8, height = 15)

```


```{r}
###### Immune, Cell cycle, Other

plot_degs <- all_degs_processes %>%
  mutate(bioprocess = fct_relevel(bioprocess, "Other", "Cell cycle", "Immune")) %>% 
  group_by(condition, bioprocess) %>% 
  summarise(n = n()) %>%   
  ungroup() %>% 
  inner_join(ann_vaccines, by = "condition") 

plot_degs_2 = plot_degs %>% 
  group_by(condition) %>% 
  summarise(total = sum(n)) %>% 
  ungroup() %>% 
    inner_join(plot_degs, by = "condition")

ggplot_degs = plot_degs_2 %>%
 ggplot() +
  aes(x = condition, fill = bioprocess, weight = n) +
  geom_bar(position = "dodge", color = "black") +
  scale_fill_manual(
    values = c("Other" = "grey90",
               "Cell cycle" = "#D7B0EE",
               "Immune" = "#5e60ce")) +
  theme_void() +
  ylab("") +
  coord_flip() +
  theme(legend.position = "right",
        axis.text.x = element_blank(),
        axis.text.y = element_text(size = 12, color = "black", angle = 0, hjust = 1)) +
  geom_text(aes(label = paste0(n, " (", round((n/total)*100, 1), "%, ", total, ")"), y = n), 
            position = position_dodge(width = 1), 
            vjust = 0.5, 
            hjust = -0.05, 
            size = 3,
            angle = 0) +
  ggtitle("DEGs - Up and down regulated, by vaccine, pvalue <0.05") +
  facet_wrap(vars(disease_vac), scales = "free")

ggsave(ggplot_degs, file = "degs_immune_other_3.png",  width = 15, height = 8)
```


```{r}
##### Immune, Other

plot_degs <- all_degs_processes_immune_nonimmune %>%
  mutate(bioprocess = fct_relevel(bioprocess, "Immune", "Other")) %>% 
  group_by(condition, bioprocess) %>% 
  summarise(n = n()) %>%   
  ungroup() %>% 
  inner_join(ann_vaccines, by = "condition") 

plot_degs_2 = plot_degs %>% 
  group_by(condition) %>% 
  summarise(total = sum(n)) %>% 
  ungroup() %>% 
    inner_join(plot_degs, by = "condition")


# Bars

ggplot_degs = plot_degs_2 %>%
  filter(bioprocess == "Immune") %>% 
 ggplot() +
  aes(x = condition, fill = bioprocess, weight = n) +
  geom_bar(position = "stack", color = "black") +
  scale_fill_manual(
    values = c("Other" = "grey90",
               "Cell cycle" = "#D7B0EE",
               "Immune" = "#5e60ce")) +
  theme_void() +
  ylab("") +
  coord_flip() +
  theme(legend.position = "right",
        axis.text.x = element_blank(),
        axis.text.y = element_text(size = 10, color = "black", angle = 0, hjust = 1)) +
  geom_text(aes(label = if_else(bioprocess == "Immune", paste0(n, " (", round((n/total)*100, 1), "%, ", total, ")"), 
                                paste0(n, " (Other)")), y = n), 
            position = position_dodge(width = 0.9), 
            vjust = 0.5, 
            hjust = -0.05, 
            size = 3,
            angle = 0) +
  ggtitle("DEGs - Up and down regulated, by vaccine, pvalue <0.05") +
  facet_wrap(vars(disease_vac), scales = "free")

ggsave(ggplot_degs, file = "degs_immune_percent_bars_2.png",  width = 15, height = 8) 


# Columns

ggplot_degs = plot_degs_2 %>%
  filter(bioprocess == "Immune") %>% 
 ggplot() +
  aes(x = condition, fill = bioprocess, weight = n) +
  geom_bar(position = "stack") +
  scale_fill_manual(
    values = c("Other" = "grey90",
               "Cell cycle" = "#D7B0EE",
               "Immune" = "#5e60ce")) +
  theme_void() +
  ylab("") +
  theme(legend.position = "right",
        axis.text.x = element_text(size = 10, color = "black", angle = 45, hjust = 1, vjust = 1),
        axis.text.y = element_text(size = 20, color = "white", angle = 0, hjust = -2, vjust = 1)) +
  geom_text(aes(label = if_else(bioprocess == "Immune", paste0(n, " (", round((n/total)*100, 1), "%)"), 
                                paste0(n, " (Other)")), y = n), 
            position = position_dodge(width = 0.9), 
            vjust = 0.5, 
            hjust = -0.05, 
            size = 3,
            angle = 90) +
  ggtitle("DEGs - Up and down regulated, by vaccine, pvalue <0.05") +
  ylim(0, 700) +
  facet_wrap(vars(disease_vac), scales = "free", nrow = 2)

ggsave(ggplot_degs, file = "degs_immune_percent_columns_2.png",  width = 10, height = 8) 



```



### Immune related only

```{r}

degs_all_updown = all_degs_p_05_vac_infected_19_12_23 %>% 
  filter(direction != "NEUTRAL") %>% 
  distinct()

ImmuneGO_Annotated_Genes = ImmuneGO_Annotated_Genes %>% 
  filter(!(process %in% c("TCR REPERTOIRE", "BCR REPERTOIRE")))



shared_genes_infection = degs_all_updown %>%
  filter(disease_vac == "I") %>% 
  select(condition, genes) %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  select(condition, genes, vaccine) %>% 
  inner_join(ImmuneGO_Annotated_Genes %>% select(genes) %>% distinct(), by = "genes") %>% 
  group_by(genes) %>% 
  summarise("Infection" = n()) %>% 
  arrange(desc("Infection"), genes) %>% 
  ungroup() %>% 
  pivot_longer(-genes, names_to = "infection_vaccination", values_to = "n_conditions") %>% 
  inner_join(ImmuneGO_Annotated_Genes , by = "genes") %>% 
  select(genes, infection_vaccination, n_conditions) %>% 
  mutate(comparison = "Infection only") %>% 
  arrange(desc(n_conditions)) %>% 
  distinct()

shared_genes_vaccines = degs_all_updown %>%
  filter(disease_vac == "V") %>% 
  select(condition, genes) %>% 
  distinct() %>% 
  inner_join(ImmuneGO_Annotated_Genes %>% select(genes) %>% distinct(), by = "genes") %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  select(condition, genes, vaccine) %>% 
  group_by(genes) %>% 
  summarise("Vaccination" = n()) %>% 
  arrange(desc("Vaccination"), genes) %>% 
  ungroup() %>% 
  pivot_longer(-genes, names_to = "infection_vaccination", values_to = "n_conditions") %>% 
  inner_join(ImmuneGO_Annotated_Genes , by = "genes") %>% 
  select(genes, infection_vaccination,n_conditions) %>% 
  mutate(comparison = "Vaccination only") %>% 
  arrange(desc(n_conditions)) %>% 
  distinct()


# Infection and vaccination
shared_genes_vaccines_infection = shared_genes_vaccines %>% 
  select(genes, n_conditions) %>% 
  inner_join(shared_genes_infection %>% select(genes, n_conditions), by = "genes") %>% 
  rename(Vaccination = n_conditions.x, Infection = n_conditions.y) %>% 
  distinct() %>% 
  pivot_longer(-genes, names_to = "infection_vaccination", values_to = "n_conditions") %>% 
  inner_join(ImmuneGO_Annotated_Genes , by = "genes") %>% 
  select(genes, infection_vaccination,n_conditions) %>% 
  mutate(comparison = "Infection and vaccination") %>% 
  distinct()
  

shared_genes_comparisons = bind_rows(shared_genes_infection, 
                                     shared_genes_vaccines, 
                                     shared_genes_vaccines_infection)

write.csv(shared_genes_comparisons, file = "shared_genes_infection_vaccination_comparisons.csv")

```


```{r}
# Vaccine and infection ----
shared_genes_vaccines_infection_plot = shared_genes_vaccines_infection %>% 
  select(genes, infection_vaccination, n_conditions, comparison) %>% 
  distinct() %>%
  slice_head(n=40) %>%  
  ggplot() +
  aes(
    x = reorder(genes, -n_conditions),
    y = n_conditions,
    fill = infection_vaccination
  ) +
  geom_col() +
  scale_fill_manual(
    values = c("Vaccination" = "#5e60ce",
              "Infection" = "#f72585")
    )+
  theme_minimal() +
  labs(fill = "Infection and vaccination") +
  xlab("Genes") + 
  ylab("Number of shared conditions") +
  theme(legend.position = "right",
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1, size = 12),
        axis.text.y = element_text(size = 7, angle = 90),
        panel.grid.minor.y = element_blank(),
        legend.text = element_text(size=12),
        legend.title = element_text(size = 12, face = "bold")) +
  ggtitle("Infection and vaccination") 

ggsave(shared_genes_vaccines_infection_plot, file = "Genes_shared_Vaccination_Infection.png", width = 10, height = 3)

#Vaccination only -----
shared_genes_vaccines_only = shared_genes_vaccines %>% 
  anti_join(shared_genes_infection, by = "genes") %>% 
  inner_join(ImmuneGO_Annotated_Genes %>% select(genes), by = "genes") %>% 
  select(genes, n_conditions) %>% 
  mutate(comparison = "Vaccination only") %>% 
  arrange(desc(n_conditions)) %>% 
  distinct()

shared_genes_vaccines_only_plot = shared_genes_vaccines_only %>% 
  select(genes, n_conditions, comparison) %>% 
  distinct() %>%
  slice_head(n=20) %>% 
  ggplot() +
  aes(x = reorder(genes, -n_conditions), 
      fill = comparison, 
      weight = n_conditions) +
  geom_bar() +
  scale_fill_manual(
    values = c("Vaccination only" = "#5e60ce",
              "Infection only" = "#f72585")) +
  theme_minimal() +
  labs(fill = "Vaccination only") +
  xlab("Genes") + 
  ylab("Number of shared conditions") +
  theme(legend.position = "right",
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1, size = 12),
        axis.text.y = element_text(size = 7, angle = 0),
        panel.grid.minor.y = element_blank(),
        legend.text = element_text(size=12),
        legend.title = element_text(size = 12, face = "bold")) +
  ggtitle("Vaccination only") 

ggsave(shared_genes_vaccines_only_plot, file = "Genes_shared_Vaccination_only.png", width = 10, height = 3)


# Infection only
shared_genes_infection_only = shared_genes_infection %>% 
  anti_join(shared_genes_vaccines_only, by = "genes") %>%
  inner_join(ImmuneGO_Annotated_Genes, by = "genes") %>% 
  select(genes, n_conditions) %>% 
  mutate(comparison = "Infection only") %>% 
  arrange(desc(n_conditions))

shared_genes_infection_only_plot = shared_genes_infection_only %>% 
  select(genes, n_conditions, comparison) %>% 
  arrange(desc(n_conditions)) %>% 
  distinct() %>%
  slice_head(n=20) %>% 
  ggplot() +
  aes(x = reorder(genes, -n_conditions), 
      fill = comparison, 
      weight = n_conditions) +
  geom_bar() +
  scale_fill_manual(
    values = c("Vaccination only" = "#56cfe1",
              "Infection only" = "#f72585")) +
  theme_minimal() +
  labs(fill = "Infection only") +
  xlab("Genes") + 
  ylab("Number of shared conditions") +
  theme(legend.position = "right",
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1, size = 12),
        axis.text.y = element_text(size = 7, angle = 0),
        panel.grid.minor.y = element_blank(),
        legend.text = element_text(size=12),
        legend.title = element_text(size = 12, face = "bold")) +
  ggtitle("Infection only") 

ggsave(shared_genes_infection_only_plot, file = "Genes_shared_Infection_only.png", width = 10, height = 3)


```


### not-immune related only

```{r}
degs_all_updown = all_degs_p_05_vac_infected_19_12_23 %>% 
  filter(direction != "NEUTRAL") %>% 
  distinct()

ImmuneGO_Annotated_Genes = ImmuneGO_Annotated_Genes_26_2_24 

## Processing -----

#Infection -----
shared_genes_infection = degs_all_updown %>%
  filter(disease_vac == "I") %>% 
  select(condition, genes) %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  select(condition, genes, vaccine) %>% 
  anti_join(ImmuneGO_Annotated_Genes %>% select(genes) %>% distinct(), by = "genes") %>% 
  group_by(genes) %>% 
  summarise("Infection" = n()) %>% 
  arrange(desc("Infection"), genes) %>% 
  ungroup() %>% 
  pivot_longer(-genes, names_to = "infection_vaccination", values_to = "n_conditions") %>% 
  anti_join(ImmuneGO_Annotated_Genes , by = "genes") %>% 
  select(genes, infection_vaccination,n_conditions) %>% 
  mutate(comparison = "Infection only") %>% 
  arrange(desc(n_conditions))

#Vaccine -----
shared_genes_vaccines = degs_all_updown %>%
  filter(disease_vac == "V") %>% 
  select(condition, genes) %>% 
  anti_join(ImmuneGO_Annotated_Genes %>% select(genes) %>% distinct(), by = "genes") %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  select(condition, genes, vaccine) %>% 
  group_by(genes) %>% 
  summarise("Vaccination" = n()) %>% 
  arrange(desc("Vaccination"), genes) %>% 
  ungroup() %>% 
  pivot_longer(-genes, names_to = "infection_vaccination", values_to = "n_conditions") %>% 
  anti_join(ImmuneGO_Annotated_Genes %>% select(genes) %>% distinct(), by = "genes") %>% 
  select(genes, infection_vaccination,n_conditions) %>% 
  mutate(comparison = "Vaccination only") %>% 
  arrange(desc(n_conditions))


# Infection and vaccination -----
shared_genes_vaccines_infection = shared_genes_vaccines %>% 
  inner_join(shared_genes_infection, by = "genes") %>% 
  select(genes, Vaccination = n_conditions.x, Infection = n_conditions.y) %>% 
  pivot_longer(-genes, names_to = "infection_vaccination", values_to = "n_conditions") %>% 
  anti_join(ImmuneGO_Annotated_Genes , by = "genes") %>% 
  select(genes, infection_vaccination,n_conditions) %>% 
  mutate(comparison = "Infection and vaccination") 

shared_genes_comparisons = bind_rows(shared_genes_infection, 
                                     shared_genes_vaccines, 
                                     shared_genes_vaccines_infection) %>% 
  arrange(desc(n_conditions))

write.csv(shared_genes_comparisons, file = "non_Immune_shared_genes_infection_vaccination_comparisons.csv")


# Plotting ------

#Infection -----
shared_genes_vaccines_infection_plot = shared_genes_vaccines_infection %>% 
  select(genes, infection_vaccination, n_conditions, comparison) %>% 
  distinct() %>%
  slice_head(n=40) %>%  
  ggplot() +
  aes(
    x = reorder(genes, -n_conditions),
    y = n_conditions,
    fill = infection_vaccination
  ) +
  geom_col() +
  scale_fill_manual(
    values = c("Vaccination" = "#5e60ce",
              "Infection" = "#f72585")
    )+
  theme_minimal() +
  labs(fill = "Infection and vaccination") +
  xlab("Genes") + 
  ylab("Number of shared conditions") +
  theme(legend.position = "right",
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1, size = 12),
        axis.text.y = element_text(size = 7, angle = 90),
        panel.grid.minor.y = element_blank(),
        legend.text = element_text(size=12),
        legend.title = element_text(size = 12, face = "bold")) +
  ggtitle("Infection and vaccination") 

ggsave(shared_genes_vaccines_infection_plot, file = "Non_immune_Genes_shared_Vaccination_Infection.png", width = 10, height = 3)

#Vaccination only -----

shared_genes_vaccines_only_plot = shared_genes_vaccines %>% 
  select(genes, infection_vaccination, n_conditions, comparison) %>% 
  arrange(desc(n_conditions)) %>% 
  distinct() %>%
  slice_head(n=20) %>% 
  ggplot() +
  aes(x = reorder(genes, -n_conditions), 
      fill = comparison, 
      weight = n_conditions) +
  geom_bar() +
  scale_fill_manual(
    values = c("Vaccination only" = "#5e60ce",
              "Infection only" = "#f72585")) +
  theme_minimal() +
  labs(fill = "Vaccination only") +
  xlab("Genes") + 
  ylab("Number of shared conditions") +
  scale_y_continuous(breaks= pretty_breaks())+
  theme(legend.position = "right",
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1, size = 12),
        axis.text.y = element_text(size = 7, angle = 0),
        panel.grid.minor.y = element_blank(),
        legend.text = element_text(size=12),
        legend.title = element_text(size = 12, face = "bold")) +
  ggtitle("Vaccination only") 

ggsave(shared_genes_vaccines_only_plot, file = "Non_immune_shared_Vaccination_only.png", width = 10, height = 3)


# Infection only -------

shared_genes_infection_only_plot = shared_genes_infection %>% 
  select(genes, infection_vaccination, n_conditions, comparison) %>% 
  arrange(desc(n_conditions)) %>% 
  distinct() %>%
  slice_head(n=20) %>% 
  ggplot() +
  aes(x = reorder(genes, -n_conditions), 
      fill = comparison, 
      weight = n_conditions) +
  geom_bar() +
  scale_fill_manual(
    values = c("Vaccination only" = "#56cfe1",
              "Infection only" = "#f72585")) +
  theme_minimal() +
  labs(fill = "Infection only") +
  xlab("Genes") + 
  ylab("Number of shared conditions") +
  theme(legend.position = "right",
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1, size = 12),
        axis.text.y = element_text(size = 7, angle = 0),
        panel.grid.minor.y = element_blank(),
        legend.text = element_text(size=12),
        legend.title = element_text(size = 12, face = "bold")) +
  ggtitle("Infection only") 

ggsave(shared_genes_infection_only_plot, file = "Non_immune_shared_Infection_only.png", width = 10, height = 3)

```


