

---
title: "RNAseq analyses - Gene set enrichment"
author: "Wasim Syed"
date: "2023-09-06"
output:
  html_document: default
  pdf_document: default
  toc: true
    toc_depth: 3  
    number_sections: true 
    theme: united
  
---

# 1. Initializing

## 1.1. Baixar pacotes

```{r eval=FALSE, include=FALSE}
install.packages("forcats") # útil para agrupar e manipular dados categóricos em análises de RNA-seq.
install.packages("stringr") # útil para trabalhar com strings no R, o que pode ser útil para manipular e filtrar nomes de genes, por exemplo.
install.packages("ggplot2") # criação de visualizações de dados
install.packages("ggbeeswarm")
install.packages("ggrepel") #é um pacote que fornece uma alternativa ao pacote ggplot2 para ajudar a evitar sobreposição de textos em gráficos.
install.packages("readr") #importação de arquivos de dados no R, como arquivos CSV, TSV e outros formatos
install.packages("tidyverse") #limpar e organizar dados no R
install.packages("dplyr") # manipulação de dados em R, útil para filtrar, selecionar, ordenar, agrupar e resumir dados
install.packages("org.Hs.eg.db")
install.packages("RColorBrewer") #paletas
install.packages("plotly") #Gráficos interativos
install.packages("msigdbr")
install.packages("ape")
install.packages("sva")
install.packages("gplots")
install.packages("tibble")
install.packages("metaRNASeq")
install.packages("editData")
install.packages("tidyverse")
install.packages("corrr")
install.packages("ggcorrplot")
install.packages("FactoMineR")
install.packages("factoextra")
install.packages("esquisse")
install.packages("corto")
install.packages("reshape2")

#Instale os pacotes do BioConductor.

if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager") #

BiocManager::install("GEOquery") # baixar e processar dados do Gene Expression Omnibus (GEO)
BiocManager::install("pheatmap") # criação de mapas de calor no R
BiocManager::install("org.Hs.eg.db") #Anotação de genes.
BiocManager::install("deseq2")
BiocManager::install("biomaRt") #Anotações de genes
BiocManager::install("celldex") # anotações de genes por tipo celular
BiocManager::install("NMF") #análise de fatorização de matrizes não negativas (NMF) em dados de expressão genética.
BiocManager::install("BiasedUrn") #Algoritmos para inferência de enriquecimento de conjuntos de genes utilizando métodos de urna viciada.
BiocManager::install("GSVA") # Análise de enriquecimento de vias genéticas usando a variação da atividade de genes ao longo de um conjunto de amostras.
BiocManager::install("sva") #Métodos para ajustar fatores de confusão em análises de expressão genética
BiocManager::install("clusterProfiler") #Fornece ferramentas para análise de enriquecimento funcional e perfilagem de conjuntos de genes
BiocManager::install("ggupset", update = FALSE)
BiocManager::install("pheatmap") #Criação de heatmaps personalizados para visualização de dados de expressão genética.
BiocManager::install("org.Hs.eg.db") #Anotações de genes
BiocManager::install("msigdbr", update = FALSE)
BiocManager::install("EnhancedVolcano")
BiocManager::install("AnnotationDbi")
install.packages("explore")

```

## 1.2. Bibliotecas

```{r eval=FALSE, include=FALSE}
library(readr)
library(explore)
library(maditr)
library(GEOquery)
library(Matrix)
library(circlize)
library(RColorBrewer)
library(celldex)
library(biomaRt)
library(org.Hs.eg.db)
library(plotly)
library(DESeq2)
library(msigdbr)
library(ape)
library(GSVA)
library(sva)
library(clusterProfiler)
library(pheatmap)
library(EnhancedVolcano)
library(metaRNASeq)
library(ggbeeswarm)
library(tidyverse)
library(corrr)
library(ggcorrplot)
library(FactoMineR)
library(factoextra)
library(esquisse)
library(corto)
library(factoextra)
library(reshape2)
library(ComplexHeatmap)
library(janitor)
```

# 2. Prepare data

```{r}
#All DEGs, not Immune filtered
plot_degs = degs_all_updown %>% 
  group_by(gse_id, condition, direction) %>% 
  summarise(n_regulated = n()) %>%
  ungroup() %>% 
  inner_join(ann_vaccines, by = "condition")

ggplot_degs = plot_degs %>%
 filter(!(direction %in% "NEUTRAL")) %>%
 ggplot() +
  aes(x = condition, fill = direction, weight = n_regulated) +
  geom_bar(position = "dodge") +
  scale_fill_manual(
    values = c(DOWN = "grey80",
    UP = "#5e60ce")
  ) +
  theme_void() +
  ylab("") +
  theme(legend.position = "right",
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1, size = 7),
        axis.text.y=element_text(size = 20, color = "white", angle = 0)) +
  geom_text(aes(label = n_regulated, y = n_regulated), 
            position = position_dodge(width = 0.9), 
            vjust = 0.2, 
            hjust = -0.2, 
            size = 3,
            angle = 90) +
  ylim(0,3000)+
  ggtitle("DEGs - Up and down regulated, by vaccine, pvalue <0.05") +
  facet_wrap(~disease_vac, scales = "free_x", nrow = 2)

print(ggplot_degs)


ggsave(ggplot_degs, file = "DEGs_up_down_barplot_pvalue05.png",  width = 10, height = 8) 


```


Immune-related degs
```{r}

```



#Gene set databases

##Blood transcription modules <https://www.nature.com/articles/ni.2789>

```{r}
#Converter ítens de "Module member genes" em linhas em uma tabela longa
btm_annotation_table_long_immune <- BTM_modules %>%
  separate_rows(`Module member genes`, sep = ",") %>% 
  filter(`Module category` == "immune")

#Salvar
write.csv(btm_annotation_table_long_immune, file = "btm_annotation_table_long_immune.csv")

#Display
head(btm_annotation_table_long_immune)

#Calculate Set size
btm_annotation_nested <- btm_annotation_table_long_immune %>%
  group_by(`Module title`) %>% #Somar e agrupar
  mutate(SetSize = n())

#Convert gene symbol column to aggregated rows
btm_annotation_nested <- btm_annotation_nested %>%
  group_by(`Module title`) %>%
  summarize(Genes = paste(`Module member genes`, collapse = ", "), .groups = "drop") %>% 
  left_join(btm_annotation_nested, by = "Module title") %>% 
  filter(Genes != "NA") %>% 
  relocate(Genes, .after="SetSize")

#Salvar
write.csv(btm_annotation_nested, file = "btm_annotation_genes.csv")

#Convert to unique values table
btm_annotation_unique = btm_annotation_nested %>% 
  select(-`Module member genes`)

btm_annotation_unique = btm_annotation_unique %>% 
  distinct() %>% 
  relocate(Genes, .after="SetSize") %>% 
  filter(Genes != "NA")

#salvar
write.csv(btm_annotation_unique, file = "btm_annotation_unique.csv")

#Display
head(btm_annotation_unique)
```

## CellMarker

Baixar tabela dos sites "<http://117.50.127.228/CellMarker/CellMarker_download_files/file/Cell_marker_Human.xlsx>" ou "<http://117.50.127.228/CellMarker/CellMarker_download_files/file/Cell_marker_Human.xlsx>"

```{r}
##################### Baixar tabela de marcadores celulares

url <- "http://117.50.127.228/CellMarker/CellMarker_download_files/file/Cell_marker_Human.xlsx" #or http://117.50.127.228/CellMarker/CellMarker_download_files/file/Cell_marker_Human.xlsx 
f <- tempfile(fileext = ".xlsx")
download.file(url, f)
cell_marker_data <- readxl::read_excel(f, 1)

##################### Sub-agrupar células por sistema imunológico

#Filtrar marcadores e células por tecido (Blood)
cells <- cell_marker_data %>%
    filter(tissue_class == "Blood") %>%
    select(cell_name, marker)

##################### Filtrar células
# Acessar todos os tipos de células no dataset de Blood
cellmarker_celltypes <- unique(cells$cell_name)

# Filtrar apenas os tipos que contêm "B cell" e plasmacell
Bcells_no = cellmarker_celltypes[grep("B cell", cellmarker_celltypes)] #44 outputs
Plasmacells_no = cellmarker_celltypes[grep("plasma", cellmarker_celltypes)] #5 outputs

# Filtrar apenas os tipos que contêm "T cell" e seus subtipos "CD4" e "CD8"
Tcells_no = cellmarker_celltypes[grep("T cell", cellmarker_celltypes)] #93 outputs
TCD4_no = cellmarker_celltypes[grep("CD4", cellmarker_celltypes)] #35 outputs
TCD8_no = cellmarker_celltypes[grep("CD8", cellmarker_celltypes)] #34 outputs
Treg_no = cellmarker_celltypes[grep("Treg", cellmarker_celltypes)] #7 outputs
Thelper_no = cellmarker_celltypes[grep("helper", cellmarker_celltypes)] #17 outputs

# Filtrar outros tipos celulares da resposta imune
Neutrophils_no = cellmarker_celltypes[grep("neutrophil", cellmarker_celltypes)] #7 outputs
Monocyte_no = cellmarker_celltypes[grep("monocyte", cellmarker_celltypes)] #14 outputs
DCells_no = cellmarker_celltypes[grep("dendritic", cellmarker_celltypes)] #25 outputs
Macrophages_no = cellmarker_celltypes[grep("macrophage", cellmarker_celltypes)] #7 outputs
NKcells_no = cellmarker_celltypes[grep("killer", cellmarker_celltypes)] # 11 outputs
Neutrophils_no = cellmarker_celltypes[grep("mastcell", cellmarker_celltypes)] #7 outputs

# Criar variáveis para cada célula para enriquecimento

# Filtrar apenas as células da resposta imune adaptativa
Bcells <- filter(cells, str_detect(cell_name, regex("B cell|plasma cell|plasmablast", ignore_case = TRUE)))
TCD4 <- filter(cells, str_detect(cell_name, regex("CD4", ignore_case = TRUE)))
TCD8 <- filter(cells, str_detect(cell_name, regex("CD8", ignore_case = TRUE)))
Treg <- filter(cells, str_detect(cell_name, regex("Treg", ignore_case = TRUE)))
Thelper <- filter(cells, str_detect(cell_name, regex("helper", ignore_case = TRUE)))


# Filtrar células da resposta imune inata
Neutrophils <- filter(cells, str_detect(cell_name, regex("neutrophil", ignore_case = TRUE)))
Monocyte <- filter(cells, str_detect(cell_name, regex("monocyte", ignore_case = TRUE)))
DCells <- filter(cells, str_detect(cell_name, regex("dendritic", ignore_case = TRUE)))
Macrophages <- filter(cells, str_detect(cell_name, regex("macrophage", ignore_case = TRUE)))
Mastcells <- filter(cells, str_detect(cell_name, regex("mast cell", ignore_case = TRUE)))
NKcells <- filter(cells, str_detect(cell_name, regex("killer", ignore_case = TRUE)))
Platelets <- filter(cells, str_detect(cell_name, regex("Platelet", ignore_case = TRUE)))
Eosinophil <- filter(cells, str_detect(cell_name, regex("Eosinophil", ignore_case = TRUE)))
Basophil <- filter(cells, str_detect(cell_name, regex("Basophil", ignore_case = TRUE)))
Granulocyte <- filter(cells, str_detect(cell_name, regex("Granulocyte", ignore_case = TRUE)))

#Agrupar todos os resultados em Immune cells
Immune_adap = rbind(Bcells, TCD4, TCD8, Treg, Thelper)
Immune_adap$Type = "Adaptive"

Immune_innate = rbind(Neutrophils, Monocyte, DCells, Macrophages, Mastcells, NKcells, Platelets, Eosinophil, Basophil, Granulocyte)
Immune_innate$Type = "Innate"

#Unir tabelas
Immune_cells = rbind(Immune_adap, Immune_innate) 

#Salvar
write.csv(Immune_cells, file = "CellMarker_ImmuneCells.csv")

esquisser(Immune_cells_markersbycells)
```

Top 20 cells by #markers

```{r}

Immune_cells_markersbycells = Immune_cells %>% 
  group_by(cell_name, Type) %>% 
  summarise(n_markers = n()) %>% 
  arrange(desc(n_markers))
Immune_cells_markersbycells_top = Immune_cells_markersbycells[1:20, ]

ggImmune_cells_cellsbyimmune = ggplot(Immune_cells_markersbycells_top) +
  aes(x = reorder(cell_name, n_markers), y = n_markers, fill = Type) +
  scale_fill_manual(values = c('Innate' = "#989898",
                               'Adaptive' = "#6DF8DF")) +
  geom_col() +
  labs(
    x = "#Markers",
    y = "Cells",
    title = "Top 20 cells by #markers"
  ) +
  coord_flip() +
  theme_minimal() +
  geom_text(aes(label = n_markers, y = n_markers), 
            position = position_dodge(width = 0.9), 
            vjust = -0.2, 
            hjust = -0.2, 
            size = 3)

ggsave(ggImmune_cells_cellsbyimmune, file= "CellMarker_Top10_cellsbyn_markers.png", width = 10, height = 5)

print(ggImmune_cells_cellsbyimmune)

```

Number of Cells by immune system

```{r}
Immune_cells_cellsbyimmune = Immune_cells %>% 
  group_by(Type, cell_name) %>% 
  summarise(n_cells = n()) %>% 
  group_by(Type) %>% 
  summarise(n_cells = n())

ggImmune_cells_cellsbyimmune = ggplot(Immune_cells_cellsbyimmune) +
  aes(x = reorder(Type, n_cells), y = n_cells, fill = Type) +
  geom_col() +
  scale_fill_manual(values = c('Innate' = "#989898",
                               'Adaptive' = "#6DF8DF")) +
  labs(
    x = "#Cells",
    y = "Immune system",
    title = "#Cells by immune system"
  ) +
  coord_flip() +
  theme_minimal() +
  geom_text(aes(label = n_cells, y = n_cells), 
            position = position_dodge(width = 0.9), 
            vjust = -0.2, 
            hjust = -0.2, 
            size = 3)
ggsave(ggImmune_cells_cellsbyimmune, file= "CellMarker_nCellsbyimmunesystem.png", width = 10, height = 2)

print(ggImmune_cells_cellsbyimmune)
```

## ImmuneGO

A ontologia de processo biológico "immune system process" (<GO:0002376>) foi selecionada para uma análise direcionada das DEGs em relação à resposta imunológica. O dataset foi baixado pelo QuickGO filtrando os campos de Taxon (H. sapiens), Gene Products (all marked, except for "unreviewed") e Aspect ("Biological Process").

Para ler mais sobre códigos de evidencia: <https://geneontology.org/docs/guide-go-evidence-codes/>

------------------------------------------------------------------------

Baixar dataset e importar

```{r}

######### BAIXAR DATASET ############### 
#Baixar dataset no site: https://www.ebi.ac.uk/QuickGO/annotations (.tsv)

#Importar
Immune_GO = read.table("~/Desktop/RNAseq temporary files/Gene set enrichment/28-08-23/QuickGO-ImmuneProcess.tsv", header = TRUE, sep = "\t")
Immune_GO$GO.NAME = toupper(Immune_GO$GO.NAME)
View(Immune_GO)

#Renomear
Immune_GO_T2G = Immune_GO %>% 
  select(GO.NAME, SYMBOL) %>% 
  arrange(GO.NAME) %>% 
  rename(Process = GO.NAME, 
         GENE = SYMBOL, 
         EVIDENCE = GO.EVIDENCE)

#Salvar
saveRDS(Immune_GO, file = "Immune_GO.rds")
write.csv(Immune_GO, file = "Immune_GO.csv")

############### DATASET DESCRIPTION ############### 
Immune_GO_genes_GO = Immune_GO %>% 
  select(SYMBOL, GO.NAME, GO.TERM, GO.EVIDENCE.CODE) %>% 
  unique() %>% 
  data.frame()

Immune_GO_genes = Immune_GO %>% 
  select(SYMBOL) %>% 
  unique() %>% 
  data.frame()

Immune_GO_GOs = Immune_GO %>% 
  select(GO.NAME) %>% 
  unique() %>% 
  data.frame()

Immune_GO_Evidence = Immune_GO %>% 
  select(GO.EVIDENCE.CODE) %>% 
  unique() %>% 
  data.frame()

#Stats
dim(Immune_GO) #4098 GOs
dim(Immune_GO_genes) #1251 GENES
dim(Immune_GO_genes_GO) #3889 GENES-TERMS
dim(Immune_GO_GOs) #336 TERMS-NAMES
dim(Immune_GO_Evidence) #12 

# Filtrar genes "NOT|involved_in" (QUALIFIER)
Immune_GO_genes_GO_Filt = Immune_GO %>%
  filter(QUALIFIER != "NOT|involved_in") %>% 
  write.csv(file="Immune_GO_genes_GO_Filt.csv") #Salvar

explore(Immune_GO_genes_GO_Filt) 
#4082 genes-terms
# 1294 Gene product ID
# 1247 gene symbols
# 3 qualifiers
# GO.EVIDENCE.CODE = 12
# 332 Terms, GO.NAME = 336
# 744 references
# 18 databases
```

### Descrevendo e organizando dataset
Annotation of GOs

```{r}
######## Anotação de GOs ########
ImmuneGO_Manual = ImmuneGO_Annotated_GO # ImmuneGOs manually annotated by Immune System, Sub-system, and cells
#Unir anotações com genes
ImmuneGO_Annotated = ImmuneGO_Manual %>% 
  merge(Immune_GO_genes_GO_Filt, by.x = "Process", by.y = "GO.NAME", all.y = F, all.x=T) %>% 
  select(SYMBOL, everything()) %>% 
  rename(Evidence = GO.EVIDENCE.CODE)

#Calculate Set size
ImmuneGO_Annotated_genes <- ImmuneGO_Annotated %>%
  group_by(Process) %>% #Somar e agrupar
  mutate(SetSize = n())

#Convert gene symbol column to aggregated rows
ImmuneGO_Annotated_genes <- ImmuneGO_Annotated_genes %>%
  group_by(Process) %>%
  summarize(Genes = paste(SYMBOL, collapse = ","), .groups = "drop") %>% 
  left_join(ImmuneGO_Annotated_genes, by = "Process") %>% 
  filter(SYMBOL != "NA") %>% 
  relocate(Genes, .after="SetSize")
  
#View(ImmuneGO_Annotated_genes)

#Convert to unique values table
ImmuneGO_Annotated_unique = ImmuneGO_Annotated_genes %>% 
  select(Process, `Gene set, short`, `Immune system`, `Immune sub-system`, `Immune tissue`, `GO.TERM`, SetSize, Genes) %>% 
  distinct() %>% 
  relocate(Genes, .after="SetSize") %>% 
  filter(Genes != "NA")

#Retirar processos irrelevantes e remover espaços em branco
ImmuneGO_Annotated_unique = ImmuneGO_Annotated_unique %>% 
  filter(`Immune sub-system` != "Erythrocyte", `Immune tissue` != "Erythrocyte") %>%  
  mutate(Genes = str_replace_all(Genes, " ", ""))

# Calcular o tamanho do gene set

ImmuneGO_Annotated_unique <- ImmuneGO_Annotated_unique %>%
  mutate(SetSize = lengths(str_split(Genes, ",")),
         Genes = str_c(unique(str_split(Genes, ",")), collapse = ","))

# ImmuneGO_Annotated_unique$SetSize <- sapply(strsplit(ImmuneGO_Annotated_unique$Genes, ","), function(x) length(unique(x)))
# ImmuneGO_Annotated_unique$Genes <- sapply(strsplit(ImmuneGO_Annotated_unique$Genes, ","), function(x) paste(unique(x), collapse = ","))

#Salvar
write.csv(ImmuneGO_Annotated_genes, file="ImmuneGO_Annotated_Genes.csv")
write.csv(ImmuneGO_Annotated_unique, file="ImmuneGO_Annotated_GO.csv")

head(ImmuneGO_Annotated_unique)
```

Agrupar genes de todos os processos do sistema imune

```{r}
ImmuneGO_Annotated_unique = ImmuneGO_Annotated_GO

#Agrupar genes de todos os processos do sistema imune -----------------------------------------
ImmuneGO_Adaptive <- ImmuneGO_Annotated_unique %>%
  filter(`Immune system` == "Adaptive")
ImmuneGO_Innate <- ImmuneGO_Annotated_unique %>%
  filter(`Immune system` == "Innate")
ImmuneGO_Complement <- ImmuneGO_Annotated_unique %>%
  filter(`Immune system` == "Complement")

ImmuneGO_Inflammation <- ImmuneGO_Annotated_unique %>%
  filter(immune_sub_system == "Inflammation")

ImmuneGO_Interferon <- ImmuneGO_Annotated_unique %>%
  filter(grepl("Interferon", gene_set_short))

ImmuneGO_ARPP = ImmuneGO_Annotated_unique %>%
  filter(immune_sub_system == "PRR")

ImmuneGO_ARPP = ImmuneGO_Annotated_unique %>%
  filter(immune_sub_system == "APC")


ImmuneGO_Innate_Macrophage <- ImmuneGO_Annotated_unique %>%
ImmuneGO_Innate_Macrophage <- ImmuneGO_Annotated_unique %>%
  filter(`Immune tissue` == "Macrophage")

ImmuneGO_Innate_Mastcell <- ImmuneGO_Annotated_unique %>%
ImmuneGO_Innate_Mastcell <- ImmuneGO_Annotated_unique %>%
  filter(`Immune tissue` == "Mast cell")
ImmuneGO_Innate_Neutrophil <- ImmuneGO_Annotated_unique %>%
  filter(`Immune tissue` == "Neutrophil")
ImmuneGO_Innate_Eosinophil <- ImmuneGO_Annotated_unique %>%
  filter(`Immune tissue` == "Eosinophil")
ImmuneGO_Innate_NK <- ImmuneGO_Annotated_unique %>%
  filter(`Immune tissue` == "NK cell")
ImmuneGO_Innate_Monocyte <- ImmuneGO_Annotated_unique %>%
  filter(`Immune tissue` == "Monocyte")
ImmuneGO_Adap_Cellular <- ImmuneGO_Adaptive %>% #Este comando somente inclui os genes de processos da resposta celular e muitos dos genes de receptor de celulas T não estão anotados. Por isso, eles serão incluidos nos codigos abaixo
  filter(`Immune sub-system` == "Cellular")
ImmuneGO_Adap_Humoral <- ImmuneGO_Adaptive %>%
  filter(`Immune sub-system` == "Humoral")


#Unificar genes e remover duplicatas
ImmuneGO_Adaptive_genes <- paste(ImmuneGO_Adaptive$Genes, collapse = ",")
ImmuneGO_Adaptive_genes <- unique(unlist(strsplit(ImmuneGO_Adaptive$Genes, ",")))

ImmuneGO_Innate_genes <- paste(ImmuneGO_Innate$Genes, collapse = ",")
ImmuneGO_Innate_genes <- unique(unlist(strsplit(ImmuneGO_Innate$Genes, ","))) 

ImmuneGO_Complement_genes <- paste(ImmuneGO_Complement$Genes, collapse = ",")
ImmuneGO_Complement_genes <- unique(unlist(strsplit(ImmuneGO_Complement$Genes, ",")))
ImmuneGO_Cellular_genes <- paste(ImmuneGO_Adap_Cellular$Genes, collapse = ",")
ImmuneGO_Cellular_genes <- unique(unlist(strsplit(ImmuneGO_Adap_Cellular$Genes, ",")))
ImmuneGO_Humoral_genes <- paste(ImmuneGO_Adap_Humoral$Genes, collapse = ",")
ImmuneGO_Humoral_genes <- unique(unlist(strsplit(ImmuneGO_Adap_Humoral$Genes, ",")))

ImmuneGO_Innate_Macrophage_genes <- paste(ImmuneGO_Innate_Macrophage$Genes, collapse = ",")
ImmuneGO_Innate_Macrophage_genes <- unique(unlist(strsplit(ImmuneGO_Innate_Macrophage$Genes, ",")))

ImmuneGO_Innate_Mastcell_genes <- paste(ImmuneGO_Innate_Mastcell$Genes, collapse = ",")
ImmuneGO_Innate_Mastcell_genes <- unique(unlist(strsplit(ImmuneGO_Innate_Mastcell$Genes, ",")))

ImmuneGO_Innate_Neutrophil_genes <- paste(ImmuneGO_Innate_Neutrophil$Genes, collapse = ",")
ImmuneGO_Innate_Neutrophil_genes <- unique(unlist(strsplit(ImmuneGO_Innate_Neutrophil$Genes, ",")))

ImmuneGO_Innate_Eosinophil_genes <- paste(ImmuneGO_Innate_Eosinophil$Genes, collapse = ",")
ImmuneGO_Innate_Eosinophil_genes <- unique(unlist(strsplit(ImmuneGO_Innate_Eosinophil$Genes, ",")))

ImmuneGO_Innate_NK_genes <- paste(ImmuneGO_Innate_NK$Genes, collapse = ",")
ImmuneGO_Innate_NK_genes <- unique(unlist(strsplit(ImmuneGO_Innate_NK$Genes, ",")))

ImmuneGO_Innate_Monocyte_genes <- paste(ImmuneGO_Innate_Monocyte$Genes, collapse = ",")

ImmuneGO_Innate_Monocyte_genes <- unique(unlist(strsplit(ImmuneGO_Innate_Monocyte$Genes, ",")))

ImmuneGO_Innate_Inflammation_genes <- paste(ImmuneGO_Inflammation $genes, collapse = ",")

ImmuneGO_Innate_Inflammation_genes <- unique(unlist(strsplit(ImmuneGO_Inflammation$genes, ",")))

# Calcular o tamanho do gene set
SetSize_adap <- length(ImmuneGO_Adaptive_genes) 
SetSize_innate <- length(ImmuneGO_Innate_genes)
SetSize_compl <- length(ImmuneGO_Complement_genes) 
SetSize_cellular <- length(ImmuneGO_Cellular_genes) 
SetSize_mast <- length(ImmuneGO_Innate_Mastcell_genes) 
SetSize_neutrophil <- length(ImmuneGO_Innate_Neutrophil_genes) 
SetSize_eosinophil <- length(ImmuneGO_Innate_Eosinophil_genes) 
SetSize_NK <- length(ImmuneGO_Innate_NK_genes) 
SetSize_Mono <- length(ImmuneGO_Innate_Monocyte_genes) 
SetSize_Macro <- length(ImmuneGO_Innate_Macrophage_genes) 

#Criar novas linhas
new_row_adap <- tibble(
  `Gene set, short` = "Adaptive immune system",
  `Process` = "ADAPTIVE IMMUNE SYSTEM", 
  `Immune system` = "Adaptive",
  `Immune sub-system` = "General",
  `Immune tissue` = "General",
  Genes = paste(ImmuneGO_Adaptive_genes, collapse = ","),
  SetSize = SetSize_adap
)

new_row_innate <- tibble(
  `Gene set, short` = "Innate immune system",
  `Process` = "INNATE IMMUNE SYSTEM", 
  `Immune system` = "Innate",
  `Immune sub-system` = "General",
  `Immune tissue` = "General",
  Genes = paste(ImmuneGO_Innate_genes, collapse = ","),
  SetSize = SetSize_innate
)

new_row_compl <- tibble(
  `Gene set, short` = "Complement immune system",
  `Process` = "COMPLEMENT IMMUNE SYSTEM", 
  `Immune system` = "Complement",
  `Immune sub-system` = "General",
  `Immune tissue` = "General",
  Genes = paste(ImmuneGO_Complement_genes, collapse = ","),
  SetSize = SetSize_compl
)

new_row_cellular <- tibble(
  `Gene set, short` = "Cellular adaptive immune system",
  `Process` = "CELLULAR ADAPTIVE IMMUNE SYSTEM", 
  `Immune system` = "Adaptive",
  `Immune sub-system` = "Cellular",
  `Immune tissue` = "General",
  `GO.TERM` = "Manual",
  Genes = paste(ImmuneGO_in_vaccines_Genes_ann_adap_cellular$SYMBOL, collapse = ","),
  SetSize = SetSize_cellular
)

new_row_humoral <- tibble(
  `Gene set, short` = "Humoral adaptive immune system",
  `Process` = "HUMORAL ADAPTIVE IMMUNE SYSTEM", 
  `Immune system` = "Adaptive",
  `Immune sub-system` = "Humoral",
  `Immune tissue` = "General",
  `GO.TERM` = "Manual",
  Genes = paste(ImmuneGO_Humoral_genes, collapse = ","),
  SetSize = SetSize_humoral
)

new_row_mast <- tibble(
  `Gene set, short` = "Mast cell",
  `Process` = "MAST CELL", 
  `Immune system` = "Innate",
  `Immune sub-system` = "Cell",
  `Immune tissue` = "Mast cell",
  `GO.TERM` = "Manual",
  Genes = paste(ImmuneGO_Innate_Mastcell_genes, collapse = ","),
  SetSize = SetSize_mast 
)

new_row_neutro <- tibble(
  `Gene set, short` = "Neutrophil",
  `Process` = "NEUTROPHIL", 
  `Immune system` = "Innate",
  `Immune sub-system` = "Cell",
  `Immune tissue` = "Neutrophil",
  `GO.TERM` = "Manual",
  Genes = paste(ImmuneGO_Innate_Neutrophil_genes , collapse = ","),
  SetSize = SetSize_neutrophil 
)

new_row_eosinophil <- tibble(
  `Gene set, short` = "Eosinophil",
  `Process` = "EOSINOPHIL", 
  `Immune system` = "Innate",
  `Immune sub-system` = "Cell",
  `Immune tissue` = "Eosinophil",
  `GO.TERM` = "Manual",
  Genes = paste(ImmuneGO_Innate_Eosinophil_genes, collapse = ","),
  SetSize = SetSize_eosinophil
)

new_row_nk <- tibble(
  `Gene set, short` = "NK cell",
  `Process` = "NK CELL", 
  `Immune system` = "Innate",
  `Immune sub-system` = "Cell",
  `Immune tissue` = "NK cell",
  `GO.TERM` = "Manual",
  Genes = paste(ImmuneGO_Innate_NK_genes, collapse = ","),
  SetSize = SetSize_NK 
)

new_row_mono <- tibble(
  `Gene set, short` = "Monocytes",
  `Process` = "MONOCYTES", 
  `Immune system` = "Innate",
  `Immune sub-system` = "Monocytes",
  `Immune tissue` = "Cell",
  `GO.TERM` = "Manual",
  Genes = paste(ImmuneGO_Innate_Monocyte_genes , collapse = ","),
  SetSize = SetSize_Mono
)

new_row_macro <- tibble(
  `Gene set, short` = "Macrophages",
  `Process` = "MACROPHAGES", 
  `Immune system` = "Innate",
  `Immune sub-system` = "Macrophages",
  `Immune tissue` = "Cell",
  `GO.TERM` = "Manual",
  Genes = paste(ImmuneGO_Innate_Macrophage_genes , collapse = ","),
  SetSize = SetSize_Macro
)


# Adicionar linhas
ImmuneGO_Annotated_unique <- bind_rows(ImmuneGO_Annotated_unique, 
                                       new_row_mast, 
                                       new_row_neutro, 
                                       new_row_eosinophil, 
                                       new_row_nk, 
                                       new_row_mono,
                                       new_row_macro) %>% 
  clean_names()


#Transformar em tabela de genes por linhas
ImmuneGO_Annotated_Genes <- ImmuneGO_Annotated_unique %>%
  separate_rows(genes, sep = ",")


# Separar genes de TCR e BCR
TCR_BCR_repertoire = ImmuneGO_Annotated_Genes %>% 
  filter(grepl("^TR", genes) & process == "ADAPTIVE IMMUNE SYSTEM" |
           grepl("^IG", genes) & process == "ADAPTIVE IMMUNE SYSTEM" |
           grepl("^TR", genes) & process == "CELLULAR ADAPTIVE IMMUNE SYSTEM" |
         grepl("^IG", genes) & process == "HUMORAL ADAPTIVE IMMUNE SYSTEM") %>% 
  mutate(process = if_else(process == "CELLULAR ADAPTIVE IMMUNE SYSTEM", "TCR REPERTOIRE", 
                           "BCR REPERTOIRE"),
         gene_set_short = if_else(process == "TCR REPERTOIRE", 
                                  "TCR REPERTOIRE", 
                           "BCR REPERTOIRE"),
         process = if_else(grepl("^TR", genes), "TCR REPERTOIRE", "BCR REPERTOIRE"))

TCR_BCR_repertoire_genes = TCR_BCR_repertoire %>% 
  select(genes) %>% 
  distinct()

ImmuneGO_Annotated_GO = ImmuneGO_Annotated_Genes %>%
  anti_join(TCR_BCR_repertoire_genes, by = "genes") %>% 
  bind_rows(TCR_BCR_repertoire) %>% 
  group_by(process) %>%
  summarize(genes = paste(genes, collapse = ",")) %>% 
  left_join(ImmuneGO_Annotated_unique %>% select(-genes) %>% distinct(), by = "process") %>% 
  arrange(desc(set_size)) %>% 
  distinct()


ImmuneGO_Annotated_Genes <- ImmuneGO_Annotated_GO %>%
  separate_rows(genes, sep = ",")

#Salvar
write.csv(ImmuneGO_Annotated_GO, file = "ImmuneGO_Annotated_GO.csv")
write.csv(ImmuneGO_Annotated_Genes, file = "ImmuneGO_Annotated_Genes.csv")

#Display
view(ImmuneGO_Annotated_GO)
view(ImmuneGO_Annotated_Genes)

```



```{r}
ImmuneGO_Annotated_GO_2 = ImmuneGO_Annotated_Genes_Nested_Interesting_12_12_23 %>%
  group_by(process) %>%
  summarize(genes = paste(genes, collapse = ",")) %>% 
  left_join(ImmuneGO_Annotated_Genes_Nested_Interesting_12_12_23 %>% select(-genes) %>% distinct(), by = "process") %>% 
  select(!c("...1", "...2")) %>% 
  distinct() %>% 
  filter(process %in% c("ANTIVIRAL AND INTERFERON", 
                        "ARPP", 
                        "B CELLS", 
                        "DENDRITIC CELLS", 
                        "INFLAMMATION", 
                        "IMMUNOGLOBULIN MEDIATED IMMUNE RESPONSE",
                        "T CELLS")) %>% 
  select(process, genes) %>% 
  distinct() %>% 
  bind_rows(ImmuneGO_Annotated_GO)


write.csv(ImmuneGO_Annotated_GO_2, file = "ImmuneGO_Annotated_GO_1_1_24.csv")

```



Unir genes da resposta celular

```{r}
#Unir genes da resposta celular, incluindo TCRs que estavam em outro processo
ImmuneGO_in_vaccines_Genes_ann_adap_cellular = filter(ImmuneGO_in_vaccines_Genes_ann, Process == "CELLULAR ADAPTIVE IMMUNE SYSTEM")
ImmuneGO_in_vaccines_Genes_ann_adap_cellular <- filter(ImmuneGO_in_vaccines_Genes_ann, 
                                                       Process == "CELLULAR ADAPTIVE IMMUNE SYSTEM" | 
                                                         grepl("^TRA|^TRB|^TRD|^TRG|^TRIL", SYMBOL)) 

ImmuneGO_in_vaccines_Genes_ann_adap_cellular$Process = "CELLULAR ADAPTIVE IMMUNE SYSTEM"
ImmuneGO_Cellular_genes <- paste(ImmuneGO_in_vaccines_Genes_ann_adap_cellular$SYMBOL, collapse = ",")
ImmuneGO_Cellular_genes <- unique(unlist(strsplit(ImmuneGO_in_vaccines_Genes_ann_adap_cellular$SYMBOL, ",")))

SetSize_cellular <- length(ImmuneGO_Cellular_genes) 
new_row_cellular <- tibble(
  `Gene set, short` = "Cellular adaptive immune system",
  `Process` = "CELLULAR ADAPTIVE IMMUNE SYSTEM", 
  `Immune system` = "Adaptive",
  `Immune sub-system` = "Cellular",
  `Immune tissue` = "General",
  Genes = paste(ImmuneGO_Cellular_genes, collapse = ","),
  SetSize = SetSize_cellular
)

ImmuneGO_Annotated_unique <- bind_rows(ImmuneGO_Annotated_unique, 
                                       new_row_cellular)

#Retirar linha antiga 
# ImmuneGO_Annotated_unique <- filter(ImmuneGO_Annotated_unique, 
#                                     !(Process == "CELLULAR ADAPTIVE IMMUNE SYSTEM" & SetSize == 384))

#Transformar em tabela de genes por linhas
ImmuneGO_Annotated_Genes <- ImmuneGO_Annotated_unique %>%
  separate_rows(Genes, sep = ",")

#Salvar
# write.csv(ImmuneGO_Annotated_unique, file = "ImmuneGO_Annotated_GO.csv")
# write.csv(ImmuneGO_Annotated_Genes, file = "ImmuneGO_Annotated_Genes.csv")
```


Separar genes do BCR e TCR

```{r}

TCR_BCR_repertoire = ImmuneGO_Annotated_Genes_Nested_Interesting %>% 
  filter(grepl("^TR", genes) & process == "ADAPTIVE IMMUNE SYSTEM" |
           grepl("^IG", genes) & process == "ADAPTIVE IMMUNE SYSTEM" |
           grepl("^TR", genes) & process == "CELLULAR ADAPTIVE IMMUNE SYSTEM" |
         grepl("^IG", genes) & process == "HUMORAL ADAPTIVE IMMUNE SYSTEM") %>% 
  mutate(process = if_else(process == "CELLULAR ADAPTIVE IMMUNE SYSTEM", "TCR REPERTOIRE", 
                           "BCR REPERTOIRE"),
         gene_set_short = if_else(process == "TCR REPERTOIRE", 
                                  "TCR REPERTOIRE", 
                           "BCR REPERTOIRE"),
         process = if_else(grepl("^TR", genes), "TCR REPERTOIRE", "BCR REPERTOIRE"))

TCR_BCR_repertoire_genes = TCR_BCR_repertoire %>% 
  select(genes) %>% 
  distinct()

ImmuneGO_Annotated_Genes_Nested_Interesting_12_12_23 = ImmuneGO_Annotated_Genes_Nested_Interesting %>%
  anti_join(TCR_BCR_repertoire_genes, by = "genes") %>% 
  bind_rows(TCR_BCR_repertoire)

write.csv(ImmuneGO_Annotated_Genes_Nested_Interesting_12_12_23, file = "ImmuneGO_Annotated_Genes_Nested_Interesting_12_12_23.csv")

```


Top 10 gene sets

```{r}
#Top 10 gene sets
ImmuneGO_Annotated_unique_top10 = ImmuneGO_Annotated_GO %>% 
  clean_names() %>% 
  arrange(desc(set_size)) %>% 
  slice_head(n = 10) %>% 
  mutate(immune_system = if_else(gene_set_short == "Adaptive Immune response", "Adaptive", immune_system))

#Gráfico de barras
ImmuneGO_top10_barplot = ggplot(ImmuneGO_Annotated_unique_top10) +
 aes(x = reorder(gene_set_short, set_size), 
     weight = set_size, 
     fill = immune_system) +  # Use reorder para reordenar as barras +
  geom_bar()+
  scale_fill_manual(
    values = c(Adaptive = "#6DF8DF",
    Complement = "#e9edc9",
    General = "#bde0fe",
    Innate = "#989898",
    Undefined = "#edede9")) +
 labs(y = "Set size", title = "ImmuneGO", subtitle = "Top 10 GO, by set size") +
 coord_flip() +
 theme_minimal()

#Salvar
ggsave(ImmuneGO_top10_barplot, file="ImmuneGO_top10_barplot.png", width = 10, height = 6)

#Display
head(ImmuneGO_top10_barplot)
```

Quais os tamanhos dos gene sets?

```{r}

# Usar cut para dividir os valores em intervalos
ImmuneGO_Annotated_GO$SetSize_intervals <- cut(ImmuneGO_Annotated_GO$SetSize, c(0, 10, 50, 100, Inf), labels = FALSE)
ImmuneGO_Annotated_GO = ImmuneGO_Annotated_GO %>% arrange(desc(SetSize))

# Contar quantas linhas caem em cada intervalo
contagem <- table(ImmuneGO_Annotated_GO$SetSize_intervals)

#Definir ordem de fatores para plotar
ImmuneGO_Annotated_GO$`Immune system` <- factor(ImmuneGO_Annotated_GO$`Immune system`, levels = c("General", "Adaptive", "Innate", "Complement", "Undefined"))

#Plotar
plot_hist_immunego = ggplot(ImmuneGO_Annotated_GO) +
  aes(x = SetSize_intervals, fill = `Immune system`, show.legend = FALSE) +
  geom_histogram(bins = 4L) +
  scale_fill_manual(
    values = c(Adaptive = "#6DF8DF",
    Complement = "#e9edc9",
    General = "#bde0fe",
    Innate = "#989898",
    Undefined = "#edede9")) +
  theme_minimal() +
  #facet_grid(vars(), vars(`Immune system`)) +
  theme(legend.position = "top",
        panel.grid.major = element_blank()) +  # Remove as gridlines secundárias
  scale_x_continuous(breaks = c(1, 2, 3, 4, 5), labels = c("0-10", "10-20", "20-50", "50-100", "100+")) +
  guides(color = "none")

#Salvar
ggsave(plot_hist_immunego, file = "plot_hist_immunego.png", width = 10, height = 5)

#Display
plot_hist_immunego
```

Qual a relação entre os genes de cada resposta imunológica?
```{r}
ImmuneGO_Annotated_Genes = ImmuneGO_Annotated_Genes %>% clean_names()

#Unir tabelas
ImmuneGO_Annotated_Genes_Nested = ImmuneGO_Annotated_Genes %>% 
  clean_names() %>% 
  group_by(process) %>% 
  arrange(desc(set_size)) %>% 
  filter(process %in% c("ADAPTIVE IMMUNE SYSTEM", "INNATE IMMUNE SYSTEM", "CELLULAR ADAPTIVE IMMUNE SYSTEM", "HUMORAL ADAPTIVE IMMUNE SYSTEM", "IMMUNOGLOBULIN MEDIATED IMMUNE RESPONSE"))

Immune_GO_adap_in = ImmuneGO_Annotated_Genes %>%
  clean_names() %>%
  mutate(
    process = case_when(
      immune_tissue == "Complement" ~ "COMPLEMENT",
      immune_sub_system == "Inflammation" ~ "INFLAMMATION",
      immune_tissue == "B cell" ~ "B CELLS",
      immune_tissue == "T cell" ~ "T CELLS",
      immune_tissue == "Dendritic cell" ~ "DENDRITIC CELLS",
      immune_tissue == "Macrophage" ~ "MACROPHAGES",
      immune_sub_system %in% c("APP", "PRR") ~ "ARPP",
      process %in% c(
        "ANTIVIRAL INNATE IMMUNE RESPONSE",
        "CELLULAR RESPONSE TO TYPE I INTERFERON",
        "CELLULAR RESPONSE TO TYPE II INTERFERON",
        "TYPE I INTERFERON-MEDIATED SIGNALING PATHWAY",
        "TYPE II INTERFERON-MEDIATED SIGNALING PATHWAY",
        "TYPE III INTERFERON-MEDIATED SIGNALING PATHWAY"
      ) ~ "ANTIVIRAL AND INTERFERON",
      TRUE ~ as.character(process)
    )
  ) %>% 
  filter(process %in% c("COMPLEMENT",
                        "INFLAMMATION",
                        "B CELLS",
                        "T CELLS",
                        "DENDRITIC CELLS",
                        "MACROPHAGES",
                        "ARPP",
                        "ANTIVIRAL AND INTERFERON"))
#Unir processos
ImmuneGO_Annotated_Genes_Nested_Interesting <- bind_rows(ImmuneGO_Annotated_Genes_Nested,
                                             Immune_GO_adap_in) %>% 
  distinct() #retirar genes duplicados por processo

#Salvar
write.csv(ImmuneGO_Annotated_Genes_Nested_Interesting, file = "ImmuneGO_Annotated_Genes_Nested_Interesting.csv")
```


Quais são os genes mais compartilhados?

```{r}
ImmuneGO_Annotated_Genes_sharedbetween_subsystem



#Top 10 gene sets
ImmuneGO_Annotated_Genes_top20 = ImmuneGO_Annotated_Genes_sharedbetween_subsystem %>% 
  clean_names() %>% 
  arrange(desc(shared)) %>% 
  slice_head(n = 20) 

#Gráfico de barras
ImmuneGO_Genes_top20_barplot = ggplot(ImmuneGO_Annotated_Genes_top20) +
 aes(x = reorder(genes, shared), 
     weight = shared) +
  geom_bar(fill="lightblue") +
 labs(y = "Shared", title = "ImmuneGO", subtitle = "Top 10 genes, shared") +
 coord_flip() +
 theme_minimal()

#Salvar
ggsave(ImmuneGO_Genes_top20_barplot, file="ImmuneGO_Genes_top20_barplot.png", width = 5, height = 10)

#Display
head(ImmuneGO_top10_barplot)


```

```{r}
############### VENN DIAGRAM
#Input
data = ImmuneGO_Annotated_Genes_Nested_Interesting

# Função para calcular a sobreposição entre pares de vacinas e a porcentagem de genes compartilhados
overlap_genes <- function(cond1, cond2, data) {
  genes_cond1 <- data$genes[data$process == cond1] #Trocar coluna
  genes_cond2 <- data$genes[data$process == cond2] #Trocar coluna
  genes_shared <- intersect(genes_cond1, genes_cond2)
  genes_notshared_cond1 <- setdiff(genes_cond1, genes_cond2)
  genes_notshared_cond2 <- setdiff(genes_cond2, genes_cond1)
  
  total_genes_cond1 <- length(genes_cond1)
  total_genes_cond2 <- length(genes_cond2)
  
  percentage_shared_cond1 <- length(genes_shared) / total_genes_cond1 * 100
  percentage_shared_cond2 <- length(genes_shared) / total_genes_cond2 * 100
  
  shared_genes <- data.frame(
    Cond1 = cond1,
    Cond2 = cond2,
    Shared = length(genes_shared),
    NotShared_cond1 = length(genes_notshared_cond1),
    NotShared_cond2 = length(genes_notshared_cond2),
    Total_Genes_Cond1 = total_genes_cond1,
    Total_Genes_Cond2 = total_genes_cond2,
    Genes_Names = paste(genes_shared, collapse = ", "),
    Percentage_Shared_Cond1 = percentage_shared_cond1,
    Percentage_Shared_Cond2 = percentage_shared_cond2
  )
  
  return(shared_genes)
}
# Obtenha uma lista de todas as vacinas únicas
unique_cond <- unique(data$process)  #trocar "study" pelo nome da coluna de interesse

# Inicialize um dataframe para armazenar os resultados
shared_genes_df <- data.frame()

# Calcular a sobreposição e porcentagem para cada par de vacinas
for (i in 1:(length(unique_cond) - 1)) {
  for (j in (i + 1):length(unique_cond)) {
    resultado_temp <- overlap_genes(unique_cond[i], unique_cond[j], data)
    shared_genes_df <- bind_rows(shared_genes_df, resultado_temp)
  }
}

# Função para realizar o teste exato de Fisher
fisher_exact_test <- function(shared, notshared1, notshared2) {
  cont_table <- matrix(c(shared, notshared1, notshared2, 0), nrow = 2)
  results_fisher <- fisher.test(cont_table)
  return(results_fisher$p.value)
}

# Aplicar a função a cada linha da tabela de resultados
shared_genes_df$pvalue <- mapply(fisher_exact_test, 
                                    shared_genes_df$Shared, 
                                    shared_genes_df$NotShared_cond1, 
                                    shared_genes_df$NotShared_cond2)


# Calcular -log(p)
shared_genes_df$`-log(p)`= -log(shared_genes_df$pvalue, 10)

#Duplicar e trocar colunas (Vacina 1 e Vacina 2)
shared_genes_df2 = shared_genes_df
shared_genes_df2[, c("Cond1", "Cond2")] <- shared_genes_df2[, c("Cond2", "Cond1")]

#Unir datasets com os dados espelhados
shared_genes_df_mirror = rbind(shared_genes_df, shared_genes_df2)

#Converter NA, NaNe inf em 0
shared_genes_df_mirror[is.na(shared_genes_df_mirror)] <- 0
shared_genes_df_mirror<- replace(shared_genes_df_mirror, shared_genes_df_mirror == "Inf", 0)

#Arredondar porcentagens
shared_genes_df_mirror = shared_genes_df_mirror %>%
  mutate(Percentage_Shared_Cond1 = round(Percentage_Shared_Cond1, 2),
         Percentage_Shared_Cond2 = round(Percentage_Shared_Cond2, 2))

#Salvar resultados
write.csv(shared_genes_df_mirror, file = "ImmuneGO_sharedgenes.csv") #Alterar

```

Heatmap
```{r}


esquisser(shared_genes_df_mirror)

matrix = shared_genes_df_mirror %>% 
  select(Cond1, Cond2, Shared) %>% 
  pivot_wider(names_from = "Cond2", 
              values_from = "Shared", 
              values_fn = mean) %>% 
  replace(is.na(.), 0) %>% 
  column_to_rownames(var="Cond1") %>% 
  as.matrix() %>% 
  round(0)


col_fun <- colorRamp2(c(0, 100), c("white", "#6DF8DF"))

#Plotar

file = "ImmuneGO_SharedGenes_total"
fileimageheatmap_grouped= paste0(file, sep ="_", "Complexheatmap.png")

png(file = fileimageheatmap_grouped, width=50, height=30,units="cm",res=800)

heatmap_plot = Heatmap(matrix,
                        top_annotation = NULL,
                        left_annotation = NULL,
                        cell_fun = function(j, i, x, y, width, height, fill) {
        grid.text(sprintf("%.1f", matrix[i, j]), x, y, gp = gpar(fontsize = 10))}, #Mostrar valores nas células
                        show_row_names = TRUE,
                        row_names_side = "left",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        col = col_fun, #color
                        # column_title = "Target", 
                        # row_title = "Source",
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 12),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = gpar(fontsize = 12), #Grande = 6, Pequeno 15
                        row_dend_width = unit(5, "cm"), #Altura do dendrograma
                        row_split = NULL, #Split row clusters (NULL se for um gene set mais específico)
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(8, "mm"),
                        show_column_dend = FALSE,
                        show_row_dend = FALSE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = unit(20, "cm"), # 20 para pequeno e grande
                        height = unit(20, "cm") # 20 para pequeno, 60 para grande
)

draw(heatmap_plot, 
     annotation_legend_side = "left", 
     heatmap_legend_side = "right")
dev.off()



```



### Verificando representação nos datasets

Total de genes representados

```{r}
#Genes representados e não representados

ImmuneGO_Annotated_Genes = ImmuneGO_Annotated_Genes_Nested_Interesting %>% 
  select(genes, process) 

ImmuneGO_Annotated_Genes_2 = ImmuneGO_Annotated_Genes_Nested_Interesting %>% 
  select(genes) %>% 
  distinct() %>% 
  mutate(process = "All") %>% 
  bind_rows(ImmuneGO_Annotated_Genes_Nested_Interesting)


ImmuneGO_Annotated_Genes_Interesting_studies = ImmuneGO_Annotated_Genes_2 %>% 
  merge(genes_raw_combined, all.x = T, all.y =F) %>% 
  select(study, process, genes) %>% 
  group_by(study, process) %>% 
  mutate(process = case_when(
  process == "ANTIVIRAL AND INTERFERON" ~ "ANTIVIRAL/IFN",
  process == "INNATE IMMUNE SYSTEM" ~ "INNATE",
  process == "HUMORAL ADAPTIVE IMMUNE SYSTEM" ~ "HUMORAL",
  process == "CELLULAR ADAPTIVE IMMUNE SYSTEM" ~ "CELLULAR",
  process == "ADAPTIVE IMMUNE SYSTEM" ~ "ADAPTIVE",
  process == "IMMUNOGLOBULIN MEDIATED IMMUNE RESPONSE" ~ "IMMUNOGLOBULIN-MEDIATED",
  TRUE ~ process)) %>% 
  mutate(in_out = 1) %>% 
  filter(!is.na(study)) %>% 
  mutate(process = factor(process, levels = c('INNATE', 
                                               'ADAPTIVE', 
                                               'CELLULAR', 
                                               'HUMORAL')),
         study = factor(study, levels = c('GSE199750', 
                                          'GSE201533',
                                          'GSE201530',
                                          'GSE189039',
                                          "GSE206023")))


######## ALL

ImmuneGO_Interesting_Genes_plot.higher <- ImmuneGO_Annotated_Genes_Interesting_studies %>%
  filter(process %in% c("ADAPTIVE", "INNATE", "HUMORAL", "CELLULAR")) %>%
  ggplot() +
  aes(x = study, y = genes, fill = study) +  # Reordenar os levels de study
  geom_tile() +
  scale_fill_manual(
    values = c("GSE199750" = "#80ffdb",
               "GSE201533" = "#5e60ce", 
               "GSE201530" = "#7400b8",
               "GSE206023" = "#4361ee",
               "GSE189039" = "#48bfe3")) +
  coord_flip() +
  theme_minimal() +
  facet_wrap(vars(process), scales = "free_x", ncol = 1) +
  labs(
    x = "GSE_ID",
    y = "#Genes",
    title = "#Genes represented in ImmuneGO, by study") + 
  theme(axis.text.x = element_text(angle = 90, size = 3, hjust = 1),
        axis.text.y = element_text(size = 20),
        strip.text = element_text(size= 20)) +
    scale_x_discrete(limits = rev(levels(ImmuneGO_Annotated_Genes_Interesting_studies$study)))

#Salvar
ggsave(ImmuneGO_Interesting_Genes_plot.higher, 
       file = "ImmuneGO_Interesting_Genes_Adaptive_Innate_plot_label.png", 
       width = 20, height = 20)


#Sem label
ImmuneGO_Interesting_Genes_plot.nolabel = ImmuneGO_Annotated_Genes_Interesting_studies %>%
  filter(process %in% c("ADAPTIVE", "INNATE", "HUMORAL", "CELLULAR")) %>% 
  ggplot() +
  aes(x = study, y = genes, fill = study) +  # Reordenar os levels de study
  geom_tile() +
  scale_fill_manual(
    values = c("GSE199750" = "#80ffdb",
               "GSE201533" = "#5e60ce", 
               "GSE201530" = "#7400b8",
               "GSE206023" = "#4361ee",
               "GSE189039" = "#48bfe3"))  +
  coord_flip() +
  theme_minimal() +
  facet_wrap(vars(process), scales = "free_x", ncol = 1)  +
  labs(
    x = "GSE_ID",
    y = "#Genes",
    title = "#Genes represented in ImmuneGO, by study") + 
  theme(axis.text.x = element_text(angle = 90, size = 0, hjust = 1),
        axis.text.y = element_text(size = 20),
        strip.text = element_text(size= 20)) +
  scale_x_discrete(limits = rev(levels(ImmuneGO_Annotated_Genes_Interesting_studies$study)))

#Salvar
ggsave(ImmuneGO_Interesting_Genes_plot.nolabel, 
       file = "ImmuneGO_Interesting_Genes_Adaptive_Innate_plot_nolabel.png", 
       width = 20, height = 20)

############### Grouped

ImmuneGO_Annotated_Genes_Nested_Interesting_grouped = ImmuneGO_Annotated_Genes_Nested_Interesting %>% 
  filter(process %in% c("INNATE IMMUNE SYSTEM", 
                        "ADAPTIVE IMMUNE SYSTEM", 
                        "CELLULAR ADAPTIVE IMMUNE SYSTEM", 
                        "HUMORAL ADAPTIVE IMMUNE SYSTEM"))

ImmuneGO_genes_indatasets.merge_grouped = genes_raw_combined %>% 
  group_by(study, genes) %>% 
  merge(ImmuneGO_Annotated_Genes_Nested_Interesting_grouped, by.x = "genes", all.x = F, all.y = T) %>%
  select(genes, study, process) %>% 
  distinct() %>% 
  group_by(study, process) %>%  
  select(-genes) %>% 
  summarise(n_genes_in = n()) %>% 
  merge(ImmuneGO_Annotated_Genes_Nested_Interesting_grouped, by = "process", all.y = F) %>% 
  select(study, process, set_size, n_genes_in) %>%
  distinct() %>% 
  filter(study != "NA")  %>% 
  summarise(study,
            process,
            set_size,
            n_genes_in,
            prop_in = (n_genes_in/set_size)*100,
            prop_in = round(prop_in, 1),
            study,
            process) %>% 
  mutate(process = case_when(
            process == "ANTIVIRAL AND INTERFERON" ~ "ANTIVIRAL/IFN",
            process == "INNATE IMMUNE SYSTEM" ~ "INNATE",
            process == "HUMORAL ADAPTIVE IMMUNE SYSTEM" ~ "HUMORAL",
            process == "CELLULAR ADAPTIVE IMMUNE SYSTEM" ~ "CELLULAR",
            process == "ADAPTIVE IMMUNE SYSTEM" ~ "ADAPTIVE",
            process == "IMMUNOGLOBULIN MEDIATED IMMUNE RESPONSE" ~ "IMMUNOGLOBULIN-MEDIATED",
            TRUE ~ process),
         process = factor(process, levels = c('INNATE', 
                                               'ADAPTIVE', 
                                               'CELLULAR', 
                                               'HUMORAL')),
         study = factor(study, levels = unique(study)))

#Plotar

ggImmuneGO_genes_indatasets.stats = ggplot(ImmuneGO_genes_indatasets.merge_grouped) +
  aes(x = reorder(study, prop_in), fill = study, weight = prop_in) +
  geom_bar() +
  scale_fill_manual(
    values = c("GSE199750" = "#80ffdb",
               "GSE201533" = "#5e60ce", 
               "GSE201530" = "#7400b8",
               "GSE206023" = "#4361ee",
               "GSE189039" = "#48bfe3")) +
  coord_flip() +
  theme_minimal() +
  geom_text(aes(label = prop_in, y = prop_in), 
            position = position_dodge(width = 0.9), 
            vjust = -0, 
            hjust = -0.2, 
            size = 3) +
  labs(
    x = "GSE_ID",
    y = "#Genes",
    title = "#Genes represented in ImmuneGO, by study"
  ) +
  facet_wrap(vars(process), ncol = 1) +
  theme(strip.text = element_text(size=8))
  
ggsave(ggImmuneGO_genes_indatasets.stats, file = "ImmuneGO_Genesrepresented_bystudy.png", width = 6, height = 10)

```



## MSigDB

Obtendo coleções

```{r}

############### Obter coleções ############### 
#Obter genes de um conjunto específico diretamente do MSigDB. 
msigdbr_collections() #Listar conjuntos

############### Hallmark ###############
hallmark <- msigdbr(species = "Homo sapiens", category = "H") %>% 
  dplyr::select(gs_name, gene_symbol) 

#Limpar texto
hallmark$gs_name <- gsub("HALLMARK_", "", hallmark$gs_name)
hallmark$gs_name <- gsub("_", " ", hallmark$gs_name)

#Conjuntos de GO
GO_BP <- msigdbr(species = "Homo sapiens", 
                 category = "C5", 
                 subcategory = "GO:BP") %>%
  select(gs_name, gene_symbol)

GO_MF <- msigdbr(species = "Homo sapiens", 
                 category = "C5", 
                 subcategory = "GO:MF") %>%
  select(gs_name, gene_symbol)

#Conjunto de marcadores de células 
cell_types <- msigdbr(species = "Homo sapiens", 
                      category = "C8") %>% 
  select(gs_name, gene_symbol)

#Conjunto de marcadores da resposta imune
immune_sig = msigdbr(species = "Homo sapiens", 
                     category = "C7", 
                     subcategory = "IMMUNESIGDB") %>% 
  select(gs_name, gene_symbol)

head(immune_sig)

#Conjunto de marcadores de vacinas
vax_sig = msigdbr(species = "Homo sapiens", 
                  category = "C7", 
                  subcategory = "VAX") %>% 
  select(gs_name, gene_symbol)

#Limpar texto
vax_sig$gs_name <- gsub("_", " ", vax_sig$gs_name)
head(vax_sig)
```

# VAX Sig DB

Os estudos foram filtrados pelo tamanho do conjunto de genes compartilhados (GSSize \> 10 ou \>50) e com pvalue (\<0.25).

Os dados foram anotados no R e depois avaliados iterativamente no excel.

```{r}
#Importar arquivo
VaxSigDB_genesets = VAX_GeneSets_Annotated_26_10_23

# Divida os dados em blocos de informações para cada vacina
vaccines_att <- split(VaxSigDB_genesets$Col2, VaxSigDB_genesets$Col1)

# Combine data
VAX_Genesets <- bind_rows(lapply(vaccines_att, function(x) data.frame(t(x))), .id = "V")

# Name a new index column
colnames(VAX_Genesets) <- paste0("V", seq_len(ncol(VAX_Genesets)))

# Transpose
rownames(VAX_Genesets) <- VAX_Genesets$V1
VAX_Genesets = VAX_Genesets %>% 
  t() %>%
  as.data.frame()

#Order columns
VAX_Genesets = VAX_Genesets %>% 
  select(STANDARD_NAME, everything()) %>% 
  filter(STANDARD_NAME != "STANDARD_NAME") 
rownames(VAX_Genesets) = VAX_Genesets$SYSTEMATIC_NAME
VAX_Genesets = VAX_Genesets %>% select(STANDARD_NAME, 
                                       DESCRIPTION_BRIEF,
                                       DESCRIPTION_FULL,
                                       PMID,
                                       AUTHORS,
                                       SYSTEMATIC_NAME,
                                       GENE_SYMBOLS,
                                       everything())

#Anotar novas colunas
VAX_Genesets <- VAX_Genesets %>%
  mutate(
    SAMPLE_SOURCE = ifelse(
      grepl("pbmc|peripheral blood mononuclear cell|peripheral blood mononuclear cells", DESCRIPTION_FULL, ignore.case = TRUE) |
      grepl("pbmc|peripheral blood mononuclear cell|peripheral blood mononuclear cells", DESCRIPTION_BRIEF, ignore.case = TRUE),
      "PBMC",
      ifelse(
        grepl("whole blood", DESCRIPTION_FULL, ignore.case = TRUE) |
        grepl("whole blood", DESCRIPTION_BRIEF, ignore.case = TRUE) |
        grepl("_BLOOD", STANDARD_NAME, ignore.case = TRUE)  ,
        "WHOLE BLOOD",
        "OTHER")),
    REGULATION = ifelse(
      grepl("up", DESCRIPTION_BRIEF, ignore.case = TRUE),
      "UP",
      ifelse(
        grepl("down", DESCRIPTION_BRIEF, ignore.case = TRUE),
        "DOWN",
        "UNKNOWN")))

#Reordenar colunas
VAX_Genesets = VAX_Genesets %>% select(SAMPLE_SOURCE, REGULATION, everything())

# Salvar
write.csv(VAX_Genesets, file = "VAXSigDB_Annotated.csv") #Continuar a anotação no Google Sheets

head(VAX_Genesets)
```

Converter tabela para tabela de genes

```{r}
############## Long table with Gene symbols

VAX_Genesets = VAXSigDB_Annotated

# Separar genes por linhas e somar numero de genes
VAX_Genes <- VAX_Genesets %>%
  mutate(GENE = GENE_SYMBOLS) %>% 
  separate_rows(GENE, sep = ",") %>% 
  select(GENE, everything()) %>%
  group_by(STANDARD_NAME) %>% #Somar e agrupar
  mutate(SetSize = n()) %>% 
  relocate(SetSize, .after = GENE_SYMBOLS)

VAX_GeneSets_annotated <- VAX_Genes %>%
  select(-GENE) %>% #Retirar coluna
  distinct() #Retirar linhas repetidas

#Salvar
write.csv(VAX_Genes, file = "VAX_Genes_Annotated_RAW.csv")
write.csv(VAX_GeneSets_annotated, file = "VAX_Genesets_Annotated_RAW.csv")

#Display
head(VAX_GeneSets_annotated)
```

Estatísticas

```{r}
########## Descriptive statistics
vaccines = VAX_GeneSets_annotated %>% 
  group_by(VACCINE, PLATFORM, `TARGET_PATHOGEN_DISEASE`, MICROBE_TYPE) %>% 
  summarise(Count = n()) %>% 
  arrange(desc(Count))

pathogens = VAX_GeneSets_annotated %>% 
  group_by(`TARGET_PATHOGEN_DISEASE`, MICROBE_TYPE) %>% 
  summarise(Count = n()) %>% 
  arrange(desc(Count))

platforms = VAX_GeneSets_annotated %>% 
  group_by(PLATFORM) %>% 
  summarise(Count = n()) %>% 
  arrange(desc(Count))

```

### Visualização

By vaccines

```{r}
vaccine_hist = ggplot(vaccines) +
  aes(x = reorder(VACCINE, Count),
      y = Count,
      fill = `MICROBE_TYPE`) + 
  ggtitle("VAX MSigDB - Vaccines, by platform and target") +
  geom_col() +
  geom_text(aes(label = Count), vjust = 0, hjust = 2, size = 3) +
  coord_flip() +
  theme_minimal() + 
  labs(fill = "Vaccine target type") +
  theme(plot.title = element_text(face = "bold", colour = "black", size = (20)),
  legend.title = element_text(face = "bold"),
  axis.title = element_text(size = (10), colour = "black"),
  axis.text = element_text(colour = "black", size = (10))) +
  facet_wrap(vars(PLATFORM), scales = "free") + 
  scale_fill_manual(values = c('VIRUS'= "#d00000",
                               'BACTERIUM' = "#16E0BD",
                               'PARASITE' = "#118ab2",
                              'MENINGOCOCCUS'	= "#16E0BD",
                             'PNEUMOCOCCUS'	= "#44C199",
                             'TUBERCULOSIS'	= "#99e2b4",
                             'F TULARENSIS'	= "#90e0ef",
                             'LEISHMANIA'	= "#118ab2",
                             'MALARIA' = "#3f37c9",
                             'EBOV'	= "#d00000",
                             'HBV'	= "#e85d04",
                             'HBV, HAV'	= "#f48c06",
                             'HIV'	= "#a4133c",
                             'HPV'	= "#f27059",
                             'INFLUENZA'	= "#ffa69e",
                             'MEASLES'	= "#c65b7c",
                             'MMR'	= "#f9b5ac",
                             'SMALLPOX'	= "#ff8c42",
                             'VEEV'	= "#e01e37",
                             'VZV'	= "#ff9b85",
                             'YF'	= "#f38375"))


ggsave(vaccine_hist, file = "VAXSigDB_Vaccines_platforms_pathogens_1.png", width = 25, height = 8)
```

By platforms

```{r}
platforms_hist = ggplot(platforms) +
  aes(x = reorder(`PLATFORM`, Count),
      y = Count,
      fill = `PLATFORM`) + 
  ggtitle("VAX MSigDB - Vaccine platforms") +
  geom_col() +
  geom_text(aes(label = Count), vjust = 0, hjust = 0, size = 3) +
  coord_flip() +
  theme_minimal() + 
  labs(fill = "Platform") +
  theme(plot.title = element_text(face = "bold", colour = "black", size = (20)),
  legend.title = element_text(face = "bold"),
  axis.title = element_text(size = (10), colour = "black"),
  axis.text = element_text(colour = "black", size = (10))) +
  #facet_wrap(vars(PLATFORM), scales = "free") + 
  scale_fill_manual(values = c('IN'= "#1e6091",
                               'LA' = "#16E0BD",
                               'VV' = "#34a0a4",
                               'CONJ' = "#76c893",
                               'VLP' = "#b5e48c", 
                               'SU' = "#2ec4b6",
                               'LA and IV' = "#cbf3f0",
                               'SUB' = "#00bbf9",
                               'PEP' = "#c0fdff"))

ggsave(platforms_hist, file = "VAXSigDB_Platforms_hist_1.png", width = 10, height = 5)
```

By pathogens

```{r}

pathogens_hist = ggplot(pathogens) +
  aes(x = reorder(`TARGET_PATHOGEN_DISEASE`, Count),
      y = Count,
      fill = `MICROBE_TYPE`) + 
  ggtitle("VAX MSigDB - Vaccines, by platform and target") +
  geom_col() +
  geom_text(aes(label = Count), vjust = 0, hjust = 0, size = 3) +
  coord_flip() +
  theme_minimal() + 
  labs(fill = "Vaccine target type") +
  theme(plot.title = element_text(face = "bold", colour = "black", size = (20)),
  legend.title = element_text(face = "bold"),
  axis.title = element_text(size = (10), colour = "black"),
  axis.text = element_text(colour = "black", size = (10))) +
  #facet_wrap(vars(PLATFORM), scales = "free") + 
  scale_fill_manual(values = c('VIRUS'= "#d00000",
                               'BACTERIUM' = "#16E0BD",
                               'PARASITE' = "#118ab2",
                              'MENINGOCOCCUS'	= "#16E0BD",
                             'PNEUMOCOCCUS'	= "#44C199",
                             'TUBERCULOSIS'	= "#99e2b4",
                             'F TULARENSIS'	= "#90e0ef",
                             'LEISHMANIA'	= "#118ab2",
                             'MALARIA' = "#3f37c9",
                             'EBOV'	= "#d00000",
                             'HBV'	= "#e85d04",
                             'HBV, HAV'	= "#f48c06",
                             'HIV'	= "#a4133c",
                             'HPV'	= "#f27059",
                             'INFLUENZA'	= "#ffa69e",
                             'MEASLES'	= "#c65b7c",
                             'MMR'	= "#f9b5ac",
                             'SMALLPOX'	= "#ff8c42",
                             'VEEV'	= "#e01e37",
                             'VZV'	= "#ff9b85",
                             'YF'	= "#f38375"))

ggsave(pathogens_hist, file = "VAXSigDB_Pathogens_hist_1.png", width = 10, height = 5)

```

### Filtragem

```{r}
unique(VAX_GeneSets_annotated$STUDY_TYPE) 
#"CORRELATION"
#"VACCINE"
#"SUBSET"
#"CHALLENGE"

unique(VAX_GeneSets_annotated$STUDY_SUBTYPE)
# "ANTIBODY"				
# "VAC ONLY"				
# "BOOSTER"				
# "VAC VS CTRL"				
# "SIGNATURE"				
# "ADVERSE EVENT"			
# "COMPARISON"				
# "RESPONDER VS NON RESPONDER"				
# "AGE"				
# "RESPONDER"				
# "ADJUVANT"				
# "NON RESPONDER"				
# "ANTIMICROBIAL ACTIVITY"				
# "CELLS"				
# "CHALLENGE"				
# "TOP DEGS"				
# "PROTECTION"				
# "TRAINING SET"				
# "EXON ANALYSIS"				
# "DOSE"				
# "VAC VS OTHER"


# Subset de vacinas para as seguintes análises comparativas
VAX_GeneSets_Blood_PBMC_VAC = VAX_GeneSets_annotated %>% 
  filter(SAMPLE_SOURCE %in% c("PBMC", "WHOLE BLOOD"), 
         STUDY_TYPE %in% c("VACCINE"))

# Subset de vacinas com resultados de comparações entre vacinas, adjuvantes, idade e sexo
VAX_GeneSets_Blood_PBMC_CORR = VAX_GeneSets_annotated %>% 
  filter(SAMPLE_SOURCE %in% c("PBMC", "WHOLE BLOOD"), 
         STUDY_TYPE %in% c("CORRELATION"))

# Subset com resultados de correlações com resposta inata, celular e humoral
VAX_GeneSets_Blood_PBMC_SUBSETS = VAX_GeneSets_annotated %>% 
  filter(SAMPLE_SOURCE %in% c("PBMC", "WHOLE BLOOD"), 
         STUDY_TYPE %in% c("SUBSET"))


# Subset de vacinas com genes associados a eventos adversos
VAX_GeneSets_Blood_PBMC_CHALLENGE = VAX_GeneSets_annotated %>% 
  filter(SAMPLE_SOURCE %in% c("PBMC", "WHOLE BLOOD"), 
         STUDY_TYPE %in% c("CHALLENGE"))

VAX_GeneSets_Blood_PBMC_VAC.stats = VAX_GeneSets_Blood_PBMC_VAC %>% 
  group_by(VACCINE, PLATFORM, REGULATION, `TARGET_PATHOGEN_DISEASE`, MICROBE_TYPE) %>% 
  summarise(Count = n()) %>% 
  arrange(desc(Count))

```


*Qual a relação entre os genes de cada dataset?*

```{r}
############### VENN DIAGRAM
#Input
data = VAX_GeneSets_Blood_PBMC_VAC %>%
  as.tibble() %>% 
  separate_rows(GENE_SYMBOLS, sep = ",") %>% 
  select(VACCINE, GENE_SYMBOLS, REGULATION) %>% 
  as.data.frame() %>% 
  mutate(VACCINE_REG = paste0(VACCINE, sep="_", REGULATION))
  
```


```{r}
# Função para calcular a sobreposição entre pares de vacinas e a porcentagem de genes compartilhados
overlap_genes <- function(cond1, cond2, data) {
  genes_cond1 <- data$GENE_SYMBOLS[data$VACCINE_REG == cond1] #Trocar coluna
  genes_cond2 <- data$GENE_SYMBOLS[data$VACCINE_REG == cond2] #Trocar coluna
  genes_shared <- intersect(genes_cond1, genes_cond2)
  genes_notshared_cond1 <- setdiff(genes_cond1, genes_cond2)
  genes_notshared_cond2 <- setdiff(genes_cond2, genes_cond1)
  
  total_genes_cond1 <- length(genes_cond1)
  total_genes_cond2 <- length(genes_cond2)
  
  percentage_shared_cond1 <- length(genes_shared) / total_genes_cond1 * 100
  percentage_shared_cond2 <- length(genes_shared) / total_genes_cond2 * 100
  
  shared_genes <- data.frame(
    Cond1 = cond1,
    Cond2 = cond2,
    Shared = length(genes_shared),
    NotShared_cond1 = length(genes_notshared_cond1),
    NotShared_cond2 = length(genes_notshared_cond2),
    Total_Genes_Cond1 = total_genes_cond1,
    Total_Genes_Cond2 = total_genes_cond2,
    Genes_Names = paste(genes_shared, collapse = ", "),
    Percentage_Shared_Cond1 = percentage_shared_cond1,
    Percentage_Shared_Cond2 = percentage_shared_cond2
  )
  
  return(shared_genes)
}
# Obtenha uma lista de todas as vacinas únicas
unique_cond <- unique(data$VACCINE_REG)  #trocar "study" pelo nome da coluna de interesse

# Inicialize um dataframe para armazenar os resultados
shared_genes_df <- data.frame()

# Calcular a sobreposição e porcentagem para cada par de vacinas
for (i in 1:(length(unique_cond) - 1)) {
  for (j in (i + 1):length(unique_cond)) {
    resultado_temp <- overlap_genes(unique_cond[i], unique_cond[j], data)
    shared_genes_df <- bind_rows(shared_genes_df, resultado_temp)
  }
}

# Função para realizar o teste exato de Fisher
fisher_exact_test <- function(shared, notshared1, notshared2) {
  cont_table <- matrix(c(shared, notshared1, notshared2, 0), nrow = 2)
  results_fisher <- fisher.test(cont_table)
  return(results_fisher$p.value)
}

# Aplicar a função a cada linha da tabela de resultados
shared_genes_df$pvalue <- mapply(fisher_exact_test, 
                                    shared_genes_df$Shared, 
                                    shared_genes_df$NotShared_cond1, 
                                    shared_genes_df$NotShared_cond2)


# Calcular -log(p)
shared_genes_df$`-log(p)`= -log(shared_genes_df$pvalue, 10)

#Duplicar e trocar colunas (Vacina 1 e Vacina 2)
shared_genes_df2 = shared_genes_df
shared_genes_df2[, c("Cond1", "Cond2")] <- shared_genes_df2[, c("Cond2", "Cond1")]

#Unir datasets com os dados espelhados
shared_genes_df_mirror = rbind(shared_genes_df, shared_genes_df2)

#Converter NA, NaNe inf em 0
shared_genes_df_mirror[is.na(shared_genes_df_mirror)] <- 0
shared_genes_df_mirror<- replace(shared_genes_df_mirror, shared_genes_df_mirror == "Inf", 0)

#Arredondar porcentagens
shared_genes_df_mirror = shared_genes_df_mirror %>%
  mutate(Percentage_Shared_Cond1 = round(Percentage_Shared_Cond1, 2),
         Percentage_Shared_Cond2 = round(Percentage_Shared_Cond2, 2))

#Salvar resultados
write.csv(shared_genes_df_mirror, file = "VAXSigDB_sharedgenes.csv") #Alterar

```

Heatmap
```{r}

matrix = shared_genes_df_mirror %>% 
  filter(pvalue <=0.10) %>% 
  select(Cond1, Cond2, Percentage_Shared_Cond1) %>% #Percentage_Shared_Cond1
  filter(grepl("DOWN", Cond1) & grepl("DOWN",Cond2)) %>% 
  pivot_wider(names_from = "Cond2", 
              values_from = "Percentage_Shared_Cond1",  #Percentage_Shared_Cond1
              values_fn = mean) %>% 
  replace(is.na(.), 0) %>% 
  column_to_rownames(var="Cond1") %>% 
  as.matrix() %>% 
  round(0)


col_fun <- colorRamp2(c(0, 100), c("white", "red"))

#Plotar

file = "VAXSig_SharedGenes_perc_DOWN-DOWN"
fileimageheatmap_grouped= paste0(file, sep ="_", "Complexheatmap.png")

png(file = fileimageheatmap_grouped, width=50, height=50,units="cm",res=800)

heatmap_plot = Heatmap(matrix,
                        top_annotation = NULL,
                        left_annotation = NULL,
                        cell_fun = function(j, i, x, y, width, height, fill) {
        grid.text(sprintf("%.1f", matrix[i, j]), x, y, gp = gpar(fontsize = 5))}, #Mostrar valores nas células
                        show_row_names = TRUE,
                        row_names_side = "left",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        col = col_fun, #color
                        # column_title = "Target", 
                        # row_title = "Source",
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 12),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = gpar(fontsize = 12), #Grande = 6, Pequeno 15
                        row_dend_width = unit(5, "cm"), #Altura do dendrograma
                        row_split = NULL, #Split row clusters (NULL se for um gene set mais específico)
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(8, "mm"),
                        show_column_dend = FALSE,
                        show_row_dend = FALSE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = unit(40, "cm"), # 20 para pequeno e grande
                        height = unit(40, "cm") # 20 para pequeno, 60 para grande
)

draw(heatmap_plot, 
     annotation_legend_side = "left", 
     heatmap_legend_side = "right")
dev.off()

```


# Representação de genes por gene set para cada condição 

## Gráfico de linhas

```{r}

Immune_GO_GSEA

################## Innate
Immune_GO_innate = Immune_GO_GSEA %>% filter(`Immune system` == "Innate")
write.csv(Immune_GO_innate, file = "Immune_GO_innate_l2fc.csv")

Immune_GO_innate_SUM <- Immune_GO_innate %>% 
  rename(Condition = study) %>% 
  group_by(Condition, Process) %>%
  mutate(Soma = n()) %>% select(Vaccine, Condition, Process, `Immune group`, `Immune response gene set, specific`, `Immune system`, Soma) %>% filter(Vaccine != "NA")
write.csv(Immune_GO_innate_SUM, file="Immune_GO_innate_SUM.csv")


################## Adaptive
#Humoral
Immune_GO_humoral = Immune_allgenes %>% filter(`Immune system` == "Humoral")
write.csv(Immune_GO_humoral, file = "Immune_GO_humoral_l2fc.csv")

Immune_GO_humoral_SUM <- Immune_GO_humoral %>% 
  rename(Condition = study) %>% 
  group_by(Condition, Process) %>%
  mutate(Soma = n()) %>% select(Vaccine, Condition, Process, `Immune group`, `Immune response gene set, specific`, `Immune system`, Soma) %>% filter(Vaccine != "NA")
write.csv(Immune_GO_humoral_SUM, file="Immune_GO_humoral_SUM.csv")

#Cellular
Immune_GO_cellular = Immune_allgenes %>% filter(`Immune system` == "Cellular")
write.csv(Immune_GO_cellular, file = "Immune_GO_cellular_l2fc.csv")

Immune_GO_cellular_SUM <- Immune_GO_cellular %>% 
  rename(Condition = study) %>% 
  group_by(Condition, Process) %>%
  mutate(Soma = n()) %>% select(Vaccine, Condition, Process, `Immune group`, `Immune response gene set, specific`, `Immune system`, Soma) %>% filter(Vaccine != "NA")
write.csv(Immune_GO_cellular_SUM, file="Immune_GO_cellular_SUM.csv")

################## General
Immune_GO_general = Immune_allgenes %>% filter(`Immune system` == "General")
write.csv(Immune_GO_general, file = "Immune_GO_general_l2fc.csv")

Immune_GO_general_SUM <- Immune_GO_general %>% 
  rename(Condition = study) %>% 
  group_by(Condition, Process) %>%
  mutate(Soma = n()) %>% select(Vaccine, Condition, Process, `Immune group`, `Immune response gene set, specific`, `Immune system`, Soma) %>% filter(Vaccine != "NA")
write.csv(Immune_GO_general_SUM, file = "Immune_GO_general_SUM.csv")

################## All immune processes and genes
Immune_allgenes_vaccines = Immune_allgenes_l2fc %>% rename(Condition = study, Genes = genes) %>% filter(Vaccine != "NA")
ann_vaccines_pca_matrix = merge(ann_vaccines_pca, matrix_data_pca, by.x = "Condition", all.x = F, all.y = F)
ann_vaccines_pca_matrix_long = ann_vaccines_pca_matrix %>% gather(Genes, L2FC, -Condition, -Vaccine, -Day, -Dose)

Immune_allgenes_vaccines_SUM <- Immune_allgenes_vaccines %>%
  group_by(Condition, Process) %>% #Somar e agrupar
  summarize(Soma = n())

Immune_allgenes_vaccines_SUM = merge(ann_vaccines_pca, Immune_allgenes_vaccines_SUM, by.x = "Condition", all.x = F, all.y = F) # Anotar vacinas
write.csv(Immune_allgenes_vaccines_SUM, file="Immune_allgenes_vaccines_SUM.csv") #Salvar

```

## Shifted bars

```{r}
ImmuneGO_GSEA_General_Innate_Adaptive = Immune_GO_GSEA %>% 
  filter(`Gene set short` %in% c("CELLULAR ADAPTIVE IMMUNE SYSTEM","HUMORAL ADAPTIVE IMMUNE SYSTEM")) %>% 
  mutate(`-log(qvalue)` = as.numeric(`-log(qvalue)`))

barchart_NES

ggplot(ImmuneGO_GSEA_General_Innate_Adaptive) +
 aes(x = Vaccine, fill = `Gene set short`, weight = NES) +
 geom_bar() +
 scale_fill_manual(values = c(`ADAPTIVE IMMUNE RESPONSE` = "#6DF8DF", `INNATE IMMUNE RESPONSE` = "#989898"
 )) +
 labs(title = "Immune system ", subtitle = "Adaptive and Innate, by NES") +
 coord_flip() +
 theme_light()


ggsave(barchart_NES, file= "barchart_NES.png", width = 10, height = 5)

```

```{r}
Immune_GO_general_SUM_Adaptive_Innate = Immune_GO_general_SUM %>% 
  filter(Process == "adaptive immune response" | Process == "innate immune response") %>% 
  arrange(desc(Process)) %>% 
  distinct()

view(Immune_GO_general_SUM_Adaptive_Innate)

write.csv(Immune_GO_general_SUM_Adaptive_Innate, file = "Immune_GO_general_SUM_Adaptive_Innate.csv")


barchart_SUM = Immune_GO_general_SUM_Adaptive_Innate %>%
 filter(Process %in% c("adaptive immune response", "innate immune response")) %>%
 ggplot() +
 aes(x = Condition, fill = Process, weight = Soma) +
 labs(title = "Immune system ", subtitle = "Adaptive and Innate") +
 geom_bar(position = "fill") +
  scale_fill_manual(values = c(`adaptive immune response` = "#6DF8DF", `innate immune response` = "#989898"
 ))+
 coord_flip() +
 theme_minimal() +
 theme(legend.position = "top") +
 facet_wrap(vars(Vaccine), scales = "free")

ggsave(barchart_SUM, file= "barchart_SUM.png", width = 10, height = 5)

```

###### Gráfico de linhas

```{r}
lines = Immune_GO_general_SUM_Adaptive_Innate %>% 
  merge(ann_vaccines, by = "Condition", all = T) %>% 
  group_by(Condition, Process) %>% 
  mutate(Day = as.numeric(Day))


esquisser(lines)

lines_plot = lines %>%
 filter(!is.na(Vaccine.x)) %>%
 filter(!is.na(Process)) %>%
 filter(!is.na(`Immune group`)) %>%
 ggplot() +
  aes(
    x = Day,
    y = Soma,
    colour = Vaccine.y,
    shape = Vaccine.x
  ) +
  geom_point(size = 4L) +
  geom_line()+
  scale_color_hue(direction = 1) +
  theme_linedraw() +
  facet_grid(
    vars(`Immune response gene set, specific`),
    vars(Dose),
    scales = "free"
  )

ggsave(lines_plot, file = "ImmuneGO_Adap_Innate_lines_plot.png", width = 10, height = 7)

```

```{r}
barchart_stack = Immune_GO_general_SUM_Adaptive_Innate %>%
 filter(Process %in% c("adaptive immune response", "innate immune response")) %>%
 ggplot() +
 aes(x = Condition, fill = Process, weight = Soma) +
 labs(title = "Immune system ", subtitle = "Adaptive and Innate") +
 geom_bar(position = "stack") +
  scale_fill_manual(values = c(`adaptive immune response` = "#6DF8DF", `innate immune response` = "#989898"
 ))+
 coord_flip() +
 theme_minimal() +
 theme(legend.position = "top") +
 facet_wrap(vars(Vaccine), scales = "free")

```

# Heatmap Cell Marker

```{r}
cellmarker_studies
cellmarker_heatmap = cellmarker_studies %>% filter(qvalue <= 0.10)
cellmarker_heatmap.wide = dcast(cellmarker_heatmap, `ID` ~ `study`, value.var = "NES", fun.aggregate = mean)
cellmarker_heatmap.wide[is.na(cellmarker_heatmap.wide)] = 0
cellmarker_heatmap.wide.ID = cellmarker_heatmap.wide %>% 
  column_to_rownames(var="ID") %>% 
  as.matrix()

#Column annotation
ann_cols = ann_vaccines 
ann_cols$Vaccines = rownames(ann_cols)

#Anotar colunas
cellmarker_heatmap.wide.ID.ANN = cellmarker_heatmap.wide.ID %>% 
  as.data.frame() %>% 
  rownames_to_column("Condition")

ann_vaccines_pheatmap = ann_vaccines %>% 
  merge(shared_heatmap.ann, by="Condition", all.x = F) %>% 
  column_to_rownames("Condition") %>% 
  select(Vaccine, Day, Dose, Type, Regimen) %>% 
  mutate(Day = as.factor(Day))

#Row annotation
ann_rows_cellmarker = Immune_cells %>% 
  select(cell_name, Type)


ann_vaccines_pheatmap


col_fun = colorRampPalette(c("blue", "white", "red"))(100)


#Visualizar
heatmap = pheatmap(
  cellmarker_heatmap.wide.ID,
  annotation_row = NA, 
  annotation_col = ann_vaccines_pheatmap,
  annotation_colors = ann_colors,
  annotation_legend = TRUE,
  clustering_method = "complete",  # Método de clustering, ajuste conforme desejado
  display_numbers = FALSE,  # Mostrar os valores nos quadrados do heatmap
  show_colnames = TRUE,  # Mostrar nomes das colunas
  cluster_cols = TRUE,
  show_rownames = TRUE,  # Mostrar nomes das linhas
  scale = "column",  # Escala - "column" para escala por coluna
  main = "CellMarker scaled data",
  border_color = "white",  # Define a cor da borda, 'NA' para omitir as bordas
  color = col_fun,  # Define a paleta de cores
  breaks = seq(-2, 2),  # Define os intervalos da paleta de cores
  fontsize_row = 5,
  fontsize_col = 5,
  cutree_rows = 5, 
  cutree_cols = 3
)

png(file= "CellMarker_pheatmap.png", width=2000, height=2000, res=315, pointsize=10)
print(heatmap)  # Substitua 'p' pelo objeto 'pheatmap'
dev.off()  # Finalize a gravação do arquivo
```