

---
title: "Gene set enrichment analysis"
author: "Wasim"
date: "2023-12-31"
output: html_document
---

## 1.2. Bibliotecas

```{r eval=FALSE, include=FALSE}
library(readr)
library(explore)
library(maditr)
library(GEOquery)
library(Matrix)
library(circlize)
library(RColorBrewer)
library(celldex)
library(biomaRt)
library(org.Hs.eg.db)
library(DESeq2)
library(msigdbr)
library(ape)
library(GSVA)
library(sva)
library(clusterProfiler)
library(pheatmap)
library(EnhancedVolcano)
library(metaRNASeq)
library(ggbeeswarm)
library(tidyverse)
library(corrr)
library(ggcorrplot)
library(FactoMineR)
library(factoextra)
library(esquisse)
library(corto)
library(factoextra)
library(reshape2)
library(ComplexHeatmap)
library(janitor)
library(stats)

```



# ORA e GSEA

O ORA e GSEA podem ser pouco sensíveis quando o pvalue/padj \< 5%. Na documentação, recomendam usar o cutoff de padj (FDR) de 25% para incluir gene sets que estejam 75% corretos - isto é, em 4 observações, 3 estão corretas -, o que aumenta o número de resultados interessantes. Para verificar se são realmente significantes, visualize no heatmap com labels de \*significancia.
Referencia: Link{<https://www.gsea-msigdb.org/gsea/doc/GSEAUserGuideFrame.html>
https://yulab-smu.top/biomedical-knowledge-mining-book/universal-api.html 

### Input

Insira a tabela longa com genes filtrados para up- ou down-regulated gerados na etapa de preparação dos dados.


```{r}
######### GSE199750
gse199750_degs_table_notfiltered$gse_id = "GSE199750"
gse199750_degs_table_updown = filter(gse199750_degs_table_notfiltered) %>% 
  select(-`...1`) %>% 
  select(gse_id, study, genes, everything()) %>% 
  mutate(Direction = case_when(
    log2FoldChange > 1 ~ "UP", 
    log2FoldChange < -1 ~ "DOWN", 
    TRUE ~ "NEUTRAL")) 

gse199750_degs_table_up = gse199750_degs_table %>% 
filter(log2FoldChange >= 1, padj <0.05)
gse199750_degs_table_down = gse199750_degs_table %>% 
  filter(log2FoldChange < -1,  padj <0.05)


############# GSE206023

gse206023_counts = GSE206023_tpm_of_48_samples %>% 
  rename(genes = Name)  

#Selecionar as versoes mais recentes de cada gene que for anotado com HGNC symbol derivado do GIDCON
Genes_ann_GSE206023 = Genbank_ID_to_HGNC_symbol %>% 
  clean_names() %>% 
  filter(grepl("\\.", gene_original)) %>% 
  arrange(desc(gene_original)) %>% 
  select(gene_original, gene_simplified, symbol) %>% 
  distinct() %>% 
  arrange(desc(gene_original)) %>% 
  separate(gene_original, c("genbank_ac", "version"), sep = "\\.", remove=F) %>% 
  group_by(gene_original) %>% 
  filter(!(gene_original == symbol)) %>% 
  group_by(symbol) %>% 
  select(symbol, everything()) %>% 
  arrange(desc(version), .by_group = T) %>% 
  filter(version == max(version, na.rm = TRUE)) %>% 
  select(gene_original, genbank_ac, version)

gse206023_degs_table_updown = gse206023_degs_table_notfiltered_annotated %>% 
  select(-`...1`) %>% 
  rename(study = Study, genes = Symbol) %>% 
  mutate(Direction = case_when(
    log2FoldChange > 1 ~ "UP", 
    log2FoldChange < -1 ~ "DOWN", 
    TRUE ~ "NEUTRAL")) %>% 
  merge(Genes_ann_GSE206023, by="Gene_original", all = T) %>% #Unir e selecionar somente genes com HGNC symbol e com versão mais recente
  select(gse_id, study, genes, baseMean, log2FoldChange, lfcSE, stat, pvalue, padj)

#Salvar
write.csv(gse206023_degs_table_updown, file="gse206023_degs_table_updown_recentGenbank.csv")

# gse206023_degs_table_up = filter(gse206023_degs_table, log2FoldChange > 1, pvalue < 0.05)
# gse206023_degs_table_down = filter(gse206023_degs_table, log2FoldChange < -1, pvalue <0.05)
```


```{r}
############# GSE201533
gse201533_degs_table_notfiltered$gse_id = "GSE201533"
gse201533_degs_table_updown = gse201533_degs_table_notfiltered %>% 
  select(-`...1`) %>%
  select(gse_id, study, genes, everything()) %>% 
  mutate(Direction = case_when(
    log2FoldChange > 1 ~ "UP", 
    log2FoldChange < -1 ~ "DOWN", 
    TRUE ~ "NEUTRAL"))

# gse201533_degs_table_up = filter(gse201533_degs_table, log2FoldChange > 1, padj <0.05)
# gse201533_degs_table_down = filter(gse201533_degs_table, log2FoldChange < -1, padj <0.05)


############# Unir tabelas de UP e DOWN
degs_allstudies_updown_notfiltered = rbind(gse199750_degs_table_updown, 
                               gse206023_degs_table_updown, 
                               gse201533_degs_table_updown)
degs_allstudies_updown_notfiltered = degs_allstudies_updown_notfiltered %>% 
  filter(!str_detect(genes, "__")) #Filtrar genes nao alinhados e com erro

#Criar coluna de Vaccine
degs_allstudies_updown <- degs_allstudies_updown_notfiltered %>%
  mutate(Vaccine = ifelse(grepl("BBIBP", study, ignore.case = TRUE), "BBIBP",
             ifelse(grepl("ZF2001", study, ignore.case = TRUE), "ZF2001",
             ifelse(grepl("ChAd", study, ignore.case = TRUE) | grepl("ChAd-BNT", study, ignore.case = TRUE), "ChAd",
             ifelse(grepl("BNT-MO", study, ignore.case = TRUE), "BNT",
             "BNT"))))) %>% 
  select(Vaccine, everything()) %>% 
  merge(ImmuneGO_Annotated_Genes, by.x = "genes", by.y = "Genes", all.x=T, all.y=F)

write.csv(degs_allstudies_updown, file="degs_allstudies_updown_notfiltered_Annotated_ImmuneGO.csv")


#Salvar tabela UP
degs_allstudies_up = degs_allstudies_updown %>% 
  filter(Direction == "UP")

write.csv(degs_allstudies_up, file="degs_allstudies_up_notfiltered.csv")

#Salvar tabela DOWN
degs_allstudies_down = degs_allstudies_updown %>% 
  filter(Direction == "DOWN") 

write.csv(degs_allstudies_down, file="degs_allstudies_down_notfiltered.csv")

############# Filter by padj and pvalue

degs_allstudies_updown_p05 = degs_allstudies_updown %>% filter(
  (gse_id %in% c("GSE199750", "GSE201533") & padj <= 0.05) |
    (gse_id == "GSE206023" & pvalue <= 0.05))

write.csv(degs_allstudies_updown_p05, file = "degs_allstudies_updown_p<05.csv")


############ Merge
degs_allstudies_updown = degs_allstudies_updown %>% 
  mutate(comparison = "not filtered")

degs_allstudies_updown_p05 = degs_allstudies_updown_p05 %>% 
  mutate(comparison = "p <= 0.05")

degs_allstudies_updown_filterspvalue = rbind(degs_allstudies_updown,
                                             degs_allstudies_updown_p05)

write.csv(degs_allstudies_updown_filterspvalue, file = "degs_allstudies_updown_filters_Annotated_ImmuneGO.csv")

############# STATISTICS

a = degs_allstudies_updown %>% 
  group_by(gse_id) %>% 
  summarise(ngenes = n()) %>% 
  mutate(comparison = "not filtered")

b = degs_allstudies_updown_p05 %>% 
  group_by(gse_id) %>% 
  summarise(ngenes = n()) %>% 
  mutate(comparison = "p <= 0.05")

plot_diff = rbind(a, b)

# Reordenar fill de acordo com o número de genes
plot_diff.05$comparison <- fct_reorder(plot_diff.05$comparison, plot_diff.05$ngenes, sum)


GSEA_ALL <- function(df, Immune_GO_T2G, CellMarker_genes) {
  resultados <- list()

  condicoes <- df$condition %>% unique() %>% as.character()

  for (condicao in condicoes) {
    # Selecionar DEGs para a condição atual
    
# Plotar o gráfico
plot_diff_bar = ggplot(plot_diff.05) +
  aes(x = reorder(gse_id, ngenes), fill = comparison, weight = ngenes) +
  geom_bar(position = "dodge") +
  scale_fill_manual(values = c(`not filtered` = "#7F7F7F", 
                               `p <= 0.05` = "#4BC2FF", 
                               `p <= 0.10` = "#F58E8E")) +
  labs(title = "Number of genes filtered by pvalue/padj", fill = "Datasets") +
  #coord_flip() +
  theme_minimal() +
  theme(
    plot.title = element_text(size = 19L, face = "bold", hjust = 0.5),
    axis.title.x = element_text(size = 15L, face = "bold")) +
  geom_text(aes(label = ngenes, y = ngenes), 
            position = position_dodge(width = 0.9), 
            vjust = -0.2, 
            hjust = -0.2, 
            size = 3) +
  coord_flip()

ggsave(plot_diff_bar, file = "Datasetsize_filteringbypvalue.png", width=10, height = 5)

print(plot_diff_bar)

```

### Visualização

```{r}
#All DEGs, not Immune filtered
plot_degs = degs_allstudies_updown_filter %>% 
  filter(filter == "p<0.05") %>% 
  group_by(gse_id, study, Direction, filter) %>% 
  summarise(n_regulated = n()) %>%
  merge(ann_vaccines, by.x = "study", by.y = "Condition")

ggplot_degs = plot_degs %>%
 filter(!(Direction %in% "NEUTRAL")) %>%
 ggplot() +
  aes(x = study, fill = Direction, weight = n_regulated) +
  geom_bar(position = "dodge") +
  scale_fill_manual(
    values = c(DOWN = "#545454",
    NEUTRAL = "#00C19F",
    UP = "#4BC2FF")
  ) +
  theme_void() +
  theme(legend.position = "top",
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1, size = 7),
        axis.text.y = element_text(vjust = 2)) +
  geom_text(aes(label = n_regulated, y = n_regulated), 
            position = position_dodge(width = 0.9), 
            vjust = -0.5, 
            hjust = 0.5, 
            size = 2) +
  ggtitle("DEGs - Up and down regulated, by vaccine, pvalue <0.05") +
  facet_grid(vars(filter), vars())

ggsave(ggplot_degs, file = "DEGs_up_down_barplot_pvalue05.png",  width = 10, height = 5) 

print(ggplot_degs)
```

```{r}
esquisser(plot_degs) 

ggplot_degs_bars = plot_degs %>%
 filter(!(Direction %in% "NEUTRAL")) %>%
 filter(filter %in% "p<0.05") %>%
 ggplot() +
  aes(x = reorder(study, Day), fill = Direction, weight = n_regulated) +
  geom_bar(position = "dodge") +
  scale_fill_manual(
    values = c(DOWN = "#0E252D",
    NEUTRAL = "#00C19F",
    UP = "#3399FF")
  ) +
  theme_minimal() +
  facet_wrap(vars(Vaccine), scales = "free_x", nrow = 2L) +
  geom_text(aes(label = n_regulated, y = n_regulated), 
            position = position_dodge(width = 0.9), 
            vjust = -0, 
            hjust = 0, 
            size = 2) +
  theme(legend.position = "top",
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1, size = 7),
        axis.text.y = element_text(vjust = 2))

ggsave(ggplot_degs_bars, file = "DEGs_up_down_barplot_pvalue05_4.png",  width = 10, height = 6) 

```


#Gene set databases

##Blood transcription modules <https://www.nature.com/articles/ni.2789>

```{r}
#Converter ítens de "Module member genes" em linhas em uma tabela longa
btm_annotation_table_long_immune <- BTM_modules %>%
  separate_rows(`Module member genes`, sep = ",") %>% 
  filter(`Module category` == "immune")

#Salvar
write.csv(btm_annotation_table_long_immune, file = "btm_annotation_table_long_immune.csv")

#Display
head(btm_annotation_table_long_immune)

#Calculate Set size
btm_annotation_nested <- btm_annotation_table_long_immune %>%
  group_by(`Module title`) %>% #Somar e agrupar
  mutate(SetSize = n())

#Convert gene symbol column to aggregated rows
btm_annotation_nested <- btm_annotation_nested %>%
  group_by(`Module title`) %>%
  summarize(Genes = paste(`Module member genes`, collapse = ", "), .groups = "drop") %>% 
  left_join(btm_annotation_nested, by = "Module title") %>% 
  filter(Genes != "NA") %>% 
  relocate(Genes, .after="SetSize")

#Salvar
write.csv(btm_annotation_nested, file = "btm_annotation_genes.csv")

#Convert to unique values table
btm_annotation_unique = btm_annotation_nested %>% 
  select(-`Module member genes`)

btm_annotation_unique = btm_annotation_unique %>% 
  distinct() %>% 
  relocate(Genes, .after="SetSize") %>% 
  filter(Genes != "NA")

#salvar
write.csv(btm_annotation_unique, file = "btm_annotation_unique.csv")

#Display
head(btm_annotation_unique)
```

## CellMarker

Baixar tabela dos sites "<http://117.50.127.228/CellMarker/CellMarker_download_files/file/Cell_marker_Human.xlsx>" ou "<http://117.50.127.228/CellMarker/CellMarker_download_files/file/Cell_marker_Human.xlsx>"

```{r}
##################### Baixar tabela de marcadores celulares

url <- "http://117.50.127.228/CellMarker/CellMarker_download_files/file/Cell_marker_Human.xlsx" #or http://117.50.127.228/CellMarker/CellMarker_download_files/file/Cell_marker_Human.xlsx 
f <- tempfile(fileext = ".xlsx")
download.file(url, f)
cell_marker_data <- readxl::read_excel(f, 1)

##################### Sub-agrupar células por sistema imunológico

#Filtrar marcadores e células por tecido (Blood)
cells <- cell_marker_data %>%
    filter(tissue_class == "Blood") %>%
    select(cell_name, marker)

##################### Filtrar células
# Acessar todos os tipos de células no dataset de Blood
cellmarker_celltypes <- unique(cells$cell_name)

# Filtrar apenas os tipos que contêm "B cell" e plasmacell
Bcells_no = cellmarker_celltypes[grep("B cell", cellmarker_celltypes)] #44 outputs
Plasmacells_no = cellmarker_celltypes[grep("plasma", cellmarker_celltypes)] #5 outputs

# Filtrar apenas os tipos que contêm "T cell" e seus subtipos "CD4" e "CD8"
Tcells_no = cellmarker_celltypes[grep("T cell", cellmarker_celltypes)] #93 outputs
TCD4_no = cellmarker_celltypes[grep("CD4", cellmarker_celltypes)] #35 outputs
TCD8_no = cellmarker_celltypes[grep("CD8", cellmarker_celltypes)] #34 outputs
Treg_no = cellmarker_celltypes[grep("Treg", cellmarker_celltypes)] #7 outputs
Thelper_no = cellmarker_celltypes[grep("helper", cellmarker_celltypes)] #17 outputs

# Filtrar outros tipos celulares da resposta imune
Neutrophils_no = cellmarker_celltypes[grep("neutrophil", cellmarker_celltypes)] #7 outputs
Monocyte_no = cellmarker_celltypes[grep("monocyte", cellmarker_celltypes)] #14 outputs
DCells_no = cellmarker_celltypes[grep("dendritic", cellmarker_celltypes)] #25 outputs
Macrophages_no = cellmarker_celltypes[grep("macrophage", cellmarker_celltypes)] #7 outputs
NKcells_no = cellmarker_celltypes[grep("killer", cellmarker_celltypes)] # 11 outputs
Neutrophils_no = cellmarker_celltypes[grep("mastcell", cellmarker_celltypes)] #7 outputs

# Criar variáveis para cada célula para enriquecimento

# Filtrar apenas as células da resposta imune adaptativa
Bcells <- filter(cells, str_detect(cell_name, regex("B cell|plasma cell|plasmablast", ignore_case = TRUE)))
TCD4 <- filter(cells, str_detect(cell_name, regex("CD4", ignore_case = TRUE)))
TCD8 <- filter(cells, str_detect(cell_name, regex("CD8", ignore_case = TRUE)))
Treg <- filter(cells, str_detect(cell_name, regex("Treg", ignore_case = TRUE)))
Thelper <- filter(cells, str_detect(cell_name, regex("helper", ignore_case = TRUE)))


# Filtrar células da resposta imune inata
Neutrophils <- filter(cells, str_detect(cell_name, regex("neutrophil", ignore_case = TRUE)))
Monocyte <- filter(cells, str_detect(cell_name, regex("monocyte", ignore_case = TRUE)))
DCells <- filter(cells, str_detect(cell_name, regex("dendritic", ignore_case = TRUE)))
Macrophages <- filter(cells, str_detect(cell_name, regex("macrophage", ignore_case = TRUE)))
Mastcells <- filter(cells, str_detect(cell_name, regex("mast cell", ignore_case = TRUE)))
NKcells <- filter(cells, str_detect(cell_name, regex("killer", ignore_case = TRUE)))
Platelets <- filter(cells, str_detect(cell_name, regex("Platelet", ignore_case = TRUE)))
Eosinophil <- filter(cells, str_detect(cell_name, regex("Eosinophil", ignore_case = TRUE)))
Basophil <- filter(cells, str_detect(cell_name, regex("Basophil", ignore_case = TRUE)))
Granulocyte <- filter(cells, str_detect(cell_name, regex("Granulocyte", ignore_case = TRUE)))

#Agrupar todos os resultados em Immune cells
Immune_adap = rbind(Bcells, TCD4, TCD8, Treg, Thelper)
Immune_adap$Type = "Adaptive"

Immune_innate = rbind(Neutrophils, Monocyte, DCells, Macrophages, Mastcells, NKcells, Platelets, Eosinophil, Basophil, Granulocyte)
Immune_innate$Type = "Innate"

#Unir tabelas
Immune_cells = rbind(Immune_adap, Immune_innate) 

#Salvar
write.csv(Immune_cells, file = "CellMarker_ImmuneCells.csv")

esquisser(Immune_cells_markersbycells)
```

Top 20 cells by #markers

```{r}

Immune_cells_markersbycells = Immune_cells %>% 
  group_by(cell_name, Type) %>% 
  summarise(n_markers = n()) %>% 
  arrange(desc(n_markers))
Immune_cells_markersbycells_top = Immune_cells_markersbycells[1:20, ]

ggImmune_cells_cellsbyimmune = ggplot(Immune_cells_markersbycells_top) +
  aes(x = reorder(cell_name, n_markers), y = n_markers, fill = Type) +
  scale_fill_manual(values = c('Innate' = "#989898",
                               'Adaptive' = "#6DF8DF")) +
  geom_col() +
  labs(
    x = "#Markers",
    y = "Cells",
    title = "Top 20 cells by #markers"
  ) +
  coord_flip() +
  theme_minimal() +
  geom_text(aes(label = n_markers, y = n_markers), 
            position = position_dodge(width = 0.9), 
            vjust = -0.2, 
            hjust = -0.2, 
            size = 3)

ggsave(ggImmune_cells_cellsbyimmune, file= "CellMarker_Top10_cellsbyn_markers.png", width = 10, height = 5)

print(ggImmune_cells_cellsbyimmune)

```

Number of Cells by immune system

```{r}
Immune_cells_cellsbyimmune = Immune_cells %>% 
  group_by(Type, cell_name) %>% 
  summarise(n_cells = n()) %>% 
  group_by(Type) %>% 
  summarise(n_cells = n())

ggImmune_cells_cellsbyimmune = ggplot(Immune_cells_cellsbyimmune) +
  aes(x = reorder(Type, n_cells), y = n_cells, fill = Type) +
  geom_col() +
  scale_fill_manual(values = c('Innate' = "#989898",
                               'Adaptive' = "#6DF8DF")) +
  labs(
    x = "#Cells",
    y = "Immune system",
    title = "#Cells by immune system"
  ) +
  coord_flip() +
  theme_minimal() +
  geom_text(aes(label = n_cells, y = n_cells), 
            position = position_dodge(width = 0.9), 
            vjust = -0.2, 
            hjust = -0.2, 
            size = 3)
ggsave(ggImmune_cells_cellsbyimmune, file= "CellMarker_nCellsbyimmunesystem.png", width = 10, height = 2)

print(ggImmune_cells_cellsbyimmune)
```

## ImmuneGO

A ontologia de processo biológico "immune system process" (<GO:0002376>) foi selecionada para uma análise direcionada das DEGs em relação à resposta imunológica. O dataset foi baixado pelo QuickGO filtrando os campos de Taxon (H. sapiens), Gene Products (all marked, except for "unreviewed") e Aspect ("Biological Process").

Para ler mais sobre códigos de evidencia: <https://geneontology.org/docs/guide-go-evidence-codes/>

------------------------------------------------------------------------

Baixar dataset e importar

```{r}

######### INPUT AQUI ######### 

#Data
all_degs_p_05_vac_infected = all_degs_p_05_vac_infected_19_12_23 %>% 
  select(!c("...1", "...12")) %>% 
  distinct()


#Gene sets
ImmuneGO_Annotated_GO_1BAIXAR DATASET ############### 
#Baixar dataset no site: https://www.ebi.ac.uk/QuickGO/annotations (.tsv)

#Importar
Immune_GO = read.table("~/Desktop/RNAseq temporary files/Gene set enrichment/28-08-23/QuickGO-ImmuneProcess.tsv", header = TRUE, sep = "\t")
Immune_GO$GO.NAME = toupper(Immune_GO$GO.NAME)
View(Immune_GO)

#Renomear
Immune_GO_T2G = Immune_GO %>% 
  select(GO.NAME, SYMBOL) %>% 
  arrange(GO.NAME) %>% 
  rename(Process = GO.NAME, 
         GENE = SYMBOL, 
         EVIDENCE = GO.EVIDENCE)

#Salvar
saveRDS(Immune_GO, file = "Immune_GO.rds")
write.csv(Immune_GO, file = "Immune_GO.csv")

############### DATASET DESCRIPTION ############### 
Immune_GO_genes_GO = Immune_GO %>% 
  select(SYMBOL, GO.NAME, GO.TERM, GO.EVIDENCE.CODE) %>% 
  unique() %>% 
  data.frame()

Immune_GO_genes = Immune_GO %>% 
  select(SYMBOL) %>% 
  unique() %>% 
  data.frame()

Immune_GO_GOs = Immune_GO %>% 
  select(GO.NAME) %>% 
  unique() %>% 
  data.frame()

Immune_GO_Evidence = Immune_GO %>% 
  select(GO.EVIDENCE.CODE) %>% 
  unique() %>% 
  data.frame()

#Stats
dim(Immune_GO) #4098 GOs
dim(Immune_GO_genes) #1251 GENES
dim(Immune_GO_genes_GO) #3889 GENES-TERMS
dim(Immune_GO_GOs) #336 TERMS-NAMES
dim(Immune_GO_Evidence) #12 

# Filtrar genes "NOT|involved_in" (QUALIFIER)
Immune_GO_genes_GO_Filt = Immune_GO %>%
  filter(QUALIFIER != "NOT|involved_in") %>% 
  write.csv(file="Immune_GO_genes_GO_Filt.csv") #Salvar

explore(Immune_GO_genes_GO_Filt) 
#4082 genes-terms
# 1294 Gene product ID
# 1247 gene symbols
# 3 qualifiers
# GO.EVIDENCE.CODE = 12
# 332 Terms, GO.NAME = 336
# 744 references
# 18 databases.
```

### Descrevendo e organizando dataset
Annotation of GOs

```{r}
######## Anotação de GOs ########
ImmuneGO_Manual = ImmuneGO_Annotated_GO # ImmuneGOs manually annotated by Immune System, Sub-system, and cells
#Unir anotações com genes
ImmuneGO_Annotated = ImmuneGO_Manual %>% 
  merge(Immune_GO_genes_GO_Filt, by.x = "Process", by.y = "GO.NAME", all.y = F, all.x=T) %>% 
  select(SYMBOL, everything()) %>% 
  rename(Evidence = GO.EVIDENCE.CODE)

#Calculate Set size
ImmuneGO_Annotated_genes <- ImmuneGO_Annotated %>%
  group_by(Process) %>% #Somar e agrupar
  mutate(SetSize = n())

#Convert gene symbol column to aggregated rows
ImmuneGO_Annotated_genes <- ImmuneGO_Annotated_genes %>%
  group_by(Process) %>%
  summarize(Genes = paste(SYMBOL, collapse = ","), .groups = "drop") %>% 
  left_join(ImmuneGO_Annotated_genes, by = "Process") %>% 
  filter(SYMBOL != "NA") %>% 
  relocate(Genes, .after="SetSize")
  
#View(ImmuneGO_Annotated_genes)

#Convert to unique values table
ImmuneGO_Annotated_unique = ImmuneGO_Annotated_genes %>% 
  select(Process, `Gene set, short`, `Immune system`, `Immune sub-system`, `Immune tissue`, `GO.TERM`, SetSize, Genes) %>% 
  distinct() %>% 
  relocate(Genes, .after="SetSize") %>% 
  filter(Genes != "NA")

#Retirar processos irrelevantes e remover espaços em branco
ImmuneGO_Annotated_unique = ImmuneGO_Annotated_unique %>% 
  filter(`Immune sub-system` != "Erythrocyte", `Immune tissue` != "Erythrocyte") %>%  
  mutate(Genes = str_replace_all(Genes, " ", ""))

# Calcular o tamanho do gene set

ImmuneGO_Annotated_unique <- ImmuneGO_Annotated_unique %>%
  mutate(SetSize = lengths(str_split(Genes, ",")),
         Genes = str_c(unique(str_split(Genes, ",")), collapse = ","))

# ImmuneGO_Annotated_unique$SetSize <- sapply(strsplit(ImmuneGO_Annotated_unique$Genes, ","), function(x) length(unique(x)))
# ImmuneGO_Annotated_unique$Genes <- sapply(strsplit(ImmuneGO_Annotated_unique$Genes, ","), function(x) paste(unique(x), collapse = ","))

#Salvar
write.csv(ImmuneGO_Annotated_genes, file="ImmuneGO_Annotated_Genes.csv")
write.csv(ImmuneGO_Annotated_unique, file="ImmuneGO_Annotated_GO.csv")

head(ImmuneGO_Annotated_unique)
```

Agrupar genes de todos os processos do sistema imune

```{r}
ImmuneGO_Annotated_unique = ImmuneGO_Annotated_GO

#Agrupar genes de todos os processos do sistema imune -----------------------------------------
ImmuneGO_Adaptive <- ImmuneGO_Annotated_unique %>%
  filter(`Immune system` == "Adaptive")
ImmuneGO_Innate <- ImmuneGO_Annotated_unique %>%
  filter(`Immune system` == "Innate")
ImmuneGO_Complement <- ImmuneGO_Annotated_unique %>%
  filter(`Immune system` == "Complement")
ImmuneGO_Adap_Cellular <- ImmuneGO_Adaptive %>% #Este comando somente inclui os genes de processos da resposta celular e muitos dos genes de receptor de celulas T não estão anotados. Por isso, eles serão incluidos nos codigos abaixo
  filter(`Immune sub-system` == "Cellular")
ImmuneGO_Adap_Humoral <- ImmuneGO_Adaptive %>%
  filter(`Immune sub-system` == "Humoral" )

View(ImmuneGO_Adap_Humoral)
```

Unificar genes de gene sets por categoria do sistema imune, remover duplicatas e calcular tamanho dos sets

```{r}

#Unificar genes e remover duplicatas
ImmuneGO_Adaptive_genes <- paste(ImmuneGO_Adaptive$Genes, collapse = ",")
ImmuneGO_Adaptive_genes <- unique(unlist(strsplit(ImmuneGO_Adaptive$Genes, ",")))
ImmuneGO_Innate_genes <- paste(ImmuneGO_Innate$Genes, collapse = ",")
ImmuneGO_Innate_genes <- unique(unlist(strsplit(ImmuneGO_Innate$Genes, ","))) 
ImmuneGO_Complement_genes <- paste(ImmuneGO_Complement$Genes, collapse = ",")
ImmuneGO_Complement_genes <- unique(unlist(strsplit(ImmuneGO_Complement$Genes, ",")))
ImmuneGO_Cellular_genes <- paste(ImmuneGO_Adap_Cellular$Genes, collapse = ",")
ImmuneGO_Cellular_genes <- unique(unlist(strsplit(ImmuneGO_Adap_Cellular$Genes, ",")))
ImmuneGO_Humoral_genes <- paste(ImmuneGO_Adap_Humoral$Genes, collapse = ",")
ImmuneGO_Humoral_genes <- unique(unlist(strsplit(ImmuneGO_Adap_Humoral$Genes, ",")))

# Calcular o tamanho do gene set
SetSize_adap <- length(ImmuneGO_Adaptive_genes) 
SetSize_innate <- length(ImmuneGO_Innate_genes)
SetSize_compl <- length(ImmuneGO_Complement_genes) 
SetSize_cellular <- length(ImmuneGO_Cellular_genes) 
SetSize_humoral <- length(ImmuneGO_Humoral_genes) 

#Criar novas linhas
new_row_adap <- tibble(
  `Gene set, short` = "Adaptive immune system",
  `Process` = "ADAPTIVE IMMUNE SYSTEM", 
  `Immune system` = "Adaptive",
  `Immune sub-system` = "General",
  `Immune tissue` = "General",
  Genes = paste(ImmuneGO_Adaptive_genes, collapse = ","),
  SetSize = SetSize_adap
)

new_row_innate <- tibble(
  `Gene set, short` = "Innate immune system",
  `Process` = "INNATE IMMUNE SYSTEM", 
  `Immune system` = "Innate",
  `Immune sub-system` = "General",
  `Immune tissue` = "General",
  Genes = paste(ImmuneGO_Innate_genes, collapse = ","),
  SetSize = SetSize_innate
)

new_row_compl <- tibble(
  `Gene set, short` = "Complement immune system",
  `Process` = "COMPLEMENT IMMUNE SYSTEM", 
  `Immune system` = "Complement",
  `Immune sub-system` = "General",
  `Immune tissue` = "General",
  Genes = paste(ImmuneGO_Complement_genes, collapse = ","),
  SetSize = SetSize_compl
)

new_row_cellular <- tibble(
  `Gene set, short` = "Cellular adaptive immune system",
  `Process` = "CELLULAR ADAPTIVE IMMUNE SYSTEM", 
  `Immune system` = "Adaptive",
  `Immune sub-system` = "Cellular",
  `Immune tissue` = "General",
  Genes = paste(ImmuneGO_in_vaccines_Genes_ann_adap_cellular$SYMBOL, collapse = ","),
  SetSize = SetSize_cellular
)

new_row_humoral <- tibble(
  `Gene set, short` = "Humoral adaptive immune system",
  `Process` = "HUMORAL ADAPTIVE IMMUNE SYSTEM", 
  `Immune system` = "Adaptive",
  `Immune sub-system` = "Humoral",
  `Immune tissue` = "General",
  Genes = paste(ImmuneGO_Humoral_genes, collapse = ","),
  SetSize = SetSize_humoral
)

# Adicionar linhas
ImmuneGO_Annotated_unique <- bind_rows(ImmuneGO_Annotated_unique, 
                                       new_row_adap, 
                                       new_row_innate, 
                                       new_row_compl, 
                                       new_row_cellular, 
                                       new_row_humoral)

#Transformar em tabela de genes por linhas
ImmuneGO_Annotated_Genes <- ImmuneGO_Annotated_unique %>%
  separate_rows(Genes, sep = ",")

#Salvar
write.csv(ImmuneGO_Annotated_unique, file = "ImmuneGO_Annotated_GO.csv")
write.csv(ImmuneGO_Annotated_Genes, file = "ImmuneGO_Annotated_Genes.csv")

#Display
head(ImmuneGO_Annotated_Genes)
head(ImmuneGO_Annotated_unique)
```

Unir genes da resposta celular

```{r}
#Unir genes da resposta celular, incluindo TCRs que estavam em outro processo
ImmuneGO_in_vaccines_Genes_ann_adap_cellular = filter(ImmuneGO_in_vaccines_Genes_ann, Process == "CELLULAR ADAPTIVE IMMUNE SYSTEM")
ImmuneGO_in_vaccines_Genes_ann_adap_cellular <- filter(ImmuneGO_in_vaccines_Genes_ann, 
                                                       Process == "CELLULAR ADAPTIVE IMMUNE SYSTEM" | 
                                                         grepl("^TRA|^TRB|^TRD|^TRG|^TRIL", SYMBOL)) 

ImmuneGO_in_vaccines_Genes_ann_adap_cellular$Process = "CELLULAR ADAPTIVE IMMUNE SYSTEM"
ImmuneGO_Cellular_genes <- paste(ImmuneGO_in_vaccines_Genes_ann_adap_cellular$SYMBOL, collapse = ",")
ImmuneGO_Cellular_genes <- unique(unlist(strsplit(ImmuneGO_in_vaccines_Genes_ann_adap_cellular$SYMBOL, ",")))

SetSize_cellular <- length(ImmuneGO_Cellular_genes) 
new_row_cellular <- tibble(
  `Gene set, short` = "Cellular adaptive immune system",
  `Process` = "CELLULAR ADAPTIVE IMMUNE SYSTEM", 
  `Immune system` = "Adaptive",
  `Immune sub-system` = "Cellular",
  `Immune tissue` = "General",
  Genes = paste(ImmuneGO_Cellular_genes, collapse = ","),
  SetSize = SetSize_cellular
)

ImmuneGO_Annotated_unique <- bind_rows(ImmuneGO_Annotated_unique, 
                                       new_row_cellular)

#Retirar linha antiga 
# ImmuneGO_Annotated_unique <- filter(ImmuneGO_Annotated_unique, 
#                                     !(Process == "CELLULAR ADAPTIVE IMMUNE SYSTEM" & SetSize == 384))

#Transformar em tabela de genes por linhas
ImmuneGO_Annotated_Genes <- ImmuneGO_Annotated_unique %>%
  separate_rows(Genes, sep = ",")

#Salvar
# write.csv(ImmuneGO_Annotated_unique, file = "ImmuneGO_Annotated_GO.csv")
# write.csv(ImmuneGO_Annotated_Genes, file = "ImmuneGO_Annotated_Genes.csv")
```


Separar genes do BCR e TCR

```{r}

TCR_BCR_repertoire = ImmuneGO_Annotated_Genes_Nested_Interesting %>% 
  filter(grepl("^TR", genes) & process == "ADAPTIVE IMMUNE SYSTEM" |
           grepl("^IG", genes) & process == "ADAPTIVE IMMUNE SYSTEM" |
           grepl("^TR", genes) & process == "CELLULAR ADAPTIVE IMMUNE SYSTEM" |
         grepl("^IG", genes) & process == "HUMORAL ADAPTIVE IMMUNE SYSTEM") %>% 
  mutate(process = if_else(process == "CELLULAR ADAPTIVE IMMUNE SYSTEM", "TCR REPERTOIRE", 
                           "BCR REPERTOIRE"),
         gene_set_short = if_else(process == "TCR REPERTOIRE", 
                                  "TCR REPERTOIRE", 
                           "BCR REPERTOIRE"),
         process = if_else(grepl("^TR", genes), "TCR REPERTOIRE", "BCR REPERTOIRE"))

TCR_BCR_repertoire_genes = TCR_BCR_repertoire %>% 
  select(genes) %>% 
  distinct()

ImmuneGO_Annotated_Genes_Nested_Interesting_12_12_23 = ImmuneGO_Annotated_Genes_Nested_Interesting %>%
  anti_join(TCR_BCR_repertoire_genes, by = "genes") %>% 
  bind_rows(TCR_BCR_repertoire)

write.csv(ImmuneGO_Annotated_Genes_Nested_Interesting_12_12_23, file = "ImmuneGO_Annotated_Genes_Nested_Interesting_12_12_24


Immune_GO_T2G = ImmuneGO_Annotated_Genes %>% 
  filter(go_term == "Manual",
         !process %in% c("ADAPTIVE IMMUNE SYSTEM",
                         "INNATE IMMUNE SYSTEM",
                         "BCR REPERTOIRE",
              3.csv")

```


Top 10 gene sets

```{r}
#Top 10 gene sets
ImmuneGO_Annotated_unique_top10 = ImmuneGO_Annotated_GO %>% 
  clean_names() %>% 
  arrange(desc(set_size)) %>% 
  slice_head(n = 10) %>% 
  mutate(immune_system = if_else(gene_set_short == "Adaptive Immune response", "Adaptive", immune_system))

#Gráfico de barras
ImmuneGO_top10_barplot = ggplot(ImmuneGO_Annotated_unique_top10) +
 aes(x = reorder(gene_set_short, set_size), 
     weight = set_size, 
     fill = immune_system) +  # Use reorder para reordenar as barras +
  geom_bar()+
  scale_fill_manual(
    values = c(Adaptive = "#6DF8DF",
    Complement = "#e9edc9",
    General = "#bde0fe",
    Innate = "#989898",
    Undefined = "#edede9")) +
 labs(y = "Set size", title = "ImmuneGO", subtitle = "Top 10 GO, by set size") +
 coord_flip() +
 theme_minimal()

#Salvar
ggsave(ImmuneGO_top10_barplot, file="ImmuneGO_top10_barplot.png", width = 10, height = 6)

#Display
head(ImmuneGO_top10_barplot)
```

Quais os tamanhos dos gene sets?

```{r}

# Usar cut para dividir os valores em intervalos
ImmuneGO_Annotated_GO$SetSize_intervals <- cut(ImmuneGO_Annotated_GO$SetSize, c(0, 10, 50, 100, Inf), labels = FALSE)
ImmuneGO_Annotated_GO = ImmuneGO_Annotated_GO %>% arrange(desc(SetSize))

# Contar quantas linhas caem em cada intervalo
contagem <- table(ImmuneGO_Annotated_GO$SetSize_intervals)

#Definir ordem de fatores para plotar
ImmuneGO_Annotated_GO$`Immune system` <- factor(ImmuneGO_Annotated_GO$`Immune system`, levels = c("General", "Adaptive", "Innate", "Complement", "Undefined"))

#Plotar
plot_hist_immunego = ggplot(ImmuneGO_Annotated_GO) +
  aes(x = SetSize_intervals, fill = `Immune system`, show.legend = FALSE) +
  geom_histogram(bins = 4L) +
  scale_fill_manual(
    values = c(Adaptive = "#6DF8DF",
    Complement = "#e9edc9",
    General = "#bde0fe",
    Innate = "#989898",
    Undefined = "#edede9")) +
  theme_minimal() +
  #facet_grid(vars(), vars(`Immune system`)) +
  theme(legend.position = "top",
        panel.grid.major = element_blank()) +  # Remove as gridlines secundárias
  scale_x_continuous(breaks = c(1, 2, 3, 4, 5), labels = c("0-10", "10-20", "20-50", "50-100", "100+")) +
  guides(color = "none")

#Salvar
ggsave(plot_hist_immunego, file = "plot_hist_immunego.png", width = 10, height = 5)

#Display
plot_hist_immunego
```

Qual a relação entre os genes de cada resposta imunológica?
```{r}
ImmuneGO_Annotated_Genes = ImmuneGO_Annotated_Genes %>% clean_names()

#Unir tabelas
ImmuneGO_Annotated_Genes_Nested = ImmuneGO_Annotated_Genes %>% 
  clean_names() %>% 
  group_by(process) %>% 
  arrange(desc(set_size)) %>% 
  filter(process %in% c("ADAPTIVE IMMUNE SYSTEM", "INNATE IMMUNE SYSTEM", "CELLULAR ADAPTIVE IMMUNE SYSTEM", "HUMORAL ADAPTIVE IMMUNE SYSTEM", "IMMUNOGLOBULIN MEDIATED IMMUNE RESPONSE"))

Immune_GO_adap_in = ImmuneGO_Annotated_Genes %>%
  clean_names() %>%
  mutate(
    process = case_when(
      immune_tissue == "Complement" ~ "COMPLEMENT",
      immune_sub_system == "Inflammation" ~ "INFLAMMATION",
      immune_tissue == "B cell" ~ "B CELLS",
      immune_tissue == "T cell" ~ "T CELLS",
      immune_tissue == "Dendritic cell" ~ "DENDRITIC CELLS",
      immune_tissue == "Macrophage" ~ "MACROPHAGES",
      immune_sub_system %in% c("APP", "PRR") ~ "ARPP",
      process %in% c(
        "ANTIVIRAL INNATE IMMUNE RESPONSE",
        "CELLULAR RESPONSE TO TYPE I INTERFERON",
        "CELLULAR RESPONSE TO TYPE II INTERFERON",
        "TYPE I INTERFERON-MEDIATED SIGNALING PATHWAY",
        "TYPE II INTERFERON-MEDIATED SIGNALING PATHWAY",
        "TYPE III INTERFERON-MEDIATED SIGNALING PATHWAY"
      ) ~ "ANTIVIRAL AND INTERFERON",
      TRUE ~ as.character(process)
    )
  ) %>% 
  filter(process %in% c("COMPLEMENT",
                        "INFLAMMATION",
                        "B CELLS",
                        "T CELLS",
                        "DENDRITIC CELLS",
                        "MACROPHAGES",
                        "ARPP",
                        "ANTIVIRAL AND INTERFERON"))
#Unir processos
ImmuneGO_Annotated_Genes_Nested_Interesting <- bind_rows(ImmuneGO_Annotated_Genes_Nested,
                                             Immune_GO_adap_in) %>% 
  distinct() #retirar genes duplicados por processo

#Salvar
write.csv(ImmuneGO_Annotated_Genes_Nested_Interesting, file = "ImmuneGO_Annotated_Genes_Nested_Interesting.csv")
```


Quais são os genes mais compartilhados?

```{r}
ImmuneGO_Annotated_Genes_sharedbetween_subsystem



#Top 10 gene sets
ImmuneGO_Annotated_Genes_top20 = ImmuneGO_Annotated_Genes_sharedbetween_subsystem %>% 
  clean_names() %>% 
  arrange(desc(shared)) %>% 
  slice_head(n = 20) 

#Gráfico de barras
ImmuneGO_Genes_top20_barplot = ggplot(ImmuneGO_Annotated_Genes_top20) +
 aes(x = reorder(genes, shared), 
     weight = shared) +
  geom_bar(fill="lightblue") +
 labs(y = "Shared", title = "ImmuneGO", subtitle = "Top 10 genes, shared") +
 coord_flip() +
 theme_minimal()

#Salvar
ggsave(ImmuneGO_Genes_top20_barplot, file="ImmuneGO_Genes_top20_barplot.png", width = 5, height = 10)

#Display
head(ImmuneGO_top10_barplot)


```

```{r}
############### VENN DIAGRAM
#Input
data = ImmuneGO_Annotated_Genes_Nested_Interesting_12_12_23

# Função para calcular a sobreposição entre pares de vacinas e a porcentagem de genes compartilhados
overlap_genes <- function(cond1, cond2, data) {
  genes_cond1 <- data$genes[data$process == cond1] #Trocar coluna
  genes_cond2 <- data$genes[data$process == cond2] #Trocar coluna
  genes_shared <- intersect(genes_cond1, genes_cond2)
  genes_notshared_cond1 <- setdiff(genes_cond1, genes_cond2)
  genes_notshared_cond2 <- setdiff(genes_cond2, genes_cond1)
  
  total_genes_cond1 <- length(genes_cond1)
  total_genes_cond2 <- length(genes_cond2)
  
  percentage_shared_cond1 <- length(genes_shared) / total_genes_cond1 * 100
  percentage_shared_cond2 <- length(genes_shared) / total_genes_cond2 * 100
  
  shared_genes <- data.frame(
    Cond1 = cond1,
    Cond2 = cond2,
    Shared = length(genes_shared),
    NotShared_cond1 = length(genes_notshared_cond1),
    NotShared_cond2 = length(genes_notshared_cond2),
    Total_Genes_Cond1 = total_genes_cond1,
    Total_Genes_Cond2 = total_genes_cond2,
    Genes_Names = paste(genes_shared, collapse = ", "),
    Percentage_Shared_Cond1 = percentage_shared_cond1,
    Percentage_Shared_Cond2 = percentage_shared_cond2
  )
  
         "TCR REPERTOIRE")) %>% 
  select(process, genes) 

CellMarker_genes = CellMarker_ImmuneCells %>% 
  select(cell_name, marker) %>% 
  tibble()

return(shared_genes)
}
# Obtenha uma lista de todas as vacinas únicas
unique_cond <- unique(data$process)  #trocar "study" pelo nome da coluna de interesse

# Inicialize um dataframe para armazenar os resultados
shared_genes_df <- data.frame()

# Calcular a sobreposição e porcentagem para cada par de vacinas
for (i in 1:(length(unique_cond) - 1)) {
  for (j in (i + 1):length(unique_cond)) {
    resultado_temp <- overlap_genes(unique_cond[i], unique_cond[j], data)
    shared_genes_df <- bind_rows(shared_genes_df, resultado_temp)
  }
}

# Função para realizar o teste exato de Fisher
fisher_exact_test <- function(shared, notshared1, notshared2) {
  cont_table <- matrix(c(shared, notshared1, notshared2, 0), nrow = 2)
  results_fisher <- fisher.test(cont_table)
  return(results_fisher$p.value)
}

# Aplicar a função a cada linha da tabela de resultados
shared_genes_df$pvalue <- mapply(fisher_exact_test, 
                                    shared_genes_df$Shared, 
                                    shared_genes_df$NotShared_cond1, 
                                    shared_genes_df$NotShared_cond2)
shared_genes_df <- shared_genes_df %>%
  mutate(qvalue = qvalue(pvalue)$qvalues)

# Calcular -log(p)
shared_genes_df$`-log(p)`= -log(shared_genes_df$pvalue, 10)

#Duplicar e trocar colunas (Vacina 1 e Vacina 2)
shared_genes_df2 = shared_genes_df
shared_genes_df2[, c("Cond1", "Cond2")] <- shared_genes_df2[, c("Cond2", "Cond1")]

#Unir datasets com os dados espelhados
shared_genes_df_mirror = rbind(shared_genes_df, shared_genes_df2)

#Converter NA, NaNe inf em 0
shared_genes_df_mirror[is.na(shared_genes_df_mirror)] <- 0
shared_genes_df_mirror<- replace(shared_genes_df_mirror, shared_genes_df_mirror == "Inf", 0)

#Arredondar porcentagens
shared_genes_df_mirror = shared_genes_df_mirror %>%
  mutate(Percentage_Shared_Cond1 = round(Percentage_Shared_Cond1, 2),
         Percentage_Shared_Cond2 = round(Percentage_Shared_Cond2, 2))

#Salvar resultados
write.csv(shared_genes_df_mirror, file = "ImmuneGO_sharedgenes.csv") #Alterar

```

### Total genes

```{r}
# Required
ann = ImmuneGO_Annotated_Genes_Nested_Interesting_12_12_23 
data = shared_genes_df_mirror
filename = "ImmuneGO_Shared_TOTAL_"

#Matriz
matrix_heatmap_df = data %>%
  filter(!Cond1 %in% c("TCR REPERTOIRE", "BCR REPERTOIRE"),
         !Cond2 %in% c("TCR REPERTOIRE", "BCR REPERTOIRE")) %>% 
  dcast(Cond1 ~ Cond2, 
        value.var = "Shared", 
        fun.aggregate = mean) %>% 
  column_to_rownames("Cond1")


#Anotar colunas
############## Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_immune_complex =  ann %>% 
  select(process, immune_system) %>% 
  distinct()

#Unir tabela de anotações
ann_immune_heatmap = matrix_heatmap_df %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "process") %>% 
  rename(process = '.') %>% 
  merge(ann_immune_complex, by = "process", all.x = F) %>% 
  arrange()


#Ordenar colunas da matriz
matrix_heatmap_ordered = matrix_heatmap_df %>% 
  rownames_to_column(var= "process") %>% 
  arrange(process) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  column_to_rownames("process") %>% 
  as.matrix()#Replace

ann_immune_heatmap = ann_immune_heatmap %>% 
  column_to_rownames("process")

dim(matrix_heatmap_ordered)
dim(ann_immune_heatmap)
```


```{r}
########## Definir Cores dos grupos
#Vaccines
ann_colors = list(
  immune_system = c("Adaptive" = "#D7B0EE",
                    "Innate" = "#56cfe1"))

col_fun <- colorRamp2(c(0, 100), c("white", "#16E0BD"))

#Criar heatmap annotation
col_ha = HeatmapAnnotation(df = ann_immune_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")
row_ha = rowAnnotation(df = ann_immune_heatmap, col = ann_colors)


#TOTAL GENES (with values)
mat = matrix_heatmap_ordered 

fileimageheatmap_ungrouped= paste0(filename, sep ="_", "VALUES_Complexheatmap.png")

png(file = fileimageheatmap_ungrouped, width=40, height=40,units="cm",res=900)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = col_ha,
                        right_annotation = row_ha,
                        na_col = "black",
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        cell_fun = function(j, i, x, y, width, height, fill) {
                        grid.text(sprintf("%.1f", mat[i, j]), x, y, gp = gpar(fontsize = 8))}, #valor das células
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "Total genes", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 8),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = gpar(fontsize = 10),
                        #row_split = 2, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        #column_split = NULL, #Split group clusters
                        column_gap = unit(8, "mm"),
                        show_column_dend = FALSE,
                        show_row_dend = FALSE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = unit(20, "cm"), 
                        height = unit(20, "cm")
)

draw(heatmap_plot_all, annotation_legend_side = "left", heatmap_legend_side = "left")

dev.off()


#TOTAL GENES (no values)
mat = matrix_heatmap_ordered 

fileimageheatmap_ungrouped= paste0(filename, sep ="_", "NOVALUES_Complexheatmap.png")

png(file = fileimageheatmap_ungrouped, width=40, height=40,units="cm",res=900)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = col_ha,
                        left_annotation = row_ha,
                        na_col = "black",
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        cell_fun = function(j, i, x, y, width, height, fill) {
                        grid.text(sprintf("%.1f", mat[i, j]), x, y, gp = gpar(fontsize = 8))}, #valor das células
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "Total genes", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 8),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = gpar(fontsize = 10),
                        #row_split = 2, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        #column_split = NULL, #Split group clusters
                        column_gap = unit(8, "mm"),
                        show_column_dend = FALSE,
                        show_row_dend = FALSE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = unit(20, "cm"), 
                        height = unit(20, "cm")
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```



### Manual
```{r}

#GSE199750
degs <- filter(all_degs_p_05_vac_infected, condition == "BNT (V1, D6)") #Não enriquecido
degs <- filter(all_degs_p_05_vac_infected, condition == "BNT (V2, D1)") #Enriquecido
degs <- filter(all_degs_p_05_vac_infected, condition == "BNT (V3, D1)") #Enriquecido
degs <- filter(all_degs_p_05_vac_infected, condition == "BNT-MO (V3, D1)") #Enriquecido
degs <- filter(all_degs_p_05_vac_infected, condition == "ChAd (V1, D6)") #Enriquecido
degs <- filter(all_degs_p_05_vac_infected, condition == "ChAd (V2, D1)") #Enriquecido
Shared identity (%)


#### Cond1
```{r}
# Required
ann = ImmuneGO_Annotated_Genes_Nested_Interesting_12_12_23 
data = shared_genes_df_mirror
filename = "ImmuneGO_Shared_IDENTITY_Cond1"

#Matriz
matrix_heatmap_df = data %>%
  filter(!Cond1 %in% c("TCR REPERTOIRE", "BCR REPERTOIRE"),
         !Cond2 %in% c("TCR REPERTOIRE", "BCR REPERTOIRE")) %>% 
  dcast(Cond1 ~ Cond2, 
        value.var = "Percentage_Shared_Cond1", 
        fun.aggregate = mean) %>% 
  column_to_rownames("Cond1")


#Anotar colunas
############## Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_immune_complex =  ann %>% 
  select(process, immune_system) %>% 
  distinct()

#Unir tabela de anotações
ann_immune_heatmap = matrix_heatmap_df %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "process") %>% 
  rename(process = '.') %>% 
  merge(ann_immune_complex, by = "process", all.x = F) %>% 
  arrange()


#Ordenar colunas da matriz
matrix_heatmap_ordered = matrix_heatmap_df %>% 
  rownames_to_column(var= "process") %>% 
  arrange(process) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  column_to_rownames("process") %>% 
  as.matrix()#Replace

ann_immune_heatmap = ann_immune_heatmap %>% 
  column_to_rownames("process")

dim(matrix_heatmap_ordered)
dim(ann_immune_heatmap)

########## Definir Cores dos grupos
#Vaccines
ann_colors = list(
  immune_system = c("Adaptive" = "#D7B0EE",
                    "Innate" = "#56cfe1"))

col_fun <- colorRamp2(c(0, 100), c("white", "#16E0BD"))

#Criar heatmap annotation
col_ha = HeatmapAnnotation(df = ann_immune_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")
row_ha = rowAnnotation(df = ann_immune_heatmap, col = ann_colors)


#TOTAL GENES (with values)
mat = matrix_heatmap_ordered 

fileimageheatmap_ungrouped= paste0(filename, sep ="_", "VALUES_Complexheatmap.png")

png(file = fileimageheatmap_ungrouped, width=40, height=40,units="cm",res=900)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = col_ha,
                        right_annotation = row_ha,
                        na_col = "black",
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        cell_fun = function(j, i, x, y, width, height, fill) {
                        grid.text(sprintf("%.1f", mat[i, j]), x, y, gp = gpar(fontsize = 8))}, #valor das células
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "Total genes", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 8),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = gpar(fontsize = 10),
                        #row_split = 2, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        #column_split = NULL, #Split group clusters
                        column_gap = unit(8, "mm"),
                        show_column_dend = FALSE,
                        show_row_dend = FALSE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = unit(20, "cm"), 
                        height = unit(20, "cm")
)

draw(heatmap_plot_all, annotation_legend_side = "left", heatmap_legend_side = "left")

dev.off()


#TOTAL GENES (no values)
mat = matrix_heatmap_ordered 

fileimageheatmap_ungrouped= paste0(filename, sep ="_", "NOVALUES_Complexheatmap.png")

png(file = fileimageheatmap_ungrouped, width=40, height=40,units="cm",res=900)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = col_ha,
                        left_annotation = row_ha,
                        na_col = "black",
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        cell_fun = function(j, i, x, y, width, height, fill) {
                        grid.text(sprintf("%.1f", mat[i, j]), x, y, gp = gpar(fontsize = 8))}, #valor das células
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "Total genes", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 8),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = gpar(fontsize = 10),
                        #row_split = 2, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        #column_split = NULL, #Split group clusters
                        column_gap = unit(8, "mm"),
                        show_column_dend = FALSE,
                        show_row_dend = FALSE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = unit(20, "cm"), 
                        height = unit(20, "cm")
)

draw(heatmap_plot_all, annotation_legend_side = "left", heatmap_legend_side = "left")

dev.off()
```
#### Cond2
```{r}
# Required
ann = ImmuneGO_Annotated_Genes_Nested_Interesting_12_12_23 
data = shared_genes_df_mirror
filename = "ImmuneGO_Shared_IDENTITY_Cond2"

#Matriz
matrix_heatmap_df = data %>%
  filter(!Cond1 %in% c("TCR REPERTOIRE", "BCR REPERTOIRE"),
         !Cond2 %in% c("TCR REPERTOIRE", "BCR REPERTOIRE")) %>% 
  dcast(Cond2 ~ Cond1, 
        value.var = "Percentage_Shared_Cond2", 
        fun.aggregate = mean) %>% 
  column_to_rownames("Cond2")


#Anotar colunas
############## Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_immune_complex =  ann %>% 
  select(process, immune_system) %>% 
  distinct()

#Unir tabela de anotações
ann_immune_heatmap = matrix_heatmap_df %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "process") %>% 
  rename(process = '.') %>% 
  merge(ann_immune_complex, by = "process", all.x = F) %>% 
  arrange()


#Ordenar colunas da matriz
matrix_heatmap_ordered = matrix_heatmap_df %>% 
  rownames_to_column(var= "process") %>% 
  arrange(process) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  column_to_rownames("process") %>% 
  as.matrix()#Replace

ann_immune_heatmap = ann_immune_heatmap %>% 
  column_to_rownames("process")

dim(matrix_heatmap_ordered)
dim(ann_immune_heatmap)

########## Definir Cores dos grupos
#Vaccines
ann_colors = list(
  immune_system = c("Adaptive" = "#D7B0EE",
                    "Innate" = "#56cfe1"))

col_fun <- colorRamp2(c(0, 100), c("white", "#16E0BD"))

#Criar heatmap annotation
col_ha = HeatmapAnnotation(df = ann_immune_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")
row_ha = rowAnnotation(df = ann_immune_heatmap, col = ann_colors)


#TOTAL GENES (with values)
mat = matrix_heatmap_ordered 

fileimageheatmap_ungrouped= paste0(filename, sep ="_", "VALUES_Complexheatmap.png")

png(file = fileimageheatmap_ungrouped, width=40, height=40,units="cm",res=900)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = col_ha,
                        right_annotation = row_ha,
                        na_col = "black",
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        cell_fun = function(j, i, x, y, width, height, fill) {
                        grid.text(sprintf("%.1f", mat[i, j]), x, y, gp = gpar(fontsize = 8))}, #valor das células
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "Total genes", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 8),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = gpar(fontsize = 10),
                        #row_split = 2, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        #column_split = NULL, #Split group clusters
                        column_gap = unit(8, "mm"),
                        show_column_dend = FALSE,
                        show_row_dend = FALSE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = unit(20, "cm"), 
                        height = unit(20, "cm")
)

draw(heatmap_plot_all, annotation_legend_side = "left", heatmap_legend_side = "left")

dev.off()


#TOTAL GENES (no values)
mat = matrix_heatmap_ordered 

fileimageheatmap_ungrouped= paste0(filename, sep ="_", "NOVALUES_Complexheatmap.png")

png(file = fileimageheatmap_ungrouped, width=40, height=40,units="cm",res=900)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = col_ha,
                        left_annotation = row_ha,
                        na_col = "black",
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        cell_fun = function(j, i, x, y, width, height, fill) {
                        grid.text(sprintf("%.1f", mat[i, j]), x, y, gp = gpar(fontsize = 8))}, #valor das células
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "Total genes", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 8),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = gpar(fontsize = 10),
                        #row_split = 2, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        #column_split = NULL, #Split group clusters
                        column_gap = unit(8, "mm"),
                        show_column_dend = FALSE,
                        show_row_dend = FALSE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = unit(20, "cm"), 
                        height = unit(20, "cm")
)

draw(heatmap_plot_all, annotation_legend_side = "left", heatmap_legend_side = "left")

dev.off()
```


### Verificando representação nos datasets

Total de genes representados

```{r}
#Genes representados e não representados

ImmuneGO_Annotated_Genes = ImmuneGO_Annotated_Genes_Nested_Interesting %>% 
  select(genes, process) 

ImmuneGO_Annotated_Genes_2 = ImmuneGO_Annotated_Genes_Nested_Interesting %>% 
  select(genes) %>% 
  distinct() %>% 
  mutate(process = "All") %>% 
  bind_rows(ImmuneGO_Annotated_Genes_Nested_Interesting)


ImmuneGO_Annotated_Genes_Interesting_studies = ImmuneGO_Annotated_Genes_2 %>% 
  merge(genes_raw_combined, all.x = T, all.y =F) %>% 
  select(study, process, genes) %>% 
  group_by(study, process) %>% 
  mutate(process = case_when(
  process == "ANTIVIRAL AND INTERFERON" ~ "ANTIVIRAL/IFN",
  process == "INNATE IMMUNE SYSTEM" ~ "INNATE",
  process == "HUMORAL ADAPTIVE IMMUNE SYSTEM" ~ "HUMORAL",
  process == "CELLULAR ADAPTIVE IMMUNE SYSTEM" ~ "CELLULAR",
  process == "ADAPTIVE IMMUNE SYSTEM" ~ "ADAPTIVE",
  process == "IMMUNOGLOBULIN MEDIATED IMMUNE RESPONSE" ~ "IMMUNOGLOBULIN-MEDIATED",
  TRUE ~ process)) %>% 
  mutate(in_out = 1) %>% 
  filter(!is.na(study)) %>% 
  mutate(process = factor(process, levels = c('INNATE', 
                                               'ADAPTIVE', 
                                               'CELLULAR', 
                                               'HUMORAL')),
         study = factor(study, levels = c('GSE199750', 
                                          'GSE201533',
                                          'GSE201530',
                                          'GSE189039',
                                          "GSE206023")))


######## ALL

ImmuneGO_Interesting_Genes_plot.higher <- ImmuneGO_Annotated_Genes_Interesting_studies %>%
  filter(process %in% c("ADAPTIVE", "INNATE", "HUMORAL", "CELLULAR")) %>%
  ggplot() +
  aes(x = study, y = genes, fill = study) +  # Reordenar os levels de study
  geom_tile() +
  scale_fill_manual(
    values = c("GSE199750" = "#80ffdb",
               "GSE201533" = "#5e60ce", 
               "GSE201530" = "#7400b8",
               "GSE206023" = "#4361ee",
               "GSE189039" = "#48bfe3")) +
  coord_flip() +
  theme_minimal() +
  facet_wrap(vars(process), scales = "free_x", ncol = 1) +
  labs(
    x = "GSE_ID",
    y = "#Genes",
    title = "#Genes represented in ImmuneGO, by study") + 
  theme(axis.text.x = element_text(angle = 90, size = 3, hjust = 1),
        axis.text.y = element_text(size = 20),
        strip.text = element_text(size= 20)) +
    scale_x_discrete(limits = rev(levels(ImmuneGO_Annotated_Genes_Interesting_studies$study)))

#Salvar
ggsave(ImmuneGO_Interesting_Genes_plot.higher, 
       file = "ImmuneGO_Interesting_Genes_Adaptive_Innate_plot_label.png", 
       width = 20, height = 20)


#Sem label
ImmuneGO_Interesting_Genes_plot.nolabel = ImmuneGO_Annotated_Genes_Interesting_studies %>%
  filter(process %in% c("ADAPTIVE", "INNATE", "HUMORAL", "CELLULAR")) %>% 
  ggplot() +
  aes(x = study, y = genes, fill = study) +  # Reordenar os levels de study
  geom_tile() +
  scale_fill_manual(
    values = c("GSE199750" = "#80ffdb",
               "GSE201533" = "#5e60ce", 
               "GSE201530" = "#7400b8",
               "GSE206023" = "#4361ee",
               "GSE189039" = "#48bfe3"))  +
  coord_flip() +
  theme_minimal() +
  facet_wrap(vars(process), scales = "free_x", ncol = 1)  +
  labs(
    x = "GSE_ID",
    y = "#Genes",
    title = "#Genes represented in ImmuneGO, by study") + 
  theme(axis.text.x = element_text(angle = 90, size = 0, hjust = 1),
        axis.text.y = element_text(size = 20),
        strip.text = element_text(size= 20)) +
  scale_x_discrete(limits = rev(levels(ImmuneGO_Annotated_Genes_Interesting_studies$study)))

#Salvar
ggsave(ImmuneGO_Interesting_Genes_plot.nolabel, 
       file = "ImmuneGO_Interesting_Genes_Adaptive_Innate_plot_nolabel.png", 
       width = 20, height = 20)

############### Grouped

ImmuneGO_Annotated_Genes_Nested_Interesting_grouped = ImmuneGO_Annotated_Genes_Nested_Interesting %>% 
  filter(process %in% c("INNATE IMMUNE SYSTEM", 
                        "ADAPTIVE IMMUNE SYSTEM", 
                        "CELLULAR ADAPTIVE IMMUNE SYSTEM", 
                        "HUMORAL ADAPTIVE IMMUNE SYSTEM"))

ImmuneGO_genes_indatasets.merge_grouped = genes_raw_combined %>% 
  group_by(study, genes) %>% 
  merge(ImmuneGO_Annotated_Genes_Nested_Interesting_grouped, by.x = "genes", all.x = F, all.y = T) %>%
  select(genes, study, process) %>% 
  distinct() %>% 
  group_by(study, process) %>%  
  select(-genes) %>% 
  summarise(n_genes_in = n()) %>% 
  merge(ImmuneGO_Annotated_Genes_Nested_Interesting_grouped, by = "process", all.y = F) %>% 
  select(study, process, set_size, n_genes_in) %>%
  distinct() %>% 
  filter(study != "NA")  %>% 
  summarise(study,
            process,
            set_size,
            n_genes_in,
            prop_in = (n_genes_in/set_size)*100,
            prop_in = round(prop_in, 1),
            study,
            process) %>% 
  mutate(process = case_when(
            process == "ANTIVIRAL AND INTERFERON" ~ "ANTIVIRAL/IFN",
            process == "INNATE IMMUNE SYSTEM" ~ "INNATE",
            process == "HUMORAL ADAPTIVE IMMUNE SYSTEM" ~ "HUMORAL",
            process == "CELLULAR ADAPTIVE IMMUNE SYSTEM" ~ "CELLULAR",
            process == "ADAPTIVE IMMUNE SYSTEM" ~ "ADAPTIVE",
            process == "IMMUNOGLOBULIN MEDIATED IMMUNE RESPONSE" ~ "IMMUNOGLOBULIN-MEDIATED",
            TRUE ~ process),
         process = factor(process, levels = c('INNATE', 
                                               'ADAPTIVE', 
                                               'CELLULAR', 
                                               'HUMORAL')),
         study = factor(study, levels = unique(study)))

#Plotar

ggImmuneGO_genes_indatasets.stats = ggplot(ImmuneGO_genes_indatasets.merge_grouped) +
  aes(x = reorder(study, prop_in), fill = study, weight = prop_in) +
  geom_bar() +
  scale_fill_manual(
    values = c("GSE199750" = "#80ffdb",
               "GSE201533" = "#5e60ce", 
               "GSE201530" = "#7400b8",
               "GSE206023" = "#4361ee",
               "GSE189039" = "#48bfe3")) +
  coord_flip() +
  theme_minimal() +
  geom_text(aes(label = prop_in, y = prop_in), 
            position = position_dodge(width = 0.9), 
            vjust = -0, 
            hjust = -0.2, 
            size = 3) +
  labs(
    x = "GSE_ID",
    y = "#Genes",
    title = "#Genes represented in ImmuneGO, by study"
  ) +
  facet_wrap(vars(process), ncol = 1) +
  theme(strip.text = element_text(size=8))
  
ggsave(ggImmuneGO_genes_indatasets.stats, file = "ImmuneGO_Genesrepresented_bystudy.png", width = 6, height = 10)

```



## MSigDB

Obtendo coleções

```{r}

############### Obter coleções ############### 
#Obter genes de um conjunto específico diretamente do MSigDB. 
msigdbr_collections() #Listar conjuntos

############### Hallmark ###############
hallmark <- msigdbr(species = "Homo sapiens", category = "H") %>% 
  dplyr::select(gs_name, gene_symbol) 

#Limpar texto
hallmark$gs_name <- gsub("HALLMARK_", "", hallmark$gs_name)
hallmark$gs_name <- gsub("_", " ", hallmark$gs_name)

#Conjuntos de GO
GO_BP <- msigdbr(species = "Homo sapiens", 
                 category = "C5", 
                 subcategory = "GO:BP") %>%
  select(gs_name, gene_symbol)

GO_MF <- msigdbr(species = "Homo sapiens", 
                 category = "C5", 
                 subcategory = "GO:MF") %>%
  select(gs_name, gene_symbol)

#Conjunto de marcadores de células 
cell_types <- msigdbr(species = "Homo sapiens", 
                      category = "C8") %>% 
  select(gs_name, gene_symbol)

#Conjunto de marcadores da resposta imune
immune_sig = msigdbr(species = "Homo sapiens", 
                     category = "C7", 
                     subcategory = "IMMUNESIGDB") %>% 
  select(gs_name, gene_symbol)

head(immune_sig)

#Conjunto de marcadores de vacinas
vax_sig = msigdbr(species = "Homo sapiens", 
                  category = "C7", 
                  subcategory = "VAX") %>% 
  select(gs_name, gene_symbol)

#Limpar texto
vax_sig$gs_name <- gsub("_", " ", vax_sig$gs_name)
head(vax_sig)
```

# VAX Sig DB

Os estudos foram filtrados pelo tamanho do conjunto de genes compartilhados (GSSize \> 10 ou \>50) e com pvalue (\<0.25).

Os dados foram anotados no R e depois avaliados iterativamente no excel.

```{r}
#Importar arquivo
VaxSigDB_genesets = VAX_GeneSets_Annotated_26_10_23

# Divida os dados em blocos de informações para cada vacina
vaccines_att <- split(VaxSigDB_genesets$Col2, VaxSigDB_genesets$Col1)

# Combine data
VAX_Genesets <- bind_rows(lapply(vaccines_att, function(x) data.frame(t(x))), .id = "V")

# Name a new index column
colnames(VAX_Genesets) <- paste0("V", seq_len(ncol(VAX_Genesets)))

# Transpose
rownames(VAX_Genesets) <- VAX_Genesets$V1
VAX_Genesets = VAX_Genesets %>% 
  t() %>%
  as.data.frame()

#Order columns
VAX_Genesets = VAX_Genesets %>% 
  select(STANDARD_NAME, everything()) %>% 
  filter(STANDARD_NAME != "STANDARD_NAME") 
rownames(VAX_Genesets) = VAX_Genesets$SYSTEMATIC_NAME
VAX_Genesets = VAX_Genesets %>% select(STANDARD_NAME, 
                                       DESCRIPTION_BRIEF,
                                       DESCRIPTION_FULL,
                                       PMID,
                                       AUTHORS,
                                       SYSTEMATIC_NAME,
                                       GENE_SYMBOLS,
                                       everything())

#Anotar novas colunas
VAX_Genesets <- VAX_Genesets %>%
  mutate(
    SAMPLE_SOURCE = ifelse(
      grepl("pbmc|peripheral blood mononuclear cell|peripheral blood mononuclear cells", DESCRIPTION_FULL, ignore.case = TRUE) |
      grepl("pbmc|peripheral blood mononuclear cell|peripheral blood mononuclear cells", DESCRIPTION_BRIEF, ignore.case = TRUE),
      "PBMC",
      ifelse(
        grepl("whole blood", DESCRIPTION_FULL, ignore.case = TRUE) |
        grepl("whole blood", DESCRIPTION_BRIEF, ignore.case = TRUE) |
        grepl("_BLOOD", STANDARD_NAME, ignore.case = TRUE)  ,
        "WHOLE BLOOD",
        "OTHER")),
    REGULATION = ifelse(
      grepl("up", DESCRIPTION_BRIEF, ignore.case = TRUE),
      "UP",
      ifelse(
        grepl("down", DESCRIPTION_BRIEF, ignore.case = TRUE),
        "DOWN",
        "UNKNOWN")))

#Reordenar colunas
VAX_Genesets = VAX_Genesets %>% select(SAMPLE_SOURCE, REGULATION, everything())

# Salvar
write.csv(VAX_Genesets, file = "VAXSigDB_Annotated.csv") #Continuar a anotação no Google Sheets

head(VAX_Genesets)
```

Converter tabela para tabela de genes

```{r}
############## Long table with Gene symbols

VAX_Genesets = VAXSigDB_Annotated

# Separar genes por linhas e somar numero de genes
VAX_Genes <- VAX_Genesets %>%
  mutate(GENE = GENE_SYMBOLS) %>% 
  separate_rows(GENE, sep = ",") %>% 
  select(GENE, everything()) %>%
  group_by(STANDARD_NAME) %>% #Somar e agrupar
  mutate(SetSize = n()) %>% 
  relocate(SetSize, .after = GENE_SYMBOLS)

VAX_GeneSets_annotated <- VAX_Genes %>%
  select(-GENE) %>% #Retirar coluna
  distinct() #Retirar linhas repetidas

#Salvar
write.csv(VAX_Genes, file = "VAX_Genes_Annotated_RAW.csv")
write.csv(VAX_GeneSets_annotated, file = "VAX_Genesets_Annotated_RAW.csv")

#Display
head(VAX_GeneSets_annotated)
```

Estatísticas

```{r}
### Descriptive statistics
vaccines = VAX_GeneSets_annotated %>% 
  group_by(VACCINE, PLATFORM, `TARGET_PATHOGEN_DISEASE`, MICROBE_TYPE) %>% 
  summarise(Count = n()) %>% 
  arrange(desc(Count))

pathogens = VAX_GeneSets_annotated %>% 
  group_by(`TARGET_PATHOGEN_DISEASE`, MICROBE_TYPE) %>% 
  summarise(Count = n()) %>% 
  arrange(desc(Count))

platforms = VAX_GeneSets_annotated %>% 
  group_by(PLATFORM) %>% 
  summarise(Count = n()) %>% 
  arrange(desc(Count))

```

### Visualização

By vaccines

```{r}
vaccine_hist = ggplot(vaccines) +
  aes(x = reorder(VACCINE, Count),
      y = Count,
      fill = `MICROBE_TYPE`) + 
  ggtitle("VAX MSigDB - Vaccines, by platform and target") +
  geom_col() +
  geom_text(aes(label = Count), vjust = 0, hjust = 2, size = 3) +
  coord_flip() +
  theme_minimal() + 
  labs(fill = "Vaccine target type") +
  theme(plot.title = element_text(face = "bold", colour = "black", size = (20)),
  legend.title = element_text(face = "bold"),
  axis.title = element_text(size = (10), colour = "black"),
  axis.text = element_text(colour = "black", size = (10))) +
  facet_wrap(vars(PLATFORM), scales = "free") + 
  scale_fill_manual(values = c('VIRUS'= "#d00000",
                               'BACTERIUM' = "#16E0BD",
                               'PARASITE' = "#118ab2",
                              'MENINGOCOCCUS'	= "#16E0BD",
                             'PNEUMOCOCCUS'	= "#44C199",
                             'TUBERCULOSIS'	= "#99e2b4",
                             'F TULARENSIS'	= "#90e0ef",
                             'LEISHMANIA'	= "#118ab2",
                             'MALARIA' = "#3f37c9",
                             'EBOV'	= "#d00000",
                             'HBV'	= "#e85d04",
                             'HBV, HAV'	= "#f48c06",
                             'HIV'	= "#a4133c",
                             'HPV'	= "#f27059",
                             'INFLUENZA'	= "#ffa69e",
                             'MEASLES'	= "#c65b7c",
                             'MMR'	= "#f9b5ac",
                             'SMALLPOX'	= "#ff8c42",
                             'VEEV'	= "#e01e37",
                             'VZV'	= "#ff9b85",
                             'YF'	= "#f38375"))


ggsave(vaccine_hist, file = "VAXSigDB_Vaccines_platforms_pathogens_1.png", width = 25, height = 8)
```

By platforms

```{r}
platforms_hist = ggplot(platforms) +
  aes(x = reorder(`PLATFORM`, Count),
      y = Count,
      fill = `PLATFORM`) + 
  ggtitle("VAX MSigDB - Vaccine platforms") +
  geom_col() +
  geom_text(aes(label = Count), vjust = 0, hjust = 0, size = 3) +
  coord_flip() +
  theme_minimal() + 
  labs(fill = "Platform") +
  theme(plot.title = element_text(face = "bold", colour = "black", size = (20)),
  legend.title = element_text(face = "bold"),
  axis.title = element_text(size = (10), colour = "black"),
  axis.text = element_text(colour = "black", size = (10))) +
  #facet_wrap(vars(PLATFORM), scales = "free") + 
  scale_fill_manual(values = c('IN'= "#1e6091",
                               'LA' = "#16E0BD",
                               'VV' = "#34a0a4",
                               'CONJ' = "#76c893",
                               'VLP' = "#b5e48c", 
                               'SU' = "#2ec4b6",
                               'LA and IV' = "#cbf3f0",
                               'SUB' = "#00bbf9",
                               'PEP' = "#c0fdff"))

ggsave(platforms_hist, file = "VAXSigDB_Platforms_hist_1.png", width = 10, height = 5)
```

By pathogens

```{r}

pathogens_hist = ggplot(pathogens) +
  aes(x = reorder(TARGET_PATHOGEN_DISEASE, Count),
      y = Count,
      fill = MICROBE_TYPE) + 
  ggtitle("VAX MSigDB - Vaccines, by platform and target") +
  geom_col() +
  geom_text(aes(label = Count), vjust = 0, hjust = 0, size = 3) +
  coord_flip() +
  theme_minimal() + 
  labs(fill = "Vaccine target type") +
  theme(plot.title = element_text(face = "bold", colour = "black", size = (20)),
  legend.title = element_text(face = "bold"),
  axis.title = element_text(size = (10), colour = "black"),
  axis.text = element_text(colour = "black", size = (10))) +
  #facet_wrap(vars(PLATFORM), scales = "free") + 
  scale_fill_manual(values = c('VIRUS'= "#d00000",
                               'BACTERIUM' = "#16E0BD",
                               'PARASITE' = "#118ab2",
                              'MENINGOCOCCUS'	= "#16E0BD",
                             'PNEUMOCOCCUS'	= "#44C199",
                             'TUBERCULOSIS'	= "#99e2b4",
                             'F TULARENSIS'	= "#90e0ef",
                             'LEISHMANIA'	= "#118ab2",
                             'MALARIA' = "#3f37c9",
                             'EBOV'	= "#d00000",
                             'HBV'	= "#e85d04",
                             'HBV, HAV'	= "#f48c06",
                             'HIV'	= "#a4133c",
                             'HPV'	= "#f27059",
                             'INFLUENZA'	= "#ffa69e",
                             'MEASLES'	= "#c65b7c",
                             'MMR'	= "#f9b5ac",
                             'SMALLPOX'	= "#ff8c42",
                             'VEEV'	= "#e01e37",
                             'VZV'	= "#ff9b85",
                             'YF'	= "#f38375"))

ggsave(pathogens_hist, file = "VAXSigDB_Pathogens_hist_1.png", width = 10, height = 5)

```

### Filtragem

```{r}
unique(VAX_GeneSets_annotated$STUDY_TYPE) 

<!-- #"CORRELATION" -->
<!-- #"VACCINE" -->
<!-- #"SUBSET" -->
<!-- #"CHALLENGE" -->

unique(VAX_GeneSets_annotated$STUDY_SUBTYPE)
<!-- # "ANTIBODY" -->
<!-- # "VAC ONLY" -->
<!-- # "BOOSTER" -->
<!-- # "VAC VS CTRL" -->
<!-- # "SIGNATURE" -->
<!-- # "ADVERSE EVENT" -->
<!-- # "COMPARISON" -->
<!-- # "RESPONDER VS NON RESPONDER" -->
<!-- # "AGE" -->
<!-- # "RESPONDER" -->
<!-- # "ADJUVANT" -->
<!-- # "NON RESPONDER" -->
<!-- # "ANTIMICROBIAL ACTIVITY" -->
<!-- # "CELLS" -->
<!-- # "CHALLENGE" -->
<!-- # "TOP DEGS" -->
<!-- # "PROTECTION" -->
<!-- # "TRAINING SET" -->
<!-- # "EXON ANALYSIS" -->
<!-- # "DOSE" -->
<!-- # "VAC VS OTHER" -->


<!-- # Subset de vacinas para as seguintes análises comparativas -->
VAX_GeneSets_Blood_PBMC_VAC = VAX_GeneSets_annotated %>% 
  filter(SAMPLE_SOURCE %in% c("PBMC", "WHOLE BLOOD"), 
         STUDY_TYPE %in% c("VACCINE"))

# Subset de vacinas com resultados de comparações entre vacinas, adjuvantes, idade e sexo
VAX_GeneSets_Blood_PBMC_CORR = VAX_GeneSets_annotated %>% 
  filter(SAMPLE_SOURCE %in% c("PBMC", "WHOLE BLOOD"), 
         STUDY_TYPE %in% c("CORRELATION"))

# Subset com resultados de correlações com resposta inata, celular e humoral
VAX_GeneSets_Blood_PBMC_SUBSETS = VAX_GeneSets_annotated %>% 
  filter(SAMPLE_SOURCE %in% c("PBMC", "WHOLE BLOOD"), 
         STUDY_TYPE %in% c("SUBSET"))


# Subset de vacinas com genes associados a eventos adversos
VAX_GeneSets_Blood_PBMC_CHALLENGE = VAX_GeneSets_annotated %>% 
  filter(SAMPLE_SOURCE %in% c("PBMC", "WHOLE BLOOD"), 
         STUDY_TYPE %in% c("CHALLENGE"))

VAX_GeneSets_Blood_PBMC_VAC.stats = VAX_GeneSets_Blood_PBMC_VAC %>% 
  group_by(VACCINE, PLATFORM, REGULATION, TARGET_PATHOGEN_DISEASE, MICROBE_TYPE) %>% 
  summarise(Count = n()) %>% 
  arrange(desc(Count))

```


*Qual a relação entre os genes de cada dataset?*

```{r}
############### VENN DIAGRAM
#Input
data = VAX_GeneSets_Blood_PBMC_VAC %>%
  as.tibble() %>% 
  separate_rows(GENE_SYMBOLS, sep = ",") %>% 
  select(VACCINE, GENE_SYMBOLS, REGULATION) %>% 
  as.data.frame() %>% 
  mutate(VACCINE_REG = paste0(VACCINE, sep="_", REGULATION))
  
```


```{r}
# Função para calcular a sobreposição entre pares de vacinas e a porcentagem de genes compartilhados
overlap_genes <- function(cond1, cond2, data) {
  genes_cond1 <- data$GENE_SYMBOLS[data$VACCINE_REG == cond1] #Trocar coluna
  genes_cond2 <- data$GENE_SYMBOLS[data$VACCINE_REG == cond2] #Trocar coluna
  genes_shared <- intersect(genes_cond1, genes_cond2)
  genes_notshared_cond1 <- setdiff(genes_cond1, genes_cond2)
  genes_notshared_cond2 <- setdiff(genes_cond2, genes_cond1)
  
  total_genes_cond1 <- length(genes_cond1)
  total_genes_cond2 <- length(genes_cond2)
  
  percentage_shared_cond1 <- length(genes_shared) / total_genes_cond1 * 100
  percentage_shared_cond2 <- length(genes_shared) / total_genes_cond2 * 100
  
  shared_genes <- data.frame(
    Cond1 = cond1,
    Cond2 = cond2,
    Shared = length(genes_shared),
    NotShared_cond1 = length(genes_notshared_cond1),
    NotShared_cond2 = length(genes_notshared_cond2),
    Total_Genes_Cond1 = total_genes_cond1,
    Total_Genes_Cond2 = total_genes_cond2,
    Genes_Names = paste(genes_shared, collapse = ", "),
    Percentage_Shared_Cond1 = percentage_shared_cond1,
    Percentage_Shared_Cond2 = percentage_shared_cond2
  )
  
  return(shared_genes)
}
# Obtenha uma lista de todas as vacinas únicas
unique_cond <- unique(data$VACCINE_REG)  #trocar "study" pelo nome da coluna de interesse

# Inicialize um dataframe para armazenar os resultados
shared_genes_df <- data.frame()

# Calcular a sobreposição e porcentagem para cada par de vacinas
for (i in 1:(length(unique_cond) - 1)) {
  for (j in (i + 1):length(unique_cond)) {
    resultado_temp <- overlap_genes(unique_cond[i], unique_cond[j], data)
    shared_genes_df <- bind_rows(shared_genes_df, resultado_temp)
  }
}

# Função para realizar o teste exato de Fisher
fisher_exact_test <- function(shared, notshared1, notshared2) {
  cont_table <- matrix(c(shared, notshared1, notshared2, 0), nrow = 2)
  results_fisher <- fisher.test(cont_table)
  return(results_fisher$p.value)
}

# Aplicar a função a cada linha da tabela de resultados
shared_genes_df$pvalue <- mapply(fisher_exact_test, 
                                    shared_genes_df$Shared, 
                                    shared_genes_df$NotShared_cond1, 
                                    shared_genes_df$NotShared_cond2)


# Calcular -log(p)
shared_genes_df$`-log(p)`= -log(shared_genes_df$pvalue, 10)

#Duplicar e trocar colunas (Vacina 1 e Vacina 2)
shared_genes_df2 = shared_genes_df
shared_genes_df2[, c("Cond1", "Cond2")] <- shared_genes_df2[, c("Cond2", "Cond1")]

#Unir datasets com os dados espelhados
shared_genes_df_mirror = rbind(shared_genes_df, shared_genes_df2)

#Converter NA, NaNe inf em 0
shared_genes_df_mirror[is.na(shared_genes_df_mirror)] <- 0
shared_genes_df_mirror<- replace(shared_genes_df_mirror, shared_genes_df_mirror == "Inf", 0)

#Arredondar porcentagens
shared_genes_df_mirror = shared_genes_df_mirror %>%
  mutate(Percentage_Shared_Cond1 = round(Percentage_Shared_Cond1, 2),
         Percentage_Shared_Cond2 = round(Percentage_Shared_Cond2, 2))

#Salvar resultados
write.csv(shared_genes_df_mirror, file = "VAXSigDB_sharedgenes.csv") #Alterar

```

Heatmap
```{r}

matrix = shared_genes_df_mirror %>% 
  filter(pvalue <=0.10) %>% 
  select(Cond1, Cond2, Percentage_Shared_Cond1) %>% #Percentage_Shared_Cond1
  filter(grepl("DOWN", Cond1) & grepl("DOWN",Cond2)) %>% 
  pivot_wider(names_from = "Cond2", 
              values_from = "Percentage_Shared_Cond1",  #Percentage_Shared_Cond1
              values_fn = mean) %>% 
  replace(is.na(.), 0) %>% 
  column_to_rownames(var="Cond1") %>% 
  as.matrix() %>% 
  round(0)


col_fun <- colorRamp2(c(0, 100), c("white", "red"))

#Plotar

file = "VAXSig_SharedGenes_perc_DOWN-DOWN"
fileimageheatmap_grouped= paste0(file, sep ="_", "Complexheatmap.png")

png(file = fileimageheatmap_grouped, width=50, height=50,units="cm",res=800)

heatmap_plot = Heatmap(matrix,
                        top_annotation = NULL,
                        left_annotation = NULL,
                        cell_fun = function(j, i, x, y, width, height, fill) {
        grid.text(sprintf("%.1f", matrix[i, j]), x, y, gp = gpar(fontsize = 5))}, #Mostrar valores nas células
                        show_row_names = TRUE,
                        row_names_side = "left",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        col = col_fun, #color
                        # column_title = "Target", 
                        # row_title = "Source",
                        column_title_gp = gpar(fontsize = 15, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = gpar(fontsize = 12),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = gpar(fontsize = 12), #Grande = 6, Pequeno 15
                        row_dend_width = unit(5, "cm"), #Altura do dendrograma
                        row_split = NULL, #Split row clusters (NULL se for um gene set mais específico)
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(8, "mm"),
                        show_column_dend = FALSE,
                        show_row_dend = FALSE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = unit(40, "cm"), # 20 para pequeno e grande
                        height = unit(40, "cm") # 20 para pequeno, 60 para grande
)

draw(heatmap_plot, 
     annotation_legend_side = "left", 
     heatmap_legend_side = "right")
dev.off()

```


# ORA e GSEA

O ORA e GSEA podem ser pouco sensíveis quando o pvalue/padj \< 5%. Na documentação, recomendam usar o cutoff de padj (FDR) de 25% para incluir gene sets que estejam 75% corretos - isto é, em 4 observações, 3 estão corretas -, o que aumenta o número de resultados interessantes. Para verificar se são realmente significantes, visualize no heatmap com labels de \*significancia.

Referencia: Link{<https://www.gsea-msigdb.org/gsea/doc/GSEAUserGuideFrame.html>}

##Input

Insira a tabela longa com genes filtrados para up- ou down-regulated gerados na etapa de preparação dos dados.

## Manual
```{r}
######### INPUT AQUI ######### 
# all_degs_p_05_vac_infected = all_degs_p_05_vac_infected_27_11_23 %>%
#   mutate(condition = case_when(condition == "H-BNT-I (D1)" ~ "BNT-I (D1)",
#                                condition == "H-BNT-I (D3)" ~ "BNT-I (D3)",
#                                condition == "H-BNT-I (D4)" ~ "BNT-I (D4)",
#                                condition == "H-I-I (D2)" ~ "I-I (D2)",
#                                condition == "I-BNT-I (D2)" ~ "I-BNT-I (D2)",
#                                condition == "I-BNT-I (D5)" ~ "I-BNT-I (D5)",
#                                condition == "H-BNT-I (D10, mild)" ~ "BNT-I (D10, mild)",
#                                condition == "H-BNT-I (D10, severe)" ~ "BNT-I (D10, severe)",
#                                condition == "H-BNT-I (D26, mild)" ~ "BNT-I (D26, mild)",
#                                condition == "H-BNT-I (D26, severe)" ~ "BNT-I (D26, severe)",
#                                condition == "H-I (D10, moderate)" ~ "I (D10, moderate)",
#                                condition == "H-I (D10, severe)" ~ "I (D10, severe)",
#                                condition == "H-I (D26, moderate)" ~ "I (D26, moderate)",
#                                condition == "H-I (D26, severe)" ~ "I (D26, severe)",
#                                condition == "H-I (D51, moderate)" ~ "I (D51, moderate)",
#                                condition == "H-I (D51, severe)" ~ "I (D51, severe)",
#                                TRUE ~ condition))
# 
# write.csv(all_degs_p_05_vac_infected_27_11_23, file ="all_degs_p_05_vac_infected_29_11_23.csv")

all_degs_p_05_vac_infected

#GSE199750
# degs <- filter(all_degs_p_05_vac_infected, condition == "BNT (V1, D6)") #Não enriquecido
# degs <- filter(all_degs_p_05_vac_infected, condition == "BNT (V2, D1)") #Enriquecido
# degs <- filter(all_degs_p_05_vac_infected, condition == "BNT (V3, D1)") #Enriquecido
# degs <- filter(all_degs_p_05_vac_infected, condition == "BNT-MO (V3, D1)") #Enriquecido
# degs <- filter(all_degs_p_05_vac_infected, condition == "ChAd (V1, D6)") #Enriquecido
# degs <- filter(all_degs_p_05_vac_infected, condition == "ChAd (V2, D1)") #Enriquecido
# degs <- filter(all_degs_p_05_vac_infected, condition == "ChAd-BNT (V3, D1)") #Não enriquecido

#GSE201533
# degs = filter(all_degs_p_05_vac_infected, condition == "ChAd (V1, D3)") #Enriquecido
# degs = filter(all_degs_p_05_vac_infected, condition == "ChAd (V1, D7)") #Enriquecido
# degs = filter(all_degs_p_05_vac_infected, condition == "ChAd (V2, D3)") #Não enriquecido
# degs = filter(all_degs_p_05_vac_infected, condition == "ChAd (V2, D7)") #Enriquecido
# degs = filter(all_degs_p_05_vac_infected, condition == "ChAd-BNT (V2, D3)") #Enriquecido
# degs = filter(all_degs_p_05_vac_infected, condition == "ChAd-BNT (V2, D7)") #Enriquecido

#GSE206023
# degs = filter(all_degs_p_05_vac_infected, condition == "BBIBP (V3, D07)") #Não enriquecido
# degs = filter(all_degs_p_05_vac_infected, condition == "BBIBP (V3, D14)") #Não enriquecido
# degs = filter(all_degs_p_05_vac_infected, condition == "BBIBP (V3, D28)") #Não enriquecido
# degs = filter(all_degs_p_05_vac_infected, condition == "ZF2001 (V3, D07)") #Enriquecido
# degs = filter(all_degs_p_05_vac_infected, condition == "ZF2001 (V3, D14)") #Não eEnriquecido
# degs = filter(all_degs_p_05_vac_infected, condition == "BBIBP (V3, D14)") #Não enriquecido
# degs = filter(all_degs_p_05_vac_infected, condition == "BBIBP (V3, D28)") #Enriquecido
# degs = filter(all_degs_p_05_vac_infected, condition == "ZF2001 (V3, D07)") #Enriquecido
# degs = filter(all_degs_p_05_vac_infected, condition == "ZF2001 (V3, D14)") #Enriquecido
# degs = filter(all_degs_p_05_vac_infected, condition == "ZF2001 (V3, D28)") #Enriquecido

#GSE201530
# degs = filter(all_degs_p_05_vac_infected, condition == "BNT-I (D1)")
# degs = filter(all_degs_p_05_vac_infected, condition == "BNT-I (D3)")
# degs = filter(all_degs_p_05_vac_infected, condition == "BNT-I (D4)")
# degs = filter(all_degs_p_05_vac_infected, condition == "I-I (D2)")
# degs = filter(all_degs_p_05_vac_infected, condition == "I-BNT-I (D2)")
# degs = filter(all_degs_p_05_vac_infected, condition == "I-BNT-I (D5)")
# degs = filter(all_degs_p_05_vac_infected, condition == "I-I (D1)")
# degs = filter(all_degs_p_05_vac_infected, condition == "I-I (D3)")
# degs = filter(all_degs_p_05_vac_infected, condition == "I-I (D5)")


#GSE189039
# degs = filter(all_degs_p_05_vac_infected, condition == "BNT-I-BNT (D51, mild)")
# degs = filter(all_degs_p_05_vac_infected, condition == "BNT-I-BNT (D51, severe)")
# degs = filter(all_degs_p_05_vac_infected, condition == "BNT-I (D10, mild)")
# degs = filter(all_degs_p_05_vac_infected, condition == "BNT-I (D10, severe)")
# degs = filter(all_degs_p_05_vac_infected, condition == "BNT-I (D26, mild)")
# degs = filter(all_degs_p_05_vac_infected, condition == "BNT-I (D26, severe)")
# degs = filter(all_degs_p_05_vac_infected, condition == "I (D10, moderate)")
# degs = filter(all_degs_p_05_vac_infected, condition == "I (D10, severe)")
# degs = filter(all_degs_p_05_vac_infected, condition == "I (D26, moderate)")
# degs = filter(all_degs_p_05_vac_infected, condition == "I (D26, severe)")
# degs = filter(all_degs_p_05_vac_infected, condition == "I (D51, moderate)")
# degs = filter(all_degs_p_05_vac_infected, condition == "I (D51, severe)")

```


```{r}


# Select columns and rank
degs = degs %>% 
  select(condition, genes, log2fold_change) %>% 
  distinct() %>% 
  mutate(rank = rank(log2fold_change,  ties.method = "random")) %>% 
  arrange(desc(rank))

#Definir nome para arquivos gerados
file_name = degs %>% 
  select(condition) %>% 
  distinct() %>% 
  as.character()

# Preparar gene lists
gene_list_ora <- degs$genes
gene_list_gsea = degs %>% select(genes, log2fold_change) %>% as.list()

######### DEGs para GSEA ######### 
degs_gene_list_lf2c <- degs$log2fold_change
names(degs_gene_list_lf2c) <- degs$genes
degs_gene_list_lf2c<-na.omit(degs_gene_list_lf2c)
degs_gene_list_lf2c = degs_gene_list_lf2c %>% sort(decreasing = TRUE)

######### DEGs

############ IMMUNE GO

immune_GO_gsea <- GSEA(geneList =  #Definir nome para arquivos gerados
file_name = degs %>% 
  select(condition) %>% 
  distinct() %>% 
  as.character()

######### DEGs para ORA ######### 
gene_list_ora <- degs$genes
######### DEGs para GSEA ######### 
degs_gene_list_lf2c <- degs$log2fold_change
names(degs_gene_list_lf2c) <- degs$genes
degs_gene_list_lf2c<-na.omit(degs_gene_list_lf2c)
degs_gene_list_lf2c = sort(degs_gene_list_lf2c, decreasing = TRUE)

############ IMMUNE GO
Immune_GO_T2G = ImmuneGO_Annotated_Genes_Nested_Interesting %>% 
  select(gene_set_short, genes)

immune_GO_gsea <- GSEA(degs_gene_list_lf2c, 
                       TERM2GENE = Immune_GO_T2G, 
                       minGSSize = 12,
                       maxGSSize = 1000,
                       pvalueCutoff = 0.25,
                       pAdjustMethod = "BH") %>% 
  as.data.frame() %>% 
  arrange(qvalue) %>% 
  mutate(condition = file_name,
         gsea_enrichment = "ImmuneGO")

#Salvar tabela
write.csv(immune_GO_gsea, file = paste(file_name, "ImmuneGO_GSEA.csv", sep = "_"))

########### CELL MARKER

CellMarker_genes = CellMarker_ImmuneCells %>% 
  select(cell_name, marker)

cell_types_gsea_immune <- GSEA(degs_gene_list_lf2c, 
                       TERM2GENE = CellMarker_genes, 
                       minGSSize = 1,
                       maxGSSize = 1000,
                       pvalueCutoff = 0.25, 
                       pAdjustMethod = "BH") %>% 
  as.data.frame() %>% 
  arrange(qvalue) %>% 
  mutate(condition = file_name,
         gsea_enrichment = "CellMarker") 

#Salvar tabela
write.csv(cell_types_gsea_immune, file=paste(file_name, "CellMarker_GSEA.csv", sep = "_"))

```

## ORA Automatizado

```{r}
# Função para realizar a análise GSEA para uma condição específica
perform_ORA <- function(df, Immune_GO_T2G) {
  resultados <- list()
  
  # Obter condições únicas na coluna "condition"
  condicoes <- df$condition %>% 
    unique() %>% 
    as.character()
  
  for (condicao in condicoes) {
    # Selecionar DEGs para a condição atual
    degs_condicao <- df %>% 
      filter(condition == condicao) %>%
      select(genes) %>%
      distinct()

    # Preparar gene list
    gene_list_ora <- degs_condicao$genes

    ############ IMMUNE GO
    immune_GO_ORA <- tryCatch({
      enricher(gene_list_ora, 
               TERM2GENE = Immune_GO_T2G, 
               minGSSize = 1,
               maxGSSize = 1000,
               pvalueCutoff = 0.25,
               pAdjustMethod = "BH") %>% 
        as.data.frame() %>% 
        arrange(qvalue) %>% 
        mutate(condition = condicao,
               ora_enrichment = "ImmuneGO")
    }, error = function(e) {
      return(NULL)  # Retorna NULL caso ocorra o erro
    })

    if (!is.null(immune_GO_ORA)) {
      resultados[[paste(condicao, "ImmuneGO", sep = "_")]] <- immune_GO_ORA
    }
  }

  # Combinar todos os resultados em um único dataframe
  resultado_final <- bind_rows(resultados)

  return(resultado_final)
}

# Exemplo de uso
df_resultados <- perform_ORA(degs_all_updown_vax_immunego, Immune_GO_T2G)

# Visualizar os resultados
print(df_resultados)

#Anotar
VAX_immune_GO_ORA = df_resultados %>% 
  clean_names() %>% 
  select(process = id, gene_ratio:ora_enrichment) %>% 
  inner_join(VAX_Genes_Annotated_RAW %>% clean_names() %>% select(condition:time, gene_symbols, study_subtype) %>%   filter(study_subtype == "VAC ONLY") %>% distinct(), by = "condition") %>% 
  distinct() %>% 
  select(process:ora_enrichment) %>% 
  distinct() %>% 
  mutate(log_q = -log10(qvalue))



write.csv(VAX_immune_GO_ORA, file = "VAX_immune_GO_ORA.csv")

view(VAX_immune_GO_ORA)
```


## GSEA Automatizado 
```{r}

autoGSEA <- function(df, TERM2GENE, geneset_name) {
  resultados <- list()

  condicoes <- df$condition %>% 
    unique() %>% 
    as.character()

  for (condicao in condicoes) {
    # Selecionar DEGs para a condição atual
    degs_condicao <- df %>% 
      filter(condition == condicao) %>%
      select(genes, log2fold_change) %>%
      distinct() %>%
      mutate(rank = rank(log2fold_change, ties.method = "random")) %>%
      arrange(desc(rank))

    # Garantir que log2fold_change seja um vetor
    gene_list_lf2c <- as.vector(degs_condicao$log2fold_change)
    names(gene_list_lf2c) <- degs_condicao$genes
    gene_list_lf2c <- na.omit(gene_list_lf2c)
    gene_list_lf2c <- sort(gene_list_lf2c, decreasing = TRUE)

    ############ IMMUNE GO
    auto_gsea <- tryCatch({
      GSEA(geneList = gene_list_lf2c,
           TERM2GENE = TERM2GENE,
           minGSSize = 1,
           maxGSSize = 1000,
           pvalueCutoff = 0.25,
           pAdjustMethod = "BH") %>%
        as.data.frame() %>%
        arrange(qvalue) %>%
        mutate(condition = condicao,
               gsea_enrichment = "ImmuneGO")
    }, error = function(e) {
      return(NULL)  # Retorna NULL caso ocorra o erro
    })

    if (!is.null(auto_gsea)) {
      resultados[[paste(condicao, geneset_name, sep = "_")]] <- auto_gsea
    }
  }

  # Combinar todos os resultados em um único dataframe
  resultado_final <- bind_rows(resultados)

  return(resultado_final)
}


TERM2GENE = CellMarker_genes
geneset_name = "CellMarker Leukocyte"



# Chamar a função para executar as análises para cada valor da coluna "condition"
all_degs_p_05_vac_infected_GSEA_ALL <- all_degs_p_05_vac_infected %>%
  select(condition, genes, log2fold_change) %>% 
  autoGSEA(TERM2GENE = TERM2GENE, 
           geneset_name = geneset_name) 

all_degs_p_05_vac_infected_GSEA_cellmarker = all_degs_p_05_vac_infected_GSEA_ALL %>% 
  rownames_to_column("delete") %>% 
  select(condition, ID, everything()) %>% 
  clean_names()

view(all_degs_p_05_vac_infected_GSEA_ALL)

#Salvar
write.csv(all_degs_p_05_vac_infected_GSEA_ALL, 
          file = "all_degs_p_05_vac_infected_CellMarker_GSEA_ALL_17-1-24.csv")

view(all_degs_p_05_vac_infected_GSEA_ALL)
```





# ssGSEA

Corto package

```{r}
#Expression matrix
#Ordenar colunas
ssgsea_gene = gse199750_counts_ready %>% as.data.frame()
ssgsea_gene.t = t(ssgsea_gene) %>% as.data.frame()
ssgsea_gene.t$sample = rownames(ssgsea_gene.t)
ssgsea_gene.t = ssgsea_gene.t %>% relocate(Sample, .before=TSPAN6) %>% arrange(Sample)
rownames(ssgsea_gene.t) = ssgsea_gene.t$sample 
ssgsea_gene.t = ssgsea_gene.t %>% select(-Sample)
ssgsea_gene = t(ssgsea_gene.t) %>% as.data.frame()

#Anotar com metadados
gse199750_metadata = arrange(gse199750_metadata, Sample)
gse199750_metadata$Sample = rownames(gse199750_metadata)
sample_names = gse199750_metadata %>% 
  select(Sample) %>% 
  as.data.frame() %>% 
  rename(sample = '.')

#Gene sets list
Immune_GO_genes_GO_Filt_GSEA = Immune_GO_genes_GO_Filt %>% select(GO.NAME, SYMBOL)
genelist <- split(Immune_GO_genes_GO_Filt_GSEA$SYMBOL, Immune_GO_genes_GO_Filt_GSEA$GO.NAME) 

# Run ssGSE
nesmat = ssgsea(ssgsea_gene, genelist)
nesmat = nesmat %>% as.data.frame()
nesmat.t = nesmat %>% t() %>% as.data.frame()
nesmat.t$Sample = rownames(nesmat.t) 
nesmat.t = nesmat.t %>% relocate(Sample, .before = `ACTIVATION OF INNATE IMMUNE RESPONSE`)
nesmat.t = cbind(sample_names, nesmat.t)
rownames(nesmat.t) = nesmat.t$Sample
nesmat.t = nesmat.t %>% select(-Sample)
nesmat = nesmat.t %>% t() %>% as.data.frame()

# Padronizar tabela
ssgsea_result = nesmat %>% as.data.frame() 
ssgsea_result$GeneSet = rownames(ssgsea_result)
ssgsea_result = ssgsea_result %>% relocate(GeneSet, .before=everything())
ssgsea_result = pivot_longer(ssgsea_result, cols = -GeneSet, names_to = "Sample", values_to = "NES")

#Converter NES em pvalue, ajustar para qvalue e calcualr -log(q)
ssgsea_result$pvalue = z2p(ssgsea_result$NES)
ssgsea_result$qvalue = p.adjust(ssgsea_result$pvalue, method = "BH")
ssgsea_result = ssgsea_result %>% mutate(`-log(q)` = - log10(qvalue))

#Unir tabelas
ssgsea_result_metadata = merge(ssgsea_result, gse199750_metadata, by="Sample", all.x=T) #Unir com metadata
ssgsea_result_metadata = ssgsea_result_metadata %>% rename(Process = GeneSet)
ssgsea_result_metadata$Study = "GSE199750"
ssgsea_result_metadata_ImmuneGO_ann = merge(ssgsea_result_metadata, ImmuneGO_Annotated_GO, by = "Process", all.x=T)

#Ordenar tabela
ssgsea_result_metadata_ImmuneGO_ann = ssgsea_result_metadata_ImmuneGO_ann %>% select(Study, Sample, FirstSecondDose, ThirdVaccine, Timepoint, Age, Sex, Process, everything())
ssgsea_result_metadata_ImmuneGO_ann = ssgsea_result_metadata_ImmuneGO_ann %>% select(-SampleN, -Tissue, -Run, -`...1`, -SubjectID)

# Organizar tabela
ssgsea_result_metadata_ImmuneGO_ann$FirstSecondDose <- gsub("ChAdOx1 nCoV-19", "ChAdOx1", ssgsea_result_metadata_ImmuneGO_ann$FirstSecondDose)

#Unificar coluna de FirstSecondDose e ThirdVaccine
# ssgsea_result_metadata_ImmuneGO_ann = ssgsea_result_metadata_ImmuneGO_ann %>%
#   mutate(Condition = paste(FirstSecondDose, ThirdVaccine, sep = "_"), FirstSecondDose) %>% 
#   mutate(Condition = gsub("_NA|_N/A|_Missing", "", Condition)) %>% 
#   relocate(Condition, .before = FirstSecondDose)

ssgsea_result_metadata_ImmuneGO_ann_union <- ssgsea_result_metadata_ImmuneGO_ann %>%
  mutate(
    Condition = ifelse(Timepoint == "V3A", paste(FirstSecondDose, ThirdVaccine, sep = "-"), paste(FirstSecondDose)),
    Condition = gsub("ChAdOx1", "ChAd", Condition),
    Condition = gsub("BNT162b2", "BNT", Condition),
    Condition = gsub("BNT-BNT", "BNT", Condition),
    Condition = gsub("mRNA-1273", "MO", Condition),
    TimepointDay = ifelse(Timepoint == "V1", "(V1, D6)", 
                         ifelse(Timepoint == "V2A", "(V2, D1)", 
                                ifelse(Timepoint == "V3A", "(V3, D1)", "(V0)")))
  ) %>% 
  mutate(
    Condition = paste(Condition, TimepointDay, sep = " ")
  )


#Anotar vacinas
ann_vaccines$Condition = rownames(ann_vaccines)

ssgsea_result_metadata_ImmuneGO_ann_union_ann = merge(ssgsea_result_metadata_ImmuneGO_ann_union, ann_vaccines, by = "Condition", all.x = TRUE, all.y = F)


View(ssgsea_result_metadata_ImmuneGO_ann_union_ann)

#Salvar
write.csv(ssgsea_result_metadata_ImmuneGO_ann_union_ann, file = "GSE199750_ssgsea_result.csv")
```

### Visualização

```{r}
test = ssgsea_result_metadata_ImmuneGO_ann_union_ann %>% filter(Process %in% c("ADAPTIVE IMMUNE RESPONSE","INNATE IMMUNE RESPONSE", "IMMUNOGLOBULIN MEDIATED IMMUNE RESPONSE", "T CELL ACTIVATION"))

esquisser(test)

ggplot(test) +
 aes(x = TimepointDay, y = NES, fill = Condition) +
 geom_boxplot() +
 scale_fill_manual(values = c(`BNT (V1, D6)` = "#6DCBF8", 
`BNT (V2, D1)` = "#6DCBF8", `BNT (V3, D1)` = "#6DCBF8", `BNT-MO (V3, D1)` = "#EDEF0F", `ChAd (V1, D6)` = "#0BE747", 
`ChAd (V2, D1)` = "#0BE747", `ChAd (V3, D1)` = "#0BE747", `ChAd-BNT (V3, D1)` = "#61FF94")) +
 labs(title = "ImmuneGO ssGSEA - GSE199750", 
 fill = "Vaccine") +
 theme_bw() +
stat_compare_means() +
 facet_wrap(vars(Process))

#Boxplot (NES)
NES_plot = ggplot(test) +
 aes(x = TimepointDay, 
     y = NES, 
     fill = Condition) +
 geom_boxplot() +
 scale_fill_manual(values = c(`BNT (V1, D6)` = "#6DDAF8", 
                              `BNT (V2, D1)` = "#6DDAF8", 
                              `BNT (V3, D1)` = "#6DDAF8", 
                              `BNT-MO (V3, D1)` = "#F8A06D", 
                              `ChAd (V1, D6)` = "#44F307", 
                              `ChAd (V2, D1)` = "#44F307", 
                              `ChAd (V3, D1)` = "#44F307", 
                              `ChAd-BNT (V3, D1)` = "#46F1CF")) +
 labs(title = "ImmuneGO ssGSEA - GSE199750", 
      fill = "Vaccine") +
 theme_bw() +
 theme(plot.title = element_text(size = 20L, face = "bold", hjust = 0.5)) +
 facet_wrap(vars(Process))

#Salvar
ggsave(NES_plot, file= "GSE199750_ssGSEA_Adap_Inn_Ig_Tcell-NES.png", width = 10, height = 6)

#Boxplot (-log(q))
logq_plot = ggplot(test) +
 aes(x = TimepointDay, 
     y = `-log(q)`, 
     fill = Condition) +
 geom_boxplot() +
 scale_fill_manual(values = c(`BNT (V1, D6)` = "#6DDAF8", 
                              `BNT (V2, D1)` = "#6DDAF8", 
                              `BNT (V3, D1)` = "#6DDAF8", 
                              `BNT-MO (V3, D1)` = "#F8A06D", 
                              `ChAd (V1, D6)` = "#44F307", 
                              `ChAd (V2, D1)` = "#44F307", 
                              `ChAd (V3, D1)` = "#44F307", 
                              `ChAd-BNT (V3, D1)` = "#46F1CF")) +
 labs(title = "ImmuneGO ssGSEA - GSE199750", 
      fill = "Vaccine") +
 theme_bw() +
 theme(plot.title = element_text(size = 20L, face = "bold", hjust = 0.5)) +
 facet_wrap(vars(Process))

logq_plot

ggsave(logq_plot, file= "GSE199750_ssGSEA_Adap_Inn_Ig_Tcell_-LOG(Q).png", width = 10, height = 6)


```

### ANOVA

```{r}
esquisser(test)

library(ggsignif)
library(rstatix)

df = test
df$Condition <- as.factor(df$Condition)

stat.test <- df %>%
  t_test(Condition ~ NES) %>%
  add_significance()


test %>%
  filter(Process %in% "INNATE IMMUNE RESPONSE") %>%
  ggboxplot(x = "Process", y = "NES",
          color = "Condition", palette = "jco") +
  facet_wrap(vars(Vaccine)) +
  stat_compare_means(method = "anova", label.y = 4) +      # Add global p-value
  stat_compare_means(label = "p.signif", method = "t.test",
                     ref.group = ".all.")    # Comparação múltipla usando Tukey HSD


# Box plots with p-values
bxp <- ggboxplot(df, x = "Process", y = "NES", fill = "#00AFBB")
stat.test <- stat.test %>% add_xy_position(x = "supp")
bxp + 
  stat_pvalue_manual(stat.test, label = "p") +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1)))

```




# Heatmap

Processamento de tabelas Extraindo genes de enriched gene sets

```{r}
#Abrir tabela de GSEA
dados = DEGs_All_studies_ImmuneGO_GSEA_1_

dados_long <- dados %>%
  mutate(core_enrichment = strsplit(core_enrichment, "/")) %>%  # Divida a coluna em listas de genes
  unnest(core_enrichment)  # Transforme a lista em várias linhas

#Filtrar genes por categorias
genes_adaptive <- dados_long %>%
  filter(`Immune system` == "Adaptive")
```

Calculando media, soma etc. por categorias em uma tabela longa

```{r}
dados = DEGs_All_studies_ImmuneGO_GSEA_1_
# Selecione as colunas de interesse
names(dados)

# Calcular a soma dos valores da coluna NES para cada combinação de Vaccine e Immune system
resultado <- dados %>%
  group_by(Vaccine, `Immune system`) %>%
  summarise(`NES sum` = sum(NES))
```

Transformando tabela de DEGs por coluna em tabela longa

```{r}
#Converter tabela de DEGs Up and Down em tabela longa

degs_all = DEGs_All_studies_DEGs_All_Studies
degs_all_long <- pivot_longer(degs_all, cols = everything(), names_to = "Vaccine", values_to = "Genes")

#Salvar
write.csv(degs_all_long, file="degs_all_long.csv")
```

Transformando tabela longa em tabela wide para pca e heatmap

Shared Genes

```{r}
#Selecionar colunas de comparação e valores. Trocar value.var pela variável de interesse ("Shared", "-log(p)")

sharedDEGs_DOWN_fisher_sig = sharedDEGs_DOWN_fisher %>% filter()

matrix_data <- dcast(sharedDEGs_DOWN_fisher, `Cond1` ~ `Cond2`, value.var = "Percentage_Shared_Cond1", fun.aggregate = mean)

#Converter coluna 1 em rownames e excluir
matrix_data$Cond1 = as.factor(matrix_data$Cond1)
rownames(matrix_data) <- matrix_data$Cond1
matrix_data <- matrix_data %>% select(-`Cond1`)

#Arredondar valores para facilitar visualização
matrix_data <- round(matrix_data, 1)
matrix_data <- replace(matrix_data, matrix_data == "NaN", 0)

#Salvar
write.csv(matrix_data, file = "heatmap_sharedDEGs_UP_shared_percent.csv")

#Visualizar
heatmaply(matrix_data, 
          main = "Shared DEGs - UP-DOWN - %identity", 
          cellnote = matrix_data, 
          cellnote_size = 8,
          cellnote_textposition = "middle right",
  scale_fill_gradient_fun = ggplot2::scale_fill_gradient2(
    low = "white", 
    high = "red", 
    midpoint = 1, 
    limits = c(0, 60)
  ))
```

Down vs UP only

```{r}
####### DOWN vs UP only ######## 
filtered_data <- shared_genes_df_mirror %>%
  filter((grepl("UP", Cond1) & grepl("DOWN", Cond2)) | (grepl("DOWN", Cond1) & grepl("UP", Cond2)))

matrix_data <- dcast(filtered_data, `Cond1` ~ `Cond2`, value.var = "Percentage_Shared_Cond1", fun.aggregate = mean)

#Converter coluna 1 em rownames e excluir
rownames(matrix_data) <- matrix_data$`Cond1`
matrix_data <- matrix_data %>% select(-`Cond1`)

#Arredondar valores para facilitar visualização
matrix_data <- round(matrix_data, 1)
matrix_data <- replace(matrix_data, matrix_data == "NaN", 0)

#Arredondar valores para facilitar visualização
matrix_data <- round(matrix_data, 1)

#Salvar
write.csv(matrix_data, file = "heatmap_sharedDEGs_UP_vs_DOWN_shared_percent.csv")

#Visualizar
heatmaply(matrix_data, 
          main = "Shared DEGs - UP vs. DOWN - %identity", 
          cellnote = matrix_data, 
          cellnote_size = 8,
          cellnote_textposition = "middle right",
  scale_fill_gradient_fun = ggplot2::scale_fill_gradient2(
    low = "white", 
    high = "red", 
    midpoint = 1, 
    limits = c(0, 100)
  )
)

```


# Circos heatmap

# Gráfico de barras

Innate

```{r}
#Definir conjunto de dados
Immune_GO_GSEA$`-log(qvalue)` = as.numeric(Immune_GO_GSEA$`-log(qvalue)`)

# Filtrar as linhas onde a coluna "Gene set" está em valores_desejados
GO_set_innate <- Immune_GO_GSEA %>% filter(`Gene set short` %in% c("INNATE IMMUNE RESPONSE")) %>% mutate(`-log(qvalue)` = as.numeric(`-log(qvalue)`))

GO_set_innate = merge(GO_set_innate, ann_vaccines_complex, by.x = "Vaccine", by.y = "Condition")
esquisser(GO_set_innate)

GO_set_innate_bars = ggplot(GO_set_innate) +
  aes(
    x = Day,
    fill = `Vaccine group`,
    colour = Vaccine,
    weight = NES
  ) +
  geom_bar(position = "dodge") +
  scale_fill_viridis_d(option = "magma", direction = 1) +
  scale_color_viridis_d(option = "magma", direction = 1) +
  labs(y = "-LOG(Q)") +
  theme_minimal() +
  facet_wrap(vars(Dose))

ggsave(GO_set_innate_bars, file = "GO_set_innate_bars.png")

```

```{r}
#Definir conjunto de dados
Immune_GO_GSEA$`-log(qvalue)` = as.numeric(Immune_GO_GSEA$`-log(qvalue)`)

# Filtrar as linhas onde a coluna "Gene set" está em valores_desejados
GO_set_adap <- Immune_GO_GSEA %>% filter(`Gene set short` %in% c("ADAPTIVE IMMUNE RESPONSE")) %>% mutate(`-log(qvalue)` = as.numeric(`-log(qvalue)`))

GO_set_adap = merge(GO_set_adap, ann_vaccines_complex, by.x = "Vaccine", by.y = "Condition")
esquisser(GO_set_adap)

GO_set_adap_bars = ggplot(GO_set_adap) +
  aes(
    x = Day,
    fill = `Vaccine group`,
    colour = Vaccine,
    weight = NES
  ) +
  geom_bar(position = "dodge") +
  scale_fill_viridis_d(option = "magma", direction = 1) +
  scale_color_viridis_d(option = "magma", direction = 1) +
  labs(y = "-LOG(Q)") +
  theme_minimal() +
  facet_wrap(vars(Dose))

ggsave(GO_set_adap_bars, file = "GO_set_adap_bars.png")

```

```{r}
GO_set_humoral <- Immune_GO_GSEA %>% filter(`Gene set short` %in% c("innate immune response", "adaptive immune response", "immunoglobulin mediated immune response", "T cell activation"))



#Definir nome para arquivos gerados
file_bar <- "Immune_GO_general"
fileimage <- paste(file_bar, "_barplot.png", sep = "") 

#Input

barras = 	GO_set_general #Trocar

#Criar gráfico de barras
GO_bars <- ggplot(barras, aes(x = NES, y = `Vaccine`, fill = `-log(qvalue)`)) +
  geom_col(position = "dodge") +
  geom_text(aes(label = round(`-log(qvalue)`, 2)),
            position = position_dodge(width = 0), hjust = 0, vjust = 0, size = 3, angle = 0) +
  theme_minimal() +
  labs(title = fileimage, size = 5,
       x = "NES",
       y = "Vaccine") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 6),
        legend.position = "right") +  # Posição da legenda à direita
  facet_wrap(~`Gene set, short`, ncol = 2) +
  scale_fill_gradient(
    low = "gray90",
    high = "red",# Cor para valores ausentes (NA)
    guide = guide_colorbar(  # Personalizar a legenda de cores
      title = "-log(qvalue)",  # Título da legenda
      title.position = "top",  # Posição do título
      title.hjust = 0.5  # Ajuste de alinhamento do título
    )
  )

GO_bars_bg = GO_bars + theme(
  plot.background = element_rect(fill = "white"),  # Fundo branco
  panel.background = element_rect(fill = "white"),  # Fundo branco
  legend.key = element_rect(fill = "white"),  # Fundo branco para a legenda
  legend.background = element_rect(fill = "white"),  # Fundo branco para a legenda
  legend.title = element_text(color = "black")  # Cor do título da legenda
)

# Salvar
ggsave(GO_bars_bg, file=fileimage, width = 10, height = 5, units = "in") #Trocar

# Display
print(GO_bars_bg)
```

## Gene L2FC by immune process to heatmap

```{r}
#Padronizar tabela
Immune_allgenes_allprocesses = Immune_allgenes_l2fc %>% 
  rename(Condition = study, Genes = genes) %>% filter(Vaccine != "NA")

#General
Immune_GO_general_innate = Immune_allgenes_allprocesses %>% 
  filter(Process == "innate immune response") %>% 
  filter(Condition != "NA")

Immune_GO_general_adaptive = Immune_allgenes_allprocesses %>% 
  rename(Condition = study, Genes = genes) %>% 
  filter(Process == "adaptive immune response",
         Condition != "NA"))

\#Humoral
Immune_GO_humoral_humoral = Immune_GO_humoral %>% 
  rename(Condition = study, Genes = genes) %>% 
  filter(Process == "humoral immune response",
         Condition != "NA"))

Immune_GO_humoral_IgMediated = Immune_GO_humoral %>% 
  rename(Condition = study, Genes = genes) %>% 
  filter(Process == "immunoglobulin mediated immune response", 
         Condition != "NA"))

#Cellular
Immune_GO_adaptive_cellular_tcellactivation = Immune_GO_cellular %>% rename(Condition = study, Genes = genes) %>% filter(Process == "T cell activation") %>% filter(Condition != "NA")
```

## Criando matriz para heatmap

```{r}
#Input
heatmap_data = Immune_GO_general_adaptive
file= "Immune_GO_general_adaptive"

#Excluir linhas com Target ou Source == NA
heatmap_data = filter(heatmap_data, Vaccine != "NA")

#Selecionar colunas de comparação e valores. Trocar value.var pela variável de interesse
names(heatmap_data)
matrix_data <- dcast(heatmap_data, `Genes` ~ `Condition`, value.var = "log2FoldChange", fun.aggregate = mean)

#Converter coluna 1 em rownames e excluir
rownames(matrix_data) <- matrix_data$`Genes`
matrix_data <- matrix_data %>% select(-`Genes`)

#Arredondar valores para facilitar visualização
matrix_data <- round(matrix_data, 1)
matrix_data <- replace(matrix_data, matrix_data == "NaN", 0)


#Definir nome para arquivo geral
write.csv(matrix_data, file=paste(file, "heatmap.csv",sep="_")) 

```

## pHeatmap

```{r}

#Definir matriz de dados
matrix_data = as.matrix(matrix_data)

cellmarker_studies

# Criar tabela de anotações para as vacinas (só rodar uma vez)
# rownames(ann_vaccines) = rownames(t(matrix_data)) # Obter condições e transpor para linhas
# ann_vaccines = select(ann_vaccines, -ann_vaccines)
#ann_vaccines = data.frame(ann_vaccines) # Converter em dataframe

# Agrupamento
# ann_vaccines$Vaccine = NA
# ann_vaccines$Dose= NA
# ann_vaccines$Day = NA
# #Anotar manualmente usando editData
# ann_vaccines = editData(ann_vaccines) 

#Cores

# # Defina os níveis do fator Vaccine para corresponderem às cores
# ann_vaccines$Vaccine <- factor(ann_vaccines$Vaccine)
# ann_vaccines$Dose = factor(ann_vaccines$Dose)
# ann_vaccines$Day <- factor(ann_vaccines$Day)

#Defina a ordem das anotações no gráfico
# ann_vaccines = select(ann_vaccines, Day, Dose, Vaccine)

# # Defina as cores correspondentes em annotation_colors
# ann_colors <- list(
#   Vaccine = c(
#   "BBIBP" = "#F46D43",
#   "BNT" = "#0096C7",
#   "ChAd" = "#99d98c",
#   "ChAd-BNT" = "#76c893",
#   "ZF2001" = "#ffd500"),
#   Dose = c(
#     "1" = "#ffea00",
#     "2" = "#ff7b00",
#     "3" = "#e01e37"
#   ),
#   Day = c(
#     "01" = "#a5ffd6",
#     "03" = "#99d98c",
#     "06" = "#00bbf9",
#     "07" = "#4cc9f0",
#     "14" = "#7b2cbf",
#     "28" = "#b5179e"))


# Defina os valores mínimos e máximos para o espectro de cores
valores_minimos <- -1  #  Substitua pelo valor mínimo desejado
valores_maximos <- 3   # Substitua pelo valor máximo desejado

# min(matrix_data)
# max(matrix_data)



# Crie o gráfico de calor com configurações personalizadas
p = pheatmap::pheatmap(
matrix_data,
annotation_colors = ann_colors,
annotation_col = ann_vaccines,
cutree_cols = 5,
cutree_rows = 7,
width = 100,  # Largura da figura (ajuste conforme necessário)
height = 1000,  # Altura da figura (ajuste conforme necessário)
fontsize_row = 2,  # Tamanho da fonte para rótulos de linhas
fontsize_col = 3,  # Tamanho da fonte para rótulos de colunas
cluster_rows = TRUE,  # Impede o agrupamento de linhas
cluster_cols = F,  # Impede o agrupamento de colunas
main = "Immune_GO_adaptive_cellular_tcellactivation",
angle_col = 90,  # Rotação dos rótulos de coluna (0 para horizontal)
cellheight = 2,  # Altura das células (ajuste conforme necessário)
legend = TRUE,
border = TRUE,
border_color = "black",
show_colnames = TRUE,
show_rownames = TRUE,
color = colorRampPalette(c("blue", "white", "red"))(100),  # Esquema de cores
breaks = c(seq(valores_minimos, -0.001, length.out = 50), seq(0.001, valores_maximos, length.out = 50)))


# Salve o gráfico como um arquivo PNG
png(file= fileimage, width=2000, height=2000, res=315, pointsize=10)
print(p)  # Substitua 'p' pelo objeto 'pheatmap'
dev.off()  # Finalize a gravação do arquivo


```

## Gráfico de barras

```{r}

#Definir conjunto de dados
gene_set = ImmuneGO_Annotated_Genes

#Definir nome para arquivos gerados
file <- "Immune_GO"
fileimage <- paste(file, "_barplot.png", sep = "") 


#Separar por vacina
# BBIBP = filter(gene_set, Vaccine == "BBIBP")
# ZF2001 = filter(gene_set, Vaccine == "ZF2001")
# BNT = filter(gene_set, Vaccine == "BNT")
# ChAd = filter(gene_set, Vaccine == "ChAd")

#Input
barras = gene_set #Trocar

#Criar gráfico de barras
genes_bars <- ggplot(barras, aes(x = Genes, y = log2FoldChange, fill = Condition)) +
  geom_col(position = "dodge") +
  geom_text(aes(label = round(log2FoldChange, 1)),
            position = position_dodge(width = 0), hjust = 0, vjust = 0, size = 1.5, angle = 0) +
  theme_minimal() +
  labs(title = "Innate immune response associated genes - BBIBP", size = 5, #Trocar
       x = "Gene",
       y = "L2FC") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, size = 6),
        legend.position = "none") + 
  facet_wrap(~Condition, ncol = 1)

genes_bars

ggsave(genes_bars, file=fileimage, width = 8, height = 16, units = "in") #Trocar
```

# PCA

<https://www.datacamp.com/tutorial/pca-analysis-r> 
<https://rpkgs.datanovia.com/factoextra/index.html> <http://www.sthda.com/english/articles/31-principal-component-methods-in-r-practical-guide/114-mca-multiple-correspondence-analysis-in-r-essentials/>
<https://statisticsglobe.com/pca-before-k-means-clustering-r>>

Bibliotecas
```{r}
library(factoextra)
library(caret)
library(stats)
library(ggfortify)
library(ggplot2)
```

# Principal component analysis - By sample

```{r}
# Criar um PCA com o p-valor de cada gene 
pca_data <- prcomp(t(assay(dds)))  # Perform PCA on the transposed count or normalized data

# Create a dataframe with PCA components and Timepointf and FirstSecondDosef
pca_df <- data.frame(PC1 = pca_data$x[, 1], PC2 = pca_data$x[, 2],
                     Timepoint = colData(dds)$Timepoint,
                    Vaccines = colData(dds)$Vaccine)

# Plot the PCA using ggplot2
ggplot(pca_df, aes(x = PC1, y = PC2, color = Vaccines, shape = Timepoint)) +
  geom_point(size = 3) +
  labs(title = "Principal component analysis - Vaccine-Timepoint", x = "PC1", y = "PC2") +
  theme_minimal()

#DESeq2 native analysis
vsd <- vst(dds, blind=FALSE)
pcaData = plotPCA(vsd,intgroup=c("Timepoint", "Vaccine"), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))

pca_plot = ggplot(pcaData, aes(PC1, PC2, color=Vaccine, shape=Timepoint)) +
  geom_point(size=2) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed() + ggtitle("Principal component analysis - Vaccines and Timepoint")

ggsave(pca_plot, file = "GSE206023_PCA.png")

```

```{r}

dados = DEGs_All_studies_ImmuneGO_GSEA_1_
dados_filtrados <- dados[, c("Vaccine", "Gene set", "NES")]

matriz_dados <- dcast(dados_filtrados, Vaccine ~ `Gene set, short`, value.var = "NES")


matriz_dados[is.na(matriz_dados)] <- 0
matriz_dados_sem_constantes <- matriz_dados[, apply(matriz_dados, 2, var) > 0.0001]
write.csv(matriz_dados_sem_constantes, file="degs_gsea_immunego_pca.csv")
resultado_pca <- prcomp(matriz_dados_sem_constantes, scale = TRUE)
fviz_pca_ind(resultado_pca,
             geom.ind = "point",
             col.ind = "cos2",
             gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"),
             title = "PCA - Componentes Principais 1 e 2")

```


### By genes

Padronizar
```{r}
install.packages("janitor")
library(janitor)
#Inputs
degs_allstudies_updown = degs_allstudies_updown_filter %>% 
  select(-ends_with(".y")) %>% 
  rename_all(~sub("\\.x$", "", .)) %>%
  clean_names() %>% 
  rename(lfcse = lfc_se,
         l2fc = log2fold_change) %>% 
  select(- set_size_intervals) %>% 
  group_by(study) %>% 
  select(vaccine, study, gse_id, genes, process, gene_set_short, immune_system, immune_sub_system, immune_tissue, go_term, set_size, everything(), filter) %>% 
  mutate(study = ifelse(study == "ChAd (V2, D2)", "ChAd (V2, D3)", study)) %>% 
  distinct() %>% 
  filter(filter == "p<0.05")



#Salvar
write.csv(degs_allstudies_updown, file = "ImmuneGO_degs_allstudies_updown_filter_p005.csv")
```

Separar processos imunológicos em objetos
```{r}
############## Innate
Immune_GO_innate = ImmuneGO_Annotated_Genes %>%  #250 genes (SetSize)
  filter(process == "INNATE IMMUNE RESPONSE" & 
         ((gse_id %in% c("GSE199750", "GSE201533") & padj <= 0.05) |
          (gse_id == "GSE206023" & pvalue <= 0.05)))

Immune_GO_Complement = degs_allstudies_updown %>% #27 genes (SetSize)
  filter(process == "COMPLEMENT ACTIVATION" & 
         ((gse_id %in% c("GSE199750", "GSE201533") & padj <= 0.05) |
          (gse_id == "GSE206023" & pvalue <= 0.05)))

Immune_GO_innate_Inflammatory = degs_allstudies_updown %>% #14 genes (SetSize)
  filter(process == "INFLAMMATORY RESPONSE TO ANTIGENIC STIMULUS" & 
         ((gse_id %in% c("GSE199750", "GSE201533") & padj <= 0.05) |
          (gse_id == "GSE206023" & pvalue <= 0.05)))

Immune_GO_innate_DendriticCells = degs_allstudies_updown %>% #11 genes (SetSize)
  filter(process %in% c("DENDRITIC CELL CHEMOTAXIS", 
                        "DENDRITIC CELL HOMEOSTASIS", 
                        "DENDRITIC CELL MIGRATION", 
                        "MYELOID DENDRITIC CELL ACTIVATION", 
                        "MYELOID DENDRITIC CELL CHEMOTAXIS", 
                        "MYELOID DENDRITIC CELL DIFFERENTIATION", 
                        "PLASMACYTOID DENDRITIC CELL ACTIVATION", 
                        "PLASMACYTOID DENDRITIC CELL DIFFERENTIATION") & 
         ((gse_id %in% c("GSE199750", "GSE201533") & padj <= 0.05) |
          (gse_id == "GSE206023" & pvalue <= 0.05)))


Immune_GO_innate_Macrophages = degs_allstudies_updown %>% 
  filter(process %in% c("MACROPHAGE ACTIVATION", 
                        "MACROPHAGE ACTIVATION INVOLVED IN IMMUNE RESPONSE", 
                        "MACROPHAGE CHEMOTAXIS", 
                        "MACROPHAGE HOMEOSTASIS",
                        "MACROPHAGE MIGRATION") & 
         ((gse_id %in% c("GSE199750", "GSE201533") & padj <= 0.05) |
          (gse_id == "GSE206023" & pvalue <= 0.05)))

Immune_GO_innate_APP = degs_allstudies_updown %>% #21 genes (SetSize)
  filter(process == "ANTIGEN PROCESSING AND PRESENTATION" & 
         ((gse_id %in% c("GSE199750", "GSE201533") & padj <= 0.05) |
          (gse_id == "GSE206023" & pvalue <= 0.05)))

Immune_GO_innate_TLR = degs_allstudies_updown %>% #14 genes (SetSize)
  filter(process == "TOLL-LIKE RECEPTOR SIGNALING PATHWAY" & 
         ((gse_id %in% c("GSE199750", "GSE201533") & padj <= 0.05) |
          (gse_id == "GSE206023" & pvalue <= 0.05)))

Immune_GO_innate_PRR = degs_allstudies_updown %>% #8 genes (SetSize)
  filter(process == "PATTERN RECOGNITION RECEPTOR SIGNALING PATHWAY" & 
         ((gse_id %in% c("GSE199750", "GSE201533") & padj <= 0.05) |
          (gse_id == "GSE206023" & pvalue <= 0.05)))
Immune_GO_innate_CPR = degs_allstudies_updown %>% #6 genes (SetSize)
  filter(process == "CYTOSOLIC PATTERN RECOGNITION RECEPTOR SIGNALING PATHWAY" & 
         ((gse_id %in% c("GSE199750", "GSE201533") & padj <= 0.05) |
          (gse_id == "GSE206023" & pvalue <= 0.05)))

############## Adaptive
Immune_GO_adaptive = degs_allstudies_updown %>% #330 genes (SetSize)
  filter(process == "ADAPTIVE IMMUNE RESPONSE" &
    ((gse_id %in% c("GSE199750", "GSE201533") & padj <= 0.05) |
    (gse_id == "GSE206023" & pvalue <= 0.05)) &
    !(gse_id %in% c("GSE201533")))

#Humoral
Immune_GO_adaptive_humoral = degs_allstudies_updown %>% #27 genes (SetSize)
  filter(process == "HUMORAL ADAPTIVE IMMUNE SYSTEM" &
    ((gse_id %in% c("GSE199750", "GSE201533") & padj <= 0.05) |
    (gse_id == "GSE206023" & pvalue <= 0.05)) &
    !(gse_id %in% c("GSE201533")))

Immune_GO_adap_IgMediated = degs_allstudies_updown %>% #62 genes (SetSize)
  filter(process == "IMMUNOGLOBULIN MEDIATED IMMUNE RESPONSE" &
    ((gse_id %in% c("GSE199750", "GSE201533") & padj <= 0.05) |
    (gse_id == "GSE206023" & pvalue <= 0.05)) &
    !(gse_id %in% c("GSE201533")))

Immune_GO_adap_Breceptorsig = degs_allstudies_updown %>% #31 genes (SetSize)
  filter(process == "B CELL RECEPTOR SIGNALING PATHWAY" &
    ((gse_id %in% c("GSE199750", "GSE201533") & padj <= 0.05) |
    (gse_id == "GSE206023" & pvalue <= 0.05)) &
    !(gse_id %in% c("GSE201533")))

Immune_GO_adap_Bactivation = degs_allstudies_updown %>% #36 genes
  filter(process == "B CELL ACTIVATION" &
    ((gse_id %in% c("GSE199750", "GSE201533") & padj <= 0.05) |
    (gse_id == "GSE206023" & pvalue <= 0.05)) &
    !(gse_id %in% c("GSE201533")))

#Cellular

Immune_GO_adaptive_cellular = degs_allstudies_updown %>% #41 genes
  filter(process == "CELLULAR ADAPTIVE IMMUNE SYSTEM" &
    ((gse_id %in% c("GSE199750", "GSE201533") & padj <= 0.05) |
    (gse_id == "GSE206023" & pvalue <= 0.05)) &
    !(gse_id %in% c("GSE201533")))

Immune_GO_adaptive_cellular_tcellactivation = degs_allstudies_updown %>% #41 genes
  filter(process == "T CELL ACTIVATION" &
    ((gse_id %in% c("GSE199750", "GSE201533") & padj <= 0.05) |
    (gse_id == "GSE206023" & pvalue <= 0.05)) &
    !(gse_id %in% c("GSE201533")))

Immune_GO_adaptive_cellular_treceptor = degs_allstudies_updown %>% #78 genes
  filter(process == "T CELL RECEPTOR SIGNALING PATHWAY" &
    ((gse_id %in% c("GSE199750", "GSE201533") & padj <= 0.05) |
    (gse_id == "GSE206023" & pvalue <= 0.05)) &
    !(gse_id %in% c("GSE201533")))

Immune_GO_adaptive_cellular_th1 = degs_allstudies_updown %>% #6 genes
  filter(process == "T-HELPER 1 TYPE IMMUNE RESPONSE" &
    ((gse_id %in% c("GSE199750", "GSE201533") & padj <= 0.05) |
    (gse_id == "GSE206023" & pvalue <= 0.05)) &
    !(gse_id %in% c("GSE201533")))

Immune_GO_adaptive_cellular_th2 = degs_allstudies_updown %>% #5 genes
  filter(process == "TYPE 2 IMMUNE RESPONSE" &
    ((gse_id %in% c("GSE199750", "GSE201533") & padj <= 0.05) |
    (gse_id == "GSE206023" & pvalue <= 0.05)) &
    !(gse_id %in% c("GSE201533")))

Immune_GO_adaptive_cellular_th17 = degs_allstudies_updown %>% #3 genes
  filter(process == "T-HELPER 17 TYPE IMMUNE RESPONSE" &
    ((gse_id %in% c("GSE199750", "GSE201533") & padj <= 0.05) |
    (gse_id == "GSE206023" & pvalue <= 0.05)) &
    !(gse_id %in% c("GSE201533")))



ImmuneGO_Genes_interest = bind_rows(
  Immune_GO_general_innate,
  Immune_GO_general_Complement,
  Immune_GO_innate_Inflammatory,
  Immune_GO_innate_Macrophages,
  Immune_GO_innate_DendriticCells,
  Immune_GO_innate_APP,
  Immune_GO_innate_TLR,
  Immune_GO_innate_PRR,
  Immune_GO_innate_CPR,
  Immune_GO_adaptive,
  Immune_GO_adaptive_humoral,
  Immune_GO_adap_IgMediated,
  Immune_GO_adap_Breceptorsig,
  Immune_GO_adap_Bactivation,
  Immune_GO_adaptive_cellular,
  Immune_GO_adaptive_cellular_tcellactivation,
  Immune_GO_adaptive_cellular_treceptor,
  Immune_GO_adaptive_cellular_th1,
  Immune_GO_adaptive_cellular_th2,
  Immune_GO_adaptive_cellular_th17)

write.csv(ImmuneGO_Genes_interest, file = "ImmuneGO_Genes_interest_bystudy.csv")

```

Processar dados
```{r}
# Immune_GO_general_innate
# Immune_GO_adaptive

Immune_GO_general_Complement
Immune_GO_innate_Macrophages
Immune_GO_innate_DendriticCells
Immune_GO_innate_Inflammatory
Immune_GO_innate_APP
Immune_GO_innate_TLR #Não usar
Immune_GO_innate_PRR #(SIGNALING PATHWAY)
Immune_GO_innate_CPR #Não usar
Immune_GO_adaptive_humoral
Immune_GO_adap_IgMediated
Immune_GO_adap_Breceptorsig
Immune_GO_adap_Bactivation
Immune_GO_adaptive_cellular
Immune_GO_adaptive_cellular_tcellactivation
Immune_GO_adaptive_cellular_treceptor


#INPUT
data_genes = Immune_GO_adaptive
filename = "Immune_GO_adaptive"
```

PCA

```{r}
#Converter de long para wide
matrix_genes = data_genes %>% 
  select(study, genes, l2fc) %>%  
  dcast(`study` ~ `genes`, 
        value.var = "l2fc", 
        fun.aggregate = mean) %>% 
  as.data.frame() %>% 
  column_to_rownames(var = "study") %>% 
  replace(is.na(.), 0)

matrix_data_pca = matrix_genes %>% 
  rownames_to_column(var = "Condition")

matrix_data_pca_ready = matrix_data_pca %>% 
  column_to_rownames(var = "Condition")

ann_vaccines_pca_matrix = ann_vaccines_pca %>% 
  merge(matrix_data_pca, 
        by.x = "Condition", 
        all.x = F, 
        all.y = F) %>% 
  select(Condition:Efficacy)

# Verifique quais colunas têm variância muito baixa
nearZeroVarCols <- nearZeroVar(matrix_data_pca_ready, saveMetrics = TRUE)
matrix_data_pca_ready <- matrix_data_pca_ready[, !nearZeroVarCols$nzv]
pca_res <- prcomp(matrix_data_pca_ready, scale. = TRUE)

# Crie o gráfico de PCA
pca_plot_ann = autoplot(pca_res, 
                    data = ann_vaccines_pca_matrix, 
                    colour = 'Vaccine', 
                    label = 1) + #1: display, 0: hide
  labs(title=filename) +
  theme_classic()
  # scale_color_manual(values = c(BBIBP = "#black", 
  #                               ZF2001 = "black",
  #                               BNT = "#56cfe1",
  #                               ChAd = "#80ffdb" ,
  #                               "ChAd-BNT" = "#72efdd")) 

pca_plot_not_ann = autoplot(pca_res, 
                    data = ann_vaccines_pca_matrix, 
                    colour = 'Vaccine', 
                    label = 0) + #1: display, 0: hide
  labs(title=filename) +
  theme_classic() +
  stat_ellipse(type = "t")

#Salvar
ggsave(pca_plot_not_ann, filename = paste0(filename, "_PCA_not-ann.png"), width = 10, height = 8)
ggsave(pca_plot_ann, filename = paste0(filename, "_PCA_ann.png"), width = 10, height = 8)

# Exiba o gráfico
print(pca_plot_ann)
print(pca_plot_not_ann)

############### KNN classification

#Determinar o número de clusters para KNN
pca_scores <- data.frame(pca_res$x[, 1:2])
ggp1 <- fviz_nbclust(pca_scores,  
                     FUNcluster = kmeans,
                     method = "wss")

ggp1 

set.seed(123)                             # Set seed for randomization
kmeans_clust <- kmeans(pca_scores,        # Perform k-means clustering
                        centers = 4) # Definir numero de clusters

#Visualizar clusters
ggp2 <- fviz_pca_ind(pca_res,
                   habillage = kmeans_clust$cluster,
                   repel = TRUE,
                   addEllipses = TRUE,
                   ellipse.type = "convex",
                   labelsize = 2) +
     guides(color = guide_legend(override.aes = list(label = ""))) +
  ggtitle(paste0(filename,  "KNN_Clustered.png"))

print(ggp2)
ggsave(ggp2, file = paste0(filename, "KNN_Clustered.png"), width = 5, height = 4)


```


```{r}
########Correlation plot
#Matriz
corr_matrix = cor(matrix_data_pca_ready) 

#Plot
corrplot = ggcorrplot(corr_matrix, hc.order = TRUE) + 
  theme(axis.text.x = element_text(angle = 90, size = 5), 
        axis.text.y = element_text(size = 5))
#Salvar
ggsave(corrplot, file = paste0(filename, "_corrplot.png"), 
       width = 20, #Grande 20, pequeno 10
       height= 20) #Grande 20, pequeno 10
print(corrplot)

#########Scree plot
data.pca = princomp(corr_matrix) #PCA
summary(data.pca) #Retornar PCs

#########Scree plot
scree_plot = fviz_eig(data.pca, 
                      addlabels = TRUE,
                      ylim = c(0, 70)) +
  geom_col(color = "#00AFBB", fill = "#00AFBB") +
  theme_classic()

#Salvar
ggsave(scree_plot, file = paste0(filename, "_screeplot.png"), width = 10, height = 3) 
print(scree_plot)

#Comp1-Comp2
loadings = data.frame(data.pca$loadings[, 1:3])
loadings$Genes = rownames(loadings)
loadings_1 = arrange(loadings, desc(Comp.1))
loadings_2 = arrange(loadings, desc(Comp.2))
loadings_3 = arrange(loadings, desc(Comp.3))

#Plot
loadings_1_plot = ggplot(loadings_1, aes(x = reorder(Genes, -Comp.1), y=Comp.1, fill = Comp.1)) + 
  ggtitle(paste0("Comp1-Comp2 Genes", filename)) + 
  ylab("Comp1") +
  xlab("Genes") +
  geom_bar(stat = "identity") + 
  theme_light() +
  theme(axis.text.x = element_text(vjust = 0.1, hjust = 1, angle = 90, size = 3), 
        axis.text.y = element_text(size = 5),
        plot.title = element_text(size = 15L, hjust = 0.5)) + 
  scale_fill_continuous(type = "viridis") #cores
  
#Salvar
ggsave(loadings_1_plot, 
       file = paste0(filename, "_genescomp1-2.png"), 
       width = 10, height = 5)

print(loadings_1_plot)
```


```{r}
# Graph of the variables
fviz_pca_var_genes = fviz_pca_var(data.pca, 
                                  col.ind = "cos2",
                                  gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"))

#Salvar
ggsave(fviz_pca_var_genes, file = paste0(filename, "_fviz_pca_var_genes.png"), width = 10)

######### COS2
cos2 = fviz_cos2(data.pca, 
                 choice = "var", 
                 axes = 1:2) + 
  theme(axis.text.x = element_text(angle=45, 
                                   size = 3)) +
  theme_light()

#Salvar
ggsave(cos2, file = paste0(filename, "_cos2.png"), width = 10, height = 5)

#Colorido
fviz_pca_var(data.pca, col.var = "cos2",
            gradient.cols = c("black", "orange", "green"),
            repel = TRUE)

#Print
print(fviz_pca_var_genes)
print(cos2)
```

### By GO term

```{r}

Immune_GO_GSEA = Immune_GO_GSEA %>% 
  clean_names()

Immune_GO_GSEA_Innate = Immune_GO_GSEA %>% filter(immune_system == "Innate")
Immune_GO_GSEA_Adaptive = Immune_GO_GSEA %>% filter(immune_system == "Adaptive")


data_ImmuneGO = Immune_GO_GSEA_Innate
filename = "Immune_GO_GSEA_Innate"

#Converter de long para wide
matrix_genes = data_ImmuneGO %>% 
  select(vaccine, gene_set_short, nes)

matrix_genes <- dcast(matrix_genes, `vaccine` ~ `gene_set_short`, 
                     value.var = "nes", 
                     fun.aggregate = mean)

#Arredondar valores para facilitar visualização
matrix_genes <- replace(matrix_genes, matrix_genes == "NaN", 0)

#Converter coluna 1 em rownames e excluir
matrix_data_pca = as.data.frame(matrix_genes)
rownames(matrix_data_pca) <- matrix_data_pca$vaccine
matrix_data_pca_ready = select(matrix_data_pca, -vaccine)
ann_vaccines_pca_matrix = merge(ann_vaccines_pca, matrix_data_pca, by.x = "Condition", by.y = "vaccine", all.x = F, all.y = F) 
ann_vaccines_pca_matrix_ready = select(ann_vaccines_pca_matrix, Condition, Vaccine, Day, Dose)

# Verifique quais colunas têm variância muito baixa
nearZeroVarCols <- nearZeroVar(matrix_data_pca_ready, saveMetrics = TRUE)
matrix_data_pca_ready <- matrix_data_pca_ready[, !nearZeroVarCols$nzv]
pca_res <- prcomp(matrix_data_pca_ready, scale. = TRUE)

# Crie o gráfico de PCA
pca_plot = autoplot(pca_res, 
                    data = ann_vaccines_pca_matrix, 
                    colour = 'Vaccine', 
                    label = 1) + 
  labs(title=filename)

# Exiba o gráfico
print(pca_plot)
ggsave(pca_plot, filename = "PCA_Immune_GO_GSEA_Adaptive_not-ann.png", width = 10, height = 8)
ggsave(pca_plot, filename = "PCA_Immune_GO_GSEA_Adaptive_annotated.png", width = 10, height = 8)


############### KNN classification

#Determinar o número de clusters para KNN
pca_scores <- data.frame(pca_res$x[, 1:2])
ggp1 <- fviz_nbclust(pca_scores,  
                     FUNcluster = kmeans,
                     method = "wss")

ggp1 

set.seed(123)                             # Set seed for randomization
kmeans_clust <- kmeans(pca_scores,        # Perform k-means clustering
                        centers = 4) # Definir numero de clusters

#Visualizar clusters
ggp2 <- fviz_pca_ind(pca_res,
                   habillage = kmeans_clust$cluster,
                   repel = TRUE,
                   addEllipses = TRUE,
                   ellipse.type = "convex",
                   labelsize = 2) +
     guides(color = guide_legend(override.aes = list(label = ""))) +
  ggtitle(paste0(filename,  "KNN_Clustered.png")) +
  scale_color_brewer(palette="Set1")

print(ggp2)
ggsave(ggp2, file = paste0(filename, "_KNN_Clustered.png"), width = 5, height = 4)


```

```{r}
#Plot
corrplot = ggcorrplot(corr_matrix, hc.order = TRUE) + 
  theme(axis.text.x = element_text(angle = 90, size = 5), 
        axis.text.y = element_text(size = 5))
#Salvar
ggsave(corrplot, file = paste0(filename, "_GSEA_corrplot.png"), 
       width = 10, #Grande 20, pequeno 10
       height= 10) #Grande 20, pequeno 10
print(corrplot)

data.pca <- princomp(corr_matrix) #PCA
summary(data.pca) #Retornar PCs


#########Scree plot
data.pca = princomp(corr_matrix) #PCA
summary(data.pca) #Retornar PCs

#########Scree plot
scree_plot = fviz_eig(data.pca, 
                      addlabels = TRUE,
                      ylim = c(0, 70)) +
  geom_col(color = "#00AFBB", fill = "#00AFBB") +
  theme_classic()

#Salvar
ggsave(scree_plot, file = paste0(filename, "_GSEA_screeplot.png"), width = 10, height = 3) 
print(scree_plot)


#Scree plot
loadings = data.frame(data.pca$loadings[, 1:3])
loadings$Genes = rownames(loadings)
loadings_1 = arrange(loadings, desc(Comp.1))
loadings_2 = arrange(loadings, desc(Comp.2))
loadings_3 = arrange(loadings, desc(Comp.3))

#Plot
loadings_1_plot = ggplot(loadings_1, aes(x = reorder(Genes, -Comp.1), y=Comp.1, fill = Comp.1)) + 
  ggtitle(paste0("Comp1-Comp2 Genes", filename)) + 
  ylab("Comp1") +
  xlab("Gene sets") +
  geom_bar(stat = "identity") + 
  theme_light() +
  theme(axis.text.x = element_text(vjust = 1, hjust = 1, angle = 45, size = 4), 
        axis.text.y = element_text(size = 5),
        plot.title = element_text(size = 15L, hjust = 0.5)) + 
  scale_fill_continuous(type = "viridis") #cores

print(loadings_1_plot)


#Salvar
ggsave(loadings_1_plot, 
       file = paste0(filename, "_GSEA_genescomp1-2.png"), 
       width = 10, height = 5)



# Graph of the variables
fviz_pca_var_genes = fviz_pca_var(data.pca, 
                                  col.ind = "cos2",
                                  gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"))
fviz_pca_var_genes

ggsave(fviz_pca_var_genes, file = "fviz_pca_var_GSEA_innate.png", width = 15)

cos2 = fviz_cos2(data.pca, choice = "var", axes = 1:2)
cos2
ggsave(cos2, file = "cos2_GSEA_innate.png", width = 7)




```
BIPLOT
```{r}
fviz_pca_biplot(pca_res,              # Visualize clusters in biplot
                      col.var = "black",
                      alpha.var = 0.1,
                      label = "all",
                      habillage = kmeans_clust$cluster,
                      repel = TRUE,
                      addEllipses = TRUE,
                      ellipse.type = "convex",
                      labelsize = 2, 
                      palette = "Dark2")
```


# Gráfico de linhas

```{r}

Immune_GO_GSEA

################## Innate
Immune_GO_innate = Immune_GO_GSEA %>% filter(`Immune system` == "Innate")
write.csv(Immune_GO_innate, file = "Immune_GO_innate_l2fc.csv")

Immune_GO_innate_SUM <- Immune_GO_innate %>% 
  rename(Condition = study) %>% 
  group_by(Condition, Process) %>%
  mutate(Soma = n()) %>% select(Vaccine, Condition, Process, `Immune group`, `Immune response gene set, specific`, `Immune system`, Soma) %>% filter(Vaccine != "NA")
write.csv(Immune_GO_innate_SUM, file="Immune_GO_innate_SUM.csv")


################## Adaptive
#Humoral
Immune_GO_humoral = Immune_allgenes %>% filter(`Immune system` == "Humoral")
write.csv(Immune_GO_humoral, file = "Immune_GO_humoral_l2fc.csv")

Immune_GO_humoral_SUM <- Immune_GO_humoral %>% 
  rename(Condition = study) %>% 
  group_by(Condition, Process) %>%
  mutate(Soma = n()) %>% select(Vaccine, Condition, Process, `Immune group`, `Immune response gene set, specific`, `Immune system`, Soma) %>% filter(Vaccine != "NA")
write.csv(Immune_GO_humoral_SUM, file="Immune_GO_humoral_SUM.csv")

#Cellular
Immune_GO_cellular = Immune_allgenes %>% filter(`Immune system` == "Cellular")
write.csv(Immune_GO_cellular, file = "Immune_GO_cellular_l2fc.csv")

Immune_GO_cellular_SUM <- Immune_GO_cellular %>% 
  rename(Condition = study) %>% 
  group_by(Condition, Process) %>%
  mutate(Soma = n()) %>% select(Vaccine, Condition, Process, `Immune group`, `Immune response gene set, specific`, `Immune system`, Soma) %>% filter(Vaccine != "NA")
write.csv(Immune_GO_cellular_SUM, file="Immune_GO_cellular_SUM.csv")

################## General
Immune_GO_general = Immune_allgenes %>% filter(`Immune system` == "General")
write.csv(Immune_GO_general, file = "Immune_GO_general_l2fc.csv")

Immune_GO_general_SUM <- Immune_GO_general %>% 
  rename(Condition = study) %>% 
  group_by(Condition, Process) %>%
  mutate(Soma = n()) %>% select(Vaccine, Condition, Process, `Immune group`, `Immune response gene set, specific`, `Immune system`, Soma) %>% filter(Vaccine != "NA")
write.csv(Immune_GO_general_SUM, file = "Immune_GO_general_SUM.csv")

################## All immune processes and genes
Immune_allgenes_vaccines = Immune_allgenes_l2fc %>% rename(Condition = study, Genes = genes) %>% filter(Vaccine != "NA")
ann_vaccines_pca_matrix = merge(ann_vaccines_pca, matrix_data_pca, by.x = "Condition", all.x = F, all.y = F)
ann_vaccines_pca_matrix_long = ann_vaccines_pca_matrix %>% gather(Genes, L2FC, -Condition, -Vaccine, -Day, -Dose)

Immune_allgenes_vaccines_SUM <- Immune_allgenes_vaccines %>%
  group_by(Condition, Process) %>% #Somar e agrupar
  summarize(Soma = n())

Immune_allgenes_vaccines_SUM = merge(ann_vaccines_pca, Immune_allgenes_vaccines_SUM, by.x = "Condition", all.x = F, all.y = F) # Anotar vacinas
write.csv(Immune_allgenes_vaccines_SUM, file="Immune_allgenes_vaccines_SUM.csv") #Salvar

```

# Shifted bars

```{r}
ImmuneGO_GSEA_General_Innate_Adaptive = Immune_GO_GSEA %>% 
  filter(`Gene set short` %in% c("CELLULAR ADAPTIVE IMMUNE SYSTEM","HUMORAL ADAPTIVE IMMUNE SYSTEM")) %>% 
  mutate(`-log(qvalue)` = as.numeric(`-log(qvalue)`))

barchart_NES

ggplot(ImmuneGO_GSEA_General_Innate_Adaptive) +
 aes(x = Vaccine, fill = `Gene set short`, weight = NES) +
 geom_bar() +
 scale_fill_manual(values = c(`ADAPTIVE IMMUNE RESPONSE` = "#6DF8DF", `INNATE IMMUNE RESPONSE` = "#989898"
 )) +
 labs(title = "Immune system ", subtitle = "Adaptive and Innate, by NES") +
 coord_flip() +
 theme_light()


ggsave(barchart_NES, file= "barchart_NES.png", width = 10, height = 5)

```

```{r}
Immune_GO_general_SUM_Adaptive_Innate = Immune_GO_general_SUM %>% 
  filter(Process == "adaptive immune response" | Process == "innate immune response") %>% 
  arrange(desc(Process)) %>% 
  distinct()

view(Immune_GO_general_SUM_Adaptive_Innate)

write.csv(Immune_GO_general_SUM_Adaptive_Innate, file = "Immune_GO_general_SUM_Adaptive_Innate.csv")


barchart_SUM = Immune_GO_general_SUM_Adaptive_Innate %>%
 filter(Process %in% c("adaptive immune response", "innate immune response")) %>%
 ggplot() +
 aes(x = Condition, fill = Process, weight = Soma) +
 labs(title = "Immune system ", subtitle = "Adaptive and Innate") +
 geom_bar(position = "fill") +
  scale_fill_manual(values = c(`adaptive immune response` = "#6DF8DF", `innate immune response` = "#989898"
 ))+
 coord_flip() +
 theme_minimal() +
 theme(legend.position = "top") +
 facet_wrap(vars(Vaccine), scales = "free")

ggsave(barchart_SUM, file= "barchart_SUM.png", width = 10, height = 5)

```

###### Gráfico de linhas

```{r}
lines = Immune_GO_general_SUM_Adaptive_Innate %>% 
  merge(ann_vaccines, by = "Condition", all = T) %>% 
  group_by(Condition, Process) %>% 
  mutate(Day = as.numeric(Day))


esquisser(lines)

lines_plot = lines %>%
 filter(!is.na(Vaccine.x)) %>%
 filter(!is.na(Process)) %>%
 filter(!is.na(`Immune group`)) %>%
 ggplot() +
  aes(
    x = Day,
    y = Soma,
    colour = Vaccine.y,
    shape = Vaccine.x
  ) +
  geom_point(size = 4L) +
  geom_line()+
  scale_color_hue(direction = 1) +
  theme_linedraw() +
  facet_grid(
    vars(`Immune response gene set, specific`),
    vars(Dose),
    scales = "free"
  )

ggsave(lines_plot, file = "ImmuneGO_Adap_Innate_lines_plot.png", width = 10, height = 7)

```

```{r}
barchart_stack = Immune_GO_general_SUM_Adaptive_Innate %>%
 filter(Process %in% c("adaptive immune response", "innate immune response")) %>%
 ggplot() +
 aes(x = Condition, fill = Process, weight = Soma) +
 labs(title = "Immune system ", subtitle = "Adaptive and Innate") +
 geom_bar(position = "stack") +
  scale_fill_manual(values = c(`adaptive immune response` = "#6DF8DF", `innate immune response` = "#989898"
 ))+
 coord_flip() +
 theme_minimal() +
 theme(legend.position = "top") +
 facet_wrap(vars(Vaccine), scales = "free")

```

# Heatmap Cell Marker

```{r}
cellmarker_studies
cellmarker_heatmap = cellmarker_studies %>% filter(qvalue <= 0.10)
cellmarker_heatmap.wide = dcast(cellmarker_heatmap, `ID` ~ `study`, value.var = "NES", fun.aggregate = mean)
cellmarker_heatmap.wide[is.na(cellmarker_heatmap.wide)] = 0
cellmarker_heatmap.wide.ID = cellmarker_heatmap.wide %>% 
  column_to_rownames(var="ID") %>% 
  as.matrix()

#Column annotation
ann_cols = ann_vaccines 
ann_cols$Vaccines = rownames(ann_cols)

#Anotar colunas
cellmarker_heatmap.wide.ID.ANN = cellmarker_heatmap.wide.ID %>% 
  as.data.frame() %>% 
  rownames_to_column("Condition")

ann_vaccines_pheatmap = ann_vaccines %>% 
  merge(shared_heatmap.ann, by="Condition", all.x = F) %>% 
  column_to_rownames("Condition") %>% 
  select(Vaccine, Day, Dose, Type, Regimen) %>% 
  mutate(Day = as.factor(Day))

#Row annotation
ann_rows_cellmarker = Immune_cells %>% 
  select(cell_name, Type)


ann_vaccines_pheatmap


col_fun = colorRampPalette(c("blue", "white", "red"))(100)


#Visualizar
heatmap = pheatmap(
  cellmarker_heatmap.wide.ID,
  annotation_row = NA, 
  annotation_col = ann_vaccines_pheatmap,
  annotation_colors = ann_colors,
  annotation_legend = TRUE,
  clustering_method = "complete",  # Método de clustering, ajuste conforme desejado
  display_numbers = FALSE,  # Mostrar os valores nos quadrados do heatmap
  show_colnames = TRUE,  # Mostrar nomes das colunas
  cluster_cols = TRUE,
  show_rownames = TRUE,  # Mostrar nomes das linhas
  scale = "column",  # Escala - "column" para escala por coluna
  main = "CellMarker scaled data",
  border_color = "white",  # Define a cor da borda, 'NA' para omitir as bordas
  color = col_fun,  # Define a paleta de cores
  breaks = seq(-2, 2),  # Define os intervalos da paleta de cores
  fontsize_row = 5,
  fontsize_col = 5,
  cutree_rows = 5, 
  cutree_cols = 3
)

png(file= "CellMarker_pheatmap.png", width=2000, height=2000, res=315, pointsize=10)
print(heatmap)  # Substitua 'p' pelo objeto 'pheatmap'
dev.off()  # Finalize a gravação do arquivo
```