---
title: "Naive and infected patients, vaccinated"
output: html_notebook
---

```{r}
library(readr)
library(qvalue)
library(explore)
library(maditr)
library(GEOquery)
library(Matrix)
library(ggplot2)
library(dplyr)
library(pheatmap)
library(RColorBrewer)
library(celldex)
library(biomaRt)
library(org.Hs.eg.db)
library(plotly)
library(DESeq2)
library(msigdbr)
library(ape)
library(GSVA)
library(sva)
library(readxl)
library(clusterProfiler)
library(pheatmap)
library(gplots)
library(EnhancedVolcano)
library(metaRNASeq)
library(ggbeeswarm)
library(editData)
library(tidyverse)
library(corrr)
library(ggcorrplot)
library(FactoMineR)
library(factoextra)
library(esquisse)
library(corto)
library(circlize)
library(ComplexHeatmap)
library(janitor)
```


# 1. Preparação dos dados

## 2. Definir padrões de figuras

####Cores
```{r}
# Definir variáveis de cores
vaxcolors = c(
  "BBIBP" = "#ffd000",
  "BNT" = "#56cfe1",
  "ChAd" = "#80ffdb",
  "ChAd-BNT" = "#72efdd",
  "ZF2001" = "#b5179e")

dose_colors = c(
    "1" = "#56cfe1",
    "2" = "#5978d4",
    "3" = "#b5179e")

day_colors = c(
    "1" = "#caf0f8",
    "3" = "#90e0ef",
    "6" = "#00b4d8",
    "7" = "#0077b6",
    "14" = "#023e8a",
    "28" = "#03045e")

type_colors = c('IN'= "#ffd000", #First generation vaccines
           'LA' = "#ffff3f",
           'CONJ' = "#f4a261",
           'VLP' = "#da1e37", #Second generation vaccines
           'SU' = "#ff4d6d",
           'PEP' = "#f48498",
           'VV' = "#72efdd", #Third generation vaccines
           'RNA' = "#56cfe1",
           'VV-RNA' = "#5e60ce",
           'IN-SU' = "#b5179e")

regimen_colors = c('HO' = "#56cfe1",
              'HE' = "#5e60ce")

immune_system_colors = c(
    "Innate" = "#faedcd",
    "Adaptive" = "#ffb5a7",
    "General" = "#99d98c",
    "Other" = "#e5e5e5",
    "Complement" = "#7400b8")

immune_subsystem_colors = c(
    "Adaptive" = "#ffb5a7",
    "Cellular" = "#ff758f",
    "Humoral" = "#ff758f",
    "Antibacterial" = "#4895ef",
    "Antifungal" = "#4361ee",
    "Antimicrobial" = "#7400b8",
    "General" = "#99d98c",
    "Antigen presentation" = "#e4c4a5",
    "APC" = "#ccd5ae",
    "Eosinophil" = "#faedcd",
    "Granulocyte" = "#faedcd",
    "Leukocyte" = "#ffb5a7",
    "Neutrophil" = "#ccd5ae",
    "Innate immune response" = "#faedcd",
    "Interferon" = "#a4c3b2",
    "Antiviral" = "#60495a",
    "Pattern recognition receptor" = "#f4f4d5",
    "Natural killer" = "#fedd96",
    "Mucosa" = "#c3dfe0",
    "Erythrocyte" = "#4895ef")



########## Colors dictionary

#Vaccines
ann_colors = list(
  vaccine = c(
  "BBIBP" = "#dee2e6",
  "BNT" = "#56cfe1",
  "ChAd" = "#80ffdb",
  "ChAd-BNT" = "#5e60ce",
  "ZF2001" = "#b5179e",
  "INFECTION" = "#ff4d6d",
  "INFECTION-VACCINE"= "#ffccd5"),
  dose = c(
    "NA" = "white",
    "0" = "white",
    "1" = "#56cfe1",
    "2" = "#5978d4",
    "3" = "#b5179e"),
  day = c(
    "0" = "white",
    "1" = "#caf0f8",
    "2" = "#ade8f4",
    "3" = "#90e0ef",
    "4" = "#6CD5EA",
    "5" = "#48cae4",
    "6" = "#00b4d8",
    "7" = "#0096c7",
    "10" = "#0087BF",
    "14" = "#0077b6",
    "26" = "#015BA0",
    "28" = "#023e8a",
    "51" = "#03045e"),
  type = c('IN'= "#e5e5e5", #1st Gen  vaccines
           'SU' = "#00a896",
           'VV' = "#72efdd", #3rd Gen vaccines
           'RNA' = "#56cfe1",
           'VV-RNA' = "#5e60ce", #Heterologous
           'IN-SU' = "#b5179e",
           'INFECTION'= "#da1e37"), #Infection
  regimen = c('HO' = "grey90",
              'HE' = "#8d99ae",
              "V-I-V" = "#36248f",
              "V-I" = "#D7B0EE",
              "I-V-I" = "#7400b8",
              "I-I" = "#5390d9",
              "I" = "#64dfdf"),
  infection = c("none" = "white",
                "1" = "#56cfe1",
                "2" = "#5e60ce"),
  variant = c("NA" = "white",
              "Beta" = "#72efdd",
              "Omicron" = "#5e60ce"),
  severity = c("A (9%) MI (81%) MO (0) S (4%)" = "#caf0f8",
               "A (29%) MI (57%) MO (14) S (0%)" = "#56cfe1",
               "A (0%) MI (89%) MO (11%) S (0%)" = "#72efdd",
               "A (9%) MI (47%) MO (21%) S (3%)" = "#64dfdf",
               "S" = "#5e60ce",
               "MO" = "#64dfdf",
               "MI" = "#72efdd",
               "NA" = "white"))

#Rows
# ann_genesets_heatmap
ann_colors_rows = list(
  `immune_system` = c(
    "Innate" = "#faedcd",
    "Adaptive" = "#ffb5a7",
    "General" = "#99d98c",
    "Other" = "#e5e5e5",
    "Complement" = "#7400b8",
    "No enrichment" = "gray90"
  ),
  `immune_sub_system` = c(
    "Adaptive" = "#ffb5a7",
    "Cellular" = "#ff758f",
    "Humoral" = "#ff758f",
    "Antibacterial" = "#4895ef",
    "Antifungal" = "#4361ee",
    "Antimicrobial" = "#7400b8",
    "General" = "#99d98c",
    "Antigen presentation" = "#e4c4a5",
    "APC" = "#ccd5ae",
    "Eosinophil" = "#faedcd",
    "Granulocyte" = "#faedcd",
    "Leukocyte" = "#ffb5a7",
    "Neutrophil" = "#ccd5ae",
    "Innate immune response" = "#faedcd",
    "Interferon" = "#a4c3b2",
    "Antiviral" = "#60495a",
    "Pattern recognition receptor" = "#f4f4d5",
    "Natural killer" = "#fedd96",
    "Mucosa" = "#c3dfe0",
    "Erythrocyte" = "#4895ef",
    "No enrichment" = "gray90"
  )
)
  

```



# 3. Seleção de estudos


```{r}
# GSE189039
# GSE201530
```

## 3.1 Baixar dados

```{r eval=FALSE, include=FALSE}
#Baixar dados de uma lista de estudos

############ Obter Tabela de counts
#GEO supplementary files. Baixar individualmente. Arquivos grandes devem ser baixados direto do navegador.
getGEOSuppFiles('GSE189039') # Baixar tabela de counts
untar('GSE189039_RAW.tar') # Descompactar tar zipped
#gunzip("GSE174621_RAW.tar") # Descompactar gunzip (.gz)


############ Metadados

filename = "GSE189039"
GSE189039 <- getGEO("GSE189039", GSEMatrix = TRUE) #Use GSEMatrix = TRUE para obter metadados de cada amostra
GSE189039 <- GSE189039[[1]] #Analisar o primeiro objeto da lista. 

#############Analisar informações sobre amostras. 

GSE189039_metadata = GSE189039 %>% 
  pData() %>% #Informações sobre amostras 
  select( #Selecionar e renomear colunas de interesse
    geo_sample = "geo_accession", 
    subject_id = "title", age = "age:ch1", 
    sex = "gender:ch1", disease_state = "diseas state:ch1", 
    severity = "severity:ch1") 

write.csv(GSE189039_metadata, file = "GSE189039_metadata.csv") #Terminar de anotar no excel

```


```{r eval=FALSE, include=FALSE}
############ Obter Tabela de counts
#GEO supplementary files. Baixar individualmente. Arquivos grandes devem ser baixados direto do navegador.
getGEOSuppFiles('GSE201530') # Baixar tabela de counts
untar('GSE201530_RAW.tar') # Descompactar tar zipped

############ Metadados

filename = "GSE201530"
GSE201530 <- getGEO("GSE201530", GSEMatrix = TRUE) #Use GSEMatrix = TRUE para obter metadados de cada amostra
GSE201530 <- GSE201530[[1]] #Analisar o primeiro objeto da lista. 

#############Analisar informações sobre amostras. 

GSE201530_metadata = GSE201530 %>% 
  pData() %>% #Informações sobre amostras 
  select(
    geo_sample = "geo_accession", 
    subject_id = "title", 
    age = "age:ch1", 
    sex = "gender:ch1", 
    disease_state = "disease state:ch1", 
    timepoint = "days after positive pcr results:ch1",
    group = "group (by covid-19 vaccination, prior infection):ch1",
    variant = "omicron sublineage:ch1") 

write.csv(GSE201530_metadata, file = "GSE201530_metadata.csv")

#Anotar no excel com dados do artigo
```

*Ler arquivos e transformar em tabela*

Abra a lista do estudo e analise suas informações. Ao descompactar, surgem diversos arquivos ".featureCounts.txt.gz", que devem ser descompactados. Para descompactar a lista de arquivos, use um loop.

```{r eval=FALSE, include=FALSE}
#Descompactar todos os arquivos
diretorio <- "~/Desktop/RNA seq - Wasim/GSE189039"
arquivos <- list.files(path = diretorio, pattern = "*.gz$", full.names = TRUE)
for(arquivo_gz in arquivos) {gunzip(arquivo_gz)}

#Ler arquivos do estudo específico
diretorio_GSE201530 <- "~/Desktop/RNA seq - Wasim/GSE201530"
diretorio_GSE189039 <- "~/Desktop/RNA seq - Wasim/GSE189039"

arquivos_txt_GSE201530 <- list.files(path = diretorio_GSE201530, pattern = "*.txt$", full.names = TRUE)
arquivos_txt_GSE189039 <- list.files(path = diretorio_GSE189039, pattern = "*.txt$", full.names = TRUE)

# Função para ler e ajustar os arquivos
ler_e_ajustar <- function(arquivo) {
  dados <- read.table(arquivo, header = FALSE, sep = "\t")
  if (ncol(dados) >= 2) {
    colnames(dados)[1:2] <- c("genes", tools::file_path_sans_ext(basename(arquivo)))
    return(dados)
  } else {
    return(NULL)
  }
}

# Ler e ajustar todos os arquivos na lista
lista_dados_GSE201530 <- lapply(arquivos_txt_GSE201530, ler_e_ajustar)
lista_dados_GSE189039 <- lapply(arquivos_txt_GSE189039, ler_e_ajustar)

# Filtrar elementos NULL da lista e unir os dataframes
counts_gse201530 <- lista_dados_GSE201530 %>%
  purrr::discard(is.null) %>%
  reduce(full_join, by = "genes")

counts_GSE189039 <- lista_dados_GSE189039 %>%
  purrr::discard(is.null) %>%
  reduce(full_join, by = "genes")

# Renomear as colunas exceto a primeira ("genes")
colnames(counts_gse201530)[-1] <- gsub("^[^_]+_(.*)", "\\1", colnames(counts_gse201530)[-1])
colnames(counts_GSE189039)[-1] <- gsub("^[^_]+_(.*)", "\\1", colnames(counts_GSE189039)[-1])


# Ler e ajustar todos os arquivos na lista
lista_dados_GSE201530 <- lapply(arquivos_txt_GSE201530, ler_e_ajustar)
lista_dados_GSE189039 <- lapply(arquivos_txt_GSE189039, ler_e_ajustar)

# Filtrar elementos NULL da lista e unir os dataframes
counts_gse201530 <- lista_dados_GSE201530 %>%
  purrr::discard(is.null) %>%
  reduce(full_join, by = "genes")

counts_GSE189039 <- lista_dados_GSE189039 %>%
  purrr::discard(is.null) %>%
  reduce(full_join, by = "genes")

# Renomear as colunas exceto a primeira ("genes")
colnames(counts_gse201530)[-1] <- gsub("^[^_]+_(.*)", "\\1", colnames(counts_gse201530)[-1])
colnames(counts_GSE189039)[-1] <- gsub("^[^_]+_(.*)", "\\1", colnames(counts_GSE189039)[-1])

#Salvar
write.csv(counts_gse201530, file = "gse201530_counts.csv")
write.csv(counts_GSE189039, file = "gse189039_counts.csv")

```



# 4. Análise de dados







## Demographics

```{r}
Demographics

esquisser(ann_vaccines_samples_4_12_23)
```





## 4.1. Análise de expressão genica diferencial

*Padronizando tabela de counts*

```{r eval=FALSE, include=FALSE}

############### GSE201530
# Ordenar tabela
gse201530_counts_ready <- counts_gse201530 %>%
  column_to_rownames("genes") %>% 
  t() %>%
  as.data.frame() %>% 
  rownames_to_column("sample") %>% 
  arrange(sample) %>% 
  column_to_rownames("sample") %>% 
  t() %>% 
  as.data.frame() %>%
  rownames_to_column("genes") %>% 
  filter(!grepl("_", genes)) %>% 
  column_to_rownames("genes")

# Salvar arquivo com rownames
saveRDS(gse201530_counts_ready, file = "gse201530_counts_ready.rds")


############### GSE189039
# Ordenar tabela
gse189039_counts_ready <- counts_GSE189039 %>%
  column_to_rownames("genes") %>% 
  t() %>%
  as.data.frame() %>% 
  rownames_to_column("sample") %>% 
  arrange(sample) %>% 
  column_to_rownames("sample") %>% 
  t() %>% 
  as.data.frame() %>%
  rownames_to_column("genes") %>% 
  filter(!grepl("_", genes)) %>% 
  column_to_rownames("genes")

# Salvar arquivo com rownames
saveRDS(gse189039_counts_ready, file = "gse189039_counts_ready.rds")

GSE189039_metadata

```

*Metadata*

A tabela de counts fornecida pelo estudo é nomeada com o Subject ID + Timepoint e não com o codigo GSM. Por isso, teremos de tratar a tabela de metadados. Além disso, ela não possui a coluna de idade. É possível manipular estes dados no excel.


*DESEQ2*

*Preparando dados*


## GSE201530

```{r}
###DESEQ2 - PF versus AZ, PF em diferentes tempos, AZ em diferentes tempos

####Definindo design

###### Padronize as variáveis 
countData <- gse201530_counts_ready
colData <- GSE201530_metadata %>%
  mutate(
    timepoint_day = factor(timepoint_day),
    age = factor(age),
    sex = factor(sex),
    disease_vac = factor(disease_vac),
    dose = factor(dose),
    disease_time = paste0(disease_vac, "_D", timepoint_day)) %>% 
  select(subject_id, everything())


###### Crie o objeto DESeqDataSet ######
# O design deve incluir todos os fatores que podem influenciar significativamente os dados. Coloque por último o fator de interesse principal - no caso, o fator de primeira ou segunda doses, pois nesta primeira etapa não serão analisados os efeitos de uma terceira dose de vacina heteróloga.

dds <- DESeqDataSetFromMatrix(countData = countData, 
                              colData = colData, 
                              design = ~ disease_time)

###### Pré-filtragem ######
#Mantenha os genes com pelo menos 10 leituras no total.

keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]

```

*Executando DESEQ2*

```{r}
###### Executar DESeq ######

#Executar DESeq com método Wald
dds <- DESeq(dds)

#Salvar
write_rds(dds,file="gse201530_dds_DESeq2_Wald_disease_vac_time.rds")

###### Obter comparações realizadas ######
resultsNames(dds)

dds = gse201530_dds_DESeq2_Wald_disease_vac_timepoint_day

###### Obtendo os DEGs totais ######
res <- results(dds) %>% 
  as.data.frame() %>% 
  filter(pvalue < 0.05)

#Salvar tabela
write.csv(res, file = "gse201530_dds_DESeq2_Wald_disease_vac_AND_timepoint_day_ALLDEGS.rds", row.names = TRUE)

```

*Comparando tempos e condições*

Nesta comparação, são encontradas DEGs de cada vacina individual nos diferentes tempos. Para comparar uma vacina com a outra em diferentes tempos, é necessário usar os argumentos "contrast" ou "name".

```{r}
#H-BNT-I
# contrast = c("disease_time", "H.BNT.I_D1", "H_D0")
# contrast = c("disease_time", "H.BNT.I_D2", "H_D0") (NAO DEU)
# contrast = c("disease_time", "H.BNT.I_D3", "H_D0")
# contrast = c("disease_time", "H.BNT.I_D4", "H_D0")

#I-BNT-I
# contrast = c("disease_time", "I.BNT.I_D2", "H_D0")
# contrast = c("disease_time", "I.BNT.I_D5", "H_D0")

#H-I
# contrast = c("disease_time", "H.I_D1", "H_D0")

#H-I-I
# contrast = c("disease_time", "H.I.I_D2", "H_D0")
#I-I
# contrast = c("disease_time", "I.I_D1", "H_D0")
# contrast = c("disease_time", "I.I_D3", "H_D0")
# contrast = c("disease_time", "I.I_D5", "H_D0")
```

*H_BNT_I* 
```{r}
######### H_BNT_I

#H_BNT_I_D1
gse201530_H_BNT_I_D1 = results(dds, contrast = c("disease_time", "H.BNT.I_D1", "H_D0")) %>% 
  as.data.frame() %>% 
  arrange(padj) %>% 
  mutate(condition = "H-BNT-I (D1)",
         day = 1,
         vaccine = "H-BNT-I",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse201530_H_BNT_I_D1, file = "gse201530_H_BNT_I_D1_DEGs_notfiltered.csv", row.names = TRUE)

#H_BNT_I_D3
gse201530_H_BNT_I_D3 = results(dds, contrast = c("disease_time", "H.BNT.I_D3", "H_D0")) %>%
  as.data.frame() %>%
  arrange(padj) %>%
  mutate(condition = "H-BNT-I (D3)",
         day = 3,
         vaccine = "H-BNT-I",
         direction = ifelse(log2FoldChange >= 1, "UP",
                            ifelse(log2FoldChange <= -1, "DOWN",
                                   "NEUTRAL"))) %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse201530_H_BNT_I_D3, file = "gse201530_H_BNT_I_D3_DEGs.csv", row.names = TRUE)

#H_BNT_I_D4
gse201530_H_BNT_I_D4 = results(dds, contrast = c("disease_time", "H.BNT.I_D4", "H_D0")) %>%
  as.data.frame() %>%
  filter(padj < 0.05) %>%
  arrange(padj) %>%
  mutate(condition = "H-BNT-I (D4)",
         day = 4,
         vaccine = "H-BNT-I",
         direction = ifelse(log2FoldChange >= 1, "UP",
                            ifelse(log2FoldChange <= -1, "DOWN",
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse201530_H_BNT_I_D4, file = "gse201530_H_BNT_I_D4_DEGs.csv", row.names = TRUE)

gse201530_H_BNT_I = bind_rows(gse201530_H_BNT_I_D1,
                            gse201530_H_BNT_I_D3,
                            gse201530_H_BNT_I_D4)

write.csv(gse201530_H_BNT_I, "gse201530_H_BNT_I.csv")

```


*I_BNT_I* 
```{r}
######### I_BNT_I

#D2
gse201530_I_BNT_I_D2 = results(dds, contrast = c("disease_time", "I.BNT.I_D2", "H_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "I-BNT-I (D2)",
         day = 2,
         vaccine = "I-BNT-I",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse201530_I_BNT_I_D2, file = "gse201530_I_BNT_I_D2_DEGs.csv")


#D5
gse201530_I_BNT_I_D5 = results(dds, contrast = c("disease_time", "I.BNT.I_D5", "H_D0")) %>%
  as.data.frame() %>%
  filter(padj < 0.05) %>%
  arrange(padj) %>%
  mutate(condition = "I-BNT-I (D5)",
         day = 5,
         vaccine = "I-BNT-I",
         direction = ifelse(log2FoldChange >= 1, "UP",
                            ifelse(log2FoldChange <= -1, "DOWN",
                                   "NEUTRAL"))) %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse201530_I_BNT_I_D5, file = "gse201530_H_BNT_I_D5_DEGs.csv")

gse201530_I_BNT_I = bind_rows(gse201530_I_BNT_I_D2,
                            gse201530_I_BNT_I_D5)

write.csv(gse201530_I_BNT_I, "gse201530_I_BNT_I.csv")

```


*H-I* 
```{r}
######### H-I

#D1
gse201530_H_I_D1 = results(dds, contrast = c("disease_time", "H.I_D1", "H_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "H-I (D1)",
         day = 1,
         vaccine = "H-I",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse201530_H_I_D1, file = "gse201530_H_I_D1_DEGs-PF.csv")


#D5
gse201530_H_I_D5 = results(dds, contrast = c("disease_time", "I.BNT.I_D5", "H_D0")) %>%
  as.data.frame() %>%
  filter(padj < 0.05) %>%
  arrange(padj) %>%
  mutate(condition = "H-I (D5)",
         day = 5,
         vaccine = "H-I",
         direction = ifelse(log2FoldChange >= 1, "UP",
                            ifelse(log2FoldChange <= -1, "DOWN",
                                   "NEUTRAL"))) %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse201530_H_I_D5, file = "gse201530_H_I_D5_DEGs-PF.csv")

gse201530_H_I = bind_rows(gse201530_I_BNT_I_D2,
                            gse201530_I_BNT_I_D5)

write.csv(gse201530_H_I, "gse201530_H_I.csv")

```

*H-I-I* 
```{r}

######### H-I-I
#D2
gse201530_H_I_I_D2 = results(dds, contrast = c("disease_time", "H.I.I_D2", "H_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "H-I-I (D2)",
         day = 2,
         vaccine = "H-I-I",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse201530_H_I_I_D2, file = "gse201530_H_I_I_D2_DEGs-PF.csv")

```

*I-I*
```{r}

######### I-I
#D1
gse201530_I_I_D1 = results(dds, contrast = c("disease_time", "I.I_D1", "H_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "I-I (D1)",
         day = 1,
         vaccine = "I-I",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse201530_I_I_D1, file = "gse201530_I_I_D1_DEGs-PF.csv")


#D3
gse201530_I_I_D3 = results(dds, contrast = c("disease_time", "I.I_D3", "H_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "I-I (D3)",
         day = 3,
         vaccine = "I-I",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse201530_I_I_D3, file = "gse201530_I_I_D3_DEGs-PF.csv")


#D5
gse201530_I_I_D5 = results(dds, contrast = c("disease_time", "I.I_D5", "H_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "I-I (D5)",
         day = 5,
         vaccine = "I-I",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse201530_I_I_D5, file = "gse201530_I_I_D5_DEGs-PF.csv")

gse201530_I_I = bind_rows(gse201530_I_I_D1,
                          gse201530_I_I_D3,
                          gse201530_I_I_D5)

```

Unir raw datasets 
```{r}
########## GSE201530

gse201530_DEGs = bind_rows(gse201530_I_I,
                           gse201530_H_I_I_D2,
                           gse201530_H_I,
                           gse201530_I_BNT_I,
                           gse201530_H_BNT_I) %>% 
  mutate(study = "GSE201530")

#Salvar
write.csv(gse201530_DEGs, file = "gse201530_DEGs_27-11-23.csv")

GSE201530_metadata_conditions = gse201530_DEGs %>% 
  select(condition:vaccine, study) %>% 
  distinct()


#Anotações
ann_vaccines_lower = ann_vaccines %>% 
  clean_names()

ann_vaccines_27_11_23 = bind_rows(ann_vaccines_lower, 
                                  GSE201530_metadata_conditions)

#Salvar
write.csv(ann_vaccines_27_11_23, file = 'ann_vaccines_27_11_23.csv')


```






## GSE189039

```{r}
###DESEQ2 - PF versus AZ, PF em diferentes tempos, AZ em diferentes tempos

####Definindo design

###### Padronize as variáveis 
countData <- gse189039_counts_ready
colData <- GSE189039_metadata %>%
  mutate(disease_vac = ifelse(disease_vac == "I", "H-I",
                              ifelse(disease_vac == "BNT-I", "H-BNT-I", disease_vac)),
         disease_vac_severity_time = paste0(disease_vac, "_", severity, "_D", timepoint_day),
         disease_vac_severity_time = factor(disease_vac_severity_time)) %>% 
  select(subject_id, everything())


###### Crie o objeto DESeqDataSet ######
dds <- DESeqDataSetFromMatrix(countData = countData, 
                              colData = colData, 
                              design = ~ disease_vac_severity_time)

###### Pré-filtragem ######
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]

###### Executar DESeq ######

#Método Wald
dds <- DESeq(dds)

#Salvar
write_rds(dds,file="gse189039_dds_DESeq2_Wald_disease_vac_severity_time.rds")

###### Obter comparações realizadas ######
resultsNames(dds)

###### Obtendo os DEGs totais ######
res <- results(dds) %>% 
  as.data.frame() %>% 
  filter(pvalue < 0.05)

#Salvar tabela
write.csv(res, file = "gse189039_dds_DESeq2_Wald_disease_vac_severity_time_ALLDEGS.rds")

```

*Comparando tempos e condições*

Nesta comparação, são encontradas DEGs de cada vacina individual nos diferentes tempos. Para comparar uma vacina com a outra em diferentes tempos, é necessário usar os argumentos "contrast" ou "name".

```{r}

## H-I
# MODERATE
# contrast = c("disease_vac_severity_time", "H.I_moderate_D10", "H_naive_vac_D0")
# contrast = c("disease_vac_severity_time", "H.I_moderate_D26", "H_naive_vac_D0")
# contrast = c("disease_vac_severity_time", "H.I_moderate_D51", "H_naive_vac_D0")

# SEVERE
# contrast = c("disease_vac_severity_time", "H.I_severe_D10", "H_naive_vac_D0")
# contrast = c("disease_vac_severity_time", "H.I_severe_D26", "H_naive_vac_D0")
# contrast = c("disease_vac_severity_time", "H.I_severe_D51", "H_naive_vac_D0")


## H-BNT-I
# MILD
# contrast = c("disease_vac_severity_time", "H.BNT.I_mild_D10", "H_naive_vac_D0")
# contrast = c("disease_vac_severity_time", "H.BNT.I_mild_D26", "H_naive_vac_D0")

# SEVERE
# contrast = c("disease_vac_severity_time", "H.BNT.I_severe_D10", "H_naive_vac_D0")
# contrast = c("disease_vac_severity_time", "H.BNT.I_severe_D26", "H_naive_vac_D0")

## BNT-I-BNT
# contrast = c("disease_vac_severity_time", "BNT.I.BNT_severe_D51", "H_naive_vac_D0")
contrast = c("disease_vac_severity_time", "BNT.I.BNT_mild_D51", "H_naive_vac_D0") (nao foi ???)

```

*H_I* 
```{r}
######### H_I

#H_I_D10
gse189039_H_I_moderate_D10 = results(dds, contrast = c("disease_vac_severity_time", "H.I_moderate_D10", "H_naive_vac_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "H-I (D10, moderate)",
         day = 10,
         vaccine = "H-I",
         severity = "moderate",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse189039_H_I_moderate_D10, file = "gse189039_H_I_moderate_D10_DEGs.csv")

#H_I_D26
gse189039_H_I_moderate_D26 = results(dds, contrast = c("disease_vac_severity_time", "H.I_moderate_D26", "H_naive_vac_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "H-I (D26, moderate)",
         day = 26,
         severity = "moderate",
         vaccine = "H-I",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse189039_H_I_moderate_D26, file = "gse189039_H_I_moderate_D26_DEGs.csv")

#H_I_D51
gse189039_H_I_moderate_D51 = results(dds, contrast = c("disease_vac_severity_time", "H.I_moderate_D51", "H_naive_vac_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "H-I (D51, moderate)",
         day = 51,
         vaccine = "H-I",
         severity = "moderate",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse189039_H_I_moderate_D51, file = "gse189039_H_I_moderate_D51_DEGs.csv")

#H_I_severe_D10
gse189039_H_I_severe_D10 = results(dds, contrast = c("disease_vac_severity_time", "H.I_severe_D10", "H_naive_vac_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "H-I (D10, severe)",
         day = 10,
         vaccine = "H-I",
         severity = "severe",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse189039_H_I_severe_D10, file = "gse189039_H_I_severe_D10_DEGs.csv")



#H_I_severe_D26
gse189039_H_I_severe_D26 = results(dds, contrast = c("disease_vac_severity_time", "H.I_severe_D26", "H_naive_vac_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "H-I (D26, severe)",
         day = 26,
         vaccine = "H-I",
         severity = "severe",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse189039_H_I_severe_D26, file = "gse189039_H_I_severe_D26_DEGs.csv")

#H_I_severe_D51
gse189039_H_I_severe_D51 = results(dds, contrast = c("disease_vac_severity_time", "H.I_severe_D51", "H_naive_vac_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "H-I (D51, severe)",
         day = 51,
         vaccine = "H-I",
         severity = "severe",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse189039_H_I_severe_D51, file = "gse189039_H_I_severe_D51_DEGs.csv")

gse189039_H_I = bind_rows(gse189039_H_I_moderate_D10,
                          gse189039_H_I_severe_D10,
                            gse189039_H_I_moderate_D26,
                          gse189039_H_I_severe_D26,
                            gse189039_H_I_moderate_D51,
                          gse189039_H_I_severe_D51)

write.csv(gse189039_H_I, "gse189039_H_I_DEGS.csv")

```


*H_BNT_I* 
```{r}
######### H-BNT-I

#H_BNT-I

#MILD
gse189039_H_BNT_I_mild_D10 = results(dds, contrast = c("disease_vac_severity_time", "H.BNT.I_mild_D10", "H_naive_vac_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "H-BNT-I (D10, mild)",
         day = 10,
         vaccine = "H-BNT-I",
         severity = "mild",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse189039_H_BNT_I_mild_D10, file = "gse189039_H_BNT_I_mild_D10_DEGs.csv")


gse189039_H_BNT_I_mild_D26 = results(dds, contrast = c("disease_vac_severity_time", "H.BNT.I_mild_D26", "H_naive_vac_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "H-BNT-I (D26, mild)",
         day = 26,
         vaccine = "H-BNT-I",
         severity = "mild",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse189039_H_BNT_I_mild_D26, file = "gse189039_H_BNT_I_mild_D26_DEGs.csv")


#SEVERE

gse189039_H_BNT_I_severe_D10 = results(dds, contrast = c("disease_vac_severity_time", "H.BNT.I_severe_D10", "H_naive_vac_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "H-BNT-I (D10, severe)",
         day = 10,
         vaccine = "H-BNT-I",
         severity = "severe",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse189039_H_BNT_I_severe_D10, file = "gse189039_H_BNT_I_severe_D10_DEGs.csv")


gse189039_H_BNT_I_severe_D26 = results(dds, contrast = c("disease_vac_severity_time", "H.BNT.I_severe_D26", "H_naive_vac_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "H-BNT-I (D26, severe)",
         day = 26,
         vaccine = "H-BNT-I",
         severity = "severe",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse189039_H_BNT_I_severe_D26, file = "gse189039_H_BNT_I_severe_D26_DEGs.csv")

#UNIR
gse189039_H_BNT_I = bind_rows(gse189039_H_BNT_I_mild_D10,
                              gse189039_H_BNT_I_severe_D10,
                              gse189039_H_BNT_I_mild_D26,
                              gse189039_H_BNT_I_severe_D26)
#Salvar
write.csv(gse189039_H_BNT_I, file = "gse189039_H_BNT_I_DEGS.csv")

```


*BNT-I-BNT*
```{r}
######### BNT-I-BNT	

# D51
gse189039_BNT_I_BNT_severe_D51 = results(dds, contrast = c("disease_vac_severity_time",
                                                         "BNT.I.BNT_severe_D51", 
                                                         "H_naive_vac_D0")) %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  arrange(padj) %>% 
  mutate(condition = "BNT-I-BNT (D51, severe)",
         day = 51,
         vaccine = "BNT-I-BNT",
         severity = "severe",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes")

#Salvar tabela
write.csv(gse189039_BNT_I_BNT_severe_D51, file = "gse189039_BNT_I_BNT_severe_D51_DEGs.csv")


#Aqui tive de usar o argumento name, pois contrast não estava funcionando (?)
gse189039_BNT_I_BNT_mild_D51 = results(dds, name= "disease_vac_severity_time_H_naive_vac_D0_vs_BNT.I.BNT_mild_D51") %>% 
  as.data.frame() %>% 
  filter(padj < 0.05) %>%
  mutate(log2FoldChange = ifelse(log2FoldChange > 0, -log2FoldChange, abs(log2FoldChange))) %>% 
  arrange(padj) %>% 
  mutate(condition = "BNT-I-BNT (D51, mild)",
         day = 51,
         vaccine = "BNT-I-BNT",
         severity = "mild",
         direction = ifelse(log2FoldChange >= 1, "UP", 
                            ifelse(log2FoldChange <= -1, "DOWN", 
                                   "NEUTRAL")))  %>% 
  rownames_to_column("genes") 

#Salvar tabela
write.csv(gse189039_BNT_I_BNT_mild_D51, file = "gse189039_BNT_I_BNT_mild_D51_DEGs.csv")

#Unir
gse189039_BNT_I_BNT = bind_rows(gse189039_BNT_I_BNT_mild_D51,
                                gse189039_BNT_I_BNT_severe_D51)


write.csv(gse189039_BNT_I_BNT, file = "gse189039_BNT_I_BNT_DEGS.csv")

```

*UNIR*
```{r}
gse189039_DEGs = bind_rows(gse189039_BNT_I_BNT,
                           gse189039_H_I,
                           gse189039_H_BNT_I) %>% 
  mutate(study = "GSE189039")



##### TODAS AS DEGS

gse_201530_189039 = bind_rows(gse201530_DEGs,
                              gse189039_DEGs) %>% 
  select(genes:condition, direction) %>% 
  merge(ann_vaccines, by = "condition", all.y = F) %>% 
  clean_names()

degs_allstudies_updown_p_05 = degs_allstudies_updown_p_05 %>% 
  clean_names() %>% 
  select(genes:direction, -vaccine, -gse_id) %>% 
  merge(ann_vaccines, by = "condition", all.y = F) %>% 
  select(-participants)


all_degs_p_05_vac_infected = bind_rows(degs_allstudies_updown_p_05,
                                       gse_201530_189039) %>% 
  mutate_all(~ replace(., is.na(.), "NA")) %>% 
  mutate(regimen = case_when(
regimen == "BNT-I-BNT" ~ "V-I-V",
regimen == "H-BNT-I" ~ "V-I",
regimen == "VAC-I" ~ "V-I",
regimen == "I-VAC-I" ~ "I-V-I",
regimen == "H-V-I" ~ "V-I",
regimen == "H-I-I" ~ "I-I",
regimen == "H-I" ~ "I",
TRUE ~ as.character(regimen)))

#Salvar
write.csv(gse_201530_189039, file = "gse_201530_189039_DEGs_27-11-23.csv")

#Salvar
write.csv(all_degs_p_05_vac_infected, file = 'all_degs_p_05_vac_infected_27-11-23.csv')
```


## GSE199750

```{r}
######### PFIZER ######### 

######### Tempos 1 versus 0 ######### 

# Obter comparação
res_1_0_bnt <- results(dds, contrast= c("Group_I_II", "BNT162b2V1", "BNT162b2V0")) # Acessar comparação
res_1_0_bnt <- subset(res_1_0_bnt, padj < 0.05) #excluir genes com padj > 5%
res_1_0_bnt <- subset(res_1_0_bnt, log2FoldChange > 1.0 | log2FoldChange < -1.0)
res_1_0_bnt <- res_1_0_bnt[order(res_1_0_bnt$padj),]  #ordenar por padj

#Obtendo DEGs UP and DOWN
res_up <- subset(res, log2FoldChange > 1.0 )
res_down <- subset(res, log2FoldChange < -1.0 )

# Criar uma tabela de DEGs
res_1_0_bnt_degs_table <- data.frame(Gene = rownames(res_1_0_bnt),
                          log2FoldChange = res_1_0_bnt$log2FoldChange,
                          pvalue = res_1_0_bnt$pvalue,
                          padj = res_1_0_bnt$padj)
#Salvar tabela
write.csv(res_1_0_bnt_degs_table, file = "GSE199750_T0-T1_DEGs-PF.csv", row.names = TRUE)
saveRDS(res_1_0_bnt_degs_table, file = "GSE199750_T0-T1_DEGs-PF.rds")

######### Tempos 2 versus 0 ######### 

# Obter comparação
res_2_0_bnt <- results(dds, contrast= c("Group_I_II","BNT162b2V2A","BNT162b2V0")) # Acessar comparação usando contrast. O valor referencia é sempre o segundo termo definido.
res_2_0_bnt <- subset(res_2_0_bnt, padj < 0.05) #excluir genes com padj > 5%
res_2_0_bnt <- subset(res_2_0_bnt, log2FoldChange > 1.0 | log2FoldChange < -1.0)
res_2_0_bnt <- res_2_0_bnt[order(res_2_0_bnt$padj),]  #ordenar por padj

#Obtendo DEGs UP and DOWN
res_up <- subset(res, log2FoldChange > 1.0 )
res_down <- subset(res, log2FoldChange < -1.0 )

# Criar uma tabela de DEGs
res_2_0_bnt_degs_table <- data.frame(Gene = rownames(res_2_0_bnt),
                          log2FoldChange = res_2_0_bnt$log2FoldChange,
                          pvalue = res_2_0_bnt$pvalue,
                          padj = res_2_0_bnt$padj)
#Salvar tabela
write.csv(res_2_0_bnt_degs_table, file = "GSE199750_T0-T2_DEGs-PF.csv", row.names = TRUE)
saveRDS(res_2_0_bnt_degs_table, file = "GSE199750_T0-T2_DEGs-PF.rds")

######### PFIZER ######### 
# Obter comparação
resultsNames(dds)
res_3_0_bnt_bnt <- results(dds, contrast= c("AllVaccines", "V3A_BNT162b2_3A_BNT162b2", "V0_BNT162b2_3A_BNT162b2"))
res_3_0_bnt_bnt <- subset(res_3_0_bnt_bnt, padj < 0.05) #excluir genes com padj > 5%
res_3_0_bnt_bnt <- subset(res_3_0_bnt_bnt, log2FoldChange > 1.0 | log2FoldChange < -1.0)
res_3_0_bnt_bnt <- res_3_0_bnt_bnt[order(res_3_0_bnt_bnt$padj),]  #ordenar por padj

#Obtendo DEGs UP and DOWN
res_3_0_bnt_bnt_up <- subset(res_3_0_bnt_bnt, log2FoldChange > 1.0 )
res_3_0_bnt_bnt_down <- subset(res_3_0_bnt_bnt, log2FoldChange < -1.0 )

# Criar uma tabela de DEGs
res_3_0_bnt_bnt_degs_table <- data.frame(res_3_0_bnt_bnt)
#Salvar tabela
write.csv(res_3_0_bnt_bnt_degs_table, file = "GSE199750_T3-T0_DEGs_2PF-1PF.csv", row.names = TRUE)

######### PFIZER ######### 
res_3_0_bnt_mo <- results(dds, contrast= c("AllVaccines", "V3A_BNT162b2_3A_mRNA-1273", "V0_BNT162b2_3A_BNT162b2"))
res_3_0_bnt_mo <- subset(res_3_0_bnt_mo, padj < 0.05) #excluir genes com padj > 5%
res_3_0_bnt_mo <- subset(res_3_0_bnt_mo, log2FoldChange > 1.0 | log2FoldChange < -1.0)
res_3_0_bnt_mo <- res_3_0_bnt_mo[order(res_3_0_bnt_mo$padj),]  #ordenar por padj

#Obtendo DEGs UP and DOWN
res_3_0_bnt_mo_up <- subset(res_3_0_bnt_mo, log2FoldChange > 1.0 )
res_3_0_bnt_mo_down <- subset(res_3_0_bnt_mo, log2FoldChange < -1.0 )

# Criar uma tabela de DEGs
res_3_0_bnt_mo_degs_table <- data.frame(res_3_0_bnt_mo)
#Salvar tabela
write.csv(res_3_0_bnt_mo_degs_table, file = "GSE199750_T3-T0_DEGs_2PF-1MO.csv", row.names = TRUE)

######### PFIZER ######### 
res_3_2_bnt_bnt <- results(dds, contrast= c("AllVaccines", "V3A_BNT162b2_3A_BNT162b2", "V2A_BNT162b2_3A_BNT162b2"))
res_3_2_bnt_bnt <- subset(res_3_2_bnt_bnt, padj < 0.05) #excluir genes com padj > 5%
res_3_2_bnt_bnt <- subset(res_3_2_bnt_bnt, log2FoldChange > 1.0 | log2FoldChange < -1.0)
res_3_2_bnt_bnt <- res_3_2_bnt_bnt[order(res_3_2_bnt_bnt$padj),]  #ordenar por padj

#Obtendo DEGs UP and DOWN
res_3_2_bnt_bnt_up <- subset(res_3_2_bnt_bnt, log2FoldChange > 1.0 )
res_3_2_bnt_bnt_down <- subset(res_3_2_bnt_bnt, log2FoldChange < -1.0 )

# Criar uma tabela de DEGs
res_3_2_bnt_bnt_degs_table <- data.frame(res_3_2_bnt_bnt)
#Salvar tabela
write.csv(res_3_2_bnt_bnt_degs_table, file = "GSE199750_T3-T2_DEGs_2PF-1PF.csv", row.names = TRUE)

######### PFIZER ######### 
res_3_2_bnt_mo <- results(dds, contrast= c("AllVaccines", "V3A_BNT162b2_3A_mRNA-1273", "V2A_BNT162b2_3A_BNT162b2"))
res_3_2_bnt_mo <- subset(res_3_2_bnt_mo, padj < 0.05) #excluir genes com padj > 5%
res_3_2_bnt_mo <- subset(res_3_2_bnt_mo, log2FoldChange > 1.0 | log2FoldChange < -1.0)
res_3_2_bnt_mo <- res_3_2_bnt_mo[order(res_3_2_bnt_mo$padj),]  #ordenar por padj

#Obtendo DEGs UP and DOWN
res_3_2_bnt_mo_up <- subset(res_3_2_bnt_mo, log2FoldChange > 1.0 )
res_3_2_bnt_mo_down <- subset(res_3_2_bnt_mo, log2FoldChange < -1.0 )

# Criar uma tabela de DEGs
res_3_2_bnt_mo_degs_table <- data.frame(res_3_2_bnt_mo)
#Salvar tabela
write.csv(res_3_2_bnt_mo_degs_table, file = "GSE199750_T3-T2_DEGs_2PF-1MO.csv", row.names = TRUE)

#### ASTRAZENECA

##### Tempos 0 e 1

resultsNames(dds)
######### ASTRAZENECA ######### 
# Obter comparação
res_1_0_chad <- results(dds, contrast= c("Group_I_II","ChAdOx1V1","ChAdOx1V0"))
res_1_0_chad <- subset(res_1_0_chad, padj < 0.05)
res_1_0_chad <- subset(res_1_0_chad, log2FoldChange > 1.0 | log2FoldChange < -1.0)
res_1_0_chad <- res_1_0_chad[order(res_1_0_chad$padj),]  #ordenar por padj

#Obtendo DEGs UP and DOWN
res_up <- subset(res, log2FoldChange > 1.0 )
res_down <- subset(res, log2FoldChange < -1.0 )

# Criar uma tabela de DEGs
res_1_0_chad_degs_table <- data.frame(Gene = rownames(res_1_0_chad),
                          log2FoldChange = res_1_0_chad$log2FoldChange,
                          pvalue = res_1_0_chad$pvalue,
                          padj = res_1_0_chad$padj)
#Salvar tabela
write.csv(res_1_0_chad_degs_table, file = "GSE199750_T0-T1_DEGs-AZ.csv", row.names = TRUE)
write.csv(res_1_0_chad_degs_table, file = "GSE199750_T0-T1_DEGs-AZ.rds")

######### Tempos 2 versus 0 ######### 
# Obter comparação
res_2_0_chad <- results(dds, contrast= c("Group_I_II","ChAdOx1V2A","ChAdOx1V0"))
res_2_0_chad <- subset(res_2_0_chad, padj < 0.05) #excluir genes com padj > 5%
res_2_0_chad <- subset(res_2_0_chad, log2FoldChange > 1.0 | log2FoldChange < -1.0)
res_2_0_chad <- res_2_0_chad[order(res_2_0_chad$padj),]  #ordenar por padj
summary(res_2_0_chad) #Up e Down

# Criar uma tabela de DEGs
res_2_0_chad_degs_table <- data.frame(Gene = rownames(res_2_0_chad),
                          log2FoldChange = res_2_0_chad$log2FoldChange,
                          pvalue = res_2_0_chad$pvalue,
                          padj = res_2_0_chad$padj)
#Salvar tabela
write.csv(res_2_0_chad_degs_table, file = "GSE199750_T0-T2_DEGs-AZ.csv", row.names = FALSE)
saveRDS(res_2_0_chad_degs_table, file = "GSE199750_T0-T2_DEGs-AZ.rds")

######### Tempos 2 versus 1 ######### 
# Obter comparação
res_2_1_chad <- results(dds, contrast= c("Group_I_II","ChAdOx1V2A","ChAdOx1V1")) # Acessar comparação usando contrast. O valor referencia é sempre o segundo termo definido.
res_2_1_chad <- subset(res_2_1_chad, padj < 0.05) #excluir genes com padj > 5%
res_2_1_chad <- subset(res_2_1_chad, log2FoldChange > 1.0 | log2FoldChange < -1.0)
res_2_1_chad <- res_2_1_chad[order(res_2_1_chad$padj),]  #ordenar por padj
summary(res_2_1_chad) #Up e Down

# Criar uma tabela de DEGs
res_2_1_chad_degs_table <- data.frame(Gene = rownames(res_2_1_chad),
                          log2FoldChange = res_2_1_chad$log2FoldChange,
                          pvalue = res_2_1_chad$pvalue,
                          padj = res_2_1_chad$padj)
#Salvar tabela
write.csv(res_2_1_chad_degs_table, file = "GSE199750_T1-T2_DEGs-AZ.csv", row.names = FALSE)

######### ASTRAZENECA ######### 
# Obter comparação
resultsNames(dds)
res_3_0_chad_bnt <- results(dds, contrast= c("AllVaccines", "V3A_ChAdOx1_3A_BNT162b2", "V0_ChAdOx1_3A_Missing"))
res_3_0_chad_bnt <- subset(res_3_0_chad_bnt, padj < 0.05) #excluir genes com padj > 5%
res_3_0_chad_bnt <- subset(res_3_0_chad_bnt, log2FoldChange > 1.0 | log2FoldChange < -1.0)
res_3_0_chad_bnt <- res_3_0_chad_bnt[order(res_3_0_chad_bnt$padj),]  #ordenar por padj

#Obtendo DEGs UP and DOWN
res_3_0_chad_bnt_up <- subset(res_3_0_chad_bnt, log2FoldChange > 1.0 )
res_3_0_chad_bnt_down <- subset(res_3_0_chad_bnt, log2FoldChange < -1.0 )

# Criar uma tabela de DEGs
res_3_0_chad_bnt_degs_table <- data.frame(res_3_0_chad_bnt)
#Salvar tabela
write.csv(res_3_0_chad_bnt_degs_table, file = "GSE199750_T3-T0_DEGs_2AZ-1PF.csv", row.names = TRUE)

######### ASTRAZENECA ######### 
# Obter comparação
resultsNames(dds)
res_3_2_chad_bnt <- results(dds, contrast= c("AllVaccines", "V3A_ChAdOx1_3A_BNT162b2", "V2A_ChAdOx1_3A_Missing"))
res_3_2_chad_bnt <- subset(res_3_2_chad_bnt, padj < 0.05) #excluir genes com padj (=False discovery rate) > 5%
res_3_2_chad_bnt <- subset(res_3_2_chad_bnt, log2FoldChange > 1.0 | log2FoldChange < -1.0)
res_3_2_chad_bnt <- res_3_2_chad_bnt[order(res_3_2_chad_bnt$padj),]  #ordenar por padj

#Obtendo DEGs UP and DOWN
res_3_2_chad_bnt_up <- subset(res_3_2_chad_bnt, log2FoldChange > 1.0 )
res_3_2_chad_bnt_down <- subset(res_3_2_chad_bnt, log2FoldChange < -1.0 )

# Criar uma tabela de DEGs
res_3_2_chad_bnt_degs_table <- data.frame(res_3_2_chad_bnt)
#Salvar tabela
write.csv(res_3_2_chad_bnt_degs_table, file = "GSE199750_T3-T2_DEGs_2AZ-1PF.csv", row.names = TRUE)
```




Combinar tabela de genes sequenciados (raw)

```{r}
#TABELA DE GENES SEQUENCIADOS

gse201530_genes = gse201530_counts_ready %>%
  rownames_to_column("genes") %>% 
  select(genes) %>% 
  mutate(study = "GSE201530")

gse199750_genes = gse199750_counts_ready %>%
  as.data.frame() %>% 
  rownames_to_column("genes") %>% 
  select(genes) %>% 
  mutate(study = "GSE199750")

gse189039_genes = gse189039_counts_ready %>%
  as.data.frame() %>% 
  rownames_to_column("genes") %>% 
  select(genes) %>% 
  mutate(study = "GSE189039")

gse201533_genes = gse201533_counts_ready %>%
  as.data.frame() %>% 
  rownames_to_column("genes") %>% 
  select(genes) %>% 
  mutate(study = "GSE201533")

gse206023_genes = gse206023_counts_long_annotated_14_11_23 %>% 
  select(genes) %>% 
  distinct() %>% 
  mutate(study = "GSE206023")


genes_raw_combined = bind_rows(gse206023_genes,
                               gse201533_genes,
                               gse189039_genes,
                               gse199750_genes,
                               gse201530_genes)


write.csv(genes_raw_combined, file = "genes_raw_combined.csv")

```


# 5. Visualização de dados

## DEGs plot

```{r}
ann_vaccines = ann_vaccines_6_12_23 %>% clean_names()

degs_updown_p_05_vac_infected = degs_updown_p_05_vac_infected %>% 
  mutate(direction_wide = case_when(log2fold_change > 0 ~ "UP", 
                                    log2fold_change < 0 ~ "DOWN",
                                    TRUE ~ "NEUTRAL")) %>% 
  select(condition:padj, direction_wide) %>% 
  merge(ann_vaccines, by = "condition", all = T)

write.csv(degs_updown_p_05_vac_infected, file = "degs_updown_p_05_vac_infected_11-12-23.csv")

all_degs = degs_updown_p_05_vac_infected %>% 
  group_by(condition, direction_wide) %>% 
  summarise(n_degs = n()) %>% 
  merge(ann_vaccines, by="condition", all = T)

# Plot 1 VACCINES

ggplot_degs_bars_vaccines = all_degs %>%
  filter(disease_vac == "V",
         day != 0,
         !(vaccine %in% c("ZF2001", "BBIBP")),
         !(condition %in% c("BNT (V1, D0)", "BNT-MO (V1, D6)"))) %>% 
  mutate(vaccine = if_else(grepl("ChAd-BNT", condition), "ChAd-BNT", vaccine),
         type = if_else(grepl("ChAd-BNT", condition), "VV-RNA", type)) %>% 
 ggplot() +
  aes(x = reorder(condition, dose), fill = direction_wide, weight = n_degs) +
  geom_bar(position = position_dodge2(width = 0.9, preserve = "single")) +
  geom_text(aes(label = n_degs, y = n_degs), 
            position = position_dodge2(width = 0.9, preserve = "single"), 
            vjust = -1, 
            hjust = 0.5, 
            size = 2) +
  scale_fill_manual(
    values = c(DOWN = "#0E252D",
    NEUTRAL = "grey90",
    UP = "#3399FF")) +
  theme_minimal() +
  facet_wrap(~type, scales = "free_x", ncol = 1) +
  theme(legend.position = "top",
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1, size = 7),
        axis.text.y = element_blank(), 
        axis.ticks.y = element_blank(),
        panel.grid = element_blank()) +
  ylim(0, 1300) + 
  ggtitle("DEGs - all studies") 

ggplot_degs_bars_vaccines
ggsave(ggplot_degs_bars_vaccines, file = "DEGs_up_down_barplot_vaccines_1.png",  width = 10, height = 6)

# PLOT 2 INFECTION AND VACCINE

ggplot_degs_bars_infection = all_degs %>%
  filter(disease_vac == "I",
         direction_wide != "NEUTRAL",
         day != 0,
         !(vaccine %in% c("ZF2001", "BBIBP")),
         !(condition %in% c("BNT-I (D0)"))) %>% 
  mutate(type = if_else(grepl("BNT", condition), "V", type),
         condition = case_when(condition == "BNT-I-BNT (D51, mild)" ~ "BNT-I-BNT (51 DPI/10 Dpi, mild)",
                             condition == "BNT-I-BNT (D51, severe)" ~ "BNT-I-BNT (51 DPI/10 Dpv, severe)",
                             TRUE ~ condition)) %>% 
 ggplot() +
  aes(x = reorder(condition, dose), fill = direction_wide, weight = n_degs) +
  geom_bar(position = position_dodge2(width = 0.9, preserve = "single")) +
  geom_text(aes(label = n_degs, y = n_degs), 
            position = position_dodge2(width = 0.9, preserve = "single"), 
            vjust = -1, 
            hjust = 0.5, 
            size = 2) +
  scale_fill_manual(
    values = c(DOWN = "#0E252D",
    NEUTRAL = "grey90",
    UP = "#3399FF")) +
  theme_minimal() +
  facet_grid(~type, scales = "free_x", space = "free_x") +
  theme(legend.position = "top",
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1, size = 7),
        axis.text.y = element_blank(), 
        axis.ticks.y = element_blank(),
        panel.grid = element_blank()) +
  ylim(0, 1300) + 
  ggtitle("DEGs - Infected, vaccinated") 

ggplot_degs_bars_infection
ggsave(ggplot_degs_bars_infection, file = "DEGs_up_down_barplot_infection_1.png",  width = 10, height = 6)



# PLOT 3 UNITED

ggplot_degs_bars_vaccines = all_degs %>%
  filter(day != 0,
         direction_wide != "NEUTRAL",
         !(vaccine %in% c("ZF2001", "BBIBP")),
         !(condition %in% c("BNT (V1, D0)", "BNT-MO (V1, D6)"))) %>% 
  mutate(vaccine = if_else(grepl("ChAd-BNT", condition), "ChAd-BNT", vaccine),
         type = if_else(grepl("ChAd-BNT", condition), "VV-RNA", type)) %>% 
  mutate(type = if_else(grepl("BNT-I", condition), "V-I", type),
         type = if_else(grepl("I-BNT", condition), "V-I", type),
         condition = case_when(condition == "BNT-I-BNT (D51, mild)" ~ "BNT-I-BNT (51 DPI/10 Dpi, mild)",
                             condition == "BNT-I-BNT (D51, severe)" ~ "BNT-I-BNT (51 DPI/10 Dpv, severe)",
                             TRUE ~ condition)) %>% 
   mutate(type = factor(type, levels = c("VV", "RNA", "VV-RNA", "I", "V-I"))) %>% 
 ggplot() +
  aes(x = reorder(condition, dose), fill = direction_wide, weight = n_degs) +
  geom_bar(position = position_dodge2(width = 0.9, preserve = "single")) +
  geom_text(aes(label = n_degs, y = n_degs), 
            position = position_dodge2(width = 0.9, preserve = "single"), 
            vjust = -1, 
            hjust = 0.5, 
            size = 2) +
  scale_fill_manual(
    values = c(DOWN = "#0E252D",
    NEUTRAL = "grey90",
    UP = "#3399FF")) +
  theme_minimal() +
  facet_grid(~type, scales = "free_x", space = "free_x") +
  theme(legend.position = "top",
        axis.text.x = element_text(angle = 45, vjust = 1, hjust=1, size = 7),
        axis.text.y = element_blank(), 
        axis.ticks.y = element_blank(),
        panel.grid = element_blank()) +
  ylim(0, 1300) + 
  ggtitle("DEGs - all studies") 

ggplot_degs_bars_vaccines
ggsave(ggplot_degs_bars_vaccines, file = "DEGs_up_down_barplot_vaccines_infection_2.png", width = 15, height = 4.5)

```


## 5.1. Volcanoplot

Os volcanoplots das duas vacinas são os mesmos, mas espelhados. Isso significa que definir o nivel de referencia no fator de vacinas é importante para determinar as DEGs up e down. \#### ASTRAZENECA

*Astrazeneca* 

```{r}

#All
volcano_res <- EnhancedVolcano(res, 
                                    lab = rownames(res),
                                    x = 'log2FoldChange',
                                    y = 'pvalue',
                                    title='Differential expression of genes - All timepoints - ChAdOx1')
# Save the plot as a PNG file
png("gse199750_volcanoplot_DEGs_allfactores_AZ.png", width = 800, height = 600)
print(volcano_res_chad)
dev.off()

#T1-T0
volcano_res_chad_1_0 <- EnhancedVolcano(res_1_0_chad, 
                                    lab = rownames(res_1_0_chad),
                                    x = 'log2FoldChange',
                                    y = 'pvalue',
                                    title='Differential expression of genes - D1 vs. D0 - ChAdOx1')
# Save the plot as a PNG file
png("gse199750_volcanoplot_DEGs_D0-D1_AZ.png", width = 800, height = 600)
print(volcano_res_chad_1_0)
dev.off()

#T2-T0
volcano_res_chad_2_0 <- EnhancedVolcano(res_time0_time2_chad, 
                                    lab = rownames(res_time0_time2_chad),
                                    x = 'log2FoldChange',
                                    y = 'pvalue',
                                    title='Differential expression of genes - D2 vs. D0 - ChAdOx1')
# Save the plot as a PNG file
png("gse199750_volcanoplot_DEGs_D0-D2_AZ.png", width = 800, height = 600)
print(volcano_res_chad_2_0)
dev.off()

#Plotar mais de um gráfico
plot_grid(volcano_res_chad, volcano_res_chad_1_0, volcano_res_chad_2_0, ncol = 3) 

```

## 5.2. Heatmap

```{r}
# teste heatmap
head(res)

#Obter DEGs do estudo
heat_test = res %>% 
  as.data.frame() %>% 
  rownames_to_column(var="genes")

#Obter counts dos DEGs
heat_count = countData %>% 
  rownames_to_column(var = "genes")

#Unir tabelas
###A tabela agora possui os valores de counts e outras estatísticas do objeto deseq2 pareadas pelos DEGs.
merged_heat = heat_test %>% 
  merge(heat_count, by = 'genes') %>% 
  arrange(padj)

#Cores
tp_cols = c("V0" = "steelblue2", "V1" = "goldenrod","V2A"="firebrick3","V3A"="grey44")
v_cols = c("ChAdOx1" = "violetred2", "BNT162b2" = "deepskyblue2","mRNA-1273" = "goldenrod")

# Transformar as counts dos DEGs em matriz e normalizar 
data2plot <- as.matrix(merged_heat[, 8:267])
zzz <- scale(t(data2plot)) %>%
  pmin(-3) %>%
  pmax(3))

#Transformar em dataframe
test <- zzz %>%
  as.data.frame() %>%
  mutate(genes = merged_heat$genes) %>%
  tidyr::gather("samples", "expression", -genes)
tail(test)

zzz_filtered = test %>% 
  subset(expression > 2 & samples == 'COVIRS_41_V0')
dim(zzz_filtered)
head(zzz_filtered)

#Heatmap annotation
ha = HeatmapAnnotation(df = colData[,c("timepoint","first_second_dosef")],
                       col = list(Timepoint = tp_cols,
                                  FirstSecondDosef = v_cols))

Heatmap(zzz, 
        show_row_names = FALSE,
        show_column_names = FALSE,
        cluster_rows=TRUE,
        cluster_columns = TRUE,
        top_annotation = ha,
        col=c("darkgreen","azure2","darkorchid4"),
        column_split=colData$Timepoint,
        column_title="Heatplot")

```

## 5.3. Diagrama de Venn com Heatmap

*Tabela de overlaps de DEGs com teste exato de fisher*

Determinar a proporção de DEGs compartilhadas entre condições diferentes e sua significância. O código abaixo deve ter como input uma tabela longa chamada "degs_all" (coluna 1: condicoes/vacinas, coluna 2: DEGs). O output será uma tabela longa com as colunas Vacina 1, Vacina 2, shared degs, not shared vacina 1, not shared vacina 2, lista com nomes das degs, valor p de fisher (teste exato de fisher).

```{r}

#Converter tabela de DEGs Up and Down em tabela longa
degs_all_updown = degs_updown_p_05_vac_infected_11_12_23 %>% 
  filter(direction_wide != "NEUTRAL") %>% 
  filter(!(vaccine %in% c("BBIBP", "ZF2001")))
#Salvar
write.csv(degs_all_updown, file= "degs_updown_p<05_vac_infected.csv")

# Tabela de dados

#Up vs Up
degs_venn_up = degs_all_updown %>% 
  filter(direction_wide == "UP") %>%
  mutate(comparison = "UP-UP") %>% 
  select(condition, genes)

#Down vs Down
degs_venn_down = degs_all_updown %>% 
  filter(direction_wide == "DOWN") %>%
  mutate(comparison = "DOWN-DOWN") %>% 
  select(condition, genes)


# DOWN-UP ALL
degs_venn_updown = degs_all_updown %>% 
  filter(direction_wide %in% c("UP", "DOWN")) %>%
  mutate(comparison = "UP_AND_DOWN") %>% 
  select(condition, genes)

#Up vs Down
degs_venn_up_vs_down = degs_all_updown %>% 
  mutate(comparison = "UP_vs_DOWN") %>% 
  select(condition, genes)

```


```{r}
#Input
data = degs_venn_updown
filename = "degs_venn_updown_<0>_Totalgenes"

# Função para calcular a sobreposição entre pares de vacinas e a porcentagem de genes compartilhados
overlap_genes <- function(cond1, cond2, data) {
  genes_cond1 <- data$genes[data$condition == cond1] #Trocar coluna
  genes_cond2 <- data$genes[data$condition == cond2] #Trocar coluna
  genes_shared <- intersect(genes_cond1, genes_cond2)
  genes_notshared_cond1 <- setdiff(genes_cond1, genes_cond2)
  genes_notshared_cond2 <- setdiff(genes_cond2, genes_cond1)
  
  total_genes_cond1 <- length(genes_cond1)
  total_genes_cond2 <- length(genes_cond2)
  
  percentage_shared_cond1 <- length(genes_shared) / total_genes_cond1 * 100
  percentage_shared_cond2 <- length(genes_shared) / total_genes_cond2 * 100
  
  shared_genes <- data.frame(
    Cond1 = cond1,
    Cond2 = cond2,
    Shared = length(genes_shared),
    NotShared_cond1 = length(genes_notshared_cond1),
    NotShared_cond2 = length(genes_notshared_cond2),
    Total_Genes_Cond1 = total_genes_cond1,
    Total_Genes_Cond2 = total_genes_cond2,
    Genes_Names = paste(genes_shared, collapse = ", "),
    Percentage_Shared_Cond1 = percentage_shared_cond1,
    Percentage_Shared_Cond2 = percentage_shared_cond2
  )
  
  return(shared_genes)
}
# Obtenha uma lista de todas as vacinas únicas
unique_cond <- unique(data$condition)  #trocar "study" pelo nome da coluna de interesse

# Inicialize um dataframe para armazenar os resultados
shared_genes_df <- data.frame()

# Calcular a sobreposição e porcentagem para cada par de vacinas
for (i in 1:(length(unique_cond) - 1)) {
  for (j in (i + 1):length(unique_cond)) {
    resultado_temp <- overlap_genes(unique_cond[i], unique_cond[j], data)
    shared_genes_df <- bind_rows(shared_genes_df, resultado_temp)
  }
}

# Função para realizar o teste exato de Fisher
fisher_exact_test <- function(shared, notshared1, notshared2) {
  cont_table <- matrix(c(shared, notshared1, notshared2, 0), nrow = 2)
  results_fisher <- fisher.test(cont_table)
  return(results_fisher$p.value)
}

# Aplicar a função a cada linha da tabela de resultados
shared_genes_df$pvalue <- mapply(fisher_exact_test, 
                                    shared_genes_df$Shared, 
                                    shared_genes_df$NotShared_cond1, 
                                    shared_genes_df$NotShared_cond2)


# Calcular qvalue (BH) e -log(FDR)
shared_genes_df <- shared_genes_df %>%
  mutate(qvalue = qvalue(pvalue)$qvalues)

# Calcular -log(q)
shared_genes_df$`log_q`= -log(shared_genes_df$qvalue, 10)


#Duplicar e trocar colunas (Vacina 1 e Vacina 2)
shared_genes_df2 = shared_genes_df
shared_genes_df2[, c("Cond1", "Cond2")] <- shared_genes_df2[, c("Cond2", "Cond1")]

#Unir datasets com os dados espelhados
shared_genes_df_mirror = rbind(shared_genes_df, shared_genes_df2)

#Converter NA, NaNe inf em 0
shared_genes_df_mirror[is.na(shared_genes_df_mirror)] <- 0
shared_genes_df_mirror<- replace(shared_genes_df_mirror, shared_genes_df_mirror == "Inf", 0)

#Arredondar porcentagens
shared_genes_df_mirror = shared_genes_df_mirror %>%
  mutate(Percentage_Shared_Cond1 = round(Percentage_Shared_Cond1, 2),
         Percentage_Shared_Cond2 = round(Percentage_Shared_Cond2, 2))

#Salvar resultados
write.csv(shared_genes_df_mirror, file = paste0("shared_", filename, "_fisher.csv")) #Alterar

#Filtrar
shared_genes_df_mirror_q10 = shared_genes_df_mirror
write.csv(shared_genes_df_mirror_q10, file = paste0("shared_", filename, "_fisher_p<10_qvalue10.csv"))

```


### Total genes

```{r}
# Required
ann = ann_vaccines_6_12_23
data = shared_genes_df_mirror_q10
filename = "degs_venn_updown_<0>_Totalgenes"

#Matriz
matrix_heatmap_df = data %>% 
  dcast(Cond1 ~ Cond2, 
        value.var = "Shared", 
        fun.aggregate = mean) %>% 
  column_to_rownames("Cond1")


#Anotar colunas
############## Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex =  ann %>% 
  select(condition:severity, regimen) %>% 
  distinct()

#Unir tabela de anotações
ann_vaccines_heatmap = matrix_heatmap_df %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "condition") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange() %>% 
  select(condition, disease_vac, regimen, type, vaccine, dose, infection, day, severity) %>% 
  mutate(day = as.factor(day),
         dose = as.factor(dose),
         infection = as.factor(infection),
         regimen = as.factor(regimen),
         severity = as.factor(severity))


#Ordenar colunas da matriz
matrix_heatmap_ordered = matrix_heatmap_df %>% 
  rownames_to_column(var= "condition") %>% 
  arrange(condition) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  column_to_rownames("condition") %>% 
  as.matrix()#Replace

ann_vaccines_heatmap = ann_vaccines_heatmap %>% 
  column_to_rownames("condition")

dim(matrix_heatmap_ordered)
dim(ann_vaccines_heatmap)
```


```{r}
########## Definir Cores dos grupos
#Vaccines
ann_colors = list(
  vaccine = c(
  "BBIBP" = "#dee2e6",
  "BNT" = "#56cfe1",
  "ChAd" = "#5e60ce",
  "ZF2001" = "#b5179e",
  "MO" = "#D7B0EE",
  "I" = "#ff4d6d",
  "H" = "grey95"),
  dose = c(
    "0" = "grey95",
    "1" = "#56cfe1",
    "2" = "#5978d4",
    "3" = "#b5179e"),
  day = c(
    "0" = "white",
    "1" = "#caf0f8",
    "2" = "#ade8f4",
    "3" = "#90e0ef",
    "4" = "#6CD5EA",
    "5" = "#48cae4",
    "6" = "#00b4d8",
    "7" = "#0096c7",
    "10" = "#0087BF",
    "14" = "#0077b6",
    "26" = "#015BA0",
    "28" = "#023e8a",
    "51" = "#03045e"),
  type = c('IN'= "#e5e5e5", #1st Gen  vaccines
           'SU' = "#00a896",
           'VV' = "#72efdd", #3rd Gen vaccines
           'RNA' = "#56cfe1",
           'SU' = "#b5179e",
           'I'= "#da1e37",
           "H" = "grey95"), #Infection
  regimen = c('HO' = "grey90",
              'HE' = "#8d99ae",
              "V-I-V" = "#36248f",
              "V-I" = "#D7B0EE",
              "I-V-I" = "#7400b8",
              "I-I" = "#5390d9",
              "I" = "#64dfdf",
              "H" = "grey95"),
  infection = c("0" = "grey95",
                "1" = "#56cfe1",
                "2" = "#5e60ce"),
  variant = c("None" = "grey95",
              "Beta" = "#72efdd",
              "Omicron" = "#5e60ce"),
  severity = c("H" = "grey95",
               "S" = "#b5179e",
               "MO" = "#5e60ce",
               "MI" = "#80ffdb",
               "NA" = "grey95"),
  sex = c("Male" = "#56cfe1",
          "Female" = "#b5179e",
          "M/F" = "#5e60ce"),
  disease_vac = c("H" = "#caf0f8",
                  "Healthy" = "#caf0f8",
                  "I" = "#da1e37",
                  "V" = "#5e60ce")
  )

col_fun <- colorRamp2(c(0, 100), c("white", "#16E0BD"))

#Criar heatmap annotation
col_ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")
row_ha = rowAnnotation(df = ann_vaccines_heatmap, col = ann_colors)


#TOTAL GENES (with values)
mat = matrix_heatmap_ordered 

fileimageheatmap_ungrouped= paste0(filename, sep ="_", "VALUES_Complexheatmap.png")

png(file = fileimageheatmap_ungrouped, width=40, height=40,units="cm",res=900)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = col_ha,
                        left_annotation = row_ha,
                        na_col = "black",
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        cell_fun = function(j, i, x, y, width, height, fill) {
                        grid.text(sprintf("%.1f", mat[i, j]), x, y, gp = gpar(fontsize = 6))}, #valor das células
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "Total genes", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 8, fontface = "bold"),
                        column_names_rot = 90,
                        column_names_gp = gpar(fontsize = 8),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = gpar(fontsize = 10),
                        row_split = 2, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_split = 2, #Split group clusters
                        column_gap = unit(8, "mm"),
                        show_column_dend = FALSE,
                        show_row_dend = FALSE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = unit(20, "cm"), 
                        height = unit(20, "cm")
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()


#TOTAL GENES (no values)
mat = matrix_heatmap_ordered 

fileimageheatmap_ungrouped= paste0(filename, sep ="_", "NOVALUES_Complexheatmap.png")

png(file = fileimageheatmap_ungrouped, width=40, height=40,units="cm",res=900)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = col_ha,
                        left_annotation = row_ha,
                        na_col = "black",
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "Total genes", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 8, fontface = "bold"),
                        column_names_rot = 90,
                        column_names_gp = gpar(fontsize = 8),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = gpar(fontsize = 10),
                        row_split = 2, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_split = 2, #Split group clusters
                        column_gap = unit(8, "mm"),
                        show_column_dend = FALSE,
                        show_row_dend = FALSE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = unit(20, "cm"), 
                        height = unit(20, "cm")
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```



### Shared identity (%)


#### Cond1
```{r}
# Required
ann = ann_vaccines_6_12_23
data = shared_genes_df_mirror_q10
filename = "degs_venn_updown_<0>_SharedIdentity_Cond1"
#Matriz
matrix_heatmap_df = data %>% 
  dcast(Cond1 ~ Cond2, 
        value.var = "Percentage_Shared_Cond1", 
        fun.aggregate = mean) %>% 
  column_to_rownames("Cond1")


#Anotar colunas
############## Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex =  ann %>% 
  select(condition:severity, regimen) %>% 
  distinct()

#Unir tabela de anotações
ann_vaccines_heatmap = matrix_heatmap_df %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "condition") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange() %>% 
  select(condition, disease_vac, regimen, type, vaccine, dose, infection, day, severity) %>% 
  mutate(day = as.factor(day),
         dose = as.factor(dose),
         infection = as.factor(infection),
         regimen = as.factor(regimen),
         severity = as.factor(severity))


#Ordenar colunas da matriz
matrix_heatmap_ordered = matrix_heatmap_df %>% 
  rownames_to_column(var= "condition") %>% 
  arrange(condition) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  column_to_rownames("condition") %>% 
  as.matrix()#Replace

ann_vaccines_heatmap = ann_vaccines_heatmap %>% 
  column_to_rownames("condition")

dim(matrix_heatmap_ordered)
dim(ann_vaccines_heatmap)
```


```{r}
########## Definir Cores dos grupos
col_fun <- colorRamp2(c(0, 100), c("white", "#16E0BD"))

#Criar heatmap annotation
col_ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")
row_ha = rowAnnotation(df = ann_vaccines_heatmap, col = ann_colors)


#TOTAL GENES (with values)
mat = matrix_heatmap_ordered 

fileimageheatmap_ungrouped= paste0(filename, sep ="_", "VALUES_Complexheatmap.png")

png(file = fileimageheatmap_ungrouped, width=40, height=40,units="cm",res=900)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = col_ha,
                        left_annotation = row_ha,
                        na_col = "black",
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        cell_fun = function(j, i, x, y, width, height, fill) {
                        grid.text(sprintf("%.1f", mat[i, j]), x, y, gp = gpar(fontsize = 6))}, #valor das células
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "Total genes", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 8, fontface = "bold"),
                        column_names_rot = 90,
                        column_names_gp = gpar(fontsize = 8),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = gpar(fontsize = 10),
                        row_split = 2, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_split = 2, #Split group clusters
                        column_gap = unit(8, "mm"),
                        show_column_dend = FALSE,
                        show_row_dend = FALSE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = unit(20, "cm"), 
                        height = unit(20, "cm")
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()


#TOTAL GENES (no values)
mat = matrix_heatmap_ordered 

fileimageheatmap_ungrouped= paste0(filename, sep ="_", "NOVALUES_Complexheatmap.png")

png(file = fileimageheatmap_ungrouped, width=40, height=40,units="cm",res=900)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = col_ha,
                        left_annotation = row_ha,
                        na_col = "black",
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "Total genes", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 8, fontface = "bold"),
                        column_names_rot = 90,
                        column_names_gp = gpar(fontsize = 8),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = gpar(fontsize = 10),
                        row_split = 2, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_split = 2, #Split group clusters
                        column_gap = unit(8, "mm"),
                        show_column_dend = FALSE,
                        show_row_dend = FALSE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = unit(20, "cm"), 
                        height = unit(20, "cm")
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```


#### Cond2
```{r}
# Required
ann = ann_vaccines_6_12_23
data = shared_genes_df_mirror_q10
filename = "degs_venn_updown_<0>_SharedIdentity_Cond2"
#Matriz
matrix_heatmap_df = data %>% 
  filter(qvalue < 0.10) %>% 
  dcast(Cond2 ~ Cond1, 
        value.var = "Percentage_Shared_Cond2", 
        fun.aggregate = mean) %>% 
  column_to_rownames("Cond2")


#Anotar colunas
############## Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex =  ann %>% 
  select(condition:severity, regimen) %>% 
  distinct()

#Unir tabela de anotações
ann_vaccines_heatmap = matrix_heatmap_df %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "condition") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange() %>% 
  select(condition, disease_vac, regimen, type, vaccine, dose, infection, day, severity) %>% 
  mutate(day = as.factor(day),
         dose = as.factor(dose),
         infection = as.factor(infection),
         regimen = as.factor(regimen),
         severity = as.factor(severity))


#Ordenar colunas da matriz
matrix_heatmap_ordered = matrix_heatmap_df %>% 
  rownames_to_column(var= "condition") %>% 
  arrange(condition) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  column_to_rownames("condition") %>% 
  as.matrix()#Replace

ann_vaccines_heatmap = ann_vaccines_heatmap %>% 
  column_to_rownames("condition")

dim(matrix_heatmap_ordered)
dim(ann_vaccines_heatmap)
```


```{r}
########## Definir Cores dos grupos
col_fun <- colorRamp2(c(0, 100), c("white", "#16E0BD"))

#Criar heatmap annotation
col_ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")
row_ha = rowAnnotation(df = ann_vaccines_heatmap, col = ann_colors)


#TOTAL GENES (with values)
mat = matrix_heatmap_ordered 

fileimageheatmap_ungrouped= paste0(filename, sep ="_", "VALUES_Complexheatmap.png")

png(file = fileimageheatmap_ungrouped, width=40, height=40,units="cm",res=900)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = col_ha,
                        left_annotation = row_ha,
                        na_col = "black",
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        cell_fun = function(j, i, x, y, width, height, fill) {
                        grid.text(sprintf("%.1f", mat[i, j]), x, y, gp = gpar(fontsize = 6))}, #valor das células
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "Identity (%)", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 8, fontface = "bold"),
                        column_names_rot = 90,
                        column_names_gp = gpar(fontsize = 8),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = gpar(fontsize = 10),
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_split = NULL, #Split group clusters
                        column_gap = unit(8, "mm"),
                        show_column_dend = FALSE,
                        show_row_dend = FALSE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = unit(20, "cm"), 
                        height = unit(20, "cm")
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()


#TOTAL GENES (no values)
mat = matrix_heatmap_ordered 

fileimageheatmap_ungrouped= paste0(filename, sep ="_", "NOVALUES_Complexheatmap.png")

png(file = fileimageheatmap_ungrouped, width=40, height=40,units="cm",res=900)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = col_ha,
                        left_annotation = row_ha,
                        na_col = "black",
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "Identity (%)", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 8, fontface = "bold"),
                        column_names_rot = 90,
                        column_names_gp = gpar(fontsize = 8),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = gpar(fontsize = 10),
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_split = NULL, #Split group clusters
                        column_gap = unit(8, "mm"),
                        show_column_dend = FALSE,
                        show_row_dend = FALSE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = unit(20, "cm"), 
                        height = unit(20, "cm")
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()
```
### LogQ
```{r}
# Required
ann = ann_vaccines_6_12_23
data = shared_genes_df_mirror_q10
filename = "degs_venn_updown_<0>_Logq_Cond1"
#Matriz
matrix_heatmap_df = data %>% 
  dcast(Cond1 ~ Cond2, 
        value.var = "log_q", 
        fun.aggregate = mean) %>% 
  column_to_rownames("Cond1")


#Anotar colunas
############## Verificar se todas as vacinas da anotação estão na matriz e unir as tabelas
ann_vaccines_complex =  ann %>% 
  select(condition:severity, regimen) %>% 
  distinct()

#Unir tabela de anotações
ann_vaccines_heatmap = matrix_heatmap_df %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "condition") %>% 
  rename(condition = '.') %>% 
  merge(ann_vaccines_complex, by = "condition", all.x = TRUE) %>% 
  arrange() %>% 
  select(condition, disease_vac, regimen, type, vaccine, dose, infection, day, severity) %>% 
  mutate(day = as.factor(day),
         dose = as.factor(dose),
         infection = as.factor(infection),
         regimen = as.factor(regimen),
         severity = as.factor(severity))


#Ordenar colunas da matriz
matrix_heatmap_ordered = matrix_heatmap_df %>% 
  rownames_to_column(var= "condition") %>% 
  arrange(condition) %>% #Trocar a variável pela qual deseja ordenar (Vaccine, Day, Condition, etc.)
  column_to_rownames("condition") %>% 
  as.matrix()#Replace

ann_vaccines_heatmap = ann_vaccines_heatmap %>% 
  column_to_rownames("condition")

dim(matrix_heatmap_ordered)
dim(ann_vaccines_heatmap)
```


```{r}
########## Definir Cores dos grupos
col_fun <- colorRamp2(c(0, 1, 3), c("white","#C1FFE5", "#16E0BD"))

#Criar heatmap annotation
col_ha = HeatmapAnnotation(df = ann_vaccines_heatmap, 
                       col = ann_colors, 
                       annotation_name_side = "left")
row_ha = rowAnnotation(df = ann_vaccines_heatmap, col = ann_colors)


#TOTAL GENES (with values)
mat = matrix_heatmap_ordered 

fileimageheatmap_ungrouped= paste0(filename, sep ="_", "VALUES_Complexheatmap.png")

png(file = fileimageheatmap_ungrouped, width=40, height=40,units="cm",res=900)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = col_ha,
                        left_annotation = row_ha,
                        na_col = "black",
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        cell_fun = function(j, i, x, y, width, height, fill) {
                        grid.text(sprintf("%.1f", mat[i, j]), x, y, gp = gpar(fontsize = 6))}, #valor das células
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "-LogQ", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 8, fontface = "bold"),
                        column_names_rot = 90,
                        column_names_gp = gpar(fontsize = 8),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = gpar(fontsize = 10),
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_split = NULL, #Split group clusters
                        column_gap = unit(8, "mm"),
                        show_column_dend = FALSE,
                        show_row_dend = FALSE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = unit(20, "cm"), 
                        height = unit(20, "cm")
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")

dev.off()


#TOTAL GENES (no values)
mat = matrix_heatmap_ordered 

fileimageheatmap_ungrouped= paste0(filename, sep ="_", "NOVALUES_Complexheatmap.png")

png(file = fileimageheatmap_ungrouped, width=40, height=40,units="cm",res=900)

heatmap_plot_all = Heatmap(mat,
                        top_annotation = col_ha,
                        left_annotation = row_ha,
                        na_col = "black",
                        show_row_names = TRUE,
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        name = "Identity (%)", #Legend
                        col = col_fun, #color
                        column_title = fileimageheatmap_ungrouped,
                        column_title_gp = gpar(fontsize = 8, fontface = "bold"),
                        column_names_rot = 90,
                        column_names_gp = gpar(fontsize = 8),
                        rect_gp = gpar(col = "gray90", lwd = 1),
                        cluster_rows = TRUE,
                        row_names_gp = gpar(fontsize = 10),
                        row_split = NULL, #Split row clusters
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_split = NULL, #Split group clusters
                        column_gap = unit(8, "mm"),
                        show_column_dend = FALSE,
                        show_row_dend = FALSE,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        column_dend_reorder = TRUE,
                        width = unit(20, "cm"), 
                        height = unit(20, "cm")
)

draw(heatmap_plot_all, annotation_legend_side = "right", heatmap_legend_side = "right")
 
dev.off()
```


*Chord diagram*

```{r}
#Chord diagram
install.packages("circlize")
library(circlize)

# sharedDEGs_UPDOWN_fisher_pvalue10
# sharedDEGs_UP_fisher_p_10_pvalue10
# sharedDEGs_DOWN_fisher_p_10_pvalue10

#Input
shared_genes_df_mirror_p10 = sharedDEGs_UP_fisher_p_10_pvalue10


#All doses
circos_alldoses = shared_genes_df_mirror_p10 %>% 
  select(Cond1, Cond2, Shared) %>% 
  mutate(Shared = as.numeric(Shared)) %>%  
  merge(ann_vaccines, by.x = "Cond1", by.y="Condition", all.x=T, all.y=F) %>%
  merge(ann_vaccines, by.x = "Cond2", by.y="Condition", all.x=T, all.y=F) %>% 
  select(Cond1, Cond2, Shared,Dose.x, Dose.y) %>% 
  rename(cond1_dose = Dose.x,
         cond2_dose = Dose.y) %>% 
  filter(Shared != 0)

circos_alldoses_mirror = circos_alldoses %>% 
  mutate(Cond1A = Cond2,
         Cond2A = Cond1) %>% 
  select(Cond1A, Cond2A, Shared, cond1_dose, cond2_dose) %>% 
  rename(Cond1 = Cond1A,
         Cond2 = Cond2A) %>% 
  rbind(circos_alldoses)

write.csv(circos_alldoses_mirror, file = "Circos_DOWN_Alldoses_pvalue10_dose1.csv")


#Dose 1
circos_dose1 = shared_genes_df_mirror_p10 %>% 
  select(Cond1, Cond2, Shared) %>% 
  mutate(Shared = as.numeric(Shared)) %>%  
  merge(ann_vaccines, by.x = "Cond1", by.y="Condition", all.x=T, all.y=F) %>%
  merge(ann_vaccines, by.x = "Cond2", by.y="Condition", all.x=T, all.y=F) %>% 
  select(Cond1, Cond2, Shared,Dose.x, Dose.y) %>% 
  rename(cond1_dose = Dose.x,
         cond2_dose = Dose.y) %>% 
  filter(cond1_dose == 1,
         cond2_dose == 1)

circos_dose1_mirror = circos_dose1 %>% 
  mutate(Cond1A = Cond2,
         Cond2A = Cond1) %>% 
  select(Cond1A, Cond2A, Shared, cond1_dose, cond2_dose) %>% 
  rename(Cond1 = Cond1A,
         Cond2 = Cond2A) %>% 
  rbind(circos_dose1)

write.csv(circos_dose1_mirror, file = "Circos_DOWN_pvalue10_dose1.csv")

#Dose 2
circos_dose2 = shared_genes_df_mirror_p10 %>% 
  select(Cond1, Cond2, Shared) %>% 
  mutate(Shared = as.numeric(Shared)) %>%  
  merge(ann_vaccines, by.x = "Cond1", by.y="Condition", all.x=T, all.y=F) %>%
  merge(ann_vaccines, by.x = "Cond2", by.y="Condition", all.x=T, all.y=F) %>% 
  select(Cond1, Cond2, Shared,Dose.x, Dose.y) %>% 
  rename(cond1_dose = Dose.x,
         cond2_dose = Dose.y) %>% 
  filter(cond1_dose == 2,
         cond2_dose == 2)
circos_dose2_mirror = circos_dose2 %>% 
  mutate(Cond1A = Cond2,
         Cond2A = Cond1) %>% 
  select(Cond1A, Cond2A, Shared, cond1_dose, cond2_dose) %>% 
  rename(Cond1 = Cond1A,
         Cond2 = Cond2A) %>% 
  rbind(circos_dose2) 

write.csv(circos_dose2_mirror, file = "Circos_DOWN_pvalue10_dose2.csv")

#Dose 3
circos_dose3 = shared_genes_df_mirror_p10 %>% 
  select(Cond1, Cond2, Shared) %>% 
  mutate(Shared = as.numeric(Shared)) %>%  
  merge(ann_vaccines, by.x = "Cond1", by.y="Condition", all.x=T, all.y=F) %>%
  merge(ann_vaccines, by.x = "Cond2", by.y="Condition", all.x=T, all.y=F) %>% 
  select(Cond1, Cond2, Shared,Dose.x, Dose.y) %>% 
  rename(cond1_dose = Dose.x,
         cond2_dose = Dose.y) %>% 
  filter(cond1_dose == 3,
         cond2_dose == 3)

 circos_dose3_mirror = circos_dose3 %>% 
  mutate(Cond1A = Cond2,
         Cond2A = Cond1) %>% 
  select(Cond1A, Cond2A, Shared, cond1_dose, cond2_dose) %>% 
  rename(Cond1 = Cond1A,
         Cond2 = Cond2A) %>% 
  rbind(circos_dose3) 

write.csv(circos_dose3_mirror, file = "Circos_DOWN_pvalue10_dose3.csv")

```



### Representação


### Visualização

Dotplot

```{r}
#### Identidade e -log(q)
install.packages("ggdendro")
library(ggdendro)
library(cowplot)
library(ggtree)
library(patchwork) 

#Filtrar valores com -log(p) >= 1
sharedDEGs_UP_DOWN_fisher_filtered <- sharedDEGs_UP_DOWN_fisher %>%
  mutate(`Significance -log(p)` = cut(`-log(p)`, 
                        breaks = c(-Inf, 1, 1.3, 2, Inf), 
                        labels = c("<1", "1-1.3", "1.3-2", ">2"))) %>%
  filter(Percentage_Shared_Cond1 >= 1, Percentage_Shared_Cond2 >= 1, `Significance -log(p)` != "<1")

#Salvar
write.csv(sharedDEGs_UP_DOWN_fisher_filtered, file = "sharedDEGs_UP_DOWN_fisher_filtered.csv")

#Visualizar
plot = ggplot(sharedDEGs_UP_DOWN_fisher, aes(x = Cond1, y = Cond2, color = Percentage_Shared_Cond1, size = Percentage_Shared_Cond1)) + 
  geom_point() +
  scale_size_continuous(range = c(2, 8)) +
  geom_text(aes(label = sprintf("%.f", Percentage_Shared_Cond1)), vjust = 0.5, hjust = 0.5, size = 2, color = "black")  +
  cowplot::theme_cowplot() + 
  theme(axis.line = element_blank(),
        axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1, size = 10),
        axis.text.y = element_text(size = 10)) +
  scale_color_gradient(low = "white", high = "#ef233c", limits = c(0, 60)) +
  ggtitle("%Identity between vaccines, Up and Down-Regulated, pvalue < 0.05" ) +
  ylab('Cond2') +
  theme(axis.ticks = element_blank()) +
  guides(color = FALSE, size = FALSE)

#Salvar
ggsave(plot, file = "sharedDEGs_UP_DOWN_fisher.png")

#Display
plot
```

