---
title: "PCA"
output: html_notebook
---

# PCA

Tutoriais <https://www.datacamp.com/tutorial/pca-analysis-r> <https://rpkgs.datanovia.com/factoextra/index.html> <http://www.sthda.com/english/articles/31-principal-component-methods-in-r-practical-guide/114-mca-multiple-correspondence-analysis-in-r-essentials/> <https://statisticsglobe.com/pca-before-k-means-clustering-r>\>

Bibliotecas

```{r}
library(tidyverse)
library(ggrepel) #load before factoextra
library(factoextra)
library(caret)
library(stats)
library(ggfortify)
library(ggplot2)
library(corrplot)
library(cowplot)
library(ggcorrplot)
library(ComplexHeatmap)
library(circlize)
library(ggExtra)
```

### Process data

```{r}

ImmuneGO_Annotated_Genes_all = ImmuneGO_Annotated_Genes

ImmuneGO_Annotated_Genes = ImmuneGO_Annotated_Genes_all %>% 
  filter(!process %in% c("TCR REPERTOIRE", "BCR REPERTOIRE"))

ImmuneGO_Genes = all_degs_p_05_vac_infected_19_12_23 %>% 
  inner_join(ImmuneGO_Annotated_Genes %>% 
               select(genes, process) %>% 
               distinct() %>% 
               select(genes),
               by = "genes") %>% 
  distinct()

nonImmuneGO_Genes = all_degs_p_05_vac_infected_19_12_23 %>% 
  select(-"...1", -"...12") %>% 
  anti_join(ImmuneGO_Annotated_Genes_all %>% 
               select(genes) %>% 
               distinct(),
               by = "genes") %>% 
  distinct()
```

```{r}
#INPUT
data_genes = ImmuneGO_Genes 
filename = "ImmuneGO_Genes"

data_annotation = ann_vaccines_19_1_24 %>% 
  mutate(day = as.factor(day),
         dose = as.factor(dose)) 

#Converter de long para wide

n_distinct(data_genes$genes)

#Matriz com anotações
ann_vaccines_pca_matrix = data_genes %>% 
  mutate(log2fold_change = as.numeric(log2fold_change)) %>% 
  select(condition, genes, log2fold_change) %>% 
  pivot_wider(., names_from = "condition", 
                     values_from = "log2fold_change") %>% 
  replace(is.na(.), 0) %>%
  column_to_rownames("genes") %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column("condition") %>% 
  inner_join(data_annotation %>% 
               select(condition, disease_vac, vaccine, day, week, dose, infection, type), by = "condition") 

#Matriz para PCA
ann_vaccines_pca_matrix_ready = ann_vaccines_pca_matrix %>% 
  column_to_rownames("condition") %>% 
  select(!disease_vac:type) %>% 
  as.matrix()

# Verifique colunas com variancia baixa
nearZeroVarCols <- nearZeroVar(ann_vaccines_pca_matrix_ready, saveMetrics = TRUE)
ann_vaccines_pca_matrix_ready <- ann_vaccines_pca_matrix_ready[, !nearZeroVarCols$nzv]
pca_res <- prcomp(ann_vaccines_pca_matrix_ready, scale. = T)

########Correlation plot
corr_matrix = cor(ann_vaccines_pca_matrix_ready %>% scale()) 
var <- get_pca_var(pca_res)

### Contribuição dos genes principais para cada PC
# Comp1
var_ordenado_top20_dim1 <- var$cos2 %>%
  as.data.frame() %>%
  arrange(desc(.[, 1])) %>%   # Ordenar pela primeira coluna em ordem decrescente
  slice_head(n = 20) %>%      # Selecionar as primeiras 50 linhas
  select(1:1) %>%                  # Selecionar as primeiras 5 colunas
  as.matrix()

corplot_dim1 = {corrplot(var_ordenado_top20_dim1, 
         is.corr = T, tl.col = 'black', 
         tl.cex = 0.5, 
         addCoef.col = 'white',
         number.cex = 0.4,
         cl.pos = 'n', 
         col = COL1('Blues'),
         col.lim = c(min(var_ordenado_top20_dim1), max(var_ordenado_top20_dim1)))  ;
        # Call the recordPlot() function to record the plot
        recordPlot()
       }

#Comp2
var_ordenado_top20_dim2 <- var$contrib %>%
  as.data.frame() %>%
  arrange(desc(.[, 2])) %>%   # Ordenar pela primeira coluna em ordem decrescente
  slice_head(n = 20) %>%      # Selecionar as primeiras 50 linhas
  select(2) %>%                  # Selecionar as primeiras 5 colunas
  as.matrix() 

corplot_dim2 = {corrplot(var_ordenado_top20_dim2, 
         is.corr = T, 
         tl.col = 'black', 
         tl.cex = 0.5, 
         addCoef.col = 'white',
         number.cex = 0.4,
         cl.pos = 'n', 
         col = COL2('PiYG'),
         col.lim = c(min(var_ordenado_top20_dim2), max(var_ordenado_top20_dim2))) ;
        # Call the recordPlot() function to record the plot
        recordPlot()
       }

dim1_dim2_corrplot = plot_grid(corplot_dim1, corplot_dim2, 
          rel_widths = c(1, 0.5, 1), 
          align = "hv",
          labels = c("Dim1", "Dim2"), 
          nrow = 1)

ggsave(dim1_dim2_corrplot, file = paste0(filename, "dim1_dim2_corrplot.png"))

# #Plot
# corrplot = ggcorrplot(corr_matrix, hc.order = TRUE) + 
#   theme(axis.text.x = element_text(angle = 90, size = 5), 
#         axis.text.y = element_text(size = 5))
# #Salvar
# ggsave(corrplot, file = paste0(filename, "_corrplot.png"), 
#        width = 20, #Grande 20, pequeno 10
#        height= 20) #Grande 20, pequeno 10
# print(corrplot)

#########Scree plot
data.pca = prcomp(corr_matrix) #PCA
summary(corr_matrix) #Retornar PCs

#########Scree plot
scree_plot = fviz_eig(data.pca, 
                      addlabels = TRUE,
                      ylim = c(0, 70)) +
  geom_col(color = "#00AFBB", fill = "#00AFBB") +
  theme_classic()



#Salvar
ggsave(scree_plot, file = paste0(filename, "_screeplot.png"), width = 10, height = 3) 
print(scree_plot)

```

```{r}
# Graph of the variables
options(ggrepel.max.overlaps = Inf)

circle_contrib= fviz_pca_var(pca_res, col.var = "cos2",
                              gradient.cols = c("black", "#DC0000FF"),
                              select.var= list(cos2 = 20), 
                              repel = T, 
                              labelsize = 3,
                              col.circle = NA) +
  xlab("") + 
  ylab("") +
  theme_minimal() +
theme(panel.grid = element_blank(),  # Remover linhas de grade
        axis.text = element_blank(),   # Remover rótulos de texto dos eixos
        axis.ticks = element_blank()) 

ggsave(circle_contrib, file = paste0(filename, "_circle_contrib.png"), width = 5, height = 5)
```

CORES

```{r}
condition_col = c(
                                "BBIBP (V3, D07)" = "#00A087FF",
                                "BBIBP (V3, D14)" = "#00A087FF",
                                "BBIBP (V3, D28)" = "#00A087FF",
                                "BNT (V1, D6)" = "#4DBBD5FF",
                                "BNT (V2, D1)" = "#4DBBD5FF",
                                "BNT (V3, D1)" = "#4DBBD5FF",
                                "BNT-I (D1)"= "#DC0000FF",
                                "BNT-I (D10, mild)"= "#DC0000FF",
                                "BNT-I (D10, severe)"= "#DC0000FF",
                                "BNT-I (D2)"= "#DC0000FF",
                                "BNT-I (D26, mild)"= "#DC0000FF",
                                "BNT-I (D26, severe)"= "#DC0000FF",
                                "BNT-I (D3)"= "#DC0000FF",
                                "BNT-I (D4)"= "#DC0000FF",
                                "BNT-I (D51, severe)"= "#DC0000FF",
                                "BNT-I-BNT (D51, mild)"= "#F5BB00",
                                "BNT-I-BNT (D51, severe)"= "#F5BB00",
                                "BNT-MO (V1, D6)"= "#72ddf7",
                                "BNT-MO (V3, D1)"= "#72ddf7",
                                "ChAd (V1, D3)"= "#3C5488FF",
                                "ChAd (V1, D6)"= "#3C5488FF",
                                "ChAd (V1, D7)"= "#3C5488FF",
                                "ChAd (V2, D1)"= "#3C5488FF",
                                "ChAd (V2, D3)"= "#3C5488FF",
                                "ChAd (V2, D7)"= "#3C5488FF",
                                "ChAd-BNT (V2, D0)"= "#4DBBD5FF",
                                "ChAd-BNT (V2, D3)"= "#4DBBD5FF",
                                "ChAd-BNT (V2, D7)"= "#4DBBD5FF",
                                "ChAd-BNT (V3, D1)"= "#4DBBD5FF",
                                "I (D1)"= "#DC0000FF",
                                "I (D10, moderate)"= "#DC0000FF",
                                "I (D10, severe)"= "#DC0000FF",
                                "I (D26, moderate)"= "#DC0000FF",
                                "I (D26, severe)"= "#DC0000FF",
                                "I (D51, moderate)"= "#DC0000FF",
                                "I (D51, severe)"= "#DC0000FF",
                                "I-BNT-I (D2)"= "#DC0000FF",
                                "I-BNT-I (D5)"= "#DC0000FF",
                                "I-I (D0)"= "#DC0000FF",
                                "I-I (D1)"= "#DC0000FF",
                                "I-I (D2)"= "#DC0000FF",
                                "I-I (D3)"= "#DC0000FF",
                                "I-I (D5)"= "#DC0000FF",
                                "ZF2001 (V3, D07)"= "#8491B4FF",
                                "ZF2001 (V3, D14)"= "#8491B4FF",
                                "ZF2001 (V3, D28)" = "#8491B4FF"
                              )
```

```{r}
############### KNN classification
#Use the screeplot generated % of explained variance for each dimension
scree_plot #Dim 1 = 66.8%, Dim 2 = 15.7%


# PC1-PC2 ------
#Determinar o número de clusters para KNN
pca_scores <- data.frame(pca_res$x[, 1:2])
fviz_nbclust(pca_scores,  
                     FUNcluster = kmeans,
                     method = "wss")

pcas = "PC1-PC2"

set.seed(666)                             # Set seed for randomization
cluster_model <- kmeans(pca_res$x[, 1:2], centers = 3)  # Ajuste o número de clusters conforme necessário
ann_vaccines_pca_matrix$cluster <- as.factor(cluster_model$cluster)


# Condition -------
pca_plot_knn_condition_lab = autoplot(pca_res, 
        data = ann_vaccines_pca_matrix, 
        colour = 'condition')  +
  stat_ellipse(aes(color = cluster, fill = cluster), 
               geom = "polygon", 
               alpha = 0.1, 
               linetype = 1,
               size = 0.3,
               type = "t") +
  labs(title=paste0(filename, "_bycondition")) + 
  xlab("PC1 (68.2%)") +
  ylab("PC2 (15.9%)") +
  scale_color_manual(values = c("1" = "#8491B4FF", 
                                "2" = "#DC0000FF", 
                                "3" = "#4DBBD5FF", 
                                "4" = "#00A087FF", 
                                "5" = "#E64B35FF", 
                                condition_col)) +  # Ellipse line colors
  scale_fill_manual(values = c("1" = "#8491B4FF", 
                                "2" = "#DC0000FF", 
                                "3" = "#4DBBD5FF", 
                                "4" = "#00A087FF", 
                                "5" = "#E64B35FF")) +  # Cluster fill colors
  theme_minimal() +
  guides(col="none") +
  theme(panel.grid = element_blank()) +
  geom_text_repel(aes(label = condition,
                      segment.colour="gray70"),
                  box.padding = 0.3,
                  size = 2) 

ggsave(pca_plot_knn_condition_lab, file = paste0(filename, pcas, "_KNN_Clustered_labels.png"), width = 7, height = 5)


# Clusters, with density plot ------

pca_plot_knn_cluster = autoplot(pca_res, 
        data = ann_vaccines_pca_matrix, 
        colour = 'cluster')  +
  stat_ellipse(aes(color = cluster, fill = cluster), 
               geom = "polygon", 
               alpha = 0.1, 
               linetype = 1,
               size = 0.3,
               type = "t") +
  labs(title=paste0(filename, "_bycondition")) + 
  xlab("PC1 (68.2%)") +
  ylab("PC2 (15.9%)") +
  scale_color_manual(values = c("1" = "#8491B4FF", 
                                "2" = "#DC0000FF", 
                                "3" = "#4DBBD5FF", 
                                "4" = "#00A087FF", 
                                "5" = "#E64B35FF", 
                                condition_col)) +  # Ellipse line colors
  scale_fill_manual(values = c("1" = "#8491B4FF", 
                                "2" = "#DC0000FF", 
                                "3" = "#4DBBD5FF", 
                                "4" = "#00A087FF", 
                                "5" = "#E64B35FF")) +  # Cluster fill colors
  theme_minimal() +
  guides(col="none") +
  theme(panel.grid = element_blank()) +
  geom_text_repel(aes(label = condition,
                      segment.colour="gray70"),
                  box.padding = 0.3,
                  size = 2) 

pca_plot_knn_cluster_marginal = ggMarginal(pca_plot_knn_cluster, type = "histogram", groupFill = T, groupColour = T, xparams = list(binwidth = 0.1, size = 0.1), yparams = list(binwidth = 0.1, size = 0.1))

ggsave(pca_plot_knn_cluster_marginal, file = paste0(filename, pcas, "_KNN_Clustered_labels_Density.png"), width = 8, height = 5)

# PC2-PC3 ------
#Determinar o número de clusters para KNN
pca_scores <- data.frame(pca_res$x[, 1:3])
pcas = "PC1-PC3"
fviz_nbclust(pca_scores,  
                     FUNcluster = kmeans,
                     method = "wss")

set.seed(666)                             # Set seed for randomization
cluster_model <- kmeans(pca_res$x[, 1:3], centers = 3)  # Ajuste o número de clusters conforme necessário
ann_vaccines_pca_matrix$cluster <- as.factor(cluster_model$cluster)


# Condition -------
pca_plot_knn_condition_lab = autoplot(pca_res, 
        data = ann_vaccines_pca_matrix, 
        colour = 'condition')  +
  stat_ellipse(aes(color = cluster, fill = cluster), 
               geom = "polygon", 
               alpha = 0.1, 
               linetype = 1,
               size = 0.3,
               type = "t") +
  labs(title=paste0(filename, "_bycondition")) + 
  # xlab("PC1 (66.8%)") + 
  # ylab("PC2 (15.7%)") +
  scale_color_manual(values = c("1" = "#5e60ce", 
                                "2" = "#f72585", 
                                "3" = "#56cfe1", 
                                "4" = "#b5e48c", 
                                "5" = "#ff8c42", 
                                condition_col)) +  # Ellipse line colors
  scale_fill_manual(values = c("1" = "#5e60ce", 
                                "2" = "#f72585", 
                                "3" = "#56cfe1", 
                                "4" = "#b5e48c", 
                                "5" = "#ff8c42")) +  # Cluster fill colors
  theme_minimal() +
  guides(col="none") +
  theme(panel.grid = element_blank()) +
  geom_text_repel(aes(label = condition,
                      segment.colour="gray70"),
                  box.padding = 0.3,
                  size = 2) 

ggsave(pca_plot_knn_condition_lab, file = paste0(filename, pcas, "_KNN_Clustered_labels_PC2_PC3.png"), width = 7, height = 5)


# Clusters, with density plot ------

pca_plot_knn_cluster = autoplot(pca_res, 
        data = ann_vaccines_pca_matrix, 
        colour = 'cluster')  +
  stat_ellipse(aes(color = cluster, fill = cluster), 
               geom = "polygon", 
               alpha = 0.1, 
               linetype = 1,
               size = 0.3,
               type = "t") +
  labs(title=paste0(filename, "_bycondition")) + 
  # xlab("PC1 (66.8%)") + 
  # ylab("PC2 (15.7%)") +
  scale_color_manual(values = c("1" = "#5e60ce", 
                                "2" = "#f72585", 
                                "3" = "#56cfe1", 
                                "4" = "#b5e48c", 
                                "5" = "#ff8c42", 
                                condition_col)) +  # Ellipse line colors
  scale_fill_manual(values = c("1" = "#5e60ce", 
                                "2" = "#f72585", 
                                "3" = "#56cfe1", 
                                "4" = "#b5e48c", 
                                "5" = "#ff8c42")) +  # Cluster fill colors
  theme_minimal() +
  guides(col="none") +
  theme(panel.grid = element_blank()) +
  geom_text_repel(aes(label = condition,
                      segment.colour="gray70"),
                  box.padding = 0.3,
                  size = 2) 

pca_plot_knn_cluster_marginal = ggMarginal(pca_plot_knn_cluster, type = "histogram", groupFill = T, groupColour = T, xparams = list(binwidth = 0.1, size = 0.1), yparams = list(binwidth = 0.1, size = 0.1))

ggsave(pca_plot_knn_cluster_marginal, file = paste0(filename, pcas, "_KNN_Clustered_labels_Density.png"), width = 8, height = 5)



```

#Heatmap

## Genes annotated

```{r}
####### INPUT
PC1 = var$cos2 %>%
  as.data.frame() %>%
  arrange(desc(Dim.1)) %>%
  filter(Dim.1 >= 0.5) %>% 
  #slice_head(n=20) %>% 
  rownames_to_column("genes") %>% 
  distinct() %>% 
  select(genes, cos2 = Dim.1) %>% 
    mutate(dimension = "PC1")

PC2 = var$cos2 %>%
  as.data.frame() %>%
  filter(Dim.2 >= 0.5) %>% 
  #slice_head(n=20) %>% 
  rownames_to_column("genes") %>% 
  distinct() %>% 
  select(genes, cos2 = Dim.2) %>% 
  mutate(dimension = "PC2")

PC1_2 = bind_rows(PC1, PC2) 
PC1_2 %>% 
  select(genes) %>% distinct()

#244 genes


PC1_2 %>% write.csv(file = "PC_1_2_cos2.csv", row.names = F)


# Gene annotation
ImmuneGO_Annotated_Genes
OtherGOs_Annotated_Genes

# INPUT

ann_genes = ImmuneGO_Annotated_Genes
pc = "PC2"
data_2 =  PC2 %>% 
  slice_max(order_by = cos2, n = 20) %>% 
  inner_join(data_genes, by = "genes")
filename_2 = paste0(filename, "_PCA_",pc, "_COVID")

######## Matrix
matrix = data_2 %>% 
  select(genes, condition, log2fold_change) %>%
  pivot_wider(names_from = "condition", 
              values_from = "log2fold_change", 
              values_fn = mean) %>% 
  replace(is.na(.), 0) %>%
  arrange(genes) %>% 
  column_to_rownames(var="genes") %>% 
  as.matrix()

######## Annotations
ann_vaccines_covid = data_2 %>% 
  select(condition) %>% 
  distinct() %>% 
  inner_join(ann_vaccines, by = "condition") %>% 
  select(condition, disease_vac, type, day) %>% 
  distinct()

#Columns
ann_cols_heatmap = matrix %>% 
  colnames() %>% 
  data.frame() %>% 
  filter(. != "genes") %>% 
  rename(condition = '.') %>% 
  inner_join(ann_vaccines_covid, by = "condition") %>% 
  distinct() %>% 
  arrange(condition) %>% 
  column_to_rownames("condition")

matrix_data = matrix %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var= "condition") %>% 
  inner_join(ann_cols_heatmap %>% rownames_to_column("condition"), by = "condition") %>% 
  select(!c(disease_vac:day)) %>% 
  arrange(condition) %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>%
  as.matrix()

#Rows Immune -----
ann_rows_immune = matrix_data %>%
  as.data.frame() %>%
  rownames_to_column("genes") %>%
  inner_join(ann_genes, by = "genes") %>%
  filter(go_term == "Manual") %>%
  select(genes, process) %>%
  distinct() %>%
  group_by(genes) %>%
  arrange(genes, factor(process, levels = c("INNATE IMMUNE SYSTEM", "ADAPTIVE IMMUNE SYSTEM", "OTHER", "ANTIMICROBIAL IMMUNE RESPONSE")), .by_group = TRUE) %>%
  mutate(i_process = row_number()) %>%
  pivot_wider(., names_from = "i_process", values_from = "process") %>%
  column_to_rownames("genes") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("index") %>%
  arrange(desc(as.numeric(index))) %>%
  column_to_rownames("index") %>%
  t() %>%
  as.data.frame() %>%
  replace(is.na(.), ".")

# Rows Not immune -----
# ann_rows_notimmune = matrix_data %>%
#   as.data.frame() %>%
#   rownames_to_column("genes") %>%
#   left_join(ann_genes %>% 
#                select(process = term, genes = symbol, annotation) %>% 
#                filter(annotation == "Major process", 
#                       process != "Immune system"), by = "genes") %>% 
#   select(genes, process) %>%
#   distinct() %>%
#   group_by(genes) %>%
#   #arrange(genes, factor(process, levels = c("INNATE IMMUNE SYSTEM", "ADAPTIVE IMMUNE SYSTEM", "OTHER", "ANTIMICROBIAL IMMUNE RESPONSE")), .by_group = TRUE) %>% 
#   mutate(i_process = row_number()) %>% 
#   pivot_wider(., names_from = "i_process", values_from = "process") %>% 
#   column_to_rownames("genes") %>% 
#   t() %>% 
#   as.data.frame() %>% 
#   rownames_to_column("index") %>% 
#   arrange(desc(index)) %>% 
#   column_to_rownames("index") %>% 
#   t() %>% 
#   as.data.frame() %>% 
#   replace(is.na(.), ".")

#Verificar dimensões do dataset
dim(matrix_data)
dim(ann_rows_immune)
dim(ann_cols_heatmap)
```

### Colors

### Plot

```{r}
#Heatmap annotation -----
ha = HeatmapAnnotation(df = ann_cols_heatmap, 
                       col = ann_vaxsig_colors,
                       annotation_name_side = "right")

row_ha = HeatmapAnnotation(df = ann_rows_immune, 
                            which = "row",
                       col = col_process_immune, #col_process_immune / col_process_notimmune 
                       show_annotation_name = F,
                       show_legend = F)

# Values colors -----
col_fun <- colorRamp2(c(-2, 0, 2), c("#9bc1bc", "white", "#ed6a5a"))


# Parameters -----
file = filename_2
mat = matrix_data
width_image = 30
height_image = 30
width = unit(15, "cm")
height = unit(10, "cm")
column_names_gp = gpar(fontsize = 10)
row_names_gp = gpar(fontsize = 10)
rect_gp = gpar(col = "gray80", lwd = 0.5)

fileimageheatmap_grouped = paste0(file, sep ="_", "Complexheatmap.png")

png(file = fileimageheatmap_grouped, 
    width = width_image, 
    height = height_image, 
    units="cm", 
    res=900)

heatmap_plot = Heatmap(mat,
                        top_annotation = ha,
                        left_annotation = row_ha,
                        show_row_names = TRUE,
                        row_names_side = "right",
                        show_heatmap_legend = TRUE,
                        heatmap_legend_param = list(direction = "horizontal"),
                        col = col_fun, #color
                        show_column_names = TRUE,
                        column_title = fileimageheatmap_grouped,
                        column_title_gp = gpar(fontsize = 10, fontface = "bold"),
                        column_names_rot = 45,
                        column_names_gp = column_names_gp,
                        rect_gp = rect_gp,
                        na_col = "white",
                        cluster_rows = TRUE,
                        row_names_gp = row_names_gp,
                        row_dend_width = unit(2, "cm"),
                        column_split = 3,
                        row_split = 2,
                        row_gap = unit(2, "mm"),
                        cluster_columns = TRUE,
                        column_gap = unit(8, "mm"),
                        show_column_dend = F,
                        show_row_dend = F,
                        row_dend_side = "left",
                        column_dend_side = "top",
                        clustering_distance_rows = "euclidean",
                        border = TRUE,
                        width = width,
                        height = height 
)

draw(heatmap_plot, 
     annotation_legend_side = "left", 
     heatmap_legend_side = "left")
dev.off()

```

### STATS

```{r}
stats_1 = covid_other_immunego_genes %>% 
  inner_join(ann_vaccines_covid_other, by = "condition") %>% 
  group_by(genes, covid_other) %>% 
  summarise(n_conditions = n()) %>% 
  filter(all(c("COVID", "OTHER") %in% covid_other)) %>% 
  arrange(desc(n_conditions), genes) %>% 
  ungroup()


ggplot_stats1 = ggplot(stats_1) +
  aes(x = reorder(genes, n_conditions), fill = covid_other, weight = n_conditions) +
  geom_bar() +
  theme_minimal() +
  scale_fill_manual(
    values = c(COVID = "#56cfe1",
    OTHER = "#343a40")
  ) +
  theme_minimal() +
  labs(fill = "Covid or non-covid") +
  xlab("Genes") + 
  ylab("Number of shared conditions") +
  theme(legend.position = "right",
        axis.text.x = element_text(angle = 45, vjust = 1.5, hjust=1, size = 8),
        axis.text.y = element_text(size = 7, angle = 0),
        panel.grid.minor.y = element_blank(),
        legend.text = element_text(size=12),
        legend.title = element_text(size = 12, face = "bold")) +
  # geom_text(aes(label = n_conditions, y = n_conditions), 
  #           position = position_dodge(width = 0.9), 
  #           vjust = 0.2, 
  #           hjust = -0.2, 
  #           size = 1,
  #           angle = 90) +
  ggtitle("Genes shared") +
  coord_flip()

ggsave(ggplot_stats1, file = "Covid_Vax_genes_shared_column.png", width = 10, height = 20)




stats_2 = matrix_data %>%
  as.data.frame() %>%
  rownames_to_column("genes") %>%
  inner_join(ImmuneGO_Annotated_Genes, by = "genes") %>%
  filter(go_term == "Manual",
         !(process %in% c("ADAPTIVE IMMUNE SYSTEM",
                          "INNATE IMMUNE SYSTEM"))) %>%
  select(genes, process, immune_system) %>%
  distinct() %>%
  group_by(genes, immune_system) %>% 
  summarise(n_processes = n()) %>% 
  arrange(desc(n_processes)) %>% 
  ungroup() %>% 
  filter(n_processes >1)
  
ggplot_stats2 = ggplot(stats_2) +
  aes(x = reorder(genes, n_processes), fill = immune_system, weight = n_processes) +
  geom_bar() +
  theme_minimal() +
  scale_fill_manual(
    values = c(Adaptive = "#6f2dbd",
    Innate = "#06d6a0",
    Complement = "#ffadc7")
  ) +
  theme_minimal() +
  labs(fill = "Immune system") +
  xlab("Genes") + 
  ylab("Number of shared conditions") +
  theme(legend.position = "right",
        axis.text.x = element_text(angle = 45, vjust = 1.5, hjust=1, size = 8),
        axis.text.y = element_text(size = 7, angle = 0),
        panel.grid.minor.y = element_blank(),
        strip.text.x.top = element_text(size = 10, face = "bold"),
        legend.text = element_text(size=12),
        legend.title = element_text(size = 12, face = "bold")) +
  # geom_text(aes(label = n_conditions, y = n_conditions), 
  #           position = position_dodge(width = 0.9), 
  #           vjust = 0.2, 
  #           hjust = -0.2, 
  #           size = 1,
  #           angle = 90) +
  ggtitle("Genes shared") +
  coord_flip() +
  facet_wrap(~immune_system, scales = "free_y")

ggsave(ggplot_stats2, file = "ImmuneGO_genes_shared_major_process.png", width = 6, height = 20)

```

PVCA of normalized raw counts

```{r}
#### PVCA of normalized raw counts -----
# BiocManager::install("ExpressionNormalizationWorkflow")
library(ExpressionNormalizationWorkflow)

#Covariants
ann_vaccines_samples = ann_vaccines_samples_4_12_23
cvrts = ann_vaccines_samples %>% 
  arrange(condition) %>%
  distinct() %>% 
  column_to_rownames("sample")

#Expression data

exprs = all_matrices_normalized_counts %>% 
  replace(is.na(.), 0) %>% 
  replace(., . == -Inf, 0) %>% 
  replace(., . == Inf, 0) %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column("sample") %>% 
  arrange(sample) %>% 
  inner_join(cvrts %>% 
               #filter(disease_vac == "V") %>% 
               rownames_to_column("sample"),
               select(sample) %>% 
               distinct(), 
             by = "sample") %>% 
  column_to_rownames("sample") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(., . == -Inf, 0) %>% 
  replace(., . == Inf, 0) %>% 
  replace(is.na(.), 0)

cvrts = cvrts %>% 
  rownames_to_column("sample") %>% 
  inner_join(exprs %>% 
               colnames() %>% 
               as.data.frame() %>% 
               select(sample = "."), 
             by = "sample") %>% 
  arrange(sample) %>% 
  column_to_rownames("sample")

# Input
cvrts_eff_var = c("day","age", "sex", "dose", "type", "regimen") #For vaccination only (no ongoing infection)
cvrts_eff_var = c("day","age", "sex", "dose", "type", "previous_vaccination","variant", "prior_infection", "severity", "disease_vac", "regimen") #For vaccination and infection
inpData <- expSetobj(exprs, cvrts) #Convert to Expression Dataset Object
pct_thrsh <- 0.75  #Threshold

#PVCA analysis
pvca = pvcAnaly(inpData, pct_thrsh, cvrts_eff_var) 

pvca_df = pvca$dat %>% 
  as.data.frame() %>% 
  t() %>%
  as.data.frame() %>% 
  rename(wavgpv = V1) %>% 
  cbind(pvca$label %>% as.data.frame() %>% rename(batch = '.')) %>% 
  as.data.frame() %>% 
  mutate(batch = as.factor(batch),
         wavgpv = as.numeric(wavgpv),
         batch = fct_rev(fct_reorder(batch, wavgpv))) %>% 
  arrange(batch, wavgpv)

# Visualize
pvca_df_plot = ggplot(pvca_df) +
  aes(x = batch, y = wavgpv) +
  geom_col(fill = "#5e60ce") +
  theme_minimal() +
  theme(axis.text.x = element_text(size = 20, angle = 45, vjust = 1, hjust=1, color = "black"),
        axis.text.y = element_text(size = 20, vjust = 1, hjust=1, color = "black")) +
  geom_text(aes(label = round(wavgpv,3), y = wavgpv), 
            position = position_dodge(width = 0.9), 
            vjust = -0.2, 
            hjust = 0.5, 
            size = 4,
            angle = 0) 

pvca_df_plot

ggsave(pvca_df_plot, file = "pvca_raw_normalized_immunegenes_allconditions.png", width = 12, height = 10)

```

PVCA of log2FC

```{r}
#### PVCA of log2FC -----

# BiocManager::install("ExpressionNormalizationWorkflow")

library(ExpressionNormalizationWorkflow)

#Covariants
cvrts = ann_vaccines %>% 
  select(gse_id, condition:previous_vaccination) %>% 
  arrange(condition) %>%
  filter(day != 0) %>% 
  distinct() %>% 
  column_to_rownames("condition")
#Expression data

degs_lg2fc_wide = degs_updown_filtered_p_05_vac_infected %>% 
  select(condition, genes, log2fold_change) %>%
  distinct() %>% 
  arrange(condition) %>% 
  pivot_wider(names_from = "condition", values_from = "log2fold_change", values_fn = mean)
  
exprs = degs_lg2fc_wide %>% 
  column_to_rownames("genes") %>% 
  replace(is.na(.), 0) %>% 
  replace(., . == -Inf, 0) %>% 
  replace(., . == Inf, 0) %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column("condition") %>% 
  arrange(condition) %>% 
  inner_join(cvrts %>% 
               rownames_to_column("condition"),
               select(condition) %>% 
               distinct(), 
             by = "condition") %>% 
  column_to_rownames("condition") %>% 
  t() %>% 
  as.data.frame() %>% 
  mutate_if(is.character, as.numeric) %>% 
  replace(., . == -Inf, 0) %>% 
  replace(., . == Inf, 0) %>% 
  replace(is.na(.), 0)

cvrts = cvrts %>% 
  rownames_to_column("condition") %>% 
  inner_join(exprs %>% 
               colnames() %>% 
               as.data.frame() %>% 
               select(condition = "."), 
             by = "condition") %>% 
  column_to_rownames("condition")

# Input
cvrts_eff_var = c("day", "dose", "prior_infection", "previous_vaccination", "severity", "disease_vac", "regimen", "type" )
inpData <- expSetobj(exprs, cvrts)
pct_thrsh <- 0.75 
pvca = pvcAnaly(inpData, pct_thrsh, cvrts_eff_var)

pvca_df = pvca$dat %>% 
  as.data.frame() %>% 
  t() %>%
  as.data.frame() %>% 
  rename(wavgpv = V1) %>% 
  cbind(pvca$label %>% as.data.frame() %>% rename(batch = '.')) %>% 
  as.data.frame() %>% 
  mutate(batch = as.factor(batch),
         wavgpv = as.numeric(wavgpv),
         batch = fct_rev(fct_reorder(batch, wavgpv))) %>% 
  arrange(batch, wavgpv)

# Visualize

pvca_df_plot = ggplot(pvca_df) +
  aes(x = batch, y = wavgpv) +
  geom_col(fill = "#5e60ce") +
  theme_minimal() +
  theme(axis.text.x = element_text(size = 20, angle = 45, vjust = 1, hjust=1, color = "black"),
        axis.title.x = element_text(size = 20,  color = "black"))

ggsave(pvca_df_plot, file = "pvca_log2foldchange.png", width = 12, height = 10)


```

# Gene expression by gender

```{r}
##### DESEQ2 Normalization -------
library(DESeq2)

exprs = all_matrices_counts %>% 
  column_to_rownames("genes") %>% 
  t() %>% 
  as.data.frame() %>% 
  round(., 0) %>% 
  rownames_to_column("sample") %>% 
  arrange(sample) %>% 
  inner_join(cvrts %>% 
               rownames_to_column("sample"),
               select(sample) %>% 
               distinct(), 
             by = "sample") %>% 
  column_to_rownames("sample") %>% 
  t() %>% 
  as.data.frame() %>% 
  exprs %>% 
  drop_na() %>% 
  mutate_if(is.character, as.integer) %>% 
  drop_na() 

cvrts = cvrts %>% 
  mutate(condition = as.factor(condition),
         age = as.factor(age),
         sex = as.factor(sex),
         vaccine = as.factor(vaccine),
         day = as.factor(day),
         gse_id = as.factor(gse_id))
  
dds <- DESeqDataSetFromMatrix(countData = a, colData = cvrts, design = ~ condition)
dds <- estimateSizeFactors(dds)
normalized_counts <- counts(dds, normalized=TRUE) %>% 
  as.data.frame()

saveRDS(normalized_counts, file = "all_matrices_normalized_counts.rds")

normalized_counts = all_matrices_normalized_counts
```

```{r}
#### PC1 -------
PC1 = var$cos2 %>%
  as.data.frame() %>%
  arrange(desc(Dim.1)) %>%
  slice_head(n = 5) %>%
  rownames_to_column("genes") %>% 
  distinct() %>% 
  select(genes, Dim.1)

PC1_genes_counts = normalized_counts %>% 
  rownames_to_column("genes") %>% 
  inner_join(PC1, by = "genes") %>% 
  pivot_longer(cols = -c("genes", "Dim.1"), names_to = "sample", values_to = "counts") %>% 
  mutate(counts = round(counts, 0)) %>% 
  select(genes, sample, counts, Dim.1) %>% 
  inner_join(ann_vaccines_samples, by = "sample") %>% 
  filter(gse_id != "GSE206023") %>% 
  group_by(condition) %>%
  filter(any(sex == 'Male') & any(sex == 'Female'),
         condition != "H (D0)") %>% 
  ungroup() %>% 
  mutate(age_inter = cut(age, breaks = c(10, 20, 30, 40, 50, 60, 70, 80, 90, 100), labels = c("10-20", "21-30", "31-40", "41-50", "51-60", "61-70", "71-80", "81-90", "91-100")))

  
write.csv(PC1_genes_counts, file = "PC1_genes_samples_counts.csv", row.names = F)


#Sex ----
pc1_5genes  = ggplot(PC1_genes_counts) +
  aes(x = condition, y = counts, fill = sex, colour = sex) +
  geom_boxplot() +
  scale_fill_hue(h = c(180, 270)) +
  scale_color_hue(h = c(180, 270)) +
  scale_y_continuous(trans = "log2") +
  theme_light() +
  theme(legend.position = "top") +
  facet_wrap(vars(genes), scales = "free", ncol =1) +
  theme(axis.text.x = element_text(size = 10, color = "black", angle = 45, vjust = 1, hjust = 1),
        axis.text.y = element_text(size = 10, color = "black"),
        strip.background = element_rect(fill = "black"),
        strip.text = element_text(color = "white", face = "bold", size = 15)) 

ggsave(pc1_5genes, file = "pc1_5genes_sex.png", width = 10, height = 20)

#Age ----

pc1_5genes_age  = ggplot(PC1_genes_counts) +
  aes(x = condition, y = counts, fill = age_inter) +
  geom_boxplot() +
  scale_fill_hue(h = c(180, 360)) +
  scale_color_hue(h = c(180, 360)) +
  scale_y_continuous(trans = "log2") +
  theme_light() +
  theme(legend.position = "top") +
  theme(axis.text.x = element_text(size = 10, color = "black", angle = 45, vjust = 1, hjust = 1),
        axis.text.y = element_text(size = 10, color = "black"),
        strip.background = element_rect(fill = "black"),
        strip.text = element_text(color = "white", face = "bold", size = 15)) +
  facet_wrap(vars(genes), scales = "free", ncol = 1)
ggsave(pc1_5genes_age, file = "pc1_5genes_age.png", width = 10, height = 20)


#### PC2 -------
PC2 = var$cos2 %>%
  as.data.frame() %>%
  arrange(desc(Dim.2)) %>%
  slice_head(n = 5) %>%
  rownames_to_column("genes") %>% 
  distinct() %>% 
  select(genes, Dim.1)

PC2_genes_counts = normalized_counts %>% 
  rownames_to_column("genes") %>% 
  inner_join(PC2, by = "genes") %>% 
  pivot_longer(cols = -c("genes", "Dim.1"), names_to = "sample", values_to = "counts") %>% 
  mutate(counts = round(counts, 0)) %>% 
  select(genes, sample, counts, Dim.1) %>% 
  inner_join(ann_vaccines_samples, by = "sample") %>% 
  filter(gse_id != "GSE206023") %>% 
  group_by(condition) %>%
  filter(any(sex == 'Male') & any(sex == 'Female'),
         condition != "H (D0)") %>% 
  ungroup() %>% 
  mutate(age_inter = cut(age, breaks = c(10, 20, 30, 40, 50, 60, 70, 80, 90, 100), labels = c("10-20", "21-30", "31-40", "41-50", "51-60", "61-70", "71-80", "81-90", "91-100")))

  
write.csv(PC2_genes_counts, file = "PC2_genes_samples_counts.csv", row.names = F)


#Sex ----
pc2_5genes  = ggplot(PC2_genes_counts) +
  aes(x = condition, y = counts, fill = sex, colour = sex) +
  geom_boxplot() +
  scale_fill_hue(h = c(180, 270)) +
  scale_color_hue(h = c(180, 270)) +
  scale_y_continuous(trans = "log2") +
  theme_light() +
  theme(legend.position = "top") +
  facet_wrap(vars(genes), scales = "free", ncol =1) +
  theme(axis.text.x = element_text(size = 10, color = "black", angle = 45, vjust = 1, hjust = 1),
        axis.text.y = element_text(size = 10, color = "black"),
        strip.background = element_rect(fill = "black"),
        strip.text = element_text(color = "white", face = "bold", size = 15))

ggsave(pc2_5genes, file = "pc2_5genes_sex.png", width = 10, height = 20)

#Age ----

pc2_5genes_age  = ggplot(PC2_genes_counts) +
  aes(x = condition, y = counts, fill = age_inter, colour = age_inter) +
  geom_boxplot() +
  scale_fill_hue(h = c(180, 360)) +
  scale_color_hue(h = c(180, 360)) +
  scale_y_continuous(trans = "log2") +
  theme_light() +
  theme(legend.position = "top") +
  theme(axis.text.x = element_text(size = 10, color = "black", angle = 45, vjust = 1, hjust = 1),
        axis.text.y = element_text(size = 10, color = "black"),
        strip.background = element_rect(fill = "black"),
        strip.text = element_text(color = "white", face = "bold", size = 15)) +
  facet_wrap(vars(genes), scales = "free", ncol = 1)
ggsave(pc2_5genes_age, file = "pc2_5genes_age.png", width = 10, height = 20)


```
